<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.11: http://docutils.sourceforge.net/" />
<title></title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 7614 2013-02-21 15:55:51Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

object[type="image/svg+xml"], object[type="application/x-shockwave-flash"] {
  overflow: hidden;
}

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title, .code .error {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left, object.align-left {
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right, object.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

img.align-center, .figure.align-center, object.align-center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: inherit }

/* div.align-center * { */
/*   text-align: left } */

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block, pre.math, pre.code {
  margin-left: 2em ;
  margin-right: 2em }

pre.code .ln { color: grey; } /* line numbers */
pre.code, code { background-color: #eeeeee }
pre.code .comment, code .comment { color: #5C6576 }
pre.code .keyword, code .keyword { color: #3B0D06; font-weight: bold }
pre.code .literal.string, code .literal.string { color: #0C5404 }
pre.code .name.builtin, code .name.builtin { color: #352B84 }
pre.code .deleted, code .deleted { background-color: #DEB0A1}
pre.code .inserted, code .inserted { background-color: #A3D289}

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

/* "booktabs" style (no vertical lines) */
table.docutils.booktabs {
  border: 0px;
  border-top: 2px solid;
  border-bottom: 2px solid;
  border-collapse: collapse;
}
table.docutils.booktabs * {
  border: 0px;
}
table.docutils.booktabs th {
  border-bottom: thin solid;
  text-align: left;
}

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
<style type="text/css">

/*
*   math2html: convert LaTeX equations to HTML output.
*
*   Copyright (C) 2009,2010 Alex Fernández
*
*   Released under the terms of the `2-Clause BSD license'_, in short:
*   Copying and distribution of this file, with or without modification,
*   are permitted in any medium without royalty provided the copyright
*   notice and this notice are preserved.
*   This file is offered as-is, without any warranty.
*
* .. _2-Clause BSD license: http://www.spdx.org/licenses/BSD-2-Clause
*
*   Based on eLyXer: convert LyX source files to HTML output.
*   http://elyxer.nongnu.org/
*/
/* --end--
* CSS file for LaTeX formulas.
*/

/* Formulas */
.formula {
	text-align: center;
	font-family: "Droid Serif", "DejaVu Serif", "STIX", serif;
	margin: 1.2em 0;
}
span.formula {
	white-space: nowrap;
}
div.formula {
	padding: 0.5ex;
	margin-left: auto;
	margin-right: auto;
}

/* Basic features */
a.eqnumber {
	display: inline-block;
	float: right;
	clear: right;
	font-weight: bold;
}
span.unknown {
	color: #800000;
}
span.ignored, span.arraydef {
	display: none;
}
.formula i {
	letter-spacing: 0.1ex;
}

/* Alignment */
.align-left, .align-l {
	text-align: left;
}
.align-right, .align-r {
	text-align: right;
}
.align-center, .align-c {
	text-align: center;
}

/* Structures */
span.overline, span.bar {
	text-decoration: overline;
}
.fraction, .fullfraction {
	display: inline-block;
	vertical-align: middle;
	text-align: center;
}
.fraction .fraction {
	font-size: 80%;
	line-height: 100%;
}
span.numerator {
	display: block;
}
span.denominator {
	display: block;
	padding: 0ex;
	border-top: thin solid;
}
sup.numerator, sup.unit {
	font-size: 70%;
	vertical-align: 80%;
}
sub.denominator, sub.unit {
	font-size: 70%;
	vertical-align: -20%;
}
span.sqrt {
	display: inline-block;
	vertical-align: middle;
	padding: 0.1ex;
}
sup.root {
	font-size: 70%;
	position: relative;
	left: 1.4ex;
}
span.radical {
	display: inline-block;
	padding: 0ex;
	font-size: 150%;
	vertical-align: top;
}
span.root {
	display: inline-block;
	border-top: thin solid;
	padding: 0ex;
	vertical-align: middle;
}
span.symbol {
	line-height: 125%;
	font-size: 125%;
}
span.bigsymbol {
	line-height: 150%;
	font-size: 150%;
}
span.largesymbol {
	font-size: 175%;
}
span.hugesymbol {
	font-size: 200%;
}
span.scripts {
	display: inline-table;
	vertical-align: middle;
}
.script {
	display: table-row;
	text-align: left;
	line-height: 150%;
}
span.limits {
	display: inline-table;
	vertical-align: middle;
}
.limit {
	display: table-row;
	line-height: 99%;
}
sup.limit, sub.limit {
	line-height: 100%;
}
span.symbolover {
	display: inline-block;
	text-align: center;
	position: relative;
	float: right;
	right: 100%;
	bottom: 0.5em;
	width: 0px;
}
span.withsymbol {
	display: inline-block;
}
span.symbolunder {
	display: inline-block;
	text-align: center;
	position: relative;
	float: right;
	right: 80%;
	top: 0.3em;
	width: 0px;
}

/* Environments */
span.array, span.bracketcases, span.binomial, span.environment {
	display: inline-table;
	text-align: center;
	border-collapse: collapse;
	margin: 0em;
	vertical-align: middle;
}
span.arrayrow, span.binomrow {
	display: table-row;
	padding: 0ex;
	border: 0ex;
}
span.arraycell, span.bracket, span.case, span.binomcell, span.environmentcell {
	display: table-cell;
	padding: 0ex 0.2ex;
	line-height: 99%;
	border: 0ex;
}
/*
* CSS file for LaTeX formulas, extra stuff:
* binomials, vertical braces, stackrel, fonts and colors.
*/

/* Inline binomials */
span.binom {
	display: inline-block;
	vertical-align: middle;
	text-align: center;
	font-size: 80%;
}
span.binomstack {
	display: block;
	padding: 0em;
}

/* Over- and underbraces */
span.overbrace {
	border-top: 2pt solid;
}
span.underbrace {
	border-bottom: 2pt solid;
}

/* Stackrel */
span.stackrel {
	display: inline-block;
	text-align: center;
}
span.upstackrel {
	display: block;
	padding: 0em;
	font-size: 80%;
	line-height: 64%;
	position: relative;
	top: 0.15em;

}
span.downstackrel {
	display: block;
	vertical-align: bottom;
	padding: 0em;
}

/* Fonts */
span.mathsf, span.textsf {
	font-style: normal;
	font-family: sans-serif;
}
span.mathrm, span.textrm {
	font-style: normal;
	font-family: serif;
}
span.text, span.textnormal {
	font-style: normal;
}
span.textipa {
	color: #008080;
}
span.fraktur {
	font-family: "Lucida Blackletter", eufm10, blackletter;
}
span.blackboard {
	font-family: Blackboard, msbm10, serif;
}
span.scriptfont {
	font-family: "Monotype Corsiva", "Apple Chancery", "URW Chancery L", cursive;
	font-style: italic;
}

/* Colors */
span.colorbox {
	display: inline-block;
	padding: 5px;
}
span.fbox {
	display: inline-block;
	border: thin solid black;
	padding: 2px;
}
span.boxed, span.framebox {
	display: inline-block;
	border: thin solid black;
	padding: 5px;
}


</style>
</head>
<body>
<div class="document">


<div class="section" id="chapter-1-introduction">
<h1>Chapter 1  Introduction</h1>
<div class="section" id="what-is-biopython">
<h2>1.1  What is Biopython?</h2>
<p>The Biopython Project is an international association of developers of
freely available Python
(<a class="reference external" href="http://www.python.org">http://www.python.org</a>) tools for
computational molecular biology. Python is an object oriented,
interpreted, flexible language that is becoming increasingly popular for
scientific computing. Python is easy to learn, has a very clear syntax
and can easily be extended with modules written in C, C++ or FORTRAN.</p>
<p>The Biopython web site
(<a class="reference external" href="http://www.biopython.org">http://www.biopython.org</a>) provides
an online resource for modules, scripts, and web links for developers of
Python-based software for bioinformatics use and research. Basically,
the goal of Biopython is to make it as easy as possible to use Python
for bioinformatics by creating high-quality, reusable modules and
classes. Biopython features include parsers for various Bioinformatics
file formats (BLAST, Clustalw, FASTA, Genbank,...), access to online
services (NCBI, Expasy,...), interfaces to common and not-so-common
programs (Clustalw, DSSP, MSMS...), a standard sequence class, various
clustering modules, a KD tree data structure etc. and even
documentation.</p>
<p>Basically, we just like to program in Python and want to make it as easy
as possible to use Python for bioinformatics by creating high-quality,
reusable modules and scripts.</p>
</div>
<div class="section" id="what-can-i-find-in-the-biopython-package">
<h2>1.2  What can I find in the Biopython package</h2>
<p>The main Biopython releases have lots of functionality, including:</p>
<ul class="simple">
<li>The ability to parse bioinformatics files into Python utilizable data
structures, including support for the following formats:<ul>
<li>Blast output – both from standalone and WWW Blast</li>
<li>Clustalw</li>
<li>FASTA</li>
<li>GenBank</li>
<li>PubMed and Medline</li>
<li>ExPASy files, like Enzyme and Prosite</li>
<li>SCOP, including ‘dom’ and ‘lin’ files</li>
<li>UniGene</li>
<li>SwissProt</li>
</ul>
</li>
<li>Files in the supported formats can be iterated over record by record
or indexed and accessed via a Dictionary interface.</li>
<li>Code to deal with popular on-line bioinformatics destinations such
as:<ul>
<li>NCBI – Blast, Entrez and PubMed services</li>
<li>ExPASy – Swiss-Prot and Prosite entries, as well as Prosite
searches</li>
</ul>
</li>
<li>Interfaces to common bioinformatics programs such as:<ul>
<li>Standalone Blast from NCBI</li>
<li>Clustalw alignment program</li>
<li>EMBOSS command line tools</li>
</ul>
</li>
<li>A standard sequence class that deals with sequences, ids on
sequences, and sequence features.</li>
<li>Tools for performing common operations on sequences, such as
translation, transcription and weight calculations.</li>
<li>Code to perform classification of data using k Nearest Neighbors,
Naive Bayes or Support Vector Machines.</li>
<li>Code for dealing with alignments, including a standard way to create
and deal with substitution matrices.</li>
<li>Code making it easy to split up parallelizable tasks into separate
processes.</li>
<li>GUI-based programs to do basic sequence manipulations, translations,
BLASTing, etc.</li>
<li>Extensive documentation and help with using the modules, including
this file, on-line wiki documentation, the web site, and the mailing
list.</li>
<li>Integration with BioSQL, a sequence database schema also supported by
the BioPerl and BioJava projects.</li>
</ul>
<p>We hope this gives you plenty of reasons to download and start using
Biopython!</p>
</div>
<div class="section" id="installing-biopython">
<h2>1.3  Installing Biopython</h2>
<p>All of the installation information for Biopython was separated from
this document to make it easier to keep updated.</p>
<p>The short version is go to our downloads page
(<a class="reference external" href="http://biopython.org/wiki/Download">http://biopython.org/wiki/Download</a>),
download and install the listed dependencies, then download and install
Biopython. Biopython runs on many platforms (Windows, Mac, and on the
various flavors of Linux and Unix). For Windows we provide pre-compiled
click-and-run installers, while for Unix and other operating systems you
must install from source as described in the included README file. This
is usually as simple as the standard commands:</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">setup</span><span class="operator">.</span><span class="name">py</span> <span class="name">build</span>
<span class="name">python</span> <span class="name">setup</span><span class="operator">.</span><span class="name">py</span> <span class="name">test</span>
<span class="name">sudo</span> <span class="name">python</span> <span class="name">setup</span><span class="operator">.</span><span class="name">py</span> <span class="name">install</span>
</pre>
<p>(You can in fact skip the build and test, and go straight to the install
– but its better to make sure everything seems to be working.)</p>
<p>The longer version of our installation instructions covers installation
of Python, Biopython dependencies and Biopython itself. It is available
in PDF
(<a class="reference external" href="http://biopython.org/DIST/docs/install/Installation.pdf">http://biopython.org/DIST/docs/install/Installation.pdf</a>)
and HTML formats
(<a class="reference external" href="http://biopython.org/DIST/docs/install/Installation.html">http://biopython.org/DIST/docs/install/Installation.html</a>).</p>
</div>
<div class="section" id="frequently-asked-questions-faq">
<h2>1.4  Frequently Asked Questions (FAQ)</h2>
<ol class="arabic">
<li><p class="first"><em>How do I cite Biopython in a scientific publication?</em></p>
<p>Please cite our application note [<a class="reference external" href="#cock2009">1</a>, Cock <em>et al.</em>,
2009] as the main Biopython reference. In addition, please cite any
publications from the following list if appropriate, in particular as
a reference for specific modules within Biopython (more information
can be found on our website):</p>
<ul class="simple">
<li>For the official project announcement: [<a class="reference external" href="#chapman2000">13</a>,
Chapman and Chang, 2000];</li>
<li>For <tt class="docutils literal">Bio.PDB</tt>: [<a class="reference external" href="#hamelryck2003a">18</a>, Hamelryck and
Manderick, 2003];</li>
<li>For <tt class="docutils literal">Bio.Cluster</tt>: [<a class="reference external" href="#dehoon2004">14</a>, De Hoon <em>et al.</em>,
2004];</li>
<li>For <tt class="docutils literal">Bio.Graphics.GenomeDiagram</tt>: [<a class="reference external" href="#pritchard2006">2</a>,
Pritchard <em>et al.</em>, 2006];</li>
<li>For <tt class="docutils literal">Bio.Phylo</tt> and <tt class="docutils literal">Bio.Phylo.PAML</tt>: [<a class="reference external" href="#talevich2012">9</a>,
Talevich <em>et al.</em>, 2012];</li>
<li>For the FASTQ file format as supported in Biopython, BioPerl,
BioRuby, BioJava, and EMBOSS: [<a class="reference external" href="#cock2010">7</a>, Cock <em>et al.</em>,
2010].</li>
</ul>
</li>
<li><p class="first"><em>How should I capitalize “Biopython”? Is “BioPython” OK?</em></p>
<p>The correct capitalization is “Biopython”, not “BioPython” (even
though that would have matched BioPerl, BioJava and BioRuby).</p>
</li>
<li><p class="first"><em>How do I find out what version of Biopython I have installed?</em></p>
<p>Use this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">Bio</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">Bio</span><span class="operator">.</span><span class="name">__version__</span>
<span class="operator">...</span>
</pre>
<p>If the “<tt class="docutils literal">import Bio</tt>” line fails, Biopython is not installed.
If the second line fails, your version is very out of date. If the
version string ends with a plus, you don’t have an official release,
but a snapshot of the in development code.</p>
</li>
<li><p class="first"><em>Where is the latest version of this document?</em></p>
<p>If you download a Biopython source code archive, it will include the
relevant version in both HTML and PDF formats. The latest published
version of this document (updated at each release) is online:</p>
<ul class="simple">
<li><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html">http://biopython.org/DIST/docs/tutorial/Tutorial.html</a></li>
<li><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.pdf">http://biopython.org/DIST/docs/tutorial/Tutorial.pdf</a></li>
</ul>
<p>If you are using the very latest unreleased code from our repository
you can find copies of the in-progress tutorial here:</p>
<ul class="simple">
<li><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial-dev.html">http://biopython.org/DIST/docs/tutorial/Tutorial-dev.html</a></li>
<li><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial-dev.pdf">http://biopython.org/DIST/docs/tutorial/Tutorial-dev.pdf</a></li>
</ul>
</li>
<li><p class="first"><em>Which “Numerical Python” do I need?</em></p>
<p>For Biopython 1.48 or earlier, you needed the old Numeric module.
For Biopython 1.49 onwards, you need the newer NumPy instead. Both
Numeric and NumPy can be installed on the same machine fine. See
also: <a class="reference external" href="http://numpy.scipy.org/">http://numpy.scipy.org/</a></p>
</li>
<li><p class="first"><em>Why is the</em> <tt class="docutils literal">Seq</tt> <em>object missing the (back) transcription &amp;
translation methods described in this Tutorial?</em></p>
<p>You need Biopython 1.49 or later. Alternatively, use the <tt class="docutils literal">Bio.Seq</tt>
module functions described in
Section <a class="reference external" href="#sec:seq-module-functions">3.14</a>.</p>
</li>
<li><p class="first"><em>Why is the</em> <tt class="docutils literal">Seq</tt> <em>object missing the upper &amp; lower methods
described in this Tutorial?</em></p>
<p>You need Biopython 1.53 or later. Alternatively, use
<tt class="docutils literal"><span class="pre">str(my_seq).upper()</span></tt> to get an upper case string. If you need a
Seq object, try <tt class="docutils literal"><span class="pre">Seq(str(my_seq).upper())</span></tt> but be careful about
blindly re-using the same alphabet.</p>
</li>
<li><p class="first"><em>Why doesn’t the</em> <tt class="docutils literal">Seq</tt> <em>object translation method support the</em>
<tt class="docutils literal">cds</tt> <em>option described in this Tutorial?</em></p>
<p>You need Biopython 1.51 or later.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal">Bio.SeqIO</tt> <em>work? It imports fine but there is no
parse function etc.</em></p>
<p>You need Biopython 1.43 or later. Older versions did contain some
related code under the <tt class="docutils literal">Bio.SeqIO</tt> name which has since been
removed - and this is why the import “works”.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal">Bio.SeqIO.read()</tt> <em>work? The module imports fine but
there is no read function!</em></p>
<p>You need Biopython 1.45 or later. Or, use
<tt class="docutils literal"><span class="pre">Bio.SeqIO.parse(...).next()</span></tt> instead.</p>
</li>
<li><p class="first"><em>Why isn’t</em> <tt class="docutils literal">Bio.AlignIO</tt> <em>present? The module import fails!</em></p>
<p>You need Biopython 1.46 or later.</p>
</li>
<li><p class="first"><em>What file formats do</em> <tt class="docutils literal">Bio.SeqIO</tt> <em>and</em> <tt class="docutils literal">Bio.AlignIO</tt> <em>read and
write?</em></p>
<p>Check the built in docstrings (<tt class="docutils literal">from Bio import SeqIO</tt>, then
<tt class="docutils literal">help(SeqIO)</tt>), or see
<a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a>
and
<a class="reference external" href="http://biopython.org/wiki/AlignIO">http://biopython.org/wiki/AlignIO</a>
on the wiki for the latest listing.</p>
</li>
<li><p class="first"><em>Why don’t the</em> <tt class="docutils literal">Bio.SeqIO</tt> <em>and</em> <tt class="docutils literal">Bio.AlignIO</tt> <em>input functions
let me provide a sequence alphabet?</em></p>
<p>You need Biopython 1.49 or later.</p>
</li>
<li><p class="first"><em>Why won’t the</em> <tt class="docutils literal">Bio.SeqIO</tt> <em>and</em> <tt class="docutils literal">Bio.AlignIO</tt> <em>functions</em>
<tt class="docutils literal">parse</tt><em>,</em> <tt class="docutils literal">read</tt> <em>and</em> <tt class="docutils literal">write</tt> <em>take filenames? They insist
on handles!</em></p>
<p>You need Biopython 1.54 or later, or just use handles explicitly
(see Section <a class="reference external" href="#sec:appendix-handles">22.1</a>). It is especially
important to remember to close output handles explicitly after
writing your data.</p>
</li>
<li><p class="first"><em>Why won’t the</em> <tt class="docutils literal">Bio.SeqIO.write()</tt> <em>and</em> <tt class="docutils literal">Bio.AlignIO.write()</tt>
<em>functions accept a single record or alignment? They insist on a list
or iterator!</em></p>
<p>You need Biopython 1.54 or later, or just wrap the item with
<tt class="docutils literal"><span class="pre">[...]</span></tt> to create a list of one element.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal"><span class="pre">str(...)</span></tt> <em>give me the full sequence of a</em> <tt class="docutils literal">Seq</tt>
<em>object?</em></p>
<p>You need Biopython 1.45 or later. Alternatively, rather than
<tt class="docutils literal">str(my_seq)</tt>, use <tt class="docutils literal">my_seq.tostring()</tt> (which will also work on
recent versions of Biopython).</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal">Bio.Blast</tt> <em>work with the latest plain text NCBI
blast output?</em></p>
<p>The NCBI keep tweaking the plain text output from the BLAST tools,
and keeping our parser up to date is/was an ongoing struggle. If you
aren’t using the latest version of Biopython, you could try
upgrading. However, we (and the NCBI) recommend you use the XML
output instead, which is designed to be read by a computer program.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal">Bio.Entrez.read()</tt> <em>work? The module imports fine
but there is no read function!</em></p>
<p>You need Biopython 1.46 or later.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal">Bio.Entrez.parse()</tt> <em>work? The module imports fine
but there is no parse function!</em></p>
<p>You need Biopython 1.52 or later.</p>
</li>
<li><p class="first"><em>Why has my script using</em> <tt class="docutils literal">Bio.Entrez.efetch()</tt> <em>stopped working?</em></p>
<p>This could be due to NCBI changes in February 2012 introducing
EFetch 2.0. First, they changed the default return modes - you
probably want to add <tt class="docutils literal"><span class="pre">retmode=&quot;text&quot;</span></tt> to your call. Second, they
are now stricter about how to provide a list of IDs – Biopython 1.59
onwards turns a list into a comma separated string automatically.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal">Bio.Blast.NCBIWWW.qblast()</tt> <em>give the same results
as the NCBI BLAST website?</em></p>
<p>You need to specify the same options – the NCBI often adjust the
default settings on the website, and they do not match the QBLAST
defaults anymore. Check things like the gap penalties and expectation
threshold.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal">Bio.Blast.NCBIXML.read()</tt> <em>work? The module imports
but there is no read function!</em></p>
<p>You need Biopython 1.50 or later. Or, use
<tt class="docutils literal"><span class="pre">Bio.Blast.NCBIXML.parse(...).next()</span></tt> instead.</p>
</li>
<li><p class="first"><em>Why doesn’t my</em> <tt class="docutils literal">SeqRecord</tt> <em>object have a</em> <tt class="docutils literal">letter_annotations</tt>
<em>attribute?</em></p>
<p>Per-letter-annotation support was added in Biopython 1.50.</p>
</li>
<li><p class="first"><em>Why can’t I slice my</em> <tt class="docutils literal">SeqRecord</tt> <em>to get a sub-record?</em></p>
<p>You need Biopython 1.50 or later.</p>
</li>
<li><p class="first"><em>Why can’t I add</em> <tt class="docutils literal">SeqRecord</tt> <em>objects together?</em></p>
<p>You need Biopython 1.53 or later.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal">Bio.SeqIO.convert()</tt> <em>or</em> <tt class="docutils literal">Bio.AlignIO.convert()</tt>
<em>work? The modules import fine but there is no convert function!</em></p>
<p>You need Biopython 1.52 or later. Alternatively, combine the
<tt class="docutils literal">parse</tt> and <tt class="docutils literal">write</tt> functions as described in this tutorial (see
Sections <a class="reference external" href="#sec:SeqIO-conversion">5.5.2</a>
and <a class="reference external" href="#sec:converting-alignments">6.2.1</a>).</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal">Bio.SeqIO.index()</tt> <em>work? The module imports fine
but there is no index function!</em></p>
<p>You need Biopython 1.52 or later.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal">Bio.SeqIO.index_db()</tt> <em>work? The module imports fine
but there is no</em> <tt class="docutils literal">index_db</tt> <em>function!</em></p>
<p>You need Biopython 1.57 or later (and a Python with SQLite3
support).</p>
</li>
<li><p class="first"><em>Where is the</em> <tt class="docutils literal">MultipleSeqAlignment</tt> <em>object? The</em> <tt class="docutils literal">Bio.Align</tt>
<em>module imports fine but this class isn’t there!</em></p>
<p>You need Biopython 1.54 or later. Alternatively, the older
<tt class="docutils literal">Bio.Align.Generic.Alignment</tt> class supports some of its
functionality, but using this is now discouraged.</p>
</li>
<li><p class="first"><em>Why can’t I run command line tools directly from the application
wrappers?</em></p>
<p>You need Biopython 1.55 or later. Alternatively, use the Python
<tt class="docutils literal">subprocess</tt> module directly.</p>
</li>
<li><p class="first"><em>I looked in a directory for code, but I couldn’t find the code that
does something. Where’s it hidden?</em></p>
<p>One thing to know is that we put code in <tt class="docutils literal">__init__.py</tt> files. If
you are not used to looking for code in this file this can be
confusing. The reason we do this is to make the imports easier for
users. For instance, instead of having to do a “repetitive” import
like <tt class="docutils literal">from Bio.GenBank import GenBank</tt>, you can just use
<tt class="docutils literal">from Bio import GenBank</tt>.</p>
</li>
<li><p class="first"><em>Why does the code from CVS seem out of date?</em></p>
<p>In late September 2009, just after the release of Biopython 1.52, we
switched from using CVS to git, a distributed version control system.
The old CVS server will remain available as a static and read only
backup, but if you want to grab the latest code, you’ll need to use
git instead. See our website for more details.</p>
</li>
</ol>
<p>For more general questions, the Python FAQ pages
<a class="reference external" href="http://www.python.org/doc/faq/">http://www.python.org/doc/faq/</a>
may be useful.</p>
</div>
</div>
<div class="section" id="chapter-2-quick-start-what-can-you-do-with-biopython">
<h1>Chapter 2  Quick Start – What can you do with Biopython?</h1>
<p>This section is designed to get you started quickly with Biopython, and
to give a general overview of what is available and how to use it. All
of the examples in this section assume that you have some general
working knowledge of Python, and that you have successfully installed
Biopython on your system. If you think you need to brush up on your
Python, the main Python web site provides quite a bit of free
documentation to get started with
(<a class="reference external" href="http://www.python.org/doc/">http://www.python.org/doc/</a>).</p>
<p>Since much biological work on the computer involves connecting with
databases on the internet, some of the examples will also require a
working internet connection in order to run.</p>
<p>Now that that is all out of the way, let’s get into what we can do with
Biopython.</p>
<div class="section" id="general-overview-of-what-biopython-provides">
<h2>2.1  General overview of what Biopython provides</h2>
<p>As mentioned in the introduction, Biopython is a set of libraries to
provide the ability to deal with “things” of interest to biologists
working on the computer. In general this means that you will need to
have at least some programming experience (in Python, of course!) or at
least an interest in learning to program. Biopython’s job is to make
your job easier as a programmer by supplying reusable libraries so that
you can focus on answering your specific question of interest, instead
of focusing on the internals of parsing a particular file format (of
course, if you want to help by writing a parser that doesn’t exist and
contributing it to Biopython, please go ahead!). So Biopython’s job is
to make you happy!</p>
<p>One thing to note about Biopython is that it often provides multiple
ways of “doing the same thing.” Things have improved in recent releases,
but this can still be frustrating as in Python there should ideally be
one right way to do something. However, this can also be a real benefit
because it gives you lots of flexibility and control over the libraries.
The tutorial helps to show you the common or easy ways to do things so
that you can just make things work. To learn more about the alternative
possibilities, look in the Cookbook (Chapter <a class="reference external" href="#chapter:cookbook">18</a>,
this has some cools tricks and tips), the Advanced section
(Chapter <a class="reference external" href="#chapter:advanced">20</a>), the built in “docstrings” (via the
Python help command, or the <a class="reference external" href="http://biopython.org/DIST/docs/api/">API
documentation</a>) or ultimately
the code itself.</p>
</div>
<div class="section" id="working-with-sequences">
<h2>2.2  Working with sequences</h2>
<p>Disputably (of course!), the central object in bioinformatics is the
sequence. Thus, we’ll start with a quick introduction to the Biopython
mechanisms for dealing with sequences, the <tt class="docutils literal">Seq</tt> object, which we’ll
discuss in more detail in Chapter <a class="reference external" href="#chapter:Bio.Seq">3</a>.</p>
<p>Most of the time when we think about sequences we have in my mind a
string of letters like ‘<tt class="docutils literal">AGTACACTGGT</tt>’. You can create such
<tt class="docutils literal">Seq</tt> object with this sequence as follows - the “<tt class="docutils literal">&gt;&gt;&gt;</tt>”
represents the Python prompt followed by what you would type in:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AGTACACTGGT&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGTACACTGGT'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_seq</span>
<span class="name">AGTACACTGGT</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">alphabet</span>
<span class="name">Alphabet</span><span class="punctuation">()</span>
</pre>
<p>What we have here is a sequence object with a <em>generic</em> alphabet -
reflecting the fact we have <em>not</em> specified if this is a DNA or protein
sequence (okay, a protein with a lot of Alanines, Glycines, Cysteines
and Threonines!). We’ll talk more about alphabets in
Chapter <a class="reference external" href="#chapter:Bio.Seq">3</a>.</p>
<p>In addition to having an alphabet, the <tt class="docutils literal">Seq</tt> object differs from the
Python string in the methods it supports. You can’t do this with a plain
string:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGTACACTGGT'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">complement</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TCATGTGACCA'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ACCAGTGTACT'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
</pre>
<p>The next most important class is the <tt class="docutils literal">SeqRecord</tt> or Sequence Record.
This holds a sequence (as a <tt class="docutils literal">Seq</tt> object) with additional annotation
including an identifier, name and description. The <tt class="docutils literal">Bio.SeqIO</tt> module
for reading and writing sequence file formats works with <tt class="docutils literal">SeqRecord</tt>
objects, which will be introduced below and covered in more detail by
Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>.</p>
<p>This covers the basic features and uses of the Biopython sequence class.
Now that you’ve got some idea of what it is like to interact with the
Biopython libraries, it’s time to delve into the fun, fun world of
dealing with biological file formats!</p>
</div>
<div class="section" id="a-usage-example">
<h2>2.3  A usage example</h2>
<p>Before we jump right into parsers and everything else to do with
Biopython, let’s set up an example to motivate everything we do and make
life more interesting. After all, if there wasn’t any biology in this
tutorial, why would you want you read it?</p>
<p>Since I love plants, I think we’re just going to have to have a plant
based example (sorry to all the fans of other organisms out there!).
Having just completed a recent trip to our local greenhouse, we’ve
suddenly developed an incredible obsession with Lady Slipper Orchids (if
you wonder why, have a look at some <a class="reference external" href="http://www.flickr.com/search/?q=lady+slipper+orchid&amp;s=int&amp;z=t">Lady Slipper Orchids photos on
Flickr</a>,
or try a <a class="reference external" href="http://images.google.com/images?q=lady%20slipper%20orchid">Google Image
Search</a>).</p>
<p>Of course, orchids are not only beautiful to look at, they are also
extremely interesting for people studying evolution and systematics. So
let’s suppose we’re thinking about writing a funding proposal to do a
molecular study of Lady Slipper evolution, and would like to see what
kind of research has already been done and how we can add to that.</p>
<p>After a little bit of reading up we discover that the Lady Slipper
Orchids are in the Orchidaceae family and the Cypripedioideae sub-family
and are made up of 5 genera: <em>Cypripedium</em>, <em>Paphiopedilum</em>,
<em>Phragmipedium</em>, <em>Selenipedium</em> and <em>Mexipedium</em>.</p>
<p>That gives us enough to get started delving for more information. So,
let’s look at how the Biopython tools can help us. We’ll start with
sequence parsing in Section <a class="reference external" href="#sec:sequence-parsing">2.4</a>, but the
orchids will be back later on as well - for example we’ll search PubMed
for papers about orchids and extract sequence data from GenBank in
Chapter <a class="reference external" href="#chapter:entrez">9</a>, extract data from Swiss-Prot from
certain orchid proteins in Chapter <a class="reference external" href="#chapter:swiss_prot">10</a>, and
work with ClustalW multiple sequence alignments of orchid proteins in
Section <a class="reference external" href="#sec:align_clustal">6.4.1</a>.</p>
</div>
<div class="section" id="parsing-sequence-file-formats">
<h2>2.4  Parsing sequence file formats</h2>
<p>A large part of much bioinformatics work involves dealing with the many
types of file formats designed to hold biological data. These files are
loaded with interesting biological data, and a special challenge is
parsing these files into a format so that you can manipulate them with
some kind of programming language. However the task of parsing these
files can be frustrated by the fact that the formats can change quite
regularly, and that formats may contain small subtleties which can break
even the most well designed parsers.</p>
<p>We are now going to briefly introduce the <tt class="docutils literal">Bio.SeqIO</tt> module – you can
find out more in Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>. We’ll start with
an online search for our friends, the lady slipper orchids. To keep this
introduction simple, we’re just using the NCBI website by hand. Let’s
just take a look through the nucleotide databases at NCBI, using an
Entrez online search
(<a class="reference external" href="http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?db=Nucleotide">http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?db=Nucleotide</a>)
for everything mentioning the text Cypripedioideae (this is the
subfamily of lady slipper orchids).</p>
<p>When this tutorial was originally written, this search gave us only 94
hits, which we saved as a FASTA formatted text file and as a GenBank
formatted text file (files
<tt class="docutils literal">`ls_orchid.fasta <span class="pre">&lt;http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta&gt;`__</span>
and
<span class="pre">```ls_orchid.gbk</span> <span class="pre">&lt;http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk&gt;`__,</span>
also included with the Biopython source code under
``docs/tutorial/examples/</tt>).</p>
<p>If you run the search today, you’ll get hundreds of results! When
following the tutorial, if you want to see the same list of genes, just
download the two files above or copy them from <tt class="docutils literal">docs/examples/</tt> in the
Biopython source code. In
Section <a class="reference external" href="#sec:connecting-with-biological-databases">2.5</a> we will
look at how to do a search like this from within Python.</p>
<div class="section" id="simple-fasta-parsing-example">
<h3>2.4.1  Simple FASTA parsing example</h3>
<p>If you open the lady slipper orchids FASTA file
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta">ls_orchid.fasta</a>
in your favourite text editor, you’ll see that the file starts like
this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">2765658</span><span class="operator">|</span><span class="name">emb</span><span class="operator">|</span><span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">CIZ78533</span> <span class="name">C</span><span class="operator">.</span><span class="name">irapeanum</span> <span class="literal number float">5.8</span><span class="name">S</span> <span class="name">rRNA</span> <span class="name">gene</span> <span class="operator word">and</span> <span class="name">ITS1</span> <span class="operator word">and</span> <span class="name">ITS2</span> <span class="name">DNA</span>
<span class="name">CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGGAATAAACGATCGAGTG</span>
<span class="name">AATCCGGAGGACCGGTGTACTCAGCTCACCGGGGGCATTGCTCCCGTGGTGACCCTGATTTGTTGTTGGG</span>
<span class="operator">...</span>
</pre>
<p>It contains 94 records, each has a line starting with “<tt class="docutils literal">&gt;</tt>”
(greater-than symbol) followed by the sequence on one or more lines. Now
try this in Python:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span>
    <span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
    <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="punctuation">)</span>
</pre>
<p>You should get something like this on your screen:</p>
<pre class="code python literal-block">
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">2765658</span><span class="operator">|</span><span class="name">emb</span><span class="operator">|</span><span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">CIZ78533</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC'</span><span class="punctuation">,</span> <span class="name">SingleLetterAlphabet</span><span class="punctuation">())</span>
<span class="literal number integer">740</span>
<span class="operator">...</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">2765564</span><span class="operator">|</span><span class="name">emb</span><span class="operator">|</span><span class="name">Z78439</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">PBZ78439</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC'</span><span class="punctuation">,</span> <span class="name">SingleLetterAlphabet</span><span class="punctuation">())</span>
<span class="literal number integer">592</span>
</pre>
<p>Notice that the FASTA format does not specify the alphabet, so
<tt class="docutils literal">Bio.SeqIO</tt> has defaulted to the rather generic
<tt class="docutils literal">SingleLetterAlphabet()</tt> rather than something DNA specific.</p>
</div>
<div class="section" id="simple-genbank-parsing-example">
<h3>2.4.2  Simple GenBank parsing example</h3>
<p>Now let’s load the GenBank file
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk">ls_orchid.gbk</a>
instead - notice that the code to do this is almost identical to the
snippet used above for the FASTA file - the only difference is we change
the filename and the format string:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span>
    <span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
    <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="punctuation">)</span>
</pre>
<p>This should give:</p>
<pre class="code python literal-block">
<span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
<span class="literal number integer">740</span>
<span class="operator">...</span>
<span class="name">Z78439</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
<span class="literal number integer">592</span>
</pre>
<p>This time <tt class="docutils literal">Bio.SeqIO</tt> has been able to choose a sensible alphabet,
IUPAC Ambiguous DNA. You’ll also notice that a shorter string has been
used as the <tt class="docutils literal">seq_record.id</tt> in this case.</p>
</div>
<div class="section" id="i-love-parsing-please-dont-stop-talking-about-it">
<h3>2.4.3  I love parsing – please don’t stop talking about it!</h3>
<p>Biopython has a lot of parsers, and each has its own little special
niches based on the sequence format it is parsing and all of that.
Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> covers <tt class="docutils literal">Bio.SeqIO</tt> in more
detail, while Chapter <a class="reference external" href="#chapter:Bio.AlignIO">6</a> introduces
<tt class="docutils literal">Bio.AlignIO</tt> for sequence alignments.</p>
<p>While the most popular file formats have parsers integrated into
<tt class="docutils literal">Bio.SeqIO</tt> and/or <tt class="docutils literal">Bio.AlignIO</tt>, for some of the rarer and unloved
file formats there is either no parser at all, or an old parser which
has not been linked in yet. Please also check the wiki pages
<a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a>
and
<a class="reference external" href="http://biopython.org/wiki/AlignIO">http://biopython.org/wiki/AlignIO</a>
for the latest information, or ask on the mailing list. The wiki pages
should include an up to date list of supported file types, and some
additional examples.</p>
<p>The next place to look for information about specific parsers and how to
do cool things with them is in the Cookbook
(Chapter <a class="reference external" href="#chapter:cookbook">18</a> of this Tutorial). If you don’t find
the information you are looking for, please consider helping out your
poor overworked documentors and submitting a cookbook entry about it!
(once you figure out how to do it, that is!)</p>
</div>
</div>
<div class="section" id="connecting-with-biological-databases">
<h2>2.5  Connecting with biological databases</h2>
<p>One of the very common things that you need to do in bioinformatics is
extract information from biological databases. It can be quite tedious
to access these databases manually, especially if you have a lot of
repetitive work to do. Biopython attempts to save you time and energy by
making some on-line databases available from Python scripts. Currently,
Biopython has code to extract information from the following databases:</p>
<ul class="simple">
<li><a class="reference external" href="http://www.ncbi.nlm.nih.gov/Entrez/">Entrez</a> (and
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/PubMed/">PubMed</a>) from the NCBI – See
Chapter <a class="reference external" href="#chapter:entrez">9</a>.</li>
<li><a class="reference external" href="http://www.expasy.org/">ExPASy</a> – See
Chapter <a class="reference external" href="#chapter:swiss_prot">10</a>.</li>
<li><a class="reference external" href="http://scop.mrc-lmb.cam.ac.uk/scop/">SCOP</a> – See the
<tt class="docutils literal">Bio.SCOP.search()</tt> function.</li>
</ul>
<p>The code in these modules basically makes it easy to write Python code
that interact with the CGI scripts on these pages, so that you can get
results in an easy to deal with format. In some cases, the results can
be tightly integrated with the Biopython parsers to make it even easier
to extract information.</p>
</div>
<div class="section" id="what-to-do-next">
<h2>2.6  What to do next</h2>
<p>Now that you’ve made it this far, you hopefully have a good
understanding of the basics of Biopython and are ready to start using it
for doing useful work. The best thing to do now is finish reading this
tutorial, and then if you want start snooping around in the source code,
and looking at the automatically generated documentation.</p>
<p>Once you get a picture of what you want to do, and what libraries in
Biopython will do it, you should take a peak at the Cookbook
(Chapter <a class="reference external" href="#chapter:cookbook">18</a>), which may have example code to do
something similar to what you want to do.</p>
<p>If you know what you want to do, but can’t figure out how to do it,
please feel free to post questions to the main Biopython list (see
<a class="reference external" href="http://biopython.org/wiki/Mailing_lists">http://biopython.org/wiki/Mailing_lists</a>).
This will not only help us answer your question, it will also allow us
to improve the documentation so it can help the next person do what you
want to do.</p>
<p>Enjoy the code!</p>
</div>
</div>
<div class="section" id="chapter-3-sequence-objects">
<h1>Chapter 3  Sequence objects</h1>
<p>Biological sequences are arguably the central object in Bioinformatics,
and in this chapter we’ll introduce the Biopython mechanism for dealing
with sequences, the <tt class="docutils literal">Seq</tt> object. Chapter <a class="reference external" href="#chapter:SeqRecord">4</a>
will introduce the related <tt class="docutils literal">SeqRecord</tt> object, which combines the
sequence information with any annotation, used again in
Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> for Sequence Input/Output.</p>
<p>Sequences are essentially strings of letters like <tt class="docutils literal">AGTACACTGGT</tt>, which
seems very natural since this is the most common way that sequences are
seen in biological file formats.</p>
<p>There are two important differences between <tt class="docutils literal">Seq</tt> objects and standard
Python strings. First of all, they have different methods. Although the
<tt class="docutils literal">Seq</tt> object supports many of the same methods as a plain string, its
<tt class="docutils literal">translate()</tt> method differs by doing biological translation, and
there are also additional biologically relevant methods like
<tt class="docutils literal">reverse_complement()</tt>. Secondly, the <tt class="docutils literal">Seq</tt> object has an important
attribute, <tt class="docutils literal">alphabet</tt>, which is an object describing what the
individual characters making up the sequence string “mean”, and how they
should be interpreted. For example, is <tt class="docutils literal">AGTACACTGGT</tt> a DNA sequence,
or just a protein sequence that happens to be rich in Alanines,
Glycines, Cysteines and Threonines?</p>
<div class="section" id="sequences-and-alphabets">
<h2>3.1  Sequences and Alphabets</h2>
<p>The alphabet object is perhaps the important thing that makes the
<tt class="docutils literal">Seq</tt> object more than just a string. The currently available
alphabets for Biopython are defined in the <tt class="docutils literal">Bio.Alphabet</tt> module.
We’ll use the IUPAC alphabets
(<a class="reference external" href="http://www.chem.qmw.ac.uk/iupac/">http://www.chem.qmw.ac.uk/iupac/</a>)
here to deal with some of our favorite objects: DNA, RNA and Proteins.</p>
<p><tt class="docutils literal">Bio.Alphabet.IUPAC</tt> provides basic definitions for proteins, DNA and
RNA, but additionally provides the ability to extend and customize the
basic definitions. For instance, for proteins, there is a basic
IUPACProtein class, but there is an additional ExtendedIUPACProtein
class providing for the additional elements “U” (or “Sec” for
selenocysteine) and “O” (or “Pyl” for pyrrolysine), plus the ambiguous
symbols “B” (or “Asx” for asparagine or aspartic acid), “Z” (or “Glx”
for glutamine or glutamic acid), “J” (or “Xle” for leucine isoleucine)
and “X” (or “Xxx” for an unknown amino acid). For DNA you’ve got choices
of IUPACUnambiguousDNA, which provides for just the basic letters,
IUPACAmbiguousDNA (which provides for ambiguity letters for every
possible situation) and ExtendedIUPACDNA, which allows letters for
modified bases. Similarly, RNA can be represented by IUPACAmbiguousRNA
or IUPACUnambiguousRNA.</p>
<p>The advantages of having an alphabet class are two fold. First, this
gives an idea of the type of information the Seq object contains.
Secondly, this provides a means of constraining the information, as a
means of type checking.</p>
<p>Now that we know what we are dealing with, let’s look at how to utilize
this class to do interesting work. You can create an ambiguous sequence
with the default generic alphabet like this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AGTACACTGGT&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGTACACTGGT'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">alphabet</span>
<span class="name">Alphabet</span><span class="punctuation">()</span>
</pre>
<p>However, where possible you should specify the alphabet explicitly when
creating your sequence objects - in this case an unambiguous DNA
alphabet object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AGTACACTGGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGTACACTGGT'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">alphabet</span>
<span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()</span>
</pre>
<p>Unless of course, this really is an amino acid sequence:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_prot</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AGTACACTGGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">protein</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_prot</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGTACACTGGT'</span><span class="punctuation">,</span> <span class="name">IUPACProtein</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_prot</span><span class="operator">.</span><span class="name">alphabet</span>
<span class="name">IUPACProtein</span><span class="punctuation">()</span>
</pre>
</div>
<div class="section" id="sequences-act-like-strings">
<h2>3.2  Sequences act like strings</h2>
<p>In many ways, we can deal with Seq objects as if they were normal Python
strings, for example getting the length, or iterating over the elements:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GATCG&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">index</span><span class="punctuation">,</span> <span class="name">letter</span> <span class="operator word">in</span> <span class="name builtin">enumerate</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">index</span><span class="punctuation">,</span> <span class="name">letter</span>
<span class="literal number integer">0</span> <span class="name">G</span>
<span class="literal number integer">1</span> <span class="name">A</span>
<span class="literal number integer">2</span> <span class="name">T</span>
<span class="literal number integer">3</span> <span class="name">C</span>
<span class="literal number integer">4</span> <span class="name">G</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
</pre>
<p>You can access elements of the sequence in the same way as for strings
(but remember, Python counts from zero!):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span> <span class="comment">#first letter</span>
<span class="name">G</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">]</span> <span class="comment">#third letter</span>
<span class="name">T</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span> <span class="comment">#last letter</span>
<span class="name">G</span>
</pre>
<p>The <tt class="docutils literal">Seq</tt> object has a <tt class="docutils literal">.count()</tt> method, just like a string. Note
that this means that like a Python string, this gives a
<em>non-overlapping</em> count:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="literal string">&quot;AAAA&quot;</span><span class="operator">.</span><span class="name">count</span><span class="punctuation">(</span><span class="literal string">&quot;AA&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AAAA&quot;</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">count</span><span class="punctuation">(</span><span class="literal string">&quot;AA&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">2</span>
</pre>
<p>For some biological uses, you may actually want an overlapping count
(i.e. 3 in this trivial example). When searching for single letters,
this makes no difference:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATCGATGGGCCTATATAGGATCGAAAATCGC'</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="punctuation">)</span>
<span class="literal number integer">32</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">count</span><span class="punctuation">(</span><span class="literal string">&quot;G&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">9</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="literal number integer">100</span> <span class="operator">*</span> <span class="name builtin">float</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="operator">.</span><span class="name">count</span><span class="punctuation">(</span><span class="literal string">&quot;G&quot;</span><span class="punctuation">)</span> <span class="operator">+</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">count</span><span class="punctuation">(</span><span class="literal string">&quot;C&quot;</span><span class="punctuation">))</span> <span class="operator">/</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="punctuation">)</span>
<span class="literal number float">46.875</span>
</pre>
<p>While you could use the above snippet of code to calculate a GC%, note
that the <tt class="docutils literal">Bio.SeqUtils</tt> module has several GC functions already built.
For example:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqUtils</span> <span class="keyword namespace">import</span> <span class="name">GC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATCGATGGGCCTATATAGGATCGAAAATCGC'</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">GC</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="punctuation">)</span>
<span class="literal number float">46.875</span>
</pre>
<p>Note that using the <tt class="docutils literal">Bio.SeqUtils.GC()</tt> function should automatically
cope with mixed case sequences and the ambiguous nucleotide S which
means G or C.</p>
<p>Also note that just like a normal Python string, the <tt class="docutils literal">Seq</tt> object is
in some ways “read-only”. If you need to edit your sequence, for example
simulating a point mutation, look at the
Section <a class="reference external" href="#sec:mutable-seq">3.12</a> below which talks about the
<tt class="docutils literal">MutableSeq</tt> object.</p>
</div>
<div class="section" id="slicing-a-sequence">
<h2>3.3  Slicing a sequence</h2>
<p>A more complicated example, let’s get a slice of the sequence:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GATCGATGGGCCTATATAGGATCGAAAATCGC&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">4</span><span class="punctuation">:</span><span class="literal number integer">12</span><span class="punctuation">]</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATGGGCC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>Two things are interesting to note. First, this follows the normal
conventions for Python strings. So the first element of the sequence is
0 (which is normal for computer science, but not so normal for biology).
When you do a slice the first item is included (i.e. 4 in this case) and
the last is excluded (12 in this case), which is the way things work in
Python, but of course not necessarily the way everyone in the world
would expect. The main goal is to stay consistent with what Python does.</p>
<p>The second thing to notice is that the slice is performed on the
sequence data string, but the new object produced is another <tt class="docutils literal">Seq</tt>
object which retains the alphabet information from the original <tt class="docutils literal">Seq</tt>
object.</p>
<p>Also like a Python string, you can do slices with a start, stop and
<em>stride</em> (the step size, which defaults to one). For example, we can get
the first, second and third codon positions of this DNA sequence:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">::</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GCTGTAGTAAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">::</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGGCATGCATC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">::</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TAGCTAAGAC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>Another stride trick you might have seen with a Python string is the use
of a -1 stride to reverse the string. You can do this with a <tt class="docutils literal">Seq</tt>
object too:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[::</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGCTAAAAGCTAGGATATATCCGGGTAGCTAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
</div>
<div class="section" id="turning-seq-objects-into-strings">
<h2>3.4  Turning Seq objects into strings</h2>
<p>If you really do just need a plain string, for example to write to a
file, or insert into a database, then this is very easy to get:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="punctuation">)</span>
<span class="literal string">'GATCGATGGGCCTATATAGGATCGAAAATCGC'</span>
</pre>
<p>Since calling <tt class="docutils literal">str()</tt> on a <tt class="docutils literal">Seq</tt> object returns the full sequence as
a string, you often don’t actually have to do this conversion
explicitly. Python does this automatically with a print statement:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_seq</span>
<span class="name">GATCGATGGGCCTATATAGGATCGAAAATCGC</span>
</pre>
<p>You can also use the <tt class="docutils literal">Seq</tt> object directly with a <tt class="docutils literal">%s</tt> placeholder
when using the Python string formatting or interpolation operator
(<tt class="docutils literal">%</tt>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">fasta_format_string</span> <span class="operator">=</span> <span class="literal string">&quot;&gt;Name</span><span class="literal string escape">\n</span><span class="literal string interpol">%s</span><span class="literal string escape">\n</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">my_seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">fasta_format_string</span>
<span class="operator">&gt;</span><span class="name">Name</span>
<span class="name">GATCGATGGGCCTATATAGGATCGAAAATCGC</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>This line of code constructs a simple FASTA format record (without
worrying about line wrapping). Section <a class="reference external" href="#sec:SeqRecord-format">4.5</a>
describes a neat way to get a FASTA formatted string from a
<tt class="docutils literal">SeqRecord</tt> object, while the more general topic of reading and
writing FASTA format sequence files is covered in
Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>.</p>
<p><em>NOTE:</em> If you are using Biopython 1.44 or older, using <tt class="docutils literal">str(my_seq)</tt>
will give just a truncated representation. Instead use
<tt class="docutils literal">my_seq.tostring()</tt> (which is still available in the current Biopython
releases for backwards compatibility):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">tostring</span><span class="punctuation">()</span>
<span class="literal string">'GATCGATGGGCCTATATAGGATCGAAAATCGC'</span>
</pre>
</div>
<div class="section" id="concatenating-or-adding-sequences">
<h2>3.5  Concatenating or adding sequences</h2>
<p>Naturally, you can in principle add any two Seq objects together - just
like you can with Python strings to concatenate them. However, you can’t
add sequences with incompatible alphabets, such as a protein sequence
and a DNA sequence:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">protein_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;EVRNAK&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">protein</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">protein_seq</span> <span class="operator">+</span> <span class="name">dna_seq</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
<span class="operator">...</span>
<span class="name exception">TypeError</span><span class="punctuation">:</span> <span class="name">Incompatible</span> <span class="name">alphabets</span> <span class="name">IUPACProtein</span><span class="punctuation">()</span> <span class="operator word">and</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()</span>
</pre>
<p>If you <em>really</em> wanted to do this, you’d have to first give both
sequences generic alphabets:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_alphabet</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">protein_seq</span><span class="operator">.</span><span class="name">alphabet</span> <span class="operator">=</span> <span class="name">generic_alphabet</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span><span class="operator">.</span><span class="name">alphabet</span> <span class="operator">=</span> <span class="name">generic_alphabet</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">protein_seq</span> <span class="operator">+</span> <span class="name">dna_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'EVRNAKACGT'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
</pre>
<p>Here is an example of adding a generic nucleotide sequence to an
unambiguous IUPAC DNA sequence, resulting in an ambiguous nucleotide
sequence:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_nucleotide</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nuc_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GATCGATGC&quot;</span><span class="punctuation">,</span> <span class="name">generic_nucleotide</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nuc_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATCGATGC'</span><span class="punctuation">,</span> <span class="name">NucleotideAlphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ACGT'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nuc_seq</span> <span class="operator">+</span> <span class="name">dna_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATCGATGCACGT'</span><span class="punctuation">,</span> <span class="name">NucleotideAlphabet</span><span class="punctuation">())</span>
</pre>
</div>
<div class="section" id="changing-case">
<h2>3.6  Changing case</h2>
<p>Python strings have very useful <tt class="docutils literal">upper</tt> and <tt class="docutils literal">lower</tt> methods for
changing the case. As of Biopython 1.53, the <tt class="docutils literal">Seq</tt> object gained
similar methods which are alphabet aware. For example,</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_dna</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;acgtACGT&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'acgtACGT'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ACGTACGT'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span><span class="operator">.</span><span class="name">lower</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'acgtacgt'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">())</span>
</pre>
<p>These are useful for doing case insensitive matching:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="literal string">&quot;GTAC&quot;</span> <span class="operator word">in</span> <span class="name">dna_seq</span>
<span class="name builtin pseudo">False</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="literal string">&quot;GTAC&quot;</span> <span class="operator word">in</span> <span class="name">dna_seq</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">()</span>
<span class="name builtin pseudo">True</span>
</pre>
<p>Note that strictly speaking the IUPAC alphabets are for upper case
sequences only, thus:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ACGT'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span><span class="operator">.</span><span class="name">lower</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'acgt'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">())</span>
</pre>
</div>
<div class="section" id="nucleotide-sequences-and-reverse-complements">
<h2>3.7  Nucleotide sequences and (reverse) complements</h2>
<p>For nucleotide sequences, you can easily obtain the complement or
reverse complement of a <tt class="docutils literal">Seq</tt> object using its built-in methods:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GATCGATGGGCCTATATAGGATCGAAAATCGC&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATCGATGGGCCTATATAGGATCGAAAATCGC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">complement</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CTAGCTACCCGGATATATCCTAGCTTTTAGCG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GCGATTTTCGATCCTATATAGGCCCATCGATC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>As mentioned earlier, an easy way to just reverse a <tt class="docutils literal">Seq</tt> object (or a
Python string) is slice it with -1 step:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[::</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGCTAAAAGCTAGGATATATCCGGGTAGCTAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>In all of these operations, the alphabet property is maintained. This is
very useful in case you accidentally end up trying to do something weird
like take the (reverse)complement of a protein sequence:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">protein_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;EVRNAK&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">protein</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">protein_seq</span><span class="operator">.</span><span class="name">complement</span><span class="punctuation">()</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
<span class="operator">...</span>
<span class="name exception">ValueError</span><span class="punctuation">:</span> <span class="name">Proteins</span> <span class="name">do</span> <span class="operator word">not</span> <span class="name">have</span> <span class="name">complements</span><span class="error">!</span>
</pre>
<p>The example in Section <a class="reference external" href="#sec:SeqIO-reverse-complement">5.5.3</a>
combines the <tt class="docutils literal">Seq</tt> object’s reverse complement method with
<tt class="docutils literal">Bio.SeqIO</tt> for sequence input/output.</p>
</div>
<div class="section" id="transcription">
<h2>3.8  Transcription</h2>
<p>Before talking about transcription, I want to try and clarify the strand
issue. Consider the following (made up) stretch of double stranded DNA
which encodes a short peptide:</p>
<div class="formula">
<br/>    <sub><i>DNA</i><i>coding</i><i>strand</i>(<i>aka</i><i>Crick</i><i>strand</i>, <i>strand</i> + 1)</sub>    <br/>5’    <tt> ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG</tt>    3’<br/>    <tt> |||||||||||||||||||||||||||||||||||||||</tt>    <br/>3’    <tt> TACCGGTAACATTACCCGGCGACTTTCCCACGGGCTATC</tt>    5’<br/>    <sub><i>DNA</i><i>template</i><i>strand</i>(<i>aka</i><i>Watson</i><i>strand</i>, <i>strand</i> − 1)</sub>    <br/><br/>    |    <br/>    <i>Transcription</i>    <br/>    ↓    <br/><br/>5’    <tt> AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG</tt>    3’<br/>    <sub><i>Single</i><i>stranded</i><i>messenger</i><i>RNA</i></sub>    <br/><br/>
</div>
<p>The actual biological transcription process works from the template
strand, doing a reverse complement (TCAG → CUGA) to give the mRNA.
However, in Biopython and bioinformatics in general, we typically work
directly with the coding strand because this means we can get the mRNA
sequence just by switching T → U.</p>
<p>Now let’s actually get down to doing a transcription in Biopython.
First, let’s create <tt class="docutils literal">Seq</tt> objects for the coding and template DNA
strands:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">template_dna</span> <span class="operator">=</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">template_dna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CTATCGGGCACCCTTTCAGCGGCCCATTACAATGGCCAT'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>These should match the figure above - remember by convention nucleotide
sequences are normally read from the 5’ to 3’ direction, while in the
figure the template strand is shown reversed.</p>
<p>Now let’s transcribe the coding strand into the corresponding mRNA,
using the <tt class="docutils literal">Seq</tt> object’s built in <tt class="docutils literal">transcribe</tt> method:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span> <span class="operator">=</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">transcribe</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousRNA</span><span class="punctuation">())</span>
</pre>
<p>As you can see, all this does is switch T → U, and adjust the alphabet.</p>
<p>If you do want to do a true biological transcription starting with the
template strand, then this becomes a two-step process:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">template_dna</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">transcribe</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousRNA</span><span class="punctuation">())</span>
</pre>
<p>The <tt class="docutils literal">Seq</tt> object also includes a back-transcription method for going
from the mRNA to the coding strand of the DNA. Again, this is a simple U
→ T substitution and associated change of alphabet:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_rna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousRNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span><span class="operator">.</span><span class="name">back_transcribe</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p><em>Note:</em> The <tt class="docutils literal">Seq</tt> object’s <tt class="docutils literal">transcribe</tt> and <tt class="docutils literal">back_transcribe</tt>
methods were added in Biopython 1.49. For older releases you would have
to use the <tt class="docutils literal">Bio.Seq</tt> module’s functions instead, see
Section <a class="reference external" href="#sec:seq-module-functions">3.14</a>.</p>
</div>
<div class="section" id="translation">
<h2>3.9  Translation</h2>
<p>Sticking with the same example discussed in the transcription section
above, now let’s translate this mRNA into the corresponding protein
sequence - again taking advantage of one of the <tt class="docutils literal">Seq</tt> object’s
biological methods:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_rna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousRNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGR*KGAR*'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">))</span>
</pre>
<p>You can also translate directly from the coding strand DNA sequence:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGR*KGAR*'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">))</span>
</pre>
<p>You should notice in the above protein sequences that in addition to the
end stop character, there is an internal stop as well. This was a
deliberate choice of example, as it gives an excuse to talk about some
optional arguments, including different translation tables (Genetic
Codes).</p>
<p>The translation tables available in Biopython are based on those <a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi">from
the NCBI</a>
(see the next section of this tutorial). By default, translation will
use the <em>standard</em> genetic code (NCBI table id 1). Suppose we are
dealing with a mitochondrial sequence. We need to tell the translation
function to use the relevant genetic code instead:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal string">&quot;Vertebrate Mitochondrial&quot;</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGRWKGAR*'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">))</span>
</pre>
<p>You can also specify the table using the NCBI table number which is
shorter, and often included in the feature annotation of GenBank files:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGRWKGAR*'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">))</span>
</pre>
<p>Now, you may want to translate the nucleotides up to the first in frame
stop codon, and then stop (as happens in nature):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGR*KGAR*'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">to_stop</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGR'</span><span class="punctuation">,</span> <span class="name">IUPACProtein</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGRWKGAR*'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name">to_stop</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGRWKGAR'</span><span class="punctuation">,</span> <span class="name">IUPACProtein</span><span class="punctuation">())</span>
</pre>
<p>Notice that when you use the <tt class="docutils literal">to_stop</tt> argument, the stop codon itself
is not translated - and the stop symbol is not included at the end of
your protein sequence.</p>
<p>You can even specify the stop symbol if you don’t like the default
asterisk:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name">stop_symbol</span><span class="operator">=</span><span class="literal string">&quot;&#64;&quot;</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGRWKGAR&#64;'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'&#64;'</span><span class="punctuation">))</span>
</pre>
<p>Now, suppose you have a complete coding sequence CDS, which is to say a
nucleotide sequence (e.g. mRNA – after any splicing) which is a whole
number of codons (i.e. the length is a multiple of three), commences
with a start codon, ends with a stop codon, and has no internal in-frame
stop codons. In general, given a complete CDS, the default translate
method will do what you want (perhaps with the <tt class="docutils literal">to_stop</tt> option).
However, what if your sequence uses a non-standard start codon? This
happens a lot in bacteria – for example the gene yaaX in <tt class="docutils literal">E. coli</tt>
K12:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_dna</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gene</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GTGAAAAAGATGCAATCTATCGTACTCGCACTTTCCCTGGTTCTGGTCGCTCCCATGGCA&quot;</span> <span class="operator">+</span> \
<span class="operator">...</span>            <span class="literal string">&quot;GCACAGGCTGCGGAAATTACGTTAGTCCCGTCAGTAAAATTACAGATAGGCGATCGTGAT&quot;</span> <span class="operator">+</span> \
<span class="operator">...</span>            <span class="literal string">&quot;AATCGTGGCTATTACTGGGATGGAGGTCACTGGCGCGACCACGGCTGGTGGAAACAACAT&quot;</span> <span class="operator">+</span> \
<span class="operator">...</span>            <span class="literal string">&quot;TATGAATGGCGAGGCAATCGCTGGCACCTACACGGACCGCCGCCACCGCCGCGCCACCAT&quot;</span> <span class="operator">+</span> \
<span class="operator">...</span>            <span class="literal string">&quot;AAGAAAGCTCCTCATGATCATCACGGCGGTCATGGTCCAGGCAAACATCACCGCTAA&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>            <span class="name">generic_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gene</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal string">&quot;Bacterial&quot;</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'VKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYWDGGHWRDH...HR*'</span><span class="punctuation">,</span>
<span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">ExtendedIUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gene</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal string">&quot;Bacterial&quot;</span><span class="punctuation">,</span> <span class="name">to_stop</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'VKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYWDGGHWRDH...HHR'</span><span class="punctuation">,</span>
<span class="name">ExtendedIUPACProtein</span><span class="punctuation">())</span>
</pre>
<p>In the bacterial genetic code <tt class="docutils literal">GTG</tt> is a valid start codon, and while
it does <em>normally</em> encode Valine, if used as a start codon it should be
translated as methionine. This happens if you tell Biopython your
sequence is a complete CDS:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gene</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal string">&quot;Bacterial&quot;</span><span class="punctuation">,</span> <span class="name">cds</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYWDGGHWRDH...HHR'</span><span class="punctuation">,</span>
<span class="name">ExtendedIUPACProtein</span><span class="punctuation">())</span>
</pre>
<p>In addition to telling Biopython to translate an alternative start codon
as methionine, using this option also makes sure your sequence really is
a valid CDS (you’ll get an exception if not).</p>
<p>The example in Section <a class="reference external" href="#sec:SeqIO-translate">18.1.3</a> combines the
<tt class="docutils literal">Seq</tt> object’s translate method with <tt class="docutils literal">Bio.SeqIO</tt> for sequence
input/output.</p>
</div>
<div class="section" id="translation-tables">
<h2>3.10  Translation Tables</h2>
<p>In the previous sections we talked about the <tt class="docutils literal">Seq</tt> object translation
method (and mentioned the equivalent function in the <tt class="docutils literal">Bio.Seq</tt> module
– see Section <a class="reference external" href="#sec:seq-module-functions">3.14</a>). Internally these
use codon table objects derived from the NCBI information at
<tt class="docutils literal"><span class="pre">`ftp://ftp.ncbi.nlm.nih.gov/entrez/misc/data/gc.prt</span></tt> &lt;<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/entrez/misc/data/gc.prt">ftp://ftp.ncbi.nlm.nih.gov/entrez/misc/data/gc.prt</a>&gt;`__,
also shown on
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi">http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi</a>
in a much more readable layout.</p>
<p>As before, let’s just focus on two choices: the Standard translation
table, and the translation table for Vertebrate Mitochondrial DNA.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Data</span> <span class="keyword namespace">import</span> <span class="name">CodonTable</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">standard_table</span> <span class="operator">=</span> <span class="name">CodonTable</span><span class="operator">.</span><span class="name">unambiguous_dna_by_name</span><span class="punctuation">[</span><span class="literal string">&quot;Standard&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mito_table</span> <span class="operator">=</span> <span class="name">CodonTable</span><span class="operator">.</span><span class="name">unambiguous_dna_by_name</span><span class="punctuation">[</span><span class="literal string">&quot;Vertebrate Mitochondrial&quot;</span><span class="punctuation">]</span>
</pre>
<p>Alternatively, these tables are labeled with ID numbers 1 and 2,
respectively:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Data</span> <span class="keyword namespace">import</span> <span class="name">CodonTable</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">standard_table</span> <span class="operator">=</span> <span class="name">CodonTable</span><span class="operator">.</span><span class="name">unambiguous_dna_by_id</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mito_table</span> <span class="operator">=</span> <span class="name">CodonTable</span><span class="operator">.</span><span class="name">unambiguous_dna_by_id</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">]</span>
</pre>
<p>You can compare the actual tables visually by printing them:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">standard_table</span>
<span class="name">Table</span> <span class="literal number integer">1</span> <span class="name">Standard</span><span class="punctuation">,</span> <span class="name">SGC0</span>

  <span class="operator">|</span>  <span class="name">T</span>      <span class="operator">|</span>  <span class="name">C</span>      <span class="operator">|</span>  <span class="name">A</span>      <span class="operator">|</span>  <span class="name">G</span>      <span class="operator">|</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTT</span> <span class="name">F</span>   <span class="operator">|</span> <span class="name">TCT</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAT</span> <span class="name">Y</span>   <span class="operator">|</span> <span class="name">TGT</span> <span class="name">C</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTC</span> <span class="name">F</span>   <span class="operator">|</span> <span class="name">TCC</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAC</span> <span class="name">Y</span>   <span class="operator">|</span> <span class="name">TGC</span> <span class="name">C</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTA</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">TCA</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAA</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">TGA</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">A</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTG</span> <span class="name">L</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">TCG</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAG</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">TGG</span> <span class="name">W</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTT</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCT</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAT</span> <span class="name">H</span>   <span class="operator">|</span> <span class="name">CGT</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTC</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCC</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAC</span> <span class="name">H</span>   <span class="operator">|</span> <span class="name">CGC</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTA</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCA</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAA</span> <span class="name">Q</span>   <span class="operator">|</span> <span class="name">CGA</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">A</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTG</span> <span class="name">L</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">CCG</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAG</span> <span class="name">Q</span>   <span class="operator">|</span> <span class="name">CGG</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATT</span> <span class="name">I</span>   <span class="operator">|</span> <span class="name">ACT</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAT</span> <span class="name">N</span>   <span class="operator">|</span> <span class="name">AGT</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATC</span> <span class="name">I</span>   <span class="operator">|</span> <span class="name">ACC</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAC</span> <span class="name">N</span>   <span class="operator">|</span> <span class="name">AGC</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATA</span> <span class="name">I</span>   <span class="operator">|</span> <span class="name">ACA</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAA</span> <span class="name">K</span>   <span class="operator">|</span> <span class="name">AGA</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">A</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATG</span> <span class="name">M</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">ACG</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAG</span> <span class="name">K</span>   <span class="operator">|</span> <span class="name">AGG</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTT</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCT</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAT</span> <span class="name">D</span>   <span class="operator">|</span> <span class="name">GGT</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTC</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCC</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAC</span> <span class="name">D</span>   <span class="operator">|</span> <span class="name">GGC</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTA</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCA</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAA</span> <span class="name">E</span>   <span class="operator">|</span> <span class="name">GGA</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">A</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTG</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCG</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAG</span> <span class="name">E</span>   <span class="operator">|</span> <span class="name">GGG</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
</pre>
<p>and:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">mito_table</span>
<span class="name">Table</span> <span class="literal number integer">2</span> <span class="name">Vertebrate</span> <span class="name">Mitochondrial</span><span class="punctuation">,</span> <span class="name">SGC1</span>

  <span class="operator">|</span>  <span class="name">T</span>      <span class="operator">|</span>  <span class="name">C</span>      <span class="operator">|</span>  <span class="name">A</span>      <span class="operator">|</span>  <span class="name">G</span>      <span class="operator">|</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTT</span> <span class="name">F</span>   <span class="operator">|</span> <span class="name">TCT</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAT</span> <span class="name">Y</span>   <span class="operator">|</span> <span class="name">TGT</span> <span class="name">C</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTC</span> <span class="name">F</span>   <span class="operator">|</span> <span class="name">TCC</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAC</span> <span class="name">Y</span>   <span class="operator">|</span> <span class="name">TGC</span> <span class="name">C</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTA</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">TCA</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAA</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">TGA</span> <span class="name">W</span>   <span class="operator">|</span> <span class="name">A</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTG</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">TCG</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAG</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">TGG</span> <span class="name">W</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTT</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCT</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAT</span> <span class="name">H</span>   <span class="operator">|</span> <span class="name">CGT</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTC</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCC</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAC</span> <span class="name">H</span>   <span class="operator">|</span> <span class="name">CGC</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTA</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCA</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAA</span> <span class="name">Q</span>   <span class="operator">|</span> <span class="name">CGA</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">A</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTG</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCG</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAG</span> <span class="name">Q</span>   <span class="operator">|</span> <span class="name">CGG</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATT</span> <span class="name">I</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">ACT</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAT</span> <span class="name">N</span>   <span class="operator">|</span> <span class="name">AGT</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATC</span> <span class="name">I</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">ACC</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAC</span> <span class="name">N</span>   <span class="operator">|</span> <span class="name">AGC</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATA</span> <span class="name">M</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">ACA</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAA</span> <span class="name">K</span>   <span class="operator">|</span> <span class="name">AGA</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">A</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATG</span> <span class="name">M</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">ACG</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAG</span> <span class="name">K</span>   <span class="operator">|</span> <span class="name">AGG</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTT</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCT</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAT</span> <span class="name">D</span>   <span class="operator">|</span> <span class="name">GGT</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTC</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCC</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAC</span> <span class="name">D</span>   <span class="operator">|</span> <span class="name">GGC</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTA</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCA</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAA</span> <span class="name">E</span>   <span class="operator">|</span> <span class="name">GGA</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">A</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTG</span> <span class="name">V</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">GCG</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAG</span> <span class="name">E</span>   <span class="operator">|</span> <span class="name">GGG</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
</pre>
<p>You may find these following properties useful – for example if you are
trying to do your own gene finding:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mito_table</span><span class="operator">.</span><span class="name">stop_codons</span>
<span class="punctuation">[</span><span class="literal string">'TAA'</span><span class="punctuation">,</span> <span class="literal string">'TAG'</span><span class="punctuation">,</span> <span class="literal string">'AGA'</span><span class="punctuation">,</span> <span class="literal string">'AGG'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mito_table</span><span class="operator">.</span><span class="name">start_codons</span>
<span class="punctuation">[</span><span class="literal string">'ATT'</span><span class="punctuation">,</span> <span class="literal string">'ATC'</span><span class="punctuation">,</span> <span class="literal string">'ATA'</span><span class="punctuation">,</span> <span class="literal string">'ATG'</span><span class="punctuation">,</span> <span class="literal string">'GTG'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mito_table</span><span class="operator">.</span><span class="name">forward_table</span><span class="punctuation">[</span><span class="literal string">&quot;ACG&quot;</span><span class="punctuation">]</span>
<span class="literal string">'T'</span>
</pre>
</div>
<div class="section" id="comparing-seq-objects">
<h2>3.11  Comparing Seq objects</h2>
<p>Sequence comparison is actually a very complicated topic, and there is
no easy way to decide if two sequences are equal. The basic problem is
the meaning of the letters in a sequence are context dependent - the
letter “A” could be part of a DNA, RNA or protein sequence. Biopython
uses alphabet objects as part of each <tt class="docutils literal">Seq</tt> object to try and capture
this information - so comparing two <tt class="docutils literal">Seq</tt> objects means considering
both the sequence strings <em>and</em> the alphabets.</p>
<p>For example, you might argue that the two DNA <tt class="docutils literal">Seq</tt> objects
<tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;,</span> IUPAC.unambiguous_dna)</tt> and
<tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;,</span> IUPAC.ambiguous_dna)</tt> should be equal, even though they
do have different alphabets. Depending on the context this could be
important.</p>
<p>This gets worse – suppose you think
<tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;,</span> IUPAC.unambiguous_dna)</tt> and <tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;)</span></tt> (i.e. the
default generic alphabet) should be equal. Then, logically,
<tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;,</span> IUPAC.protein)</tt> and <tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;)</span></tt> should also be equal.
Now, in logic if <em>A</em>=<em>B</em> and <em>B</em>=<em>C</em>, by transitivity we expect
<em>A</em>=<em>C</em>. So for logical consistency we’d require
<tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;,</span> IUPAC.unambiguous_dna)</tt> and
<tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;,</span> IUPAC.protein)</tt> to be equal – which most people would
agree is just not right. This transitivity problem would also have
implications for using <tt class="docutils literal">Seq</tt> objects as Python dictionary keys.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq1</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq2</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
</pre>
<p>So, what does Biopython do? Well, the equality test is the default for
Python objects – it tests to see if they are the same object in memory.
This is a very strict test:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq1</span> <span class="operator">==</span> <span class="name">seq2</span>
<span class="name builtin pseudo">False</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq1</span> <span class="operator">==</span> <span class="name">seq1</span>
<span class="name builtin pseudo">True</span>
</pre>
<p>If you actually want to do this, you can be more explicit by using the
Python <tt class="docutils literal">id</tt> function,</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">id</span><span class="punctuation">(</span><span class="name">seq1</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="name builtin">id</span><span class="punctuation">(</span><span class="name">seq2</span><span class="punctuation">)</span>
<span class="name builtin pseudo">False</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">id</span><span class="punctuation">(</span><span class="name">seq1</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="name builtin">id</span><span class="punctuation">(</span><span class="name">seq1</span><span class="punctuation">)</span>
<span class="name builtin pseudo">True</span>
</pre>
<p>Now, in every day use, your sequences will probably all have the same
alphabet, or at least all be the same type of sequence (all DNA, all
RNA, or all protein). What you probably want is to just compare the
sequences as strings – so do this explicitly:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">seq1</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">seq2</span><span class="punctuation">)</span>
<span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">seq1</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">seq1</span><span class="punctuation">)</span>
<span class="name builtin pseudo">True</span>
</pre>
<p>As an extension to this, while you can use a Python dictionary with
<tt class="docutils literal">Seq</tt> objects as keys, it is generally more useful to use the sequence
a string for the key. See also Section <a class="reference external" href="#sec:seq-to-string">3.4</a>.</p>
</div>
<div class="section" id="mutableseq-objects">
<h2>3.12  MutableSeq objects</h2>
<p>Just like the normal Python string, the <tt class="docutils literal">Seq</tt> object is “read only”,
or in Python terminology, immutable. Apart from wanting the <tt class="docutils literal">Seq</tt>
object to act like a string, this is also a useful default since in many
biological applications you want to ensure you are not changing your
sequence data:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GCCATTGTAATGGGCCGCTGAAAGGGTGCCCGA&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
</pre>
<p>Observe what happens if you try to edit the sequence:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">5</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="literal string">&quot;G&quot;</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
<span class="operator">...</span>
<span class="name exception">TypeError</span><span class="punctuation">:</span> <span class="literal string">'Seq'</span> <span class="name builtin">object</span> <span class="name">does</span> <span class="operator word">not</span> <span class="name">support</span> <span class="name">item</span> <span class="name">assignment</span>
</pre>
<p>However, you can convert it into a mutable sequence (a <tt class="docutils literal">MutableSeq</tt>
object) and do pretty much anything you want with it:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span> <span class="operator">=</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">tomutable</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span>
<span class="name">MutableSeq</span><span class="punctuation">(</span><span class="literal string">'GCCATTGTAATGGGCCGCTGAAAGGGTGCCCGA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>Alternatively, you can create a <tt class="docutils literal">MutableSeq</tt> object directly from a
string:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">MutableSeq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span> <span class="operator">=</span> <span class="name">MutableSeq</span><span class="punctuation">(</span><span class="literal string">&quot;GCCATTGTAATGGGCCGCTGAAAGGGTGCCCGA&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
</pre>
<p>Either way will give you a sequence object which can be changed:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span>
<span class="name">MutableSeq</span><span class="punctuation">(</span><span class="literal string">'GCCATTGTAATGGGCCGCTGAAAGGGTGCCCGA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span><span class="punctuation">[</span><span class="literal number integer">5</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="literal string">&quot;C&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span>
<span class="name">MutableSeq</span><span class="punctuation">(</span><span class="literal string">'GCCATCGTAATGGGCCGCTGAAAGGGTGCCCGA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span><span class="operator">.</span><span class="name">remove</span><span class="punctuation">(</span><span class="literal string">&quot;T&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span>
<span class="name">MutableSeq</span><span class="punctuation">(</span><span class="literal string">'GCCACGTAATGGGCCGCTGAAAGGGTGCCCGA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span><span class="operator">.</span><span class="name">reverse</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span>
<span class="name">MutableSeq</span><span class="punctuation">(</span><span class="literal string">'AGCCCGTGGGAAAGTCGCCGGGTAATGCACCG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>Do note that unlike the <tt class="docutils literal">Seq</tt> object, the <tt class="docutils literal">MutableSeq</tt> object’s
methods like <tt class="docutils literal">reverse_complement()</tt> and <tt class="docutils literal">reverse()</tt> act in-situ!</p>
<p>An important technical difference between mutable and immutable objects
in Python means that you can’t use a <tt class="docutils literal">MutableSeq</tt> object as a
dictionary key, but you can use a Python string or a <tt class="docutils literal">Seq</tt> object in
this way.</p>
<p>Once you have finished editing your a <tt class="docutils literal">MutableSeq</tt> object, it’s easy
to get back to a read-only <tt class="docutils literal">Seq</tt> object should you need to:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">new_seq</span> <span class="operator">=</span> <span class="name">mutable_seq</span><span class="operator">.</span><span class="name">toseq</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">new_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGCCCGTGGGAAAGTCGCCGGGTAATGCACCG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>You can also get a string from a <tt class="docutils literal">MutableSeq</tt> object just like from a
<tt class="docutils literal">Seq</tt> object (Section <a class="reference external" href="#sec:seq-to-string">3.4</a>).</p>
</div>
<div class="section" id="unknownseq-objects">
<h2>3.13  UnknownSeq objects</h2>
<p>The <tt class="docutils literal">UnknownSeq</tt> object is a subclass of the basic <tt class="docutils literal">Seq</tt> object and
its purpose is to represent a sequence where we know the length, but not
the actual letters making it up. You could of course use a normal
<tt class="docutils literal">Seq</tt> object in this situation, but it wastes rather a lot of memory
to hold a string of a million “N” characters when you could just store a
single letter “N” and the desired length as an integer.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">UnknownSeq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk</span> <span class="operator">=</span> <span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">Alphabet</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'?'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">unk</span>
<span class="error">????????????????????</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">unk</span><span class="punctuation">)</span>
<span class="literal number integer">20</span>
</pre>
<p>You can of course specify an alphabet, meaning for nucleotide sequences
the letter defaults to “N” and for proteins “X”, rather than just “?”.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">UnknownSeq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_dna</span> <span class="operator">=</span> <span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span><span class="operator">=</span><span class="name">IUPAC</span><span class="operator">.</span><span class="name">ambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_dna</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'N'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">unk_dna</span>
<span class="name">NNNNNNNNNNNNNNNNNNNN</span>
</pre>
<p>You can use all the usual <tt class="docutils literal">Seq</tt> object methods too, note these give
back memory saving <tt class="docutils literal">UnknownSeq</tt> objects where appropriate as you might
expect:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_dna</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'N'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_dna</span><span class="operator">.</span><span class="name">complement</span><span class="punctuation">()</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'N'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_dna</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'N'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_dna</span><span class="operator">.</span><span class="name">transcribe</span><span class="punctuation">()</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">IUPACAmbiguousRNA</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'N'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_protein</span> <span class="operator">=</span> <span class="name">unk_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_protein</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">ProteinAlphabet</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'X'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">unk_protein</span>
<span class="name">XXXXXX</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">unk_protein</span><span class="punctuation">)</span>
<span class="literal number integer">6</span>
</pre>
<p>You may be able to find a use for the <tt class="docutils literal">UnknownSeq</tt> object in your own
code, but it is more likely that you will first come across them in a
<tt class="docutils literal">SeqRecord</tt> object created by <tt class="docutils literal">Bio.SeqIO</tt> (see
Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>). Some sequence file formats don’t
always include the actual sequence, for example GenBank and EMBL files
may include a list of features but for the sequence just present the
contig information. Alternatively, the QUAL files used in sequencing
work hold quality scores but they <em>never</em> contain a sequence – instead
there is a partner FASTA file which <em>does</em> have the sequence.</p>
</div>
<div class="section" id="working-with-directly-strings">
<h2>3.14  Working with directly strings</h2>
<p>To close this chapter, for those you who <em>really</em> don’t want to use the
sequence objects (or who prefer a functional programming style to an
object orientated one), there are module level functions in <tt class="docutils literal">Bio.Seq</tt>
will accept plain Python strings, <tt class="docutils literal">Seq</tt> objects (including
<tt class="docutils literal">UnknownSeq</tt> objects) or <tt class="docutils literal">MutableSeq</tt> objects:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">reverse_complement</span><span class="punctuation">,</span> <span class="name">transcribe</span><span class="punctuation">,</span> <span class="name">back_transcribe</span><span class="punctuation">,</span> <span class="name">translate</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_string</span> <span class="operator">=</span> <span class="literal string">&quot;GCTGTTATGGGTCGTTGGAAGGGTGGTCGTGCTGCTGGTTAG&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name">my_string</span><span class="punctuation">)</span>
<span class="literal string">'CTAACCAGCAGCACGACCACCCTTCCAACGACCCATAACAGC'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">transcribe</span><span class="punctuation">(</span><span class="name">my_string</span><span class="punctuation">)</span>
<span class="literal string">'GCUGUUAUGGGUCGUUGGAAGGGUGGUCGUGCUGCUGGUUAG'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">back_transcribe</span><span class="punctuation">(</span><span class="name">my_string</span><span class="punctuation">)</span>
<span class="literal string">'GCTGTTATGGGTCGTTGGAAGGGTGGTCGTGCTGCTGGTTAG'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">translate</span><span class="punctuation">(</span><span class="name">my_string</span><span class="punctuation">)</span>
<span class="literal string">'AVMGRWKGGRAAG*'</span>
</pre>
<p>You are, however, encouraged to work with <tt class="docutils literal">Seq</tt> objects by default.</p>
</div>
</div>
<div class="section" id="chapter-4-sequence-annotation-objects">
<h1>Chapter 4  Sequence annotation objects</h1>
<p>Chapter <a class="reference external" href="#chapter:Bio.Seq">3</a> introduced the sequence classes.
Immediately “above” the <tt class="docutils literal">Seq</tt> class is the Sequence Record or
<tt class="docutils literal">SeqRecord</tt> class, defined in the <tt class="docutils literal">Bio.SeqRecord</tt> module. This class
allows higher level features such as identifiers and features (as
<tt class="docutils literal">SeqFeature</tt> objects) to be associated with the sequence, and is used
throughout the sequence input/output interface <tt class="docutils literal">Bio.SeqIO</tt> described
fully in Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>.</p>
<p>If you are only going to be working with simple data like FASTA files,
you can probably skip this chapter for now. If on the other hand you are
going to be using richly annotated sequence data, say from GenBank or
EMBL files, this information is quite important.</p>
<p>While this chapter should cover most things to do with the <tt class="docutils literal">SeqRecord</tt>
and <tt class="docutils literal">SeqFeature</tt> objects in this chapter, you may also want to read
the <tt class="docutils literal">SeqRecord</tt> wiki page
(<a class="reference external" href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a>),
and the built in documentation (also online –
<a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SeqRecord.SeqRecord-class.html">SeqRecord</a>
and
<a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SeqFeature.SeqFeature-class.html">SeqFeature</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">SeqRecord</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<div class="section" id="the-seqrecord-object">
<h2>4.1  The SeqRecord object</h2>
<p>The <tt class="docutils literal">SeqRecord</tt> (Sequence Record) class is defined in the
<tt class="docutils literal">Bio.SeqRecord</tt> module. This class allows higher level features such
as identifiers and features to be associated with a sequence (see
Chapter <a class="reference external" href="#chapter:Bio.Seq">3</a>), and is the basic data type for the
<tt class="docutils literal">Bio.SeqIO</tt> sequence input/output interface (see
Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>).</p>
<p>The <tt class="docutils literal">SeqRecord</tt> class itself is quite simple, and offers the following
information as attributes:</p>
<dl class="docutils">
<dt><strong>.seq</strong></dt>
<dd>– The sequence itself, typically a <tt class="docutils literal">Seq</tt> object.</dd>
<dt><strong>.id</strong></dt>
<dd>– The primary ID used to identify the sequence – a string. In most
cases this is something like an accession number.</dd>
<dt><strong>.name</strong></dt>
<dd>– A “common” name/id for the sequence – a string. In some cases this
will be the same as the accession number, but it could also be a
clone name. I think of this as being analogous to the LOCUS id in a
GenBank record.</dd>
<dt><strong>.description</strong></dt>
<dd>– A human readable description or expressive name for the sequence –
a string.</dd>
<dt><strong>.letter_annotations</strong></dt>
<dd>– Holds per-letter-annotations using a (restricted) dictionary of
additional information about the letters in the sequence. The keys
are the name of the information, and the information is contained in
the value as a Python sequence (i.e. a list, tuple or string) with
the same length as the sequence itself. This is often used for
quality scores (e.g.
Section <a class="reference external" href="#sec:FASTQ-filtering-example">18.1.6</a>) or secondary
structure information (e.g. from Stockholm/PFAM alignment files).</dd>
<dt><strong>.annotations</strong></dt>
<dd>– A dictionary of additional information about the sequence. The
keys are the name of the information, and the information is
contained in the value. This allows the addition of more
“unstructured” information to the sequence.</dd>
<dt><strong>.features</strong></dt>
<dd>– A list of <tt class="docutils literal">SeqFeature</tt> objects with more structured information
about the features on a sequence (e.g. position of genes on a
genome, or domains on a protein sequence). The structure of sequence
features is described below in
Section <a class="reference external" href="#sec:seq_features">4.3</a>.</dd>
<dt><strong>.dbxrefs</strong></dt>
<dd>– A list of database cross-references as strings.</dd>
</dl>
</div>
<div class="section" id="creating-a-seqrecord">
<h2>4.2  Creating a SeqRecord</h2>
<p>Using a <tt class="docutils literal">SeqRecord</tt> object is not very complicated, since all of the
information is presented as attributes of the class. Usually you won’t
create a <tt class="docutils literal">SeqRecord</tt> “by hand”, but instead use <tt class="docutils literal">Bio.SeqIO</tt> to read
in a sequence file for you (see Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> and
the examples below). However, creating <tt class="docutils literal">SeqRecord</tt> can be quite
simple.</p>
<div class="section" id="seqrecord-objects-from-scratch">
<h3>4.2.1  SeqRecord objects from scratch</h3>
<p>To create a <tt class="docutils literal">SeqRecord</tt> at a minimum you just need a <tt class="docutils literal">Seq</tt> object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GATC&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">simple_seq</span><span class="punctuation">)</span>
</pre>
<p>Additionally, you can also pass the id, name and description to the
initialization function, but if not they will be set as strings
indicating they are unknown, and can be modified subsequently:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">id</span>
<span class="literal string">'&lt;unknown id&gt;'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">id</span> <span class="operator">=</span> <span class="literal string">&quot;AC12345&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;Made up sequence I wish I could write a paper about&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">description</span>
<span class="name">Made</span> <span class="name">up</span> <span class="name">sequence</span> <span class="name">I</span> <span class="name">wish</span> <span class="name">I</span> <span class="name">could</span> <span class="name">write</span> <span class="name">a</span> <span class="name">paper</span> <span class="name">about</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATC'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
</pre>
<p>Including an identifier is very important if you want to output your
<tt class="docutils literal">SeqRecord</tt> to a file. You would normally include this when creating
the object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GATC&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">simple_seq</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;AC12345&quot;</span><span class="punctuation">)</span>
</pre>
<p>As mentioned above, the <tt class="docutils literal">SeqRecord</tt> has an dictionary attribute
<tt class="docutils literal">annotations</tt>. This is used for any miscellaneous annotations that
doesn’t fit under one of the other more specific attributes. Adding
annotations is easy, and just involves dealing directly with the
annotation dictionary:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;evidence&quot;</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="literal string">&quot;None. I just made it up.&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">annotations</span>
<span class="punctuation">{</span><span class="literal string">'evidence'</span><span class="punctuation">:</span> <span class="literal string">'None. I just made it up.'</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;evidence&quot;</span><span class="punctuation">]</span>
<span class="name builtin pseudo">None</span><span class="operator">.</span> <span class="name">I</span> <span class="name">just</span> <span class="name">made</span> <span class="name">it</span> <span class="name">up</span><span class="operator">.</span>
</pre>
<p>Working with per-letter-annotations is similar, <tt class="docutils literal">letter_annotations</tt>
is a dictionary like attribute which will let you assign any Python
sequence (i.e. a string, list or tuple) which has the same length as the
sequence:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">40</span><span class="punctuation">,</span><span class="literal number integer">40</span><span class="punctuation">,</span><span class="literal number integer">38</span><span class="punctuation">,</span><span class="literal number integer">30</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">letter_annotations</span>
<span class="punctuation">{</span><span class="literal string">'phred_quality'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">38</span><span class="punctuation">,</span> <span class="literal number integer">30</span><span class="punctuation">]}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">38</span><span class="punctuation">,</span> <span class="literal number integer">30</span><span class="punctuation">]</span>
</pre>
<p>The <tt class="docutils literal">dbxrefs</tt> and <tt class="docutils literal">features</tt> attributes are just Python lists, and
should be used to store strings and <tt class="docutils literal">SeqFeature</tt> objects (discussed
later in this chapter) respectively.</p>
</div>
<div class="section" id="seqrecord-objects-from-fasta-files">
<h3>4.2.2  SeqRecord objects from FASTA files</h3>
<p>This example uses a fairly large FASTA file containing the whole
sequence for <em>Yersinia pestis biovar Microtus</em> str. 91001 plasmid pPCP1,
originally downloaded from the NCBI. This file is included with the
Biopython unit tests under the GenBank folder, or online
<a class="reference external" href="http://biopython.org/SRC/biopython/Tests/GenBank/NC_005816.fna">NC_005816.fna</a>
from our website.</p>
<p>The file starts like this - and you can check there is only one record
present (i.e. only one line starting with a greater than symbol):</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">45478711</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NC_005816</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="name">Yersinia</span> <span class="name">pestis</span> <span class="name">biovar</span> <span class="name">Microtus</span> <span class="operator">...</span> <span class="name">pPCP1</span><span class="punctuation">,</span> <span class="name">complete</span> <span class="name">sequence</span>
<span class="name">TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGGGGGTAATCTGCTCTCC</span>
<span class="operator">...</span>
</pre>
<p>Back in Chapter <a class="reference external" href="#chapter:quick-start">2</a> you will have seen the
function <tt class="docutils literal"><span class="pre">Bio.SeqIO.parse(...)</span></tt> used to loop over all the records in a
file as <tt class="docutils literal">SeqRecord</tt> objects. The <tt class="docutils literal">Bio.SeqIO</tt> module has a sister
function for use on files which contain just one record which we’ll use
here (see Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> for details):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.fna&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'</span><span class="punctuation">,</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|45478711|ref|NC_005816.1|'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'gi|45478711|ref|NC_005816.1|'</span><span class="punctuation">,</span>
<span class="name">description</span><span class="operator">=</span><span class="literal string">'gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus ... sequence'</span><span class="punctuation">,</span>
<span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[])</span>
</pre>
<p>Now, let’s have a look at the key attributes of this <tt class="docutils literal">SeqRecord</tt>
individually – starting with the <tt class="docutils literal">seq</tt> attribute which gives you a
<tt class="docutils literal">Seq</tt> object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'</span><span class="punctuation">,</span> <span class="name">SingleLetterAlphabet</span><span class="punctuation">())</span>
</pre>
<p>Here <tt class="docutils literal">Bio.SeqIO</tt> has defaulted to a generic alphabet, rather than
guessing that this is DNA. If you know in advance what kind of sequence
your FASTA file contains, you can tell <tt class="docutils literal">Bio.SeqIO</tt> which alphabet to
use (see Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>).</p>
<p>Next, the identifiers and description:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span>
<span class="literal string">'gi|45478711|ref|NC_005816.1|'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'gi|45478711|ref|NC_005816.1|'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">description</span>
<span class="literal string">'gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus ... pPCP1, complete sequence'</span>
</pre>
<p>As you can see above, the first word of the FASTA record’s title line
(after removing the greater than symbol) is used for both the <tt class="docutils literal">id</tt> and
<tt class="docutils literal">name</tt> attributes. The whole title line (after removing the greater
than symbol) is used for the record description. This is deliberate,
partly for backwards compatibility reasons, but it also makes sense if
you have a FASTA file like this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">Yersinia</span> <span class="name">pestis</span> <span class="name">biovar</span> <span class="name">Microtus</span> <span class="name builtin">str</span><span class="operator">.</span> <span class="literal number integer">91001</span> <span class="name">plasmid</span> <span class="name">pPCP1</span>
<span class="name">TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGGGGGTAATCTGCTCTCC</span>
<span class="operator">...</span>
</pre>
<p>Note that none of the other annotation attributes get populated when
reading a FASTA file:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="punctuation">[]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">annotations</span>
<span class="punctuation">{}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">letter_annotations</span>
<span class="punctuation">{}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span>
<span class="punctuation">[]</span>
</pre>
<p>In this case our example FASTA file was from the NCBI, and they have a
fairly well defined set of conventions for formatting their FASTA lines.
This means it would be possible to parse this information and extract
the GI number and accession for example. However, FASTA files from other
sources vary, so this isn’t possible in general.</p>
</div>
<div class="section" id="seqrecord-objects-from-genbank-files">
<h3>4.2.3  SeqRecord objects from GenBank files</h3>
<p>As in the previous example, we’re going to look at the whole sequence
for <em>Yersinia pestis biovar Microtus</em> str. 91001 plasmid pPCP1,
originally downloaded from the NCBI, but this time as a GenBank file.
Again, this file is included with the Biopython unit tests under the
GenBank folder, or online
<a class="reference external" href="http://biopython.org/SRC/biopython/Tests/GenBank/NC_005816.gb">NC_005816.gb</a>
from our website.</p>
<p>This file contains a single record (i.e. only one LOCUS line) and
starts:</p>
<pre class="code python literal-block">
<span class="name">LOCUS</span>       <span class="name">NC_005816</span>               <span class="literal number integer">9609</span> <span class="name">bp</span>    <span class="name">DNA</span>     <span class="name">circular</span> <span class="name">BCT</span> <span class="literal number integer">21</span><span class="operator">-</span><span class="name">JUL</span><span class="operator">-</span><span class="literal number integer">2008</span>
<span class="name">DEFINITION</span>  <span class="name">Yersinia</span> <span class="name">pestis</span> <span class="name">biovar</span> <span class="name">Microtus</span> <span class="name builtin">str</span><span class="operator">.</span> <span class="literal number integer">91001</span> <span class="name">plasmid</span> <span class="name">pPCP1</span><span class="punctuation">,</span> <span class="name">complete</span>
            <span class="name">sequence</span><span class="operator">.</span>
<span class="name">ACCESSION</span>   <span class="name">NC_005816</span>
<span class="name">VERSION</span>     <span class="name">NC_005816</span><span class="operator">.</span><span class="literal number integer">1</span>  <span class="name">GI</span><span class="punctuation">:</span><span class="literal number integer">45478711</span>
<span class="name">PROJECT</span>     <span class="name">GenomeProject</span><span class="punctuation">:</span><span class="literal number integer">10638</span>
<span class="operator">...</span>
</pre>
<p>Again, we’ll use <tt class="docutils literal">Bio.SeqIO</tt> to read this file in, and the code is
almost identical to that for used above for the FASTA file (see
Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> for details):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'</span><span class="punctuation">,</span>
<span class="name">IUPACAmbiguousDNA</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'NC_005816.1'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'NC_005816'</span><span class="punctuation">,</span>
<span class="name">description</span><span class="operator">=</span><span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span><span class="punctuation">,</span>
<span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">'Project:10638'</span><span class="punctuation">])</span>
</pre>
<p>You should be able to spot some differences already! But taking the
attributes individually, the sequence string is the same as before, but
this time <tt class="docutils literal">Bio.SeqIO</tt> has been able to automatically assign a more
specific alphabet (see Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> for
details):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>The <tt class="docutils literal">name</tt> comes from the LOCUS line, while the <tt class="docutils literal">id</tt> includes the
version suffix. The description comes from the DEFINITION line:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span>
<span class="literal string">'NC_005816.1'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'NC_005816'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">description</span>
<span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span>
</pre>
<p>GenBank files don’t have any per-letter annotations:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">letter_annotations</span>
<span class="punctuation">{}</span>
</pre>
<p>Most of the annotations information gets recorded in the <tt class="docutils literal">annotations</tt>
dictionary, for example:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">)</span>
<span class="literal number integer">11</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;source&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Yersinia pestis biovar Microtus str. 91001'</span>
</pre>
<p>The <tt class="docutils literal">dbxrefs</tt> list gets populated from any PROJECT or DBLINK lines:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="punctuation">[</span><span class="literal string">'Project:10638'</span><span class="punctuation">]</span>
</pre>
<p>Finally, and perhaps most interestingly, all the entries in the features
table (e.g. the genes or CDS features) get recorded as <tt class="docutils literal">SeqFeature</tt>
objects in the <tt class="docutils literal">features</tt> list.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">)</span>
<span class="literal number integer">29</span>
</pre>
<p>We’ll talk about <tt class="docutils literal">SeqFeature</tt> objects next, in
Section <a class="reference external" href="#sec:seq_features">4.3</a>.</p>
</div>
</div>
<div class="section" id="feature-location-and-position-objects">
<h2>4.3  Feature, location and position objects</h2>
<div class="section" id="seqfeature-objects">
<h3>4.3.1  SeqFeature objects</h3>
<p>Sequence features are an essential part of describing a sequence. Once
you get beyond the sequence itself, you need some way to organize and
easily get at the more “abstract” information that is known about the
sequence. While it is probably impossible to develop a general sequence
feature class that will cover everything, the Biopython <tt class="docutils literal">SeqFeature</tt>
class attempts to encapsulate as much of the information about the
sequence as possible. The design is heavily based on the GenBank/EMBL
feature tables, so if you understand how they look, you’ll probably have
an easier time grasping the structure of the Biopython classes.</p>
<p>The key idea about each <tt class="docutils literal">SeqFeature</tt> object is to describe a region on
a parent sequence, typically a <tt class="docutils literal">SeqRecord</tt> object. That region is
described with a location object, typically a range between two
positions (see Section <a class="reference external" href="#sec:locations">4.3.2</a> below).</p>
<p>The <tt class="docutils literal">SeqFeature</tt> class has a number of attributes, so first we’ll list
them and their general features, and then later in the chapter work
through examples to show how this applies to a real life example. The
attributes of a SeqFeature are:</p>
<dl class="docutils">
<dt><strong>.type</strong></dt>
<dd>– This is a textual description of the type of feature (for
instance, this will be something like ‘CDS’ or ‘gene’).</dd>
<dt><strong>.location</strong></dt>
<dd><p class="first">– The location of the <tt class="docutils literal">SeqFeature</tt> on the sequence that you are
dealing with, see Section <a class="reference external" href="#sec:locations">4.3.2</a> below. The
<tt class="docutils literal">SeqFeature</tt> delegates much of its functionality to the location
object, and includes a number of shortcut attributes for properties
of the location:</p>
<dl class="last docutils">
<dt><strong>.ref</strong></dt>
<dd>– shorthand for <tt class="docutils literal">.location.ref</tt> – any (different) reference
sequence the location is referring to. Usually just None.</dd>
<dt><strong>.ref_db</strong></dt>
<dd>– shorthand for <tt class="docutils literal">.location.ref_db</tt> – specifies the database
any identifier in <tt class="docutils literal">.ref</tt> refers to. Usually just None.</dd>
<dt><strong>.strand</strong></dt>
<dd>– shorthand for <tt class="docutils literal">.location.strand</tt> – the strand on the
sequence that the feature is located on. For double stranded
nucleotide sequence this may either be 1 for the top strand, ?1
for the bottom strand, 0 if the strand is important but is
unknown, or <tt class="docutils literal">None</tt> if it doesn’t matter. This is None for
proteins, or single stranded sequences.</dd>
</dl>
</dd>
<dt><strong>.qualifiers</strong></dt>
<dd>– This is a Python dictionary of additional information about the
feature. The key is some kind of terse one-word description of what
the information contained in the value is about, and the value is
the actual information. For example, a common key for a qualifier
might be “evidence” and the value might be “computational
(non-experimental).” This is just a way to let the person who is
looking at the feature know that it has not be experimentally
(i. e. in a wet lab) confirmed. Note that other the value will be a
list of strings (even when there is only one string). This is a
reflection of the feature tables in GenBank/EMBL files.</dd>
<dt><strong>.sub_features</strong></dt>
<dd>– This used to be used to represent features with complicated
locations like ‘joins’ in GenBank/EMBL files. This has been
deprecated with the introduction of the <tt class="docutils literal">CompoundLocation</tt> object,
and should now be ignored.</dd>
</dl>
</div>
<div class="section" id="positions-and-locations">
<h3>4.3.2  Positions and locations</h3>
<p>The key idea about each <tt class="docutils literal">SeqFeature</tt> object is to describe a region on
a parent sequence, for which we use a location object, typically
describing a range between two positions. Two try to clarify the
terminology we’re using:</p>
<dl class="docutils">
<dt><strong>position</strong></dt>
<dd>– This refers to a single position on a sequence, which may be fuzzy
or not. For instance, 5, 20, <tt class="docutils literal">&lt;100</tt> and <tt class="docutils literal">&gt;200</tt> are all
positions.</dd>
<dt><strong>location</strong></dt>
<dd>– A location is region of sequence bounded by some positions. For
instance 5..20 (i. e. 5 to 20) is a location.</dd>
</dl>
<p>I just mention this because sometimes I get confused between the two.</p>
<div class="section" id="featurelocation-object">
<h4>4.3.2.1  FeatureLocation object</h4>
<p>Unless you work with eukaryotic genes, most <tt class="docutils literal">SeqFeature</tt> locations are
extremely simple - you just need start and end coordinates and a strand.
That’s essentially all the basic <tt class="docutils literal">FeatureLocation</tt> object does.</p>
<p>In practise of course, things can be more complicated. First of all we
have to handle compound locations made up of several regions. Secondly,
the positions themselves may be fuzzy (inexact).</p>
</div>
<div class="section" id="compoundlocation-object">
<h4>4.3.2.2  CompoundLocation object</h4>
<p>Biopython 1.62 introduced the <tt class="docutils literal">CompoundLocation</tt> as part of a
restructuring of how complex locations made up of multiple regions are
represented. The main usage is for handling ‘join’ locations in
EMBL/GenBank files.</p>
</div>
<div class="section" id="fuzzy-positions">
<h4>4.3.2.3  Fuzzy Positions</h4>
<p>So far we’ve only used simple positions. One complication in dealing
with feature locations comes in the positions themselves. In biology
many times things aren’t entirely certain (as much as us wet lab
biologists try to make them certain!). For instance, you might do a
dinucleotide priming experiment and discover that the start of mRNA
transcript starts at one of two sites. This is very useful information,
but the complication comes in how to represent this as a position. To
help us deal with this, we have the concept of fuzzy positions.
Basically there are several types of fuzzy positions, so we have five
classes do deal with them:</p>
<dl class="docutils">
<dt><strong>ExactPosition</strong></dt>
<dd>– As its name suggests, this class represents a position which is
specified as exact along the sequence. This is represented as just a
number, and you can get the position by looking at the <tt class="docutils literal">position</tt>
attribute of the object.</dd>
<dt><strong>BeforePosition</strong></dt>
<dd>– This class represents a fuzzy position that occurs prior to some
specified site. In GenBank/EMBL notation, this is represented as
something like <tt class="docutils literal">`&lt;13'</tt>, signifying that the real position is
located somewhere less than 13. To get the specified upper boundary,
look at the <tt class="docutils literal">position</tt> attribute of the object.</dd>
<dt><strong>AfterPosition</strong></dt>
<dd>– Contrary to <tt class="docutils literal">BeforePosition</tt>, this class represents a position
that occurs after some specified site. This is represented in
GenBank as <tt class="docutils literal">`&gt;13'</tt>, and like <tt class="docutils literal">BeforePosition</tt>, you get the
boundary number by looking at the <tt class="docutils literal">position</tt> attribute of the
object.</dd>
<dt><strong>WithinPosition</strong></dt>
<dd>– Occasionally used for GenBank/EMBL locations, this class models a
position which occurs somewhere between two specified nucleotides.
In GenBank/EMBL notation, this would be represented as ‘(1.5)’, to
represent that the position is somewhere within the range 1 to 5. To
get the information in this class you have to look at two
attributes. The <tt class="docutils literal">position</tt> attribute specifies the lower boundary
of the range we are looking at, so in our example case this would be
one. The <tt class="docutils literal">extension</tt> attribute specifies the range to the higher
boundary, so in this case it would be 4. So <tt class="docutils literal">object.position</tt> is
the lower boundary and <tt class="docutils literal">object.position + object.extension</tt> is the
upper boundary.</dd>
<dt><strong>OneOfPosition</strong></dt>
<dd>– Occasionally used for GenBank/EMBL locations, this class deals
with a position where several possible values exist, for instance
you could use this if the start codon was unclear and there where
two candidates for the start of the gene. Alternatively, that might
be handled explicitly as two related gene features.</dd>
<dt><strong>UnknownPosition</strong></dt>
<dd>– This class deals with a position of unknown location. This is not
used in GenBank/EMBL, but corresponds to the ‘?’ feature coordinate
used in UniProt.</dd>
</dl>
<p>Here’s an example where we create a location with fuzzy end points:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqFeature</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">start_pos</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="operator">.</span><span class="name">AfterPosition</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">end_pos</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="operator">.</span><span class="name">BetweenPosition</span><span class="punctuation">(</span><span class="literal number integer">9</span><span class="punctuation">,</span> <span class="name">left</span><span class="operator">=</span><span class="literal number integer">8</span><span class="punctuation">,</span> <span class="name">right</span><span class="operator">=</span><span class="literal number integer">9</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_location</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="operator">.</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="name">start_pos</span><span class="punctuation">,</span> <span class="name">end_pos</span><span class="punctuation">)</span>
</pre>
<p>Note that the details of some of the fuzzy-locations changed in
Biopython 1.59, in particular for BetweenPosition and WithinPosition you
must now make it explicit which integer position should be used for
slicing etc. For a start position this is generally the lower (left)
value, while for an end position this would generally be the higher
(right) value.</p>
<p>If you print out a <tt class="docutils literal">FeatureLocation</tt> object, you can get a nice
representation of the information:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_location</span>
<span class="punctuation">[</span><span class="operator">&gt;</span><span class="literal number integer">5</span><span class="punctuation">:(</span><span class="literal number integer">8</span><span class="operator">^</span><span class="literal number integer">9</span><span class="punctuation">)]</span>
</pre>
<p>We can access the fuzzy start and end positions using the start and end
attributes of the location:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_location</span><span class="operator">.</span><span class="name">start</span>
<span class="name">AfterPosition</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_location</span><span class="operator">.</span><span class="name">start</span>
<span class="operator">&gt;</span><span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_location</span><span class="operator">.</span><span class="name">end</span>
<span class="name">BetweenPosition</span><span class="punctuation">(</span><span class="literal number integer">9</span><span class="punctuation">,</span> <span class="name">left</span><span class="operator">=</span><span class="literal number integer">8</span><span class="punctuation">,</span> <span class="name">right</span><span class="operator">=</span><span class="literal number integer">9</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_location</span><span class="operator">.</span><span class="name">end</span>
<span class="punctuation">(</span><span class="literal number integer">8</span><span class="operator">^</span><span class="literal number integer">9</span><span class="punctuation">)</span>
</pre>
<p>If you don’t want to deal with fuzzy positions and just want numbers,
they are actually subclasses of integers so should work like integers:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">int</span><span class="punctuation">(</span><span class="name">my_location</span><span class="operator">.</span><span class="name">start</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">int</span><span class="punctuation">(</span><span class="name">my_location</span><span class="operator">.</span><span class="name">end</span><span class="punctuation">)</span>
<span class="literal number integer">9</span>
</pre>
<p>For compatibility with older versions of Biopython you can ask for the
<tt class="docutils literal">nofuzzy_start</tt> and <tt class="docutils literal">nofuzzy_end</tt> attributes of the location which
are plain integers:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_location</span><span class="operator">.</span><span class="name">nofuzzy_start</span>
<span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_location</span><span class="operator">.</span><span class="name">nofuzzy_end</span>
<span class="literal number integer">9</span>
</pre>
<p>Notice that this just gives you back the position attributes of the
fuzzy locations.</p>
<p>Similarly, to make it easy to create a position without worrying about
fuzzy positions, you can just pass in numbers to the <tt class="docutils literal">FeaturePosition</tt>
constructors, and you’ll get back out <tt class="docutils literal">ExactPosition</tt> objects:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exact_location</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="operator">.</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="literal number integer">9</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">exact_location</span>
<span class="punctuation">[</span><span class="literal number integer">5</span><span class="punctuation">:</span><span class="literal number integer">9</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exact_location</span><span class="operator">.</span><span class="name">start</span>
<span class="name">ExactPosition</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">int</span><span class="punctuation">(</span><span class="name">exact_location</span><span class="operator">.</span><span class="name">start</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exact_location</span><span class="operator">.</span><span class="name">nofuzzy_start</span>
<span class="literal number integer">5</span>
</pre>
<p>That is most of the nitty gritty about dealing with fuzzy positions in
Biopython. It has been designed so that dealing with fuzziness is not
that much more complicated than dealing with exact positions, and
hopefully you find that true!</p>
</div>
<div class="section" id="location-testing">
<h4>4.3.2.4  Location testing</h4>
<p>You can use the Python keyword <tt class="docutils literal">in</tt> with a <tt class="docutils literal">SeqFeature</tt> or location
object to see if the base/residue for a parent coordinate is within the
feature/location or not.</p>
<p>For example, suppose you have a SNP of interest and you want to know
which features this SNP is within, and lets suppose this SNP is at index
4350 (Python counting!). Here is a simple brute force solution where we
just check all the features one by one in a loop:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_snp</span> <span class="operator">=</span> <span class="literal number integer">4350</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">feature</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">my_snp</span> <span class="operator word">in</span> <span class="name">feature</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="name">feature</span><span class="operator">.</span><span class="name">type</span><span class="punctuation">,</span> <span class="name">feature</span><span class="operator">.</span><span class="name">qualifiers</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">'db_xref'</span><span class="punctuation">)</span>
<span class="operator">...</span>
<span class="name">source</span> <span class="punctuation">[</span><span class="literal string">'taxon:229193'</span><span class="punctuation">]</span>
<span class="name">gene</span> <span class="punctuation">[</span><span class="literal string">'GeneID:2767712'</span><span class="punctuation">]</span>
<span class="name">CDS</span> <span class="punctuation">[</span><span class="literal string">'GI:45478716'</span><span class="punctuation">,</span> <span class="literal string">'GeneID:2767712'</span><span class="punctuation">]</span>
</pre>
<p>Note that gene and CDS features from GenBank or EMBL files defined with
joins are the union of the exons – they do not cover any introns.</p>
</div>
</div>
<div class="section" id="sequence-described-by-a-feature-or-location">
<h3>4.3.3  Sequence described by a feature or location</h3>
<p>A <tt class="docutils literal">SeqFeature</tt> or location object doesn’t directly contain a sequence,
instead the location (see Section <a class="reference external" href="#sec:locations">4.3.2</a>)
describes how to get this from the parent sequence. For example consider
a (short) gene sequence with location 5:18 on the reverse strand, which
in GenBank/EMBL notation using 1-based counting would be
<tt class="docutils literal"><span class="pre">complement(6..18)</span></tt>, like this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqFeature</span> <span class="keyword namespace">import</span> <span class="name">SeqFeature</span><span class="punctuation">,</span> <span class="name">FeatureLocation</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">example_parent</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACCGAGACGGCAAAGGCTAGCATAGGTATGAGACTTCCTTCCTGCCAGTGCTGAGGAACTGGGAGCCTAC&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">example_feature</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="punctuation">(</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="literal number integer">18</span><span class="punctuation">),</span> <span class="name builtin">type</span><span class="operator">=</span><span class="literal string">&quot;gene&quot;</span><span class="punctuation">,</span> <span class="name">strand</span><span class="operator">=-</span><span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>You could take the parent sequence, slice it to extract 5:18, and then
take the reverse complement. If you are using Biopython 1.59 or later,
the feature location’s start and end are integer like so this works:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">feature_seq</span> <span class="operator">=</span> <span class="name">example_parent</span><span class="punctuation">[</span><span class="name">example_feature</span><span class="operator">.</span><span class="name">location</span><span class="operator">.</span><span class="name">start</span><span class="punctuation">:</span><span class="name">example_feature</span><span class="operator">.</span><span class="name">location</span><span class="operator">.</span><span class="name">end</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">feature_seq</span>
<span class="name">AGCCTTTGCCGTC</span>
</pre>
<p>This is a simple example so this isn’t too bad – however once you have
to deal with compound features (joins) this is rather messy. Instead,
the <tt class="docutils literal">SeqFeature</tt> object has an <tt class="docutils literal">extract</tt> method to take care of all
this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">feature_seq</span> <span class="operator">=</span> <span class="name">example_feature</span><span class="operator">.</span><span class="name">extract</span><span class="punctuation">(</span><span class="name">example_parent</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">feature_seq</span>
<span class="name">AGCCTTTGCCGTC</span>
</pre>
<p>The length of a <tt class="docutils literal">SeqFeature</tt> or location matches that of the region of
sequence it describes.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">example_feature</span><span class="operator">.</span><span class="name">extract</span><span class="punctuation">(</span><span class="name">example_parent</span><span class="punctuation">)</span>
<span class="name">AGCCTTTGCCGTC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">example_feature</span><span class="operator">.</span><span class="name">extract</span><span class="punctuation">(</span><span class="name">example_parent</span><span class="punctuation">))</span>
<span class="literal number integer">13</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">example_feature</span><span class="punctuation">)</span>
<span class="literal number integer">13</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">example_feature</span><span class="operator">.</span><span class="name">location</span><span class="punctuation">)</span>
<span class="literal number integer">13</span>
</pre>
<p>For simple <tt class="docutils literal">FeatureLocation</tt> objects the length is just the difference
between the start and end positions. However, for a <tt class="docutils literal">CompoundLocation</tt>
the length is the sum of the constituent regions.</p>
</div>
</div>
<div class="section" id="references">
<h2>4.4  References</h2>
<p>Another common annotation related to a sequence is a reference to a
journal or other published work dealing with the sequence. We have a
fairly simple way of representing a Reference in Biopython – we have a
<tt class="docutils literal">Bio.SeqFeature.Reference</tt> class that stores the relevant information
about a reference as attributes of an object.</p>
<p>The attributes include things that you would expect to see in a
reference like <tt class="docutils literal">journal</tt>, <tt class="docutils literal">title</tt> and <tt class="docutils literal">authors</tt>. Additionally, it
also can hold the <tt class="docutils literal">medline_id</tt> and <tt class="docutils literal">pubmed_id</tt> and a <tt class="docutils literal">comment</tt>
about the reference. These are all accessed simply as attributes of the
object.</p>
<p>A reference also has a <tt class="docutils literal">location</tt> object so that it can specify a
particular location on the sequence that the reference refers to. For
instance, you might have a journal that is dealing with a particular
gene located on a BAC, and want to specify that it only refers to this
position exactly. The <tt class="docutils literal">location</tt> is a potentially fuzzy location, as
described in section <a class="reference external" href="#sec:locations">4.3.2</a>.</p>
<p>Any reference objects are stored as a list in the <tt class="docutils literal">SeqRecord</tt> object’s
<tt class="docutils literal">annotations</tt> dictionary under the key “references”. That’s all there
is too it. References are meant to be easy to deal with, and hopefully
general enough to cover lots of usage cases.</p>
</div>
<div class="section" id="the-format-method">
<h2>4.5  The format method</h2>
<p>The <tt class="docutils literal">format()</tt> method of the <tt class="docutils literal">SeqRecord</tt> class gives a string
containing your record formatted using one of the output file formats
supported by <tt class="docutils literal">Bio.SeqIO</tt>, such as FASTA:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_protein</span>

<span class="name">record</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD&quot;</span> \
                      <span class="operator">+</span><span class="literal string">&quot;GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK&quot;</span> \
                      <span class="operator">+</span><span class="literal string">&quot;NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM&quot;</span> \
                      <span class="operator">+</span><span class="literal string">&quot;SSAC&quot;</span><span class="punctuation">,</span> <span class="name">generic_protein</span><span class="punctuation">),</span>
                   <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;gi|14150838|gb|AAK54648.1|AF376133_1&quot;</span><span class="punctuation">,</span>
                   <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;chalcone synthase [Cucumis sativus]&quot;</span><span class="punctuation">)</span>

<span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>which should give:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">14150838</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AAK54648</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF376133_1</span> <span class="name">chalcone</span> <span class="name">synthase</span> <span class="punctuation">[</span><span class="name">Cucumis</span> <span class="name">sativus</span><span class="punctuation">]</span>
<span class="name">MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD</span>
<span class="name">GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK</span>
<span class="name">NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM</span>
<span class="name">SSAC</span>
</pre>
<p>This <tt class="docutils literal">format</tt> method takes a single mandatory argument, a lower case
string which is supported by <tt class="docutils literal">Bio.SeqIO</tt> as an output format (see
Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>). However, some of the file formats
<tt class="docutils literal">Bio.SeqIO</tt> can write to <em>require</em> more than one record (typically the
case for multiple sequence alignment formats), and thus won’t work via
this <tt class="docutils literal">format()</tt> method. See also
Section <a class="reference external" href="#sec:Bio.SeqIO-and-StringIO">5.5.4</a>.</p>
</div>
<div class="section" id="slicing-a-seqrecord">
<h2>4.6  Slicing a SeqRecord</h2>
<p>You can slice a <tt class="docutils literal">SeqRecord</tt>, to give you a new <tt class="docutils literal">SeqRecord</tt> covering
just part of the sequence. What is important here is that any per-letter
annotations are also sliced, and any features which fall completely
within the new sequence are preserved (with their locations adjusted).</p>
<p>For example, taking the same GenBank file used earlier:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'</span><span class="punctuation">,</span>
<span class="name">IUPACAmbiguousDNA</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'NC_005816.1'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'NC_005816'</span><span class="punctuation">,</span>
<span class="name">description</span><span class="operator">=</span><span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span><span class="punctuation">,</span>
<span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">'Project:10638'</span><span class="punctuation">])</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="literal number integer">9609</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">)</span>
<span class="literal number integer">41</span>
</pre>
<p>For this example we’re going to focus in on the <tt class="docutils literal">pim</tt> gene,
<tt class="docutils literal">YP_pPCP05</tt>. If you have a look at the GenBank file directly you’ll
find this gene/CDS has location string <tt class="docutils literal"><span class="pre">4343..4780</span></tt>, or in Python
counting <tt class="docutils literal">4342:4780</tt>. From looking at the file you can work out that
these are the twelfth and thirteenth entries in the file, so in Python
zero-based counting they are entries 11 and 12 in the <tt class="docutils literal">features</tt> list:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">[</span><span class="literal number integer">20</span><span class="punctuation">]</span>
<span class="name builtin">type</span><span class="punctuation">:</span> <span class="name">gene</span>
<span class="name">location</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">4342</span><span class="punctuation">:</span><span class="literal number integer">4780</span><span class="punctuation">](</span><span class="operator">+</span><span class="punctuation">)</span>
<span class="name">qualifiers</span><span class="punctuation">:</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">db_xref</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'GeneID:2767712'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'pim'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">locus_tag</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'YP_pPCP05'</span><span class="punctuation">]</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">[</span><span class="literal number integer">21</span><span class="punctuation">]</span>
<span class="name builtin">type</span><span class="punctuation">:</span> <span class="name">CDS</span>
<span class="name">location</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">4342</span><span class="punctuation">:</span><span class="literal number integer">4780</span><span class="punctuation">](</span><span class="operator">+</span><span class="punctuation">)</span>
<span class="name">qualifiers</span><span class="punctuation">:</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">codon_start</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'1'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">db_xref</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'GI:45478716'</span><span class="punctuation">,</span> <span class="literal string">'GeneID:2767712'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'pim'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">locus_tag</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'YP_pPCP05'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">note</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'similar to many previously sequenced pesticin immunity ...'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">product</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'pesticin immunity protein'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">protein_id</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'NP_995571.1'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">transl_table</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'11'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">translation</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'MGGGMISKLFCLALIFLSSSGLAEKNTYTAKDILQNLELNTFGNSLSH...'</span><span class="punctuation">]</span>
</pre>
<p>Let’s slice this parent record from 4300 to 4800 (enough to include the
<tt class="docutils literal">pim</tt> gene/CDS), and see how many features we get:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">4300</span><span class="punctuation">:</span><span class="literal number integer">4800</span><span class="punctuation">]</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ATAAATAGATTATTCCAAATAATTTATTTATGTAAGAACAGGATGGGAGGGGGA...TTA'</span><span class="punctuation">,</span>
<span class="name">IUPACAmbiguousDNA</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'NC_005816.1'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'NC_005816'</span><span class="punctuation">,</span>
<span class="name">description</span><span class="operator">=</span><span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span><span class="punctuation">,</span>
<span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[])</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">sub_record</span><span class="punctuation">)</span>
<span class="literal number integer">500</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">sub_record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">)</span>
<span class="literal number integer">2</span>
</pre>
<p>Our sub-record just has two features, the gene and CDS entries for
<tt class="docutils literal">YP_pPCP05</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="name builtin">type</span><span class="punctuation">:</span> <span class="name">gene</span>
<span class="name">location</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">42</span><span class="punctuation">:</span><span class="literal number integer">480</span><span class="punctuation">](</span><span class="operator">+</span><span class="punctuation">)</span>
<span class="name">qualifiers</span><span class="punctuation">:</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">db_xref</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'GeneID:2767712'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'pim'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">locus_tag</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'YP_pPCP05'</span><span class="punctuation">]</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">[</span><span class="literal number integer">20</span><span class="punctuation">]</span>
<span class="name builtin">type</span><span class="punctuation">:</span> <span class="name">CDS</span>
<span class="name">location</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">42</span><span class="punctuation">:</span><span class="literal number integer">480</span><span class="punctuation">](</span><span class="operator">+</span><span class="punctuation">)</span>
<span class="name">qualifiers</span><span class="punctuation">:</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">codon_start</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'1'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">db_xref</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'GI:45478716'</span><span class="punctuation">,</span> <span class="literal string">'GeneID:2767712'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'pim'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">locus_tag</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'YP_pPCP05'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">note</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'similar to many previously sequenced pesticin immunity ...'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">product</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'pesticin immunity protein'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">protein_id</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'NP_995571.1'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">transl_table</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'11'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">translation</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'MGGGMISKLFCLALIFLSSSGLAEKNTYTAKDILQNLELNTFGNSLSH...'</span><span class="punctuation">]</span>
</pre>
<p>Notice that their locations have been adjusted to reflect the new parent
sequence!</p>
<p>While Biopython has done something sensible and hopefully intuitive with
the features (and any per-letter annotation), for the other annotation
it is impossible to know if this still applies to the sub-sequence or
not. To avoid guessing, the <tt class="docutils literal">annotations</tt> and <tt class="docutils literal">dbxrefs</tt> are omitted
from the sub-record, and it is up to you to transfer any relevant
information as appropriate.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">annotations</span>
<span class="punctuation">{}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="punctuation">[]</span>
</pre>
<p>The same point could be made about the record <tt class="docutils literal">id</tt>, <tt class="docutils literal">name</tt> and
<tt class="docutils literal">description</tt>, but for practicality these are preserved:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">id</span>
<span class="literal string">'NC_005816.1'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'NC_005816'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">description</span>
<span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span>
</pre>
<p>This illustrates the problem nicely though, our new sub-record is <em>not</em>
the complete sequence of the plasmid, so the description is wrong! Let’s
fix this and then view the sub-record as a reduced GenBank file using
the <tt class="docutils literal">format</tt> method described above in
Section <a class="reference external" href="#sec:SeqRecord-format">4.5</a>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, partial.&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>See Sections <a class="reference external" href="#sec:FASTQ-slicing-off-primer">18.1.7</a>
and <a class="reference external" href="#sec:FASTQ-slicing-off-adaptor">18.1.8</a> for some FASTQ
examples where the per-letter annotations (the read quality scores) are
also sliced.</p>
</div>
<div class="section" id="adding-seqrecord-objects">
<h2>4.7  Adding SeqRecord objects</h2>
<p>You can add <tt class="docutils literal">SeqRecord</tt> objects together, giving a new <tt class="docutils literal">SeqRecord</tt>.
What is important here is that any common per-letter annotations are
also added, all the features are preserved (with their locations
adjusted), and any other common annotation is also kept (like the id,
name and description).</p>
<p>For an example with per-letter annotation, we’ll use the first record in
a FASTQ file. Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> will explain the
<tt class="docutils literal">SeqIO</tt> functions:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;example.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="literal number integer">25</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">CCCTTCTTGTCTTCAGCGTTTCTCC</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">18</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">22</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span>
<span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">23</span><span class="punctuation">,</span> <span class="literal number integer">23</span><span class="punctuation">]</span>
</pre>
<p>Let’s suppose this was Roche 454 data, and that from other information
you think the <tt class="docutils literal">TTT</tt> should be only <tt class="docutils literal">TT</tt>. We can make a new edited
record by first slicing the <tt class="docutils literal">SeqRecord</tt> before and after the “extra”
third <tt class="docutils literal">T</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">left</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[:</span><span class="literal number integer">20</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">left</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">CCCTTCTTGTCTTCAGCGTT</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">left</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">18</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">22</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">right</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">21</span><span class="punctuation">:]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">right</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">CTCC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">right</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">23</span><span class="punctuation">,</span> <span class="literal number integer">23</span><span class="punctuation">]</span>
</pre>
<p>Now add the two parts together:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">edited</span> <span class="operator">=</span> <span class="name">left</span> <span class="operator">+</span> <span class="name">right</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">edited</span><span class="punctuation">)</span>
<span class="literal number integer">24</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">edited</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">CCCTTCTTGTCTTCAGCGTTCTCC</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">edited</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">18</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">22</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span>
<span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">23</span><span class="punctuation">,</span> <span class="literal number integer">23</span><span class="punctuation">]</span>
</pre>
<p>Easy and intuitive? We hope so! You can make this shorter with just:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">edited</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[:</span><span class="literal number integer">20</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">21</span><span class="punctuation">:]</span>
</pre>
<p>Now, for an example with features, we’ll use a GenBank file. Suppose you
have a circular genome:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'</span><span class="punctuation">,</span>
<span class="name">IUPACAmbiguousDNA</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'NC_005816.1'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'NC_005816'</span><span class="punctuation">,</span>
<span class="name">description</span><span class="operator">=</span><span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span><span class="punctuation">,</span>
<span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">'Project:10638'</span><span class="punctuation">])</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="literal number integer">9609</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">)</span>
<span class="literal number integer">41</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="punctuation">[</span><span class="literal string">'Project:58037'</span><span class="punctuation">]</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">annotations</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'comment'</span><span class="punctuation">,</span> <span class="literal string">'sequence_version'</span><span class="punctuation">,</span> <span class="literal string">'source'</span><span class="punctuation">,</span> <span class="literal string">'taxonomy'</span><span class="punctuation">,</span> <span class="literal string">'keywords'</span><span class="punctuation">,</span> <span class="literal string">'references'</span><span class="punctuation">,</span>
<span class="literal string">'accessions'</span><span class="punctuation">,</span> <span class="literal string">'data_file_division'</span><span class="punctuation">,</span> <span class="literal string">'date'</span><span class="punctuation">,</span> <span class="literal string">'organism'</span><span class="punctuation">,</span> <span class="literal string">'gi'</span><span class="punctuation">]</span>
</pre>
<p>You can shift the origin like this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">2000</span><span class="punctuation">:]</span> <span class="operator">+</span> <span class="name">record</span><span class="punctuation">[:</span><span class="literal number integer">2000</span><span class="punctuation">]</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATACGCAGTCATATTTTTTACACAATTCTCTAATCCCGACAAGGTCGTAGGTC...GGA'</span><span class="punctuation">,</span>
<span class="name">IUPACAmbiguousDNA</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'NC_005816.1'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'NC_005816'</span><span class="punctuation">,</span>
<span class="name">description</span><span class="operator">=</span><span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span><span class="punctuation">,</span>
<span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[])</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">shifted</span><span class="punctuation">)</span>
<span class="literal number integer">9609</span>
</pre>
<p>Note that this isn’t perfect in that some annotation like the database
cross references and one of the features (the source feature) have been
lost:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">shifted</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">)</span>
<span class="literal number integer">40</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="punctuation">[]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span><span class="operator">.</span><span class="name">annotations</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[]</span>
</pre>
<p>This is because the <tt class="docutils literal">SeqRecord</tt> slicing step is cautious in what
annotation it preserves (erroneously propagating annotation can cause
major problems). If you want to keep the database cross references or
the annotations dictionary, this must be done explicitly:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span><span class="operator">.</span><span class="name">dbxrefs</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span><span class="punctuation">[:]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span><span class="operator">.</span><span class="name">annotations</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">annotations</span><span class="operator">.</span><span class="name">copy</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="punctuation">[</span><span class="literal string">'Project:10638'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span><span class="operator">.</span><span class="name">annotations</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'comment'</span><span class="punctuation">,</span> <span class="literal string">'sequence_version'</span><span class="punctuation">,</span> <span class="literal string">'source'</span><span class="punctuation">,</span> <span class="literal string">'taxonomy'</span><span class="punctuation">,</span> <span class="literal string">'keywords'</span><span class="punctuation">,</span> <span class="literal string">'references'</span><span class="punctuation">,</span>
<span class="literal string">'accessions'</span><span class="punctuation">,</span> <span class="literal string">'data_file_division'</span><span class="punctuation">,</span> <span class="literal string">'date'</span><span class="punctuation">,</span> <span class="literal string">'organism'</span><span class="punctuation">,</span> <span class="literal string">'gi'</span><span class="punctuation">]</span>
</pre>
<p>Also note that in an example like this, you should probably change the
record identifiers since the NCBI references refer to the <em>original</em>
unmodified sequence.</p>
</div>
<div class="section" id="reverse-complementing-seqrecord-objects">
<h2>4.8  Reverse-complementing SeqRecord objects</h2>
<p>One of the new features in Biopython 1.57 was the <tt class="docutils literal">SeqRecord</tt> object’s
<tt class="docutils literal">reverse_complement</tt> method. This tries to balance easy of use with
worries about what to do with the annotation in the reverse complemented
record.</p>
<p>For the sequence, this uses the Seq object’s reverse complement method.
Any features are transferred with the location and strand recalculated.
Likewise any per-letter-annotation is also copied but reversed (which
makes sense for typical examples like quality scores). However, transfer
of most annotation is problematical.</p>
<p>For instance, if the record ID was an accession, that accession should
not really apply to the reverse complemented sequence, and transferring
the identifier by default could easily cause subtle data corruption in
downstream analysis. Therefore by default, the <tt class="docutils literal">SeqRecord</tt>’s id,
name, description, annotations and database cross references are all
<em>not</em> transferred by default.</p>
<p>The <tt class="docutils literal">SeqRecord</tt> object’s <tt class="docutils literal">reverse_complement</tt> method takes a number
of optional arguments corresponding to properties of the record. Setting
these arguments to <tt class="docutils literal">True</tt> means copy the old values, while <tt class="docutils literal">False</tt>
means drop the old values and use the default value. You can
alternatively provide the new desired value instead.</p>
<p>Consider this example record:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">)</span>
<span class="name">NC_005816</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="literal number integer">9609</span> <span class="literal number integer">41</span> <span class="literal number integer">1</span> <span class="literal number integer">11</span>
</pre>
<p>Here we take the reverse complement and specify a new identifier – but
notice how most of the annotation is dropped (but not the features):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">rc</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;TESTING&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">rc</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rc</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rc</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rc</span><span class="operator">.</span><span class="name">dbxrefs</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rc</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">)</span>
<span class="name">TESTING</span> <span class="literal number integer">9609</span> <span class="literal number integer">41</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span>
</pre>
</div>
</div>
<div class="section" id="chapter-5-sequence-input-output">
<h1>Chapter 5  Sequence Input/Output</h1>
<p>In this chapter we’ll discuss in more detail the <tt class="docutils literal">Bio.SeqIO</tt> module,
which was briefly introduced in Chapter <a class="reference external" href="#chapter:quick-start">2</a>
and also used in Chapter <a class="reference external" href="#chapter:SeqRecord">4</a>. This aims to
provide a simple interface for working with assorted sequence file
formats in a uniform way. See also the <tt class="docutils literal">Bio.SeqIO</tt> wiki page
(<a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a>),
and the built in documentation (also
<a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SeqIO-module.html">online</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>The “catch” is that you have to work with <tt class="docutils literal">SeqRecord</tt> objects (see
Chapter <a class="reference external" href="#chapter:SeqRecord">4</a>), which contain a <tt class="docutils literal">Seq</tt> object
(see Chapter <a class="reference external" href="#chapter:Bio.Seq">3</a>) plus annotation like an
identifier and description.</p>
<div class="section" id="parsing-or-reading-sequences">
<h2>5.1  Parsing or Reading Sequences</h2>
<p>The workhorse function <tt class="docutils literal">Bio.SeqIO.parse()</tt> is used to read in sequence
data as SeqRecord objects. This function expects two arguments:</p>
<ol class="arabic simple">
<li>The first argument is a <em>handle</em> to read the data from, or a
filename. A handle is typically a file opened for reading, but could
be the output from a command line program, or data downloaded from
the internet (see Section <a class="reference external" href="#sec:SeqIO_Online">5.3</a>). See
Section <a class="reference external" href="#sec:appendix-handles">22.1</a> for more about handles.</li>
<li>The second argument is a lower case string specifying sequence format
– we don’t try and guess the file format for you! See
<a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a>
for a full listing of supported formats.</li>
</ol>
<p>There is an optional argument <tt class="docutils literal">alphabet</tt> to specify the alphabet to be
used. This is useful for file formats like FASTA where otherwise
<tt class="docutils literal">Bio.SeqIO</tt> will default to a generic alphabet.</p>
<p>The <tt class="docutils literal">Bio.SeqIO.parse()</tt> function returns an <em>iterator</em> which gives
<tt class="docutils literal">SeqRecord</tt> objects. Iterators are typically used in a for loop as
shown below.</p>
<p>Sometimes you’ll find yourself dealing with files which contain only a
single record. For this situation use the function <tt class="docutils literal">Bio.SeqIO.read()</tt>
which takes the same arguments. Provided there is one and only one
record in the file, this is returned as a <tt class="docutils literal">SeqRecord</tt> object.
Otherwise an exception is raised.</p>
<div class="section" id="reading-sequence-files">
<h3>5.1.1  Reading Sequence Files</h3>
<p>In general <tt class="docutils literal">Bio.SeqIO.parse()</tt> is used to read in sequence files as
<tt class="docutils literal">SeqRecord</tt> objects, and is typically used with a for loop like this:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span>
    <span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
    <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="punctuation">)</span>
</pre>
<p>The above example is repeated from the introduction in
Section <a class="reference external" href="#sec:sequence-parsing">2.4</a>, and will load the orchid DNA
sequences in the FASTA format file
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta">ls_orchid.fasta</a>.
If instead you wanted to load a GenBank format file like
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk">ls_orchid.gbk</a>
then all you need to do is change the filename and the format string:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span>
    <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span>
    <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="punctuation">)</span>
</pre>
<p>Similarly, if you wanted to read in a file in another file format, then
assuming <tt class="docutils literal">Bio.SeqIO.parse()</tt> supports it you would just need to change
the format string as appropriate, for example “swiss” for SwissProt
files or “embl” for EMBL text files. There is a full listing on the wiki
page
(<a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a>)
and in the built in documentation (also
<a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SeqIO-module.html">online</a>).</p>
<p>Another very common way to use a Python iterator is within a list
comprehension (or a generator expression). For example, if all you
wanted to extract from the file was a list of the record identifiers we
can easily do this with the following list comprehension:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">identifiers</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span> <span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">identifiers</span>
<span class="punctuation">[</span><span class="literal string">'Z78533.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78532.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78531.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78530.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78529.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78527.1'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Z78439.1'</span><span class="punctuation">]</span>
</pre>
<p>There are more examples using <tt class="docutils literal">SeqIO.parse()</tt> in a list comprehension
like this in Section <a class="reference external" href="#seq:sequence-parsing-plus-pylab">18.2</a> (e.g.
for plotting sequence lengths or GC%).</p>
</div>
<div class="section" id="iterating-over-the-records-in-a-sequence-file">
<h3>5.1.2  Iterating over the records in a sequence file</h3>
<p>In the above examples, we have usually used a for loop to iterate over
all the records one by one. You can use the for loop with all sorts of
Python objects (including lists, tuples and strings) which support the
iteration interface.</p>
<p>The object returned by <tt class="docutils literal">Bio.SeqIO</tt> is actually an iterator which
returns <tt class="docutils literal">SeqRecord</tt> objects. You get to see each record in turn, but
once and only once. The plus point is that an iterator can save you
memory when dealing with large files.</p>
<p>Instead of using a for loop, can also use the <tt class="docutils literal">.next()</tt> method of an
iterator to step through the entries, like this:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">record_iterator</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>

<span class="name">first_record</span> <span class="operator">=</span> <span class="name">record_iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">id</span>
<span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">description</span>

<span class="name">second_record</span> <span class="operator">=</span> <span class="name">record_iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="name">second_record</span><span class="operator">.</span><span class="name">id</span>
<span class="keyword">print</span> <span class="name">second_record</span><span class="operator">.</span><span class="name">description</span>
</pre>
<p>Note that if you try and use <tt class="docutils literal">.next()</tt> and there are no more results,
you’ll get the special <tt class="docutils literal">StopIteration</tt> exception.</p>
<p>One special case to consider is when your sequence files have multiple
records, but you only want the first one. In this situation the
following code is very concise:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">first_record</span>  <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
</pre>
<p>A word of warning here – using the <tt class="docutils literal">.next()</tt> method like this will
silently ignore any additional records in the file. If your files have
<em>one and only one</em> record, like some of the online examples later in
this chapter, or a GenBank file for a single chromosome, then use the
new <tt class="docutils literal">Bio.SeqIO.read()</tt> function instead. This will check there are no
extra unexpected records present.</p>
</div>
<div class="section" id="getting-a-list-of-the-records-in-a-sequence-file">
<h3>5.1.3  Getting a list of the records in a sequence file</h3>
<p>In the previous section we talked about the fact that
<tt class="docutils literal">Bio.SeqIO.parse()</tt> gives you a <tt class="docutils literal">SeqRecord</tt> iterator, and that you
get the records one by one. Very often you need to be able to access the
records in any order. The Python <tt class="docutils literal">list</tt> data type is perfect for this,
and we can turn the record iterator into a list of <tt class="docutils literal">SeqRecord</tt> objects
using the built-in Python function <tt class="docutils literal">list()</tt> like so:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">))</span>

<span class="keyword">print</span> <span class="literal string">&quot;Found </span><span class="literal string interpol">%i</span><span class="literal string"> records&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">)</span>

<span class="keyword">print</span> <span class="literal string">&quot;The last record&quot;</span>
<span class="name">last_record</span> <span class="operator">=</span> <span class="name">records</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span> <span class="comment">#using Python's list tricks</span>
<span class="keyword">print</span> <span class="name">last_record</span><span class="operator">.</span><span class="name">id</span>
<span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">last_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">last_record</span><span class="punctuation">)</span>

<span class="keyword">print</span> <span class="literal string">&quot;The first record&quot;</span>
<span class="name">first_record</span> <span class="operator">=</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span> <span class="comment">#remember, Python counts from zero</span>
<span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">id</span>
<span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">first_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">first_record</span><span class="punctuation">)</span>
</pre>
<p>Giving:</p>
<pre class="code python literal-block">
<span class="name">Found</span> <span class="literal number integer">94</span> <span class="name">records</span>
<span class="name">The</span> <span class="name">last</span> <span class="name">record</span>
<span class="name">Z78439</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
<span class="literal number integer">592</span>
<span class="name">The</span> <span class="name">first</span> <span class="name">record</span>
<span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
<span class="literal number integer">740</span>
</pre>
<p>You can of course still use a for loop with a list of <tt class="docutils literal">SeqRecord</tt>
objects. Using a list is much more flexible than an iterator (for
example, you can determine the number of records from the length of the
list), but does need more memory because it will hold all the records in
memory at once.</p>
</div>
<div class="section" id="extracting-data">
<h3>5.1.4  Extracting data</h3>
<p>The <tt class="docutils literal">SeqRecord</tt> object and its annotation structures are described
more fully in Chapter <a class="reference external" href="#chapter:SeqRecord">4</a>. As an example of how
annotations are stored, we’ll look at the output from parsing the first
record in the GenBank file
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk">ls_orchid.gbk</a>.</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">record_iterator</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="name">first_record</span> <span class="operator">=</span> <span class="name">record_iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="name">first_record</span>
</pre>
<p>That should give something like this:</p>
<pre class="code python literal-block">
<span class="name">ID</span><span class="punctuation">:</span> <span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">Name</span><span class="punctuation">:</span> <span class="name">Z78533</span>
<span class="name">Description</span><span class="punctuation">:</span> <span class="name">C</span><span class="operator">.</span><span class="name">irapeanum</span> <span class="literal number float">5.8</span><span class="name">S</span> <span class="name">rRNA</span> <span class="name">gene</span> <span class="operator word">and</span> <span class="name">ITS1</span> <span class="operator word">and</span> <span class="name">ITS2</span> <span class="name">DNA</span><span class="operator">.</span>
<span class="name">Number</span> <span class="name">of</span> <span class="name">features</span><span class="punctuation">:</span> <span class="literal number integer">5</span>
<span class="operator">/</span><span class="name">sequence_version</span><span class="operator">=</span><span class="literal number integer">1</span>
<span class="operator">/</span><span class="name">source</span><span class="operator">=</span><span class="name">Cypripedium</span> <span class="name">irapeanum</span>
<span class="operator">/</span><span class="name">taxonomy</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">'Eukaryota'</span><span class="punctuation">,</span> <span class="literal string">'Viridiplantae'</span><span class="punctuation">,</span> <span class="literal string">'Streptophyta'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Cypripedium'</span><span class="punctuation">]</span>
<span class="operator">/</span><span class="name">keywords</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">'5.8S ribosomal RNA'</span><span class="punctuation">,</span> <span class="literal string">'5.8S rRNA gene'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'ITS1'</span><span class="punctuation">,</span> <span class="literal string">'ITS2'</span><span class="punctuation">]</span>
<span class="operator">/</span><span class="name">references</span><span class="operator">=</span><span class="punctuation">[</span><span class="operator">...</span><span class="punctuation">]</span>
<span class="operator">/</span><span class="name">accessions</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">'Z78533'</span><span class="punctuation">]</span>
<span class="operator">/</span><span class="name">data_file_division</span><span class="operator">=</span><span class="name">PLN</span>
<span class="operator">/</span><span class="name">date</span><span class="operator">=</span><span class="literal number integer">30</span><span class="operator">-</span><span class="name">NOV</span><span class="operator">-</span><span class="literal number integer">2006</span>
<span class="operator">/</span><span class="name">organism</span><span class="operator">=</span><span class="name">Cypripedium</span> <span class="name">irapeanum</span>
<span class="operator">/</span><span class="name">gi</span><span class="operator">=</span><span class="literal number integer">2765658</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>This gives a human readable summary of most of the annotation data for
the <tt class="docutils literal">SeqRecord</tt>. For this example we’re going to use the
<tt class="docutils literal">.annotations</tt> attribute which is just a Python dictionary. The
contents of this annotations dictionary were shown when we printed the
record above. You can also print them out directly:</p>
<pre class="code python literal-block">
<span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">annotations</span>
</pre>
<p>Like any Python dictionary, you can easily get a list of the keys:</p>
<pre class="code python literal-block">
<span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">annotations</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
</pre>
<p>or values:</p>
<pre class="code python literal-block">
<span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">annotations</span><span class="operator">.</span><span class="name">values</span><span class="punctuation">()</span>
</pre>
<p>In general, the annotation values are strings, or lists of strings. One
special case is any references in the file get stored as reference
objects.</p>
<p>Suppose you wanted to extract a list of the species from the
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk">ls_orchid.gbk</a>
GenBank file. The information we want, <em>Cypripedium irapeanum</em>, is held
in the annotations dictionary under ‘source’ and ‘organism’, which we
can access like this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;source&quot;</span><span class="punctuation">]</span>
<span class="name">Cypripedium</span> <span class="name">irapeanum</span>
</pre>
<p>or:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;organism&quot;</span><span class="punctuation">]</span>
<span class="name">Cypripedium</span> <span class="name">irapeanum</span>
</pre>
<p>In general, ‘organism’ is used for the scientific name (in Latin, e.g.
<em>Arabidopsis thaliana</em>), while ‘source’ will often be the common name
(e.g. thale cress). In this example, as is often the case, the two
fields are identical.</p>
<p>Now let’s go through all the records, building up a list of the species
each orchid sequence is from:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">all_species</span> <span class="operator">=</span> <span class="punctuation">[]</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">):</span>
    <span class="name">all_species</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;organism&quot;</span><span class="punctuation">])</span>
<span class="keyword">print</span> <span class="name">all_species</span>
</pre>
<p>Another way of writing this code is to use a list comprehension:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">all_species</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;organism&quot;</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> \
               <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)]</span>
<span class="keyword">print</span> <span class="name">all_species</span>
</pre>
<p>In either case, the result is:</p>
<pre class="code python literal-block">
<span class="punctuation">[</span><span class="literal string">'Cypripedium irapeanum'</span><span class="punctuation">,</span> <span class="literal string">'Cypripedium californicum'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Paphiopedilum barbatum'</span><span class="punctuation">]</span>
</pre>
<p>Great. That was pretty easy because GenBank files are annotated in a
standardised way.</p>
<p>Now, let’s suppose you wanted to extract a list of the species from a
FASTA file, rather than the GenBank file. The bad news is you will have
to write some code to extract the data you want from the record’s
description line - if the information is in the file in the first place!
Our example FASTA format file
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta">ls_orchid.fasta</a>
starts like this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">2765658</span><span class="operator">|</span><span class="name">emb</span><span class="operator">|</span><span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">CIZ78533</span> <span class="name">C</span><span class="operator">.</span><span class="name">irapeanum</span> <span class="literal number float">5.8</span><span class="name">S</span> <span class="name">rRNA</span> <span class="name">gene</span> <span class="operator word">and</span> <span class="name">ITS1</span> <span class="operator word">and</span> <span class="name">ITS2</span> <span class="name">DNA</span>
<span class="name">CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGGAATAAACGATCGAGTG</span>
<span class="name">AATCCGGAGGACCGGTGTACTCAGCTCACCGGGGGCATTGCTCCCGTGGTGACCCTGATTTGTTGTTGGG</span>
<span class="operator">...</span>
</pre>
<p>You can check by hand, but for every record the species name is in the
description line as the second word. This means if we break up each
record’s <tt class="docutils literal">.description</tt> at the spaces, then the species is there as
field number one (field zero is the record identifier). That means we
can do this:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">all_species</span> <span class="operator">=</span> <span class="punctuation">[]</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">):</span>
    <span class="name">all_species</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">description</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">()[</span><span class="literal number integer">1</span><span class="punctuation">])</span>
<span class="keyword">print</span> <span class="name">all_species</span>
</pre>
<p>This gives:</p>
<pre class="code python literal-block">
<span class="punctuation">[</span><span class="literal string">'C.irapeanum'</span><span class="punctuation">,</span> <span class="literal string">'C.californicum'</span><span class="punctuation">,</span> <span class="literal string">'C.fasciculatum'</span><span class="punctuation">,</span> <span class="literal string">'C.margaritaceum'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'P.barbatum'</span><span class="punctuation">]</span>
</pre>
<p>The concise alternative using list comprehensions would be:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">all_species</span> <span class="operator">==</span> <span class="punctuation">[</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">description</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">()[</span><span class="literal number integer">1</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> \
                <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)]</span>
<span class="keyword">print</span> <span class="name">all_species</span>
</pre>
<p>In general, extracting information from the FASTA description line is
not very nice. If you can get your sequences in a well annotated file
format like GenBank or EMBL, then this sort of annotation information is
much easier to deal with.</p>
</div>
</div>
<div class="section" id="parsing-sequences-from-compressed-files">
<h2>5.2  Parsing sequences from compressed files</h2>
<p>In the previous section, we looked at parsing sequence data from a file.
Instead of using a filename, you can give <tt class="docutils literal">Bio.SeqIO</tt> a handle (see
Section <a class="reference external" href="#sec:appendix-handles">22.1</a>), and in this section we’ll
use handles to parse sequence from compressed files.</p>
<p>As you’ll have seen above, we can use <tt class="docutils literal">Bio.SeqIO.read()</tt> or
<tt class="docutils literal">Bio.SeqIO.parse()</tt> with a filename - for instance this quick example
calculates the total length of the sequences in a multiple record
GenBank file using a generator expression:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">sum</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">))</span>
<span class="literal number integer">67518</span>
</pre>
<p>Here we use a file handle instead, using the <tt class="docutils literal">with</tt> statement (Python
2.5 or later) to close the handle automatically:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">__future__</span> <span class="keyword namespace">import</span> <span class="name">with_statement</span> <span class="comment">#Needed on Python 2.5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">with</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">)</span> <span class="keyword">as</span> <span class="name">handle</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name builtin">sum</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">))</span>
<span class="literal number integer">67518</span>
</pre>
<p>Or, the old fashioned way where you manually close the handle:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">sum</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">))</span>
<span class="literal number integer">67518</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>Now, suppose we have a gzip compressed file instead? These are very
commonly used on Linux. We can use Python’s <tt class="docutils literal">gzip</tt> module to open the
compressed file for reading - which gives us a handle object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">gzip</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">gzip</span><span class="operator">.</span><span class="name">open</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk.gz&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;r&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">sum</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">))</span>
<span class="literal number integer">67518</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>Similarly if we had a bzip2 compressed file (sadly the function name
isn’t quite as consistent):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">bz2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">bz2</span><span class="operator">.</span><span class="name">BZ2File</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk.bz2&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;r&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">sum</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">))</span>
<span class="literal number integer">67518</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>If you are using Python 2.7 or later, the <tt class="docutils literal">with</tt>-version works for
gzip and bz2 as well. Unfortunately this is broken on older versions of
Python (<a class="reference external" href="http://bugs.python.org/issue3860">Issue 3860</a>) and you’d get
an <tt class="docutils literal">AttributeError</tt> about <tt class="docutils literal">__exit__</tt> being missing.</p>
<p>There is a gzip (GNU Zip) variant called BGZF (Blocked GNU Zip Format),
which can be treated like an ordinary gzip file for reading, but has
advantages for random access later which we’ll talk about later in
Section <a class="reference external" href="#sec:SeqIO-index-bgzf">5.4.4</a>.</p>
</div>
<div class="section" id="parsing-sequences-from-the-net">
<h2>5.3  Parsing sequences from the net</h2>
<p>In the previous sections, we looked at parsing sequence data from a file
(using a filename or handle), and from compressed files (using a
handle). Here we’ll use <tt class="docutils literal">Bio.SeqIO</tt> with another type of handle, a
network connection, to download and parse sequences from the internet.</p>
<p>Note that just because you <em>can</em> download sequence data and parse it
into a <tt class="docutils literal">SeqRecord</tt> object in one go doesn’t mean this is a good idea.
In general, you should probably download sequences <em>once</em> and save them
to a file for reuse.</p>
<div class="section" id="parsing-genbank-records-from-the-net">
<h3>5.3.1  Parsing GenBank records from the net</h3>
<p>Section <a class="reference external" href="#sec:efetch">9.6</a> talks about the Entrez EFetch interface
in more detail, but for now let’s just connect to the NCBI and get a few
<em>Opuntia</em> (prickly-pear) sequences from GenBank using their GI numbers.</p>
<p>First of all, let’s fetch just one record. If you don’t care about the
annotations and features downloading a FASTA file is a good choice as
these are compact. Now remember, when you expect the handle to contain
one and only one record, use the <tt class="docutils literal">Bio.SeqIO.read()</tt> function:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;6273291&quot;</span><span class="punctuation">)</span>
<span class="name">seq_record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> with </span><span class="literal string interpol">%i</span><span class="literal string"> features&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">))</span>
</pre>
<p>Expected output:</p>
<pre class="code python literal-block">
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273291</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191665</span> <span class="keyword">with</span> <span class="literal number integer">0</span> <span class="name">features</span>
</pre>
<p>The NCBI will also let you ask for the file in other formats, in
particular as a GenBank file. Until Easter 2009, the Entrez EFetch API
let you use “genbank” as the return type, however the NCBI now insist on
using the official return types of “gb” (or “gp” for proteins) as
described on <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/efetchseq_help.html">EFetch for Sequence and other Molecular Biology
Databases</a>.
As a result, in Biopython 1.50 onwards, we support “gb” as an alias for
“genbank” in <tt class="docutils literal">Bio.SeqIO</tt>.</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;6273291&quot;</span><span class="punctuation">)</span>
<span class="name">seq_record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">)</span> <span class="comment">#using &quot;gb&quot; as an alias for &quot;genbank&quot;</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> with </span><span class="literal string interpol">%i</span><span class="literal string"> features&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">))</span>
</pre>
<p>The expected output of this example is:</p>
<pre class="code python literal-block">
<span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
</pre>
<p>Notice this time we have three features.</p>
<p>Now let’s fetch several records. This time the handle contains multiple
records, so we must use the <tt class="docutils literal">Bio.SeqIO.parse()</tt> function:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">,</span> \
                       <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;6273291,6273290,6273289&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">description</span><span class="punctuation">[:</span><span class="literal number integer">50</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">&quot;...&quot;</span>
    <span class="keyword">print</span> <span class="literal string">&quot;Sequence length </span><span class="literal string interpol">%i</span><span class="literal string">,&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="punctuation">),</span>
    <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%i</span><span class="literal string"> features,&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">),</span>
    <span class="keyword">print</span> <span class="literal string">&quot;from: </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;source&quot;</span><span class="punctuation">]</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>That should give the following output:</p>
<pre class="code python literal-block">
<span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="name">Opuntia</span> <span class="name">marenae</span> <span class="name">rpl16</span> <span class="name">gene</span><span class="punctuation">;</span> <span class="name">chloroplast</span> <span class="name">gene</span> <span class="keyword">for</span> <span class="name">c</span><span class="operator">...</span>
<span class="name">Sequence</span> <span class="name">length</span> <span class="literal number integer">902</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">features</span><span class="punctuation">,</span> <span class="name">from</span><span class="punctuation">:</span> <span class="name">chloroplast</span> <span class="name">Opuntia</span> <span class="name">marenae</span>
<span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="name">Opuntia</span> <span class="name">clavata</span> <span class="name">rpl16</span> <span class="name">gene</span><span class="punctuation">;</span> <span class="name">chloroplast</span> <span class="name">gene</span> <span class="keyword">for</span> <span class="name">c</span><span class="operator">...</span>
<span class="name">Sequence</span> <span class="name">length</span> <span class="literal number integer">899</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">features</span><span class="punctuation">,</span> <span class="name">from</span><span class="punctuation">:</span> <span class="name">chloroplast</span> <span class="name">Grusonia</span> <span class="name">clavata</span>
<span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="name">Opuntia</span> <span class="name">bradtiana</span> <span class="name">rpl16</span> <span class="name">gene</span><span class="punctuation">;</span> <span class="name">chloroplast</span> <span class="name">gene</span> <span class="keyword">for</span><span class="operator">...</span>
<span class="name">Sequence</span> <span class="name">length</span> <span class="literal number integer">899</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">features</span><span class="punctuation">,</span> <span class="name">from</span><span class="punctuation">:</span> <span class="name">chloroplast</span> <span class="name">Opuntia</span> <span class="name">bradtianaa</span>
</pre>
<p>See Chapter <a class="reference external" href="#chapter:entrez">9</a> for more about the <tt class="docutils literal">Bio.Entrez</tt>
module, and make sure to read about the NCBI guidelines for using Entrez
(Section <a class="reference external" href="#sec:entrez-guidelines">9.1</a>).</p>
</div>
<div class="section" id="parsing-swissprot-sequences-from-the-net">
<h3>5.3.2  Parsing SwissProt sequences from the net</h3>
<p>Now let’s use a handle to download a SwissProt file from ExPASy,
something covered in more depth in
Chapter <a class="reference external" href="#chapter:swiss_prot">10</a>. As mentioned above, when you
expect the handle to contain one and only one record, use the
<tt class="docutils literal">Bio.SeqIO.read()</tt> function:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_sprot_raw</span><span class="punctuation">(</span><span class="literal string">&quot;O23729&quot;</span><span class="punctuation">)</span>
<span class="name">seq_record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;swiss&quot;</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span>
<span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">name</span>
<span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">description</span>
<span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Length </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;keywords&quot;</span><span class="punctuation">]</span>
</pre>
<p>Assuming your network connection is OK, you should get back:</p>
<pre class="code python literal-block">
<span class="name">O23729</span>
<span class="name">CHS3_BROFI</span>
<span class="name">RecName</span><span class="punctuation">:</span> <span class="name">Full</span><span class="operator">=</span><span class="name">Chalcone</span> <span class="name">synthase</span> <span class="literal number integer">3</span><span class="punctuation">;</span> <span class="name">EC</span><span class="operator">=</span><span class="literal number float">2.3</span><span class="operator">.</span><span class="literal number float">1.74</span><span class="punctuation">;</span> <span class="name">AltName</span><span class="punctuation">:</span> <span class="name">Full</span><span class="operator">=</span><span class="name">Naringenin</span><span class="operator">-</span><span class="name">chalcone</span> <span class="name">synthase</span> <span class="literal number integer">3</span><span class="punctuation">;</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAPAMEEIRQAQRAEGPAAVLAIGTSTPPNALYQADYPDYYFRITKSEHLTELK...GAE'</span><span class="punctuation">,</span> <span class="name">ProteinAlphabet</span><span class="punctuation">())</span>
<span class="name">Length</span> <span class="literal number integer">394</span>
<span class="punctuation">[</span><span class="literal string">'Acyltransferase'</span><span class="punctuation">,</span> <span class="literal string">'Flavonoid biosynthesis'</span><span class="punctuation">,</span> <span class="literal string">'Transferase'</span><span class="punctuation">]</span>
</pre>
</div>
</div>
<div class="section" id="sequence-files-as-dictionaries">
<h2>5.4  Sequence files as Dictionaries</h2>
<p>We’re now going to introduce three related functions in the
<tt class="docutils literal">Bio.SeqIO</tt> module which allow dictionary like random access to a
multi-sequence file. There is a trade off here between flexibility and
memory usage. In summary:</p>
<ul class="simple">
<li><tt class="docutils literal">Bio.SeqIO.to_dict()</tt> is the most flexible but also the most memory
demanding option (see Section <a class="reference external" href="#SeqIO:to_dict">5.4.1</a>). This is
basically a helper function to build a normal Python <tt class="docutils literal">dictionary</tt>
with each entry held as a <tt class="docutils literal">SeqRecord</tt> object in memory, allowing
you to modify the records.</li>
<li><tt class="docutils literal">Bio.SeqIO.index()</tt> is a useful middle ground, acting like a read
only dictionary and parsing sequences into <tt class="docutils literal">SeqRecord</tt> objects on
demand (see Section <a class="reference external" href="#sec:SeqIO-index">5.4.2</a>).</li>
<li><tt class="docutils literal">Bio.SeqIO.index_db()</tt> also acts like a read only dictionary but
stores the identifiers and file offsets in a file on disk (as an
SQLite3 database), meaning it has very low memory requirements (see
Section <a class="reference external" href="#sec:SeqIO-index-db">5.4.3</a>), but will be a little bit
slower.</li>
</ul>
<p>See the discussion for an broad overview
(Section <a class="reference external" href="#sec:SeqIO-indexing-discussion">5.4.5</a>).</p>
<div class="section" id="sequence-files-as-dictionaries-in-memory">
<h3>5.4.1  Sequence files as Dictionaries – In memory</h3>
<p>The next thing that we’ll do with our ubiquitous orchid files is to show
how to index them and access them like a database using the Python
<tt class="docutils literal">dictionary</tt> data type (like a hash in Perl). This is very useful for
moderately large files where you only need to access certain elements of
the file, and makes for a nice quick ’n dirty database. For dealing with
larger files where memory becomes a problem, see
Section <a class="reference external" href="#sec:SeqIO-index">5.4.2</a> below.</p>
<p>You can use the function <tt class="docutils literal">Bio.SeqIO.to_dict()</tt> to make a SeqRecord
dictionary (in memory). By default this will use each record’s
identifier (i.e. the <tt class="docutils literal">.id</tt> attribute) as the key. Let’s try this using
our GenBank file:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">to_dict</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">))</span>
</pre>
<p>There is just one required argument for <tt class="docutils literal">Bio.SeqIO.to_dict()</tt>, a list
or generator giving <tt class="docutils literal">SeqRecord</tt> objects. Here we have just used the
output from the <tt class="docutils literal">SeqIO.parse</tt> function. As the name suggests, this
returns a Python dictionary.</p>
<p>Since this variable <tt class="docutils literal">orchid_dict</tt> is an ordinary Python dictionary, we
can look at all of the keys we have available:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">orchid_dict</span><span class="punctuation">)</span>
<span class="literal number integer">94</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'Z78484.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78464.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78455.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78442.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78532.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78453.1'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Z78471.1'</span><span class="punctuation">]</span>
</pre>
<p>If you really want to, you can even look at all the records at once:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">values</span><span class="punctuation">()</span> <span class="comment">#lots of output!</span>
<span class="operator">...</span>
</pre>
<p>We can access a single <tt class="docutils literal">SeqRecord</tt> object via the keys and manipulate
the object as normal:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq_record</span> <span class="operator">=</span> <span class="name">orchid_dict</span><span class="punctuation">[</span><span class="literal string">&quot;Z78475.1&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">description</span>
<span class="name">P</span><span class="operator">.</span><span class="name">supardii</span> <span class="literal number float">5.8</span><span class="name">S</span> <span class="name">rRNA</span> <span class="name">gene</span> <span class="operator word">and</span> <span class="name">ITS1</span> <span class="operator word">and</span> <span class="name">ITS2</span> <span class="name">DNA</span><span class="operator">.</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGT'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>So, it is very easy to create an in memory “database” of our GenBank
records. Next we’ll try this for the FASTA file instead.</p>
<p>Note that those of you with prior Python experience should all be able
to construct a dictionary like this “by hand”. However, typical
dictionary construction methods will not deal with the case of repeated
keys very nicely. Using the <tt class="docutils literal">Bio.SeqIO.to_dict()</tt> will explicitly
check for duplicate keys, and raise an exception if any are found.</p>
<div class="section" id="specifying-the-dictionary-keys">
<h4>5.4.1.1  Specifying the dictionary keys</h4>
<p>Using the same code as above, but for the FASTA file instead:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">to_dict</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="keyword">print</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
</pre>
<p>This time the keys are:</p>
<pre class="code python literal-block">
<span class="punctuation">[</span><span class="literal string">'gi|2765596|emb|Z78471.1|PDZ78471'</span><span class="punctuation">,</span> <span class="literal string">'gi|2765646|emb|Z78521.1|CCZ78521'</span><span class="punctuation">,</span> <span class="operator">...</span>
 <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'gi|2765613|emb|Z78488.1|PTZ78488'</span><span class="punctuation">,</span> <span class="literal string">'gi|2765583|emb|Z78458.1|PHZ78458'</span><span class="punctuation">]</span>
</pre>
<p>You should recognise these strings from when we parsed the FASTA file
earlier in Section <a class="reference external" href="#sec:fasta-parsing">2.4.1</a>. Suppose you would
rather have something else as the keys - like the accession numbers.
This brings us nicely to <tt class="docutils literal">SeqIO.to_dict()</tt>’s optional argument
<tt class="docutils literal">key_function</tt>, which lets you define what to use as the dictionary
key for your records.</p>
<p>First you must write your own function to return the key you want (as a
string) when given a <tt class="docutils literal">SeqRecord</tt> object. In general, the details of
function will depend on the sort of input records you are dealing with.
But for our orchids, we can just split up the record’s identifier using
the “pipe” character (the vertical line) and return the fourth entry
(field three):</p>
<pre class="code python literal-block">
<span class="keyword">def</span> <span class="name function">get_accession</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">):</span>
    <span class="literal string doc">&quot;&quot;&quot;&quot;Given a SeqRecord, return the accession number as a string.

    e.g. &quot;gi|2765613|emb|Z78488.1|PTZ78488&quot; -&gt; &quot;Z78488.1&quot;
    &quot;&quot;&quot;</span>
    <span class="name">parts</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">(</span><span class="literal string">&quot;|&quot;</span><span class="punctuation">)</span>
    <span class="keyword">assert</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">parts</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="literal number integer">5</span> <span class="operator word">and</span> <span class="name">parts</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="literal string">&quot;gi&quot;</span> <span class="operator word">and</span> <span class="name">parts</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="literal string">&quot;emb&quot;</span>
    <span class="keyword">return</span> <span class="name">parts</span><span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">]</span>
</pre>
<p>Then we can give this function to the <tt class="docutils literal">SeqIO.to_dict()</tt> function to
use in building the dictionary:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">to_dict</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">),</span> <span class="name">key_function</span><span class="operator">=</span><span class="name">get_accession</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
</pre>
<p>Finally, as desired, the new dictionary keys:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'Z78484.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78464.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78455.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78442.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78532.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78453.1'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Z78471.1'</span><span class="punctuation">]</span>
</pre>
<p>Not too complicated, I hope!</p>
</div>
<div class="section" id="indexing-a-dictionary-using-the-seguid-checksum">
<h4>5.4.1.2  Indexing a dictionary using the SEGUID checksum</h4>
<p>To give another example of working with dictionaries of <tt class="docutils literal">SeqRecord</tt>
objects, we’ll use the SEGUID checksum function. This is a relatively
recent checksum, and collisions should be very rare (i.e. two different
sequences with the same checksum), an improvement on the CRC64 checksum.</p>
<p>Once again, working with the orchids GenBank file:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqUtils.CheckSum</span> <span class="keyword namespace">import</span> <span class="name">seguid</span>
<span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">seguid</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
</pre>
<p>This should give:</p>
<pre class="code python literal-block">
<span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="name">JUEoWn6DPhgZ9nAyowsgtoD9TTo</span>
<span class="name">Z78532</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="name">MN</span><span class="operator">/</span><span class="name">s0q9zDoCVEEc</span><span class="operator">+</span><span class="name">k</span><span class="operator">/</span><span class="name">IFwCNF2pY</span>
<span class="operator">...</span>
<span class="name">Z78439</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="name">H</span><span class="operator">+</span><span class="name">JfaShya</span><span class="operator">/</span><span class="literal number integer">4</span><span class="name">yyAj7IbMqgNkxdxQ</span>
</pre>
<p>Now, recall the <tt class="docutils literal">Bio.SeqIO.to_dict()</tt> function’s <tt class="docutils literal">key_function</tt>
argument expects a function which turns a <tt class="docutils literal">SeqRecord</tt> into a string.
We can’t use the <tt class="docutils literal">seguid()</tt> function directly because it expects to be
given a <tt class="docutils literal">Seq</tt> object (or a string). However, we can use Python’s
<tt class="docutils literal">lambda</tt> feature to create a “one off” function to give to
<tt class="docutils literal">Bio.SeqIO.to_dict()</tt> instead:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqUtils.CheckSum</span> <span class="keyword namespace">import</span> <span class="name">seguid</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seguid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">to_dict</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>                             <span class="keyword">lambda</span> <span class="name">rec</span> <span class="punctuation">:</span> <span class="name">seguid</span><span class="punctuation">(</span><span class="name">rec</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">seguid_dict</span><span class="punctuation">[</span><span class="literal string">&quot;MN/s0q9zDoCVEEc+k/IFwCNF2pY&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span>
<span class="name">Z78532</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">description</span>
<span class="name">C</span><span class="operator">.</span><span class="name">californicum</span> <span class="literal number float">5.8</span><span class="name">S</span> <span class="name">rRNA</span> <span class="name">gene</span> <span class="operator word">and</span> <span class="name">ITS1</span> <span class="operator word">and</span> <span class="name">ITS2</span> <span class="name">DNA</span><span class="operator">.</span>
</pre>
<p>That should have retrieved the record <tt class="docutils literal">Z78532.1</tt>, the second entry in
the file.</p>
</div>
</div>
<div class="section" id="sequence-files-as-dictionaries-indexed-files">
<h3>5.4.2  Sequence files as Dictionaries – Indexed files</h3>
<p>As the previous couple of examples tried to illustrate, using
<tt class="docutils literal">Bio.SeqIO.to_dict()</tt> is very flexible. However, because it holds
everything in memory, the size of file you can work with is limited by
your computer’s RAM. In general, this will only work on small to medium
files.</p>
<p>For larger files you should consider <tt class="docutils literal">Bio.SeqIO.index()</tt>, which works
a little differently. Although it still returns a dictionary like
object, this does <em>not</em> keep <em>everything</em> in memory. Instead, it just
records where each record is within the file – when you ask for a
particular record, it then parses it on demand.</p>
<p>As an example, let’s use the same GenBank file as before:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">orchid_dict</span><span class="punctuation">)</span>
<span class="literal number integer">94</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'Z78484.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78464.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78455.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78442.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78532.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78453.1'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Z78471.1'</span><span class="punctuation">]</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq_record</span> <span class="operator">=</span> <span class="name">orchid_dict</span><span class="punctuation">[</span><span class="literal string">&quot;Z78475.1&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">description</span>
<span class="name">P</span><span class="operator">.</span><span class="name">supardii</span> <span class="literal number float">5.8</span><span class="name">S</span> <span class="name">rRNA</span> <span class="name">gene</span> <span class="operator word">and</span> <span class="name">ITS1</span> <span class="operator word">and</span> <span class="name">ITS2</span> <span class="name">DNA</span><span class="operator">.</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGT'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>Note that <tt class="docutils literal">Bio.SeqIO.index()</tt> won’t take a handle, but only a
filename. There are good reasons for this, but it is a little technical.
The second argument is the file format (a lower case string as used in
the other <tt class="docutils literal">Bio.SeqIO</tt> functions). You can use many other simple file
formats, including FASTA and FASTQ files (see the example in
Section <a class="reference external" href="#sec:fastq-indexing">18.1.11</a>). However, alignment formats
like PHYLIP or Clustal are not supported. Finally as an optional
argument you can supply an alphabet, or a key function.</p>
<p>Here is the same example using the FASTA file - all we change is the
filename and the format name:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">orchid_dict</span><span class="punctuation">)</span>
<span class="literal number integer">94</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'gi|2765596|emb|Z78471.1|PDZ78471'</span><span class="punctuation">,</span> <span class="literal string">'gi|2765646|emb|Z78521.1|CCZ78521'</span><span class="punctuation">,</span> <span class="operator">...</span>
 <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'gi|2765613|emb|Z78488.1|PTZ78488'</span><span class="punctuation">,</span> <span class="literal string">'gi|2765583|emb|Z78458.1|PHZ78458'</span><span class="punctuation">]</span>
</pre>
<div class="section" id="id1">
<h4>5.4.2.1  Specifying the dictionary keys</h4>
<p>Suppose you want to use the same keys as before? Much like with the
<tt class="docutils literal">Bio.SeqIO.to_dict()</tt> example in
Section <a class="reference external" href="#seq:seqio-todict-functionkey">5.4.1.1</a>, you’ll need to
write a tiny function to map from the FASTA identifier (as a string) to
the key you want:</p>
<pre class="code python literal-block">
<span class="keyword">def</span> <span class="name function">get_acc</span><span class="punctuation">(</span><span class="name">identifier</span><span class="punctuation">):</span>
    <span class="literal string doc">&quot;&quot;&quot;&quot;Given a SeqRecord identifier string, return the accession number as a string.

    e.g. &quot;gi|2765613|emb|Z78488.1|PTZ78488&quot; -&gt; &quot;Z78488.1&quot;
    &quot;&quot;&quot;</span>
    <span class="name">parts</span> <span class="operator">=</span> <span class="name">identifier</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">(</span><span class="literal string">&quot;|&quot;</span><span class="punctuation">)</span>
    <span class="keyword">assert</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">parts</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="literal number integer">5</span> <span class="operator word">and</span> <span class="name">parts</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="literal string">&quot;gi&quot;</span> <span class="operator word">and</span> <span class="name">parts</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="literal string">&quot;emb&quot;</span>
    <span class="keyword">return</span> <span class="name">parts</span><span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">]</span>
</pre>
<p>Then we can give this function to the <tt class="docutils literal">Bio.SeqIO.index()</tt> function to
use in building the dictionary:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">,</span> <span class="name">key_function</span><span class="operator">=</span><span class="name">get_acc</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'Z78484.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78464.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78455.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78442.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78532.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78453.1'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Z78471.1'</span><span class="punctuation">]</span>
</pre>
<p>Easy when you know how?</p>
</div>
<div class="section" id="getting-the-raw-data-for-a-record">
<h4>5.4.2.2  Getting the raw data for a record</h4>
<p>The dictionary-like object from <tt class="docutils literal">Bio.SeqIO.index()</tt> gives you each
entry as a <tt class="docutils literal">SeqRecord</tt> object. However, it is sometimes useful to be
able to get the original raw data straight from the file. For this use
the <tt class="docutils literal">get_raw()</tt> method which takes a single argument (the record
identifier) and returns a string (extracted from the file without
modification).</p>
<p>A motivating example is extracting a subset of a records from a large
file where either <tt class="docutils literal">Bio.SeqIO.write()</tt> does not (yet) support the
output file format (e.g. the plain text SwissProt file format) or where
you need to preserve the text exactly (e.g. GenBank or EMBL output from
Biopython does not yet preserve every last bit of annotation).</p>
<p>Let’s suppose you have download the whole of UniProt in the plain text
SwissPort file format from their FTP site
(<a class="reference external" href="ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz">ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz</a>)
and uncompressed it as the file <tt class="docutils literal">uniprot_sprot.dat</tt>, and you want to
extract just a few records from it:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">uniprot</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;uniprot_sprot.dat&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;swiss&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;selected.dat&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">acc</span> <span class="operator word">in</span> <span class="punctuation">[</span><span class="literal string">&quot;P33487&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;P19801&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;P13689&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;Q8JZQ5&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;Q9TRC7&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="name">handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">uniprot</span><span class="operator">.</span><span class="name">get_raw</span><span class="punctuation">(</span><span class="name">acc</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>There is a longer example in Section <a class="reference external" href="#sec:SeqIO-sort">18.1.5</a>
using the <tt class="docutils literal">SeqIO.index()</tt> function to sort a large sequence file
(without loading everything into memory at once).</p>
</div>
</div>
<div class="section" id="sequence-files-as-dictionaries-database-indexed-files">
<h3>5.4.3  Sequence files as Dictionaries – Database indexed files</h3>
<p>Biopython 1.57 introduced an alternative, <tt class="docutils literal">Bio.SeqIO.index_db()</tt>,
which can work on even extremely large files since it stores the record
information as a file on disk (using an SQLite3 database) rather than in
memory. Also, you can index multiple files together (providing all the
record identifiers are unique).</p>
<p>The <tt class="docutils literal">Bio.SeqIO.index()</tt> function takes three required arguments:</p>
<ul class="simple">
<li>Index filename, we suggest using something ending <tt class="docutils literal">.idx</tt>. This
index file is actually an SQLite3 database.</li>
<li>List of sequence filenames to index (or a single filename)</li>
<li>File format (lower case string as used in the rest of the <tt class="docutils literal">SeqIO</tt>
module).</li>
</ul>
<p>As an example, consider the GenBank flat file releases from the NCBI FTP
site,
<a class="reference external" href="ftp://ftp.ncbi.nih.gov/genbank/">ftp://ftp.ncbi.nih.gov/genbank/</a>,
which are gzip compressed GenBank files. As of GenBank release 182,
there are 16 files making up the viral sequences, <tt class="docutils literal">gbvrl1.seq</tt>, …,
<tt class="docutils literal">gbvrl16.seq</tt>, containing in total almost one million records. You can
index them like this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">files</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">&quot;gbvrl</span><span class="literal string interpol">%i</span><span class="literal string">.seq&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">16</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gb_vrl</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index_db</span><span class="punctuation">(</span><span class="literal string">&quot;gbvrl.idx&quot;</span><span class="punctuation">,</span> <span class="name">files</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%i</span><span class="literal string"> sequences indexed&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">gb_vrl</span><span class="punctuation">)</span>
<span class="literal number integer">958086</span> <span class="name">sequences</span> <span class="name">indexed</span>
</pre>
<p>That takes about two minutes to run on my machine. If you rerun it then
the index file (here <tt class="docutils literal">gbvrl.idx</tt>) is reloaded in under a second. You
can use the index as a read only Python dictionary - without having to
worry about which file the sequence comes from, e.g.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">gb_vrl</span><span class="punctuation">[</span><span class="literal string">&quot;GQ333173.1&quot;</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">description</span>
<span class="name">HIV</span><span class="operator">-</span><span class="literal number integer">1</span> <span class="name">isolate</span> <span class="name">F12279A1</span> <span class="keyword namespace">from</span> <span class="name namespace">Uganda</span> <span class="name namespace">gag</span> <span class="name namespace">protein</span> <span class="punctuation">(</span><span class="name">gag</span><span class="punctuation">)</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">partial</span> <span class="name">cds</span><span class="operator">.</span>
</pre>
<div class="section" id="id2">
<h4>5.4.3.1  Getting the raw data for a record</h4>
<p>Just as with the <tt class="docutils literal">Bio.SeqIO.index()</tt> function discussed above in
Section <a class="reference external" href="#sec:seqio-index-getraw">5.4.2.2</a>, the dictionary like
object also lets you get at the raw text of each record:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">gb_vrl</span><span class="operator">.</span><span class="name">get_raw</span><span class="punctuation">(</span><span class="literal string">&quot;GQ333173.1&quot;</span><span class="punctuation">)</span>
<span class="name">LOCUS</span>       <span class="name">GQ333173</span>                 <span class="literal number integer">459</span> <span class="name">bp</span>    <span class="name">DNA</span>     <span class="name">linear</span>   <span class="name">VRL</span> <span class="literal number integer">21</span><span class="operator">-</span><span class="name">OCT</span><span class="operator">-</span><span class="literal number integer">2009</span>
<span class="name">DEFINITION</span>  <span class="name">HIV</span><span class="operator">-</span><span class="literal number integer">1</span> <span class="name">isolate</span> <span class="name">F12279A1</span> <span class="keyword namespace">from</span> <span class="name namespace">Uganda</span> <span class="name namespace">gag</span> <span class="name namespace">protein</span> <span class="punctuation">(</span><span class="name">gag</span><span class="punctuation">)</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">partial</span>
            <span class="name">cds</span><span class="operator">.</span>
<span class="name">ACCESSION</span>   <span class="name">GQ333173</span>
<span class="operator">...</span>
<span class="operator">//</span>
</pre>
</div>
</div>
<div class="section" id="indexing-compressed-files">
<h3>5.4.4  Indexing compressed files</h3>
<p>Very often when you are indexing a sequence file it can be quite large –
so you may want to compress it on disk. Unfortunately efficient random
access is difficult with the more common file formats like gzip and
bzip2. In this setting, BGZF (Blocked GNU Zip Format) can be very
helpful. This is a variant of gzip (and can be decompressed using
standard gzip tools) popularised by the BAM file format,
<a class="reference external" href="http://samtools.sourceforge.net/">samtools</a>, and
<a class="reference external" href="http://samtools.sourceforge.net/tabix.shtml">tabix</a>.</p>
<p>To create a BGZF compressed file you can use the command line tool
<tt class="docutils literal">bgzip</tt> which comes with samtools. In our examples we use a filename
extension <tt class="docutils literal">*.bgz</tt>, so they can be distinguished from normal gzipped
files (named <tt class="docutils literal">*.gz</tt>). You can also use the <tt class="docutils literal">Bio.bgzf</tt> module to read
and write BGZF files from within Python.</p>
<p>The <tt class="docutils literal">Bio.SeqIO.index()</tt> and <tt class="docutils literal">Bio.SeqIO.index_db()</tt> can both be used
with BGZF compressed files. For example, if you started with an
uncompressed GenBank file:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">orchid_dict</span><span class="punctuation">)</span>
<span class="literal number integer">94</span>
</pre>
<p>You could compress this (while keeping the original file) at the command
line using the following command – but don’t worry, the compressed file
is already included with the other example files:</p>
<pre class="code python literal-block">
<span class="error">$</span> <span class="name">bgzip</span> <span class="operator">-</span><span class="name">c</span> <span class="name">ls_orchid</span><span class="operator">.</span><span class="name">gbk</span> <span class="operator">&gt;</span> <span class="name">ls_orchid</span><span class="operator">.</span><span class="name">gbk</span><span class="operator">.</span><span class="name">bgz</span>
</pre>
<p>You can use the compressed file in exactly the same way:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk.bgz&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">orchid_dict</span><span class="punctuation">)</span>
<span class="literal number integer">94</span>
</pre>
<p>or:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index_db</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk.bgz.idx&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;ls_orchid.gbk.bgz&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">orchid_dict</span><span class="punctuation">)</span>
<span class="literal number integer">94</span>
</pre>
<p>The <tt class="docutils literal">SeqIO</tt> indexing automatically detects the BGZF compression. Note
that you can’t use the same index file for the uncompressed and
compressed files.</p>
</div>
<div class="section" id="discussion">
<h3>5.4.5  Discussion</h3>
<p>So, which of these methods should you use and why? It depends on what
you are trying to do (and how much data you are dealing with). However,
in general picking <tt class="docutils literal">Bio.SeqIO.index()</tt> is a good starting point. If
you are dealing with millions of records, multiple files, or repeated
analyses, then look at <tt class="docutils literal">Bio.SeqIO.index_db()</tt>.</p>
<p>Reasons to choose <tt class="docutils literal">Bio.SeqIO.to_dict()</tt> over either
<tt class="docutils literal">Bio.SeqIO.index()</tt> or <tt class="docutils literal">Bio.SeqIO.index_db()</tt> boil down to a need
for flexibility despite its high memory needs. The advantage of storing
the <tt class="docutils literal">SeqRecord</tt> objects in memory is they can be changed, added to, or
removed at will. In addition to the downside of high memory consumption,
indexing can also take longer because all the records must be fully
parsed.</p>
<p>Both <tt class="docutils literal">Bio.SeqIO.index()</tt> and <tt class="docutils literal">Bio.SeqIO.index_db()</tt> only parse
records on demand. When indexing, they scan the file once looking for
the start of each record and do as little work as possible to extract
the identifier.</p>
<p>Reasons to choose <tt class="docutils literal">Bio.SeqIO.index()</tt> over <tt class="docutils literal">Bio.SeqIO.index_db()</tt>
include:</p>
<ul class="simple">
<li>Faster to build the index (more noticeable in simple file formats)</li>
<li>Slightly faster access as SeqRecord objects (but the difference is
only really noticeable for simple to parse file formats).</li>
<li>Can use any immutable Python object as the dictionary keys (e.g. a
tuple of strings, or a frozen set) not just strings.</li>
<li>Don’t need to worry about the index database being out of date if the
sequence file being indexed has changed.</li>
</ul>
<p>Reasons to choose <tt class="docutils literal">Bio.SeqIO.index_db()</tt> over <tt class="docutils literal">Bio.SeqIO.index()</tt>
include:</p>
<ul class="simple">
<li>Not memory limited – this is already important with files from second
generation sequencing where 10s of millions of sequences are common,
and using <tt class="docutils literal">Bio.SeqIO.index()</tt> can require more than 4GB of RAM and
therefore a 64bit version of Python.</li>
<li>Because the index is kept on disk, it can be reused. Although
building the index database file takes longer, if you have a script
which will be rerun on the same datafiles in future, this could save
time in the long run.</li>
<li>Indexing multiple files together</li>
<li>The <tt class="docutils literal">get_raw()</tt> method can be much faster, since for most file
formats the length of each record is stored as well as its offset.</li>
</ul>
</div>
</div>
<div class="section" id="writing-sequence-files">
<h2>5.5  Writing Sequence Files</h2>
<p>We’ve talked about using <tt class="docutils literal">Bio.SeqIO.parse()</tt> for sequence input
(reading files), and now we’ll look at <tt class="docutils literal">Bio.SeqIO.write()</tt> which is
for sequence output (writing files). This is a function taking three
arguments: some <tt class="docutils literal">SeqRecord</tt> objects, a handle or filename to write to,
and a sequence format.</p>
<p>Here is an example, where we start by creating a few <tt class="docutils literal">SeqRecord</tt>
objects the hard way (by hand, rather than by loading them from a file):</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_protein</span>

<span class="name">rec1</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;SSAC&quot;</span><span class="punctuation">,</span> <span class="name">generic_protein</span><span class="punctuation">),</span>
                 <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;gi|14150838|gb|AAK54648.1|AF376133_1&quot;</span><span class="punctuation">,</span>
                 <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;chalcone synthase [Cucumis sativus]&quot;</span><span class="punctuation">)</span>

<span class="name">rec2</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;YPDYYFRITNREHKAELKEKFQRMCDKSMIKKRYMYLTEEILKENPSMCEYMAPSLDARQ&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;DMVVVEIPKLGKEAAVKAIKEWGQ&quot;</span><span class="punctuation">,</span> <span class="name">generic_protein</span><span class="punctuation">),</span>
                 <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;gi|13919613|gb|AAK33142.1|&quot;</span><span class="punctuation">,</span>
                 <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;chalcone synthase [Fragaria vesca subsp. bracteata]&quot;</span><span class="punctuation">)</span>

<span class="name">rec3</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;MVTVEEFRRAQCAEGPATVMAIGTATPSNCVDQSTYPDYYFRITNSEHKVELKEKFKRMC&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;EKSMIKKRYMHLTEEILKENPNICAYMAPSLDARQDIVVVEVPKLGKEAAQKAIKEWGQP&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;KSKITHLVFCTTSGVDMPGCDYQLTKLLGLRPSVKRFMMYQQGCFAGGTVLRMAKDLAEN&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;NKGARVLVVCSEITAVTFRGPNDTHLDSLVGQALFGDGAAAVIIGSDPIPEVERPLFELV&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;SAAQTLLPDSEGAIDGHLREVGLTFHLLKDVPGLISKNIEKSLVEAFQPLGISDWNSLFW&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;IAHPGGPAILDQVELKLGLKQEKLKATRKVLSNYGNMSSACVLFILDEMRKASAKEGLGT&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;TGEGLEWGVLFGFGPGLTVETVVLHSVAT&quot;</span><span class="punctuation">,</span> <span class="name">generic_protein</span><span class="punctuation">),</span>
                 <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;gi|13925890|gb|AAK49457.1|&quot;</span><span class="punctuation">,</span>
                 <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;chalcone synthase [Nicotiana tabacum]&quot;</span><span class="punctuation">)</span>

<span class="name">my_records</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">rec1</span><span class="punctuation">,</span> <span class="name">rec2</span><span class="punctuation">,</span> <span class="name">rec3</span><span class="punctuation">]</span>
</pre>
<p>Now we have a list of <tt class="docutils literal">SeqRecord</tt> objects, we’ll write them to a FASTA
format file:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">my_records</span><span class="punctuation">,</span> <span class="literal string">&quot;my_example.faa&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>And if you open this file in your favourite text editor it should look
like this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">14150838</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AAK54648</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF376133_1</span> <span class="name">chalcone</span> <span class="name">synthase</span> <span class="punctuation">[</span><span class="name">Cucumis</span> <span class="name">sativus</span><span class="punctuation">]</span>
<span class="name">MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD</span>
<span class="name">GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK</span>
<span class="name">NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM</span>
<span class="name">SSAC</span>
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">13919613</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AAK33142</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="name">chalcone</span> <span class="name">synthase</span> <span class="punctuation">[</span><span class="name">Fragaria</span> <span class="name">vesca</span> <span class="name">subsp</span><span class="operator">.</span> <span class="name">bracteata</span><span class="punctuation">]</span>
<span class="name">YPDYYFRITNREHKAELKEKFQRMCDKSMIKKRYMYLTEEILKENPSMCEYMAPSLDARQ</span>
<span class="name">DMVVVEIPKLGKEAAVKAIKEWGQ</span>
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">13925890</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AAK49457</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="name">chalcone</span> <span class="name">synthase</span> <span class="punctuation">[</span><span class="name">Nicotiana</span> <span class="name">tabacum</span><span class="punctuation">]</span>
<span class="name">MVTVEEFRRAQCAEGPATVMAIGTATPSNCVDQSTYPDYYFRITNSEHKVELKEKFKRMC</span>
<span class="name">EKSMIKKRYMHLTEEILKENPNICAYMAPSLDARQDIVVVEVPKLGKEAAQKAIKEWGQP</span>
<span class="name">KSKITHLVFCTTSGVDMPGCDYQLTKLLGLRPSVKRFMMYQQGCFAGGTVLRMAKDLAEN</span>
<span class="name">NKGARVLVVCSEITAVTFRGPNDTHLDSLVGQALFGDGAAAVIIGSDPIPEVERPLFELV</span>
<span class="name">SAAQTLLPDSEGAIDGHLREVGLTFHLLKDVPGLISKNIEKSLVEAFQPLGISDWNSLFW</span>
<span class="name">IAHPGGPAILDQVELKLGLKQEKLKATRKVLSNYGNMSSACVLFILDEMRKASAKEGLGT</span>
<span class="name">TGEGLEWGVLFGFGPGLTVETVVLHSVAT</span>
</pre>
<p>Suppose you wanted to know how many records the <tt class="docutils literal">Bio.SeqIO.write()</tt>
function wrote to the handle? If your records were in a list you could
just use <tt class="docutils literal">len(my_records)</tt>, however you can’t do that when your
records come from a generator/iterator. The <tt class="docutils literal">Bio.SeqIO.write()</tt>
function returns the number of <tt class="docutils literal">SeqRecord</tt> objects written to the
file.</p>
<p><em>Note</em> - If you tell the <tt class="docutils literal">Bio.SeqIO.write()</tt> function to write to a
file that already exists, the old file will be overwritten without any
warning.</p>
<div class="section" id="round-trips">
<h3>5.5.1  Round trips</h3>
<p>Some people like their parsers to be “round-tripable”, meaning if you
read in a file and write it back out again it is unchanged. This
requires that the parser must extract enough information to reproduce
the original file <em>exactly</em>. <tt class="docutils literal">Bio.SeqIO</tt> does <em>not</em> aim to do this.</p>
<p>As a trivial example, any line wrapping of the sequence data in FASTA
files is allowed. An identical <tt class="docutils literal">SeqRecord</tt> would be given from parsing
the following two examples which differ only in their line breaks:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">YAL068C</span><span class="operator">-</span><span class="literal number float">7235.2170</span> <span class="name">Putative</span> <span class="name">promoter</span> <span class="name">sequence</span>
<span class="name">TACGAGAATAATTTCTCATCATCCAGCTTTAACACAAAATTCGCACAGTTTTCGTTAAGA</span>
<span class="name">GAACTTAACATTTTCTTATGACGTAAATGAAGTTTATATATAAATTTCCTTTTTATTGGA</span>

<span class="operator">&gt;</span><span class="name">YAL068C</span><span class="operator">-</span><span class="literal number float">7235.2170</span> <span class="name">Putative</span> <span class="name">promoter</span> <span class="name">sequence</span>
<span class="name">TACGAGAATAATTTCTCATCATCCAGCTTTAACACAAAATTCGCA</span>
<span class="name">CAGTTTTCGTTAAGAGAACTTAACATTTTCTTATGACGTAAATGA</span>
<span class="name">AGTTTATATATAAATTTCCTTTTTATTGGA</span>
</pre>
<p>To make a round-tripable FASTA parser you would need to keep track of
where the sequence line breaks occurred, and this extra information is
usually pointless. Instead Biopython uses a default line wrapping of 60
characters on output. The same problem with white space applies in many
other file formats too. Another issue in some cases is that Biopython
does not (yet) preserve every last bit of annotation (e.g. GenBank and
EMBL).</p>
<p>Occasionally preserving the original layout (with any quirks it may
have) is important. See Section <a class="reference external" href="#sec:seqio-index-getraw">5.4.2.2</a>
about the <tt class="docutils literal">get_raw()</tt> method of the <tt class="docutils literal">Bio.SeqIO.index()</tt>
dictionary-like object for one potential solution.</p>
</div>
<div class="section" id="converting-between-sequence-file-formats">
<h3>5.5.2  Converting between sequence file formats</h3>
<p>In previous example we used a list of <tt class="docutils literal">SeqRecord</tt> objects as input to
the <tt class="docutils literal">Bio.SeqIO.write()</tt> function, but it will also accept a
<tt class="docutils literal">SeqRecord</tt> iterator like we get from <tt class="docutils literal">Bio.SeqIO.parse()</tt> – this
lets us do file conversion by combining these two functions.</p>
<p>For this example we’ll read in the GenBank format file
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk">ls_orchid.gbk</a>
and write it out in FASTA format:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;my_example.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Converted </span><span class="literal string interpol">%i</span><span class="literal string"> records&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>Still, that is a little bit complicated. So, because file conversion is
such a common task, there is a helper function letting you replace that
with just:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;my_example.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Converted </span><span class="literal string interpol">%i</span><span class="literal string"> records&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>The <tt class="docutils literal">Bio.SeqIO.convert()</tt> function will take handles <em>or</em> filenames.
Watch out though – if the output file already exists, it will overwrite
it! To find out more, see the built in help:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>In principle, just by changing the filenames and the format names, this
code could be used to convert between any file formats available in
Biopython. However, writing some formats requires information (e.g.
quality scores) which other files formats don’t contain. For example,
while you can turn a FASTQ file into a FASTA file, you can’t do the
reverse. See also Sections <a class="reference external" href="#sec:SeqIO-fastq-conversion">18.1.9</a>
and <a class="reference external" href="#sec:SeqIO-fasta-qual-conversion">18.1.10</a> in the cookbook
chapter which looks at inter-converting between different FASTQ formats.</p>
<p>Finally, as an added incentive for using the <tt class="docutils literal">Bio.SeqIO.convert()</tt>
function (on top of the fact your code will be shorter), doing it this
way may also be faster! The reason for this is the convert function can
take advantage of several file format specific optimisations and tricks.</p>
</div>
<div class="section" id="converting-a-file-of-sequences-to-their-reverse-complements">
<h3>5.5.3  Converting a file of sequences to their reverse complements</h3>
<p>Suppose you had a file of nucleotide sequences, and you wanted to turn
it into a file containing their reverse complement sequences. This time
a little bit of work is required to transform the <tt class="docutils literal">SeqRecord</tt> objects
we get from our input file into something suitable for saving to our
output file.</p>
<p>To start with, we’ll use <tt class="docutils literal">Bio.SeqIO.parse()</tt> to load some nucleotide
sequences from a file, then print out their reverse complements using
the <tt class="docutils literal">Seq</tt> object’s built in <tt class="docutils literal">.reverse_complement()</tt> method (see
Section <a class="reference external" href="#sec:seq-reverse-complement">3.7</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
</pre>
<p>Now, if we want to save these reverse complements to a file, we’ll need
to make <tt class="docutils literal">SeqRecord</tt> objects. We can use the <tt class="docutils literal">SeqRecord</tt> object’s
built in <tt class="docutils literal">.reverse_complement()</tt> method (see
Section <a class="reference external" href="#sec:SeqRecord-reverse-complement">4.8</a>) but we must
decide how to name our new records.</p>
<p>This is an excellent place to demonstrate the power of list
comprehensions which make a list in memory:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">rec</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;rc_&quot;</span><span class="operator">+</span><span class="name">rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;reverse complement&quot;</span><span class="punctuation">)</span> \
<span class="operator">...</span>            <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">)</span>
<span class="literal number integer">94</span>
</pre>
<p>Now list comprehensions have a nice trick up their sleeves, you can add
a conditional statement:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">rec</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;rc_&quot;</span><span class="operator">+</span><span class="name">rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;reverse complement&quot;</span><span class="punctuation">)</span> \
<span class="operator">...</span>            <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span><span class="operator">&lt;</span><span class="literal number integer">700</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">)</span>
<span class="literal number integer">18</span>
</pre>
<p>That would create an in memory list of reverse complement records where
the sequence length was under 700 base pairs. However, we can do exactly
the same with a generator expression - but with the advantage that this
does not create a list of all the records in memory at once:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;rc_&quot;</span><span class="operator">+</span><span class="name">rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;reverse complement&quot;</span><span class="punctuation">)</span> \
<span class="operator">...</span>           <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span><span class="operator">&lt;</span><span class="literal number integer">700</span><span class="punctuation">)</span>
</pre>
<p>As a complete example:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;rc_&quot;</span><span class="operator">+</span><span class="name">rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;reverse complement&quot;</span><span class="punctuation">)</span> \
<span class="operator">...</span>            <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span><span class="operator">&lt;</span><span class="literal number integer">700</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;rev_comp.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">18</span>
</pre>
<p>There is a related example in
Section <a class="reference external" href="#sec:SeqIO-translate">18.1.3</a>, translating each record in
a FASTA file from nucleotides to amino acids.</p>
</div>
<div class="section" id="getting-your-seqrecord-objects-as-formatted-strings">
<h3>5.5.4  Getting your SeqRecord objects as formatted strings</h3>
<p>Suppose that you don’t really want to write your records to a file or
handle – instead you want a string containing the records in a
particular file format. The <tt class="docutils literal">Bio.SeqIO</tt> interface is based on handles,
but Python has a useful built in module which provides a string based
handle.</p>
<p>For an example of how you might use this, let’s load in a bunch of
<tt class="docutils literal">SeqRecord</tt> objects from our orchids GenBank file, and create a string
containing the records in FASTA format:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">StringIO</span> <span class="keyword namespace">import</span> <span class="name">StringIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="name">out_handle</span> <span class="operator">=</span> <span class="name">StringIO</span><span class="punctuation">()</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">out_handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="name">fasta_data</span> <span class="operator">=</span> <span class="name">out_handle</span><span class="operator">.</span><span class="name">getvalue</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="name">fasta_data</span>
</pre>
<p>This isn’t entirely straightforward the first time you see it! On the
bright side, for the special case where you would like a string
containing a <em>single</em> record in a particular file format, use the the
<tt class="docutils literal">SeqRecord</tt> class’ <tt class="docutils literal">format()</tt> method (see
Section <a class="reference external" href="#sec:SeqRecord-format">4.5</a>).</p>
<p>Note that although we don’t encourage it, you <em>can</em> use the <tt class="docutils literal">format()</tt>
method to write to a file, for example something like this:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">out_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid_long.tab&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">):</span>
    <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="literal number integer">100</span><span class="punctuation">:</span>
        <span class="name">out_handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;tab&quot;</span><span class="punctuation">))</span>
<span class="name">out_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>While this style of code will work for a simple sequential file format
like FASTA or the simple tab separated format used here, it will <em>not</em>
work for more complex or interlaced file formats. This is why we still
recommend using <tt class="docutils literal">Bio.SeqIO.write()</tt>, as in the following example:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="literal number integer">100</span><span class="punctuation">)</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;ls_orchid.tab&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;tab&quot;</span><span class="punctuation">)</span>
</pre>
<p>Making a single call to <tt class="docutils literal"><span class="pre">SeqIO.write(...)</span></tt> is also much quicker than
multiple calls to the <tt class="docutils literal"><span class="pre">SeqRecord.format(...)</span></tt> method.</p>
</div>
</div>
</div>
<div class="section" id="chapter-6-multiple-sequence-alignment-objects">
<h1>Chapter 6  Multiple Sequence Alignment objects</h1>
<p>This chapter is about Multiple Sequence Alignments, by which we mean a
collection of multiple sequences which have been aligned together –
usually with the insertion of gap characters, and addition of leading or
trailing gaps – such that all the sequence strings are the same length.
Such an alignment can be regarded as a matrix of letters, where each row
is held as a <tt class="docutils literal">SeqRecord</tt> object internally.</p>
<p>We will introduce the <tt class="docutils literal">MultipleSeqAlignment</tt> object which holds this
kind of data, and the <tt class="docutils literal">Bio.AlignIO</tt> module for reading and writing
them as various file formats (following the design of the <tt class="docutils literal">Bio.SeqIO</tt>
module from the previous chapter). Note that both <tt class="docutils literal">Bio.SeqIO</tt> and
<tt class="docutils literal">Bio.AlignIO</tt> can read and write sequence alignment files. The
appropriate choice will depend largely on what you want to do with the
data.</p>
<p>The final part of this chapter is about our command line wrappers for
common multiple sequence alignment tools like ClustalW and MUSCLE.</p>
<div class="section" id="parsing-or-reading-sequence-alignments">
<h2>6.1  Parsing or Reading Sequence Alignments</h2>
<p>We have two functions for reading in sequence alignments,
<tt class="docutils literal">Bio.AlignIO.read()</tt> and <tt class="docutils literal">Bio.AlignIO.parse()</tt> which following the
convention introduced in <tt class="docutils literal">Bio.SeqIO</tt> are for files containing one or
multiple alignments respectively.</p>
<p>Using <tt class="docutils literal">Bio.AlignIO.parse()</tt> will return an <em>iterator</em> which gives
<tt class="docutils literal">MultipleSeqAlignment</tt> objects. Iterators are typically used in a for
loop. Examples of situations where you will have multiple different
alignments include resampled alignments from the PHYLIP tool
<tt class="docutils literal">seqboot</tt>, or multiple pairwise alignments from the EMBOSS tools
<tt class="docutils literal">water</tt> or <tt class="docutils literal">needle</tt>, or Bill Pearson’s FASTA tools.</p>
<p>However, in many situations you will be dealing with files which contain
only a single alignment. In this case, you should use the
<tt class="docutils literal">Bio.AlignIO.read()</tt> function which returns a single
<tt class="docutils literal">MultipleSeqAlignment</tt> object.</p>
<p>Both functions expect two mandatory arguments:</p>
<ol class="arabic simple">
<li>The first argument is a <em>handle</em> to read the data from, typically an
open file (see Section <a class="reference external" href="#sec:appendix-handles">22.1</a>), or a
filename.</li>
<li>The second argument is a lower case string specifying the alignment
format. As in <tt class="docutils literal">Bio.SeqIO</tt> we don’t try and guess the file format
for you! See
<a class="reference external" href="http://biopython.org/wiki/AlignIO">http://biopython.org/wiki/AlignIO</a>
for a full listing of supported formats.</li>
</ol>
<p>There is also an optional <tt class="docutils literal">seq_count</tt> argument which is discussed in
Section <a class="reference external" href="#sec:AlignIO-count-argument">6.1.3</a> below for dealing with
ambiguous file formats which may contain more than one alignment.</p>
<p>A further optional <tt class="docutils literal">alphabet</tt> argument allowing you to specify the
expected alphabet. This can be useful as many alignment file formats do
not explicitly label the sequences as RNA, DNA or protein – which means
<tt class="docutils literal">Bio.AlignIO</tt> will default to using a generic alphabet.</p>
<div class="section" id="single-alignments">
<h3>6.1.1  Single Alignments</h3>
<p>As an example, consider the following annotation rich protein alignment
in the PFAM or Stockholm file format:</p>
<pre class="code python literal-block">
<span class="comment"># STOCKHOLM 1.0</span>
<span class="comment">#=GS COATB_BPIKE/30-81  AC P03620.1</span>
<span class="comment">#=GS COATB_BPIKE/30-81  DR PDB; 1ifl ; 1-52;</span>
<span class="comment">#=GS Q9T0Q8_BPIKE/1-52  AC Q9T0Q8.1</span>
<span class="comment">#=GS COATB_BPI22/32-83  AC P15416.1</span>
<span class="comment">#=GS COATB_BPM13/24-72  AC P69541.1</span>
<span class="comment">#=GS COATB_BPM13/24-72  DR PDB; 2cpb ; 1-49;</span>
<span class="comment">#=GS COATB_BPM13/24-72  DR PDB; 2cps ; 1-49;</span>
<span class="comment">#=GS COATB_BPZJ2/1-49   AC P03618.1</span>
<span class="comment">#=GS Q9T0Q9_BPFD/1-49   AC Q9T0Q9.1</span>
<span class="comment">#=GS Q9T0Q9_BPFD/1-49   DR PDB; 1nh4 A; 1-49;</span>
<span class="comment">#=GS COATB_BPIF1/22-73  AC P03619.2</span>
<span class="comment">#=GS COATB_BPIF1/22-73  DR PDB; 1ifk ; 1-50;</span>
<span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>             <span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span>
<span class="comment">#=GR COATB_BPIKE/30-81  SS    -HHHHHHHHHHHHHH--HHHHHHHH--HHHHHHHHHHHHHHHHHHHHH----</span>
<span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>             <span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span>
<span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>             <span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span>
<span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>             <span class="name">AEGDDP</span><span class="operator">...</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span>
<span class="comment">#=GR COATB_BPM13/24-72  SS    ---S-T...CHCHHHHCCCCTCCCTTCHHHHHHHHHHHHHHHHHHHHCTT--</span>
<span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>              <span class="name">AEGDDP</span><span class="operator">...</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span>
<span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>              <span class="name">AEGDDP</span><span class="operator">...</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span>
<span class="comment">#=GR Q9T0Q9_BPFD/1-49   SS    ------...-HHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH--</span>
<span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>             <span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span>
<span class="comment">#=GR COATB_BPIF1/22-73  SS    XX-HHHH--HHHHHH--HHHHHHH--HHHHHHHHHHHHHHHHHHHHHHH---</span>
<span class="comment">#=GC SS_cons                  XHHHHHHHHHHHHHHHCHHHHHHHHCHHHHHHHHHHHHHHHHHHHHHHHC--</span>
<span class="comment">#=GC seq_cons                 AEssss...AptAhDSLpspAT-hIu.sWshVsslVsAsluIKLFKKFsSKA</span>
<span class="operator">//</span>
</pre>
<p>This is the seed alignment for the Phage_Coat_Gp8 (PF05371) PFAM
entry, downloaded from a now out of date release of PFAM from
<a class="reference external" href="http://pfam.sanger.ac.uk/">http://pfam.sanger.ac.uk/</a>. We can
load this file as follows (assuming it has been saved to disk as
“PF05371_seed.sth” in the current working directory):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
</pre>
<p>This code will print out a summary of the alignment:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">52</span> <span class="name">columns</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKL</span><span class="operator">...</span><span class="name">SRA</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKL</span><span class="operator">...</span><span class="name">SRA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>You’ll notice in the above output the sequences have been truncated. We
could instead write our own code to format this as we please by
iterating over the rows as <tt class="docutils literal">SeqRecord</tt> objects:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;Alignment length </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">get_alignment_length</span><span class="punctuation">()</span>
<span class="name">Alignment</span> <span class="name">length</span> <span class="literal number integer">52</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> - </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span> <span class="operator">-</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span> <span class="operator">-</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span> <span class="operator">-</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="operator">-</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span> <span class="operator">-</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="operator">-</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span> <span class="operator">-</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>You could also use the alignment object’s <tt class="docutils literal">format</tt> method to show it
in a particular file format – see
Section <a class="reference external" href="#sec:alignment-format-method">6.2.2</a> for details.</p>
<p>Did you notice in the raw file above that several of the sequences
include database cross-references to the PDB and the associated known
secondary structure? Try this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span> <span class="punctuation">[</span><span class="literal string">'PDB; 1ifl ; 1-52;'</span><span class="punctuation">]</span>
<span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span> <span class="punctuation">[</span><span class="literal string">'PDB; 2cpb ; 1-49;'</span><span class="punctuation">,</span> <span class="literal string">'PDB; 2cps ; 1-49;'</span><span class="punctuation">]</span>
<span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span> <span class="punctuation">[</span><span class="literal string">'PDB; 1nh4 A; 1-49;'</span><span class="punctuation">]</span>
<span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span> <span class="punctuation">[</span><span class="literal string">'PDB; 1ifk ; 1-50;'</span><span class="punctuation">]</span>
</pre>
<p>To have a look at all the sequence annotation, try this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span>
</pre>
<p>Sanger provide a nice web interface at
<a class="reference external" href="http://pfam.sanger.ac.uk/family?acc=PF05371">http://pfam.sanger.ac.uk/family?acc=PF05371</a>
which will actually let you download this alignment in several other
formats. This is what the file looks like in the FASTA file format:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span>
<span class="operator">&gt;</span><span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span>
<span class="operator">&gt;</span><span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span>
<span class="operator">&gt;</span><span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span>
<span class="operator">&gt;</span><span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span>
<span class="operator">&gt;</span><span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span>
<span class="operator">&gt;</span><span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
<span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span>
</pre>
<p>Note the website should have an option about showing gaps as periods
(dots) or dashes, we’ve shown dashes above. Assuming you download and
save this as file “PF05371_seed.faa” then you can load it with almost
exactly the same code:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.faa&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name">alignment</span>
</pre>
<p>All that has changed in this code is the filename and the format string.
You’ll get the same output as before, the sequences and record
identifiers are the same. However, as you should expect, if you check
each <tt class="docutils literal">SeqRecord</tt> there is no annotation nor database cross-references
because these are not included in the FASTA file format.</p>
<p>Note that rather than using the Sanger website, you could have used
<tt class="docutils literal">Bio.AlignIO</tt> to convert the original Stockholm format file into a
FASTA file yourself (see below).</p>
<p>With any supported file format, you can load an alignment in exactly the
same way just by changing the format string. For example, use “phylip”
for PHYLIP files, “nexus” for NEXUS files or “emboss” for the alignments
output by the EMBOSS tools. There is a full listing on the wiki page
(<a class="reference external" href="http://biopython.org/wiki/AlignIO">http://biopython.org/wiki/AlignIO</a>)
and in the built in documentation (also
<a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.AlignIO-module.html">online</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">AlignIO</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="multiple-alignments">
<h3>6.1.2  Multiple Alignments</h3>
<p>The previous section focused on reading files containing a single
alignment. In general however, files can contain more than one
alignment, and to read these files we must use the
<tt class="docutils literal">Bio.AlignIO.parse()</tt> function.</p>
<p>Suppose you have a small alignment in PHYLIP format:</p>
<pre class="code python literal-block">
    <span class="literal number integer">5</span>    <span class="literal number integer">6</span>
<span class="name">Alpha</span>     <span class="name">AACAAC</span>
<span class="name">Beta</span>      <span class="name">AACCCC</span>
<span class="name">Gamma</span>     <span class="name">ACCAAC</span>
<span class="name">Delta</span>     <span class="name">CCACCA</span>
<span class="name">Epsilon</span>   <span class="name">CCAAAC</span>
</pre>
<p>If you wanted to bootstrap a phylogenetic tree using the PHYLIP tools,
one of the steps would be to create a set of many resampled alignments
using the tool <tt class="docutils literal">bootseq</tt>. This would give output something like this,
which has been abbreviated for conciseness:</p>
<pre class="code python literal-block">
    <span class="literal number integer">5</span>     <span class="literal number integer">6</span>
<span class="name">Alpha</span>     <span class="name">AAACCA</span>
<span class="name">Beta</span>      <span class="name">AAACCC</span>
<span class="name">Gamma</span>     <span class="name">ACCCCA</span>
<span class="name">Delta</span>     <span class="name">CCCAAC</span>
<span class="name">Epsilon</span>   <span class="name">CCCAAA</span>
    <span class="literal number integer">5</span>     <span class="literal number integer">6</span>
<span class="name">Alpha</span>     <span class="name">AAACAA</span>
<span class="name">Beta</span>      <span class="name">AAACCC</span>
<span class="name">Gamma</span>     <span class="name">ACCCAA</span>
<span class="name">Delta</span>     <span class="name">CCCACC</span>
<span class="name">Epsilon</span>   <span class="name">CCCAAA</span>
    <span class="literal number integer">5</span>     <span class="literal number integer">6</span>
<span class="name">Alpha</span>     <span class="name">AAAAAC</span>
<span class="name">Beta</span>      <span class="name">AAACCC</span>
<span class="name">Gamma</span>     <span class="name">AACAAC</span>
<span class="name">Delta</span>     <span class="name">CCCCCA</span>
<span class="name">Epsilon</span>   <span class="name">CCCAAC</span>
<span class="operator">...</span>
    <span class="literal number integer">5</span>     <span class="literal number integer">6</span>
<span class="name">Alpha</span>     <span class="name">AAAACC</span>
<span class="name">Beta</span>      <span class="name">ACCCCC</span>
<span class="name">Gamma</span>     <span class="name">AAAACC</span>
<span class="name">Delta</span>     <span class="name">CCCCAA</span>
<span class="name">Epsilon</span>   <span class="name">CAAACC</span>
</pre>
<p>If you wanted to read this in using <tt class="docutils literal">Bio.AlignIO</tt> you could use:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignments</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;resampled.phy&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phylip&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">alignment</span> <span class="operator word">in</span> <span class="name">alignments</span><span class="punctuation">:</span>
    <span class="keyword">print</span> <span class="name">alignment</span>
    <span class="keyword">print</span>
</pre>
<p>This would give the following output, again abbreviated for display:</p>
<pre class="code python literal-block">
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">5</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">6</span> <span class="name">columns</span>
<span class="name">AAACCA</span> <span class="name">Alpha</span>
<span class="name">AAACCC</span> <span class="name">Beta</span>
<span class="name">ACCCCA</span> <span class="name">Gamma</span>
<span class="name">CCCAAC</span> <span class="name">Delta</span>
<span class="name">CCCAAA</span> <span class="name">Epsilon</span>

<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">5</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">6</span> <span class="name">columns</span>
<span class="name">AAACAA</span> <span class="name">Alpha</span>
<span class="name">AAACCC</span> <span class="name">Beta</span>
<span class="name">ACCCAA</span> <span class="name">Gamma</span>
<span class="name">CCCACC</span> <span class="name">Delta</span>
<span class="name">CCCAAA</span> <span class="name">Epsilon</span>

<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">5</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">6</span> <span class="name">columns</span>
<span class="name">AAAAAC</span> <span class="name">Alpha</span>
<span class="name">AAACCC</span> <span class="name">Beta</span>
<span class="name">AACAAC</span> <span class="name">Gamma</span>
<span class="name">CCCCCA</span> <span class="name">Delta</span>
<span class="name">CCCAAC</span> <span class="name">Epsilon</span>

<span class="operator">...</span>

<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">5</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">6</span> <span class="name">columns</span>
<span class="name">AAAACC</span> <span class="name">Alpha</span>
<span class="name">ACCCCC</span> <span class="name">Beta</span>
<span class="name">AAAACC</span> <span class="name">Gamma</span>
<span class="name">CCCCAA</span> <span class="name">Delta</span>
<span class="name">CAAACC</span> <span class="name">Epsilon</span>
</pre>
<p>As with the function <tt class="docutils literal">Bio.SeqIO.parse()</tt>, using
<tt class="docutils literal">Bio.AlignIO.parse()</tt> returns an iterator. If you want to keep all the
alignments in memory at once, which will allow you to access them in any
order, then turn the iterator into a list:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignments</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">AlignIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;resampled.phy&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phylip&quot;</span><span class="punctuation">))</span>
<span class="name">last_align</span> <span class="operator">=</span> <span class="name">alignments</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="name">first_align</span> <span class="operator">=</span> <span class="name">alignments</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="ambiguous-alignments">
<h3>6.1.3  Ambiguous Alignments</h3>
<p>Many alignment file formats can explicitly store more than one
alignment, and the division between each alignment is clear. However,
when a general sequence file format has been used there is no such block
structure. The most common such situation is when alignments have been
saved in the FASTA file format. For example consider the following:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAG</span><span class="operator">--</span><span class="name">G</span>
<span class="operator">&gt;</span><span class="name">Beta</span>
<span class="name">ACTACCGCTAGCTCAGAAG</span>
<span class="operator">&gt;</span><span class="name">Gamma</span>
<span class="name">ACTACGGCTAGCACAGAAG</span>
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAGG</span><span class="operator">--</span>
<span class="operator">&gt;</span><span class="name">Beta</span>
<span class="name">ACTACCGCTAGCTCAGAAG</span>
<span class="operator">&gt;</span><span class="name">Gamma</span>
<span class="name">ACTACGGCTAGCACAGAAG</span>
</pre>
<p>This could be a single alignment containing six sequences (with repeated
identifiers). Or, judging from the identifiers, this is probably two
different alignments each with three sequences, which happen to all have
the same length.</p>
<p>What about this next example?</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAG</span><span class="operator">--</span><span class="name">G</span>
<span class="operator">&gt;</span><span class="name">Beta</span>
<span class="name">ACTACCGCTAGCTCAGAAG</span>
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAGG</span><span class="operator">--</span>
<span class="operator">&gt;</span><span class="name">Gamma</span>
<span class="name">ACTACGGCTAGCACAGAAG</span>
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAGG</span><span class="operator">--</span>
<span class="operator">&gt;</span><span class="name">Delta</span>
<span class="name">ACTACGGCTAGCACAGAAG</span>
</pre>
<p>Again, this could be a single alignment with six sequences. However this
time based on the identifiers we might guess this is three pairwise
alignments which by chance have all got the same lengths.</p>
<p>This final example is similar:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAG</span><span class="operator">--</span><span class="name">G</span>
<span class="operator">&gt;</span><span class="name">XXX</span>
<span class="name">ACTACCGCTAGCTCAGAAG</span>
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAGG</span>
<span class="operator">&gt;</span><span class="name">YYY</span>
<span class="name">ACTACGGCAAGCACAGG</span>
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="operator">--</span><span class="name">ACTACGAC</span><span class="operator">--</span><span class="name">TAGCTCAGG</span>
<span class="operator">&gt;</span><span class="name">ZZZ</span>
<span class="name">GGACTACGACAATAGCTCAGG</span>
</pre>
<p>In this third example, because of the differing lengths, this cannot be
treated as a single alignment containing all six records. However, it
could be three pairwise alignments.</p>
<p>Clearly trying to store more than one alignment in a FASTA file is not
ideal. However, if you are forced to deal with these as input files
<tt class="docutils literal">Bio.AlignIO</tt> can cope with the most common situation where all the
alignments have the same number of records. One example of this is a
collection of pairwise alignments, which can be produced by the EMBOSS
tools <tt class="docutils literal">needle</tt> and <tt class="docutils literal">water</tt> – although in this situation,
<tt class="docutils literal">Bio.AlignIO</tt> should be able to understand their native output using
“emboss” as the format string.</p>
<p>To interpret these FASTA examples as several separate alignments, we can
use <tt class="docutils literal">Bio.AlignIO.parse()</tt> with the optional <tt class="docutils literal">seq_count</tt> argument
which specifies how many sequences are expected in each alignment (in
these examples, 3, 2 and 2 respectively). For example, using the third
example as the input data:</p>
<pre class="code python literal-block">
<span class="keyword">for</span> <span class="name">alignment</span> <span class="operator word">in</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">,</span> <span class="name">seq_count</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="literal string">&quot;Alignment length </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">get_alignment_length</span><span class="punctuation">()</span>
    <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">:</span>
        <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> - </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>
    <span class="keyword">print</span>
</pre>
<p>giving:</p>
<pre class="code python literal-block">
<span class="name">Alignment</span> <span class="name">length</span> <span class="literal number integer">19</span>
<span class="name">ACTACGACTAGCTCAG</span><span class="operator">--</span><span class="name">G</span> <span class="operator">-</span> <span class="name">Alpha</span>
<span class="name">ACTACCGCTAGCTCAGAAG</span> <span class="operator">-</span> <span class="name">XXX</span>

<span class="name">Alignment</span> <span class="name">length</span> <span class="literal number integer">17</span>
<span class="name">ACTACGACTAGCTCAGG</span> <span class="operator">-</span> <span class="name">Alpha</span>
<span class="name">ACTACGGCAAGCACAGG</span> <span class="operator">-</span> <span class="name">YYY</span>

<span class="name">Alignment</span> <span class="name">length</span> <span class="literal number integer">21</span>
<span class="operator">--</span><span class="name">ACTACGAC</span><span class="operator">--</span><span class="name">TAGCTCAGG</span> <span class="operator">-</span> <span class="name">Alpha</span>
<span class="name">GGACTACGACAATAGCTCAGG</span> <span class="operator">-</span> <span class="name">ZZZ</span>
</pre>
<p>Using <tt class="docutils literal">Bio.AlignIO.read()</tt> or <tt class="docutils literal">Bio.AlignIO.parse()</tt> without the
<tt class="docutils literal">seq_count</tt> argument would give a single alignment containing all six
records for the first two examples. For the third example, an exception
would be raised because the lengths differ preventing them being turned
into a single alignment.</p>
<p>If the file format itself has a block structure allowing <tt class="docutils literal">Bio.AlignIO</tt>
to determine the number of sequences in each alignment directly, then
the <tt class="docutils literal">seq_count</tt> argument is not needed. If it is supplied, and doesn’t
agree with the file contents, an error is raised.</p>
<p>Note that this optional <tt class="docutils literal">seq_count</tt> argument assumes each alignment in
the file has the same number of sequences. Hypothetically you may come
across stranger situations, for example a FASTA file containing several
alignments each with a different number of sequences – although I would
love to hear of a real world example of this. Assuming you cannot get
the data in a nicer file format, there is no straight forward way to
deal with this using <tt class="docutils literal">Bio.AlignIO</tt>. In this case, you could consider
reading in the sequences themselves using <tt class="docutils literal">Bio.SeqIO</tt> and batching
them together to create the alignments as appropriate.</p>
</div>
</div>
<div class="section" id="writing-alignments">
<h2>6.2  Writing Alignments</h2>
<p>We’ve talked about using <tt class="docutils literal">Bio.AlignIO.read()</tt> and
<tt class="docutils literal">Bio.AlignIO.parse()</tt> for alignment input (reading files), and now
we’ll look at <tt class="docutils literal">Bio.AlignIO.write()</tt> which is for alignment output
(writing files). This is a function taking three arguments: some
<tt class="docutils literal">MultipleSeqAlignment</tt> objects (or for backwards compatibility the
obsolete <tt class="docutils literal">Alignment</tt> objects), a handle or filename to write to, and a
sequence format.</p>
<p>Here is an example, where we start by creating a few
<tt class="docutils literal">MultipleSeqAlignment</tt> objects the hard way (by hand, rather than by
loading them from a file). Note we create some <tt class="docutils literal">SeqRecord</tt> objects to
construct the alignment from.</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_dna</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Align</span> <span class="keyword namespace">import</span> <span class="name">MultipleSeqAlignment</span>

<span class="name">align1</span> <span class="operator">=</span> <span class="name">MultipleSeqAlignment</span><span class="punctuation">([</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACTGCTAGCTAG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Alpha&quot;</span><span class="punctuation">),</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACT-CTAGCTAG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Beta&quot;</span><span class="punctuation">),</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACTGCTAGDTAG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Gamma&quot;</span><span class="punctuation">),</span>
         <span class="punctuation">])</span>

<span class="name">align2</span> <span class="operator">=</span> <span class="name">MultipleSeqAlignment</span><span class="punctuation">([</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GTCAGC-AG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Delta&quot;</span><span class="punctuation">),</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GACAGCTAG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Epsilon&quot;</span><span class="punctuation">),</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GTCAGCTAG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Zeta&quot;</span><span class="punctuation">),</span>
         <span class="punctuation">])</span>

<span class="name">align3</span> <span class="operator">=</span> <span class="name">MultipleSeqAlignment</span><span class="punctuation">([</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACTAGTACAGCTG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Eta&quot;</span><span class="punctuation">),</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACTAGTACAGCT-&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Theta&quot;</span><span class="punctuation">),</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;-CTACTACAGGTG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Iota&quot;</span><span class="punctuation">),</span>
         <span class="punctuation">])</span>

<span class="name">my_alignments</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">align1</span><span class="punctuation">,</span> <span class="name">align2</span><span class="punctuation">,</span> <span class="name">align3</span><span class="punctuation">]</span>
</pre>
<p>Now we have a list of <tt class="docutils literal">Alignment</tt> objects, we’ll write them to a
PHYLIP format file:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">AlignIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">my_alignments</span><span class="punctuation">,</span> <span class="literal string">&quot;my_example.phy&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phylip&quot;</span><span class="punctuation">)</span>
</pre>
<p>And if you open this file in your favourite text editor it should look
like this:</p>
<pre class="code python literal-block">
 <span class="literal number integer">3</span> <span class="literal number integer">12</span>
<span class="name">Alpha</span>      <span class="name">ACTGCTAGCT</span> <span class="name">AG</span>
<span class="name">Beta</span>       <span class="name">ACT</span><span class="operator">-</span><span class="name">CTAGCT</span> <span class="name">AG</span>
<span class="name">Gamma</span>      <span class="name">ACTGCTAGDT</span> <span class="name">AG</span>
 <span class="literal number integer">3</span> <span class="literal number integer">9</span>
<span class="name">Delta</span>      <span class="name">GTCAGC</span><span class="operator">-</span><span class="name">AG</span>
<span class="name">Epislon</span>    <span class="name">GACAGCTAG</span>
<span class="name">Zeta</span>       <span class="name">GTCAGCTAG</span>
 <span class="literal number integer">3</span> <span class="literal number integer">13</span>
<span class="name">Eta</span>        <span class="name">ACTAGTACAG</span> <span class="name">CTG</span>
<span class="name">Theta</span>      <span class="name">ACTAGTACAG</span> <span class="name">CT</span><span class="operator">-</span>
<span class="name">Iota</span>       <span class="operator">-</span><span class="name">CTACTACAG</span> <span class="name">GTG</span>
</pre>
<p>Its more common to want to load an existing alignment, and save that,
perhaps after some simple manipulation like removing certain rows or
columns.</p>
<p>Suppose you wanted to know how many alignments the
<tt class="docutils literal">Bio.AlignIO.write()</tt> function wrote to the handle? If your alignments
were in a list like the example above, you could just use
<tt class="docutils literal">len(my_alignments)</tt>, however you can’t do that when your records come
from a generator/iterator. Therefore the <tt class="docutils literal">Bio.AlignIO.write()</tt>
function returns the number of alignments written to the file.</p>
<p><em>Note</em> - If you tell the <tt class="docutils literal">Bio.AlignIO.write()</tt> function to write to a
file that already exists, the old file will be overwritten without any
warning.</p>
<div class="section" id="converting-between-sequence-alignment-file-formats">
<h3>6.2.1  Converting between sequence alignment file formats</h3>
<p>Converting between sequence alignment file formats with <tt class="docutils literal">Bio.AlignIO</tt>
works in the same way as converting between sequence file formats with
<tt class="docutils literal">Bio.SeqIO</tt> (Section <a class="reference external" href="#sec:SeqIO-conversion">5.5.2</a>). We load
generally the alignment(s) using <tt class="docutils literal">Bio.AlignIO.parse()</tt> and then save
them using the <tt class="docutils literal">Bio.AlignIO.write()</tt> – or just use the
<tt class="docutils literal">Bio.AlignIO.convert()</tt> helper function.</p>
<p>For this example, we’ll load the PFAM/Stockholm format file used earlier
and save it as a Clustal W format file:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PF05371_seed.aln&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Converted </span><span class="literal string interpol">%i</span><span class="literal string"> alignments&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>Or, using <tt class="docutils literal">Bio.AlignIO.parse()</tt> and <tt class="docutils literal">Bio.AlignIO.write()</tt>:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignments</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">alignments</span><span class="punctuation">,</span> <span class="literal string">&quot;PF05371_seed.aln&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Converted </span><span class="literal string interpol">%i</span><span class="literal string"> alignments&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>The <tt class="docutils literal">Bio.AlignIO.write()</tt> function expects to be given multiple
alignment objects. In the example above we gave it the alignment
iterator returned by <tt class="docutils literal">Bio.AlignIO.parse()</tt>.</p>
<p>In this case, we know there is only one alignment in the file so we
could have used <tt class="docutils literal">Bio.AlignIO.read()</tt> instead, but notice we have to
pass this alignment to <tt class="docutils literal">Bio.AlignIO.write()</tt> as a single element list:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="name">AlignIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">([</span><span class="name">alignment</span><span class="punctuation">],</span> <span class="literal string">&quot;PF05371_seed.aln&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
</pre>
<p>Either way, you should end up with the same new Clustal W format file
“PF05371_seed.aln” with the following content:</p>
<pre class="code python literal-block">
<span class="name">CLUSTAL</span> <span class="name">X</span> <span class="punctuation">(</span><span class="literal number float">1.81</span><span class="punctuation">)</span> <span class="name">multiple</span> <span class="name">sequence</span> <span class="name">alignment</span>


<span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>                   <span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSS</span>
<span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>                   <span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVS</span>
<span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>                   <span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSS</span>
<span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>                   <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTS</span>
<span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>                    <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFAS</span>
<span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>                    <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTS</span>
<span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>                   <span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVS</span>

<span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>                   <span class="name">KA</span>
<span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>                   <span class="name">RA</span>
<span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>                   <span class="name">KA</span>
<span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>                   <span class="name">KA</span>
<span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>                    <span class="name">KA</span>
<span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>                    <span class="name">KA</span>
<span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>                   <span class="name">RA</span>
</pre>
<p>Alternatively, you could make a PHYLIP format file which we’ll name
“PF05371_seed.phy”:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">AlignIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PF05371_seed.phy&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phylip&quot;</span><span class="punctuation">)</span>
</pre>
<p>This time the output looks like this:</p>
<pre class="code python literal-block">
 <span class="literal number integer">7</span> <span class="literal number integer">52</span>
<span class="name">COATB_BPIK</span> <span class="name">AEPNAATNYA</span> <span class="name">TEAMDSLKTQ</span> <span class="name">AIDLISQTWP</span> <span class="name">VVTTVVVAGL</span> <span class="name">VIRLFKKFSS</span>
<span class="name">Q9T0Q8_BPI</span> <span class="name">AEPNAATNYA</span> <span class="name">TEAMDSLKTQ</span> <span class="name">AIDLISQTWP</span> <span class="name">VVTTVVVAGL</span> <span class="name">VIKLFKKFVS</span>
<span class="name">COATB_BPI2</span> <span class="name">DGTSTATSYA</span> <span class="name">TEAMNSLKTQ</span> <span class="name">ATDLIDQTWP</span> <span class="name">VVTSVAVAGL</span> <span class="name">AIRLFKKFSS</span>
<span class="name">COATB_BPM1</span> <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">A</span> <span class="name">KAAFNSLQAS</span> <span class="name">ATEYIGYAWA</span> <span class="name">MVVVIVGATI</span> <span class="name">GIKLFKKFTS</span>
<span class="name">COATB_BPZJ</span> <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">A</span> <span class="name">KAAFDSLQAS</span> <span class="name">ATEYIGYAWA</span> <span class="name">MVVVIVGATI</span> <span class="name">GIKLFKKFAS</span>
<span class="name">Q9T0Q9_BPF</span> <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">A</span> <span class="name">KAAFDSLQAS</span> <span class="name">ATEYIGYAWA</span> <span class="name">MVVVIVGATI</span> <span class="name">GIKLFKKFTS</span>
<span class="name">COATB_BPIF</span> <span class="name">FAADDATSQA</span> <span class="name">KAAFDSLTAQ</span> <span class="name">ATEMSGYAWA</span> <span class="name">LVVLVVGATV</span> <span class="name">GIKLFKKFVS</span>

           <span class="name">KA</span>
           <span class="name">RA</span>
           <span class="name">KA</span>
           <span class="name">KA</span>
           <span class="name">KA</span>
           <span class="name">KA</span>
           <span class="name">RA</span>
</pre>
<p>One of the big handicaps of the PHYLIP alignment file format is that the
sequence identifiers are strictly truncated at ten characters. In this
example, as you can see the resulting names are still unique - but they
are not very readable. In this particular case, there is no clear way to
compress the identifiers, but for the sake of argument you may want to
assign your own names or numbering system. This following bit of code
manipulates the record identifiers before saving the output:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="name">name_mapping</span> <span class="operator">=</span> <span class="punctuation">{}</span>
<span class="keyword">for</span> <span class="name">i</span><span class="punctuation">,</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name builtin">enumerate</span><span class="punctuation">(</span><span class="name">alignment</span><span class="punctuation">):</span>
    <span class="name">name_mapping</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span>
    <span class="name">record</span><span class="operator">.</span><span class="name">id</span> <span class="operator">=</span> <span class="literal string">&quot;seq</span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">i</span>
<span class="keyword">print</span> <span class="name">name_mapping</span>

<span class="name">AlignIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">([</span><span class="name">alignment</span><span class="punctuation">],</span> <span class="literal string">&quot;PF05371_seed.phy&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phylip&quot;</span><span class="punctuation">)</span>
</pre>
<p>This code used a Python dictionary to record a simple mapping from the
new sequence system to the original identifier:</p>
<pre class="code python literal-block">
<span class="punctuation">{</span><span class="literal number integer">0</span><span class="punctuation">:</span> <span class="literal string">'COATB_BPIKE/30-81'</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">:</span> <span class="literal string">'Q9T0Q8_BPIKE/1-52'</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">:</span> <span class="literal string">'COATB_BPI22/32-83'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">}</span>
</pre>
<p>Here is the new PHYLIP format output:</p>
<pre class="code python literal-block">
 <span class="literal number integer">7</span> <span class="literal number integer">52</span>
<span class="name">seq0</span>       <span class="name">AEPNAATNYA</span> <span class="name">TEAMDSLKTQ</span> <span class="name">AIDLISQTWP</span> <span class="name">VVTTVVVAGL</span> <span class="name">VIRLFKKFSS</span>
<span class="name">seq1</span>       <span class="name">AEPNAATNYA</span> <span class="name">TEAMDSLKTQ</span> <span class="name">AIDLISQTWP</span> <span class="name">VVTTVVVAGL</span> <span class="name">VIKLFKKFVS</span>
<span class="name">seq2</span>       <span class="name">DGTSTATSYA</span> <span class="name">TEAMNSLKTQ</span> <span class="name">ATDLIDQTWP</span> <span class="name">VVTSVAVAGL</span> <span class="name">AIRLFKKFSS</span>
<span class="name">seq3</span>       <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">A</span> <span class="name">KAAFNSLQAS</span> <span class="name">ATEYIGYAWA</span> <span class="name">MVVVIVGATI</span> <span class="name">GIKLFKKFTS</span>
<span class="name">seq4</span>       <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">A</span> <span class="name">KAAFDSLQAS</span> <span class="name">ATEYIGYAWA</span> <span class="name">MVVVIVGATI</span> <span class="name">GIKLFKKFAS</span>
<span class="name">seq5</span>       <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">A</span> <span class="name">KAAFDSLQAS</span> <span class="name">ATEYIGYAWA</span> <span class="name">MVVVIVGATI</span> <span class="name">GIKLFKKFTS</span>
<span class="name">seq6</span>       <span class="name">FAADDATSQA</span> <span class="name">KAAFDSLTAQ</span> <span class="name">ATEMSGYAWA</span> <span class="name">LVVLVVGATV</span> <span class="name">GIKLFKKFVS</span>

           <span class="name">KA</span>
           <span class="name">RA</span>
           <span class="name">KA</span>
           <span class="name">KA</span>
           <span class="name">KA</span>
           <span class="name">KA</span>
           <span class="name">RA</span>
</pre>
<p>In general, because of the identifier limitation, working with PHYLIP
file formats shouldn’t be your first choice. Using the PFAM/Stockholm
format on the other hand allows you to record a lot of additional
annotation too.</p>
</div>
<div class="section" id="getting-your-alignment-objects-as-formatted-strings">
<h3>6.2.2  Getting your alignment objects as formatted strings</h3>
<p>The <tt class="docutils literal">Bio.AlignIO</tt> interface is based on handles, which means if you
want to get your alignment(s) into a string in a particular file format
you need to do a little bit more work (see below). However, you will
probably prefer to take advantage of the alignment object’s <tt class="docutils literal">format()</tt>
method. This takes a single mandatory argument, a lower case string
which is supported by <tt class="docutils literal">Bio.AlignIO</tt> as an output format. For example:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
</pre>
<p>As described in Section <a class="reference external" href="#sec:SeqRecord-format">4.5</a>), the
<tt class="docutils literal">SeqRecord</tt> object has a similar method using output formats supported
by <tt class="docutils literal">Bio.SeqIO</tt>.</p>
<p>Internally the <tt class="docutils literal">format()</tt> method is using the <tt class="docutils literal">StringIO</tt> string
based handle and calling <tt class="docutils literal">Bio.AlignIO.write()</tt>. You can do this in
your own code if for example you are using an older version of
Biopython:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">StringIO</span> <span class="keyword namespace">import</span> <span class="name">StringIO</span>

<span class="name">alignments</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>

<span class="name">out_handle</span> <span class="operator">=</span> <span class="name">StringIO</span><span class="punctuation">()</span>
<span class="name">AlignIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">alignments</span><span class="punctuation">,</span> <span class="name">out_handle</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
<span class="name">clustal_data</span> <span class="operator">=</span> <span class="name">out_handle</span><span class="operator">.</span><span class="name">getvalue</span><span class="punctuation">()</span>

<span class="keyword">print</span> <span class="name">clustal_data</span>
</pre>
</div>
</div>
<div class="section" id="manipulating-alignments">
<h2>6.3  Manipulating Alignments</h2>
<p>Now that we’ve covered loading and saving alignments, we’ll look at what
else you can do with them.</p>
<div class="section" id="slicing-alignments">
<h3>6.3.1  Slicing alignments</h3>
<p>First of all, in some senses the alignment objects act like a Python
<tt class="docutils literal">list</tt> of <tt class="docutils literal">SeqRecord</tt> objects (the rows). With this model in mind
hopefully the actions of <tt class="docutils literal">len()</tt> (the number of rows) and iteration
(each row as a <tt class="docutils literal">SeqRecord</tt>) make sense:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;Number of rows: </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">alignment</span><span class="punctuation">)</span>
<span class="name">Number</span> <span class="name">of</span> <span class="name">rows</span><span class="punctuation">:</span> <span class="literal number integer">7</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> - </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span> <span class="operator">-</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span> <span class="operator">-</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span> <span class="operator">-</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="operator">-</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span> <span class="operator">-</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="operator">-</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span> <span class="operator">-</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>You can also use the list-like <tt class="docutils literal">append</tt> and <tt class="docutils literal">extend</tt> methods to add
more rows to the alignment (as <tt class="docutils literal">SeqRecord</tt> objects). Keeping the list
metaphor in mind, simple slicing of the alignment should also make sense
- it selects some of the rows giving back another alignment object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">52</span> <span class="name">columns</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKL</span><span class="operator">...</span><span class="name">SRA</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKL</span><span class="operator">...</span><span class="name">SRA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">:</span><span class="literal number integer">7</span><span class="punctuation">]</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">4</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">52</span> <span class="name">columns</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKL</span><span class="operator">...</span><span class="name">SRA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>What if you wanted to select by column? Those of you who have used the
NumPy matrix or array objects won’t be surprised at this - you use a
double index.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number integer">6</span><span class="punctuation">]</span>
<span class="name">T</span>
</pre>
<p>Using two integer indices pulls out a single letter, short hand for
this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">]</span>
<span class="name">T</span>
</pre>
<p>You can pull out a single column as a string like this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[:,</span><span class="literal number integer">6</span><span class="punctuation">]</span>
<span class="name">TTT</span><span class="operator">---</span><span class="name">T</span>
</pre>
<p>You can also select a range of columns. For example, to pick out those
same three rows we extracted earlier, but take just their first six
columns:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">:</span><span class="literal number integer">6</span><span class="punctuation">,:</span><span class="literal number integer">6</span><span class="punctuation">]</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">6</span> <span class="name">columns</span>
<span class="name">AEGDDP</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
</pre>
<p>Leaving the first index as <tt class="docutils literal">:</tt> means take all the rows:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[:,:</span><span class="literal number integer">6</span><span class="punctuation">]</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">6</span> <span class="name">columns</span>
<span class="name">AEPNAA</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAA</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">DGTSTA</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AEGDDP</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>This brings us to a neat way to remove a section. Notice columns 7, 8
and 9 which are gaps in three of the seven sequences:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[:,</span><span class="literal number integer">6</span><span class="punctuation">:</span><span class="literal number integer">9</span><span class="punctuation">]</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">3</span> <span class="name">columns</span>
<span class="name">TNY</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">TNY</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">TSY</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="operator">---</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="operator">---</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="operator">---</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">TSQ</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>Again, you can slice to get everything after the ninth column:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[:,</span><span class="literal number integer">9</span><span class="punctuation">:]</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">43</span> <span class="name">columns</span>
<span class="name">ATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">ATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">ATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>Now, the interesting thing is that addition of alignment objects works
by column. This lets you do this as a way to remove a block of columns:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">edited</span> <span class="operator">=</span> <span class="name">alignment</span><span class="punctuation">[:,:</span><span class="literal number integer">6</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="name">alignment</span><span class="punctuation">[:,</span><span class="literal number integer">9</span><span class="punctuation">:]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">edited</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">49</span> <span class="name">columns</span>
<span class="name">AEPNAAATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAAATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">DGTSTAATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AEGDDPAKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDPAKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDPAKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDAAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>Another common use of alignment addition would be to combine alignments
for several different genes into a meta-alignment. Watch out though -
the identifiers need to match up (see
Section <a class="reference external" href="#sec:SeqRecord-addition">4.7</a> for how adding <tt class="docutils literal">SeqRecord</tt>
objects works). You may find it helpful to first sort the alignment rows
alphabetically by id:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">edited</span><span class="operator">.</span><span class="name">sort</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">edited</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">49</span> <span class="name">columns</span>
<span class="name">DGTSTAATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">FAADDAAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
<span class="name">AEPNAAATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEGDDPAKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDPAKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEPNAAATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">AEGDDPAKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
</pre>
<p>Note that you can only add two alignments together if they have the same
number of rows.</p>
</div>
<div class="section" id="alignments-as-arrays">
<h3>6.3.2  Alignments as arrays</h3>
<p>Depending on what you are doing, it can be more useful to turn the
alignment object into an array of letters – and you can do this with
NumPy:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">numpy</span> <span class="keyword namespace">as</span> <span class="name namespace">np</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align_array</span> <span class="operator">=</span> <span class="name">np</span><span class="operator">.</span><span class="name">array</span><span class="punctuation">([</span><span class="name builtin">list</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">],</span> <span class="name">np</span><span class="operator">.</span><span class="name">character</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align_array</span><span class="operator">.</span><span class="name">shape</span>
<span class="punctuation">(</span><span class="literal number integer">7</span><span class="punctuation">,</span> <span class="literal number integer">52</span><span class="punctuation">)</span>
</pre>
<p>If you will be working heavily with the columns, you can tell NumPy to
store the array by column (as in Fortran) rather then its default of by
row (as in C):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align_array</span> <span class="operator">=</span> <span class="name">np</span><span class="operator">.</span><span class="name">array</span><span class="punctuation">([</span><span class="name builtin">list</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">],</span> <span class="name">np</span><span class="operator">.</span><span class="name">character</span><span class="punctuation">,</span> <span class="name">order</span><span class="operator">=</span><span class="literal string">&quot;F&quot;</span><span class="punctuation">)</span>
</pre>
<p>Note that this leaves the original Biopython alignment object and the
NumPy array in memory as separate objects - editing one will not update
the other!</p>
</div>
</div>
<div class="section" id="alignment-tools">
<h2>6.4  Alignment Tools</h2>
<p>There are <em>lots</em> of algorithms out there for aligning sequences, both
pairwise alignments and multiple sequence alignments. These calculations
are relatively slow, and you generally wouldn’t want to write such an
algorithm in Python. Instead, you can use Biopython to invoke a command
line tool on your behalf. Normally you would:</p>
<ol class="arabic simple">
<li>Prepare an input file of your unaligned sequences, typically this
will be a FASTA file which you might create using <tt class="docutils literal">Bio.SeqIO</tt> (see
Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>).</li>
<li>Call the command line tool to process this input file, typically via
one of Biopython’s command line wrappers (which we’ll discuss here).</li>
<li>Read the output from the tool, i.e. your aligned sequences, typically
using <tt class="docutils literal">Bio.AlignIO</tt> (see earlier in this chapter).</li>
</ol>
<p>All the command line wrappers we’re going to talk about in this chapter
follow the same style. You create a command line object specifying the
options (e.g. the input filename and the output filename), then invoke
this command line via a Python operating system call (e.g. using the
<tt class="docutils literal">subprocess</tt> module).</p>
<p>Most of these wrappers are defined in the <tt class="docutils literal">Bio.Align.Applications</tt>
module:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">Bio.Align.Applications</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">dir</span><span class="punctuation">(</span><span class="name">Bio</span><span class="operator">.</span><span class="name">Align</span><span class="operator">.</span><span class="name">Applications</span><span class="punctuation">)</span>
<span class="operator">...</span>
<span class="punctuation">[</span><span class="literal string">'ClustalwCommandline'</span><span class="punctuation">,</span> <span class="literal string">'DialignCommandline'</span><span class="punctuation">,</span> <span class="literal string">'MafftCommandline'</span><span class="punctuation">,</span> <span class="literal string">'MuscleCommandline'</span><span class="punctuation">,</span>
<span class="literal string">'PrankCommandline'</span><span class="punctuation">,</span> <span class="literal string">'ProbconsCommandline'</span><span class="punctuation">,</span> <span class="literal string">'TCoffeeCommandline'</span> <span class="operator">...</span><span class="punctuation">]</span>
</pre>
<p>(Ignore the entries starting with an underscore – these have special
meaning in Python.) The module <tt class="docutils literal">Bio.Emboss.Applications</tt> has wrappers
for some of the <a class="reference external" href="http://emboss.sourceforge.net/">EMBOSS suite</a>,
including <tt class="docutils literal">needle</tt> and <tt class="docutils literal">water</tt>, which are described below in
Section <a class="reference external" href="#seq:emboss-needle-water">6.4.5</a>, and wrappers for the
EMBOSS packaged versions of the PHYLIP tools (which EMBOSS refer to as
one of their EMBASSY packages - third party tools with an EMBOSS style
interface). We won’t explore all these alignment tools here in the
section, just a sample, but the same principles apply.</p>
<div class="section" id="clustalw">
<h3>6.4.1  ClustalW</h3>
<p>ClustalW is a popular command line tool for multiple sequence alignment
(there is also a graphical interface called ClustalX). Biopython’s
<tt class="docutils literal">Bio.Align.Applications</tt> module has a wrapper for this alignment tool
(and several others).</p>
<p>Before trying to use ClustalW from within Python, you should first try
running the ClustalW tool yourself by hand at the command line, to
familiarise yourself the other options. You’ll find the Biopython
wrapper is very faithful to the actual command line API:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">ClustalwCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">ClustalwCommandline</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>For the most basic usage, all you need is to have a FASTA input file,
such as
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/opuntia.fasta">opuntia.fasta</a>
(available online or in the Doc/examples subdirectory of the Biopython
source code). This is a small FASTA file containing seven prickly-pear
DNA sequences (from the cactus family <em>Opuntia</em>).</p>
<p>By default ClustalW will generate an alignment and guide tree file with
names based on the input FASTA file, in this case <tt class="docutils literal">opuntia.aln</tt> and
<tt class="docutils literal">opuntia.dnd</tt>, but you can override this or make it explicit:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">ClustalwCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cline</span> <span class="operator">=</span> <span class="name">ClustalwCommandline</span><span class="punctuation">(</span><span class="literal string">&quot;clustalw2&quot;</span><span class="punctuation">,</span> <span class="name">infile</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">cline</span>
<span class="name">clustalw2</span> <span class="operator">-</span><span class="name">infile</span><span class="operator">=</span><span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span>
</pre>
<p>Notice here we have given the executable name as <tt class="docutils literal">clustalw2</tt>,
indicating we have version two installed, which has a different filename
to version one (<tt class="docutils literal">clustalw</tt>, the default). Fortunately both versions
support the same set of arguments at the command line (and indeed,
should be functionally identical).</p>
<p>You may find that even though you have ClustalW installed, the above
command doesn’t work – you may get a message about “command not found”
(especially on Windows). This indicated that the ClustalW executable is
not on your PATH (an environment variable, a list of directories to be
searched). You can either update your PATH setting to include the
location of your copy of ClustalW tools (how you do this will depend on
your OS), or simply type in the full path of the tool. For example:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">os</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">ClustalwCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clustalw_exe</span> <span class="operator">=</span> <span class="literal string">r&quot;C:\Program Files\new clustal\clustalw2.exe&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clustalw_cline</span> <span class="operator">=</span> <span class="name">ClustalwCommandline</span><span class="punctuation">(</span><span class="name">clustalw_exe</span><span class="punctuation">,</span> <span class="name">infile</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">)</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">assert</span> <span class="name">os</span><span class="operator">.</span><span class="name">path</span><span class="operator">.</span><span class="name">isfile</span><span class="punctuation">(</span><span class="name">clustalw_exe</span><span class="punctuation">),</span> <span class="literal string">&quot;Clustal W executable missing&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span> <span class="name">stderr</span> <span class="operator">=</span> <span class="name">clustalw_cline</span><span class="punctuation">()</span>
</pre>
<p>Remember, in Python strings <tt class="docutils literal">\n</tt> and <tt class="docutils literal">\t</tt> are by default interpreted
as a new line and a tab – which is why we’re put a letter “r” at the
start for a raw string that isn’t translated in this way. This is
generally good practice when specifying a Windows style file name.</p>
<p>Internally this uses the <tt class="docutils literal">subprocess</tt> module which is now the
recommended way to run another program in Python. This replaces older
options like the <tt class="docutils literal">os.system()</tt> and the <tt class="docutils literal">os.popen*</tt> functions.</p>
<p>Now, at this point it helps to know about how command line tools “work”.
When you run a tool at the command line, it will often print text output
directly to screen. This text can be captured or redirected, via two
“pipes”, called standard output (the normal results) and standard error
(for error messages and debug messages). There is also standard input,
which is any text fed into the tool. These names get shortened to stdin,
stdout and stderr. When the tool finishes, it has a return code (an
integer), which by convention is zero for success.</p>
<p>When you run the command line tool like this via the Biopython wrapper,
it will wait for it to finish, and check the return code. If this is non
zero (indicating an error), an exception is raised. The wrapper then
returns two strings, stdout and stderr.</p>
<p>In the case of ClustalW, when run at the command line all the important
output is written directly to the output files. Everything normally
printed to screen while you wait (via stdout or stderr) is boring and
can be ignored (assuming it worked).</p>
<p>What we care about are the two output files, the alignment and the guide
tree. We didn’t tell ClustalW what filenames to use, but it defaults to
picking names based on the input file. In this case the output should be
in the file <tt class="docutils literal">opuntia.aln</tt>. You should be able to work out how to read
in the alignment using <tt class="docutils literal">Bio.AlignIO</tt> by now:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;opuntia.aln&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">align</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">906</span> <span class="name">columns</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273285</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191659</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273284</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191658</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273287</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191661</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
<span class="name">TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273286</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191660</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273290</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273289</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273291</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
</pre>
<p>In case you are interested (and this is an aside from the main thrust of
this chapter), the <tt class="docutils literal">opuntia.dnd</tt> file ClustalW creates is just a
standard Newick tree file, and <tt class="docutils literal">Bio.Phylo</tt> can parse these:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Phylo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;opuntia.dnd&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;newick&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_ascii</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
                             <span class="name">_______________</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273291</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191665</span>
  <span class="name">__________________________</span><span class="operator">|</span>
 <span class="operator">|</span>                          <span class="operator">|</span>   <span class="name">______</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273290</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191664</span>
 <span class="operator">|</span>                          <span class="operator">|</span><span class="name">__</span><span class="operator">|</span>
 <span class="operator">|</span>                             <span class="operator">|</span><span class="name">_____</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273289</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191663</span>
 <span class="operator">|</span>
<span class="name">_</span><span class="operator">|</span><span class="name">_________________</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273287</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191661</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191661</span>
 <span class="operator">|</span>
 <span class="operator">|</span><span class="name">__________</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273286</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191660</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191660</span>
 <span class="operator">|</span>
 <span class="operator">|</span>    <span class="name">__</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273285</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191659</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191659</span>
 <span class="operator">|</span><span class="name">___</span><span class="operator">|</span>
     <span class="operator">|</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273284</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191658</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191658</span>
</pre>
<p>Chapter <a class="reference external" href="#sec:Phylo">13</a> covers Biopython’s support for phylogenetic
trees in more depth.</p>
</div>
<div class="section" id="muscle">
<h3>6.4.2  MUSCLE</h3>
<p>MUSCLE is a more recent multiple sequence alignment tool than ClustalW,
and Biopython also has a wrapper for it under the
<tt class="docutils literal">Bio.Align.Applications</tt> module. As before, we recommend you try using
MUSCLE from the command line before trying it from within Python, as the
Biopython wrapper is very faithful to the actual command line API:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">MuscleCommandline</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>For the most basic usage, all you need is to have a FASTA input file,
such as
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/opuntia.fasta">opuntia.fasta</a>
(available online or in the Doc/examples subdirectory of the Biopython
source code). You can then tell MUSCLE to read in this FASTA file, and
write the alignment to an output file:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">,</span> <span class="name">out</span><span class="operator">=</span><span class="literal string">&quot;opuntia.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">cline</span>
<span class="name">muscle</span> <span class="operator">-</span><span class="operator word">in</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span> <span class="operator">-</span><span class="name">out</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">txt</span>
</pre>
<p>Note that MUSCLE uses “-in” and “-out” but in Biopython we have to use
“input” and “out” as the keyword arguments or property names. This is
because “in” is a reserved word in Python.</p>
<p>By default MUSCLE will output the alignment as a FASTA file (using
gapped sequences). The <tt class="docutils literal">Bio.AlignIO</tt> module should be able to read
this alignment using <tt class="docutils literal"><span class="pre">format=&quot;fasta&quot;</span></tt>. You can also ask for
ClustalW-like output:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">,</span> <span class="name">out</span><span class="operator">=</span><span class="literal string">&quot;opuntia.aln&quot;</span><span class="punctuation">,</span> <span class="name">clw</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">cline</span>
<span class="name">muscle</span> <span class="operator">-</span><span class="operator word">in</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span> <span class="operator">-</span><span class="name">out</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">aln</span> <span class="operator">-</span><span class="name">clw</span>
</pre>
<p>Or, strict ClustalW output where the original ClustalW header line is
used for maximum compatibility:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">,</span> <span class="name">out</span><span class="operator">=</span><span class="literal string">&quot;opuntia.aln&quot;</span><span class="punctuation">,</span> <span class="name">clwstrict</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">cline</span>
<span class="name">muscle</span> <span class="operator">-</span><span class="operator word">in</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span> <span class="operator">-</span><span class="name">out</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">aln</span> <span class="operator">-</span><span class="name">clwstrict</span>
</pre>
<p>The <tt class="docutils literal">Bio.AlignIO</tt> module should be able to read these alignments using
<tt class="docutils literal"><span class="pre">format=&quot;clustal&quot;</span></tt>.</p>
<p>MUSCLE can also output in GCG MSF format (using the <tt class="docutils literal">msf</tt> argument),
but Biopython can’t currently parse that, or using HTML which would give
a human readable web page (not suitable for parsing).</p>
<p>You can also set the other optional parameters, for example the maximum
number of iterations. See the built in help for details.</p>
<p>You would then run MUSCLE command line string as described above for
ClustalW, and parse the output using <tt class="docutils literal">Bio.AlignIO</tt> to get an alignment
object.</p>
</div>
<div class="section" id="muscle-using-stdout">
<h3>6.4.3  MUSCLE using stdout</h3>
<p>Using a MUSCLE command line as in the examples above will write the
alignment to a file. This means there will be no important information
written to the standard out (stdout) or standard error (stderr) handles.
However, by default MUSCLE will write the alignment to standard output
(stdout). We can take advantage of this to avoid having a temporary
output file! For example:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">muscle_cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">muscle_cline</span>
<span class="name">muscle</span> <span class="operator">-</span><span class="operator word">in</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span>
</pre>
<p>If we run this via the wrapper, we get back the output as a string. In
order to parse this we can use <tt class="docutils literal">StringIO</tt> to turn it into a handle.
Remember that MUSCLE defaults to using FASTA as the output format:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">muscle_cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span> <span class="name">stderr</span> <span class="operator">=</span> <span class="name">muscle_cline</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">StringIO</span> <span class="keyword namespace">import</span> <span class="name">StringIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">StringIO</span><span class="punctuation">(</span><span class="name">stdout</span><span class="punctuation">),</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">align</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">906</span> <span class="name">columns</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273289</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191663</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273291</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191665</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273290</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191664</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273287</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191661</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191661</span>
<span class="name">TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273286</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191660</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191660</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273285</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191659</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191659</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273284</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191658</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191658</span>
</pre>
<p>The above approach is fairly simple, but if you are dealing with very
large output text the fact that all of stdout and stderr is loaded into
memory as a string can be a potential drawback. Using the <tt class="docutils literal">subprocess</tt>
module we can work directly with handles instead:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">subprocess</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">muscle_cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">child</span> <span class="operator">=</span> <span class="name">subprocess</span><span class="operator">.</span><span class="name">Popen</span><span class="punctuation">(</span><span class="name builtin">str</span><span class="punctuation">(</span><span class="name">muscle_cline</span><span class="punctuation">),</span>
<span class="operator">...</span>                          <span class="name">stdout</span><span class="operator">=</span><span class="name">subprocess</span><span class="operator">.</span><span class="name">PIPE</span><span class="punctuation">,</span>
<span class="operator">...</span>                          <span class="name">stderr</span><span class="operator">=</span><span class="name">subprocess</span><span class="operator">.</span><span class="name">PIPE</span><span class="punctuation">,</span>
<span class="operator">...</span>                          <span class="name">shell</span><span class="operator">=</span><span class="punctuation">(</span><span class="name">sys</span><span class="operator">.</span><span class="name">platform</span><span class="operator">!=</span><span class="literal string">&quot;win32&quot;</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">child</span><span class="operator">.</span><span class="name">stdout</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">align</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">906</span> <span class="name">columns</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273289</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191663</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273291</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191665</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273290</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191664</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273287</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191661</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191661</span>
<span class="name">TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273286</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191660</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191660</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273285</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191659</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191659</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273284</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191658</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191658</span>
</pre>
</div>
<div class="section" id="muscle-using-stdin-and-stdout">
<h3>6.4.4  MUSCLE using stdin and stdout</h3>
<p>We don’t actually <em>need</em> to have our FASTA input sequences prepared in a
file, because by default MUSCLE will read in the input sequence from
standard input! Note this is a bit more advanced and fiddly, so don’t
bother with this technique unless you need to.</p>
<p>First, we’ll need some unaligned sequences in memory as <tt class="docutils literal">SeqRecord</tt>
objects. For this demonstration I’m going to use a filtered version of
the original FASTA file (using a generator expression), taking just six
of the seven sequences:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">r</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="operator">&lt;</span> <span class="literal number integer">900</span><span class="punctuation">)</span>
</pre>
<p>Then we create the MUSCLE command line, leaving the input and output to
their defaults (stdin and stdout). I’m also going to ask for strict
ClustalW format as for the output.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">muscle_cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name">clwstrict</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">muscle_cline</span>
<span class="name">muscle</span> <span class="operator">-</span><span class="name">clwstrict</span>
</pre>
<p>Now for the fiddly bits using the <tt class="docutils literal">subprocess</tt> module, stdin and
stdout:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">subprocess</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">sys</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">child</span> <span class="operator">=</span> <span class="name">subprocess</span><span class="operator">.</span><span class="name">Popen</span><span class="punctuation">(</span><span class="name builtin">str</span><span class="punctuation">(</span><span class="name">cline</span><span class="punctuation">),</span>
<span class="operator">...</span>                          <span class="name">stdin</span><span class="operator">=</span><span class="name">subprocess</span><span class="operator">.</span><span class="name">PIPE</span><span class="punctuation">,</span>
<span class="operator">...</span>                          <span class="name">stdout</span><span class="operator">=</span><span class="name">subprocess</span><span class="operator">.</span><span class="name">PIPE</span><span class="punctuation">,</span>
<span class="operator">...</span>                          <span class="name">stderr</span><span class="operator">=</span><span class="name">subprocess</span><span class="operator">.</span><span class="name">PIPE</span><span class="punctuation">,</span>
<span class="operator">...</span>                          <span class="name">shell</span><span class="operator">=</span><span class="punctuation">(</span><span class="name">sys</span><span class="operator">.</span><span class="name">platform</span><span class="operator">!=</span><span class="literal string">&quot;win32&quot;</span><span class="punctuation">))</span>
</pre>
<p>That should start MUSCLE, but it will be sitting waiting for its FASTA
input sequences, which we must supply via its stdin handle:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">child</span><span class="operator">.</span><span class="name">stdin</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">6</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">child</span><span class="operator">.</span><span class="name">stdin</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>After writing the six sequences to the handle, MUSCLE will still be
waiting to see if that is all the FASTA sequences or not – so we must
signal that this is all the input data by closing the handle. At that
point MUSCLE should start to run, and we can ask for the output:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">child</span><span class="operator">.</span><span class="name">stdout</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">align</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">6</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">900</span> <span class="name">columns</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273290</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273289</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273287</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191661</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273286</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191660</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273285</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191659</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19165</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273284</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191658</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19165</span>
</pre>
<p>Wow! There we are with a new alignment of just the six records, without
having created a temporary FASTA input file, or a temporary alignment
output file. However, a word of caution: Dealing with errors with this
style of calling external programs is much more complicated. It also
becomes far harder to diagnose problems, because you can’t try running
MUSCLE manually outside of Biopython (because you don’t have the input
file to supply). There can also be subtle cross platform issues (e.g.
Windows versus Linux), and how you run your script can have an impact
(e.g. at the command line, from IDLE or an IDE, or as a GUI script).
These are all generic Python issues though, and not specific to
Biopython.</p>
<p>If you find working directly with <tt class="docutils literal">subprocess</tt> like this scary, there
is an alternative. If you execute the tool with <tt class="docutils literal">muscle_cline()</tt> you
can supply any standard input as a big string,
<tt class="docutils literal"><span class="pre">muscle_cline(stdin=...)</span></tt>. So, provided your data isn’t very big, you
can prepare the FASTA input in memory as a string using <tt class="docutils literal">StringIO</tt>
(see Section <a class="reference external" href="#sec:appendix-handles">22.1</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">r</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="operator">&lt;</span> <span class="literal number integer">900</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">StringIO</span> <span class="keyword namespace">import</span> <span class="name">StringIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">StringIO</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">6</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">data</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">getvalue</span><span class="punctuation">()</span>
</pre>
<p>You can then run the tool and parse the alignment as follows:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span> <span class="name">stderr</span> <span class="operator">=</span> <span class="name">muscle_cline</span><span class="punctuation">(</span><span class="name">stdin</span><span class="operator">=</span><span class="name">data</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">StringIO</span><span class="punctuation">(</span><span class="name">stdout</span><span class="punctuation">),</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">align</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">6</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">900</span> <span class="name">columns</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273290</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273289</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273287</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191661</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273286</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191660</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273285</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191659</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19165</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273284</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191658</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19165</span>
</pre>
<p>You might find this easier, but it does require more memory (RAM) for
the strings used for the input FASTA and output Clustal formatted data.</p>
</div>
<div class="section" id="emboss-needle-and-water">
<h3>6.4.5  EMBOSS needle and water</h3>
<p>The <a class="reference external" href="http://emboss.sourceforge.net/">EMBOSS</a> suite includes the
<tt class="docutils literal">water</tt> and <tt class="docutils literal">needle</tt> tools for Smith-Waterman algorithm local
alignment, and Needleman-Wunsch global alignment. The tools share the
same style interface, so switching between the two is trivial – we’ll
just use <tt class="docutils literal">needle</tt> here.</p>
<p>Suppose you want to do a global pairwise alignment between two
sequences, prepared in FASTA format as follows:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">HBA_HUMAN</span>
<span class="name">MVLSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHFDLSHGSAQVKGHG</span>
<span class="name">KKVADALTNAVAHVDDMPNALSALSDLHAHKLRVDPVNFKLLSHCLLVTLAAHLPAEFTP</span>
<span class="name">AVHASLDKFLASVSTVLTSKYR</span>
</pre>
<p>in a file <tt class="docutils literal">alpha.fasta</tt>, and secondly in a file <tt class="docutils literal">beta.fasta</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">HBB_HUMAN</span>
<span class="name">MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGNPK</span>
<span class="name">VKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHHFG</span>
<span class="name">KEFTPPVQAAYQKVVAGVANALAHKYH</span>
</pre>
<p>Let’s start by creating a complete <tt class="docutils literal">needle</tt> command line object in one
go:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Emboss.Applications</span> <span class="keyword namespace">import</span> <span class="name">NeedleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span> <span class="operator">=</span> <span class="name">NeedleCommandline</span><span class="punctuation">(</span><span class="name">asequence</span><span class="operator">=</span><span class="literal string">&quot;alpha.faa&quot;</span><span class="punctuation">,</span> <span class="name">bsequence</span><span class="operator">=</span><span class="literal string">&quot;beta.faa&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                  <span class="name">gapopen</span><span class="operator">=</span><span class="literal number integer">10</span><span class="punctuation">,</span> <span class="name">gapextend</span><span class="operator">=</span><span class="literal number float">0.5</span><span class="punctuation">,</span> <span class="name">outfile</span><span class="operator">=</span><span class="literal string">&quot;needle.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">needle_cline</span>
<span class="name">needle</span> <span class="operator">-</span><span class="name">outfile</span><span class="operator">=</span><span class="name">needle</span><span class="operator">.</span><span class="name">txt</span> <span class="operator">-</span><span class="name">asequence</span><span class="operator">=</span><span class="name">alpha</span><span class="operator">.</span><span class="name">faa</span> <span class="operator">-</span><span class="name">bsequence</span><span class="operator">=</span><span class="name">beta</span><span class="operator">.</span><span class="name">faa</span> <span class="operator">-</span><span class="name">gapopen</span><span class="operator">=</span><span class="literal number integer">10</span> <span class="operator">-</span><span class="name">gapextend</span><span class="operator">=</span><span class="literal number float">0.5</span>
</pre>
<p>Why not try running this by hand at the command prompt? You should see
it does a pairwise comparison and records the output in the file
<tt class="docutils literal">needle.txt</tt> (in the default EMBOSS alignment file format).</p>
<p>Even if you have EMBOSS installed, running this command may not work –
you might get a message about “command not found” (especially on
Windows). This probably means that the EMBOSS tools are not on your PATH
environment variable. You can either update your PATH setting, or simply
tell Biopython the full path to the tool, for example:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Emboss.Applications</span> <span class="keyword namespace">import</span> <span class="name">NeedleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span> <span class="operator">=</span> <span class="name">NeedleCommandline</span><span class="punctuation">(</span><span class="literal string">r&quot;C:\EMBOSS\needle.exe&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                  <span class="name">asequence</span><span class="operator">=</span><span class="literal string">&quot;alpha.faa&quot;</span><span class="punctuation">,</span> <span class="name">bsequence</span><span class="operator">=</span><span class="literal string">&quot;beta.faa&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                  <span class="name">gapopen</span><span class="operator">=</span><span class="literal number integer">10</span><span class="punctuation">,</span> <span class="name">gapextend</span><span class="operator">=</span><span class="literal number float">0.5</span><span class="punctuation">,</span> <span class="name">outfile</span><span class="operator">=</span><span class="literal string">&quot;needle.txt&quot;</span><span class="punctuation">)</span>
</pre>
<p>Remember in Python that for a default string <tt class="docutils literal">\n</tt> or <tt class="docutils literal">\t</tt> means a
new line or a tab – which is why we’re put a letter “r” at the start for
a raw string.</p>
<p>At this point it might help to try running the EMBOSS tools yourself by
hand at the command line, to familiarise yourself the other options and
compare them to the Biopython help text:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Emboss.Applications</span> <span class="keyword namespace">import</span> <span class="name">NeedleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">NeedleCommandline</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>Note that you can also specify (or change or look at) the settings like
this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Emboss.Applications</span> <span class="keyword namespace">import</span> <span class="name">NeedleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span> <span class="operator">=</span> <span class="name">NeedleCommandline</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span><span class="operator">.</span><span class="name">asequence</span><span class="operator">=</span><span class="literal string">&quot;alpha.faa&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span><span class="operator">.</span><span class="name">bsequence</span><span class="operator">=</span><span class="literal string">&quot;beta.faa&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span><span class="operator">.</span><span class="name">gapopen</span><span class="operator">=</span><span class="literal number integer">10</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span><span class="operator">.</span><span class="name">gapextend</span><span class="operator">=</span><span class="literal number float">0.5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span><span class="operator">.</span><span class="name">outfile</span><span class="operator">=</span><span class="literal string">&quot;needle.txt&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">needle_cline</span>
<span class="name">needle</span> <span class="operator">-</span><span class="name">outfile</span><span class="operator">=</span><span class="name">needle</span><span class="operator">.</span><span class="name">txt</span> <span class="operator">-</span><span class="name">asequence</span><span class="operator">=</span><span class="name">alpha</span><span class="operator">.</span><span class="name">faa</span> <span class="operator">-</span><span class="name">bsequence</span><span class="operator">=</span><span class="name">beta</span><span class="operator">.</span><span class="name">faa</span> <span class="operator">-</span><span class="name">gapopen</span><span class="operator">=</span><span class="literal number integer">10</span> <span class="operator">-</span><span class="name">gapextend</span><span class="operator">=</span><span class="literal number float">0.5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">needle_cline</span><span class="operator">.</span><span class="name">outfile</span>
<span class="name">needle</span><span class="operator">.</span><span class="name">txt</span>
</pre>
<p>Next we want to use Python to run this command for us. As explained
above, for full control, we recommend you use the built in Python
<tt class="docutils literal">subprocess</tt> module, but for simple usage the wrapper object usually
suffices:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span> <span class="name">stderr</span> <span class="operator">=</span> <span class="name">needle_cline</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">stdout</span> <span class="operator">+</span> <span class="name">stderr</span>
<span class="name">Needleman</span><span class="operator">-</span><span class="name">Wunsch</span> <span class="keyword">global</span> <span class="name">alignment</span> <span class="name">of</span> <span class="name">two</span> <span class="name">sequences</span>
</pre>
<p>Next we can load the output file with <tt class="docutils literal">Bio.AlignIO</tt> as discussed
earlier in this chapter, as the <tt class="docutils literal">emboss</tt> format:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;needle.txt&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;emboss&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">align</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">2</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">149</span> <span class="name">columns</span>
<span class="name">MV</span><span class="operator">-</span><span class="name">LSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTY</span><span class="operator">...</span><span class="name">KYR</span> <span class="name">HBA_HUMAN</span>
<span class="name">MVHLTPEEKSAVTALWGKV</span><span class="operator">--</span><span class="name">NVDEVGGEALGRLLVVYPWTQRF</span><span class="operator">...</span><span class="name">KYH</span> <span class="name">HBB_HUMAN</span>
</pre>
<p>In this example, we told EMBOSS to write the output to a file, but you
<em>can</em> tell it to write the output to stdout instead (useful if you don’t
want a temporary output file to get rid of – use <tt class="docutils literal">stdout=True</tt> rather
than the <tt class="docutils literal">outfile</tt> argument), and also to read <em>one</em> of the one of the
inputs from stdin (e.g. <tt class="docutils literal"><span class="pre">asequence=&quot;stdin&quot;</span></tt>, much like in the MUSCLE
example in the section above).</p>
<p>This has only scratched the surface of what you can do with <tt class="docutils literal">needle</tt>
and <tt class="docutils literal">water</tt>. One useful trick is that the second file can contain
multiple sequences (say five), and then EMBOSS will do five pairwise
alignments.</p>
<p>Note - Biopython includes its own pairwise alignment code in the
<tt class="docutils literal">Bio.pairwise2</tt> module (written in C for speed, but with a pure Python
fallback available too). This doesn’t work with alignment objects, so we
have not covered it within this chapter. See the module’s docstring
(built in help) for details.</p>
</div>
</div>
</div>
<div class="section" id="chapter-7-blast">
<h1>Chapter 7  BLAST</h1>
<p>Hey, everybody loves BLAST right? I mean, geez, how can get it get any
easier to do comparisons between one of your sequences and every other
sequence in the known world? But, of course, this section isn’t about
how cool BLAST is, since we already know that. It is about the problem
with BLAST – it can be really difficult to deal with the volume of data
generated by large runs, and to automate BLAST runs in general.</p>
<p>Fortunately, the Biopython folks know this only too well, so they’ve
developed lots of tools for dealing with BLAST and making things much
easier. This section details how to use these tools and do useful things
with them.</p>
<p>Dealing with BLAST can be split up into two steps, both of which can be
done from within Biopython. Firstly, running BLAST for your query
sequence(s), and getting some output. Secondly, parsing the BLAST output
in Python for further analysis.</p>
<p>Your first introduction to running BLAST was probably via the NCBI
web-service. In fact, there are lots of ways you can run BLAST, which
can be categorised several ways. The most important distinction is
running BLAST locally (on your own machine), and running BLAST remotely
(on another machine, typically the NCBI servers). We’re going to start
this chapter by invoking the NCBI online BLAST service from within a
Python script.</p>
<p><em>NOTE</em>: The following Chapter <a class="reference external" href="#chapter:searchio">8</a> describes
<tt class="docutils literal">Bio.SearchIO</tt>, an <em>experimental</em> module in Biopython. We intend this
to ultimately replace the older <tt class="docutils literal">Bio.Blast</tt> module, as it provides a
more general framework handling other related sequence searching tools
as well. However, until that is declared stable, for production code
please continue to use the <tt class="docutils literal">Bio.Blast</tt> module for dealing with NCBI
BLAST.</p>
<div class="section" id="running-blast-over-the-internet">
<h2>7.1  Running BLAST over the Internet</h2>
<p>We use the function <tt class="docutils literal">qblast()</tt> in the <tt class="docutils literal">Bio.Blast.NCBIWWW</tt> module
call the online version of BLAST. This has three non-optional arguments:</p>
<ul class="simple">
<li>The first argument is the blast program to use for the search, as a
lower case string. The options and descriptions of the programs are
available at
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/BLAST/blast_program.shtml">http://www.ncbi.nlm.nih.gov/BLAST/blast_program.shtml</a>.
Currently <tt class="docutils literal">qblast</tt> only works with blastn, blastp, blastx, tblast
and tblastx.</li>
<li>The second argument specifies the databases to search against. Again,
the options for this are available on the NCBI web pages at
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/BLAST/blast_databases.shtml">http://www.ncbi.nlm.nih.gov/BLAST/blast_databases.shtml</a>.</li>
<li>The third argument is a string containing your query sequence. This
can either be the sequence itself, the sequence in fasta format, or
an identifier like a GI number.</li>
</ul>
<p>The <tt class="docutils literal">qblast</tt> function also take a number of other option arguments
which are basically analogous to the different parameters you can set on
the BLAST web page. We’ll just highlight a few of them here:</p>
<ul class="simple">
<li>The <tt class="docutils literal">qblast</tt> function can return the BLAST results in various
formats, which you can choose with the optional <tt class="docutils literal">format_type</tt>
keyword: <tt class="docutils literal">&quot;HTML&quot;</tt>, <tt class="docutils literal">&quot;Text&quot;</tt>, <tt class="docutils literal">&quot;ASN.1&quot;</tt>, or <tt class="docutils literal">&quot;XML&quot;</tt>. The
default is <tt class="docutils literal">&quot;XML&quot;</tt>, as that is the format expected by the parser,
described in section <a class="reference external" href="#sec:parsing-blast">7.3</a> below.</li>
<li>The argument <tt class="docutils literal">expect</tt> sets the expectation or e-value threshold.</li>
</ul>
<p>For more about the optional BLAST arguments, we refer you to the NCBI’s
own documentation, or that built into Biopython:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIWWW</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">NCBIWWW</span><span class="operator">.</span><span class="name">qblast</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>Note that the default settings on the NCBI BLAST website are not quite
the same as the defaults on QBLAST. If you get different results, you’ll
need to check the parameters (e.g. the expectation value threshold and
the gap values).</p>
<p>For example, if you have a nucleotide sequence you want to search
against the nucleotide database (nt) using BLASTN, and you know the GI
number of your query sequence, you can use:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIWWW</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name">NCBIWWW</span><span class="operator">.</span><span class="name">qblast</span><span class="punctuation">(</span><span class="literal string">&quot;blastn&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;nt&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;8332116&quot;</span><span class="punctuation">)</span>
</pre>
<p>Alternatively, if we have our query sequence already in a FASTA
formatted file, we just need to open the file and read in this record as
a string, and use that as the query argument:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIWWW</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">fasta_string</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta&quot;</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name">NCBIWWW</span><span class="operator">.</span><span class="name">qblast</span><span class="punctuation">(</span><span class="literal string">&quot;blastn&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;nt&quot;</span><span class="punctuation">,</span> <span class="name">fasta_string</span><span class="punctuation">)</span>
</pre>
<p>We could also have read in the FASTA file as a <tt class="docutils literal">SeqRecord</tt> and then
supplied just the sequence itself:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIWWW</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta&quot;</span><span class="punctuation">,</span> <span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name">NCBIWWW</span><span class="operator">.</span><span class="name">qblast</span><span class="punctuation">(</span><span class="literal string">&quot;blastn&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;nt&quot;</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
</pre>
<p>Supplying just the sequence means that BLAST will assign an identifier
for your sequence automatically. You might prefer to use the
<tt class="docutils literal">SeqRecord</tt> object’s format method to make a fasta string (which will
include the existing identifier):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIWWW</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta&quot;</span><span class="punctuation">,</span> <span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name">NCBIWWW</span><span class="operator">.</span><span class="name">qblast</span><span class="punctuation">(</span><span class="literal string">&quot;blastn&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;nt&quot;</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
</pre>
<p>This approach makes more sense if you have your sequence(s) in a
non-FASTA file format which you can extract using <tt class="docutils literal">Bio.SeqIO</tt> (see
Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>).</p>
<p>Whatever arguments you give the <tt class="docutils literal">qblast()</tt> function, you should get
back your results in a handle object (by default in XML format). The
next step would be to parse the XML output into Python objects
representing the search results (Section <a class="reference external" href="#sec:parsing-blast">7.3</a>),
but you might want to save a local copy of the output file first. I find
this especially useful when debugging my code that extracts info from
the BLAST results (because re-running the online search is slow and
wastes the NCBI computer time).</p>
<p>We need to be a bit careful since we can use <tt class="docutils literal">result_handle.read()</tt> to
read the BLAST output only once – calling <tt class="docutils literal">result_handle.read()</tt> again
returns an empty string.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">save_file</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;my_blast.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">save_file</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">save_file</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>After doing this, the results are in the file <tt class="docutils literal">my_blast.xml</tt> and the
original handle has had all its data extracted (so we closed it).
However, the <tt class="docutils literal">parse</tt> function of the BLAST parser (described
in <a class="reference external" href="#sec:parsing-blast">7.3</a>) takes a file-handle-like object, so
we can just open the saved file for input:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;my_blast.xml&quot;</span><span class="punctuation">)</span>
</pre>
<p>Now that we’ve got the BLAST results back into a handle again, we are
ready to do something with them, so this leads us right into the parsing
section (see Section <a class="reference external" href="#sec:parsing-blast">7.3</a> below). You may want
to jump ahead to that now ….</p>
</div>
<div class="section" id="running-blast-locally">
<h2>7.2  Running BLAST locally</h2>
<div class="section" id="introduction">
<h3>7.2.1  Introduction</h3>
<p>Running BLAST locally (as opposed to over the internet, see
Section <a class="reference external" href="#sec:running-www-blast">7.1</a>) has at least major two
advantages:</p>
<ul class="simple">
<li>Local BLAST may be faster than BLAST over the internet;</li>
<li>Local BLAST allows you to make your own database to search for
sequences against.</li>
</ul>
<p>Dealing with proprietary or unpublished sequence data can be another
reason to run BLAST locally. You may not be allowed to redistribute the
sequences, so submitting them to the NCBI as a BLAST query would not be
an option.</p>
<p>Unfortunately, there are some major drawbacks too – installing all the
bits and getting it setup right takes some effort:</p>
<ul class="simple">
<li>Local BLAST requires command line tools to be installed.</li>
<li>Local BLAST requires (large) BLAST databases to be setup (and
potentially kept up to date).</li>
</ul>
<p>To further confuse matters there are at least four different standalone
BLAST packages, and there are also other tools which can produce
imitation BLAST output files, such as BLAT.</p>
</div>
<div class="section" id="standalone-ncbi-legacy-blast">
<h3>7.2.2  Standalone NCBI “legacy” BLAST</h3>
<p><a class="reference external" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">NCBI “legacy”
BLAST</a>
included command line tools <tt class="docutils literal">blastall</tt>, <tt class="docutils literal">blastpgp</tt> and <tt class="docutils literal">rpsblast</tt>.
This was the most widely used standalone BLAST tool up until its
replacement BLAST+ was released by the NCBI.</p>
<p>The <tt class="docutils literal">Bio.Blast.Applications</tt> module has wrappers for the “legacy” NCBI
BLAST tools like <tt class="docutils literal">blastall</tt>, <tt class="docutils literal">blastpgp</tt> and <tt class="docutils literal">rpsblast</tt>, and there
are also helper functions in <tt class="docutils literal">Bio.Blast.NCBIStandalone</tt>. These are now
considered obsolete, and will be deprecated and eventually removed from
Biopython as people move over to the replacement BLAST+ suite.</p>
<p>To try and avoid confusion, we will not cover calling these old tools
from Biopython in this tutorial. Have a look at the older edition of
this tutorial included with Biopython 1.52 if you are curious (look at
the Tutorial PDF or HTML file in the Doc directory within
<tt class="docutils literal"><span class="pre">biopython-1.52.tar.gz</span></tt> or <tt class="docutils literal"><span class="pre">biopython-1.52.zip</span></tt>).</p>
</div>
<div class="section" id="standalone-ncbi-blast">
<h3>7.2.3  Standalone NCBI BLAST+</h3>
<p><a class="reference external" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">NCBI “new”
BLAST+</a>
was released in 2009. This replaces the old NCBI “legacy” BLAST package.
The <tt class="docutils literal">Bio.Blast.Applications</tt> module has wrappers for these “new” tools
like <tt class="docutils literal">blastn</tt>, <tt class="docutils literal">blastp</tt>, <tt class="docutils literal">blastx</tt>, <tt class="docutils literal">tblastn</tt>, <tt class="docutils literal">tblastx</tt> (which
all used to be handled by <tt class="docutils literal">blastall</tt>), <tt class="docutils literal">psiblast</tt> (replacing
<tt class="docutils literal">blastpgp</tt>) and <tt class="docutils literal">rpsblast</tt> and <tt class="docutils literal">rpstblastn</tt> (which replace the old
<tt class="docutils literal">rpsblast</tt>). We don’t include a wrapper for the <tt class="docutils literal">makeblastdb</tt> used
in BLAST+ to build a local BLAST database from FASTA file, nor the
equivalent tool <tt class="docutils literal">formatdb</tt> in “legacy” BLAST.</p>
<p>This section will show briefly how to use these tools from within
Python. If you have already read or tried the alignment tool examples in
Section <a class="reference external" href="#sec:alignment-tools">6.4</a> this should all seem quite
straightforward. First, we construct a command line string (as you would
type in at the command line prompt if running standalone BLAST by hand).
Then we can execute this command from within Python.</p>
<p>For example, taking a FASTA file of gene nucleotide sequences, you might
want to run a BLASTX (translation) search against the non-redundant (NR)
protein database. Assuming you (or your systems administrator) has
downloaded and installed the NR database, you might run:</p>
<pre class="code python literal-block">
<span class="name">blastx</span> <span class="operator">-</span><span class="name">query</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span> <span class="operator">-</span><span class="name">db</span> <span class="name">nr</span> <span class="operator">-</span><span class="name">out</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">xml</span> <span class="operator">-</span><span class="name">evalue</span> <span class="literal number float">0.001</span> <span class="operator">-</span><span class="name">outfmt</span> <span class="literal number integer">5</span>
</pre>
<p>This should run BLASTX against the NR database, using an expectation
cut-off value of 0.001 and produce XML output to the specified file
(which we can then parse). On my computer this takes about six minutes -
a good reason to save the output to a file so you and repeat any
analysis as needed.</p>
<p>From within Biopython we can use the NCBI BLASTX wrapper from the
<tt class="docutils literal">Bio.Blast.Applications</tt> module to build the command line string, and
run it:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast.Applications</span> <span class="keyword namespace">import</span> <span class="name">NcbiblastxCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">NcbiblastxCommandline</span><span class="punctuation">)</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blastx_cline</span> <span class="operator">=</span> <span class="name">NcbiblastxCommandline</span><span class="punctuation">(</span><span class="name">query</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">,</span> <span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nr&quot;</span><span class="punctuation">,</span> <span class="name">evalue</span><span class="operator">=</span><span class="literal number float">0.001</span><span class="punctuation">,</span>
<span class="operator">...</span>                                      <span class="name">outfmt</span><span class="operator">=</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="name">out</span><span class="operator">=</span><span class="literal string">&quot;opuntia.xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blastx_cline</span>
<span class="name">NcbiblastxCommandline</span><span class="punctuation">(</span><span class="name">cmd</span><span class="operator">=</span><span class="literal string">'blastx'</span><span class="punctuation">,</span> <span class="name">out</span><span class="operator">=</span><span class="literal string">'opuntia.xml'</span><span class="punctuation">,</span> <span class="name">outfmt</span><span class="operator">=</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="name">query</span><span class="operator">=</span><span class="literal string">'opuntia.fasta'</span><span class="punctuation">,</span>
<span class="name">db</span><span class="operator">=</span><span class="literal string">'nr'</span><span class="punctuation">,</span> <span class="name">evalue</span><span class="operator">=</span><span class="literal number float">0.001</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blastx_cline</span>
<span class="name">blastx</span> <span class="operator">-</span><span class="name">out</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">xml</span> <span class="operator">-</span><span class="name">outfmt</span> <span class="literal number integer">5</span> <span class="operator">-</span><span class="name">query</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span> <span class="operator">-</span><span class="name">db</span> <span class="name">nr</span> <span class="operator">-</span><span class="name">evalue</span> <span class="literal number float">0.001</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span> <span class="name">stderr</span> <span class="operator">=</span> <span class="name">blastx_cline</span><span class="punctuation">()</span>
</pre>
<p>In this example there shouldn’t be any output from BLASTX to the
terminal, so stdout and stderr should be empty. You may want to check
the output file <tt class="docutils literal">opuntia.xml</tt> has been created.</p>
<p>As you may recall from earlier examples in the tutorial, the
<tt class="docutils literal">opuntia.fasta</tt> contains seven sequences, so the BLAST XML output
should contain multiple results. Therefore use
<tt class="docutils literal">Bio.Blast.NCBIXML.parse()</tt> to parse it as described below in
Section <a class="reference external" href="#sec:parsing-blast">7.3</a>.</p>
</div>
<div class="section" id="wu-blast-and-ab-blast">
<h3>7.2.4  WU-BLAST and AB-BLAST</h3>
<p>You may also come across <a class="reference external" href="http://blast.wustl.edu/">Washington University
BLAST</a> (WU-BLAST), and its successor,
<a class="reference external" href="http://blast.advbiocomp.com">Advanced Biocomputing BLAST</a> (AB-BLAST,
released in 2009, not free/open source). These packages include the
command line tools <tt class="docutils literal"><span class="pre">wu-blastall</span></tt> and <tt class="docutils literal"><span class="pre">ab-blastall</span></tt>.</p>
<p>Biopython does not currently provide wrappers for calling these tools,
but should be able to parse any NCBI compatible output from them.</p>
</div>
</div>
<div class="section" id="parsing-blast-output">
<h2>7.3  Parsing BLAST output</h2>
<p>As mentioned above, BLAST can generate output in various formats, such
as XML, HTML, and plain text. Originally, Biopython had parsers for
BLAST plain text and HTML output, as these were the only output formats
offered at the time. Unfortunately, the BLAST output in these formats
kept changing, each time breaking the Biopython parsers. Our HTML BLAST
parser has been removed, but the plain text BLAST parser is still
available (see Section <a class="reference external" href="#sec:parsing-blast-deprecated">7.5</a>). Use
it at your own risk, it may or may not work, depending on which BLAST
version you’re using.</p>
<p>As keeping up with changes in BLAST became a hopeless endeavor,
especially with users running different BLAST versions, we now recommend
to parse the output in XML format, which can be generated by recent
versions of BLAST. Not only is the XML output more stable than the plain
text and HTML output, it is also much easier to parse automatically,
making Biopython a whole lot more stable.</p>
<p>You can get BLAST output in XML format in various ways. For the parser,
it doesn’t matter how the output was generated, as long as it is in the
XML format.</p>
<ul class="simple">
<li>You can use Biopython to run BLAST over the internet, as described in
section <a class="reference external" href="#sec:running-www-blast">7.1</a>.</li>
<li>You can use Biopython to run BLAST locally, as described in
section <a class="reference external" href="#sec:running-local-blast">7.2</a>.</li>
<li>You can do the BLAST search yourself on the NCBI site through your
web browser, and then save the results. You need to choose XML as the
format in which to receive the results, and save the final BLAST page
you get (you know, the one with all of the interesting results!) to a
file.</li>
<li>You can also run BLAST locally without using Biopython, and save the
output in a file. Again, you need to choose XML as the format in
which to receive the results.</li>
</ul>
<p>The important point is that you do not have to use Biopython scripts to
fetch the data in order to be able to parse it. Doing things in one of
these ways, you then need to get a handle to the results. In Python, a
handle is just a nice general way of describing input to any info source
so that the info can be retrieved using <tt class="docutils literal">read()</tt> and <tt class="docutils literal">readline()</tt>
functions (see Section sec:appendix-handles).</p>
<p>If you followed the code above for interacting with BLAST through a
script, then you already have <tt class="docutils literal">result_handle</tt>, the handle to the BLAST
results. For example, using a GI number to do an online search:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIWWW</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name">NCBIWWW</span><span class="operator">.</span><span class="name">qblast</span><span class="punctuation">(</span><span class="literal string">&quot;blastn&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;nt&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;8332116&quot;</span><span class="punctuation">)</span>
</pre>
<p>If instead you ran BLAST some other way, and have the BLAST output (in
XML format) in the file <tt class="docutils literal">my_blast.xml</tt>, all you need to do is to open
the file for reading:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;my_blast.xml&quot;</span><span class="punctuation">)</span>
</pre>
<p>Now that we’ve got a handle, we are ready to parse the output. The code
to parse it is really quite small. If you expect a single BLAST result
(i.e. you used a single query):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIXML</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">NCBIXML</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">)</span>
</pre>
<p>or, if you have lots of results (i.e. multiple query sequences):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIXML</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_records</span> <span class="operator">=</span> <span class="name">NCBIXML</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">)</span>
</pre>
<p>Just like <tt class="docutils literal">Bio.SeqIO</tt> and <tt class="docutils literal">Bio.AlignIO</tt> (see
Chapters <a class="reference external" href="#chapter:Bio.SeqIO">5</a>
and <a class="reference external" href="#chapter:Bio.AlignIO">6</a>), we have a pair of input functions,
<tt class="docutils literal">read</tt> and <tt class="docutils literal">parse</tt>, where <tt class="docutils literal">read</tt> is for when you have exactly one
object, and <tt class="docutils literal">parse</tt> is an iterator for when you can have lots of
objects – but instead of getting <tt class="docutils literal">SeqRecord</tt> or
<tt class="docutils literal">MultipleSeqAlignment</tt> objects, we get BLAST record objects.</p>
<p>To be able to handle the situation where the BLAST file may be huge,
containing thousands of results, <tt class="docutils literal">NCBIXML.parse()</tt> returns an
iterator. In plain English, an iterator allows you to step through the
BLAST output, retrieving BLAST records one by one for each BLAST search
result:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIXML</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_records</span> <span class="operator">=</span> <span class="name">NCBIXML</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="comment"># ... do something with blast_record</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="comment"># ... do something with blast_record</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="comment"># ... do something with blast_record</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
  <span class="name">File</span> <span class="literal string">&quot;&lt;stdin&gt;&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="operator">&lt;</span><span class="name">module</span><span class="operator">&gt;</span>
<span class="name exception">StopIteration</span>
<span class="comment"># No further records</span>
</pre>
<p>Or, you can use a <tt class="docutils literal">for</tt>-loop:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">blast_record</span> <span class="operator word">in</span> <span class="name">blast_records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="comment"># Do something with blast_record</span>
</pre>
<p>Note though that you can step through the BLAST records only once.
Usually, from each BLAST record you would save the information that you
are interested in. If you want to save all returned BLAST records, you
can convert the iterator into a list:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_records</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">blast_records</span><span class="punctuation">)</span>
</pre>
<p>Now you can access each BLAST record in the list with an index as usual.
If your BLAST file is huge though, you may run into memory problems
trying to save them all in a list.</p>
<p>Usually, you’ll be running one BLAST search at a time. Then, all you
need to do is to pick up the first (and only) BLAST record in
<tt class="docutils literal">blast_records</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIXML</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_records</span> <span class="operator">=</span> <span class="name">NCBIXML</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
</pre>
<p>or more elegantly:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIXML</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">NCBIXML</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">)</span>
</pre>
<p>I guess by now you’re wondering what is in a BLAST record.</p>
</div>
<div class="section" id="the-blast-record-class">
<h2>7.4  The BLAST record class</h2>
<p>A BLAST Record contains everything you might ever want to extract from
the BLAST output. Right now we’ll just show an example of how to get
some info out of the BLAST report, but if you want something in
particular that is not described here, look at the info on the record
class in detail, and take a gander into the code or automatically
generated documentation – the docstrings have lots of good info about
what is stored in each piece of information.</p>
<p>To continue with our example, let’s just print out some summary info
about all hits in our blast report greater than a particular threshold.
The following code does this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">E_VALUE_THRESH</span> <span class="operator">=</span> <span class="literal number float">0.04</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">alignment</span> <span class="operator word">in</span> <span class="name">blast_record</span><span class="operator">.</span><span class="name">alignments</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">hsp</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">hsps</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">if</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">expect</span> <span class="operator">&lt;</span> <span class="name">E_VALUE_THRESH</span><span class="punctuation">:</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'****Alignment****'</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'sequence:'</span><span class="punctuation">,</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">title</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'length:'</span><span class="punctuation">,</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">length</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'e value:'</span><span class="punctuation">,</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">expect</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">query</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">'...'</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">match</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">'...'</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">sbjct</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">'...'</span>
</pre>
<p>This will print out summary reports like the following:</p>
<pre class="code python literal-block">
<span class="operator">****</span><span class="name">Alignment</span><span class="operator">****</span>
<span class="name">sequence</span><span class="punctuation">:</span> <span class="operator">&gt;</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF283004</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF283004</span> <span class="name">Arabidopsis</span> <span class="name">thaliana</span> <span class="name">cold</span> <span class="name">acclimation</span> <span class="name">protein</span> <span class="name">WCOR413</span><span class="operator">-</span><span class="name">like</span> <span class="name">protein</span>
<span class="name">alpha</span> <span class="name">form</span> <span class="name">mRNA</span><span class="punctuation">,</span> <span class="name">complete</span> <span class="name">cds</span>
<span class="name">length</span><span class="punctuation">:</span> <span class="literal number integer">783</span>
<span class="name">e</span> <span class="name">value</span><span class="punctuation">:</span> <span class="literal number float">0.034</span>
<span class="name">tacttgttgatattggatcgaacaaactggagaaccaacatgctcacgtcacttttagtcccttacatattcctc</span><span class="operator">...</span>
<span class="operator">|||||||||</span> <span class="operator">|</span> <span class="operator">|||||||||||</span> <span class="operator">||</span> <span class="operator">||||</span>  <span class="operator">||</span> <span class="operator">||</span> <span class="operator">||||||||</span> <span class="operator">||||||</span> <span class="operator">|</span>  <span class="operator">|</span> <span class="operator">||||||||</span> <span class="operator">|||</span> <span class="operator">||...</span>
<span class="name">tacttgttggtgttggatcgaaccaattggaagacgaatatgctcacatcacttctcattccttacatcttcttc</span><span class="operator">...</span>
</pre>
<p>Basically, you can do anything you want to with the info in the BLAST
report once you have parsed it. This will, of course, depend on what you
want to use it for, but hopefully this helps you get started on doing
what you need to do!</p>
<p>An important consideration for extracting information from a BLAST
report is the type of objects that the information is stored in. In
Biopython, the parsers return <tt class="docutils literal">Record</tt> objects, either <tt class="docutils literal">Blast</tt> or
<tt class="docutils literal">PSIBlast</tt> depending on what you are parsing. These objects are
defined in <tt class="docutils literal">Bio.Blast.Record</tt> and are quite complete.</p>
<p>Here are my attempts at UML class diagrams for the <tt class="docutils literal">Blast</tt> and
<tt class="docutils literal">PSIBlast</tt> record classes. If you are good at UML and see
mistakes/improvements that can be made, please let me know. The Blast
class diagram is shown in Figure <a class="reference external" href="#fig:blastrecord">7.4</a>.</p>
<p><img alt="image1" src="../images/BlastRecord.png" /></p>
<p>The PSIBlast record object is similar, but has support for the rounds
that are used in the iteration steps of PSIBlast. The class diagram for
PSIBlast is shown in Figure <a class="reference external" href="#fig:psiblastrecord">7.4</a>.</p>
<p><img alt="image2" src="../images/PSIBlastRecord.png" /></p>
</div>
<div class="section" id="deprecated-blast-parsers">
<h2>7.5  Deprecated BLAST parsers</h2>
<p>Older versions of Biopython had parsers for BLAST output in plain text
or HTML format. Over the years, we discovered that it is very hard to
maintain these parsers in working order. Basically, any small change to
the BLAST output in newly released BLAST versions tends to cause the
plain text and HTML parsers to break. We therefore recommend parsing
BLAST output in XML format, as described in
section <a class="reference external" href="#sec:parsing-blast">7.3</a>.</p>
<p>Depending on which BLAST versions or programs you’re using, our plain
text BLAST parser may or may not work. Use it at your own risk!</p>
<div class="section" id="parsing-plain-text-blast-output">
<h3>7.5.1  Parsing plain-text BLAST output</h3>
<p>The plain text BLAST parser is located in <tt class="docutils literal">Bio.Blast.NCBIStandalone</tt>.</p>
<p>As with the XML parser, we need to have a handle object that we can pass
to the parser. The handle must implement the <tt class="docutils literal">readline()</tt> method and
do this properly. The common ways to get such a handle are to either use
the provided <tt class="docutils literal">blastall</tt> or <tt class="docutils literal">blastpgp</tt> functions to run the local
blast, or to run a local blast via the command line, and then do
something like the following:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;my_file_of_blast_output.txt&quot;</span><span class="punctuation">)</span>
</pre>
<p>Well, now that we’ve got a handle (which we’ll call <tt class="docutils literal">result_handle</tt>),
we are ready to parse it. This can be done with the following code:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIStandalone</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_parser</span> <span class="operator">=</span> <span class="name">NCBIStandalone</span><span class="operator">.</span><span class="name">BlastParser</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_parser</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">)</span>
</pre>
<p>This will parse the BLAST report into a Blast Record class (either a
Blast or a PSIBlast record, depending on what you are parsing) so that
you can extract the information from it. In our case, let’s just use
print out a quick summary of all of the alignments greater than some
threshold value.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">E_VALUE_THRESH</span> <span class="operator">=</span> <span class="literal number float">0.04</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">alignment</span> <span class="operator word">in</span> <span class="name">blast_record</span><span class="operator">.</span><span class="name">alignments</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">hsp</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">hsps</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">if</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">expect</span> <span class="operator">&lt;</span> <span class="name">E_VALUE_THRESH</span><span class="punctuation">:</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'****Alignment****'</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'sequence:'</span><span class="punctuation">,</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">title</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'length:'</span><span class="punctuation">,</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">length</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'e value:'</span><span class="punctuation">,</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">expect</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">query</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">'...'</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">match</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">'...'</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">sbjct</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">'...'</span>
</pre>
<p>If you also read the section <a class="reference external" href="#sec:parsing-blast">7.3</a> on parsing
BLAST XML output, you’ll notice that the above code is identical to what
is found in that section. Once you parse something into a record class
you can deal with it independent of the format of the original BLAST
info you were parsing. Pretty snazzy!</p>
<p>Sure, parsing one record is great, but I’ve got a BLAST file with tons
of records – how can I parse them all? Well, fear not, the answer lies
in the very next section.</p>
</div>
<div class="section" id="parsing-a-plain-text-blast-file-full-of-blast-runs">
<h3>7.5.2  Parsing a plain-text BLAST file full of BLAST runs</h3>
<p>We can do this using the blast iterator. To set up an iterator, we first
set up a parser, to parse our blast reports in Blast Record objects:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIStandalone</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_parser</span> <span class="operator">=</span> <span class="name">NCBIStandalone</span><span class="operator">.</span><span class="name">BlastParser</span><span class="punctuation">()</span>
</pre>
<p>Then we will assume we have a handle to a bunch of blast records, which
we’ll call <tt class="docutils literal">result_handle</tt>. Getting a handle is described in full
detail above in the blast parsing sections.</p>
<p>Now that we’ve got a parser and a handle, we are ready to set up the
iterator with the following command:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_iterator</span> <span class="operator">=</span> <span class="name">NCBIStandalone</span><span class="operator">.</span><span class="name">Iterator</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">,</span> <span class="name">blast_parser</span><span class="punctuation">)</span>
</pre>
<p>The second option, the parser, is optional. If we don’t supply a parser,
then the iterator will just return the raw BLAST reports one at a time.</p>
<p>Now that we’ve got an iterator, we start retrieving blast records
(generated by our parser) using <tt class="docutils literal">next()</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
</pre>
<p>Each call to next will return a new record that we can deal with. Now we
can iterate through this records and generate our old favorite, a nice
little blast report:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">blast_record</span> <span class="operator word">in</span> <span class="name">blast_iterator</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">E_VALUE_THRESH</span> <span class="operator">=</span> <span class="literal number float">0.04</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">alignment</span> <span class="operator word">in</span> <span class="name">blast_record</span><span class="operator">.</span><span class="name">alignments</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">for</span> <span class="name">hsp</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">hsps</span><span class="punctuation">:</span>
<span class="operator">...</span>             <span class="keyword">if</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">expect</span> <span class="operator">&lt;</span> <span class="name">E_VALUE_THRESH</span><span class="punctuation">:</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="literal string">'****Alignment****'</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="literal string">'sequence:'</span><span class="punctuation">,</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">title</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="literal string">'length:'</span><span class="punctuation">,</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">length</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="literal string">'e value:'</span><span class="punctuation">,</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">expect</span>
<span class="operator">...</span>                 <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">hsp</span><span class="operator">.</span><span class="name">query</span><span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="literal number integer">75</span><span class="punctuation">:</span>
<span class="operator">...</span>                     <span class="name">dots</span> <span class="operator">=</span> <span class="literal string">'...'</span>
<span class="operator">...</span>                 <span class="keyword">else</span><span class="punctuation">:</span>
<span class="operator">...</span>                     <span class="name">dots</span> <span class="operator">=</span> <span class="literal string">''</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">query</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="name">dots</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">match</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="name">dots</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">sbjct</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="name">dots</span>
</pre>
<p>The iterator allows you to deal with huge blast records without any
memory problems, since things are read in one at a time. I have parsed
tremendously huge files without any problems using this.</p>
</div>
<div class="section" id="finding-a-bad-record-somewhere-in-a-huge-plain-text-blast-file">
<h3>7.5.3  Finding a bad record somewhere in a huge plain-text BLAST file</h3>
<p>One really ugly problem that happens to me is that I’ll be parsing a
huge blast file for a while, and the parser will bomb out with a
ValueError. This is a serious problem, since you can’t tell if the
ValueError is due to a parser problem, or a problem with the BLAST. To
make it even worse, you have no idea where the parse failed, so you
can’t just ignore the error, since this could be ignoring an important
data point.</p>
<p>We used to have to make a little script to get around this problem, but
the <tt class="docutils literal">Bio.Blast</tt> module now includes a <tt class="docutils literal">BlastErrorParser</tt> which
really helps make this easier. The <tt class="docutils literal">BlastErrorParser</tt> works very
similar to the regular <tt class="docutils literal">BlastParser</tt>, but it adds an extra layer of
work by catching ValueErrors that are generated by the parser, and
attempting to diagnose the errors.</p>
<p>Let’s take a look at using this parser – first we define the file we are
going to parse and the file to write the problem reports to:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">os</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_file</span> <span class="operator">=</span> <span class="name">os</span><span class="operator">.</span><span class="name">path</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">os</span><span class="operator">.</span><span class="name">getcwd</span><span class="punctuation">(),</span> <span class="literal string">&quot;blast_out&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;big_blast.out&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">error_file</span> <span class="operator">=</span> <span class="name">os</span><span class="operator">.</span><span class="name">path</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">os</span><span class="operator">.</span><span class="name">getcwd</span><span class="punctuation">(),</span> <span class="literal string">&quot;blast_out&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;big_blast.problems&quot;</span><span class="punctuation">)</span>
</pre>
<p>Now we want to get a <tt class="docutils literal">BlastErrorParser</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIStandalone</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">error_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="name">error_file</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_error_parser</span> <span class="operator">=</span> <span class="name">NCBIStandalone</span><span class="operator">.</span><span class="name">BlastErrorParser</span><span class="punctuation">(</span><span class="name">error_handle</span><span class="punctuation">)</span>
</pre>
<p>Notice that the parser take an optional argument of a handle. If a
handle is passed, then the parser will write any blast records which
generate a ValueError to this handle. Otherwise, these records will not
be recorded.</p>
<p>Now we can use the <tt class="docutils literal">BlastErrorParser</tt> just like a regular blast
parser. Specifically, we might want to make an iterator that goes
through our blast records one at a time and parses them with the error
parser:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="name">blast_file</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">iterator</span> <span class="operator">=</span> <span class="name">NCBIStandalone</span><span class="operator">.</span><span class="name">Iterator</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">,</span> <span class="name">blast_error_parser</span><span class="punctuation">)</span>
</pre>
<p>We can read these records one a time, but now we can catch and deal with
errors that are due to problems with Blast (and not with the parser
itself):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">try</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">next_record</span> <span class="operator">=</span> <span class="name">iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="operator">...</span> <span class="keyword">except</span> <span class="name">NCBIStandalone</span><span class="operator">.</span><span class="name">LowQualityBlastError</span><span class="punctuation">,</span> <span class="name">info</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;LowQualityBlastError detected in id </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">info</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
</pre>
<p>The <tt class="docutils literal">.next()</tt> method is normally called indirectly via a <tt class="docutils literal">for</tt>-loop.
Right now the <tt class="docutils literal">BlastErrorParser</tt> can generate the following errors:</p>
<ul class="simple">
<li><tt class="docutils literal">ValueError</tt> – This is the same error generated by the regular
BlastParser, and is due to the parser not being able to parse a
specific file. This is normally either due to a bug in the parser, or
some kind of discrepancy between the version of BLAST you are using
and the versions the parser is able to handle.</li>
<li><tt class="docutils literal">LowQualityBlastError</tt> – When BLASTing a sequence that is of really
bad quality (for example, a short sequence that is basically a
stretch of one nucleotide), it seems that Blast ends up masking out
the entire sequence and ending up with nothing to parse. In this case
it will produce a truncated report that causes the parser to generate
a ValueError. <tt class="docutils literal">LowQualityBlastError</tt> is reported in these cases.
This error returns an info item with the following information:<ul>
<li><tt class="docutils literal">item[0]</tt> – The error message</li>
<li><tt class="docutils literal">item[1]</tt> – The id of the input record that caused the error.
This is really useful if you want to record all of the records
that are causing problems.</li>
</ul>
</li>
</ul>
<p>As mentioned, with each error generated, the BlastErrorParser will write
the offending record to the specified <tt class="docutils literal">error_handle</tt>. You can then go
ahead and look and these and deal with them as you see fit. Either you
will be able to debug the parser with a single blast report, or will
find out problems in your blast runs. Either way, it will definitely be
a useful experience!</p>
<p>Hopefully the <tt class="docutils literal">BlastErrorParser</tt> will make it much easier to debug and
deal with large Blast files.</p>
</div>
</div>
<div class="section" id="dealing-with-psi-blast">
<h2>7.6  Dealing with PSI-BLAST</h2>
<p>You can run the standalone version of PSI-BLAST (the legacy NCBI command
line tool <tt class="docutils literal">blastpgp</tt>, or its replacement <tt class="docutils literal">psiblast</tt>) using the
wrappers in <tt class="docutils literal">Bio.Blast.Applications</tt> module.</p>
<p>At the time of writing, the NCBI do not appear to support tools running
a PSI-BLAST search via the internet.</p>
<p>Note that the <tt class="docutils literal">Bio.Blast.NCBIXML</tt> parser can read the XML output from
current versions of PSI-BLAST, but information like which sequences in
each iteration is new or reused isn’t present in the XML file. If you
care about this information you may have more joy with the plain text
output and the <tt class="docutils literal">PSIBlastParser</tt> in <tt class="docutils literal">Bio.Blast.NCBIStandalone</tt>.</p>
</div>
<div class="section" id="dealing-with-rps-blast">
<h2>7.7  Dealing with RPS-BLAST</h2>
<p>You can run the standalone version of RPS-BLAST (either the legacy NCBI
command line tool <tt class="docutils literal">rpsblast</tt>, or its replacement with the same name)
using the wrappers in <tt class="docutils literal">Bio.Blast.Applications</tt> module.</p>
<p>At the time of writing, the NCBI do not appear to support tools running
an RPS-BLAST search via the internet.</p>
<p>You can use the <tt class="docutils literal">Bio.Blast.NCBIXML</tt> parser to read the XML output from
current versions of RPS-BLAST.</p>
</div>
</div>
<div class="section" id="chapter-8-blast-and-other-sequence-search-tools-experimental-code">
<h1>Chapter 8  BLAST and other sequence search tools (<em>experimental code</em>)</h1>
<p><em>WARNING</em>: This chapter of the Tutorial describes an <em>experimental</em>
module in Biopython. It is being included in Biopython and documented
here in the tutorial in a pre-final state to allow a period of feedback
and refinement before we declare it stable. Until then the details will
probably change, and any scripts using the current <tt class="docutils literal">Bio.SearchIO</tt>
would need to be updated. Please keep this in mind! For stable code
working with NCBI BLAST, please continue to use Bio.Blast described in
the preceding Chapter <a class="reference external" href="#chapter:blast">7</a>.</p>
<p>Biological sequence identification is an integral part of
bioinformatics. Several tools are available for this, each with their
own algorithms and approaches, such as BLAST (arguably the most
popular), FASTA, HMMER, and many more. In general, these tools usually
use your sequence to search a database of potential matches. With the
growing number of known sequences (hence the growing number of potential
matches), interpreting the results becomes increasingly hard as there
could be hundreds or even thousands of potential matches. Naturally,
manual interpretation of these searches’ results is out of the question.
Moreover, you often need to work with several sequence search tools,
each with its own statistics, conventions, and output format. Imagine
how daunting it would be when you need to work with multiple sequences
using multiple search tools.</p>
<p>We know this too well ourselves, which is why we created the
<tt class="docutils literal">Bio.SearchIO</tt> submodule in Biopython. <tt class="docutils literal">Bio.SearchIO</tt> allows you to
extract information from your search results in a convenient way, while
also dealing with the different standards and conventions used by
different search tools. The name <tt class="docutils literal">SearchIO</tt> is a homage to BioPerl’s
module of the same name.</p>
<p>In this chapter, we’ll go through the main features of <tt class="docutils literal">Bio.SearchIO</tt>
to show what it can do for you. We’ll use two popular search tools along
the way: BLAST and BLAT. They are used merely for illustrative purposes,
and you should be able to adapt the workflow to any other search tools
supported by <tt class="docutils literal">Bio.SearchIO</tt> in a breeze. You’re very welcome to follow
along with the search output files we’ll be using. The BLAST output file
can be downloaded
<a class="reference external" href="http://biopython.org/SRC/Tests/Tutorial/my_blast.xml">here</a>, and the
BLAT output file
<a class="reference external" href="http://biopython.org/SRC/Tests/Tutorial/my_blat.psl">here</a>. Both
output files were generated using this sequence:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">mystery_seq</span>
<span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG</span>
</pre>
<p>The BLAST result is an XML file generated using <tt class="docutils literal">blastn</tt> against the
NCBI <tt class="docutils literal">refseq_rna</tt> database. For BLAT, the sequence database was the
February 2009 <tt class="docutils literal">hg19</tt> human genome draft and the output format is PSL.</p>
<p>We’ll start from an introduction to the <tt class="docutils literal">Bio.SearchIO</tt> object model.
The model is the representation of your search results, thus it is core
to <tt class="docutils literal">Bio.SearchIO</tt> itself. After that, we’ll check out the main
functions in <tt class="docutils literal">Bio.SearchIO</tt> that you may often use.</p>
<p>Now that we’re all set, let’s go to the first step: introducing the core
object model.</p>
<div class="section" id="the-searchio-object-model">
<h2>8.1  The SearchIO object model</h2>
<p>Despite the wildly differing output styles among many sequence search
tools, it turns out that their underlying concept is similar:</p>
<ul class="simple">
<li>The output file may contain results from one or more search queries.</li>
<li>In each search query, you will see one or more hits from the given
search database.</li>
<li>In each database hit, you will see one or more regions containing the
actual sequence alignment between your query sequence and the
database sequence.</li>
<li>Some programs like BLAT or Exonerate may further split these regions
into several alignment fragments (or blocks in BLAT and possibly
exons in exonerate). This is not something you always see, as
programs like BLAST and HMMER do not do this.</li>
</ul>
<p>Realizing this generality, we decided use it as base for creating the
<tt class="docutils literal">Bio.SearchIO</tt> object model. The object model consists of a nested
hierarchy of Python objects, each one representing one concept outlined
above. These objects are:</p>
<ul class="simple">
<li><tt class="docutils literal">QueryResult</tt>, to represent a single search query.</li>
<li><tt class="docutils literal">Hit</tt>, to represent a single database hit. <tt class="docutils literal">Hit</tt> objects are
contained within <tt class="docutils literal">QueryResult</tt> and in each <tt class="docutils literal">QueryResult</tt> there is
zero or more <tt class="docutils literal">Hit</tt> objects.</li>
<li><tt class="docutils literal">HSP</tt> (short for high-scoring pair), to represent region(s) of
significant alignments between query and hit sequences. <tt class="docutils literal">HSP</tt>
objects are contained within <tt class="docutils literal">Hit</tt> objects and each <tt class="docutils literal">Hit</tt> has one
or more <tt class="docutils literal">HSP</tt> objects.</li>
<li><tt class="docutils literal">HSPFragment</tt>, to represent a single contiguous alignment between
query and hit sequences. <tt class="docutils literal">HSPFragment</tt> objects are contained within
<tt class="docutils literal">HSP</tt> objects. Most sequence search tools like BLAST and HMMER
unify <tt class="docutils literal">HSP</tt> and <tt class="docutils literal">HSPFragment</tt> objects as each <tt class="docutils literal">HSP</tt> will only
have a single <tt class="docutils literal">HSPFragment</tt>. However there are tools like BLAT and
Exonerate that produce <tt class="docutils literal">HSP</tt> containing multiple <tt class="docutils literal">HSPFragment</tt>.
Don’t worry if this seems a tad confusing now, we’ll elaborate more
on these two objects later on.</li>
</ul>
<p>These four objects are the ones you will interact with when you use
<tt class="docutils literal">Bio.SearchIO</tt>. They are created using one of the main
<tt class="docutils literal">Bio.SearchIO</tt> methods: <tt class="docutils literal">read</tt>, <tt class="docutils literal">parse</tt>, <tt class="docutils literal">index</tt>, or
<tt class="docutils literal">index_db</tt>. The details of these methods are provided in later
sections. For this section, we’ll only be using read and parse. These
functions behave similarly to their <tt class="docutils literal">Bio.SeqIO</tt> and <tt class="docutils literal">Bio.AlignIO</tt>
counterparts:</p>
<ul class="simple">
<li><tt class="docutils literal">read</tt> is used for search output files with a single query and
returns a <tt class="docutils literal">QueryResult</tt> object</li>
<li><tt class="docutils literal">parse</tt> is used for search output files with multiple queries and
returns a generator that yields <tt class="docutils literal">QueryResult</tt> objects</li>
</ul>
<p>With that settled, let’s start probing each <tt class="docutils literal">Bio.SearchIO</tt> object,
beginning with <tt class="docutils literal">QueryResult</tt>.</p>
<div class="section" id="queryresult">
<h3>8.1.1  QueryResult</h3>
<p>The QueryResult object represents a single search query and contains
zero or more Hit objects. Let’s see what it looks like using the BLAST
file we have:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blast.xml'</span><span class="punctuation">,</span> <span class="literal string">'blast-xml'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blast_qresult</span>
<span class="name">Program</span><span class="punctuation">:</span> <span class="name">blastn</span> <span class="punctuation">(</span><span class="literal number float">2.2</span><span class="operator">.</span><span class="literal number integer">27</span><span class="operator">+</span><span class="punctuation">)</span>
  <span class="name">Query</span><span class="punctuation">:</span> <span class="literal number integer">42291</span> <span class="punctuation">(</span><span class="literal number integer">61</span><span class="punctuation">)</span>
         <span class="name">mystery_seq</span>
 <span class="name">Target</span><span class="punctuation">:</span> <span class="name">refseq_rna</span>
   <span class="name">Hits</span><span class="punctuation">:</span> <span class="operator">----</span>  <span class="operator">-----</span>  <span class="operator">----------------------------------------------------------</span>
            <span class="comment">#  # HSP  ID + description</span>
         <span class="operator">----</span>  <span class="operator">-----</span>  <span class="operator">----------------------------------------------------------</span>
            <span class="literal number integer">0</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
            <span class="literal number integer">1</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171311</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035856</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
            <span class="literal number integer">2</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270133242</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032573</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Macaca</span> <span class="name">mulatta</span> <span class="name">microRNA</span> <span class="operator">...</span>
            <span class="literal number integer">3</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171322</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035857</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
            <span class="literal number integer">4</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171267</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035851</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
            <span class="literal number integer">5</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205330</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030198</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
            <span class="literal number integer">6</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205302</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030191</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">51.</span><span class="operator">..</span>
            <span class="literal number integer">7</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171259</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035850</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
            <span class="literal number integer">8</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205451</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030222</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">51.</span><span class="operator">..</span>
            <span class="literal number integer">9</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171447</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035871</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">10</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171276</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035852</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">11</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205290</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030188</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">51.</span><span class="operator">..</span>
           <span class="literal number integer">12</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171354</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035860</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">13</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205281</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030186</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
           <span class="literal number integer">14</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205298</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030190</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
           <span class="literal number integer">15</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171394</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035865</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">16</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205429</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030218</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">51.</span><span class="operator">..</span>
           <span class="literal number integer">17</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205423</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030217</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
           <span class="literal number integer">18</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171401</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035866</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">19</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270133247</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032574</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Macaca</span> <span class="name">mulatta</span> <span class="name">microRNA</span> <span class="operator">...</span>
           <span class="literal number integer">20</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205309</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030193</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
           <span class="literal number integer">21</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270132717</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032716</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Macaca</span> <span class="name">mulatta</span> <span class="name">microRNA</span> <span class="operator">...</span>
           <span class="literal number integer">22</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171437</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035870</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">23</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270133306</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032587</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Macaca</span> <span class="name">mulatta</span> <span class="name">microRNA</span> <span class="operator">...</span>
           <span class="literal number integer">24</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171428</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035869</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">25</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171211</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035845</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">26</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171153</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035838</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">27</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171146</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035837</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">28</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270133254</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032575</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Macaca</span> <span class="name">mulatta</span> <span class="name">microRNA</span> <span class="operator">...</span>
           <span class="literal number integer">29</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205445</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030221</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">51.</span><span class="operator">..</span>
           <span class="operator">~~~</span>
           <span class="literal number integer">97</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">356517317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_003527287</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">PREDICTED</span><span class="punctuation">:</span> <span class="name">Glycine</span> <span class="name">ma</span><span class="operator">...</span>
           <span class="literal number integer">98</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">297814701</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_002875188</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Arabidopsis</span> <span class="name">lyrata</span> <span class="name">su</span><span class="operator">...</span>
           <span class="literal number integer">99</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">397513516</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_003827011</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">PREDICTED</span><span class="punctuation">:</span> <span class="name">Pan</span> <span class="name">panisc</span><span class="operator">...</span>
</pre>
<p>We’ve just begun to scratch the surface of the object model, but you can
see that there’s already some useful information. By invoking <tt class="docutils literal">print</tt>
on the <tt class="docutils literal">QueryResult</tt> object, you can see:</p>
<ul class="simple">
<li>The program name and version (blastn version 2.2.27+)</li>
<li>The query ID, description, and its sequence length (ID is 42291,
description is ‘mystery_seq’, and it is 61 nucleotides long)</li>
<li>The target database to search against (refseq_rna)</li>
<li>A quick overview of the resulting hits. For our query sequence, there
are 100 potential hits (numbered 0–99 in the table). For each hit, we
can also see how many HSPs it contains, its ID, and a snippet of its
description. Notice here that <tt class="docutils literal">Bio.SearchIO</tt> truncates the hit
table overview, by showing only hits numbered 0–29, and then 97–99.</li>
</ul>
<p>Now let’s check our BLAT results using the same procedure as above:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blat.psl'</span><span class="punctuation">,</span> <span class="literal string">'blat-psl'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blat_qresult</span>
<span class="name">Program</span><span class="punctuation">:</span> <span class="name">blat</span> <span class="punctuation">(</span><span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">version</span><span class="operator">&gt;</span><span class="punctuation">)</span>
  <span class="name">Query</span><span class="punctuation">:</span> <span class="name">mystery_seq</span> <span class="punctuation">(</span><span class="literal number integer">61</span><span class="punctuation">)</span>
         <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
 <span class="name">Target</span><span class="punctuation">:</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">target</span><span class="operator">&gt;</span>
   <span class="name">Hits</span><span class="punctuation">:</span> <span class="operator">----</span>  <span class="operator">-----</span>  <span class="operator">----------------------------------------------------------</span>
            <span class="comment">#  # HSP  ID + description</span>
         <span class="operator">----</span>  <span class="operator">-----</span>  <span class="operator">----------------------------------------------------------</span>
            <span class="literal number integer">0</span>     <span class="literal number integer">17</span>  <span class="name">chr19</span>  <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
</pre>
<p>You’ll immediately notice that there are some differences. Some of these
are caused by the way PSL format stores its details, as you’ll see. The
rest are caused by the genuine program and target database differences
between our BLAST and BLAT searches:</p>
<ul class="simple">
<li>The program name and version. <tt class="docutils literal">Bio.SearchIO</tt> knows that the program
is BLAT, but in the output file there is no information regarding the
program version so it defaults to ‘&lt;unknown version&gt;’.</li>
<li>The query ID, description, and its sequence length. Notice here that
these details are slightly different from the ones we saw in BLAST.
The ID is ‘mystery_seq’ instead of 42991, there is no known
description, but the query length is still 61. This is actually a
difference introduced by the file formats themselves. BLAST sometimes
creates its own query IDs and uses your original ID as the sequence
description.</li>
<li>The target database is not known, as it is not stated in the BLAT
output file.</li>
<li>And finally, the list of hits we have is completely different. Here,
we see that our query sequence only hits the ‘chr19’ database entry,
but in it we see 17 HSP regions. This should not be surprising
however, given that we are using a different program, each with its
own target database.</li>
</ul>
<p>All the details you saw when invoking the <tt class="docutils literal">print</tt> method can be
accessed individually using Python’s object attribute access notation
(a.k.a. the dot notation). There are also other format-specific
attributes that you can access using the same method.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">blast_qresult</span><span class="operator">.</span><span class="name">program</span><span class="punctuation">,</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">version</span><span class="punctuation">)</span>
<span class="name">blastn</span> <span class="literal number float">2.2</span><span class="operator">.</span><span class="literal number integer">27</span><span class="operator">+</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">blat_qresult</span><span class="operator">.</span><span class="name">program</span><span class="punctuation">,</span> <span class="name">blat_qresult</span><span class="operator">.</span><span class="name">version</span><span class="punctuation">)</span>
<span class="name">blat</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">version</span><span class="operator">&gt;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">param_evalue_threshold</span>    <span class="comment"># blast-xml specific</span>
<span class="literal number float">10.0</span>
</pre>
<p>For a complete list of accessible attributes, you can check each
format-specific documentation. Here are the ones <a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SearchIO.BlastIO-module.html">for
BLAST</a>
and for
<a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SearchIO.BlatIO-module.html">BLAT</a>.</p>
<p>Having looked at using <tt class="docutils literal">print</tt> on <tt class="docutils literal">QueryResult</tt> objects, let’s drill
down deeper. What exactly is a <tt class="docutils literal">QueryResult</tt>? In terms of Python
objects, <tt class="docutils literal">QueryResult</tt> is a hybrid between a list and a dictionary. In
other words, it is a container object with all the convenient features
of lists and dictionaries.</p>
<p>Like Python lists and dictionaries, <tt class="docutils literal">QueryResult</tt> objects are
iterable. Each iteration returns a <tt class="docutils literal">Hit</tt> object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">hit</span> <span class="operator word">in</span> <span class="name">blast_qresult</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">hit</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|262205317|ref|NR_030195.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|301171311|ref|NR_035856.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|270133242|ref|NR_032573.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|301171322|ref|NR_035857.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">2</span> <span class="name">hsps</span><span class="punctuation">)</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|301171267|ref|NR_035851.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>To check how many items (hits) a <tt class="docutils literal">QueryResult</tt> has, you can simply
invoke Python’s <tt class="docutils literal">len</tt> method:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">blast_qresult</span><span class="punctuation">)</span>
<span class="literal number integer">100</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">blat_qresult</span><span class="punctuation">)</span>
<span class="literal number integer">1</span>
</pre>
<p>Like Python lists, you can retrieve items (hits) from a <tt class="docutils literal">QueryResult</tt>
using the slice notation:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>        <span class="comment"># retrieves the top hit</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|262205317|ref|NR_030195.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span>       <span class="comment"># retrieves the last hit</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|397513516|ref|XM_003827011.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
</pre>
<p>To retrieve multiple hits, you can slice <tt class="docutils literal">QueryResult</tt> objects using
the slice notation as well. In this case, the slice will return a new
<tt class="docutils literal">QueryResult</tt> object containing only the sliced hits:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_slice</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="punctuation">[:</span><span class="literal number integer">3</span><span class="punctuation">]</span>     <span class="comment"># slices the first three hits</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blast_slice</span>
<span class="name">Program</span><span class="punctuation">:</span> <span class="name">blastn</span> <span class="punctuation">(</span><span class="literal number float">2.2</span><span class="operator">.</span><span class="literal number integer">27</span><span class="operator">+</span><span class="punctuation">)</span>
  <span class="name">Query</span><span class="punctuation">:</span> <span class="literal number integer">42291</span> <span class="punctuation">(</span><span class="literal number integer">61</span><span class="punctuation">)</span>
         <span class="name">mystery_seq</span>
 <span class="name">Target</span><span class="punctuation">:</span> <span class="name">refseq_rna</span>
   <span class="name">Hits</span><span class="punctuation">:</span> <span class="operator">----</span>  <span class="operator">-----</span>  <span class="operator">----------------------------------------------------------</span>
            <span class="comment">#  # HSP  ID + description</span>
         <span class="operator">----</span>  <span class="operator">-----</span>  <span class="operator">----------------------------------------------------------</span>
            <span class="literal number integer">0</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
            <span class="literal number integer">1</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171311</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035856</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
            <span class="literal number integer">2</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270133242</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032573</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Macaca</span> <span class="name">mulatta</span> <span class="name">microRNA</span> <span class="operator">...</span>
</pre>
<p>Like Python dictionaries, you can also retrieve hits using the hit’s ID.
This is particularly useful if you know a given hit ID exists within a
search query results:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="punctuation">[</span><span class="literal string">'gi|262205317|ref|NR_030195.1|'</span><span class="punctuation">]</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|262205317|ref|NR_030195.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
</pre>
<p>You can also get a full list of <tt class="docutils literal">Hit</tt> objects using <tt class="docutils literal">hits</tt> and a
full list of <tt class="docutils literal">Hit</tt> IDs using <tt class="docutils literal">hit_keys</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">hits</span>
<span class="punctuation">[</span><span class="operator">...</span><span class="punctuation">]</span>       <span class="comment"># list of all hits</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">hit_keys</span>
<span class="punctuation">[</span><span class="operator">...</span><span class="punctuation">]</span>       <span class="comment"># list of all hit IDs</span>
</pre>
<p>What if you just want to check whether a particular hit is present in
the query results? You can do a simple Python membership test using the
<tt class="docutils literal">in</tt> keyword:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="literal string">'gi|262205317|ref|NR_030195.1|'</span> <span class="operator word">in</span> <span class="name">blast_qresult</span>
<span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="literal string">'gi|262205317|ref|NR_030194.1|'</span> <span class="operator word">in</span> <span class="name">blast_qresult</span>
<span class="name builtin pseudo">False</span>
</pre>
<p>Sometimes, knowing whether a hit is present is not enough; you also want
to know the rank of the hit. Here, the <tt class="docutils literal">index</tt> method comes to the
rescue:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">'gi|301171437|ref|NR_035870.1|'</span><span class="punctuation">)</span>
<span class="literal number integer">22</span>
</pre>
<p>Remember that we’re using Python’s indexing style here, which is
zero-based. This means our hit above is ranked at no. 23, not 22.</p>
<p>Also, note that the hit rank you see here is based on the native hit
ordering present in the original search output file. Different search
tools may order these hits based on different criteria.</p>
<p>If the native hit ordering doesn’t suit your taste, you can use the
<tt class="docutils literal">sort</tt> method of the <tt class="docutils literal">QueryResult</tt> object. It is very similar to
Python’s <tt class="docutils literal">list.sort</tt> method, with the addition of an option to create
a new sorted <tt class="docutils literal">QueryResult</tt> object or not.</p>
<p>Here is an example of using <tt class="docutils literal">QueryResult.sort</tt> to sort the hits based
on each hit’s full sequence length. For this particular sort, we’ll set
the <tt class="docutils literal">in_place</tt> flag to <tt class="docutils literal">False</tt> so that sorting will return a new
<tt class="docutils literal">QueryResult</tt> object and leave our initial object unsorted. We’ll also
set the <tt class="docutils literal">reverse</tt> flag to <tt class="docutils literal">True</tt> so that we sort in descending
order.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">hit</span> <span class="operator word">in</span> <span class="name">blast_qresult</span><span class="punctuation">[:</span><span class="literal number integer">5</span><span class="punctuation">]:</span>   <span class="comment"># id and sequence length of the first five hits</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">hit</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">hit</span><span class="operator">.</span><span class="name">seq_len</span>
<span class="operator">...</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">61</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171311</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035856</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">60</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270133242</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032573</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">85</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171322</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035857</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">86</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171267</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035851</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">80</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="name">sort_key</span> <span class="operator">=</span> <span class="keyword">lambda</span> <span class="name">hit</span><span class="punctuation">:</span> <span class="name">hit</span><span class="operator">.</span><span class="name">seq_len</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sorted_qresult</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">sort</span><span class="punctuation">(</span><span class="name">key</span><span class="operator">=</span><span class="name">sort_key</span><span class="punctuation">,</span> <span class="name">reverse</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">in_place</span><span class="operator">=</span><span class="name builtin pseudo">False</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">hit</span> <span class="operator word">in</span> <span class="name">sorted_qresult</span><span class="punctuation">[:</span><span class="literal number integer">5</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">hit</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">hit</span><span class="operator">.</span><span class="name">seq_len</span>
<span class="operator">...</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">397513516</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_003827011</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">6002</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">390332045</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_776818</span><span class="operator">.</span><span class="literal number integer">2</span><span class="operator">|</span> <span class="literal number integer">4082</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">390332043</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_003723358</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">4079</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">356517317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_003527287</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">3251</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">356543101</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_003539954</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">2936</span>
</pre>
<p>The advantage of having the <tt class="docutils literal">in_place</tt> flag here is that we’re
preserving the native ordering, so we may use it again later. You should
note that this is not the default behavior of <tt class="docutils literal">QueryResult.sort</tt>,
however, which is why we needed to set the <tt class="docutils literal">in_place</tt> flag to <tt class="docutils literal">True</tt>
explicitly.</p>
<p>At this point, you’ve known enough about <tt class="docutils literal">QueryResult</tt> objects to make
it work for you. But before we go on to the next object in the
<tt class="docutils literal">Bio.SearchIO</tt> model, let’s take a look at two more sets of methods
that could make it even easier to work with <tt class="docutils literal">QueryResult</tt> objects: the
<tt class="docutils literal">filter</tt> and <tt class="docutils literal">map</tt> methods.</p>
<p>If you’re familiar with Python’s list comprehensions, generator
expressions or the built in <tt class="docutils literal">filter</tt> and <tt class="docutils literal">map</tt> functions, you’ll
know how useful they are for working with list-like objects (if you’re
not, check them out!). You can use these built in methods to manipulate
<tt class="docutils literal">QueryResult</tt> objects, but you’ll end up with regular Python lists and
lose the ability to do more interesting manipulations.</p>
<p>That’s why, <tt class="docutils literal">QueryResult</tt> objects provide its own flavor of <tt class="docutils literal">filter</tt>
and <tt class="docutils literal">map</tt> methods. Analogous to <tt class="docutils literal">filter</tt>, there are <tt class="docutils literal">hit_filter</tt>
and <tt class="docutils literal">hsp_filter</tt> methods. As their name implies, these methods filter
its <tt class="docutils literal">QueryResult</tt> object either on its <tt class="docutils literal">Hit</tt> objects or <tt class="docutils literal">HSP</tt>
objects. Similarly, analogous to <tt class="docutils literal">map</tt>, <tt class="docutils literal">QueryResult</tt> objects also
provide the <tt class="docutils literal">hit_map</tt> and <tt class="docutils literal">hsp_map</tt> methods. These methods apply a
given function to all hits or HSPs in a <tt class="docutils literal">QueryResult</tt> object,
respectively.</p>
<p>Let’s see these methods in action, beginning with <tt class="docutils literal">hit_filter</tt>. This
method accepts a callback function that checks whether a given <tt class="docutils literal">Hit</tt>
object passes the condition you set or not. In other words, the function
must accept as its argument a single <tt class="docutils literal">Hit</tt> object and returns <tt class="docutils literal">True</tt>
or <tt class="docutils literal">False</tt>.</p>
<p>Here is an example of using <tt class="docutils literal">hit_filter</tt> to filter out <tt class="docutils literal">Hit</tt> objects
that only have one HSP:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">filter_func</span> <span class="operator">=</span> <span class="keyword">lambda</span> <span class="name">hit</span><span class="punctuation">:</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">hit</span><span class="operator">.</span><span class="name">hsps</span><span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="literal number integer">1</span>     <span class="comment"># the callback function</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">blast_qresult</span><span class="punctuation">)</span>      <span class="comment"># no. of hits before filtering</span>
<span class="literal number integer">100</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">filtered_qresult</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">hit_filter</span><span class="punctuation">(</span><span class="name">filter_func</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">filtered_qresult</span><span class="punctuation">)</span>   <span class="comment"># no. of hits after filtering</span>
<span class="literal number integer">37</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">hit</span> <span class="operator word">in</span> <span class="name">filtered_qresult</span><span class="punctuation">[:</span><span class="literal number integer">5</span><span class="punctuation">]:</span>    <span class="comment"># quick check for the hit lengths</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">hit</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">hit</span><span class="operator">.</span><span class="name">hsps</span><span class="punctuation">)</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171322</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035857</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">2</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205330</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030198</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">2</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171447</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035871</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">2</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205298</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030190</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">2</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270132717</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032716</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">2</span>
</pre>
<p><tt class="docutils literal">hsp_filter</tt> works the same as <tt class="docutils literal">hit_filter</tt>, only instead of looking
at the <tt class="docutils literal">Hit</tt> objects, it performs filtering on the <tt class="docutils literal">HSP</tt> objects in
each hits.</p>
<p>As for the <tt class="docutils literal">map</tt> methods, they too accept a callback function as their
arguments. However, instead of returning <tt class="docutils literal">True</tt> or <tt class="docutils literal">False</tt>, the
callback function must return the modified <tt class="docutils literal">Hit</tt> or <tt class="docutils literal">HSP</tt> object
(depending on whether you’re using <tt class="docutils literal">hit_map</tt> or <tt class="docutils literal">hsp_map</tt>).</p>
<p>Let’s see an example where we’re using <tt class="docutils literal">hit_map</tt> to rename the hit
IDs:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">def</span> <span class="name function">map_func</span><span class="punctuation">(</span><span class="name">hit</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="name">hit</span><span class="operator">.</span><span class="name">id</span> <span class="operator">=</span> <span class="name">hit</span><span class="operator">.</span><span class="name">id</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">(</span><span class="literal string">'|'</span><span class="punctuation">)[</span><span class="literal number integer">3</span><span class="punctuation">]</span>   <span class="comment"># renames 'gi|301171322|ref|NR_035857.1|' to 'NR_035857.1'</span>
<span class="operator">...</span>     <span class="keyword">return</span> <span class="name">hit</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mapped_qresult</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">hit_map</span><span class="punctuation">(</span><span class="name">map_func</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">hit</span> <span class="operator word">in</span> <span class="name">mapped_qresult</span><span class="punctuation">[:</span><span class="literal number integer">5</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">hit</span><span class="operator">.</span><span class="name">id</span>
<span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">NR_035856</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">NR_032573</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">NR_035857</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">NR_035851</span><span class="operator">.</span><span class="literal number integer">1</span>
</pre>
<p>Again, <tt class="docutils literal">hsp_map</tt> works the same as <tt class="docutils literal">hit_map</tt>, but on <tt class="docutils literal">HSP</tt> objects
instead of <tt class="docutils literal">Hit</tt> objects.</p>
</div>
<div class="section" id="hit">
<h3>8.1.2  Hit</h3>
<p><tt class="docutils literal">Hit</tt> objects represent all query results from a single database
entry. They are the second-level container in the <tt class="docutils literal">Bio.SearchIO</tt>
object hierarchy. You’ve seen that they are contained by <tt class="docutils literal">QueryResult</tt>
objects, but they themselves contain <tt class="docutils literal">HSP</tt> objects.</p>
<p>Let’s see what they look like, beginning with our BLAST search:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blast.xml'</span><span class="punctuation">,</span> <span class="literal string">'blast-xml'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hit</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">]</span>    <span class="comment"># fourth hit from the query result</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blast_hit</span>
<span class="name">Query</span><span class="punctuation">:</span> <span class="literal number integer">42291</span>
       <span class="name">mystery_seq</span>
  <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171322</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035857</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="punctuation">(</span><span class="literal number integer">86</span><span class="punctuation">)</span>
       <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span> <span class="name">mir</span><span class="operator">-</span><span class="literal number integer">520</span><span class="name">c</span> <span class="punctuation">(</span><span class="name">MIR520C</span><span class="punctuation">),</span> <span class="name">microRNA</span>
 <span class="name">HSPs</span><span class="punctuation">:</span> <span class="operator">----</span>  <span class="operator">--------</span>  <span class="operator">---------</span>  <span class="operator">------</span>  <span class="operator">---------------</span>  <span class="operator">---------------------</span>
          <span class="comment">#   E-value  Bit score    Span      Query range              Hit range</span>
       <span class="operator">----</span>  <span class="operator">--------</span>  <span class="operator">---------</span>  <span class="operator">------</span>  <span class="operator">---------------</span>  <span class="operator">---------------------</span>
          <span class="literal number integer">0</span>   <span class="literal number float">8.9e-20</span>     <span class="literal number float">100.47</span>      <span class="literal number integer">60</span>           <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>                <span class="punctuation">[</span><span class="literal number integer">13</span><span class="punctuation">:</span><span class="literal number integer">73</span><span class="punctuation">]</span>
          <span class="literal number integer">1</span>   <span class="literal number float">3.3e-06</span>      <span class="literal number float">55.39</span>      <span class="literal number integer">60</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>                <span class="punctuation">[</span><span class="literal number integer">13</span><span class="punctuation">:</span><span class="literal number integer">73</span><span class="punctuation">]</span>
</pre>
<p>You see that we’ve got the essentials covered here:</p>
<ul class="simple">
<li>The query ID and description is present. A hit is always tied to a
query, so we want to keep track of the originating query as well.
These values can be accessed from a hit using the <tt class="docutils literal">query_id</tt> and
<tt class="docutils literal">query_description</tt> attributes.</li>
<li>We also have the unique hit ID, description, and full sequence
lengths. They can be accessed using <tt class="docutils literal">id</tt>, <tt class="docutils literal">description</tt>, and
<tt class="docutils literal">seq_len</tt>, respectively.</li>
<li>Finally, there’s a table containing quick information about the HSPs
this hit contains. In each row, we’ve got the important HSP details
listed: the HSP index, its e-value, its bit score, its span (the
alignment length including gaps), its query coordinates, and its hit
coordinates.</li>
</ul>
<p>Now let’s contrast this with the BLAT search. Remember that in the BLAT
search we had one hit with 17 HSPs.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blat.psl'</span><span class="punctuation">,</span> <span class="literal string">'blat-psl'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hit</span> <span class="operator">=</span> <span class="name">blat_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>      <span class="comment"># the only hit</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blat_hit</span>
<span class="name">Query</span><span class="punctuation">:</span> <span class="name">mystery_seq</span>
       <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
  <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">chr19</span> <span class="punctuation">(</span><span class="literal number integer">59128983</span><span class="punctuation">)</span>
       <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
 <span class="name">HSPs</span><span class="punctuation">:</span> <span class="operator">----</span>  <span class="operator">--------</span>  <span class="operator">---------</span>  <span class="operator">------</span>  <span class="operator">---------------</span>  <span class="operator">---------------------</span>
          <span class="comment">#   E-value  Bit score    Span      Query range              Hit range</span>
       <span class="operator">----</span>  <span class="operator">--------</span>  <span class="operator">---------</span>  <span class="operator">------</span>  <span class="operator">---------------</span>  <span class="operator">---------------------</span>
          <span class="literal number integer">0</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54204480</span><span class="punctuation">:</span><span class="literal number integer">54204541</span><span class="punctuation">]</span>
          <span class="literal number integer">1</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54233104</span><span class="punctuation">:</span><span class="literal number integer">54264463</span><span class="punctuation">]</span>
          <span class="literal number integer">2</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54254477</span><span class="punctuation">:</span><span class="literal number integer">54260071</span><span class="punctuation">]</span>
          <span class="literal number integer">3</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54210720</span><span class="punctuation">:</span><span class="literal number integer">54210780</span><span class="punctuation">]</span>
          <span class="literal number integer">4</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54198476</span><span class="punctuation">:</span><span class="literal number integer">54198536</span><span class="punctuation">]</span>
          <span class="literal number integer">5</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54265610</span><span class="punctuation">:</span><span class="literal number integer">54265671</span><span class="punctuation">]</span>
          <span class="literal number integer">6</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54238143</span><span class="punctuation">:</span><span class="literal number integer">54240175</span><span class="punctuation">]</span>
          <span class="literal number integer">7</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54189735</span><span class="punctuation">:</span><span class="literal number integer">54189795</span><span class="punctuation">]</span>
          <span class="literal number integer">8</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54185425</span><span class="punctuation">:</span><span class="literal number integer">54185486</span><span class="punctuation">]</span>
          <span class="literal number integer">9</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54197657</span><span class="punctuation">:</span><span class="literal number integer">54197717</span><span class="punctuation">]</span>
         <span class="literal number integer">10</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54255662</span><span class="punctuation">:</span><span class="literal number integer">54255723</span><span class="punctuation">]</span>
         <span class="literal number integer">11</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54201651</span><span class="punctuation">:</span><span class="literal number integer">54201712</span><span class="punctuation">]</span>
         <span class="literal number integer">12</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">8</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54206009</span><span class="punctuation">:</span><span class="literal number integer">54206061</span><span class="punctuation">]</span>
         <span class="literal number integer">13</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>          <span class="punctuation">[</span><span class="literal number integer">10</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54178987</span><span class="punctuation">:</span><span class="literal number integer">54179038</span><span class="punctuation">]</span>
         <span class="literal number integer">14</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">8</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54212018</span><span class="punctuation">:</span><span class="literal number integer">54212071</span><span class="punctuation">]</span>
         <span class="literal number integer">15</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">8</span><span class="punctuation">:</span><span class="literal number integer">51</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54234278</span><span class="punctuation">:</span><span class="literal number integer">54234321</span><span class="punctuation">]</span>
         <span class="literal number integer">16</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">8</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54238143</span><span class="punctuation">:</span><span class="literal number integer">54238196</span><span class="punctuation">]</span>
</pre>
<p>Here, we’ve got a similar level of detail as with the BLAST hit we saw
earlier. There are some differences worth explaining, though:</p>
<ul class="simple">
<li>The e-value and bit score column values. As BLAT HSPs do not have
e-values and bit scores, the display defaults to ‘?’.</li>
<li>What about the span column? The span values is meant to display the
complete alignment length, which consists of all residues and any
gaps that may be present. The PSL format do not have this information
readily available and <tt class="docutils literal">Bio.SearchIO</tt> does not attempt to try guess
what it is, so we get a ‘?’ similar to the e-value and bit score
columns.</li>
</ul>
<p>In terms of Python objects, <tt class="docutils literal">Hit</tt> behaves almost the same as Python
lists, but contain <tt class="docutils literal">HSP</tt> objects exclusively. If you’re familiar with
lists, you should encounter no difficulties working with the <tt class="docutils literal">Hit</tt>
object.</p>
<p>Just like Python lists, <tt class="docutils literal">Hit</tt> objects are iterable, and each iteration
returns one <tt class="docutils literal">HSP</tt> object it contains:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">hsp</span> <span class="operator word">in</span> <span class="name">blast_hit</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">hsp</span>
<span class="name">HSP</span><span class="punctuation">(</span><span class="name">hit_id</span><span class="operator">=</span><span class="literal string">'gi|301171322|ref|NR_035857.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">fragments</span><span class="punctuation">)</span>
<span class="name">HSP</span><span class="punctuation">(</span><span class="name">hit_id</span><span class="operator">=</span><span class="literal string">'gi|301171322|ref|NR_035857.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">fragments</span><span class="punctuation">)</span>
</pre>
<p>You can invoke <tt class="docutils literal">len</tt> on a <tt class="docutils literal">Hit</tt> to see how many <tt class="docutils literal">HSP</tt> objects it
has:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">blast_hit</span><span class="punctuation">)</span>
<span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">blat_hit</span><span class="punctuation">)</span>
<span class="literal number integer">17</span>
</pre>
<p>You can use the slice notation on <tt class="docutils literal">Hit</tt> objects, whether to retrieve
single <tt class="docutils literal">HSP</tt> or multiple <tt class="docutils literal">HSP</tt> objects. Like <tt class="docutils literal">QueryResult</tt>, if you
slice for multiple <tt class="docutils literal">HSP</tt>, a new <tt class="docutils literal">Hit</tt> object will be returned
containing only the sliced <tt class="docutils literal">HSP</tt> objects:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hit</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>                 <span class="comment"># retrieve single items</span>
<span class="name">HSP</span><span class="punctuation">(</span><span class="name">hit_id</span><span class="operator">=</span><span class="literal string">'chr19'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'mystery_seq'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">fragments</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sliced_hit</span> <span class="operator">=</span> <span class="name">blat_hit</span><span class="punctuation">[</span><span class="literal number integer">4</span><span class="punctuation">:</span><span class="literal number integer">9</span><span class="punctuation">]</span>  <span class="comment"># retrieve multiple items</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">sliced_hit</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">sliced_hit</span>
<span class="name">Query</span><span class="punctuation">:</span> <span class="name">mystery_seq</span>
       <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
  <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">chr19</span> <span class="punctuation">(</span><span class="literal number integer">59128983</span><span class="punctuation">)</span>
       <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
 <span class="name">HSPs</span><span class="punctuation">:</span> <span class="operator">----</span>  <span class="operator">--------</span>  <span class="operator">---------</span>  <span class="operator">------</span>  <span class="operator">---------------</span>  <span class="operator">---------------------</span>
          <span class="comment">#   E-value  Bit score    Span      Query range              Hit range</span>
       <span class="operator">----</span>  <span class="operator">--------</span>  <span class="operator">---------</span>  <span class="operator">------</span>  <span class="operator">---------------</span>  <span class="operator">---------------------</span>
          <span class="literal number integer">0</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54198476</span><span class="punctuation">:</span><span class="literal number integer">54198536</span><span class="punctuation">]</span>
          <span class="literal number integer">1</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54265610</span><span class="punctuation">:</span><span class="literal number integer">54265671</span><span class="punctuation">]</span>
          <span class="literal number integer">2</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54238143</span><span class="punctuation">:</span><span class="literal number integer">54240175</span><span class="punctuation">]</span>
          <span class="literal number integer">3</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54189735</span><span class="punctuation">:</span><span class="literal number integer">54189795</span><span class="punctuation">]</span>
          <span class="literal number integer">4</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54185425</span><span class="punctuation">:</span><span class="literal number integer">54185486</span><span class="punctuation">]</span>
</pre>
<p>You can also sort the <tt class="docutils literal">HSP</tt> inside a <tt class="docutils literal">Hit</tt>, using the exact same
arguments like the sort method you saw in the <tt class="docutils literal">QueryResult</tt> object.</p>
<p>Finally, there are also the <tt class="docutils literal">filter</tt> and <tt class="docutils literal">map</tt> methods you can use
on <tt class="docutils literal">Hit</tt> objects. Unlike in the <tt class="docutils literal">QueryResult</tt> object, <tt class="docutils literal">Hit</tt>
objects only have one variant of <tt class="docutils literal">filter</tt> (<tt class="docutils literal">Hit.filter</tt>) and one
variant of <tt class="docutils literal">map</tt> (<tt class="docutils literal">Hit.map</tt>). Both of <tt class="docutils literal">Hit.filter</tt> and <tt class="docutils literal">Hit.map</tt>
work on the <tt class="docutils literal">HSP</tt> objects a <tt class="docutils literal">Hit</tt> has.</p>
</div>
<div class="section" id="hsp">
<h3>8.1.3  HSP</h3>
<p><tt class="docutils literal">HSP</tt> (high-scoring pair) represents region(s) in the hit sequence
that contains significant alignment(s) to the query sequence. It
contains the actual match between your query sequence and a database
entry. As this match is determined by the sequence search tool’s
algorithms, the <tt class="docutils literal">HSP</tt> object contains the bulk of the statistics
computed by the search tool. This also makes the distinction between
<tt class="docutils literal">HSP</tt> objects from different search tools more apparent compared to
the differences you’ve seen in <tt class="docutils literal">QueryResult</tt> or <tt class="docutils literal">Hit</tt> objects.</p>
<p>Let’s see some examples from our BLAST and BLAT searches. We’ll look at
the BLAST HSP first:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blast.xml'</span><span class="punctuation">,</span> <span class="literal string">'blast-xml'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">]</span>    <span class="comment"># first hit, first hsp</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blast_hsp</span>
      <span class="name">Query</span><span class="punctuation">:</span> <span class="literal number integer">42291</span> <span class="name">mystery_seq</span>
        <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number integer">520</span><span class="name">b</span> <span class="punctuation">(</span><span class="name">MIR520</span><span class="operator">...</span>
<span class="name">Query</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Hit</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="name">Quick</span> <span class="name">stats</span><span class="punctuation">:</span> <span class="name">evalue</span> <span class="literal number float">4.9e-23</span><span class="punctuation">;</span> <span class="name">bitscore</span> <span class="literal number float">111.29</span>
  <span class="name">Fragments</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="punctuation">(</span><span class="literal number integer">61</span> <span class="name">columns</span><span class="punctuation">)</span>
     <span class="name">Query</span> <span class="operator">-</span> <span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG</span>
             <span class="operator">|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||</span>
       <span class="name">Hit</span> <span class="operator">-</span> <span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG</span>
</pre>
<p>Just like <tt class="docutils literal">QueryResult</tt> and <tt class="docutils literal">Hit</tt>, invoking <tt class="docutils literal">print</tt> on an <tt class="docutils literal">HSP</tt>
shows its general details:</p>
<ul class="simple">
<li>There are the query and hit IDs and descriptions. We need these to
identify our <tt class="docutils literal">HSP</tt>.</li>
<li>We’ve also got the matching range of the query and hit sequences. The
slice notation we’re using here is an indication that the range is
displayed using Python’s indexing style (zero-based, half open). The
number inside the parenthesis denotes the strand. In this case, both
sequences have the plus strand.</li>
<li>Some quick statistics are available: the e-value and bitscore.</li>
<li>There is information about the HSP fragments. Ignore this for now; it
will be explained later on.</li>
<li>And finally, we have the query and hit sequence alignment itself.</li>
</ul>
<p>These details can be accessed on their own using the dot notation, just
like in <tt class="docutils literal">QueryResult</tt> and <tt class="docutils literal">Hit</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">query_range</span>
<span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">61</span><span class="punctuation">)</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">evalue</span>
<span class="literal number float">4.91307e-23</span>
</pre>
<p>They’re not the only attributes available, though. <tt class="docutils literal">HSP</tt> objects come
with a default set of properties that makes it easy to probe their
various details. Here are some examples:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">hit_start</span>         <span class="comment"># start coordinate of the hit sequence</span>
<span class="literal number integer">0</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">query_span</span>        <span class="comment"># how many residues in the query sequence</span>
<span class="literal number integer">61</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">aln_span</span>          <span class="comment"># how long the alignment is</span>
<span class="literal number integer">61</span>
</pre>
<p>Check out the <tt class="docutils literal">HSP</tt>
<a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SearchIO._model.hsp-module.html">documentation</a>
for a full list of these predefined properties.</p>
<p>Furthermore, each sequence search tool usually computes its own
statistics / details for its <tt class="docutils literal">HSP</tt> objects. For example, an XML BLAST
search also outputs the number of gaps and identical residues. These
attributes can be accessed like so:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">gap_num</span>       <span class="comment"># number of gaps</span>
<span class="literal number integer">0</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">ident_num</span>     <span class="comment"># number of identical residues</span>
<span class="literal number integer">61</span>
</pre>
<p>These details are format-specific; they may not be present in other
formats. To see which details are available for a given sequence search
tool, you should check the format’s documentation in <tt class="docutils literal">Bio.SearchIO</tt>.
Alternatively, you may also use <tt class="docutils literal">.__dict__.keys()</tt> for a quick list of
what’s available:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">__dict__</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'bitscore'</span><span class="punctuation">,</span> <span class="literal string">'evalue'</span><span class="punctuation">,</span> <span class="literal string">'ident_num'</span><span class="punctuation">,</span> <span class="literal string">'gap_num'</span><span class="punctuation">,</span> <span class="literal string">'bitscore_raw'</span><span class="punctuation">,</span> <span class="literal string">'pos_num'</span><span class="punctuation">,</span> <span class="literal string">'_items'</span><span class="punctuation">]</span>
</pre>
<p>Finally, you may have noticed that the <tt class="docutils literal">query</tt> and <tt class="docutils literal">hit</tt> attributes
of our HSP are not just regular strings:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">query</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTT...GGG'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'aligned query sequence'</span><span class="punctuation">,</span> <span class="name">description</span><span class="operator">=</span><span class="literal string">'mystery_seq'</span><span class="punctuation">,</span> <span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[])</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">hit</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTT...GGG'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|262205317|ref|NR_030195.1|'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'aligned hit sequence'</span><span class="punctuation">,</span> <span class="name">description</span><span class="operator">=</span><span class="literal string">'Homo sapiens microRNA 520b (MIR520B), microRNA'</span><span class="punctuation">,</span> <span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[])</span>
</pre>
<p>They are <tt class="docutils literal">SeqRecord</tt> objects you saw earlier in
Section <a class="reference external" href="#chapter:SeqRecord">4</a>! This means that you can do all
sorts of interesting things you can do with <tt class="docutils literal">SeqRecord</tt> objects on
<tt class="docutils literal">HSP.query</tt> and/or <tt class="docutils literal">HSP.hit</tt>.</p>
<p>It should not surprise you now that the <tt class="docutils literal">HSP</tt> object has an
<tt class="docutils literal">alignment</tt> property which is a <tt class="docutils literal">MultipleSeqAlignment</tt> object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">aln</span>
<span class="name">DNAAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">2</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">61</span> <span class="name">columns</span>
<span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAG</span><span class="operator">...</span><span class="name">GGG</span> <span class="literal number integer">42291</span>
<span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAG</span><span class="operator">...</span><span class="name">GGG</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>
</pre>
<p>Having probed the BLAST HSP, let’s now take a look at HSPs from our BLAT
results for a different kind of HSP. As usual, we’ll begin by invoking
<tt class="docutils literal">print</tt> on it:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blat.psl'</span><span class="punctuation">,</span> <span class="literal string">'blat-psl'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span> <span class="operator">=</span> <span class="name">blat_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">]</span>       <span class="comment"># first hit, first hsp</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blat_hsp</span>
      <span class="name">Query</span><span class="punctuation">:</span> <span class="name">mystery_seq</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
        <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">chr19</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
<span class="name">Query</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Hit</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">54204480</span><span class="punctuation">:</span><span class="literal number integer">54204541</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="name">Quick</span> <span class="name">stats</span><span class="punctuation">:</span> <span class="name">evalue</span> <span class="error">?</span><span class="punctuation">;</span> <span class="name">bitscore</span> <span class="error">?</span>
  <span class="name">Fragments</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="punctuation">(</span><span class="error">?</span> <span class="name">columns</span><span class="punctuation">)</span>
</pre>
<p>Some of the outputs you may have already guessed. We have the query and
hit IDs and descriptions and the sequence coordinates. Values for evalue
and bitscore is ‘?’ as BLAT HSPs do not have these attributes. But The
biggest difference here is that you don’t see any sequence alignments
displayed. If you look closer, PSL formats themselves do not have any
hit or query sequences, so <tt class="docutils literal">Bio.SearchIO</tt> won’t create any sequence or
alignment objects. What happens if you try to access <tt class="docutils literal">HSP.query</tt>,
<tt class="docutils literal">HSP.hit</tt>, or <tt class="docutils literal">HSP.aln</tt>? You’ll get the default values for these
attributes, which is <tt class="docutils literal">None</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">hit</span> <span class="operator word">is</span> <span class="name builtin pseudo">None</span>
<span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">query</span> <span class="operator word">is</span> <span class="name builtin pseudo">None</span>
<span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">aln</span> <span class="operator word">is</span> <span class="name builtin pseudo">None</span>
<span class="name builtin pseudo">True</span>
</pre>
<p>This does not affect other attributes, though. For example, you can
still access the length of the query or hit alignment. Despite not
displaying any attributes, the PSL format still have this information so
<tt class="docutils literal">Bio.SearchIO</tt> can extract them:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">query_span</span>     <span class="comment"># length of query match</span>
<span class="literal number integer">61</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">hit_span</span>       <span class="comment"># length of hit match</span>
<span class="literal number integer">61</span>
</pre>
<p>Other format-specific attributes are still present as well:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">score</span>          <span class="comment"># PSL score</span>
<span class="literal number integer">61</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">mismatch_num</span>   <span class="comment"># the mismatch column</span>
<span class="literal number integer">0</span>
</pre>
<p>So far so good? Things get more interesting when you look at another
‘variant’ of HSP present in our BLAT results. You might recall that in
BLAT searches, sometimes we get our results separated into ‘blocks’.
These blocks are essentially alignment fragments that may have some
intervening sequence between them.</p>
<p>Let’s take a look at a BLAT HSP that contains multiple blocks to see how
<tt class="docutils literal">Bio.SearchIO</tt> deals with this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span> <span class="operator">=</span> <span class="name">blat_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">1</span><span class="punctuation">]</span>      <span class="comment"># first hit, second hsp</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blat_hsp2</span>
      <span class="name">Query</span><span class="punctuation">:</span> <span class="name">mystery_seq</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
        <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">chr19</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
<span class="name">Query</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Hit</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">54233104</span><span class="punctuation">:</span><span class="literal number integer">54264463</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="name">Quick</span> <span class="name">stats</span><span class="punctuation">:</span> <span class="name">evalue</span> <span class="error">?</span><span class="punctuation">;</span> <span class="name">bitscore</span> <span class="error">?</span>
  <span class="name">Fragments</span><span class="punctuation">:</span> <span class="operator">---</span>  <span class="operator">--------------</span>  <span class="operator">----------------------</span>  <span class="operator">----------------------</span>
               <span class="comment">#            Span             Query range               Hit range</span>
             <span class="operator">---</span>  <span class="operator">--------------</span>  <span class="operator">----------------------</span>  <span class="operator">----------------------</span>
               <span class="literal number integer">0</span>               <span class="error">?</span>                  <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">18</span><span class="punctuation">]</span>     <span class="punctuation">[</span><span class="literal number integer">54233104</span><span class="punctuation">:</span><span class="literal number integer">54233122</span><span class="punctuation">]</span>
               <span class="literal number integer">1</span>               <span class="error">?</span>                 <span class="punctuation">[</span><span class="literal number integer">18</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>     <span class="punctuation">[</span><span class="literal number integer">54264420</span><span class="punctuation">:</span><span class="literal number integer">54264463</span><span class="punctuation">]</span>
</pre>
<p>What’s happening here? We still some essential details covered: the IDs
and descriptions, the coordinates, and the quick statistics are similar
to what you’ve seen before. But the fragments detail is all different.
Instead of showing ‘Fragments: 1’, we now have a table with two data
rows.</p>
<p>This is how <tt class="docutils literal">Bio.SearchIO</tt> deals with HSPs having multiple fragments.
As mentioned before, an HSP alignment may be separated by intervening
sequences into fragments. The intervening sequences are not part of the
query-hit match, so they should not be considered part of query nor hit
sequence. However, they do affect how we deal with sequence coordinates,
so we can’t ignore them.</p>
<p>Take a look at the hit coordinate of the HSP above. In the
<tt class="docutils literal">Hit range:</tt> field, we see that the coordinate is
<tt class="docutils literal">[54233104:54264463]</tt>. But looking at the table rows, we see that not
the entire region spanned by this coordinate matches our query.
Specifically, the intervening region spans from <tt class="docutils literal">54233122</tt> to
<tt class="docutils literal">54264420</tt>.</p>
<p>Why then, is the query coordinates seem to be contiguous, you ask? This
is perfectly fine. In this case it means that the query match is
contiguous (no intervening regions), while the hit match is not.</p>
<p>All these attributes are accessible from the HSP directly, by the way:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">hit_range</span>         <span class="comment"># hit start and end coordinates of the entire HSP</span>
<span class="punctuation">(</span><span class="literal number integer">54233104</span><span class="punctuation">,</span> <span class="literal number integer">54264463</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">hit_range_all</span>     <span class="comment"># hit start and end coordinates of each fragment</span>
<span class="punctuation">[(</span><span class="literal number integer">54233104</span><span class="punctuation">,</span> <span class="literal number integer">54233122</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="literal number integer">54264420</span><span class="punctuation">,</span> <span class="literal number integer">54264463</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">hit_span</span>          <span class="comment"># hit span of the entire HSP</span>
<span class="literal number integer">31359</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">hit_span_all</span>      <span class="comment"># hit span of each fragment</span>
<span class="punctuation">[</span><span class="literal number integer">18</span><span class="punctuation">,</span> <span class="literal number integer">43</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">hit_inter_ranges</span>  <span class="comment"># start and end coordinates of intervening regions in the hit sequence</span>
<span class="punctuation">[(</span><span class="literal number integer">54233122</span><span class="punctuation">,</span> <span class="literal number integer">54264420</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">hit_inter_spans</span>   <span class="comment"># span of intervening regions in the hit sequence</span>
<span class="punctuation">[</span><span class="literal number integer">31298</span><span class="punctuation">]</span>
</pre>
<p>Most of these attributes are not readily available from the PSL file we
have, but <tt class="docutils literal">Bio.SearchIO</tt> calculates them for you on the fly when you
parse the PSL file. All it needs are the start and end coordinates of
each fragment.</p>
<p>What about the <tt class="docutils literal">query</tt>, <tt class="docutils literal">hit</tt>, and <tt class="docutils literal">aln</tt> attributes? If the HSP
has multiple fragments, you won’t be able to use these attributes as
they only fetch single <tt class="docutils literal">SeqRecord</tt> or <tt class="docutils literal">MultipleSeqAlignment</tt>
objects. However, you can use their <tt class="docutils literal">*_all</tt> counterparts:
<tt class="docutils literal">query_all</tt>, <tt class="docutils literal">hit_all</tt>, and <tt class="docutils literal">aln_all</tt>. These properties will
return a list containing <tt class="docutils literal">SeqRecord</tt> or <tt class="docutils literal">MultipleSeqAlignment</tt>
objects from each of the HSP fragment. There are other attributes that
behave similarly, i.e. they only work for HSPs with one fragment. Check
out the <tt class="docutils literal">HSP</tt>
<a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SearchIO._model.hsp-module.html">documentation</a>
for a full list.</p>
<p>Finally, to check whether you have multiple fragments or not, you can
use the <tt class="docutils literal">is_fragmented</tt> property like so:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">is_fragmented</span>     <span class="comment"># BLAT HSP with 2 fragments</span>
<span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">is_fragmented</span>      <span class="comment"># BLAT HSP from earlier, with one fragment</span>
<span class="name builtin pseudo">False</span>
</pre>
<p>Before we move on, you should also know that we can use the slice
notation on <tt class="docutils literal">HSP</tt> objects, just like <tt class="docutils literal">QueryResult</tt> or <tt class="docutils literal">Hit</tt>
objects. When you use this notation, you’ll get an <tt class="docutils literal">HSPFragment</tt>
object in return, the last component of the object model.</p>
</div>
<div class="section" id="hspfragment">
<h3>8.1.4  HSPFragment</h3>
<p><tt class="docutils literal">HSPFragment</tt> represents a single, contiguous match between the query
and hit sequences. You could consider it the core of the object model
and search result, since it is the presence of these fragments that
determine whether your search have results or not.</p>
<p>In most cases, you don’t have to deal with <tt class="docutils literal">HSPFragment</tt> objects
directly since not that many sequence search tools fragment their HSPs.
When you do have to deal with them, what you should remember is that
<tt class="docutils literal">HSPFragment</tt> objects were written with to be as compact as possible.
In most cases, they only contain attributes directly related to
sequences: strands, reading frames, alphabets, coordinates, the
sequences themselves, and their IDs and descriptions.</p>
<p>These attributes are readily shown when you invoke <tt class="docutils literal">print</tt> on an
<tt class="docutils literal">HSPFragment</tt>. Here’s an example, taken from our BLAST search:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blast.xml'</span><span class="punctuation">,</span> <span class="literal string">'blast-xml'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_frag</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">]</span>    <span class="comment"># first hit, first hsp, first fragment</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blast_frag</span>
      <span class="name">Query</span><span class="punctuation">:</span> <span class="literal number integer">42291</span> <span class="name">mystery_seq</span>
        <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number integer">520</span><span class="name">b</span> <span class="punctuation">(</span><span class="name">MIR520</span><span class="operator">...</span>
<span class="name">Query</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Hit</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Fragments</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="punctuation">(</span><span class="literal number integer">61</span> <span class="name">columns</span><span class="punctuation">)</span>
     <span class="name">Query</span> <span class="operator">-</span> <span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG</span>
             <span class="operator">|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||</span>
       <span class="name">Hit</span> <span class="operator">-</span> <span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG</span>
</pre>
<p>At this level, the BLAT fragment looks quite similar to the BLAST
fragment, save for the query and hit sequences which are not present:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blat.psl'</span><span class="punctuation">,</span> <span class="literal string">'blat-psl'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_frag</span> <span class="operator">=</span> <span class="name">blat_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">]</span>    <span class="comment"># first hit, first hsp, first fragment</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blat_frag</span>
      <span class="name">Query</span><span class="punctuation">:</span> <span class="name">mystery_seq</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
        <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">chr19</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
<span class="name">Query</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Hit</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">54204480</span><span class="punctuation">:</span><span class="literal number integer">54204541</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Fragments</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="punctuation">(</span><span class="error">?</span> <span class="name">columns</span><span class="punctuation">)</span>
</pre>
<p>In all cases, these attributes are accessible using our favorite dot
notation. Some examples:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_frag</span><span class="operator">.</span><span class="name">query_start</span>      <span class="comment"># query start coordinate</span>
<span class="literal number integer">0</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_frag</span><span class="operator">.</span><span class="name">hit_strand</span>       <span class="comment"># hit sequence strand</span>
<span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_frag</span><span class="operator">.</span><span class="name">hit</span>              <span class="comment"># hit sequence, as a SeqRecord object</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTT...GGG'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|262205317|ref|NR_030195.1|'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'aligned hit sequence'</span><span class="punctuation">,</span> <span class="name">description</span><span class="operator">=</span><span class="literal string">'Homo sapiens microRNA 520b (MIR520B), microRNA'</span><span class="punctuation">,</span> <span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[])</span>
</pre>
</div>
</div>
<div class="section" id="a-note-about-standards-and-conventions">
<h2>8.2  A note about standards and conventions</h2>
<p>Before we move on to the main functions, there is something you ought to
know about the standards <tt class="docutils literal">Bio.SearchIO</tt> uses. If you’ve worked with
multiple sequence search tools, you might have had to deal with the many
different ways each program deals with things like sequence coordinates.
It might not have been a pleasant experience as these search tools
usually have their own standards. For example, one tools might use
one-based coordinates, while the other uses zero-based coordinates. Or,
one program might reverse the start and end coordinates if the strand is
minus, while others don’t. In short, these often creates unnecessary
mess must be dealt with.</p>
<p>We realize this problem ourselves and we intend to address it in
<tt class="docutils literal">Bio.SearchIO</tt>. After all, one of the goals of <tt class="docutils literal">Bio.SearchIO</tt> is to
create a common, easy to use interface to deal with various search
output files. This means creating standards that extend beyond the
object model you just saw.</p>
<p>Now, you might complain, &quot;Not another standard!&quot;. Well, eventually we
have to choose one convention or the other, so this is necessary. Plus,
we’re not creating something entirely new here; just adopting a standard
we think is best for a Python programmer (it is Biopython, after all).</p>
<p>There are three implicit standards that you can expect when working with
<tt class="docutils literal">Bio.SearchIO</tt>:</p>
<ul class="simple">
<li>The first one pertains to sequence coordinates. In <tt class="docutils literal">Bio.SearchIO</tt>,
all sequence coordinates follows Python’s coordinate style:
zero-based and half open. For example, if in a BLAST XML output file
the start and end coordinates of an HSP are 10 and 28, they would
become 9 and 28 in <tt class="docutils literal">Bio.SearchIO</tt>. The start coordinate becomes 9
because Python indices start from zero, while the end coordinate
remains 28 as Python slices omit the last item in an interval.</li>
<li>The second is on sequence coordinate orders. In <tt class="docutils literal">Bio.SearchIO</tt>,
start coordinates are always less than or equal to end coordinates.
This isn’t always the case with all sequence search tools, as some of
them have larger start coordinates when the sequence strand is minus.</li>
<li>The last one is on strand and reading frame values. For strands,
there are only four valid choices: <tt class="docutils literal">1</tt> (plus strand), <tt class="docutils literal"><span class="pre">-1</span></tt> (minus
strand), <tt class="docutils literal">0</tt> (protein sequences), and <tt class="docutils literal">None</tt> (no strand). For
reading frames, the valid choices are integers from <tt class="docutils literal"><span class="pre">-3</span></tt> to <tt class="docutils literal">3</tt>
and <tt class="docutils literal">None</tt>.</li>
</ul>
<p>Note that these standards only exist in <tt class="docutils literal">Bio.SearchIO</tt> objects. If you
write <tt class="docutils literal">Bio.SearchIO</tt> objects into an output format, <tt class="docutils literal">Bio.SearchIO</tt>
will use the format’s standard for the output. It does not force its
standard over to your output file.</p>
</div>
<div class="section" id="reading-search-output-files">
<h2>8.3  Reading search output files</h2>
<p>There are two functions you can use for reading search output files into
<tt class="docutils literal">Bio.SearchIO</tt> objects: <tt class="docutils literal">read</tt> and <tt class="docutils literal">parse</tt>. They’re essentially
similar to <tt class="docutils literal">read</tt> and <tt class="docutils literal">parse</tt> functions in other submodules like
<tt class="docutils literal">Bio.SeqIO</tt> or <tt class="docutils literal">Bio.AlignIO</tt>. In both cases, you need to supply the
search output file name and the file format name, both as Python
strings. You can check the documentation for a list of format names
<tt class="docutils literal">Bio.SearchIO</tt> recognizes.</p>
<p><tt class="docutils literal">Bio.SearchIO.read</tt> is used for reading search output files with only
one query and returns a <tt class="docutils literal">QueryResult</tt> object. You’ve seen <tt class="docutils literal">read</tt>
used in our previous examples. What you haven’t seen is that <tt class="docutils literal">read</tt>
may also accept additional keyword arguments, depending on the file
format.</p>
<p>Here are some examples. In the first one, we use <tt class="docutils literal">read</tt> just like
previously to read a BLAST tabular output file. In the second one, we
use a keyword argument to modify so it parses the BLAST tabular variant
with comments in it:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_003.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresult</span>
<span class="name">QueryResult</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">hits</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresult2</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_007.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">,</span> <span class="name">comments</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresult2</span>
<span class="name">QueryResult</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">hits</span><span class="punctuation">)</span>
</pre>
<p>These keyword arguments differs among file formats. Check the format
documentation to see if it has keyword arguments that modifies its
parser’s behavior.</p>
<p>As for the <tt class="docutils literal">Bio.SearchIO.parse</tt>, it is used for reading search output
files with any number of queries. The function returns a generator
object that yields a <tt class="docutils literal">QueryResult</tt> object in each iteration. Like
<tt class="docutils literal">Bio.SearchIO.read</tt>, it also accepts format-specific keyword
arguments:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresults</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_001.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">qresult</span> <span class="operator word">in</span> <span class="name">qresults</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">qresult</span><span class="operator">.</span><span class="name">id</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">16080617</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NP_391444</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">11464971</span><span class="punctuation">:</span><span class="literal number integer">4</span><span class="operator">-</span><span class="literal number integer">101</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresults2</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_005.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">,</span> <span class="name">comments</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">qresult</span> <span class="operator word">in</span> <span class="name">qresults2</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">qresult</span><span class="operator">.</span><span class="name">id</span>
<span class="name">random_s00</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">16080617</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NP_391444</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">11464971</span><span class="punctuation">:</span><span class="literal number integer">4</span><span class="operator">-</span><span class="literal number integer">101</span>
</pre>
</div>
<div class="section" id="dealing-with-large-search-output-files-with-indexing">
<h2>8.4  Dealing with large search output files with indexing</h2>
<p>Sometimes, you’re handed a search output file containing hundreds or
thousands of queries that you need to parse. You can of course use
<tt class="docutils literal">Bio.SearchIO.parse</tt> for this file, but that would be grossly
inefficient if you need to access only a few of the queries. This is
because <tt class="docutils literal">parse</tt> will parse all queries it sees before it fetches your
query of interest.</p>
<p>In this case, the ideal choice would be to index the file using
<tt class="docutils literal">Bio.SearchIO.index</tt> or <tt class="docutils literal">Bio.SearchIO.index_db</tt>. If the names sound
familiar, it’s because you’ve seen them before in
Section <a class="reference external" href="#sec:SeqIO-index">5.4.2</a>. These functions also behave
similarly to their <tt class="docutils literal">Bio.SeqIO</tt> counterparts, with the addition of
format-specific keyword arguments.</p>
<p>Here are some examples. You can use <tt class="docutils literal">index</tt> with just the filename and
format name:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idx</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_001.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">sorted</span><span class="punctuation">(</span><span class="name">idx</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">())</span>
<span class="punctuation">[</span><span class="literal string">'gi|11464971:4-101'</span><span class="punctuation">,</span> <span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idx</span><span class="punctuation">[</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">]</span>
<span class="name">QueryResult</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">hits</span><span class="punctuation">)</span>
</pre>
<p>Or also with the format-specific keyword argument:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idx</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_005.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">,</span> <span class="name">comments</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">sorted</span><span class="punctuation">(</span><span class="name">idx</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">())</span>
<span class="punctuation">[</span><span class="literal string">'gi|11464971:4-101'</span><span class="punctuation">,</span> <span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">,</span> <span class="literal string">'random_s00'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idx</span><span class="punctuation">[</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">]</span>
<span class="name">QueryResult</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">hits</span><span class="punctuation">)</span>
</pre>
<p>Or with the <tt class="docutils literal">key_function</tt> argument, as in <tt class="docutils literal">Bio.SeqIO</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">key_function</span> <span class="operator">=</span> <span class="keyword">lambda</span> <span class="name builtin">id</span><span class="punctuation">:</span> <span class="name builtin">id</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">()</span>    <span class="comment"># capitalizes the keys</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idx</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_001.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">,</span> <span class="name">key_function</span><span class="operator">=</span><span class="name">key_function</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">sorted</span><span class="punctuation">(</span><span class="name">idx</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">())</span>
<span class="punctuation">[</span><span class="literal string">'GI|11464971:4-101'</span><span class="punctuation">,</span> <span class="literal string">'GI|16080617|REF|NP_391444.1|'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idx</span><span class="punctuation">[</span><span class="literal string">'GI|16080617|REF|NP_391444.1|'</span><span class="punctuation">]</span>
<span class="name">QueryResult</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">hits</span><span class="punctuation">)</span>
</pre>
<p><tt class="docutils literal">Bio.SearchIO.index_db</tt> works like as <tt class="docutils literal">index</tt>, only it writes the
query offsets into an SQLite database file.</p>
</div>
<div class="section" id="writing-and-converting-search-output-files">
<h2>8.5  Writing and converting search output files</h2>
<p>It is occasionally useful to be able to manipulate search results from
an output file and write it again to a new file. <tt class="docutils literal">Bio.SearchIO</tt>
provides a <tt class="docutils literal">write</tt> function that lets you do exactly this. It takes as
its arguments an iterable returning <tt class="docutils literal">QueryResult</tt> objects, the output
filename to write to, the format name to write to, and optionally some
format-specific keyword arguments. It returns a four-item tuple, which
denotes the number or <tt class="docutils literal">QueryResult</tt>, <tt class="docutils literal">Hit</tt>, <tt class="docutils literal">HSP</tt>, and
<tt class="docutils literal">HSPFragment</tt> objects that were written.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresults</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">'mirna.xml'</span><span class="punctuation">,</span> <span class="literal string">'blast-xml'</span><span class="punctuation">)</span>     <span class="comment"># read XML file</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">qresults</span><span class="punctuation">,</span> <span class="literal string">'results.tab'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">)</span>    <span class="comment"># write to tabular file</span>
<span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">239</span><span class="punctuation">,</span> <span class="literal number integer">277</span><span class="punctuation">,</span> <span class="literal number integer">277</span><span class="punctuation">)</span>
</pre>
<p>You should note different file formats require different attributes of
the <tt class="docutils literal">QueryResult</tt>, <tt class="docutils literal">Hit</tt>, <tt class="docutils literal">HSP</tt> and <tt class="docutils literal">HSPFragment</tt> objects. If
these attributes are not present, writing won’t work. In other words,
you can’t always write to the output format that you want. For example,
if you read a BLAST XML file, you wouldn’t be able to write the results
to a PSL file as PSL files require attributes not calculated by BLAST
(e.g. the number of repeat matches). You can always set these attributes
manually, if you really want to write to PSL, though.</p>
<p>Like <tt class="docutils literal">read</tt>, <tt class="docutils literal">parse</tt>, <tt class="docutils literal">index</tt>, and <tt class="docutils literal">index_db</tt>, <tt class="docutils literal">write</tt> also
accepts format-specific keyword arguments. Check out the documentation
for a complete list of formats <tt class="docutils literal">Bio.SearchIO</tt> can write to and their
arguments.</p>
<p>Finally, <tt class="docutils literal">Bio.SearchIO</tt> also provides a <tt class="docutils literal">convert</tt> function, which is
simply a shortcut for <tt class="docutils literal">Bio.SearchIO.parse</tt> and <tt class="docutils literal">Bio.SearchIO.write</tt>.
Using the convert function, our example above would be:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">'mirna.xml'</span><span class="punctuation">,</span> <span class="literal string">'blast-xml'</span><span class="punctuation">,</span> <span class="literal string">'results.tab'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">)</span>
<span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">239</span><span class="punctuation">,</span> <span class="literal number integer">277</span><span class="punctuation">,</span> <span class="literal number integer">277</span><span class="punctuation">)</span>
</pre>
<p>As <tt class="docutils literal">convert</tt> uses <tt class="docutils literal">write</tt>, it is only limited to format conversions
that have all the required attributes. Here, the BLAST XML file provides
all the default values a BLAST tabular file requires, so it works just
fine. However, other format conversions are less likely to work since
you need to manually assign the required attributes first.</p>
</div>
</div>
<div class="section" id="hapter-9-accessing-ncbis-entrez-databases">
<h1>hapter 9  Accessing NCBI’s Entrez databases</h1>
<p>Entrez
(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/Entrez">http://www.ncbi.nlm.nih.gov/Entrez</a>)
is a data retrieval system that provides users access to NCBI’s
databases such as PubMed, GenBank, GEO, and many others. You can access
Entrez from a web browser to manually enter queries, or you can use
Biopython’s <tt class="docutils literal">Bio.Entrez</tt> module for programmatic access to Entrez. The
latter allows you for example to search PubMed or download GenBank
records from within a Python script.</p>
<p>The <tt class="docutils literal">Bio.Entrez</tt> module makes use of the Entrez Programming Utilities
(also known as EUtils), consisting of eight tools that are described in
detail on NCBI’s page at
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/utils/">http://www.ncbi.nlm.nih.gov/entrez/utils/</a>.
Each of these tools corresponds to one Python function in the
<tt class="docutils literal">Bio.Entrez</tt> module, as described in the sections below. This module
makes sure that the correct URL is used for the queries, and that not
more than one request is made every three seconds, as required by NCBI.</p>
<p>The output returned by the Entrez Programming Utilities is typically in
XML format. To parse such output, you have several options:</p>
<ol class="arabic simple">
<li>Use <tt class="docutils literal">Bio.Entrez</tt>’s parser to parse the XML output into a Python
object;</li>
<li>Use the DOM (Document Object Model) parser in Python’s standard
library;</li>
<li>Use the SAX (Simple API for XML) parser in Python’s standard library;</li>
<li>Read the XML output as raw text, and parse it by string searching and
manipulation.</li>
</ol>
<p>For the DOM and SAX parsers, see the Python documentation. The parser in
<tt class="docutils literal">Bio.Entrez</tt> is discussed below.</p>
<p>NCBI uses DTD (Document Type Definition) files to describe the structure
of the information contained in XML files. Most of the DTD files used by
NCBI are included in the Biopython distribution. The <tt class="docutils literal">Bio.Entrez</tt>
parser makes use of the DTD files when parsing an XML file returned by
NCBI Entrez.</p>
<p>Occasionally, you may find that the DTD file associated with a specific
XML file is missing in the Biopython distribution. In particular, this
may happen when NCBI updates its DTD files. If this happens,
<tt class="docutils literal">Entrez.read</tt> will show a warning message with the name and URL of the
missing DTD file. The parser will proceed to access the missing DTD file
through the internet, allowing the parsing of the XML file to continue.
However, the parser is much faster if the DTD file is available locally.
For this purpose, please download the DTD file from the URL in the
warning message and place it in the directory
<tt class="docutils literal"><span class="pre">...site-packages/Bio/Entrez/DTDs</span></tt>, containing the other DTD files. If
you don’t have write access to this directory, you can also place the
DTD file in <tt class="docutils literal"><span class="pre">~/.biopython/Bio/Entrez/DTDs</span></tt>, where <tt class="docutils literal">~</tt> represents
your home directory. Since this directory is read before the directory
<tt class="docutils literal"><span class="pre">...site-packages/Bio/Entrez/DTDs</span></tt>, you can also put newer versions of
DTD files there if the ones in <tt class="docutils literal"><span class="pre">...site-packages/Bio/Entrez/DTDs</span></tt>
become outdated. Alternatively, if you installed Biopython from source,
you can add the DTD file to the source code’s <tt class="docutils literal">Bio/Entrez/DTDs</tt>
directory, and reinstall Biopython. This will install the new DTD file
in the correct location together with the other DTD files.</p>
<p>The Entrez Programming Utilities can also generate output in other
formats, such as the Fasta or GenBank file formats for sequence
databases, or the MedLine format for the literature database, discussed
in Section <a class="reference external" href="#sec:entrez-specialized-parsers">9.12</a>.</p>
<div class="section" id="entrez-guidelines">
<h2>9.1  Entrez Guidelines</h2>
<p>Before using Biopython to access the NCBI’s online resources (via
<tt class="docutils literal">Bio.Entrez</tt> or some of the other modules), please read the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/books/NBK25497/#chapter2.Usage_Guidelines_and_Requiremen">NCBI’s
Entrez User
Requirements</a>.
If the NCBI finds you are abusing their systems, they can and will ban
your access!</p>
<p>To paraphrase:</p>
<ul>
<li><p class="first">For any series of more than 100 requests, do this at weekends or
outside USA peak times. This is up to you to obey.</p>
</li>
<li><p class="first">Use the
<a class="reference external" href="http://eutils.ncbi.nlm.nih.gov">http://eutils.ncbi.nlm.nih.gov</a>
address, not the standard NCBI Web address. Biopython uses this web
address.</p>
</li>
<li><p class="first">Make no more than three requests every seconds (relaxed from at most
one request every three seconds in early 2009). This is automatically
enforced by Biopython.</p>
</li>
<li><p class="first">Use the optional email parameter so the NCBI can contact you if there
is a problem. You can either explicitly set this as a parameter with
each call to Entrez (e.g. include <tt class="docutils literal"><span class="pre">email=&quot;A.N.Other&#64;example.com&quot;</span></tt>
in the argument list), or you can set a global email address:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>
</pre>
<p><tt class="docutils literal">Bio.Entrez</tt> will then use this email address with each call to
Entrez. The <tt class="docutils literal">example.com</tt> address is a reserved domain name
specifically for documentation (RFC 2606). Please DO NOT use a random
email – it’s better not to give an email at all. The email parameter
will be mandatory from June 1, 2010. In case of excessive usage, NCBI
will attempt to contact a user at the e-mail address provided prior
to blocking access to the E-utilities.</p>
</li>
<li><p class="first">If you are using Biopython within some larger software suite, use the
tool parameter to specify this. You can either explicitly set the
tool name as a parameter with each call to Entrez (e.g. include
<tt class="docutils literal"><span class="pre">tool=&quot;MyLocalScript&quot;</span></tt> in the argument list), or you can set a
global tool name:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">tool</span> <span class="operator">=</span> <span class="literal string">&quot;MyLocalScript&quot;</span>
</pre>
<p>The tool parameter will default to Biopython.</p>
</li>
<li><p class="first">For large queries, the NCBI also recommend using their session
history feature (the WebEnv session cookie string, see
Section <a class="reference external" href="#sec:entrez-webenv">9.15</a>). This is only slightly more
complicated.</p>
</li>
</ul>
<p>In conclusion, be sensible with your usage levels. If you plan to
download lots of data, consider other options. For example, if you want
easy access to all the human genes, consider fetching each chromosome by
FTP as a GenBank file, and importing these into your own BioSQL database
(see Section <a class="reference external" href="#sec:BioSQL">18.5</a>).</p>
</div>
<div class="section" id="einfo-obtaining-information-about-the-entrez-databases">
<h2>9.2  EInfo: Obtaining information about the Entrez databases</h2>
<p>EInfo provides field index term counts, last update, and available links
for each of NCBI’s databases. In addition, you can use EInfo to obtain a
list of all database names accessible through the Entrez utilities:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">einfo</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
</pre>
<p>The variable <tt class="docutils literal">result</tt> now contains a list of databases in XML format:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">result</span>
<span class="operator">&lt;</span><span class="error">?</span><span class="name">xml</span> <span class="name">version</span><span class="operator">=</span><span class="literal string">&quot;1.0&quot;</span><span class="error">?</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="error">!</span><span class="name">DOCTYPE</span> <span class="name">eInfoResult</span> <span class="name">PUBLIC</span> <span class="literal string">&quot;-//NLM//DTD eInfoResult, 11 May 2002//EN&quot;</span>
 <span class="literal string">&quot;http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd&quot;</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="name">eInfoResult</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="name">DbList</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pubmed</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">protein</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nucleotide</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nuccore</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nucgss</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nucest</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">structure</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">genome</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">books</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">cancerchromosomes</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">cdd</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">gap</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">domains</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">gene</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">genomeprj</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">gensat</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">geo</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">gds</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">homologene</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">journals</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">mesh</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">ncbisearch</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nlmcatalog</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">omia</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">omim</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pmc</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">popset</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">probe</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">proteinclusters</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pcassay</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pccompound</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pcsubstance</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">snp</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">taxonomy</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">toolkit</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">unigene</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">unists</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
<span class="operator">&lt;/</span><span class="name">DbList</span><span class="operator">&gt;</span>
<span class="operator">&lt;/</span><span class="name">eInfoResult</span><span class="operator">&gt;</span>
</pre>
<p>Since this is a fairly simple XML file, we could extract the information
it contains simply by string searching. Using <tt class="docutils literal">Bio.Entrez</tt>’s parser
instead, we can directly parse this XML file into a Python object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">einfo</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>Now <tt class="docutils literal">record</tt> is a dictionary with exactly one key:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">u'DbList'</span><span class="punctuation">]</span>
</pre>
<p>The values stored in this key is the list of database names shown in the
XML above:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DbList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'pubmed'</span><span class="punctuation">,</span> <span class="literal string">'protein'</span><span class="punctuation">,</span> <span class="literal string">'nucleotide'</span><span class="punctuation">,</span> <span class="literal string">'nuccore'</span><span class="punctuation">,</span> <span class="literal string">'nucgss'</span><span class="punctuation">,</span> <span class="literal string">'nucest'</span><span class="punctuation">,</span>
 <span class="literal string">'structure'</span><span class="punctuation">,</span> <span class="literal string">'genome'</span><span class="punctuation">,</span> <span class="literal string">'books'</span><span class="punctuation">,</span> <span class="literal string">'cancerchromosomes'</span><span class="punctuation">,</span> <span class="literal string">'cdd'</span><span class="punctuation">,</span> <span class="literal string">'gap'</span><span class="punctuation">,</span>
 <span class="literal string">'domains'</span><span class="punctuation">,</span> <span class="literal string">'gene'</span><span class="punctuation">,</span> <span class="literal string">'genomeprj'</span><span class="punctuation">,</span> <span class="literal string">'gensat'</span><span class="punctuation">,</span> <span class="literal string">'geo'</span><span class="punctuation">,</span> <span class="literal string">'gds'</span><span class="punctuation">,</span> <span class="literal string">'homologene'</span><span class="punctuation">,</span>
 <span class="literal string">'journals'</span><span class="punctuation">,</span> <span class="literal string">'mesh'</span><span class="punctuation">,</span> <span class="literal string">'ncbisearch'</span><span class="punctuation">,</span> <span class="literal string">'nlmcatalog'</span><span class="punctuation">,</span> <span class="literal string">'omia'</span><span class="punctuation">,</span> <span class="literal string">'omim'</span><span class="punctuation">,</span> <span class="literal string">'pmc'</span><span class="punctuation">,</span>
 <span class="literal string">'popset'</span><span class="punctuation">,</span> <span class="literal string">'probe'</span><span class="punctuation">,</span> <span class="literal string">'proteinclusters'</span><span class="punctuation">,</span> <span class="literal string">'pcassay'</span><span class="punctuation">,</span> <span class="literal string">'pccompound'</span><span class="punctuation">,</span>
 <span class="literal string">'pcsubstance'</span><span class="punctuation">,</span> <span class="literal string">'snp'</span><span class="punctuation">,</span> <span class="literal string">'taxonomy'</span><span class="punctuation">,</span> <span class="literal string">'toolkit'</span><span class="punctuation">,</span> <span class="literal string">'unigene'</span><span class="punctuation">,</span> <span class="literal string">'unists'</span><span class="punctuation">]</span>
</pre>
<p>For each of these databases, we can use EInfo again to obtain more
information:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">einfo</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DbInfo&quot;</span><span class="punctuation">][</span><span class="literal string">&quot;Description&quot;</span><span class="punctuation">]</span>
<span class="literal string">'PubMed bibliographic record'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DbInfo&quot;</span><span class="punctuation">][</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal string">'17989604'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DbInfo&quot;</span><span class="punctuation">][</span><span class="literal string">&quot;LastUpdate&quot;</span><span class="punctuation">]</span>
<span class="literal string">'2008/05/24 06:45'</span>
</pre>
<p>Try <tt class="docutils literal"><span class="pre">record[&quot;DbInfo&quot;].keys()</span></tt> for other information stored in this
record. One of the most useful is a list of possible search fields for
use with ESearch:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">field</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DbInfo&quot;</span><span class="punctuation">][</span><span class="literal string">&quot;FieldList&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%(Name)s</span><span class="literal string">, </span><span class="literal string interpol">%(FullName)s</span><span class="literal string">, </span><span class="literal string interpol">%(Description)s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">field</span>
<span class="name">ALL</span><span class="punctuation">,</span> <span class="name">All</span> <span class="name">Fields</span><span class="punctuation">,</span> <span class="name">All</span> <span class="name">terms</span> <span class="keyword namespace">from</span> <span class="name namespace">all</span> <span class="name namespace">searchable</span> <span class="name namespace">fields</span>
<span class="name">UID</span><span class="punctuation">,</span> <span class="name">UID</span><span class="punctuation">,</span> <span class="name">Unique</span> <span class="name">number</span> <span class="name">assigned</span> <span class="name">to</span> <span class="name">publication</span>
<span class="name">FILT</span><span class="punctuation">,</span> <span class="name">Filter</span><span class="punctuation">,</span> <span class="name">Limits</span> <span class="name">the</span> <span class="name">records</span>
<span class="name">TITL</span><span class="punctuation">,</span> <span class="name">Title</span><span class="punctuation">,</span> <span class="name">Words</span> <span class="operator word">in</span> <span class="name">title</span> <span class="name">of</span> <span class="name">publication</span>
<span class="name">WORD</span><span class="punctuation">,</span> <span class="name">Text</span> <span class="name">Word</span><span class="punctuation">,</span> <span class="name">Free</span> <span class="name">text</span> <span class="name">associated</span> <span class="keyword">with</span> <span class="name">publication</span>
<span class="name">MESH</span><span class="punctuation">,</span> <span class="name">MeSH</span> <span class="name">Terms</span><span class="punctuation">,</span> <span class="name">Medical</span> <span class="name">Subject</span> <span class="name">Headings</span> <span class="name">assigned</span> <span class="name">to</span> <span class="name">publication</span>
<span class="name">MAJR</span><span class="punctuation">,</span> <span class="name">MeSH</span> <span class="name">Major</span> <span class="name">Topic</span><span class="punctuation">,</span> <span class="name">MeSH</span> <span class="name">terms</span> <span class="name">of</span> <span class="name">major</span> <span class="name">importance</span> <span class="name">to</span> <span class="name">publication</span>
<span class="name">AUTH</span><span class="punctuation">,</span> <span class="name">Author</span><span class="punctuation">,</span> <span class="name">Author</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span> <span class="name">of</span> <span class="name">publication</span>
<span class="name">JOUR</span><span class="punctuation">,</span> <span class="name">Journal</span><span class="punctuation">,</span> <span class="name">Journal</span> <span class="name">abbreviation</span> <span class="name">of</span> <span class="name">publication</span>
<span class="name">AFFL</span><span class="punctuation">,</span> <span class="name">Affiliation</span><span class="punctuation">,</span> <span class="name">Author</span><span class="literal string">'s institutional affiliation and address</span>
<span class="operator">...</span>
</pre>
<p>That’s a long list, but indirectly this tells you that for the PubMed
database, you can do things like <tt class="docutils literal">Jones[AUTH]</tt> to search the author
field, or <tt class="docutils literal">Sanger[AFFL]</tt> to restrict to authors at the Sanger Centre.
This can be very handy - especially if you are not so familiar with a
particular database.</p>
</div>
<div class="section" id="esearch-searching-the-entrez-databases">
<h2>9.3  ESearch: Searching the Entrez databases</h2>
<p>To search any of these databases, we use <tt class="docutils literal">Bio.Entrez.esearch()</tt>. For
example, let’s search in PubMed for publications related to Biopython:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;biopython&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'19304878'</span><span class="punctuation">,</span> <span class="literal string">'18606172'</span><span class="punctuation">,</span> <span class="literal string">'16403221'</span><span class="punctuation">,</span> <span class="literal string">'16377612'</span><span class="punctuation">,</span> <span class="literal string">'14871861'</span><span class="punctuation">,</span> <span class="literal string">'14630660'</span><span class="punctuation">,</span> <span class="literal string">'12230038'</span><span class="punctuation">]</span>
</pre>
<p>In this output, you see seven PubMed IDs (including 19304878 which is
the PMID for the Biopython application note), which can be retrieved by
EFetch (see section <a class="reference external" href="#sec:efetch">9.6</a>).</p>
<p>You can also use ESearch to search GenBank. Here we’ll do a quick search
for the <em>matK</em> gene in <em>Cypripedioideae</em> orchids (see
Section <a class="reference external" href="#sec:entrez-einfo">9.2</a> about EInfo for one way to find
out which fields you can search in each Entrez database):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Cypripedioideae[Orgn] AND matK[Gene]&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal string">'25'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'126789333'</span><span class="punctuation">,</span> <span class="literal string">'37222967'</span><span class="punctuation">,</span> <span class="literal string">'37222966'</span><span class="punctuation">,</span> <span class="literal string">'37222965'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'61585492'</span><span class="punctuation">]</span>
</pre>
<p>Each of the IDs (126789333, 37222967, 37222966, …) is a GenBank
identifier. See section <a class="reference external" href="#sec:efetch">9.6</a> for information on how
to actually download these GenBank records.</p>
<p>Note that instead of a species name like <tt class="docutils literal">Cypripedioideae[Orgn]</tt>, you
can restrict the search using an NCBI taxon identifier, here this would
be <tt class="docutils literal">txid158330[Orgn]</tt>. This isn’t currently documented on the ESearch
help page - the NCBI explained this in reply to an email query. You can
often deduce the search term formatting by playing with the Entrez web
interface. For example, including <tt class="docutils literal">complete[prop]</tt> in a genome search
restricts to just completed genomes.</p>
<p>As a final example, let’s get a list of computational journal titles:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;journals&quot;</span><span class="punctuation">,</span> <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;computational&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal string">'16'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'30367'</span><span class="punctuation">,</span> <span class="literal string">'33843'</span><span class="punctuation">,</span> <span class="literal string">'33823'</span><span class="punctuation">,</span> <span class="literal string">'32989'</span><span class="punctuation">,</span> <span class="literal string">'33190'</span><span class="punctuation">,</span> <span class="literal string">'33009'</span><span class="punctuation">,</span> <span class="literal string">'31986'</span><span class="punctuation">,</span>
 <span class="literal string">'34502'</span><span class="punctuation">,</span> <span class="literal string">'8799'</span><span class="punctuation">,</span> <span class="literal string">'22857'</span><span class="punctuation">,</span> <span class="literal string">'32675'</span><span class="punctuation">,</span> <span class="literal string">'20258'</span><span class="punctuation">,</span> <span class="literal string">'33859'</span><span class="punctuation">,</span> <span class="literal string">'32534'</span><span class="punctuation">,</span>
 <span class="literal string">'32357'</span><span class="punctuation">,</span> <span class="literal string">'32249'</span><span class="punctuation">]</span>
</pre>
<p>Again, we could use EFetch to obtain more information for each of these
journal IDs.</p>
<p>ESearch has many useful options — see the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/esearch_help.html">ESearch help
page</a>
for more information.</p>
</div>
<div class="section" id="epost-uploading-a-list-of-identifiers">
<h2>9.4  EPost: Uploading a list of identifiers</h2>
<p>EPost uploads a list of UIs for use in subsequent search strategies; see
the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/epost_help.html">EPost help
page</a>
for more information. It is available from Biopython through the
<tt class="docutils literal">Bio.Entrez.epost()</tt> function.</p>
<p>To give an example of when this is useful, suppose you have a long list
of IDs you want to download using EFetch (maybe sequences, maybe
citations – anything). When you make a request with EFetch your list of
IDs, the database etc, are all turned into a long URL sent to the
server. If your list of IDs is long, this URL gets long, and long URLs
can break (e.g. some proxies don’t cope well).</p>
<p>Instead, you can break this up into two steps, first uploading the list
of IDs using EPost (this uses an “HTML post” internally, rather than an
“HTML get”, getting round the long URL problem). With the history
support, you can then refer to this long list of IDs, and download the
associated data with EFetch.</p>
<p>Let’s look at a simple example to see how EPost works – uploading some
PubMed identifiers:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">id_list</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">&quot;19304878&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;18606172&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;16403221&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;16377612&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;14871861&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;14630660&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">epost</span><span class="punctuation">(</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;,&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">id_list</span><span class="punctuation">))</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&lt;</span><span class="error">?</span><span class="name">xml</span> <span class="name">version</span><span class="operator">=</span><span class="literal string">&quot;1.0&quot;</span><span class="error">?</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="error">!</span><span class="name">DOCTYPE</span> <span class="name">ePostResult</span> <span class="name">PUBLIC</span> <span class="literal string">&quot;-//NLM//DTD ePostResult, 11 May 2002//EN&quot;</span>
 <span class="literal string">&quot;http://www.ncbi.nlm.nih.gov/entrez/query/DTD/ePost_020511.dtd&quot;</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="name">ePostResult</span><span class="operator">&gt;</span>
 <span class="operator">&lt;</span><span class="name">QueryKey</span><span class="operator">&gt;</span><span class="literal number integer">1</span><span class="operator">&lt;/</span><span class="name">QueryKey</span><span class="operator">&gt;</span>
 <span class="operator">&lt;</span><span class="name">WebEnv</span><span class="operator">&gt;</span><span class="name">NCID_01_206841095_130</span><span class="operator">.</span><span class="literal number float">14.22</span><span class="operator">.</span><span class="literal number integer">101</span><span class="name">_9001_1242061629</span><span class="operator">&lt;/</span><span class="name">WebEnv</span><span class="operator">&gt;</span>
<span class="operator">&lt;/</span><span class="name">ePostResult</span><span class="operator">&gt;</span>
</pre>
<p>The returned XML includes two important strings, <tt class="docutils literal">QueryKey</tt> and
<tt class="docutils literal">WebEnv</tt> which together define your history session. You would extract
these values for use with another Entrez call such as EFetch:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">id_list</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">&quot;19304878&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;18606172&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;16403221&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;16377612&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;14871861&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;14630660&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">search_results</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">epost</span><span class="punctuation">(</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;,&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">id_list</span><span class="punctuation">)))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">webenv</span> <span class="operator">=</span> <span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;WebEnv&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">query_key</span> <span class="operator">=</span> <span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;QueryKey&quot;</span><span class="punctuation">]</span>
</pre>
<p>Section <a class="reference external" href="#sec:entrez-webenv">9.15</a> shows how to use the history
feature.</p>
</div>
<div class="section" id="esummary-retrieving-summaries-from-primary-ids">
<h2>9.5  ESummary: Retrieving summaries from primary IDs</h2>
<p>ESummary retrieves document summaries from a list of primary IDs (see
the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/esummary_help.html">ESummary help
page</a>
for more information). In Biopython, ESummary is available as
<tt class="docutils literal">Bio.Entrez.esummary()</tt>. Using the search result above, we can for
example find out more about the journal with ID 30367:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esummary</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;journals&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;30367&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Id&quot;</span><span class="punctuation">]</span>
<span class="literal string">'30367'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Title&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Computational biology and chemistry'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Publisher&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Pergamon,'</span>
</pre>
</div>
<div class="section" id="efetch-downloading-full-records-from-entrez">
<h2>9.6  EFetch: Downloading full records from Entrez</h2>
<p>EFetch is what you use when you want to retrieve a full record from
Entrez. This covers several possible databases, as described on the main
<a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/entrez/query/static/efetch_help.html">EFetch Help
page</a>.</p>
<p>For most of their databases, the NCBI support several different file
formats. Requesting a specific file format from Entrez using
<tt class="docutils literal">Bio.Entrez.efetch()</tt> requires specifying the <tt class="docutils literal">rettype</tt> and/or
<tt class="docutils literal">retmode</tt> optional arguments. The different combinations are described
for each database type on the pages linked to on <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/efetch_help.html">NCBI efetch
webpage</a>
(e.g.
<a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/corehtml/query/static/efetchlit_help.html">literature</a>,
<a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/corehtml/query/static/efetchseq_help.html">sequences</a>
and
<a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/corehtml/query/static/efetchtax_help.html">taxonomy</a>).</p>
<p>One common usage is downloading sequences in the FASTA or
GenBank/GenPept plain text formats (which can then be parsed with
<tt class="docutils literal">Bio.SeqIO</tt>, see Sections <a class="reference external" href="#sec:SeqIO_GenBank_Online">5.3.1</a>
and <a class="reference external" href="#sec:efetch">9.6</a>). From the <em>Cypripedioideae</em> example above,
we can download GenBank record 186972394 using <tt class="docutils literal">Bio.Entrez.efetch</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;186972394&quot;</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="name">LOCUS</span>       <span class="name">EU490707</span>                <span class="literal number integer">1302</span> <span class="name">bp</span>    <span class="name">DNA</span>     <span class="name">linear</span>   <span class="name">PLN</span> <span class="literal number oct">05</span><span class="operator">-</span><span class="name">MAY</span><span class="operator">-</span><span class="literal number integer">2008</span>
<span class="name">DEFINITION</span>  <span class="name">Selenipedium</span> <span class="name">aequinoctiale</span> <span class="name">maturase</span> <span class="name">K</span> <span class="punctuation">(</span><span class="name">matK</span><span class="punctuation">)</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">partial</span> <span class="name">cds</span><span class="punctuation">;</span>
            <span class="name">chloroplast</span><span class="operator">.</span>
<span class="name">ACCESSION</span>   <span class="name">EU490707</span>
<span class="name">VERSION</span>     <span class="name">EU490707</span><span class="operator">.</span><span class="literal number integer">1</span>  <span class="name">GI</span><span class="punctuation">:</span><span class="literal number integer">186972394</span>
<span class="name">KEYWORDS</span>    <span class="operator">.</span>
<span class="name">SOURCE</span>      <span class="name">chloroplast</span> <span class="name">Selenipedium</span> <span class="name">aequinoctiale</span>
  <span class="name">ORGANISM</span>  <span class="name">Selenipedium</span> <span class="name">aequinoctiale</span>
            <span class="name">Eukaryota</span><span class="punctuation">;</span> <span class="name">Viridiplantae</span><span class="punctuation">;</span> <span class="name">Streptophyta</span><span class="punctuation">;</span> <span class="name">Embryophyta</span><span class="punctuation">;</span> <span class="name">Tracheophyta</span><span class="punctuation">;</span>
            <span class="name">Spermatophyta</span><span class="punctuation">;</span> <span class="name">Magnoliophyta</span><span class="punctuation">;</span> <span class="name">Liliopsida</span><span class="punctuation">;</span> <span class="name">Asparagales</span><span class="punctuation">;</span> <span class="name">Orchidaceae</span><span class="punctuation">;</span>
            <span class="name">Cypripedioideae</span><span class="punctuation">;</span> <span class="name">Selenipedium</span><span class="operator">.</span>
<span class="name">REFERENCE</span>   <span class="literal number integer">1</span>  <span class="punctuation">(</span><span class="name">bases</span> <span class="literal number integer">1</span> <span class="name">to</span> <span class="literal number integer">1302</span><span class="punctuation">)</span>
  <span class="name">AUTHORS</span>   <span class="name">Neubig</span><span class="punctuation">,</span><span class="name">K</span><span class="operator">.</span><span class="name">M</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Whitten</span><span class="punctuation">,</span><span class="name">W</span><span class="operator">.</span><span class="name">M</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Carlsward</span><span class="punctuation">,</span><span class="name">B</span><span class="operator">.</span><span class="name">S</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Blanco</span><span class="punctuation">,</span><span class="name">M</span><span class="operator">.</span><span class="name">A</span><span class="operator">.</span><span class="punctuation">,</span>
            <span class="name">Endara</span><span class="punctuation">,</span><span class="name">C</span><span class="operator">.</span><span class="name">L</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Williams</span><span class="punctuation">,</span><span class="name">N</span><span class="operator">.</span><span class="name">H</span><span class="operator">.</span> <span class="operator word">and</span> <span class="name">Moore</span><span class="punctuation">,</span><span class="name">M</span><span class="operator">.</span><span class="name">J</span><span class="operator">.</span>
  <span class="name">TITLE</span>     <span class="name">Phylogenetic</span> <span class="name">utility</span> <span class="name">of</span> <span class="name">ycf1</span> <span class="operator word">in</span> <span class="name">orchids</span>
  <span class="name">JOURNAL</span>   <span class="name">Unpublished</span>
<span class="name">REFERENCE</span>   <span class="literal number integer">2</span>  <span class="punctuation">(</span><span class="name">bases</span> <span class="literal number integer">1</span> <span class="name">to</span> <span class="literal number integer">1302</span><span class="punctuation">)</span>
  <span class="name">AUTHORS</span>   <span class="name">Neubig</span><span class="punctuation">,</span><span class="name">K</span><span class="operator">.</span><span class="name">M</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Whitten</span><span class="punctuation">,</span><span class="name">W</span><span class="operator">.</span><span class="name">M</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Carlsward</span><span class="punctuation">,</span><span class="name">B</span><span class="operator">.</span><span class="name">S</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Blanco</span><span class="punctuation">,</span><span class="name">M</span><span class="operator">.</span><span class="name">A</span><span class="operator">.</span><span class="punctuation">,</span>
            <span class="name">Endara</span><span class="punctuation">,</span><span class="name">C</span><span class="operator">.</span><span class="name">L</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Williams</span><span class="punctuation">,</span><span class="name">N</span><span class="operator">.</span><span class="name">H</span><span class="operator">.</span> <span class="operator word">and</span> <span class="name">Moore</span><span class="punctuation">,</span><span class="name">M</span><span class="operator">.</span><span class="name">J</span><span class="operator">.</span>
  <span class="name">TITLE</span>     <span class="name">Direct</span> <span class="name">Submission</span>
  <span class="name">JOURNAL</span>   <span class="name">Submitted</span> <span class="punctuation">(</span><span class="literal number integer">14</span><span class="operator">-</span><span class="name">FEB</span><span class="operator">-</span><span class="literal number integer">2008</span><span class="punctuation">)</span> <span class="name">Department</span> <span class="name">of</span> <span class="name">Botany</span><span class="punctuation">,</span> <span class="name">University</span> <span class="name">of</span>
            <span class="name">Florida</span><span class="punctuation">,</span> <span class="literal number integer">220</span> <span class="name">Bartram</span> <span class="name">Hall</span><span class="punctuation">,</span> <span class="name">Gainesville</span><span class="punctuation">,</span> <span class="name">FL</span> <span class="literal number integer">32611</span><span class="operator">-</span><span class="literal number integer">8526</span><span class="punctuation">,</span> <span class="name">USA</span>
<span class="name">FEATURES</span>             <span class="name">Location</span><span class="operator">/</span><span class="name">Qualifiers</span>
     <span class="name">source</span>          <span class="literal number float">1.</span><span class="operator">.</span><span class="literal number integer">1302</span>
                     <span class="operator">/</span><span class="name">organism</span><span class="operator">=</span><span class="literal string">&quot;Selenipedium aequinoctiale&quot;</span>
                     <span class="operator">/</span><span class="name">organelle</span><span class="operator">=</span><span class="literal string">&quot;plastid:chloroplast&quot;</span>
                     <span class="operator">/</span><span class="name">mol_type</span><span class="operator">=</span><span class="literal string">&quot;genomic DNA&quot;</span>
                     <span class="operator">/</span><span class="name">specimen_voucher</span><span class="operator">=</span><span class="literal string">&quot;FLAS:Blanco 2475&quot;</span>
                     <span class="operator">/</span><span class="name">db_xref</span><span class="operator">=</span><span class="literal string">&quot;taxon:256374&quot;</span>
     <span class="name">gene</span>            <span class="operator">&lt;</span><span class="literal number float">1.</span><span class="operator">.&gt;</span><span class="literal number integer">1302</span>
                     <span class="operator">/</span><span class="name">gene</span><span class="operator">=</span><span class="literal string">&quot;matK&quot;</span>
     <span class="name">CDS</span>             <span class="operator">&lt;</span><span class="literal number float">1.</span><span class="operator">.&gt;</span><span class="literal number integer">1302</span>
                     <span class="operator">/</span><span class="name">gene</span><span class="operator">=</span><span class="literal string">&quot;matK&quot;</span>
                     <span class="operator">/</span><span class="name">codon_start</span><span class="operator">=</span><span class="literal number integer">1</span>
                     <span class="operator">/</span><span class="name">transl_table</span><span class="operator">=</span><span class="literal number integer">11</span>
                     <span class="operator">/</span><span class="name">product</span><span class="operator">=</span><span class="literal string">&quot;maturase K&quot;</span>
                     <span class="operator">/</span><span class="name">protein_id</span><span class="operator">=</span><span class="literal string">&quot;ACC99456.1&quot;</span>
                     <span class="operator">/</span><span class="name">db_xref</span><span class="operator">=</span><span class="literal string">&quot;GI:186972395&quot;</span>
                     <span class="operator">/</span><span class="name">translation</span><span class="operator">=</span><span class="literal string">&quot;IFYEPVEIFGYDNKSSLVLVKRLITRMYQQNFLISSVNDSNQKG</span>
                     <span class="name">FWGHKHFFSSHFSSQMVSEGFGVILEIPFSSQLVSSLEEKKIPKYQNLRSIHSIFPFL</span>
                     <span class="name">EDKFLHLNYVSDLLIPHPIHLEILVQILQCRIKDVPSLHLLRLLFHEYHNLNSLITSK</span>
                     <span class="name">KFIYAFSKRKKRFLWLLYNSYVYECEYLFQFLRKQSSYLRSTSSGVFLERTHLYVKIE</span>
                     <span class="name">HLLVVCCNSFQRILCFLKDPFMHYVRYQGKAILASKGTLILMKKWKFHLVNFWQSYFH</span>
                     <span class="name">FWSQPYRIHIKQLSNYSFSFLGYFSSVLENHLVVRNQMLENSFIINLLTKKFDTIAPV</span>
                     <span class="name">ISLIGSLSKAQFCTVLGHPISKPIWTDFSDSDILDRFCRICRNLCRYHSGSSKKQVLY</span>
                     <span class="name">RIKYILRLSCARTLARKHKSTVRTFMRRLGSGLLEEFFMEEE</span><span class="literal string">&quot;</span>
<span class="name">ORIGIN</span>
        <span class="literal number integer">1</span> <span class="name">attttttacg</span> <span class="name">aacctgtgga</span> <span class="name">aatttttggt</span> <span class="name">tatgacaata</span> <span class="name">aatctagttt</span> <span class="name">agtacttgtg</span>
       <span class="literal number integer">61</span> <span class="name">aaacgtttaa</span> <span class="name">ttactcgaat</span> <span class="name">gtatcaacag</span> <span class="name">aattttttga</span> <span class="name">tttcttcggt</span> <span class="name">taatgattct</span>
      <span class="literal number integer">121</span> <span class="name">aaccaaaaag</span> <span class="name">gattttgggg</span> <span class="name">gcacaagcat</span> <span class="name">tttttttctt</span> <span class="name">ctcatttttc</span> <span class="name">ttctcaaatg</span>
      <span class="literal number integer">181</span> <span class="name">gtatcagaag</span> <span class="name">gttttggagt</span> <span class="name">cattctggaa</span> <span class="name">attccattct</span> <span class="name">cgtcgcaatt</span> <span class="name">agtatcttct</span>
      <span class="literal number integer">241</span> <span class="name">cttgaagaaa</span> <span class="name">aaaaaatacc</span> <span class="name">aaaatatcag</span> <span class="name">aatttacgat</span> <span class="name">ctattcattc</span> <span class="name">aatatttccc</span>
      <span class="literal number integer">301</span> <span class="name">tttttagaag</span> <span class="name">acaaattttt</span> <span class="name">acatttgaat</span> <span class="name">tatgtgtcag</span> <span class="name">atctactaat</span> <span class="name">accccatccc</span>
      <span class="literal number integer">361</span> <span class="name">atccatctgg</span> <span class="name">aaatcttggt</span> <span class="name">tcaaatcctt</span> <span class="name">caatgccgga</span> <span class="name">tcaaggatgt</span> <span class="name">tccttctttg</span>
      <span class="literal number integer">421</span> <span class="name">catttattgc</span> <span class="name">gattgctttt</span> <span class="name">ccacgaatat</span> <span class="name">cataatttga</span> <span class="name">atagtctcat</span> <span class="name">tacttcaaag</span>
      <span class="literal number integer">481</span> <span class="name">aaattcattt</span> <span class="name">acgccttttc</span> <span class="name">aaaaagaaag</span> <span class="name">aaaagattcc</span> <span class="name">tttggttact</span> <span class="name">atataattct</span>
      <span class="literal number integer">541</span> <span class="name">tatgtatatg</span> <span class="name">aatgcgaata</span> <span class="name">tctattccag</span> <span class="name">tttcttcgta</span> <span class="name">aacagtcttc</span> <span class="name">ttatttacga</span>
      <span class="literal number integer">601</span> <span class="name">tcaacatctt</span> <span class="name">ctggagtctt</span> <span class="name">tcttgagcga</span> <span class="name">acacatttat</span> <span class="name">atgtaaaaat</span> <span class="name">agaacatctt</span>
      <span class="literal number integer">661</span> <span class="name">ctagtagtgt</span> <span class="name">gttgtaattc</span> <span class="name">ttttcagagg</span> <span class="name">atcctatgct</span> <span class="name">ttctcaagga</span> <span class="name">tcctttcatg</span>
      <span class="literal number integer">721</span> <span class="name">cattatgttc</span> <span class="name">gatatcaagg</span> <span class="name">aaaagcaatt</span> <span class="name">ctggcttcaa</span> <span class="name">agggaactct</span> <span class="name">tattctgatg</span>
      <span class="literal number integer">781</span> <span class="name">aagaaatgga</span> <span class="name">aatttcatct</span> <span class="name">tgtgaatttt</span> <span class="name">tggcaatctt</span> <span class="name">attttcactt</span> <span class="name">ttggtctcaa</span>
      <span class="literal number integer">841</span> <span class="name">ccgtatagga</span> <span class="name">ttcatataaa</span> <span class="name">gcaattatcc</span> <span class="name">aactattcct</span> <span class="name">tctcttttct</span> <span class="name">ggggtatttt</span>
      <span class="literal number integer">901</span> <span class="name">tcaagtgtac</span> <span class="name">tagaaaatca</span> <span class="name">tttggtagta</span> <span class="name">agaaatcaaa</span> <span class="name">tgctagagaa</span> <span class="name">ttcatttata</span>
      <span class="literal number integer">961</span> <span class="name">ataaatcttc</span> <span class="name">tgactaagaa</span> <span class="name">attcgatacc</span> <span class="name">atagccccag</span> <span class="name">ttatttctct</span> <span class="name">tattggatca</span>
     <span class="literal number integer">1021</span> <span class="name">ttgtcgaaag</span> <span class="name">ctcaattttg</span> <span class="name">tactgtattg</span> <span class="name">ggtcatccta</span> <span class="name">ttagtaaacc</span> <span class="name">gatctggacc</span>
     <span class="literal number integer">1081</span> <span class="name">gatttctcgg</span> <span class="name">attctgatat</span> <span class="name">tcttgatcga</span> <span class="name">ttttgccgga</span> <span class="name">tatgtagaaa</span> <span class="name">tctttgtcgt</span>
     <span class="literal number integer">1141</span> <span class="name">tatcacagcg</span> <span class="name">gatcctcaaa</span> <span class="name">aaaacaggtt</span> <span class="name">ttgtatcgta</span> <span class="name">taaaatatat</span> <span class="name">acttcgactt</span>
     <span class="literal number integer">1201</span> <span class="name">tcgtgtgcta</span> <span class="name">gaactttggc</span> <span class="name">acggaaacat</span> <span class="name">aaaagtacag</span> <span class="name">tacgcacttt</span> <span class="name">tatgcgaaga</span>
     <span class="literal number integer">1261</span> <span class="name">ttaggttcgg</span> <span class="name">gattattaga</span> <span class="name">agaattcttt</span> <span class="name">atggaagaag</span> <span class="name">aa</span>
<span class="operator">//</span>
</pre>
<p>The arguments <tt class="docutils literal"><span class="pre">rettype=&quot;gb&quot;</span></tt> and <tt class="docutils literal"><span class="pre">retmode=&quot;text&quot;</span></tt> let us download
this record in the GenBank format.</p>
<p>Note that until Easter 2009, the Entrez EFetch API let you use “genbank”
as the return type, however the NCBI now insist on using the official
return types of “gb” or “gbwithparts” (or “gp” for proteins) as
described on online. Also not that until Feb 2012, the Entrez EFetch API
would default to returning plain text files, but now defaults to XML.</p>
<p>Alternatively, you could for example use <tt class="docutils literal"><span class="pre">rettype=&quot;fasta&quot;</span></tt> to get the
Fasta-format; see the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/efetchseq_help.html">EFetch Sequences Help
page</a>
for other options. Remember – the available formats depend on which
database you are downloading from - see the main <a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/entrez/query/static/efetch_help.html">EFetch Help
page</a>.</p>
<p>If you fetch the record in one of the formats accepted by <tt class="docutils literal">Bio.SeqIO</tt>
(see Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>), you could directly parse it
into a <tt class="docutils literal">SeqRecord</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span><span class="punctuation">,</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;186972394&quot;</span><span class="punctuation">,</span><span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span>
<span class="name">ID</span><span class="punctuation">:</span> <span class="name">EU490707</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">Name</span><span class="punctuation">:</span> <span class="name">EU490707</span>
<span class="name">Description</span><span class="punctuation">:</span> <span class="name">Selenipedium</span> <span class="name">aequinoctiale</span> <span class="name">maturase</span> <span class="name">K</span> <span class="punctuation">(</span><span class="name">matK</span><span class="punctuation">)</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">partial</span> <span class="name">cds</span><span class="punctuation">;</span> <span class="name">chloroplast</span><span class="operator">.</span>
<span class="name">Number</span> <span class="name">of</span> <span class="name">features</span><span class="punctuation">:</span> <span class="literal number integer">3</span>
<span class="operator">...</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ATTTTTTACGAACCTGTGGAAATTTTTGGTTATGACAATAAATCTAGTTTAGTA...GAA'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>Note that a more typical use would be to save the sequence data to a
local file, and <em>then</em> parse it with <tt class="docutils literal">Bio.SeqIO</tt>. This can save you
having to re-download the same file repeatedly while working on your
script, and places less load on the NCBI’s servers. For example:</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">os</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="name">filename</span> <span class="operator">=</span> <span class="literal string">&quot;gi_186972394.gbk&quot;</span>
<span class="keyword">if</span> <span class="operator word">not</span> <span class="name">os</span><span class="operator">.</span><span class="name">path</span><span class="operator">.</span><span class="name">isfile</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">):</span>
    <span class="comment"># Downloading...</span>
    <span class="name">net_handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;186972394&quot;</span><span class="punctuation">,</span><span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
    <span class="name">out_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
    <span class="name">out_handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">net_handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">())</span>
    <span class="name">out_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
    <span class="name">net_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
    <span class="keyword">print</span> <span class="literal string">&quot;Saved&quot;</span>

<span class="keyword">print</span> <span class="literal string">&quot;Parsing...&quot;</span>
<span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name">record</span>
</pre>
<p>To get the output in XML format, which you can parse using the
<tt class="docutils literal">Bio.Entrez.read()</tt> function, use <tt class="docutils literal"><span class="pre">retmode=&quot;xml&quot;</span></tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;186972394&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;GBSeq_definition&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Selenipedium aequinoctiale maturase K (matK) gene, partial cds; chloroplast'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;GBSeq_source&quot;</span><span class="punctuation">]</span>
<span class="literal string">'chloroplast Selenipedium aequinoctiale'</span>
</pre>
<p>So, that dealt with sequences. For examples of parsing file formats
specific to the other databases (e.g. the <tt class="docutils literal">MEDLINE</tt> format used in
PubMed), see Section <a class="reference external" href="#sec:entrez-specialized-parsers">9.12</a>.</p>
<p>If you want to perform a search with <tt class="docutils literal">Bio.Entrez.esearch()</tt>, and then
download the records with <tt class="docutils literal">Bio.Entrez.efetch()</tt>, you should use the
WebEnv history feature – see Section <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
</div>
<div class="section" id="elink-searching-for-related-items-in-ncbi-entrez">
<h2>9.7  ELink: Searching for related items in NCBI Entrez</h2>
<p>ELink, available from Biopython as <tt class="docutils literal">Bio.Entrez.elink()</tt>, can be used
to find related items in the NCBI Entrez databases. For example, you can
us this to find nucleotide entries for an entry in the gene database,
and other cool stuff.</p>
<p>Let’s use ELink to find articles related to the Biopython application
note published in <em>Bioinformatics</em> in 2009. The PubMed ID of this
article is 19304878:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pmid</span> <span class="operator">=</span> <span class="literal string">&quot;19304878&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">elink</span><span class="punctuation">(</span><span class="name">dbfrom</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="name">pmid</span><span class="punctuation">))</span>
</pre>
<p>The <tt class="docutils literal">record</tt> variable consists of a Python list, one for each database
in which we searched. Since we specified only one PubMed ID to search
for, <tt class="docutils literal">record</tt> contains only one item. This item is a dictionary
containing information about our search term, as well as all the related
items that were found:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;DbFrom&quot;</span><span class="punctuation">]</span>
<span class="literal string">'pubmed'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'19304878'</span><span class="punctuation">]</span>
</pre>
<p>The <tt class="docutils literal">&quot;LinkSetDb&quot;</tt> key contains the search results, stored as a list
consisting of one item for each target database. In our search results,
we only find hits in the PubMed database (although sub-divided into
categories):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">])</span>
<span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">linksetdb</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">linksetdb</span><span class="punctuation">[</span><span class="literal string">&quot;DbTo&quot;</span><span class="punctuation">],</span> <span class="name">linksetdb</span><span class="punctuation">[</span><span class="literal string">&quot;LinkName&quot;</span><span class="punctuation">],</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">linksetdb</span><span class="punctuation">[</span><span class="literal string">&quot;Link&quot;</span><span class="punctuation">])</span>
<span class="operator">...</span>
<span class="name">pubmed</span> <span class="name">pubmed_pubmed</span> <span class="literal number integer">110</span>
<span class="name">pubmed</span> <span class="name">pubmed_pubmed_combined</span> <span class="literal number integer">6</span>
<span class="name">pubmed</span> <span class="name">pubmed_pubmed_five</span> <span class="literal number integer">6</span>
<span class="name">pubmed</span> <span class="name">pubmed_pubmed_reviews</span> <span class="literal number integer">5</span>
<span class="name">pubmed</span> <span class="name">pubmed_pubmed_reviews_five</span> <span class="literal number integer">5</span>
</pre>
<p>The actual search results are stored as under the <tt class="docutils literal">&quot;Link&quot;</tt> key. In
total, 110 items were found under standard search. Let’s now at the
first search result:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Link&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">u'Id'</span><span class="punctuation">:</span> <span class="literal string">'19304878'</span><span class="punctuation">}</span>
</pre>
<p>This is the article we searched for, which doesn’t help us much, so
let’s look at the second search result:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Link&quot;</span><span class="punctuation">][</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">u'Id'</span><span class="punctuation">:</span> <span class="literal string">'14630660'</span><span class="punctuation">}</span>
</pre>
<p>This paper, with PubMed ID 14630660, is about the Biopython PDB parser.</p>
<p>We can use a loop to print out all PubMed IDs:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">link</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Link&quot;</span><span class="punctuation">]</span> <span class="punctuation">:</span> <span class="keyword">print</span> <span class="name">link</span><span class="punctuation">[</span><span class="literal string">&quot;Id&quot;</span><span class="punctuation">]</span>
<span class="literal number integer">19304878</span>
<span class="literal number integer">14630660</span>
<span class="literal number integer">18689808</span>
<span class="literal number integer">17121776</span>
<span class="literal number integer">16377612</span>
<span class="literal number integer">12368254</span>
<span class="operator">......</span>
</pre>
<p>Now that was nice, but personally I am often more interested to find out
if a paper has been cited. Well, ELink can do that too – at least for
journals in Pubmed Central (see
Section <a class="reference external" href="#sec:elink-citations">9.15.3</a>).</p>
<p>For help on ELink, see the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/elink_help.html">ELink help
page</a>.
There is an entire sub-page just for the <a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/corehtml/query/static/entrezlinks.html">link
names</a>,
describing how different databases can be cross referenced.</p>
</div>
<div class="section" id="egquery-global-query-counts-for-search-terms">
<h2>9.8  EGQuery: Global Query - counts for search terms</h2>
<p>EGQuery provides counts for a search term in each of the Entrez
databases (i.e. a global query). This is particularly useful to find out
how many items your search terms would find in each database without
actually performing lots of separate searches with ESearch (see the
example in <a class="reference external" href="#subsec:entrez_example_genbank">9.14.2</a> below).</p>
<p>In this example, we use <tt class="docutils literal">Bio.Entrez.egquery()</tt> to obtain the counts
for “Biopython”:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">egquery</span><span class="punctuation">(</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;biopython&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">row</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;eGQueryResult&quot;</span><span class="punctuation">]:</span> <span class="keyword">print</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;DbName&quot;</span><span class="punctuation">],</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="operator">...</span>
<span class="name">pubmed</span> <span class="literal number integer">6</span>
<span class="name">pmc</span> <span class="literal number integer">62</span>
<span class="name">journals</span> <span class="literal number integer">0</span>
<span class="operator">...</span>
</pre>
<p>See the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/egquery_help.html">EGQuery help
page</a>
for more information.</p>
</div>
<div class="section" id="espell-obtaining-spelling-suggestions">
<h2>9.9  ESpell: Obtaining spelling suggestions</h2>
<p>ESpell retrieves spelling suggestions. In this example, we use
<tt class="docutils literal">Bio.Entrez.espell()</tt> to obtain the correct spelling of Biopython:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">espell</span><span class="punctuation">(</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;biopythooon&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;Query&quot;</span><span class="punctuation">]</span>
<span class="literal string">'biopythooon'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;CorrectedQuery&quot;</span><span class="punctuation">]</span>
<span class="literal string">'biopython'</span>
</pre>
<p>See the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/espell_help.html">ESpell help
page</a>
for more information. The main use of this is for GUI tools to provide
automatic suggestions for search terms.</p>
</div>
<div class="section" id="parsing-huge-entrez-xml-files">
<h2>9.10  Parsing huge Entrez XML files</h2>
<p>The <tt class="docutils literal">Entrez.read</tt> function reads the entire XML file returned by
Entrez into a single Python object, which is kept in memory. To parse
Entrez XML files too large to fit in memory, you can use the function
<tt class="docutils literal">Entrez.parse</tt>. This is a generator function that reads records in the
XML file one by one. This function is only useful if the XML file
reflects a Python list object (in other words, if <tt class="docutils literal">Entrez.read</tt> on a
computer with infinite memory resources would return a Python list).</p>
<p>For example, you can download the entire Entrez Gene database for a
given organism as a file from NCBI’s ftp site. These files can be very
large. As an example, on September 4, 2009, the file
<tt class="docutils literal">Homo_sapiens.ags.gz</tt>, containing the Entrez Gene database for human,
had a size of 116576 kB. This file, which is in the <tt class="docutils literal">ASN</tt> format, can
be converted into an XML file using NCBI’s <tt class="docutils literal">gene2xml</tt> program (see
NCBI’s ftp site for more information):</p>
<pre class="code python literal-block">
<span class="name">gene2xml</span> <span class="operator">-</span><span class="name">b</span> <span class="name">T</span> <span class="operator">-</span><span class="name">i</span> <span class="name">Homo_sapiens</span><span class="operator">.</span><span class="name">ags</span> <span class="operator">-</span><span class="name">o</span> <span class="name">Homo_sapiens</span><span class="operator">.</span><span class="name">xml</span>
</pre>
<p>The resulting XML file has a size of 6.1 GB. Attempting <tt class="docutils literal">Entrez.read</tt>
on this file will result in a <tt class="docutils literal">MemoryError</tt> on many computers.</p>
<p>The XML file <tt class="docutils literal">Homo_sapiens.xml</tt> consists of a list of Entrez gene
records, each corresponding to one Entrez gene in human.
<tt class="docutils literal">Entrez.parse</tt> retrieves these gene records one by one. You can then
print out or store the relevant information in each record by iterating
over the records. For example, this script iterates over the Entrez gene
records and prints out the gene numbers and names for all current genes:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;Homo_sapiens.xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">status</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">'Entrezgene_track-info'</span><span class="punctuation">][</span><span class="literal string">'Gene-track'</span><span class="punctuation">][</span><span class="literal string">'Gene-track_status'</span><span class="punctuation">]</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">status</span><span class="operator">.</span><span class="name">attributes</span><span class="punctuation">[</span><span class="literal string">'value'</span><span class="punctuation">]</span><span class="operator">==</span><span class="literal string">'discontinued'</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">continue</span>
<span class="operator">...</span>     <span class="name">geneid</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">'Entrezgene_track-info'</span><span class="punctuation">][</span><span class="literal string">'Gene-track'</span><span class="punctuation">][</span><span class="literal string">'Gene-track_geneid'</span><span class="punctuation">]</span>
<span class="operator">...</span>     <span class="name">genename</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">'Entrezgene_gene'</span><span class="punctuation">][</span><span class="literal string">'Gene-ref'</span><span class="punctuation">][</span><span class="literal string">'Gene-ref_locus'</span><span class="punctuation">]</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">geneid</span><span class="punctuation">,</span> <span class="name">genename</span>
</pre>
<p>This will print:</p>
<pre class="code python literal-block">
<span class="literal number integer">1</span> <span class="name">A1BG</span>
<span class="literal number integer">2</span> <span class="name">A2M</span>
<span class="literal number integer">3</span> <span class="name">A2MP</span>
<span class="literal number integer">8</span> <span class="name">AA</span>
<span class="literal number integer">9</span> <span class="name">NAT1</span>
<span class="literal number integer">10</span> <span class="name">NAT2</span>
<span class="literal number integer">11</span> <span class="name">AACP</span>
<span class="literal number integer">12</span> <span class="name">SERPINA3</span>
<span class="literal number integer">13</span> <span class="name">AADAC</span>
<span class="literal number integer">14</span> <span class="name">AAMP</span>
<span class="literal number integer">15</span> <span class="name">AANAT</span>
<span class="literal number integer">16</span> <span class="name">AARS</span>
<span class="literal number integer">17</span> <span class="name">AAVS1</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="handling-errors">
<h2>9.11  Handling errors</h2>
<p>Three things can go wrong when parsing an XML file:</p>
<ul class="simple">
<li>The file may not be an XML file to begin with;</li>
<li>The file may end prematurely or otherwise be corrupted;</li>
<li>The file may be correct XML, but contain items that are not
represented in the associated DTD.</li>
</ul>
<p>The first case occurs if, for example, you try to parse a Fasta file as
if it were an XML file:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.fna&quot;</span><span class="punctuation">)</span> <span class="comment"># a Fasta file</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
  <span class="name">File</span> <span class="literal string">&quot;&lt;stdin&gt;&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="operator">&lt;</span><span class="name">module</span><span class="operator">&gt;</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/__init__.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">257</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">read</span>
    <span class="name">record</span> <span class="operator">=</span> <span class="name">handler</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/Parser.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">164</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">read</span>
    <span class="keyword">raise</span> <span class="name">NotXMLError</span><span class="punctuation">(</span><span class="name">e</span><span class="punctuation">)</span>
<span class="name">Bio</span><span class="operator">.</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">Parser</span><span class="operator">.</span><span class="name">NotXMLError</span><span class="punctuation">:</span> <span class="name">Failed</span> <span class="name">to</span> <span class="name">parse</span> <span class="name">the</span> <span class="name">XML</span> <span class="name">data</span> <span class="punctuation">(</span><span class="name">syntax</span> <span class="name">error</span><span class="punctuation">:</span> <span class="name">line</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">column</span> <span class="literal number integer">0</span><span class="punctuation">)</span><span class="operator">.</span> <span class="name">Please</span> <span class="name">make</span> <span class="name">sure</span> <span class="name">that</span> <span class="name">the</span> <span class="name builtin">input</span> <span class="name">data</span> <span class="name">are</span> <span class="operator word">in</span> <span class="name">XML</span> <span class="name">format</span><span class="operator">.</span>
</pre>
<p>Here, the parser didn’t find the <tt class="docutils literal"><span class="pre">&lt;?xml</span> ...</tt> tag with which an XML
file is supposed to start, and therefore decides (correctly) that the
file is not an XML file.</p>
<p>When your file is in the XML format but is corrupted (for example, by
ending prematurely), the parser will raise a CorruptedXMLError. Here is
an example of an XML file that ends prematurely:</p>
<pre class="code python literal-block">
<span class="operator">&lt;</span><span class="error">?</span><span class="name">xml</span> <span class="name">version</span><span class="operator">=</span><span class="literal string">&quot;1.0&quot;</span><span class="error">?</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="error">!</span><span class="name">DOCTYPE</span> <span class="name">eInfoResult</span> <span class="name">PUBLIC</span> <span class="literal string">&quot;-//NLM//DTD eInfoResult, 11 May 2002//EN&quot;</span> <span class="literal string">&quot;http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd&quot;</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="name">eInfoResult</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="name">DbList</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pubmed</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">protein</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nucleotide</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nuccore</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nucgss</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nucest</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">structure</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">genome</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">books</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">cancerchromosomes</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">cdd</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
</pre>
<p>which will generate the following traceback:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
  <span class="name">File</span> <span class="literal string">&quot;&lt;stdin&gt;&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="operator">&lt;</span><span class="name">module</span><span class="operator">&gt;</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/__init__.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">257</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">read</span>
    <span class="name">record</span> <span class="operator">=</span> <span class="name">handler</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/Parser.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">160</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">read</span>
    <span class="keyword">raise</span> <span class="name">CorruptedXMLError</span><span class="punctuation">(</span><span class="name">e</span><span class="punctuation">)</span>
<span class="name">Bio</span><span class="operator">.</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">Parser</span><span class="operator">.</span><span class="name">CorruptedXMLError</span><span class="punctuation">:</span> <span class="name">Failed</span> <span class="name">to</span> <span class="name">parse</span> <span class="name">the</span> <span class="name">XML</span> <span class="name">data</span> <span class="punctuation">(</span><span class="name">no</span> <span class="name">element</span> <span class="name">found</span><span class="punctuation">:</span> <span class="name">line</span> <span class="literal number integer">16</span><span class="punctuation">,</span> <span class="name">column</span> <span class="literal number integer">0</span><span class="punctuation">)</span><span class="operator">.</span> <span class="name">Please</span> <span class="name">make</span> <span class="name">sure</span> <span class="name">that</span> <span class="name">the</span> <span class="name builtin">input</span> <span class="name">data</span> <span class="name">are</span> <span class="operator word">not</span> <span class="name">corrupted</span><span class="operator">.</span>

<span class="operator">&gt;&gt;&gt;</span>
</pre>
<p>Note that the error message tells you at what point in the XML file the
error was detected.</p>
<p>The third type of error occurs if the XML file contains tags that do not
have a description in the corresponding DTD file. This is an example of
such an XML file:</p>
<pre class="code python literal-block">
<span class="operator">&lt;</span><span class="error">?</span><span class="name">xml</span> <span class="name">version</span><span class="operator">=</span><span class="literal string">&quot;1.0&quot;</span><span class="error">?</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="error">!</span><span class="name">DOCTYPE</span> <span class="name">eInfoResult</span> <span class="name">PUBLIC</span> <span class="literal string">&quot;-//NLM//DTD eInfoResult, 11 May 2002//EN&quot;</span> <span class="literal string">&quot;http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd&quot;</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="name">eInfoResult</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbInfo</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pubmed</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">MenuName</span><span class="operator">&gt;</span><span class="name">PubMed</span><span class="operator">&lt;/</span><span class="name">MenuName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">Description</span><span class="operator">&gt;</span><span class="name">PubMed</span> <span class="name">bibliographic</span> <span class="name">record</span><span class="operator">&lt;/</span><span class="name">Description</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">Count</span><span class="operator">&gt;</span><span class="literal number integer">20161961</span><span class="operator">&lt;/</span><span class="name">Count</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">LastUpdate</span><span class="operator">&gt;</span><span class="literal number integer">2010</span><span class="operator">/</span><span class="literal number integer">09</span><span class="operator">/</span><span class="literal number integer">10</span> <span class="literal number oct">04</span><span class="punctuation">:</span><span class="literal number integer">52</span><span class="operator">&lt;/</span><span class="name">LastUpdate</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">FieldList</span><span class="operator">&gt;</span>
                <span class="operator">&lt;</span><span class="name">Field</span><span class="operator">&gt;</span>
<span class="operator">...</span>
                <span class="operator">&lt;/</span><span class="name">Field</span><span class="operator">&gt;</span>
        <span class="operator">&lt;/</span><span class="name">FieldList</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DocsumList</span><span class="operator">&gt;</span>
                <span class="operator">&lt;</span><span class="name">Docsum</span><span class="operator">&gt;</span>
                        <span class="operator">&lt;</span><span class="name">DsName</span><span class="operator">&gt;</span><span class="name">PubDate</span><span class="operator">&lt;/</span><span class="name">DsName</span><span class="operator">&gt;</span>
                        <span class="operator">&lt;</span><span class="name">DsType</span><span class="operator">&gt;</span><span class="literal number integer">4</span><span class="operator">&lt;/</span><span class="name">DsType</span><span class="operator">&gt;</span>
                        <span class="operator">&lt;</span><span class="name">DsTypeName</span><span class="operator">&gt;</span><span class="name">string</span><span class="operator">&lt;/</span><span class="name">DsTypeName</span><span class="operator">&gt;</span>
                <span class="operator">&lt;/</span><span class="name">Docsum</span><span class="operator">&gt;</span>
                <span class="operator">&lt;</span><span class="name">Docsum</span><span class="operator">&gt;</span>
                        <span class="operator">&lt;</span><span class="name">DsName</span><span class="operator">&gt;</span><span class="name">EPubDate</span><span class="operator">&lt;/</span><span class="name">DsName</span><span class="operator">&gt;</span>
<span class="operator">...</span>
        <span class="operator">&lt;/</span><span class="name">DbInfo</span><span class="operator">&gt;</span>
<span class="operator">&lt;/</span><span class="name">eInfoResult</span><span class="operator">&gt;</span>
</pre>
<p>In this file, for some reason the tag <tt class="docutils literal">&lt;DocsumList&gt;</tt> (and several
others) are not listed in the DTD file <tt class="docutils literal">eInfo_020511.dtd</tt>, which is
specified on the second line as the DTD for this XML file. By default,
the parser will stop and raise a ValidationError if it cannot find some
tag in the DTD:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;einfo3.xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
  <span class="name">File</span> <span class="literal string">&quot;&lt;stdin&gt;&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="operator">&lt;</span><span class="name">module</span><span class="operator">&gt;</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/__init__.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">257</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">read</span>
    <span class="name">record</span> <span class="operator">=</span> <span class="name">handler</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/Parser.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">154</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">read</span>
    <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">parser</span><span class="operator">.</span><span class="name">ParseFile</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/Parser.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">246</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">startElementHandler</span>
    <span class="keyword">raise</span> <span class="name">ValidationError</span><span class="punctuation">(</span><span class="name">name</span><span class="punctuation">)</span>
<span class="name">Bio</span><span class="operator">.</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">Parser</span><span class="operator">.</span><span class="name">ValidationError</span><span class="punctuation">:</span> <span class="name">Failed</span> <span class="name">to</span> <span class="name">find</span> <span class="name">tag</span> <span class="literal string">'DocsumList'</span> <span class="operator word">in</span> <span class="name">the</span> <span class="name">DTD</span><span class="operator">.</span> <span class="name">To</span> <span class="name">skip</span> <span class="name builtin">all</span> <span class="name">tags</span> <span class="name">that</span> <span class="name">are</span> <span class="operator word">not</span> <span class="name">represented</span> <span class="operator word">in</span> <span class="name">the</span> <span class="name">DTD</span><span class="punctuation">,</span> <span class="name">please</span> <span class="name">call</span> <span class="name">Bio</span><span class="operator">.</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span> <span class="operator word">or</span> <span class="name">Bio</span><span class="operator">.</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">parse</span> <span class="keyword">with</span> <span class="name">validate</span><span class="operator">=</span><span class="name builtin pseudo">False</span><span class="operator">.</span>
</pre>
<p>Optionally, you can instruct the parser to skip such tags instead of
raising a ValidationError. This is done by calling <tt class="docutils literal">Entrez.read</tt> or
<tt class="docutils literal">Entrez.parse</tt> with the argument <tt class="docutils literal">validate</tt> equal to False:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;einfo3.xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span><span class="name">validate</span><span class="operator">=</span><span class="name builtin pseudo">False</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span>
</pre>
<p>Of course, the information contained in the XML tags that are not in the
DTD are not present in the record returned by <tt class="docutils literal">Entrez.read</tt>.</p>
</div>
<div class="section" id="specialized-parsers">
<h2>9.12  Specialized parsers</h2>
<p>The <tt class="docutils literal">Bio.Entrez.read()</tt> function can parse most (if not all) XML
output returned by Entrez. Entrez typically allows you to retrieve
records in other formats, which may have some advantages compared to the
XML format in terms of readability (or download size).</p>
<p>To request a specific file format from Entrez using
<tt class="docutils literal">Bio.Entrez.efetch()</tt> requires specifying the <tt class="docutils literal">rettype</tt> and/or
<tt class="docutils literal">retmode</tt> optional arguments. The different combinations are described
for each database type on the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/efetch_help.html">NCBI efetch
webpage</a>.</p>
<p>One obvious case is you may prefer to download sequences in the FASTA or
GenBank/GenPept plain text formats (which can then be parsed with
<tt class="docutils literal">Bio.SeqIO</tt>, see Sections <a class="reference external" href="#sec:SeqIO_GenBank_Online">5.3.1</a>
and <a class="reference external" href="#sec:efetch">9.6</a>). For the literature databases, Biopython
contains a parser for the <tt class="docutils literal">MEDLINE</tt> format used in PubMed.</p>
<div class="section" id="parsing-medline-records">
<h3>9.12.1  Parsing Medline records</h3>
<p>You can find the Medline parser in <tt class="docutils literal">Bio.Medline</tt>. Suppose we want to
parse the file <tt class="docutils literal">pubmed_result1.txt</tt>, containing one Medline record.
You can find this file in Biopython’s <tt class="docutils literal">Tests\Medline</tt> directory. The
file looks like this:</p>
<pre class="code python literal-block">
<span class="name">PMID</span><span class="operator">-</span> <span class="literal number integer">12230038</span>
<span class="name">OWN</span> <span class="operator">-</span> <span class="name">NLM</span>
<span class="name">STAT</span><span class="operator">-</span> <span class="name">MEDLINE</span>
<span class="name">DA</span>  <span class="operator">-</span> <span class="literal number integer">20020916</span>
<span class="name">DCOM</span><span class="operator">-</span> <span class="literal number integer">20030606</span>
<span class="name">LR</span>  <span class="operator">-</span> <span class="literal number integer">20041117</span>
<span class="name">PUBM</span><span class="operator">-</span> <span class="name">Print</span>
<span class="name">IS</span>  <span class="operator">-</span> <span class="literal number integer">1467</span><span class="operator">-</span><span class="literal number integer">5463</span> <span class="punctuation">(</span><span class="name">Print</span><span class="punctuation">)</span>
<span class="name">VI</span>  <span class="operator">-</span> <span class="literal number integer">3</span>
<span class="name">IP</span>  <span class="operator">-</span> <span class="literal number integer">3</span>
<span class="name">DP</span>  <span class="operator">-</span> <span class="literal number integer">2002</span> <span class="name">Sep</span>
<span class="name">TI</span>  <span class="operator">-</span> <span class="name">The</span> <span class="name">Bio</span><span class="operator">*</span> <span class="name">toolkits</span><span class="operator">--</span><span class="name">a</span> <span class="name">brief</span> <span class="name">overview</span><span class="operator">.</span>
<span class="name">PG</span>  <span class="operator">-</span> <span class="literal number integer">296</span><span class="operator">-</span><span class="literal number integer">302</span>
<span class="name">AB</span>  <span class="operator">-</span> <span class="name">Bioinformatics</span> <span class="name">research</span> <span class="operator word">is</span> <span class="name">often</span> <span class="name">difficult</span> <span class="name">to</span> <span class="name">do</span> <span class="keyword">with</span> <span class="name">commercial</span> <span class="name">software</span><span class="operator">.</span> <span class="name">The</span>
      <span class="name">Open</span> <span class="name">Source</span> <span class="name">BioPerl</span><span class="punctuation">,</span> <span class="name">BioPython</span> <span class="operator word">and</span> <span class="name">Biojava</span> <span class="name">projects</span> <span class="name">provide</span> <span class="name">toolkits</span> <span class="keyword">with</span>
<span class="operator">...</span>
</pre>
<p>We first open the file and then parse it:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Medline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">input</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;pubmed_result1.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Medline</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="punctuation">)</span>
</pre>
<p>The <tt class="docutils literal">record</tt> now contains the Medline record as a Python dictionary:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;PMID&quot;</span><span class="punctuation">]</span>
<span class="literal string">'12230038'</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;AB&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Bioinformatics research is often difficult to do with commercial software.</span>
<span class="name">The</span> <span class="name">Open</span> <span class="name">Source</span> <span class="name">BioPerl</span><span class="punctuation">,</span> <span class="name">BioPython</span> <span class="operator word">and</span> <span class="name">Biojava</span> <span class="name">projects</span> <span class="name">provide</span> <span class="name">toolkits</span> <span class="keyword">with</span>
<span class="name">multiple</span> <span class="name">functionality</span> <span class="name">that</span> <span class="name">make</span> <span class="name">it</span> <span class="name">easier</span> <span class="name">to</span> <span class="name">create</span> <span class="name">customised</span> <span class="name">pipelines</span> <span class="operator word">or</span>
<span class="name">analysis</span><span class="operator">.</span> <span class="name">This</span> <span class="name">review</span> <span class="name">briefly</span> <span class="name">compares</span> <span class="name">the</span> <span class="name">quirks</span> <span class="name">of</span> <span class="name">the</span> <span class="name">underlying</span> <span class="name">languages</span>
<span class="operator word">and</span> <span class="name">the</span> <span class="name">functionality</span><span class="punctuation">,</span> <span class="name">documentation</span><span class="punctuation">,</span> <span class="name">utility</span> <span class="operator word">and</span> <span class="name">relative</span> <span class="name">advantages</span> <span class="name">of</span> <span class="name">the</span>
<span class="name">Bio</span> <span class="name">counterparts</span><span class="punctuation">,</span> <span class="name">particularly</span> <span class="keyword namespace">from</span> <span class="name namespace">the</span> <span class="name namespace">point</span> <span class="name namespace">of</span> <span class="name namespace">view</span> <span class="name namespace">of</span> <span class="name namespace">the</span> <span class="name namespace">beginning</span>
<span class="name">biologist</span> <span class="name">programmer</span><span class="operator">.</span><span class="literal string">'</span>
</pre>
<p>The key names used in a Medline record can be rather obscure; use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
</pre>
<p>for a brief summary.</p>
<p>To parse a file containing multiple Medline records, you can use the
<tt class="docutils literal">parse</tt> function instead:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Medline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">input</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;pubmed_result2.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Medline</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;TI&quot;</span><span class="punctuation">]</span>
<span class="name">A</span> <span class="name">high</span> <span class="name">level</span> <span class="name">interface</span> <span class="name">to</span> <span class="name">SCOP</span> <span class="operator word">and</span> <span class="name">ASTRAL</span> <span class="name">implemented</span> <span class="operator word">in</span> <span class="name">python</span><span class="operator">.</span>
<span class="name">GenomeDiagram</span><span class="punctuation">:</span> <span class="name">a</span> <span class="name">python</span> <span class="name">package</span> <span class="keyword">for</span> <span class="name">the</span> <span class="name">visualization</span> <span class="name">of</span> <span class="name">large</span><span class="operator">-</span><span class="name">scale</span> <span class="name">genomic</span> <span class="name">data</span><span class="operator">.</span>
<span class="name">Open</span> <span class="name">source</span> <span class="name">clustering</span> <span class="name">software</span><span class="operator">.</span>
<span class="name">PDB</span> <span class="name builtin">file</span> <span class="name">parser</span> <span class="operator word">and</span> <span class="name">structure</span> <span class="keyword">class</span> <span class="name class">implemented</span> <span class="operator word">in</span> <span class="name">Python</span><span class="operator">.</span>
</pre>
<p>Instead of parsing Medline records stored in files, you can also parse
Medline records downloaded by <tt class="docutils literal">Bio.Entrez.efetch</tt>. For example, let’s
look at all Medline records in PubMed related to Biopython:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;biopython&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'19304878'</span><span class="punctuation">,</span> <span class="literal string">'18606172'</span><span class="punctuation">,</span> <span class="literal string">'16403221'</span><span class="punctuation">,</span> <span class="literal string">'16377612'</span><span class="punctuation">,</span> <span class="literal string">'14871861'</span><span class="punctuation">,</span> <span class="literal string">'14630660'</span><span class="punctuation">,</span> <span class="literal string">'12230038'</span><span class="punctuation">]</span>
</pre>
<p>We now use <tt class="docutils literal">Bio.Entrez.efetch</tt> to download these Medline records:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idlist</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span><span class="name builtin">id</span><span class="operator">=</span><span class="name">idlist</span><span class="punctuation">,</span><span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;medline&quot;</span><span class="punctuation">,</span><span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
</pre>
<p>Here, we specify <tt class="docutils literal"><span class="pre">rettype=&quot;medline&quot;,</span> <span class="pre">retmode=&quot;text&quot;</span></tt> to obtain the
Medline records in plain-text Medline format. Now we use <tt class="docutils literal">Bio.Medline</tt>
to parse these records:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Medline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Medline</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;AU&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Cock PJ'</span><span class="punctuation">,</span> <span class="literal string">'Antao T'</span><span class="punctuation">,</span> <span class="literal string">'Chang JT'</span><span class="punctuation">,</span> <span class="literal string">'Chapman BA'</span><span class="punctuation">,</span> <span class="literal string">'Cox CJ'</span><span class="punctuation">,</span> <span class="literal string">'Dalke A'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'de Hoon MJ'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Munteanu CR'</span><span class="punctuation">,</span> <span class="literal string">'Gonzalez-Diaz H'</span><span class="punctuation">,</span> <span class="literal string">'Magalhaes AL'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Casbon JA'</span><span class="punctuation">,</span> <span class="literal string">'Crooks GE'</span><span class="punctuation">,</span> <span class="literal string">'Saqi MA'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Pritchard L'</span><span class="punctuation">,</span> <span class="literal string">'White JA'</span><span class="punctuation">,</span> <span class="literal string">'Birch PR'</span><span class="punctuation">,</span> <span class="literal string">'Toth IK'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'de Hoon MJ'</span><span class="punctuation">,</span> <span class="literal string">'Imoto S'</span><span class="punctuation">,</span> <span class="literal string">'Nolan J'</span><span class="punctuation">,</span> <span class="literal string">'Miyano S'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Hamelryck T'</span><span class="punctuation">,</span> <span class="literal string">'Manderick B'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Mangalam H'</span><span class="punctuation">]</span>
</pre>
<p>For comparison, here we show an example using the XML format:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idlist</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span><span class="name builtin">id</span><span class="operator">=</span><span class="name">idlist</span><span class="punctuation">,</span><span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;medline&quot;</span><span class="punctuation">,</span><span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;MedlineCitation&quot;</span><span class="punctuation">][</span><span class="literal string">&quot;Article&quot;</span><span class="punctuation">][</span><span class="literal string">&quot;ArticleTitle&quot;</span><span class="punctuation">]</span>
<span class="name">Biopython</span><span class="punctuation">:</span> <span class="name">freely</span> <span class="name">available</span> <span class="name">Python</span> <span class="name">tools</span> <span class="keyword">for</span> <span class="name">computational</span> <span class="name">molecular</span> <span class="name">biology</span> <span class="operator word">and</span>
 <span class="name">bioinformatics</span><span class="operator">.</span>
<span class="name">Enzymes</span><span class="operator">/</span><span class="name">non</span><span class="operator">-</span><span class="name">enzymes</span> <span class="name">classification</span> <span class="name">model</span> <span class="name">complexity</span> <span class="name">based</span> <span class="name">on</span> <span class="name">composition</span><span class="punctuation">,</span> <span class="name">sequence</span><span class="punctuation">,</span>
 <span class="literal number integer">3</span><span class="name">D</span> <span class="operator word">and</span> <span class="name">topological</span> <span class="name">indices</span><span class="operator">.</span>
<span class="name">A</span> <span class="name">high</span> <span class="name">level</span> <span class="name">interface</span> <span class="name">to</span> <span class="name">SCOP</span> <span class="operator word">and</span> <span class="name">ASTRAL</span> <span class="name">implemented</span> <span class="operator word">in</span> <span class="name">python</span><span class="operator">.</span>
<span class="name">GenomeDiagram</span><span class="punctuation">:</span> <span class="name">a</span> <span class="name">python</span> <span class="name">package</span> <span class="keyword">for</span> <span class="name">the</span> <span class="name">visualization</span> <span class="name">of</span> <span class="name">large</span><span class="operator">-</span><span class="name">scale</span> <span class="name">genomic</span> <span class="name">data</span><span class="operator">.</span>
<span class="name">Open</span> <span class="name">source</span> <span class="name">clustering</span> <span class="name">software</span><span class="operator">.</span>
<span class="name">PDB</span> <span class="name builtin">file</span> <span class="name">parser</span> <span class="operator word">and</span> <span class="name">structure</span> <span class="keyword">class</span> <span class="name class">implemented</span> <span class="operator word">in</span> <span class="name">Python</span><span class="operator">.</span>
<span class="name">The</span> <span class="name">Bio</span><span class="operator">*</span> <span class="name">toolkits</span><span class="operator">--</span><span class="name">a</span> <span class="name">brief</span> <span class="name">overview</span><span class="operator">.</span>
</pre>
<p>Note that in both of these examples, for simplicity we have naively
combined ESearch and EFetch. In this situation, the NCBI would expect
you to use their history feature, as illustrated in
Section <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
</div>
<div class="section" id="parsing-geo-records">
<h3>9.12.2  Parsing GEO records</h3>
<p>GEO (<a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/">Gene Expression Omnibus</a>) is
a data repository of high-throughput gene expression and hybridization
array data. The <tt class="docutils literal">Bio.Geo</tt> module can be used to parse GEO-formatted
data.</p>
<p>The following code fragment shows how to parse the example GEO file
<tt class="docutils literal">GSE16.txt</tt> into a record and print the record:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Geo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;GSE16.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Geo</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span>
</pre>
<p>You can search the “gds” database (GEO datasets) with ESearch:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span> <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;gds&quot;</span><span class="punctuation">,</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;GSE16&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'200000016'</span><span class="punctuation">,</span> <span class="literal string">'100000028'</span><span class="punctuation">]</span>
</pre>
<p>From the Entrez website, UID “200000016” is GDS16 while the other hit
“100000028” is for the associated platform, GPL28. Unfortunately, at the
time of writing the NCBI don’t seem to support downloading GEO files
using Entrez (not as XML, nor in the <em>Simple Omnibus Format in Text</em>
(SOFT) format).</p>
<p>However, it is actually pretty straight forward to download the GEO
files by FTP from
<a class="reference external" href="ftp://ftp.ncbi.nih.gov/pub/geo/">ftp://ftp.ncbi.nih.gov/pub/geo/</a>
instead. In this case you might want
<a class="reference external" href="ftp://ftp.ncbi.nih.gov/pub/geo/DATA/SOFT/by_series/GSE16/GSE16_family.soft.gz">ftp://ftp.ncbi.nih.gov/pub/geo/DATA/SOFT/by_series/GSE16/GSE16_family.soft.gz</a>
(a compressed file, see the Python module gzip).</p>
</div>
<div class="section" id="parsing-unigene-records">
<h3>9.12.3  Parsing UniGene records</h3>
<p>UniGene is an NCBI database of the transcriptome, with each UniGene
record showing the set of transcripts that are associated with a
particular gene in a specific organism. A typical UniGene record looks
like this:</p>
<pre class="code python literal-block">
<span class="name">ID</span>          <span class="name">Hs</span><span class="operator">.</span><span class="literal number integer">2</span>
<span class="name">TITLE</span>       <span class="name">N</span><span class="operator">-</span><span class="name">acetyltransferase</span> <span class="literal number integer">2</span> <span class="punctuation">(</span><span class="name">arylamine</span> <span class="name">N</span><span class="operator">-</span><span class="name">acetyltransferase</span><span class="punctuation">)</span>
<span class="name">GENE</span>        <span class="name">NAT2</span>
<span class="name">CYTOBAND</span>    <span class="literal number integer">8</span><span class="name">p22</span>
<span class="name">GENE_ID</span>     <span class="literal number integer">10</span>
<span class="name">LOCUSLINK</span>   <span class="literal number integer">10</span>
<span class="name">HOMOL</span>       <span class="name">YES</span>
<span class="name">EXPRESS</span>      <span class="name">bone</span><span class="operator">|</span> <span class="name">connective</span> <span class="name">tissue</span><span class="operator">|</span> <span class="name">intestine</span><span class="operator">|</span> <span class="name">liver</span><span class="operator">|</span> <span class="name">liver</span> <span class="name">tumor</span><span class="operator">|</span> <span class="name">normal</span><span class="operator">|</span> <span class="name">soft</span> <span class="name">tissue</span><span class="operator">/</span><span class="name">muscle</span> <span class="name">tissue</span> <span class="name">tumor</span><span class="operator">|</span> <span class="name">adult</span>
<span class="name">RESTR_EXPR</span>   <span class="name">adult</span>
<span class="name">CHROMOSOME</span>  <span class="literal number integer">8</span>
<span class="name">STS</span>         <span class="name">ACC</span><span class="operator">=</span><span class="name">PMC310725P3</span> <span class="name">UNISTS</span><span class="operator">=</span><span class="literal number integer">272646</span>
<span class="name">STS</span>         <span class="name">ACC</span><span class="operator">=</span><span class="name">WIAF</span><span class="operator">-</span><span class="literal number integer">2120</span> <span class="name">UNISTS</span><span class="operator">=</span><span class="literal number integer">44576</span>
<span class="name">STS</span>         <span class="name">ACC</span><span class="operator">=</span><span class="name">G59899</span> <span class="name">UNISTS</span><span class="operator">=</span><span class="literal number integer">137181</span>
<span class="operator">...</span>
<span class="name">STS</span>         <span class="name">ACC</span><span class="operator">=</span><span class="name">GDB</span><span class="punctuation">:</span><span class="literal number integer">187676</span> <span class="name">UNISTS</span><span class="operator">=</span><span class="literal number integer">155563</span>
<span class="name">PROTSIM</span>     <span class="name">ORG</span><span class="operator">=</span><span class="literal number integer">10090</span><span class="punctuation">;</span> <span class="name">PROTGI</span><span class="operator">=</span><span class="literal number integer">6754794</span><span class="punctuation">;</span> <span class="name">PROTID</span><span class="operator">=</span><span class="name">NP_035004</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">PCT</span><span class="operator">=</span><span class="literal number float">76.55</span><span class="punctuation">;</span> <span class="name">ALN</span><span class="operator">=</span><span class="literal number integer">288</span>
<span class="name">PROTSIM</span>     <span class="name">ORG</span><span class="operator">=</span><span class="literal number integer">9796</span><span class="punctuation">;</span> <span class="name">PROTGI</span><span class="operator">=</span><span class="literal number integer">149742490</span><span class="punctuation">;</span> <span class="name">PROTID</span><span class="operator">=</span><span class="name">XP_001487907</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">PCT</span><span class="operator">=</span><span class="literal number float">79.66</span><span class="punctuation">;</span> <span class="name">ALN</span><span class="operator">=</span><span class="literal number integer">288</span>
<span class="name">PROTSIM</span>     <span class="name">ORG</span><span class="operator">=</span><span class="literal number integer">9986</span><span class="punctuation">;</span> <span class="name">PROTGI</span><span class="operator">=</span><span class="literal number integer">126722851</span><span class="punctuation">;</span> <span class="name">PROTID</span><span class="operator">=</span><span class="name">NP_001075655</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">PCT</span><span class="operator">=</span><span class="literal number float">76.90</span><span class="punctuation">;</span> <span class="name">ALN</span><span class="operator">=</span><span class="literal number integer">288</span>
<span class="operator">...</span>
<span class="name">PROTSIM</span>     <span class="name">ORG</span><span class="operator">=</span><span class="literal number integer">9598</span><span class="punctuation">;</span> <span class="name">PROTGI</span><span class="operator">=</span><span class="literal number integer">114619004</span><span class="punctuation">;</span> <span class="name">PROTID</span><span class="operator">=</span><span class="name">XP_519631</span><span class="operator">.</span><span class="literal number integer">2</span><span class="punctuation">;</span> <span class="name">PCT</span><span class="operator">=</span><span class="literal number float">98.28</span><span class="punctuation">;</span> <span class="name">ALN</span><span class="operator">=</span><span class="literal number integer">288</span>

<span class="name">SCOUNT</span>      <span class="literal number integer">38</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">BC067218</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g45501306</span><span class="punctuation">;</span> <span class="name">PID</span><span class="operator">=</span><span class="name">g45501307</span><span class="punctuation">;</span> <span class="name">SEQTYPE</span><span class="operator">=</span><span class="name">mRNA</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">NM_000015</span><span class="operator">.</span><span class="literal number integer">2</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g116295259</span><span class="punctuation">;</span> <span class="name">PID</span><span class="operator">=</span><span class="name">g116295260</span><span class="punctuation">;</span> <span class="name">SEQTYPE</span><span class="operator">=</span><span class="name">mRNA</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">D90042</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g219415</span><span class="punctuation">;</span> <span class="name">PID</span><span class="operator">=</span><span class="name">g219416</span><span class="punctuation">;</span> <span class="name">SEQTYPE</span><span class="operator">=</span><span class="name">mRNA</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">D90040</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g219411</span><span class="punctuation">;</span> <span class="name">PID</span><span class="operator">=</span><span class="name">g219412</span><span class="punctuation">;</span> <span class="name">SEQTYPE</span><span class="operator">=</span><span class="name">mRNA</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">BC015878</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g16198419</span><span class="punctuation">;</span> <span class="name">PID</span><span class="operator">=</span><span class="name">g16198420</span><span class="punctuation">;</span> <span class="name">SEQTYPE</span><span class="operator">=</span><span class="name">mRNA</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">CR407631</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g47115198</span><span class="punctuation">;</span> <span class="name">PID</span><span class="operator">=</span><span class="name">g47115199</span><span class="punctuation">;</span> <span class="name">SEQTYPE</span><span class="operator">=</span><span class="name">mRNA</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">BG569293</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g13576946</span><span class="punctuation">;</span> <span class="name">CLONE</span><span class="operator">=</span><span class="name">IMAGE</span><span class="punctuation">:</span><span class="literal number integer">4722596</span><span class="punctuation">;</span> <span class="name">END</span><span class="operator">=</span><span class="literal number integer">5</span><span class="literal string">'; LID=6989; SEQTYPE=EST; TRACE=44157214</span>
<span class="operator">...</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">AU099534</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g13550663</span><span class="punctuation">;</span> <span class="name">CLONE</span><span class="operator">=</span><span class="name">HSI08034</span><span class="punctuation">;</span> <span class="name">END</span><span class="operator">=</span><span class="literal number integer">5</span><span class="literal string">'; LID=8800; SEQTYPE=EST</span>
<span class="operator">//</span>
</pre>
<p>This particular record shows the set of transcripts (shown in the
<tt class="docutils literal">SEQUENCE</tt> lines) that originate from the human gene NAT2, encoding en
N-acetyltransferase. The <tt class="docutils literal">PROTSIM</tt> lines show proteins with
significant similarity to NAT2, whereas the <tt class="docutils literal">STS</tt> lines show the
corresponding sequence-tagged sites in the genome.</p>
<p>To parse UniGene files, use the <tt class="docutils literal">Bio.UniGene</tt> module:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">UniGene</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">input</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;myunigenefile.data&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">UniGene</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="punctuation">)</span>
</pre>
<p>The <tt class="docutils literal">record</tt> returned by <tt class="docutils literal">UniGene.read</tt> is a Python object with
attributes corresponding to the fields in the UniGene record. For
example,</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">ID</span>
<span class="literal string">&quot;Hs.2&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">title</span>
<span class="literal string">&quot;N-acetyltransferase 2 (arylamine N-acetyltransferase)&quot;</span>
</pre>
<p>The <tt class="docutils literal">EXPRESS</tt> and <tt class="docutils literal">RESTR_EXPR</tt> lines are stored as Python lists of
strings:</p>
<pre class="code python literal-block">
<span class="punctuation">[</span><span class="literal string">'bone'</span><span class="punctuation">,</span> <span class="literal string">'connective tissue'</span><span class="punctuation">,</span> <span class="literal string">'intestine'</span><span class="punctuation">,</span> <span class="literal string">'liver'</span><span class="punctuation">,</span> <span class="literal string">'liver tumor'</span><span class="punctuation">,</span> <span class="literal string">'normal'</span><span class="punctuation">,</span> <span class="literal string">'soft tissue/muscle tissue tumor'</span><span class="punctuation">,</span> <span class="literal string">'adult'</span><span class="punctuation">]</span>
</pre>
<p>Specialized objects are returned for the <tt class="docutils literal">STS</tt>, <tt class="docutils literal">PROTSIM</tt>, and
<tt class="docutils literal">SEQUENCE</tt> lines, storing the keys shown in each line as attributes:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">sts</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">acc</span>
<span class="literal string">'PMC310725P3'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">sts</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">unists</span>
<span class="literal string">'272646'</span>
</pre>
<p>and similarly for the <tt class="docutils literal">PROTSIM</tt> and <tt class="docutils literal">SEQUENCE</tt> lines.</p>
<p>To parse a file containing more than one UniGene record, use the
<tt class="docutils literal">parse</tt> function in <tt class="docutils literal">Bio.UniGene</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">UniGene</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">input</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;unigenerecords.data&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">UniGene</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">ID</span>
</pre>
</div>
</div>
<div class="section" id="using-a-proxy">
<h2>9.13  Using a proxy</h2>
<p>Normally you won’t have to worry about using a proxy, but if this is an
issue on your network here is how to deal with it. Internally,
<tt class="docutils literal">Bio.Entrez</tt> uses the standard Python library <tt class="docutils literal">urllib</tt> for accessing
the NCBI servers. This will check an environment variable called
<tt class="docutils literal">http_proxy</tt> to configure any simple proxy automatically.
Unfortunately this module does not support the use of proxies which
require authentication.</p>
<p>You may choose to set the <tt class="docutils literal">http_proxy</tt> environment variable once (how
you do this will depend on your operating system). Alternatively you can
set this within Python at the start of your script, for example:</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">os</span>
<span class="name">os</span><span class="operator">.</span><span class="name">environ</span><span class="punctuation">[</span><span class="literal string">&quot;http_proxy&quot;</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="literal string">&quot;http://proxyhost.example.com:8080&quot;</span>
</pre>
<p>See the <a class="reference external" href="http://www.python.org/doc/lib/module-urllib.html">urllib
documentation</a> for
more details.</p>
</div>
<div class="section" id="examples">
<h2>9.14  Examples</h2>
<div class="section" id="pubmed-and-medline">
<h3>9.14.1  PubMed and Medline</h3>
<p>If you are in the medical field or interested in human issues (and many
times even if you are not!), PubMed
(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/PubMed/">http://www.ncbi.nlm.nih.gov/PubMed/</a>)
is an excellent source of all kinds of goodies. So like other things,
we’d like to be able to grab information from it and use it in Python
scripts.</p>
<p>In this example, we will query PubMed for all articles having to do with
orchids (see section <a class="reference external" href="#sec:orchids">2.3</a> for our motivation). We
first check how many of such articles there are:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">egquery</span><span class="punctuation">(</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;orchid&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">row</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;eGQueryResult&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;DbName&quot;</span><span class="punctuation">]</span><span class="operator">==</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal number integer">463</span>
</pre>
<p>Now we use the <tt class="docutils literal">Bio.Entrez.efetch</tt> function to download the PubMed IDs
of these 463 articles:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;orchid&quot;</span><span class="punctuation">,</span> <span class="name">retmax</span><span class="operator">=</span><span class="literal number integer">463</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idlist</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">idlist</span>
</pre>
<p>This returns a Python list containing all of the PubMed IDs of articles
related to orchids:</p>
<pre class="code python literal-block">
<span class="punctuation">[</span><span class="literal string">'18680603'</span><span class="punctuation">,</span> <span class="literal string">'18665331'</span><span class="punctuation">,</span> <span class="literal string">'18661158'</span><span class="punctuation">,</span> <span class="literal string">'18627489'</span><span class="punctuation">,</span> <span class="literal string">'18627452'</span><span class="punctuation">,</span> <span class="literal string">'18612381'</span><span class="punctuation">,</span>
<span class="literal string">'18594007'</span><span class="punctuation">,</span> <span class="literal string">'18591784'</span><span class="punctuation">,</span> <span class="literal string">'18589523'</span><span class="punctuation">,</span> <span class="literal string">'18579475'</span><span class="punctuation">,</span> <span class="literal string">'18575811'</span><span class="punctuation">,</span> <span class="literal string">'18575690'</span><span class="punctuation">,</span>
<span class="operator">...</span>
</pre>
<p>Now that we’ve got them, we obviously want to get the corresponding
Medline records and extract the information from them. Here, we’ll
download the Medline records in the Medline flat-file format, and use
the <tt class="docutils literal">Bio.Medline</tt> module to parse them:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Medline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="name">idlist</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;medline&quot;</span><span class="punctuation">,</span>
                           <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Medline</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>NOTE - We’ve just done a separate search and fetch here, the NCBI much
prefer you to take advantage of their history support in this situation.
See Section <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
<p>Keep in mind that <tt class="docutils literal">records</tt> is an iterator, so you can iterate through
the records only once. If you want to save the records, you can convert
them to a list:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">)</span>
</pre>
<p>Let’s now iterate over the records to print out some information about
each record:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;title:&quot;</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;TI&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;?&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;authors:&quot;</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;AU&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;?&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;source:&quot;</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;SO&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;?&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="keyword">print</span>
</pre>
<p>The output for this looks like:</p>
<pre class="code python literal-block">
<span class="name">title</span><span class="punctuation">:</span> <span class="name">Sex</span> <span class="name">pheromone</span> <span class="name">mimicry</span> <span class="operator word">in</span> <span class="name">the</span> <span class="name">early</span> <span class="name">spider</span> <span class="name">orchid</span> <span class="punctuation">(</span><span class="name">ophrys</span> <span class="name">sphegodes</span><span class="punctuation">):</span>
<span class="name">patterns</span> <span class="name">of</span> <span class="name">hydrocarbons</span> <span class="keyword">as</span> <span class="name">the</span> <span class="name">key</span> <span class="name">mechanism</span> <span class="keyword">for</span> <span class="name">pollination</span> <span class="name">by</span> <span class="name">sexual</span>
<span class="name">deception</span> <span class="punctuation">[</span><span class="name">In</span> <span class="name">Process</span> <span class="name">Citation</span><span class="punctuation">]</span>
<span class="name">authors</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'Schiestl FP'</span><span class="punctuation">,</span> <span class="literal string">'Ayasse M'</span><span class="punctuation">,</span> <span class="literal string">'Paulus HF'</span><span class="punctuation">,</span> <span class="literal string">'Lofstedt C'</span><span class="punctuation">,</span> <span class="literal string">'Hansson BS'</span><span class="punctuation">,</span>
<span class="literal string">'Ibarra F'</span><span class="punctuation">,</span> <span class="literal string">'Francke W'</span><span class="punctuation">]</span>
<span class="name">source</span><span class="punctuation">:</span> <span class="name">J</span> <span class="name">Comp</span> <span class="name">Physiol</span> <span class="punctuation">[</span><span class="name">A</span><span class="punctuation">]</span> <span class="literal number integer">2000</span> <span class="name">Jun</span><span class="punctuation">;</span><span class="literal number integer">186</span><span class="punctuation">(</span><span class="literal number integer">6</span><span class="punctuation">):</span><span class="literal number integer">567</span><span class="operator">-</span><span class="literal number integer">74</span>
</pre>
<p>Especially interesting to note is the list of authors, which is returned
as a standard Python list. This makes it easy to manipulate and search
using standard Python tools. For instance, we could loop through a whole
bunch of entries searching for a particular author with code like the
following:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">search_author</span> <span class="operator">=</span> <span class="literal string">&quot;Waits T&quot;</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="operator word">not</span> <span class="literal string">&quot;AU&quot;</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">continue</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">search_author</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;AU&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="literal string">&quot;Author </span><span class="literal string interpol">%s</span><span class="literal string"> found: </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">search_author</span><span class="punctuation">,</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;SO&quot;</span><span class="punctuation">])</span>
</pre>
<p>Hopefully this section gave you an idea of the power and flexibility of
the Entrez and Medline interfaces and how they can be used together.</p>
</div>
<div class="section" id="searching-downloading-and-parsing-entrez-nucleotide-records">
<h3>9.14.2  Searching, downloading, and parsing Entrez Nucleotide records</h3>
<p>Here we’ll show a simple example of performing a remote Entrez query. In
section <a class="reference external" href="#sec:orchids">2.3</a> of the parsing examples, we talked
about using NCBI’s Entrez website to search the NCBI nucleotide
databases for info on Cypripedioideae, our friends the lady slipper
orchids. Now, we’ll look at how to automate that process using a Python
script. In this example, we’ll just show how to connect, get the
results, and parse them, with the Entrez module doing all of the work.</p>
<p>First, we use EGQuery to find out the number of results we will get
before actually downloading them. EGQuery will tell us how many search
results were found in each of the databases, but for this example we are
only interested in nucleotides:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">egquery</span><span class="punctuation">(</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Cypripedioideae&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">row</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;eGQueryResult&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;DbName&quot;</span><span class="punctuation">]</span><span class="operator">==</span><span class="literal string">&quot;nuccore&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal number integer">814</span>
</pre>
<p>So, we expect to find 814 Entrez Nucleotide records (this is the number
I obtained in 2008; it is likely to increase in the future). If you find
some ridiculously high number of hits, you may want to reconsider if you
really want to download all of them, which is our next step:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Cypripedioideae&quot;</span><span class="punctuation">,</span> <span class="name">retmax</span><span class="operator">=</span><span class="literal number integer">814</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>Here, <tt class="docutils literal">record</tt> is a Python dictionary containing the search results
and some auxiliary information. Just for information, let’s look at what
is stored in this dictionary:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">u'Count'</span><span class="punctuation">,</span> <span class="literal string">u'RetMax'</span><span class="punctuation">,</span> <span class="literal string">u'IdList'</span><span class="punctuation">,</span> <span class="literal string">u'TranslationSet'</span><span class="punctuation">,</span> <span class="literal string">u'RetStart'</span><span class="punctuation">,</span> <span class="literal string">u'QueryTranslation'</span><span class="punctuation">]</span>
</pre>
<p>First, let’s check how many results were found:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal string">'814'</span>
</pre>
<p>which is the number we expected. The 814 results are stored in
<tt class="docutils literal"><span class="pre">record['IdList']</span></tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">])</span>
<span class="literal number integer">814</span>
</pre>
<p>Let’s look at the first five results:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">][:</span><span class="literal number integer">5</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'187237168'</span><span class="punctuation">,</span> <span class="literal string">'187372713'</span><span class="punctuation">,</span> <span class="literal string">'187372690'</span><span class="punctuation">,</span> <span class="literal string">'187372688'</span><span class="punctuation">,</span> <span class="literal string">'187372686'</span><span class="punctuation">]</span>
</pre>
<p>We can download these records using <tt class="docutils literal">efetch</tt>. While you could download
these records one by one, to reduce the load on NCBI’s servers, it is
better to fetch a bunch of records at the same time, shown below.
However, in this situation you should ideally be using the history
feature described later in Section <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idlist</span> <span class="operator">=</span> <span class="literal string">&quot;,&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">][:</span><span class="literal number integer">5</span><span class="punctuation">])</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">idlist</span>
<span class="literal number integer">187237168</span><span class="punctuation">,</span><span class="literal number integer">187372713</span><span class="punctuation">,</span><span class="literal number integer">187372690</span><span class="punctuation">,</span><span class="literal number integer">187372688</span><span class="punctuation">,</span><span class="literal number integer">187372686</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="name">idlist</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
</pre>
<p>Each of these records corresponds to one GenBank record.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">u'GBSeq_moltype'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_source'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_sequence'</span><span class="punctuation">,</span>
 <span class="literal string">u'GBSeq_primary-accession'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_definition'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_accession-version'</span><span class="punctuation">,</span>
 <span class="literal string">u'GBSeq_topology'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_length'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_feature-table'</span><span class="punctuation">,</span>
 <span class="literal string">u'GBSeq_create-date'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_other-seqids'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_division'</span><span class="punctuation">,</span>
 <span class="literal string">u'GBSeq_taxonomy'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_references'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_update-date'</span><span class="punctuation">,</span>
 <span class="literal string">u'GBSeq_organism'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_locus'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_strandedness'</span><span class="punctuation">]</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;GBSeq_primary-accession&quot;</span><span class="punctuation">]</span>
<span class="name">DQ110336</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;GBSeq_other-seqids&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'gb|DQ110336.1|'</span><span class="punctuation">,</span> <span class="literal string">'gi|187237168'</span><span class="punctuation">]</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;GBSeq_definition&quot;</span><span class="punctuation">]</span>
<span class="name">Cypripedium</span> <span class="name">calceolus</span> <span class="name">voucher</span> <span class="name">Davis</span> <span class="literal number oct">03</span><span class="operator">-</span><span class="literal number oct">03</span> <span class="name">A</span> <span class="name">maturase</span> <span class="punctuation">(</span><span class="name">matR</span><span class="punctuation">)</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">partial</span> <span class="name">cds</span><span class="punctuation">;</span>
<span class="name">mitochondrial</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;GBSeq_organism&quot;</span><span class="punctuation">]</span>
<span class="name">Cypripedium</span> <span class="name">calceolus</span>
</pre>
<p>You could use this to quickly set up searches – but for heavy usage, see
Section <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
</div>
<div class="section" id="searching-downloading-and-parsing-genbank-records">
<h3>9.14.3  Searching, downloading, and parsing GenBank records</h3>
<p>The GenBank record format is a very popular method of holding
information about sequences, sequence features, and other associated
sequence information. The format is a good way to get information from
the NCBI databases at
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/">http://www.ncbi.nlm.nih.gov/</a>.</p>
<p>In this example we’ll show how to query the NCBI databases,to retrieve
the records from the query, and then parse them using <tt class="docutils literal">Bio.SeqIO</tt> -
something touched on in Section <a class="reference external" href="#sec:SeqIO_GenBank_Online">5.3.1</a>.
For simplicity, this example <em>does not</em> take advantage of the WebEnv
history feature – see Section <a class="reference external" href="#sec:entrez-webenv">9.15</a> for this.</p>
<p>First, we want to make a query and find out the ids of the records to
retrieve. Here we’ll do a quick search for one of our favorite
organisms, <em>Opuntia</em> (prickly-pear cacti). We can do quick search and
get back the GIs (GenBank identifiers) for all of the corresponding
records. First we check how many records there are:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">egquery</span><span class="punctuation">(</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Opuntia AND rpl16&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">row</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;eGQueryResult&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;DbName&quot;</span><span class="punctuation">]</span><span class="operator">==</span><span class="literal string">&quot;nuccore&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="operator">...</span>
<span class="literal number integer">9</span>
</pre>
<p>Now we download the list of GenBank identifiers:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nuccore&quot;</span><span class="punctuation">,</span> <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Opuntia AND rpl16&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gi_list</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gi_list</span>
<span class="punctuation">[</span><span class="literal string">'57240072'</span><span class="punctuation">,</span> <span class="literal string">'57240071'</span><span class="punctuation">,</span> <span class="literal string">'6273287'</span><span class="punctuation">,</span> <span class="literal string">'6273291'</span><span class="punctuation">,</span> <span class="literal string">'6273290'</span><span class="punctuation">,</span> <span class="literal string">'6273289'</span><span class="punctuation">,</span> <span class="literal string">'6273286'</span><span class="punctuation">,</span>
<span class="literal string">'6273285'</span><span class="punctuation">,</span> <span class="literal string">'6273284'</span><span class="punctuation">]</span>
</pre>
<p>Now we use these GIs to download the GenBank records - note that with
older versions of Biopython you had to supply a comma separated list of
GI numbers to Entrez, as of Biopython 1.59 you can pass a list and this
is converted for you:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gi_str</span> <span class="operator">=</span> <span class="literal string">&quot;,&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">gi_list</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nuccore&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="name">gi_str</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
</pre>
<p>If you want to look at the raw GenBank files, you can read from this
handle and print out the result:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">text</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">text</span>
<span class="name">LOCUS</span>       <span class="name">AY851612</span>                 <span class="literal number integer">892</span> <span class="name">bp</span>    <span class="name">DNA</span>     <span class="name">linear</span>   <span class="name">PLN</span> <span class="literal number integer">10</span><span class="operator">-</span><span class="name">APR</span><span class="operator">-</span><span class="literal number integer">2007</span>
<span class="name">DEFINITION</span>  <span class="name">Opuntia</span> <span class="name">subulata</span> <span class="name">rpl16</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">intron</span><span class="punctuation">;</span> <span class="name">chloroplast</span><span class="operator">.</span>
<span class="name">ACCESSION</span>   <span class="name">AY851612</span>
<span class="name">VERSION</span>     <span class="name">AY851612</span><span class="operator">.</span><span class="literal number integer">1</span>  <span class="name">GI</span><span class="punctuation">:</span><span class="literal number integer">57240072</span>
<span class="name">KEYWORDS</span>    <span class="operator">.</span>
<span class="name">SOURCE</span>      <span class="name">chloroplast</span> <span class="name">Austrocylindropuntia</span> <span class="name">subulata</span>
  <span class="name">ORGANISM</span>  <span class="name">Austrocylindropuntia</span> <span class="name">subulata</span>
            <span class="name">Eukaryota</span><span class="punctuation">;</span> <span class="name">Viridiplantae</span><span class="punctuation">;</span> <span class="name">Streptophyta</span><span class="punctuation">;</span> <span class="name">Embryophyta</span><span class="punctuation">;</span> <span class="name">Tracheophyta</span><span class="punctuation">;</span>
            <span class="name">Spermatophyta</span><span class="punctuation">;</span> <span class="name">Magnoliophyta</span><span class="punctuation">;</span> <span class="name">eudicotyledons</span><span class="punctuation">;</span> <span class="name">core</span> <span class="name">eudicotyledons</span><span class="punctuation">;</span>
            <span class="name">Caryophyllales</span><span class="punctuation">;</span> <span class="name">Cactaceae</span><span class="punctuation">;</span> <span class="name">Opuntioideae</span><span class="punctuation">;</span> <span class="name">Austrocylindropuntia</span><span class="operator">.</span>
<span class="name">REFERENCE</span>   <span class="literal number integer">1</span>  <span class="punctuation">(</span><span class="name">bases</span> <span class="literal number integer">1</span> <span class="name">to</span> <span class="literal number integer">892</span><span class="punctuation">)</span>
  <span class="name">AUTHORS</span>   <span class="name">Butterworth</span><span class="punctuation">,</span><span class="name">C</span><span class="operator">.</span><span class="name">A</span><span class="operator">.</span> <span class="operator word">and</span> <span class="name">Wallace</span><span class="punctuation">,</span><span class="name">R</span><span class="operator">.</span><span class="name">S</span><span class="operator">.</span>
<span class="operator">...</span>
</pre>
<p>In this case, we are just getting the raw records. To get the records in
a more Python-friendly form, we can use <tt class="docutils literal">Bio.SeqIO</tt> to parse the
GenBank data into <tt class="docutils literal">SeqRecord</tt> objects, including <tt class="docutils literal">SeqFeature</tt>
objects (see Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nuccore&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="name">gi_str</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">)</span>
</pre>
<p>We can now step through the records and look at the information we are
interested in:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="operator">...</span>    <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">, length </span><span class="literal string interpol">%i</span><span class="literal string">, with </span><span class="literal string interpol">%i</span><span class="literal string"> features&quot;</span> \
<span class="operator">&gt;&gt;&gt;</span> <span class="operator">...</span>           <span class="operator">%</span> <span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">name</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">))</span>
<span class="name">AY851612</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">892</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AY851611</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">881</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191661</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">895</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191665</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">902</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191664</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">899</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191663</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">899</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191660</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">893</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191659</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">894</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191658</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">896</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
</pre>
<p>Using these automated query retrieval functionality is a big plus over
doing things by hand. Although the module should obey the NCBI’s max
three queries per second rule, the NCBI have other recommendations like
avoiding peak hours. See Section <a class="reference external" href="#sec:entrez-guidelines">9.1</a>. In
particular, please note that for simplicity, this example does not use
the WebEnv history feature. You should use this for any non-trivial
search and download work, see Section <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
<p>Finally, if plan to repeat your analysis, rather than downloading the
files from the NCBI and parsing them immediately (as shown in this
example), you should just download the records <em>once</em> and save them to
your hard disk, and then parse the local file.</p>
</div>
<div class="section" id="finding-the-lineage-of-an-organism">
<h3>9.14.4  Finding the lineage of an organism</h3>
<p>Staying with a plant example, let’s now find the lineage of the
Cypripedioideae orchid family. First, we search the Taxonomy database
for Cypripedioideae, which yields exactly one NCBI taxonomy identifier:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;Taxonomy&quot;</span><span class="punctuation">,</span> <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Cypripedioideae&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'158330'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="literal string">'158330'</span>
</pre>
<p>Now, we use <tt class="docutils literal">efetch</tt> to download this entry in the Taxonomy database,
and then parse it:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;Taxonomy&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;158330&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>Again, this record stores lots of information:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">u'Lineage'</span><span class="punctuation">,</span> <span class="literal string">u'Division'</span><span class="punctuation">,</span> <span class="literal string">u'ParentTaxId'</span><span class="punctuation">,</span> <span class="literal string">u'PubDate'</span><span class="punctuation">,</span> <span class="literal string">u'LineageEx'</span><span class="punctuation">,</span>
 <span class="literal string">u'CreateDate'</span><span class="punctuation">,</span> <span class="literal string">u'TaxId'</span><span class="punctuation">,</span> <span class="literal string">u'Rank'</span><span class="punctuation">,</span> <span class="literal string">u'GeneticCode'</span><span class="punctuation">,</span> <span class="literal string">u'ScientificName'</span><span class="punctuation">,</span>
 <span class="literal string">u'MitoGeneticCode'</span><span class="punctuation">,</span> <span class="literal string">u'UpdateDate'</span><span class="punctuation">]</span>
</pre>
<p>We can get the lineage directly from this record:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Lineage&quot;</span><span class="punctuation">]</span>
<span class="literal string">'cellular organisms; Eukaryota; Viridiplantae; Streptophyta; Streptophytina;</span>
 <span class="name">Embryophyta</span><span class="punctuation">;</span> <span class="name">Tracheophyta</span><span class="punctuation">;</span> <span class="name">Euphyllophyta</span><span class="punctuation">;</span> <span class="name">Spermatophyta</span><span class="punctuation">;</span> <span class="name">Magnoliophyta</span><span class="punctuation">;</span>
 <span class="name">Liliopsida</span><span class="punctuation">;</span> <span class="name">Asparagales</span><span class="punctuation">;</span> <span class="name">Orchidaceae</span><span class="literal string">'</span>
</pre>
<p>The record data contains much more than just the information shown here
- for example look under <tt class="docutils literal">&quot;LineageEx&quot;</tt> instead of <tt class="docutils literal">&quot;Lineage&quot;</tt> and
you’ll get the NCBI taxon identifiers of the lineage entries too.</p>
</div>
</div>
<div class="section" id="using-the-history-and-webenv">
<h2>9.15  Using the history and WebEnv</h2>
<p>Often you will want to make a series of linked queries. Most typically,
running a search, perhaps refining the search, and then retrieving
detailed search results. You <em>can</em> do this by making a series of
separate calls to Entrez. However, the NCBI prefer you to take advantage
of their history support - for example combining ESearch and EFetch.</p>
<p>Another typical use of the history support would be to combine EPost and
EFetch. You use EPost to upload a list of identifiers, which starts a
new history session. You then download the records with EFetch by
referring to the session (instead of the identifiers).</p>
<div class="section" id="searching-for-and-downloading-sequences-using-the-history">
<h3>9.15.1  Searching for and downloading sequences using the history</h3>
<p>Suppose we want to search and download all the <em>Opuntia</em> rpl16
nucleotide sequences, and store them in a FASTA file. As shown in
Section <a class="reference external" href="#sec:entrez-search-fetch-genbank">9.14.3</a>, we can naively
combine <tt class="docutils literal">Bio.Entrez.esearch()</tt> to get a list of GI numbers, and then
call <tt class="docutils literal">Bio.Entrez.efetch()</tt> to download them all.</p>
<p>However, the approved approach is to run the search with the history
feature. Then, we can fetch the results by reference to the search
results - which the NCBI can anticipate and cache.</p>
<p>To do this, call <tt class="docutils literal">Bio.Entrez.esearch()</tt> as normal, but with the
additional argument of <tt class="docutils literal"><span class="pre">usehistory=&quot;y&quot;</span></tt>,</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;history.user&#64;example.com&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">search_handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Opuntia[orgn] and rpl16&quot;</span><span class="punctuation">,</span>
                                   <span class="name">usehistory</span><span class="operator">=</span><span class="literal string">&quot;y&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">search_results</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">search_handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">search_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>When you get the XML output back, it will still include the usual search
results:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gi_list</span> <span class="operator">=</span> <span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">count</span> <span class="operator">=</span> <span class="name builtin">int</span><span class="punctuation">(</span><span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">])</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">assert</span> <span class="name">count</span> <span class="operator">==</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">gi_list</span><span class="punctuation">)</span>
</pre>
<p>However, you also get given two additional pieces of information, the
<tt class="docutils literal">WebEnv</tt> session cookie, and the <tt class="docutils literal">QueryKey</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">webenv</span> <span class="operator">=</span> <span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;WebEnv&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">query_key</span> <span class="operator">=</span> <span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;QueryKey&quot;</span><span class="punctuation">]</span>
</pre>
<p>Having stored these values in variables <tt class="docutils literal">session_cookie</tt> and
<tt class="docutils literal">query_key</tt> we can use them as parameters to <tt class="docutils literal">Bio.Entrez.efetch()</tt>
instead of giving the GI numbers as identifiers.</p>
<p>While for small searches you might be OK downloading everything at once,
it is better to download in batches. You use the <tt class="docutils literal">retstart</tt> and
<tt class="docutils literal">retmax</tt> parameters to specify which range of search results you want
returned (starting entry using zero-based counting, and maximum number
of results to return). For example,</p>
<pre class="code python literal-block">
<span class="name">batch_size</span> <span class="operator">=</span> <span class="literal number integer">3</span>
<span class="name">out_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;orchid_rpl16.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">start</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="name">count</span><span class="punctuation">,</span><span class="name">batch_size</span><span class="punctuation">):</span>
    <span class="name">end</span> <span class="operator">=</span> <span class="name builtin">min</span><span class="punctuation">(</span><span class="name">count</span><span class="punctuation">,</span> <span class="name">start</span><span class="operator">+</span><span class="name">batch_size</span><span class="punctuation">)</span>
    <span class="keyword">print</span> <span class="literal string">&quot;Going to download record </span><span class="literal string interpol">%i</span><span class="literal string"> to </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">start</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">end</span><span class="punctuation">)</span>
    <span class="name">fetch_handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">,</span>
                                 <span class="name">retstart</span><span class="operator">=</span><span class="name">start</span><span class="punctuation">,</span> <span class="name">retmax</span><span class="operator">=</span><span class="name">batch_size</span><span class="punctuation">,</span>
                                 <span class="name">webenv</span><span class="operator">=</span><span class="name">webenv</span><span class="punctuation">,</span> <span class="name">query_key</span><span class="operator">=</span><span class="name">query_key</span><span class="punctuation">)</span>
    <span class="name">data</span> <span class="operator">=</span> <span class="name">fetch_handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
    <span class="name">fetch_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
    <span class="name">out_handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
<span class="name">out_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>For illustrative purposes, this example downloaded the FASTA records in
batches of three. Unless you are downloading genomes or chromosomes, you
would normally pick a larger batch size.</p>
</div>
<div class="section" id="searching-for-and-downloading-abstracts-using-the-history">
<h3>9.15.2  Searching for and downloading abstracts using the history</h3>
<p>Here is another history example, searching for papers published in the
last year about the <em>Opuntia</em>, and then downloading them into a file in
MedLine format:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;history.user&#64;example.com&quot;</span>
<span class="name">search_results</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span>
                                            <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Opuntia[ORGN]&quot;</span><span class="punctuation">,</span>
                                            <span class="name">reldate</span><span class="operator">=</span><span class="literal number integer">365</span><span class="punctuation">,</span> <span class="name">datetype</span><span class="operator">=</span><span class="literal string">&quot;pdat&quot;</span><span class="punctuation">,</span>
                                            <span class="name">usehistory</span><span class="operator">=</span><span class="literal string">&quot;y&quot;</span><span class="punctuation">))</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name builtin">int</span><span class="punctuation">(</span><span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">])</span>
<span class="keyword">print</span> <span class="literal string">&quot;Found </span><span class="literal string interpol">%i</span><span class="literal string"> results&quot;</span> <span class="operator">%</span> <span class="name">count</span>

<span class="name">batch_size</span> <span class="operator">=</span> <span class="literal number integer">10</span>
<span class="name">out_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;recent_orchid_papers.txt&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">start</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="name">count</span><span class="punctuation">,</span><span class="name">batch_size</span><span class="punctuation">):</span>
    <span class="name">end</span> <span class="operator">=</span> <span class="name builtin">min</span><span class="punctuation">(</span><span class="name">count</span><span class="punctuation">,</span> <span class="name">start</span><span class="operator">+</span><span class="name">batch_size</span><span class="punctuation">)</span>
    <span class="keyword">print</span> <span class="literal string">&quot;Going to download record </span><span class="literal string interpol">%i</span><span class="literal string"> to </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">start</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">end</span><span class="punctuation">)</span>
    <span class="name">fetch_handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span>
                                 <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;medline&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">,</span>
                                 <span class="name">retstart</span><span class="operator">=</span><span class="name">start</span><span class="punctuation">,</span> <span class="name">retmax</span><span class="operator">=</span><span class="name">batch_size</span><span class="punctuation">,</span>
                                 <span class="name">webenv</span><span class="operator">=</span><span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;WebEnv&quot;</span><span class="punctuation">],</span>
                                 <span class="name">query_key</span><span class="operator">=</span><span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;QueryKey&quot;</span><span class="punctuation">])</span>
    <span class="name">data</span> <span class="operator">=</span> <span class="name">fetch_handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
    <span class="name">fetch_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
    <span class="name">out_handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
<span class="name">out_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>At the time of writing, this gave 28 matches - but because this is a
date dependent search, this will of course vary. As described in
Section <a class="reference external" href="#subsec:entrez-and-medline">9.12.1</a> above, you can then
use <tt class="docutils literal">Bio.Medline</tt> to parse the saved records.</p>
</div>
<div class="section" id="searching-for-citations">
<h3>9.15.3  Searching for citations</h3>
<p>Back in Section <a class="reference external" href="#sec:elink">9.7</a> we mentioned ELink can be used to
search for citations of a given paper. Unfortunately this only covers
journals indexed for PubMed Central (doing it for all the journals in
PubMed would mean a lot more work for the NIH). Let’s try this for the
Biopython PDB parser paper, PubMed ID 14630660:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pmid</span> <span class="operator">=</span> <span class="literal string">&quot;14630660&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">results</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">elink</span><span class="punctuation">(</span><span class="name">dbfrom</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pmc&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                    <span class="name">LinkName</span><span class="operator">=</span><span class="literal string">&quot;pubmed_pmc_refs&quot;</span><span class="punctuation">,</span> <span class="name">from_uid</span><span class="operator">=</span><span class="name">pmid</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pmc_ids</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">link</span><span class="punctuation">[</span><span class="literal string">&quot;Id&quot;</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name">link</span> <span class="operator word">in</span> <span class="name">results</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Link&quot;</span><span class="punctuation">]]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pmc_ids</span>
<span class="punctuation">[</span><span class="literal string">'2744707'</span><span class="punctuation">,</span> <span class="literal string">'2705363'</span><span class="punctuation">,</span> <span class="literal string">'2682512'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'1190160'</span><span class="punctuation">]</span>
</pre>
<p>Great - eleven articles. But why hasn’t the Biopython application note
been found (PubMed ID 19304878)? Well, as you might have guessed from
the variable names, there are not actually PubMed IDs, but PubMed
Central IDs. Our application note is the third citing paper in that
list, PMCID 2682512.</p>
<p>So, what if (like me) you’d rather get back a list of PubMed IDs? Well
we can call ELink again to translate them. This becomes a two step
process, so by now you should expect to use the history feature to
accomplish it (Section <a class="reference external" href="#sec:entrez-webenv">9.15</a>).</p>
<p>But first, taking the more straightforward approach of making a second
(separate) call to ELink:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">results2</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">elink</span><span class="punctuation">(</span><span class="name">dbfrom</span><span class="operator">=</span><span class="literal string">&quot;pmc&quot;</span><span class="punctuation">,</span> <span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name">LinkName</span><span class="operator">=</span><span class="literal string">&quot;pmc_pubmed&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                     <span class="name">from_uid</span><span class="operator">=</span><span class="literal string">&quot;,&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">pmc_ids</span><span class="punctuation">)))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pubmed_ids</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">link</span><span class="punctuation">[</span><span class="literal string">&quot;Id&quot;</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name">link</span> <span class="operator word">in</span> <span class="name">results2</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Link&quot;</span><span class="punctuation">]]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pubmed_ids</span>
<span class="punctuation">[</span><span class="literal string">'19698094'</span><span class="punctuation">,</span> <span class="literal string">'19450287'</span><span class="punctuation">,</span> <span class="literal string">'19304878'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'15985178'</span><span class="punctuation">]</span>
</pre>
<p>This time you can immediately spot the Biopython application note as the
third hit (PubMed ID 19304878).</p>
<p>Now, let’s do that all again but with the history … <em>TODO</em>.</p>
<p>And finally, don’t forget to include your <em>own</em> email address in the
Entrez calls.</p>
</div>
</div>
</div>
<div class="section" id="chapter-10-swiss-prot-and-expasy">
<h1>Chapter 10  Swiss-Prot and ExPASy</h1>
<div class="section" id="parsing-swiss-prot-files">
<h2>10.1  Parsing Swiss-Prot files</h2>
<p>Swiss-Prot
(<a class="reference external" href="http://www.expasy.org/sprot">http://www.expasy.org/sprot</a>) is a
hand-curated database of protein sequences. Biopython can parse the
“plain text” Swiss-Prot file format, which is still used for the UniProt
Knowledgebase which combined Swiss-Prot, TrEMBL and PIR-PSD. We do not
(yet) support the UniProtKB XML file format.</p>
<div class="section" id="parsing-swiss-prot-records">
<h3>10.1.1  Parsing Swiss-Prot records</h3>
<p>In Section <a class="reference external" href="#sec:SeqIO_ExPASy_and_SwissProt">5.3.2</a>, we described
how to extract the sequence of a Swiss-Prot record as a <tt class="docutils literal">SeqRecord</tt>
object. Alternatively, you can store the Swiss-Prot record in a
<tt class="docutils literal">Bio.SwissProt.Record</tt> object, which in fact stores the complete
information contained in the Swiss-Prot record. In this Section, we
describe how to extract <tt class="docutils literal">Bio.SwissProt.Record</tt> objects from a
Swiss-Prot file.</p>
<p>To parse a Swiss-Prot record, we first get a handle to a Swiss-Prot
record. There are several ways to do so, depending on where and how the
Swiss-Prot record is stored:</p>
<ul>
<li><p class="first">Open a Swiss-Prot file locally:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;myswissprotfile.dat&quot;</span><span class="punctuation">)</span>
</pre>
</li>
<li><p class="first">Open a gzipped Swiss-Prot file:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">gzip</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">gzip</span><span class="operator">.</span><span class="name">open</span><span class="punctuation">(</span><span class="literal string">&quot;myswissprotfile.dat.gz&quot;</span><span class="punctuation">)</span>
</pre>
</li>
<li><p class="first">Open a Swiss-Prot file over the internet:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">urllib</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">urllib</span><span class="operator">.</span><span class="name">urlopen</span><span class="punctuation">(</span><span class="literal string">&quot;http://www.somelocation.org/data/someswissprotfile.dat&quot;</span><span class="punctuation">)</span>
</pre>
</li>
<li><p class="first">Open a Swiss-Prot file over the internet from the ExPASy database
(see section <a class="reference external" href="#subsec:expasy_swissprot">10.5.1</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_sprot_raw</span><span class="punctuation">(</span><span class="name">myaccessionnumber</span><span class="punctuation">)</span>
</pre>
</li>
</ul>
<p>The key point is that for the parser, it doesn’t matter how the handle
was created, as long as it points to data in the Swiss-Prot format.</p>
<p>We can use <tt class="docutils literal">Bio.SeqIO</tt> as described in
Section <a class="reference external" href="#sec:SeqIO_ExPASy_and_SwissProt">5.3.2</a> to get file format
agnostic <tt class="docutils literal">SeqRecord</tt> objects. Alternatively, we can use
<tt class="docutils literal">Bio.SwissProt</tt> get <tt class="docutils literal">Bio.SwissProt.Record</tt> objects, which are a much
closer match to the underlying file format.</p>
<p>To read one Swiss-Prot record from the handle, we use the function
<tt class="docutils literal">read()</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SwissProt</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SwissProt</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>This function should be used if the handle points to exactly one
Swiss-Prot record. It raises a <tt class="docutils literal">ValueError</tt> if no Swiss-Prot record
was found, and also if more than one record was found.</p>
<p>We can now print out some information about this record:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">description</span>
<span class="literal string">'RecName: Full=Chalcone synthase 3; EC=2.3.1.74; AltName: Full=Naringenin-chalcone synthase 3;'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">ref</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">references</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;authors:&quot;</span><span class="punctuation">,</span> <span class="name">ref</span><span class="operator">.</span><span class="name">authors</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;title:&quot;</span><span class="punctuation">,</span> <span class="name">ref</span><span class="operator">.</span><span class="name">title</span>
<span class="operator">...</span>
<span class="name">authors</span><span class="punctuation">:</span> <span class="name">Liew</span> <span class="name">C</span><span class="operator">.</span><span class="name">F</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Lim</span> <span class="name">S</span><span class="operator">.</span><span class="name">H</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Loh</span> <span class="name">C</span><span class="operator">.</span><span class="name">S</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Goh</span> <span class="name">C</span><span class="operator">.</span><span class="name">J</span><span class="operator">.</span><span class="punctuation">;</span>
<span class="name">title</span><span class="punctuation">:</span> <span class="literal string">&quot;Molecular cloning and sequence analysis of chalcone synthase cDNAs of</span>
<span class="name">Bromheadia</span> <span class="name">finlaysoniana</span><span class="operator">.</span><span class="literal string">&quot;;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">organism_classification</span>
<span class="punctuation">[</span><span class="literal string">'Eukaryota'</span><span class="punctuation">,</span> <span class="literal string">'Viridiplantae'</span><span class="punctuation">,</span> <span class="literal string">'Streptophyta'</span><span class="punctuation">,</span> <span class="literal string">'Embryophyta'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Bromheadia'</span><span class="punctuation">]</span>
</pre>
<p>To parse a file that contains more than one Swiss-Prot record, we use
the <tt class="docutils literal">parse</tt> function instead. This function allows us to iterate over
the records in the file.</p>
<p>For example, let’s parse the full Swiss-Prot database and collect all
the descriptions. You can download this from the <a class="reference external" href="ftp://ftp.expasy.org/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz">ExPAYs FTP
site</a>
as a single gzipped-file <tt class="docutils literal">uniprot_sprot.dat.gz</tt> (about 300MB). This is
a compressed file containing a single file, <tt class="docutils literal">uniprot_sprot.dat</tt> (over
1.5GB).</p>
<p>As described at the start of this section, you can use the Python
library <tt class="docutils literal">gzip</tt> to open and uncompress a <tt class="docutils literal">.gz</tt> file, like this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">gzip</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">gzip</span><span class="operator">.</span><span class="name">open</span><span class="punctuation">(</span><span class="literal string">&quot;uniprot_sprot.dat.gz&quot;</span><span class="punctuation">)</span>
</pre>
<p>However, uncompressing a large file takes time, and each time you open
the file for reading in this way, it has to be decompressed on the fly.
So, if you can spare the disk space you’ll save time in the long run if
you first decompress the file to disk, to get the <tt class="docutils literal">uniprot_sprot.dat</tt>
file inside. Then you can open the file for reading as usual:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;uniprot_sprot.dat&quot;</span><span class="punctuation">)</span>
</pre>
<p>As of June 2009, the full Swiss-Prot database downloaded from ExPASy
contained 468851 Swiss-Prot records. One concise way to build up a list
of the record descriptions is with a list comprehension:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SwissProt</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;uniprot_sprot.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">descriptions</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">record</span><span class="operator">.</span><span class="name">description</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SwissProt</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">descriptions</span><span class="punctuation">)</span>
<span class="literal number integer">468851</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">descriptions</span><span class="punctuation">[:</span><span class="literal number integer">5</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'RecName: Full=Protein MGF 100-1R;'</span><span class="punctuation">,</span>
 <span class="literal string">'RecName: Full=Protein MGF 100-1R;'</span><span class="punctuation">,</span>
 <span class="literal string">'RecName: Full=Protein MGF 100-1R;'</span><span class="punctuation">,</span>
 <span class="literal string">'RecName: Full=Protein MGF 100-1R;'</span><span class="punctuation">,</span>
 <span class="literal string">'RecName: Full=Protein MGF 100-2L;'</span><span class="punctuation">]</span>
</pre>
<p>Or, using a for loop over the record iterator:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SwissProt</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">descriptions</span> <span class="operator">=</span> <span class="punctuation">[]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;uniprot_sprot.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SwissProt</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="name">descriptions</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">description</span><span class="punctuation">)</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">descriptions</span><span class="punctuation">)</span>
<span class="literal number integer">468851</span>
</pre>
<p>Because this is such a large input file, either way takes about eleven
minutes on my new desktop computer (using the uncompressed
<tt class="docutils literal">uniprot_sprot.dat</tt> file as input).</p>
<p>It is equally easy to extract any kind of information you’d like from
Swiss-Prot records. To see the members of a Swiss-Prot record, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">dir</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="punctuation">[</span><span class="literal string">'__doc__'</span><span class="punctuation">,</span> <span class="literal string">'__init__'</span><span class="punctuation">,</span> <span class="literal string">'__module__'</span><span class="punctuation">,</span> <span class="literal string">'accessions'</span><span class="punctuation">,</span> <span class="literal string">'annotation_update'</span><span class="punctuation">,</span>
<span class="literal string">'comments'</span><span class="punctuation">,</span> <span class="literal string">'created'</span><span class="punctuation">,</span> <span class="literal string">'cross_references'</span><span class="punctuation">,</span> <span class="literal string">'data_class'</span><span class="punctuation">,</span> <span class="literal string">'description'</span><span class="punctuation">,</span>
<span class="literal string">'entry_name'</span><span class="punctuation">,</span> <span class="literal string">'features'</span><span class="punctuation">,</span> <span class="literal string">'gene_name'</span><span class="punctuation">,</span> <span class="literal string">'host_organism'</span><span class="punctuation">,</span> <span class="literal string">'keywords'</span><span class="punctuation">,</span>
<span class="literal string">'molecule_type'</span><span class="punctuation">,</span> <span class="literal string">'organelle'</span><span class="punctuation">,</span> <span class="literal string">'organism'</span><span class="punctuation">,</span> <span class="literal string">'organism_classification'</span><span class="punctuation">,</span>
<span class="literal string">'references'</span><span class="punctuation">,</span> <span class="literal string">'seqinfo'</span><span class="punctuation">,</span> <span class="literal string">'sequence'</span><span class="punctuation">,</span> <span class="literal string">'sequence_length'</span><span class="punctuation">,</span>
<span class="literal string">'sequence_update'</span><span class="punctuation">,</span> <span class="literal string">'taxonomy_id'</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="parsing-the-swiss-prot-keyword-and-category-list">
<h3>10.1.2  Parsing the Swiss-Prot keyword and category list</h3>
<p>Swiss-Prot also distributes a file <tt class="docutils literal">keywlist.txt</tt>, which lists the
keywords and categories used in Swiss-Prot. The file contains entries in
the following form:</p>
<pre class="code python literal-block">
<span class="name">ID</span>   <span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span><span class="operator">.</span>
<span class="name">AC</span>   <span class="name">KW</span><span class="operator">-</span><span class="literal number oct">0001</span>
<span class="name">DE</span>   <span class="name">Protein</span> <span class="name">which</span> <span class="name">contains</span> <span class="name">at</span> <span class="name">least</span> <span class="name">one</span> <span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span> <span class="name">iron</span><span class="operator">-</span><span class="name">sulfur</span> <span class="name">cluster</span><span class="punctuation">:</span> <span class="literal number integer">2</span> <span class="name">iron</span>
<span class="name">DE</span>   <span class="name">atoms</span> <span class="name">complexed</span> <span class="name">to</span> <span class="literal number integer">2</span> <span class="name">inorganic</span> <span class="name">sulfides</span> <span class="operator word">and</span> <span class="literal number integer">4</span> <span class="name">sulfur</span> <span class="name">atoms</span> <span class="name">of</span>
<span class="name">DE</span>   <span class="name">cysteines</span> <span class="keyword namespace">from</span> <span class="name namespace">the</span> <span class="name namespace">protein.</span>
<span class="name">SY</span>   <span class="name">Fe2S2</span><span class="punctuation">;</span> <span class="punctuation">[</span><span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span><span class="punctuation">]</span> <span class="name">cluster</span><span class="punctuation">;</span> <span class="punctuation">[</span><span class="name">Fe2S2</span><span class="punctuation">]</span> <span class="name">cluster</span><span class="punctuation">;</span> <span class="name">Fe2</span><span class="operator">/</span><span class="name">S2</span> <span class="punctuation">(</span><span class="name">inorganic</span><span class="punctuation">)</span> <span class="name">cluster</span><span class="punctuation">;</span>
<span class="name">SY</span>   <span class="name">Di</span><span class="operator">-</span><span class="name">mu</span><span class="operator">-</span><span class="name">sulfido</span><span class="operator">-</span><span class="name">diiron</span><span class="punctuation">;</span> <span class="literal number integer">2</span> <span class="name">iron</span><span class="punctuation">,</span> <span class="literal number integer">2</span> <span class="name">sulfur</span> <span class="name">cluster</span> <span class="name">binding</span><span class="operator">.</span>
<span class="name">GO</span>   <span class="name">GO</span><span class="punctuation">:</span><span class="literal number oct">0051537</span><span class="punctuation">;</span> <span class="literal number integer">2</span> <span class="name">iron</span><span class="punctuation">,</span> <span class="literal number integer">2</span> <span class="name">sulfur</span> <span class="name">cluster</span> <span class="name">binding</span>
<span class="name">HI</span>   <span class="name">Ligand</span><span class="punctuation">:</span> <span class="name">Iron</span><span class="punctuation">;</span> <span class="name">Iron</span><span class="operator">-</span><span class="name">sulfur</span><span class="punctuation">;</span> <span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span><span class="operator">.</span>
<span class="name">HI</span>   <span class="name">Ligand</span><span class="punctuation">:</span> <span class="name">Metal</span><span class="operator">-</span><span class="name">binding</span><span class="punctuation">;</span> <span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span><span class="operator">.</span>
<span class="name">CA</span>   <span class="name">Ligand</span><span class="operator">.</span>
<span class="operator">//</span>
<span class="name">ID</span>   <span class="literal number integer">3</span><span class="name">D</span><span class="operator">-</span><span class="name">structure</span><span class="operator">.</span>
<span class="name">AC</span>   <span class="name">KW</span><span class="operator">-</span><span class="literal number oct">0002</span>
<span class="name">DE</span>   <span class="name">Protein</span><span class="punctuation">,</span> <span class="operator word">or</span> <span class="name">part</span> <span class="name">of</span> <span class="name">a</span> <span class="name">protein</span><span class="punctuation">,</span> <span class="name">whose</span> <span class="name">three</span><span class="operator">-</span><span class="name">dimensional</span> <span class="name">structure</span> <span class="name">has</span>
<span class="name">DE</span>   <span class="name">been</span> <span class="name">resolved</span> <span class="name">experimentally</span> <span class="punctuation">(</span><span class="keyword">for</span> <span class="name">example</span> <span class="name">by</span> <span class="name">X</span><span class="operator">-</span><span class="name">ray</span> <span class="name">crystallography</span> <span class="operator word">or</span>
<span class="name">DE</span>   <span class="name">NMR</span> <span class="name">spectroscopy</span><span class="punctuation">)</span> <span class="operator word">and</span> <span class="name">whose</span> <span class="name">coordinates</span> <span class="name">are</span> <span class="name">available</span> <span class="operator word">in</span> <span class="name">the</span> <span class="name">PDB</span>
<span class="name">DE</span>   <span class="name">database</span><span class="operator">.</span> <span class="name">Can</span> <span class="name">also</span> <span class="name">be</span> <span class="name">used</span> <span class="keyword">for</span> <span class="name">theoretical</span> <span class="name">models</span><span class="operator">.</span>
<span class="name">HI</span>   <span class="name">Technical</span> <span class="name">term</span><span class="punctuation">:</span> <span class="literal number integer">3</span><span class="name">D</span><span class="operator">-</span><span class="name">structure</span><span class="operator">.</span>
<span class="name">CA</span>   <span class="name">Technical</span> <span class="name">term</span><span class="operator">.</span>
<span class="operator">//</span>
<span class="name">ID</span>   <span class="literal number integer">3</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">4</span><span class="name">S</span><span class="operator">.</span>
<span class="operator">...</span>
</pre>
<p>The entries in this file can be parsed by the <tt class="docutils literal">parse</tt> function in the
<tt class="docutils literal">Bio.SwissProt.KeyWList</tt> module. Each entry is then stored as a
<tt class="docutils literal">Bio.SwissProt.KeyWList.Record</tt>, which is a Python dictionary.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SwissProt</span> <span class="keyword namespace">import</span> <span class="name">KeyWList</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;keywlist.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">KeyWList</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">'ID'</span><span class="punctuation">]</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">'DE'</span><span class="punctuation">]</span>
</pre>
<p>This prints</p>
<pre class="code python literal-block">
<span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span><span class="operator">.</span>
<span class="name">Protein</span> <span class="name">which</span> <span class="name">contains</span> <span class="name">at</span> <span class="name">least</span> <span class="name">one</span> <span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span> <span class="name">iron</span><span class="operator">-</span><span class="name">sulfur</span> <span class="name">cluster</span><span class="punctuation">:</span> <span class="literal number integer">2</span> <span class="name">iron</span> <span class="name">atoms</span>
<span class="name">complexed</span> <span class="name">to</span> <span class="literal number integer">2</span> <span class="name">inorganic</span> <span class="name">sulfides</span> <span class="operator word">and</span> <span class="literal number integer">4</span> <span class="name">sulfur</span> <span class="name">atoms</span> <span class="name">of</span> <span class="name">cysteines</span> <span class="keyword namespace">from</span> <span class="name namespace">the</span>
<span class="name">protein</span><span class="operator">.</span>
<span class="operator">...</span>
</pre>
</div>
</div>
<div class="section" id="parsing-prosite-records">
<h2>10.2  Parsing Prosite records</h2>
<p>Prosite is a database containing protein domains, protein families,
functional sites, as well as the patterns and profiles to recognize
them. Prosite was developed in parallel with Swiss-Prot. In Biopython, a
Prosite record is represented by the <tt class="docutils literal">Bio.ExPASy.Prosite.Record</tt>
class, whose members correspond to the different fields in a Prosite
record.</p>
<p>In general, a Prosite file can contain more than one Prosite records.
For example, the full set of Prosite records, which can be downloaded as
a single file (<tt class="docutils literal">prosite.dat</tt>) from the <a class="reference external" href="ftp://ftp.expasy.org/databases/prosite/prosite.dat">ExPASy FTP
site</a>, contains
2073 records (version 20.24 released on 4 December 2007). To parse such
a file, we again make use of an iterator:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Prosite</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;myprositefile.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Prosite</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>We can now take the records one at a time and print out some
information. For example, using the file containing the complete Prosite
database, we’d find</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Prosite</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;prosite.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Prosite</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">accession</span>
<span class="literal string">'PS00001'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'ASN_GLYCOSYLATION'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">pdoc</span>
<span class="literal string">'PDOC00001'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">accession</span>
<span class="literal string">'PS00004'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'CAMP_PHOSPHO_SITE'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">pdoc</span>
<span class="literal string">'PDOC00004'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">accession</span>
<span class="literal string">'PS00005'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'PKC_PHOSPHO_SITE'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">pdoc</span>
<span class="literal string">'PDOC00005'</span>
</pre>
<p>and so on. If you’re interested in how many Prosite records there are,
you could use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Prosite</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;prosite.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Prosite</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">n</span> <span class="operator">=</span> <span class="literal number integer">0</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span> <span class="name">n</span><span class="operator">+=</span><span class="literal number integer">1</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">n</span>
<span class="literal number integer">2073</span>
</pre>
<p>To read exactly one Prosite from the handle, you can use the <tt class="docutils literal">read</tt>
function:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Prosite</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;mysingleprositerecord.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Prosite</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>This function raises a ValueError if no Prosite record is found, and
also if more than one Prosite record is found.</p>
</div>
<div class="section" id="parsing-prosite-documentation-records">
<h2>10.3  Parsing Prosite documentation records</h2>
<p>In the Prosite example above, the <tt class="docutils literal">record.pdoc</tt> accession numbers
<tt class="docutils literal">'PDOC00001'</tt>, <tt class="docutils literal">'PDOC00004'</tt>, <tt class="docutils literal">'PDOC00005'</tt> and so on refer to
Prosite documentation. The Prosite documentation records are available
from ExPASy as individual files, and as one file (<tt class="docutils literal">prosite.doc</tt>)
containing all Prosite documentation records.</p>
<p>We use the parser in <tt class="docutils literal">Bio.ExPASy.Prodoc</tt> to parse Prosite
documentation records. For example, to create a list of all accession
numbers of Prosite documentation record, you can use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Prodoc</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;prosite.doc&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Prodoc</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">accessions</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">record</span><span class="operator">.</span><span class="name">accession</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">]</span>
</pre>
<p>Again a <tt class="docutils literal">read()</tt> function is provided to read exactly one Prosite
documentation record from the handle.</p>
</div>
<div class="section" id="parsing-enzyme-records">
<h2>10.4  Parsing Enzyme records</h2>
<p>ExPASy’s Enzyme database is a repository of information on enzyme
nomenclature. A typical Enzyme record looks as follows:</p>
<pre class="code python literal-block">
<span class="name">ID</span>   <span class="literal number float">3.1</span><span class="operator">.</span><span class="literal number float">1.34</span>
<span class="name">DE</span>   <span class="name">Lipoprotein</span> <span class="name">lipase</span><span class="operator">.</span>
<span class="name">AN</span>   <span class="name">Clearing</span> <span class="name">factor</span> <span class="name">lipase</span><span class="operator">.</span>
<span class="name">AN</span>   <span class="name">Diacylglycerol</span> <span class="name">lipase</span><span class="operator">.</span>
<span class="name">AN</span>   <span class="name">Diglyceride</span> <span class="name">lipase</span><span class="operator">.</span>
<span class="name">CA</span>   <span class="name">Triacylglycerol</span> <span class="operator">+</span> <span class="name">H</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">)</span><span class="name">O</span> <span class="operator">=</span> <span class="name">diacylglycerol</span> <span class="operator">+</span> <span class="name">a</span> <span class="name">carboxylate</span><span class="operator">.</span>
<span class="name">CC</span>   <span class="operator">-</span><span class="error">!</span><span class="operator">-</span> <span class="name">Hydrolyzes</span> <span class="name">triacylglycerols</span> <span class="operator word">in</span> <span class="name">chylomicrons</span> <span class="operator word">and</span> <span class="name">very</span> <span class="name">low</span><span class="operator">-</span><span class="name">density</span>
<span class="name">CC</span>       <span class="name">lipoproteins</span> <span class="punctuation">(</span><span class="name">VLDL</span><span class="punctuation">)</span><span class="operator">.</span>
<span class="name">CC</span>   <span class="operator">-</span><span class="error">!</span><span class="operator">-</span> <span class="name">Also</span> <span class="name">hydrolyzes</span> <span class="name">diacylglycerol</span><span class="operator">.</span>
<span class="name">PR</span>   <span class="name">PROSITE</span><span class="punctuation">;</span> <span class="name">PDOC00110</span><span class="punctuation">;</span>
<span class="name">DR</span>   <span class="name">P11151</span><span class="punctuation">,</span> <span class="name">LIPL_BOVIN</span> <span class="punctuation">;</span>  <span class="name">P11153</span><span class="punctuation">,</span> <span class="name">LIPL_CAVPO</span> <span class="punctuation">;</span>  <span class="name">P11602</span><span class="punctuation">,</span> <span class="name">LIPL_CHICK</span> <span class="punctuation">;</span>
<span class="name">DR</span>   <span class="name">P55031</span><span class="punctuation">,</span> <span class="name">LIPL_FELCA</span> <span class="punctuation">;</span>  <span class="name">P06858</span><span class="punctuation">,</span> <span class="name">LIPL_HUMAN</span> <span class="punctuation">;</span>  <span class="name">P11152</span><span class="punctuation">,</span> <span class="name">LIPL_MOUSE</span> <span class="punctuation">;</span>
<span class="name">DR</span>   <span class="name">O46647</span><span class="punctuation">,</span> <span class="name">LIPL_MUSVI</span> <span class="punctuation">;</span>  <span class="name">P49060</span><span class="punctuation">,</span> <span class="name">LIPL_PAPAN</span> <span class="punctuation">;</span>  <span class="name">P49923</span><span class="punctuation">,</span> <span class="name">LIPL_PIG</span>   <span class="punctuation">;</span>
<span class="name">DR</span>   <span class="name">Q06000</span><span class="punctuation">,</span> <span class="name">LIPL_RAT</span>   <span class="punctuation">;</span>  <span class="name">Q29524</span><span class="punctuation">,</span> <span class="name">LIPL_SHEEP</span> <span class="punctuation">;</span>
<span class="operator">//</span>
</pre>
<p>In this example, the first line shows the EC (Enzyme Commission) number
of lipoprotein lipase (second line). Alternative names of lipoprotein
lipase are &quot;clearing factor lipase&quot;, &quot;diacylglycerol lipase&quot;, and
&quot;diglyceride lipase&quot; (lines 3 through 5). The line starting with &quot;CA&quot;
shows the catalytic activity of this enzyme. Comment lines start with
&quot;CC&quot;. The &quot;PR&quot; line shows references to the Prosite Documentation
records, and the &quot;DR&quot; lines show references to Swiss-Prot records. Not
of these entries are necessarily present in an Enzyme record.</p>
<p>In Biopython, an Enzyme record is represented by the
<tt class="docutils literal">Bio.ExPASy.Enzyme.Record</tt> class. This record derives from a Python
dictionary and has keys corresponding to the two-letter codes used in
Enzyme files. To read an Enzyme file containing one Enzyme record, use
the <tt class="docutils literal">read</tt> function in <tt class="docutils literal">Bio.ExPASy.Enzyme</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Enzyme</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;lipoprotein.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Enzyme</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;ID&quot;</span><span class="punctuation">]</span>
<span class="literal string">'3.1.1.34'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DE&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Lipoprotein lipase.'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;AN&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Clearing factor lipase.'</span><span class="punctuation">,</span> <span class="literal string">'Diacylglycerol lipase.'</span><span class="punctuation">,</span> <span class="literal string">'Diglyceride lipase.'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;CA&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Triacylglycerol + H(2)O = diacylglycerol + a carboxylate.'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;PR&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'PDOC00110'</span><span class="punctuation">]</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;CC&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Hydrolyzes triacylglycerols in chylomicrons and very low-density lipoproteins</span>
<span class="punctuation">(</span><span class="name">VLDL</span><span class="punctuation">)</span><span class="operator">.</span><span class="literal string">', '</span><span class="name">Also</span> <span class="name">hydrolyzes</span> <span class="name">diacylglycerol</span><span class="operator">.</span><span class="literal string">']</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DR&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[[</span><span class="literal string">'P11151'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_BOVIN'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'P11153'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_CAVPO'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'P11602'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_CHICK'</span><span class="punctuation">],</span>
<span class="punctuation">[</span><span class="literal string">'P55031'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_FELCA'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'P06858'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_HUMAN'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'P11152'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_MOUSE'</span><span class="punctuation">],</span>
<span class="punctuation">[</span><span class="literal string">'O46647'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_MUSVI'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'P49060'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_PAPAN'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'P49923'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_PIG'</span><span class="punctuation">],</span>
<span class="punctuation">[</span><span class="literal string">'Q06000'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_RAT'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'Q29524'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_SHEEP'</span><span class="punctuation">]]</span>
</pre>
<p>The <tt class="docutils literal">read</tt> function raises a ValueError if no Enzyme record is found,
and also if more than one Enzyme record is found.</p>
<p>The full set of Enzyme records can be downloaded as a single file
(<tt class="docutils literal">enzyme.dat</tt>) from the <a class="reference external" href="ftp://ftp.expasy.org/databases/enzyme/enzyme.dat">ExPASy FTP
site</a>, containing
4877 records (release of 3 March 2009). To parse such a file containing
multiple Enzyme records, use the <tt class="docutils literal">parse</tt> function in
<tt class="docutils literal">Bio.ExPASy.Enzyme</tt> to obtain an iterator:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Enzyme</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;enzyme.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Enzyme</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>We can now iterate over the records one at a time. For example, we can
make a list of all EC numbers for which an Enzyme record is available:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ecnumbers</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;ID&quot;</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="accessing-the-expasy-server">
<h2>10.5  Accessing the ExPASy server</h2>
<p>Swiss-Prot, Prosite, and Prosite documentation records can be downloaded
from the ExPASy web server at
<a class="reference external" href="http://www.expasy.org">http://www.expasy.org</a>. Six kinds of
queries are available from ExPASy:</p>
<dl class="docutils">
<dt><strong>get_prodoc_entry</strong></dt>
<dd>To download a Prosite documentation record in HTML format</dd>
<dt><strong>get_prosite_entry</strong></dt>
<dd>To download a Prosite record in HTML format</dd>
<dt><strong>get_prosite_raw</strong></dt>
<dd>To download a Prosite or Prosite documentation record in raw format</dd>
<dt><strong>get_sprot_raw</strong></dt>
<dd>To download a Swiss-Prot record in raw format</dd>
<dt><strong>sprot_search_ful</strong></dt>
<dd>To search for a Swiss-Prot record</dd>
<dt><strong>sprot_search_de</strong></dt>
<dd>To search for a Swiss-Prot record</dd>
</dl>
<p>To access this web server from a Python script, we use the
<tt class="docutils literal">Bio.ExPASy</tt> module.</p>
<div class="section" id="retrieving-a-swiss-prot-record">
<h3>10.5.1  Retrieving a Swiss-Prot record</h3>
<p>Let’s say we are looking at chalcone synthases for Orchids (see
section <a class="reference external" href="#sec:orchids">2.3</a> for some justification for looking for
interesting things about orchids). Chalcone synthase is involved in
flavanoid biosynthesis in plants, and flavanoids make lots of cool
things like pigment colors and UV protectants.</p>
<p>If you do a search on Swiss-Prot, you can find three orchid proteins for
Chalcone Synthase, id numbers O23729, O23730, O23731. Now, let’s write a
script which grabs these, and parses out some interesting information.</p>
<p>First, we grab the records, using the <tt class="docutils literal">get_sprot_raw()</tt> function of
<tt class="docutils literal">Bio.ExPASy</tt>. This function is very nice since you can feed it an id
and get back a handle to a raw text record (no html to mess with!). We
can the use <tt class="docutils literal">Bio.SwissProt.read</tt> to pull out the Swiss-Prot record, or
<tt class="docutils literal">Bio.SeqIO.read</tt> to get a SeqRecord. The following code accomplishes
what I just wrote:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SwissProt</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="name">accessions</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">&quot;O23729&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;O23730&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;O23731&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">[]</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">accession</span> <span class="operator word">in</span> <span class="name">accessions</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_sprot_raw</span><span class="punctuation">(</span><span class="name">accession</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="name">record</span> <span class="operator">=</span> <span class="name">SwissProt</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="name">records</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
</pre>
<p>If the accession number you provided to <tt class="docutils literal">ExPASy.get_sprot_raw</tt> does
not exist, then <tt class="docutils literal">SwissProt.read(handle)</tt> will raise a <tt class="docutils literal">ValueError</tt>.
You can catch <tt class="docutils literal">ValueException</tt> exceptions to detect invalid accession
numbers:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">accession</span> <span class="operator word">in</span> <span class="name">accessions</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_sprot_raw</span><span class="punctuation">(</span><span class="name">accession</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="keyword">try</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="name">record</span> <span class="operator">=</span> <span class="name">SwissProt</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="keyword">except</span> <span class="name">ValueException</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="literal string">&quot;WARNING: Accession </span><span class="literal string interpol">%s</span><span class="literal string"> not found&quot;</span> <span class="operator">%</span> <span class="name">accession</span>
<span class="operator">...</span>     <span class="name">records</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="searching-swiss-prot">
<h3>10.5.2  Searching Swiss-Prot</h3>
<p>Now, you may remark that I knew the records’ accession numbers
beforehand. Indeed, <tt class="docutils literal">get_sprot_raw()</tt> needs either the entry name or
an accession number. When you don’t have them handy, you can use one of
the <tt class="docutils literal">sprot_search_de()</tt> or <tt class="docutils literal">sprot_search_ful()</tt> functions.</p>
<p><tt class="docutils literal">sprot_search_de()</tt> searches in the ID, DE, GN, OS and OG lines;
<tt class="docutils literal">sprot_search_ful()</tt> searches in (nearly) all the fields. They are
detailed on
<a class="reference external" href="http://www.expasy.org/cgi-bin/sprot-search-de">http://www.expasy.org/cgi-bin/sprot-search-de</a>
and
<a class="reference external" href="http://www.expasy.org/cgi-bin/sprot-search-ful">http://www.expasy.org/cgi-bin/sprot-search-ful</a>
respectively. Note that they don’t search in TrEMBL by default (argument
<tt class="docutils literal">trembl</tt>). Note also that they return html pages; however, accession
numbers are quite easily extractable:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">re</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">sprot_search_de</span><span class="punctuation">(</span><span class="literal string">&quot;Orchid Chalcone Synthase&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="comment"># or:</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="comment"># handle = ExPASy.sprot_search_ful(&quot;Orchid and {Chalcone Synthase}&quot;)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">html_results</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">if</span> <span class="literal string">&quot;Number of sequences found&quot;</span> <span class="operator word">in</span> <span class="name">html_results</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">ids</span> <span class="operator">=</span> <span class="name">re</span><span class="operator">.</span><span class="name">findall</span><span class="punctuation">(</span><span class="literal string">r'HREF=&quot;/uniprot/(\w+)&quot;'</span><span class="punctuation">,</span> <span class="name">html_results</span><span class="punctuation">)</span>
<span class="operator">...</span> <span class="keyword">else</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">ids</span> <span class="operator">=</span> <span class="name">re</span><span class="operator">.</span><span class="name">findall</span><span class="punctuation">(</span><span class="literal string">r'href=&quot;/cgi-bin/niceprot\.pl\?(\w+)&quot;'</span><span class="punctuation">,</span> <span class="name">html_results</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="retrieving-prosite-and-prosite-documentation-records">
<h3>10.5.3  Retrieving Prosite and Prosite documentation records</h3>
<p>Prosite and Prosite documentation records can be retrieved either in
HTML format, or in raw format. To parse Prosite and Prosite
documentation records with Biopython, you should retrieve the records in
raw format. For other purposes, however, you may be interested in these
records in HTML format.</p>
<p>To retrieve a Prosite or Prosite documentation record in raw format, use
<tt class="docutils literal">get_prosite_raw()</tt>. For example, to download a Prosite record and
print it out in raw text format, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_prosite_raw</span><span class="punctuation">(</span><span class="literal string">'PS00001'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">text</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">text</span>
</pre>
<p>To retrieve a Prosite record and parse it into a <tt class="docutils literal">Bio.Prosite.Record</tt>
object, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Prosite</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_prosite_raw</span><span class="punctuation">(</span><span class="literal string">'PS00001'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Prosite</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>The same function can be used to retrieve a Prosite documentation record
and parse it into a <tt class="docutils literal">Bio.ExPASy.Prodoc.Record</tt> object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Prodoc</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_prosite_raw</span><span class="punctuation">(</span><span class="literal string">'PDOC00001'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Prodoc</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>For non-existing accession numbers, <tt class="docutils literal">ExPASy.get_prosite_raw</tt> returns a
handle to an emptry string. When faced with an empty string,
<tt class="docutils literal">Prosite.read</tt> and <tt class="docutils literal">Prodoc.read</tt> will raise a ValueError. You can
catch these exceptions to detect invalid accession numbers.</p>
<p>The functions <tt class="docutils literal">get_prosite_entry()</tt> and <tt class="docutils literal">get_prodoc_entry()</tt> are
used to download Prosite and Prosite documentation records in HTML
format. To create a web page showing one Prosite record, you can use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_prosite_entry</span><span class="punctuation">(</span><span class="literal string">'PS00001'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">html</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">output</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;myprositerecord.html&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">output</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">html</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">output</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>and similarly for a Prosite documentation record:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_prodoc_entry</span><span class="punctuation">(</span><span class="literal string">'PDOC00001'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">html</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">output</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;myprodocrecord.html&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">output</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">html</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">output</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>For these functions, an invalid accession number returns an error
message in HTML format.</p>
</div>
</div>
<div class="section" id="scanning-the-prosite-database">
<h2>10.6  Scanning the Prosite database</h2>
<p><a class="reference external" href="http://www.expasy.org/tools/scanprosite/">ScanProsite</a> allows you to
scan protein sequences online against the Prosite database by providing
a UniProt or PDB sequence identifier or the sequence itself. For more
information about ScanProsite, please see the <a class="reference external" href="http://www.expasy.org/tools/scanprosite/scanprosite-doc.html">ScanProsite
documentation</a>
as well as the <a class="reference external" href="http://www.expasy.org/tools/scanprosite/ScanPrositeREST.html">documentation for programmatic access of
ScanProsite</a>.</p>
<p>You can use Biopython’s <tt class="docutils literal">Bio.ExPASy.ScanProsite</tt> module to scan the
Prosite database from Python. This module both helps you to access
ScanProsite programmatically, and to parse the results returned by
ScanProsite. To scan for Prosite patterns in the following protein
sequence:</p>
<pre class="code python literal-block">
<span class="name">MEHKEVVLLLLLFLKSGQGEPLDDYVNTQGASLFSVTKKQLGAGSIEECAAKCEEDEEFT</span>
<span class="name">CRAFQYHSKEQQCVIMAENRKSSIIIRMRDVVLFEKKVYLSECKTGNGKNYRGTMSKTKN</span>
</pre>
<p>you can use the following code:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sequence</span> <span class="operator">=</span> <span class="literal string">&quot;MEHKEVVLLLLLFLKSGQGEPLDDYVNTQGASLFSVTKKQLGAGSIEECAAKCEEDEEFT</span>
<span class="name">CRAFQYHSKEQQCVIMAENRKSSIIIRMRDVVLFEKKVYLSECKTGNGKNYRGTMSKTKN</span><span class="literal string">&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">ScanProsite</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ScanProsite</span><span class="operator">.</span><span class="name">scan</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">sequence</span><span class="punctuation">)</span>
</pre>
<p>By executing <tt class="docutils literal">handle.read()</tt>, you can obtain the search results in raw
XML format. Instead, let’s use <tt class="docutils literal">Bio.ExPASy.ScanProsite.read</tt> to parse
the raw XML into a Python object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span> <span class="operator">=</span> <span class="name">ScanProsite</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">type</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">)</span>
<span class="operator">&lt;</span><span class="keyword">class</span> <span class="error">'</span><span class="name class">Bio</span><span class="operator">.</span><span class="name">ExPASy</span><span class="operator">.</span><span class="name">ScanProsite</span><span class="operator">.</span><span class="name">Record</span><span class="literal string">'&gt;</span>
</pre>
<p>A <tt class="docutils literal">Bio.ExPASy.ScanProsite.Record</tt> object is derived from a list, with
each element in the list storing one ScanProsite hit. This object also
stores the number of hits, as well as the number of search sequences, as
returned by ScanProsite. This ScanProsite search resulted in six hits:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="operator">.</span><span class="name">n_seq</span>
<span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="operator">.</span><span class="name">n_match</span>
<span class="literal number integer">6</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">)</span>
<span class="literal number integer">6</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'signature_ac'</span><span class="punctuation">:</span> <span class="literal string">u'PS50948'</span><span class="punctuation">,</span> <span class="literal string">'level'</span><span class="punctuation">:</span> <span class="literal string">u'0'</span><span class="punctuation">,</span> <span class="literal string">'stop'</span><span class="punctuation">:</span> <span class="literal number integer">98</span><span class="punctuation">,</span> <span class="literal string">'sequence_ac'</span><span class="punctuation">:</span> <span class="literal string">u'USERSEQ1'</span><span class="punctuation">,</span> <span class="literal string">'start'</span><span class="punctuation">:</span> <span class="literal number integer">16</span><span class="punctuation">,</span> <span class="literal string">'score'</span><span class="punctuation">:</span> <span class="literal string">u'8.873'</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'start'</span><span class="punctuation">:</span> <span class="literal number integer">37</span><span class="punctuation">,</span> <span class="literal string">'stop'</span><span class="punctuation">:</span> <span class="literal number integer">39</span><span class="punctuation">,</span> <span class="literal string">'sequence_ac'</span><span class="punctuation">:</span> <span class="literal string">u'USERSEQ1'</span><span class="punctuation">,</span> <span class="literal string">'signature_ac'</span><span class="punctuation">:</span> <span class="literal string">u'PS00005'</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'start'</span><span class="punctuation">:</span> <span class="literal number integer">45</span><span class="punctuation">,</span> <span class="literal string">'stop'</span><span class="punctuation">:</span> <span class="literal number integer">48</span><span class="punctuation">,</span> <span class="literal string">'sequence_ac'</span><span class="punctuation">:</span> <span class="literal string">u'USERSEQ1'</span><span class="punctuation">,</span> <span class="literal string">'signature_ac'</span><span class="punctuation">:</span> <span class="literal string">u'PS00006'</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'start'</span><span class="punctuation">:</span> <span class="literal number integer">60</span><span class="punctuation">,</span> <span class="literal string">'stop'</span><span class="punctuation">:</span> <span class="literal number integer">62</span><span class="punctuation">,</span> <span class="literal string">'sequence_ac'</span><span class="punctuation">:</span> <span class="literal string">u'USERSEQ1'</span><span class="punctuation">,</span> <span class="literal string">'signature_ac'</span><span class="punctuation">:</span> <span class="literal string">u'PS00005'</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="punctuation">[</span><span class="literal number integer">4</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'start'</span><span class="punctuation">:</span> <span class="literal number integer">80</span><span class="punctuation">,</span> <span class="literal string">'stop'</span><span class="punctuation">:</span> <span class="literal number integer">83</span><span class="punctuation">,</span> <span class="literal string">'sequence_ac'</span><span class="punctuation">:</span> <span class="literal string">u'USERSEQ1'</span><span class="punctuation">,</span> <span class="literal string">'signature_ac'</span><span class="punctuation">:</span> <span class="literal string">u'PS00004'</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="punctuation">[</span><span class="literal number integer">5</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'start'</span><span class="punctuation">:</span> <span class="literal number integer">106</span><span class="punctuation">,</span> <span class="literal string">'stop'</span><span class="punctuation">:</span> <span class="literal number integer">111</span><span class="punctuation">,</span> <span class="literal string">'sequence_ac'</span><span class="punctuation">:</span> <span class="literal string">u'USERSEQ1'</span><span class="punctuation">,</span> <span class="literal string">'signature_ac'</span><span class="punctuation">:</span> <span class="literal string">u'PS00008'</span><span class="punctuation">}</span>
</pre>
<p>Other ScanProsite parameters can be passed as keyword arguments; see the
<a class="reference external" href="http://www.expasy.org/tools/scanprosite/ScanPrositeREST.html">documentation for programmatic access of
ScanProsite</a>
for more information. As an example, passing <tt class="docutils literal">lowscore=1</tt> to include
matches with low level scores lets use find one additional hit:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ScanProsite</span><span class="operator">.</span><span class="name">scan</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">sequence</span><span class="punctuation">,</span> <span class="name">lowscore</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span> <span class="operator">=</span> <span class="name">ScanProsite</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="operator">.</span><span class="name">n_match</span>
<span class="literal number integer">7</span>
</pre>
</div>
</div>
<div class="section" id="chapter-11-going-3d-the-pdb-module">
<h1>Chapter 11  Going 3D: The PDB module</h1>
<p>Bio.PDB is a Biopython module that focuses on working with crystal
structures of biological macromolecules. Among other things, Bio.PDB
includes a PDBParser class that produces a Structure object, which can
be used to access the atomic data in the file in a convenient manner.
There is limited support for parsing the information contained in the
PDB header.</p>
<div class="section" id="reading-and-writing-crystal-structure-files">
<h2>11.1  Reading and writing crystal structure files</h2>
<div class="section" id="reading-a-pdb-file">
<h3>11.1.1  Reading a PDB file</h3>
<p>First we create a <tt class="docutils literal">PDBParser</tt> object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.PDB.PDBParser</span> <span class="keyword namespace">import</span> <span class="name">PDBParser</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">p</span> <span class="operator">=</span> <span class="name">PDBParser</span><span class="punctuation">(</span><span class="name">PERMISSIVE</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>The <tt class="docutils literal">PERMISSIVE</tt> flag indicates that a number of common problems (see
<a class="reference external" href="#problem%20structures">11.7.1</a>) associated with PDB files will be
ignored (but note that some atoms and/or residues will be missing). If
the flag is not present a <tt class="docutils literal">PDBConstructionException</tt> will be generated
if any problems are detected during the parse operation.</p>
<p>The Structure object is then produced by letting the <tt class="docutils literal">PDBParser</tt>
object parse a PDB file (the PDB file in this case is called
’pdb1fat.ent’, ’1fat’ is a user defined name for the structure):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">structure_id</span> <span class="operator">=</span> <span class="literal string">&quot;1fat&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">filename</span> <span class="operator">=</span> <span class="literal string">&quot;pdb1fat.ent&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">s</span> <span class="operator">=</span> <span class="name">p</span><span class="operator">.</span><span class="name">get_structure</span><span class="punctuation">(</span><span class="name">structure_id</span><span class="punctuation">,</span> <span class="name">filename</span><span class="punctuation">)</span>
</pre>
<p>You can extract the header and trailer (simple lists of strings) of the
PDB file from the PDBParser object with the <tt class="docutils literal">get_header</tt> and
<tt class="docutils literal">get_trailer</tt> methods. Note however that many PDB files contain
headers with incomplete or erroneous information. Many of the errors
have been fixed in the equivalent mmCIF files. <em>Hence, if you are
interested in the header information, it is a good idea to extract
information from mmCIF files using the</em> <tt class="docutils literal">MMCIF2Dict</tt> <em>tool described
below, instead of parsing the PDB header.</em></p>
<p>Now that is clarified, let’s return to parsing the PDB header. The
structure object has an attribute called <tt class="docutils literal">header</tt> which is a Python
dictionary that maps header records to their values.</p>
<p>Example:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">resolution</span> <span class="operator">=</span> <span class="name">structure</span><span class="operator">.</span><span class="name">header</span><span class="punctuation">[</span><span class="literal string">'resolution'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">keywords</span> <span class="operator">=</span> <span class="name">structure</span><span class="operator">.</span><span class="name">header</span><span class="punctuation">[</span><span class="literal string">'keywords'</span><span class="punctuation">]</span>
</pre>
<p>The available keys are <tt class="docutils literal">name</tt>, <tt class="docutils literal">head</tt>, <tt class="docutils literal">deposition_date</tt>,
<tt class="docutils literal">release_date</tt>, <tt class="docutils literal">structure_method</tt>, <tt class="docutils literal">resolution</tt>,
<tt class="docutils literal">structure_reference</tt> (which maps to a list of references),
<tt class="docutils literal">journal_reference</tt>, <tt class="docutils literal">author</tt>, and <tt class="docutils literal">compound</tt> (which maps to a
dictionary with various information about the crystallized compound).</p>
<p>The dictionary can also be created without creating a <tt class="docutils literal">Structure</tt>
object, ie. directly from the PDB file:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">file</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span><span class="literal string">'r'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">header_dict</span> <span class="operator">=</span> <span class="name">parse_pdb_header</span><span class="punctuation">(</span><span class="name builtin">file</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">file</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
</div>
<div class="section" id="reading-an-mmcif-file">
<h3>11.1.2  Reading an mmCIF file</h3>
<p>Similarly to the case the case of PDB files, first create an
<tt class="docutils literal">MMCIFParser</tt> object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.PDB.MMCIFParser</span> <span class="keyword namespace">import</span> <span class="name">MMCIFParser</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">parser</span> <span class="operator">=</span> <span class="name">MMCIFParser</span><span class="punctuation">()</span>
</pre>
<p>Then use this parser to create a structure object from the mmCIF file:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">structure</span> <span class="operator">=</span> <span class="name">parser</span><span class="operator">.</span><span class="name">get_structure</span><span class="punctuation">(</span><span class="literal string">'1fat'</span><span class="punctuation">,</span> <span class="literal string">'1fat.cif'</span><span class="punctuation">)</span>
</pre>
<p>To have some more low level access to an mmCIF file, you can use the
<tt class="docutils literal">MMCIF2Dict</tt> class to create a Python dictionary that maps all mmCIF
tags in an mmCIF file to their values. If there are multiple values
(like in the case of tag <tt class="docutils literal">_atom_site.Cartn_y</tt>, which holds the <em>y</em>
coordinates of all atoms), the tag is mapped to a list of values. The
dictionary is created from the mmCIF file as follows:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.PDB.MMCIF2Dict</span> <span class="keyword namespace">import</span> <span class="name">MMCIF2Dict</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mmcif_dict</span> <span class="operator">=</span> <span class="name">MMCIF2Dict</span><span class="punctuation">(</span><span class="literal string">'1FAT.cif'</span><span class="punctuation">)</span>
</pre>
<p>Example: get the solvent content from an mmCIF file:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sc</span> <span class="operator">=</span> <span class="name">mmcif_dict</span><span class="punctuation">[</span><span class="literal string">'_exptl_crystal.density_percent_sol'</span><span class="punctuation">]</span>
</pre>
<p>Example: get the list of the <em>y</em> coordinates of all atoms</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">y_list</span> <span class="operator">=</span> <span class="name">mmcif_dict</span><span class="punctuation">[</span><span class="literal string">'_atom_site.Cartn_y'</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="reading-files-in-the-pdb-xml-format">
<h3>11.1.3  Reading files in the PDB XML format</h3>
<p>That’s not yet supported, but we are definitely planning to support that
in the future (it’s not a lot of work). Contact the Biopython developers
(<a class="reference external" href="mailto:biopython-dev&#64;biopython.org">biopython-dev&#64;biopython.org</a>)
if you need this).</p>
</div>
<div class="section" id="writing-pdb-files">
<h3>11.1.4  Writing PDB files</h3>
<p>Use the PDBIO class for this. It’s easy to write out specific parts of a
structure too, of course.</p>
<p>Example: saving a structure</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">io</span> <span class="operator">=</span> <span class="name">PDBIO</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">io</span><span class="operator">.</span><span class="name">set_structure</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">io</span><span class="operator">.</span><span class="name">save</span><span class="punctuation">(</span><span class="literal string">'out.pdb'</span><span class="punctuation">)</span>
</pre>
<p>If you want to write out a part of the structure, make use of the
<tt class="docutils literal">Select</tt> class (also in <tt class="docutils literal">PDBIO</tt>). Select has four methods:</p>
<ul class="simple">
<li><tt class="docutils literal">accept_model(model)</tt></li>
<li><tt class="docutils literal">accept_chain(chain)</tt></li>
<li><tt class="docutils literal">accept_residue(residue)</tt></li>
<li><tt class="docutils literal">accept_atom(atom)</tt></li>
</ul>
<p>By default, every method returns 1 (which means the
model/chain/residue/atom is included in the output). By subclassing
<tt class="docutils literal">Select</tt> and returning 0 when appropriate you can exclude models,
chains, etc. from the output. Cumbersome maybe, but very powerful. The
following code only writes out glycine residues:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">class</span> <span class="name class">GlySelect</span><span class="punctuation">(</span><span class="name">Select</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">def</span> <span class="name function">accept_residue</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">,</span> <span class="name">residue</span><span class="punctuation">):</span>
<span class="operator">...</span>         <span class="keyword">if</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_name</span><span class="punctuation">()</span><span class="operator">==</span><span class="literal string">'GLY'</span><span class="punctuation">:</span>
<span class="operator">...</span>             <span class="keyword">return</span> <span class="name builtin pseudo">True</span>
<span class="operator">...</span>         <span class="keyword">else</span><span class="punctuation">:</span>
<span class="operator">...</span>             <span class="keyword">return</span> <span class="name builtin pseudo">False</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">io</span> <span class="operator">=</span> <span class="name">PDBIO</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">io</span><span class="operator">.</span><span class="name">set_structure</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">io</span><span class="operator">.</span><span class="name">save</span><span class="punctuation">(</span><span class="literal string">'gly_only.pdb'</span><span class="punctuation">,</span> <span class="name">GlySelect</span><span class="punctuation">())</span>
</pre>
<p>If this is all too complicated for you, the <tt class="docutils literal">Dice</tt> module contains a
handy <tt class="docutils literal">extract</tt> function that writes out all residues in a chain
between a start and end residue.</p>
</div>
</div>
<div class="section" id="structure-representation">
<h2>11.2  Structure representation</h2>
<p>The overall layout of a <tt class="docutils literal">Structure</tt> object follows the so-called SMCRA
(Structure/Model/Chain/Residue/Atom) architecture:</p>
<ul class="simple">
<li>A structure consists of models</li>
<li>A model consists of chains</li>
<li>A chain consists of residues</li>
<li>A residue consists of atoms</li>
</ul>
<p>This is the way many structural biologists/bioinformaticians think about
structure, and provides a simple but efficient way to deal with
structure. Additional stuff is essentially added when needed. A UML
diagram of the <tt class="docutils literal">Structure</tt> object (forget about the <tt class="docutils literal">Disordered</tt>
classes for now) is shown in Fig. <a class="reference external" href="#fig:smcra">11.1</a>. Such a data
structure is not necessarily best suited for the representation of the
macromolecular content of a structure, but it is absolutely necessary
for a good interpretation of the data present in a file that describes
the structure (typically a PDB or MMCIF file). If this hierarchy cannot
represent the contents of a structure file, it is fairly certain that
the file contains an error or at least does not describe the structure
unambiguously. If a SMCRA data structure cannot be generated, there is
reason to suspect a problem. Parsing a PDB file can thus be used to
detect likely problems. We will give several examples of this in section
<a class="reference external" href="#problem%20structures">11.7.1</a>.</p>
<hr class="docutils" />
<p><img alt="image3" src="../images/smcra.png" /></p>
<p>Figure 11.1: UML diagram of SMCRA architecture of the <tt class="docutils literal">Structure</tt> class used to represent a macromolecular structure. Full lines with diamonds denote aggregation, full lines with arrows denote referencing, full lines with triangles denote inheritance and dashed lines with triangles denote interface realization.</p>
<hr class="docutils" />
<p>Structure, Model, Chain and Residue are all subclasses of the Entity
base class. The Atom class only (partly) implements the Entity interface
(because an Atom does not have children).</p>
<p>For each Entity subclass, you can extract a child by using a unique id
for that child as a key (e.g. you can extract an Atom object from a
Residue object by using an atom name string as a key, you can extract a
Chain object from a Model object by using its chain identifier as a
key).</p>
<p>Disordered atoms and residues are represented by DisorderedAtom and
DisorderedResidue classes, which are both subclasses of the
DisorderedEntityWrapper base class. They hide the complexity associated
with disorder and behave exactly as Atom and Residue objects.</p>
<p>In general, a child Entity object (i.e. Atom, Residue, Chain, Model) can
be extracted from its parent (i.e. Residue, Chain, Model, Structure,
respectively) by using an id as a key.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">child_entity</span> <span class="operator">=</span> <span class="name">parent_entity</span><span class="punctuation">[</span><span class="name">child_id</span><span class="punctuation">]</span>
</pre>
<p>You can also get a list of all child Entities of a parent Entity object.
Note that this list is sorted in a specific way (e.g. according to chain
identifier for Chain objects in a Model object).</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">child_list</span> <span class="operator">=</span> <span class="name">parent_entity</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">()</span>
</pre>
<p>You can also get the parent from a child:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">parent_entity</span> <span class="operator">=</span> <span class="name">child_entity</span><span class="operator">.</span><span class="name">get_parent</span><span class="punctuation">()</span>
</pre>
<p>At all levels of the SMCRA hierarchy, you can also extract a <em>full id</em>.
The full id is a tuple containing all id’s starting from the top object
(Structure) down to the current object. A full id for a Residue object
e.g. is something like:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">full_id</span> <span class="operator">=</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_full_id</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">full_id</span>
<span class="punctuation">(</span><span class="literal string">&quot;1abc&quot;</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal string">&quot;A&quot;</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">&quot;&quot;</span><span class="punctuation">,</span> <span class="literal number integer">10</span><span class="punctuation">,</span> <span class="literal string">&quot;A&quot;</span><span class="punctuation">))</span>
</pre>
<p>This corresponds to:</p>
<ul class="simple">
<li>The Structure with id &quot;1abc&quot;</li>
<li>The Model with id 0</li>
<li>The Chain with id &quot;A&quot;</li>
<li>The Residue with id (&quot; &quot;, 10, &quot;A&quot;).</li>
</ul>
<p>The Residue id indicates that the residue is not a hetero-residue (nor a
water) because it has a blank hetero field, that its sequence identifier
is 10 and that its insertion code is &quot;A&quot;.</p>
<p>To get the entity’s id, use the <tt class="docutils literal">get_id</tt> method:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">entity</span><span class="operator">.</span><span class="name">get_id</span><span class="punctuation">()</span>
</pre>
<p>You can check if the entity has a child with a given id by using the
<tt class="docutils literal">has_id</tt> method:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">entity</span><span class="operator">.</span><span class="name">has_id</span><span class="punctuation">(</span><span class="name">entity_id</span><span class="punctuation">)</span>
</pre>
<p>The length of an entity is equal to its number of children:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nr_children</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">entity</span><span class="punctuation">)</span>
</pre>
<p>It is possible to delete, rename, add, etc. child entities from a parent
entity, but this does not include any sanity checks (e.g. it is possible
to add two residues with the same id to one chain). This really should
be done via a nice Decorator class that includes integrity checking, but
you can take a look at the code (Entity.py) if you want to use the raw
interface.</p>
<div class="section" id="structure">
<h3>11.2.1  Structure</h3>
<p>The Structure object is at the top of the hierarchy. Its id is a user
given string. The Structure contains a number of Model children. Most
crystal structures (but not all) contain a single model, while NMR
structures typically consist of several models. Disorder in crystal
structures of large parts of molecules can also result in several
models.</p>
</div>
<div class="section" id="model">
<h3>11.2.2  Model</h3>
<p>The id of the Model object is an integer, which is derived from the
position of the model in the parsed file (they are automatically
numbered starting from 0). Crystal structures generally have only one
model (with id 0), while NMR files usually have several models. Whereas
many PDB parsers assume that there is only one model, the <tt class="docutils literal">Structure</tt>
class in <tt class="docutils literal">Bio.PDB</tt> is designed such that it can easily handle PDB
files with more than one model.</p>
<p>As an example, to get the first model from a Structure object, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">first_model</span> <span class="operator">=</span> <span class="name">structure</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
</pre>
<p>The Model object stores a list of Chain children.</p>
</div>
<div class="section" id="chain">
<h3>11.2.3  Chain</h3>
<p>The id of a Chain object is derived from the chain identifier in the
PDB/mmCIF file, and is a single character (typically a letter). Each
Chain in a Model object has a unique id. As an example, to get the Chain
object with identifier “A” from a Model object, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">chain_A</span> <span class="operator">=</span> <span class="name">model</span><span class="punctuation">[</span><span class="literal string">&quot;A&quot;</span><span class="punctuation">]</span>
</pre>
<p>The Chain object stores a list of Residue children.</p>
</div>
<div class="section" id="residue">
<h3>11.2.4  Residue</h3>
<p>A residue id is a tuple with three elements:</p>
<ul>
<li><p class="first">The <strong>hetero-field</strong> (hetfield): this is</p>
<ul class="simple">
<li><tt class="docutils literal">'W'</tt> in the case of a water molecule;</li>
<li><tt class="docutils literal">'H_'</tt> followed by the residue name for other hetero residues
(e.g. <tt class="docutils literal">'H_GLC'</tt> in the case of a glucose molecule);</li>
<li>blank for standard amino and nucleic acids.</li>
</ul>
<p>This scheme is adopted for reasons described in section
<a class="reference external" href="#hetero%20problems">11.4.1</a>.</p>
</li>
<li><p class="first">The <strong>sequence identifier</strong> (resseq), an integer describing the
position of the residue in the chain (e.g., 100);</p>
</li>
<li><p class="first">The <strong>insertion code</strong> (icode); a string, e.g. ’A’. The insertion
code is sometimes used to preserve a certain desirable residue
numbering scheme. A Ser 80 insertion mutant (inserted e.g. between a
Thr 80 and an Asn 81 residue) could e.g. have sequence identifiers
and insertion codes as follows: Thr 80 A, Ser 80 B, Asn 81. In this
way the residue numbering scheme stays in tune with that of the wild
type structure.</p>
</li>
</ul>
<p>The id of the above glucose residue would thus be
<tt class="docutils literal">(’H_GLC’, 100, ’A’)</tt>. If the hetero-flag and insertion code are
blank, the sequence identifier alone can be used:</p>
<pre class="code python literal-block">
<span class="comment"># Full id</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">=</span><span class="name">chain</span><span class="punctuation">[(</span><span class="literal string">' '</span><span class="punctuation">,</span> <span class="literal number integer">100</span><span class="punctuation">,</span> <span class="literal string">' '</span><span class="punctuation">)]</span>
<span class="comment"># Shortcut id</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">=</span><span class="name">chain</span><span class="punctuation">[</span><span class="literal number integer">100</span><span class="punctuation">]</span>
</pre>
<p>The reason for the hetero-flag is that many, many PDB files use the same
sequence identifier for an amino acid and a hetero-residue or a water,
which would create obvious problems if the hetero-flag was not used.</p>
<p>Unsurprisingly, a Residue object stores a set of Atom children. It also
contains a string that specifies the residue name (e.g. “ASN”) and the
segment identifier of the residue (well known to X-PLOR users, but not
used in the construction of the SMCRA data structure).</p>
<p>Let’s look at some examples. Asn 10 with a blank insertion code would
have residue id <tt class="docutils literal">(’ ’, 10, ’ ’)</tt>. Water 10 would have residue id
<tt class="docutils literal">(’W’, 10, ’ ’)</tt>. A glucose molecule (a hetero residue with residue
name GLC) with sequence identifier 10 would have residue id
<tt class="docutils literal">(’H_GLC’, 10, ’ ’)</tt>. In this way, the three residues (with the same
insertion code and sequence identifier) can be part of the same chain
because their residue id’s are distinct.</p>
<p>In most cases, the hetflag and insertion code fields will be blank, e.g.
<tt class="docutils literal">(’ ’, 10, ’ ’)</tt>. In these cases, the sequence identifier can be used
as a shortcut for the full id:</p>
<pre class="code python literal-block">
<span class="comment"># use full id</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">res10</span> <span class="operator">=</span> <span class="name">chain</span><span class="punctuation">[(</span><span class="literal string">' '</span><span class="punctuation">,</span> <span class="literal number integer">10</span><span class="punctuation">,</span> <span class="literal string">' '</span><span class="punctuation">)]</span>
<span class="comment"># use shortcut</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">res10</span> <span class="operator">=</span> <span class="name">chain</span><span class="punctuation">[</span><span class="literal number integer">10</span><span class="punctuation">]</span>
</pre>
<p>Each Residue object in a Chain object should have a unique id. However,
disordered residues are dealt with in a special way, as described in
section <a class="reference external" href="#point%20mutations">11.3.3</a>.</p>
<p>A Residue object has a number of additional methods:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_resname</span><span class="punctuation">()</span>       <span class="comment"># returns the residue name, e.g. &quot;ASN&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">.</span><span class="name">is_disordered</span><span class="punctuation">()</span>     <span class="comment"># returns 1 if the residue has disordered atoms</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_segid</span><span class="punctuation">()</span>         <span class="comment"># returns the SEGID, e.g. &quot;CHN1&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">.</span><span class="name">has_id</span><span class="punctuation">(</span><span class="name">name</span><span class="punctuation">)</span>        <span class="comment"># test if a residue has a certain atom</span>
</pre>
<p>You can use <tt class="docutils literal">is_aa(residue)</tt> to test if a Residue object is an amino
acid.</p>
</div>
<div class="section" id="atom">
<h3>11.2.5  Atom</h3>
<p>The Atom object stores the data associated with an atom, and has no
children. The id of an atom is its atom name (e.g. “OG” for the side
chain oxygen of a Ser residue). An Atom id needs to be unique in a
Residue. Again, an exception is made for disordered atoms, as described
in section <a class="reference external" href="#disordered%20atoms">11.3.2</a>.</p>
<p>The atom id is simply the atom name (eg. <tt class="docutils literal">’CA’</tt>). In practice, the
atom name is created by stripping all spaces from the atom name in the
PDB file.</p>
<p>However, in PDB files, a space can be part of an atom name. Often,
calcium atoms are called <tt class="docutils literal"><span class="pre">’CA..’</span></tt> in order to distinguish them from Cα
atoms (which are called <tt class="docutils literal">’.CA.’</tt>). In cases were stripping the spaces
would create problems (ie. two atoms called <tt class="docutils literal">’CA’</tt> in the same
residue) the spaces are kept.</p>
<p>In a PDB file, an atom name consists of 4 chars, typically with leading
and trailing spaces. Often these spaces can be removed for ease of use
(e.g. an amino acid C α atom is labeled “.CA.” in a PDB file, where the
dots represent spaces). To generate an atom name (and thus an atom id)
the spaces are removed, unless this would result in a name collision in
a Residue (i.e. two Atom objects with the same atom name and id). In the
latter case, the atom name including spaces is tried. This situation can
e.g. happen when one residue contains atoms with names “.CA.” and
“CA..”, although this is not very likely.</p>
<p>The atomic data stored includes the atom name, the atomic coordinates
(including standard deviation if present), the B factor (including
anisotropic B factors and standard deviation if present), the altloc
specifier and the full atom name including spaces. Less used items like
the atom element number or the atomic charge sometimes specified in a
PDB file are not stored.</p>
<p>To manipulate the atomic coordinates, use the <tt class="docutils literal">transform</tt> method of
the <tt class="docutils literal">Atom</tt> object. Use the <tt class="docutils literal">set_coord</tt> method to specify the atomic
coordinates directly.</p>
<p>An Atom object has the following additional methods:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_name</span><span class="punctuation">()</span>       <span class="comment"># atom name (spaces stripped, e.g. &quot;CA&quot;)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_id</span><span class="punctuation">()</span>         <span class="comment"># id (equals atom name)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_coord</span><span class="punctuation">()</span>      <span class="comment"># atomic coordinates</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>     <span class="comment"># atomic coordinates as Vector object</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_bfactor</span><span class="punctuation">()</span>    <span class="comment"># isotropic B factor</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_occupancy</span><span class="punctuation">()</span>  <span class="comment"># occupancy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_altloc</span><span class="punctuation">()</span>     <span class="comment"># alternative location specifier</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_sigatm</span><span class="punctuation">()</span>     <span class="comment"># standard deviation of atomic parameters</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_siguij</span><span class="punctuation">()</span>     <span class="comment"># standard deviation of anisotropic B factor</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_anisou</span><span class="punctuation">()</span>     <span class="comment"># anisotropic B factor</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_fullname</span><span class="punctuation">()</span>   <span class="comment"># atom name (with spaces, e.g. &quot;.CA.&quot;)</span>
</pre>
<p>To represent the atom coordinates, siguij, anisotropic B factor and
sigatm Numpy arrays are used.</p>
<p>The <tt class="docutils literal">get_vector</tt> method returns a <tt class="docutils literal">Vector</tt> object representation of
the coordinates of the <tt class="docutils literal">Atom</tt> object, allowing you to do vector
operations on atomic coordinates. <tt class="docutils literal">Vector</tt> implements the full set of
3D vector operations, matrix multiplication (left and right) and some
advanced rotation-related operations as well.</p>
<p>As an example of the capabilities of Bio.PDB’s <tt class="docutils literal">Vector</tt> module,
suppose that you would like to find the position of a Gly residue’s Cβ
atom, if it had one. Rotating the N atom of the Gly residue along the
Cα-C bond over -120 degrees roughly puts it in the position of a virtual
Cβ atom. Here’s how to do it, making use of the <tt class="docutils literal">rotaxis</tt> method
(which can be used to construct a rotation around a certain axis) of the
<tt class="docutils literal">Vector</tt> module:</p>
<pre class="code python literal-block">
<span class="comment"># get atom coordinates as vectors</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">n</span> <span class="operator">=</span> <span class="name">residue</span><span class="punctuation">[</span><span class="literal string">'N'</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">c</span> <span class="operator">=</span> <span class="name">residue</span><span class="punctuation">[</span><span class="literal string">'C'</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ca</span> <span class="operator">=</span> <span class="name">residue</span><span class="punctuation">[</span><span class="literal string">'CA'</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="comment"># center at origin</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">n</span> <span class="operator">=</span> <span class="name">n</span> <span class="operator">-</span> <span class="name">ca</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">c</span> <span class="operator">=</span> <span class="name">c</span> <span class="operator">-</span> <span class="name">ca</span>
<span class="comment"># find rotation matrix that rotates n</span>
<span class="comment"># -120 degrees along the ca-c vector</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">rot</span> <span class="operator">=</span> <span class="name">rotaxis</span><span class="punctuation">(</span><span class="operator">-</span><span class="name">pi</span> <span class="operator">*</span> <span class="literal number float">120.0</span><span class="operator">/</span><span class="literal number float">180.0</span><span class="punctuation">,</span> <span class="name">c</span><span class="punctuation">)</span>
<span class="comment"># apply rotation to ca-n vector</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cb_at_origin</span> <span class="operator">=</span> <span class="name">n</span><span class="operator">.</span><span class="name">left_multiply</span><span class="punctuation">(</span><span class="name">rot</span><span class="punctuation">)</span>
<span class="comment"># put on top of ca atom</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cb</span> <span class="operator">=</span> <span class="name">cb_at_origin</span><span class="operator">+</span><span class="name">ca</span>
</pre>
<p>This example shows that it’s possible to do some quite nontrivial vector
operations on atomic data, which can be quite useful. In addition to all
the usual vector operations (cross (use <tt class="docutils literal">*</tt><tt class="docutils literal">*</tt>), and dot (use
<tt class="docutils literal">*</tt>) product, angle, norm, etc.) and the above mentioned <tt class="docutils literal">rotaxis</tt>
function, the <tt class="docutils literal">Vector</tt> module also has methods to rotate (<tt class="docutils literal">rotmat</tt>)
or reflect (<tt class="docutils literal">refmat</tt>) one vector on top of another.</p>
</div>
<div class="section" id="extracting-a-specific-atom-residue-chain-model-from-a-structure">
<h3>11.2.6  Extracting a specific <tt class="docutils literal">Atom/Residue/Chain/Model</tt> from a Structure</h3>
<p>These are some examples:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">structure</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">chain</span> <span class="operator">=</span> <span class="name">model</span><span class="punctuation">[</span><span class="literal string">'A'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span> <span class="operator">=</span> <span class="name">chain</span><span class="punctuation">[</span><span class="literal number integer">100</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atom</span> <span class="operator">=</span> <span class="name">residue</span><span class="punctuation">[</span><span class="literal string">'CA'</span><span class="punctuation">]</span>
</pre>
<p>Note that you can use a shortcut:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atom</span> <span class="operator">=</span> <span class="name">structure</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">'A'</span><span class="punctuation">][</span><span class="literal number integer">100</span><span class="punctuation">][</span><span class="literal string">'CA'</span><span class="punctuation">]</span>
</pre>
</div>
</div>
<div class="section" id="disorder">
<h2>11.3  Disorder</h2>
<p>Bio.PDB can handle both disordered atoms and point mutations (i.e. a Gly
and an Ala residue in the same position).</p>
<div class="section" id="general-approach">
<h3>11.3.1  General approach</h3>
<p>Disorder should be dealt with from two points of view: the atom and the
residue points of view. In general, we have tried to encapsulate all the
complexity that arises from disorder. If you just want to loop over all
Cα atoms, you do not care that some residues have a disordered side
chain. On the other hand it should also be possible to represent
disorder completely in the data structure. Therefore, disordered atoms
or residues are stored in special objects that behave as if there is no
disorder. This is done by only representing a subset of the disordered
atoms or residues. Which subset is picked (e.g. which of the two
disordered OG side chain atom positions of a Ser residue is used) can be
specified by the user.</p>
</div>
<div class="section" id="disordered-atoms">
<h3>11.3.2  Disordered atoms</h3>
<p>Disordered atoms are represented by ordinary <tt class="docutils literal">Atom</tt> objects, but all
<tt class="docutils literal">Atom</tt> objects that represent the same physical atom are stored in a
<tt class="docutils literal">DisorderedAtom</tt> object (see Fig. <a class="reference external" href="#fig:smcra">11.1</a>). Each
<tt class="docutils literal">Atom</tt> object in a <tt class="docutils literal">DisorderedAtom</tt> object can be uniquely indexed
using its altloc specifier. The <tt class="docutils literal">DisorderedAtom</tt> object forwards all
uncaught method calls to the selected Atom object, by default the one
that represents the atom with the highest occupancy. The user can of
course change the selected <tt class="docutils literal">Atom</tt> object, making use of its altloc
specifier. In this way atom disorder is represented correctly without
much additional complexity. In other words, if you are not interested in
atom disorder, you will not be bothered by it.</p>
<p>Each disordered atom has a characteristic altloc identifier. You can
specify that a <tt class="docutils literal">DisorderedAtom</tt> object should behave like the <tt class="docutils literal">Atom</tt>
object associated with a specific altloc identifier:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atom</span><span class="operator">.</span><span class="name">disordered_select</span><span class="punctuation">(</span><span class="literal string">'A'</span><span class="punctuation">)</span> <span class="comment"># select altloc A atom</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">atom</span><span class="operator">.</span><span class="name">get_altloc</span><span class="punctuation">()</span>
<span class="literal string">&quot;A&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atom</span><span class="operator">.</span><span class="name">disordered_select</span><span class="punctuation">(</span><span class="literal string">'B'</span><span class="punctuation">)</span> <span class="comment"># select altloc B atom</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">atom</span><span class="operator">.</span><span class="name">get_altloc</span><span class="punctuation">()</span>
<span class="literal string">&quot;B&quot;</span>
</pre>
</div>
<div class="section" id="disordered-residues">
<h3>11.3.3  Disordered residues</h3>
<div class="section" id="common-case">
<h4>Common case</h4>
<p>The most common case is a residue that contains one or more disordered
atoms. This is evidently solved by using DisorderedAtom objects to
represent the disordered atoms, and storing the DisorderedAtom object in
a Residue object just like ordinary Atom objects. The DisorderedAtom
will behave exactly like an ordinary atom (in fact the atom with the
highest occupancy) by forwarding all uncaught method calls to one of the
Atom objects (the selected Atom object) it contains.</p>
</div>
<div class="section" id="point-mutations">
<h4>Point mutations</h4>
<p>A special case arises when disorder is due to a point mutation, i.e.
when two or more point mutants of a polypeptide are present in the
crystal. An example of this can be found in PDB structure 1EN2.</p>
<p>Since these residues belong to a different residue type (e.g. let’s say
Ser 60 and Cys 60) they should not be stored in a single <tt class="docutils literal">Residue</tt>
object as in the common case. In this case, each residue is represented
by one <tt class="docutils literal">Residue</tt> object, and both <tt class="docutils literal">Residue</tt> objects are stored in a
single <tt class="docutils literal">DisorderedResidue</tt> object (see Fig. <a class="reference external" href="#fig:smcra">11.1</a>).</p>
<p>The <tt class="docutils literal">DisorderedResidue</tt> object forwards all uncaught methods to the
selected <tt class="docutils literal">Residue</tt> object (by default the last <tt class="docutils literal">Residue</tt> object
added), and thus behaves like an ordinary residue. Each <tt class="docutils literal">Residue</tt>
object in a <tt class="docutils literal">DisorderedResidue</tt> object can be uniquely identified by
its residue name. In the above example, residue Ser 60 would have id
“SER” in the <tt class="docutils literal">DisorderedResidue</tt> object, while residue Cys 60 would
have id “CYS”. The user can select the active <tt class="docutils literal">Residue</tt> object in a
<tt class="docutils literal">DisorderedResidue</tt> object via this id.</p>
<p>Example: suppose that a chain has a point mutation at position 10,
consisting of a Ser and a Cys residue. Make sure that residue 10 of this
chain behaves as the Cys residue.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span> <span class="operator">=</span> <span class="name">chain</span><span class="punctuation">[</span><span class="literal number integer">10</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">.</span><span class="name">disordered_select</span><span class="punctuation">(</span><span class="literal string">'CYS'</span><span class="punctuation">)</span>
</pre>
<p>In addition, you can get a list of all <tt class="docutils literal">Atom</tt> objects (ie. all
<tt class="docutils literal">DisorderedAtom</tt> objects are ’unpacked’ to their individual <tt class="docutils literal">Atom</tt>
objects) using the <tt class="docutils literal">get_unpacked_list</tt> method of a
<tt class="docutils literal">(Disordered)Residue</tt> object.</p>
</div>
</div>
</div>
<div class="section" id="hetero-residues">
<h2>11.4  Hetero residues</h2>
<div class="section" id="associated-problems">
<h3>11.4.1  Associated problems</h3>
<p>A common problem with hetero residues is that several hetero and
non-hetero residues present in the same chain share the same sequence
identifier (and insertion code). Therefore, to generate a unique id for
each hetero residue, waters and other hetero residues are treated in a
different way.</p>
<p>Remember that Residue object have the tuple (hetfield, resseq, icode) as
id. The hetfield is blank (“ ”) for amino and nucleic acids, and a
string for waters and other hetero residues. The content of the hetfield
is explained below.</p>
</div>
<div class="section" id="water-residues">
<h3>11.4.2  Water residues</h3>
<p>The hetfield string of a water residue consists of the letter “W”. So a
typical residue id for a water is (“W”, 1, “ ”).</p>
</div>
<div class="section" id="other-hetero-residues">
<h3>11.4.3  Other hetero residues</h3>
<p>The hetfield string for other hetero residues starts with “H_” followed
by the residue name. A glucose molecule e.g. with residue name “GLC”
would have hetfield “H_GLC”. Its residue id could e.g. be (“H_GLC”, 1,
“ ”).</p>
</div>
</div>
<div class="section" id="navigating-through-a-structure-object">
<h2>11.5  Navigating through a Structure object</h2>
<div class="section" id="parse-a-pdb-file-and-extract-some-model-chain-residue-and-atom-objects">
<h3>Parse a PDB file, and extract some Model, Chain, Residue and Atom objects</h3>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.PDB.PDBParser</span> <span class="keyword namespace">import</span> <span class="name">PDBParser</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">parser</span> <span class="operator">=</span> <span class="name">PDBParser</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">structure</span> <span class="operator">=</span> <span class="name">parser</span><span class="operator">.</span><span class="name">get_structure</span><span class="punctuation">(</span><span class="literal string">&quot;test&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;1fat.pdb&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">structure</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">chain</span> <span class="operator">=</span> <span class="name">model</span><span class="punctuation">[</span><span class="literal string">&quot;A&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span> <span class="operator">=</span> <span class="name">chain</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atom</span> <span class="operator">=</span> <span class="name">residue</span><span class="punctuation">[</span><span class="literal string">&quot;CA&quot;</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="iterating-through-all-atoms-of-a-structure">
<h3>Iterating through all atoms of a structure</h3>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">p</span> <span class="operator">=</span> <span class="name">PDBParser</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">structure</span> <span class="operator">=</span> <span class="name">p</span><span class="operator">.</span><span class="name">get_structure</span><span class="punctuation">(</span><span class="literal string">'X'</span><span class="punctuation">,</span> <span class="literal string">'pdb1fat.ent'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">model</span> <span class="operator word">in</span> <span class="name">structure</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">chain</span> <span class="operator word">in</span> <span class="name">model</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">for</span> <span class="name">residue</span> <span class="operator word">in</span> <span class="name">chain</span><span class="punctuation">:</span>
<span class="operator">...</span>             <span class="keyword">for</span> <span class="name">atom</span> <span class="operator word">in</span> <span class="name">residue</span><span class="punctuation">:</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="name">atom</span>
<span class="operator">...</span>
</pre>
<p>There is a shortcut if you want to iterate over all atoms in a
structure:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atoms</span> <span class="operator">=</span> <span class="name">structure</span><span class="operator">.</span><span class="name">get_atoms</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">atom</span> <span class="operator word">in</span> <span class="name">atoms</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">atom</span>
<span class="operator">...</span>
</pre>
<p>Similarly, to iterate over all atoms in a chain, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atoms</span> <span class="operator">=</span> <span class="name">chain</span><span class="operator">.</span><span class="name">get_atoms</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">atom</span> <span class="operator word">in</span> <span class="name">atoms</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">atom</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="iterating-over-all-residues-of-a-model">
<h3>Iterating over all residues of a model</h3>
<p>or if you want to iterate over all residues in a model:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residues</span> <span class="operator">=</span> <span class="name">model</span><span class="operator">.</span><span class="name">get_residues</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">residue</span> <span class="operator word">in</span> <span class="name">residues</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">residue</span>
<span class="operator">...</span>
</pre>
<p>You can also use the <tt class="docutils literal">Selection.unfold_entities</tt> function to get all
residues from a structure:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">res_list</span> <span class="operator">=</span> <span class="name">Selection</span><span class="operator">.</span><span class="name">unfold_entities</span><span class="punctuation">(</span><span class="name">structure</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">)</span>
</pre>
<p>or to get all atoms from a chain:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atom_list</span> <span class="operator">=</span> <span class="name">Selection</span><span class="operator">.</span><span class="name">unfold_entities</span><span class="punctuation">(</span><span class="name">chain</span><span class="punctuation">,</span> <span class="literal string">'A'</span><span class="punctuation">)</span>
</pre>
<p>Obviously, <tt class="docutils literal">A=atom, R=residue, C=chain, M=model, S=structure</tt>. You can
use this to go up in the hierarchy, e.g. to get a list of (unique)
<tt class="docutils literal">Residue</tt> or <tt class="docutils literal">Chain</tt> parents from a list of <tt class="docutils literal">Atoms</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue_list</span> <span class="operator">=</span> <span class="name">Selection</span><span class="operator">.</span><span class="name">unfold_entities</span><span class="punctuation">(</span><span class="name">atom_list</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">chain_list</span> <span class="operator">=</span> <span class="name">Selection</span><span class="operator">.</span><span class="name">unfold_entities</span><span class="punctuation">(</span><span class="name">atom_list</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">)</span>
</pre>
<p>For more info, see the API documentation.</p>
</div>
<div class="section" id="extract-a-hetero-residue-from-a-chain-e-g-a-glucose-glc-moiety-with-resseq-10">
<h3>Extract a hetero residue from a chain (e.g. a glucose (GLC) moiety with resseq 10)</h3>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue_id</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="literal string">&quot;H_GLC&quot;</span><span class="punctuation">,</span> <span class="literal number integer">10</span><span class="punctuation">,</span> <span class="literal string">&quot; &quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span> <span class="operator">=</span> <span class="name">chain</span><span class="punctuation">[</span><span class="name">residue_id</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="print-all-hetero-residues-in-chain">
<h3>Print all hetero residues in chain</h3>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">residue</span> <span class="operator word">in</span> <span class="name">chain</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>    <span class="name">residue_id</span> <span class="operator">=</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_id</span><span class="punctuation">()</span>
<span class="operator">...</span>    <span class="name">hetfield</span> <span class="operator">=</span> <span class="name">residue_id</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">...</span>    <span class="keyword">if</span> <span class="name">hetfield</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">==</span><span class="literal string">&quot;H&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>        <span class="keyword">print</span> <span class="name">residue_id</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="print-out-the-coordinates-of-all-ca-atoms-in-a-structure-with-b-factor-greater-than-50">
<h3>Print out the coordinates of all CA atoms in a structure with B factor greater than 50</h3>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">model</span> <span class="operator word">in</span> <span class="name">structure</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">chain</span> <span class="operator word">in</span> <span class="name">model</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>         <span class="keyword">for</span> <span class="name">residue</span> <span class="operator word">in</span> <span class="name">chain</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>             <span class="keyword">if</span> <span class="name">residue</span><span class="operator">.</span><span class="name">has_id</span><span class="punctuation">(</span><span class="literal string">&quot;CA&quot;</span><span class="punctuation">):</span>
<span class="operator">...</span>                 <span class="name">ca</span> <span class="operator">=</span> <span class="name">residue</span><span class="punctuation">[</span><span class="literal string">&quot;CA&quot;</span><span class="punctuation">]</span>
<span class="operator">...</span>                 <span class="keyword">if</span> <span class="name">ca</span><span class="operator">.</span><span class="name">get_bfactor</span><span class="punctuation">()</span> <span class="operator">&gt;</span> <span class="literal number float">50.0</span><span class="punctuation">:</span>
<span class="operator">...</span>                     <span class="keyword">print</span> <span class="name">ca</span><span class="operator">.</span><span class="name">get_coord</span><span class="punctuation">()</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="print-out-all-the-residues-that-contain-disordered-atoms">
<h3>Print out all the residues that contain disordered atoms</h3>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">model</span> <span class="operator word">in</span> <span class="name">structure</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">chain</span> <span class="operator word">in</span> <span class="name">model</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>         <span class="keyword">for</span> <span class="name">residue</span> <span class="operator word">in</span> <span class="name">chain</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>             <span class="keyword">if</span> <span class="name">residue</span><span class="operator">.</span><span class="name">is_disordered</span><span class="punctuation">():</span>
<span class="operator">...</span>                 <span class="name">resseq</span> <span class="operator">=</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_id</span><span class="punctuation">()[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="operator">...</span>                 <span class="name">resname</span> <span class="operator">=</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_resname</span><span class="punctuation">()</span>
<span class="operator">...</span>                 <span class="name">model_id</span> <span class="operator">=</span> <span class="name">model</span><span class="operator">.</span><span class="name">get_id</span><span class="punctuation">()</span>
<span class="operator">...</span>                 <span class="name">chain_id</span> <span class="operator">=</span> <span class="name">chain</span><span class="operator">.</span><span class="name">get_id</span><span class="punctuation">()</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="name">model_id</span><span class="punctuation">,</span> <span class="name">chain_id</span><span class="punctuation">,</span> <span class="name">resname</span><span class="punctuation">,</span> <span class="name">resseq</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="loop-over-all-disordered-atoms-and-select-all-atoms-with-altloc-a-if-present">
<h3>Loop over all disordered atoms, and select all atoms with altloc A (if present)</h3>
<p>This will make sure that the SMCRA data structure will behave as if only
the atoms with altloc A are present.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">model</span> <span class="operator word">in</span> <span class="name">structure</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">chain</span> <span class="operator word">in</span> <span class="name">model</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>         <span class="keyword">for</span> <span class="name">residue</span> <span class="operator word">in</span> <span class="name">chain</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>             <span class="keyword">if</span> <span class="name">residue</span><span class="operator">.</span><span class="name">is_disordered</span><span class="punctuation">():</span>
<span class="operator">...</span>                 <span class="keyword">for</span> <span class="name">atom</span> <span class="operator word">in</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>                     <span class="keyword">if</span> <span class="name">atom</span><span class="operator">.</span><span class="name">is_disordered</span><span class="punctuation">():</span>
<span class="operator">...</span>                         <span class="keyword">if</span> <span class="name">atom</span><span class="operator">.</span><span class="name">disordered_has_id</span><span class="punctuation">(</span><span class="literal string">&quot;A&quot;</span><span class="punctuation">):</span>
<span class="operator">...</span>                             <span class="name">atom</span><span class="operator">.</span><span class="name">disordered_select</span><span class="punctuation">(</span><span class="literal string">&quot;A&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="extracting-polypeptides-from-a-structure-object">
<h3>Extracting polypeptides from a <tt class="docutils literal">Structure</tt> object</h3>
<p>To extract polypeptides from a structure, construct a list of
<tt class="docutils literal">Polypeptide</tt> objects from a <tt class="docutils literal">Structure</tt> object using
<tt class="docutils literal">PolypeptideBuilder</tt> as follows:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model_nr</span> <span class="operator">=</span> <span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">polypeptide_list</span> <span class="operator">=</span> <span class="name">build_peptides</span><span class="punctuation">(</span><span class="name">structure</span><span class="punctuation">,</span> <span class="name">model_nr</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">polypeptide</span> <span class="operator word">in</span> <span class="name">polypeptide_list</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">polypeptide</span>
<span class="operator">...</span>
</pre>
<p>A Polypeptide object is simply a UserList of Residue objects, and is
always created from a single Model (in this case model 1). You can use
the resulting <tt class="docutils literal">Polypeptide</tt> object to get the sequence as a <tt class="docutils literal">Seq</tt>
object or to get a list of Cα atoms as well. Polypeptides can be built
using a C-N or a Cα-Cα distance criterion.</p>
<p>Example:</p>
<pre class="code python literal-block">
<span class="comment"># Using C-N</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ppb</span><span class="operator">=</span><span class="name">PPBuilder</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pp</span> <span class="operator word">in</span> <span class="name">ppb</span><span class="operator">.</span><span class="name">build_peptides</span><span class="punctuation">(</span><span class="name">structure</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pp</span><span class="operator">.</span><span class="name">get_sequence</span><span class="punctuation">()</span>
<span class="operator">...</span>
<span class="comment"># Using CA-CA</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ppb</span><span class="operator">=</span><span class="name">CaPPBuilder</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pp</span> <span class="operator word">in</span> <span class="name">ppb</span><span class="operator">.</span><span class="name">build_peptides</span><span class="punctuation">(</span><span class="name">structure</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pp</span><span class="operator">.</span><span class="name">get_sequence</span><span class="punctuation">()</span>
<span class="operator">...</span>
</pre>
<p>Note that in the above case only model 0 of the structure is considered
by <tt class="docutils literal">PolypeptideBuilder</tt>. However, it is possible to use
<tt class="docutils literal">PolypeptideBuilder</tt> to build <tt class="docutils literal">Polypeptide</tt> objects from <tt class="docutils literal">Model</tt>
and <tt class="docutils literal">Chain</tt> objects as well.</p>
</div>
<div class="section" id="obtaining-the-sequence-of-a-structure">
<h3>Obtaining the sequence of a structure</h3>
<p>The first thing to do is to extract all polypeptides from the structure
(as above). The sequence of each polypeptide can then easily be obtained
from the <tt class="docutils literal">Polypeptide</tt> objects. The sequence is represented as a
Biopython <tt class="docutils literal">Seq</tt> object, and its alphabet is defined by a
<tt class="docutils literal">ProteinAlphabet</tt> object.</p>
<p>Example:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq</span> <span class="operator">=</span> <span class="name">polypeptide</span><span class="operator">.</span><span class="name">get_sequence</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'SNVVE...'</span><span class="punctuation">,</span> <span class="operator">&lt;</span><span class="keyword">class</span> <span class="name class">Bio</span><span class="operator">.</span><span class="name">Alphabet</span><span class="operator">.</span><span class="name">ProteinAlphabet</span><span class="operator">&gt;</span><span class="punctuation">)</span>
</pre>
</div>
</div>
<div class="section" id="analyzing-structures">
<h2>11.6  Analyzing structures</h2>
<div class="section" id="measuring-distances">
<h3>11.6.1  Measuring distances</h3>
<p>The minus operator for atoms has been overloaded to return the distance
between two atoms.</p>
<pre class="code python literal-block">
<span class="comment"># Get some atoms</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ca1</span> <span class="operator">=</span> <span class="name">residue1</span><span class="punctuation">[</span><span class="literal string">'CA'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ca2</span> <span class="operator">=</span> <span class="name">residue2</span><span class="punctuation">[</span><span class="literal string">'CA'</span><span class="punctuation">]</span>
<span class="comment"># Simply subtract the atoms to get their distance</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">distance</span> <span class="operator">=</span> <span class="name">ca1</span><span class="operator">-</span><span class="name">ca2</span>
</pre>
</div>
<div class="section" id="measuring-angles">
<h3>11.6.2  Measuring angles</h3>
<p>Use the vector representation of the atomic coordinates, and the
<tt class="docutils literal">calc_angle</tt> function from the <tt class="docutils literal">Vector</tt> module:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector1</span> <span class="operator">=</span> <span class="name">atom1</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector2</span> <span class="operator">=</span> <span class="name">atom2</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector3</span> <span class="operator">=</span> <span class="name">atom3</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">angle</span> <span class="operator">=</span> <span class="name">calc_angle</span><span class="punctuation">(</span><span class="name">vector1</span><span class="punctuation">,</span> <span class="name">vector2</span><span class="punctuation">,</span> <span class="name">vector3</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="measuring-torsion-angles">
<h3>11.6.3  Measuring torsion angles</h3>
<p>Use the vector representation of the atomic coordinates, and the
<tt class="docutils literal">calc_dihedral</tt> function from the <tt class="docutils literal">Vector</tt> module:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector1</span> <span class="operator">=</span> <span class="name">atom1</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector2</span> <span class="operator">=</span> <span class="name">atom2</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector3</span> <span class="operator">=</span> <span class="name">atom3</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector4</span> <span class="operator">=</span> <span class="name">atom4</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">angle</span> <span class="operator">=</span> <span class="name">calc_dihedral</span><span class="punctuation">(</span><span class="name">vector1</span><span class="punctuation">,</span> <span class="name">vector2</span><span class="punctuation">,</span> <span class="name">vector3</span><span class="punctuation">,</span> <span class="name">vector4</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="determining-atom-atom-contacts">
<h3>11.6.4  Determining atom-atom contacts</h3>
<p>Use <tt class="docutils literal">NeighborSearch</tt> to perform neighbor lookup. The neighbor lookup
is done using a KD tree module written in C (see <tt class="docutils literal">Bio.KDTree</tt>), making
it very fast. It also includes a fast method to find all point pairs
within a certain distance of each other.</p>
</div>
<div class="section" id="superimposing-two-structures">
<h3>11.6.5  Superimposing two structures</h3>
<p>Use a <tt class="docutils literal">Superimposer</tt> object to superimpose two coordinate sets. This
object calculates the rotation and translation matrix that rotates two
lists of atoms on top of each other in such a way that their RMSD is
minimized. Of course, the two lists need to contain the same number of
atoms. The <tt class="docutils literal">Superimposer</tt> object can also apply the
rotation/translation to a list of atoms. The rotation and translation
are stored as a tuple in the <tt class="docutils literal">rotran</tt> attribute of the
<tt class="docutils literal">Superimposer</tt> object (note that the rotation is right multiplying!).
The RMSD is stored in the <tt class="docutils literal">rmsd</tt> attribute.</p>
<p>The algorithm used by <tt class="docutils literal">Superimposer</tt> comes from [<a class="reference external" href="#golub1989">17</a>,
Golub &amp; Van Loan] and makes use of singular value decomposition (this is
implemented in the general <tt class="docutils literal">Bio.SVDSuperimposer</tt> module).</p>
<p>Example:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sup</span> <span class="operator">=</span> <span class="name">Superimposer</span><span class="punctuation">()</span>
<span class="comment"># Specify the atom lists</span>
<span class="comment"># 'fixed' and 'moving' are lists of Atom objects</span>
<span class="comment"># The moving atoms will be put on the fixed atoms</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sup</span><span class="operator">.</span><span class="name">set_atoms</span><span class="punctuation">(</span><span class="name">fixed</span><span class="punctuation">,</span> <span class="name">moving</span><span class="punctuation">)</span>
<span class="comment"># Print rotation/translation/rmsd</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">sup</span><span class="operator">.</span><span class="name">rotran</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">sup</span><span class="operator">.</span><span class="name">rms</span>
<span class="comment"># Apply rotation/translation to the moving atoms</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sup</span><span class="operator">.</span><span class="name">apply</span><span class="punctuation">(</span><span class="name">moving</span><span class="punctuation">)</span>
</pre>
<p>To superimpose two structures based on their active sites, use the
active site atoms to calculate the rotation/translation matrices (as
above), and apply these to the whole molecule.</p>
</div>
<div class="section" id="mapping-the-residues-of-two-related-structures-onto-each-other">
<h3>11.6.6  Mapping the residues of two related structures onto each other</h3>
<p>First, create an alignment file in FASTA format, then use the
<tt class="docutils literal">StructureAlignment</tt> class. This class can also be used for alignments
with more than two structures.</p>
</div>
<div class="section" id="calculating-the-half-sphere-exposure">
<h3>11.6.7  Calculating the Half Sphere Exposure</h3>
<p>Half Sphere Exposure (HSE) is a new, 2D measure of solvent exposure
[<a class="reference external" href="#hamelryck2005">20</a>]. Basically, it counts the number of Cα atoms
around a residue in the direction of its side chain, and in the opposite
direction (within a radius of 13 Å). Despite its simplicity, it
outperforms many other measures of solvent exposure.</p>
<p>HSE comes in two flavors: HSEα and HSEβ. The former only uses the Cα
atom positions, while the latter uses the Cα and Cβ atom positions. The
HSE measure is calculated by the <tt class="docutils literal">HSExposure</tt> class, which can also
calculate the contact number. The latter class has methods which return
dictionaries that map a <tt class="docutils literal">Residue</tt> object to its corresponding HSEα,
HSEβ and contact number values.</p>
<p>Example:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">structure</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">hse</span> <span class="operator">=</span> <span class="name">HSExposure</span><span class="punctuation">()</span>
<span class="comment"># Calculate HSEalpha</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exp_ca</span> <span class="operator">=</span> <span class="name">hse</span><span class="operator">.</span><span class="name">calc_hs_exposure</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">option</span><span class="operator">=</span><span class="literal string">'CA3'</span><span class="punctuation">)</span>
<span class="comment"># Calculate HSEbeta</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exp_cb</span><span class="operator">=</span><span class="name">hse</span><span class="operator">.</span><span class="name">calc_hs_exposure</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">option</span><span class="operator">=</span><span class="literal string">'CB'</span><span class="punctuation">)</span>
<span class="comment"># Calculate classical coordination number</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exp_fs</span> <span class="operator">=</span> <span class="name">hse</span><span class="operator">.</span><span class="name">calc_fs_exposure</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">)</span>
<span class="comment"># Print HSEalpha for a residue</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">exp_ca</span><span class="punctuation">[</span><span class="name">some_residue</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="determining-the-secondary-structure">
<h3>11.6.8  Determining the secondary structure</h3>
<p>For this functionality, you need to install DSSP (and obtain a license
for it — free for academic use, see
<a class="reference external" href="http://www.cmbi.kun.nl/gv/dssp/">http://www.cmbi.kun.nl/gv/dssp/</a>).
Then use the <tt class="docutils literal">DSSP</tt> class, which maps <tt class="docutils literal">Residue</tt> objects to their
secondary structure (and accessible surface area). The DSSP codes are
listed in Table <a class="reference external" href="#cap:DSSP-codes">11.1</a>. Note that DSSP (the program,
and thus by consequence the class) cannot handle multiple models!</p>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="22%" />
<col width="78%" />
</colgroup>
<tbody valign="top">
<tr><td>Code</td>
<td>Secondary structure</td>
</tr>
<tr><td>H</td>
<td>α-helix</td>
</tr>
<tr><td>B</td>
<td>Isolated β-bridge residue</td>
</tr>
<tr><td>E</td>
<td>Strand</td>
</tr>
<tr><td>G</td>
<td>3-10 helix</td>
</tr>
<tr><td>I</td>
<td>Π-helix</td>
</tr>
<tr><td>T</td>
<td>Turn</td>
</tr>
<tr><td>S</td>
<td>Bend</td>
</tr>
<tr><td><ul class="first last simple">
<li></li>
</ul>
</td>
<td>Other</td>
</tr>
</tbody>
</table>
<p>Table 11.1: DSSP codes in Bio.PDB.</p>
<hr class="docutils" />
<p>The <tt class="docutils literal">DSSP</tt> class can also be used to calculate the accessible surface
area of a residue. But see also section
<a class="reference external" href="#subsec:residue_depth">11.6.9</a>.</p>
</div>
<div class="section" id="calculating-the-residue-depth">
<h3>11.6.9  Calculating the residue depth</h3>
<p>Residue depth is the average distance of a residue’s atoms from the
solvent accessible surface. It’s a fairly new and very powerful
parameterization of solvent accessibility. For this functionality, you
need to install Michel Sanner’s MSMS program
(<a class="reference external" href="http://www.scripps.edu/pub/olson-web/people/sanner/html/msms_home.html">http://www.scripps.edu/pub/olson-web/people/sanner/html/msms_home.html</a>).
Then use the <tt class="docutils literal">ResidueDepth</tt> class. This class behaves as a dictionary
which maps <tt class="docutils literal">Residue</tt> objects to corresponding (residue depth, Cα
depth) tuples. The Cα depth is the distance of a residue’s Cα atom to
the solvent accessible surface.</p>
<p>Example:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">structure</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">rd</span> <span class="operator">=</span> <span class="name">ResidueDepth</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">pdb_file</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue_depth</span><span class="punctuation">,</span> <span class="name">ca_depth</span><span class="operator">=</span><span class="name">rd</span><span class="punctuation">[</span><span class="name">some_residue</span><span class="punctuation">]</span>
</pre>
<p>You can also get access to the molecular surface itself (via the
<tt class="docutils literal">get_surface</tt> function), in the form of a Numeric Python array with
the surface points.</p>
</div>
</div>
<div class="section" id="common-problems-in-pdb-files">
<h2>11.7  Common problems in PDB files</h2>
<p>It is well known that many PDB files contain semantic errors (not the
structures themselves, but their representation in PDB files). Bio.PDB
tries to handle this in two ways. The PDBParser object can behave in two
ways: a restrictive way and a permissive way, which is the default.</p>
<p>Example:</p>
<pre class="code python literal-block">
<span class="comment"># Permissive parser</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">parser</span> <span class="operator">=</span> <span class="name">PDBParser</span><span class="punctuation">(</span><span class="name">PERMISSIVE</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">parser</span> <span class="operator">=</span> <span class="name">PDBParser</span><span class="punctuation">()</span> <span class="comment"># The same (default)</span>
<span class="comment"># Strict parser</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">strict_parser</span> <span class="operator">=</span> <span class="name">PDBParser</span><span class="punctuation">(</span><span class="name">PERMISSIVE</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">)</span>
</pre>
<p>In the permissive state (DEFAULT), PDB files that obviously contain
errors are “corrected” (i.e. some residues or atoms are left out). These
errors include:</p>
<ul class="simple">
<li>Multiple residues with the same identifier</li>
<li>Multiple atoms with the same identifier (taking into account the
altloc identifier)</li>
</ul>
<p>These errors indicate real problems in the PDB file (for details see
[<a class="reference external" href="#hamelryck2003a">18</a>, Hamelryck and Manderick, 2003]). In the
restrictive state, PDB files with errors cause an exception to occur.
This is useful to find errors in PDB files.</p>
<p>Some errors however are automatically corrected. Normally each
disordered atom should have a non-blank altloc identifier. However,
there are many structures that do not follow this convention, and have a
blank and a non-blank identifier for two disordered positions of the
same atom. This is automatically interpreted in the right way.</p>
<p>Sometimes a structure contains a list of residues belonging to chain A,
followed by residues belonging to chain B, and again followed by
residues belonging to chain A, i.e. the chains are ’broken’. This is
also correctly interpreted.</p>
<div class="section" id="id3">
<h3>11.7.1  Examples</h3>
<p>The PDBParser/Structure class was tested on about 800 structures (each
belonging to a unique SCOP superfamily). This takes about 20 minutes, or
on average 1.5 seconds per structure. Parsing the structure of the large
ribosomal subunit (1FKK), which contains about 64000 atoms, takes 10
seconds on a 1000 MHz PC.</p>
<p>Three exceptions were generated in cases where an unambiguous data
structure could not be built. In all three cases, the likely cause is an
error in the PDB file that should be corrected. Generating an exception
in these cases is much better than running the chance of incorrectly
describing the structure in a data structure.</p>
<div class="section" id="duplicate-residues">
<h4>11.7.1.1  Duplicate residues</h4>
<p>One structure contains two amino acid residues in one chain with the
same sequence identifier (resseq 3) and icode. Upon inspection it was
found that this chain contains the residues Thr A3, …, Gly A202, Leu A3,
Glu A204. Clearly, Leu A3 should be Leu A203. A couple of similar
situations exist for structure 1FFK (which e.g. contains Gly B64, Met
B65, Glu B65, Thr B67, i.e. residue Glu B65 should be Glu B66).</p>
</div>
<div class="section" id="duplicate-atoms">
<h4>11.7.1.2  Duplicate atoms</h4>
<p>Structure 1EJG contains a Ser/Pro point mutation in chain A at position
22. In turn, Ser 22 contains some disordered atoms. As expected, all
atoms belonging to Ser 22 have a non-blank altloc specifier (B or C).
All atoms of Pro 22 have altloc A, except the N atom which has a blank
altloc. This generates an exception, because all atoms belonging to two
residues at a point mutation should have non-blank altloc. It turns out
that this atom is probably shared by Ser and Pro 22, as Ser 22 misses
the N atom. Again, this points to a problem in the file: the N atom
should be present in both the Ser and the Pro residue, in both cases
associated with a suitable altloc identifier.</p>
</div>
</div>
<div class="section" id="automatic-correction">
<h3>11.7.2  Automatic correction</h3>
<p>Some errors are quite common and can be easily corrected without much
risk of making a wrong interpretation. These cases are listed below.</p>
<div class="section" id="a-blank-altloc-for-a-disordered-atom">
<h4>11.7.2.1  A blank altloc for a disordered atom</h4>
<p>Normally each disordered atom should have a non-blank altloc identifier.
However, there are many structures that do not follow this convention,
and have a blank and a non-blank identifier for two disordered positions
of the same atom. This is automatically interpreted in the right way.</p>
</div>
<div class="section" id="broken-chains">
<h4>11.7.2.2  Broken chains</h4>
<p>Sometimes a structure contains a list of residues belonging to chain A,
followed by residues belonging to chain B, and again followed by
residues belonging to chain A, i.e. the chains are “broken”. This is
correctly interpreted.</p>
</div>
</div>
<div class="section" id="fatal-errors">
<h3>11.7.3  Fatal errors</h3>
<p>Sometimes a PDB file cannot be unambiguously interpreted. Rather than
guessing and risking a mistake, an exception is generated, and the user
is expected to correct the PDB file. These cases are listed below.</p>
<div class="section" id="id4">
<h4>11.7.3.1  Duplicate residues</h4>
<p>All residues in a chain should have a unique id. This id is generated
based on:</p>
<ul class="simple">
<li>The sequence identifier (resseq).</li>
<li>The insertion code (icode).</li>
<li>The hetfield string (“W” for waters and “H_” followed by the residue
name for other hetero residues)</li>
<li>The residue names of the residues in the case of point mutations (to
store the Residue objects in a DisorderedResidue object).</li>
</ul>
<p>If this does not lead to a unique id something is quite likely wrong,
and an exception is generated.</p>
</div>
<div class="section" id="id5">
<h4>11.7.3.2  Duplicate atoms</h4>
<p>All atoms in a residue should have a unique id. This id is generated
based on:</p>
<ul class="simple">
<li>The atom name (without spaces, or with spaces if a problem arises).</li>
<li>The altloc specifier.</li>
</ul>
<p>If this does not lead to a unique id something is quite likely wrong,
and an exception is generated.</p>
</div>
</div>
</div>
<div class="section" id="accessing-the-protein-data-bank">
<h2>11.8  Accessing the Protein Data Bank</h2>
<div class="section" id="downloading-structures-from-the-protein-data-bank">
<h3>11.8.1  Downloading structures from the Protein Data Bank</h3>
<p>Structures can be downloaded from the PDB (Protein Data Bank) by using
the <tt class="docutils literal">retrieve_pdb_file</tt> method on a <tt class="docutils literal">PDBList</tt> object. The argument
for this method is the PDB identifier of the structure.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pdbl</span> <span class="operator">=</span> <span class="name">PDBList</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pdbl</span><span class="operator">.</span><span class="name">retrieve_pdb_file</span><span class="punctuation">(</span><span class="literal string">'1FAT'</span><span class="punctuation">)</span>
</pre>
<p>The <tt class="docutils literal">PDBList</tt> class can also be used as a command-line tool:</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">PDBList</span><span class="operator">.</span><span class="name">py</span> <span class="literal number integer">1</span><span class="name">fat</span>
</pre>
<p>The downloaded file will be called <tt class="docutils literal">pdb1fat.ent</tt> and stored in the
current working directory. Note that the <tt class="docutils literal">retrieve_pdb_file</tt> method
also has an optional argument <tt class="docutils literal">pdir</tt> that specifies a specific
directory in which to store the downloaded PDB files.</p>
<p>The <tt class="docutils literal">retrieve_pdb_file</tt> method also has some options to specify the
compression format used for the download, and the program used for local
decompression (default <tt class="docutils literal">.Z</tt> format and <tt class="docutils literal">gunzip</tt>). In addition, the
PDB ftp site can be specified upon creation of the <tt class="docutils literal">PDBList</tt> object.
By default, the server of the Worldwide Protein Data Bank
(<a class="reference external" href="ftp://ftp.wwpdb.org/pub/pdb/data/structures/divided/pdb/">ftp://ftp.wwpdb.org/pub/pdb/data/structures/divided/pdb/</a>)
is used. See the API documentation for more details. Thanks again to
Kristian Rother for donating this module.</p>
</div>
<div class="section" id="downloading-the-entire-pdb">
<h3>11.8.2  Downloading the entire PDB</h3>
<p>The following commands will store all PDB files in the <tt class="docutils literal">/data/pdb</tt>
directory:</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">PDBList</span><span class="operator">.</span><span class="name">py</span> <span class="name builtin">all</span> <span class="operator">/</span><span class="name">data</span><span class="operator">/</span><span class="name">pdb</span>

<span class="name">python</span> <span class="name">PDBList</span><span class="operator">.</span><span class="name">py</span> <span class="name builtin">all</span> <span class="operator">/</span><span class="name">data</span><span class="operator">/</span><span class="name">pdb</span> <span class="operator">-</span><span class="name">d</span>
</pre>
<p>The API method for this is called <tt class="docutils literal">download_entire_pdb</tt>. Adding the
<tt class="docutils literal"><span class="pre">-d</span></tt> option will store all files in the same directory. Otherwise,
they are sorted into PDB-style subdirectories according to their PDB
ID’s. Depending on the traffic, a complete download will take 2-4 days.</p>
</div>
<div class="section" id="keeping-a-local-copy-of-the-pdb-up-to-date">
<h3>11.8.3  Keeping a local copy of the PDB up to date</h3>
<p>This can also be done using the <tt class="docutils literal">PDBList</tt> object. One simply creates a
<tt class="docutils literal">PDBList</tt> object (specifying the directory where the local copy of the
PDB is present) and calls the <tt class="docutils literal">update_pdb</tt> method:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pl</span> <span class="operator">=</span> <span class="name">PDBList</span><span class="punctuation">(</span><span class="name">pdb</span><span class="operator">=</span><span class="literal string">'/data/pdb'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pl</span><span class="operator">.</span><span class="name">update_pdb</span><span class="punctuation">()</span>
</pre>
<p>One can of course make a weekly <tt class="docutils literal">cronjob</tt> out of this to keep the
local copy automatically up-to-date. The PDB ftp site can also be
specified (see API documentation).</p>
<p><tt class="docutils literal">PDBList</tt> has some additional methods that can be of use. The
<tt class="docutils literal">get_all_obsolete</tt> method can be used to get a list of all obsolete
PDB entries. The <tt class="docutils literal">changed_this_week</tt> method can be used to obtain the
entries that were added, modified or obsoleted during the current week.
For more info on the possibilities of <tt class="docutils literal">PDBList</tt>, see the API
documentation.</p>
</div>
</div>
<div class="section" id="general-questions">
<h2>11.9  General questions</h2>
<div class="section" id="how-well-tested-is-bio-pdb">
<h3>11.9.1  How well tested is Bio.PDB?</h3>
<p>Pretty well, actually. Bio.PDB has been extensively tested on nearly
5500 structures from the PDB - all structures seemed to be parsed
correctly. More details can be found in the Bio.PDB Bioinformatics
article. Bio.PDB has been used/is being used in many research projects
as a reliable tool. In fact, I’m using Bio.PDB almost daily for research
purposes and continue working on improving it and adding new features.</p>
</div>
<div class="section" id="how-fast-is-it">
<h3>11.9.2  How fast is it?</h3>
<p>The <tt class="docutils literal">PDBParser</tt> performance was tested on about 800 structures (each
belonging to a unique SCOP superfamily). This takes about 20 minutes, or
on average 1.5 seconds per structure. Parsing the structure of the large
ribosomal subunit (1FKK), which contains about 64000 atoms, takes 10
seconds on a 1000 MHz PC. In short: it’s more than fast enough for many
applications.</p>
</div>
<div class="section" id="is-there-support-for-molecular-graphics">
<h3>11.9.3  Is there support for molecular graphics?</h3>
<p>Not directly, mostly since there are quite a few Python based/Python
aware solutions already, that can potentially be used with Bio.PDB. My
choice is Pymol, BTW (I’ve used this successfully with Bio.PDB, and
there will probably be specific PyMol modules in Bio.PDB soon/some day).
Python based/aware molecular graphics solutions include:</p>
<ul class="simple">
<li>PyMol:
<a class="reference external" href="http://pymol.sourceforge.net/">http://pymol.sourceforge.net/</a></li>
<li>Chimera:
<a class="reference external" href="http://www.cgl.ucsf.edu/chimera/">http://www.cgl.ucsf.edu/chimera/</a></li>
<li>PMV:
<a class="reference external" href="http://www.scripps.edu/~sanner/python/">http://www.scripps.edu/~sanner/python/</a></li>
<li>Coot:
<a class="reference external" href="http://www.ysbl.york.ac.uk/~emsley/coot/">http://www.ysbl.york.ac.uk/~emsley/coot/</a></li>
<li>CCP4mg:
<a class="reference external" href="http://www.ysbl.york.ac.uk/~lizp/molgraphics.html">http://www.ysbl.york.ac.uk/~lizp/molgraphics.html</a></li>
<li>mmLib:
<a class="reference external" href="http://pymmlib.sourceforge.net/">http://pymmlib.sourceforge.net/</a></li>
<li>VMD:
<a class="reference external" href="http://www.ks.uiuc.edu/Research/vmd/">http://www.ks.uiuc.edu/Research/vmd/</a></li>
<li>MMTK:
<a class="reference external" href="http://starship.python.net/crew/hinsen/MMTK/">http://starship.python.net/crew/hinsen/MMTK/</a></li>
</ul>
</div>
<div class="section" id="whos-using-bio-pdb">
<h3>11.9.4  Who’s using Bio.PDB?</h3>
<p>Bio.PDB was used in the construction of DISEMBL, a web server that
predicts disordered regions in proteins
(<a class="reference external" href="http://dis.embl.de/">http://dis.embl.de/</a>), and COLUMBA, a
website that provides annotated protein structures
(<a class="reference external" href="http://www.columba-db.de/">http://www.columba-db.de/</a>). Bio.PDB
has also been used to perform a large scale search for active sites
similarities between protein structures in the PDB
[<a class="reference external" href="#hamelryck2003b">19</a>, Hamelryck, 2003], and to develop a new
algorithm that identifies linear secondary structure elements
[<a class="reference external" href="#majumdar2005">26</a>, Majumdar <em>et al.</em>, 2005].</p>
<p>Judging from requests for features and information, Bio.PDB is also used
by several LPCs (Large Pharmaceutical Companies :-).</p>
</div>
</div>
</div>
<div class="section" id="chapter-12-bio-popgen-population-genetics">
<h1>Chapter 12  Bio.PopGen: Population genetics</h1>
<p>Bio.PopGen is a Biopython module supporting population genetics,
available in Biopython 1.44 onwards.</p>
<p>The medium term objective for the module is to support widely used data
formats, applications and databases. This module is currently under
intense development and support for new features should appear at a
rather fast pace. Unfortunately this might also entail some instability
on the API, especially if you are using a development version. APIs that
are made available on our official public releases should be much more
stable.</p>
<div class="section" id="genepop">
<h2>12.1  GenePop</h2>
<p>GenePop
(<a class="reference external" href="http://genepop.curtin.edu.au/">http://genepop.curtin.edu.au/</a>)
is a popular population genetics software package supporting
Hardy-Weinberg tests, linkage desiquilibrium, population diferentiation,
basic statistics, <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 and migration estimates, among
others. GenePop does not supply sequence based statistics as it doesn’t
handle sequence data. The GenePop file format is supported by a wide
range of other population genetic software applications, thus making it
a relevant format in the population genetics field.</p>
<p>Bio.PopGen provides a parser and generator of GenePop file format.
Utilities to manipulate the content of a record are also provided. Here
is an example on how to read a GenePop file (you can find example
GenePop data files in the Test/PopGen directory of Biopython):</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen</span> <span class="keyword namespace">import</span> <span class="name">GenePop</span>

<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;example.gen&quot;</span><span class="punctuation">)</span>
<span class="name">rec</span> <span class="operator">=</span> <span class="name">GenePop</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>This will read a file called example.gen and parse it. If you do print
rec, the record will be output again, in GenePop format.</p>
<p>The most important information in rec will be the loci names and
population information (but there is more – use help(GenePop.Record) to
check the API documentation). Loci names can be found on rec.loci_list.
Population information can be found on rec.populations. Populations is a
list with one element per population. Each element is itself a list of
individuals, each individual is a pair composed by individual name and a
list of alleles (2 per marker), here is an example for rec.populations:</p>
<pre class="code python literal-block">
<span class="punctuation">[</span>
    <span class="punctuation">[</span>
        <span class="punctuation">(</span><span class="literal string">'Ind1'</span><span class="punctuation">,</span> <span class="punctuation">[(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">),</span>    <span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="literal number integer">200</span><span class="punctuation">,</span> <span class="literal number integer">201</span><span class="punctuation">)],</span>
        <span class="punctuation">(</span><span class="literal string">'Ind2'</span><span class="punctuation">,</span> <span class="punctuation">[(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name builtin pseudo">None</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="name builtin pseudo">None</span><span class="punctuation">,</span> <span class="name builtin pseudo">None</span><span class="punctuation">)],</span>
    <span class="punctuation">],</span>
    <span class="punctuation">[</span>
        <span class="punctuation">(</span><span class="literal string">'Other1'</span><span class="punctuation">,</span> <span class="punctuation">[(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">),</span>  <span class="punctuation">(</span><span class="literal number integer">4</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="literal number integer">200</span><span class="punctuation">,</span> <span class="literal number integer">200</span><span class="punctuation">)],</span>
    <span class="punctuation">]</span>
<span class="punctuation">]</span>
</pre>
<p>So we have two populations, the first with two individuals, the second
with only one. The first individual of the first population is called
Ind1, allelic information for each of the 3 loci follows. Please note
that for any locus, information might be missing (see as an example,
Ind2 above).</p>
<p>A few utility functions to manipulate GenePop records are made
available, here is an example:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen</span> <span class="keyword namespace">import</span> <span class="name">GenePop</span>

<span class="comment">#Imagine that you have loaded rec, as per the code snippet above...</span>

<span class="name">rec</span><span class="operator">.</span><span class="name">remove_population</span><span class="punctuation">(</span><span class="name">pos</span><span class="punctuation">)</span>
<span class="comment">#Removes a population from a record, pos is the population position in</span>
<span class="comment">#  rec.populations, remember that it starts on position 0.</span>
<span class="comment">#  rec is altered.</span>

<span class="name">rec</span><span class="operator">.</span><span class="name">remove_locus_by_position</span><span class="punctuation">(</span><span class="name">pos</span><span class="punctuation">)</span>
<span class="comment">#Removes a locus by its position, pos is the locus position in</span>
<span class="comment">#  rec.loci_list, remember that it starts on position 0.</span>
<span class="comment">#  rec is altered.</span>

<span class="name">rec</span><span class="operator">.</span><span class="name">remove_locus_by_name</span><span class="punctuation">(</span><span class="name">name</span><span class="punctuation">)</span>
<span class="comment">#Removes a locus by its name, name is the locus name as in</span>
<span class="comment">#  rec.loci_list. If the name doesn't exist the function fails</span>
<span class="comment">#  silently.</span>
<span class="comment">#  rec is altered.</span>

<span class="name">rec_loci</span> <span class="operator">=</span> <span class="name">rec</span><span class="operator">.</span><span class="name">split_in_loci</span><span class="punctuation">()</span>
<span class="comment">#Splits a record in loci, that is, for each loci, it creates a new</span>
<span class="comment">#  record, with a single loci and all populations.</span>
<span class="comment">#  The result is returned in a dictionary, being each key the locus name.</span>
<span class="comment">#  The value is the GenePop record.</span>
<span class="comment">#  rec is not altered.</span>

<span class="name">rec_pops</span> <span class="operator">=</span>  <span class="name">rec</span><span class="operator">.</span><span class="name">split_in_pops</span><span class="punctuation">(</span><span class="name">pop_names</span><span class="punctuation">)</span>
<span class="comment">#Splits a record in populations, that is, for each population, it creates</span>
<span class="comment">#  a new record, with a single population and all loci.</span>
<span class="comment">#  The result is returned in a dictionary, being each key</span>
<span class="comment">#  the population name. As population names are not available in GenePop,</span>
<span class="comment">#  they are passed in array (pop_names).</span>
<span class="comment">#  The value of each dictionary entry is the GenePop record.</span>
<span class="comment">#  rec is not altered.</span>
</pre>
<p>GenePop does not support population names, a limitation which can be
cumbersome at times. Functionality to enable population names is
currently being planned for Biopython. These extensions won’t break
compatibility in any way with the standard format. In the medium term,
we would also like to support the GenePop web service.</p>
</div>
<div class="section" id="coalescent-simulation">
<h2>12.2  Coalescent simulation</h2>
<p>A coalescent simulation is a backward model of population genetics with
relation to time. A simulation of ancestry is done until the Most Recent
Common Ancestor (MRCA) is found. This ancestry relationship starting on
the MRCA and ending on the current generation sample is sometimes called
a genealogy. Simple cases assume a population of constant size in time,
haploidy, no population structure, and simulate the alleles of a single
locus under no selection pressure.</p>
<p>Coalescent theory is used in many fields like selection detection,
estimation of demographic parameters of real populations or disease gene
mapping.</p>
<p>The strategy followed in the Biopython implementation of the coalescent
was not to create a new, built-in, simulator from scratch but to use an
existing one, SIMCOAL2
(<a class="reference external" href="http://cmpg.unibe.ch/software/simcoal2/">http://cmpg.unibe.ch/software/simcoal2/</a>).
SIMCOAL2 allows for, among others, population structure, multiple
demographic events, simulation of multiple types of loci (SNPs,
sequences, STRs/microsatellites and RFLPs) with recombination, diploidy
multiple chromosomes or ascertainment bias. Notably SIMCOAL2 doesn’t
support any selection model. We recommend reading SIMCOAL2’s
documentation, available in the link above.</p>
<p>The input for SIMCOAL2 is a file specifying the desired demography and
genome, the output is a set of files (typically around 1000) with the
simulated genomes of a sample of individuals per subpopulation. This set
of files can be used in many ways, like to compute confidence intervals
where which certain statistics (e.g., <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 or Tajima D) are
expected to lie. Real population genetics datasets statistics can then
be compared to those confidence intervals.</p>
<p>Biopython coalescent code allows to create demographic scenarios and
genomes and to run SIMCOAL2.</p>
<div class="section" id="creating-scenarios">
<h3>12.2.1  Creating scenarios</h3>
<p>Creating a scenario involves both creating a demography and a chromosome
structure. In many cases (e.g. when doing Approximate Bayesian
Computations – ABC) it is important to test many parameter variations
(e.g. vary the effective population size, <span class="formula"><i>N</i><sub><i>e</i></sub></span>
, between 10,
50, 500 and 1000 individuals). The code provided allows for the
simulation of scenarios with different demographic parameters very
easily.</p>
<p>Below we see how we can create scenarios and then how simulate them.</p>
<div class="section" id="demography">
<h4>12.2.1.1  Demography</h4>
<p>A few predefined demographies are built-in, all have two shared
parameters: sample size (called sample_size on the template, see below
for its use) per deme and deme size, i.e. subpopulation size
(pop_size). All demographies are available as templates where all
parameters can be varied, each template has a system name. The
prefedined demographies/templates are:</p>
<dl class="docutils">
<dt><strong>Single population, constant size</strong></dt>
<dd>The standard parameters are enough to specify it. Template name:
simple.</dd>
<dt><strong>Single population, bottleneck</strong></dt>
<dd>As seen on figure <a class="reference external" href="#fig:bottle">12.2.1.1</a>. The parameters are
current population size (pop_size on template ne3 on figure), time
of expansion, given as the generation in the past when it occurred
(expand_gen), effective population size during bottleneck (ne2),
time of contraction (contract_gen) and original size in the remote
past (ne3). Template name: bottle.</dd>
<dt><strong>Island model</strong></dt>
<dd>The typical island model. The total number of demes is specified by
total_demes and the migration rate by mig. Template name island.</dd>
<dt><strong>Stepping stone model - 1 dimension</strong></dt>
<dd>The stepping stone model in 1 dimension, extremes disconnected. The
total number of demes is total_demes, migration rate is mig.
Template name is ssm_1d.</dd>
<dt><strong>Stepping stone model - 2 dimensions</strong></dt>
<dd>The stepping stone model in 2 dimensions, extremes disconnected. The
parameters are x for the horizontal dimension and y for the vertical
(being the total number of demes x times y), migration rate is mig.
Template name is ssm_2d.</dd>
</dl>
<p><img alt="image4" src="../images/bottle.png" /></p>
<p>In our first example, we will generate a template for a single
population, constant size model with a sample size of 30 and a deme size
of 500. The code for this is:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen.SimCoal.Template</span> <span class="keyword namespace">import</span> <span class="name">generate_simcoal_from_template</span>

<span class="name">generate_simcoal_from_template</span><span class="punctuation">(</span><span class="literal string">'simple'</span><span class="punctuation">,</span>
    <span class="punctuation">[(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="punctuation">[(</span><span class="literal string">'SNP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">24</span><span class="punctuation">,</span> <span class="literal number float">0.0005</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])])],</span>
    <span class="punctuation">[(</span><span class="literal string">'sample_size'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">30</span><span class="punctuation">]),</span>
    <span class="punctuation">(</span><span class="literal string">'pop_size'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">100</span><span class="punctuation">])])</span>
</pre>
<p>Executing this code snippet will generate a file on the current
directory called simple_100_300.par this file can be given as input to
SIMCOAL2 to simulate the demography (below we will see how Biopython can
take care of calling SIMCOAL2).</p>
<p>This code consists of a single function call, let’s discuss it parameter
by parameter.</p>
<p>The first parameter is the template id (from the list above). We are
using the id ’simple’ which is the template for a single population of
constant size along time.</p>
<p>The second parameter is the chromosome structure. Please ignore it for
now, it will be explained in the next section.</p>
<p>The third parameter is a list of all required parameters (recall that
the simple model only needs sample_size and pop_size) and possible
values (in this case each parameter only has a possible value).</p>
<p>Now, let’s consider an example where we want to generate several island
models, and we are interested in varying the number of demes: 10, 50 and
100 with a migration rate of 1%. Sample size and deme size will be the
same as before. Here is the code:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen.SimCoal.Template</span> <span class="keyword namespace">import</span> <span class="name">generate_simcoal_from_template</span>

<span class="name">generate_simcoal_from_template</span><span class="punctuation">(</span><span class="literal string">'island'</span><span class="punctuation">,</span>
    <span class="punctuation">[(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="punctuation">[(</span><span class="literal string">'SNP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">24</span><span class="punctuation">,</span> <span class="literal number float">0.0005</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])])],</span>
    <span class="punctuation">[(</span><span class="literal string">'sample_size'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">30</span><span class="punctuation">]),</span>
    <span class="punctuation">(</span><span class="literal string">'pop_size'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">100</span><span class="punctuation">]),</span>
    <span class="punctuation">(</span><span class="literal string">'mig'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number float">0.01</span><span class="punctuation">]),</span>
    <span class="punctuation">(</span><span class="literal string">'total_demes'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">10</span><span class="punctuation">,</span> <span class="literal number integer">50</span><span class="punctuation">,</span> <span class="literal number integer">100</span><span class="punctuation">])])</span>
</pre>
<p>In this case, 3 files will be generated: island_100_0.01_100_30.par,
island_10_0.01_100_30.par and island_50_0.01_100_30.par. Notice
the rule to make file names: template name, followed by parameter values
in reverse order.</p>
<p>A few, arguably more esoteric template demographies exist (please check
the Bio/PopGen/SimCoal/data directory on Biopython source tree).
Furthermore it is possible for the user to create new templates. That
functionality will be discussed in a future version of this document.</p>
</div>
<div class="section" id="chromosome-structure">
<h4>12.2.1.2  Chromosome structure</h4>
<p>We strongly recommend reading SIMCOAL2 documentation to understand the
full potential available in modeling chromosome structures. In this
subsection we only discuss how to implement chromosome structures using
the Biopython interface, not the underlying SIMCOAL2 capabilities.</p>
<p>We will start by implementing a single chromosome, with 24 SNPs with a
recombination rate immediately on the right of each locus of 0.0005 and
a minimum frequency of the minor allele of 0. This will be specified by
the following list (to be passed as second parameter to the function
generate_simcoal_from_template):</p>
<pre class="code python literal-block">
<span class="punctuation">[(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="punctuation">[(</span><span class="literal string">'SNP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">24</span><span class="punctuation">,</span> <span class="literal number float">0.0005</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])])]</span>
</pre>
<p>This is actually the chromosome structure used in the above examples.</p>
<p>The chromosome structure is represented by a list of chromosomes, each
chromosome (i.e., each element in the list) is composed by a tuple (a
pair): the first element is the number of times the chromosome is to be
repeated (as there might be interest in repeating the same chromosome
many times). The second element is a list of the actual components of
the chromosome. Each element is again a pair, the first member is the
locus type and the second element the parameters for that locus type.
Confused? Before showing more examples let’s review the example above:
We have a list with one element (thus one chromosome), the chromosome is
a single instance (therefore not to be repeated), it is composed of 24
SNPs, with a recombination rate of 0.0005 between each consecutive SNP,
the minimum frequency of the minor allele is 0.0 (i.e, it can be absent
from a certain population).</p>
<p>Let’s see a more complicated example:</p>
<pre class="code python literal-block">
<span class="punctuation">[</span>
  <span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="punctuation">[</span>
       <span class="punctuation">(</span><span class="literal string">'SNP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">24</span><span class="punctuation">,</span> <span class="literal number float">0.0005</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])</span>
      <span class="punctuation">]</span>
  <span class="punctuation">),</span>
  <span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="punctuation">[</span>
       <span class="punctuation">(</span><span class="literal string">'DNA'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">10</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.00005</span><span class="punctuation">,</span> <span class="literal number float">0.33</span><span class="punctuation">]),</span>
       <span class="punctuation">(</span><span class="literal string">'RFLP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.0001</span><span class="punctuation">]),</span>
       <span class="punctuation">(</span><span class="literal string">'MICROSAT'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.001</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])</span>
      <span class="punctuation">]</span>
  <span class="punctuation">)</span>
<span class="punctuation">]</span>
</pre>
<p>We start by having 5 chromosomes with the same structure as above (i.e.,
24 SNPs). We then have 2 chromosomes which have a DNA sequence with 10
nucleotides, 0.0 recombination rate, 0.0005 mutation rate, and a
transition rate of 0.33. Then we have an RFLP with 0.0 recombination
rate to the next locus and a 0.0001 mutation rate. Finally we have a
microsatellite (or STR), with 0.0 recombination rate to the next locus
(note, that as this is a single microsatellite which has no loci
following, this recombination rate here is irrelevant), with a mutation
rate of 0.001, geometric parameter of 0.0 and a range constraint of 0.0
(for information about this parameters please consult the SIMCOAL2
documentation, you can use them to simulate various mutation models,
including the typical – for microsatellites – stepwise mutation model
among others).</p>
</div>
</div>
<div class="section" id="running-simcoal2">
<h3>12.2.2  Running SIMCOAL2</h3>
<p>We now discuss how to run SIMCOAL2 from inside Biopython. It is required
that the binary for SIMCOAL2 is called simcoal2 (or simcoal2.exe on
Windows based platforms), please note that the typical name when
downloading the program is in the format simcoal2_x_y. As such, when
installing SIMCOAL2 you will need to rename of the downloaded executable
so that Biopython can find it.</p>
<p>It is possible to run SIMCOAL2 on files that were not generated using
the method above (e.g., writing a parameter file by hand), but we will
show an example by creating a model using the framework presented above.</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen.SimCoal.Template</span> <span class="keyword namespace">import</span> <span class="name">generate_simcoal_from_template</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen.SimCoal.Controller</span> <span class="keyword namespace">import</span> <span class="name">SimCoalController</span>


<span class="name">generate_simcoal_from_template</span><span class="punctuation">(</span><span class="literal string">'simple'</span><span class="punctuation">,</span>
    <span class="punctuation">[</span>
      <span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="punctuation">[</span>
           <span class="punctuation">(</span><span class="literal string">'SNP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">24</span><span class="punctuation">,</span> <span class="literal number float">0.0005</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])</span>
          <span class="punctuation">]</span>
      <span class="punctuation">),</span>
      <span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="punctuation">[</span>
           <span class="punctuation">(</span><span class="literal string">'DNA'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">10</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.00005</span><span class="punctuation">,</span> <span class="literal number float">0.33</span><span class="punctuation">]),</span>
           <span class="punctuation">(</span><span class="literal string">'RFLP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.0001</span><span class="punctuation">]),</span>
           <span class="punctuation">(</span><span class="literal string">'MICROSAT'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.001</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])</span>
          <span class="punctuation">]</span>
      <span class="punctuation">)</span>
    <span class="punctuation">],</span>
    <span class="punctuation">[(</span><span class="literal string">'sample_size'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">30</span><span class="punctuation">]),</span>
    <span class="punctuation">(</span><span class="literal string">'pop_size'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">100</span><span class="punctuation">])])</span>

<span class="name">ctrl</span> <span class="operator">=</span> <span class="name">SimCoalController</span><span class="punctuation">(</span><span class="literal string">'.'</span><span class="punctuation">)</span>
<span class="name">ctrl</span><span class="operator">.</span><span class="name">run_simcoal</span><span class="punctuation">(</span><span class="literal string">'simple_100_30.par'</span><span class="punctuation">,</span> <span class="literal number integer">50</span><span class="punctuation">)</span>
</pre>
<p>The lines of interest are the last two (plus the new import). Firstly a
controller for the application is created. The directory where the
binary is located has to be specified.</p>
<p>The simulator is then run on the last line: we know, from the rules
explained above, that the input file name is simple_100_30.par for the
simulation parameter file created. We then specify that we want to run
50 independent simulations, by default Biopython requests a simulation
of diploid data, but a third parameter can be added to simulate haploid
data (adding as a parameter the string ’0’). SIMCOAL2 will now run
(please note that this can take quite a lot of time) and will create a
directory with the simulation results. The results can now be analysed
(typically studying the data with Arlequin3). In the future Biopython
might support reading the Arlequin3 format and thus allowing for the
analysis of SIMCOAL2 data inside Biopython.</p>
</div>
</div>
<div class="section" id="other-applications">
<h2>12.3  Other applications</h2>
<p>Here we discuss interfaces and utilities to deal with population
genetics’ applications which arguably have a smaller user base.</p>
<div class="section" id="fdist-detecting-selection-and-molecular-adaptation">
<h3>12.3.1  FDist: Detecting selection and molecular adaptation</h3>
<p>FDist is a selection detection application suite based on computing
(i.e. simulating) a “neutral” confidence interval based on
<span class="formula"><i>F</i><sub><i>st</i></sub></span>
 and heterozygosity. Markers (which can be SNPs,
microsatellites, AFLPs among others) which lie outside the “neutral”
interval are to be considered as possible candidates for being under
selection.</p>
<p>FDist is mainly used when the number of markers is considered enough to
estimate an average <span class="formula"><i>F</i><sub><i>st</i></sub></span>
, but not enough to either have
outliers calculated from the dataset directly or, with even more markers
for which the relative positions in the genome are known, to use
approaches based on, e.g., Extended Haplotype Heterozygosity (EHH).</p>
<p>The typical usage pattern for FDist is as follows:</p>
<ol class="arabic simple">
<li>Import a dataset from an external format into FDist format.</li>
<li>Compute average <span class="formula"><i>F</i><sub><i>st</i></sub></span>
. This is done by datacal inside
FDist.</li>
<li>Simulate “neutral” markers based on the average <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 and
expected number of total populations. This is the core operation,
done by fdist inside FDist.</li>
<li>Calculate the confidence interval, based on the desired confidence
boundaries (typically 95% or 99%). This is done by cplot and is
mainly used to plot the interval.</li>
<li>Assess each marker status against the simulation “neutral” confidence
interval. Done by pv. This is used to detect the outlier status of
each marker against the simulation.</li>
</ol>
<p>We will now discuss each step with illustrating example code (for this
example to work FDist binaries have to be on the executable PATH).</p>
<p>The FDist data format is application specific and is not used at all by
other applications, as such you will probably have to convert your data
for use with FDist. Biopython can help you do this. Here is an example
converting from GenePop format to FDist format (along with imports that
will be needed on examples further below):</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen</span> <span class="keyword namespace">import</span> <span class="name">GenePop</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen</span> <span class="keyword namespace">import</span> <span class="name">FDist</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen.FDist</span> <span class="keyword namespace">import</span> <span class="name">Controller</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen.FDist.Utils</span> <span class="keyword namespace">import</span> <span class="name">convert_genepop_to_fdist</span>

<span class="name">gp_rec</span> <span class="operator">=</span> <span class="name">GenePop</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;example.gen&quot;</span><span class="punctuation">))</span>
<span class="name">fd_rec</span> <span class="operator">=</span> <span class="name">convert_genepop_to_fdist</span><span class="punctuation">(</span><span class="name">gp_rec</span><span class="punctuation">)</span>
<span class="name">in_file</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;infile&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="name">in_file</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name builtin">str</span><span class="punctuation">(</span><span class="name">fd_rec</span><span class="punctuation">))</span>
<span class="name">in_file</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>In this code we simply parse a GenePop file and convert it to a FDist
record.</p>
<p>Printing an FDist record will generate a string that can be directly
saved to a file and supplied to FDist. FDist requires the input file to
be called infile, therefore we save the record on a file with that name.</p>
<p>The most important fields on a FDist record are: num_pops, the number
of populations; num_loci, the number of loci and loci_data with the
marker data itself. Most probably the details of the record are of no
interest to the user, as the record only purpose is to be passed to
FDist.</p>
<p>The next step is to calculate the average <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 of the
dataset (along with the sample size):</p>
<pre class="code python literal-block">
<span class="name">ctrl</span> <span class="operator">=</span> <span class="name">Controller</span><span class="operator">.</span><span class="name">FDistController</span><span class="punctuation">()</span>
<span class="name">fst</span><span class="punctuation">,</span> <span class="name">samp_size</span> <span class="operator">=</span> <span class="name">ctrl</span><span class="operator">.</span><span class="name">run_datacal</span><span class="punctuation">()</span>
</pre>
<p>On the first line we create an object to control the call of FDist
suite, this object will be used further on in order to call other suite
applications.</p>
<p>On the second line we call the datacal application which computes the
average <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 and the sample size. It is worth noting that
the <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 computed by datacal is a <em>variation</em> of Weir and
Cockerham’s θ.</p>
<p>We can now call the main fdist application in order to simulate neutral
markers.</p>
<pre class="code python literal-block">
<span class="name">sim_fst</span> <span class="operator">=</span> <span class="name">ctrl</span><span class="operator">.</span><span class="name">run_fdist</span><span class="punctuation">(</span><span class="name">npops</span> <span class="operator">=</span> <span class="literal number integer">15</span><span class="punctuation">,</span> <span class="name">nsamples</span> <span class="operator">=</span> <span class="name">fd_rec</span><span class="operator">.</span><span class="name">num_pops</span><span class="punctuation">,</span> <span class="name">fst</span> <span class="operator">=</span> <span class="name">fst</span><span class="punctuation">,</span>
    <span class="name">sample_size</span> <span class="operator">=</span> <span class="name">samp_size</span><span class="punctuation">,</span> <span class="name">mut</span> <span class="operator">=</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">num_sims</span> <span class="operator">=</span> <span class="literal number integer">40000</span><span class="punctuation">)</span>
</pre>
<dl class="docutils">
<dt><strong>npops</strong></dt>
<dd>Number of populations existing in nature. This is really a
“guestimate”. Has to be lower than 100.</dd>
<dt><strong>nsamples</strong></dt>
<dd>Number of populations sampled, has to be lower than npops.</dd>
<dt><strong>fst</strong></dt>
<dd>Average <span class="formula"><i>F</i><sub><i>st</i></sub></span>
.</dd>
<dt><strong>sample_size</strong></dt>
<dd>Average number of individuals sampled on each population.</dd>
<dt><strong>mut</strong></dt>
<dd>Mutation model: 0 - Infinite alleles; 1 - Stepwise mutations</dd>
<dt><strong>num_sims</strong></dt>
<dd>Number of simulations to perform. Typically a number around 40000
will be OK, but if you get a confidence interval that looks sharp
(this can be detected when plotting the confidence interval computed
below) the value can be increased (a suggestion would be steps of
10000 simulations).</dd>
</dl>
<p>The confusion in wording between number of samples and sample size stems
from the original application.</p>
<p>A file named out.dat will be created with the simulated heterozygosities
and <span class="formula"><i>F</i><sub><i>st</i></sub></span>
s, it will have as many lines as the number of
simulations requested.</p>
<p>Note that fdist returns the average <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 that it was
<em>capable</em> of simulating, for more details about this issue please read
below the paragraph on approximating the desired average
<span class="formula"><i>F</i><sub><i>st</i></sub></span>
.</p>
<p>The next (optional) step is to calculate the confidence interval:</p>
<pre class="code python literal-block">
<span class="name">cpl_interval</span> <span class="operator">=</span> <span class="name">ctrl</span><span class="operator">.</span><span class="name">run_cplot</span><span class="punctuation">(</span><span class="name">ci</span><span class="operator">=</span><span class="literal number float">0.99</span><span class="punctuation">)</span>
</pre>
<p>You can only call cplot after having run fdist.</p>
<p>This will calculate the confidence intervals (99% in this case) for a
previous fdist run. A list of quadruples is returned. The first element
represents the heterozygosity, the second the lower bound of
<span class="formula"><i>F</i><sub><i>st</i></sub></span>
 confidence interval for that heterozygosity, the third
the average and the fourth the upper bound. This can be used to trace
the confidence interval contour. This list is also written to a file,
out.cpl.</p>
<p>The main purpose of this step is return a set of points which can be
easily used to plot a confidence interval. It can be skipped if the
objective is only to assess the status of each marker against the
simulation, which is the next step...</p>
<pre class="code python literal-block">
<span class="name">pv_data</span> <span class="operator">=</span> <span class="name">ctrl</span><span class="operator">.</span><span class="name">run_pv</span><span class="punctuation">()</span>
</pre>
<p>You can only call cplot after having run datacal and fdist.</p>
<p>This will use the simulated markers to assess the status of each
individual real marker. A list, in the same order than the loci_list
that is on the FDist record (which is in the same order that the GenePop
record) is returned. Each element in the list is a quadruple, the
fundamental member of each quadruple is the last element (regarding the
other elements, please refer to the pv documentation – for the sake of
simplicity we will not discuss them here) which returns the probability
of the simulated <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 being lower than the marker
<span class="formula"><i>F</i><sub><i>st</i></sub></span>
. Higher values would indicate a stronger candidate for
positive selection, lower values a candidate for balancing selection,
and intermediate values a possible neutral marker. What is “higher”,
“lower” or “intermediate” is really a subjective issue, but taking a
“confidence interval” approach and considering a 95% confidence
interval, “higher” would be between 0.95 and 1.0, “lower” between 0.0
and 0.05 and “intermediate” between 0.05 and 0.95.</p>
<div class="section" id="approximating-the-desired-average-f-st">
<h4>12.3.1.1  Approximating the desired average <span class="formula"><i>F</i><sub><i>st</i></sub></span>
</h4>
<p>Fdist tries to approximate the desired average <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 by doing
a coalescent simulation using migration rates based on the formula</p>
<div class="formula">
<i>N</i><sub><i>m</i></sub> = <span class="fraction"><span class="ignored">(</span><span class="numerator">1 − <i>F</i><sub><i>st</i></sub></span><span class="ignored">)/(</span><span class="denominator">4<i>F</i><sub><i>st</i></sub></span><span class="ignored">)</span></span>
</div>
<p>This formula assumes a few premises like an infinite number of
populations.</p>
<p>In practice, when the number of populations is low, the mutation model
is stepwise and the sample size increases, fdist will not be able to
simulate an acceptable approximate average <span class="formula"><i>F</i><sub><i>st</i></sub></span>
.</p>
<p>To address that, a function is provided to iteratively approach the
desired value by running several fdists in sequence. This approach is
computationally more intensive than running a single fdist run, but
yields good results. The following code runs fdist approximating the
desired <span class="formula"><i>F</i><sub><i>st</i></sub></span>
:</p>
<pre class="code python literal-block">
<span class="name">sim_fst</span> <span class="operator">=</span> <span class="name">ctrl</span><span class="operator">.</span><span class="name">run_fdist_force_fst</span><span class="punctuation">(</span><span class="name">npops</span> <span class="operator">=</span> <span class="literal number integer">15</span><span class="punctuation">,</span> <span class="name">nsamples</span> <span class="operator">=</span> <span class="name">fd_rec</span><span class="operator">.</span><span class="name">num_pops</span><span class="punctuation">,</span>
    <span class="name">fst</span> <span class="operator">=</span> <span class="name">fst</span><span class="punctuation">,</span> <span class="name">sample_size</span> <span class="operator">=</span> <span class="name">samp_size</span><span class="punctuation">,</span> <span class="name">mut</span> <span class="operator">=</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">num_sims</span> <span class="operator">=</span> <span class="literal number integer">40000</span><span class="punctuation">,</span>
    <span class="name">limit</span> <span class="operator">=</span> <span class="literal number float">0.05</span><span class="punctuation">)</span>
</pre>
<p>The only new optional parameter, when comparing with run_fdist, is
limit which is the desired maximum error. run_fdist can (and probably
should) be safely replaced with run_fdist_force_fst.</p>
</div>
<div class="section" id="final-notes">
<h4>12.3.1.2  Final notes</h4>
<p>The process to determine the average <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 can be more
sophisticated than the one presented here. For more information we refer
you to the FDist README file. Biopython’s code can be used to implement
more sophisticated approaches.</p>
</div>
</div>
</div>
<div class="section" id="future-developments">
<h2>12.4  Future Developments</h2>
<p>The most desired future developments would be the ones you add yourself
;) .</p>
<p>That being said, already existing fully functional code is currently
being incorporated in Bio.PopGen, that code covers the applications
FDist and SimCoal2, the HapMap and UCSC Table Browser databases and some
simple statistics like <span class="formula"><i>F</i><sub><i>st</i></sub></span>
, or allele counts.</p>
</div>
</div>
<div class="section" id="chapter-13-phylogenetics-with-bio-phylo">
<h1>Chapter 13  Phylogenetics with Bio.Phylo</h1>
<p>The Bio.Phylo module was introduced in Biopython 1.54. Following the
lead of SeqIO and AlignIO, it aims to provide a common way to work with
phylogenetic trees independently of the source data format, as well as a
consistent API for I/O operations.</p>
<p>Bio.Phylo is described in an open-access journal article
[<a class="reference external" href="#talevich2012">9</a>, Talevich <em>et al.</em>, 2012], which you might also
find helpful.</p>
<div class="section" id="demo-whats-in-a-tree">
<h2>13.1  Demo: What’s in a Tree?</h2>
<p>To get acquainted with the module, let’s start with a tree that we’ve
already constructed, and inspect it a few different ways. Then we’ll
colorize the branches, to use a special phyloXML feature, and finally
save it.</p>
<p>In a terminal, create a simple Newick file using your favorite text
editor:</p>
<pre class="code python literal-block">
<span class="operator">%</span> <span class="name">cat</span> <span class="operator">&gt;</span> <span class="name">simple</span><span class="operator">.</span><span class="name">dnd</span> <span class="operator">&lt;&lt;</span><span class="name">EOF</span>
<span class="operator">&gt;</span> <span class="punctuation">(((</span><span class="name">A</span><span class="punctuation">,</span><span class="name">B</span><span class="punctuation">),(</span><span class="name">C</span><span class="punctuation">,</span><span class="name">D</span><span class="punctuation">)),(</span><span class="name">E</span><span class="punctuation">,</span><span class="name">F</span><span class="punctuation">,</span><span class="name">G</span><span class="punctuation">));</span>
<span class="operator">&gt;</span> <span class="name">EOF</span>
</pre>
<p>This tree has no branch lengths, only a topology and labelled terminals.
(If you have a real tree file available, you can follow this demo using
that instead.)</p>
<p>Launch the Python interpreter of your choice:</p>
<pre class="code python literal-block">
<span class="operator">%</span> <span class="name">ipython</span> <span class="operator">-</span><span class="name">pylab</span>
</pre>
<p>For interactive work, launching the IPython interpreter with the
<tt class="docutils literal"><span class="pre">-pylab</span></tt> flag enables <strong>matplotlib</strong> integration, so graphics will pop
up automatically. We’ll use that during this demo.</p>
<p>Now, within Python, read the tree file, giving the file name and the
name of the format.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Phylo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;simple.dnd&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;newick&quot;</span><span class="punctuation">)</span>
</pre>
<p>Printing the tree object as a string gives us a look at the entire
object hierarchy.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">tree</span>

<span class="name">Tree</span><span class="punctuation">(</span><span class="name">weight</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">rooted</span><span class="operator">=</span><span class="name builtin pseudo">False</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;&quot;</span><span class="punctuation">)</span>
    <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">)</span>
        <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">)</span>
                <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;A&quot;</span><span class="punctuation">)</span>
                <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;B&quot;</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">)</span>
                <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;C&quot;</span><span class="punctuation">)</span>
                <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;D&quot;</span><span class="punctuation">)</span>
        <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;E&quot;</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;F&quot;</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;G&quot;</span><span class="punctuation">)</span>
</pre>
<p>The <tt class="docutils literal">Tree</tt> object contains global information about the tree, such as
whether it’s rooted or unrooted. It has one root clade, and under that,
it’s nested lists of clades all the way down to the tips.</p>
<p>The function <tt class="docutils literal">draw_ascii</tt> creates a simple ASCII-art (plain text)
dendrogram. This is a convenient visualization for interactive
exploration, in case better graphical tools aren’t available.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_ascii</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
                                                    <span class="name">________________________</span> <span class="name">A</span>
                           <span class="name">________________________</span><span class="operator">|</span>
                          <span class="operator">|</span>                        <span class="operator">|</span><span class="name">________________________</span> <span class="name">B</span>
  <span class="name">________________________</span><span class="operator">|</span>
 <span class="operator">|</span>                        <span class="operator">|</span>                         <span class="name">________________________</span> <span class="name">C</span>
 <span class="operator">|</span>                        <span class="operator">|</span><span class="name">________________________</span><span class="operator">|</span>
<span class="name">_</span><span class="operator">|</span>                                                 <span class="operator">|</span><span class="name">________________________</span> <span class="name">D</span>
 <span class="operator">|</span>
 <span class="operator">|</span>                         <span class="name">________________________</span> <span class="name">E</span>
 <span class="operator">|</span>                        <span class="operator">|</span>
 <span class="operator">|</span><span class="name">________________________</span><span class="operator">|</span><span class="name">________________________</span> <span class="name">F</span>
                          <span class="operator">|</span>
                          <span class="operator">|</span><span class="name">________________________</span> <span class="name">G</span>
</pre>
<p>If you have <strong>matplotlib</strong> or <strong>pylab</strong> installed, you can create a
graphic using the <tt class="docutils literal">draw</tt> function (see Fig.
<a class="reference external" href="#fig:phylo-simple-draw">13.1</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">rooted</span> <span class="operator">=</span> <span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
</pre>
<p><img alt="image5" src="../images/phylo-simple-draw.png" /></p>
<div class="section" id="coloring-branches-within-a-tree">
<h3>13.1.1  Coloring branches within a tree</h3>
<p>The functions <tt class="docutils literal">draw</tt> and <tt class="docutils literal">draw_graphviz</tt> support the display of
different colors and branch widths in a tree. As of Biopython 1.59, the
<tt class="docutils literal">color</tt> and <tt class="docutils literal">width</tt> attributes are available on the basic Clade
object and there’s nothing extra required to use them. Both attributes
refer to the branch leading the given clade, and apply recursively, so
all descendent branches will also inherit the assigned width and color
values during display.</p>
<p>In earlier versions of Biopython, these were special features of
PhyloXML trees, and using the attributes required first converting the
tree to a subclass of the basic tree object called Phylogeny, from the
Bio.Phylo.PhyloXML module.</p>
<p>In Biopython 1.55 and later, this is a convenient tree method:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">tree</span><span class="operator">.</span><span class="name">as_phyloxml</span><span class="punctuation">()</span>
</pre>
<p>In Biopython 1.54, you can accomplish the same thing with one extra
import:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Phylo.PhyloXML</span> <span class="keyword namespace">import</span> <span class="name">Phylogeny</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylogeny</span><span class="operator">.</span><span class="name">from_tree</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
</pre>
<p>Note that the file formats Newick and Nexus don’t support branch colors
or widths, so if you use these attributes in Bio.Phylo, you will only be
able to save the values in PhyloXML format. (You can still save a tree
as Newick or Nexus, but the color and width values will be skipped in
the output file.)</p>
<p>Now we can begin assigning colors. First, we’ll color the root clade
gray. We can do that by assigning the 24-bit color value as an RGB
triple, an HTML-style hex string, or the name of one of the predefined
colors.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">root</span><span class="operator">.</span><span class="name">color</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="literal number integer">128</span><span class="punctuation">,</span> <span class="literal number integer">128</span><span class="punctuation">,</span> <span class="literal number integer">128</span><span class="punctuation">)</span>
</pre>
<p>Or:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">root</span><span class="operator">.</span><span class="name">color</span> <span class="operator">=</span> <span class="literal string">&quot;#808080&quot;</span>
</pre>
<p>Or:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">root</span><span class="operator">.</span><span class="name">color</span> <span class="operator">=</span> <span class="literal string">&quot;gray&quot;</span>
</pre>
<p>Colors for a clade are treated as cascading down through the entire
clade, so when we colorize the root here, it turns the whole tree gray.
We can override that by assigning a different color lower down on the
tree.</p>
<p>Let’s target the most recent common ancestor (MRCA) of the nodes named
“E” and “F”. The <tt class="docutils literal">common_ancestor</tt> method returns a reference to that
clade in the original tree, so when we color that clade “salmon”, the
color will show up in the original tree.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mrca</span> <span class="operator">=</span> <span class="name">tree</span><span class="operator">.</span><span class="name">common_ancestor</span><span class="punctuation">({</span><span class="literal string">&quot;name&quot;</span><span class="punctuation">:</span> <span class="literal string">&quot;E&quot;</span><span class="punctuation">},</span> <span class="punctuation">{</span><span class="literal string">&quot;name&quot;</span><span class="punctuation">:</span> <span class="literal string">&quot;F&quot;</span><span class="punctuation">})</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mrca</span><span class="operator">.</span><span class="name">color</span> <span class="operator">=</span> <span class="literal string">&quot;salmon&quot;</span>
</pre>
<p>If we happened to know exactly where a certain clade is in the tree, in
terms of nested list entries, we can jump directly to that position in
the tree by indexing it. Here, the index <tt class="docutils literal">[0,1]</tt> refers to the second
child of the first child of the root.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">clade</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="literal number integer">1</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">color</span> <span class="operator">=</span> <span class="literal string">&quot;blue&quot;</span>
</pre>
<p>Finally, show our work (see Fig. <a class="reference external" href="#fig:phylo-color-draw">13.1.1</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
</pre>
<p><img alt="image6" src="../images/phylo-color-draw.png" /></p>
<p>Note that a clade’s color includes the branch leading to that clade, as
well as its descendents. The common ancestor of E and F turns out to be
just under the root, and with this coloring we can see exactly where the
root of the tree is.</p>
<p>My, we’ve accomplished a lot! Let’s take a break here and save our work.
Call the <tt class="docutils literal">write</tt> function with a file name or handle — here we use
standard output, to see what would be written — and the format
<tt class="docutils literal">phyloxml</tt>. PhyloXML saves the colors we assigned, so you can open
this phyloXML file in another tree viewer like Archaeopteryx, and the
colors will show up there, too.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">sys</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">,</span> <span class="name">sys</span><span class="operator">.</span><span class="name">stdout</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>

<span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">phyloxml</span> <span class="name">xmlns</span><span class="punctuation">:</span><span class="name">phy</span><span class="operator">=</span><span class="literal string">&quot;http://www.phyloxml.org&quot;</span><span class="operator">&gt;</span>
  <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">phylogeny</span> <span class="name">rooted</span><span class="operator">=</span><span class="literal string">&quot;true&quot;</span><span class="operator">&gt;</span>
    <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">clade</span><span class="operator">&gt;</span>
      <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">branch_length</span><span class="operator">&gt;</span><span class="literal number float">1.0</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">branch_length</span><span class="operator">&gt;</span>
      <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">color</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">red</span><span class="operator">&gt;</span><span class="literal number integer">128</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">red</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">green</span><span class="operator">&gt;</span><span class="literal number integer">128</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">green</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">blue</span><span class="operator">&gt;</span><span class="literal number integer">128</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">blue</span><span class="operator">&gt;</span>
      <span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">color</span><span class="operator">&gt;</span>
      <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">clade</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">branch_length</span><span class="operator">&gt;</span><span class="literal number float">1.0</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">branch_length</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">clade</span><span class="operator">&gt;</span>
          <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">branch_length</span><span class="operator">&gt;</span><span class="literal number float">1.0</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">branch_length</span><span class="operator">&gt;</span>
          <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">clade</span><span class="operator">&gt;</span>
            <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">name</span><span class="operator">&gt;</span><span class="name">A</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">name</span><span class="operator">&gt;</span>
            <span class="operator">...</span>
</pre>
<p>The rest of this chapter covers the core functionality of Bio.Phylo in
greater detail. For more examples of using Bio.Phylo, see the cookbook
page on Biopython.org:</p>
<p><a class="reference external" href="http://biopython.org/wiki/Phylo_cookbook">http://biopython.org/wiki/Phylo_cookbook</a></p>
</div>
</div>
<div class="section" id="i-o-functions">
<h2>13.2  I/O functions</h2>
<p>Like SeqIO and AlignIO, Phylo handles file input and output through four
functions: <tt class="docutils literal">parse</tt>, <tt class="docutils literal">read</tt>, <tt class="docutils literal">write</tt> and <tt class="docutils literal">convert</tt>, all of which
support the tree file formats Newick, NEXUS, phyloXML and NeXML.</p>
<p>The <tt class="docutils literal">read</tt> function parses a single tree in the given file and returns
it. Careful; it will raise an error if the file contains more than one
tree, or no trees.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Phylo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;Tests/Nexus/int_node_labels.nwk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;newick&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">tree</span>
</pre>
<p>(Example files are available in the <tt class="docutils literal">Tests/Nexus/</tt> and
<tt class="docutils literal">Tests/PhyloXML/</tt> directories of the Biopython distribution.)</p>
<p>To handle multiple (or an unknown number of) trees, use the <tt class="docutils literal">parse</tt>
function iterates through each of the trees in the given file:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">trees</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;Tests/PhyloXML/phyloxml_examples.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">tree</span> <span class="operator word">in</span> <span class="name">trees</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">tree</span>
</pre>
<p>Write a tree or iterable of trees back to file with the <tt class="docutils literal">write</tt>
function:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">trees</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">Phylo</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;phyloxml_examples.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree1</span> <span class="operator">=</span> <span class="name">trees</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">others</span> <span class="operator">=</span> <span class="name">trees</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">:]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">tree1</span><span class="punctuation">,</span> <span class="literal string">&quot;tree1.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">others</span><span class="punctuation">,</span> <span class="literal string">&quot;other_trees.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">12</span>
</pre>
<p>Convert files between any of the supported formats with the <tt class="docutils literal">convert</tt>
function:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;tree1.dnd&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;newick&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;tree1.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;nexml&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;other_trees.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;other_trees.nex&quot;</span><span class="punctuation">,</span> <span class="literal string">'nexus&quot;)</span>
<span class="literal number integer">12</span>
</pre>
<p>To use strings as input or output instead of actual files, use
<tt class="docutils literal">StringIO</tt> as you would with SeqIO and AlignIO:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Phylo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">StringIO</span> <span class="keyword namespace">import</span> <span class="name">StringIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">StringIO</span><span class="punctuation">(</span><span class="literal string">&quot;(((A,B),(C,D)),(E,F,G));&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;newick&quot;</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="view-and-export-trees">
<h2>13.3  View and export trees</h2>
<p>The simplest way to get an overview of a <tt class="docutils literal">Tree</tt> object is to <tt class="docutils literal">print</tt>
it:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;Tests/PhyloXML/example.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">tree</span>
<span class="name">Phylogeny</span><span class="punctuation">(</span><span class="name">rooted</span><span class="operator">=</span><span class="literal string">'True'</span><span class="punctuation">,</span> <span class="name">description</span><span class="operator">=</span><span class="literal string">'phyloXML allows to use either a &quot;branch_length&quot;</span>
<span class="name">attribute</span><span class="operator">...</span><span class="literal string">', name='</span><span class="name">example</span> <span class="keyword namespace">from</span> <span class="name namespace">Prof.</span> <span class="name namespace">Joe</span> <span class="name namespace">Felsenstein</span><span class="literal string">'s book &quot;Inferring Phyl...'</span><span class="punctuation">)</span>
    <span class="name">Clade</span><span class="punctuation">()</span>
        <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal string">'0.06'</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal string">'0.102'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'A'</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal string">'0.23'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'B'</span><span class="punctuation">)</span>
        <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal string">'0.4'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'C'</span><span class="punctuation">)</span>
</pre>
<p>This is essentially an outline of the object hierarchy Biopython uses to
represent a tree. But more likely, you’d want to see a drawing of the
tree. There are three functions to do this.</p>
<p>As we saw in the demo, <tt class="docutils literal">draw_ascii</tt> prints an ascii-art drawing of the
tree (a rooted phylogram) to standard output, or an open file handle if
given. Not all of the available information about the tree is shown, but
it provides a way to quickly view the tree without relying on any
external dependencies.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;example.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_ascii</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
             <span class="name">__________________</span> <span class="name">A</span>
  <span class="name">__________</span><span class="operator">|</span>
<span class="name">_</span><span class="operator">|</span>          <span class="operator">|</span><span class="name">___________________________________________</span> <span class="name">B</span>
 <span class="operator">|</span>
 <span class="operator">|</span><span class="name">___________________________________________________________________________</span> <span class="name">C</span>
</pre>
<p>The <tt class="docutils literal">draw</tt> function draws a more attractive image using the matplotlib
library. See the API documentation for details on the arguments it
accepts to customize the output.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;example.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">,</span> <span class="name">branch_labels</span><span class="operator">=</span><span class="keyword">lambda</span> <span class="name">c</span><span class="punctuation">:</span> <span class="name">c</span><span class="operator">.</span><span class="name">branch_length</span><span class="punctuation">)</span>
</pre>
<p><img alt="image7" src="../images/phylo-draw-example.png" /></p>
<p><tt class="docutils literal">draw_graphviz</tt> draws an unrooted cladogram, but requires that you
have Graphviz, PyDot or PyGraphviz, NetworkX, and matplotlib (or pylab)
installed. Using the same example as above, and the <tt class="docutils literal">dot</tt> program
included with Graphviz, let’s draw a rooted tree (see
Fig. <a class="reference external" href="#fig:phylo-dot">13.3</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;example.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_graphviz</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">,</span> <span class="name">prog</span><span class="operator">=</span><span class="literal string">'dot'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">pylab</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pylab</span><span class="operator">.</span><span class="name">show</span><span class="punctuation">()</span>                    <span class="comment"># Displays the tree in an interactive viewer</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pylab</span><span class="operator">.</span><span class="name">savefig</span><span class="punctuation">(</span><span class="literal string">'phylo-dot.png'</span><span class="punctuation">)</span>  <span class="comment"># Creates a PNG file of the same graphic</span>
</pre>
<p><img alt="image8" src="../images/phylo-dot.png" /></p>
<p>(Tip: If you execute IPython with the <tt class="docutils literal"><span class="pre">-pylab</span></tt> option, calling
<tt class="docutils literal">draw_graphviz</tt> causes the matplotlib viewer to launch automatically
without manually calling <tt class="docutils literal">show()</tt>.)</p>
<p>This exports the tree object to a NetworkX graph, uses Graphviz to lay
out the nodes, and displays it using matplotlib. There are a number of
keyword arguments that can modify the resulting diagram, including most
of those accepted by the NetworkX functions <tt class="docutils literal">networkx.draw</tt> and
<tt class="docutils literal">networkx.draw_graphviz</tt>.</p>
<p>The display is also affected by the <tt class="docutils literal">rooted</tt> attribute of the given
tree object. Rooted trees are shown with a “head” on each branch
indicating direction (see Fig. <a class="reference external" href="#fig:phylo-rooted">13.3</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;simple.dnd&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;newick&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">rooted</span> <span class="operator">=</span> <span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_graphiz</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
</pre>
<p><img alt="image9" src="../images/phylo-rooted.png" /></p>
<p>The “prog” argument specifies the Graphviz engine used for layout. The
default, <tt class="docutils literal">twopi</tt>, behaves well for any size tree, reliably avoiding
crossed branches. The <tt class="docutils literal">neato</tt> program may draw more attractive
moderately-sized trees, but sometimes will cross branches (see
Fig. <a class="reference external" href="#fig:phylo-color">13.3</a>). The <tt class="docutils literal">dot</tt> program may be useful
with small trees, but tends to do surprising things with the layout of
larger trees.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_graphviz</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">,</span> <span class="name">prog</span><span class="operator">=</span><span class="literal string">&quot;neato&quot;</span><span class="punctuation">)</span>
</pre>
<p><img alt="image10" src="../images/phylo-color.png" /></p>
<p>This viewing mode is particularly handy for exploring larger trees,
because the matplotlib viewer can zoom in on a selected region, thinning
out a cluttered graphic.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;apaf.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_graphviz</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">,</span> <span class="name">prog</span><span class="operator">=</span><span class="literal string">&quot;neato&quot;</span><span class="punctuation">,</span> <span class="name">node_size</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">)</span>
</pre>
<p><img alt="image11" src="../images/phylo-apaf.png" /> <img alt="image12" src="../images/phylo-apaf-zoom.png" /></p>
<p>Note that branch lengths are not displayed accurately, because Graphviz
ignores them when creating the node layouts. The branch lengths are
retained when exporting a tree as a NetworkX graph object
(<tt class="docutils literal">to_networkx</tt>), however.</p>
<p>See the Phylo page on the Biopython wiki
(<a class="reference external" href="http://biopython.org/wiki/Phylo">http://biopython.org/wiki/Phylo</a>)
for descriptions and examples of the more advanced functionality in
<tt class="docutils literal">draw_ascii</tt>, <tt class="docutils literal">draw_graphviz</tt> and <tt class="docutils literal">to_networkx</tt>.</p>
</div>
<div class="section" id="using-tree-and-clade-objects">
<h2>13.4  Using Tree and Clade objects</h2>
<p>The <tt class="docutils literal">Tree</tt> objects produced by <tt class="docutils literal">parse</tt> and <tt class="docutils literal">read</tt> are containers
for recursive sub-trees, attached to the <tt class="docutils literal">Tree</tt> object at the <tt class="docutils literal">root</tt>
attribute (whether or not the phylogenic tree is actually considered
rooted). A <tt class="docutils literal">Tree</tt> has globally applied information for the phylogeny,
such as rootedness, and a reference to a single <tt class="docutils literal">Clade</tt>; a <tt class="docutils literal">Clade</tt>
has node- and clade-specific information, such as branch length, and a
list of its own descendent <tt class="docutils literal">Clade</tt> instances, attached at the
<tt class="docutils literal">clades</tt> attribute.</p>
<p>So there is a distinction between <tt class="docutils literal">tree</tt> and <tt class="docutils literal">tree.root</tt>. In
practice, though, you rarely need to worry about it. To smooth over the
difference, both <tt class="docutils literal">Tree</tt> and <tt class="docutils literal">Clade</tt> inherit from <tt class="docutils literal">TreeMixin</tt>,
which contains the implementations for methods that would be commonly
used to search, inspect or modify a tree or any of its clades. This
means that almost all of the methods supported by <tt class="docutils literal">tree</tt> are also
available on <tt class="docutils literal">tree.root</tt> and any clade below it. (<tt class="docutils literal">Clade</tt> also has a
<tt class="docutils literal">root</tt> property, which returns the clade object itself.)</p>
<div class="section" id="search-and-traversal-methods">
<h3>13.4.1  Search and traversal methods</h3>
<p>For convenience, we provide a couple of simplified methods that return
all external or internal nodes directly as a list:</p>
<dl class="docutils">
<dt><strong>``get_terminals``</strong></dt>
<dd>makes a list of all of this tree’s terminal (leaf) nodes.</dd>
<dt><strong>``get_nonterminals``</strong></dt>
<dd>makes a list of all of this tree’s nonterminal (internal) nodes.</dd>
</dl>
<p>These both wrap a method with full control over tree traversal,
<tt class="docutils literal">find_clades</tt>. Two more traversal methods, <tt class="docutils literal">find_elements</tt> and
<tt class="docutils literal">find_any</tt>, rely on the same core functionality and accept the same
arguments, which we’ll call a “target specification” for lack of a
better description. These specify which objects in the tree will be
matched and returned during iteration. The first argument can be any of
the following types:</p>
<ul>
<li><p class="first">A <strong>TreeElement instance</strong>, which tree elements will match by
identity — so searching with a Clade instance as the target will find
that clade in the tree;</p>
</li>
<li><p class="first">A <strong>string</strong>, which matches tree elements’ string representation — in
particular, a clade’s <tt class="docutils literal">name</tt> <em>(added in Biopython 1.56)</em>;</p>
</li>
<li><p class="first">A <strong>class</strong> or <strong>type</strong>, where every tree element of the same type
(or sub-type) will be matched;</p>
</li>
<li><p class="first">A <strong>dictionary</strong> where keys are tree element attributes and values
are matched to the corresponding attribute of each tree element. This
one gets even more elaborate:</p>
<ul class="simple">
<li>If an <tt class="docutils literal">int</tt> is given, it matches numerically equal attributes,
e.g. 1 will match 1 or 1.0</li>
<li>If a boolean is given (True or False), the corresponding attribute
value is evaluated as a boolean and checked for the same</li>
<li><tt class="docutils literal">None</tt> matches <tt class="docutils literal">None</tt></li>
<li>If a string is given, the value is treated as a regular expression
(which must match the whole string in the corresponding element
attribute, not just a prefix). A given string without special
regex characters will match string attributes exactly, so if you
don’t use regexes, don’t worry about it. For example, in a tree
with clade names Foo1, Foo2 and Foo3,
<tt class="docutils literal"><span class="pre">tree.find_clades({&quot;name&quot;:</span> <span class="pre">&quot;Foo1&quot;})</span></tt> matches Foo1,
<tt class="docutils literal">{&quot;name&quot;: <span class="pre">&quot;Foo.*&quot;}</span></tt> matches all three clades, and
<tt class="docutils literal">{&quot;name&quot;: &quot;Foo&quot;}</tt> doesn’t match anything.</li>
</ul>
<p>Since floating-point arithmetic can produce some strange behavior, we
don’t support matching <tt class="docutils literal">float</tt>s directly. Instead, use the
boolean <tt class="docutils literal">True</tt> to match every element with a nonzero value in the
specified attribute, then filter on that attribute manually with an
inequality (or exact number, if you like living dangerously).</p>
<p>If the dictionary contains multiple entries, a matching element must
match each of the given attribute values — think “and”, not “or”.</p>
</li>
<li><p class="first">A <strong>function</strong> taking a single argument (it will be applied to each
element in the tree), returning True or False. For convenience,
LookupError, AttributeError and ValueError are silenced, so this
provides another safe way to search for floating-point values in the
tree, or some more complex characteristic.</p>
</li>
</ul>
<p>After the target, there are two optional keyword arguments:</p>
<dl class="docutils">
<dt><strong>terminal</strong></dt>
<dd>— A boolean value to select for or against terminal clades (a.k.a.
leaf nodes): True searches for only terminal clades, False for
non-terminal (internal) clades, and the default, None, searches both
terminal and non-terminal clades, as well as any tree elements
lacking the <tt class="docutils literal">is_terminal</tt> method.</dd>
<dt><strong>order</strong></dt>
<dd>— Tree traversal order: <tt class="docutils literal">&quot;preorder&quot;</tt> (default) is depth-first
search, <tt class="docutils literal">&quot;postorder&quot;</tt> is DFS with child nodes preceding parents,
and <tt class="docutils literal">&quot;level&quot;</tt> is breadth-first search.</dd>
</dl>
<p>Finally, the methods accept arbitrary keyword arguments which are
treated the same way as a dictionary target specification: keys indicate
the name of the element attribute to search for, and the argument value
(string, integer, None or boolean) is compared to the value of each
attribute found. If no keyword arguments are given, then any TreeElement
types are matched. The code for this is generally shorter than passing a
dictionary as the target specification:
<tt class="docutils literal"><span class="pre">tree.find_clades({&quot;name&quot;:</span> <span class="pre">&quot;Foo1&quot;})</span></tt> can be shortened to
<tt class="docutils literal"><span class="pre">tree.find_clades(name=&quot;Foo1&quot;)</span></tt>.</p>
<p>(In Biopython 1.56 or later, this can be even shorter:
<tt class="docutils literal"><span class="pre">tree.find_clades(&quot;Foo1&quot;)</span></tt>)</p>
<p>Now that we’ve mastered target specifications, here are the methods used
to traverse a tree:</p>
<dl class="docutils">
<dt><strong>``find_clades``</strong></dt>
<dd><p class="first">Find each clade containing a matching element. That is, find each
element as with <tt class="docutils literal">find_elements</tt>, but return the corresponding
clade object. (This is usually what you want.)</p>
<p class="last">The result is an iterable through all matching objects, searching
depth-first by default. This is not necessarily the same order as
the elements appear in the Newick, Nexus or XML source file!</p>
</dd>
<dt><strong>``find_elements``</strong></dt>
<dd>Find all tree elements matching the given attributes, and return the
matching elements themselves. Simple Newick trees don’t have complex
sub-elements, so this behaves the same as <tt class="docutils literal">find_clades</tt> on them.
PhyloXML trees often do have complex objects attached to clades, so
this method is useful for extracting those.</dd>
<dt><strong>``find_any``</strong></dt>
<dd>Return the first element found by <tt class="docutils literal">find_elements()</tt>, or None. This
is also useful for checking whether any matching element exists in
the tree, and can be used in a conditional.</dd>
</dl>
<p>Two more methods help navigating between nodes in the tree:</p>
<dl class="docutils">
<dt><strong>``get_path``</strong></dt>
<dd>List the clades directly between the tree root (or current clade)
and the given target. Returns a list of all clade objects along this
path, ending with the given target, but excluding the root clade.</dd>
<dt><strong>``trace``</strong></dt>
<dd>List of all clade object between two targets in this tree. Excluding
start, including finish.</dd>
</dl>
</div>
<div class="section" id="information-methods">
<h3>13.4.2  Information methods</h3>
<p>These methods provide information about the whole tree (or any clade).</p>
<dl class="docutils">
<dt><strong>``common_ancestor``</strong></dt>
<dd>Find the most recent common ancestor of all the given targets. (This
will be a Clade object). If no target is given, returns the root of
the current clade (the one this method is called from); if 1 target
is given, this returns the target itself. However, if any of the
specified targets are not found in the current tree (or clade), an
exception is raised.</dd>
<dt><strong>``count_terminals``</strong></dt>
<dd>Counts the number of terminal (leaf) nodes within the tree.</dd>
<dt><strong>``depths``</strong></dt>
<dd>Create a mapping of tree clades to depths. The result is a
dictionary where the keys are all of the Clade instances in the
tree, and the values are the distance from the root to each clade
(including terminals). By default the distance is the cumulative
branch length leading to the clade, but with the
<tt class="docutils literal">unit_branch_lengths=True</tt> option, only the number of branches
(levels in the tree) is counted.</dd>
<dt><strong>``distance``</strong></dt>
<dd>Calculate the sum of the branch lengths between two targets. If only
one target is specified, the other is the root of this tree.</dd>
<dt><strong>``total_branch_length``</strong></dt>
<dd>Calculate the sum of all the branch lengths in this tree. This is
usually just called the “length” of the tree in phylogenetics, but
we use a more explicit name to avoid confusion with Python
terminology.</dd>
</dl>
<p>The rest of these methods are boolean checks:</p>
<dl class="docutils">
<dt><strong>``is_bifurcating``</strong></dt>
<dd>True if the tree is strictly bifurcating; i.e. all nodes have either
2 or 0 children (internal or external, respectively). The root may
have 3 descendents and still be considered part of a bifurcating
tree.</dd>
<dt><strong>``is_monophyletic``</strong></dt>
<dd>Test if all of the given targets comprise a complete subclade —
i.e., there exists a clade such that its terminals are the same set
as the given targets. The targets should be terminals of the tree.
For convenience, this method returns the common ancestor (MCRA) of
the targets if they are monophyletic (instead of the value
<tt class="docutils literal">True</tt>), and <tt class="docutils literal">False</tt> otherwise.</dd>
<dt><strong>``is_parent_of``</strong></dt>
<dd>True if target is a descendent of this tree — not required to be a
direct descendent. To check direct descendents of a clade, simply
use list membership testing: <tt class="docutils literal">if subclade in clade: ...</tt></dd>
<dt><strong>``is_preterminal``</strong></dt>
<dd>True if all direct descendents are terminal; False if any direct
descendent is not terminal.</dd>
</dl>
</div>
<div class="section" id="modification-methods">
<h3>13.4.3  Modification methods</h3>
<p>These methods modify the tree in-place. If you want to keep the original
tree intact, make a complete copy of the tree first, using Python’s
<tt class="docutils literal">copy</tt> module:</p>
<pre class="code python literal-block">
<span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'example.xml'</span><span class="punctuation">,</span> <span class="literal string">'phyloxml'</span><span class="punctuation">)</span>
<span class="keyword namespace">import</span> <span class="name namespace">copy</span>
<span class="name">newtree</span> <span class="operator">=</span> <span class="name">copy</span><span class="operator">.</span><span class="name">deepcopy</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
</pre>
<dl class="docutils">
<dt><strong>``collapse``</strong></dt>
<dd>Deletes the target from the tree, relinking its children to its
parent.</dd>
<dt><strong>``collapse_all``</strong></dt>
<dd>Collapse all the descendents of this tree, leaving only terminals.
Branch lengths are preserved, i.e. the distance to each terminal
stays the same. With a target specification (see above), collapses
only the internal nodes matching the specification.</dd>
<dt><strong>``ladderize``</strong></dt>
<dd>Sort clades in-place according to the number of terminal nodes.
Deepest clades are placed last by default. Use <tt class="docutils literal">reverse=True</tt> to
sort clades deepest-to-shallowest.</dd>
<dt><strong>``prune``</strong></dt>
<dd>Prunes a terminal clade from the tree. If taxon is from a
bifurcation, the connecting node will be collapsed and its branch
length added to remaining terminal node. This might no longer be a
meaningful value.</dd>
<dt><strong>``root_with_outgroup``</strong></dt>
<dd><p class="first">Reroot this tree with the outgroup clade containing the given
targets, i.e. the common ancestor of the outgroup. This method is
only available on Tree objects, not Clades.</p>
<p>If the outgroup is identical to self.root, no change occurs. If the
outgroup clade is terminal (e.g. a single terminal node is given as
the outgroup), a new bifurcating root clade is created with a
0-length branch to the given outgroup. Otherwise, the internal node
at the base of the outgroup becomes a trifurcating root for the
whole tree. If the original root was bifurcating, it is dropped from
the tree.</p>
<p class="last">In all cases, the total branch length of the tree stays the same.</p>
</dd>
<dt><strong>``root_at_midpoint``</strong></dt>
<dd>Reroot this tree at the calculated midpoint between the two most
distant tips of the tree. (This uses <tt class="docutils literal">root_with_outgroup</tt> under
the hood.)</dd>
<dt><strong>``split``</strong></dt>
<dd>Generate <em>n</em> (default 2) new descendants. In a species tree, this is
a speciation event. New clades have the given <tt class="docutils literal">branch_length</tt> and
the same name as this clade’s root plus an integer suffix (counting
from 0) — for example, splitting a clade named “A” produces the
sub-clades “A0” and “A1”.</dd>
</dl>
<p>See the Phylo page on the Biopython wiki
(<a class="reference external" href="http://biopython.org/wiki/Phylo">http://biopython.org/wiki/Phylo</a>)
for more examples of using the available methods.</p>
</div>
<div class="section" id="features-of-phyloxml-trees">
<h3>13.4.4  Features of PhyloXML trees</h3>
<p>The phyloXML file format includes fields for annotating trees with
additional data types and visual cues.</p>
<p>See the PhyloXML page on the Biopython wiki
(<a class="reference external" href="http://biopython.org/wiki/PhyloXML">http://biopython.org/wiki/PhyloXML</a>)
for descriptions and examples of using the additional annotation
features provided by PhyloXML.</p>
</div>
</div>
<div class="section" id="running-external-applications">
<h2>13.5  Running external applications</h2>
<p>While Bio.Phylo doesn’t infer trees from alignments itself, there are
third-party programs available that do. These are supported through the
module <tt class="docutils literal">Bio.Phylo.Applications</tt>, using the same general framework as
<tt class="docutils literal">Bio.Emboss.Applications</tt>, <tt class="docutils literal">Bio.Align.Applications</tt> and others.</p>
<p>Biopython 1.58 introduced a wrapper for PhyML
(<a class="reference external" href="http://www.atgc-montpellier.fr/phyml/">http://www.atgc-montpellier.fr/phyml/</a>).
The program accepts an input alignment in <tt class="docutils literal"><span class="pre">phylip-relaxed</span></tt> format
(that’s Phylip format, but without the 10-character limit on taxon
names) and a variety of options. A quick example:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Phylo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Phylo.Applications</span> <span class="keyword namespace">import</span> <span class="name">PhymlCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cmd</span> <span class="operator">=</span> <span class="name">PhymlCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">'Tests/Phylip/random.phy'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">out_log</span><span class="punctuation">,</span> <span class="name">err_log</span> <span class="operator">=</span> <span class="name">cmd</span><span class="punctuation">()</span>
</pre>
<p>This generates a tree file and a stats file with the names
[<em>input filename</em>]<tt class="docutils literal">_phyml_tree.txt</tt> and
[<em>input filename</em>]<tt class="docutils literal">_phyml_stats.txt</tt>. The tree file is in Newick
format:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'Tests/Phylip/random.phy_phyml_tree.txt'</span><span class="punctuation">,</span> <span class="literal string">'newick'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_ascii</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
</pre>
<p>A similar wrapper for RAxML
(<a class="reference external" href="http://sco.h-its.org/exelixis/software.html">http://sco.h-its.org/exelixis/software.html</a>)
was added in Biopython 1.60.</p>
<p>Note that some popular Phylip programs, including <tt class="docutils literal">dnaml</tt> and
<tt class="docutils literal">protml</tt>, are already available through the EMBOSS wrappers in
<tt class="docutils literal">Bio.Emboss.Applications</tt> if you have the Phylip extensions to EMBOSS
installed on your system. See Section <a class="reference external" href="#sec:alignment-tools">6.4</a>
for some examples and clues on how to use programs like these.</p>
</div>
<div class="section" id="paml-integration">
<h2>13.6  PAML integration</h2>
<p>Biopython 1.58 brought support for PAML
(<a class="reference external" href="http://abacus.gene.ucl.ac.uk/software/paml.html">http://abacus.gene.ucl.ac.uk/software/paml.html</a>),
a suite of programs for phylogenetic analysis by maximum likelihood.
Currently the programs codeml, baseml and yn00 are implemented. Due to
PAML’s usage of control files rather than command line arguments to
control runtime options, usage of this wrapper strays from the format of
other application wrappers in Biopython.</p>
<p>A typical workflow would be to initialize a PAML object, specifying an
alignment file, a tree file, an output file and a working directory.
Next, runtime options are set via the <tt class="docutils literal">set_options()</tt> method or by
reading an existing control file. Finally, the program is run via the
<tt class="docutils literal">run()</tt> method and the output file is automatically parsed to a
results dictionary.</p>
<p>Here is an example of typical usage of codeml:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Phylo.PAML</span> <span class="keyword namespace">import</span> <span class="name">codeml</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cml</span> <span class="operator">=</span> <span class="name">codeml</span><span class="operator">.</span><span class="name">Codeml</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cml</span><span class="operator">.</span><span class="name">alignment</span> <span class="operator">=</span> <span class="literal string">&quot;Tests/PAML/alignment.phylip&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cml</span><span class="operator">.</span><span class="name">tree</span> <span class="operator">=</span> <span class="literal string">&quot;Tests/PAML/species.tree&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cml</span><span class="operator">.</span><span class="name">out_file</span> <span class="operator">=</span> <span class="literal string">&quot;results.out&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cml</span><span class="operator">.</span><span class="name">working_dir</span> <span class="operator">=</span> <span class="literal string">&quot;./scratch&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cml</span><span class="operator">.</span><span class="name">set_options</span><span class="punctuation">(</span><span class="name">seqtype</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">verbose</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">noisy</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">RateAncestor</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">model</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">NSsites</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">],</span>
<span class="operator">...</span>         <span class="name">CodonFreq</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">cleandata</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">fix_alpha</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">kappa</span><span class="operator">=</span><span class="literal number float">4.54006</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">results</span> <span class="operator">=</span> <span class="name">cml</span><span class="operator">.</span><span class="name">run</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ns_sites</span> <span class="operator">=</span> <span class="name">results</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;NSsites&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m0</span> <span class="operator">=</span> <span class="name">ns_sites</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m0_params</span> <span class="operator">=</span> <span class="name">m0</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;parameters&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m0_params</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;omega&quot;</span><span class="punctuation">)</span>
</pre>
<p>Existing output files may be parsed as well using a module’s <tt class="docutils literal">read()</tt>
function:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">results</span> <span class="operator">=</span> <span class="name">codeml</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;Tests/PAML/Results/codeml/codeml_NSsites_all.out&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">results</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;lnL max&quot;</span><span class="punctuation">)</span>
</pre>
<p>Detailed documentation for this new module currently lives on the
Biopython wiki:
<a class="reference external" href="http://biopython.org/wiki/PAML">http://biopython.org/wiki/PAML</a></p>
</div>
<div class="section" id="future-plans">
<h2>13.7  Future plans</h2>
<p>Bio.Phylo is under active development. Here are some features we might
add in future releases:</p>
<dl class="docutils">
<dt><strong>New methods</strong></dt>
<dd><p class="first">Generally useful functions for operating on Tree or Clade objects
appear on the Biopython wiki first, so that casual users can test
them and decide if they’re useful before we add them to Bio.Phylo:</p>
<p class="last"><a class="reference external" href="http://biopython.org/wiki/Phylo_cookbook">http://biopython.org/wiki/Phylo_cookbook</a></p>
</dd>
<dt><strong>Bio.Nexus port</strong></dt>
<dd><p class="first">Much of this module was written during Google Summer of Code 2009,
under the auspices of NESCent, as a project to implement Python
support for the phyloXML data format (see
<a class="reference external" href="#sec:PhyloXML">13.4.4</a>). Support for Newick and Nexus formats
was added by porting part of the existing Bio.Nexus module to the
new classes used by Bio.Phylo.</p>
<p class="last">Currently, Bio.Nexus contains some useful features that have not yet
been ported to Bio.Phylo classes — notably, calculating a consensus
tree. If you find some functionality lacking in Bio.Phylo, try
poking throught Bio.Nexus to see if it’s there instead.</p>
</dd>
</dl>
<p>We’re open to any suggestions for improving the functionality and
usability of this module; just let us know on the mailing list or our
bug database.</p>
</div>
</div>
<div class="section" id="chapter-14-sequence-motif-analysis-using-bio-motifs">
<h1>Chapter 14  Sequence motif analysis using Bio.motifs</h1>
<p>This chapter gives an overview of the functionality of the
<tt class="docutils literal">Bio.motifs</tt> package included in Biopython. It is intended for people
who are involved in the analysis of sequence motifs, so I’ll assume that
you are familiar with basic notions of motif analysis. In case something
is unclear, please look at Section <a class="reference external" href="#sec:links">14.8</a> for some
relevant links.</p>
<p>Most of this chapter describes the new <tt class="docutils literal">Bio.motifs</tt> package included
in Biopython 1.61 onwards, which is replacing the older <tt class="docutils literal">Bio.Motif</tt>
package introduced with Biopython 1.50, which was in turn based on two
older former Biopython modules, <tt class="docutils literal">Bio.AlignAce</tt> and <tt class="docutils literal">Bio.MEME</tt>. It
provides most of their functionality with a unified motif object
implementation.</p>
<p>Speaking of other libraries, if you are reading this you might be
interested in <a class="reference external" href="http://fraenkel.mit.edu/TAMO/">TAMO</a>, another python
library designed to deal with sequence motifs. It supports more
<em>de-novo</em> motif finders, but it is not a part of Biopython and has some
restrictions on commercial use.</p>
<div class="section" id="motif-objects">
<h2>14.1  Motif objects</h2>
<p>Since we are interested in motif analysis, we need to take a look at
<tt class="docutils literal">Motif</tt> objects in the first place. For that we need to import the
Bio.motifs library:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">motifs</span>
</pre>
<p>and we can start creating our first motif objects. We can either create
a <tt class="docutils literal">Motif</tt> object from a list of instances of the motif, or we can
obtain a <tt class="docutils literal">Motif</tt> object by parsing a file from a motif database or
motif finding software.</p>
<div class="section" id="creating-a-motif-from-instances">
<h3>14.1.1  Creating a motif from instances</h3>
<p>Suppose we have these instances of a DNA motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">instances</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TACAA&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>              <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TACGC&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>              <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TACAC&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>              <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TACCC&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>              <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AACCC&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>              <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AATGC&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>              <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AATGC&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>             <span class="punctuation">]</span>
</pre>
<p>then we can create a Motif object as follows:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">create</span><span class="punctuation">(</span><span class="name">instances</span><span class="punctuation">)</span>
</pre>
<p>The instances are saved in an attribute <tt class="docutils literal">m.instances</tt>, which is
essentially a Python list with some added functionality, as described
below. Printing out the Motif object shows the instances from which it
was constructed:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m</span>
<span class="name">TACAA</span>
<span class="name">TACGC</span>
<span class="name">TACAC</span>
<span class="name">TACCC</span>
<span class="name">AACCC</span>
<span class="name">AATGC</span>
<span class="name">AATGC</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>The length of the motif defined as the sequence length, which should be
the same for all instances:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">m</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
</pre>
<p>The Motif object has an attribute <tt class="docutils literal">.counts</tt> containing the counts of
each nucleotide at each position. Printing this counts matrix shows it
in an easily readable format:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">3.00</span>   <span class="literal number float">7.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">2.00</span>   <span class="literal number float">1.00</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">5.00</span>   <span class="literal number float">2.00</span>   <span class="literal number float">6.00</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">3.00</span>   <span class="literal number float">0.00</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">4.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">2.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>You can access these counts as a dictionary:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[</span><span class="literal string">'A'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">7</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">]</span>
</pre>
<p>but you can also think of it as a 2D array with the nucleotide as the
first dimension and the position as the second dimension:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[</span><span class="literal string">'T'</span><span class="punctuation">,</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="literal number integer">4</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[</span><span class="literal string">'T'</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">]</span>
<span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[</span><span class="literal string">'T'</span><span class="punctuation">,</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="literal number integer">0</span>
</pre>
<p>You can also directly access columns of the counts matrix</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[:,</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number integer">3</span><span class="punctuation">}</span>
</pre>
<p>Instead of the nucleotide itself, you can also use the index of the
nucleotide in the sorted letters in the alphabet of the motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span>
<span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="operator">.</span><span class="name">letters</span>
<span class="literal string">'GATC'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">sorted</span><span class="punctuation">(</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="operator">.</span><span class="name">letters</span><span class="punctuation">)</span>
<span class="punctuation">[</span><span class="literal string">'A'</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[</span><span class="literal string">'A'</span><span class="punctuation">,:]</span>
<span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">7</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">,:]</span>
<span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">7</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>The motif has an associated consensus sequence, defined as the sequence
of letters along the positions of the motif for which the largest value
in the corresponding columns of the <tt class="docutils literal">.counts</tt> matrix is obtained:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TACGC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>as well as an anticonsensus sequence, corresponding to the smallest
values in the columns of the <tt class="docutils literal">.counts</tt> matrix:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">anticonsensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GGGTG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>You can also ask for a degenerate consensus sequence, in which ambiguous
nucleotides are used for positions where there are multiple nucleotides
with high counts:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">degenerate_consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'WACVC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>Here, W and R follow the IUPAC nucleotide ambiguity codes: W is either A
or T, and V is A, C, or G [<a class="reference external" href="#cornish1985">10</a>]. The degenerate
consensus sequence is constructed following the rules specified by
Cavener [<a class="reference external" href="#cavener1987">11</a>].</p>
<p>We can also get the reverse complement of a motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">r</span> <span class="operator">=</span> <span class="name">m</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">r</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GCGTA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">r</span><span class="operator">.</span><span class="name">degenerate_consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GBGTW'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">r</span>
<span class="name">TTGTA</span>
<span class="name">GCGTA</span>
<span class="name">GTGTA</span>
<span class="name">GGGTA</span>
<span class="name">GGGTT</span>
<span class="name">GCATT</span>
<span class="name">GCATT</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>The reverse complement and the degenerate consensus sequence are only
defined for DNA motifs.</p>
</div>
<div class="section" id="reading-motifs">
<h3>14.1.2  Reading motifs</h3>
<p>Creating motifs from instances by hand is a bit boring, so it’s useful
to have some I/O functions for reading and writing motifs. There are no
really well established standards for storing motifs, but there’s a
couple of formats which are more used than others. The most important
distinction is whether the motif representation is based on instances or
on some version of PWM matrix.</p>
<div class="section" id="jaspar">
<h4>JASPAR</h4>
<p>One of the most popular motif databases
<a class="reference external" href="http://jaspar.genereg.net">JASPAR</a> stores motifs either as a list of
instances, or as a frequency matrix. As an example, these are the
beginning and ending lines of the JASPAR <tt class="docutils literal">Arnt.sites</tt> file showing
known binding sites of the mouse helix-loop-helix transcription factor
Arnt:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">MA0004</span> <span class="name">ARNT</span>    <span class="literal number integer">1</span>
<span class="name">CACGTGatgtcctc</span>
<span class="operator">&gt;</span><span class="name">MA0004</span> <span class="name">ARNT</span>    <span class="literal number integer">2</span>
<span class="name">CACGTGggaggtac</span>
<span class="operator">&gt;</span><span class="name">MA0004</span> <span class="name">ARNT</span>    <span class="literal number integer">3</span>
<span class="name">CACGTGccgcgcgc</span>
<span class="operator">...</span>
<span class="operator">&gt;</span><span class="name">MA0004</span> <span class="name">ARNT</span>    <span class="literal number integer">18</span>
<span class="name">AACGTGacagccctcc</span>
<span class="operator">&gt;</span><span class="name">MA0004</span> <span class="name">ARNT</span>    <span class="literal number integer">19</span>
<span class="name">AACGTGcacatcgtcc</span>
<span class="operator">&gt;</span><span class="name">MA0004</span> <span class="name">ARNT</span>    <span class="literal number integer">20</span>
<span class="name">aggaatCGCGTGc</span>
</pre>
<p>The parts of the sequence in capital letters are the motif instances
that were found to align to each other.</p>
<p>We can create a <tt class="docutils literal">Motif</tt> object from these instances as follows:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">motifs</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">arnt</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;Arnt.sites&quot;</span><span class="punctuation">),</span> <span class="literal string">&quot;sites&quot;</span><span class="punctuation">)</span>
</pre>
<p>The instances from which this motif was created is stored in the
<tt class="docutils literal">.instances</tt> property:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[:</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CACGTG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()),</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CACGTG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()),</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CACGTG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">instance</span> <span class="operator word">in</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">instance</span>
<span class="operator">...</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">AACGTG</span>
<span class="name">AACGTG</span>
<span class="name">AACGTG</span>
<span class="name">AACGTG</span>
<span class="name">CGCGTG</span>
</pre>
<p>The counts matrix of this motif is automatically calculated from the
instances:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">counts</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">4.00</span>  <span class="literal number float">19.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="name">C</span><span class="punctuation">:</span>  <span class="literal number float">16.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>   <span class="literal number float">0.00</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>The JASPAR database also makes motifs available directly as a count
matrix, without the instances from which it was created. For example,
this is the JASPAR file <tt class="docutils literal">SRF.pfm</tt> containing the count matrix for the
human SRF transcription factor:</p>
<pre class="code python literal-block">
 <span class="literal number integer">2</span>  <span class="literal number integer">9</span>  <span class="literal number integer">0</span>  <span class="literal number integer">1</span> <span class="literal number integer">32</span>  <span class="literal number integer">3</span> <span class="literal number integer">46</span>  <span class="literal number integer">1</span> <span class="literal number integer">43</span> <span class="literal number integer">15</span>  <span class="literal number integer">2</span>  <span class="literal number integer">2</span>
 <span class="literal number integer">1</span> <span class="literal number integer">33</span> <span class="literal number integer">45</span> <span class="literal number integer">45</span>  <span class="literal number integer">1</span>  <span class="literal number integer">1</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">1</span>  <span class="literal number integer">0</span>  <span class="literal number integer">1</span>
<span class="literal number integer">39</span>  <span class="literal number integer">2</span>  <span class="literal number integer">1</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span> <span class="literal number integer">44</span> <span class="literal number integer">43</span>
 <span class="literal number integer">4</span>  <span class="literal number integer">2</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span> <span class="literal number integer">13</span> <span class="literal number integer">42</span>  <span class="literal number integer">0</span> <span class="literal number integer">45</span>  <span class="literal number integer">3</span> <span class="literal number integer">30</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>
</pre>
<p>We can create a motif for this count matrix as follows:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">srf</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;SRF.pfm&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;pfm&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">srf</span><span class="operator">.</span><span class="name">counts</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>      <span class="literal number integer">6</span>      <span class="literal number integer">7</span>      <span class="literal number integer">8</span>      <span class="literal number integer">9</span>     <span class="literal number integer">10</span>     <span class="literal number integer">11</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">2.00</span>   <span class="literal number float">9.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>  <span class="literal number float">32.00</span>   <span class="literal number float">3.00</span>  <span class="literal number float">46.00</span>   <span class="literal number float">1.00</span>  <span class="literal number float">43.00</span>  <span class="literal number float">15.00</span>   <span class="literal number float">2.00</span>   <span class="literal number float">2.00</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">1.00</span>  <span class="literal number float">33.00</span>  <span class="literal number float">45.00</span>  <span class="literal number float">45.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>
<span class="name">G</span><span class="punctuation">:</span>  <span class="literal number float">39.00</span>   <span class="literal number float">2.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">44.00</span>  <span class="literal number float">43.00</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">4.00</span>   <span class="literal number float">2.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">13.00</span>  <span class="literal number float">42.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">45.00</span>   <span class="literal number float">3.00</span>  <span class="literal number float">30.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>As this motif was created from the counts matrix directly, it has no
instances associated with it:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">srf</span><span class="operator">.</span><span class="name">instances</span>
<span class="name builtin pseudo">None</span>
</pre>
<p>We can now ask for the consensus sequence of these two motifs:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">counts</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">CACGTG</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">srf</span><span class="operator">.</span><span class="name">counts</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">GCCCATATATGG</span>
</pre>
</div>
<div class="section" id="meme">
<h4>MEME</h4>
<p>MEME [<a class="reference external" href="#bailey1994">12</a>] is a tool for discovering motifs in a
group of related DNA or protein sequences. It takes as input a group of
DNA or protein sequences and outputs as many motifs as requested.
Therefore, in contrast to JASPAR files, MEME output files typically
contain multiple motifs. This is an example.</p>
<p>At the top of an output file generated by MEME shows some background
information about the MEME and the version of MEME used:</p>
<pre class="code python literal-block">
<span class="operator">********************************************************************************</span>
<span class="name">MEME</span> <span class="operator">-</span> <span class="name">Motif</span> <span class="name">discovery</span> <span class="name">tool</span>
<span class="operator">********************************************************************************</span>
<span class="name">MEME</span> <span class="name">version</span> <span class="literal number float">3.0</span> <span class="punctuation">(</span><span class="name">Release</span> <span class="name">date</span><span class="punctuation">:</span> <span class="literal number integer">2004</span><span class="operator">/</span><span class="literal number integer">08</span><span class="operator">/</span><span class="literal number integer">18</span> <span class="literal number integer">09</span><span class="punctuation">:</span><span class="literal number oct">07</span><span class="punctuation">:</span><span class="literal number oct">01</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>Further down, the input set of training sequences is recapitulated:</p>
<pre class="code python literal-block">
<span class="operator">********************************************************************************</span>
<span class="name">TRAINING</span> <span class="name">SET</span>
<span class="operator">********************************************************************************</span>
<span class="name">DATAFILE</span><span class="operator">=</span> <span class="name">INO_up800</span><span class="operator">.</span><span class="name">s</span>
<span class="name">ALPHABET</span><span class="operator">=</span> <span class="name">ACGT</span>
<span class="name">Sequence</span> <span class="name">name</span>            <span class="name">Weight</span> <span class="name">Length</span>  <span class="name">Sequence</span> <span class="name">name</span>            <span class="name">Weight</span> <span class="name">Length</span>
<span class="operator">-------------</span>            <span class="operator">------</span> <span class="operator">------</span>  <span class="operator">-------------</span>            <span class="operator">------</span> <span class="operator">------</span>
<span class="name">CHO1</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>  <span class="name">CHO2</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>
<span class="name">FAS1</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>  <span class="name">FAS2</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>
<span class="name">ACC1</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>  <span class="name">INO1</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>
<span class="name">OPI3</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>
<span class="operator">********************************************************************************</span>
</pre>
<p>and the exact command line that was used:</p>
<pre class="code python literal-block">
<span class="operator">********************************************************************************</span>
<span class="name">COMMAND</span> <span class="name">LINE</span> <span class="name">SUMMARY</span>
<span class="operator">********************************************************************************</span>
<span class="name">This</span> <span class="name">information</span> <span class="name">can</span> <span class="name">also</span> <span class="name">be</span> <span class="name">useful</span> <span class="operator word">in</span> <span class="name">the</span> <span class="name">event</span> <span class="name">you</span> <span class="name">wish</span> <span class="name">to</span> <span class="name">report</span> <span class="name">a</span>
<span class="name">problem</span> <span class="keyword">with</span> <span class="name">the</span> <span class="name">MEME</span> <span class="name">software</span><span class="operator">.</span>

<span class="name">command</span><span class="punctuation">:</span> <span class="name">meme</span> <span class="operator">-</span><span class="name">mod</span> <span class="name">oops</span> <span class="operator">-</span><span class="name">dna</span> <span class="operator">-</span><span class="name">revcomp</span> <span class="operator">-</span><span class="name">nmotifs</span> <span class="literal number integer">2</span> <span class="operator">-</span><span class="name">bfile</span> <span class="name">yeast</span><span class="operator">.</span><span class="name">nc</span><span class="operator">.</span><span class="literal number float">6.</span><span class="name">freq</span> <span class="name">INO_up800</span><span class="operator">.</span><span class="name">s</span>
<span class="operator">...</span>
</pre>
<p>Next is detailed information on each motif that was found:</p>
<pre class="code python literal-block">
<span class="operator">********************************************************************************</span>
<span class="name">MOTIF</span>  <span class="literal number integer">1</span>        <span class="name">width</span> <span class="operator">=</span>   <span class="literal number integer">12</span>   <span class="name">sites</span> <span class="operator">=</span>   <span class="literal number integer">7</span>   <span class="name">llr</span> <span class="operator">=</span> <span class="literal number integer">95</span>   <span class="name">E</span><span class="operator">-</span><span class="name">value</span> <span class="operator">=</span> <span class="literal number float">2.0e-001</span>
<span class="operator">********************************************************************************</span>
<span class="operator">--------------------------------------------------------------------------------</span>
        <span class="name">Motif</span> <span class="literal number integer">1</span> <span class="name">Description</span>
<span class="operator">--------------------------------------------------------------------------------</span>
<span class="name">Simplified</span>        <span class="name">A</span>  <span class="punctuation">:::</span><span class="literal number integer">9</span><span class="punctuation">:</span><span class="name">a</span><span class="punctuation">::::</span><span class="literal number integer">3</span><span class="punctuation">:</span>
<span class="name">pos</span><span class="operator">.-</span><span class="name">specific</span>     <span class="name">C</span>  <span class="punctuation">::</span><span class="name">a</span><span class="punctuation">:</span><span class="literal number integer">9</span><span class="punctuation">:</span><span class="literal number integer">11691</span><span class="name">a</span>
<span class="name">probability</span>       <span class="name">G</span>  <span class="punctuation">::::</span><span class="literal number integer">1</span><span class="punctuation">::</span><span class="literal number integer">94</span><span class="punctuation">:</span><span class="literal number integer">4</span><span class="punctuation">:</span>
<span class="name">matrix</span>            <span class="name">T</span>  <span class="name">aa</span><span class="punctuation">:</span><span class="literal number integer">1</span><span class="punctuation">::</span><span class="literal number integer">9</span><span class="punctuation">::</span><span class="literal number integer">11</span><span class="punctuation">:</span>
</pre>
<p>To parse this file (stored as <tt class="docutils literal">meme.dna.oops.txt</tt>), use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;meme.dna.oops.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;meme&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>The <tt class="docutils literal">motifs.parse</tt> command reads the complete file directly, so you
can close the file after calling <tt class="docutils literal">motifs.parse</tt>. The header
information is stored in attributes:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">version</span>
<span class="literal string">'3.0'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">datafile</span>
<span class="literal string">'INO_up800.s'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">command</span>
<span class="literal string">'meme -mod oops -dna -revcomp -nmotifs 2 -bfile yeast.nc.6.freq INO_up800.s'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">alphabet</span>
<span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">sequences</span>
<span class="punctuation">[</span><span class="literal string">'CHO1'</span><span class="punctuation">,</span> <span class="literal string">'CHO2'</span><span class="punctuation">,</span> <span class="literal string">'FAS1'</span><span class="punctuation">,</span> <span class="literal string">'FAS2'</span><span class="punctuation">,</span> <span class="literal string">'ACC1'</span><span class="punctuation">,</span> <span class="literal string">'INO1'</span><span class="punctuation">,</span> <span class="literal string">'OPI3'</span><span class="punctuation">]</span>
</pre>
<p>The record is an object of the <tt class="docutils literal">Bio.motifs.meme.Record</tt> class. The
class inherits from list, and you can think of <tt class="docutils literal">record</tt> as a list of
Motif objects:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">TTCACATGCCGC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">degenerate_consensus</span>
<span class="name">TTCACATGSCNC</span>
</pre>
<p>In addition to these generic motif attributes, each motif also stores
its specific information as calculated by MEME. For example,</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">num_occurrences</span>
<span class="literal number integer">7</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">length</span>
<span class="literal number integer">12</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">evalue</span> <span class="operator">=</span> <span class="name">motif</span><span class="operator">.</span><span class="name">evalue</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%3.1g</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">evalue</span>
<span class="literal number float">0.2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'Motif 1'</span>
</pre>
<p>In addition to using an index into the record, as we did above, you can
also find it by its name:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">'Motif 1'</span><span class="punctuation">]</span>
</pre>
<p>Each motif has an attribute <tt class="docutils literal">.instances</tt> with the sequence instances
in which the motif was found, providing some information on each
instance:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">)</span>
<span class="literal number integer">7</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="name">Instance</span><span class="punctuation">(</span><span class="literal string">'TTCACATGCCGC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">motif_name</span>
<span class="literal string">'Motif 1'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">sequence_name</span>
<span class="literal string">'INO1'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">start</span>
<span class="literal number integer">620</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">strand</span>
<span class="literal string">'-'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">length</span>
<span class="literal number integer">12</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pvalue</span> <span class="operator">=</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">pvalue</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%5.3g</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">pvalue</span>
<span class="literal number float">1.85e-08</span>
</pre>
</div>
<div class="section" id="mast">
<h4>MAST</h4>
</div>
<div class="section" id="transfac">
<h4>TRANSFAC</h4>
<p>TRANSFAC is a manually curated database of transcription factors,
together with their genomic binding sites and DNA binding profiles
[<a class="reference external" href="#matys2003">27</a>]. While the file format used in the TRANSFAC
database is nowadays also used by others, we will refer to it as the
TRANSFAC file format.</p>
<p>A minimal file in the TRANSFAC format looks as follows:</p>
<pre class="code python literal-block">
<span class="name">ID</span>  <span class="name">motif1</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">S</span>
<span class="literal number oct">02</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">R</span>
<span class="literal number oct">03</span>      <span class="literal number integer">3</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">1</span>      <span class="name">A</span>
<span class="literal number oct">04</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">05</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">A</span>
<span class="literal number oct">06</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">4</span>      <span class="literal number integer">1</span>      <span class="name">G</span>
<span class="literal number oct">07</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">4</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">08</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="name">T</span>
<span class="literal number integer">09</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">10</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="name">K</span>
<span class="literal number integer">11</span>      <span class="literal number integer">0</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="name">Y</span>
<span class="literal number integer">12</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="literal number integer">1</span>      <span class="name">G</span>
<span class="operator">//</span>
</pre>
<p>This file shows the frequency matrix of motif <tt class="docutils literal">motif1</tt> of 12
nucleotides. In general, one file in the TRANSFAC format can contain
multiple motifs. For example, this is the contents of the example
TRANSFAC file <tt class="docutils literal">transfac.dat</tt>:</p>
<pre class="code python literal-block">
<span class="name">VV</span>  <span class="name">EXAMPLE</span> <span class="name">January</span> <span class="literal number integer">15</span><span class="punctuation">,</span> <span class="literal number integer">2013</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="name">ID</span>  <span class="name">motif1</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">S</span>
<span class="literal number oct">02</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">R</span>
<span class="literal number oct">03</span>      <span class="literal number integer">3</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">1</span>      <span class="name">A</span>
<span class="operator">...</span>
<span class="literal number integer">11</span>      <span class="literal number integer">0</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="name">Y</span>
<span class="literal number integer">12</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="literal number integer">1</span>      <span class="name">G</span>
<span class="operator">//</span>
<span class="name">ID</span>  <span class="name">motif2</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">R</span>
<span class="literal number oct">02</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">S</span>
<span class="operator">...</span>
<span class="literal number integer">09</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="name">T</span>
<span class="literal number integer">10</span>      <span class="literal number integer">0</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="name">Y</span>
<span class="operator">//</span>
</pre>
<p>To parse a TRANSFAC file, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;transfac.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;TRANSFAC&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>The overall version number, if available, is stored as
<tt class="docutils literal">record.version</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">version</span>
<span class="literal string">'EXAMPLE January 15, 2013'</span>
</pre>
<p>Each motif in <tt class="docutils literal">record</tt> is in instance of the
<tt class="docutils literal">Bio.motifs.transfac.Motif</tt> class, which inherits both from the
<tt class="docutils literal">Bio.motifs.Motif</tt> class and from a Python dictionary. The dictionary
uses the two-letter keys to store any additional information about the
motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">degenerate_consensus</span> <span class="comment"># Using the Bio.motifs.Motif method</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'SRACAGGTGKYG'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="punctuation">[</span><span class="literal string">'ID'</span><span class="punctuation">]</span> <span class="comment"># Using motif as a dictionary</span>
<span class="literal string">'motif1'</span>
</pre>
<p>TRANSFAC files are typically much more elaborate than this example,
containing lots of additional information about the motif. Table
<a class="reference external" href="#table:transfaccodes">14.1.2</a> lists the two-letter field codes that
are commonly found in TRANSFAC files:</p>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr><td>Table 14.1: Fields commonly found in TRANSFAC files</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="16%" />
<col width="84%" />
</colgroup>
<tbody valign="top">
<tr><td><tt class="docutils literal">AC</tt></td>
<td>Accession number</td>
</tr>
<tr><td><tt class="docutils literal">AS</tt></td>
<td>Accession numbers, secondary</td>
</tr>
<tr><td><tt class="docutils literal">BA</tt></td>
<td>Statistical basis</td>
</tr>
<tr><td><tt class="docutils literal">BF</tt></td>
<td>Binding factors</td>
</tr>
<tr><td><tt class="docutils literal">BS</tt></td>
<td>Factor binding sites underlying the matrix</td>
</tr>
<tr><td><tt class="docutils literal">CC</tt></td>
<td>Comments</td>
</tr>
<tr><td><tt class="docutils literal">CO</tt></td>
<td>Copyright notice</td>
</tr>
<tr><td><tt class="docutils literal">DE</tt></td>
<td>Short factor description</td>
</tr>
<tr><td><tt class="docutils literal">DR</tt></td>
<td>External databases</td>
</tr>
<tr><td><tt class="docutils literal">DT</tt></td>
<td>Date created/updated</td>
</tr>
<tr><td><tt class="docutils literal">HC</tt></td>
<td>Subfamilies</td>
</tr>
<tr><td><tt class="docutils literal">HP</tt></td>
<td>Superfamilies</td>
</tr>
<tr><td><tt class="docutils literal">ID</tt></td>
<td>Identifier</td>
</tr>
<tr><td><tt class="docutils literal">NA</tt></td>
<td>Name of the binding factor</td>
</tr>
<tr><td><tt class="docutils literal">OC</tt></td>
<td>Taxonomic classification</td>
</tr>
<tr><td><tt class="docutils literal">OS</tt></td>
<td>Species/Taxon</td>
</tr>
<tr><td><tt class="docutils literal">OV</tt></td>
<td>Older version</td>
</tr>
<tr><td><tt class="docutils literal">PV</tt></td>
<td>Preferred version</td>
</tr>
<tr><td><tt class="docutils literal">TY</tt></td>
<td>Type</td>
</tr>
<tr><td><tt class="docutils literal">XX</tt></td>
<td>Empty line; these are not stored in the Record.</td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<p>Each motif also has an attribute <tt class="docutils literal">.references</tt> containing the
references associated with the motif, using these two-letter keys:</p>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr><td>Table 14.2: Fields used to store references in TRANSFAC files</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="32%" />
<col width="68%" />
</colgroup>
<tbody valign="top">
<tr><td><tt class="docutils literal">RN</tt></td>
<td>Reference number</td>
</tr>
<tr><td><tt class="docutils literal">RA</tt></td>
<td>Reference authors</td>
</tr>
<tr><td><tt class="docutils literal">RL</tt></td>
<td>Reference data</td>
</tr>
<tr><td><tt class="docutils literal">RT</tt></td>
<td>Reference title</td>
</tr>
<tr><td><tt class="docutils literal">RX</tt></td>
<td>PubMed ID</td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<p>Printing the motifs writes them out in their native TRANSFAC format:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span>
<span class="name">VV</span>  <span class="name">EXAMPLE</span> <span class="name">January</span> <span class="literal number integer">15</span><span class="punctuation">,</span> <span class="literal number integer">2013</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="name">ID</span>  <span class="name">motif1</span>
<span class="name">XX</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">S</span>
<span class="literal number oct">02</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">R</span>
<span class="literal number oct">03</span>      <span class="literal number integer">3</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">1</span>      <span class="name">A</span>
<span class="literal number oct">04</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">05</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">A</span>
<span class="literal number oct">06</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">4</span>      <span class="literal number integer">1</span>      <span class="name">G</span>
<span class="literal number oct">07</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">4</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">08</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="name">T</span>
<span class="literal number integer">09</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">10</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="name">K</span>
<span class="literal number integer">11</span>      <span class="literal number integer">0</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="name">Y</span>
<span class="literal number integer">12</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="literal number integer">1</span>      <span class="name">G</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="name">ID</span>  <span class="name">motif2</span>
<span class="name">XX</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">R</span>
<span class="literal number oct">02</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">S</span>
<span class="literal number oct">03</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">04</span>      <span class="literal number integer">3</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">1</span>      <span class="name">A</span>
<span class="literal number oct">05</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">4</span>      <span class="literal number integer">1</span>      <span class="name">G</span>
<span class="literal number oct">06</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">A</span>
<span class="literal number oct">07</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">4</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">08</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">09</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="name">T</span>
<span class="literal number integer">10</span>      <span class="literal number integer">0</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="name">Y</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>You can export the motifs in the TRANSFAC format by capturing this
output in a string and saving it in a file:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">text</span> <span class="operator">=</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;mytransfacfile.dat&quot;</span><span class="punctuation">,</span> <span class="literal string">'w'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">text</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
</div>
</div>
<div class="section" id="writing-motifs">
<h3>14.1.3  Writing motifs</h3>
<p>Speaking of exporting, let’s look at export functions in general. To
export a motif in the JASPAR <tt class="docutils literal">.pfm</tt> format, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;pfm&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">3</span>       <span class="literal number integer">7</span>       <span class="literal number integer">0</span>       <span class="literal number integer">2</span>       <span class="literal number integer">1</span>
<span class="literal number integer">0</span>       <span class="literal number integer">0</span>       <span class="literal number integer">5</span>       <span class="literal number integer">2</span>       <span class="literal number integer">6</span>
<span class="literal number integer">0</span>       <span class="literal number integer">0</span>       <span class="literal number integer">0</span>       <span class="literal number integer">3</span>       <span class="literal number integer">0</span>
<span class="literal number integer">4</span>       <span class="literal number integer">0</span>       <span class="literal number integer">2</span>       <span class="literal number integer">0</span>       <span class="literal number integer">0</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>To write the motif in a TRANSFAC-like matrix format, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;transfac&quot;</span><span class="punctuation">)</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">3</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">4</span>      <span class="name">W</span>
<span class="literal number oct">02</span>      <span class="literal number integer">7</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">A</span>
<span class="literal number oct">03</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">2</span>      <span class="name">C</span>
<span class="literal number oct">04</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">0</span>      <span class="name">V</span>
<span class="literal number oct">05</span>      <span class="literal number integer">1</span>      <span class="literal number integer">6</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>To write out multiple motifs, you can use <tt class="docutils literal">motifs.write</tt>. This
function can be used regardless of whether the motifs originated from a
TRANSFAC file. For example,</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">two_motifs</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">arnt</span><span class="punctuation">,</span> <span class="name">srf</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">two_motifs</span><span class="punctuation">,</span> <span class="literal string">'transfac'</span><span class="punctuation">)</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">4</span>     <span class="literal number integer">16</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">02</span>     <span class="literal number integer">19</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>      <span class="name">A</span>
<span class="literal number oct">03</span>      <span class="literal number integer">0</span>     <span class="literal number integer">20</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">04</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>     <span class="literal number integer">20</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number oct">05</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>     <span class="literal number integer">20</span>      <span class="name">T</span>
<span class="literal number oct">06</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>     <span class="literal number integer">20</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>     <span class="literal number integer">39</span>      <span class="literal number integer">4</span>      <span class="name">G</span>
<span class="literal number oct">02</span>      <span class="literal number integer">9</span>     <span class="literal number integer">33</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="name">C</span>
<span class="literal number oct">03</span>      <span class="literal number integer">0</span>     <span class="literal number integer">45</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">04</span>      <span class="literal number integer">1</span>     <span class="literal number integer">45</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">05</span>     <span class="literal number integer">32</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>     <span class="literal number integer">13</span>      <span class="name">A</span>
<span class="literal number oct">06</span>      <span class="literal number integer">3</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>     <span class="literal number integer">42</span>      <span class="name">T</span>
<span class="literal number oct">07</span>     <span class="literal number integer">46</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">A</span>
<span class="literal number integer">08</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>     <span class="literal number integer">45</span>      <span class="name">T</span>
<span class="literal number integer">09</span>     <span class="literal number integer">43</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="name">A</span>
<span class="literal number integer">10</span>     <span class="literal number integer">15</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>     <span class="literal number integer">30</span>      <span class="name">T</span>
<span class="literal number integer">11</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>     <span class="literal number integer">44</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">12</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>     <span class="literal number integer">43</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
</div>
<div class="section" id="creating-a-sequence-logo">
<h3>14.1.4  Creating a sequence logo</h3>
<p>If we have internet access, we can create a
<a class="reference external" href="http://weblogo.berkeley.edu">weblogo</a>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">weblogo</span><span class="punctuation">(</span><span class="literal string">&quot;Arnt.png&quot;</span><span class="punctuation">)</span>
</pre>
<p>We should get our logo saved as a PNG in the specified file.</p>
</div>
</div>
<div class="section" id="position-weight-matrices">
<h2>14.2  Position-Weight Matrices</h2>
<p>The <tt class="docutils literal">.counts</tt> attribute of a Motif object shows how often each
nucleotide appeared at each position along the alignment. We can
normalize this matrix by dividing by the number of instances in the
alignment, resulting in the probability of each nucleotide at each
position along the alignment. We refer to these probabilities as the
position-weight matrix. However, beware that in the literature this term
may also be used to refer to the position-specific scoring matrix, which
we discuss below.</p>
<p>Usually, pseudocounts are added to each position before normalizing.
This avoids overfitting of the position-weight matrix to the limited
number of motif instances in the alignment, and can also prevent
probabilities from becoming zero. To add a fixed pseudocount to all
nucleotides at all positions, specify a number for the <tt class="docutils literal">pseudocounts</tt>
argument:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pwm</span> <span class="operator">=</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="operator">.</span><span class="name">normalize</span><span class="punctuation">(</span><span class="name">pseudocounts</span><span class="operator">=</span><span class="literal number float">0.5</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">pwm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.39</span>   <span class="literal number float">0.83</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.28</span>   <span class="literal number float">0.17</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.61</span>   <span class="literal number float">0.28</span>   <span class="literal number float">0.72</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.39</span>   <span class="literal number float">0.06</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.50</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.28</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.06</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>Alternatively, <tt class="docutils literal">pseudocounts</tt> can be a dictionary specifying the
pseudocounts for each nucleotide. For example, as the GC content of the
human genome is about 40%, you may want to choose the pseudocounts
accordingly:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pwm</span> <span class="operator">=</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="operator">.</span><span class="name">normalize</span><span class="punctuation">(</span><span class="name">pseudocounts</span><span class="operator">=</span><span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span><span class="literal number float">0.6</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.4</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.4</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.6</span><span class="punctuation">})</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">pwm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.40</span>   <span class="literal number float">0.84</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.29</span>   <span class="literal number float">0.18</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.60</span>   <span class="literal number float">0.27</span>   <span class="literal number float">0.71</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.38</span>   <span class="literal number float">0.04</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.51</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.29</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.07</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>The position-weight matrix has its own methods to calculate the
consensus, anticonsensus, and degenerate consensus sequences:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pwm</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TACGC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pwm</span><span class="operator">.</span><span class="name">anticonsensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GGGTG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pwm</span><span class="operator">.</span><span class="name">degenerate_consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'WACNC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>Note that due to the pseudocounts, the degenerate consensus sequence
calculated from the position-weight matrix is slightly different from
the degenerate consensus sequence calculated from the instances in the
motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">degenerate_consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'WACVC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>The reverse complement of the position-weight matrix can be calculated
directly from the <tt class="docutils literal">pwm</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">rpwm</span> <span class="operator">=</span> <span class="name">pwm</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">rpwm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.29</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.51</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.38</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.04</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.71</span>   <span class="literal number float">0.27</span>   <span class="literal number float">0.60</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.04</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.18</span>   <span class="literal number float">0.29</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.84</span>   <span class="literal number float">0.40</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
</div>
<div class="section" id="position-specific-scoring-matrices">
<h2>14.3  Position-Specific Scoring Matrices</h2>
<p>Using the background distribution and PWM with pseudo-counts added, it’s
easy to compute the log-odds ratios, telling us what are the log odds of
a particular symbol to be coming from a motif against the background. We
can use the <tt class="docutils literal">.log_odds()</tt> method on the position-weight matrix:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pssm</span> <span class="operator">=</span> <span class="name">pwm</span><span class="operator">.</span><span class="name">log_odds</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">pssm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.68</span>   <span class="literal number float">1.76</span>  <span class="operator">-</span><span class="literal number float">1.91</span>   <span class="literal number float">0.21</span>  <span class="operator">-</span><span class="literal number float">0.49</span>
<span class="name">C</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">2.49</span>  <span class="operator">-</span><span class="literal number float">2.49</span>   <span class="literal number float">1.26</span>   <span class="literal number float">0.09</span>   <span class="literal number float">1.51</span>
<span class="name">G</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">2.49</span>  <span class="operator">-</span><span class="literal number float">2.49</span>  <span class="operator">-</span><span class="literal number float">2.49</span>   <span class="literal number float">0.60</span>  <span class="operator">-</span><span class="literal number float">2.49</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">1.03</span>  <span class="operator">-</span><span class="literal number float">1.91</span>   <span class="literal number float">0.21</span>  <span class="operator">-</span><span class="literal number float">1.91</span>  <span class="operator">-</span><span class="literal number float">1.91</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>Here we can see positive values for symbols more frequent in the motif
than in the background and negative for symbols more frequent in the
background. 0.0 means that it’s equally likely to see a symbol in the
background and in the motif.</p>
<p>This assumes that A, C, G, and T are equally likely in the background.
To calculate the position-specific scoring matrix against a background
with unequal probabilities for A, C, G, T, use the <tt class="docutils literal">background</tt>
argument. For example, against a background with a 40% GC content, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">background</span> <span class="operator">=</span> <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span><span class="literal number float">0.3</span><span class="punctuation">,</span><span class="literal string">'C'</span><span class="punctuation">:</span><span class="literal number float">0.2</span><span class="punctuation">,</span><span class="literal string">'G'</span><span class="punctuation">:</span><span class="literal number float">0.2</span><span class="punctuation">,</span><span class="literal string">'T'</span><span class="punctuation">:</span><span class="literal number float">0.3</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pssm</span> <span class="operator">=</span> <span class="name">pwm</span><span class="operator">.</span><span class="name">log_odds</span><span class="punctuation">(</span><span class="name">background</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">pssm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.42</span>   <span class="literal number float">1.49</span>  <span class="operator">-</span><span class="literal number float">2.17</span>  <span class="operator">-</span><span class="literal number float">0.05</span>  <span class="operator">-</span><span class="literal number float">0.75</span>
<span class="name">C</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">2.17</span>  <span class="operator">-</span><span class="literal number float">2.17</span>   <span class="literal number float">1.58</span>   <span class="literal number float">0.42</span>   <span class="literal number float">1.83</span>
<span class="name">G</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">2.17</span>  <span class="operator">-</span><span class="literal number float">2.17</span>  <span class="operator">-</span><span class="literal number float">2.17</span>   <span class="literal number float">0.92</span>  <span class="operator">-</span><span class="literal number float">2.17</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.77</span>  <span class="operator">-</span><span class="literal number float">2.17</span>  <span class="operator">-</span><span class="literal number float">0.05</span>  <span class="operator">-</span><span class="literal number float">2.17</span>  <span class="operator">-</span><span class="literal number float">2.17</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>The maximum and minimum score obtainable from the PSSM are stored in the
<tt class="docutils literal">.max</tt> and <tt class="docutils literal">.min</tt> properties:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">max</span>
<span class="literal number float">6.59</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">min</span>
<span class="operator">-</span><span class="literal number float">10.85</span>
</pre>
<p>The mean and standard deviation of the PSSM scores with respect to a
specific background are calculated by the <tt class="docutils literal">.mean</tt> and <tt class="docutils literal">.std</tt>
methods.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mean</span> <span class="operator">=</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">mean</span><span class="punctuation">(</span><span class="name">background</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">std</span> <span class="operator">=</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">std</span><span class="punctuation">(</span><span class="name">background</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;mean = </span><span class="literal string interpol">%0.2f</span><span class="literal string">, standard deviation = </span><span class="literal string interpol">%0.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">mean</span><span class="punctuation">,</span> <span class="name">std</span><span class="punctuation">)</span>
<span class="name">mean</span> <span class="operator">=</span> <span class="literal number float">3.21</span><span class="punctuation">,</span> <span class="name">standard</span> <span class="name">deviation</span> <span class="operator">=</span> <span class="literal number float">2.59</span>
</pre>
<p>A uniform background is used if <tt class="docutils literal">background</tt> is not specified. The
mean is particularly important, as its value is equal to the
Kullback-Leibler divergence or relative entropy, and is a measure for
the information content of the motif compared to the background. As in
Biopython the base-2 logarithm is used in the calculation of the
log-odds scores, the information content has units of bits.</p>
<p>The <tt class="docutils literal">.reverse_complement</tt>, <tt class="docutils literal">.consensus</tt>, <tt class="docutils literal">.anticonsensus</tt>, and
<tt class="docutils literal">.degenerate_consensus</tt> methods can be applied directly to PSSM
objects.</p>
</div>
<div class="section" id="searching-for-instances">
<h2>14.4  Searching for instances</h2>
<p>The most frequent use for a motif is to find its instances in some
sequence. For the sake of this section, we will use an artificial
sequence like this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">test_seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TACACTGCATTACAACCCAAGCATTA&quot;</span><span class="punctuation">,</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">)</span>
<span class="literal number integer">26</span>
</pre>
<div class="section" id="searching-for-exact-matches">
<h3>14.4.1  Searching for exact matches</h3>
<p>The simplest way to find instances, is to look for exact matches of the
true instances of the motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">seq</span> <span class="operator word">in</span> <span class="name">m</span><span class="operator">.</span><span class="name">instances</span><span class="operator">.</span><span class="name">search</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pos</span><span class="punctuation">,</span> <span class="name">seq</span>
<span class="operator">...</span>
<span class="literal number integer">0</span> <span class="name">TACAC</span>
<span class="literal number integer">10</span> <span class="name">TACAA</span>
<span class="literal number integer">13</span> <span class="name">AACCC</span>
</pre>
<p>We can do the same with the reverse complement (to find instances on the
complementary strand):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">seq</span> <span class="operator word">in</span> <span class="name">r</span><span class="operator">.</span><span class="name">instances</span><span class="operator">.</span><span class="name">search</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pos</span><span class="punctuation">,</span> <span class="name">seq</span>
<span class="operator">...</span>
<span class="literal number integer">6</span> <span class="name">GCATT</span>
<span class="literal number integer">20</span> <span class="name">GCATT</span>
</pre>
</div>
<div class="section" id="searching-for-matches-using-the-pssm-score">
<h3>14.4.2  Searching for matches using the PSSM score</h3>
<p>It’s just as easy to look for positions, giving rise to high log-odds
scores against our motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">position</span><span class="punctuation">,</span> <span class="name">score</span> <span class="operator word">in</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">search</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">,</span> <span class="name">threshold</span><span class="operator">=</span><span class="literal number float">3.0</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;Position </span><span class="literal string interpol">%d</span><span class="literal string">: score = </span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">position</span><span class="punctuation">,</span> <span class="name">score</span><span class="punctuation">)</span>
<span class="operator">...</span>
<span class="name">Position</span> <span class="literal number integer">0</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">5.622</span>
<span class="name">Position</span> <span class="operator">-</span><span class="literal number integer">20</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">4.601</span>
<span class="name">Position</span> <span class="literal number integer">10</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">3.037</span>
<span class="name">Position</span> <span class="literal number integer">13</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">5.738</span>
<span class="name">Position</span> <span class="operator">-</span><span class="literal number integer">6</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">4.601</span>
</pre>
<p>The negative positions refer to instances of the motif found on the
reverse strand of the test sequence, and follow the Python convention on
negative indices. Therefore, the instance of the motif at <tt class="docutils literal">pos</tt> is
located at <tt class="docutils literal">test_seq[pos:pos+len(m)]</tt> both for positive and for
negative values of <tt class="docutils literal">pos</tt>.</p>
<p>You may notice the threshold parameter, here set arbitrarily to 3.0.
This is in <em>log</em><sub>2</sub>, so we are now looking only for words, which
are eight times more likely to occur under the motif model than in the
background. The default threshold is 0.0, which selects everything that
looks more like the motif than the background.</p>
<p>You can also calculate the scores at all positions along the sequence:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">calculate</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">)</span>
<span class="name">array</span><span class="punctuation">([</span>  <span class="literal number float">5.62230396</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">5.6796999</span> <span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">3.43177247</span><span class="punctuation">,</span>   <span class="literal number float">0.93827754</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">6.84962511</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">2.04066086</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">10.84962463</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">3.65614533</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">0.03370807</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">3.91102552</span><span class="punctuation">,</span>   <span class="literal number float">3.03734159</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">2.14918518</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">0.6016975</span> <span class="punctuation">,</span>   <span class="literal number float">5.7381525</span> <span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">0.50977498</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">3.56422281</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">8.73414803</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">0.09919716</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">0.6016975</span> <span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">2.39429784</span><span class="punctuation">,</span>
       <span class="operator">-</span><span class="literal number float">10.84962463</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">3.65614533</span><span class="punctuation">],</span> <span class="name">dtype</span><span class="operator">=</span><span class="name">float32</span><span class="punctuation">)</span>
</pre>
<p>In general, this is the fastest way to calculate PSSM scores. The scores
returned by <tt class="docutils literal">pssm.calculate</tt> are for the forward strand only. To
obtain the scores on the reverse strand, you can take the reverse
complement of the PSSM:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">rpssm</span> <span class="operator">=</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">rpssm</span><span class="operator">.</span><span class="name">calculate</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">)</span>
<span class="name">array</span><span class="punctuation">([</span> <span class="operator">-</span><span class="literal number float">9.43458748</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">3.06172252</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">7.18665981</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">7.76216221</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">2.04066086</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">4.26466274</span><span class="punctuation">,</span>   <span class="literal number float">4.60124254</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">4.2480607</span> <span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">8.73414803</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">2.26503372</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">6.49598789</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">5.64668512</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">8.73414803</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">10.84962463</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">4.82356262</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">4.82356262</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">5.64668512</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">8.73414803</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">4.15613794</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">5.6796999</span> <span class="punctuation">,</span>
         <span class="literal number float">4.60124254</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">4.2480607</span> <span class="punctuation">],</span> <span class="name">dtype</span><span class="operator">=</span><span class="name">float32</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="selecting-a-score-threshold">
<h3>14.4.3  Selecting a score threshold</h3>
<p>If you want to use a less arbitrary way of selecting thresholds, you can
explore the distribution of PSSM scores. Since the space for a score
distribution grows exponentially with motif length, we are using an
approximation with a given precision to keep computation cost
manageable:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">distribution</span> <span class="operator">=</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">distribution</span><span class="punctuation">(</span><span class="name">background</span><span class="operator">=</span><span class="name">background</span><span class="punctuation">,</span> <span class="name">precision</span><span class="operator">=</span><span class="literal number integer">10</span><span class="operator">**</span><span class="literal number integer">4</span><span class="punctuation">)</span>
</pre>
<p>The <tt class="docutils literal">distribution</tt> object can be used to determine a number of
different thresholds. We can specify the requested false-positive rate
(probability of “finding” a motif instance in background generated
sequence):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">threshold</span> <span class="operator">=</span> <span class="name">distribution</span><span class="operator">.</span><span class="name">threshold_fpr</span><span class="punctuation">(</span><span class="literal number float">0.01</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">threshold</span>
<span class="literal number float">4.009</span>
</pre>
<p>or the false-negative rate (probability of “not finding” an instance
generated from the motif):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">threshold</span> <span class="operator">=</span> <span class="name">distribution</span><span class="operator">.</span><span class="name">threshold_fnr</span><span class="punctuation">(</span><span class="literal number float">0.1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">threshold</span>
<span class="operator">-</span><span class="literal number float">0.510</span>
</pre>
<p>or a threshold (approximately) satisfying some relation between the
false-positive rate and the false-negative rate (fnr/fpr≃ <em>t</em>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">threshold</span> <span class="operator">=</span> <span class="name">distribution</span><span class="operator">.</span><span class="name">threshold_balanced</span><span class="punctuation">(</span><span class="literal number integer">1000</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">threshold</span>
<span class="literal number float">6.241</span>
</pre>
<p>or a threshold satisfying (roughly) the equality between the
false-positive rate and the −<em>log</em> of the information content (as used
in patser software by Hertz and Stormo):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">threshold</span> <span class="operator">=</span> <span class="name">distribution</span><span class="operator">.</span><span class="name">threshold_patser</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">threshold</span>
<span class="literal number float">0.346</span>
</pre>
<p>For example, in case of our motif, you can get the threshold giving you
exactly the same results (for this sequence) as searching for instances
with balanced threshold with rate of 1000.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">threshold</span> <span class="operator">=</span> <span class="name">distribution</span><span class="operator">.</span><span class="name">threshold_fpr</span><span class="punctuation">(</span><span class="literal number float">0.01</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">threshold</span>
<span class="literal number float">4.009</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">position</span><span class="punctuation">,</span> <span class="name">score</span> <span class="operator word">in</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">search</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">,</span><span class="name">threshold</span><span class="operator">=</span><span class="name">threshold</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;Position </span><span class="literal string interpol">%d</span><span class="literal string">: score = </span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">position</span><span class="punctuation">,</span> <span class="name">score</span><span class="punctuation">)</span>
<span class="operator">...</span>
<span class="name">Position</span> <span class="literal number integer">0</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">5.622</span>
<span class="name">Position</span> <span class="operator">-</span><span class="literal number integer">20</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">4.601</span>
<span class="name">Position</span> <span class="literal number integer">13</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">5.738</span>
<span class="name">Position</span> <span class="operator">-</span><span class="literal number integer">6</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">4.601</span>
</pre>
</div>
</div>
<div class="section" id="each-motif-object-has-an-associated-position-specific-scoring-matrix">
<h2>14.5  Each motif object has an associated Position-Specific Scoring Matrix</h2>
<p>To facilitate searching for potential TFBSs using PSSMs, both the
position-weight matrix and the position-specific scoring matrix are
associated with each motif. Using the Arnt motif as an example:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">motifs</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;Arnt.sites&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">'sites'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">counts</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">4.00</span>  <span class="literal number float">19.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="name">C</span><span class="punctuation">:</span>  <span class="literal number float">16.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>   <span class="literal number float">0.00</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pwm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.20</span>   <span class="literal number float">0.95</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.80</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.05</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">0.32</span>   <span class="literal number float">1.93</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">1.68</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="literal number float">2.00</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="operator">-</span><span class="name">inf</span>  <span class="operator">-</span><span class="literal number float">2.32</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="literal number float">2.00</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="literal number float">2.00</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="literal number float">2.00</span>   <span class="operator">-</span><span class="name">inf</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>The negative infinities appear here because the corresponding entry in
the frequency matrix is 0, and we are using zero pseudocounts by
default:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">letter</span> <span class="operator word">in</span> <span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">: </span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">letter</span><span class="punctuation">,</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pseudocounts</span><span class="punctuation">[</span><span class="name">letter</span><span class="punctuation">])</span>
<span class="operator">...</span>
<span class="name">A</span><span class="punctuation">:</span> <span class="literal number float">0.00</span>
<span class="name">C</span><span class="punctuation">:</span> <span class="literal number float">0.00</span>
<span class="name">G</span><span class="punctuation">:</span> <span class="literal number float">0.00</span>
<span class="name">T</span><span class="punctuation">:</span> <span class="literal number float">0.00</span>
</pre>
<p>If you change the <tt class="docutils literal">.pseudocounts</tt> attribute, the position-frequency
matrix and the position-specific scoring matrix are recalculated
automatically:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pseudocounts</span> <span class="operator">=</span> <span class="literal number float">3.0</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">letter</span> <span class="operator word">in</span> <span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">: </span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">letter</span><span class="punctuation">,</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pseudocounts</span><span class="punctuation">[</span><span class="name">letter</span><span class="punctuation">])</span>
<span class="operator">...</span>
<span class="name">A</span><span class="punctuation">:</span> <span class="literal number float">3.00</span>
<span class="name">C</span><span class="punctuation">:</span> <span class="literal number float">3.00</span>
<span class="name">G</span><span class="punctuation">:</span> <span class="literal number float">3.00</span>
<span class="name">T</span><span class="punctuation">:</span> <span class="literal number float">3.00</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pwm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.22</span>   <span class="literal number float">0.69</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.59</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.72</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.12</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.72</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.72</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.72</span>   <span class="literal number float">0.09</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">0.19</span>   <span class="literal number float">1.46</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">1.25</span>  <span class="operator">-</span><span class="literal number float">1.42</span>   <span class="literal number float">1.52</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>
<span class="name">G</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.00</span>  <span class="operator">-</span><span class="literal number float">1.42</span>   <span class="literal number float">1.52</span>  <span class="operator">-</span><span class="literal number float">1.42</span>   <span class="literal number float">1.52</span>
<span class="name">T</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>   <span class="literal number float">1.52</span>  <span class="operator">-</span><span class="literal number float">1.42</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>You can also set the <tt class="docutils literal">.pseudocounts</tt> to a dictionary over the four
nucleotides if you want to use different pseudocounts for them. Setting
<tt class="docutils literal">motif.pseudocounts</tt> to <tt class="docutils literal">None</tt> resets it to its default value of
zero.</p>
<p>The position-specific scoring matrix depends on the background
distribution, which is uniform by default:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">letter</span> <span class="operator word">in</span> <span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">: </span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">letter</span><span class="punctuation">,</span> <span class="name">motif</span><span class="operator">.</span><span class="name">background</span><span class="punctuation">[</span><span class="name">letter</span><span class="punctuation">])</span>
<span class="operator">...</span>
<span class="name">A</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
<span class="name">C</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
<span class="name">G</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
<span class="name">T</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
</pre>
<p>Again, if you modify the background distribution, the position-specific
scoring matrix is recalculated:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">background</span> <span class="operator">=</span> <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.2</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.3</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.3</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.2</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.13</span>   <span class="literal number float">1.78</span>  <span class="operator">-</span><span class="literal number float">1.09</span>  <span class="operator">-</span><span class="literal number float">1.09</span>  <span class="operator">-</span><span class="literal number float">1.09</span>  <span class="operator">-</span><span class="literal number float">1.09</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.98</span>  <span class="operator">-</span><span class="literal number float">1.68</span>   <span class="literal number float">1.26</span>  <span class="operator">-</span><span class="literal number float">1.68</span>  <span class="operator">-</span><span class="literal number float">1.68</span>  <span class="operator">-</span><span class="literal number float">1.68</span>
<span class="name">G</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">1.68</span>  <span class="operator">-</span><span class="literal number float">1.26</span>  <span class="operator">-</span><span class="literal number float">1.68</span>   <span class="literal number float">1.26</span>  <span class="operator">-</span><span class="literal number float">1.68</span>   <span class="literal number float">1.26</span>
<span class="name">T</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">1.09</span>  <span class="operator">-</span><span class="literal number float">1.09</span>  <span class="operator">-</span><span class="literal number float">1.09</span>  <span class="operator">-</span><span class="literal number float">1.09</span>   <span class="literal number float">1.85</span>  <span class="operator">-</span><span class="literal number float">1.09</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>Setting <tt class="docutils literal">motif.background</tt> to <tt class="docutils literal">None</tt> resets it to a uniform
distribution:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">background</span> <span class="operator">=</span> <span class="name builtin pseudo">None</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">letter</span> <span class="operator word">in</span> <span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">: </span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">letter</span><span class="punctuation">,</span> <span class="name">motif</span><span class="operator">.</span><span class="name">background</span><span class="punctuation">[</span><span class="name">letter</span><span class="punctuation">])</span>
<span class="operator">...</span>
<span class="name">A</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
<span class="name">C</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
<span class="name">G</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
<span class="name">T</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
</pre>
<p>If you set <tt class="docutils literal">motif.background</tt> equal to a single value, it will be
interpreted as the GC content:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">background</span> <span class="operator">=</span> <span class="literal number float">0.8</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">letter</span> <span class="operator word">in</span> <span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">: </span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">letter</span><span class="punctuation">,</span> <span class="name">motif</span><span class="operator">.</span><span class="name">background</span><span class="punctuation">[</span><span class="name">letter</span><span class="punctuation">])</span>
<span class="operator">...</span>
<span class="name">A</span><span class="punctuation">:</span> <span class="literal number float">0.10</span>
<span class="name">C</span><span class="punctuation">:</span> <span class="literal number float">0.40</span>
<span class="name">G</span><span class="punctuation">:</span> <span class="literal number float">0.40</span>
<span class="name">T</span><span class="punctuation">:</span> <span class="literal number float">0.10</span>
</pre>
<p>Note that you can now calculate the mean of the PSSM scores over the
background against which it was computed:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span><span class="operator">.</span><span class="name">mean</span><span class="punctuation">(</span><span class="name">motif</span><span class="operator">.</span><span class="name">background</span><span class="punctuation">)</span>
<span class="literal number float">4.703928</span>
</pre>
<p>as well as its standard deviation:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span><span class="operator">.</span><span class="name">std</span><span class="punctuation">(</span><span class="name">motif</span><span class="operator">.</span><span class="name">background</span><span class="punctuation">)</span>
<span class="literal number float">3.290900</span>
</pre>
<p>and its distribution:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">distribution</span> <span class="operator">=</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span><span class="operator">.</span><span class="name">distribution</span><span class="punctuation">(</span><span class="name">background</span><span class="operator">=</span><span class="name">motif</span><span class="operator">.</span><span class="name">background</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">threshold</span> <span class="operator">=</span> <span class="name">distribution</span><span class="operator">.</span><span class="name">threshold_fpr</span><span class="punctuation">(</span><span class="literal number float">0.01</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">threshold</span>
<span class="literal number float">3.854375</span>
</pre>
<p>Note that the position-weight matrix and the position-specific scoring
matrix are recalculated each time you call <tt class="docutils literal">motif.pwm</tt> or
<tt class="docutils literal">motif.pssm</tt>, respectively. If speed is an issue and you want to use
the PWM or PSSM repeatedly, you can save them as a variable, as in</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pssm</span> <span class="operator">=</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span>
</pre>
</div>
<div class="section" id="comparing-motifs">
<h2>14.6  Comparing motifs</h2>
<p>Once we have more than one motif, we might want to compare them.</p>
<p>Before we start comparing motifs, I should point out that motif
boundaries are usually quite arbitrary. This means we often need to
compare motifs of different lengths, so comparison needs to involve some
kind of alignment. This means we have to take into account two things:</p>
<ul class="simple">
<li>alignment of motifs</li>
<li>some function to compare aligned motifs</li>
</ul>
<p>To align the motifs, we use ungapped alignment of PSSMs and substitute
zeros for any missing columns at the beginning and end of the matrices.
This means that effectively we are using the background distribution for
columns missing from the PSSM. The distance function then returns the
minimal distance between motifs, as well as the corresponding offset in
their alignment.</p>
<p>To give an exmaple, let us first load another motif, which is similar to
our test motif <tt class="docutils literal">m</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m_reb1</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;REB1.pfm&quot;</span><span class="punctuation">),</span> <span class="literal string">&quot;pfm&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m_reb1</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GTTACCCGG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m_reb1</span><span class="operator">.</span><span class="name">counts</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>      <span class="literal number integer">6</span>      <span class="literal number integer">7</span>      <span class="literal number integer">8</span>
<span class="name">A</span><span class="punctuation">:</span>  <span class="literal number float">30.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span> <span class="literal number float">100.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">15.00</span>
<span class="name">C</span><span class="punctuation">:</span>  <span class="literal number float">10.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span> <span class="literal number float">100.00</span> <span class="literal number float">100.00</span> <span class="literal number float">100.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">15.00</span>
<span class="name">G</span><span class="punctuation">:</span>  <span class="literal number float">50.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">60.00</span>  <span class="literal number float">55.00</span>
<span class="name">T</span><span class="punctuation">:</span>  <span class="literal number float">10.00</span> <span class="literal number float">100.00</span> <span class="literal number float">100.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">40.00</span>  <span class="literal number float">15.00</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>To make the motifs comparable, we choose the same values for the
pseudocounts and the background distribution as our motif <tt class="docutils literal">m</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m_reb1</span><span class="operator">.</span><span class="name">pseudocounts</span> <span class="operator">=</span> <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span><span class="literal number float">0.6</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.4</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.4</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.6</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m_reb1</span><span class="operator">.</span><span class="name">background</span> <span class="operator">=</span> <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span><span class="literal number float">0.3</span><span class="punctuation">,</span><span class="literal string">'C'</span><span class="punctuation">:</span><span class="literal number float">0.2</span><span class="punctuation">,</span><span class="literal string">'G'</span><span class="punctuation">:</span><span class="literal number float">0.2</span><span class="punctuation">,</span><span class="literal string">'T'</span><span class="punctuation">:</span><span class="literal number float">0.3</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pssm_reb1</span> <span class="operator">=</span> <span class="name">m_reb1</span><span class="operator">.</span><span class="name">pssm</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">pssm_reb1</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>      <span class="literal number integer">6</span>      <span class="literal number integer">7</span>      <span class="literal number integer">8</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>   <span class="literal number float">1.72</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">0.97</span>
<span class="name">C</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">0.97</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>   <span class="literal number float">2.30</span>   <span class="literal number float">2.30</span>   <span class="literal number float">2.30</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">0.41</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">1.30</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>   <span class="literal number float">1.57</span>   <span class="literal number float">1.44</span>
<span class="name">T</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">1.53</span>   <span class="literal number float">1.72</span>   <span class="literal number float">1.72</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>   <span class="literal number float">0.41</span>  <span class="operator">-</span><span class="literal number float">0.97</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>We’ll compare these motifs using the Pearson correlation. Since we want
it to resemble a distance measure, we actually take 1−<em>r</em>, where <em>r</em>
is the Pearson correlation coefficient (PCC):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">distance</span><span class="punctuation">,</span> <span class="name">offset</span> <span class="operator">=</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">dist_pearson</span><span class="punctuation">(</span><span class="name">pssm_reb1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;distance = </span><span class="literal string interpol">%5.3g</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">distance</span>
<span class="name">distance</span> <span class="operator">=</span> <span class="literal number float">0.239</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">offset</span>
<span class="operator">-</span><span class="literal number integer">2</span>
</pre>
<p>This means that the best PCC between motif <tt class="docutils literal">m</tt> and <tt class="docutils literal">m_reb1</tt> is
obtained with the following alignment:</p>
<pre class="code python literal-block">
<span class="name">m</span><span class="punctuation">:</span>      <span class="name">bbTACGCbb</span>
<span class="name">m_reb1</span><span class="punctuation">:</span> <span class="name">GTTACCCGG</span>
</pre>
<p>where <tt class="docutils literal">b</tt> stands for background distribution. The PCC itself is
roughly 1−0.239=0.761.</p>
</div>
<div class="section" id="de-novo-motif-finding">
<h2>14.7  <em>De novo</em> motif finding</h2>
<p>Currently, Biopython has only limited support for <em>de novo</em> motif
finding. Namely, we support running and parsing of AlignAce and MEME.
Since the number of motif finding tools is growing rapidly,
contributions of new parsers are welcome.</p>
<div class="section" id="id6">
<h3>14.7.1  MEME</h3>
<p>Let’s assume, you have run MEME on sequences of your choice with your
favorite parameters and saved the output in the file <tt class="docutils literal">meme.out</tt>. You
can retrieve the motifs reported by MEME by running the following piece
of code:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">motifs</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;meme.out&quot;</span><span class="punctuation">),</span> <span class="literal string">&quot;meme&quot;</span><span class="punctuation">)</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span>
<span class="punctuation">[</span><span class="operator">&lt;</span><span class="name">Bio</span><span class="operator">.</span><span class="name">motifs</span><span class="operator">.</span><span class="name">meme</span><span class="operator">.</span><span class="name">Motif</span> <span class="name builtin">object</span> <span class="name">at</span> <span class="literal number hex">0xc356b0</span><span class="operator">&gt;</span><span class="punctuation">]</span>
</pre>
<p>Besides the most wanted list of motifs, the result object contains more
useful information, accessible through properties with self-explanatory
names:</p>
<ul class="simple">
<li><tt class="docutils literal">.alphabet</tt></li>
<li><tt class="docutils literal">.datafile</tt></li>
<li><tt class="docutils literal">.sequence_names</tt></li>
<li><tt class="docutils literal">.version</tt></li>
<li><tt class="docutils literal">.command</tt></li>
</ul>
<p>The motifs returned by the MEME Parser can be treated exactly like
regular Motif objects (with instances), they also provide some extra
functionality, by adding additional information about the instances.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CTCAATCGTA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">sequence_name</span>
<span class="literal string">'SEQ10;'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">start</span>
<span class="literal number integer">3</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">strand</span>
<span class="literal string">'+'</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">pvalue</span>
<span class="literal number float">8.71e-07</span>
</pre>
</div>
<div class="section" id="alignace">
<h3>14.7.2  AlignAce</h3>
<p>We can do very similar things with the AlignACE program. Assume, you
have your output in the file <tt class="docutils literal">alignace.out</tt>. You can parse your output
with the following code:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">motifs</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsA</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;alignace.out&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;alignace&quot;</span><span class="punctuation">)</span>
</pre>
<p>Again, your motifs behave as they should:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsA</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TCTACGATTGAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>In fact you can even see, that AlignAce found a very similar motif as
MEME. It is just a longer version of a reverse complement of the MEME
motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TACGATTGAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>If you have AlignAce installed on the same machine, you can also run it
directly from Biopython. A short example of how this can be done is
shown below (other parameters can be specified as keyword parameters):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">command</span><span class="operator">=</span><span class="literal string">&quot;/opt/bin/AlignACE&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">input_file</span><span class="operator">=</span><span class="literal string">&quot;test.fa&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.motifs.applications</span> <span class="keyword namespace">import</span> <span class="name">AlignAceCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cmd</span> <span class="operator">=</span> <span class="name">AlignAceCommandline</span><span class="punctuation">(</span><span class="name">cmd</span><span class="operator">=</span><span class="name">command</span><span class="punctuation">,</span><span class="name builtin">input</span><span class="operator">=</span><span class="name">input_file</span><span class="punctuation">,</span><span class="name">gcback</span><span class="operator">=</span><span class="literal number float">0.6</span><span class="punctuation">,</span><span class="name">numcols</span><span class="operator">=</span><span class="literal number integer">10</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span><span class="name">stderr</span><span class="operator">=</span> <span class="name">cmd</span><span class="punctuation">()</span>
</pre>
<p>Since AlignAce prints all of its output to standard output, you can get
to your motifs by parsing the first part of the result:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifs</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">stdout</span><span class="punctuation">,</span><span class="literal string">&quot;alignace&quot;</span><span class="punctuation">)</span>
</pre>
</div>
</div>
<div class="section" id="useful-links">
<h2>14.8  Useful links</h2>
<ul class="simple">
<li><a class="reference external" href="http://en.wikipedia.org/wiki/Sequence_motif">Sequence motif</a> in
wikipedia</li>
<li><a class="reference external" href="http://en.wikipedia.org/wiki/Position_weight_matrix">PWM</a> in
wikipedia</li>
<li><a class="reference external" href="http://en.wikipedia.org/wiki/Consensus_sequence">Consensus
sequence</a> in
wikipedia</li>
<li><a class="reference external" href="http://bio.cs.washington.edu/assessment/">Comparison of different motif finding
programs</a></li>
</ul>
</div>
<div class="section" id="obsolete-bio-motif-module">
<h2>14.9  Obsolete Bio.Motif module</h2>
<p>The rest of this chapter above describes the <tt class="docutils literal">Bio.motifs</tt> package
included in Biopython 1.61 onwards, which is replacing the older
<tt class="docutils literal">Bio.Motif</tt> package introduced with Biopython 1.50, which was in turn
based on two older former Biopython modules, <tt class="docutils literal">Bio.AlignAce</tt> and
<tt class="docutils literal">Bio.MEME</tt>.</p>
<p>To allow for a smooth transition, the older <tt class="docutils literal">Bio.Motif</tt> package will
be maintained in parallel with its replacement <tt class="docutils literal">Bio.motifs</tt> at least
two more releases, and at least one year.</p>
<div class="section" id="id7">
<h3>14.9.1  Motif objects</h3>
<p>Since we are interested in motif analysis, we need to take a look at
<tt class="docutils literal">Motif</tt> objects in the first place. For that we need to import the
Motif library:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Motif</span>
</pre>
<p>and we can start creating our first motif objects. Let’s create a DNA
motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span> <span class="operator">=</span> <span class="name">Motif</span><span class="operator">.</span><span class="name">Motif</span><span class="punctuation">(</span><span class="name">alphabet</span><span class="operator">=</span><span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
</pre>
<p>This is for now just an empty container, so let’s add some sequences to
our newly created motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">add_instance</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TATAA&quot;</span><span class="punctuation">,</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">add_instance</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TATTA&quot;</span><span class="punctuation">,</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">add_instance</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TATAA&quot;</span><span class="punctuation">,</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">add_instance</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TATAA&quot;</span><span class="punctuation">,</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">))</span>
</pre>
<p>Now we have a full <tt class="docutils literal">Motif</tt> instance, so we can try to get some basic
information about it. Let’s start with length and consensus sequence:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">m</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">consensus</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TATAA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>In case of DNA motifs, we can also get a reverse complement of a motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">consensus</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TTATA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name">m</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">i</span>
<span class="name">TTATA</span>
<span class="name">TAATA</span>
<span class="name">TTATA</span>
<span class="name">TTATA</span>
</pre>
<p>We can also calculate the information content of a motif with a simple
call:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%0.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">m</span><span class="operator">.</span><span class="name">ic</span><span class="punctuation">()</span>
<span class="literal number float">5.27</span>
</pre>
<p>This gives us a number of bits of information provided by the motif,
which tells us how much differs from background.</p>
<p>The most common representation of a motif is a PWM (Position Weight
Matrix). It summarizes the probabilities of finding any symbol (in this
case nucleotide) in any position of a motif. It can be computed by
calling the <tt class="docutils literal">.pwm()</tt> method:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">pwm</span><span class="punctuation">()</span>
<span class="punctuation">[{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.85</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.85</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.85</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.65</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.25</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.85</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">}]</span>
</pre>
<p>The probabilities in the motif’s PWM are based on the counts in the
instances, but we can see, that even though there were no Gs and no Cs
in the instances, we still have non-zero probabilities assigned to them.
These come from pseudo-counts which are, roughly speaking, a commonly
used way to acknowledge the incompleteness of our knowledge and avoid
technical problems with calculating logarithms of 0.</p>
<p>We can control the way that pseudo-counts are added with two properties
of Motif objects <tt class="docutils literal">.background</tt> is the probability distribution over
all symbols in the alphabet that we assume represents background,
non-motif sequences (usually based on the GC content of the respective
genome). It is by default set to a uniform distribution upon creation of
a motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">background</span>
<span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.25</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.25</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.25</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.25</span><span class="punctuation">}</span>
</pre>
<p>The other parameter is <tt class="docutils literal">.beta</tt>, which states the amount of
pseudo-counts we should add to the PWM. By default it is set to 1.0,</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">beta</span>
<span class="literal number float">1.0</span>
</pre>
<p>so that the total input of pseudo-counts is equal to that of one
instance.</p>
<p>Using the background distribution and pwm with pseudo-counts added, it’s
easy to compute the log-odds ratios, telling us what are the log odds of
a particular symbol to be coming from a motif against the background. We
can use the <tt class="docutils literal">.log_odds()</tt> method:</p>
<pre class="code python literal-block">
 <span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">log_odds</span><span class="punctuation">()</span>
<span class="punctuation">[{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">1.7655347463629771</span><span class="punctuation">,</span>
  <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">1.7655347463629771</span><span class="punctuation">,</span>
  <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">1.7655347463629771</span><span class="punctuation">,</span>
  <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">1.3785116232537298</span><span class="punctuation">,</span>
  <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.0</span><span class="punctuation">,</span>
  <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">1.7655347463629771</span><span class="punctuation">,</span>
  <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">}</span>
<span class="punctuation">]</span>
</pre>
<p>Here we can see positive values for symbols more frequent in the motif
than in the background and negative for symbols more frequent in the
background. 0.0 means that it’s equally likely to see a symbol in
background and in the motif (e.g. ‘T’ in the second-last position).</p>
<div class="section" id="reading-and-writing">
<h4>14.9.1.1  Reading and writing</h4>
<p>Creating motifs from instances by hand is a bit boring, so it’s useful
to have some I/O functions for reading and writing motifs. There are no
really well established standards for storing motifs, but there’s a
couple of formats which are more used than others. The most important
distinction is whether the motif representation is based on instances or
on some version of PWM matrix. On of the most popular motif databases
<a class="reference external" href="http://jaspar.genereg.net">JASPAR</a> stores motifs in both formats, so
let’s look at how we can import JASPAR motifs from instances:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Motif</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">arnt</span> <span class="operator">=</span> <span class="name">Motif</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;Arnt.sites&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;jaspar-sites&quot;</span><span class="punctuation">)</span>
</pre>
<p>and from a count matrix:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">srf</span> <span class="operator">=</span> <span class="name">Motif</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;SRF.pfm&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;jaspar-pfm&quot;</span><span class="punctuation">)</span>
</pre>
<p>The <tt class="docutils literal">arnt</tt> and <tt class="docutils literal">srf</tt> motifs can both do the same things for us, but
they use different internal representations of the motif. We can tell
that by inspecting the <tt class="docutils literal">has_counts</tt> and <tt class="docutils literal">has_instances</tt> properties:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">has_instances</span>
<span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">srf</span><span class="operator">.</span><span class="name">has_instances</span>
<span class="name builtin pseudo">False</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">srf</span><span class="operator">.</span><span class="name">has_counts</span>
<span class="name builtin pseudo">True</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">srf</span><span class="operator">.</span><span class="name">counts</span>
<span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">9</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">32</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">46</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">43</span><span class="punctuation">,</span> <span class="literal number integer">15</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">],</span>
 <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">33</span><span class="punctuation">,</span> <span class="literal number integer">45</span><span class="punctuation">,</span> <span class="literal number integer">45</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">],</span>
 <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">39</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">44</span><span class="punctuation">,</span> <span class="literal number integer">43</span><span class="punctuation">],</span>
 <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">4</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">13</span><span class="punctuation">,</span> <span class="literal number integer">42</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">45</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">30</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">]}</span>
</pre>
<p>There are conversion functions, which can help us convert between
different representations:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">make_counts_from_instances</span><span class="punctuation">()</span>
<span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">8</span><span class="punctuation">,</span> <span class="literal number integer">38</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">],</span>
 <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">32</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">],</span>
 <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">40</span><span class="punctuation">],</span>
 <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">]}</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="name">srf</span><span class="operator">.</span><span class="name">make_instances_from_counts</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GGGAAAAAAAGG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()),</span>
 <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GGCCAAATAAGG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()),</span>
 <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GACCAAATAAGG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()),</span>
<span class="operator">....</span>
</pre>
<p>The important thing to remember here is that the method
<tt class="docutils literal">make_instances_from_counts()</tt> creates fake instances, because usually
there are very many possible sets of instances which give rise to the
same pwm, and if we have only the count matrix, we cannot reconstruct
the original one. This does not make any difference if we are using the
PWM as the representation of the motif, but one should be careful with
exporting instances from count-based motifs.</p>
<p>Speaking of exporting, let’s look at export functions. We can export to
fasta:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;</span><span class="name">instance0</span>
<span class="name">TATAA</span>
<span class="operator">&gt;</span><span class="name">instance1</span>
<span class="name">TATTA</span>
<span class="operator">&gt;</span><span class="name">instance2</span>
<span class="name">TATAA</span>
<span class="operator">&gt;</span><span class="name">instance3</span>
<span class="name">TATAA</span>
</pre>
<p>or to TRANSFAC-like matrix format (used by some motif processing
software)</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;transfac&quot;</span><span class="punctuation">)</span>
<span class="name">XX</span>
<span class="name">TY</span> <span class="name">Motif</span>
<span class="name">ID</span>
<span class="name">BF</span> <span class="name">undef</span>
<span class="name">P0</span> <span class="name">G</span> <span class="name">A</span> <span class="name">T</span> <span class="name">C</span>
<span class="literal number oct">01</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span> <span class="literal number integer">4</span> <span class="literal number integer">0</span>
<span class="literal number oct">02</span> <span class="literal number integer">0</span> <span class="literal number integer">4</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span>
<span class="literal number oct">03</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span> <span class="literal number integer">4</span> <span class="literal number integer">0</span>
<span class="literal number oct">04</span> <span class="literal number integer">0</span> <span class="literal number integer">3</span> <span class="literal number integer">1</span> <span class="literal number integer">0</span>
<span class="literal number oct">05</span> <span class="literal number integer">0</span> <span class="literal number integer">4</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span>
<span class="name">XX</span>
</pre>
<p>Finally, if we have internet access, we can create a
<a class="reference external" href="http://weblogo.berkeley.edu">weblogo</a>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">weblogo</span><span class="punctuation">(</span><span class="literal string">&quot;Arnt.png&quot;</span><span class="punctuation">)</span>
</pre>
<p>We should get our logo saved as a png in the specified file.</p>
</div>
</div>
<div class="section" id="id8">
<h3>14.9.2  Searching for instances</h3>
<p>The most frequent use for a motif is to find its instances in some
sequence. For the sake of this section, we will use an artificial
sequence like this:</p>
<pre class="code python literal-block">
<span class="name">test_seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TATGATGTAGTATAATATAATTATAA&quot;</span><span class="punctuation">,</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">)</span>
</pre>
<p>The simplest way to find instances, is to look for exact matches of the
true instances of the motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">seq</span> <span class="operator word">in</span> <span class="name">m</span><span class="operator">.</span><span class="name">search_instances</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">seq</span><span class="operator">.</span><span class="name">tostring</span><span class="punctuation">()</span>
<span class="operator">...</span>
<span class="literal number integer">10</span> <span class="name">TATAA</span>
<span class="literal number integer">15</span> <span class="name">TATAA</span>
<span class="literal number integer">21</span> <span class="name">TATAA</span>
</pre>
<p>We can do the same with the reverse complement (to find instances on the
complementary strand):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">seq</span> <span class="operator word">in</span> <span class="name">m</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">search_instances</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">seq</span><span class="operator">.</span><span class="name">tostring</span><span class="punctuation">()</span>
<span class="operator">...</span>
<span class="literal number integer">12</span> <span class="name">TAATA</span>
<span class="literal number integer">20</span> <span class="name">TTATA</span>
</pre>
<p>It’s just as easy to look for positions, giving rise to high log-odds
scores against our motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">score</span> <span class="operator word">in</span> <span class="name">m</span><span class="operator">.</span><span class="name">search_pwm</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">,</span><span class="name">threshold</span><span class="operator">=</span><span class="literal number float">5.0</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">score</span>
<span class="operator">...</span>
<span class="literal number integer">10</span> <span class="literal number float">8.44065060871</span>
<span class="operator">-</span><span class="literal number integer">12</span> <span class="literal number float">7.06213898545</span>
<span class="literal number integer">15</span> <span class="literal number float">8.44065060871</span>
<span class="operator">-</span><span class="literal number integer">20</span> <span class="literal number float">8.44065060871</span>
<span class="literal number integer">21</span> <span class="literal number float">8.44065060871</span>
</pre>
<p>You may notice the threshold parameter, here set arbitrarily to 5.0.
This is in <em>log</em><sub>2</sub>, so we are now looking only for words, which
are 32 times more likely to occur under the motif model than in the
background. The default threshold is 0.0, which selects everything that
looks more like the motif than the background.</p>
<p>If you want to use a less arbitrary way of selecting thresholds, you can
explore the <tt class="docutils literal">Motif.score_distribution</tt> class implementing an
distribution of scores for a given motif. Since the space for a score
distribution grows exponentially with motif length, we are using an
approximation with a given precision to keep computation cost
manageable:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sd</span> <span class="operator">=</span> <span class="name">Motif</span><span class="operator">.</span><span class="name">score_distribution</span><span class="punctuation">(</span><span class="name">m</span><span class="punctuation">,</span><span class="name">precision</span><span class="operator">=</span><span class="literal number integer">10</span><span class="operator">**</span><span class="literal number integer">4</span><span class="punctuation">)</span>
</pre>
<p>The sd object can be used to determine a number of different thresholds.</p>
<p>We can specify the requested false-positive rate (probability of
“finding” a motif instance in background generated sequence):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sd</span><span class="operator">.</span><span class="name">threshold_fpr</span><span class="punctuation">(</span><span class="literal number float">0.01</span><span class="punctuation">)</span>
<span class="literal number float">4.3535838726139886</span>
</pre>
<p>or the false-negative rate (probability of “not finding” an instance
generated from the motif):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sd</span><span class="operator">.</span><span class="name">threshold_fnr</span><span class="punctuation">(</span><span class="literal number float">0.1</span><span class="punctuation">)</span>
<span class="literal number float">0.26651713652234044</span>
</pre>
<p>or a threshold (approximately) satisfying some relation between fpr and
fnr <em>fnr</em>/<em>fpr</em>≃ <em>t</em>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sd</span><span class="operator">.</span><span class="name">threshold_balanced</span><span class="punctuation">(</span><span class="literal number integer">1000</span><span class="punctuation">)</span>
<span class="literal number float">8.4406506087056368</span>
</pre>
<p>or a threshold satisfying (roughly) the equality between the
false-positive rate and the −<em>log</em> of the information content (as used
in patser software by Hertz and Stormo).</p>
<p>For example, in case of our motif, you can get the threshold giving you
exactly the same results (for this sequence) as searching for instances
with balanced threshold with rate of 1000.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">score</span> <span class="operator word">in</span> <span class="name">m</span><span class="operator">.</span><span class="name">search_pwm</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">,</span><span class="name">threshold</span><span class="operator">=</span><span class="name">sd</span><span class="operator">.</span><span class="name">threshold_balanced</span><span class="punctuation">(</span><span class="literal number integer">1000</span><span class="punctuation">)):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">score</span>
<span class="operator">...</span>
<span class="literal number integer">10</span> <span class="literal number float">8.44065060871</span>
<span class="literal number integer">15</span> <span class="literal number float">8.44065060871</span>
<span class="operator">-</span><span class="literal number integer">20</span> <span class="literal number float">8.44065060871</span>
<span class="literal number integer">21</span> <span class="literal number float">8.44065060871</span>
</pre>
</div>
<div class="section" id="id9">
<h3>14.9.3  Comparing motifs</h3>
<p>Once we have more than one motif, we might want to compare them. For
that, we have currently three different methods of <tt class="docutils literal">Bio.Motif</tt>
objects.</p>
<p>Before we start comparing motifs, I should point out that motif
boundaries are usually quite arbitrary. This means, that we often need
to compare motifs of different lengths, so comparison needs to involve
some kind of alignment. This means, that we have to take into account
two things:</p>
<ul class="simple">
<li>alignment of motifs</li>
<li>some function to compare aligned motifs</li>
</ul>
<p>In <tt class="docutils literal">Bio.Motif</tt> we have 3 different functions for motif comparison,
which are based on the same idea behind motif alignment, but use
different functions to compare aligned motifs. Briefly speaking, we are
using ungapped alignment of PWMs and substitute the missing columns at
the beginning and end of the matrices with background distribution. All
three comparison functions are written in such a way, that they can be
interpreted as distance measures, however only one (<tt class="docutils literal">dist_dpq</tt>)
satisfies the triangle inequality. All of them return the minimal
distance and the corresponding offset between motifs.</p>
<p>To show how these functions work, let us first load another motif, which
is similar to our test motif <tt class="docutils literal">m</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ubx</span><span class="operator">=</span><span class="name">Motif</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;Ubx.pfm&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;jaspar-pfm&quot;</span><span class="punctuation">)</span>
<span class="operator">&lt;</span><span class="name">Bio</span><span class="operator">.</span><span class="name">Motif</span><span class="operator">.</span><span class="name">Motif</span><span class="operator">.</span><span class="name">Motif</span> <span class="name builtin">object</span> <span class="name">at</span> <span class="literal number hex">0xc29b90</span><span class="operator">&gt;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ubx</span><span class="operator">.</span><span class="name">consensus</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TAAT'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>The first function we’ll use to compare these motifs is based on Pearson
correlation. Since we want it to resemble a distance measure, we
actually take 1−<em>r</em>, where <em>r</em> is the Pearson correlation coefficient
(PCC):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">dist_pearson</span><span class="punctuation">(</span><span class="name">ubx</span><span class="punctuation">)</span>
<span class="punctuation">(</span><span class="literal number float">0.41740393308237722</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">)</span>
</pre>
<p>This means, that the best PCC between motif <tt class="docutils literal">m</tt> and <tt class="docutils literal">Ubx</tt> is
obtained with the following alignment:</p>
<pre class="code python literal-block">
<span class="name">bbTAAT</span>
<span class="name">TATAAb</span>
</pre>
<p>where <tt class="docutils literal">b</tt> stands for background distribution. The PCC itself is
roughly 1−0.42=0.58. If we try the reverse complement of the Ubx motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">dist_pearson</span><span class="punctuation">(</span><span class="name">ubx</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">())</span>
<span class="punctuation">(</span><span class="literal number float">0.25784180151584823</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>We can see that the PCC is better (almost 0.75), and the alignment is
also different:</p>
<pre class="code python literal-block">
<span class="name">bATTA</span>
<span class="name">TATAA</span>
</pre>
<p>There are two other functions: <tt class="docutils literal">dist_dpq</tt>, which is a true metric
(satisfying traingle inequality) based on the Kullback-Leibler
divergence</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">dist_dpq</span><span class="punctuation">(</span><span class="name">ubx</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">())</span>
<span class="punctuation">(</span><span class="literal number float">0.49292358382899853</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>and the <tt class="docutils literal">dist_product</tt> method, which is based on the product of
probabilities which can be interpreted as the probability of
independently generating the same instance by both motifs.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">dist_product</span><span class="punctuation">(</span><span class="name">ubx</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">())</span>
<span class="punctuation">(</span><span class="literal number float">0.16224587301064275</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="id10">
<h3>14.9.4  <em>De novo</em> motif finding</h3>
<p>Currently, Biopython has only limited support for <em>de novo</em> motif
finding. Namely, we support running and parsing of AlignAce and MEME.
Since the number of motif finding tools is growing rapidly,
contributions of new parsers are welcome.</p>
<div class="section" id="id11">
<h4>14.9.4.1  MEME</h4>
<p>Let’s assume, you have run MEME on sequences of your choice with your
favorite parameters and saved the output in the file <tt class="docutils literal">meme.out</tt>. You
can retrieve the motifs reported by MEME by running the following piece
of code:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">Motif</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;meme.out&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;MEME&quot;</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span>
<span class="punctuation">[</span><span class="operator">&lt;</span><span class="name">Bio</span><span class="operator">.</span><span class="name">Motif</span><span class="operator">.</span><span class="name">MEMEMotif</span><span class="operator">.</span><span class="name">MEMEMotif</span> <span class="name builtin">object</span> <span class="name">at</span> <span class="literal number hex">0xc356b0</span><span class="operator">&gt;</span><span class="punctuation">]</span>
</pre>
<p>Besides the most wanted list of motifs, the result object contains more
useful information, accessible through properties with self-explanatory
names:</p>
<ul class="simple">
<li><tt class="docutils literal">.alphabet</tt></li>
<li><tt class="docutils literal">.datafile</tt></li>
<li><tt class="docutils literal">.sequence_names</tt></li>
<li><tt class="docutils literal">.version</tt></li>
<li><tt class="docutils literal">.command</tt></li>
</ul>
<p>The motifs returned by MEMEParser can be treated exactly like regular
Motif objects (with instances), they also provide some extra
functionality, by adding additional information about the instances.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">consensus</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CTCAATCGTA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">pvalue</span>
<span class="literal number float">8.71e-07</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">sequence_name</span>
<span class="literal string">'SEQ10;'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">start</span>
<span class="literal number integer">3</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">strand</span>
<span class="literal string">'+'</span>
</pre>
</div>
<div class="section" id="id12">
<h4>14.9.4.2  AlignAce</h4>
<p>We can do very similar things with AlignACE program. Assume, you have
your output in the file <tt class="docutils literal">alignace.out</tt>. You can parse your output with
the following code:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsA</span><span class="operator">=</span><span class="name builtin">list</span><span class="punctuation">(</span><span class="name">Motif</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;alignace.out&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;AlignAce&quot;</span><span class="punctuation">))</span>
</pre>
<p>Again, your motifs behave as they should:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsA</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">consensus</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TCTACGATTGAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>In fact you can even see, that AlignAce found a very similar motif as
MEME, it is just a longer version of a reverse complement of MEME motif:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">consensus</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TACGATTGAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>If you have AlignAce installed on the same machine, you can also run it
directly from Biopython. Short example of how this can be done is shown
below (other parameters can be specified as keyword parameters):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">command</span><span class="operator">=</span><span class="literal string">&quot;/opt/bin/AlignACE&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">input_file</span><span class="operator">=</span><span class="literal string">&quot;test.fa&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Motif.Applications</span> <span class="keyword namespace">import</span> <span class="name">AlignAceCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cmd</span> <span class="operator">=</span> <span class="name">AlignAceCommandline</span><span class="punctuation">(</span><span class="name">cmd</span><span class="operator">=</span><span class="name">command</span><span class="punctuation">,</span><span class="name builtin">input</span><span class="operator">=</span><span class="name">input_file</span><span class="punctuation">,</span><span class="name">gcback</span><span class="operator">=</span><span class="literal number float">0.6</span><span class="punctuation">,</span><span class="name">numcols</span><span class="operator">=</span><span class="literal number integer">10</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span><span class="name">stderr</span><span class="operator">=</span> <span class="name">cmd</span><span class="punctuation">()</span>
</pre>
<p>Since AlignAce prints all its output to standard output, you can get to
your motifs by parsing the first part of the result:</p>
<pre class="code python literal-block">
<span class="name">motifs</span><span class="operator">=</span><span class="name builtin">list</span><span class="punctuation">(</span><span class="name">Motif</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">stdout</span><span class="punctuation">,</span><span class="literal string">&quot;AlignAce&quot;</span><span class="punctuation">))</span>
</pre>
</div>
</div>
</div>
</div>
<div class="section" id="chapter-15-cluster-analysis">
<h1>Chapter 15  Cluster analysis</h1>
<p>Cluster analysis is the grouping of items into clusters based on the
similarity of the items to each other. In bioinformatics, clustering is
widely used in gene expression data analysis to find groups of genes
with similar gene expression profiles. This may identify functionally
related genes, as well as suggest the function of presently unknown
genes.</p>
<p>The Biopython module <tt class="docutils literal">Bio.Cluster</tt> provides commonly used clustering
algorithms and was designed with the application to gene expression data
in mind. However, this module can also be used for cluster analysis of
other types of data. <tt class="docutils literal">Bio.Cluster</tt> and the underlying C Clustering
Library is described by De Hoon <em>et al.</em> [<a class="reference external" href="#dehoon2004">14</a>].</p>
<p>The following four clustering approaches are implemented in
<tt class="docutils literal">Bio.Cluster</tt>:</p>
<ul class="simple">
<li>Hierarchical clustering (pairwise centroid-, single-, complete-, and
average-linkage);</li>
<li><em>k</em>-means, <em>k</em>-medians, and <em>k</em>-medoids clustering;</li>
<li>Self-Organizing Maps;</li>
<li>Principal Component Analysis.</li>
</ul>
<div class="section" id="data-representation">
<h2>Data representation</h2>
<p>The data to be clustered are represented by a <em>n</em> × <em>m</em> Numerical Python
array <tt class="docutils literal">data</tt>. Within the context of gene expression data clustering,
typically the rows correspond to different genes whereas the columns
correspond to different experimental conditions. The clustering
algorithms in <tt class="docutils literal">Bio.Cluster</tt> can be applied both to rows (genes) and to
columns (experiments).</p>
</div>
<div class="section" id="missing-values">
<h2>Missing values</h2>
<p>Often in microarray experiments, some of the data values are missing,
which is indicated by an additional <em>n</em> × <em>m</em> Numerical Python integer
array <tt class="docutils literal">mask</tt>. If <tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, then <tt class="docutils literal">data[i,j]</tt> is missing and
is ignored in the analysis.</p>
</div>
<div class="section" id="random-number-generator">
<h2>Random number generator</h2>
<p>The <em>k</em>-means/medians/medoids clustering algorithms and Self-Organizing
Maps (SOMs) include the use of a random number generator. The uniform
random number generator in <tt class="docutils literal">Bio.Cluster</tt> is based on the algorithm by
L’Ecuyer [<a class="reference external" href="#lecuyer1988">25</a>], while random numbers following the
binomial distribution are generated using the BTPE algorithm by
Kachitvichyanukul and Schmeiser [<a class="reference external" href="#kachitvichyanukul1988">23</a>]. The
random number generator is initialized automatically during its first
call. As this random number generator uses a combination of two
multiplicative linear congruential generators, two (integer) seeds are
needed for initialization, for which we use the system-supplied random
number generator <tt class="docutils literal">rand</tt> (in the C standard library). We initialize
this generator by calling <tt class="docutils literal">srand</tt> with the epoch time in seconds, and
use the first two random numbers generated by <tt class="docutils literal">rand</tt> as seeds for the
uniform random number generator in <tt class="docutils literal">Bio.Cluster</tt>.</p>
</div>
<div class="section" id="distance-functions">
<h2>15.1  Distance functions</h2>
<p>In order to cluster items into groups based on their similarity, we
should first define what exactly we mean by <em>similar</em>. <tt class="docutils literal">Bio.Cluster</tt>
provides eight distance functions, indicated by a single character, to
measure similarity, or conversely, distance:</p>
<ul class="simple">
<li><tt class="docutils literal">'e'</tt>: Euclidean distance;</li>
<li><tt class="docutils literal">'b'</tt>: City-block distance.</li>
<li><tt class="docutils literal">'c'</tt>: Pearson correlation coefficient;</li>
<li><tt class="docutils literal">'a'</tt>: Absolute value of the Pearson correlation coefficient;</li>
<li><tt class="docutils literal">'u'</tt>: Uncentered Pearson correlation (equivalent to the cosine of
the angle between two data vectors);</li>
<li><tt class="docutils literal">'x'</tt>: Absolute uncentered Pearson correlation;</li>
<li><tt class="docutils literal">'s'</tt>: Spearman’s rank correlation;</li>
<li><tt class="docutils literal">'k'</tt>: Kendall’s τ.</li>
</ul>
<p>The first two are true distance functions that satisfy the triangle
inequality:</p>
<div class="formula">
<i>d</i><span class="symbol">(</span><u><i>u</i></u>, <u><i>v</i></u><span class="symbol">)</span> ≤ <i>d</i><span class="symbol">(</span><u><i>u</i></u>, <u><i>w</i></u><span class="symbol">)</span> + <i>d</i><span class="symbol">(</span><u><i>w</i></u>, <u><i>v</i></u><span class="symbol">)</span><span class="textrm"> for all </span><u><i>u</i></u>, <u><i>v</i></u>, <u><i>w</i></u>, 
</div>
<p>and are therefore refered to as <em>metrics</em>. In everyday language, this
means that the shortest distance between two points is a straight line.</p>
<p>The remaining six distance measures are related to the correlation
coefficient, where the distance <em>d</em> is defined in terms of the
correlation <em>r</em> by <em>d</em>=1−<em>r</em>. Note that these distance functions are
<em>semi-metrics</em> that do not satisfy the triangle inequality. For example,
for</p>
<div class="formula">
<u><i>u</i></u> = <span class="symbol">(</span>1, 0,  − 1<span class="symbol">)</span>;
</div>
<div class="formula">
<u><i>v</i></u> = <span class="symbol">(</span>1, 1, 0<span class="symbol">)</span>;
</div>
<div class="formula">
<u><i>w</i></u> = <span class="symbol">(</span>0, 1, 1<span class="symbol">)</span>;
</div>
<p>we find a Pearson distance <em>d</em>(<em>u</em>,<em>w</em>) = 1.8660, while
<em>d</em>(<em>u</em>,<em>v</em>)+<em>d</em>(<em>v</em>,<em>w</em>) = 1.6340.</p>
<div class="section" id="euclidean-distance">
<h3>Euclidean distance</h3>
<p>In <tt class="docutils literal">Bio.Cluster</tt>, we define the Euclidean distance as</p>
<div class="formula">
<i>d</i> = <span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="limits"><sup class="limit"><i>n</i></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><span class="symbol">(</span><i>x</i><sub><i>i</i></sub> − <i>y</i><sub><i>i</i></sub><span class="symbol">)</span><sup>2</sup>.
</div>
<p>Only those terms are included in the summation for which both
<em>x</em><sub>*i*</sub> and <em>y</em><sub>*i*</sub> are present, and the denominator <em>n</em>
is chosen accordingly. As the expression data <em>x</em><sub>*i*</sub> and
<em>y</em><sub>*i*</sub> are subtracted directly from each other, we should make
sure that the expression data are properly normalized when using the
Euclidean distance.</p>
</div>
<div class="section" id="city-block-distance">
<h3>City-block distance</h3>
<p>The city-block distance, alternatively known as the Manhattan distance,
is related to the Euclidean distance. Whereas the Euclidean distance
corresponds to the length of the shortest path between two points, the
city-block distance is the sum of distances along each dimension. As
gene expression data tend to have missing values, in <tt class="docutils literal">Bio.Cluster</tt> we
define the city-block distance as the sum of distances divided by the
number of dimensions:</p>
<div class="formula">
<i>d</i> = <span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="limits"><sup class="limit"><i>n</i></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><span class="symbol">|</span><i>x</i><sub><i>i</i></sub> − <i>y</i><sub><i>i</i></sub><span class="symbol">|</span>.
</div>
<p>This is equal to the distance you would have to walk between two points
in a city, where you have to walk along city blocks. As for the
Euclidean distance, the expression data are subtracted directly from
each other, and we should therefore make sure that they are properly
normalized.</p>
</div>
<div class="section" id="the-pearson-correlation-coefficient">
<h3>The Pearson correlation coefficient</h3>
<p>The Pearson correlation coefficient is defined as</p>
<div class="formula">
<i>r</i> = <span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="limits"><sup class="limit"><i>n</i></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="fraction"><span class="ignored">(</span><span class="numerator"><i>x</i><sub><i>i</i></sub> − <span class="bar"><i>x</i></span></span><span class="ignored">)/(</span><span class="denominator"><i>σ</i><sub><i>x</i></sub></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span><span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="fraction"><span class="ignored">(</span><span class="numerator"><i>y</i><sub><i>i</i></sub> − <span class="bar"><i>y</i></span></span><span class="ignored">)/(</span><span class="denominator"><i>σ</i><sub><i>y</i></sub></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>, 
</div>
<p>in which x, ȳ are the sample mean of <em>x</em> and <em>y</em> respectively, and
σ<sub>*x*</sub>, σ<sub>*y*</sub> are the sample standard deviation of <em>x</em>
and <em>y</em>. The Pearson correlation coefficient is a measure for how well a
straight line can be fitted to a scatterplot of <em>x</em> and <em>y</em>. If all the
points in the scatterplot lie on a straight line, the Pearson
correlation coefficient is either +1 or -1, depending on whether the
slope of line is positive or negative. If the Pearson correlation
coefficient is equal to zero, there is no correlation between <em>x</em> and
<em>y</em>.</p>
<p>The <em>Pearson distance</em> is then defined as</p>
<div class="formula">
<i>d</i><sub><span class="textrm">P</span></sub> ≡ 1 − <i>r</i>.
</div>
<p>As the Pearson correlation coefficient lies between -1 and 1, the
Pearson distance lies between 0 and 2.</p>
</div>
<div class="section" id="absolute-pearson-correlation">
<h3>Absolute Pearson correlation</h3>
<p>By taking the absolute value of the Pearson correlation, we find a
number between 0 and 1. If the absolute value is 1, all the points in
the scatter plot lie on a straight line with either a positive or a
negative slope. If the absolute value is equal to zero, there is no
correlation between <em>x</em> and <em>y</em>.</p>
<p>The corresponding distance is defined as</p>
<div class="formula">
<i>d</i><sub><span class="textrm"></span><i>A</i></sub> ≡ 1 − <span class="symbol">|</span><i>r</i><span class="symbol">|</span>, 
</div>
<p>where <em>r</em> is the Pearson correlation coefficient. As the absolute value
of the Pearson correlation coefficient lies between 0 and 1, the
corresponding distance lies between 0 and 1 as well.</p>
<p>In the context of gene expression experiments, the absolute correlation
is equal to 1 if the gene expression profiles of two genes are either
exactly the same or exactly opposite. The absolute correlation
coefficient should therefore be used with care.</p>
</div>
<div class="section" id="uncentered-correlation-cosine-of-the-angle">
<h3>Uncentered correlation (cosine of the angle)</h3>
<p>In some cases, it may be preferable to use the <em>uncentered correlation</em>
instead of the regular Pearson correlation coefficient. The uncentered
correlation is defined as</p>
<div class="formula">
<i>r</i><sub><span class="textrm"></span><i>U</i></sub> = <span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="limits"><sup class="limit"><i>n</i></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="fraction"><span class="ignored">(</span><span class="numerator"><i>x</i><sub><i>i</i></sub></span><span class="ignored">)/(</span><span class="denominator"><i>σ</i><span class="scripts"><sup class="script">(0)</sup><sub class="script"><i>x</i></sub></span></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span><span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="fraction"><span class="ignored">(</span><span class="numerator"><i>y</i><sub><i>i</i></sub></span><span class="ignored">)/(</span><span class="denominator"><i>σ</i><span class="scripts"><sup class="script">(0)</sup><sub class="script"><i>y</i></sub></span></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>, 
</div>
<p>where</p>
<div class="formula">
<span class="environment"><span class="arrayrow">
<span class="arraycell align-r">
<i>σ</i><span class="scripts"><sup class="script">(0)</sup><sub class="script"><i>x</i></sub></span>
</span>
<span class="arraycell align-c">
 = 
</span>
<span class="arraycell align-l">
<span class="sqrt"><span class="radical">√</span><span class="ignored">(</span><span class="root"><span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="limits"><sup class="limit"><i>n</i></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><i>x</i><span class="scripts"><sup class="script">2</sup><sub class="script"><i>i</i></sub></span></span><span class="ignored">)</span></span>;
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-r">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-l">
 
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-r">
<i>σ</i><span class="scripts"><sup class="script">(0)</sup><sub class="script"><i>y</i></sub></span>
</span>
<span class="arraycell align-c">
 = 
</span>
<span class="arraycell align-l">
<span class="sqrt"><span class="radical">√</span><span class="ignored">(</span><span class="root"><span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="limits"><sup class="limit"><i>n</i></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><i>y</i><span class="scripts"><sup class="script">2</sup><sub class="script"><i>i</i></sub></span></span><span class="ignored">)</span></span>.
</span>

</span>
</span>
</div>
<p>This is the same expression as for the regular Pearson correlation
coefficient, except that the sample means x, ȳ are set equal to zero.
The uncentered correlation may be appropriate if there is a zero
reference state. For instance, in the case of gene expression data given
in terms of log-ratios, a log-ratio equal to zero corresponds to the
green and red signal being equal, which means that the experimental
manipulation did not affect the gene expression.</p>
<p>The distance corresponding to the uncentered correlation coefficient is
defined as</p>
<div class="formula">
<i>d</i><sub><span class="mbox">U</span></sub> ≡ 1 − <i>r</i><sub><span class="mbox">U</span></sub>, 
</div>
<p>where <em>r</em><sub>U</sub> is the uncentered correlation. As the uncentered
correlation coefficient lies between -1 and 1, the corresponding
distance lies between 0 and 2.</p>
<p>The uncentered correlation is equal to the cosine of the angle of the
two data vectors in <em>n</em>-dimensional space, and is often referred to as
such.</p>
</div>
<div class="section" id="absolute-uncentered-correlation">
<h3>Absolute uncentered correlation</h3>
<p>As for the regular Pearson correlation, we can define a distance measure
using the absolute value of the uncentered correlation:</p>
<div class="formula">
<i>d</i><sub><span class="mbox">AU</span></sub> ≡ 1 − <span class="symbol">|</span><i>r</i><sub><span class="mbox">U</span></sub><span class="symbol">|</span>, 
</div>
<p>where <em>r</em><sub>U</sub> is the uncentered correlation coefficient. As the
absolute value of the uncentered correlation coefficient lies between 0
and 1, the corresponding distance lies between 0 and 1 as well.</p>
<p>Geometrically, the absolute value of the uncentered correlation is equal
to the cosine between the supporting lines of the two data vectors
(i.e., the angle without taking the direction of the vectors into
consideration).</p>
</div>
<div class="section" id="spearman-rank-correlation">
<h3>Spearman rank correlation</h3>
<p>The Spearman rank correlation is an example of a non-parametric
similarity measure, and tends to be more robust against outliers than
the Pearson correlation.</p>
<p>To calculate the Spearman rank correlation, we replace each data value
by their rank if we would order the data in each vector by their value.
We then calculate the Pearson correlation between the two rank vectors
instead of the data vectors.</p>
<p>As in the case of the Pearson correlation, we can define a distance
measure corresponding to the Spearman rank correlation as</p>
<div class="formula">
<i>d</i><sub><span class="mbox">S</span></sub> ≡ 1 − <i>r</i><sub><span class="mbox">S</span></sub>, 
</div>
<p>where <em>r</em><sub>S</sub> is the Spearman rank correlation.</p>
</div>
<div class="section" id="kendalls">
<h3>Kendall’s τ</h3>
<p>Kendall’s τ is another example of a non-parametric similarity measure.
It is similar to the Spearman rank correlation, but instead of the ranks
themselves only the relative ranks are used to calculate τ (see Snedecor
&amp; Cochran [<a class="reference external" href="#snedecor1989">29</a>]).</p>
<p>We can define a distance measure corresponding to Kendall’s τ as</p>
<div class="formula">
<i>d</i><sub><span class="mbox">K</span></sub> ≡ 1 − <i>τ</i>.
</div>
<p>As Kendall’s τ is always between -1 and 1, the corresponding distance
will be between 0 and 2.</p>
</div>
<div class="section" id="weighting">
<h3>Weighting</h3>
<p>For most of the distance functions available in <tt class="docutils literal">Bio.Cluster</tt>, a
weight vector can be applied. The weight vector contains weights for the
items in the data vector. If the weight for item <em>i</em> is <em>w</em><sub>*i*</sub>,
then that item is treated as if it occurred <em>w</em><sub>*i*</sub> times in the
data. The weight do not have to be integers. For the Spearman rank
correlation and Kendall’s τ, weights do not have a well-defined meaning
and are therefore not implemented.</p>
</div>
<div class="section" id="calculating-the-distance-matrix">
<h3>Calculating the distance matrix</h3>
<p>The distance matrix is a square matrix with all pairwise distances
between the items in <tt class="docutils literal">data</tt>, and can be calculated by the function
<tt class="docutils literal">distancematrix</tt> in the <tt class="docutils literal">Bio.Cluster</tt> module:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">distancematrix</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">matrix</span> <span class="operator">=</span> <span class="name">distancematrix</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>where the following arguments are defined:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt> (required)</dt>
<dd><p class="first last">Array containing the data for the items.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">Array of integers showing which data are missing. If
<tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, then <tt class="docutils literal">data[i,j]</tt> is missing. If <tt class="docutils literal"><span class="pre">mask==None</span></tt>,
then all data are present.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">weight</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">The weights to be used when calculating distances. If
<tt class="docutils literal"><span class="pre">weight==None</span></tt>, then equal weights are assumed.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">Determines if the distances between the rows of <tt class="docutils literal">data</tt> are to be
calculated (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>), or between the columns of <tt class="docutils literal">data</tt>
(<tt class="docutils literal"><span class="pre">transpose==1</span></tt>).</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (default: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">Defines the distance function to be used (see
<a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
</ul>
<p>To save memory, the distance matrix is returned as a list of 1D arrays.
The number of columns in each row is equal to the row number. Hence, the
first row has zero elements. An example of the return value is</p>
<pre class="code python literal-block">
<span class="punctuation">[</span><span class="name">array</span><span class="punctuation">([]),</span>
 <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">1.</span><span class="punctuation">]),</span>
 <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">7.</span><span class="punctuation">,</span> <span class="literal number float">3.</span><span class="punctuation">]),</span>
 <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">4.</span><span class="punctuation">,</span> <span class="literal number float">2.</span><span class="punctuation">,</span> <span class="literal number float">6.</span><span class="punctuation">])]</span>
</pre>
<p>This corresponds to the distance matrix</p>
<div class="formula">
<span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎜</span></span><span class="arrayrow"><span class="bracket align-left">⎜</span></span><span class="arrayrow"><span class="bracket align-left">⎜</span></span><span class="arrayrow"><span class="bracket align-left">⎜</span></span><span class="arrayrow"><span class="bracket align-left">⎜</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="array"><span class="arrayrow">
<span class="arraycell align-c">
0
</span>
<span class="arraycell align-c">
1
</span>
<span class="arraycell align-c">
7
</span>
<span class="arraycell align-c">
4
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-c">
1
</span>
<span class="arraycell align-c">
0
</span>
<span class="arraycell align-c">
3
</span>
<span class="arraycell align-c">
2
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-c">
7
</span>
<span class="arraycell align-c">
3
</span>
<span class="arraycell align-c">
0
</span>
<span class="arraycell align-c">
6
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-c">
4
</span>
<span class="arraycell align-c">
2
</span>
<span class="arraycell align-c">
6
</span>
<span class="arraycell align-c">
0
</span>

</span>
</span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎟</span></span><span class="arrayrow"><span class="bracket align-right">⎟</span></span><span class="arrayrow"><span class="bracket align-right">⎟</span></span><span class="arrayrow"><span class="bracket align-right">⎟</span></span><span class="arrayrow"><span class="bracket align-right">⎟</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>.
</div>
</div>
</div>
<div class="section" id="calculating-cluster-properties">
<h2>15.2  Calculating cluster properties</h2>
<div class="section" id="calculating-the-cluster-centroids">
<h3>Calculating the cluster centroids</h3>
<p>The centroid of a cluster can be defined either as the mean or as the
median of each dimension over all cluster items. The function
<tt class="docutils literal">clustercentroids</tt> in <tt class="docutils literal">Bio.Cluster</tt> can be used to calculate either:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">clustercentroids</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cdata</span><span class="punctuation">,</span> <span class="name">cmask</span> <span class="operator">=</span> <span class="name">clustercentroids</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>where the following arguments are defined:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt> (required)</dt>
<dd><p class="first last">Array containing the data for the items.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">Array of integers showing which data are missing. If
<tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, then <tt class="docutils literal">data[i,j]</tt> is missing. If <tt class="docutils literal"><span class="pre">mask==None</span></tt>,
then all data are present.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">clusterid</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">Vector of integers showing to which cluster each item belongs. If
<tt class="docutils literal">clusterid</tt> is <tt class="docutils literal">None</tt>, then all items are assumed to belong to
the same cluster.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (default: <tt class="docutils literal">'a'</tt>)</dt>
<dd><p class="first last">Specifies whether the arithmetic mean (<tt class="docutils literal"><span class="pre">method=='a'</span></tt>) or the
median (<tt class="docutils literal"><span class="pre">method=='m'</span></tt>) is used to calculate the cluster center.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">Determines if the centroids of the rows of <tt class="docutils literal">data</tt> are to be
calculated (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>), or the centroids of the columns of
<tt class="docutils literal">data</tt> (<tt class="docutils literal"><span class="pre">transpose==1</span></tt>).</p>
</dd>
</dl>
</li>
</ul>
<p>This function returns the tuple <tt class="docutils literal">(cdata, cmask)</tt>. The centroid data
are stored in the 2D Numerical Python array <tt class="docutils literal">cdata</tt>, with missing data
indicated by the 2D Numerical Python integer array <tt class="docutils literal">cmask</tt>. The
dimensions of these arrays are (number of clusters, number of columns)
if <tt class="docutils literal">transpose</tt> is <tt class="docutils literal">0</tt>, or (number of rows, number of clusters) if
<tt class="docutils literal">transpose</tt> is <tt class="docutils literal">1</tt>. Each row (if <tt class="docutils literal">transpose</tt> is <tt class="docutils literal">0</tt>) or column
(if <tt class="docutils literal">transpose</tt> is <tt class="docutils literal">1</tt>) contains the averaged data corresponding to
the centroid of each cluster.</p>
</div>
<div class="section" id="calculating-the-distance-between-clusters">
<h3>Calculating the distance between clusters</h3>
<p>Given a distance function between <em>items</em>, we can define the distance
between two <em>clusters</em> in several ways. The distance between the
arithmetic means of the two clusters is used in pairwise
centroid-linkage clustering and in <em>k</em>-means clustering. In <em>k</em>-medoids
clustering, the distance between the medians of the two clusters is used
instead. The shortest pairwise distance between items of the two
clusters is used in pairwise single-linkage clustering, while the
longest pairwise distance is used in pairwise maximum-linkage
clustering. In pairwise average-linkage clustering, the distance between
two clusters is defined as the average over the pairwise distances.</p>
<p>To calculate the distance between two clusters, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">clusterdistance</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">distance</span> <span class="operator">=</span> <span class="name">clusterdistance</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>where the following arguments are defined:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt> (required)</dt>
<dd><p class="first last">Array containing the data for the items.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">Array of integers showing which data are missing. If
<tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, then <tt class="docutils literal">data[i,j]</tt> is missing. If <tt class="docutils literal"><span class="pre">mask==None</span></tt>,
then all data are present.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">weight</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">The weights to be used when calculating distances. If
<tt class="docutils literal"><span class="pre">weight==None</span></tt>, then equal weights are assumed.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">index1</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">A list containing the indices of the items belonging to the first
cluster. A cluster containing only one item <em>i</em> can be represented
either as a list <tt class="docutils literal">[i]</tt>, or as an integer <tt class="docutils literal">i</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">index2</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">A list containing the indices of the items belonging to the second
cluster. A cluster containing only one items <em>i</em> can be represented
either as a list <tt class="docutils literal">[i]</tt>, or as an integer <tt class="docutils literal">i</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (default: <tt class="docutils literal">'a'</tt>)</dt>
<dd><p class="first last">Specifies how the distance between clusters is defined:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal">'a'</tt>: Distance between the two cluster centroids (arithmetic
mean);</li>
<li><tt class="docutils literal">'m'</tt>: Distance between the two cluster centroids (median);</li>
<li><tt class="docutils literal">'s'</tt>: Shortest pairwise distance between items in the two
clusters;</li>
<li><tt class="docutils literal">'x'</tt>: Longest pairwise distance between items in the two
clusters;</li>
<li><tt class="docutils literal">'v'</tt>: Average over the pairwise distances between items in the
two clusters.</li>
</ul>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (default: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">Defines the distance function to be used (see
<a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">If <tt class="docutils literal"><span class="pre">transpose==0</span></tt>, calculate the distance between the rows of
<tt class="docutils literal">data</tt>. If <tt class="docutils literal"><span class="pre">transpose==1</span></tt>, calculate the distance between the
columns of <tt class="docutils literal">data</tt>.</p>
</dd>
</dl>
</li>
</ul>
</div>
</div>
<div class="section" id="partitioning-algorithms">
<h2>15.3  Partitioning algorithms</h2>
<p>Partitioning algorithms divide items into <em>k</em> clusters such that the sum
of distances over the items to their cluster centers is minimal. The
number of clusters <em>k</em> is specified by the user. Three partitioning
algorithms are available in <tt class="docutils literal">Bio.Cluster</tt>:</p>
<ul class="simple">
<li><em>k</em>-means clustering</li>
<li><em>k</em>-medians clustering</li>
<li><em>k</em>-medoids clustering</li>
</ul>
<p>These algorithms differ in how the cluster center is defined. In
<em>k</em>-means clustering, the cluster center is defined as the mean data
vector averaged over all items in the cluster. Instead of the mean, in
<em>k</em>-medians clustering the median is calculated for each dimension in
the data vector. Finally, in <em>k</em>-medoids clustering the cluster center
is defined as the item which has the smallest sum of distances to the
other items in the cluster. This clustering algorithm is suitable for
cases in which the distance matrix is known but the original data matrix
is not available, for example when clustering proteins based on their
structural similarity.</p>
<p>The expectation-maximization (EM) algorithm is used to find this
partitioning into <em>k</em> groups. In the initialization of the EM algorithm,
we randomly assign items to clusters. To ensure that no empty clusters
are produced, we use the binomial distribution to randomly choose the
number of items in each cluster to be one or more. We then randomly
permute the cluster assignments to items such that each item has an
equal probability to be in any cluster. Each cluster is thus guaranteed
to contain at least one item.</p>
<p>We then iterate:</p>
<ul class="simple">
<li>Calculate the centroid of each cluster, defined as either the mean,
the median, or the medoid of the cluster;</li>
<li>Calculate the distances of each item to the cluster centers;</li>
<li>For each item, determine which cluster centroid is closest;</li>
<li>Reassign each item to its closest cluster, or stop the iteration if
no further item reassignments take place.</li>
</ul>
<p>To avoid clusters becoming empty during the iteration, in <em>k</em>-means and
<em>k</em>-medians clustering the algorithm keeps track of the number of items
in each cluster, and prohibits the last remaining item in a cluster from
being reassigned to a different cluster. For <em>k</em>-medoids clustering,
such a check is not needed, as the item that functions as the cluster
centroid has a zero distance to itself, and will therefore never be
closer to a different cluster.</p>
<p>As the initial assignment of items to clusters is done randomly, usually
a different clustering solution is found each time the EM algorithm is
executed. To find the optimal clustering solution, the <em>k</em>-means
algorithm is repeated many times, each time starting from a different
initial random clustering. The sum of distances of the items to their
cluster center is saved for each run, and the solution with the smallest
value of this sum will be returned as the overall clustering solution.</p>
<p>How often the EM algorithm should be run depends on the number of items
being clustered. As a rule of thumb, we can consider how often the
optimal solution was found; this number is returned by the partitioning
algorithms as implemented in this library. If the optimal solution was
found many times, it is unlikely that better solutions exist than the
one that was found. However, if the optimal solution was found only
once, there may well be other solutions with a smaller within-cluster
sum of distances. If the number of items is large (more than several
hundreds), it may be difficult to find the globally optimal solution.</p>
<p>The EM algorithm terminates when no further reassignments take place. We
noticed that for some sets of initial cluster assignments, the EM
algorithm fails to converge due to the same clustering solution
reappearing periodically after a small number of iteration steps. We
therefore check for the occurrence of such periodic solutions during the
iteration. After a given number of iteration steps, the current
clustering result is saved as a reference. By comparing the clustering
result after each subsequent iteration step to the reference state, we
can determine if a previously encountered clustering result is found. In
such a case, the iteration is halted. If after a given number of
iterations the reference state has not yet been encountered, the current
clustering solution is saved to be used as the new reference state.
Initially, ten iteration steps are executed before resaving the
reference state. This number of iteration steps is doubled each time, to
ensure that periodic behavior with longer periods can also be detected.</p>
<div class="section" id="k-means-and-k-medians">
<h3><em>k</em>-means and <em>k</em>-medians</h3>
<p>The <em>k</em>-means and <em>k</em>-medians algorithms are implemented as the function
<tt class="docutils literal">kcluster</tt> in <tt class="docutils literal">Bio.Cluster</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">kcluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clusterid</span><span class="punctuation">,</span> <span class="name">error</span><span class="punctuation">,</span> <span class="name">nfound</span> <span class="operator">=</span> <span class="name">kcluster</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>where the following arguments are defined:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt> (required)</dt>
<dd><p class="first last">Array containing the data for the items.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">nclusters</tt> (default: <tt class="docutils literal">2</tt>)</dt>
<dd><p class="first last">The number of clusters <em>k</em>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">Array of integers showing which data are missing. If
<tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, then <tt class="docutils literal">data[i,j]</tt> is missing. If <tt class="docutils literal"><span class="pre">mask==None</span></tt>,
then all data are present.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">weight</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">The weights to be used when calculating distances. If
<tt class="docutils literal"><span class="pre">weight==None</span></tt>, then equal weights are assumed.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">Determines if rows (<tt class="docutils literal">transpose</tt> is <tt class="docutils literal">0</tt>) or columns
(<tt class="docutils literal">transpose</tt> is <tt class="docutils literal">1</tt>) are to be clustered.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">npass</tt> (default: <tt class="docutils literal">1</tt>)</dt>
<dd><p class="first last">The number of times the <em>k</em>-means/-medians clustering algorithm is
performed, each time with a different (random) initial condition. If
<tt class="docutils literal">initialid</tt> is given, the value of <tt class="docutils literal">npass</tt> is ignored and the
clustering algorithm is run only once, as it behaves
deterministically in that case.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (default: <tt class="docutils literal">a</tt>)</dt>
<dd><p class="first last">describes how the center of a cluster is found:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">method=='a'</span></tt>: arithmetic mean (<em>k</em>-means clustering);</li>
<li><tt class="docutils literal"><span class="pre">method=='m'</span></tt>: median (<em>k</em>-medians clustering).</li>
</ul>
<p>For other values of <tt class="docutils literal">method</tt>, the arithmetic mean is used.</p>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (default: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">Defines the distance function to be used (see
<a class="reference external" href="#sec:distancefunctions">15.1</a>). Whereas all eight distance
measures are accepted by <tt class="docutils literal">kcluster</tt>, from a theoretical viewpoint
it is best to use the Euclidean distance for the <em>k</em>-means algorithm,
and the city-block distance for <em>k</em>-medians.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">initialid</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">Specifies the initial clustering to be used for the EM algorithm. If
<tt class="docutils literal"><span class="pre">initialid==None</span></tt>, then a different random initial clustering is
used for each of the <tt class="docutils literal">npass</tt> runs of the EM algorithm. If
<tt class="docutils literal">initialid</tt> is not <tt class="docutils literal">None</tt>, then it should be equal to a 1D array
containing the cluster number (between <tt class="docutils literal">0</tt> and <tt class="docutils literal"><span class="pre">nclusters-1</span></tt>) for
each item. Each cluster should contain at least one item. With the
initial clustering specified, the EM algorithm is deterministic.</p>
</dd>
</dl>
</li>
</ul>
<p>This function returns a tuple <tt class="docutils literal">(clusterid, error, nfound)</tt>, where
<tt class="docutils literal">clusterid</tt> is an integer array containing the number of the cluster
to which each row or cluster was assigned, <tt class="docutils literal">error</tt> is the
within-cluster sum of distances for the optimal clustering solution, and
<tt class="docutils literal">nfound</tt> is the number of times this optimal solution was found.</p>
</div>
<div class="section" id="k-medoids-clustering">
<h3><em>k</em>-medoids clustering</h3>
<p>The <tt class="docutils literal">kmedoids</tt> routine performs <em>k</em>-medoids clustering on a given set
of items, using the distance matrix and the number of clusters passed by
the user:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">kmedoids</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clusterid</span><span class="punctuation">,</span> <span class="name">error</span><span class="punctuation">,</span> <span class="name">nfound</span> <span class="operator">=</span> <span class="name">kmedoids</span><span class="punctuation">(</span><span class="name">distance</span><span class="punctuation">)</span>
</pre>
<p>where the following arguments are defined: , nclusters=2, npass=1,
initialid=None)|</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">distance</tt> (required)</dt>
<dd><p class="first last">The matrix containing the distances between the items; this matrix
can be specified in three ways:</p>
</dd>
</dl>
<ul>
<li><p class="first">as a 2D Numerical Python array (in which only the left-lower part
of the array will be accessed):</p>
<pre class="code python literal-block">
<span class="name">distance</span> <span class="operator">=</span> <span class="name">array</span><span class="punctuation">([[</span><span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">1.1</span><span class="punctuation">,</span> <span class="literal number float">2.3</span><span class="punctuation">],</span>
                  <span class="punctuation">[</span><span class="literal number float">1.1</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">],</span>
                  <span class="punctuation">[</span><span class="literal number float">2.3</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">]])</span>
</pre>
</li>
<li><p class="first">as a 1D Numerical Python array containing consecutively the
distances in the left-lower part of the distance matrix:</p>
<pre class="code python literal-block">
<span class="name">distance</span> <span class="operator">=</span> <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">1.1</span><span class="punctuation">,</span> <span class="literal number float">2.3</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">])</span>
</pre>
</li>
<li><p class="first">as a list containing the rows of the left-lower part of the
distance matrix:</p>
<pre class="code python literal-block">
<span class="name">distance</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">array</span><span class="punctuation">([]</span><span class="operator">|</span><span class="punctuation">,</span>
            <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">1.1</span><span class="punctuation">]),</span>
            <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">2.3</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">])</span>
           <span class="punctuation">]</span>
</pre>
</li>
</ul>
<p>These three expressions correspond to the same distance matrix.</p>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">nclusters</tt> (default: <tt class="docutils literal">2</tt>)</dt>
<dd><p class="first last">The number of clusters <em>k</em>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">npass</tt> (default: <tt class="docutils literal">1</tt>)</dt>
<dd><p class="first last">The number of times the <em>k</em>-medoids clustering algorithm is
performed, each time with a different (random) initial condition. If
<tt class="docutils literal">initialid</tt> is given, the value of <tt class="docutils literal">npass</tt> is ignored, as the
clustering algorithm behaves deterministically in that case.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">initialid</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">Specifies the initial clustering to be used for the EM algorithm. If
<tt class="docutils literal"><span class="pre">initialid==None</span></tt>, then a different random initial clustering is
used for each of the <tt class="docutils literal">npass</tt> runs of the EM algorithm. If
<tt class="docutils literal">initialid</tt> is not <tt class="docutils literal">None</tt>, then it should be equal to a 1D array
containing the cluster number (between <tt class="docutils literal">0</tt> and <tt class="docutils literal"><span class="pre">nclusters-1</span></tt>) for
each item. Each cluster should contain at least one item. With the
initial clustering specified, the EM algorithm is deterministic.</p>
</dd>
</dl>
</li>
</ul>
<p>This function returns a tuple <tt class="docutils literal">(clusterid, error, nfound)</tt>, where
<tt class="docutils literal">clusterid</tt> is an array containing the number of the cluster to which
each item was assigned, <tt class="docutils literal">error</tt> is the within-cluster sum of distances
for the optimal <em>k</em>-medoids clustering solution, and <tt class="docutils literal">nfound</tt> is the
number of times the optimal solution was found. Note that the cluster
number in <tt class="docutils literal">clusterid</tt> is defined as the item number of the item
representing the cluster centroid.</p>
</div>
</div>
<div class="section" id="hierarchical-clustering">
<h2>15.4  Hierarchical clustering</h2>
<p>Hierarchical clustering methods are inherently different from the
<em>k</em>-means clustering method. In hierarchical clustering, the similarity
in the expression profile between genes or experimental conditions are
represented in the form of a tree structure. This tree structure can be
shown graphically by programs such as Treeview and Java Treeview, which
has contributed to the popularity of hierarchical clustering in the
analysis of gene expression data.</p>
<p>The first step in hierarchical clustering is to calculate the distance
matrix, specifying all the distances between the items to be clustered.
Next, we create a node by joining the two closest items. Subsequent
nodes are created by pairwise joining of items or nodes based on the
distance between them, until all items belong to the same node. A tree
structure can then be created by retracing which items and nodes were
merged. Unlike the EM algorithm, which is used in <em>k</em>-means clustering,
the complete process of hierarchical clustering is deterministic.</p>
<p>Several flavors of hierarchical clustering exist, which differ in how
the distance between subnodes is defined in terms of their members. In
<tt class="docutils literal">Bio.Cluster</tt>, pairwise single, maximum, average, and centroid linkage
are available.</p>
<ul class="simple">
<li>In pairwise single-linkage clustering, the distance between two nodes
is defined as the shortest distance among the pairwise distances
between the members of the two nodes.</li>
<li>In pairwise maximum-linkage clustering, alternatively known as
pairwise complete-linkage clustering, the distance between two nodes
is defined as the longest distance among the pairwise distances
between the members of the two nodes.</li>
<li>In pairwise average-linkage clustering, the distance between two
nodes is defined as the average over all pairwise distances between
the items of the two nodes.</li>
<li>In pairwise centroid-linkage clustering, the distance between two
nodes is defined as the distance between their centroids. The
centroids are calculated by taking the mean over all the items in a
cluster. As the distance from each newly formed node to existing
nodes and items need to be calculated at each step, the computing
time of pairwise centroid-linkage clustering may be significantly
longer than for the other hierarchical clustering methods. Another
peculiarity is that (for a distance measure based on the Pearson
correlation), the distances do not necessarily increase when going up
in the clustering tree, and may even decrease. This is caused by an
inconsistency between the centroid calculation and the distance
calculation when using the Pearson correlation: Whereas the Pearson
correlation effectively normalizes the data for the distance
calculation, no such normalization occurs for the centroid
calculation.</li>
</ul>
<p>For pairwise single-, complete-, and average-linkage clustering, the
distance between two nodes can be found directly from the distances
between the individual items. Therefore, the clustering algorithm does
not need access to the original gene expression data, once the distance
matrix is known. For pairwise centroid-linkage clustering, however, the
centroids of newly formed subnodes can only be calculated from the
original data and not from the distance matrix.</p>
<p>The implementation of pairwise single-linkage hierarchical clustering is
based on the SLINK algorithm (R. Sibson, 1973), which is much faster and
more memory-efficient than a straightforward implementation of pairwise
single-linkage clustering. The clustering result produced by this
algorithm is identical to the clustering solution found by the
conventional single-linkage algorithm. The single-linkage hierarchical
clustering algorithm implemented in this library can be used to cluster
large gene expression data sets, for which conventional hierarchical
clustering algorithms fail due to excessive memory requirements and
running time.</p>
<div class="section" id="representing-a-hierarchical-clustering-solution">
<h3>Representing a hierarchical clustering solution</h3>
<p>The result of hierarchical clustering consists of a tree of nodes, in
which each node joins two items or subnodes. Usually, we are not only
interested in which items or subnodes are joined at each node, but also
in their similarity (or distance) as they are joined. To store one node
in the hierarchical clustering tree, we make use of the class <tt class="docutils literal">Node</tt>,
which defined in <tt class="docutils literal">Bio.Cluster</tt>. An instance of <tt class="docutils literal">Node</tt> has three
attributes:</p>
<ul class="simple">
<li><tt class="docutils literal">left</tt></li>
<li><tt class="docutils literal">right</tt></li>
<li><tt class="docutils literal">distance</tt></li>
</ul>
<p>Here, <tt class="docutils literal">left</tt> and <tt class="docutils literal">right</tt> are integers referring to the two items or
subnodes that are joined at this node, and <tt class="docutils literal">distance</tt> is the distance
between them. The items being clustered are numbered from 0 to (number
of items − 1), while clusters are numbered from -1 to −(number of
items−1). Note that the number of nodes is one less than the number of
items.</p>
<p>To create a new <tt class="docutils literal">Node</tt> object, we need to specify <tt class="docutils literal">left</tt> and
<tt class="docutils literal">right</tt>; <tt class="docutils literal">distance</tt> is optional.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">Node</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number integer">3</span><span class="punctuation">)</span>
<span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">):</span> <span class="literal number integer">0</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number integer">3</span><span class="punctuation">,</span><span class="literal number float">0.91</span><span class="punctuation">)</span>
<span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">):</span> <span class="literal number float">0.91</span>
</pre>
<p>The attributes <tt class="docutils literal">left</tt>, <tt class="docutils literal">right</tt>, and <tt class="docutils literal">distance</tt> of an existing
<tt class="docutils literal">Node</tt> object can be modified directly:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">node</span> <span class="operator">=</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">4</span><span class="punctuation">,</span><span class="literal number integer">5</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">node</span><span class="operator">.</span><span class="name">left</span> <span class="operator">=</span> <span class="literal number integer">6</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">node</span><span class="operator">.</span><span class="name">right</span> <span class="operator">=</span> <span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">node</span><span class="operator">.</span><span class="name">distance</span> <span class="operator">=</span> <span class="literal number float">0.73</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">node</span>
<span class="punctuation">(</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">):</span> <span class="literal number float">0.73</span>
</pre>
<p>An error is raised if <tt class="docutils literal">left</tt> and <tt class="docutils literal">right</tt> are not integers, or if
<tt class="docutils literal">distance</tt> cannot be converted to a floating-point value.</p>
<p>The Python class <tt class="docutils literal">Tree</tt> represents a full hierarchical clustering
solution. A <tt class="docutils literal">Tree</tt> object can be created from a list of <tt class="docutils literal">Node</tt>
objects:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">Node</span><span class="punctuation">,</span> <span class="name">Tree</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nodes</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number float">0.2</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="literal number integer">3</span><span class="punctuation">,</span><span class="literal number float">0.5</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number integer">4</span><span class="punctuation">,</span><span class="literal number float">0.6</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="operator">-</span><span class="literal number integer">3</span><span class="punctuation">,</span><span class="literal number float">0.9</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Tree</span><span class="punctuation">(</span><span class="name">nodes</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">tree</span>
<span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">):</span> <span class="literal number float">0.2</span>
<span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">):</span> <span class="literal number float">0.5</span>
<span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">4</span><span class="punctuation">):</span> <span class="literal number float">0.6</span>
<span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">3</span><span class="punctuation">):</span> <span class="literal number float">0.9</span>
</pre>
<p>The <tt class="docutils literal">Tree</tt> initializer checks if the list of nodes is a valid
hierarchical clustering result:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nodes</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number float">0.2</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number float">0.5</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Tree</span><span class="punctuation">(</span><span class="name">nodes</span><span class="punctuation">)</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
  <span class="name">File</span> <span class="literal string">&quot;&lt;stdin&gt;&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="error">?</span>
<span class="name exception">ValueError</span><span class="punctuation">:</span> <span class="name">Inconsistent</span> <span class="name">tree</span>
</pre>
<p>Individual nodes in a <tt class="docutils literal">Tree</tt> object can be accessed using square
brackets:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nodes</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number float">0.2</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number float">0.5</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Tree</span><span class="punctuation">(</span><span class="name">nodes</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">):</span> <span class="literal number float">0.2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">):</span> <span class="literal number float">0.5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">):</span> <span class="literal number float">0.5</span>
</pre>
<p>As a <tt class="docutils literal">Tree</tt> object is read-only, we cannot change individual nodes in
a <tt class="docutils literal">Tree</tt> object. However, we can convert the tree to a list of nodes,
modify this list, and create a new tree from this list:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Tree</span><span class="punctuation">([</span><span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number float">0.1</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number float">0.5</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number integer">3</span><span class="punctuation">,</span><span class="literal number float">0.9</span><span class="punctuation">)])</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">tree</span>
<span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">):</span> <span class="literal number float">0.1</span>
<span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">):</span> <span class="literal number float">0.5</span>
<span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">):</span> <span class="literal number float">0.9</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nodes</span> <span class="operator">=</span> <span class="name">tree</span><span class="punctuation">[:]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nodes</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number float">0.2</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nodes</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">left</span> <span class="operator">=</span> <span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Tree</span><span class="punctuation">(</span><span class="name">nodes</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">tree</span>
<span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">):</span> <span class="literal number float">0.2</span>
<span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">):</span> <span class="literal number float">0.5</span>
<span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">):</span> <span class="literal number float">0.9</span>
</pre>
<p>This guarantees that any <tt class="docutils literal">Tree</tt> object is always well-formed.</p>
<p>To display a hierarchical clustering solution with visualization
programs such as Java Treeview, it is better to scale all node distances
such that they are between zero and one. This can be accomplished by
calling the <tt class="docutils literal">scale</tt> method on an existing <tt class="docutils literal">Tree</tt> object:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">scale</span><span class="punctuation">()</span>
</pre>
<p>This method takes no arguments, and returns <tt class="docutils literal">None</tt>.</p>
<p>After hierarchical clustering, the items can be grouped into <em>k</em>
clusters based on the tree structure stored in the <tt class="docutils literal">Tree</tt> object by
cutting the tree:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clusterid</span> <span class="operator">=</span> <span class="name">tree</span><span class="operator">.</span><span class="name">cut</span><span class="punctuation">(</span><span class="name">nclusters</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>where <tt class="docutils literal">nclusters</tt> (defaulting to <tt class="docutils literal">1</tt>) is the desired number of
clusters <em>k</em>. This method ignores the top <em>k</em>−1 linking events in the
tree structure, resulting in <em>k</em> separated clusters of items. The number
of clusters <em>k</em> should be positive, and less than or equal to the number
of items. This method returns an array <tt class="docutils literal">clusterid</tt> containing the
number of the cluster to which each item is assigned.</p>
</div>
<div class="section" id="performing-hierarchical-clustering">
<h3>Performing hierarchical clustering</h3>
<p>To perform hierarchical clustering, use the <tt class="docutils literal">treecluster</tt> function in
<tt class="docutils literal">Bio.Cluster</tt>.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">treecluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">treecluster</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>where the following arguments are defined:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt></dt>
<dd><p class="first last">Array containing the data for the items.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">Array of integers showing which data are missing. If
<tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, then <tt class="docutils literal">data[i,j]</tt> is missing. If <tt class="docutils literal"><span class="pre">mask==None</span></tt>,
then all data are present.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">weight</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">The weights to be used when calculating distances. If
<tt class="docutils literal"><span class="pre">weight==None</span></tt>, then equal weights are assumed.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">Determines if rows (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>) or columns (<tt class="docutils literal"><span class="pre">transpose==1</span></tt>)
are to be clustered.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (default: <tt class="docutils literal">'m'</tt>)</dt>
<dd><p class="first last">defines the linkage method to be used:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">method=='s'</span></tt>: pairwise single-linkage clustering</li>
<li><tt class="docutils literal"><span class="pre">method=='m'</span></tt>: pairwise maximum- (or complete-) linkage
clustering</li>
<li><tt class="docutils literal"><span class="pre">method=='c'</span></tt>: pairwise centroid-linkage clustering</li>
<li><tt class="docutils literal"><span class="pre">method=='a'</span></tt>: pairwise average-linkage clustering</li>
</ul>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (default: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">Defines the distance function to be used (see
<a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
</ul>
<p>To apply hierarchical clustering on a precalculated distance matrix,
specify the <tt class="docutils literal">distancematrix</tt> argument when calling <tt class="docutils literal">treecluster</tt>
function instead of the <tt class="docutils literal">data</tt> argument:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">treecluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">treecluster</span><span class="punctuation">(</span><span class="name">distancematrix</span><span class="operator">=</span><span class="name">distance</span><span class="punctuation">)</span>
</pre>
<p>In this case, the following arguments are defined:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">distancematrix</tt></dt>
<dd><p class="first last">The distance matrix, which can be specified in three ways:</p>
</dd>
</dl>
<ul>
<li><p class="first">as a 2D Numerical Python array (in which only the left-lower part
of the array will be accessed):</p>
<pre class="code python literal-block">
<span class="name">distance</span> <span class="operator">=</span> <span class="name">array</span><span class="punctuation">([[</span><span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">1.1</span><span class="punctuation">,</span> <span class="literal number float">2.3</span><span class="punctuation">],</span>
                  <span class="punctuation">[</span><span class="literal number float">1.1</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">],</span>
                  <span class="punctuation">[</span><span class="literal number float">2.3</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">]])</span>
</pre>
</li>
<li><p class="first">as a 1D Numerical Python array containing consecutively the
distances in the left-lower part of the distance matrix:</p>
<pre class="code python literal-block">
<span class="name">distance</span> <span class="operator">=</span> <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">1.1</span><span class="punctuation">,</span> <span class="literal number float">2.3</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">])</span>
</pre>
</li>
<li><p class="first">as a list containing the rows of the left-lower part of the
distance matrix:</p>
<pre class="code python literal-block">
<span class="name">distance</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">array</span><span class="punctuation">([]),</span>
            <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">1.1</span><span class="punctuation">]),</span>
            <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">2.3</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">])</span>
</pre>
</li>
</ul>
<p>These three expressions correspond to the same distance matrix. As
<tt class="docutils literal">treecluster</tt> may shuffle the values in the distance matrix as part
of the clustering algorithm, be sure to save this array in a
different variable before calling <tt class="docutils literal">treecluster</tt> if you need it
later.</p>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt></dt>
<dd><p class="first last">The linkage method to be used:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">method=='s'</span></tt>: pairwise single-linkage clustering</li>
<li><tt class="docutils literal"><span class="pre">method=='m'</span></tt>: pairwise maximum- (or complete-) linkage
clustering</li>
<li><tt class="docutils literal"><span class="pre">method=='a'</span></tt>: pairwise average-linkage clustering</li>
</ul>
<p>While pairwise single-, maximum-, and average-linkage clustering can
be calculated from the distance matrix alone, pairwise
centroid-linkage cannot.</p>
</li>
</ul>
<p>When calling <tt class="docutils literal">treecluster</tt>, either <tt class="docutils literal">data</tt> or <tt class="docutils literal">distancematrix</tt>
should be <tt class="docutils literal">None</tt>.</p>
<p>This function returns a <tt class="docutils literal">Tree</tt> object. This object contains (number of
items − 1) nodes, where the number of items is the number of rows if
rows were clustered, or the number of columns if columns were clustered.
Each node describes a pairwise linking event, where the node attributes
<tt class="docutils literal">left</tt> and <tt class="docutils literal">right</tt> each contain the number of one item or subnode,
and <tt class="docutils literal">distance</tt> the distance between them. Items are numbered from 0 to
(number of items − 1), while clusters are numbered -1 to −(number of
items−1).</p>
</div>
</div>
<div class="section" id="self-organizing-maps">
<h2>15.5  Self-Organizing Maps</h2>
<p>Self-Organizing Maps (SOMs) were invented by Kohonen to describe neural
networks (see for instance Kohonen, 1997 [<a class="reference external" href="#kohonen1997">24</a>]).
Tamayo (1999) first applied Self-Organizing Maps to gene expression data
[<a class="reference external" href="#tamayo1999">30</a>].</p>
<p>SOMs organize items into clusters that are situated in some topology.
Usually a rectangular topology is chosen. The clusters generated by SOMs
are such that neighboring clusters in the topology are more similar to
each other than clusters far from each other in the topology.</p>
<p>The first step to calculate a SOM is to randomly assign a data vector to
each cluster in the topology. If rows are being clustered, then the
number of elements in each data vector is equal to the number of
columns.</p>
<p>An SOM is then generated by taking rows one at a time, and finding which
cluster in the topology has the closest data vector. The data vector of
that cluster, as well as those of the neighboring clusters, are adjusted
using the data vector of the row under consideration. The adjustment is
given by</p>
<div class="formula">
Δ<u><i>x</i></u><sub><span class="textrm">cell</span></sub> = <i>τ</i>⋅<span class="symbol">(</span><u><i>x</i></u><sub><span class="textrm">row</span></sub> − <u><i>x</i></u><sub><span class="textrm">cell</span></sub><span class="symbol">)</span>.
</div>
<p>The parameter τ is a parameter that decreases at each iteration step. We
have used a simple linear function of the iteration step:</p>
<div class="formula">
<i>τ</i> = <i>τ</i><sub><span class="textrm">init</span></sub>⋅<span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span>1 − <span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>, 
</div>
<p>τ<sub>init</sub> is the initial value of τ as specified by the user, <em>i</em>
is the number of the current iteration step, and <em>n</em> is the total number
of iteration steps to be performed. While changes are made rapidly in
the beginning of the iteration, at the end of iteration only small
changes are made.</p>
<p>All clusters within a radius <em>R</em> are adjusted to the gene under
consideration. This radius decreases as the calculation progresses as</p>
<div class="formula">
<i>R</i> = <i>R</i><sub><span class="textrm">max</span></sub>⋅<span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span>1 − <span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>, 
</div>
<p>in which the maximum radius is defined as</p>
<div class="formula">
<i>R</i><sub><span class="textrm">max</span></sub> = <span class="sqrt"><span class="radical">√</span><span class="ignored">(</span><span class="root"><i>N</i><span class="scripts"><sup class="script">2</sup><sub class="script"><i>x</i></sub></span> + <i>N</i><span class="scripts"><sup class="script">2</sup><sub class="script"><i>y</i></sub></span></span><span class="ignored">)</span></span>, 
</div>
<p>where (<em>N</em><sub>*x*</sub>, <em>N</em><sub>*y*</sub>) are the dimensions of the
rectangle defining the topology.</p>
<p>The function <tt class="docutils literal">somcluster</tt> implements the complete algorithm to
calculate a Self-Organizing Map on a rectangular grid. First it
initializes the random number generator. The node data are then
initialized using the random number generator. The order in which genes
or microarrays are used to modify the SOM is also randomized. The total
number of iterations in the SOM algorithm is specified by the user.</p>
<p>To run <tt class="docutils literal">somcluster</tt>, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">somcluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clusterid</span><span class="punctuation">,</span> <span class="name">celldata</span> <span class="operator">=</span> <span class="name">somcluster</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>where the following arguments are defined:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt> (required)</dt>
<dd><p class="first last">Array containing the data for the items.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">Array of integers showing which data are missing. If
<tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, then <tt class="docutils literal">data[i,j]</tt> is missing. If <tt class="docutils literal"><span class="pre">mask==None</span></tt>,
then all data are present.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">weight</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">contains the weights to be used when calculating distances. If
<tt class="docutils literal"><span class="pre">weight==None</span></tt>, then equal weights are assumed.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">Determines if rows (<tt class="docutils literal">transpose</tt> is <tt class="docutils literal">0</tt>) or columns
(<tt class="docutils literal">transpose</tt> is <tt class="docutils literal">1</tt>) are to be clustered.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">nxgrid, nygrid</tt> (default: <tt class="docutils literal">2, 1</tt>)</dt>
<dd><p class="first last">The number of cells horizontally and vertically in the rectangular
grid on which the Self-Organizing Map is calculated.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">inittau</tt> (default: <tt class="docutils literal">0.02</tt>)</dt>
<dd><p class="first last">The initial value for the parameter τ that is used in the SOM
algorithm. The default value for <tt class="docutils literal">inittau</tt> is 0.02, which was used
in Michael Eisen’s Cluster/TreeView program.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">niter</tt> (default: <tt class="docutils literal">1</tt>)</dt>
<dd><p class="first last">The number of iterations to be performed.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (default: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">Defines the distance function to be used (see
<a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
</ul>
<p>This function returns the tuple <tt class="docutils literal">(clusterid, celldata)</tt>:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">clusterid</tt>:</dt>
<dd><p class="first last">An array with two columns, where the number of rows is equal to the
number of items that were clustered. Each row contains the <em>x</em> and
<em>y</em> coordinates of the cell in the rectangular SOM grid to which the
item was assigned.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">celldata</tt>:</dt>
<dd><p class="first last">An array with dimensions (<tt class="docutils literal">nxgrid</tt>, <tt class="docutils literal">nygrid</tt>, number of columns)
if rows are being clustered, or (<tt class="docutils literal">nxgrid</tt>, <tt class="docutils literal">nygrid</tt>, number of
rows) if columns are being clustered. Each element <tt class="docutils literal"><span class="pre">[ix][iy]</span></tt> of
this array is a 1D vector containing the gene expression data for the
centroid of the cluster in the grid cell with coordinates
<tt class="docutils literal"><span class="pre">[ix][iy]</span></tt>.</p>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="principal-component-analysis">
<h2>15.6  Principal Component Analysis</h2>
<p>Principal Component Analysis (PCA) is a widely used technique for
analyzing multivariate data. A practical example of applying Principal
Component Analysis to gene expression data is presented by Yeung and
Ruzzo (2001) [<a class="reference external" href="#yeung2001">33</a>].</p>
<p>In essence, PCA is a coordinate transformation in which each row in the
data matrix is written as a linear sum over basis vectors called
principal components, which are ordered and chosen such that each
maximally explains the remaining variance in the data vectors. For
example, an <em>n</em> × 3 data matrix can be represented as an ellipsoidal
cloud of <em>n</em> points in three dimensional space. The first principal
component is the longest axis of the ellipsoid, the second principal
component the second longest axis of the ellipsoid, and the third
principal component is the shortest axis. Each row in the data matrix
can be reconstructed as a suitable linear combination of the principal
components. However, in order to reduce the dimensionality of the data,
usually only the most important principal components are retained. The
remaining variance present in the data is then regarded as unexplained
variance.</p>
<p>The principal components can be found by calculating the eigenvectors of
the covariance matrix of the data. The corresponding eigenvalues
determine how much of the variance present in the data is explained by
each principal component.</p>
<p>Before applying principal component analysis, typically the mean is
subtracted from each column in the data matrix. In the example above,
this effectively centers the ellipsoidal cloud around its centroid in 3D
space, with the principal components describing the variation of points
in the ellipsoidal cloud with respect to their centroid.</p>
<p>The function <tt class="docutils literal">pca</tt> below first uses the singular value decomposition
to calculate the eigenvalues and eigenvectors of the data matrix. The
singular value decomposition is implemented as a translation in C of the
Algol procedure <tt class="docutils literal">svd</tt> [<a class="reference external" href="#golub1971">16</a>], which uses Householder
bidiagonalization and a variant of the QR algorithm. The principal
components, the coordinates of each data vector along the principal
components, and the eigenvalues corresponding to the principal
components are then evaluated and returned in decreasing order of the
magnitude of the eigenvalue. If data centering is desired, the mean
should be subtracted from each column in the data matrix before calling
the <tt class="docutils literal">pca</tt> routine.</p>
<p>To apply Principal Component Analysis to a rectangular matrix <tt class="docutils literal">data</tt>,
use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">pca</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">columnmean</span><span class="punctuation">,</span> <span class="name">coordinates</span><span class="punctuation">,</span> <span class="name">components</span><span class="punctuation">,</span> <span class="name">eigenvalues</span> <span class="operator">=</span> <span class="name">pca</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>This function returns a tuple
<tt class="docutils literal">columnmean, coordinates, components, eigenvalues</tt>:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">columnmean</tt></dt>
<dd><p class="first last">Array containing the mean over each column in <tt class="docutils literal">data</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">coordinates</tt></dt>
<dd><p class="first last">The coordinates of each row in <tt class="docutils literal">data</tt> with respect to the
principal components.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">components</tt></dt>
<dd><p class="first last">The principal components.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">eigenvalues</tt></dt>
<dd><p class="first last">The eigenvalues corresponding to each of the principal components.</p>
</dd>
</dl>
</li>
</ul>
<p>The original matrix <tt class="docutils literal">data</tt> can be recreated by calculating
<tt class="docutils literal">columnmean +&nbsp; dot(coordinates, components)</tt>.</p>
</div>
<div class="section" id="handling-cluster-treeview-type-files">
<h2>15.7  Handling Cluster/TreeView-type files</h2>
<p>Cluster/TreeView are GUI-based codes for clustering gene expression
data. They were originally written by <a class="reference external" href="http://rana.lbl.gov">Michael
Eisen</a> while at Stanford University.
<tt class="docutils literal">Bio.Cluster</tt> contains functions for reading and writing data files
that correspond to the format specified for Cluster/TreeView. In
particular, by saving a clustering result in that format, TreeView can
be used to visualize the clustering results. We recommend using Alok
Saldanha’s
<a class="reference external" href="http://jtreeview.sourceforge.net/">http://jtreeview.sourceforge.net/</a>Java
TreeView program, which can display hierarchical as well as <em>k</em>-means
clustering results.</p>
<p>An object of the class <tt class="docutils literal">Record</tt> contains all information stored in a
Cluster/TreeView-type data file. To store the information contained in
the data file in a <tt class="docutils literal">Record</tt> object, we first open the file and then
read it:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Cluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;mydatafile.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Cluster</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>This two-step process gives you some flexibility in the source of the
data. For example, you can use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">gzip</span> <span class="comment"># Python standard library</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">gzip</span><span class="operator">.</span><span class="name">open</span><span class="punctuation">(</span><span class="literal string">&quot;mydatafile.txt.gz&quot;</span><span class="punctuation">)</span>
</pre>
<p>to open a gzipped file, or</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">urllib</span> <span class="comment"># Python standard library</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">urllib</span><span class="operator">.</span><span class="name">urlopen</span><span class="punctuation">(</span><span class="literal string">&quot;http://somewhere.org/mydatafile.txt&quot;</span><span class="punctuation">)</span>
</pre>
<p>to open a file stored on the Internet before calling <tt class="docutils literal">read</tt>.</p>
<p>The <tt class="docutils literal">read</tt> command reads the tab-delimited text file
<tt class="docutils literal">mydatafile.txt</tt> containing gene expression data in the format
specified for Michael Eisen’s Cluster/TreeView program. For a
description of this file format, see the manual to Cluster/TreeView. It
is available at <a class="reference external" href="http://rana.lbl.gov/manuals/ClusterTreeView.pdf">Michael Eisen’s lab
website</a> and at <a class="reference external" href="http://bonsai.ims.u-tokyo.ac.jp/~mdehoon/software/cluster/cluster3.pdf">our
website</a>.</p>
<p>A <tt class="docutils literal">Record</tt> object has the following attributes:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt></dt>
<dd><p class="first last">The data array containing the gene expression data. Genes are stored
row-wise, while microarrays are stored column-wise.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt></dt>
<dd><p class="first last">This array shows which elements in the <tt class="docutils literal">data</tt> array, if any, are
missing. If <tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, then <tt class="docutils literal">data[i,j]</tt> is missing. If no
data were found to be missing, <tt class="docutils literal">mask</tt> is set to <tt class="docutils literal">None</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">geneid</tt></dt>
<dd><p class="first last">This is a list containing a unique description for each gene (i.e.,
ORF numbers).</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">genename</tt></dt>
<dd><p class="first last">This is a list containing a description for each gene (i.e., gene
name). If not present in the data file, <tt class="docutils literal">genename</tt> is set to
<tt class="docutils literal">None</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">gweight</tt></dt>
<dd><p class="first last">The weights that are to be used to calculate the distance in
expression profile between genes. If not present in the data file,
<tt class="docutils literal">gweight</tt> is set to <tt class="docutils literal">None</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">gorder</tt></dt>
<dd><p class="first last">The preferred order in which genes should be stored in an output
file. If not present in the data file, <tt class="docutils literal">gorder</tt> is set to <tt class="docutils literal">None</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">expid</tt></dt>
<dd><p class="first last">This is a list containing a description of each microarray, e.g.
experimental condition.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">eweight</tt></dt>
<dd><p class="first last">The weights that are to be used to calculate the distance in
expression profile between microarrays. If not present in the data
file, <tt class="docutils literal">eweight</tt> is set to <tt class="docutils literal">None</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">eorder</tt></dt>
<dd><p class="first last">The preferred order in which microarrays should be stored in an
output file. If not present in the data file, <tt class="docutils literal">eorder</tt> is set to
<tt class="docutils literal">None</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">uniqid</tt></dt>
<dd><p class="first last">The string that was used instead of UNIQID in the data file.</p>
</dd>
</dl>
</li>
</ul>
<p>After loading a <tt class="docutils literal">Record</tt> object, each of these attributes can be
accessed and modified directly. For example, the data can be
log-transformed by taking the logarithm of <tt class="docutils literal">record.data</tt>.</p>
<div class="section" id="id13">
<h3>Calculating the distance matrix</h3>
<p>To calculate the distance matrix between the items stored in the record,
use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">matrix</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">distancematrix</span><span class="punctuation">()</span>
</pre>
<p>where the following arguments are defined:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">Determines if the distances between the rows of <tt class="docutils literal">data</tt> are to be
calculated (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>), or between the columns of <tt class="docutils literal">data</tt>
(<tt class="docutils literal"><span class="pre">transpose==1</span></tt>).</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (default: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">Defines the distance function to be used (see
<a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
</ul>
<p>This function returns the distance matrix as a list of rows, where the
number of columns of each row is equal to the row number (see section
<a class="reference external" href="#subsec:distancematrix">15.1</a>).</p>
</div>
<div class="section" id="id14">
<h3>Calculating the cluster centroids</h3>
<p>To calculate the centroids of clusters of items stored in the record,
use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cdata</span><span class="punctuation">,</span> <span class="name">cmask</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">clustercentroids</span><span class="punctuation">()</span>
</pre>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">clusterid</tt> (default: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">Vector of integers showing to which cluster each item belongs. If
<tt class="docutils literal">clusterid</tt> is not given, then all items are assumed to belong to
the same cluster.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (default: <tt class="docutils literal">'a'</tt>)</dt>
<dd><p class="first last">Specifies whether the arithmetic mean (<tt class="docutils literal"><span class="pre">method=='a'</span></tt>) or the
median (<tt class="docutils literal"><span class="pre">method=='m'</span></tt>) is used to calculate the cluster center.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">Determines if the centroids of the rows of <tt class="docutils literal">data</tt> are to be
calculated (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>), or the centroids of the columns of
<tt class="docutils literal">data</tt> (<tt class="docutils literal"><span class="pre">transpose==1</span></tt>).</p>
</dd>
</dl>
</li>
</ul>
<p>This function returns the tuple <tt class="docutils literal">cdata, cmask</tt>; see section
<a class="reference external" href="#subsec:clustercentroids">15.2</a> for a description.</p>
</div>
<div class="section" id="id15">
<h3>Calculating the distance between clusters</h3>
<p>To calculate the distance between clusters of items stored in the
record, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">distance</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">clusterdistance</span><span class="punctuation">()</span>
</pre>
<p>where the following arguments are defined:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">index1</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">A list containing the indices of the items belonging to the first
cluster. A cluster containing only one item <em>i</em> can be represented
either as a list <tt class="docutils literal">[i]</tt>, or as an integer <tt class="docutils literal">i</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">index2</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">A list containing the indices of the items belonging to the second
cluster. A cluster containing only one item <em>i</em> can be represented
either as a list <tt class="docutils literal">[i]</tt>, or as an integer <tt class="docutils literal">i</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (default: <tt class="docutils literal">'a'</tt>)</dt>
<dd><p class="first last">Specifies how the distance between clusters is defined:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal">'a'</tt>: Distance between the two cluster centroids (arithmetic
mean);</li>
<li><tt class="docutils literal">'m'</tt>: Distance between the two cluster centroids (median);</li>
<li><tt class="docutils literal">'s'</tt>: Shortest pairwise distance between items in the two
clusters;</li>
<li><tt class="docutils literal">'x'</tt>: Longest pairwise distance between items in the two
clusters;</li>
<li><tt class="docutils literal">'v'</tt>: Average over the pairwise distances between items in the
two clusters.</li>
</ul>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (default: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">Defines the distance function to be used (see
<a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">If <tt class="docutils literal"><span class="pre">transpose==0</span></tt>, calculate the distance between the rows of
<tt class="docutils literal">data</tt>. If <tt class="docutils literal"><span class="pre">transpose==1</span></tt>, calculate the distance between the
columns of <tt class="docutils literal">data</tt>.</p>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id16">
<h3>Performing hierarchical clustering</h3>
<p>To perform hierarchical clustering on the items stored in the record,
use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">treecluster</span><span class="punctuation">()</span>
</pre>
<p>where the following arguments are defined:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">Determines if rows (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>) or columns (<tt class="docutils literal"><span class="pre">transpose==1</span></tt>)
are to be clustered.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (default: <tt class="docutils literal">'m'</tt>)</dt>
<dd><p class="first last">defines the linkage method to be used:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">method=='s'</span></tt>: pairwise single-linkage clustering</li>
<li><tt class="docutils literal"><span class="pre">method=='m'</span></tt>: pairwise maximum- (or complete-) linkage
clustering</li>
<li><tt class="docutils literal"><span class="pre">method=='c'</span></tt>: pairwise centroid-linkage clustering</li>
<li><tt class="docutils literal"><span class="pre">method=='a'</span></tt>: pairwise average-linkage clustering</li>
</ul>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (default: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">Defines the distance function to be used (see
<a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt></dt>
<dd><p class="first last">Determines if genes or microarrays are being clustered. If
<tt class="docutils literal"><span class="pre">transpose==0</span></tt>, genes (rows) are being clustered. If
<tt class="docutils literal"><span class="pre">transpose==1</span></tt>, microarrays (columns) are clustered.</p>
</dd>
</dl>
</li>
</ul>
<p>This function returns a <tt class="docutils literal">Tree</tt> object. This object contains (number of
items − 1) nodes, where the number of items is the number of rows if
rows were clustered, or the number of columns if columns were clustered.
Each node describes a pairwise linking event, where the node attributes
<tt class="docutils literal">left</tt> and <tt class="docutils literal">right</tt> each contain the number of one item or subnode,
and <tt class="docutils literal">distance</tt> the distance between them. Items are numbered from 0 to
(number of items − 1), while clusters are numbered -1 to −(number of
items−1).</p>
</div>
<div class="section" id="performing-k-means-or-k-medians-clustering">
<h3>Performing <em>k</em>-means or <em>k</em>-medians clustering</h3>
<p>To perform <em>k</em>-means or <em>k</em>-medians clustering on the items stored in
the record, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clusterid</span><span class="punctuation">,</span> <span class="name">error</span><span class="punctuation">,</span> <span class="name">nfound</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">kcluster</span><span class="punctuation">()</span>
</pre>
<p>where the following arguments are defined:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">nclusters</tt> (default: <tt class="docutils literal">2</tt>)</dt>
<dd><p class="first last">The number of clusters <em>k</em>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">Determines if rows (<tt class="docutils literal">transpose</tt> is <tt class="docutils literal">0</tt>) or columns
(<tt class="docutils literal">transpose</tt> is <tt class="docutils literal">1</tt>) are to be clustered.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">npass</tt> (default: <tt class="docutils literal">1</tt>)</dt>
<dd><p class="first last">The number of times the <em>k</em>-means/-medians clustering algorithm is
performed, each time with a different (random) initial condition. If
<tt class="docutils literal">initialid</tt> is given, the value of <tt class="docutils literal">npass</tt> is ignored and the
clustering algorithm is run only once, as it behaves
deterministically in that case.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (default: <tt class="docutils literal">a</tt>)</dt>
<dd><p class="first last">describes how the center of a cluster is found:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">method=='a'</span></tt>: arithmetic mean (<em>k</em>-means clustering);</li>
<li><tt class="docutils literal"><span class="pre">method=='m'</span></tt>: median (<em>k</em>-medians clustering).</li>
</ul>
<p>For other values of <tt class="docutils literal">method</tt>, the arithmetic mean is used.</p>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (default: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">Defines the distance function to be used (see
<a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
</ul>
<p>This function returns a tuple <tt class="docutils literal">(clusterid, error, nfound)</tt>, where
<tt class="docutils literal">clusterid</tt> is an integer array containing the number of the cluster
to which each row or cluster was assigned, <tt class="docutils literal">error</tt> is the
within-cluster sum of distances for the optimal clustering solution, and
<tt class="docutils literal">nfound</tt> is the number of times this optimal solution was found.</p>
</div>
<div class="section" id="calculating-a-self-organizing-map">
<h3>Calculating a Self-Organizing Map</h3>
<p>To calculate a Self-Organizing Map of the items stored in the record,
use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clusterid</span><span class="punctuation">,</span> <span class="name">celldata</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">somcluster</span><span class="punctuation">()</span>
</pre>
<p>where the following arguments are defined:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (default: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">Determines if rows (<tt class="docutils literal">transpose</tt> is <tt class="docutils literal">0</tt>) or columns
(<tt class="docutils literal">transpose</tt> is <tt class="docutils literal">1</tt>) are to be clustered.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">nxgrid, nygrid</tt> (default: <tt class="docutils literal">2, 1</tt>)</dt>
<dd><p class="first last">The number of cells horizontally and vertically in the rectangular
grid on which the Self-Organizing Map is calculated.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">inittau</tt> (default: <tt class="docutils literal">0.02</tt>)</dt>
<dd><p class="first last">The initial value for the parameter τ that is used in the SOM
algorithm. The default value for <tt class="docutils literal">inittau</tt> is 0.02, which was used
in Michael Eisen’s Cluster/TreeView program.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">niter</tt> (default: <tt class="docutils literal">1</tt>)</dt>
<dd><p class="first last">The number of iterations to be performed.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (default: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">Defines the distance function to be used (see
<a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
</ul>
<p>This function returns the tuple <tt class="docutils literal">(clusterid, celldata)</tt>:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">clusterid</tt>:</dt>
<dd><p class="first last">An array with two columns, where the number of rows is equal to the
number of items that were clustered. Each row contains the <em>x</em> and
<em>y</em> coordinates of the cell in the rectangular SOM grid to which the
item was assigned.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">celldata</tt>:</dt>
<dd><p class="first last">An array with dimensions (<tt class="docutils literal">nxgrid</tt>, <tt class="docutils literal">nygrid</tt>, number of columns)
if rows are being clustered, or (<tt class="docutils literal">nxgrid</tt>, <tt class="docutils literal">nygrid</tt>, number of
rows) if columns are being clustered. Each element <tt class="docutils literal"><span class="pre">[ix][iy]</span></tt> of
this array is a 1D vector containing the gene expression data for the
centroid of the cluster in the grid cell with coordinates
<tt class="docutils literal"><span class="pre">[ix][iy]</span></tt>.</p>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="saving-the-clustering-result">
<h3>Saving the clustering result</h3>
<p>To save the clustering result, use</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">save</span><span class="punctuation">(</span><span class="name">jobname</span><span class="punctuation">,</span> <span class="name">geneclusters</span><span class="punctuation">,</span> <span class="name">expclusters</span><span class="punctuation">)</span>
</pre>
<p>where the following arguments are defined:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">jobname</tt></dt>
<dd><p class="first last">The string <tt class="docutils literal">jobname</tt> is used as the base name for names of the
files that are to be saved.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">geneclusters</tt></dt>
<dd><p class="first last">This argument describes the gene (row-wise) clustering result. In
case of <em>k</em>-means clustering, this is a 1D array containing the
number of the cluster each gene belongs to. It can be calculated
using <tt class="docutils literal">kcluster</tt>. In case of hierarchical clustering,
<tt class="docutils literal">geneclusters</tt> is a <tt class="docutils literal">Tree</tt> object.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">expclusters</tt></dt>
<dd><p class="first last">This argument describes the (column-wise) clustering result for the
experimental conditions. In case of <em>k</em>-means clustering, this is a
1D array containing the number of the cluster each experimental
condition belongs to. It can be calculated using <tt class="docutils literal">kcluster</tt>. In
case of hierarchical clustering, <tt class="docutils literal">expclusters</tt> is a <tt class="docutils literal">Tree</tt>
object.</p>
</dd>
</dl>
</li>
</ul>
<p>This method writes the text file <tt class="docutils literal">jobname.cdt</tt>, <tt class="docutils literal">jobname.gtr</tt>,
<tt class="docutils literal">jobname.atr</tt>, <tt class="docutils literal"><span class="pre">jobname*.kgg</span></tt>, and/or <tt class="docutils literal"><span class="pre">jobname*.kag</span></tt> for
subsequent reading by the Java TreeView program. If <tt class="docutils literal">geneclusters</tt> and
<tt class="docutils literal">expclusters</tt> are both <tt class="docutils literal">None</tt>, this method only writes the text file
<tt class="docutils literal">jobname.cdt</tt>; this file can subsequently be read into a new
<tt class="docutils literal">Record</tt> object.</p>
</div>
</div>
<div class="section" id="example-calculation">
<h2>15.8  Example calculation</h2>
<p>This is an example of a hierarchical clustering calculation, using
single linkage clustering for genes and maximum linkage clustering for
experimental conditions. As the Euclidean distance is being used for
gene clustering, it is necessary to scale the node distances
<tt class="docutils literal">genetree</tt> such that they are all between zero and one. This is needed
for the Java TreeView code to display the tree diagram correctly. To
cluster the experimental conditions, the uncentered correlation is being
used. No scaling is needed in this case, as the distances in <tt class="docutils literal">exptree</tt>
are already between zero and two. The example data <tt class="docutils literal">cyano.txt</tt> can be
found in the <tt class="docutils literal">data</tt> subdirectory.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Cluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;cyano.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Cluster</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">genetree</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">treecluster</span><span class="punctuation">(</span><span class="name">method</span><span class="operator">=</span><span class="literal string">'s'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">genetree</span><span class="operator">.</span><span class="name">scale</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exptree</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">treecluster</span><span class="punctuation">(</span><span class="name">dist</span><span class="operator">=</span><span class="literal string">'u'</span><span class="punctuation">,</span> <span class="name">transpose</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">save</span><span class="punctuation">(</span><span class="literal string">&quot;cyano_result&quot;</span><span class="punctuation">,</span> <span class="name">genetree</span><span class="punctuation">,</span> <span class="name">exptree</span><span class="punctuation">)</span>
</pre>
<p>This will create the files <tt class="docutils literal">cyano_result.cdt</tt>, <tt class="docutils literal">cyano_result.gtr</tt>,
and <tt class="docutils literal">cyano_result.atr</tt>.</p>
<p>Similarly, we can save a <em>k</em>-means clustering solution:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Cluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;cyano.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Cluster</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="punctuation">(</span><span class="name">geneclusters</span><span class="punctuation">,</span> <span class="name">error</span><span class="punctuation">,</span> <span class="name">ifound</span><span class="punctuation">)</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">kcluster</span><span class="punctuation">(</span><span class="name">nclusters</span><span class="operator">=</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="name">npass</span><span class="operator">=</span><span class="literal number integer">1000</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="punctuation">(</span><span class="name">expclusters</span><span class="punctuation">,</span> <span class="name">error</span><span class="punctuation">,</span> <span class="name">ifound</span><span class="punctuation">)</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">kcluster</span><span class="punctuation">(</span><span class="name">nclusters</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name">npass</span><span class="operator">=</span><span class="literal number integer">100</span><span class="punctuation">,</span> <span class="name">transpose</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">save</span><span class="punctuation">(</span><span class="literal string">&quot;cyano_result&quot;</span><span class="punctuation">,</span> <span class="name">geneclusters</span><span class="punctuation">,</span> <span class="name">expclusters</span><span class="punctuation">)</span>
</pre>
<p>This will create the files <tt class="docutils literal">cyano_result_K_G2_A2.cdt</tt>,
<tt class="docutils literal">cyano_result_K_G2.kgg</tt>, and <tt class="docutils literal">cyano_result_K_A2.kag</tt>.</p>
</div>
<div class="section" id="auxiliary-functions">
<h2>15.9  Auxiliary functions</h2>
<p><tt class="docutils literal">median(data)</tt> returns the median of the 1D array <tt class="docutils literal">data</tt>.</p>
<p><tt class="docutils literal">mean(data)</tt> returns the mean of the 1D array <tt class="docutils literal">data</tt>.</p>
<p><tt class="docutils literal">version()</tt> returns the version number of the underlying C Clustering
Library as a string.</p>
</div>
</div>
<div class="section" id="chapter-16-supervised-learning-methods">
<h1>Chapter 16  Supervised learning methods</h1>
<p>Note the supervised learning methods described in this chapter all
require Numerical Python (numpy) to be installed.</p>
<div class="section" id="the-logistic-regression-model">
<h2>16.1  The Logistic Regression Model</h2>
<div class="section" id="background-and-purpose">
<h3>16.1.1  Background and Purpose</h3>
<p>Logistic regression is a supervised learning approach that attempts to
distinguish <em>K</em> classes from each other using a weighted sum of some
predictor variables <em>x</em><sub>*i*</sub>. The logistic regression model is
used to calculate the weights β<sub>*i*</sub> of the predictor variables.
In Biopython, the logistic regression model is currently implemented for
two classes only (<em>K</em> = 2); the number of predictor variables has no
predefined limit.</p>
<p>As an example, let’s try to predict the operon structure in bacteria. An
operon is a set of adjacent genes on the same strand of DNA that are
transcribed into a single mRNA molecule. Translation of the single mRNA
molecule then yields the individual proteins. For <em>Bacillus subtilis</em>,
whose data we will be using, the average number of genes in an operon is
about 2.4.</p>
<p>As a first step in understanding gene regulation in bacteria, we need to
know the operon structure. For about 10% of the genes in <em>Bacillus
subtilis</em>, the operon structure is known from experiments. A supervised
learning method can be used to predict the operon structure for the
remaining 90% of the genes.</p>
<p>For such a supervised learning approach, we need to choose some
predictor variables <em>x</em><sub>*i*</sub> that can be measured easily and are
somehow related to the operon structure. One predictor variable might be
the distance in base pairs between genes. Adjacent genes belonging to
the same operon tend to be separated by a relatively short distance,
whereas adjacent genes in different operons tend to have a larger space
between them to allow for promoter and terminator sequences. Another
predictor variable is based on gene expression measurements. By
definition, genes belonging to the same operon have equal gene
expression profiles, while genes in different operons are expected to
have different expression profiles. In practice, the measured expression
profiles of genes in the same operon are not quite identical due to the
presence of measurement errors. To assess the similarity in the gene
expression profiles, we assume that the measurement errors follow a
normal distribution and calculate the corresponding log-likelihood
score.</p>
<p>We now have two predictor variables that we can use to predict if two
adjacent genes on the same strand of DNA belong to the same operon:</p>
<ul class="simple">
<li><em>x</em><sub>1</sub>: the number of base pairs between them;</li>
<li><em>x</em><sub>2</sub>: their similarity in expression profile.</li>
</ul>
<p>In a logistic regression model, we use a weighted sum of these two
predictors to calculate a joint score <em>S</em>:</p>
<div class="formula">
<i>S</i> = <i>β</i><sub>0</sub> + <i>β</i><sub>1</sub><i>x</i><sub>1</sub> + <i>β</i><sub>2</sub><i>x</i><sub>2</sub>.
</div>
<p>The logistic regression model gives us appropriate values for the
parameters β<sub>0</sub>, β<sub>1</sub>, β<sub>2</sub> using two sets of
example genes:</p>
<ul class="simple">
<li>OP: Adjacent genes, on the same strand of DNA, known to belong to the
same operon;</li>
<li>NOP: Adjacent genes, on the same strand of DNA, known to belong to
different operons.</li>
</ul>
<p>In the logistic regression model, the probability of belonging to a
class depends on the score via the logistic function. For the two
classes OP and NOP, we can write this as</p>
<div class="formula">
<span class="environment"><span class="arrayrow">
<span class="arraycell align-r">
Pr(<span class="mathrm">OP</span>|<i>x</i><sub>1</sub>, <i>x</i><sub>2</sub>)
</span>
<span class="arraycell align-c">
 = 
</span>
<span class="arraycell align-l">
<span class="fraction"><span class="ignored">(</span><span class="numerator">exp(<i>β</i><sub>0</sub> + <i>β</i><sub>1</sub><i>x</i><sub>1</sub> + <i>β</i><sub>2</sub><i>x</i><sub>2</sub>)</span><span class="ignored">)/(</span><span class="denominator">1 + exp(<i>β</i><sub>0</sub> + <i>β</i><sub>1</sub><i>x</i><sub>1</sub> + <i>β</i><sub>2</sub><i>x</i><sub>2</sub>)</span><span class="ignored">)</span></span> <a class="eqnumber" name="eq:OP"> </a>
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-r">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-l">
 
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-r">
Pr(<span class="mathrm">NOP</span>|<i>x</i><sub>1</sub>, <i>x</i><sub>2</sub>)
</span>
<span class="arraycell align-c">
 = 
</span>
<span class="arraycell align-l">
<span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator">1 + exp(<i>β</i><sub>0</sub> + <i>β</i><sub>1</sub><i>x</i><sub>1</sub> + <i>β</i><sub>2</sub><i>x</i><sub>2</sub>)</span><span class="ignored">)</span></span> <a class="eqnumber" name="eq:NOP"> </a>
</span>

</span>
</span>
</div>
<p>Using a set of gene pairs for which it is known whether they belong to
the same operon (class OP) or to different operons (class NOP), we can
calculate the weights β<sub>0</sub>, β<sub>1</sub>, β<sub>2</sub> by
maximizing the log-likelihood corresponding to the probability functions
(<a class="reference external" href="#eq:OP">16.2</a>) and (<a class="reference external" href="#eq:NOP">16.3</a>).</p>
</div>
<div class="section" id="training-the-logistic-regression-model">
<h3>16.1.2  Training the logistic regression model</h3>
<div class="system-message">
<p class="system-message-title">System Message: ERROR/3 (<tt class="docutils">allinone.rst</tt>, line 16520)</p>
Document or section may not begin with a transition.</div>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr><td>Table 16.1: Adjacent gene pairs known to belong to the same operon (class OP) or to different operons (class NOP). Intergene distances are negative if the two genes overlap.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="35%" />
<col width="38%" />
<col width="8%" />
</colgroup>
<tbody valign="top">
<tr><td>Gene pair</td>
<td>Intergene distance (<em>x</em><sub>1</sub>)</td>
<td>Gene expression score (<em>x</em><sub>2</sub>)</td>
<td>Class</td>
</tr>
<tr><td><em>cotJA</em> — <em>cotJB</em></td>
<td>-53</td>
<td>-200.78</td>
<td>OP</td>
</tr>
<tr><td><em>yesK</em> — <em>yesL</em></td>
<td>117</td>
<td>-267.14</td>
<td>OP</td>
</tr>
<tr><td><em>lplA</em> — <em>lplB</em></td>
<td>57</td>
<td>-163.47</td>
<td>OP</td>
</tr>
<tr><td><em>lplB</em> — <em>lplC</em></td>
<td>16</td>
<td>-190.30</td>
<td>OP</td>
</tr>
<tr><td><em>lplC</em> — <em>lplD</em></td>
<td>11</td>
<td>-220.94</td>
<td>OP</td>
</tr>
<tr><td><em>lplD</em> — <em>yetF</em></td>
<td>85</td>
<td>-193.94</td>
<td>OP</td>
</tr>
<tr><td><em>yfmT</em> — <em>yfmS</em></td>
<td>16</td>
<td>-182.71</td>
<td>OP</td>
</tr>
<tr><td><em>yfmF</em> — <em>yfmE</em></td>
<td>15</td>
<td>-180.41</td>
<td>OP</td>
</tr>
<tr><td><em>citS</em> — <em>citT</em></td>
<td>-26</td>
<td>-181.73</td>
<td>OP</td>
</tr>
<tr><td><em>citM</em> — <em>yflN</em></td>
<td>58</td>
<td>-259.87</td>
<td>OP</td>
</tr>
<tr><td><em>yfiI</em> — <em>yfiJ</em></td>
<td>126</td>
<td>-414.53</td>
<td>NOP</td>
</tr>
<tr><td><em>lipB</em> — <em>yfiQ</em></td>
<td>191</td>
<td>-249.57</td>
<td>NOP</td>
</tr>
<tr><td><em>yfiU</em> — <em>yfiV</em></td>
<td>113</td>
<td>-265.28</td>
<td>NOP</td>
</tr>
<tr><td><em>yfhH</em> — <em>yfhI</em></td>
<td>145</td>
<td>-312.99</td>
<td>NOP</td>
</tr>
<tr><td><em>cotY</em> — <em>cotX</em></td>
<td>154</td>
<td>-213.83</td>
<td>NOP</td>
</tr>
<tr><td><em>yjoB</em> — <em>rapA</em></td>
<td>147</td>
<td>-380.85</td>
<td>NOP</td>
</tr>
<tr><td><em>ptsI</em> — <em>splA</em></td>
<td>93</td>
<td>-291.13</td>
<td>NOP</td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<p>Table <a class="reference external" href="#table:training">16.1</a> lists some of the <em>Bacillus subtilis</em>
gene pairs for which the operon structure is known. Let’s calculate the
logistic regression model from these data:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">LogisticRegression</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">xs</span> <span class="operator">=</span> <span class="punctuation">[[</span><span class="operator">-</span><span class="literal number integer">53</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">200.78</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">117</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">267.14</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">57</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">163.47</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">16</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">190.30</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">11</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">220.94</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">85</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">193.94</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">16</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">182.71</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">15</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">180.41</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">26</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">181.73</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">58</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">259.87</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">126</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">414.53</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">191</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">249.57</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">113</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">265.28</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">145</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">312.99</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">154</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">213.83</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">147</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">380.85</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">93</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">291.13</span><span class="punctuation">]]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ys</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">train</span><span class="punctuation">(</span><span class="name">xs</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">)</span>
</pre>
<p>Here, <tt class="docutils literal">xs</tt> and <tt class="docutils literal">ys</tt> are the training data: <tt class="docutils literal">xs</tt> contains the
predictor variables for each gene pair, and <tt class="docutils literal">ys</tt> specifies if the gene
pair belongs to the same operon (<tt class="docutils literal">1</tt>, class OP) or different operons
(<tt class="docutils literal">0</tt>, class NOP). The resulting logistic regression model is stored in
<tt class="docutils literal">model</tt>, which contains the weights β<sub>0</sub>, β<sub>1</sub>, and
β<sub>2</sub>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span><span class="operator">.</span><span class="name">beta</span>
<span class="punctuation">[</span><span class="literal number float">8.9830290157144681</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">0.035968960444850887</span><span class="punctuation">,</span> <span class="literal number float">0.02181395662983519</span><span class="punctuation">]</span>
</pre>
<p>Note that β<sub>1</sub> is negative, as gene pairs with a shorter
intergene distance have a higher probability of belonging to the same
operon (class OP). On the other hand, β<sub>2</sub> is positive, as gene
pairs belonging to the same operon typically have a higher similarity
score of their gene expression profiles. The parameter β<sub>0</sub> is
positive due to the higher prevalence of operon gene pairs than
non-operon gene pairs in the training data.</p>
<p>The function <tt class="docutils literal">train</tt> has two optional arguments: <tt class="docutils literal">update_fn</tt> and
<tt class="docutils literal">typecode</tt>. The <tt class="docutils literal">update_fn</tt> can be used to specify a callback
function, taking as arguments the iteration number and the
log-likelihood. With the callback function, we can for example track the
progress of the model calculation (which uses a Newton-Raphson iteration
to maximize the log-likelihood function of the logistic regression
model):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">def</span> <span class="name function">show_progress</span><span class="punctuation">(</span><span class="name">iteration</span><span class="punctuation">,</span> <span class="name">loglikelihood</span><span class="punctuation">):</span>
        <span class="keyword">print</span> <span class="literal string">&quot;Iteration:&quot;</span><span class="punctuation">,</span> <span class="name">iteration</span><span class="punctuation">,</span> <span class="literal string">&quot;Log-likelihood function:&quot;</span><span class="punctuation">,</span> <span class="name">loglikelihood</span>
<span class="operator">&gt;&gt;&gt;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">train</span><span class="punctuation">(</span><span class="name">xs</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">,</span> <span class="name">update_fn</span><span class="operator">=</span><span class="name">show_progress</span><span class="punctuation">)</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">11.7835020695</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">7.15886767672</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">2</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">5.76877209868</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">3</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">5.11362294338</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">4</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">4.74870642433</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">5</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">4.50026077146</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">6</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">4.31127773737</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">7</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">4.16015043396</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">8</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">4.03561719785</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">9</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.93073282192</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">10</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.84087660929</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">11</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.76282560605</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">12</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.69425027154</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">13</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.6334178602</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">14</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.57900855837</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">15</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.52999671386</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">16</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.48557145163</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">17</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.44508206139</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">18</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.40799948447</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">19</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.3738885624</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">20</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.3423876581</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">21</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.31319343769</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">22</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.2860493346</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">23</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.2607366863</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">24</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.23706784091</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">25</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.21488073614</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">26</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.19403459259</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">27</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.17440646052</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">28</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.15588842703</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">29</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.13838533947</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">30</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.12181293595</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">31</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.10609629966</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">32</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.09116857282</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">33</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.07696988017</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">34</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.06344642288</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">35</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.05054971191</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">36</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.03823591619</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">37</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.02646530573</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">38</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.01520177394</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">39</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.00441242601</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">40</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.99406722296</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">41</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.98413867259</span>
</pre>
<p>The iteration stops once the increase in the log-likelihood function is
less than 0.01. If no convergence is reached after 500 iterations, the
<tt class="docutils literal">train</tt> function returns with an <tt class="docutils literal">AssertionError</tt>.</p>
<p>The optional keyword <tt class="docutils literal">typecode</tt> can almost always be ignored. This
keyword allows the user to choose the type of Numeric matrix to use. In
particular, to avoid memory problems for very large problems, it may be
necessary to use single-precision floats (Float8, Float16, etc.) rather
than double, which is used by default.</p>
</div>
<div class="section" id="using-the-logistic-regression-model-for-classification">
<h3>16.1.3  Using the logistic regression model for classification</h3>
<p>Classification is performed by calling the <tt class="docutils literal">classify</tt> function. Given
a logistic regression model and the values for <em>x</em><sub>1</sub> and
<em>x</em><sub>2</sub> (e.g. for a gene pair of unknown operon structure), the
<tt class="docutils literal">classify</tt> function returns <tt class="docutils literal">1</tt> or <tt class="docutils literal">0</tt>, corresponding to class OP
and class NOP, respectively. For example, let’s consider the gene pairs
<em>yxcE</em>, <em>yxcD</em> and <em>yxiB</em>, <em>yxiA</em>:</p>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr><td>Table 16.2: Adjacent gene pairs of unknown operon status.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="20%" />
<col width="38%" />
<col width="41%" />
</colgroup>
<tbody valign="top">
<tr><td>Gene pair</td>
<td>Intergene distance <em>x</em><sub>1</sub></td>
<td>Gene expression score <em>x</em><sub>2</sub></td>
</tr>
<tr><td><em>yxcE</em> — <em>yxcD</em></td>
<td>6</td>
<td>-173.143442352</td>
</tr>
<tr><td><em>yxiB</em> — <em>yxiA</em></td>
<td>309</td>
<td>-271.005880394</td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<p>The logistic regression model classifies <em>yxcE</em>, <em>yxcD</em> as belonging to
the same operon (class OP), while <em>yxiB</em>, <em>yxiA</em> are predicted to belong
to different operons:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;yxcE, yxcD:&quot;</span><span class="punctuation">,</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">,</span><span class="operator">-</span><span class="literal number float">173.143442352</span><span class="punctuation">])</span>
<span class="name">yxcE</span><span class="punctuation">,</span> <span class="name">yxcD</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;yxiB, yxiA:&quot;</span><span class="punctuation">,</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">309</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">271.005880394</span><span class="punctuation">])</span>
<span class="name">yxiB</span><span class="punctuation">,</span> <span class="name">yxiA</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
</pre>
<p>(which, by the way, agrees with the biological literature).</p>
<p>To find out how confident we can be in these predictions, we can call
the <tt class="docutils literal">calculate</tt> function to obtain the probabilities (equations
(<a class="reference external" href="#eq:OP">16.2</a>) and (<a class="reference external" href="#eq:NOP">16.3</a>)) for class OP and NOP. For
<em>yxcE</em>, <em>yxcD</em> we find</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">q</span><span class="punctuation">,</span> <span class="name">p</span> <span class="operator">=</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">calculate</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">,</span><span class="operator">-</span><span class="literal number float">173.143442352</span><span class="punctuation">])</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;class OP: probability =&quot;</span><span class="punctuation">,</span> <span class="name">p</span><span class="punctuation">,</span> <span class="literal string">&quot;class NOP: probability =&quot;</span><span class="punctuation">,</span> <span class="name">q</span>
<span class="keyword">class</span> <span class="name class">OP</span><span class="punctuation">:</span> <span class="name">probability</span> <span class="operator">=</span> <span class="literal number float">0.993242163503</span> <span class="keyword">class</span> <span class="name class">NOP</span><span class="punctuation">:</span> <span class="name">probability</span> <span class="operator">=</span> <span class="literal number float">0.00675783649744</span>
</pre>
<p>and for <em>yxiB</em>, <em>yxiA</em></p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">q</span><span class="punctuation">,</span> <span class="name">p</span> <span class="operator">=</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">calculate</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">309</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">271.005880394</span><span class="punctuation">])</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;class OP: probability =&quot;</span><span class="punctuation">,</span> <span class="name">p</span><span class="punctuation">,</span> <span class="literal string">&quot;class NOP: probability =&quot;</span><span class="punctuation">,</span> <span class="name">q</span>
<span class="keyword">class</span> <span class="name class">OP</span><span class="punctuation">:</span> <span class="name">probability</span> <span class="operator">=</span> <span class="literal number float">0.000321211251817</span> <span class="keyword">class</span> <span class="name class">NOP</span><span class="punctuation">:</span> <span class="name">probability</span> <span class="operator">=</span> <span class="literal number float">0.999678788748</span>
</pre>
<p>To get some idea of the prediction accuracy of the logistic regression
model, we can apply it to the training data:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">ys</span><span class="punctuation">)):</span>
        <span class="keyword">print</span> <span class="literal string">&quot;True:&quot;</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">],</span> <span class="literal string">&quot;Predicted:&quot;</span><span class="punctuation">,</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">xs</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">])</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
</pre>
<p>showing that the prediction is correct for all but one of the gene
pairs. A more reliable estimate of the prediction accuracy can be found
from a leave-one-out analysis, in which the model is recalculated from
the training data after removing the gene to be predicted:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">ys</span><span class="punctuation">)):</span>
        <span class="name">model</span> <span class="operator">=</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">train</span><span class="punctuation">(</span><span class="name">xs</span><span class="punctuation">[:</span><span class="name">i</span><span class="punctuation">]</span><span class="operator">+</span><span class="name">xs</span><span class="punctuation">[</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">:],</span> <span class="name">ys</span><span class="punctuation">[:</span><span class="name">i</span><span class="punctuation">]</span><span class="operator">+</span><span class="name">ys</span><span class="punctuation">[</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">:])</span>
        <span class="keyword">print</span> <span class="literal string">&quot;True:&quot;</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">],</span> <span class="literal string">&quot;Predicted:&quot;</span><span class="punctuation">,</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">xs</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">])</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
</pre>
<p>The leave-one-out analysis shows that the prediction of the logistic
regression model is incorrect for only two of the gene pairs, which
corresponds to a prediction accuracy of 88%.</p>
</div>
<div class="section" id="logistic-regression-linear-discriminant-analysis-and-support-vector-machines">
<h3>16.1.4  Logistic Regression, Linear Discriminant Analysis, and Support Vector Machines</h3>
<p>The logistic regression model is similar to linear discriminant
analysis. In linear discriminant analysis, the class probabilities also
follow equations (<a class="reference external" href="#eq:OP">16.2</a>) and (<a class="reference external" href="#eq:NOP">16.3</a>). However,
instead of estimating the coefficients β directly, we first fit a normal
distribution to the predictor variables <em>x</em>. The coefficients β are then
calculated from the means and covariances of the normal distribution. If
the distribution of <em>x</em> is indeed normal, then we expect linear
discriminant analysis to perform better than the logistic regression
model. The logistic regression model, on the other hand, is more robust
to deviations from normality.</p>
<p>Another similar approach is a support vector machine with a linear
kernel. Such an SVM also uses a linear combination of the predictors,
but estimates the coefficients β from the predictor variables <em>x</em> near
the boundary region between the classes. If the logistic regression
model (equations (<a class="reference external" href="#eq:OP">16.2</a>) and (<a class="reference external" href="#eq:NOP">16.3</a>)) is a good
description for <em>x</em> away from the boundary region, we expect the
logistic regression model to perform better than an SVM with a linear
kernel, as it relies on more data. If not, an SVM with a linear kernel
may perform better.</p>
<p>Trevor Hastie, Robert Tibshirani, and Jerome Friedman: <em>The Elements of
Statistical Learning. Data Mining, Inference, and Prediction</em>. Springer
Series in Statistics, 2001. Chapter 4.4.</p>
</div>
</div>
<div class="section" id="k-nearest-neighbors">
<h2>16.2  <em>k</em>-Nearest Neighbors</h2>
<div class="section" id="id17">
<h3>16.2.1  Background and purpose</h3>
<p>The <em>k</em>-nearest neighbors method is a supervised learning approach that
does not need to fit a model to the data. Instead, data points are
classified based on the categories of the <em>k</em> nearest neighbors in the
training data set.</p>
<p>In Biopython, the <em>k</em>-nearest neighbors method is available in
<tt class="docutils literal">Bio.kNN</tt>. To illustrate the use of the <em>k</em>-nearest neighbor method in
Biopython, we will use the same operon data set as in section
<a class="reference external" href="#sec:LogisticRegression">16.1</a>.</p>
</div>
<div class="section" id="initializing-a-k-nearest-neighbors-model">
<h3>16.2.2  Initializing a <em>k</em>-nearest neighbors model</h3>
<p>Using the data in Table <a class="reference external" href="#table:training">16.1</a>, we create and
initialize a <em>k</em>-nearest neighbors model as follows:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">kNN</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">k</span> <span class="operator">=</span> <span class="literal number integer">3</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">train</span><span class="punctuation">(</span><span class="name">xs</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">,</span> <span class="name">k</span><span class="punctuation">)</span>
</pre>
<p>where <tt class="docutils literal">xs</tt> and <tt class="docutils literal">ys</tt> are the same as in Section
<a class="reference external" href="#subsec:LogisticRegressionTraining">16.1.2</a>. Here, <tt class="docutils literal">k</tt> is the
number of neighbors <em>k</em> that will be considered for the classification.
For classification into two classes, choosing an odd number for <em>k</em> lets
you avoid tied votes. The function name <tt class="docutils literal">train</tt> is a bit of a
misnomer, since no model training is done: this function simply stores
<tt class="docutils literal">xs</tt>, <tt class="docutils literal">ys</tt>, and <tt class="docutils literal">k</tt> in <tt class="docutils literal">model</tt>.</p>
</div>
<div class="section" id="using-a-k-nearest-neighbors-model-for-classification">
<h3>16.2.3  Using a <em>k</em>-nearest neighbors model for classification</h3>
<p>To classify new data using the <em>k</em>-nearest neighbors model, we use the
<tt class="docutils literal">classify</tt> function. This function takes a data point
(<em>x</em><sub>1</sub>,<em>x</em><sub>2</sub>) and finds the <em>k</em>-nearest neighbors in
the training data set <tt class="docutils literal">xs</tt>. The data point (<em>x</em><sub>1</sub>,
<em>x</em><sub>2</sub>) is then classified based on which category (<tt class="docutils literal">ys</tt>)
occurs most among the <em>k</em> neighbors.</p>
<p>For the example of the gene pairs <em>yxcE</em>, <em>yxcD</em> and <em>yxiB</em>, <em>yxiA</em>, we
find:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">173.143442352</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;yxcE, yxcD:&quot;</span><span class="punctuation">,</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">x</span><span class="punctuation">)</span>
<span class="name">yxcE</span><span class="punctuation">,</span> <span class="name">yxcD</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">309</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">271.005880394</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;yxiB, yxiA:&quot;</span><span class="punctuation">,</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">x</span><span class="punctuation">)</span>
<span class="name">yxiB</span><span class="punctuation">,</span> <span class="name">yxiA</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
</pre>
<p>In agreement with the logistic regression model, <em>yxcE</em>, <em>yxcD</em> are
classified as belonging to the same operon (class OP), while <em>yxiB</em>,
<em>yxiA</em> are predicted to belong to different operons.</p>
<p>The <tt class="docutils literal">classify</tt> function lets us specify both a distance function and a
weight function as optional arguments. The distance function affects
which <em>k</em> neighbors are chosen as the nearest neighbors, as these are
defined as the neighbors with the smallest distance to the query point
(<em>x</em>, <em>y</em>). By default, the Euclidean distance is used. Instead, we
could for example use the city-block (Manhattan) distance:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">def</span> <span class="name function">cityblock</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">,</span> <span class="name">x2</span><span class="punctuation">):</span>
<span class="operator">...</span>    <span class="keyword">assert</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">)</span><span class="operator">==</span><span class="literal number integer">2</span>
<span class="operator">...</span>    <span class="keyword">assert</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">x2</span><span class="punctuation">)</span><span class="operator">==</span><span class="literal number integer">2</span>
<span class="operator">...</span>    <span class="name">distance</span> <span class="operator">=</span> <span class="name builtin">abs</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">-</span><span class="name">x2</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">])</span> <span class="operator">+</span> <span class="name builtin">abs</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span><span class="operator">-</span><span class="name">x2</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">])</span>
<span class="operator">...</span>    <span class="keyword">return</span> <span class="name">distance</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">173.143442352</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;yxcE, yxcD:&quot;</span><span class="punctuation">,</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">x</span><span class="punctuation">,</span> <span class="name">distance_fn</span> <span class="operator">=</span> <span class="name">cityblock</span><span class="punctuation">)</span>
<span class="name">yxcE</span><span class="punctuation">,</span> <span class="name">yxcD</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
</pre>
<p>The weight function can be used for weighted voting. For example, we may
want to give closer neighbors a higher weight than neighbors that are
further away:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">def</span> <span class="name function">weight</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">,</span> <span class="name">x2</span><span class="punctuation">):</span>
<span class="operator">...</span>    <span class="keyword">assert</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">)</span><span class="operator">==</span><span class="literal number integer">2</span>
<span class="operator">...</span>    <span class="keyword">assert</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">x2</span><span class="punctuation">)</span><span class="operator">==</span><span class="literal number integer">2</span>
<span class="operator">...</span>    <span class="keyword">return</span> <span class="name">exp</span><span class="punctuation">(</span><span class="operator">-</span><span class="name builtin">abs</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">-</span><span class="name">x2</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">])</span> <span class="operator">-</span> <span class="name builtin">abs</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span><span class="operator">-</span><span class="name">x2</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]))</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">173.143442352</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;yxcE, yxcD:&quot;</span><span class="punctuation">,</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">x</span><span class="punctuation">,</span> <span class="name">weight_fn</span> <span class="operator">=</span> <span class="name">weight</span><span class="punctuation">)</span>
<span class="name">yxcE</span><span class="punctuation">,</span> <span class="name">yxcD</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
</pre>
<p>By default, all neighbors are given an equal weight.</p>
<p>To find out how confident we can be in these predictions, we can call
the <tt class="docutils literal">calculate</tt> function, which will calculate the total weight
assigned to the classes OP and NOP. For the default weighting scheme,
this reduces to the number of neighbors in each category. For <em>yxcE</em>,
<em>yxcD</em>, we find</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">173.143442352</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">weight</span> <span class="operator">=</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">calculate</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">x</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;class OP: weight =&quot;</span><span class="punctuation">,</span> <span class="name">weight</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">],</span> <span class="literal string">&quot;class NOP: weight =&quot;</span><span class="punctuation">,</span> <span class="name">weight</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="keyword">class</span> <span class="name class">OP</span><span class="punctuation">:</span> <span class="name">weight</span> <span class="operator">=</span> <span class="literal number float">0.0</span> <span class="keyword">class</span> <span class="name class">NOP</span><span class="punctuation">:</span> <span class="name">weight</span> <span class="operator">=</span> <span class="literal number float">3.0</span>
</pre>
<p>which means that all three neighbors of <tt class="docutils literal">x1</tt>, <tt class="docutils literal">x2</tt> are in the NOP
class. As another example, for <em>yesK</em>, <em>yesL</em> we find</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">117</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">267.14</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">weight</span> <span class="operator">=</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">calculate</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">x</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;class OP: weight =&quot;</span><span class="punctuation">,</span> <span class="name">weight</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">],</span> <span class="literal string">&quot;class NOP: weight =&quot;</span><span class="punctuation">,</span> <span class="name">weight</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="keyword">class</span> <span class="name class">OP</span><span class="punctuation">:</span> <span class="name">weight</span> <span class="operator">=</span> <span class="literal number float">2.0</span> <span class="keyword">class</span> <span class="name class">NOP</span><span class="punctuation">:</span> <span class="name">weight</span> <span class="operator">=</span> <span class="literal number float">1.0</span>
</pre>
<p>which means that two neighbors are operon pairs and one neighbor is a
non-operon pair.</p>
<p>To get some idea of the prediction accuracy of the <em>k</em>-nearest neighbors
approach, we can apply it to the training data:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">ys</span><span class="punctuation">)):</span>
        <span class="keyword">print</span> <span class="literal string">&quot;True:&quot;</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">],</span> <span class="literal string">&quot;Predicted:&quot;</span><span class="punctuation">,</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">xs</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">])</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
</pre>
<p>showing that the prediction is correct for all but two of the gene
pairs. A more reliable estimate of the prediction accuracy can be found
from a leave-one-out analysis, in which the model is recalculated from
the training data after removing the gene to be predicted:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">ys</span><span class="punctuation">)):</span>
        <span class="name">model</span> <span class="operator">=</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">train</span><span class="punctuation">(</span><span class="name">xs</span><span class="punctuation">[:</span><span class="name">i</span><span class="punctuation">]</span><span class="operator">+</span><span class="name">xs</span><span class="punctuation">[</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">:],</span> <span class="name">ys</span><span class="punctuation">[:</span><span class="name">i</span><span class="punctuation">]</span><span class="operator">+</span><span class="name">ys</span><span class="punctuation">[</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">:])</span>
        <span class="keyword">print</span> <span class="literal string">&quot;True:&quot;</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">],</span> <span class="literal string">&quot;Predicted:&quot;</span><span class="punctuation">,</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">xs</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">])</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
</pre>
<p>The leave-one-out analysis shows that <em>k</em>-nearest neighbors model is
correct for 13 out of 17 gene pairs, which corresponds to a prediction
accuracy of 76%.</p>
</div>
</div>
<div class="section" id="naive-bayes">
<h2>16.3  Naïve Bayes</h2>
<p>This section will describe the <tt class="docutils literal">Bio.NaiveBayes</tt> module.</p>
</div>
<div class="section" id="maximum-entropy">
<h2>16.4  Maximum Entropy</h2>
<p>This section will describe the <tt class="docutils literal">Bio.MaximumEntropy</tt> module.</p>
</div>
<div class="section" id="markov-models">
<h2>16.5  Markov Models</h2>
<p>This section will describe the <tt class="docutils literal">Bio.MarkovModel</tt> and/or
<tt class="docutils literal">Bio.HMM.MarkovModel</tt> modules.</p>
</div>
</div>
<div class="section" id="chapter-17-graphics-including-genomediagram">
<h1>Chapter 17  Graphics including GenomeDiagram</h1>
<p>The <tt class="docutils literal">Bio.Graphics</tt> module depends on the third party Python library
<a class="reference external" href="http://www.reportlab.org">ReportLab</a>. Although focused on producing
PDF files, ReportLab can also create encapsulated postscript (EPS) and
(SVG) files. In addition to these vector based images, provided certain
further dependencies such as the <a class="reference external" href="http://www.pythonware.com/products/pil/">Python Imaging Library
(PIL)</a> are installed,
ReportLab can also output bitmap images (including JPEG, PNG, GIF, BMP
and PICT formats).</p>
<div class="section" id="genomediagram">
<h2>17.1  GenomeDiagram</h2>
<div class="section" id="id18">
<h3>17.1.1  Introduction</h3>
<p>The <tt class="docutils literal">Bio.Graphics.GenomeDiagram</tt> module was added to Biopython 1.50,
having previously been available as a separate Python module dependent
on Biopython. GenomeDiagram is described in the Bioinformatics journal
publication by Pritchard et al. (2006) [<a class="reference external" href="#pritchard2006">2</a>], which
includes some examples images. There is a PDF copy of the old manual
here,
<a class="reference external" href="http://biopython.org/DIST/docs/GenomeDiagram/userguide.pdf">http://biopython.org/DIST/docs/GenomeDiagram/userguide.pdf</a>
which has some more examples. As the name might suggest, GenomeDiagram
was designed for drawing whole genomes, in particular prokaryotic
genomes, either as linear diagrams (optionally broken up into fragments
to fit better) or as circular wheel diagrams. Have a look at Figure 2 in
Toth <em>et al.</em> (2006) [<a class="reference external" href="#toth2006">3</a>] for a good example. It proved
also well suited to drawing quite detailed figures for smaller genomes
such as phage, plasmids or mitochrondia, for example see Figures 1 and 2
in Van der Auwera <em>et al.</em> (2009) [<a class="reference external" href="#vanderauwera2009">4</a>] (shown
with additional manual editing).</p>
<p>This module is easiest to use if you have your genome loaded as a
<tt class="docutils literal">SeqRecord</tt> object containing lots of <tt class="docutils literal">SeqFeature</tt> objects - for
example as loaded from a GenBank file (see
Chapters <a class="reference external" href="#chapter:SeqRecord">4</a> and <a class="reference external" href="#chapter:Bio.SeqIO">5</a>).</p>
</div>
<div class="section" id="diagrams-tracks-feature-sets-and-features">
<h3>17.1.2  Diagrams, tracks, feature-sets and features</h3>
<p>GenomeDiagram uses a nested set of objects. At the top level, you have a
diagram object representing a sequence (or sequence region) along the
horizontal axis (or circle). A diagram can contain one or more tracks,
shown stacked vertically (or radially on circular diagrams). These will
typically all have the same length and represent the same sequence
region. You might use one track to show the gene locations, another to
show regulatory regions, and a third track to show the GC percentage.
The most commonly used type of track will contain features, bundled
together in feature-sets. You might choose to use one feature-set for
all your CDS features, and another for tRNA features. This isn’t
required - they can all go in the same feature-set, but it makes it
easier to update the properties of just selected features (e.g. make all
the tRNA features red).</p>
<p>There are two main ways to build up a complete diagram. Firstly, the top
down approach where you create a diagram object, and then using its
methods add track(s), and use the track methods to add feature-set(s),
and use their methods to add the features. Secondly, you can create the
individual objects separately (in whatever order suits your code), and
then combine them.</p>
</div>
<div class="section" id="a-top-down-example">
<h3>17.1.3  A top down example</h3>
<p>We’re going to draw a whole genome from a <tt class="docutils literal">SeqRecord</tt> object read in
from a GenBank file (see Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>). This
example uses the pPCP1 plasmid from <em>Yersinia pestis biovar Microtus</em>,
the file is included with the Biopython unit tests under the GenBank
folder, or online
<a class="reference external" href="http://biopython.org/SRC/biopython/Tests/GenBank/NC_005816.gb">NC_005816.gb</a>
from our website.</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib</span> <span class="keyword namespace">import</span> <span class="name">colors</span>
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.units</span> <span class="keyword namespace">import</span> <span class="name">cm</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">GenomeDiagram</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
</pre>
<p>We’re using a top down approach, so after loading in our sequence we
next create an empty diagram, then add an (empty) track, and to that add
an (empty) feature set:</p>
<pre class="code python literal-block">
<span class="name">gd_diagram</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Diagram</span><span class="punctuation">(</span><span class="literal string">&quot;Yersinia pestis biovar Microtus plasmid pPCP1&quot;</span><span class="punctuation">)</span>
<span class="name">gd_track_for_features</span> <span class="operator">=</span> <span class="name">gd_diagram</span><span class="operator">.</span><span class="name">new_track</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;Annotated Features&quot;</span><span class="punctuation">)</span>
<span class="name">gd_feature_set</span> <span class="operator">=</span> <span class="name">gd_track_for_features</span><span class="operator">.</span><span class="name">new_set</span><span class="punctuation">()</span>
</pre>
<p>Now the fun part - we take each gene <tt class="docutils literal">SeqFeature</tt> object in our
<tt class="docutils literal">SeqRecord</tt>, and use it to generate a feature on the diagram. We’re
going to color them blue, alternating between a dark blue and a light
blue.</p>
<pre class="code python literal-block">
<span class="keyword">for</span> <span class="name">feature</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">:</span>
    <span class="keyword">if</span> <span class="name">feature</span><span class="operator">.</span><span class="name">type</span> <span class="operator">!=</span> <span class="literal string">&quot;gene&quot;</span><span class="punctuation">:</span>
        <span class="comment">#Exclude this feature</span>
        <span class="keyword">continue</span>
    <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">gd_feature_set</span><span class="punctuation">)</span> <span class="operator">%</span> <span class="literal number integer">2</span> <span class="operator">==</span> <span class="literal number integer">0</span><span class="punctuation">:</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">blue</span>
    <span class="keyword">else</span><span class="punctuation">:</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">lightblue</span>
    <span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
</pre>
<p>Now we come to actually making the output file. This happens in two
steps, first we call the <tt class="docutils literal">draw</tt> method, which creates all the shapes
using ReportLab objects. Then we call the <tt class="docutils literal">write</tt> method which renders
these to the requested file format. Note you can output in multiple file
formats:</p>
<pre class="code python literal-block">
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;linear&quot;</span><span class="punctuation">,</span> <span class="name">orientation</span><span class="operator">=</span><span class="literal string">&quot;landscape&quot;</span><span class="punctuation">,</span> <span class="name">pagesize</span><span class="operator">=</span><span class="literal string">'A4'</span><span class="punctuation">,</span>
                <span class="name">fragments</span><span class="operator">=</span><span class="literal number integer">4</span><span class="punctuation">,</span> <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">))</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PDF&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear.eps&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;EPS&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear.svg&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;SVG&quot;</span><span class="punctuation">)</span>
</pre>
<p>Also, provided you have the dependencies installed, you can also do
bitmaps, for example:</p>
<pre class="code python literal-block">
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear.png&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PNG&quot;</span><span class="punctuation">)</span>
</pre>
<p><img alt="image13" src="../images/plasmid_linear.png" /></p>
<p>Notice that the <tt class="docutils literal">fragments</tt> argument which we set to four controls how
many pieces the genome gets broken up into.</p>
<p>If you want to do a circular figure, then try this:</p>
<pre class="code python literal-block">
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;circular&quot;</span><span class="punctuation">,</span> <span class="name">circular</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">pagesize</span><span class="operator">=</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">,</span><span class="literal number integer">20</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">),</span>
                <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">),</span> <span class="name">circle_core</span><span class="operator">=</span><span class="literal number float">0.7</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_circular.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PDF&quot;</span><span class="punctuation">)</span>
</pre>
<p><img alt="image14" src="../images/plasmid_circular.png" /></p>
<p>These figures are not very exciting, but we’ve only just got started.</p>
</div>
<div class="section" id="a-bottom-up-example">
<h3>17.1.4  A bottom up example</h3>
<p>Now let’s produce exactly the same figures, but using the bottom up
approach. This means we create the different objects directly (and this
can be done in almost any order) and then combine them.</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib</span> <span class="keyword namespace">import</span> <span class="name">colors</span>
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.units</span> <span class="keyword namespace">import</span> <span class="name">cm</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">GenomeDiagram</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>

<span class="comment">#Create the feature set and its feature objects,</span>
<span class="name">gd_feature_set</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">FeatureSet</span><span class="punctuation">()</span>
<span class="keyword">for</span> <span class="name">feature</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">:</span>
    <span class="keyword">if</span> <span class="name">feature</span><span class="operator">.</span><span class="name">type</span> <span class="operator">!=</span> <span class="literal string">&quot;gene&quot;</span><span class="punctuation">:</span>
        <span class="comment">#Exclude this feature</span>
        <span class="keyword">continue</span>
    <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">gd_feature_set</span><span class="punctuation">)</span> <span class="operator">%</span> <span class="literal number integer">2</span> <span class="operator">==</span> <span class="literal number integer">0</span><span class="punctuation">:</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">blue</span>
    <span class="keyword">else</span><span class="punctuation">:</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">lightblue</span>
    <span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="comment">#(this for loop is the same as in the previous example)</span>

<span class="comment">#Create a track, and a diagram</span>
<span class="name">gd_track_for_features</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Track</span><span class="punctuation">(</span><span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;Annotated Features&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Diagram</span><span class="punctuation">(</span><span class="literal string">&quot;Yersinia pestis biovar Microtus plasmid pPCP1&quot;</span><span class="punctuation">)</span>

<span class="comment">#Now have to glue the bits together...</span>
<span class="name">gd_track_for_features</span><span class="operator">.</span><span class="name">add_set</span><span class="punctuation">(</span><span class="name">gd_feature_set</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">add_track</span><span class="punctuation">(</span><span class="name">gd_track_for_features</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>You can now call the <tt class="docutils literal">draw</tt> and <tt class="docutils literal">write</tt> methods as before to produce
a linear or circular diagram, using the code at the end of the top-down
example above. The figures should be identical.</p>
</div>
<div class="section" id="features-without-a-seqfeature">
<h3>17.1.5  Features without a SeqFeature</h3>
<p>In the above example we used a <tt class="docutils literal">SeqRecord</tt>’s <tt class="docutils literal">SeqFeature</tt> objects
to build our diagram (see also Section <a class="reference external" href="#sec:seq_features">4.3</a>).
Sometimes you won’t have <tt class="docutils literal">SeqFeature</tt> objects, but just the
coordinates for a feature you want to draw. You have to create minimal
<tt class="docutils literal">SeqFeature</tt> object, but this is easy:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqFeature</span> <span class="keyword namespace">import</span> <span class="name">SeqFeature</span><span class="punctuation">,</span> <span class="name">FeatureLocation</span>
<span class="name">my_seq_feature</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="punctuation">(</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="literal number integer">50</span><span class="punctuation">,</span><span class="literal number integer">100</span><span class="punctuation">),</span><span class="name">strand</span><span class="operator">=+</span><span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>For strand, use <tt class="docutils literal">+1</tt> for the forward strand, <tt class="docutils literal"><span class="pre">-1</span></tt> for the reverse
strand, and <tt class="docutils literal">None</tt> for both. Here is a short self contained example:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqFeature</span> <span class="keyword namespace">import</span> <span class="name">SeqFeature</span><span class="punctuation">,</span> <span class="name">FeatureLocation</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">GenomeDiagram</span>
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.units</span> <span class="keyword namespace">import</span> <span class="name">cm</span>

<span class="name">gdd</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Diagram</span><span class="punctuation">(</span><span class="literal string">'Test Diagram'</span><span class="punctuation">)</span>
<span class="name">gdt_features</span> <span class="operator">=</span> <span class="name">gdd</span><span class="operator">.</span><span class="name">new_track</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">greytrack</span><span class="operator">=</span><span class="name builtin pseudo">False</span><span class="punctuation">)</span>
<span class="name">gds_features</span> <span class="operator">=</span> <span class="name">gdt_features</span><span class="operator">.</span><span class="name">new_set</span><span class="punctuation">()</span>

<span class="comment">#Add three features to show the strand options,</span>
<span class="name">feature</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="punctuation">(</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="literal number integer">25</span><span class="punctuation">,</span> <span class="literal number integer">125</span><span class="punctuation">),</span> <span class="name">strand</span><span class="operator">=+</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="name">gds_features</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;Forward&quot;</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="name">feature</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="punctuation">(</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="literal number integer">150</span><span class="punctuation">,</span> <span class="literal number integer">250</span><span class="punctuation">),</span> <span class="name">strand</span><span class="operator">=</span><span class="name builtin pseudo">None</span><span class="punctuation">)</span>
<span class="name">gds_features</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;Strandless&quot;</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="name">feature</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="punctuation">(</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="literal number integer">275</span><span class="punctuation">,</span> <span class="literal number integer">375</span><span class="punctuation">),</span> <span class="name">strand</span><span class="operator">=-</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="name">gds_features</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;Reverse&quot;</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>

<span class="name">gdd</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">format</span><span class="operator">=</span><span class="literal string">'linear'</span><span class="punctuation">,</span> <span class="name">pagesize</span><span class="operator">=</span><span class="punctuation">(</span><span class="literal number integer">15</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">,</span><span class="literal number integer">4</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">),</span> <span class="name">fragments</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">,</span>
         <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="literal number integer">400</span><span class="punctuation">)</span>
<span class="name">gdd</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;GD_labels_default.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;pdf&quot;</span><span class="punctuation">)</span>
</pre>
<p>The top part of the image in the next subsection shows the output (in
the default feature color, pale green).</p>
<p>Notice that we have used the <tt class="docutils literal">name</tt> argument here to specify the
caption text for these features. This is discussed in more detail next.</p>
</div>
<div class="section" id="feature-captions">
<h3>17.1.6  Feature captions</h3>
<p>Recall we used the following (where <tt class="docutils literal">feature</tt> was a <tt class="docutils literal">SeqFeature</tt>
object) to add a feature to the diagram:</p>
<pre class="code python literal-block">
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
</pre>
<p>In the example above the <tt class="docutils literal">SeqFeature</tt> annotation was used to pick a
sensible caption for the features. By default the following possible
entries under the <tt class="docutils literal">SeqFeature</tt> object’s qualifiers dictionary are
used: <tt class="docutils literal">gene</tt>, <tt class="docutils literal">label</tt>, <tt class="docutils literal">name</tt>, <tt class="docutils literal">locus_tag</tt>, and <tt class="docutils literal">product</tt>.
More simply, you can specify a name directly:</p>
<pre class="code python literal-block">
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;My Gene&quot;</span><span class="punctuation">)</span>
</pre>
<p>In addition to the caption text for each feature’s label, you can also
choose the font, position (this defaults to the start of the sigil, you
can also choose the middle or at the end) and orientation (for linear
diagrams only, where this defaults to rotated by 45 degrees):</p>
<pre class="code python literal-block">
<span class="comment">#Large font, parallel with the track</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;green&quot;</span><span class="punctuation">,</span>
                           <span class="name">label_size</span><span class="operator">=</span><span class="literal number integer">25</span><span class="punctuation">,</span> <span class="name">label_angle</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">)</span>

<span class="comment">#Very small font, perpendicular to the track (towards it)</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;purple&quot;</span><span class="punctuation">,</span>
                           <span class="name">label_position</span><span class="operator">=</span><span class="literal string">&quot;end&quot;</span><span class="punctuation">,</span>
                           <span class="name">label_size</span><span class="operator">=</span><span class="literal number integer">4</span><span class="punctuation">,</span> <span class="name">label_angle</span><span class="operator">=</span><span class="literal number integer">90</span><span class="punctuation">)</span>

<span class="comment">#Small font, perpendicular to the track (away from it)</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;blue&quot;</span><span class="punctuation">,</span>
                           <span class="name">label_position</span><span class="operator">=</span><span class="literal string">&quot;middle&quot;</span><span class="punctuation">,</span>
                           <span class="name">label_size</span><span class="operator">=</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="name">label_angle</span><span class="operator">=-</span><span class="literal number integer">90</span><span class="punctuation">)</span>
</pre>
<p>Combining each of these three fragments with the complete example in the
previous section should give something like this:</p>
<p><img alt="image15" src="../images/GD_sigil_labels.png" /></p>
<p>We’ve not shown it here, but you can also set <tt class="docutils literal">label_color</tt> to control
the label’s color (used in Section <a class="reference external" href="#sec:gd_nice_example">17.1.9</a>).</p>
<p>You’ll notice the default font is quite small - this makes sense because
you will usually be drawing many (small) features on a page, not just a
few large ones as shown here.</p>
</div>
<div class="section" id="feature-sigils">
<h3>17.1.7  Feature sigils</h3>
<p>The examples above have all just used the default sigil for the feature,
a plain box, which was all that was available in the last publicly
released standalone version of GenomeDiagram. Arrow sigils were included
when GenomeDiagram was added to Biopython 1.50:</p>
<pre class="code python literal-block">
<span class="comment">#Default uses a BOX sigil</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">)</span>

<span class="comment">#You can make this explicit:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;BOX&quot;</span><span class="punctuation">)</span>

<span class="comment">#Or opt for an arrow:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">)</span>
</pre>
<p>Biopython 1.61 added three more sigils,</p>
<pre class="code python literal-block">
<span class="comment">#Box with corners cut off (making it an octagon)</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;OCTO&quot;</span><span class="punctuation">)</span>

<span class="comment">#Box with jagged edges (useful for showing breaks in contains)</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;JAGGY&quot;</span><span class="punctuation">)</span>

<span class="comment">#Arrow which spans the axis with strand used only for direction</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;BIGARROW&quot;</span><span class="punctuation">)</span>
</pre>
<p>These are shown below. Most sigils fit into a bounding box (as given by
the default BOX sigil), either above or below the axis for the forward
or reverse strand, or straddling it (double the height) for strand-less
features. The BIGARROW sigil is different, always straddling the axis
with the direction taken from the feature’s stand.</p>
<p><img alt="image16" src="../images/GD_sigils.png" /></p>
</div>
<div class="section" id="arrow-sigils">
<h3>17.1.8  Arrow sigils</h3>
<p>We introduced the arrow sigils in the previous section. There are two
additional options to adjust the shapes of the arrows, firstly the
thickness of the arrow shaft, given as a proportion of the height of the
bounding box:</p>
<pre class="code python literal-block">
<span class="comment">#Full height shafts, giving pointed boxes:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;brown&quot;</span><span class="punctuation">,</span>
                           <span class="name">arrowshaft_height</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">)</span>
<span class="comment">#Or, thin shafts:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;teal&quot;</span><span class="punctuation">,</span>
                           <span class="name">arrowshaft_height</span><span class="operator">=</span><span class="literal number float">0.2</span><span class="punctuation">)</span>
<span class="comment">#Or, very thin shafts:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;darkgreen&quot;</span><span class="punctuation">,</span>
                           <span class="name">arrowshaft_height</span><span class="operator">=</span><span class="literal number float">0.1</span><span class="punctuation">)</span>
</pre>
<p>The results are shown below:</p>
<p><img alt="image17" src="../images/GD_sigil_arrow_shafts.png" /></p>
<p>Secondly, the length of the arrow head - given as a proportion of the
height of the bounding box (defaulting to 0.5, or 50%):</p>
<pre class="code python literal-block">
<span class="comment">#Short arrow heads:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;blue&quot;</span><span class="punctuation">,</span>
                           <span class="name">arrowhead_length</span><span class="operator">=</span><span class="literal number float">0.25</span><span class="punctuation">)</span>
<span class="comment">#Or, longer arrow heads:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;orange&quot;</span><span class="punctuation">,</span>
                           <span class="name">arrowhead_length</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="comment">#Or, very very long arrow heads (i.e. all head, no shaft, so triangles):</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;red&quot;</span><span class="punctuation">,</span>
                           <span class="name">arrowhead_length</span><span class="operator">=</span><span class="literal number integer">10000</span><span class="punctuation">)</span>
</pre>
<p>The results are shown below:</p>
<p><img alt="image18" src="../images/GD_sigil_arrow_heads.png" /></p>
<p>Biopython 1.61 adds a new <tt class="docutils literal">BIGARROW</tt> sigil which always stradles the
axis, pointing left for the reverse strand or right otherwise:</p>
<pre class="code python literal-block">
<span class="comment">#A large arrow straddling the axis:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;BIGARROW&quot;</span><span class="punctuation">)</span>
</pre>
<p>All the shaft and arrow head options shown above for the <tt class="docutils literal">ARROW</tt> sigil
can be used for the <tt class="docutils literal">BIGARROW</tt> sigil too.</p>
</div>
<div class="section" id="a-nice-example">
<h3>17.1.9  A nice example</h3>
<p>Now let’s return to the pPCP1 plasmid from <em>Yersinia pestis biovar
Microtus</em>, and the top down approach used in
Section <a class="reference external" href="#sec:gd_top_down">17.1.3</a>, but take advantage of the sigil
options we’ve now discussed. This time we’ll use arrows for the genes,
and overlay them with strand-less features (as plain boxes) showing the
position of some restriction digest sites.</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib</span> <span class="keyword namespace">import</span> <span class="name">colors</span>
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.units</span> <span class="keyword namespace">import</span> <span class="name">cm</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">GenomeDiagram</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqFeature</span> <span class="keyword namespace">import</span> <span class="name">SeqFeature</span><span class="punctuation">,</span> <span class="name">FeatureLocation</span>

<span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>

<span class="name">gd_diagram</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Diagram</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>
<span class="name">gd_track_for_features</span> <span class="operator">=</span> <span class="name">gd_diagram</span><span class="operator">.</span><span class="name">new_track</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;Annotated Features&quot;</span><span class="punctuation">)</span>
<span class="name">gd_feature_set</span> <span class="operator">=</span> <span class="name">gd_track_for_features</span><span class="operator">.</span><span class="name">new_set</span><span class="punctuation">()</span>

<span class="keyword">for</span> <span class="name">feature</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">:</span>
    <span class="keyword">if</span> <span class="name">feature</span><span class="operator">.</span><span class="name">type</span> <span class="operator">!=</span> <span class="literal string">&quot;gene&quot;</span><span class="punctuation">:</span>
        <span class="comment">#Exclude this feature</span>
        <span class="keyword">continue</span>
    <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">gd_feature_set</span><span class="punctuation">)</span> <span class="operator">%</span> <span class="literal number integer">2</span> <span class="operator">==</span> <span class="literal number integer">0</span><span class="punctuation">:</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">blue</span>
    <span class="keyword">else</span><span class="punctuation">:</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">lightblue</span>
    <span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span>
                               <span class="name">color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span>
                               <span class="name">label_size</span> <span class="operator">=</span> <span class="literal number integer">14</span><span class="punctuation">,</span> <span class="name">label_angle</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">)</span>

<span class="comment">#I want to include some strandless features, so for an example</span>
<span class="comment">#will use EcoRI recognition sites etc.</span>
<span class="keyword">for</span> <span class="name">site</span><span class="punctuation">,</span> <span class="name">name</span><span class="punctuation">,</span> <span class="name">color</span> <span class="operator word">in</span> <span class="punctuation">[(</span><span class="literal string">&quot;GAATTC&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;EcoRI&quot;</span><span class="punctuation">,</span><span class="name">colors</span><span class="operator">.</span><span class="name">green</span><span class="punctuation">),</span>
                          <span class="punctuation">(</span><span class="literal string">&quot;CCCGGG&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;SmaI&quot;</span><span class="punctuation">,</span><span class="name">colors</span><span class="operator">.</span><span class="name">orange</span><span class="punctuation">),</span>
                          <span class="punctuation">(</span><span class="literal string">&quot;AAGCTT&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;HindIII&quot;</span><span class="punctuation">,</span><span class="name">colors</span><span class="operator">.</span><span class="name">red</span><span class="punctuation">),</span>
                          <span class="punctuation">(</span><span class="literal string">&quot;GGATCC&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;BamHI&quot;</span><span class="punctuation">,</span><span class="name">colors</span><span class="operator">.</span><span class="name">purple</span><span class="punctuation">)]:</span>
    <span class="name">index</span> <span class="operator">=</span> <span class="literal number integer">0</span>
    <span class="keyword">while</span> <span class="name builtin pseudo">True</span><span class="punctuation">:</span>
        <span class="name">index</span>  <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">find</span><span class="punctuation">(</span><span class="name">site</span><span class="punctuation">,</span> <span class="name">start</span><span class="operator">=</span><span class="name">index</span><span class="punctuation">)</span>
        <span class="keyword">if</span> <span class="name">index</span> <span class="operator">==</span> <span class="operator">-</span><span class="literal number integer">1</span> <span class="punctuation">:</span> <span class="keyword">break</span>
        <span class="name">feature</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="punctuation">(</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="name">index</span><span class="punctuation">,</span> <span class="name">index</span><span class="operator">+</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">site</span><span class="punctuation">)))</span>
        <span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="name">name</span><span class="punctuation">,</span>
                                   <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">label_size</span> <span class="operator">=</span> <span class="literal number integer">10</span><span class="punctuation">,</span>
                                   <span class="name">label_color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">)</span>
        <span class="name">index</span> <span class="operator">+=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">site</span><span class="punctuation">)</span>

<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;linear&quot;</span><span class="punctuation">,</span> <span class="name">pagesize</span><span class="operator">=</span><span class="literal string">'A4'</span><span class="punctuation">,</span> <span class="name">fragments</span><span class="operator">=</span><span class="literal number integer">4</span><span class="punctuation">,</span>
                <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">))</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear_nice.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PDF&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear_nice.eps&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;EPS&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear_nice.svg&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;SVG&quot;</span><span class="punctuation">)</span>

<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;circular&quot;</span><span class="punctuation">,</span> <span class="name">circular</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">pagesize</span><span class="operator">=</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">,</span><span class="literal number integer">20</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">),</span>
                <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">),</span> <span class="name">circle_core</span> <span class="operator">=</span> <span class="literal number float">0.5</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_circular_nice.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PDF&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_circular_nice.eps&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;EPS&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_circular_nice.svg&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;SVG&quot;</span><span class="punctuation">)</span>
</pre>
<p>And the output:</p>
<p><img alt="image19" src="../images/plasmid_linear_nice.png" /></p>
<p><img alt="image20" src="../images/plasmid_circular_nice.png" /></p>
</div>
<div class="section" id="multiple-tracks">
<h3>17.1.10  Multiple tracks</h3>
<p>All the examples so far have used a single track, but you can have more
than one track – for example show the genes on one, and repeat regions
on another. In this example we’re going to show three phage genomes side
by side to scale, inspired by Figure 6 in Proux <em>e</em>t al. (2002)
[<a class="reference external" href="#proux2002">5</a>]. We’ll need the GenBank files for the following
three phage:</p>
<ul class="simple">
<li><tt class="docutils literal">NC_002703</tt> – Lactococcus phage Tuc2009, complete genome (38347 bp)</li>
<li><tt class="docutils literal">AF323668</tt> – Bacteriophage bIL285, complete genome (35538 bp)</li>
<li><tt class="docutils literal">NC_003212</tt> – <em>Listeria innocua</em> Clip11262, complete genome, of
which we are focussing only on integrated prophage 5 (similar
length).</li>
</ul>
<p>You can download these using Entrez if you like, see
Section <a class="reference external" href="#sec:efetch">9.6</a> for more details. For the third record
we’ve worked out where the phage is integrated into the genome, and
slice the record to extract it (with the features preserved, see
Section <a class="reference external" href="#sec:SeqRecord-slicing">4.6</a>), and must also reverse
complement to match the orientation of the first two phage (again
preserving the features, see
Section <a class="reference external" href="#sec:SeqRecord-reverse-complement">4.8</a>):</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>

<span class="name">A_rec</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_002703.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">)</span>
<span class="name">B_rec</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;AF323668.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">)</span>
<span class="name">C_rec</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_003212.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">)[</span><span class="literal number integer">2587879</span><span class="punctuation">:</span><span class="literal number integer">2625807</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name">name</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
</pre>
<p>The figure we are imitating used different colors for different gene
functions. One way to do this is to edit the GenBank file to record
color preferences for each feature - something <a class="reference external" href="http://www.sanger.ac.uk/resources/software/artemis/">Sanger’s Artemis
editor</a> does,
and which GenomeDiagram should understand. Here however, we’ll just hard
code three lists of colors.</p>
<p>Note that the annotation in the GenBank files doesn’t exactly match that
shown in Proux <em>et al.</em>, they have drawn some unannotated genes.</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.colors</span> <span class="keyword namespace">import</span> <span class="name">red</span><span class="punctuation">,</span> <span class="name">grey</span><span class="punctuation">,</span> <span class="name">orange</span><span class="punctuation">,</span> <span class="name">green</span><span class="punctuation">,</span> <span class="name">brown</span><span class="punctuation">,</span> <span class="name">blue</span><span class="punctuation">,</span> <span class="name">lightblue</span><span class="punctuation">,</span> <span class="name">purple</span>

<span class="name">A_colors</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">red</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">7</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">orange</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">orange</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">11</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">green</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">4</span> \
         <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">green</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">,</span> <span class="name">green</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">brown</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">blue</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">4</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">lightblue</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span> \
         <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">,</span> <span class="name">lightblue</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">purple</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span>
<span class="name">B_colors</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">red</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">6</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">8</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">orange</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">orange</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">21</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">green</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span> \
         <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">brown</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">4</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">blue</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">3</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">lightblue</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">3</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">purple</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span>
<span class="name">C_colors</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">30</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">green</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">brown</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">4</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">blue</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">,</span> <span class="name">blue</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">lightblue</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> \
         <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span>
</pre>
<p>Now to draw them – this time we add three tracks to the diagram, and
also notice they are given different start/end values to reflect their
different lengths (this requires Biopython 1.59 or later).</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">GenomeDiagram</span>

<span class="name">name</span> <span class="operator">=</span> <span class="literal string">&quot;Proux Fig 6&quot;</span>
<span class="name">gd_diagram</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Diagram</span><span class="punctuation">(</span><span class="name">name</span><span class="punctuation">)</span>
<span class="name">max_len</span> <span class="operator">=</span> <span class="literal number integer">0</span>
<span class="keyword">for</span> <span class="name">record</span><span class="punctuation">,</span> <span class="name">gene_colors</span> <span class="operator word">in</span> <span class="name builtin">zip</span><span class="punctuation">([</span><span class="name">A_rec</span><span class="punctuation">,</span> <span class="name">B_rec</span><span class="punctuation">,</span> <span class="name">C_rec</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="name">A_colors</span><span class="punctuation">,</span> <span class="name">B_colors</span><span class="punctuation">,</span> <span class="name">C_colors</span><span class="punctuation">]):</span>
    <span class="name">max_len</span> <span class="operator">=</span> <span class="name builtin">max</span><span class="punctuation">(</span><span class="name">max_len</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">))</span>
    <span class="name">gd_track_for_features</span> <span class="operator">=</span> <span class="name">gd_diagram</span><span class="operator">.</span><span class="name">new_track</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span>
                            <span class="name">name</span><span class="operator">=</span><span class="name">record</span><span class="operator">.</span><span class="name">name</span><span class="punctuation">,</span>
                            <span class="name">greytrack</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span>
                            <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">))</span>
    <span class="name">gd_feature_set</span> <span class="operator">=</span> <span class="name">gd_track_for_features</span><span class="operator">.</span><span class="name">new_set</span><span class="punctuation">()</span>

    <span class="name">i</span> <span class="operator">=</span> <span class="literal number integer">0</span>
    <span class="keyword">for</span> <span class="name">feature</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">:</span>
        <span class="keyword">if</span> <span class="name">feature</span><span class="operator">.</span><span class="name">type</span> <span class="operator">!=</span> <span class="literal string">&quot;gene&quot;</span><span class="punctuation">:</span>
            <span class="comment">#Exclude this feature</span>
            <span class="keyword">continue</span>
        <span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span>
                                   <span class="name">color</span><span class="operator">=</span><span class="name">gene_colors</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">],</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span>
                                   <span class="name">name</span> <span class="operator">=</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">),</span>
                                   <span class="name">label_position</span><span class="operator">=</span><span class="literal string">&quot;start&quot;</span><span class="punctuation">,</span>
                                   <span class="name">label_size</span> <span class="operator">=</span> <span class="literal number integer">6</span><span class="punctuation">,</span> <span class="name">label_angle</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">)</span>
        <span class="name">i</span><span class="operator">+=</span><span class="literal number integer">1</span>

<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;linear&quot;</span><span class="punctuation">,</span> <span class="name">pagesize</span><span class="operator">=</span><span class="literal string">'A4'</span><span class="punctuation">,</span> <span class="name">fragments</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">,</span>
                <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="name">max_len</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">name</span> <span class="operator">+</span> <span class="literal string">&quot;.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PDF&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">name</span> <span class="operator">+</span> <span class="literal string">&quot;.eps&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;EPS&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">name</span> <span class="operator">+</span> <span class="literal string">&quot;.svg&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;SVG&quot;</span><span class="punctuation">)</span>
</pre>
<p>The result:</p>
<p><img alt="image21" src="../images/three_track_simple.png" /></p>
<p>I did wonder why in the original manuscript there were no red or orange
genes marked in the bottom phage. Another important point is here the
phage are shown with different lengths - this is because they are all
drawn to the same scale (they <em>are</em> different lengths).</p>
<p>The key difference from the published figure is they have color-coded
links between similar proteins – which is what we will do in the next
section.</p>
</div>
<div class="section" id="cross-links-between-tracks">
<h3>17.1.11  Cross-Links between tracks</h3>
<p>Biopython 1.59 added the ability to draw cross links between tracks -
both simple linear diagrams as we will show here, but also linear
diagrams split into fragments and circular diagrams.</p>
<p>Continuing the example from the previous section inspired by Figure 6
from Proux <em>et al.</em> 2002 [<a class="reference external" href="#proux2002">5</a>], we would need a list of
cross links between pairs of genes, along with a score or color to use.
Realistically you might extract this from a BLAST file computationally,
but here I have manually typed them in.</p>
<p>My naming convention continues to refer to the three phage as A, B and
C. Here are the links we want to show between A and B, given as a list
of tuples (percentage similarity score, gene in A, gene in B).</p>
<pre class="code python literal-block">
<span class="comment">#Tuc2009 (NC_002703) vs bIL285 (AF323668)</span>
<span class="name">A_vs_B</span> <span class="operator">=</span> <span class="punctuation">[</span>
    <span class="punctuation">(</span><span class="literal number integer">99</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_01&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;int&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">33</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_03&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf4&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">94</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_05&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf6&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">100</span><span class="punctuation">,</span><span class="literal string">&quot;Tuc2009_06&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf7&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">97</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_07&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf8&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">98</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_08&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf9&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">98</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_09&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf10&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">100</span><span class="punctuation">,</span><span class="literal string">&quot;Tuc2009_10&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf12&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">100</span><span class="punctuation">,</span><span class="literal string">&quot;Tuc2009_11&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf13&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">94</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_12&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf14&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">87</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_13&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf15&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">94</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_14&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf16&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">94</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_15&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf17&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">88</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_17&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;rusA&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">91</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_18&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf20&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">93</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_19&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf22&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">71</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_20&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf23&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">51</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_22&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf27&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">97</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_23&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf28&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">88</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_24&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf29&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_26&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf38&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">19</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_46&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf52&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">77</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_48&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf54&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">91</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_49&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf55&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">95</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_52&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf60&quot;</span><span class="punctuation">),</span>
<span class="punctuation">]</span>
</pre>
<p>Likewise for B and C:</p>
<pre class="code python literal-block">
<span class="comment">#bIL285 (AF323668) vs Listeria innocua prophage 5 (in NC_003212)</span>
<span class="name">B_vs_C</span> <span class="operator">=</span> <span class="punctuation">[</span>
    <span class="punctuation">(</span><span class="literal number integer">42</span><span class="punctuation">,</span> <span class="literal string">&quot;orf39&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2581&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">31</span><span class="punctuation">,</span> <span class="literal string">&quot;orf40&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2580&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">49</span><span class="punctuation">,</span> <span class="literal string">&quot;orf41&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2579&quot;</span><span class="punctuation">),</span> <span class="comment">#terL</span>
    <span class="punctuation">(</span><span class="literal number integer">54</span><span class="punctuation">,</span> <span class="literal string">&quot;orf42&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2578&quot;</span><span class="punctuation">),</span> <span class="comment">#portal</span>
    <span class="punctuation">(</span><span class="literal number integer">55</span><span class="punctuation">,</span> <span class="literal string">&quot;orf43&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2577&quot;</span><span class="punctuation">),</span> <span class="comment">#protease</span>
    <span class="punctuation">(</span><span class="literal number integer">33</span><span class="punctuation">,</span> <span class="literal string">&quot;orf44&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2576&quot;</span><span class="punctuation">),</span> <span class="comment">#mhp</span>
    <span class="punctuation">(</span><span class="literal number integer">51</span><span class="punctuation">,</span> <span class="literal string">&quot;orf46&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2575&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">33</span><span class="punctuation">,</span> <span class="literal string">&quot;orf47&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2574&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal string">&quot;orf48&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2573&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">25</span><span class="punctuation">,</span> <span class="literal string">&quot;orf49&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2572&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">50</span><span class="punctuation">,</span> <span class="literal string">&quot;orf50&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2571&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">48</span><span class="punctuation">,</span> <span class="literal string">&quot;orf51&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2570&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">24</span><span class="punctuation">,</span> <span class="literal string">&quot;orf52&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2568&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">30</span><span class="punctuation">,</span> <span class="literal string">&quot;orf53&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2567&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">28</span><span class="punctuation">,</span> <span class="literal string">&quot;orf54&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2566&quot;</span><span class="punctuation">),</span>
<span class="punctuation">]</span>
</pre>
<p>For the first and last phage these identifiers are locus tags, for the
middle phage there are no locus tags so I’ve used gene names instead.
The following little helper function lets us lookup a feature using
either a locus tag or gene name:</p>
<pre class="code python literal-block">
<span class="keyword">def</span> <span class="name function">get_feature</span><span class="punctuation">(</span><span class="name">features</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="punctuation">,</span> <span class="name">tags</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">&quot;locus_tag&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;gene&quot;</span><span class="punctuation">]):</span>
    <span class="literal string doc">&quot;&quot;&quot;Search list of SeqFeature objects for an identifier under the given tags.&quot;&quot;&quot;</span>
    <span class="keyword">for</span> <span class="name">f</span> <span class="operator word">in</span> <span class="name">features</span><span class="punctuation">:</span>
        <span class="keyword">for</span> <span class="name">key</span> <span class="operator word">in</span> <span class="name">tags</span><span class="punctuation">:</span>
            <span class="comment">#tag may not be present in this feature</span>
            <span class="keyword">for</span> <span class="name">x</span> <span class="operator word">in</span> <span class="name">f</span><span class="operator">.</span><span class="name">qualifiers</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="name">key</span><span class="punctuation">,</span> <span class="punctuation">[]):</span>
                <span class="keyword">if</span> <span class="name">x</span> <span class="operator">==</span> <span class="name builtin">id</span><span class="punctuation">:</span>
                     <span class="keyword">return</span> <span class="name">f</span>
    <span class="keyword">raise</span> <span class="name exception">KeyError</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="punctuation">)</span>
</pre>
<p>We can now turn those list of identifier pairs into SeqFeature pairs,
and thus find their location co-ordinates. We can now add all that code
and the following snippet to the previous example (just before the
<tt class="docutils literal"><span class="pre">gd_diagram.draw(...)</span></tt> line – see the finished example script
<a class="reference external" href="http://biopython.org/SRC/biopython/Doc/examples/Proux_et_al_2002_Figure_6.py">Proux_et_al_2002_Figure_6.py</a>
included in the <tt class="docutils literal">Doc/examples</tt> folder of the Biopython source code) to
add cross links to the figure:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics.GenomeDiagram</span> <span class="keyword namespace">import</span> <span class="name">CrossLink</span>
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib</span> <span class="keyword namespace">import</span> <span class="name">colors</span>
<span class="comment">#Note it might have been clearer to assign the track numbers explicitly...</span>
<span class="keyword">for</span> <span class="name">rec_X</span><span class="punctuation">,</span> <span class="name">tn_X</span><span class="punctuation">,</span> <span class="name">rec_Y</span><span class="punctuation">,</span> <span class="name">tn_Y</span><span class="punctuation">,</span> <span class="name">X_vs_Y</span> <span class="operator word">in</span> <span class="punctuation">[(</span><span class="name">A_rec</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">,</span> <span class="name">B_rec</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name">A_vs_B</span><span class="punctuation">),</span>
                                         <span class="punctuation">(</span><span class="name">B_rec</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name">C_rec</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">B_vs_C</span><span class="punctuation">)]:</span>
    <span class="name">track_X</span> <span class="operator">=</span> <span class="name">gd_diagram</span><span class="operator">.</span><span class="name">tracks</span><span class="punctuation">[</span><span class="name">tn_X</span><span class="punctuation">]</span>
    <span class="name">track_Y</span> <span class="operator">=</span> <span class="name">gd_diagram</span><span class="operator">.</span><span class="name">tracks</span><span class="punctuation">[</span><span class="name">tn_Y</span><span class="punctuation">]</span>
    <span class="keyword">for</span> <span class="name">score</span><span class="punctuation">,</span> <span class="name">id_X</span><span class="punctuation">,</span> <span class="name">id_Y</span> <span class="operator word">in</span> <span class="name">X_vs_Y</span><span class="punctuation">:</span>
        <span class="name">feature_X</span> <span class="operator">=</span> <span class="name">get_feature</span><span class="punctuation">(</span><span class="name">rec_X</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">,</span> <span class="name">id_X</span><span class="punctuation">)</span>
        <span class="name">feature_Y</span> <span class="operator">=</span> <span class="name">get_feature</span><span class="punctuation">(</span><span class="name">rec_Y</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">,</span> <span class="name">id_Y</span><span class="punctuation">)</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">linearlyInterpolatedColor</span><span class="punctuation">(</span><span class="name">colors</span><span class="operator">.</span><span class="name">white</span><span class="punctuation">,</span> <span class="name">colors</span><span class="operator">.</span><span class="name">firebrick</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">100</span><span class="punctuation">,</span> <span class="name">score</span><span class="punctuation">)</span>
        <span class="name">link_xy</span> <span class="operator">=</span> <span class="name">CrossLink</span><span class="punctuation">((</span><span class="name">track_X</span><span class="punctuation">,</span> <span class="name">feature_X</span><span class="operator">.</span><span class="name">location</span><span class="operator">.</span><span class="name">start</span><span class="punctuation">,</span> <span class="name">feature_X</span><span class="operator">.</span><span class="name">location</span><span class="operator">.</span><span class="name">end</span><span class="punctuation">),</span>
                            <span class="punctuation">(</span><span class="name">track_Y</span><span class="punctuation">,</span> <span class="name">feature_Y</span><span class="operator">.</span><span class="name">location</span><span class="operator">.</span><span class="name">start</span><span class="punctuation">,</span> <span class="name">feature_Y</span><span class="operator">.</span><span class="name">location</span><span class="operator">.</span><span class="name">end</span><span class="punctuation">),</span>
                            <span class="name">color</span><span class="punctuation">,</span> <span class="name">colors</span><span class="operator">.</span><span class="name">lightgrey</span><span class="punctuation">)</span>
        <span class="name">gd_diagram</span><span class="operator">.</span><span class="name">cross_track_links</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">link_xy</span><span class="punctuation">)</span>
</pre>
<p>There are several important pieces to this code. First the
<tt class="docutils literal">GenomeDiagram</tt> object has a <tt class="docutils literal">cross_track_links</tt> attribute which is
just a list of <tt class="docutils literal">CrossLink</tt> objects. Each <tt class="docutils literal">CrossLink</tt> object takes
two sets of track-specific co-ordinates (here given as tuples, you can
alternatively use a <tt class="docutils literal">GenomeDiagram.Feature</tt> object instead). You can
optionally supply a colour, border color, and say if this link should be
drawn flipped (useful for showing inversions).</p>
<p>You can also see how we turn the BLAST percentage identity score into a
colour, interpolating between white (0%) and a dark red (100%). In this
example we don’t have any problems with overlapping cross-links. One way
to tackle that is to use transparency in ReportLab, by using colors with
their alpha channel set. However, this kind of shaded color scheme
combined with overlap transparency would be difficult to interpret. The
result:</p>
<p><img alt="image22" src="../images/three_track_cl.png" /></p>
<p>There is still a lot more that can be done within Biopython to help
improve this figure. First of all, the cross links in this case are
between proteins which are drawn in a strand specific manor. It can help
to add a background region (a feature using the ‘BOX’ sigil) on the
feature track to extend the cross link. Also, we could reduce the
vertical height of the feature tracks to allocate more to the links
instead – one way to do that is to allocate space for empty tracks.
Furthermore, in cases like this where there are no large gene overlaps,
we can use the axis-straddling <tt class="docutils literal">BIGARROW</tt> sigil, which allows us to
further reduce the vertical space needed for the track. These
improvements are demonstrated in the example script
<a class="reference external" href="http://biopython.org/SRC/biopython/Doc/examples/Proux_et_al_2002_Figure_6.py">Proux_et_al_2002_Figure_6.py</a>
included in the <tt class="docutils literal">Doc/examples</tt> folder of the Biopython source code.
The result:</p>
<p><img alt="image23" src="../images/three_track_cl2a.png" /></p>
<p>Beyond that, finishing touches you might want to do manually in a vector
image editor include fine tuning the placement of gene labels, and
adding other custom annotation such as highlighting particular regions.</p>
<p>Although not really necessary in this example since none of the
cross-links overlap, using a transparent color in ReportLab is a very
useful technique for superimposing multiple links. However, in this case
a shaded color scheme should be avoided.</p>
</div>
<div class="section" id="further-options">
<h3>17.1.12  Further options</h3>
<p>You can control the tick marks to show the scale – after all every graph
should show its units, and the number of the grey-track labels.</p>
<p>Also, we have only used the <tt class="docutils literal">FeatureSet</tt> so far. GenomeDiagram also
has a <tt class="docutils literal">GraphSet</tt> which can be used for show line graphs, bar charts
and heat plots (e.g. to show plots of GC% on a track parallel to the
features).</p>
<p>These options are not covered here yet, so for now we refer you to the
<a class="reference external" href="http://biopython.org/DIST/docs/GenomeDiagram/userguide.pdf">User Guide
(PDF)</a>
included with the standalone version of GenomeDiagram (but please read
the next section first), and the docstrings.</p>
</div>
<div class="section" id="converting-old-code">
<h3>17.1.13  Converting old code</h3>
<p>If you have old code written using the standalone version of
GenomeDiagram, and you want to switch it over to using the new version
included with Biopython then you will have to make a few changes - most
importantly to your import statements.</p>
<p>Also, the older version of GenomeDiagram used only the UK spellings of
color and center (colour and centre). As part of the integration into
Biopython, both forms can now be used for argument names. However, at
some point in the future the UK spellings may be deprecated.</p>
<p>For example, if you used to have:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">GenomeDiagram</span> <span class="keyword namespace">import</span> <span class="name">GDFeatureSet</span><span class="punctuation">,</span> <span class="name">GDDiagram</span>
<span class="name">gdd</span> <span class="operator">=</span> <span class="name">GDDiagram</span><span class="punctuation">(</span><span class="literal string">&quot;An example&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>you could just switch the import statements like this:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics.GenomeDiagram</span> <span class="keyword namespace">import</span> <span class="name">FeatureSet</span> <span class="keyword">as</span> <span class="name">GDFeatureSet</span><span class="punctuation">,</span> <span class="name">Diagram</span> <span class="keyword">as</span> <span class="name">GDDiagram</span>
<span class="name">gdd</span> <span class="operator">=</span> <span class="name">GDDiagram</span><span class="punctuation">(</span><span class="literal string">&quot;An example&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>and hopefully that should be enough. In the long term you might want to
switch to the new names, but you would have to change more of your code:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics.GenomeDiagram</span> <span class="keyword namespace">import</span> <span class="name">FeatureSet</span><span class="punctuation">,</span> <span class="name">Diagram</span>
<span class="name">gdd</span> <span class="operator">=</span> <span class="name">Diagram</span><span class="punctuation">(</span><span class="literal string">&quot;An example&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>or:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">GenomeDiagram</span>
<span class="name">gdd</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Diagram</span><span class="punctuation">(</span><span class="literal string">&quot;An example&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>If you run into difficulties, please ask on the Biopython mailing list
for advice. One catch is that we have not included the old module
<tt class="docutils literal">GenomeDiagram.GDUtilities</tt> yet. This included a number of GC% related
functions, which will probably be merged under <tt class="docutils literal">Bio.SeqUtils</tt> later
on.</p>
</div>
</div>
<div class="section" id="chromosomes">
<h2>17.2  Chromosomes</h2>
<p>The <tt class="docutils literal">Bio.Graphics.BasicChromosome</tt> module allows drawing of
chromosomes. There is an example in Jupe <em>et al.</em> (2012)
[<a class="reference external" href="#jupe2012">6</a>] (open access) using colors to highlight different
gene families.</p>
<div class="section" id="simple-chromosomes">
<h3>17.2.1  Simple Chromosomes</h3>
<p>Here is a very simple example - for which we’ll use <em>Arabidopsis
thaliana</em>.</p>
<p>You can skip this bit, but first I downloaded the five sequenced
chromosomes from the NCBI’s FTP site
<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/Arabidopsis_thaliana">ftp://ftp.ncbi.nlm.nih.gov/genomes/Arabidopsis_thaliana</a>
and then parsed them with <tt class="docutils literal">Bio.SeqIO</tt> to find out their lengths. You
could use the GenBank files for this, but it is faster to use the FASTA
files for the whole chromosomes:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">entries</span> <span class="operator">=</span> <span class="punctuation">[(</span><span class="literal string">&quot;Chr I&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_I/NC_003070.fna&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr II&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_II/NC_003071.fna&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr III&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_III/NC_003074.fna&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr IV&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_IV/NC_003075.fna&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr V&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_V/NC_003076.fna&quot;</span><span class="punctuation">)]</span>
<span class="keyword">for</span> <span class="punctuation">(</span><span class="name">name</span><span class="punctuation">,</span> <span class="name">filename</span><span class="punctuation">)</span> <span class="operator word">in</span> <span class="name">entries</span><span class="punctuation">:</span>
   <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
   <span class="keyword">print</span> <span class="name">name</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
</pre>
<p>This gave the lengths of the five chromosomes, which we’ll now use in
the following short demonstration of the <tt class="docutils literal">BasicChromosome</tt> module:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.units</span> <span class="keyword namespace">import</span> <span class="name">cm</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">BasicChromosome</span>

<span class="name">entries</span> <span class="operator">=</span> <span class="punctuation">[(</span><span class="literal string">&quot;Chr I&quot;</span><span class="punctuation">,</span> <span class="literal number integer">30432563</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr II&quot;</span><span class="punctuation">,</span> <span class="literal number integer">19705359</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr III&quot;</span><span class="punctuation">,</span> <span class="literal number integer">23470805</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr IV&quot;</span><span class="punctuation">,</span> <span class="literal number integer">18585042</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr V&quot;</span><span class="punctuation">,</span> <span class="literal number integer">26992728</span><span class="punctuation">)]</span>

<span class="name">max_len</span> <span class="operator">=</span> <span class="literal number integer">30432563</span> <span class="comment">#Could compute this</span>
<span class="name">telomere_length</span> <span class="operator">=</span> <span class="literal number integer">1000000</span> <span class="comment">#For illustration</span>

<span class="name">chr_diagram</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">Organism</span><span class="punctuation">()</span>
<span class="name">chr_diagram</span><span class="operator">.</span><span class="name">page_size</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="literal number float">29.7</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">,</span> <span class="literal number integer">21</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">)</span> <span class="comment">#A4 landscape</span>

<span class="keyword">for</span> <span class="name">name</span><span class="punctuation">,</span> <span class="name">length</span> <span class="operator word">in</span> <span class="name">entries</span><span class="punctuation">:</span>
    <span class="name">cur_chromosome</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">Chromosome</span><span class="punctuation">(</span><span class="name">name</span><span class="punctuation">)</span>
    <span class="comment">#Set the scale to the MAXIMUM length plus the two telomeres in bp,</span>
    <span class="comment">#want the same scale used on all five chromosomes so they can be</span>
    <span class="comment">#compared to each other</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">scale_num</span> <span class="operator">=</span> <span class="name">max_len</span> <span class="operator">+</span> <span class="literal number integer">2</span> <span class="operator">*</span> <span class="name">telomere_length</span>

    <span class="comment">#Add an opening telomere</span>
    <span class="name">start</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">TelomereSegment</span><span class="punctuation">()</span>
    <span class="name">start</span><span class="operator">.</span><span class="name">scale</span> <span class="operator">=</span> <span class="name">telomere_length</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">start</span><span class="punctuation">)</span>

    <span class="comment">#Add a body - using bp as the scale length here.</span>
    <span class="name">body</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">ChromosomeSegment</span><span class="punctuation">()</span>
    <span class="name">body</span><span class="operator">.</span><span class="name">scale</span> <span class="operator">=</span> <span class="name">length</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">body</span><span class="punctuation">)</span>

    <span class="comment">#Add a closing telomere</span>
    <span class="name">end</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">TelomereSegment</span><span class="punctuation">(</span><span class="name">inverted</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
    <span class="name">end</span><span class="operator">.</span><span class="name">scale</span> <span class="operator">=</span> <span class="name">telomere_length</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">end</span><span class="punctuation">)</span>

    <span class="comment">#This chromosome is done</span>
    <span class="name">chr_diagram</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">cur_chromosome</span><span class="punctuation">)</span>

<span class="name">chr_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="literal string">&quot;simple_chrom.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;Arabidopsis thaliana&quot;</span><span class="punctuation">)</span>
</pre>
<p>This should create a very simple PDF file, shown here:</p>
<p><img alt="image24" src="../images/simple_chrom.png" /></p>
<p>This example is deliberately short and sweet. The next example shows the
location of features of interest.</p>
</div>
<div class="section" id="annotated-chromosomes">
<h3>17.2.2  Annotated Chromosomes</h3>
<p>Continuing from the previous example, let’s also show the tRNA genes.
We’ll get their locations by parsing the GenBank files for the five
<em>Arabidopsis thaliana</em> chromosomes. You’ll need to download these files
from the NCBI FTP site
<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/Arabidopsis_thaliana">ftp://ftp.ncbi.nlm.nih.gov/genomes/Arabidopsis_thaliana</a>,
and preserve the subdirectory names or edit the paths below:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.units</span> <span class="keyword namespace">import</span> <span class="name">cm</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">BasicChromosome</span>

<span class="name">entries</span> <span class="operator">=</span> <span class="punctuation">[(</span><span class="literal string">&quot;Chr I&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_I/NC_003070.gbk&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr II&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_II/NC_003071.gbk&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr III&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_III/NC_003074.gbk&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr IV&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_IV/NC_003075.gbk&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr V&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_V/NC_003076.gbk&quot;</span><span class="punctuation">)]</span>

<span class="name">max_len</span> <span class="operator">=</span> <span class="literal number integer">30432563</span> <span class="comment">#Could compute this</span>
<span class="name">telomere_length</span> <span class="operator">=</span> <span class="literal number integer">1000000</span> <span class="comment">#For illustration</span>

<span class="name">chr_diagram</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">Organism</span><span class="punctuation">()</span>
<span class="name">chr_diagram</span><span class="operator">.</span><span class="name">page_size</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="literal number float">29.7</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">,</span> <span class="literal number integer">21</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">)</span> <span class="comment">#A4 landscape</span>

<span class="keyword">for</span> <span class="name">index</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="name">name</span><span class="punctuation">,</span> <span class="name">filename</span><span class="punctuation">)</span> <span class="operator word">in</span> <span class="name builtin">enumerate</span><span class="punctuation">(</span><span class="name">entries</span><span class="punctuation">):</span>
    <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span><span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
    <span class="name">length</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
    <span class="name">features</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">f</span> <span class="keyword">for</span> <span class="name">f</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span> <span class="keyword">if</span> <span class="name">f</span><span class="operator">.</span><span class="name">type</span><span class="operator">==</span><span class="literal string">&quot;tRNA&quot;</span><span class="punctuation">]</span>
    <span class="comment">#Record an Artemis style integer color in the feature's qualifiers,</span>
    <span class="comment">#1 = Black, 2 = Red, 3 = Green, 4 = blue, 5 =cyan, 6 = purple</span>
    <span class="keyword">for</span> <span class="name">f</span> <span class="operator word">in</span> <span class="name">features</span><span class="punctuation">:</span> <span class="name">f</span><span class="operator">.</span><span class="name">qualifiers</span><span class="punctuation">[</span><span class="literal string">&quot;color&quot;</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">index</span><span class="operator">+</span><span class="literal number integer">2</span><span class="punctuation">]</span>

    <span class="name">cur_chromosome</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">Chromosome</span><span class="punctuation">(</span><span class="name">name</span><span class="punctuation">)</span>
    <span class="comment">#Set the scale to the MAXIMUM length plus the two telomeres in bp,</span>
    <span class="comment">#want the same scale used on all five chromosomes so they can be</span>
    <span class="comment">#compared to each other</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">scale_num</span> <span class="operator">=</span> <span class="name">max_len</span> <span class="operator">+</span> <span class="literal number integer">2</span> <span class="operator">*</span> <span class="name">telomere_length</span>

    <span class="comment">#Add an opening telomere</span>
    <span class="name">start</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">TelomereSegment</span><span class="punctuation">()</span>
    <span class="name">start</span><span class="operator">.</span><span class="name">scale</span> <span class="operator">=</span> <span class="name">telomere_length</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">start</span><span class="punctuation">)</span>

    <span class="comment">#Add a body - again using bp as the scale length here.</span>
    <span class="name">body</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">AnnotatedChromosomeSegment</span><span class="punctuation">(</span><span class="name">length</span><span class="punctuation">,</span> <span class="name">features</span><span class="punctuation">)</span>
    <span class="name">body</span><span class="operator">.</span><span class="name">scale</span> <span class="operator">=</span> <span class="name">length</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">body</span><span class="punctuation">)</span>

    <span class="comment">#Add a closing telomere</span>
    <span class="name">end</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">TelomereSegment</span><span class="punctuation">(</span><span class="name">inverted</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
    <span class="name">end</span><span class="operator">.</span><span class="name">scale</span> <span class="operator">=</span> <span class="name">telomere_length</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">end</span><span class="punctuation">)</span>

    <span class="comment">#This chromosome is done</span>
    <span class="name">chr_diagram</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">cur_chromosome</span><span class="punctuation">)</span>

<span class="name">chr_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="literal string">&quot;tRNA_chrom.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;Arabidopsis thaliana&quot;</span><span class="punctuation">)</span>
</pre>
<p>It might warn you about the labels being too close together - have a
look at the forward strand (right hand side) of Chr I, but it should
create a colorful PDF file, shown here:</p>
<p><img alt="image25" src="../images/tRNA_chrom.png" /></p>
</div>
</div>
</div>
<div class="section" id="chapter-18-cookbook-cool-things-to-do-with-it">
<h1>Chapter 18  Cookbook – Cool things to do with it</h1>
<p>Biopython now has two collections of “cookbook” examples – this chapter
(which has been included in this tutorial for many years and has
gradually grown), and
<a class="reference external" href="http://biopython.org/wiki/Category:Cookbook">http://biopython.org/wiki/Category:Cookbook</a>
which is a user contributed collection on our wiki.</p>
<p>We’re trying to encourage Biopython users to contribute their own
examples to the wiki. In addition to helping the community, one direct
benefit of sharing an example like this is that you could also get some
feedback on the code from other Biopython users and developers - which
could help you improve all your Python code.</p>
<p>In the long term, we may end up moving all of the examples in this
chapter to the wiki, or elsewhere within the tutorial.</p>
<div class="section" id="working-with-sequence-files">
<h2>18.1  Working with sequence files</h2>
<p>This section shows some more examples of sequence input/output, using
the <tt class="docutils literal">Bio.SeqIO</tt> module described in
Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>.</p>
<div class="section" id="filtering-a-sequence-file">
<h3>18.1.1  Filtering a sequence file</h3>
<p>Often you’ll have a large file with many sequences in it (e.g. FASTA
file or genes, or a FASTQ or SFF file of reads), a separate shorter list
of the IDs for a subset of sequences of interest, and want to make a new
sequence file for this subset.</p>
<p>Let’s say the list of IDs is in a simple text file, as the first word on
each line. This could be a tabular file where the first column is the
ID. Try something like this:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">input_file</span> <span class="operator">=</span> <span class="literal string">&quot;big_file.sff&quot;</span>
<span class="name">id_file</span> <span class="operator">=</span> <span class="literal string">&quot;short_list.txt&quot;</span>
<span class="name">output_file</span> <span class="operator">=</span> <span class="literal string">&quot;short_list.sff&quot;</span>
<span class="name">wanted</span> <span class="operator">=</span> <span class="name builtin">set</span><span class="punctuation">(</span><span class="name">line</span><span class="operator">.</span><span class="name">rstrip</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string escape">\n</span><span class="literal string">&quot;</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">(</span><span class="name builtin pseudo">None</span><span class="punctuation">,</span><span class="literal number integer">1</span><span class="punctuation">)[</span><span class="literal number integer">0</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name">line</span> <span class="operator word">in</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="name">id_file</span><span class="punctuation">))</span>
<span class="keyword">print</span> <span class="literal string">&quot;Found </span><span class="literal string interpol">%i</span><span class="literal string"> unique identifiers in </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">wanted</span><span class="punctuation">),</span> <span class="name">id_file</span><span class="punctuation">)</span>
<span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">r</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">input_file</span><span class="punctuation">,</span> <span class="literal string">&quot;sff&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name">r</span><span class="operator">.</span><span class="name">id</span> <span class="operator word">in</span> <span class="name">wanted</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">output_file</span><span class="punctuation">,</span> <span class="literal string">&quot;sff&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> records from </span><span class="literal string interpol">%s</span><span class="literal string"> to </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">count</span><span class="punctuation">,</span> <span class="name">input_file</span><span class="punctuation">,</span> <span class="name">output_file</span><span class="punctuation">)</span>
<span class="keyword">if</span> <span class="name">count</span> <span class="operator">&lt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">wanted</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="literal string">&quot;Warning </span><span class="literal string interpol">%i</span><span class="literal string"> IDs not found in </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">wanted</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">count</span><span class="punctuation">,</span> <span class="name">input_file</span><span class="punctuation">)</span>
</pre>
<p>Note that we use a Python <tt class="docutils literal">set</tt> rather than a <tt class="docutils literal">list</tt>, this makes
testing membership faster.</p>
</div>
<div class="section" id="producing-randomised-genomes">
<h3>18.1.2  Producing randomised genomes</h3>
<p>Let’s suppose you are looking at genome sequence, hunting for some
sequence feature – maybe extreme local GC% bias, or possible restriction
digest sites. Once you’ve got your Python code working on the real
genome it may be sensible to try running the same search on randomised
versions of the same genome for statistical analysis (after all, any
“features” you’ve found could just be there just by chance).</p>
<p>For this discussion, we’ll use the GenBank file for the pPCP1 plasmid
from <em>Yersinia pestis biovar Microtus</em>. The file is included with the
Biopython unit tests under the GenBank folder, or you can get it from
our website,
<a class="reference external" href="http://biopython.org/SRC/biopython/Tests/GenBank/NC_005816.gb">NC_005816.gb</a>.
This file contains one and only one record, so we can read it in as a
<tt class="docutils literal">SeqRecord</tt> using the <tt class="docutils literal">Bio.SeqIO.read()</tt> function:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">original_rec</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
</pre>
<p>So, how can we generate a shuffled versions of the original sequence? I
would use the built in Python <tt class="docutils literal">random</tt> module for this, in particular
the function <tt class="docutils literal">random.shuffle</tt> – but this works on a Python list. Our
sequence is a <tt class="docutils literal">Seq</tt> object, so in order to shuffle it we need to turn
it into a list:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">random</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nuc_list</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">original_rec</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">random</span><span class="operator">.</span><span class="name">shuffle</span><span class="punctuation">(</span><span class="name">nuc_list</span><span class="punctuation">)</span> <span class="comment">#acts in situ!</span>
</pre>
<p>Now, in order to use <tt class="docutils literal">Bio.SeqIO</tt> to output the shuffled sequence, we
need to construct a new <tt class="docutils literal">SeqRecord</tt> with a new <tt class="docutils literal">Seq</tt> object using
this shuffled list. In order to do this, we need to turn the list of
nucleotides (single letter strings) into a long string – the standard
Python way to do this is with the string object’s join method.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shuffled_rec</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">nuc_list</span><span class="punctuation">),</span> <span class="name">original_rec</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">),</span>
<span class="operator">...</span>                          <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Shuffled&quot;</span><span class="punctuation">,</span> <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;Based on </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">original_rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>
</pre>
<p>Let’s put all these pieces together to make a complete Python script
which generates a single FASTA file containing 30 randomly shuffled
versions of the original sequence.</p>
<p>This first version just uses a big for loop and writes out the records
one by one (using the <tt class="docutils literal">SeqRecord</tt>’s format method described in
Section <a class="reference external" href="#sec:Bio.SeqIO-and-StringIO">5.5.4</a>):</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">random</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>

<span class="name">original_rec</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>

<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;shuffled.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">30</span><span class="punctuation">):</span>
    <span class="name">nuc_list</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">original_rec</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
    <span class="name">random</span><span class="operator">.</span><span class="name">shuffle</span><span class="punctuation">(</span><span class="name">nuc_list</span><span class="punctuation">)</span>
    <span class="name">shuffled_rec</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">nuc_list</span><span class="punctuation">),</span> <span class="name">original_rec</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">),</span> \
                             <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Shuffled</span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">),</span> \
                             <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;Based on </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">original_rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>
    <span class="name">handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">shuffled_rec</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>Personally I prefer the following version using a function to shuffle
the record and a generator expression instead of the for loop:</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">random</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>

<span class="keyword">def</span> <span class="name function">make_shuffle_record</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">,</span> <span class="name">new_id</span><span class="punctuation">):</span>
    <span class="name">nuc_list</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
    <span class="name">random</span><span class="operator">.</span><span class="name">shuffle</span><span class="punctuation">(</span><span class="name">nuc_list</span><span class="punctuation">)</span>
    <span class="keyword">return</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">nuc_list</span><span class="punctuation">),</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">),</span> \
           <span class="name builtin">id</span><span class="operator">=</span><span class="name">new_id</span><span class="punctuation">,</span> <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;Based on </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">original_rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>

<span class="name">original_rec</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="name">shuffled_recs</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">make_shuffle_record</span><span class="punctuation">(</span><span class="name">original_rec</span><span class="punctuation">,</span> <span class="literal string">&quot;Shuffled</span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">))</span> \
                 <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">30</span><span class="punctuation">))</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;shuffled.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">shuffled_recs</span><span class="punctuation">,</span> <span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
</div>
<div class="section" id="translating-a-fasta-file-of-cds-entries">
<h3>18.1.3  Translating a FASTA file of CDS entries</h3>
<p>Suppose you’ve got an input file of CDS entries for some organism, and
you want to generate a new FASTA file containing their protein
sequences. i.e. Take each nucleotide sequence from the original file,
and translate it. Back in Section <a class="reference external" href="#sec:translation">3.9</a> we saw
how to use the <tt class="docutils literal">Seq</tt> object’s <tt class="docutils literal">translate method</tt>, and the optional
<tt class="docutils literal">cds</tt> argument which enables correct translation of alternative start
codons.</p>
<p>We can combine this with <tt class="docutils literal">Bio.SeqIO</tt> as shown in the reverse
complement example in
Section <a class="reference external" href="#sec:SeqIO-reverse-complement">5.5.3</a>. The key point is
that for each nucleotide <tt class="docutils literal">SeqRecord</tt>, we need to create a protein
<tt class="docutils literal">SeqRecord</tt> - and take care of naming it.</p>
<p>You can write you own function to do this, choosing suitable protein
identifiers for your sequences, and the appropriate genetic code. In
this example we just use the default table and add a prefix to the
identifier:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="keyword">def</span> <span class="name function">make_protein_record</span><span class="punctuation">(</span><span class="name">nuc_record</span><span class="punctuation">):</span>
    <span class="literal string doc">&quot;&quot;&quot;Returns a new SeqRecord with the translated sequence (default table).&quot;&quot;&quot;</span>
    <span class="keyword">return</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span> <span class="operator">=</span> <span class="name">nuc_record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">cds</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">),</span> \
                     <span class="name builtin">id</span> <span class="operator">=</span> <span class="literal string">&quot;trans_&quot;</span> <span class="operator">+</span> <span class="name">nuc_record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> \
                     <span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;translation of CDS, using default table&quot;</span><span class="punctuation">)</span>
</pre>
<p>We can then use this function to turn the input nucleotide records into
protein records ready for output. An elegant way and memory efficient
way to do this is with a generator expression:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">proteins</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">make_protein_record</span><span class="punctuation">(</span><span class="name">nuc_rec</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">nuc_rec</span> <span class="operator word">in</span> \
            <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;coding_sequences.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">proteins</span><span class="punctuation">,</span> <span class="literal string">&quot;translations.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>This should work on any FASTA file of complete coding sequences. If you
are working on partial coding sequences, you may prefer to use
<tt class="docutils literal">nuc_record.seq.translate(to_stop=True)</tt> in the example above, as this
wouldn’t check for a valid start codon etc.</p>
</div>
<div class="section" id="making-the-sequences-in-a-fasta-file-upper-case">
<h3>18.1.4  Making the sequences in a FASTA file upper case</h3>
<p>Often you’ll get data from collaborators as FASTA files, and sometimes
the sequences can be in a mixture of upper and lower case. In some cases
this is deliberate (e.g. lower case for poor quality regions), but
usually it is not important. You may want to edit the file to make
everything consistent (e.g. all upper case), and you can do this easily
using the <tt class="docutils literal">upper()</tt> method of the <tt class="docutils literal">SeqRecord</tt> object (added in
Biopython 1.55):</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">()</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;mixed.fas&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;upper.fas&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Converted </span><span class="literal string interpol">%i</span><span class="literal string"> records to upper case&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>How does this work? The first line is just importing the <tt class="docutils literal">Bio.SeqIO</tt>
module. The second line is the interesting bit – this is a Python
generator expression which gives an upper case version of each record
parsed from the input file (<tt class="docutils literal">mixed.fas</tt>). In the third line we give
this generator expression to the <tt class="docutils literal">Bio.SeqIO.write()</tt> function and it
saves the new upper cases records to our output file (<tt class="docutils literal">upper.fas</tt>).</p>
<p>The reason we use a generator expression (rather than a list or list
comprehension) is this means only one record is kept in memory at a
time. This can be really important if you are dealing with large files
with millions of entries.</p>
</div>
<div class="section" id="sorting-a-sequence-file">
<h3>18.1.5  Sorting a sequence file</h3>
<p>Suppose you wanted to sort a sequence file by length (e.g. a set of
contigs from an assembly), and you are working with a file format like
FASTA or FASTQ which <tt class="docutils literal">Bio.SeqIO</tt> can read, write (and index).</p>
<p>If the file is small enough, you can load it all into memory at once as
a list of <tt class="docutils literal">SeqRecord</tt> objects, sort the list, and save it:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">records</span><span class="operator">.</span><span class="name">sort</span><span class="punctuation">(</span><span class="name builtin">cmp</span><span class="operator">=</span><span class="keyword">lambda</span> <span class="name">x</span><span class="punctuation">,</span><span class="name">y</span><span class="punctuation">:</span> <span class="name builtin">cmp</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">x</span><span class="punctuation">),</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">y</span><span class="punctuation">)))</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;sorted_orchids.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>The only clever bit is specifying a comparison function for how to sort
the records (here we sort them by length). If you wanted the longest
records first, you could flip the comparison or use the reverse
argument:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">records</span><span class="operator">.</span><span class="name">sort</span><span class="punctuation">(</span><span class="name builtin">cmp</span><span class="operator">=</span><span class="keyword">lambda</span> <span class="name">x</span><span class="punctuation">,</span><span class="name">y</span><span class="punctuation">:</span> <span class="name builtin">cmp</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">y</span><span class="punctuation">),</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">x</span><span class="punctuation">)))</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;sorted_orchids.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>Now that’s pretty straight forward - but what happens if you have a very
large file and you can’t load it all into memory like this? For example,
you might have some next-generation sequencing reads to sort by length.
This can be solved using the <tt class="docutils literal">Bio.SeqIO.index()</tt> function.</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="comment">#Get the lengths and ids, and sort on length</span>
<span class="name">len_and_ids</span> <span class="operator">=</span> <span class="name builtin">sorted</span><span class="punctuation">((</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">),</span> <span class="name">rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> \
                     <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">ids</span> <span class="operator">=</span> <span class="name builtin">reversed</span><span class="punctuation">([</span><span class="name builtin">id</span> <span class="keyword">for</span> <span class="punctuation">(</span><span class="name">length</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="punctuation">)</span> <span class="operator word">in</span> <span class="name">len_and_ids</span><span class="punctuation">])</span>
<span class="keyword">del</span> <span class="name">len_and_ids</span> <span class="comment">#free this memory</span>
<span class="name">record_index</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">record_index</span><span class="punctuation">[</span><span class="name builtin">id</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name builtin">id</span> <span class="operator word">in</span> <span class="name">ids</span><span class="punctuation">)</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;sorted.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>First we scan through the file once using <tt class="docutils literal">Bio.SeqIO.parse()</tt>,
recording the record identifiers and their lengths in a list of tuples.
We then sort this list to get them in length order, and discard the
lengths. Using this sorted list of identifiers <tt class="docutils literal">Bio.SeqIO.index()</tt>
allows us to retrieve the records one by one, and we pass them to
<tt class="docutils literal">Bio.SeqIO.write()</tt> for output.</p>
<p>These examples all use <tt class="docutils literal">Bio.SeqIO</tt> to parse the records into
<tt class="docutils literal">SeqRecord</tt> objects which are output using <tt class="docutils literal">Bio.SeqIO.write()</tt>. What
if you want to sort a file format which <tt class="docutils literal">Bio.SeqIO.write()</tt> doesn’t
support, like the plain text SwissProt format? Here is an alternative
solution using the <tt class="docutils literal">get_raw()</tt> method added to <tt class="docutils literal">Bio.SeqIO.index()</tt>
in Biopython 1.54 (see Section <a class="reference external" href="#sec:seqio-index-getraw">5.4.2.2</a>).</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="comment">#Get the lengths and ids, and sort on length</span>
<span class="name">len_and_ids</span> <span class="operator">=</span> <span class="name builtin">sorted</span><span class="punctuation">((</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">),</span> <span class="name">rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> \
                     <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">ids</span> <span class="operator">=</span> <span class="name builtin">reversed</span><span class="punctuation">([</span><span class="name builtin">id</span> <span class="keyword">for</span> <span class="punctuation">(</span><span class="name">length</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="punctuation">)</span> <span class="operator word">in</span> <span class="name">len_and_ids</span><span class="punctuation">])</span>
<span class="keyword">del</span> <span class="name">len_and_ids</span> <span class="comment">#free this memory</span>
<span class="name">record_index</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;sorted.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name builtin">id</span> <span class="operator word">in</span> <span class="name">ids</span><span class="punctuation">:</span>
    <span class="name">handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">record_index</span><span class="operator">.</span><span class="name">get_raw</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="punctuation">))</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>As a bonus, because it doesn’t parse the data into <tt class="docutils literal">SeqRecord</tt> objects
a second time it should be faster.</p>
</div>
<div class="section" id="simple-quality-filtering-for-fastq-files">
<h3>18.1.6  Simple quality filtering for FASTQ files</h3>
<p>The FASTQ file format was introduced at Sanger and is now widely used
for holding nucleotide sequencing reads together with their quality
scores. FASTQ files (and the related QUAL files) are an excellent
example of per-letter-annotation, because for each nucleotide in the
sequence there is an associated quality score. Any per-letter-annotation
is held in a <tt class="docutils literal">SeqRecord</tt> in the <tt class="docutils literal">letter_annotations</tt> dictionary as a
list, tuple or string (with the same number of elements as the sequence
length).</p>
<p>One common task is taking a large set of sequencing reads and filtering
them (or cropping them) based on their quality scores. The following
example is very simplistic, but should illustrate the basics of working
with quality data in a <tt class="docutils literal">SeqRecord</tt> object. All we are going to do here
is read in a file of FASTQ data, and filter it to pick out only those
records whose PHRED quality scores are all above some threshold (here
20).</p>
<p>For this example we’ll use some real data downloaded from the ENA
sequence read archive,
<a class="reference external" href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz">ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz</a>
(2MB) which unzips to a 19MB file <tt class="docutils literal">SRR020192.fastq</tt>. This is some
Roche 454 GS FLX single end data from virus infected California sea
lions (see
<a class="reference external" href="http://www.ebi.ac.uk/ena/data/view/SRS004476">http://www.ebi.ac.uk/ena/data/view/SRS004476</a>
for details).</p>
<p>First, let’s count the reads:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">count</span> <span class="operator">=</span> <span class="literal number integer">0</span>
<span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">):</span>
    <span class="name">count</span> <span class="operator">+=</span> <span class="literal number integer">1</span>
<span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>Now let’s do a simple filtering for a minimum PHRED quality of 20:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">good_reads</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> \
              <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span> \
              <span class="keyword">if</span> <span class="name builtin">min</span><span class="punctuation">(</span><span class="name">rec</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">])</span> <span class="operator">&gt;=</span> <span class="literal number integer">20</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">good_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;good_quality.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>This pulled out only 14580 reads out of the 41892 present. A more
sensible thing to do would be to quality trim the reads, but this is
intended as an example only.</p>
<p>FASTQ files can contain millions of entries, so it is best to avoid
loading them all into memory at once. This example uses a generator
expression, which means only one <tt class="docutils literal">SeqRecord</tt> is created at a time -
avoiding any memory limitations.</p>
</div>
<div class="section" id="trimming-off-primer-sequences">
<h3>18.1.7  Trimming off primer sequences</h3>
<p>For this example we’re going to pretend that <tt class="docutils literal">GATGACGGTGT</tt> is a 5’
primer sequence we want to look for in some FASTQ formatted read data.
As in the example above, we’ll use the <tt class="docutils literal">SRR020192.fastq</tt> file
downloaded from the ENA
(<a class="reference external" href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz">ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz</a>).
The same approach would work with any other supported file format (e.g.
FASTA files).</p>
<p>This code uses <tt class="docutils literal">Bio.SeqIO</tt> with a generator expression (to avoid
loading all the sequences into memory at once), and the <tt class="docutils literal">Seq</tt> object’s
<tt class="docutils literal">startswith</tt> method to see if the read starts with the primer
sequence:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">primer_reads</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> \
                <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span> \
                <span class="keyword">if</span> <span class="name">rec</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">startswith</span><span class="punctuation">(</span><span class="literal string">&quot;GATGACGGTGT&quot;</span><span class="punctuation">))</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">primer_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;with_primer.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>That should find 13819 reads from <tt class="docutils literal">SRR014849.fastq</tt> and save them to a
new FASTQ file, <tt class="docutils literal">with_primer.fastq</tt>.</p>
<p>Now suppose that instead you wanted to make a FASTQ file containing
these reads but with the primer sequence removed? That’s just a small
change as we can slice the <tt class="docutils literal">SeqRecord</tt> (see
Section <a class="reference external" href="#sec:SeqRecord-slicing">4.6</a>) to remove the first eleven
letters (the length of our primer):</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">trimmed_primer_reads</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">[</span><span class="literal number integer">11</span><span class="punctuation">:]</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> \
                        <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span> \
                        <span class="keyword">if</span> <span class="name">rec</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">startswith</span><span class="punctuation">(</span><span class="literal string">&quot;GATGACGGTGT&quot;</span><span class="punctuation">))</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">trimmed_primer_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;with_primer_trimmed.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>Again, that should pull out the 13819 reads from <tt class="docutils literal">SRR020192.fastq</tt>,
but this time strip off the first ten characters, and save them to
another new FASTQ file, <tt class="docutils literal">with_primer_trimmed.fastq</tt>.</p>
<p>Finally, suppose you want to create a new FASTQ file where these reads
have their primer removed, but all the other reads are kept as they
were? If we want to still use a generator expression, it is probably
clearest to define our own trim function:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">def</span> <span class="name function">trim_primer</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">,</span> <span class="name">primer</span><span class="punctuation">):</span>
    <span class="keyword">if</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">startswith</span><span class="punctuation">(</span><span class="name">primer</span><span class="punctuation">):</span>
        <span class="keyword">return</span> <span class="name">record</span><span class="punctuation">[</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">primer</span><span class="punctuation">):]</span>
    <span class="keyword">else</span><span class="punctuation">:</span>
        <span class="keyword">return</span> <span class="name">record</span>

<span class="name">trimmed_reads</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">trim_primer</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">,</span> <span class="literal string">&quot;GATGACGGTGT&quot;</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> \
                 <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">))</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">trimmed_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>This takes longer, as this time the output file contains all 41892
reads. Again, we’re used a generator expression to avoid any memory
problems. You could alternatively use a generator function rather than a
generator expression.</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">def</span> <span class="name function">trim_primers</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">primer</span><span class="punctuation">):</span>
    <span class="literal string doc">&quot;&quot;&quot;Removes perfect primer sequences at start of reads.

    This is a generator function, the records argument should
    be a list or iterator returning SeqRecord objects.
    &quot;&quot;&quot;</span>
    <span class="name">len_primer</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">primer</span><span class="punctuation">)</span> <span class="comment">#cache this for later</span>
    <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
        <span class="keyword">if</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">startswith</span><span class="punctuation">(</span><span class="name">primer</span><span class="punctuation">):</span>
            <span class="keyword">yield</span> <span class="name">record</span><span class="punctuation">[</span><span class="name">len_primer</span><span class="punctuation">:]</span>
        <span class="keyword">else</span><span class="punctuation">:</span>
            <span class="keyword">yield</span> <span class="name">record</span>

<span class="name">original_reads</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="name">trimmed_reads</span> <span class="operator">=</span> <span class="name">trim_primers</span><span class="punctuation">(</span><span class="name">original_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;GATGACGGTGT&quot;</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">trimmed_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>This form is more flexible if you want to do something more complicated
where only some of the records are retained – as shown in the next
example.</p>
</div>
<div class="section" id="trimming-off-adaptor-sequences">
<h3>18.1.8  Trimming off adaptor sequences</h3>
<p>This is essentially a simple extension to the previous example. We are
going to going to pretend <tt class="docutils literal">GATGACGGTGT</tt> is an adaptor sequence in some
FASTQ formatted read data, again the <tt class="docutils literal">SRR020192.fastq</tt> file from the
NCBI
(<a class="reference external" href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz">ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz</a>).</p>
<p>This time however, we will look for the sequence <em>anywhere</em> in the
reads, not just at the very beginning:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>

<span class="keyword">def</span> <span class="name function">trim_adaptors</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">adaptor</span><span class="punctuation">):</span>
    <span class="literal string doc">&quot;&quot;&quot;Trims perfect adaptor sequences.

    This is a generator function, the records argument should
    be a list or iterator returning SeqRecord objects.
    &quot;&quot;&quot;</span>
    <span class="name">len_adaptor</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">adaptor</span><span class="punctuation">)</span> <span class="comment">#cache this for later</span>
    <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
        <span class="name">index</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">find</span><span class="punctuation">(</span><span class="name">adaptor</span><span class="punctuation">)</span>
        <span class="keyword">if</span> <span class="name">index</span> <span class="operator">==</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">:</span>
            <span class="comment">#adaptor not found, so won't trim</span>
            <span class="keyword">yield</span> <span class="name">record</span>
        <span class="keyword">else</span><span class="punctuation">:</span>
            <span class="comment">#trim off the adaptor</span>
            <span class="keyword">yield</span> <span class="name">record</span><span class="punctuation">[</span><span class="name">index</span><span class="operator">+</span><span class="name">len_adaptor</span><span class="punctuation">:]</span>

<span class="name">original_reads</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="name">trimmed_reads</span> <span class="operator">=</span> <span class="name">trim_adaptors</span><span class="punctuation">(</span><span class="name">original_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;GATGACGGTGT&quot;</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">trimmed_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>Because we are using a FASTQ input file in this example, the
<tt class="docutils literal">SeqRecord</tt> objects have per-letter-annotation for the quality scores.
By slicing the <tt class="docutils literal">SeqRecord</tt> object the appropriate scores are used on
the trimmed records, so we can output them as a FASTQ file too.</p>
<p>Compared to the output of the previous example where we only looked for
a primer/adaptor at the start of each read, you may find some of the
trimmed reads are quite short after trimming (e.g. if the adaptor was
found in the middle rather than near the start). So, let’s add a minimum
length requirement as well:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>

<span class="keyword">def</span> <span class="name function">trim_adaptors</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">adaptor</span><span class="punctuation">,</span> <span class="name">min_len</span><span class="punctuation">):</span>
    <span class="literal string doc">&quot;&quot;&quot;Trims perfect adaptor sequences, checks read length.

    This is a generator function, the records argument should
    be a list or iterator returning SeqRecord objects.
    &quot;&quot;&quot;</span>
    <span class="name">len_adaptor</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">adaptor</span><span class="punctuation">)</span> <span class="comment">#cache this for later</span>
    <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
        <span class="name">len_record</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span> <span class="comment">#cache this for later</span>
        <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span> <span class="operator">&lt;</span> <span class="name">min_len</span><span class="punctuation">:</span>
           <span class="comment">#Too short to keep</span>
           <span class="keyword">continue</span>
        <span class="name">index</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">find</span><span class="punctuation">(</span><span class="name">adaptor</span><span class="punctuation">)</span>
        <span class="keyword">if</span> <span class="name">index</span> <span class="operator">==</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">:</span>
            <span class="comment">#adaptor not found, so won't trim</span>
            <span class="keyword">yield</span> <span class="name">record</span>
        <span class="keyword">elif</span> <span class="name">len_record</span> <span class="operator">-</span> <span class="name">index</span> <span class="operator">-</span> <span class="name">len_adaptor</span> <span class="operator">&gt;=</span> <span class="name">min_len</span><span class="punctuation">:</span>
            <span class="comment">#after trimming this will still be long enough</span>
            <span class="keyword">yield</span> <span class="name">record</span><span class="punctuation">[</span><span class="name">index</span><span class="operator">+</span><span class="name">len_adaptor</span><span class="punctuation">:]</span>

<span class="name">original_reads</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="name">trimmed_reads</span> <span class="operator">=</span> <span class="name">trim_adaptors</span><span class="punctuation">(</span><span class="name">original_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;GATGACGGTGT&quot;</span><span class="punctuation">,</span> <span class="literal number integer">100</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">trimmed_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>By changing the format names, you could apply this to FASTA files
instead. This code also could be extended to do a fuzzy match instead of
an exact match (maybe using a pairwise alignment, or taking into account
the read quality scores), but that will be much slower.</p>
</div>
<div class="section" id="converting-fastq-files">
<h3>18.1.9  Converting FASTQ files</h3>
<p>Back in Section <a class="reference external" href="#sec:SeqIO-conversion">5.5.2</a> we showed how to use
<tt class="docutils literal">Bio.SeqIO</tt> to convert between two file formats. Here we’ll go into a
little more detail regarding FASTQ files which are used in second
generation DNA sequencing. Please refer to Cock <em>et al.</em> (2009)
[<a class="reference external" href="#cock2010">7</a>] for a longer description. FASTQ files store both
the DNA sequence (as a string) and the associated read qualities.</p>
<p>PHRED scores (used in most FASTQ files, and also in QUAL files, ACE
files and SFF files) have become a <em>de facto</em> standard for representing
the probability of a sequencing error (here denoted by <em>P</em><sub>*e*</sub>)
at a given base using a simple base ten log transformation:</p>
<div class="formula">
<i>Q</i><sub><span class="textrm">PHRED</span></sub> =  − 10 × <span class="textrm"> log</span><sub>10</sub>(<i>P</i><sub><i>e</i></sub>)
</div>
<p>This means a wrong read (<em>P</em><sub>*e*</sub> = 1) gets a PHRED quality of 0,
while a very good read like <em>P</em><sub>*e*</sub> = 0.00001 gets a PHRED
quality of 50. While for raw sequencing data qualities higher than this
are rare, with post processing such as read mapping or assembly,
qualities of up to about 90 are possible (indeed, the MAQ tool allows
for PHRED scores in the range 0 to 93 inclusive).</p>
<p>The FASTQ format has the potential to become a <em>de facto</em> standard for
storing the letters and quality scores for a sequencing read in a single
plain text file. The only fly in the ointment is that there are at least
three versions of the FASTQ format which are incompatible and difficult
to distinguish...</p>
<ol class="arabic simple">
<li>The original Sanger FASTQ format uses PHRED qualities encoded with an
ASCII offset of 33. The NCBI are using this format in their Short
Read Archive. We call this the <tt class="docutils literal">fastq</tt> (or <tt class="docutils literal"><span class="pre">fastq-sanger</span></tt>) format
in <tt class="docutils literal">Bio.SeqIO</tt>.</li>
<li>Solexa (later bought by Illumina) introduced their own version using
Solexa qualities encoded with an ASCII offset of 64. We call this the
<tt class="docutils literal"><span class="pre">fastq-solexa</span></tt> format.</li>
<li>Illumina pipeline 1.3 onwards produces FASTQ files with PHRED
qualities (which is more consistent), but encoded with an ASCII
offset of 64. We call this the <tt class="docutils literal"><span class="pre">fastq-illumina</span></tt> format.</li>
</ol>
<p>The Solexa quality scores are defined using a different log
transformation:</p>
<div class="formula">
<i>Q</i><sub><span class="textrm">Solexa</span></sub> =  − 10 × <span class="textrm"> log</span><sub>10</sub><span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="fraction"><span class="ignored">(</span><span class="numerator"><i>P</i><sub><i>e</i></sub></span><span class="ignored">)/(</span><span class="denominator">1 − <i>P</i><sub><i>e</i></sub></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>
</div>
<p>Given Solexa/Illumina have now moved to using PHRED scores in version
1.3 of their pipeline, the Solexa quality scores will gradually fall out
of use. If you equate the error estimates (<em>P</em><sub>*e*</sub>) these two
equations allow conversion between the two scoring systems - and
Biopython includes functions to do this in the <tt class="docutils literal">Bio.SeqIO.QualityIO</tt>
module, which are called if you use <tt class="docutils literal">Bio.SeqIO</tt> to convert an old
Solexa/Illumina file into a standard Sanger FASTQ file:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;solexa.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq-solexa&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;standard.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
</pre>
<p>If you want to convert a new Illumina 1.3+ FASTQ file, all that gets
changed is the ASCII offset because although encoded differently the
scores are all PHRED qualities:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;illumina.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq-illumina&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;standard.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
</pre>
<p>Note that using <tt class="docutils literal">Bio.SeqIO.convert()</tt> like this is <em>much</em> faster than
combining <tt class="docutils literal">Bio.SeqIO.parse()</tt> and <tt class="docutils literal">Bio.SeqIO.write()</tt> because
optimised code is used for converting between FASTQ variants (and also
for FASTQ to FASTA conversion).</p>
<p>For good quality reads, PHRED and Solexa scores are approximately equal,
which means since both the <tt class="docutils literal"><span class="pre">fasta-solexa</span></tt> and <tt class="docutils literal"><span class="pre">fastq-illumina</span></tt>
formats use an ASCII offset of 64 the files are almost the same. This
was a deliberate design choice by Illumina, meaning applications
expecting the old <tt class="docutils literal"><span class="pre">fasta-solexa</span></tt> style files will probably be OK using
the newer <tt class="docutils literal"><span class="pre">fastq-illumina</span></tt> files (on good data). Of course, both
variants are very different from the original FASTQ standard as used by
Sanger, the NCBI, and elsewhere (format name <tt class="docutils literal">fastq</tt> or
<tt class="docutils literal"><span class="pre">fastq-sanger</span></tt>).</p>
<p>For more details, see the built in help (also
<a class="reference external" href="http://www.biopython.org/DIST/docs/api/Bio.SeqIO.QualityIO-module.html">online</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqIO</span> <span class="keyword namespace">import</span> <span class="name">QualityIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">QualityIO</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="converting-fasta-and-qual-files-into-fastq-files">
<h3>18.1.10  Converting FASTA and QUAL files into FASTQ files</h3>
<p>FASTQ files hold <em>both</em> sequences and their quality strings. FASTA files
hold <em>just</em> sequences, while QUAL files hold <em>just</em> the qualities.
Therefore a single FASTQ file can be converted to or from <em>paired</em> FASTA
and QUAL files.</p>
<p>Going from FASTQ to FASTA is easy:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;example.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;example.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>Going from FASTQ to QUAL is also easy:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;example.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;example.qual&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;qual&quot;</span><span class="punctuation">)</span>
</pre>
<p>However, the reverse is a little more tricky. You can use
<tt class="docutils literal">Bio.SeqIO.parse()</tt> to iterate over the records in a <em>single</em> file,
but in this case we have two input files. There are several strategies
possible, but assuming that the two files are really paired the most
memory efficient way is to loop over both together. The code is a little
fiddly, so we provide a function called <tt class="docutils literal">PairedFastaQualIterator</tt> in
the <tt class="docutils literal">Bio.SeqIO.QualityIO</tt> module to do this. This takes two handles
(the FASTA file and the QUAL file) and returns a <tt class="docutils literal">SeqRecord</tt> iterator:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqIO.QualityIO</span> <span class="keyword namespace">import</span> <span class="name">PairedFastaQualIterator</span>
<span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">PairedFastaQualIterator</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;example.fasta&quot;</span><span class="punctuation">),</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;example.qual&quot;</span><span class="punctuation">)):</span>
   <span class="keyword">print</span> <span class="name">record</span>
</pre>
<p>This function will check that the FASTA and QUAL files are consistent
(e.g. the records are in the same order, and have the same sequence
length). You can combine this with the <tt class="docutils literal">Bio.SeqIO.write()</tt> function to
convert a pair of FASTA and QUAL files into a single FASTQ files:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqIO.QualityIO</span> <span class="keyword namespace">import</span> <span class="name">PairedFastaQualIterator</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;temp.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span> <span class="comment">#w=write</span>
<span class="name">records</span> <span class="operator">=</span> <span class="name">PairedFastaQualIterator</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;example.fasta&quot;</span><span class="punctuation">),</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;example.qual&quot;</span><span class="punctuation">))</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="literal string">&quot;Converted </span><span class="literal string interpol">%i</span><span class="literal string"> records&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
</div>
<div class="section" id="indexing-a-fastq-file">
<h3>18.1.11  Indexing a FASTQ file</h3>
<p>FASTQ files are often very large, with millions of reads in them. Due to
the sheer amount of data, you can’t load all the records into memory at
once. This is why the examples above (filtering and trimming) iterate
over the file looking at just one <tt class="docutils literal">SeqRecord</tt> at a time.</p>
<p>However, sometimes you can’t use a big loop or an iterator - you may
need random access to the reads. Here the <tt class="docutils literal">Bio.SeqIO.index()</tt> function
may prove very helpful, as it allows you to access any read in the FASTQ
file by its name (see Section <a class="reference external" href="#sec:SeqIO-index">5.4.2</a>).</p>
<p>Again we’ll use the <tt class="docutils literal">SRR020192.fastq</tt> file from the ENA
(<a class="reference external" href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz">ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz</a>),
although this is actually quite a small FASTQ file with less than 50,000
reads:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">fq_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">fq_dict</span><span class="punctuation">)</span>
<span class="literal number integer">41892</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">fq_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()[:</span><span class="literal number integer">4</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'SRR020192.38240'</span><span class="punctuation">,</span> <span class="literal string">'SRR020192.23181'</span><span class="punctuation">,</span> <span class="literal string">'SRR020192.40568'</span><span class="punctuation">,</span> <span class="literal string">'SRR020192.23186'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">fq_dict</span><span class="punctuation">[</span><span class="literal string">&quot;SRR020192.23186&quot;</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GTCCCAGTATTCGGATTTGTCTGCCAAAACAATGAAATTGACACAGTTTACAAC...CCG'</span><span class="punctuation">,</span> <span class="name">SingleLetterAlphabet</span><span class="punctuation">())</span>
</pre>
<p>When testing this on a FASTQ file with seven million reads, indexing
took about a minute, but record access was almost instant.</p>
<p>The example in Section <a class="reference external" href="#sec:SeqIO-sort">18.1.5</a> show how you can
use the <tt class="docutils literal">Bio.SeqIO.index()</tt> function to sort a large FASTA file – this
could also be used on FASTQ files.</p>
</div>
<div class="section" id="converting-sff-files">
<h3>18.1.12  Converting SFF files</h3>
<p>If you work with 454 (Roche) sequence data, you will probably have
access to the raw data as a Standard Flowgram Format (SFF) file. This
contains the sequence reads (called bases) with quality scores and the
original flow information.</p>
<p>A common task is to convert from SFF to a pair of FASTA and QUAL files,
or to a single FASTQ file. These operations are trivial using the
<tt class="docutils literal">Bio.SeqIO.convert()</tt> function (see
Section <a class="reference external" href="#sec:SeqIO-conversion">5.5.2</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;E3MFGYR02_random_10_reads.sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;reads.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">10</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;E3MFGYR02_random_10_reads.sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;reads.qual&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;qual&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">10</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;E3MFGYR02_random_10_reads.sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;reads.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">10</span>
</pre>
<p>Remember the convert function returns the number of records, in this
example just ten. This will give you the <em>untrimmed</em> reads, where the
leading and trailing poor quality sequence or adaptor will be in lower
case. If you want the <em>trimmed</em> reads (using the clipping information
recorded within the SFF file) use this:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;E3MFGYR02_random_10_reads.sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;sff-trim&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">10</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;E3MFGYR02_random_10_reads.sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;sff-trim&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.qual&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;qual&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">10</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;E3MFGYR02_random_10_reads.sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;sff-trim&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">10</span>
</pre>
<p>If you run Linux, you could ask Roche for a copy of their “off
instrument” tools (often referred to as the Newbler tools). This offers
an alternative way to do SFF to FASTA or QUAL conversion at the command
line (but currently FASTQ output is not supported), e.g.</p>
<pre class="code python literal-block">
<span class="error">$</span> <span class="name">sffinfo</span> <span class="operator">-</span><span class="name">seq</span> <span class="operator">-</span><span class="name">notrim</span> <span class="name">E3MFGYR02_random_10_reads</span><span class="operator">.</span><span class="name">sff</span> <span class="operator">&gt;</span> <span class="name">reads</span><span class="operator">.</span><span class="name">fasta</span>
<span class="error">$</span> <span class="name">sffinfo</span> <span class="operator">-</span><span class="name">qual</span> <span class="operator">-</span><span class="name">notrim</span> <span class="name">E3MFGYR02_random_10_reads</span><span class="operator">.</span><span class="name">sff</span> <span class="operator">&gt;</span> <span class="name">reads</span><span class="operator">.</span><span class="name">qual</span>
<span class="error">$</span> <span class="name">sffinfo</span> <span class="operator">-</span><span class="name">seq</span> <span class="operator">-</span><span class="name">trim</span> <span class="name">E3MFGYR02_random_10_reads</span><span class="operator">.</span><span class="name">sff</span> <span class="operator">&gt;</span> <span class="name">trimmed</span><span class="operator">.</span><span class="name">fasta</span>
<span class="error">$</span> <span class="name">sffinfo</span> <span class="operator">-</span><span class="name">qual</span> <span class="operator">-</span><span class="name">trim</span> <span class="name">E3MFGYR02_random_10_reads</span><span class="operator">.</span><span class="name">sff</span> <span class="operator">&gt;</span> <span class="name">trimmed</span><span class="operator">.</span><span class="name">qual</span>
</pre>
<p>The way Biopython uses mixed case sequence strings to represent the
trimming points deliberately mimics what the Roche tools do.</p>
<p>For more information on the Biopython SFF support, consult the built in
help:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqIO</span> <span class="keyword namespace">import</span> <span class="name">SffIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">SffIO</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="identifying-open-reading-frames">
<h3>18.1.13  Identifying open reading frames</h3>
<p>A very simplistic first step at identifying possible genes is to look
for open reading frames (ORFs). By this we mean look in all six frames
for long regions without stop codons – an ORF is just a region of
nucleotides with no in frame stop codons.</p>
<p>Of course, to find a gene you would also need to worry about locating a
start codon, possible promoters – and in Eukaryotes there are introns to
worry about too. However, this approach is still useful in viruses and
Prokaryotes.</p>
<p>To show how you might approach this with Biopython, we’ll need a
sequence to search, and as an example we’ll again use the bacterial
plasmid – although this time we’ll start with a plain FASTA file with no
pre-marked genes:
<a class="reference external" href="http://biopython.org/SRC/biopython/Tests/GenBank/NC_005816.fna">NC_005816.fna</a>.
This is a bacterial sequence, so we’ll want to use NCBI codon table 11
(see Section <a class="reference external" href="#sec:translation">3.9</a> about translation).</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.fna&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">table</span> <span class="operator">=</span> <span class="literal number integer">11</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">min_pro_len</span> <span class="operator">=</span> <span class="literal number integer">100</span>
</pre>
<p>Here is a neat trick using the <tt class="docutils literal">Seq</tt> object’s <tt class="docutils literal">split</tt> method to get
a list of all the possible ORF translations in the six reading frames:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">strand</span><span class="punctuation">,</span> <span class="name">nuc</span> <span class="operator word">in</span> <span class="punctuation">[(</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">())]:</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">frame</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">):</span>
<span class="operator">...</span>         <span class="name">length</span> <span class="operator">=</span> <span class="literal number integer">3</span> <span class="operator">*</span> <span class="punctuation">((</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">frame</span><span class="punctuation">)</span> <span class="operator">//</span> <span class="literal number integer">3</span><span class="punctuation">)</span> <span class="comment">#Multiple of three</span>
<span class="operator">...</span>         <span class="keyword">for</span> <span class="name">pro</span> <span class="operator word">in</span> <span class="name">nuc</span><span class="punctuation">[</span><span class="name">frame</span><span class="punctuation">:</span><span class="name">frame</span><span class="operator">+</span><span class="name">length</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">(</span><span class="literal string">&quot;*&quot;</span><span class="punctuation">):</span>
<span class="operator">...</span>             <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">pro</span><span class="punctuation">)</span> <span class="operator">&gt;=</span> <span class="name">min_pro_len</span><span class="punctuation">:</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">...</span><span class="literal string interpol">%s</span><span class="literal string"> - length </span><span class="literal string interpol">%i</span><span class="literal string">, strand </span><span class="literal string interpol">%i</span><span class="literal string">, frame </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> \
<span class="operator">...</span>                       <span class="operator">%</span> <span class="punctuation">(</span><span class="name">pro</span><span class="punctuation">[:</span><span class="literal number integer">30</span><span class="punctuation">],</span> <span class="name">pro</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">3</span><span class="punctuation">:],</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">pro</span><span class="punctuation">),</span> <span class="name">strand</span><span class="punctuation">,</span> <span class="name">frame</span><span class="punctuation">)</span>
<span class="name">GCLMKKSSIVATIITILSGSANAASSQLIP</span><span class="operator">...</span><span class="name">YRF</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">315</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">0</span>
<span class="name">KSGELRQTPPASSTLHLRLILQRSGVMMEL</span><span class="operator">...</span><span class="name">NPE</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">285</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">1</span>
<span class="name">GLNCSFFSICNWKFIDYINRLFQIIYLCKN</span><span class="operator">...</span><span class="name">YYH</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">176</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">1</span>
<span class="name">VKKILYIKALFLCTVIKLRRFIFSVNNMKF</span><span class="operator">...</span><span class="name">DLP</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">165</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">1</span>
<span class="name">NQIQGVICSPDSGEFMVTFETVMEIKILHK</span><span class="operator">...</span><span class="name">GVA</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">355</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">2</span>
<span class="name">RRKEHVSKKRRPQKRPRRRRFFHRLRPPDE</span><span class="operator">...</span><span class="name">PTR</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">128</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">2</span>
<span class="name">TGKQNSCQMSAIWQLRQNTATKTRQNRARI</span><span class="operator">...</span><span class="name">AIK</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">100</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">2</span>
<span class="name">QGSGYAFPHASILSGIAMSHFYFLVLHAVK</span><span class="operator">...</span><span class="name">CSD</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">114</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">0</span>
<span class="name">IYSTSEHTGEQVMRTLDEVIASRSPESQTR</span><span class="operator">...</span><span class="name">FHV</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">111</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">0</span>
<span class="name">WGKLQVIGLSMWMVLFSQRFDDWLNEQEDA</span><span class="operator">...</span><span class="name">ESK</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">125</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">1</span>
<span class="name">RGIFMSDTMVVNGSGGVPAFLFSGSTLSSY</span><span class="operator">...</span><span class="name">LLK</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">361</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">1</span>
<span class="name">WDVKTVTGVLHHPFHLTFSLCPEGATQSGR</span><span class="operator">...</span><span class="name">VKR</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">111</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">1</span>
<span class="name">LSHTVTDFTDQMAQVGLCQCVNVFLDEVTG</span><span class="operator">...</span><span class="name">KAA</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">107</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">2</span>
<span class="name">RALTGLSAPGIRSQTSCDRLRELRYVPVSL</span><span class="operator">...</span><span class="name">PLQ</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">119</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">2</span>
</pre>
<p>Note that here we are counting the frames from the 5’ end (start) of
<em>each</em> strand. It is sometimes easier to always count from the 5’ end
(start) of the <em>forward</em> strand.</p>
<p>You could easily edit the above loop based code to build up a list of
the candidate proteins, or convert this to a list comprehension. Now,
one thing this code doesn’t do is keep track of where the proteins are.</p>
<p>You could tackle this in several ways. For example, the following code
tracks the locations in terms of the protein counting, and converts back
to the parent sequence by multiplying by three, then adjusting for the
frame and strand:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="name">table</span> <span class="operator">=</span> <span class="literal number integer">11</span>
<span class="name">min_pro_len</span> <span class="operator">=</span> <span class="literal number integer">100</span>

<span class="keyword">def</span> <span class="name function">find_orfs_with_trans</span><span class="punctuation">(</span><span class="name">seq</span><span class="punctuation">,</span> <span class="name">trans_table</span><span class="punctuation">,</span> <span class="name">min_protein_length</span><span class="punctuation">):</span>
    <span class="name">answer</span> <span class="operator">=</span> <span class="punctuation">[]</span>
    <span class="name">seq_len</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq</span><span class="punctuation">)</span>
    <span class="keyword">for</span> <span class="name">strand</span><span class="punctuation">,</span> <span class="name">nuc</span> <span class="operator word">in</span> <span class="punctuation">[(</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">seq</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">seq</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">())]:</span>
        <span class="keyword">for</span> <span class="name">frame</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">):</span>
            <span class="name">trans</span> <span class="operator">=</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">nuc</span><span class="punctuation">[</span><span class="name">frame</span><span class="punctuation">:]</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">trans_table</span><span class="punctuation">))</span>
            <span class="name">trans_len</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">trans</span><span class="punctuation">)</span>
            <span class="name">aa_start</span> <span class="operator">=</span> <span class="literal number integer">0</span>
            <span class="name">aa_end</span> <span class="operator">=</span> <span class="literal number integer">0</span>
            <span class="keyword">while</span> <span class="name">aa_start</span> <span class="operator">&lt;</span> <span class="name">trans_len</span><span class="punctuation">:</span>
                <span class="name">aa_end</span> <span class="operator">=</span> <span class="name">trans</span><span class="operator">.</span><span class="name">find</span><span class="punctuation">(</span><span class="literal string">&quot;*&quot;</span><span class="punctuation">,</span> <span class="name">aa_start</span><span class="punctuation">)</span>
                <span class="keyword">if</span> <span class="name">aa_end</span> <span class="operator">==</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">:</span>
                    <span class="name">aa_end</span> <span class="operator">=</span> <span class="name">trans_len</span>
                <span class="keyword">if</span> <span class="name">aa_end</span><span class="operator">-</span><span class="name">aa_start</span> <span class="operator">&gt;=</span> <span class="name">min_protein_length</span><span class="punctuation">:</span>
                    <span class="keyword">if</span> <span class="name">strand</span> <span class="operator">==</span> <span class="literal number integer">1</span><span class="punctuation">:</span>
                        <span class="name">start</span> <span class="operator">=</span> <span class="name">frame</span><span class="operator">+</span><span class="name">aa_start</span><span class="operator">*</span><span class="literal number integer">3</span>
                        <span class="name">end</span> <span class="operator">=</span> <span class="name builtin">min</span><span class="punctuation">(</span><span class="name">seq_len</span><span class="punctuation">,</span><span class="name">frame</span><span class="operator">+</span><span class="name">aa_end</span><span class="operator">*</span><span class="literal number integer">3</span><span class="operator">+</span><span class="literal number integer">3</span><span class="punctuation">)</span>
                    <span class="keyword">else</span><span class="punctuation">:</span>
                        <span class="name">start</span> <span class="operator">=</span> <span class="name">seq_len</span><span class="operator">-</span><span class="name">frame</span><span class="operator">-</span><span class="name">aa_end</span><span class="operator">*</span><span class="literal number integer">3</span><span class="operator">-</span><span class="literal number integer">3</span>
                        <span class="name">end</span> <span class="operator">=</span> <span class="name">seq_len</span><span class="operator">-</span><span class="name">frame</span><span class="operator">-</span><span class="name">aa_start</span><span class="operator">*</span><span class="literal number integer">3</span>
                    <span class="name">answer</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">((</span><span class="name">start</span><span class="punctuation">,</span> <span class="name">end</span><span class="punctuation">,</span> <span class="name">strand</span><span class="punctuation">,</span>
                                   <span class="name">trans</span><span class="punctuation">[</span><span class="name">aa_start</span><span class="punctuation">:</span><span class="name">aa_end</span><span class="punctuation">]))</span>
                <span class="name">aa_start</span> <span class="operator">=</span> <span class="name">aa_end</span><span class="operator">+</span><span class="literal number integer">1</span>
    <span class="name">answer</span><span class="operator">.</span><span class="name">sort</span><span class="punctuation">()</span>
    <span class="keyword">return</span> <span class="name">answer</span>

<span class="name">orf_list</span> <span class="operator">=</span> <span class="name">find_orfs_with_trans</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">,</span> <span class="name">table</span><span class="punctuation">,</span> <span class="name">min_pro_len</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">start</span><span class="punctuation">,</span> <span class="name">end</span><span class="punctuation">,</span> <span class="name">strand</span><span class="punctuation">,</span> <span class="name">pro</span> <span class="operator word">in</span> <span class="name">orf_list</span><span class="punctuation">:</span>
    <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">...</span><span class="literal string interpol">%s</span><span class="literal string"> - length </span><span class="literal string interpol">%i</span><span class="literal string">, strand </span><span class="literal string interpol">%i</span><span class="literal string">, </span><span class="literal string interpol">%i</span><span class="literal string">:</span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> \
          <span class="operator">%</span> <span class="punctuation">(</span><span class="name">pro</span><span class="punctuation">[:</span><span class="literal number integer">30</span><span class="punctuation">],</span> <span class="name">pro</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">3</span><span class="punctuation">:],</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">pro</span><span class="punctuation">),</span> <span class="name">strand</span><span class="punctuation">,</span> <span class="name">start</span><span class="punctuation">,</span> <span class="name">end</span><span class="punctuation">)</span>
</pre>
<p>And the output:</p>
<pre class="code python literal-block">
<span class="name">NQIQGVICSPDSGEFMVTFETVMEIKILHK</span><span class="operator">...</span><span class="name">GVA</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">355</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">41</span><span class="punctuation">:</span><span class="literal number integer">1109</span>
<span class="name">WDVKTVTGVLHHPFHLTFSLCPEGATQSGR</span><span class="operator">...</span><span class="name">VKR</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">111</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">491</span><span class="punctuation">:</span><span class="literal number integer">827</span>
<span class="name">KSGELRQTPPASSTLHLRLILQRSGVMMEL</span><span class="operator">...</span><span class="name">NPE</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">285</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">1030</span><span class="punctuation">:</span><span class="literal number integer">1888</span>
<span class="name">RALTGLSAPGIRSQTSCDRLRELRYVPVSL</span><span class="operator">...</span><span class="name">PLQ</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">119</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2830</span><span class="punctuation">:</span><span class="literal number integer">3190</span>
<span class="name">RRKEHVSKKRRPQKRPRRRRFFHRLRPPDE</span><span class="operator">...</span><span class="name">PTR</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">128</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">3470</span><span class="punctuation">:</span><span class="literal number integer">3857</span>
<span class="name">GLNCSFFSICNWKFIDYINRLFQIIYLCKN</span><span class="operator">...</span><span class="name">YYH</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">176</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">4249</span><span class="punctuation">:</span><span class="literal number integer">4780</span>
<span class="name">RGIFMSDTMVVNGSGGVPAFLFSGSTLSSY</span><span class="operator">...</span><span class="name">LLK</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">361</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">4814</span><span class="punctuation">:</span><span class="literal number integer">5900</span>
<span class="name">VKKILYIKALFLCTVIKLRRFIFSVNNMKF</span><span class="operator">...</span><span class="name">DLP</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">165</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">5923</span><span class="punctuation">:</span><span class="literal number integer">6421</span>
<span class="name">LSHTVTDFTDQMAQVGLCQCVNVFLDEVTG</span><span class="operator">...</span><span class="name">KAA</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">107</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">5974</span><span class="punctuation">:</span><span class="literal number integer">6298</span>
<span class="name">GCLMKKSSIVATIITILSGSANAASSQLIP</span><span class="operator">...</span><span class="name">YRF</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">315</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">6654</span><span class="punctuation">:</span><span class="literal number integer">7602</span>
<span class="name">IYSTSEHTGEQVMRTLDEVIASRSPESQTR</span><span class="operator">...</span><span class="name">FHV</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">111</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">7788</span><span class="punctuation">:</span><span class="literal number integer">8124</span>
<span class="name">WGKLQVIGLSMWMVLFSQRFDDWLNEQEDA</span><span class="operator">...</span><span class="name">ESK</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">125</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">8087</span><span class="punctuation">:</span><span class="literal number integer">8465</span>
<span class="name">TGKQNSCQMSAIWQLRQNTATKTRQNRARI</span><span class="operator">...</span><span class="name">AIK</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">100</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">8741</span><span class="punctuation">:</span><span class="literal number integer">9044</span>
<span class="name">QGSGYAFPHASILSGIAMSHFYFLVLHAVK</span><span class="operator">...</span><span class="name">CSD</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">114</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">9264</span><span class="punctuation">:</span><span class="literal number integer">9609</span>
</pre>
<p>If you comment out the sort statement, then the protein sequences will
be shown in the same order as before, so you can check this is doing the
same thing. Here we have sorted them by location to make it easier to
compare to the actual annotation in the GenBank file (as visualised in
Section <a class="reference external" href="#sec:gd_nice_example">17.1.9</a>).</p>
<p>If however all you want to find are the locations of the open reading
frames, then it is a waste of time to translate every possible codon,
including doing the reverse complement to search the reverse strand too.
All you need to do is search for the possible stop codons (and their
reverse complements). Using regular expressions is an obvious approach
here (see the Python module <tt class="docutils literal">re</tt>). These are an extremely powerful
(but rather complex) way of describing search strings, which are
supported in lots of programming languages and also command line tools
like <tt class="docutils literal">grep</tt> as well). You can find whole books about this topic!</p>
</div>
</div>
<div class="section" id="sequence-parsing-plus-simple-plots">
<h2>18.2  Sequence parsing plus simple plots</h2>
<p>This section shows some more examples of sequence parsing, using the
<tt class="docutils literal">Bio.SeqIO</tt> module described in Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>,
plus the Python library matplotlib’s <tt class="docutils literal">pylab</tt> plotting interface (see
<a class="reference external" href="http://matplotlib.sourceforge.net/">the matplotlib website for a
tutorial</a>). Note that to follow
these examples you will need matplotlib installed - but without it you
can still try the data parsing bits.</p>
<div class="section" id="histogram-of-sequence-lengths">
<h3>18.2.1  Histogram of sequence lengths</h3>
<p>There are lots of times when you might want to visualise the
distribution of sequence lengths in a dataset – for example the range of
contig sizes in a genome assembly project. In this example we’ll reuse
our orchid FASTA file
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta">ls_orchid.fasta</a>
which has only 94 sequences.</p>
<p>First of all, we will use <tt class="docutils literal">Bio.SeqIO</tt> to parse the FASTA file and
compile a list of all the sequence lengths. You could do this with a for
loop, but I find a list comprehension more pleasing:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sizes</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">),</span> <span class="name builtin">min</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">),</span> <span class="name builtin">max</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">)</span>
<span class="punctuation">(</span><span class="literal number integer">94</span><span class="punctuation">,</span> <span class="literal number integer">572</span><span class="punctuation">,</span> <span class="literal number integer">789</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sizes</span>
<span class="punctuation">[</span><span class="literal number integer">740</span><span class="punctuation">,</span> <span class="literal number integer">753</span><span class="punctuation">,</span> <span class="literal number integer">748</span><span class="punctuation">,</span> <span class="literal number integer">744</span><span class="punctuation">,</span> <span class="literal number integer">733</span><span class="punctuation">,</span> <span class="literal number integer">718</span><span class="punctuation">,</span> <span class="literal number integer">730</span><span class="punctuation">,</span> <span class="literal number integer">704</span><span class="punctuation">,</span> <span class="literal number integer">740</span><span class="punctuation">,</span> <span class="literal number integer">709</span><span class="punctuation">,</span> <span class="literal number integer">700</span><span class="punctuation">,</span> <span class="literal number integer">726</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal number integer">592</span><span class="punctuation">]</span>
</pre>
<p>Now that we have the lengths of all the genes (as a list of integers),
we can use the matplotlib histogram function to display it.</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">sizes</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)]</span>

<span class="keyword namespace">import</span> <span class="name namespace">pylab</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">hist</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">,</span> <span class="name">bins</span><span class="operator">=</span><span class="literal number integer">20</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">title</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string interpol">%i</span><span class="literal string"> orchid sequences</span><span class="literal string escape">\n</span><span class="literal string">Lengths </span><span class="literal string interpol">%i</span><span class="literal string"> to </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> \
            <span class="operator">%</span> <span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">),</span><span class="name builtin">min</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">),</span><span class="name builtin">max</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">)))</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">xlabel</span><span class="punctuation">(</span><span class="literal string">&quot;Sequence length (bp)&quot;</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">ylabel</span><span class="punctuation">(</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">show</span><span class="punctuation">()</span>
</pre>
<p>That should pop up a new window containing the following graph:</p>
<p><img alt="image26" src="../images/hist_plot.png" /></p>
<p>Notice that most of these orchid sequences are about 740 bp long, and
there could be two distinct classes of sequence here with a subset of
shorter sequences.</p>
<p><em>Tip:</em> Rather than using <tt class="docutils literal">pylab.show()</tt> to show the plot in a window,
you can also use <tt class="docutils literal"><span class="pre">pylab.savefig(...)</span></tt> to save the figure to a file
(e.g. as a PNG or PDF).</p>
</div>
<div class="section" id="plot-of-sequence-gc">
<h3>18.2.2  Plot of sequence GC%</h3>
<p>Another easily calculated quantity of a nucleotide sequence is the GC%.
You might want to look at the GC% of all the genes in a bacterial genome
for example, and investigate any outliers which could have been recently
acquired by horizontal gene transfer. Again, for this example we’ll
reuse our orchid FASTA file
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta">ls_orchid.fasta</a>.</p>
<p>First of all, we will use <tt class="docutils literal">Bio.SeqIO</tt> to parse the FASTA file and
compile a list of all the GC percentages. Again, you could do this with
a for loop, but I prefer this:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqUtils</span> <span class="keyword namespace">import</span> <span class="name">GC</span>

<span class="name">gc_values</span> <span class="operator">=</span> <span class="name builtin">sorted</span><span class="punctuation">(</span><span class="name">GC</span><span class="punctuation">(</span><span class="name">rec</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
</pre>
<p>Having read in each sequence and calculated the GC%, we then sorted them
into ascending order. Now we’ll take this list of floating point values
and plot them with matplotlib:</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">pylab</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">plot</span><span class="punctuation">(</span><span class="name">gc_values</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">title</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string interpol">%i</span><span class="literal string"> orchid sequences</span><span class="literal string escape">\n</span><span class="literal string">GC</span><span class="literal string interpol">%%</span><span class="literal string"> </span><span class="literal string interpol">%0.1f</span><span class="literal string"> to </span><span class="literal string interpol">%0.1f</span><span class="literal string">&quot;</span> \
            <span class="operator">%</span> <span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">gc_values</span><span class="punctuation">),</span><span class="name builtin">min</span><span class="punctuation">(</span><span class="name">gc_values</span><span class="punctuation">),</span><span class="name builtin">max</span><span class="punctuation">(</span><span class="name">gc_values</span><span class="punctuation">)))</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">xlabel</span><span class="punctuation">(</span><span class="literal string">&quot;Genes&quot;</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">ylabel</span><span class="punctuation">(</span><span class="literal string">&quot;GC%&quot;</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">show</span><span class="punctuation">()</span>
</pre>
<p>As in the previous example, that should pop up a new window containing a
graph:</p>
<p><img alt="image27" src="../images/gc_plot.png" /></p>
<p>If you tried this on the full set of genes from one organism, you’d
probably get a much smoother plot than this.</p>
</div>
<div class="section" id="nucleotide-dot-plots">
<h3>18.2.3  Nucleotide dot plots</h3>
<p>A dot plot is a way of visually comparing two nucleotide sequences for
similarity to each other. A sliding window is used to compare short
sub-sequences to each other, often with a mis-match threshold. Here for
simplicity we’ll only look for perfect matches (shown in black in the
plot below).</p>
<p>To start off, we’ll need two sequences. For the sake of argument, we’ll
just take the first two from our orchid FASTA file
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta">ls_orchid.fasta</a>:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">)</span>
<span class="name">record_iterator</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="name">rec_one</span> <span class="operator">=</span> <span class="name">record_iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="name">rec_two</span> <span class="operator">=</span> <span class="name">record_iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>We’re going to show two approaches. Firstly, a simple naive
implementation which compares all the window sized sub-sequences to each
other to compiles a similarity matrix. You could construct a matrix or
array object, but here we just use a list of lists of booleans created
with a nested list comprehension:</p>
<pre class="code python literal-block">
<span class="name">window</span> <span class="operator">=</span> <span class="literal number integer">7</span>
<span class="name">seq_one</span> <span class="operator">=</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">rec_one</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">()</span>
<span class="name">seq_two</span> <span class="operator">=</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">rec_two</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">()</span>
<span class="name">data</span> <span class="operator">=</span> <span class="punctuation">[[(</span><span class="name">seq_one</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">:</span><span class="name">i</span><span class="operator">+</span><span class="name">window</span><span class="punctuation">]</span> <span class="operator">&lt;&gt;</span> <span class="name">seq_two</span><span class="punctuation">[</span><span class="name">j</span><span class="punctuation">:</span><span class="name">j</span><span class="operator">+</span><span class="name">window</span><span class="punctuation">])</span> \
        <span class="keyword">for</span> <span class="name">j</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_one</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">window</span><span class="punctuation">)]</span> \
       <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_two</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">window</span><span class="punctuation">)]</span>
</pre>
<p>Note that we have <em>not</em> checked for reverse complement matches here. Now
we’ll use the matplotlib’s <tt class="docutils literal">pylab.imshow()</tt> function to display this
data, first requesting the gray color scheme so this is done in black
and white:</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">pylab</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">gray</span><span class="punctuation">()</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">imshow</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">xlabel</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> (length </span><span class="literal string interpol">%i</span><span class="literal string"> bp)&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">rec_one</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec_one</span><span class="punctuation">)))</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">ylabel</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> (length </span><span class="literal string interpol">%i</span><span class="literal string"> bp)&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">rec_two</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec_two</span><span class="punctuation">)))</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">title</span><span class="punctuation">(</span><span class="literal string">&quot;Dot plot using window size </span><span class="literal string interpol">%i</span><span class="literal string escape">\n</span><span class="literal string">(allowing no mis-matches)&quot;</span> <span class="operator">%</span> <span class="name">window</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">show</span><span class="punctuation">()</span>
</pre>
<p>That should pop up a new window containing a graph like this:</p>
<p><img alt="image28" src="../images/dot_plot.png" /></p>
<p>As you might have expected, these two sequences are very similar with a
partial line of window sized matches along the diagonal. There are no
off diagonal matches which would be indicative of inversions or other
interesting events.</p>
<p>The above code works fine on small examples, but there are two problems
applying this to larger sequences, which we will address below. First
off all, this brute force approach to the all against all comparisons is
very slow. Instead, we’ll compile dictionaries mapping the window sized
sub-sequences to their locations, and then take the set intersection to
find those sub-sequences found in both sequences. This uses more memory,
but is <em>much</em> faster. Secondly, the <tt class="docutils literal">pylab.imshow()</tt> function is
limited in the size of matrix it can display. As an alternative, we’ll
use the <tt class="docutils literal">pylab.scatter()</tt> function.</p>
<p>We start by creating dictionaries mapping the window-sized sub-sequences
to locations:</p>
<pre class="code python literal-block">
<span class="name">window</span> <span class="operator">=</span> <span class="literal number integer">7</span>
<span class="name">dict_one</span> <span class="operator">=</span> <span class="punctuation">{}</span>
<span class="name">dict_two</span> <span class="operator">=</span> <span class="punctuation">{}</span>
<span class="keyword">for</span> <span class="punctuation">(</span><span class="name">seq</span><span class="punctuation">,</span> <span class="name">section_dict</span><span class="punctuation">)</span> <span class="operator word">in</span> <span class="punctuation">[(</span><span class="name builtin">str</span><span class="punctuation">(</span><span class="name">rec_one</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">(),</span> <span class="name">dict_one</span><span class="punctuation">),</span>
                            <span class="punctuation">(</span><span class="name builtin">str</span><span class="punctuation">(</span><span class="name">rec_two</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">(),</span> <span class="name">dict_two</span><span class="punctuation">)]:</span>
    <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">window</span><span class="punctuation">):</span>
        <span class="name">section</span> <span class="operator">=</span> <span class="name">seq</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">:</span><span class="name">i</span><span class="operator">+</span><span class="name">window</span><span class="punctuation">]</span>
        <span class="keyword">try</span><span class="punctuation">:</span>
            <span class="name">section_dict</span><span class="punctuation">[</span><span class="name">section</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">i</span><span class="punctuation">)</span>
        <span class="keyword">except</span> <span class="name exception">KeyError</span><span class="punctuation">:</span>
            <span class="name">section_dict</span><span class="punctuation">[</span><span class="name">section</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">i</span><span class="punctuation">]</span>
<span class="comment">#Now find any sub-sequences found in both sequences</span>
<span class="comment">#(Python 2.3 would require slightly different code here)</span>
<span class="name">matches</span> <span class="operator">=</span> <span class="name builtin">set</span><span class="punctuation">(</span><span class="name">dict_one</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">intersection</span><span class="punctuation">(</span><span class="name">dict_two</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%i</span><span class="literal string"> unique matches&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">matches</span><span class="punctuation">)</span>
</pre>
<p>In order to use the <tt class="docutils literal">pylab.scatter()</tt> we need separate lists for the
<em>x</em> and <em>y</em> co-ordinates:</p>
<pre class="code python literal-block">
<span class="comment">#Create lists of x and y co-ordinates for scatter plot</span>
<span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[]</span>
<span class="name">y</span> <span class="operator">=</span> <span class="punctuation">[]</span>
<span class="keyword">for</span> <span class="name">section</span> <span class="operator word">in</span> <span class="name">matches</span><span class="punctuation">:</span>
    <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name">dict_one</span><span class="punctuation">[</span><span class="name">section</span><span class="punctuation">]:</span>
        <span class="keyword">for</span> <span class="name">j</span> <span class="operator word">in</span> <span class="name">dict_two</span><span class="punctuation">[</span><span class="name">section</span><span class="punctuation">]:</span>
            <span class="name">x</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">i</span><span class="punctuation">)</span>
            <span class="name">y</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">j</span><span class="punctuation">)</span>
</pre>
<p>We are now ready to draw the revised dot plot as a scatter plot:</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">pylab</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">cla</span><span class="punctuation">()</span> <span class="comment">#clear any prior graph</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">gray</span><span class="punctuation">()</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">scatter</span><span class="punctuation">(</span><span class="name">x</span><span class="punctuation">,</span><span class="name">y</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">xlim</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec_one</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">window</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">ylim</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec_two</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">window</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">xlabel</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> (length </span><span class="literal string interpol">%i</span><span class="literal string"> bp)&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">rec_one</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec_one</span><span class="punctuation">)))</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">ylabel</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> (length </span><span class="literal string interpol">%i</span><span class="literal string"> bp)&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">rec_two</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec_two</span><span class="punctuation">)))</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">title</span><span class="punctuation">(</span><span class="literal string">&quot;Dot plot using window size </span><span class="literal string interpol">%i</span><span class="literal string escape">\n</span><span class="literal string">(allowing no mis-matches)&quot;</span> <span class="operator">%</span> <span class="name">window</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">show</span><span class="punctuation">()</span>
</pre>
<p>That should pop up a new window containing a graph like this:</p>
<p><img alt="image29" src="../images/dot_plot_scatter.png" /></p>
<p>Personally I find this second plot much easier to read! Again note that
we have <em>not</em> checked for reverse complement matches here – you could
extend this example to do this, and perhaps plot the forward matches in
one color and the reverse matches in another.</p>
</div>
<div class="section" id="plotting-the-quality-scores-of-sequencing-read-data">
<h3>18.2.4  Plotting the quality scores of sequencing read data</h3>
<p>If you are working with second generation sequencing data, you may want
to try plotting the quality data. Here is an example using two FASTQ
files containing paired end reads, <tt class="docutils literal">SRR001666_1.fastq</tt> for the forward
reads, and <tt class="docutils literal">SRR001666_2.fastq</tt> for the reverse reads. These were
downloaded from the ENA sequence read archive FTP site
(<a class="reference external" href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR001/SRR001666/SRR001666_1.fastq.gz">ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR001/SRR001666/SRR001666_1.fastq.gz</a>
and
<a class="reference external" href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR001/SRR001666/SRR001666_2.fastq.gz">ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR001/SRR001666/SRR001666_2.fastq.gz</a>),
and are from <em>E. coli</em> – see
<a class="reference external" href="http://www.ebi.ac.uk/ena/data/view/SRR001666">http://www.ebi.ac.uk/ena/data/view/SRR001666</a>
for details. In the following code the <tt class="docutils literal"><span class="pre">pylab.subplot(...)</span></tt> function
is used in order to show the forward and reverse qualities on two
subplots, side by side. There is also a little bit of code to only plot
the first fifty reads.</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">pylab</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">for</span> <span class="name">subfigure</span> <span class="operator word">in</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">]:</span>
    <span class="name">filename</span> <span class="operator">=</span> <span class="literal string">&quot;SRR001666_</span><span class="literal string interpol">%i</span><span class="literal string">.fastq&quot;</span> <span class="operator">%</span> <span class="name">subfigure</span>
    <span class="name">pylab</span><span class="operator">.</span><span class="name">subplot</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name">subfigure</span><span class="punctuation">)</span>
    <span class="keyword">for</span> <span class="name">i</span><span class="punctuation">,</span><span class="name">record</span> <span class="operator word">in</span> <span class="name builtin">enumerate</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)):</span>
        <span class="keyword">if</span> <span class="name">i</span> <span class="operator">&gt;=</span> <span class="literal number integer">50</span> <span class="punctuation">:</span> <span class="keyword">break</span> <span class="comment">#trick!</span>
        <span class="name">pylab</span><span class="operator">.</span><span class="name">plot</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">])</span>
    <span class="name">pylab</span><span class="operator">.</span><span class="name">ylim</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="literal number integer">45</span><span class="punctuation">)</span>
    <span class="name">pylab</span><span class="operator">.</span><span class="name">ylabel</span><span class="punctuation">(</span><span class="literal string">&quot;PHRED quality score&quot;</span><span class="punctuation">)</span>
    <span class="name">pylab</span><span class="operator">.</span><span class="name">xlabel</span><span class="punctuation">(</span><span class="literal string">&quot;Position&quot;</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">savefig</span><span class="punctuation">(</span><span class="literal string">&quot;SRR001666.png&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Done&quot;</span>
</pre>
<p>You should note that we are using the <tt class="docutils literal">Bio.SeqIO</tt> format name
<tt class="docutils literal">fastq</tt> here because the NCBI has saved these reads using the standard
Sanger FASTQ format with PHRED scores. However, as you might guess from
the read lengths, this data was from an Illumina Genome Analyzer and was
probably originally in one of the two Solexa/Illumina FASTQ variant file
formats instead.</p>
<p>This example uses the <tt class="docutils literal"><span class="pre">pylab.savefig(...)</span></tt> function instead of
<tt class="docutils literal"><span class="pre">pylab.show(...)</span></tt>, but as mentioned before both are useful. Here is
the result:</p>
<p><img alt="image30" src="../images/SRR001666.png" /></p>
</div>
</div>
<div class="section" id="dealing-with-alignments">
<h2>18.3  Dealing with alignments</h2>
<p>This section can been seen as a follow on to
Chapter <a class="reference external" href="#chapter:Bio.AlignIO">6</a>.</p>
<div class="section" id="calculating-summary-information">
<h3>18.3.1  Calculating summary information</h3>
<p>Once you have an alignment, you are very likely going to want to find
out information about it. Instead of trying to have all of the functions
that can generate information about an alignment in the alignment object
itself, we’ve tried to separate out the functionality into separate
classes, which act on the alignment.</p>
<p>Getting ready to calculate summary information about an object is quick
to do. Let’s say we’ve got an alignment object called <tt class="docutils literal">alignment</tt>, for
example read in using <tt class="docutils literal"><span class="pre">Bio.AlignIO.read(...)</span></tt> as described in
Chapter <a class="reference external" href="#chapter:Bio.AlignIO">6</a>. All we need to do to get an
object that will calculate summary information is:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Align</span> <span class="keyword namespace">import</span> <span class="name">AlignInfo</span>
<span class="name">summary_align</span> <span class="operator">=</span> <span class="name">AlignInfo</span><span class="operator">.</span><span class="name">SummaryInfo</span><span class="punctuation">(</span><span class="name">alignment</span><span class="punctuation">)</span>
</pre>
<p>The <tt class="docutils literal">summary_align</tt> object is very useful, and will do the following
neat things for you:</p>
<ol class="arabic simple">
<li>Calculate a quick consensus sequence – see
section <a class="reference external" href="#sec:consensus">18.3.2</a></li>
<li>Get a position specific score matrix for the alignment – see
section <a class="reference external" href="#sec:pssm">18.3.3</a></li>
<li>Calculate the information content for the alignment – see
section <a class="reference external" href="#sec:getting_info_content">18.3.4</a></li>
<li>Generate information on substitutions in the alignment –
section <a class="reference external" href="#sec:sub_matrix">18.4</a> details using this to generate a
substitution matrix.</li>
</ol>
</div>
<div class="section" id="calculating-a-quick-consensus-sequence">
<h3>18.3.2  Calculating a quick consensus sequence</h3>
<p>The <tt class="docutils literal">SummaryInfo</tt> object, described in
section <a class="reference external" href="#sec:summary_info">18.3.1</a>, provides functionality to
calculate a quick consensus of an alignment. Assuming we’ve got a
<tt class="docutils literal">SummaryInfo</tt> object called <tt class="docutils literal">summary_align</tt> we can calculate a
consensus by doing:</p>
<pre class="code python literal-block">
<span class="name">consensus</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">dumb_consensus</span><span class="punctuation">()</span>
</pre>
<p>As the name suggests, this is a really simple consensus calculator, and
will just add up all of the residues at each point in the consensus, and
if the most common value is higher than some threshold value will add
the common residue to the consensus. If it doesn’t reach the threshold,
it adds an ambiguity character to the consensus. The returned consensus
object is Seq object whose alphabet is inferred from the alphabets of
the sequences making up the consensus. So doing a <tt class="docutils literal">print consensus</tt>
would give:</p>
<pre class="code python literal-block">
<span class="name">consensus</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TATACATNAAAGNAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT</span>
<span class="operator">...</span><span class="literal string">', IUPACAmbiguousDNA())</span>
</pre>
<p>You can adjust how <tt class="docutils literal">dumb_consensus</tt> works by passing optional
parameters:</p>
<dl class="docutils">
<dt><strong>the threshold</strong></dt>
<dd>This is the threshold specifying how common a particular residue has
to be at a position before it is added. The default is 0.7 (meaning
70%).</dd>
<dt><strong>the ambiguous character</strong></dt>
<dd>This is the ambiguity character to use. The default is ’N’.</dd>
<dt><strong>the consensus alphabet</strong></dt>
<dd>This is the alphabet to use for the consensus sequence. If an
alphabet is not specified than we will try to guess the alphabet
based on the alphabets of the sequences in the alignment.</dd>
</dl>
</div>
<div class="section" id="position-specific-score-matrices">
<h3>18.3.3  Position Specific Score Matrices</h3>
<p>Position specific score matrices (PSSMs) summarize the alignment
information in a different way than a consensus, and may be useful for
different tasks. Basically, a PSSM is a count matrix. For each column in
the alignment, the number of each alphabet letters is counted and
totaled. The totals are displayed relative to some representative
sequence along the left axis. This sequence may be the consesus
sequence, but can also be any sequence in the alignment. For instance
for the alignment,</p>
<pre class="code python literal-block">
<span class="name">GTATC</span>
<span class="name">AT</span><span class="operator">--</span><span class="name">C</span>
<span class="name">CTGTC</span>
</pre>
<p>the PSSM is:</p>
<pre class="code python literal-block">
  <span class="name">G</span> <span class="name">A</span> <span class="name">T</span> <span class="name">C</span>
<span class="name">G</span> <span class="literal number integer">1</span> <span class="literal number integer">1</span> <span class="literal number integer">0</span> <span class="literal number integer">1</span>
<span class="name">T</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span> <span class="literal number integer">3</span> <span class="literal number integer">0</span>
<span class="name">A</span> <span class="literal number integer">1</span> <span class="literal number integer">1</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span>
<span class="name">T</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span> <span class="literal number integer">2</span> <span class="literal number integer">0</span>
<span class="name">C</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span> <span class="literal number integer">3</span>
</pre>
<p>Let’s assume we’ve got an alignment object called <tt class="docutils literal">c_align</tt>. To get a
PSSM with the consensus sequence along the side we first get a summary
object and calculate the consensus sequence:</p>
<pre class="code python literal-block">
<span class="name">summary_align</span> <span class="operator">=</span> <span class="name">AlignInfo</span><span class="operator">.</span><span class="name">SummaryInfo</span><span class="punctuation">(</span><span class="name">c_align</span><span class="punctuation">)</span>
<span class="name">consensus</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">dumb_consensus</span><span class="punctuation">()</span>
</pre>
<p>Now, we want to make the PSSM, but ignore any <tt class="docutils literal">N</tt> ambiguity residues
when calculating this:</p>
<pre class="code python literal-block">
<span class="name">my_pssm</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">pos_specific_score_matrix</span><span class="punctuation">(</span><span class="name">consensus</span><span class="punctuation">,</span>
                                                  <span class="name">chars_to_ignore</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">'N'</span><span class="punctuation">])</span>
</pre>
<p>Two notes should be made about this:</p>
<ol class="arabic">
<li><p class="first">To maintain strictness with the alphabets, you can only include
characters along the top of the PSSM that are in the alphabet of the
alignment object. Gaps are not included along the top axis of the
PSSM.</p>
</li>
<li><p class="first">The sequence passed to be displayed along the left side of the axis
does not need to be the consensus. For instance, if you wanted to
display the second sequence in the alignment along this axis, you
would need to do:</p>
<pre class="code python literal-block">
<span class="name">second_seq</span> <span class="operator">=</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">get_seq_by_num</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="name">my_pssm</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">pos_specific_score_matrix</span><span class="punctuation">(</span><span class="name">second_seq</span>
                                                  <span class="name">chars_to_ignore</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">'N'</span><span class="punctuation">])</span>
</pre>
</li>
</ol>
<p>The command above returns a <tt class="docutils literal">PSSM</tt> object. To print out the PSSM as we
showed above, we simply need to do a <tt class="docutils literal">print my_pssm</tt>, which gives:</p>
<pre class="code python literal-block">
    <span class="name">A</span>   <span class="name">C</span>   <span class="name">G</span>   <span class="name">T</span>
<span class="name">T</span>  <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">7.0</span>
<span class="name">A</span>  <span class="literal number float">7.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span>
<span class="name">T</span>  <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">7.0</span>
<span class="name">A</span>  <span class="literal number float">7.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span>
<span class="name">C</span>  <span class="literal number float">0.0</span> <span class="literal number float">7.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span>
<span class="name">A</span>  <span class="literal number float">7.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span>
<span class="name">T</span>  <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">7.0</span>
<span class="name">T</span>  <span class="literal number float">1.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">6.0</span>
<span class="operator">...</span>
</pre>
<p>You can access any element of the PSSM by subscripting like
<tt class="docutils literal"><span class="pre">your_pssm[sequence_number][residue_count_name]</span></tt>. For instance, to get
the counts for the ’A’ residue in the second element of the above PSSM
you would do:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_pssm</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">][</span><span class="literal string">&quot;A&quot;</span><span class="punctuation">]</span>
<span class="literal number float">7.0</span>
</pre>
<p>The structure of the PSSM class hopefully makes it easy both to access
elements and to pretty print the matrix.</p>
</div>
<div class="section" id="information-content">
<h3>18.3.4  Information Content</h3>
<p>A potentially useful measure of evolutionary conservation is the
information content of a sequence.</p>
<p>A useful introduction to information theory targeted towards molecular
biologists can be found at
<a class="reference external" href="http://www.lecb.ncifcrf.gov/~toms/paper/primer/">http://www.lecb.ncifcrf.gov/~toms/paper/primer/</a>.
For our purposes, we will be looking at the information content of a
consesus sequence, or a portion of a consensus sequence. We calculate
information content at a particular column in a multiple sequence
alignment using the following formula:</p>
<div class="formula">
<i>IC</i><sub><i>j</i></sub> = <span class="limits"><sup class="limit"><i>N</i><sub><i>a</i></sub></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><i>P</i><sub><i>ij</i></sub><span class="mathrm">log</span><span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="fraction"><span class="ignored">(</span><span class="numerator"><i>P</i><sub><i>ij</i></sub></span><span class="ignored">)/(</span><span class="denominator"><i>Q</i><sub><i>i</i></sub></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>
</div>
<p>where:</p>
<ul class="simple">
<li><em>IC</em><sub>*j*</sub> – The information content for the <em>j</em>-th column in
an alignment.</li>
<li><em>N</em><sub>*a*</sub> – The number of letters in the alphabet.</li>
<li><em>P</em><sub>*ij*</sub> – The frequency of a particular letter <em>i</em> in the
<em>j</em>-th column (i. e. if G occurred 3 out of 6 times in an aligment
column, this would be 0.5)</li>
<li><em>Q</em><sub>*i*</sub> – The expected frequency of a letter <em>i</em>. This is an
optional argument, usage of which is left at the user’s discretion.
By default, it is automatically assigned to 0.05 = 1/20 for a protein
alphabet, and 0.25 = 1/4 for a nucleic acid alphabet. This is for
geting the information content without any assumption of prior
distributions. When assuming priors, or when using a non-standard
alphabet, you should supply the values for <em>Q</em><sub>*i*</sub>.</li>
</ul>
<p>Well, now that we have an idea what information content is being
calculated in Biopython, let’s look at how to get it for a particular
region of the alignment.</p>
<p>First, we need to use our alignment to get an alignment summary object,
which we’ll assume is called <tt class="docutils literal">summary_align</tt> (see
section <a class="reference external" href="#sec:summary_info">18.3.1</a>) for instructions on how to get
this. Once we’ve got this object, calculating the information content
for a region is as easy as:</p>
<pre class="code python literal-block">
<span class="name">info_content</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">information_content</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="literal number integer">30</span><span class="punctuation">,</span>
                                                 <span class="name">chars_to_ignore</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">'N'</span><span class="punctuation">])</span>
</pre>
<p>Wow, that was much easier then the formula above made it look! The
variable <tt class="docutils literal">info_content</tt> now contains a float value specifying the
information content over the specified region (from 5 to 30 of the
alignment). We specifically ignore the ambiguity residue ’N’ when
calculating the information content, since this value is not included in
our alphabet (so we shouldn’t be interested in looking at it!).</p>
<p>As mentioned above, we can also calculate relative information content
by supplying the expected frequencies:</p>
<pre class="code python literal-block">
<span class="name">expect_freq</span> <span class="operator">=</span> <span class="punctuation">{</span>
    <span class="literal string">'A'</span> <span class="punctuation">:</span> <span class="operator">.</span><span class="literal number integer">3</span><span class="punctuation">,</span>
    <span class="literal string">'G'</span> <span class="punctuation">:</span> <span class="operator">.</span><span class="literal number integer">2</span><span class="punctuation">,</span>
    <span class="literal string">'T'</span> <span class="punctuation">:</span> <span class="operator">.</span><span class="literal number integer">3</span><span class="punctuation">,</span>
    <span class="literal string">'C'</span> <span class="punctuation">:</span> <span class="operator">.</span><span class="literal number integer">2</span><span class="punctuation">}</span>
</pre>
<p>The expected should not be passed as a raw dictionary, but instead by
passed as a <tt class="docutils literal">SubsMat.FreqTable</tt> object (see
section <a class="reference external" href="#sec:freq_table">20.2.2</a> for more information about
FreqTables). The FreqTable object provides a standard for associating
the dictionary with an Alphabet, similar to how the Biopython Seq class
works.</p>
<p>To create a FreqTable object, from the frequency dictionary you just
need to do:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SubsMat</span> <span class="keyword namespace">import</span> <span class="name">FreqTable</span>

<span class="name">e_freq_table</span> <span class="operator">=</span> <span class="name">FreqTable</span><span class="operator">.</span><span class="name">FreqTable</span><span class="punctuation">(</span><span class="name">expect_freq</span><span class="punctuation">,</span> <span class="name">FreqTable</span><span class="operator">.</span><span class="name">FREQ</span><span class="punctuation">,</span>
                                   <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
</pre>
<p>Now that we’ve got that, calculating the relative information content
for our region of the alignment is as simple as:</p>
<pre class="code python literal-block">
<span class="name">info_content</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">information_content</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="literal number integer">30</span><span class="punctuation">,</span>
                                                 <span class="name">e_freq_table</span> <span class="operator">=</span> <span class="name">e_freq_table</span><span class="punctuation">,</span>
                                                 <span class="name">chars_to_ignore</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">'N'</span><span class="punctuation">])</span>
</pre>
<p>Now, <tt class="docutils literal">info_content</tt> will contain the relative information content over
the region in relation to the expected frequencies.</p>
<p>The value return is calculated using base 2 as the logarithm base in the
formula above. You can modify this by passing the parameter <tt class="docutils literal">log_base</tt>
as the base you want:</p>
<pre class="code python literal-block">
<span class="name">info_content</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">information_content</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="literal number integer">30</span><span class="punctuation">,</span> <span class="name">log_base</span> <span class="operator">=</span> <span class="literal number integer">10</span><span class="punctuation">,</span>
                                                 <span class="name">chars_to_ignore</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">'N'</span><span class="punctuation">])</span>
</pre>
<p>Well, now you are ready to calculate information content. If you want to
try applying this to some real life problems, it would probably be best
to dig into the literature on information content to get an idea of how
it is used. Hopefully your digging won’t reveal any mistakes made in
coding this function!</p>
</div>
</div>
<div class="section" id="substitution-matrices">
<h2>18.4  Substitution Matrices</h2>
<p>Substitution matrices are an extremely important part of everyday
bioinformatics work. They provide the scoring terms for classifying how
likely two different residues are to substitute for each other. This is
essential in doing sequence comparisons. The book “Biological Sequence
Analysis” by Durbin et al. provides a really nice introduction to
Substitution Matrices and their uses. Some famous substitution matrices
are the PAM and BLOSUM series of matrices.</p>
<p>Biopython provides a ton of common substitution matrices, and also
provides functionality for creating your own substitution matrices.</p>
<div class="section" id="using-common-substitution-matrices">
<h3>18.4.1  Using common substitution matrices</h3>
</div>
<div class="section" id="creating-your-own-substitution-matrix-from-an-alignment">
<h3>18.4.2  Creating your own substitution matrix from an alignment</h3>
<p>A very cool thing that you can do easily with the substitution matrix
classes is to create your own substitution matrix from an alignment. In
practice, this is normally done with protein alignments. In this
example, we’ll first get a Biopython alignment object and then get a
summary object to calculate info about the alignment. The file
containing <a class="reference external" href="examples/protein.aln">protein.aln</a> (also available online
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/protein.aln">here</a>)
contains the Clustalw alignment output.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Alphabet</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align</span> <span class="keyword namespace">import</span> <span class="name">AlignInfo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">filename</span> <span class="operator">=</span> <span class="literal string">&quot;protein.aln&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">alpha</span> <span class="operator">=</span> <span class="name">Alphabet</span><span class="operator">.</span><span class="name">Gapped</span><span class="punctuation">(</span><span class="name">IUPAC</span><span class="operator">.</span><span class="name">protein</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">c_align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">,</span> <span class="name">alphabet</span><span class="operator">=</span><span class="name">alpha</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">summary_align</span> <span class="operator">=</span> <span class="name">AlignInfo</span><span class="operator">.</span><span class="name">SummaryInfo</span><span class="punctuation">(</span><span class="name">c_align</span><span class="punctuation">)</span>
</pre>
<p>Sections <a class="reference external" href="#sec:align_clustal">6.4.1</a>
and <a class="reference external" href="#sec:summary_info">18.3.1</a> contain more information on doing
this.</p>
<p>Now that we’ve got our <tt class="docutils literal">summary_align</tt> object, we want to use it to
find out the number of times different residues substitute for each
other. To make the example more readable, we’ll focus on only amino
acids with polar charged side chains. Luckily, this can be done easily
when generating a replacement dictionary, by passing in all of the
characters that should be ignored. Thus we’ll create a dictionary of
replacements for only charged polar amino acids using:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">replace_info</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">replacement_dictionary</span><span class="punctuation">([</span><span class="literal string">&quot;G&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;A&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;V&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;L&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;I&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                                      <span class="literal string">&quot;M&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;P&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;F&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;W&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;S&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                                      <span class="literal string">&quot;T&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;N&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;Q&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;Y&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;C&quot;</span><span class="punctuation">])</span>
</pre>
<p>This information about amino acid replacements is represented as a
python dictionary which will look something like (the order can vary):</p>
<pre class="code python literal-block">
<span class="punctuation">{(</span><span class="literal string">'R'</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">):</span> <span class="literal number float">2079.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'R'</span><span class="punctuation">,</span> <span class="literal string">'H'</span><span class="punctuation">):</span> <span class="literal number float">17.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'R'</span><span class="punctuation">,</span> <span class="literal string">'K'</span><span class="punctuation">):</span> <span class="literal number float">103.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'R'</span><span class="punctuation">,</span> <span class="literal string">'E'</span><span class="punctuation">):</span> <span class="literal number float">2.0</span><span class="punctuation">,</span>
<span class="punctuation">(</span><span class="literal string">'R'</span><span class="punctuation">,</span> <span class="literal string">'D'</span><span class="punctuation">):</span> <span class="literal number float">2.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'H'</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">):</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'D'</span><span class="punctuation">,</span> <span class="literal string">'H'</span><span class="punctuation">):</span> <span class="literal number float">15.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'K'</span><span class="punctuation">,</span> <span class="literal string">'K'</span><span class="punctuation">):</span> <span class="literal number float">3218.0</span><span class="punctuation">,</span>
<span class="punctuation">(</span><span class="literal string">'K'</span><span class="punctuation">,</span> <span class="literal string">'H'</span><span class="punctuation">):</span> <span class="literal number float">24.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'H'</span><span class="punctuation">,</span> <span class="literal string">'K'</span><span class="punctuation">):</span> <span class="literal number float">8.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'E'</span><span class="punctuation">,</span> <span class="literal string">'H'</span><span class="punctuation">):</span> <span class="literal number float">15.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'H'</span><span class="punctuation">,</span> <span class="literal string">'H'</span><span class="punctuation">):</span> <span class="literal number float">1235.0</span><span class="punctuation">,</span>
<span class="punctuation">(</span><span class="literal string">'H'</span><span class="punctuation">,</span> <span class="literal string">'E'</span><span class="punctuation">):</span> <span class="literal number float">18.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'H'</span><span class="punctuation">,</span> <span class="literal string">'D'</span><span class="punctuation">):</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'K'</span><span class="punctuation">,</span> <span class="literal string">'D'</span><span class="punctuation">):</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'K'</span><span class="punctuation">,</span> <span class="literal string">'E'</span><span class="punctuation">):</span> <span class="literal number float">9.0</span><span class="punctuation">,</span>
<span class="punctuation">(</span><span class="literal string">'D'</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">):</span> <span class="literal number float">48.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'E'</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">):</span> <span class="literal number float">2.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'D'</span><span class="punctuation">,</span> <span class="literal string">'K'</span><span class="punctuation">):</span> <span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'E'</span><span class="punctuation">,</span> <span class="literal string">'K'</span><span class="punctuation">):</span> <span class="literal number float">45.0</span><span class="punctuation">,</span>
<span class="punctuation">(</span><span class="literal string">'K'</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">):</span> <span class="literal number float">130.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'E'</span><span class="punctuation">,</span> <span class="literal string">'D'</span><span class="punctuation">):</span> <span class="literal number float">241.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'E'</span><span class="punctuation">,</span> <span class="literal string">'E'</span><span class="punctuation">):</span> <span class="literal number float">3305.0</span><span class="punctuation">,</span>
<span class="punctuation">(</span><span class="literal string">'D'</span><span class="punctuation">,</span> <span class="literal string">'E'</span><span class="punctuation">):</span> <span class="literal number float">270.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'D'</span><span class="punctuation">,</span> <span class="literal string">'D'</span><span class="punctuation">):</span> <span class="literal number float">2360.0</span><span class="punctuation">}</span>
</pre>
<p>This information gives us our accepted number of replacements, or how
often we expect different things to substitute for each other. It turns
out, amazingly enough, that this is all of the information we need to go
ahead and create a substitution matrix. First, we use the replacement
dictionary information to create an Accepted Replacement Matrix (ARM):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SubsMat</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_arm</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">SeqMat</span><span class="punctuation">(</span><span class="name">replace_info</span><span class="punctuation">)</span>
</pre>
<p>With this accepted replacement matrix, we can go right ahead and create
our log odds matrix (i. e. a standard type Substitution Matrix):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_lom</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">make_log_odds_matrix</span><span class="punctuation">(</span><span class="name">my_arm</span><span class="punctuation">)</span>
</pre>
<p>The log odds matrix you create is customizable with the following
optional arguments:</p>
<ul class="simple">
<li><tt class="docutils literal">exp_freq_table</tt> – You can pass a table of expected frequencies for
each alphabet. If supplied, this will be used instead of the passed
accepted replacement matrix when calculate expected replacments.</li>
<li><tt class="docutils literal">logbase</tt> - The base of the logarithm taken to create the log odd
matrix. Defaults to base 10.</li>
<li><tt class="docutils literal">factor</tt> - The factor to multiply each matrix entry by. This
defaults to 10, which normally makes the matrix numbers easy to work
with.</li>
<li><tt class="docutils literal">round_digit</tt> - The digit to round to in the matrix. This defaults
to 0 (i. e. no digits).</li>
</ul>
<p>Once you’ve got your log odds matrix, you can display it prettily using
the function <tt class="docutils literal">print_mat</tt>. Doing this on our created matrix gives:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_lom</span><span class="operator">.</span><span class="name">print_mat</span><span class="punctuation">()</span>
<span class="name">D</span>   <span class="literal number integer">2</span>
<span class="name">E</span>  <span class="operator">-</span><span class="literal number integer">1</span>   <span class="literal number integer">1</span>
<span class="name">H</span>  <span class="operator">-</span><span class="literal number integer">5</span>  <span class="operator">-</span><span class="literal number integer">4</span>   <span class="literal number integer">3</span>
<span class="name">K</span> <span class="operator">-</span><span class="literal number integer">10</span>  <span class="operator">-</span><span class="literal number integer">5</span>  <span class="operator">-</span><span class="literal number integer">4</span>   <span class="literal number integer">1</span>
<span class="name">R</span>  <span class="operator">-</span><span class="literal number integer">4</span>  <span class="operator">-</span><span class="literal number integer">8</span>  <span class="operator">-</span><span class="literal number integer">4</span>  <span class="operator">-</span><span class="literal number integer">2</span>   <span class="literal number integer">2</span>
   <span class="name">D</span>   <span class="name">E</span>   <span class="name">H</span>   <span class="name">K</span>   <span class="name">R</span>
</pre>
<p>Very nice. Now we’ve got our very own substitution matrix to play with!</p>
</div>
</div>
<div class="section" id="biosql-storing-sequences-in-a-relational-database">
<h2>18.5  BioSQL – storing sequences in a relational database</h2>
<p><a class="reference external" href="http://www.biosql.org/">BioSQL</a> is a joint effort between the
<a class="reference external" href="http://open-bio.org/">OBF</a> projects (BioPerl, BioJava etc) to
support a shared database schema for storing sequence data. In theory,
you could load a GenBank file into the database with BioPerl, then using
Biopython extract this from the database as a record object with
features - and get more or less the same thing as if you had loaded the
GenBank file directly as a SeqRecord using <tt class="docutils literal">Bio.SeqIO</tt>
(Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>).</p>
<p>Biopython’s BioSQL module is currently documented at
<a class="reference external" href="http://biopython.org/wiki/BioSQL">http://biopython.org/wiki/BioSQL</a>
which is part of our wiki pages.</p>
</div>
</div>
<div class="section" id="chapter-19-the-biopython-testing-framework">
<h1>Chapter 19  The Biopython testing framework</h1>
<p>Biopython has a regression testing framework (the file <tt class="docutils literal">run_tests.py</tt>)
based on <a class="reference external" href="http://docs.python.org/library/unittest.html">unittest</a>,
the standard unit testing framework for Python. Providing comprehensive
tests for modules is one of the most important aspects of making sure
that the Biopython code is as bug-free as possible before going out. It
also tends to be one of the most undervalued aspects of contributing.
This chapter is designed to make running the Biopython tests and writing
good test code as easy as possible. Ideally, every module that goes into
Biopython should have a test (and should also have documentation!). All
our developers, and anyone installing Biopython from source, are
strongly encouraged to run the unit tests.</p>
<div class="section" id="running-the-tests">
<h2>19.1  Running the tests</h2>
<p>When you download the Biopython source code, or check it out from our
source code repository, you should find a subdirectory call <tt class="docutils literal">Tests</tt>.
This contains the key script <tt class="docutils literal">run_tests.py</tt>, lots of individual
scripts named <tt class="docutils literal">test_XXX.py</tt>, a subdirectory called <tt class="docutils literal">output</tt> and lots
of other subdirectories which contain input files for the test suite.</p>
<p>As part of building and installing Biopython you will typically run the
full test suite at the command line from the Biopython source top level
directory using the following:</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">setup</span><span class="operator">.</span><span class="name">py</span> <span class="name">test</span>
</pre>
<p>This is actually equivalent to going to the <tt class="docutils literal">Tests</tt> subdirectory and
running:</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">run_tests</span><span class="operator">.</span><span class="name">py</span>
</pre>
<p>You’ll often want to run just some of the tests, and this is done like
this:</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">run_tests</span><span class="operator">.</span><span class="name">py</span> <span class="name">test_SeqIO</span><span class="operator">.</span><span class="name">py</span> <span class="name">test_AlignIO</span><span class="operator">.</span><span class="name">py</span>
</pre>
<p>When giving the list of tests, the <tt class="docutils literal">.py</tt> extension is optional, so you
can also just type:</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">run_tests</span><span class="operator">.</span><span class="name">py</span> <span class="name">test_SeqIO</span> <span class="name">test_AlignIO</span>
</pre>
<p>To run the docstring tests (see section <a class="reference external" href="#section:doctest">19.3</a>),
you can use</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">run_tests</span><span class="operator">.</span><span class="name">py</span> <span class="name">doctest</span>
</pre>
<p>By default, <tt class="docutils literal">run_tests.py</tt> runs all tests, including the docstring
tests.</p>
<p>If an individual test is failing, you can also try running it directly,
which may give you more information.</p>
<p>Importantly, note that the individual unit tests come in two types:</p>
<ul class="simple">
<li>Simple print-and-compare scripts. These unit tests are essentially
short example Python programs, which print out various output text.
For a test file named <tt class="docutils literal">test_XXX.py</tt> there will be a matching text
file called <tt class="docutils literal">test_XXX</tt> under the <tt class="docutils literal">output</tt> subdirectory which
contains the expected output. All that the test framework does to is
run the script, and check the output agrees.</li>
<li>Standard <tt class="docutils literal">unittest</tt>- based tests. These will <tt class="docutils literal">import unittest</tt>
and then define <tt class="docutils literal">unittest.TestCase</tt> classes, each with one or more
sub-tests as methods starting with <tt class="docutils literal">test_</tt> which check some
specific aspect of the code. These tests should not print any output
directly.</li>
</ul>
<p>Currently, about half of the Biopython tests are <tt class="docutils literal">unittest</tt>-style
tests, and half are print-and-compare tests.</p>
<p>Running a simple print-and-compare test directly will usually give lots
of output on screen, but does not check the output matches the expected
output. If the test is failing with an exception error, it should be
very easy to locate where exactly the script is failing. For an example
of a print-and-compare test, try:</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">test_SeqIO</span><span class="operator">.</span><span class="name">py</span>
</pre>
<p>The <tt class="docutils literal">unittest</tt>-based tests instead show you exactly which
sub-section(s) of the test are failing. For example,</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">test_Cluster</span><span class="operator">.</span><span class="name">py</span>
</pre>
</div>
<div class="section" id="writing-tests">
<h2>19.2  Writing tests</h2>
<p>Let’s say you want to write some tests for a module called <tt class="docutils literal">Biospam</tt>.
This can be a module you wrote, or an existing module that doesn’t have
any tests yet. In the examples below, we assume that <tt class="docutils literal">Biospam</tt> is a
module that does simple math.</p>
<p>Each Biopython test can have three important files and directories
involved with it:</p>
<ol class="arabic simple">
<li><tt class="docutils literal">test_Biospam.py</tt> – The actual test code for your module.</li>
<li><tt class="docutils literal">Biospam</tt> [optional]– A directory where any necessary input files
will be located. Any output files that will be generated should also
be written here (and preferably cleaned up after the tests are done)
to prevent clogging up the main Tests directory.</li>
<li><tt class="docutils literal">output/Biospam</tt> – [for print-and-compare tests only] This file
contains the expected output from running <tt class="docutils literal">test_Biospam.py</tt>. This
file is not needed for <tt class="docutils literal">unittest</tt>-style tests, since there the
validation is done in the test script <tt class="docutils literal">test_Biospam.py</tt> itself.</li>
</ol>
<p>It’s up to you to decide whether you want to write a print-and-compare
test script or a <tt class="docutils literal">unittest</tt>-style test script. The important thing is
that you cannot mix these two styles in a single test script.
Particularly, don’t use <tt class="docutils literal">unittest</tt> features in a print-and-compare
test.</p>
<p>Any script with a <tt class="docutils literal">test_</tt> prefix in the <tt class="docutils literal">Tests</tt> directory will be
found and run by <tt class="docutils literal">run_tests.py</tt>. Below, we show an example test script
<tt class="docutils literal">test_Biospam.py</tt> both for a print-and-compare test and for a
<tt class="docutils literal">unittest</tt>-based test. If you put this script in the Biopython
<tt class="docutils literal">Tests</tt> directory, then <tt class="docutils literal">run_tests.py</tt> will find it and execute the
tests contained in it:</p>
<pre class="code python literal-block">
<span class="error">$</span> <span class="name">python</span> <span class="name">run_tests</span><span class="operator">.</span><span class="name">py</span>
<span class="name">test_Ace</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_AlignIO</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_BioSQL</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_BioSQL_SeqIO</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_Biospam</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_CAPS</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_Clustalw</span> <span class="operator">...</span> <span class="name">ok</span>
</pre>
<p>…</p>
<pre class="code python literal-block">
<span class="operator">----------------------------------------------------------------------</span>
<span class="name">Ran</span> <span class="literal number integer">107</span> <span class="name">tests</span> <span class="operator word">in</span> <span class="literal number float">86.127</span> <span class="name">seconds</span>
</pre>
<div class="section" id="writing-a-print-and-compare-test">
<h3>19.2.1  Writing a print-and-compare test</h3>
<p>A print-and-compare style test should be much simpler for beginners or
novices to write - essentially it is just an example script using your
new module.</p>
<p>Here is what you should do to make a print-and-compare test for the
<tt class="docutils literal">Biospam</tt> module.</p>
<ol class="arabic">
<li><p class="first">Write a script called <tt class="docutils literal">test_Biospam.py</tt></p>
<ul class="simple">
<li>This script should live in the Tests directory</li>
<li>The script should test all of the important functionality of the
module (the more you test the better your test is, of course!).</li>
<li>Try to avoid anything which might be platform specific, such as
printing floating point numbers without using an explicit
formatting string to avoid having too many decimal places
(different platforms can give very slightly different values).</li>
</ul>
</li>
<li><p class="first">If the script requires files to do the testing, these should go in
the directory Tests/Biospam (if you just need something generic, like
a FASTA sequence file, or a GenBank record, try and use an existing
sample input file instead).</p>
</li>
<li><p class="first">Write out the test output and verify the output to be correct.</p>
<p>There are two ways to do this:</p>
<ol class="arabic simple">
<li>The long way:<ul>
<li>Run the script and write its output to a file. On UNIX
(including Linux and Mac OS X) machines, you would do something
like: <tt class="docutils literal">python test_Biospam.py &gt; test_Biospam</tt> which would
write the output to the file <tt class="docutils literal">test_Biospam</tt>.</li>
<li>Manually look at the file <tt class="docutils literal">test_Biospam</tt> to make sure the
output is correct. When you are sure it is all right and there
are no bugs, you need to quickly edit the <tt class="docutils literal">test_Biospam</tt> file
so that the first line is: ‘<tt class="docutils literal">test_Biospam</tt>’ (no quotes).</li>
<li>copy the <tt class="docutils literal">test_Biospam</tt> file to the directory Tests/output</li>
</ul>
</li>
<li>The quick way:<ul>
<li>Run <tt class="docutils literal">python run_tests.py <span class="pre">-g</span> test_Biospam.py</tt>. The regression
testing framework is nifty enough that it’ll put the output in
the right place in just the way it likes it.</li>
<li>Go to the output (which should be in
<tt class="docutils literal">Tests/output/test_Biospam</tt>) and double check the output to
make sure it is all correct.</li>
</ul>
</li>
</ol>
</li>
<li><p class="first">Now change to the Tests directory and run the regression tests with
<tt class="docutils literal">python run_tests.py</tt>. This will run all of the tests, and you
should see your test run (and pass!).</p>
</li>
<li><p class="first">That’s it! Now you’ve got a nice test for your module ready to check
in, or submit to Biopython. Congratulations!</p>
</li>
</ol>
<p>As an example, the <tt class="docutils literal">test_Biospam.py</tt> test script to test the
<tt class="docutils literal">addition</tt> and <tt class="docutils literal">multiplication</tt> functions in the <tt class="docutils literal">Biospam</tt> module
could look as follows:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Biospam</span>

<span class="keyword">print</span> <span class="literal string">&quot;2 + 3 =&quot;</span><span class="punctuation">,</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">addition</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;9 - 1 =&quot;</span><span class="punctuation">,</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">addition</span><span class="punctuation">(</span><span class="literal number integer">9</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;2 * 3 =&quot;</span><span class="punctuation">,</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">multiplication</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;9 * (- 1) =&quot;</span><span class="punctuation">,</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">multiplication</span><span class="punctuation">(</span><span class="literal number integer">9</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>We generate the corresponding output with
<tt class="docutils literal">python run_tests.py <span class="pre">-g</span> test_Biospam.py</tt>, and check the output file
<tt class="docutils literal">output/test_Biospam</tt>:</p>
<pre class="code python literal-block">
<span class="name">test_Biospam</span>
<span class="literal number integer">2</span> <span class="operator">+</span> <span class="literal number integer">3</span> <span class="operator">=</span> <span class="literal number integer">5</span>
<span class="literal number integer">9</span> <span class="operator">-</span> <span class="literal number integer">1</span> <span class="operator">=</span> <span class="literal number integer">8</span>
<span class="literal number integer">2</span> <span class="operator">*</span> <span class="literal number integer">3</span> <span class="operator">=</span> <span class="literal number integer">6</span>
<span class="literal number integer">9</span> <span class="operator">*</span> <span class="punctuation">(</span><span class="operator">-</span> <span class="literal number integer">1</span><span class="punctuation">)</span> <span class="operator">=</span> <span class="operator">-</span><span class="literal number integer">9</span>
</pre>
<p>Often, the difficulty with larger print-and-compare tests is to keep
track which line in the output corresponds to which command in the test
script. For this purpose, it is important to print out some markers to
help you match lines in the input script with the generated output.</p>
</div>
<div class="section" id="writing-a-unittest-based-test">
<h3>19.2.2  Writing a unittest-based test</h3>
<p>We want all the modules in Biopython to have unit tests, and a simple
print-and-compare test is better than no test at all. However, although
there is a steeper learning curve, using the <tt class="docutils literal">unittest</tt> framework
gives a more structured result, and if there is a test failure this can
clearly pinpoint which part of the test is going wrong. The sub-tests
can also be run individually which is helpful for testing or debugging.</p>
<p>The <tt class="docutils literal">unittest</tt>-framework has been included with Python since version
2.1, and is documented in the Python Library Reference (which I know you
are keeping under your pillow, as recommended). There is also <a class="reference external" href="http://docs.python.org/library/unittest.html">online
documentaion for
unittest</a>. If you are
familiar with the <tt class="docutils literal">unittest</tt> system (or something similar like the
nose test framework), you shouldn’t have any trouble. You may find
looking at the existing example within Biopython helpful too.</p>
<p>Here’s a minimal <tt class="docutils literal">unittest</tt>-style test script for <tt class="docutils literal">Biospam</tt>, which
you can copy and paste to get started:</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">unittest</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Biospam</span>

<span class="keyword">class</span> <span class="name class">BiospamTestAddition</span><span class="punctuation">(</span><span class="name">unittest</span><span class="operator">.</span><span class="name">TestCase</span><span class="punctuation">):</span>

    <span class="keyword">def</span> <span class="name function">test_addition1</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">addition</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="literal number integer">5</span><span class="punctuation">)</span>

    <span class="keyword">def</span> <span class="name function">test_addition2</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">addition</span><span class="punctuation">(</span><span class="literal number integer">9</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="literal number integer">8</span><span class="punctuation">)</span>

<span class="keyword">class</span> <span class="name class">BiospamTestDivision</span><span class="punctuation">(</span><span class="name">unittest</span><span class="operator">.</span><span class="name">TestCase</span><span class="punctuation">):</span>

    <span class="keyword">def</span> <span class="name function">test_division1</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">division</span><span class="punctuation">(</span><span class="literal number float">3.0</span><span class="punctuation">,</span> <span class="literal number float">2.0</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertAlmostEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="literal number float">1.5</span><span class="punctuation">)</span>

    <span class="keyword">def</span> <span class="name function">test_division2</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">division</span><span class="punctuation">(</span><span class="literal number float">10.0</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">2.0</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertAlmostEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">5.0</span><span class="punctuation">)</span>


<span class="keyword">if</span> <span class="name">__name__</span> <span class="operator">==</span> <span class="literal string">&quot;__main__&quot;</span><span class="punctuation">:</span>
    <span class="name">runner</span> <span class="operator">=</span> <span class="name">unittest</span><span class="operator">.</span><span class="name">TextTestRunner</span><span class="punctuation">(</span><span class="name">verbosity</span> <span class="operator">=</span> <span class="literal number integer">2</span><span class="punctuation">)</span>
    <span class="name">unittest</span><span class="operator">.</span><span class="name">main</span><span class="punctuation">(</span><span class="name">testRunner</span><span class="operator">=</span><span class="name">runner</span><span class="punctuation">)</span>
</pre>
<p>In the division tests, we use <tt class="docutils literal">assertAlmostEqual</tt> instead of
<tt class="docutils literal">assertEqual</tt> to avoid tests failing due to roundoff errors; see the
<tt class="docutils literal">unittest</tt> chapter in the Python documentation for details and for
other functionality available in <tt class="docutils literal">unittest</tt> (<a class="reference external" href="http://docs.python.org/library/unittest.html">online
reference</a>).</p>
<p>These are the key points of <tt class="docutils literal">unittest</tt>-based tests:</p>
<ul>
<li><p class="first">Test cases are stored in classes that derive from
<tt class="docutils literal">unittest.TestCase</tt> and cover one basic aspect of your code</p>
</li>
<li><p class="first">You can use methods <tt class="docutils literal">setUp</tt> and <tt class="docutils literal">tearDown</tt> for any repeated code
which should be run before and after each test method. For example,
the <tt class="docutils literal">setUp</tt> method might be used to create an instance of the
object you are testing, or open a file handle. The <tt class="docutils literal">tearDown</tt>
should do any “tidying up”, for example closing the file handle.</p>
</li>
<li><p class="first">The tests are prefixed with <tt class="docutils literal">test_</tt> and each test should cover one
specific part of what you are trying to test. You can have as many
tests as you want in a class.</p>
</li>
<li><p class="first">At the end of the test script, you can use</p>
<pre class="code python literal-block">
<span class="keyword">if</span> <span class="name">__name__</span> <span class="operator">==</span> <span class="literal string">&quot;__main__&quot;</span><span class="punctuation">:</span>
    <span class="name">runner</span> <span class="operator">=</span> <span class="name">unittest</span><span class="operator">.</span><span class="name">TextTestRunner</span><span class="punctuation">(</span><span class="name">verbosity</span> <span class="operator">=</span> <span class="literal number integer">2</span><span class="punctuation">)</span>
    <span class="name">unittest</span><span class="operator">.</span><span class="name">main</span><span class="punctuation">(</span><span class="name">testRunner</span><span class="operator">=</span><span class="name">runner</span><span class="punctuation">)</span>
</pre>
<p>to execute the tests when the script is run by itself (rather than
imported from <tt class="docutils literal">run_tests.py</tt>). If you run this script, then you’ll
see something like the following:</p>
<pre class="code python literal-block">
<span class="error">$</span> <span class="name">python</span> <span class="name">test_BiospamMyModule</span><span class="operator">.</span><span class="name">py</span>
<span class="name">test_addition1</span> <span class="punctuation">(</span><span class="name">__main__</span><span class="operator">.</span><span class="name">TestAddition</span><span class="punctuation">)</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_addition2</span> <span class="punctuation">(</span><span class="name">__main__</span><span class="operator">.</span><span class="name">TestAddition</span><span class="punctuation">)</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_division1</span> <span class="punctuation">(</span><span class="name">__main__</span><span class="operator">.</span><span class="name">TestDivision</span><span class="punctuation">)</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_division2</span> <span class="punctuation">(</span><span class="name">__main__</span><span class="operator">.</span><span class="name">TestDivision</span><span class="punctuation">)</span> <span class="operator">...</span> <span class="name">ok</span>

<span class="operator">----------------------------------------------------------------------</span>
<span class="name">Ran</span> <span class="literal number integer">4</span> <span class="name">tests</span> <span class="operator word">in</span> <span class="literal number float">0.059</span><span class="name">s</span>

<span class="name">OK</span>
</pre>
</li>
<li><p class="first">To indicate more clearly what each test is doing, you can add
docstrings to each test. These are shown when running the tests,
which can be useful information if a test is failing.</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">unittest</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Biospam</span>

<span class="keyword">class</span> <span class="name class">BiospamTestAddition</span><span class="punctuation">(</span><span class="name">unittest</span><span class="operator">.</span><span class="name">TestCase</span><span class="punctuation">):</span>

    <span class="keyword">def</span> <span class="name function">test_addition1</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="literal string doc">&quot;&quot;&quot;An addition test&quot;&quot;&quot;</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">addition</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="literal number integer">5</span><span class="punctuation">)</span>

    <span class="keyword">def</span> <span class="name function">test_addition2</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="literal string doc">&quot;&quot;&quot;A second addition test&quot;&quot;&quot;</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">addition</span><span class="punctuation">(</span><span class="literal number integer">9</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="literal number integer">8</span><span class="punctuation">)</span>

<span class="keyword">class</span> <span class="name class">BiospamTestDivision</span><span class="punctuation">(</span><span class="name">unittest</span><span class="operator">.</span><span class="name">TestCase</span><span class="punctuation">):</span>

    <span class="keyword">def</span> <span class="name function">test_division1</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="literal string doc">&quot;&quot;&quot;Now let's check division&quot;&quot;&quot;</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">division</span><span class="punctuation">(</span><span class="literal number float">3.0</span><span class="punctuation">,</span> <span class="literal number float">2.0</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertAlmostEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="literal number float">1.5</span><span class="punctuation">)</span>

    <span class="keyword">def</span> <span class="name function">test_division2</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="literal string doc">&quot;&quot;&quot;A second division test&quot;&quot;&quot;</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">division</span><span class="punctuation">(</span><span class="literal number float">10.0</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">2.0</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertAlmostEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">5.0</span><span class="punctuation">)</span>


<span class="keyword">if</span> <span class="name">__name__</span> <span class="operator">==</span> <span class="literal string">&quot;__main__&quot;</span><span class="punctuation">:</span>
    <span class="name">runner</span> <span class="operator">=</span> <span class="name">unittest</span><span class="operator">.</span><span class="name">TextTestRunner</span><span class="punctuation">(</span><span class="name">verbosity</span> <span class="operator">=</span> <span class="literal number integer">2</span><span class="punctuation">)</span>
    <span class="name">unittest</span><span class="operator">.</span><span class="name">main</span><span class="punctuation">(</span><span class="name">testRunner</span><span class="operator">=</span><span class="name">runner</span><span class="punctuation">)</span>
</pre>
<p>Running the script will now show you:</p>
<pre class="code python literal-block">
<span class="error">$</span> <span class="name">python</span> <span class="name">test_BiospamMyModule</span><span class="operator">.</span><span class="name">py</span>
<span class="name">An</span> <span class="name">addition</span> <span class="name">test</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">A</span> <span class="name">second</span> <span class="name">addition</span> <span class="name">test</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">Now</span> <span class="name">let</span><span class="literal string">'s check division ... ok</span>
<span class="name">A</span> <span class="name">second</span> <span class="name">division</span> <span class="name">test</span> <span class="operator">...</span> <span class="name">ok</span>

<span class="operator">----------------------------------------------------------------------</span>
<span class="name">Ran</span> <span class="literal number integer">4</span> <span class="name">tests</span> <span class="operator word">in</span> <span class="literal number float">0.001</span><span class="name">s</span>

<span class="name">OK</span>
</pre>
</li>
</ul>
<p>If your module contains docstring tests (see section
<a class="reference external" href="#section:doctest">19.3</a>), you may want to include those in the tests
to be run. You can do so as follows by modifying the code under
<tt class="docutils literal">if __name__ == &quot;__main__&quot;:</tt> to look like this:</p>
<pre class="code python literal-block">
<span class="keyword">if</span> <span class="name">__name__</span> <span class="operator">==</span> <span class="literal string">&quot;__main__&quot;</span><span class="punctuation">:</span>
    <span class="name">unittest_suite</span> <span class="operator">=</span> <span class="name">unittest</span><span class="operator">.</span><span class="name">TestLoader</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">loadTestsFromName</span><span class="punctuation">(</span><span class="literal string">&quot;test_Biospam&quot;</span><span class="punctuation">)</span>
    <span class="name">doctest_suite</span> <span class="operator">=</span> <span class="name">doctest</span><span class="operator">.</span><span class="name">DocTestSuite</span><span class="punctuation">(</span><span class="name">Biospam</span><span class="punctuation">)</span>
    <span class="name">suite</span> <span class="operator">=</span> <span class="name">unittest</span><span class="operator">.</span><span class="name">TestSuite</span><span class="punctuation">((</span><span class="name">unittest_suite</span><span class="punctuation">,</span> <span class="name">doctest_suite</span><span class="punctuation">))</span>
    <span class="name">runner</span> <span class="operator">=</span> <span class="name">unittest</span><span class="operator">.</span><span class="name">TextTestRunner</span><span class="punctuation">(</span><span class="name">sys</span><span class="operator">.</span><span class="name">stdout</span><span class="punctuation">,</span> <span class="name">verbosity</span> <span class="operator">=</span> <span class="literal number integer">2</span><span class="punctuation">)</span>
    <span class="name">runner</span><span class="operator">.</span><span class="name">run</span><span class="punctuation">(</span><span class="name">suite</span><span class="punctuation">)</span>
</pre>
<p>This is only relevant if you want to run the docstring tests when you
execute <tt class="docutils literal">python test_Biospam.py</tt>; with <tt class="docutils literal">python run_tests.py</tt>, the
docstring tests are run automatically (assuming they are included in the
list of docstring tests in <tt class="docutils literal">run_tests.py</tt>, see the section below).</p>
</div>
</div>
<div class="section" id="writing-doctests">
<h2>19.3  Writing doctests</h2>
<p>Python modules, classes and functions support built in documentation
using docstrings. The <a class="reference external" href="http://docs.python.org/library/doctest.html">doctest
framework</a> (included
with Python) allows the developer to embed working examples in the
docstrings, and have these examples automatically tested.</p>
<p>Currently only a small part of Biopython includes doctests. The
<tt class="docutils literal">run_tests.py</tt> script takes care of running the doctests. For this
purpose, at the top of the <tt class="docutils literal">run_tests.py</tt> script is a manually
compiled list of modules to test, which allows us to skip modules with
optional external dependencies which may not be installed (e.g. the
Reportlab and NumPy libraries). So, if you’ve added some doctests to the
docstrings in a Biopython module, in order to have them included in the
Biopython test suite, you must update <tt class="docutils literal">run_tests.py</tt> to include your
module. Currently, the relevant part of <tt class="docutils literal">run_tests.py</tt> looks as
follows:</p>
<pre class="code python literal-block">
<span class="comment"># This is the list of modules containing docstring tests.</span>
<span class="comment"># If you develop docstring tests for other modules, please add</span>
<span class="comment"># those modules here.</span>
<span class="name">DOCTEST_MODULES</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">&quot;Bio.Seq&quot;</span><span class="punctuation">,</span>
                   <span class="literal string">&quot;Bio.SeqRecord&quot;</span><span class="punctuation">,</span>
                   <span class="literal string">&quot;Bio.SeqIO&quot;</span><span class="punctuation">,</span>
                   <span class="literal string">&quot;...&quot;</span><span class="punctuation">,</span>
                  <span class="punctuation">]</span>
<span class="comment">#Silently ignore any doctests for modules requiring numpy!</span>
<span class="keyword">try</span><span class="punctuation">:</span>
    <span class="keyword namespace">import</span> <span class="name namespace">numpy</span>
    <span class="name">DOCTEST_MODULES</span><span class="operator">.</span><span class="name">extend</span><span class="punctuation">([</span><span class="literal string">&quot;Bio.Statistics.lowess&quot;</span><span class="punctuation">])</span>
<span class="keyword">except</span> <span class="name exception">ImportError</span><span class="punctuation">:</span>
    <span class="keyword">pass</span>
</pre>
<p>Note that we regard doctests primarily as documentation, so you should
stick to typical usage. Generally complicated examples dealing with
error conditions and the like would be best left to a dedicated unit
test.</p>
<p>Note that if you want to write doctests involving file parsing, defining
the file location complicates matters. Ideally use relative paths
assuming the code will be run from the <tt class="docutils literal">Tests</tt> directory, see the
<tt class="docutils literal">Bio.SeqIO</tt> doctests for an example of this.</p>
<p>To run the docstring tests only, use</p>
<pre class="code python literal-block">
<span class="error">$</span> <span class="name">python</span> <span class="name">run_tests</span><span class="operator">.</span><span class="name">py</span> <span class="name">doctest</span>
</pre>
</div>
</div>
<div class="section" id="chapter-20-advanced">
<h1>Chapter 20  Advanced</h1>
<div class="section" id="parser-design">
<h2>20.1  Parser Design</h2>
<p>Many of the older Biopython parsers were built around an event-oriented
design that includes Scanner and Consumer objects.</p>
<p>Scanners take input from a data source and analyze it line by line,
sending off an event whenever it recognizes some information in the
data. For example, if the data includes information about an organism
name, the scanner may generate an <tt class="docutils literal">organism_name</tt> event whenever it
encounters a line containing the name.</p>
<p>Consumers are objects that receive the events generated by Scanners.
Following the previous example, the consumer receives the
<tt class="docutils literal">organism_name</tt> event, and the processes it in whatever manner
necessary in the current application.</p>
<p>This is a very flexible framework, which is advantageous if you want to
be able to parse a file format into more than one representation. For
example, the <tt class="docutils literal">Bio.GenBank</tt> module uses this to construct either
<tt class="docutils literal">SeqRecord</tt> objects or file-format-specific record objects.</p>
<p>More recently, many of the parsers added for <tt class="docutils literal">Bio.SeqIO</tt> and
<tt class="docutils literal">Bio.AlignIO</tt> take a much simpler approach, but only generate a single
object representation (<tt class="docutils literal">SeqRecord</tt> and <tt class="docutils literal">MultipleSeqAlignment</tt>
objects respectively). In some cases the <tt class="docutils literal">Bio.SeqIO</tt> parsers actually
wrap another Biopython parser - for example, the <tt class="docutils literal">Bio.SwissProt</tt>
parser produces SwissProt format specific record objects, which get
converted into <tt class="docutils literal">SeqRecord</tt> objects.</p>
</div>
<div class="section" id="id19">
<h2>20.2  Substitution Matrices</h2>
<div class="section" id="subsmat">
<h3>20.2.1  SubsMat</h3>
<p>This module provides a class and a few routines for generating
substitution matrices, similar to BLOSUM or PAM matrices, but based on
user-provided data. Additionally, you may select a matrix from
MatrixInfo.py, a collection of established substitution matrices. The
<tt class="docutils literal">SeqMat</tt> class derives from a dictionary:</p>
<pre class="code python literal-block">
<span class="keyword">class</span> <span class="name class">SeqMat</span><span class="punctuation">(</span><span class="name builtin">dict</span><span class="punctuation">)</span>
</pre>
<p>The dictionary is of the form
<tt class="docutils literal"><span class="pre">{(i1,j1):n1,</span> <span class="pre">(i1,j2):n2,...,(ik,jk):nk}</span></tt> where i, j are alphabet
letters, and n is a value.</p>
<ol class="arabic">
<li><p class="first">Attributes</p>
<ol class="arabic simple">
<li><tt class="docutils literal">self.alphabet</tt>: a class as defined in Bio.Alphabet</li>
<li><tt class="docutils literal">self.ab_list</tt>: a list of the alphabet’s letters, sorted. Needed
mainly for internal purposes</li>
</ol>
</li>
<li><p class="first">Methods</p>
<ol class="arabic">
<li><pre class="code python first literal-block">
<span class="name">__init__</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">,</span><span class="name">data</span><span class="operator">=</span><span class="name builtin pseudo">None</span><span class="punctuation">,</span><span class="name">alphabet</span><span class="operator">=</span><span class="name builtin pseudo">None</span><span class="punctuation">,</span> <span class="name">mat_name</span><span class="operator">=</span><span class="literal string">''</span><span class="punctuation">,</span> <span class="name">build_later</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">):</span>
</pre>
<ol class="arabic simple">
<li><tt class="docutils literal">data</tt>: can be either a dictionary, or another SeqMat
instance.</li>
<li><tt class="docutils literal">alphabet</tt>: a Bio.Alphabet instance. If not provided,
construct an alphabet from data.</li>
<li><tt class="docutils literal">mat_name</tt>: matrix name, such as &quot;BLOSUM62&quot; or &quot;PAM250&quot;</li>
<li><tt class="docutils literal">build_later</tt>: default false. If true, user may supply only
alphabet and empty dictionary, if intending to build the matrix
later. this skips the sanity check of alphabet size vs. matrix
size.</li>
</ol>
</li>
<li><pre class="code python first literal-block">
<span class="name">entropy</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">,</span><span class="name">obs_freq_mat</span><span class="punctuation">)</span>
</pre>
<ol class="arabic simple">
<li><tt class="docutils literal">obs_freq_mat</tt>: an observed frequency matrix. Returns the
matrix’s entropy, based on the frequency in <tt class="docutils literal">obs_freq_mat</tt>.
The matrix instance should be LO or SUBS.</li>
</ol>
</li>
<li><pre class="code python first literal-block">
<span class="name builtin">sum</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">)</span>
</pre>
<p>Calculates the sum of values for each letter in the matrix’s
alphabet, and returns it as a dictionary of the form
<tt class="docutils literal">{i1: s1, i2: <span class="pre">s2,...,in:sn}</span></tt>, where:</p>
<ul class="simple">
<li>i: an alphabet letter;</li>
<li>s: sum of all values in a half-matrix for that letter;</li>
<li>n: number of letters in alphabet.</li>
</ul>
</li>
<li><pre class="code python first literal-block">
<span class="name">print_mat</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">,</span><span class="name">f</span><span class="punctuation">,</span><span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;</span><span class="literal string interpol">%4d</span><span class="literal string">&quot;</span><span class="punctuation">,</span><span class="name">bottomformat</span><span class="operator">=</span><span class="literal string">&quot;</span><span class="literal string interpol">%4s</span><span class="literal string">&quot;</span><span class="punctuation">,</span><span class="name">alphabet</span><span class="operator">=</span><span class="name builtin pseudo">None</span><span class="punctuation">)</span>
</pre>
<p>prints the matrix to file handle f. <tt class="docutils literal">format</tt> is the format field
for the matrix values; <tt class="docutils literal">bottomformat</tt> is the format field for
the bottom row, containing matrix letters. Example output for a
3-letter alphabet matrix:</p>
<pre class="code python literal-block">
<span class="name">A</span> <span class="literal number integer">23</span>
<span class="name">B</span> <span class="literal number integer">12</span> <span class="literal number integer">34</span>
<span class="name">C</span> <span class="literal number integer">7</span>  <span class="literal number integer">22</span>  <span class="literal number integer">27</span>
  <span class="name">A</span>   <span class="name">B</span>   <span class="name">C</span>
</pre>
<p>The <tt class="docutils literal">alphabet</tt> optional argument is a string of all characters
in the alphabet. If supplied, the order of letters along the axes
is taken from the string, rather than by alphabetical order.</p>
</li>
</ol>
</li>
<li><p class="first">Usage</p>
<p>The following section is laid out in the order by which most people
wish to generate a log-odds matrix. Of course, interim matrices can
be generated and investigated. Most people just want a log-odds
matrix, that’s all.</p>
<ol class="arabic">
<li><p class="first">Generating an Accepted Replacement Matrix</p>
<p>Initially, you should generate an accepted replacement matrix
(ARM) from your data. The values in ARM are the counted number of
replacements according to your data. The data could be a set of
pairs or multiple alignments. So for instance if Alanine was
replaced by Cysteine 10 times, and Cysteine by Alanine 12 times,
the corresponding ARM entries would be:</p>
<pre class="code python literal-block">
<span class="punctuation">(</span><span class="literal string">'A'</span><span class="punctuation">,</span><span class="literal string">'C'</span><span class="punctuation">):</span> <span class="literal number integer">10</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'C'</span><span class="punctuation">,</span><span class="literal string">'A'</span><span class="punctuation">):</span> <span class="literal number integer">12</span>
</pre>
<p>as order doesn’t matter, user can already provide only one entry:</p>
<pre class="code python literal-block">
<span class="punctuation">(</span><span class="literal string">'A'</span><span class="punctuation">,</span><span class="literal string">'C'</span><span class="punctuation">):</span> <span class="literal number integer">22</span>
</pre>
<p>A SeqMat instance may be initialized with either a full (first
method of counting: 10, 12) or half (the latter method, 22)
matrices. A full protein alphabet matrix would be of the size
20x20 = 400. A half matrix of that alphabet would be 20x20/2 +
20/2 = 210. That is because same-letter entries don’t change. (The
matrix diagonal). Given an alphabet size of N:</p>
<ol class="arabic simple">
<li>Full matrix size:N*N</li>
<li>Half matrix size: N(N+1)/2</li>
</ol>
<p>The SeqMat constructor automatically generates a half-matrix, if a
full matrix is passed. If a half matrix is passed, letters in the
key should be provided in alphabetical order: (’A’,’C’) and not
(’C’,A’).</p>
<p>At this point, if all you wish to do is generate a log-odds
matrix, please go to the section titled Example of Use. The
following text describes the nitty-gritty of internal functions,
to be used by people who wish to investigate their
nucleotide/amino-acid frequency data more thoroughly.</p>
</li>
<li><p class="first">Generating the observed frequency matrix (OFM)</p>
<p>Use:</p>
<pre class="code python literal-block">
<span class="name">OFM</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">_build_obs_freq_mat</span><span class="punctuation">(</span><span class="name">ARM</span><span class="punctuation">)</span>
</pre>
<p>The OFM is generated from the ARM, only instead of replacement
counts, it contains replacement frequencies.</p>
</li>
<li><p class="first">Generating an expected frequency matrix (EFM)</p>
<p>Use:</p>
<pre class="code python literal-block">
<span class="name">EFM</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">_build_exp_freq_mat</span><span class="punctuation">(</span><span class="name">OFM</span><span class="punctuation">,</span><span class="name">exp_freq_table</span><span class="punctuation">)</span>
</pre>
<ol class="arabic simple">
<li><tt class="docutils literal">exp_freq_table</tt>: should be a FreqTable instance. See
section <a class="reference external" href="#sec:freq_table">20.2.2</a> for detailed information
on FreqTable. Briefly, the expected frequency table has the
frequencies of appearance for each member of the alphabet. It
is implemented as a dictionary with the alphabet letters as
keys, and each letter’s frequency as a value. Values sum to 1.</li>
</ol>
<p>The expected frequency table can (and generally should) be
generated from the observed frequency matrix. So in most cases you
will generate <tt class="docutils literal">exp_freq_table</tt> using:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exp_freq_table</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">_exp_freq_table_from_obs_freq</span><span class="punctuation">(</span><span class="name">OFM</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">EFM</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">_build_exp_freq_mat</span><span class="punctuation">(</span><span class="name">OFM</span><span class="punctuation">,</span><span class="name">exp_freq_table</span><span class="punctuation">)</span>
</pre>
<p>But you can supply your own <tt class="docutils literal">exp_freq_table</tt>, if you wish</p>
</li>
<li><p class="first">Generating a substitution frequency matrix (SFM)</p>
<p>Use:</p>
<pre class="code python literal-block">
<span class="name">SFM</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">_build_subs_mat</span><span class="punctuation">(</span><span class="name">OFM</span><span class="punctuation">,</span><span class="name">EFM</span><span class="punctuation">)</span>
</pre>
<p>Accepts an OFM, EFM. Provides the division product of the
corresponding values.</p>
</li>
<li><p class="first">Generating a log-odds matrix (LOM)</p>
<p>Use:</p>
<pre class="code python literal-block">
<span class="name">LOM</span><span class="operator">=</span><span class="name">SubsMat</span><span class="operator">.</span><span class="name">_build_log_odds_mat</span><span class="punctuation">(</span><span class="name">SFM</span><span class="punctuation">[,</span><span class="name">logbase</span><span class="operator">=</span><span class="literal number integer">10</span><span class="punctuation">,</span><span class="name">factor</span><span class="operator">=</span><span class="literal number float">10.0</span><span class="punctuation">,</span><span class="name">round_digit</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">])</span>
</pre>
<ol class="arabic simple">
<li>Accepts an SFM.</li>
<li><tt class="docutils literal">logbase</tt>: base of the logarithm used to generate the
log-odds values.</li>
<li><tt class="docutils literal">factor</tt>: factor used to multiply the log-odds values. Each
entry is generated by log(LOM[key])*factor And rounded to the
<tt class="docutils literal">round_digit</tt> place after the decimal point, if required.</li>
</ol>
</li>
</ol>
</li>
<li><p class="first">Example of use</p>
<p>As most people would want to generate a log-odds matrix, with minimum
hassle, SubsMat provides one function which does it all:</p>
<pre class="code python literal-block">
<span class="name">make_log_odds_matrix</span><span class="punctuation">(</span><span class="name">acc_rep_mat</span><span class="punctuation">,</span><span class="name">exp_freq_table</span><span class="operator">=</span><span class="name builtin pseudo">None</span><span class="punctuation">,</span><span class="name">logbase</span><span class="operator">=</span><span class="literal number integer">10</span><span class="punctuation">,</span>
                      <span class="name">factor</span><span class="operator">=</span><span class="literal number float">10.0</span><span class="punctuation">,</span><span class="name">round_digit</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">):</span>
</pre>
<ol class="arabic simple">
<li><tt class="docutils literal">acc_rep_mat</tt>: user provided accepted replacements matrix</li>
<li><tt class="docutils literal">exp_freq_table</tt>: expected frequencies table. Used if provided,
if not, generated from the <tt class="docutils literal">acc_rep_mat</tt>.</li>
<li><tt class="docutils literal">logbase</tt>: base of logarithm for the log-odds matrix. Default
base 10.</li>
<li><tt class="docutils literal">round_digit</tt>: number after decimal digit to which result should
be rounded. Default zero.</li>
</ol>
</li>
</ol>
</div>
<div class="section" id="freqtable">
<h3>20.2.2  FreqTable</h3>
<pre class="code python literal-block">
<span class="name">FreqTable</span><span class="operator">.</span><span class="name">FreqTable</span><span class="punctuation">(</span><span class="name">UserDict</span><span class="operator">.</span><span class="name">UserDict</span><span class="punctuation">)</span>
</pre>
<ol class="arabic">
<li><p class="first">Attributes:</p>
<ol class="arabic simple">
<li><tt class="docutils literal">alphabet</tt>: A Bio.Alphabet instance.</li>
<li><tt class="docutils literal">data</tt>: frequency dictionary</li>
<li><tt class="docutils literal">count</tt>: count dictionary (in case counts are provided).</li>
</ol>
</li>
<li><p class="first">Functions:</p>
<ol class="arabic simple">
<li><tt class="docutils literal">read_count(f)</tt>: read a count file from stream f. Then convert
to frequencies</li>
<li><tt class="docutils literal">read_freq(f)</tt>: read a frequency data file from stream f. Of
course, we then don’t have the counts, but it is usually the
letter frquencies which are interesting.</li>
</ol>
</li>
<li><p class="first">Example of use: The expected count of the residues in the database is
sitting in a file, whitespace delimited, in the following format
(example given for a 3-letter alphabet):</p>
<pre class="code python literal-block">
<span class="name">A</span>   <span class="literal number integer">35</span>
<span class="name">B</span>   <span class="literal number integer">65</span>
<span class="name">C</span>   <span class="literal number integer">100</span>
</pre>
<p>And will be read using the <tt class="docutils literal">FreqTable.read_count(file_handle)</tt>
function.</p>
<p>An equivalent frequency file:</p>
<pre class="code python literal-block">
<span class="name">A</span>  <span class="literal number float">0.175</span>
<span class="name">B</span>  <span class="literal number float">0.325</span>
<span class="name">C</span>  <span class="literal number float">0.5</span>
</pre>
<p>Conversely, the residue frequencies or counts can be passed as a
dictionary. Example of a count dictionary (3-letter alphabet):</p>
<pre class="code python literal-block">
<span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number integer">35</span><span class="punctuation">,</span> <span class="literal string">'B'</span><span class="punctuation">:</span> <span class="literal number integer">65</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number integer">100</span><span class="punctuation">}</span>
</pre>
<p>Which means that an expected data count would give a 0.5 frequency
for ’C’, a 0.325 probability of ’B’ and a 0.175 probability of ’A’
out of 200 total, sum of A, B and C)</p>
<p>A frequency dictionary for the same data would be:</p>
<pre class="code python literal-block">
<span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.175</span><span class="punctuation">,</span> <span class="literal string">'B'</span><span class="punctuation">:</span> <span class="literal number float">0.325</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.5</span><span class="punctuation">}</span>
</pre>
<p>Summing up to 1.</p>
<p>When passing a dictionary as an argument, you should indicate whether
it is a count or a frequency dictionary. Therefore the FreqTable
class constructor requires two arguments: the dictionary itself, and
FreqTable.COUNT or FreqTable.FREQ indicating counts or frequencies,
respectively.</p>
<p>Read expected counts. readCount will already generate the frequencies
Any one of the following may be done to geerate the frequency table
(ftab):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">SubsMat</span> <span class="keyword namespace">import</span> <span class="operator">*</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ftab</span> <span class="operator">=</span> <span class="name">FreqTable</span><span class="operator">.</span><span class="name">FreqTable</span><span class="punctuation">(</span><span class="name">my_frequency_dictionary</span><span class="punctuation">,</span><span class="name">FreqTable</span><span class="operator">.</span><span class="name">FREQ</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ftab</span> <span class="operator">=</span> <span class="name">FreqTable</span><span class="operator">.</span><span class="name">FreqTable</span><span class="punctuation">(</span><span class="name">my_count_dictionary</span><span class="punctuation">,</span><span class="name">FreqTable</span><span class="operator">.</span><span class="name">COUNT</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ftab</span> <span class="operator">=</span> <span class="name">FreqTable</span><span class="operator">.</span><span class="name">read_count</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">'myCountFile'</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ftab</span> <span class="operator">=</span> <span class="name">FreqTable</span><span class="operator">.</span><span class="name">read_frequency</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">'myFrequencyFile'</span><span class="punctuation">))</span>
</pre>
</li>
</ol>
</div>
</div>
</div>
<div class="section" id="chapter-21-where-to-go-from-here-contributing-to-biopython">
<h1>Chapter 21  Where to go from here – contributing to Biopython</h1>
<div class="section" id="bug-reports-feature-requests">
<h2>21.1  Bug Reports + Feature Requests</h2>
<p>Getting feedback on the Biopython modules is very important to us.
Open-source projects like this benefit greatly from feedback,
bug-reports (and patches!) from a wide variety of contributors.</p>
<p>The main forums for discussing feature requests and potential bugs are
the <a class="reference external" href="http://biopython.org/wiki/Mailing_lists">Biopython mailing
lists</a>:</p>
<ul class="simple">
<li><a class="reference external" href="mailto:biopython&#64;biopython.org">biopython&#64;biopython.org</a> – An
unmoderated list for discussion of anything to do with Biopython.</li>
<li><a class="reference external" href="mailto:biopython-dev&#64;biopython.org">biopython-dev&#64;biopython.org</a>
– A more development oriented list that is mainly used by developers
(but anyone is free to contribute!).</li>
</ul>
<p>Additionally, if you think you’ve found a bug, you can submit it to our
bug-tracking page at
<a class="reference external" href="http://redmine.open-bio.org/projects/biopython">http://redmine.open-bio.org/projects/biopython</a>.
This way, it won’t get buried in anyone’s Inbox and forgotten about.</p>
</div>
<div class="section" id="mailing-lists-and-helping-newcomers">
<h2>21.2  Mailing lists and helping newcomers</h2>
<p>We encourage all our uses to sign up to the main Biopython mailing list.
Once you’ve got the hang of an area of Biopython, we’d encourage you to
help answer questions from beginners. After all, you were a beginner
once.</p>
</div>
<div class="section" id="contributing-documentation">
<h2>21.3  Contributing Documentation</h2>
<p>We’re happy to take feedback or contributions - either via a bug-report
or on the Mailing List. While reading this tutorial, perhaps you noticed
some topics you were interested in which were missing, or not clearly
explained. There is also Biopython’s built in documentation (the
docstrings, these are also
<a class="reference external" href="http://biopython.org/DIST/docs/api">online</a>), where again, you may
be able to help fill in any blanks.</p>
</div>
<div class="section" id="contributing-cookbook-examples">
<h2>21.4  Contributing cookbook examples</h2>
<p>As explained in Chapter <a class="reference external" href="#chapter:cookbook">18</a>, Biopython now has
a wiki collection of user contributed “cookbook” examples,
<a class="reference external" href="http://biopython.org/wiki/Category:Cookbook">http://biopython.org/wiki/Category:Cookbook</a>
– maybe you can add to this?</p>
</div>
<div class="section" id="maintaining-a-distribution-for-a-platform">
<h2>21.5  Maintaining a distribution for a platform</h2>
<p>We currently provide source code archives (suitable for any OS, if you
have the right build tools installed), and Windows Installers which are
just click and run. This covers all the major operating systems.</p>
<p>Most major Linux distributions have volunteers who take these source
code releases, and compile them into packages for Linux users to easily
install (taking care of dependencies etc). This is really great and we
are of course very grateful. If you would like to contribute to this
work, please find out more about how your Linux distribution handles
this.</p>
<p>Below are some tips for certain platforms to maybe get people started
with helping out:</p>
<dl class="docutils">
<dt><strong>Windows</strong></dt>
<dd><p class="first">– Windows products typically have a nice graphical installer that
installs all of the essential components in the right place. We use
Distutils to create a installer of this type fairly easily.</p>
<p>You must first make sure you have a C compiler on your Windows
computer, and that you can compile and install things (this is the
hard bit - see the Biopython installation instructions for info on
how to do this).</p>
<p>Once you are setup with a C compiler, making the installer just
requires doing:</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">setup</span><span class="operator">.</span><span class="name">py</span> <span class="name">bdist_wininst</span>
</pre>
<p class="last">Now you’ve got a Windows installer. Congrats! At the moment we have
no trouble shipping installers built on 32 bit windows. If anyone
would like to look into supporting 64 bit Windows that would be
great.</p>
</dd>
<dt><strong>RPMs</strong></dt>
<dd><p class="first">– RPMs are pretty popular package systems on some Linux platforms.
There is lots of documentation on RPMs available at
<a class="reference external" href="http://www.rpm.org">http://www.rpm.org</a> to help you get
started with them. To create an RPM for your platform is really
easy. You just need to be able to build the package from source
(having a C compiler that works is thus essential) – see the
Biopython installation instructions for more info on this.</p>
<p>To make the RPM, you just need to do:</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">setup</span><span class="operator">.</span><span class="name">py</span> <span class="name">bdist_rpm</span>
</pre>
<p class="last">This will create an RPM for your specific platform and a source RPM
in the directory <tt class="docutils literal">dist</tt>. This RPM should be good and ready to go,
so this is all you need to do! Nice and easy.</p>
</dd>
<dt><strong>Macintosh</strong></dt>
<dd>– Since Apple moved to Mac OS X, things have become much easier on
the Mac. We generally treat it as just another Unix variant, and
installing Biopython from source is just as easy as on Linux. The
easiest way to get all the GCC compilers etc installed is to install
Apple’s X-Code. We might be able to provide click and run installers
for Mac OS X, but to date there hasn’t been any demand.</dd>
</dl>
<p>Once you’ve got a package, please test it on your system to make sure it
installs everything in a good way and seems to work properly. Once you
feel good about it, send it off to one of the Biopython developers
(write to our main mailing list at <a class="reference external" href="mailto:biopython&#64;biopython.org">biopython&#64;biopython.org</a> if you’re not
sure who to send it to) and you’ve done it. Thanks!</p>
</div>
<div class="section" id="contributing-unit-tests">
<h2>21.6  Contributing Unit Tests</h2>
<p>Even if you don’t have any new functionality to add to Biopython, but
you want to write some code, please consider extending our unit test
coverage. We’ve devoted all of Chapter <a class="reference external" href="#sec:regr_test">19</a> to this
topic.</p>
</div>
<div class="section" id="contributing-code">
<h2>21.7  Contributing Code</h2>
<p>There are no barriers to joining Biopython code development other than
an interest in creating biology-related code in Python. The best place
to express an interest is on the Biopython mailing lists – just let us
know you are interested in coding and what kind of stuff you want to
work on. Normally, we try to have some discussion on modules before
coding them, since that helps generate good ideas – then just feel free
to jump right in and start coding!</p>
<p>The main Biopython release tries to be fairly uniform and interworkable,
to make it easier for users. You can read about some of (fairly
informal) coding style guidelines we try to use in Biopython in the
contributing documentation at
<a class="reference external" href="http://biopython.org/wiki/Contributing">http://biopython.org/wiki/Contributing</a>.
We also try to add code to the distribution along with tests (see
Chapter <a class="reference external" href="#sec:regr_test">19</a> for more info on the regression
testing framework) and documentation, so that everything can stay as
workable and well documented as possible (including docstrings). This
is, of course, the most ideal situation, under many situations you’ll be
able to find other people on the list who will be willing to help add
documentation or more tests for your code once you make it available.
So, to end this paragraph like the last, feel free to start working!</p>
<p>Please note that to make a code contribution you must have the legal
right to contribute it and license it under the Biopython license. If
you wrote it all yourself, and it is not based on any other code, this
shouldn’t be a problem. However, there are issues if you want to
contribute a derivative work - for example something based on GPL or
LPGL licenced code would not be compatible with our license. If you have
any queries on this, please discuss the issue on the biopython-dev
mailing list.</p>
<p>Another point of concern for any additions to Biopython regards any
build time or run time dependencies. Generally speaking, writing code to
interact with a standalone tool (like BLAST, EMBOSS or ClustalW) doesn’t
present a big problem. However, any dependency on another library - even
a Python library (especially one needed in order to compile and install
Biopython like NumPy) would need further discussion.</p>
<p>Additionally, if you have code that you don’t think fits in the
distribution, but that you want to make available, we maintain Script
Central
(<a class="reference external" href="http://biopython.org/wiki/Scriptcentral">http://biopython.org/wiki/Scriptcentral</a>)
which has pointers to freely available code in Python for
bioinformatics.</p>
<p>Hopefully this documentation has got you excited enough about Biopython
to try it out (and most importantly, contribute!). Thanks for reading
all the way through!</p>
</div>
</div>
<div class="section" id="chapter-22-appendix-useful-stuff-about-python">
<h1>Chapter 22  Appendix: Useful stuff about Python</h1>
<p>If you haven’t spent a lot of time programming in Python, many questions
and problems that come up in using Biopython are often related to Python
itself. This section tries to present some ideas and code that come up
often (at least for us!) while using the Biopython libraries. If you
have any suggestions for useful pointers that could go here, please
contribute!</p>
<div class="section" id="what-the-heck-is-a-handle">
<h2>22.1  What the heck is a handle?</h2>
<p>Handles are mentioned quite frequently throughout this documentation,
and are also fairly confusing (at least to me!). Basically, you can
think of a handle as being a “wrapper” around text information.</p>
<p>Handles provide (at least) two benefits over plain text information:</p>
<ol class="arabic simple">
<li>They provide a standard way to deal with information stored in
different ways. The text information can be in a file, or in a string
stored in memory, or the output from a command line program, or at
some remote website, but the handle provides a common way of dealing
with information in all of these formats.</li>
<li>They allow text information to be read incrementally, instead of all
at once. This is really important when you are dealing with huge text
files which would use up all of your memory if you had to load them
all.</li>
</ol>
<p>Handles can deal with text information that is being read (e. g. reading
from a file) or written (e. g. writing information to a file). In the
case of a “read” handle, commonly used functions are <tt class="docutils literal">read()</tt>, which
reads the entire text information from the handle, and <tt class="docutils literal">readline()</tt>,
which reads information one line at a time. For “write” handles, the
function <tt class="docutils literal">write()</tt> is regularly used.</p>
<p>The most common usage for handles is reading information from a file,
which is done using the built-in Python function <tt class="docutils literal">open</tt>. Here, we open
a handle to the file <a class="reference external" href="examples/m_cold.fasta">m_cold.fasta</a> (also
available online
<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/m_cold.fasta">here</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;r&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">readline</span><span class="punctuation">()</span>
<span class="literal string">&quot;&gt;gi|8332116|gb|BE037100.1|BE037100 MP14H09 MP Mesembryanthemum ...</span><span class="literal string escape">\n</span><span class="literal string">&quot;</span>
</pre>
<p>Handles are regularly used in Biopython for passing information to
parsers. For example, since Biopython 1.54 the main functions in
<tt class="docutils literal">Bio.SeqIO</tt> and <tt class="docutils literal">Bio.AlignIO</tt> have allowed you to use a filename
instead of a handle:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
</pre>
<p>On older versions of Biopython you had to use a handle, e.g.</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;r&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>This pattern is still useful - for example suppose you have a gzip
compressed FASTA file you want to parse:</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">gzip</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name">gzip</span><span class="operator">.</span><span class="name">open</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta.gz&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>See Section <a class="reference external" href="#sec:SeqIO_compressed">5.2</a> for more examples like
this, including reading bzip2 compressed files.</p>
<div class="section" id="creating-a-handle-from-a-string">
<h3>22.1.1  Creating a handle from a string</h3>
<p>One useful thing is to be able to turn information contained in a string
into a handle. The following example shows how to do this using
<tt class="docutils literal">cStringIO</tt> from the Python standard library:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_info</span> <span class="operator">=</span> <span class="literal string">'A string</span><span class="literal string escape">\n</span><span class="literal string"> with multiple lines.'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_info</span>
<span class="name">A</span> <span class="name">string</span>
 <span class="keyword">with</span> <span class="name">multiple</span> <span class="name">lines</span><span class="operator">.</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">StringIO</span> <span class="keyword namespace">import</span> <span class="name">StringIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_info_handle</span> <span class="operator">=</span> <span class="name">StringIO</span><span class="punctuation">(</span><span class="name">my_info</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">first_line</span> <span class="operator">=</span> <span class="name">my_info_handle</span><span class="operator">.</span><span class="name">readline</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">first_line</span>
<span class="name">A</span> <span class="name">string</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">second_line</span> <span class="operator">=</span> <span class="name">my_info_handle</span><span class="operator">.</span><span class="name">readline</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">second_line</span>
 <span class="keyword">with</span> <span class="name">multiple</span> <span class="name">lines</span><span class="operator">.</span>
</pre>
</div>
</div>
</div>
<div class="section" id="id20">
<h1>References</h1>
<dl class="docutils">
<dt>[1]</dt>
<dd>Peter J. A. Cock, Tiago Antao, Jeffrey T. Chang, Brad A. Chapman,
Cymon J. Cox, Andrew Dalke, Iddo Friedberg, Thomas Hamelryck, Frank
Kauff, Bartek Wilczynski, Michiel J. L. de Hoon: “Biopython: freely
available Python tools for computational molecular biology and
bioinformatics”. <em>Bioinformatics</em> <strong>25</strong> (11), 1422–1423 (2009).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/btp163">doi:10.1093/bioinformatics/btp163</a>,</dd>
<dt>[2]</dt>
<dd>Leighton Pritchard, Jennifer A. White, Paul R.J. Birch, Ian K. Toth:
“GenomeDiagram: a python package for the visualization of
large-scale genomic data”. <em>Bioinformatics</em> <strong>22</strong> (5): 616–617
(2006).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/btk021">doi:10.1093/bioinformatics/btk021</a>,</dd>
<dt>[3]</dt>
<dd>Ian K. Toth, Leighton Pritchard, Paul R. J. Birch: “Comparative
genomics reveals what makes an enterobacterial plant pathogen”.
<em>Annual Review of Phytopathology</em> <strong>44</strong>: 305–336 (2006).
<a class="reference external" href="http://dx.doi.org/10.1146/annurev.phyto.44.070505.143444">doi:10.1146/annurev.phyto.44.070505.143444</a>,</dd>
<dt>[4]</dt>
<dd>Géraldine A. van der Auwera, Jaroslaw E. Król, Haruo Suzuki, Brian
Foster, Rob van Houdt, Celeste J. Brown, Max Mergeay, Eva M. Top:
“Plasmids captured in C. metallidurans CH34: defining the PromA
family of broad-host-range plasmids”. <em>Antonie van Leeuwenhoek</em>
<strong>96</strong> (2): 193–204 (2009).
<a class="reference external" href="http://dx.doi.org/10.1007/s10482-009-9316-9">doi:10.1007/s10482-009-9316-9</a></dd>
<dt>[5]</dt>
<dd>Caroline Proux, Douwe van Sinderen, Juan Suarez, Pilar Garcia,
Victor Ladero, Gerald F. Fitzgerald, Frank Desiere, Harald Brüssow:
“The dilemma of phage taxonomy illustrated by comparative genomics
of Sfi21-Like Siphoviridae in lactic acid bacteria”. <em>Journal of
Bacteriology</em> <strong>184</strong> (21): 6026–6036 (2002).
<a class="reference external" href="http://dx.doi.org/10.1128/JB.184.21.6026-6036.2002">http://dx.doi.org/10.1128/JB.184.21.6026-6036.2002</a></dd>
<dt>[6]</dt>
<dd>Florian Jupe, Leighton Pritchard, Graham J. Etherington, Katrin
MacKenzie, Peter JA Cock, Frank Wright, Sanjeev Kumar Sharma1, Dan
Bolser, Glenn J Bryan, Jonathan DG Jones, Ingo Hein: “Identification
and localisation of the NB-LRR gene family within the potato
genome”. <em>BMC Genomics</em> <strong>13</strong>: 75 (2012).
<a class="reference external" href="http://dx.doi.org/10.1186/1471-2164-13-75">http://dx.doi.org/10.1186/1471-2164-13-75</a></dd>
<dt>[7]</dt>
<dd>Peter J. A. Cock, Christopher J. Fields, Naohisa Goto, Michael L.
Heuer, Peter M. Rice: “The Sanger FASTQ file format for sequences
with quality scores, and the Solexa/Illumina FASTQ variants”.
<em>Nucleic Acids Research</em> <strong>38</strong> (6): 1767–1771 (2010).
<a class="reference external" href="http://dx.doi.org/10.1093/nar/gkp1137">doi:10.1093/nar/gkp1137</a></dd>
<dt>[8]</dt>
<dd>Patrick O. Brown, David Botstein: “Exploring the new world of the
genome with DNA microarrays”. <em>Nature Genetics</em> <strong>21</strong> (Supplement
1), 33–37 (1999).
<a class="reference external" href="http://dx.doi.org/10.1038/4462">doi:10.1038/4462</a></dd>
<dt>[9]</dt>
<dd>Eric Talevich, Brandon M. Invergo, Peter J.A. Cock, Brad A. Chapman:
“Bio.Phylo: A unified toolkit for processing, analyzing and
visualizing phylogenetic trees in Biopython”. <em>BMC Bioinformatics</em>
<strong>13</strong>: 209 (2012).
<a class="reference external" href="http://dx.doi.org/10.1186/1471-2105-13-209">doi:10.1186/1471-2105-13-209</a></dd>
<dt>[10]</dt>
<dd>Athel Cornish-Bowden: “Nomenclature for incompletely specified bases
in nucleic acid sequences: Recommendations 1984.” <em>Nucleic Acids
Research</em> <strong>13</strong> (9): 3021–3030 (1985).
<a class="reference external" href="http://dx.doi.org/10.1093/nar/13.9.3021">doi:10.1093/nar/13.9.3021</a></dd>
<dt>[11]</dt>
<dd>Douglas R. Cavener: “Comparison of the consensus sequence flanking
translational start sites in Drosophila and vertebrates.” <em>Nucleic
Acids Research</em> <strong>15</strong> (4): 1353–1361 (1987).
<a class="reference external" href="http://dx.doi.org/10.1093/nar/15.4.1353">doi:10.1093/nar/15.4.1353</a></dd>
<dt>[12]</dt>
<dd>Timothy L. Bailey and Charles Elkan: “Fitting a mixture model by
expectation maximization to discover motifs in biopolymers”,
<em>Proceedings of the Second International Conference on Intelligent
Systems for Molecular Biology</em> 28–36. AAAI Press, Menlo Park,
California (1994).</dd>
<dt>[13]</dt>
<dd>Brad Chapman and Jeff Chang: “Biopython: Python tools for
computational biology”. <em>ACM SIGBIO Newsletter</em> <strong>20</strong> (2): 15–19
(August 2000).</dd>
<dt>[14]</dt>
<dd>Michiel J. L. de Hoon, Seiya Imoto, John Nolan, Satoru Miyano: “Open
source clustering software”. <em>Bioinformatics</em> <strong>20</strong> (9): 1453–1454
(2004).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/bth078">doi:10.1093/bioinformatics/bth078</a></dd>
<dt>[15]</dt>
<dd>Michiel B. Eisen, Paul T. Spellman, Patrick O. Brown, David
Botstein: “Cluster analysis and display of genome-wide expression
patterns”. <em>Proceedings of the National Academy of Science USA</em>
<strong>95</strong> (25): 14863–14868 (1998).
<a class="reference external" href="http://dx.doi.org/10.1073/pnas.96.19.10943-c">doi:10.1073/pnas.96.19.10943-c</a></dd>
<dt>[16]</dt>
<dd>Gene H. Golub, Christian Reinsch: “Singular value decomposition and
least squares solutions”. In <em>Handbook for Automatic Computation</em>,
<strong>2</strong>, (Linear Algebra) (J. H. Wilkinson and C. Reinsch, eds),
134–151. New York: Springer-Verlag (1971).</dd>
<dt>[17]</dt>
<dd>Gene H. Golub, Charles F. Van Loan: <em>Matrix computations</em>, 2nd
edition (1989).</dd>
<dt>[18]</dt>
<dd>Thomas Hamelryck and Bernard Manderick: 11PDB parser and structure
class implemented in Python”. <em>Bioinformatics</em>, <strong>19</strong> (17):
2308–2310 (2003) <a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/btg299">doi:
10.1093/bioinformatics/btg299</a>.</dd>
<dt>[19]</dt>
<dd>Thomas Hamelryck: “Efficient identification of side-chain patterns
using a multidimensional index tree”. <em>Proteins</em> <strong>51</strong> (1): 96–108
(2003).
<a class="reference external" href="http://dx.doi.org/10.1002/prot.10338">doi:10.1002/prot.10338</a></dd>
<dt>[20]</dt>
<dd>Thomas Hamelryck: “An amino acid has two sides; A new 2D measure
provides a different view of solvent exposure”. <em>Proteins</em> <strong>59</strong>
(1): 29–48 (2005).
<a class="reference external" href="http://dx.doi.org/10.1002/prot.20379">doi:10.1002/prot.20379</a>.</dd>
<dt>[21]</dt>
<dd>John A. Hartiga. <em>Clustering algorithms</em>. New York: Wiley (1975).</dd>
<dt>[22]</dt>
<dd>Anil L. Jain, Richard C. Dubes: <em>Algorithms for clustering data</em>.
Englewood Cliffs, N.J.: Prentice Hall (1988).</dd>
<dt>[23]</dt>
<dd>Voratas Kachitvichyanukul, Bruce W. Schmeiser: Binomial Random
Variate Generation. <em>Communications of the ACM</em> <strong>31</strong> (2): 216–222
(1988).
<a class="reference external" href="http://dx.doi.org/10.1145/42372.42381">doi:10.1145/42372.42381</a></dd>
<dt>[24]</dt>
<dd>Teuvo Kohonen: “Self-organizing maps”, 2nd Edition. Berlin; New
York: Springer-Verlag (1997).</dd>
<dt>[25]</dt>
<dd>Pierre L’Ecuyer: “Efficient and Portable Combined Random Number
Generators.” <em>Communications of the ACM</em> <strong>31</strong> (6): 742–749,774
(1988).
<a class="reference external" href="http://dx.doi.org/10.1145/62959.62969">doi:10.1145/62959.62969</a></dd>
<dt>[26]</dt>
<dd>Indraneel Majumdar, S. Sri Krishna, Nick V. Grishin: “PALSSE: A
program to delineate linear secondary structural elements from
protein structures.” <em>BMC Bioinformatics</em>, <strong>6</strong>: 202 (2005).
<a class="reference external" href="http://dx.doi.org/10.1186/1471-2105-6-202">doi:10.1186/1471-2105-6-202</a>.</dd>
<dt>[27]</dt>
<dd>V. Matys, E. Fricke, R. Geffers, E. G?ssling, M. Haubrock, R. Hehl,
K. Hornischer, D. Karas, A.E. Kel, O.V. Kel-Margoulis, D.U. Kloos,
S. Land, B. Lewicki-Potapov, H. Michael, R. Münch, I. Reuter, S.
Rotert, H. Saxel, M. Scheer, S. Thiele, E. Wingender E: “TRANSFAC:
transcriptional regulation, from patterns to profiles.” Nucleic
Acids Research <strong>31</strong> (1): 374–378 (2003).
<a class="reference external" href="http://dx.doi.org/10.1093/nar/gkg108">doi:10.1093/nar/gkg108</a></dd>
<dt>[28]</dt>
<dd>Robin Sibson: “SLINK: An optimally efficient algorithm for the
single-link cluster method”. <em>The Computer Journal</em> <strong>16</strong> (1):
30–34 (1973).
<a class="reference external" href="http://dx.doi.org/10.1093/comjnl/16.1.30">doi:10.1093/comjnl/16.1.30</a></dd>
<dt>[29]</dt>
<dd>George W. Snedecor, William G. Cochran: <em>Statistical methods</em>. Ames,
Iowa: Iowa State University Press (1989).</dd>
<dt>[30]</dt>
<dd>Pablo Tamayo, Donna Slonim, Jill Mesirov, Qing Zhu, Sutisak
Kitareewan, Ethan Dmitrovsky, Eric S. Lander, Todd R. Golub:
“Interpreting patterns of gene expression with self-organizing maps:
Methods and application to hematopoietic differentiation”.
<em>Proceedings of the National Academy of Science USA</em> <strong>96</strong> (6):
2907–2912 (1999).
<a class="reference external" href="http://dx.doi.org/10.1073/pnas.96.6.2907">doi:10.1073/pnas.96.6.2907</a></dd>
<dt>[31]</dt>
<dd>Robert C. Tryon, Daniel E. Bailey: <em>Cluster analysis</em>. New York:
McGraw-Hill (1970).</dd>
<dt>[32]</dt>
<dd>John W. Tukey: “Exploratory data analysis”. Reading, Mass.:
Addison-Wesley Pub. Co. (1977).</dd>
<dt>[33]</dt>
<dd>Ka Yee Yeung, Walter L. Ruzzo: “Principal Component Analysis for
clustering gene expression data”. <em>Bioinformatics</em> <strong>17</strong> (9):
763–774 (2001).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/17.9.763">doi:10.1093/bioinformatics/17.9.763</a></dd>
<dt>[34]</dt>
<dd>Alok Saldanha: “Java Treeview—extensible visualization of microarray
data”. <em>Bioinformatics</em> <strong>20</strong> (17): 3246–3248 (2004).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/bth349">http://dx.doi.org/10.1093/bioinformatics/bth349</a></dd>
</dl>
</div>
</div>
</body>
</html>
