<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Chapter 1 Introduction &mdash; biopython_en 0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="biopython_en 0.1 documentation" href="index.html" />
    <link rel="next" title="Chapter 2 Quick Start – What can you do with Biopython?" href="chr02.html" />
    <link rel="prev" title="Welcome to biopython_cn’s documentation!" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="chr02.html" title="Chapter 2 Quick Start – What can you do with Biopython?"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to biopython_cn’s documentation!"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">biopython_en 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="chapter-1-introduction">
<h1>Chapter 1  Introduction<a class="headerlink" href="#chapter-1-introduction" title="Permalink to this headline">¶</a></h1>
<div class="section" id="what-is-biopython">
<h2>1.1  What is Biopython?<a class="headerlink" href="#what-is-biopython" title="Permalink to this headline">¶</a></h2>
<p>The Biopython Project is an international association of developers of
freely available Python
(<a class="reference external" href="http://www.python.org">http://www.python.org</a>) tools for
computational molecular biology. Python is an object oriented,
interpreted, flexible language that is becoming increasingly popular for
scientific computing. Python is easy to learn, has a very clear syntax
and can easily be extended with modules written in C, C++ or FORTRAN.</p>
<p>The Biopython web site
(<a class="reference external" href="http://www.biopython.org">http://www.biopython.org</a>) provides
an online resource for modules, scripts, and web links for developers of
Python-based software for bioinformatics use and research. Basically,
the goal of Biopython is to make it as easy as possible to use Python
for bioinformatics by creating high-quality, reusable modules and
classes. Biopython features include parsers for various Bioinformatics
file formats (BLAST, Clustalw, FASTA, Genbank,...), access to online
services (NCBI, Expasy,...), interfaces to common and not-so-common
programs (Clustalw, DSSP, MSMS...), a standard sequence class, various
clustering modules, a KD tree data structure etc. and even
documentation.</p>
<p>Basically, we just like to program in Python and want to make it as easy
as possible to use Python for bioinformatics by creating high-quality,
reusable modules and scripts.</p>
</div>
<div class="section" id="what-can-i-find-in-the-biopython-package">
<h2>1.2  What can I find in the Biopython package<a class="headerlink" href="#what-can-i-find-in-the-biopython-package" title="Permalink to this headline">¶</a></h2>
<p>The main Biopython releases have lots of functionality, including:</p>
<ul class="simple">
<li>The ability to parse bioinformatics files into Python utilizable data
structures, including support for the following formats:<ul>
<li>Blast output – both from standalone and WWW Blast</li>
<li>Clustalw</li>
<li>FASTA</li>
<li>GenBank</li>
<li>PubMed and Medline</li>
<li>ExPASy files, like Enzyme and Prosite</li>
<li>SCOP, including ‘dom’ and ‘lin’ files</li>
<li>UniGene</li>
<li>SwissProt</li>
</ul>
</li>
<li>Files in the supported formats can be iterated over record by record
or indexed and accessed via a Dictionary interface.</li>
<li>Code to deal with popular on-line bioinformatics destinations such
as:<ul>
<li>NCBI – Blast, Entrez and PubMed services</li>
<li>ExPASy – Swiss-Prot and Prosite entries, as well as Prosite
searches</li>
</ul>
</li>
<li>Interfaces to common bioinformatics programs such as:<ul>
<li>Standalone Blast from NCBI</li>
<li>Clustalw alignment program</li>
<li>EMBOSS command line tools</li>
</ul>
</li>
<li>A standard sequence class that deals with sequences, ids on
sequences, and sequence features.</li>
<li>Tools for performing common operations on sequences, such as
translation, transcription and weight calculations.</li>
<li>Code to perform classification of data using k Nearest Neighbors,
Naive Bayes or Support Vector Machines.</li>
<li>Code for dealing with alignments, including a standard way to create
and deal with substitution matrices.</li>
<li>Code making it easy to split up parallelizable tasks into separate
processes.</li>
<li>GUI-based programs to do basic sequence manipulations, translations,
BLASTing, etc.</li>
<li>Extensive documentation and help with using the modules, including
this file, on-line wiki documentation, the web site, and the mailing
list.</li>
<li>Integration with BioSQL, a sequence database schema also supported by
the BioPerl and BioJava projects.</li>
</ul>
<p>We hope this gives you plenty of reasons to download and start using
Biopython!</p>
</div>
<div class="section" id="installing-biopython">
<h2>1.3  Installing Biopython<a class="headerlink" href="#installing-biopython" title="Permalink to this headline">¶</a></h2>
<p>All of the installation information for Biopython was separated from
this document to make it easier to keep updated.</p>
<p>The short version is go to our downloads page
(<a class="reference external" href="http://biopython.org/wiki/Download">http://biopython.org/wiki/Download</a>),
download and install the listed dependencies, then download and install
Biopython. Biopython runs on many platforms (Windows, Mac, and on the
various flavors of Linux and Unix). For Windows we provide pre-compiled
click-and-run installers, while for Unix and other operating systems you
must install from source as described in the included README file. This
is usually as simple as the standard commands:</p>
<div class="code python highlight-python"><pre>python setup.py build
python setup.py test
sudo python setup.py install</pre>
</div>
<p>(You can in fact skip the build and test, and go straight to the install
– but its better to make sure everything seems to be working.)</p>
<p>The longer version of our installation instructions covers installation
of Python, Biopython dependencies and Biopython itself. It is available
in PDF
(<a class="reference external" href="http://biopython.org/DIST/docs/install/Installation.pdf">http://biopython.org/DIST/docs/install/Installation.pdf</a>)
and HTML formats
(<a class="reference external" href="http://biopython.org/DIST/docs/install/Installation.html">http://biopython.org/DIST/docs/install/Installation.html</a>).</p>
</div>
<div class="section" id="frequently-asked-questions-faq">
<h2>1.4  Frequently Asked Questions (FAQ)<a class="headerlink" href="#frequently-asked-questions-faq" title="Permalink to this headline">¶</a></h2>
<ol class="arabic">
<li><p class="first"><em>How do I cite Biopython in a scientific publication?</em></p>
<p>Please cite our application note [<a class="reference external" href="#cock2009">1</a>, Cock <em>et al.</em>,
2009] as the main Biopython reference. In addition, please cite any
publications from the following list if appropriate, in particular as
a reference for specific modules within Biopython (more information
can be found on our website):</p>
<ul class="simple">
<li>For the official project announcement: [<a class="reference external" href="#chapman2000">13</a>,
Chapman and Chang, 2000];</li>
<li>For <tt class="docutils literal"><span class="pre">Bio.PDB</span></tt>: [<a class="reference external" href="#hamelryck2003a">18</a>, Hamelryck and
Manderick, 2003];</li>
<li>For <tt class="docutils literal"><span class="pre">Bio.Cluster</span></tt>: [<a class="reference external" href="#dehoon2004">14</a>, De Hoon <em>et al.</em>,
2004];</li>
<li>For <tt class="docutils literal"><span class="pre">Bio.Graphics.GenomeDiagram</span></tt>: [<a class="reference external" href="#pritchard2006">2</a>,
Pritchard <em>et al.</em>, 2006];</li>
<li>For <tt class="docutils literal"><span class="pre">Bio.Phylo</span></tt> and <tt class="docutils literal"><span class="pre">Bio.Phylo.PAML</span></tt>: [<a class="reference external" href="#talevich2012">9</a>,
Talevich <em>et al.</em>, 2012];</li>
<li>For the FASTQ file format as supported in Biopython, BioPerl,
BioRuby, BioJava, and EMBOSS: [<a class="reference external" href="#cock2010">7</a>, Cock <em>et al.</em>,
2010].</li>
</ul>
</li>
<li><p class="first"><em>How should I capitalize “Biopython”? Is “BioPython” OK?</em></p>
<p>The correct capitalization is “Biopython”, not “BioPython” (even
though that would have matched BioPerl, BioJava and BioRuby).</p>
</li>
<li><p class="first"><em>How do I find out what version of Biopython I have installed?</em></p>
<p>Use this:</p>
<div class="code python highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">Bio</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">Bio</span><span class="o">.</span><span class="n">__version__</span>
<span class="gp">...</span>
</pre></div>
</div>
<p>If the “<tt class="docutils literal"><span class="pre">import</span> <span class="pre">Bio</span></tt>” line fails, Biopython is not installed.
If the second line fails, your version is very out of date. If the
version string ends with a plus, you don’t have an official release,
but a snapshot of the in development code.</p>
</li>
<li><p class="first"><em>Where is the latest version of this document?</em></p>
<p>If you download a Biopython source code archive, it will include the
relevant version in both HTML and PDF formats. The latest published
version of this document (updated at each release) is online:</p>
<ul class="simple">
<li><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html">http://biopython.org/DIST/docs/tutorial/Tutorial.html</a></li>
<li><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.pdf">http://biopython.org/DIST/docs/tutorial/Tutorial.pdf</a></li>
</ul>
<p>If you are using the very latest unreleased code from our repository
you can find copies of the in-progress tutorial here:</p>
<ul class="simple">
<li><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial-dev.html">http://biopython.org/DIST/docs/tutorial/Tutorial-dev.html</a></li>
<li><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial-dev.pdf">http://biopython.org/DIST/docs/tutorial/Tutorial-dev.pdf</a></li>
</ul>
</li>
<li><p class="first"><em>Which “Numerical Python” do I need?</em></p>
<p>For Biopython 1.48 or earlier, you needed the old Numeric module.
For Biopython 1.49 onwards, you need the newer NumPy instead. Both
Numeric and NumPy can be installed on the same machine fine. See
also: <a class="reference external" href="http://numpy.scipy.org/">http://numpy.scipy.org/</a></p>
</li>
<li><p class="first"><em>Why is the</em> <tt class="docutils literal"><span class="pre">Seq</span></tt> <em>object missing the (back) transcription &amp;
translation methods described in this Tutorial?</em></p>
<p>You need Biopython 1.49 or later. Alternatively, use the <tt class="docutils literal"><span class="pre">Bio.Seq</span></tt>
module functions described in
Section <a class="reference external" href="#sec:seq-module-functions">3.14</a>.</p>
</li>
<li><p class="first"><em>Why is the</em> <tt class="docutils literal"><span class="pre">Seq</span></tt> <em>object missing the upper &amp; lower methods
described in this Tutorial?</em></p>
<p>You need Biopython 1.53 or later. Alternatively, use
<tt class="docutils literal"><span class="pre">str(my_seq).upper()</span></tt> to get an upper case string. If you need a
Seq object, try <tt class="docutils literal"><span class="pre">Seq(str(my_seq).upper())</span></tt> but be careful about
blindly re-using the same alphabet.</p>
</li>
<li><p class="first"><em>Why doesn’t the</em> <tt class="docutils literal"><span class="pre">Seq</span></tt> <em>object translation method support the</em>
<tt class="docutils literal"><span class="pre">cds</span></tt> <em>option described in this Tutorial?</em></p>
<p>You need Biopython 1.51 or later.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> <em>work? It imports fine but there is no
parse function etc.</em></p>
<p>You need Biopython 1.43 or later. Older versions did contain some
related code under the <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> name which has since been
removed - and this is why the import “works”.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal"><span class="pre">Bio.SeqIO.read()</span></tt> <em>work? The module imports fine but
there is no read function!</em></p>
<p>You need Biopython 1.45 or later. Or, use
<tt class="docutils literal"><span class="pre">Bio.SeqIO.parse(...).next()</span></tt> instead.</p>
</li>
<li><p class="first"><em>Why isn’t</em> <tt class="docutils literal"><span class="pre">Bio.AlignIO</span></tt> <em>present? The module import fails!</em></p>
<p>You need Biopython 1.46 or later.</p>
</li>
<li><p class="first"><em>What file formats do</em> <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> <em>and</em> <tt class="docutils literal"><span class="pre">Bio.AlignIO</span></tt> <em>read and
write?</em></p>
<p>Check the built in docstrings (<tt class="docutils literal"><span class="pre">from</span> <span class="pre">Bio</span> <span class="pre">import</span> <span class="pre">SeqIO</span></tt>, then
<tt class="docutils literal"><span class="pre">help(SeqIO)</span></tt>), or see
<a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a>
and
<a class="reference external" href="http://biopython.org/wiki/AlignIO">http://biopython.org/wiki/AlignIO</a>
on the wiki for the latest listing.</p>
</li>
<li><p class="first"><em>Why don’t the</em> <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> <em>and</em> <tt class="docutils literal"><span class="pre">Bio.AlignIO</span></tt> <em>input functions
let me provide a sequence alphabet?</em></p>
<p>You need Biopython 1.49 or later.</p>
</li>
<li><p class="first"><em>Why won’t the</em> <tt class="docutils literal"><span class="pre">Bio.SeqIO</span></tt> <em>and</em> <tt class="docutils literal"><span class="pre">Bio.AlignIO</span></tt> <em>functions</em>
<tt class="docutils literal"><span class="pre">parse</span></tt><em>,</em> <tt class="docutils literal"><span class="pre">read</span></tt> <em>and</em> <tt class="docutils literal"><span class="pre">write</span></tt> <em>take filenames? They insist
on handles!</em></p>
<p>You need Biopython 1.54 or later, or just use handles explicitly
(see Section <a class="reference external" href="#sec:appendix-handles">22.1</a>). It is especially
important to remember to close output handles explicitly after
writing your data.</p>
</li>
<li><p class="first"><em>Why won’t the</em> <tt class="docutils literal"><span class="pre">Bio.SeqIO.write()</span></tt> <em>and</em> <tt class="docutils literal"><span class="pre">Bio.AlignIO.write()</span></tt>
<em>functions accept a single record or alignment? They insist on a list
or iterator!</em></p>
<p>You need Biopython 1.54 or later, or just wrap the item with
<tt class="docutils literal"><span class="pre">[...]</span></tt> to create a list of one element.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal"><span class="pre">str(...)</span></tt> <em>give me the full sequence of a</em> <tt class="docutils literal"><span class="pre">Seq</span></tt>
<em>object?</em></p>
<p>You need Biopython 1.45 or later. Alternatively, rather than
<tt class="docutils literal"><span class="pre">str(my_seq)</span></tt>, use <tt class="docutils literal"><span class="pre">my_seq.tostring()</span></tt> (which will also work on
recent versions of Biopython).</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal"><span class="pre">Bio.Blast</span></tt> <em>work with the latest plain text NCBI
blast output?</em></p>
<p>The NCBI keep tweaking the plain text output from the BLAST tools,
and keeping our parser up to date is/was an ongoing struggle. If you
aren’t using the latest version of Biopython, you could try
upgrading. However, we (and the NCBI) recommend you use the XML
output instead, which is designed to be read by a computer program.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal"><span class="pre">Bio.Entrez.read()</span></tt> <em>work? The module imports fine
but there is no read function!</em></p>
<p>You need Biopython 1.46 or later.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal"><span class="pre">Bio.Entrez.parse()</span></tt> <em>work? The module imports fine
but there is no parse function!</em></p>
<p>You need Biopython 1.52 or later.</p>
</li>
<li><p class="first"><em>Why has my script using</em> <tt class="docutils literal"><span class="pre">Bio.Entrez.efetch()</span></tt> <em>stopped working?</em></p>
<p>This could be due to NCBI changes in February 2012 introducing
EFetch 2.0. First, they changed the default return modes - you
probably want to add <tt class="docutils literal"><span class="pre">retmode=&quot;text&quot;</span></tt> to your call. Second, they
are now stricter about how to provide a list of IDs – Biopython 1.59
onwards turns a list into a comma separated string automatically.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal"><span class="pre">Bio.Blast.NCBIWWW.qblast()</span></tt> <em>give the same results
as the NCBI BLAST website?</em></p>
<p>You need to specify the same options – the NCBI often adjust the
default settings on the website, and they do not match the QBLAST
defaults anymore. Check things like the gap penalties and expectation
threshold.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal"><span class="pre">Bio.Blast.NCBIXML.read()</span></tt> <em>work? The module imports
but there is no read function!</em></p>
<p>You need Biopython 1.50 or later. Or, use
<tt class="docutils literal"><span class="pre">Bio.Blast.NCBIXML.parse(...).next()</span></tt> instead.</p>
</li>
<li><p class="first"><em>Why doesn’t my</em> <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> <em>object have a</em> <tt class="docutils literal"><span class="pre">letter_annotations</span></tt>
<em>attribute?</em></p>
<p>Per-letter-annotation support was added in Biopython 1.50.</p>
</li>
<li><p class="first"><em>Why can’t I slice my</em> <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> <em>to get a sub-record?</em></p>
<p>You need Biopython 1.50 or later.</p>
</li>
<li><p class="first"><em>Why can’t I add</em> <tt class="docutils literal"><span class="pre">SeqRecord</span></tt> <em>objects together?</em></p>
<p>You need Biopython 1.53 or later.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal"><span class="pre">Bio.SeqIO.convert()</span></tt> <em>or</em> <tt class="docutils literal"><span class="pre">Bio.AlignIO.convert()</span></tt>
<em>work? The modules import fine but there is no convert function!</em></p>
<p>You need Biopython 1.52 or later. Alternatively, combine the
<tt class="docutils literal"><span class="pre">parse</span></tt> and <tt class="docutils literal"><span class="pre">write</span></tt> functions as described in this tutorial (see
Sections <a class="reference external" href="#sec:SeqIO-conversion">5.5.2</a>
and <a class="reference external" href="#sec:converting-alignments">6.2.1</a>).</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal"><span class="pre">Bio.SeqIO.index()</span></tt> <em>work? The module imports fine
but there is no index function!</em></p>
<p>You need Biopython 1.52 or later.</p>
</li>
<li><p class="first"><em>Why doesn’t</em> <tt class="docutils literal"><span class="pre">Bio.SeqIO.index_db()</span></tt> <em>work? The module imports fine
but there is no</em> <tt class="docutils literal"><span class="pre">index_db</span></tt> <em>function!</em></p>
<p>You need Biopython 1.57 or later (and a Python with SQLite3
support).</p>
</li>
<li><p class="first"><em>Where is the</em> <tt class="docutils literal"><span class="pre">MultipleSeqAlignment</span></tt> <em>object? The</em> <tt class="docutils literal"><span class="pre">Bio.Align</span></tt>
<em>module imports fine but this class isn’t there!</em></p>
<p>You need Biopython 1.54 or later. Alternatively, the older
<tt class="docutils literal"><span class="pre">Bio.Align.Generic.Alignment</span></tt> class supports some of its
functionality, but using this is now discouraged.</p>
</li>
<li><p class="first"><em>Why can’t I run command line tools directly from the application
wrappers?</em></p>
<p>You need Biopython 1.55 or later. Alternatively, use the Python
<tt class="docutils literal"><span class="pre">subprocess</span></tt> module directly.</p>
</li>
<li><p class="first"><em>I looked in a directory for code, but I couldn’t find the code that
does something. Where’s it hidden?</em></p>
<p>One thing to know is that we put code in <tt class="docutils literal"><span class="pre">__init__.py</span></tt> files. If
you are not used to looking for code in this file this can be
confusing. The reason we do this is to make the imports easier for
users. For instance, instead of having to do a “repetitive” import
like <tt class="docutils literal"><span class="pre">from</span> <span class="pre">Bio.GenBank</span> <span class="pre">import</span> <span class="pre">GenBank</span></tt>, you can just use
<tt class="docutils literal"><span class="pre">from</span> <span class="pre">Bio</span> <span class="pre">import</span> <span class="pre">GenBank</span></tt>.</p>
</li>
<li><p class="first"><em>Why does the code from CVS seem out of date?</em></p>
<p>In late September 2009, just after the release of Biopython 1.52, we
switched from using CVS to git, a distributed version control system.
The old CVS server will remain available as a static and read only
backup, but if you want to grab the latest code, you’ll need to use
git instead. See our website for more details.</p>
</li>
</ol>
<p>For more general questions, the Python FAQ pages
<a class="reference external" href="http://www.python.org/doc/faq/">http://www.python.org/doc/faq/</a>
may be useful.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Chapter 1  Introduction</a><ul>
<li><a class="reference internal" href="#what-is-biopython">1.1  What is Biopython?</a></li>
<li><a class="reference internal" href="#what-can-i-find-in-the-biopython-package">1.2  What can I find in the Biopython package</a></li>
<li><a class="reference internal" href="#installing-biopython">1.3  Installing Biopython</a></li>
<li><a class="reference internal" href="#frequently-asked-questions-faq">1.4  Frequently Asked Questions (FAQ)</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to biopython_cn&#8217;s documentation!</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="chr02.html"
                        title="next chapter">Chapter 2  Quick Start – What can you do with Biopython?</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/chr01.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="chr02.html" title="Chapter 2 Quick Start – What can you do with Biopython?"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to biopython_cn’s documentation!"
             >previous</a> |</li>
        <li><a href="index.html">biopython_en 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, biopythoners.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.
    </div>
  </body>
</html>