<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.11: http://docutils.sourceforge.net/" />
<title></title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 7614 2013-02-21 15:55:51Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

object[type="image/svg+xml"], object[type="application/x-shockwave-flash"] {
  overflow: hidden;
}

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title, .code .error {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left, object.align-left {
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right, object.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

img.align-center, .figure.align-center, object.align-center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: inherit }

/* div.align-center * { */
/*   text-align: left } */

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block, pre.math, pre.code {
  margin-left: 2em ;
  margin-right: 2em }

pre.code .ln { color: grey; } /* line numbers */
pre.code, code { background-color: #eeeeee }
pre.code .comment, code .comment { color: #5C6576 }
pre.code .keyword, code .keyword { color: #3B0D06; font-weight: bold }
pre.code .literal.string, code .literal.string { color: #0C5404 }
pre.code .name.builtin, code .name.builtin { color: #352B84 }
pre.code .deleted, code .deleted { background-color: #DEB0A1}
pre.code .inserted, code .inserted { background-color: #A3D289}

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

/* "booktabs" style (no vertical lines) */
table.docutils.booktabs {
  border: 0px;
  border-top: 2px solid;
  border-bottom: 2px solid;
  border-collapse: collapse;
}
table.docutils.booktabs * {
  border: 0px;
}
table.docutils.booktabs th {
  border-bottom: thin solid;
  text-align: left;
}

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
<style type="text/css">

/*
*   math2html: convert LaTeX equations to HTML output.
*
*   Copyright (C) 2009,2010 Alex Fernández
*
*   Released under the terms of the `2-Clause BSD license'_, in short:
*   Copying and distribution of this file, with or without modification,
*   are permitted in any medium without royalty provided the copyright
*   notice and this notice are preserved.
*   This file is offered as-is, without any warranty.
*
* .. _2-Clause BSD license: http://www.spdx.org/licenses/BSD-2-Clause
*
*   Based on eLyXer: convert LyX source files to HTML output.
*   http://elyxer.nongnu.org/
*/
/* --end--
* CSS file for LaTeX formulas.
*/

/* Formulas */
.formula {
	text-align: center;
	font-family: "Droid Serif", "DejaVu Serif", "STIX", serif;
	margin: 1.2em 0;
}
span.formula {
	white-space: nowrap;
}
div.formula {
	padding: 0.5ex;
	margin-left: auto;
	margin-right: auto;
}

/* Basic features */
a.eqnumber {
	display: inline-block;
	float: right;
	clear: right;
	font-weight: bold;
}
span.unknown {
	color: #800000;
}
span.ignored, span.arraydef {
	display: none;
}
.formula i {
	letter-spacing: 0.1ex;
}

/* Alignment */
.align-left, .align-l {
	text-align: left;
}
.align-right, .align-r {
	text-align: right;
}
.align-center, .align-c {
	text-align: center;
}

/* Structures */
span.overline, span.bar {
	text-decoration: overline;
}
.fraction, .fullfraction {
	display: inline-block;
	vertical-align: middle;
	text-align: center;
}
.fraction .fraction {
	font-size: 80%;
	line-height: 100%;
}
span.numerator {
	display: block;
}
span.denominator {
	display: block;
	padding: 0ex;
	border-top: thin solid;
}
sup.numerator, sup.unit {
	font-size: 70%;
	vertical-align: 80%;
}
sub.denominator, sub.unit {
	font-size: 70%;
	vertical-align: -20%;
}
span.sqrt {
	display: inline-block;
	vertical-align: middle;
	padding: 0.1ex;
}
sup.root {
	font-size: 70%;
	position: relative;
	left: 1.4ex;
}
span.radical {
	display: inline-block;
	padding: 0ex;
	font-size: 150%;
	vertical-align: top;
}
span.root {
	display: inline-block;
	border-top: thin solid;
	padding: 0ex;
	vertical-align: middle;
}
span.symbol {
	line-height: 125%;
	font-size: 125%;
}
span.bigsymbol {
	line-height: 150%;
	font-size: 150%;
}
span.largesymbol {
	font-size: 175%;
}
span.hugesymbol {
	font-size: 200%;
}
span.scripts {
	display: inline-table;
	vertical-align: middle;
}
.script {
	display: table-row;
	text-align: left;
	line-height: 150%;
}
span.limits {
	display: inline-table;
	vertical-align: middle;
}
.limit {
	display: table-row;
	line-height: 99%;
}
sup.limit, sub.limit {
	line-height: 100%;
}
span.symbolover {
	display: inline-block;
	text-align: center;
	position: relative;
	float: right;
	right: 100%;
	bottom: 0.5em;
	width: 0px;
}
span.withsymbol {
	display: inline-block;
}
span.symbolunder {
	display: inline-block;
	text-align: center;
	position: relative;
	float: right;
	right: 80%;
	top: 0.3em;
	width: 0px;
}

/* Environments */
span.array, span.bracketcases, span.binomial, span.environment {
	display: inline-table;
	text-align: center;
	border-collapse: collapse;
	margin: 0em;
	vertical-align: middle;
}
span.arrayrow, span.binomrow {
	display: table-row;
	padding: 0ex;
	border: 0ex;
}
span.arraycell, span.bracket, span.case, span.binomcell, span.environmentcell {
	display: table-cell;
	padding: 0ex 0.2ex;
	line-height: 99%;
	border: 0ex;
}
/*
* CSS file for LaTeX formulas, extra stuff:
* binomials, vertical braces, stackrel, fonts and colors.
*/

/* Inline binomials */
span.binom {
	display: inline-block;
	vertical-align: middle;
	text-align: center;
	font-size: 80%;
}
span.binomstack {
	display: block;
	padding: 0em;
}

/* Over- and underbraces */
span.overbrace {
	border-top: 2pt solid;
}
span.underbrace {
	border-bottom: 2pt solid;
}

/* Stackrel */
span.stackrel {
	display: inline-block;
	text-align: center;
}
span.upstackrel {
	display: block;
	padding: 0em;
	font-size: 80%;
	line-height: 64%;
	position: relative;
	top: 0.15em;

}
span.downstackrel {
	display: block;
	vertical-align: bottom;
	padding: 0em;
}

/* Fonts */
span.mathsf, span.textsf {
	font-style: normal;
	font-family: sans-serif;
}
span.mathrm, span.textrm {
	font-style: normal;
	font-family: serif;
}
span.text, span.textnormal {
	font-style: normal;
}
span.textipa {
	color: #008080;
}
span.fraktur {
	font-family: "Lucida Blackletter", eufm10, blackletter;
}
span.blackboard {
	font-family: Blackboard, msbm10, serif;
}
span.scriptfont {
	font-family: "Monotype Corsiva", "Apple Chancery", "URW Chancery L", cursive;
	font-style: italic;
}

/* Colors */
span.colorbox {
	display: inline-block;
	padding: 5px;
}
span.fbox {
	display: inline-block;
	border: thin solid black;
	padding: 2px;
}
span.boxed, span.framebox {
	display: inline-block;
	border: thin solid black;
	padding: 5px;
}


</style>
</head>
<body>
<div class="document">


<div class="section" id="id1">
<h1>第1章 介绍</h1>
<div class="section" id="biopython">
<h2>1.1  什么是Biopython？</h2>
<p>Biopython工程是一个使用Python来开发计算分子生物学工具的国际团体。(<a class="reference external" href="http://www.python.org">http://www.python.org</a>)
Python是一种面向对象的、解释型的、灵活的语言，在计算机科学中日益流行。Python易学，语法明晰，并且能很容易的使用以C，C++或
者FORTRAN编写的模块实现扩展。</p>
<p>Biopython官网(<a class="reference external" href="http://www.biopython.org">http://www.biopython.org</a>)为使用和研究生物信息学的开发者提供了一个在线的
资源库，包括模块、脚本以及一些基于Python的软件的网站链接。一般来讲，Biopython致力于通过创造高质量的和可重复利用的模块及
类，从而使得Python在生物信息学中的应用变得更加容易。Biopython的特点包括解析各种生物信息学格式的文件(BLAST， Clustalw， FASTA，
Genbank...)，访问在线的服务器(NCBI，Expasy...)，常见和不那么常见程序的接口(Clustalw， DSSP，MSMS...)，标准的序列类，各
种收集的模块，KD树数据结构等等，还有一些文档。</p>
<p>基本来说，我们喜欢使用Python来编程，并且希望通过创建高质量、可复用的模块和脚本来使得Python在生物信息学中的应用变得容易。</p>
</div>
<div class="section" id="id2">
<h2>1.2  在Biopython包中我能发现什么？</h2>
<p>主要的Biopython发行版本有很多种功能，包括：</p>
<ul class="simple">
<li>将生物信息学文件解析为Python可用的数据结构，包含以下支持的格式：<ul>
<li>Blast输出结果 – standalone和在线Blast</li>
<li>Clustalw</li>
<li>FASTA</li>
<li>GenBank</li>
<li>PubMed和Medline</li>
<li>ExPASy文件, 如Enzyme和Prosite</li>
<li>SCOP, 包括‘dom’和‘lin’文件</li>
<li>UniGene</li>
<li>SwissProt</li>
</ul>
</li>
<li>被支持格式的文件可以通过记录来重复或者通过字典界面来索引。</li>
<li>处理常见的生物信息学在线数据库的代码：<ul>
<li>NCBI – Blast, Entrez和PubMed服务</li>
<li>ExPASy – Swiss-Prot和Prosite条目, 包括Prosite搜索</li>
</ul>
</li>
<li>常见生物信息学程序的接口，例如：<ul>
<li>NCBI的Standalone Blast</li>
<li>Clustalw比对程序</li>
<li>EMBOSS命令行工具</li>
</ul>
</li>
<li>一个能处理序列、ID和序列特征的标准序列类。</li>
<li>对序列实现常规操作的工具，如翻译，转录和权重计算。</li>
<li>利用k最近邻接、Bayes或SVM对数据进行分类的代码。</li>
<li>处理比对的代码，包括创建和处理替换矩阵的标准方法。</li>
<li>分发并行任务到不同进程的代码。</li>
<li>实现序列的基本操作，翻译以及BLAST等功能的GUI程序。</li>
<li>使用这些模块的详细文档和帮助，包括此文件，在线的wiki文档，网站和邮件列表。</li>
<li>整合BioSQL，一个也被BioPerl和BioJava支持的数据库架构。</li>
</ul>
<p>我们希望这些能给你足够的理由去下载并开始使用Biopython！</p>
</div>
<div class="section" id="id3">
<h2>1.3  安装Biopython</h2>
<p>Biopython的所有安装信息在此文档中分开，以便于更容易保持更新。</p>
<p>简短的版本去我们的下载页面(<a class="reference external" href="http://biopython.org/wiki/Download">http://biopython.org/wiki/Download</a>),
下载并安装所列举的dependencies，然后下载并安装Biopython。Biopython能在多种平台上运行（Windows，Mac，各种版本的Linux和Unix）。
对于Windows我们提供预编译的一键式安装程序，而对Unix和其他操作系统，你必须按照附带的README文件从源开始安装。这通常
是很简单的，只要标准命令：</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">setup</span><span class="operator">.</span><span class="name">py</span> <span class="name">build</span>
<span class="name">python</span> <span class="name">setup</span><span class="operator">.</span><span class="name">py</span> <span class="name">test</span>
<span class="name">sudo</span> <span class="name">python</span> <span class="name">setup</span><span class="operator">.</span><span class="name">py</span> <span class="name">install</span>
</pre>
<p>（事实上你可以跳过build和test，直接install。但最好是确保所有的东西看起来都没问题。）</p>
<p>我们的安装说明的详细版本包括了Python的安装，Biopython dependencies的安装以及Biopython本身的安装。可从PDF
(<a class="reference external" href="http://biopython.org/DIST/docs/install/Installation.pdf">http://biopython.org/DIST/docs/install/Installation.pdf</a>)
和HTML格式获得。
(<a class="reference external" href="http://biopython.org/DIST/docs/install/Installation.html">http://biopython.org/DIST/docs/install/Installation.html</a>)。</p>
</div>
<div class="section" id="faq">
<h2>1.4  常见问答（FAQ）</h2>
<ol class="arabic">
<li><p class="first"><em>在科学出版中我怎样引用Biopython？</em></p>
<p>请引用我们的应用笔记 [<a class="reference external" href="#cock2009">1</a>, Cock <em>et al.</em> ,  2009] 作为主要的Biopython参考文献。另外，如果可以，请
引用以下任意出版物，特别是作为Biopython特定模块的参考文献的话。
（更多信息可在我们网站上获得）：</p>
<ul class="simple">
<li>对于官方项目声明: [<a class="reference external" href="#chapman2000">13</a>,
Chapman and Chang, 2000];</li>
<li>对于 <tt class="docutils literal">Bio.PDB</tt>: [<a class="reference external" href="#hamelryck2003a">18</a>, Hamelryck and
Manderick, 2003];</li>
<li>对于 <tt class="docutils literal">Bio.Cluster</tt>: [<a class="reference external" href="#dehoon2004">14</a>, De Hoon <em>et al.</em>,
2004];</li>
<li>对于 <tt class="docutils literal">Bio.Graphics.GenomeDiagram</tt>: [<a class="reference external" href="#pritchard2006">2</a>,
Pritchard <em>et al.</em>, 2006];</li>
<li>对于 <tt class="docutils literal">Bio.Phylo</tt> 和 <tt class="docutils literal">Bio.Phylo.PAML</tt>: [<a class="reference external" href="#talevich2012">9</a>,
Talevich <em>et al.</em>, 2012];</li>
<li>对于在Biopython，BioPerl，BioRuby，BioJava和EMBOSS支持的FASTQ格
式文件：[<a class="reference external" href="#cock2010">7</a>, Cock <em>et al.</em>, 2010].</li>
</ul>
</li>
<li><p class="first"><em>我该怎样以大写字母写“Biopython”？写成“BioPython”可以吗？</em></p>
<p>正确的大写是“Biopython”而不是“BioPython”（虽然对于BioPerl，BioRuby
和BioJava是这样）。</p>
</li>
<li><p class="first"><em>我怎样查看自己安装的Biopython的版本？</em></p>
<p>使用以下代码：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">Bio</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">Bio</span><span class="operator">.</span><span class="name">__version__</span>
<span class="operator">...</span>
</pre>
<p>如果 “<tt class="docutils literal">import Bio</tt>” 这行报错，说明Biopython未被安装。如果第二行报错，
你的版本已经很过时了。如果版本号以“+”号结束，说明你用的并不是官方版本，而
是开发代码的快照。</p>
</li>
<li><p class="first"><em>此文档的最新版本在哪里？</em></p>
<p>如果你下载的是一个Biopython源代码包，那么它将包含此文档HTML和PDF两种格式
的相应版本。此文档最新出版的版本可通过在线获得（每个版本的更新）：</p>
<ul class="simple">
<li><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html">http://biopython.org/DIST/docs/tutorial/Tutorial.html</a></li>
<li><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.pdf">http://biopython.org/DIST/docs/tutorial/Tutorial.pdf</a></li>
</ul>
<p>如果你使用的是从我们库中获得的尚未发布的最新代码，你可以在这里找到还在开发中
的教程的拷贝：</p>
<ul class="simple">
<li><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial-dev.html">http://biopython.org/DIST/docs/tutorial/Tutorial-dev.html</a></li>
<li><a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial-dev.pdf">http://biopython.org/DIST/docs/tutorial/Tutorial-dev.pdf</a></li>
</ul>
</li>
<li><p class="first"><em>我需要哪一个“Numerical Python”？</em></p>
<p>对于Biopython 1.48或更早的版本，你需要老的Numeric模块。对于Biopython 1.49
及更高的版本，你需要更新的NumPy来代替。Numeric和NumPy都可以在同一台机器上安
装。也可以访问： <a class="reference external" href="http://numpy.scipy.org/">http://numpy.scipy.org/</a></p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Seq</tt> <em>对象缺少了这篇教程里的（反向）transcription和translation方法？</em></p>
<p>你需要Biopython 1.49或更新的版本。或者，使用以下 <a class="reference external" href="#sec:seq-module-functions">3.14</a> 部分中的 <tt class="docutils literal">Bio.Seq</tt> 模块
功能。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Seq</tt> <em>对象缺少了这篇教程中的upper和lower方法？</em></p>
<p>你需要Biopython 1.53或更新版本。或者，使用 <tt class="docutils literal"><span class="pre">str(my_seq).upper()</span></tt> 来获得
大写字符串。如果你需要一个Seq对象，试试 <tt class="docutils literal"><span class="pre">Seq(str(my_seq).upper())</span></tt> ，但是
要小心重用相同的字母。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Seq</tt> <em>对象的translation方法不支持本教程中描述的</em> <tt class="docutils literal">cds</tt> <em>选项？</em></p>
<p>你需要Biopython 1.51或更新版本。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Bio.SeqIO</tt> <em>不能正常工作？它导入正常但是没有解析函数等。</em></p>
<p>你需要Biopython 1.43或更新版本。较老的版本确实包含了一些相关的代码在 <tt class="docutils literal">Bio.SeqIO</tt> 下面但是后来就被移除了——这就是为什么import是正常的。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Bio.SeqIO.read()</tt> <em>不能正常工作？该模块导入正常但是并没有read函数！</em></p>
<p>你需要Biopython 1.45或更新的版本。或者，使用 <tt class="docutils literal"><span class="pre">Bio.SeqIO.parse(...).next()</span></tt> 来代替。</p>
</li>
<li><p class="first"><em>为什么没有</em> <tt class="docutils literal">Bio.AlignIO</tt> <em>？模块导入失败！</em></p>
<p>你需要Biopython 1.46或更新的版本。</p>
</li>
<li><p class="first"><tt class="docutils literal">Bio.SeqIO</tt> <em>和</em> <tt class="docutils literal">Bio.AlignIO</tt> <em>读写什么样的文件格式？</em></p>
<p>请检查内建文档（<tt class="docutils literal">from Bio import SeqIO</tt>，然后 <tt class="docutils literal">help(SeqIO)</tt> ），或见wiki上的最
新条目：
<a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a>
以及
<a class="reference external" href="http://biopython.org/wiki/AlignIO">http://biopython.org/wiki/AlignIO</a></p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Bio.SeqIO</tt> <em>和</em> <tt class="docutils literal">Bio.AlignIO</tt> <em>的input函数不让我提供一个序列字母？</em></p>
<p>你需要Biopython 1.49或更新版本。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Bio.SeqIO</tt> <em>和</em> <tt class="docutils literal">Bio.AlignIO</tt> <em>函数</em> <tt class="docutils literal">parse</tt> <em>，</em> <tt class="docutils literal">read</tt> <em>和</em> <tt class="docutils literal">write</tt> <em>不能使用文件名？它们坚持句柄！</em></p>
<p>你需要Biopython 1.54或更新的版本。或者明确使用句柄。
(见 Section <a class="reference external" href="#sec:appendix-handles">22.1</a>). 一定要记得当你写完数据后关闭输
出句柄。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Bio.SeqIO.write()</tt> <em>和</em> <tt class="docutils literal">Bio.AlignIO.write()</tt> <em>函数不接受单个记录
或比对？它们坚持需要一个列表或迭代器！</em></p>
<p>你需要Biopython 1.54或更新版本，或将该条目以 <tt class="docutils literal"><span class="pre">[...]</span></tt> 包起来形成一个单元素的列表。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal"><span class="pre">str(...)</span></tt> <em>不给我一个</em> <tt class="docutils literal">Seq</tt> <em>对象的全序列？</em></p>
<p>你需要Biopython 1.45或更新的版本。或者，与其使用 <tt class="docutils literal">str(my_seq)</tt>，不如试试 <tt class="docutils literal">my_seq.tostring()</tt> 这也能在最近的Biopython版本上工作）。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Bio.Blast</tt> <em>不能处理最新的NCBI blast输出文本文件结果？</em></p>
<p>NCBI在不断的调整BLAST工具的纯文本输出，导致我们的解析器需要不断更新。
如果你没使用最新版本的Biopython，你可以试试升级。但是，我们（还有NCBI）推荐你使用
HTML格式输出来代替，因为HTML是设计给电脑程序读取的。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Bio.Entrez.read()</tt> <em>不能正常工作？模块导入正常但是没有read函数！</em></p>
<p>你需要Biopython 1.46或更新的版本。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Bio.Entrez.parse()</tt> <em>不能正常工作？模块导入正常但是没有parse函数！</em></p>
<p>你需要Biopython 1.52或更新的版本。</p>
</li>
<li><p class="first"><em>为什么我的脚本使用了</em> <tt class="docutils literal">Bio.Entrez.efetch()</tt> <em>便停止工作了？</em></p>
<p>这可能是由于NCBI在2012年2月引进EFetch 2.0后发生了改变。首先，他们改变了默认的返回方式——
你可能想添加 <tt class="docutils literal"><span class="pre">retmode=&quot;text&quot;</span></tt> 到你的call。其次，他们对于怎么提供一个ID列表变得更加严格——
Biopython 1.59及之后版本或自动将一个列表转换成逗号分隔的字符串。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Bio.Blast.NCBIWWW.qblast()</tt> <em>没有给出与NCBI BLAST网站上相同的结果？</em></p>
<p>你需要指定相同的选项——NCBI经常调整网站上的默认设置，并且他们不再匹配QBLAST的默认设置了。
请检查gap罚分和期望值阈值。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Bio.Blast.NCBIXML.read()</tt> <em>不正常工作？模块导入了但是没有read函数！</em></p>
<p>你需要Biopython 1.50或更新的版本。或者，使用 <tt class="docutils literal"><span class="pre">Bio.Blast.NCBIXML.parse(...).next()</span></tt> 代替。</p>
</li>
<li><p class="first"><em>为什么我的</em> <tt class="docutils literal">SeqRecord</tt> <em>对象没有一个</em> <tt class="docutils literal">letter_annotations</tt> <em>的属性？</em></p>
<p>Per-letter-annotation已经被加入到Biopython 1.50中。</p>
</li>
<li><p class="first"><em>为什么我无法切片我的</em> <tt class="docutils literal">SeqRecord</tt> <em>来获取一个子记录？</em>
你需要Biopython 1.50或更新版本。</p>
</li>
<li><p class="first"><em>为什么我无法一起添加</em> <tt class="docutils literal">SeqRecord</tt> <em>对象？</em></p>
<p>你需要Biopython 1.53或更新版本。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Bio.SeqIO.convert()</tt> <em>或</em> <tt class="docutils literal">Bio.AlignIO.convert()</tt> <em>不能正常工作？模块导入
正常但是没有convert函数！</em></p>
<p>你需要Biopython 1.52或更新版本。或者，按以下教程中描述的结合 <tt class="docutils literal">parse</tt> 和 <tt class="docutils literal">write</tt> 函数。
（见 Sections <a class="reference external" href="#sec:SeqIO-conversion">5.5.2</a> 和 <a class="reference external" href="#sec:converting-alignments">6.2.1</a>）。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Bio.SeqIO.index()</tt> <em>不能正常工作？模块导入正常但是没有index函数！</em></p>
<p>你需要Biopython 1.52或更新版本。</p>
</li>
<li><p class="first"><em>为什么</em> <tt class="docutils literal">Bio.SeqIO.index_db()</tt> <em>不能正常工作？模块导入正常但是没有</em> <tt class="docutils literal">index_db</tt> <em>函数！</em></p>
<p>你需要Biopython 1.57或更新版本。（有SQLite3的Python支持）</p>
</li>
<li><p class="first"><tt class="docutils literal">MultipleSeqAlignment</tt> <em>对象在哪里？</em> <tt class="docutils literal">Bio.Align</tt> <em>模块导入正常但是这个类不在那里！</em></p>
<p>你需要Biopython 1.54或更新版本。或者，较早的 <tt class="docutils literal">Bio.Align.Generic.Alignment</tt> 类支持它的一些功能，
但是现在不推荐使用这个。</p>
</li>
<li><p class="first"><em>为什么我不能直接从应用程序包装器上运行命令行工具？</em></p>
<p>你需要Biopython 1.55或更新版本。或者，直接使用Python的 <tt class="docutils literal">subprocess</tt> 模块。</p>
</li>
<li><p class="first"><em>我看到过一个代码的目录，但是我找不到那个能干嘛的代码了。它藏在哪儿了？</em></p>
<p>我们知道，我们的代码存放在 <tt class="docutils literal">__init__.py</tt> 文件里。如果你此前没有在这个文件里寻找代码那么这可能会
让人困惑。我们这样做的原因是为了让用户更容易导入。比如，不一定要像 <tt class="docutils literal">from Bio.GenBank import GenBank</tt>
来导入一个“repetitive”，你仅需使用 <tt class="docutils literal">from Bio import GenBank</tt> 就行。</p>
</li>
<li><p class="first"><em>为什么CVS的代码貌似过期了？</em></p>
<p>2009年9月下旬，在Biopython 1.52发布之后，我们从使用CVS转变为使用git，git是一个分散式的版本控制系统。
旧的CVS服务仍可作为静态和只读备份，但是如果你想获取最新的代码，你需要使用git。详见我们的网站获取更多
信息。</p>
</li>
</ol>
<p>对于更一般的问题，Python FAQ页面 <a class="reference external" href="http://www.python.org/doc/faq/">http://www.python.org/doc/faq/</a>
可能会有帮助。</p>
</div>
</div>
<div class="section" id="id4">
<h1>第2章 快速开始 —— 你能用Biopython做什么？</h1>
<p>此部分旨在能让你快速开始Biopython，并给你一个大概的了解什么可用以及如何使用它。
此部分的所有例子都会假设你有Python的基础知识，并且前提是你已经在你系统上
安装了Biopython。如果你认为你需要认真复习Python，主流的Python网站提供了相当多
的免费文档，你可以从以下网站开始(<a class="reference external" href="http://www.python.org/doc/">http://www.python.org/doc/</a>)。</p>
<p>由于计算机上大量的生物学工作涉及到网上的数据库，某些例子也会需要联网才能完成。</p>
<p>让我们看看我们能用Biopython做什么吧。</p>
<div class="section" id="id5">
<h2>2.1  Biopython功能概览</h2>
<p>正如介绍中提到的，Biopython是一个库的集合，这个库能在计算机上工作的生物学家解决感兴趣的事情。一般来说，你至少需要一点编程经验（当然是Python！）或至少
有兴趣学习编程。Biopython的任务就是通过提供可重复利用的库，让编程人员的工作变得更加容易，可以使你能集中精力解决你所感兴趣的问题，而不用花太多精力去完成一个解析特殊文件格式的构件（当然，如果你想帮我们写一个原本不存在的解析器并把它贡献给Biopython，请继续！）。所以Biopython的工作是让你更加轻松！</p>
<p>值得一提的是，Biopython通常能给出多种方式来解决“相同的事情”。在最近的版本中，
情况有所改善，但这仍可让人沮丧，因为在理想的Python中应该只有一种正确的方式
去解决问题。但是，这也可以成为一个真正的好处，因为它给了你很多灵活性和对库的
控制。本教程给你展示普通的或简单的方式去处理问题以便于你能自己处理事情。想要
学习更多替代的方法，请查看Cookbook（第 <a class="reference external" href="#chapter:cookbook">18</a> 章,
这里有一些很酷的技巧和提示），进阶部分(第 <a class="reference external" href="#chapter:advanced">20</a> 章)，
内建“文档”（通过Python help命令），或者 <a class="reference external" href="http://biopython.org/DIST/docs/api/">API 文档</a> )
或者代码本身。</p>
</div>
<div class="section" id="id6">
<h2>2.2  处理序列</h2>
<p>生物信息学的中心对象是序列。因此，我们先快速开始介绍一下Biopython处理序列的机制，主要是 <tt class="docutils literal">Seq</tt> 对象，这个我们也将会在第 <a class="reference external" href="#chapter:Bio.Seq">3</a> 章中详细讨论。</p>
<p>大多数时候当我们想到一条序列时，在我们脑海中都会有一串类似于‘<tt class="docutils literal">AGTACACTGGT</tt>’的
字母串。你可以按以下步骤创建一个 <tt class="docutils literal">Seq</tt> 对象——“<tt class="docutils literal">&gt;&gt;&gt;</tt>”表示Python提示符后紧跟你要
输入的内容：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AGTACACTGGT&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGTACACTGGT'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_seq</span>
<span class="name">AGTACACTGGT</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">alphabet</span>
<span class="name">Alphabet</span><span class="punctuation">()</span>
</pre>
<p>这里是一个由 <em>通用</em> 字母表组成的序列对象——说明我们还 <em>没有</em> 指定它是一条DNA还是蛋白质
序列（好吧，一个有很多的Ala，Gly，Cys和Thr蛋白质序列！（幽默））。在第 <a class="reference external" href="#chapter:Bio.Seq">3</a> 章我们将讨论更
多关于字母表。</p>
<p>除了有一个字母表， <tt class="docutils literal">Seq</tt> 对象支持不同于Python的字符串方法。你不能对一个纯字符串做以下处理：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGTACACTGGT'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">complement</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TCATGTGACCA'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ACCAGTGTACT'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
</pre>
<p>另一个最重要的类是 <tt class="docutils literal">SeqRecord</tt> 或Sequence Record。它保留了一条序列（作为 <tt class="docutils literal">Seq</tt> 对象）
额外的注释信息，包括ID，name和description。用于读写序列文件格式的 <tt class="docutils literal">Bio.SeqIO</tt> 模块
能与 <tt class="docutils literal">SeqRecord</tt> 对象一起工作，稍后我们将会介绍，详细内容在第 <a class="reference external" href="#chapter:Bio.SeqIO">5</a> 章。</p>
<p>这涵盖了基本的功能和Biopython序列类的使用。现在你应该有一些想法像怎么和Biopython库互动，是时候去钻研它的乐趣，探讨处理生物学文件格式的有趣世界了！</p>
</div>
<div class="section" id="id7">
<h2>2.3  用法示例</h2>
<p>在我们跳到语法解析器和Biopython处理的其他所有事之前，让我们先建立一个例子来激励我们所做的事情并让生活变得更加有趣。毕竟，如果没有任何生物学在这篇教程里，那你为什么还要读它呢？</p>
<p>因为我喜欢植物，我想我们就来一个基于植物的例子吧（对其他生物的爱好者对不起了！）。刚刚结束了我们当地的一个温室的旅程，我们对Lady Slipper Orchids突然有一个难以置信
的迷恋（如果你想知道为什么，请看看一些 <a class="reference external" href="http://www.flickr.com/search/?q=lady+slipper+orchid&amp;s=int&amp;z=t">Lady Slipper Orchids的照片</a>，
并试试 <a class="reference external" href="http://images.google.com/images?q=lady%20slipper%20orchid">Google图片搜索</a>）。</p>
<p>当然，兰花不仅仅只有外观好看，它们也极力吸引着人们研究其进化和系统分类学。假设我们正在考虑写一份关于Lady Slipper Orchids进化研究的基金方案，我们就会想看看别人已经做了什么样的研究，然后我们能够增加一些什么内容。</p>
<p>经过一些阅读之后，我们发现Lady Slipper Orchids属于兰科拖鞋兰亚科并且由5个属组成：<em>Cypripedium</em>，<em>Paphiopedilum</em>，<em>Phragmipedium</em>，<em>Selenipedium</em> 和 <em>Mexipedium</em>。</p>
<p>这已经给了我们足够多的信息来探究更多的东西。现在，让我们看看Biopython工具能起到怎样的作用。
我们从一条从 <a class="reference external" href="#sec:sequence-parsing">2.4</a> 部分解析出来的序列开始， 但是我们稍后还是回到兰花上来——比如我们将在PubMed上搜索有关兰花的文章然后在GenBank上提取序列（第
<a class="reference external" href="#chapter:entrez">9</a> 章），从Swiss-Prot上提取特定的兰花蛋白数据（第<a class="reference external" href="#chapter:swiss_prot">10</a> 章），最后在<a class="reference external" href="#sec:align_clustal">6.4.1</a> 部分我们用ClustalW对兰花蛋白进行多序列比对。</p>
</div>
<div class="section" id="id8">
<h2>2.4  解析序列文件格式</h2>
<p>很多生物信息学工作的一大部分都会涉及到处理各种包含有生物学数据的文件格式类型。这些文件保存了有趣的生物学数据，因而一个特殊的挑战是需要将这些文件解析成你能使用某种编程语言操作的格式。然而这些解析工作有时会让人感到失望，因为这些格式有可能经常改变，而一个细微的改变也有可能让设计得最好的解析器失去作用。</p>
<p>我们现在开始简单地介绍 <tt class="docutils literal">Bio.SeqIO</tt> 模块——你可以在第<a class="reference external" href="#chapter:Bio.SeqIO">5</a> 章中查看更多。
我们从在线搜索我们的朋友——Lady Slipper Orchids——开始。为尽量保持简单，我们仅仅手动使用NCBI网站。我们先看看NCBI上的nucleotide库，使用在线的Entrez搜索
( <a class="reference external" href="http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?db=Nucleotide">http://www.ncbi.nlm.nih.gov:80/entrez/query.fcgi?db=Nucleotide</a> )
包含Cypripedioideae所有东西（这是Lady Slipper Orchids的亚科）。</p>
<p>当本教程最初编写时，这个搜索仅给我们找到了94条匹配的信息，我们将结果保存为FASTA格式文本文件和
GenBank格式文本文件（文件 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta">ls_orchid.fasta</a>
和 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk">ls_orchid.gbk</a>，
也包含在Biopython源代码包下 <tt class="docutils literal">docs/tutorial/examples/</tt> ）。</p>
<p>如果你现在搜索，你将会获得几百个的匹配结果！跟着教程，如果你想要看看相同的基因列表，请下载上面两个文件或者从Biopython源代码中拷贝 <tt class="docutils literal">docs/examples/</tt> 。在
<a class="reference external" href="#sec:connecting-with-biological-databases">2.5</a> 部分我们将会看到怎样使用Python做类似的搜索。</p>
<div class="section" id="fasta">
<h3>2.4.1  简单的FASTA解析示例</h3>
<p>如果你用你喜好的文本编辑器打开了lady slipper orchids的FASTA文件 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta">ls_orchid.fasta</a>，
你会看到文件开头像这样：</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">2765658</span><span class="operator">|</span><span class="name">emb</span><span class="operator">|</span><span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">CIZ78533</span> <span class="name">C</span><span class="operator">.</span><span class="name">irapeanum</span> <span class="literal number float">5.8</span><span class="name">S</span> <span class="name">rRNA</span> <span class="name">gene</span> <span class="operator word">and</span> <span class="name">ITS1</span> <span class="operator word">and</span> <span class="name">ITS2</span> <span class="name">DNA</span>
<span class="name">CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGGAATAAACGATCGAGTG</span>
<span class="name">AATCCGGAGGACCGGTGTACTCAGCTCACCGGGGGCATTGCTCCCGTGGTGACCCTGATTTGTTGTTGGG</span>
<span class="operator">...</span>
</pre>
<p>它包含有94条记录，每一行都以“<tt class="docutils literal">&gt;</tt>”开头，（大于号）紧随其后的是一行或多行序列。现在试试以下Python代码：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span>
    <span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
    <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="punctuation">)</span>
</pre>
<p>你应该会得到类似这样的一些东西出现在屏幕上：</p>
<pre class="code python literal-block">
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">2765658</span><span class="operator">|</span><span class="name">emb</span><span class="operator">|</span><span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">CIZ78533</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC'</span><span class="punctuation">,</span> <span class="name">SingleLetterAlphabet</span><span class="punctuation">())</span>
<span class="literal number integer">740</span>
<span class="operator">...</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">2765564</span><span class="operator">|</span><span class="name">emb</span><span class="operator">|</span><span class="name">Z78439</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">PBZ78439</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC'</span><span class="punctuation">,</span> <span class="name">SingleLetterAlphabet</span><span class="punctuation">())</span>
<span class="literal number integer">592</span>
</pre>
<p>注意FASTA文件并没有指定字母表，因此 <tt class="docutils literal">Bio.SeqIO</tt> 默认使用相当通用的 <tt class="docutils literal">SingleLetterAlphabet()</tt> 而不是DNA序列特有的。</p>
</div>
<div class="section" id="genbank">
<h3>2.4.2  简单的GenBank解析示例</h3>
<p>现在我们来加载一个GenBank文件 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk">ls_orchid.gbk</a>
——注意这里的代码与上面处理FASTA文件的代码几乎完全相同——仅有的不同之处是我们改变了文件名和格式的字符串：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span>
    <span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
    <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="punctuation">)</span>
</pre>
<p>这段代码应该会给出：</p>
<pre class="code python literal-block">
<span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
<span class="literal number integer">740</span>
<span class="operator">...</span>
<span class="name">Z78439</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
<span class="literal number integer">592</span>
</pre>
<p>这一次 <tt class="docutils literal">Bio.SeqIO</tt> 能够选择一个合理的字母表，IUPAC Ambiguous DNA。你应该注意到了这个例子中有一个较短的字符串被作为 <tt class="docutils literal">seq_record.id</tt> 。</p>
</div>
<div class="section" id="id9">
<h3>2.4.3  我爱解析——请别停止讨论它！</h3>
<p>Biopython有很多的解析器，基于它们所解析的文件格式，每一个都有自己独特的作用。第<cite>5 &lt;#chapter:Bio.SeqIO&gt;`__章包含 ``Bio.SeqIO`</cite> 更详细的内容，而第
<a class="reference external" href="#chapter:Bio.AlignIO">6</a> 章将介绍用于序列比对的 <tt class="docutils literal">Bio.AlignIO</tt> 。</p>
<p>由于最主流的文件格式都有解析器整合在 <tt class="docutils literal">Bio.SeqIO</tt> 和/或 <tt class="docutils literal">Bio.AlignIO</tt> 中，对于一些比较罕见的或者不被人们喜爱的文件格式，要么根本就没有解析器，要么就是一些没有链接的老的解析器。请到wiki页面 <a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a>
以及 <a class="reference external" href="http://biopython.org/wiki/AlignIO">http://biopython.org/wiki/AlignIO</a> 查看最新信息，或者咨询邮件列表。wiki页面上应该包含了支持文件类型的最新列表，还有一些附加的例子。</p>
<p>另一个查找特定解析器信息和如何很酷的使用它们的地方就是Cookbook（本教程的第 <a class="reference external" href="#chapter:cookbook">18</a> 章）。如果你没有找到你要的信息，请考虑及时帮帮你那可怜的过劳的文档，并提交一份cookbook entry！（一旦你知道怎么做了，那就是了！）</p>
</div>
</div>
<div class="section" id="id10">
<h2>2.5  连接生物学数据库</h2>
<p>在生物信息学中你需要做的很普遍的事情之一是从生物学数据库中提取信息。手动访问这些数据库可能会非常枯燥乏味，尤其
是当你有很多重复的工作要做的时候。Biopython试图通过用Python脚本访问一些可用的在线数据库来节省你的时间和精力。目前，
Biopython有从以下数据库中获取信息的代码：</p>
<ul class="simple">
<li>NCBI的 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/Entrez/">Entrez</a> （和 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/PubMed/">PubMed</a>）
——见第 <a class="reference external" href="#chapter:entrez">9</a> 章。</li>
<li><a class="reference external" href="http://www.expasy.org/">ExPASy</a> ——见第 <a class="reference external" href="#chapter:swiss_prot">10</a> 章。</li>
<li><a class="reference external" href="http://scop.mrc-lmb.cam.ac.uk/scop/">SCOP</a>——见 <tt class="docutils literal">Bio.SCOP.search()</tt> 方法。</li>
</ul>
<p>使用模块里的代码基本上可以容易地写出与这些页面中CGI脚本交互的Python代码，因此你能很方便地获得想要的结果。在某些情况下，结果能很好地整合到Biopython解析器中从而使得提取信息更加简单。</p>
</div>
<div class="section" id="id11">
<h2>2.6  下一步做什么</h2>
<p>现在你已经做到这一步，你应该对基本的Biopython有一个很好的了解，并准备好开始用它完成一些有用的工作。现在最好先完成
阅读本教程，然后如果你可能会想看看源码以及文档。</p>
<p>一旦你知道你想做什么，以及Biopython能完成它的库，你应该看看Cookbook（第 <a class="reference external" href="#chapter:cookbook">18</a> 章），
在这里可能会有一些类似你工作的示例代码。</p>
<p>如果你知道你想要做什么，但是还没弄明白怎么去做，请随时将你的问题贴出到主要的Biopython列表中（见
<a class="reference external" href="http://biopython.org/wiki/Mailing_lists">http://biopython.org/wiki/Mailing_lists</a>）。这不仅方便我们回答你的
问题，也有助于我们改进文档以便于它能帮到下一个和你做同样工作的人。</p>
<p>请享受代码吧！</p>
</div>
</div>
<div class="section" id="id12">
<h1>第3章     生物序列对象</h1>
<p>生物学序列以绝对的优势成为生物信息学研究的重点对象。这一章我们将简要介绍
Biopython处理这些序列的机制-- <tt class="docutils literal">Seq</tt> 对象。第 <a class="reference external" href="#chapter:SeqRecord">4</a>
章将要引入与此相关的 <tt class="docutils literal">SeqRecord</tt> 对象，此对象可以将序列信息和注释结合起来，
用于第 <a class="reference external" href="#chapter:Bio.SeqIO">5</a> 章序列的输入/输出。</p>
<p>序列实质上就是由字母构成的字符串，比如 <tt class="docutils literal">AGTACACTGGT</tt> ，看起来很自然，因为
这就是序列在生物学文件中的常用代表格式。</p>
<p><tt class="docutils literal">Seq</tt> 对象和标准的Python字符串有两个明显的不同。首先，它们使用不同的方法。
尽管``Seq``对象支持常规字符串的很多方法，但是它的 <tt class="docutils literal">translate()</tt> 方法在做
生物学翻译时是不同的。相似的还有其他的生物学相关的方法，比如 <tt class="docutils literal">reverse_complement()</tt> 。
其次， <tt class="docutils literal">Seq</tt> 对象具有一个重要的属性-- <tt class="docutils literal">alphabet</tt> ，这一对象用于描述由单个
字母构成的序列字符串的 “mean” （意义），以及如何解释这一字符串。例如， <tt class="docutils literal">AGTACACTGGT</tt>
序列是个DNA序列还是一段富含Alanines, Glycines, Cysteines and Threonines
的蛋白质序列？</p>
<div class="section" id="id13">
<h2>3.1  序列和字母表</h2>
<p>字母对象可能是使得``Seq``对象不仅仅是字符串的重要因素。目前，Biopython的字母表
定义在``Bio.Alphabet``模块。我们将会使用IUPAC字母表
（ <a class="reference external" href="http://www.chem.qmw.ac.uk/iupac/">http://www.chem.qmw.ac.uk/iupac/</a> ）
来处理我们比较青睐的对象：DNA、RNA和蛋白质对象。</p>
<p><tt class="docutils literal">Bio.Alphabet.IUPAC</tt> 提供了蛋白质、DNA和RNA的基本定义，并且提供了扩展
和定制基本定义的功能。例如蛋白质，有一个基本的IUPACProtein类，另外还有
一个ExtendedIUPACProtein类。这个类包含除20种常见氨基酸外的其他氨基酸元素，
比如 “U” （或 “Sec”代表硒代半胱氨酸），“O”（或“Pyl” 代表吡咯赖氨酸），
还有歧意字母“B” （或“Asx”代表天冬酰胺或者天冬氨酸），“Z”（或“Glx”代表谷
氨酰胺或者谷氨酸），“J” （或“Xle”代表亮氨酸或异亮氨酸），“X” （或“Xxx”代
表未知氨基酸）。同理，对于DNA有IUPACUnambiguousDNA、IUPACAmbiguousDNA和
ExtendedIUPACDNA类，分别提供基本字母，每种可能下的歧意字母和修饰后的碱基。
同样地，RNA可以使用IUPACAmbiguousRNA和IUPACUnambiguousRNA类。</p>
<p>使用字母表类有两方面的优势。首先，明确了Seq对象包含的信息的类型；其次通过
类型检查，它提供了约束信息的工具。</p>
<p>我们已经知道了将要处理的对象，现在让我们看看怎么使用这些类做一些有意思的事情。
你可以创建一条有通用字母组成的模糊序列，如下：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AGTACACTGGT&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGTACACTGGT'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">alphabet</span>
<span class="name">Alphabet</span><span class="punctuation">()</span>
</pre>
<p>然而，如果可能，你要在创建序列对象的时候就尽量明确指定字母的类型，如下创建一条
明确的DNA字母表对象。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AGTACACTGGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGTACACTGGT'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">alphabet</span>
<span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()</span>
</pre>
<p>当然,除非这真的是一个氨基酸序列：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_prot</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AGTACACTGGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">protein</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_prot</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGTACACTGGT'</span><span class="punctuation">,</span> <span class="name">IUPACProtein</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_prot</span><span class="operator">.</span><span class="name">alphabet</span>
<span class="name">IUPACProtein</span><span class="punctuation">()</span>
</pre>
</div>
<div class="section" id="id14">
<h2>3.2  序列表现的像字符串一样</h2>
<p>在许多时候，我们可以讲Seq对象处理成正常的Python字符串，比如取序列长度，迭代
元素：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GATCG&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">index</span><span class="punctuation">,</span> <span class="name">letter</span> <span class="operator word">in</span> <span class="name builtin">enumerate</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">index</span><span class="punctuation">,</span> <span class="name">letter</span>
<span class="literal number integer">0</span> <span class="name">G</span>
<span class="literal number integer">1</span> <span class="name">A</span>
<span class="literal number integer">2</span> <span class="name">T</span>
<span class="literal number integer">3</span> <span class="name">C</span>
<span class="literal number integer">4</span> <span class="name">G</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
</pre>
<p>你可以像字符串那样获取序列的元素（但是请记住，Python计数从0开始）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span> <span class="comment">#first letter</span>
<span class="name">G</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">]</span> <span class="comment">#third letter</span>
<span class="name">T</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span> <span class="comment">#last letter</span>
<span class="name">G</span>
</pre>
<p><tt class="docutils literal">Seq</tt> 对象有一个 <tt class="docutils literal">.count()</tt> 方法，类似于字符串。记住这意味就像Python的
字符串一样进行着非重叠的计数。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="literal string">&quot;AAAA&quot;</span><span class="operator">.</span><span class="name">count</span><span class="punctuation">(</span><span class="literal string">&quot;AA&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AAAA&quot;</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">count</span><span class="punctuation">(</span><span class="literal string">&quot;AA&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">2</span>
</pre>
<p>但是在某些生物学上，你可能需要使用重叠计数（就像上面的例子中如果重复计
数结果将为3）。当计算耽搁字母出现的次数时，重叠计数和非重叠计数没有差别。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATCGATGGGCCTATATAGGATCGAAAATCGC'</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="punctuation">)</span>
<span class="literal number integer">32</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">count</span><span class="punctuation">(</span><span class="literal string">&quot;G&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">9</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="literal number integer">100</span> <span class="operator">*</span> <span class="name builtin">float</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="operator">.</span><span class="name">count</span><span class="punctuation">(</span><span class="literal string">&quot;G&quot;</span><span class="punctuation">)</span> <span class="operator">+</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">count</span><span class="punctuation">(</span><span class="literal string">&quot;C&quot;</span><span class="punctuation">))</span> <span class="operator">/</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="punctuation">)</span>
<span class="literal number float">46.875</span>
</pre>
<p>你当然可以使用上面的代码段计算GC含量，但是记住 <tt class="docutils literal">Bio.SeqUtils</tt> 模块已经
建立了好几个GC函数，类如：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqUtils</span> <span class="keyword namespace">import</span> <span class="name">GC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATCGATGGGCCTATATAGGATCGAAAATCGC'</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">GC</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="punctuation">)</span>
<span class="literal number float">46.875</span>
</pre>
<p>注意在使用 <tt class="docutils literal">Bio.SeqUtils.GC()</tt> 函数时会自动处理序列和可代表G或者C的歧意核苷酸
字母S混合的情况。</p>
<p>然后还要注意，就像正常的Python字符串， <tt class="docutils literal">Seq</tt> 对象在某些方式下是只读的。如果需要
编辑序列，比如模拟点突变，请看后续的 <a class="reference external" href="#sec:mutable-seq">3.12</a> 章节中讲述的
<tt class="docutils literal">MutableSeq</tt> 对象。</p>
</div>
<div class="section" id="id15">
<h2>3.3  切取序列</h2>
<p>一个较为复杂的例子，让我们切取序列。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GATCGATGGGCCTATATAGGATCGAAAATCGC&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">4</span><span class="punctuation">:</span><span class="literal number integer">12</span><span class="punctuation">]</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATGGGCC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>要注意两个有意思的地方。首先，序列第一个元素从0开始，这是符合Python字符串的规则的。
这在计算机科学上是普遍现象，但在生物学上不是这样。当你做切片的时候，第一项包含了
（比如例子中的4），而最后一项去除了（例子中的12）。这是Python的规则，但当然这不是
世界上所有人都希望的。主要是为了和Python保持一致。</p>
<dl class="docutils">
<dt>第二个需要注意的地方是，切片是在序列数据字符串上执行的，但是产生的新对象保留了原始</dt>
<dd><tt class="docutils literal">Seq</tt> 对象的字母表信息。</dd>
</dl>
<p>同样和Python字符串一样，你可以通过设置起始位置、终止位置和 <em>步幅</em> （间隔数，默认为1）
进行切片。例如，我们可以分别获取下面DNA序列密码子第一、第二、第三位的碱基。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">::</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GCTGTAGTAAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">::</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGGCATGCATC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">::</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TAGCTAAGAC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>你可能已经注意到Python字符串中的另一个奇特步幅设定：使用-1返回倒序字符串切片。
当然以也可以使用 <tt class="docutils literal">Seq</tt> 对象来完成。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[::</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGCTAAAAGCTAGGATATATCCGGGTAGCTAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
</div>
<div class="section" id="id16">
<h2>3.4  将序列对象转换成字符串</h2>
<p>如果你仅仅需要一个单纯的字符串，就像写入文件或者插入数据库，这事很容易就
可以实现的：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">my_seq</span><span class="punctuation">)</span>
<span class="literal string">'GATCGATGGGCCTATATAGGATCGAAAATCGC'</span>
</pre>
<p>尽管对 <tt class="docutils literal">Seq</tt> 对象调用 <tt class="docutils literal">str()</tt> 方法将以字符串的形式返回全长序列，但是你经常不需要
特地做这个转换。当使用print打印声明是，Python会自动转换。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_seq</span>
<span class="name">GATCGATGGGCCTATATAGGATCGAAAATCGC</span>
</pre>
<p>当你进行Python字符串格式化或者插入操作符（ <tt class="docutils literal">%</tt> ）时，
可以直接把 <tt class="docutils literal">Seq</tt> 对象和 <tt class="docutils literal">%s</tt> 占位符一起使用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">fasta_format_string</span> <span class="operator">=</span> <span class="literal string">&quot;&gt;Name</span><span class="literal string escape">\n</span><span class="literal string interpol">%s</span><span class="literal string escape">\n</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">my_seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">fasta_format_string</span>
<span class="operator">&gt;</span><span class="name">Name</span>
<span class="name">GATCGATGGGCCTATATAGGATCGAAAATCGC</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>这一行代码展示的是一个简单的FASTA格式的记录（不用关心自动换行）。
<a class="reference external" href="#sec:SeqRecord-format">4.5</a> 部分将介绍一个简洁的方式从 <tt class="docutils literal">SeqRecord</tt>
对象中获取FASTA格式的字符串，更详细的读写FASTA格式的序列文件将在第
<a class="reference external" href="#chapter:Bio.SeqIO">5</a> 章介绍。</p>
<p><em>注意：</em>  如果你使用Biopython 1.44或者更旧的版本，使用 <tt class="docutils literal">str(my_seq)</tt> 只会
返回一个截短了的序列。这时候可以使用 <tt class="docutils literal">my_seq.tostring()</tt> ，为了保持向后兼
容性，这一方法在当前的Biopython版本中还有保留。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">tostring</span><span class="punctuation">()</span>
<span class="literal string">'GATCGATGGGCCTATATAGGATCGAAAATCGC'</span>
</pre>
</div>
<div class="section" id="id17">
<h2>3.5  连接或添加序列</h2>
<p>当然，原则上你可以将任何两个Seq对象加在一起，就像Python字符串一样去连接它们。
但是你不能将两个不相容的字母表加在一起，比如蛋白质序列和核苷酸序列就不能简单
叠加。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">protein_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;EVRNAK&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">protein</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">protein_seq</span> <span class="operator">+</span> <span class="name">dna_seq</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
<span class="operator">...</span>
<span class="name exception">TypeError</span><span class="punctuation">:</span> <span class="name">Incompatible</span> <span class="name">alphabets</span> <span class="name">IUPACProtein</span><span class="punctuation">()</span> <span class="operator word">and</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()</span>
</pre>
<p>如果你 <em>真的</em> 想这么做，你必须首先将两个序列转换成通用字母表。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_alphabet</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">protein_seq</span><span class="operator">.</span><span class="name">alphabet</span> <span class="operator">=</span> <span class="name">generic_alphabet</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span><span class="operator">.</span><span class="name">alphabet</span> <span class="operator">=</span> <span class="name">generic_alphabet</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">protein_seq</span> <span class="operator">+</span> <span class="name">dna_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'EVRNAKACGT'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
</pre>
<p>这里有个例子是将通用核苷酸序列加到明确的IUPAC DNA序列上，最后生成一段
模糊的核苷酸序列。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_nucleotide</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nuc_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GATCGATGC&quot;</span><span class="punctuation">,</span> <span class="name">generic_nucleotide</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nuc_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATCGATGC'</span><span class="punctuation">,</span> <span class="name">NucleotideAlphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ACGT'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nuc_seq</span> <span class="operator">+</span> <span class="name">dna_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATCGATGCACGT'</span><span class="punctuation">,</span> <span class="name">NucleotideAlphabet</span><span class="punctuation">())</span>
</pre>
</div>
<div class="section" id="id18">
<h2>3.6  改变大小写</h2>
<p>Python字符串具有很有用的转换大小写的 <tt class="docutils literal">upper</tt> 和 <tt class="docutils literal">lower</tt> 方法。从
Biopython 1.53起， <tt class="docutils literal">Seq</tt> 对象也获取了类似的方法应用于字母表。例如:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_dna</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;acgtACGT&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'acgtACGT'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ACGTACGT'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span><span class="operator">.</span><span class="name">lower</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'acgtacgt'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">())</span>
</pre>
<p>这在不区分大小写进行匹配的时候很有用。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="literal string">&quot;GTAC&quot;</span> <span class="operator word">in</span> <span class="name">dna_seq</span>
<span class="name builtin pseudo">False</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="literal string">&quot;GTAC&quot;</span> <span class="operator word">in</span> <span class="name">dna_seq</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">()</span>
<span class="name builtin pseudo">True</span>
</pre>
<p>注意，严格地说IUPAC字母表仅仅是对于大写字母构成的序列的，因此：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ACGT'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">dna_seq</span><span class="operator">.</span><span class="name">lower</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'acgt'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">())</span>
</pre>
</div>
<div class="section" id="id19">
<h2>3.7  核苷酸序列和（反向）互补序列</h2>
<p>对于核苷酸序列，你可以使用 <tt class="docutils literal">Seq</tt> 对象内置的方法很容易地获得 <tt class="docutils literal">Seq</tt>
的互补或反向互补序列。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GATCGATGGGCCTATATAGGATCGAAAATCGC&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATCGATGGGCCTATATAGGATCGAAAATCGC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">complement</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CTAGCTACCCGGATATATCCTAGCTTTTAGCG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GCGATTTTCGATCCTATATAGGCCCATCGATC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>在前面的方法中，使用切片的-1的步长可以很容易的获取一个 <tt class="docutils literal">Seq</tt> 对象的反向序列。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[::</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGCTAAAAGCTAGGATATATCCGGGTAGCTAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>在所有这些操作中，字母的属性一直保留着。这是非常有用的，以防你不小心做一些
奇怪的事情，比如获取蛋白质序列的（反向）互补序列。</p>
<pre class="code python literal-block">
   <span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
   <span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
   <span class="operator">&gt;&gt;&gt;</span> <span class="name">protein_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;EVRNAK&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">protein</span><span class="punctuation">)</span>
   <span class="operator">&gt;&gt;&gt;</span> <span class="name">protein_seq</span><span class="operator">.</span><span class="name">complement</span><span class="punctuation">()</span>
   <span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
   <span class="operator">...</span>
   <span class="name exception">ValueError</span><span class="punctuation">:</span> <span class="name">Proteins</span> <span class="name">do</span> <span class="operator word">not</span> <span class="name">have</span> <span class="name">complements</span><span class="error">!</span>

 \ <span class="literal string backtick">`5.5.3 &lt;#sec:SeqIO-reverse-complement&gt;`</span><span class="name">__</span> <span class="error">部分的例子将</span> <span class="literal string backtick">``</span><span class="name">Seq</span><span class="literal string backtick">``</span> <span class="error">对象的反向互补</span>
<span class="error">方法和</span> <span class="literal string backtick">``</span><span class="name">Bio</span><span class="operator">.</span><span class="name">SeqIO</span><span class="literal string backtick">``</span> <span class="error">对于序列的输入</span><span class="operator">/</span><span class="error">输出方法结合起来。</span>
</pre>
</div>
<div class="section" id="id20">
<h2>3.8  转录</h2>
<p>在谈论转录之前，我想先说明一下链的问题。考虑以下（编造的）编码短肽的双链DNA的延伸：</p>
<div class="formula">
<br/>    <sub><i>DNA</i><i>coding</i><i>strand</i>(<i>aka</i><i>Crick</i><i>strand</i>, <i>strand</i> + 1)</sub>    <br/>5’    <tt> ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG</tt>    3’<br/>    <tt> |||||||||||||||||||||||||||||||||||||||</tt>    <br/>3’    <tt> TACCGGTAACATTACCCGGCGACTTTCCCACGGGCTATC</tt>    5’<br/>    <sub><i>DNA</i><i>template</i><i>strand</i>(<i>aka</i><i>Watson</i><i>strand</i>, <i>strand</i> − 1)</sub>    <br/><br/>    |    <br/>    <i>Transcription</i>    <br/>    ↓    <br/><br/>5’    <tt> AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG</tt>    3’<br/>    <sub><i>Single</i><i>stranded</i><i>messenger</i><i>RNA</i></sub>    <br/><br/>
</div>
<p>实际的生物学上的转录过程是将模板链反向互补（TCAG → CUGA）生成mRNA。但是，
在Biopython和生物信息学领域，我们通常会直接利用编码链，因为我们可以通过
T → U的转换获得mRNA。</p>
<p>现在让我们着手真实地使用Biopython做一个转录。首先，让我们分别创建DNA序列的
编码链和模板链的 <tt class="docutils literal">Seq</tt> 对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">template_dna</span> <span class="operator">=</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">template_dna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CTATCGGGCACCCTTTCAGCGGCCCATTACAATGGCCAT'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>这是和上面的图表相一致的，记住按照惯例核苷酸序列通常是从5’到3’端方向的，
而图中所示的模板链是反向的。</p>
<p>现在让我们使用 <tt class="docutils literal">Seq</tt> 对象内置的 <tt class="docutils literal">transcribe</tt> 方法将编码链转录成对应的mRNA：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span> <span class="operator">=</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">transcribe</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousRNA</span><span class="punctuation">())</span>
</pre>
<p>就如你看到的，这里做的全部工作是将T → U转换，并调整字母表。</p>
<p>如果你确实想从模板链去做一个真正的生物学上的转录，需要两步：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">template_dna</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">transcribe</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousRNA</span><span class="punctuation">())</span>
</pre>
<p><tt class="docutils literal">Seq</tt> 对象还包含了从mRNA逆向转录为DNA编码链的方法。同样，这仅仅是从U
→ T的替代并伴随着字母表的变化：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_rna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousRNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span><span class="operator">.</span><span class="name">back_transcribe</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p><em>注意：</em> <tt class="docutils literal">Seq</tt> 对象的 <tt class="docutils literal">transcribe</tt> 和 <tt class="docutils literal">back_transcribe</tt> 方法直到
Biopython 1.49版本才出现，在较早的版本中你需要使用 <tt class="docutils literal">Bio.Seq</tt> 模块的函
数替代，详见 <a class="reference external" href="#sec:seq-module-functions">3.14</a> 部分。</p>
</div>
<div class="section" id="id21">
<h2>3.9  翻译</h2>
<p>继续使用在转录那个小节中的例子，现在让我们将这个mRNA翻译成相对应的
蛋白质序列，利用的是 <tt class="docutils literal">Seq</tt> 对象众多生物学方法中的一个：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_rna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousRNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">messenger_rna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGR*KGAR*'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">))</span>
</pre>
<p>你也可以直接从编码DNA链进行翻译：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGR*KGAR*'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">))</span>
</pre>
<p>你应该注意到在上面的蛋白质序列中，除了末尾的终止符外，在序列中间还有一个终止符。
其实这是一个精心选择的例子，因为由它我们可以引申讲一下可选参数，包括不同的翻译
表（遗传密码）。</p>
<p>Biopython上可用的翻译表是基于 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi">NCBI</a>
（参考这个教程的下一个部分）。默认情况下，翻译使用的是 <em>标准</em> 遗传密码（NCBI上table id 1)。
假设我们需要翻译一个线粒体序列，我们就需要告诉翻译函数使用相关的遗传密码：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal string">&quot;Vertebrate Mitochondrial&quot;</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGRWKGAR*'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">))</span>
</pre>
<p>你也可以利用NCBI上表格的标号来指定所使用的遗传密码，这样更简洁一些，
在GenBank文件的特征注释中经常包含表格的标号：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGRWKGAR*'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">))</span>
</pre>
<p>现在你可能想将上面的核苷酸序列仅翻译到阅读框的第一个终止密码子，然后停止
（这更符合自然现象）。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGR*KGAR*'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">to_stop</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGR'</span><span class="punctuation">,</span> <span class="name">IUPACProtein</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGRWKGAR*'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name">to_stop</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGRWKGAR'</span><span class="punctuation">,</span> <span class="name">IUPACProtein</span><span class="punctuation">())</span>
</pre>
<p>注意到当你使用 <tt class="docutils literal">to_stop</tt> 参数时，终止密码子本身是不翻译的，终止的符号也是
不显现在蛋白质序列中的。</p>
<p>如果你不喜欢默认的星号作为终止符号，你也可以自己指定终止符。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">coding_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name">stop_symbol</span><span class="operator">=</span><span class="literal string">&quot;&#64;&quot;</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAIVMGRWKGAR&#64;'</span><span class="punctuation">,</span> <span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">IUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'&#64;'</span><span class="punctuation">))</span>
</pre>
<p>现在假设你有一条完整的编码序列CDS，这是一种核苷酸序列（例如mRNA剪切以后），
序列全长都是密码子（也就是长度是3的倍数），开始于起始密码子，终止于终止密
码子，阅读框内没有内部的终止密码子。通常情况下，给你一条完整的CDS，默认的
翻译方法即可以翻译出你想要的（有时使用 <tt class="docutils literal">to_stop</tt> 选项）。但是，如果序列使
用的是非标准的起始密码子呢？这种情况在细菌中很常见，比如 <tt class="docutils literal">E. coli</tt>
K12中的基因yaaX：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_dna</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gene</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GTGAAAAAGATGCAATCTATCGTACTCGCACTTTCCCTGGTTCTGGTCGCTCCCATGGCA&quot;</span> <span class="operator">+</span> \
<span class="operator">...</span>            <span class="literal string">&quot;GCACAGGCTGCGGAAATTACGTTAGTCCCGTCAGTAAAATTACAGATAGGCGATCGTGAT&quot;</span> <span class="operator">+</span> \
<span class="operator">...</span>            <span class="literal string">&quot;AATCGTGGCTATTACTGGGATGGAGGTCACTGGCGCGACCACGGCTGGTGGAAACAACAT&quot;</span> <span class="operator">+</span> \
<span class="operator">...</span>            <span class="literal string">&quot;TATGAATGGCGAGGCAATCGCTGGCACCTACACGGACCGCCGCCACCGCCGCGCCACCAT&quot;</span> <span class="operator">+</span> \
<span class="operator">...</span>            <span class="literal string">&quot;AAGAAAGCTCCTCATGATCATCACGGCGGTCATGGTCCAGGCAAACATCACCGCTAA&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>            <span class="name">generic_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gene</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal string">&quot;Bacterial&quot;</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'VKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYWDGGHWRDH...HHR*'</span><span class="punctuation">,</span>
<span class="name">HasStopCodon</span><span class="punctuation">(</span><span class="name">ExtendedIUPACProtein</span><span class="punctuation">(),</span> <span class="literal string">'*'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gene</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal string">&quot;Bacterial&quot;</span><span class="punctuation">,</span> <span class="name">to_stop</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'VKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYWDGGHWRDH...HHR'</span><span class="punctuation">,</span>
<span class="name">ExtendedIUPACProtein</span><span class="punctuation">())</span>
</pre>
<p>在细菌遗传密码中 <tt class="docutils literal">GTG</tt> 是个有效的起始密码子。 <em>正常情况下</em> 编码缬氨酸，
如果作为起始密码子，则翻译成甲硫氨酸。当你告诉Biopython你的序列是完整CDS时，
这事将会发生。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gene</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="operator">=</span><span class="literal string">&quot;Bacterial&quot;</span><span class="punctuation">,</span> <span class="name">cds</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYWDGGHWRDH...HHR'</span><span class="punctuation">,</span>
<span class="name">ExtendedIUPACProtein</span><span class="punctuation">())</span>
</pre>
<p>除了告诉Biopython翻译时使用另一种起始密码子编码甲硫氨酸外，使用这一选项同样能
确保你的序列是个真实有效的CDS（如果不是将会抛出异常）。</p>
<dl class="docutils">
<dt>第 <a class="reference external" href="#sec:SeqIO-translate">18.1.3</a> 章的例子将把 <tt class="docutils literal">Seq</tt> 对象的翻译方法和</dt>
<dd><tt class="docutils literal">Bio.SeqIO</tt> 对象的对于序列的输入/输出方法结合起来。</dd>
</dl>
</div>
<div class="section" id="id22">
<h2>3.10  翻译表</h2>
<p>在前面的章节中我们讨论了 <tt class="docutils literal">Seq</tt> 对象的转录方法（并且提到了 <tt class="docutils literal">Bio.Seq</tt> 模块
中的等效函数--参见第 <a class="reference external" href="#sec:seq-module-functions">3.14</a> 章节）。实质上
使用的这些密码子表对象来自与NCBI的 <tt class="docutils literal"><span class="pre">`ftp://ftp.ncbi.nlm.nih.gov/entrez/misc/data/gc.prt</span></tt>
&lt;<a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/entrez/misc/data/gc.prt">ftp://ftp.ncbi.nlm.nih.gov/entrez/misc/data/gc.prt</a>&gt;`__ ，还有
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi">http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi</a>
以一种更易读的形式呈现。</p>
<p>和前面一样，让我们仅仅关注两个选择：标准的翻译表和脊椎动物线粒体DNA的翻译表。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Data</span> <span class="keyword namespace">import</span> <span class="name">CodonTable</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">standard_table</span> <span class="operator">=</span> <span class="name">CodonTable</span><span class="operator">.</span><span class="name">unambiguous_dna_by_name</span><span class="punctuation">[</span><span class="literal string">&quot;Standard&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mito_table</span> <span class="operator">=</span> <span class="name">CodonTable</span><span class="operator">.</span><span class="name">unambiguous_dna_by_name</span><span class="punctuation">[</span><span class="literal string">&quot;Vertebrate Mitochondrial&quot;</span><span class="punctuation">]</span>
</pre>
<p>另一种方式，这些表也可以分别以标号1和2来标识：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Data</span> <span class="keyword namespace">import</span> <span class="name">CodonTable</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">standard_table</span> <span class="operator">=</span> <span class="name">CodonTable</span><span class="operator">.</span><span class="name">unambiguous_dna_by_id</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mito_table</span> <span class="operator">=</span> <span class="name">CodonTable</span><span class="operator">.</span><span class="name">unambiguous_dna_by_id</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">]</span>
</pre>
<p>你可以在打印后直观地比较这些实际的翻译表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">standard_table</span>
<span class="name">Table</span> <span class="literal number integer">1</span> <span class="name">Standard</span><span class="punctuation">,</span> <span class="name">SGC0</span>

  <span class="operator">|</span>  <span class="name">T</span>      <span class="operator">|</span>  <span class="name">C</span>      <span class="operator">|</span>  <span class="name">A</span>      <span class="operator">|</span>  <span class="name">G</span>      <span class="operator">|</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTT</span> <span class="name">F</span>   <span class="operator">|</span> <span class="name">TCT</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAT</span> <span class="name">Y</span>   <span class="operator">|</span> <span class="name">TGT</span> <span class="name">C</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTC</span> <span class="name">F</span>   <span class="operator">|</span> <span class="name">TCC</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAC</span> <span class="name">Y</span>   <span class="operator">|</span> <span class="name">TGC</span> <span class="name">C</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTA</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">TCA</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAA</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">TGA</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">A</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTG</span> <span class="name">L</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">TCG</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAG</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">TGG</span> <span class="name">W</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTT</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCT</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAT</span> <span class="name">H</span>   <span class="operator">|</span> <span class="name">CGT</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTC</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCC</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAC</span> <span class="name">H</span>   <span class="operator">|</span> <span class="name">CGC</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTA</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCA</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAA</span> <span class="name">Q</span>   <span class="operator">|</span> <span class="name">CGA</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">A</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTG</span> <span class="name">L</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">CCG</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAG</span> <span class="name">Q</span>   <span class="operator">|</span> <span class="name">CGG</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATT</span> <span class="name">I</span>   <span class="operator">|</span> <span class="name">ACT</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAT</span> <span class="name">N</span>   <span class="operator">|</span> <span class="name">AGT</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATC</span> <span class="name">I</span>   <span class="operator">|</span> <span class="name">ACC</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAC</span> <span class="name">N</span>   <span class="operator">|</span> <span class="name">AGC</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATA</span> <span class="name">I</span>   <span class="operator">|</span> <span class="name">ACA</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAA</span> <span class="name">K</span>   <span class="operator">|</span> <span class="name">AGA</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">A</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATG</span> <span class="name">M</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">ACG</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAG</span> <span class="name">K</span>   <span class="operator">|</span> <span class="name">AGG</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTT</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCT</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAT</span> <span class="name">D</span>   <span class="operator">|</span> <span class="name">GGT</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTC</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCC</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAC</span> <span class="name">D</span>   <span class="operator">|</span> <span class="name">GGC</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTA</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCA</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAA</span> <span class="name">E</span>   <span class="operator">|</span> <span class="name">GGA</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">A</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTG</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCG</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAG</span> <span class="name">E</span>   <span class="operator">|</span> <span class="name">GGG</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
</pre>
<p>和</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">mito_table</span>
<span class="name">Table</span> <span class="literal number integer">2</span> <span class="name">Vertebrate</span> <span class="name">Mitochondrial</span><span class="punctuation">,</span> <span class="name">SGC1</span>

  <span class="operator">|</span>  <span class="name">T</span>      <span class="operator">|</span>  <span class="name">C</span>      <span class="operator">|</span>  <span class="name">A</span>      <span class="operator">|</span>  <span class="name">G</span>      <span class="operator">|</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTT</span> <span class="name">F</span>   <span class="operator">|</span> <span class="name">TCT</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAT</span> <span class="name">Y</span>   <span class="operator">|</span> <span class="name">TGT</span> <span class="name">C</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTC</span> <span class="name">F</span>   <span class="operator">|</span> <span class="name">TCC</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAC</span> <span class="name">Y</span>   <span class="operator">|</span> <span class="name">TGC</span> <span class="name">C</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTA</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">TCA</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAA</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">TGA</span> <span class="name">W</span>   <span class="operator">|</span> <span class="name">A</span>
<span class="name">T</span> <span class="operator">|</span> <span class="name">TTG</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">TCG</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">TAG</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">TGG</span> <span class="name">W</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTT</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCT</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAT</span> <span class="name">H</span>   <span class="operator">|</span> <span class="name">CGT</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTC</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCC</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAC</span> <span class="name">H</span>   <span class="operator">|</span> <span class="name">CGC</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTA</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCA</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAA</span> <span class="name">Q</span>   <span class="operator">|</span> <span class="name">CGA</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">A</span>
<span class="name">C</span> <span class="operator">|</span> <span class="name">CTG</span> <span class="name">L</span>   <span class="operator">|</span> <span class="name">CCG</span> <span class="name">P</span>   <span class="operator">|</span> <span class="name">CAG</span> <span class="name">Q</span>   <span class="operator">|</span> <span class="name">CGG</span> <span class="name">R</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATT</span> <span class="name">I</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">ACT</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAT</span> <span class="name">N</span>   <span class="operator">|</span> <span class="name">AGT</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATC</span> <span class="name">I</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">ACC</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAC</span> <span class="name">N</span>   <span class="operator">|</span> <span class="name">AGC</span> <span class="name">S</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATA</span> <span class="name">M</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">ACA</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAA</span> <span class="name">K</span>   <span class="operator">|</span> <span class="name">AGA</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">A</span>
<span class="name">A</span> <span class="operator">|</span> <span class="name">ATG</span> <span class="name">M</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">ACG</span> <span class="name">T</span>   <span class="operator">|</span> <span class="name">AAG</span> <span class="name">K</span>   <span class="operator">|</span> <span class="name">AGG</span> <span class="name">Stop</span><span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTT</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCT</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAT</span> <span class="name">D</span>   <span class="operator">|</span> <span class="name">GGT</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">T</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTC</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCC</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAC</span> <span class="name">D</span>   <span class="operator">|</span> <span class="name">GGC</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">C</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTA</span> <span class="name">V</span>   <span class="operator">|</span> <span class="name">GCA</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAA</span> <span class="name">E</span>   <span class="operator">|</span> <span class="name">GGA</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">A</span>
<span class="name">G</span> <span class="operator">|</span> <span class="name">GTG</span> <span class="name">V</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span><span class="operator">|</span> <span class="name">GCG</span> <span class="name">A</span>   <span class="operator">|</span> <span class="name">GAG</span> <span class="name">E</span>   <span class="operator">|</span> <span class="name">GGG</span> <span class="name">G</span>   <span class="operator">|</span> <span class="name">G</span>
<span class="operator">--+---------+---------+---------+---------+--</span>
</pre>
<p>你会发现下面的特性很有用，比如当你查找新基因时：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mito_table</span><span class="operator">.</span><span class="name">stop_codons</span>
<span class="punctuation">[</span><span class="literal string">'TAA'</span><span class="punctuation">,</span> <span class="literal string">'TAG'</span><span class="punctuation">,</span> <span class="literal string">'AGA'</span><span class="punctuation">,</span> <span class="literal string">'AGG'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mito_table</span><span class="operator">.</span><span class="name">start_codons</span>
<span class="punctuation">[</span><span class="literal string">'ATT'</span><span class="punctuation">,</span> <span class="literal string">'ATC'</span><span class="punctuation">,</span> <span class="literal string">'ATA'</span><span class="punctuation">,</span> <span class="literal string">'ATG'</span><span class="punctuation">,</span> <span class="literal string">'GTG'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mito_table</span><span class="operator">.</span><span class="name">forward_table</span><span class="punctuation">[</span><span class="literal string">&quot;ACG&quot;</span><span class="punctuation">]</span>
<span class="literal string">'T'</span>
</pre>
</div>
<div class="section" id="seq">
<h2>3.11  比较Seq对象</h2>
<p>序列之间的比较实际上是一个比较复杂的话题，没有简单的方法来判断两个序列是等同的。
核心的问题是字母的意义是依赖于上下文的。字母 “A” 既可以是DNA、RNA也可以使蛋白质序
列的一部分。 Biopython在 <tt class="docutils literal">Seq</tt> 对象中包含了字母表对象，以此尝试获得这些信息。所
以比较两个 <tt class="docutils literal">Seq</tt> 对象意味着既要考虑两个序列的字符串 <em>又要</em> 考虑字母表。</p>
<p>举个例子，你可能会觉得 <tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;,</span> IUPAC.unambiguous_dna)</tt> 和
<tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;,</span> IUPAC.ambiguous_dna)</tt> 这两个DNA <tt class="docutils literal">Seq</tt> 对象是一样的，尽管它们确实具
有不同的字母表。根据上下文来判断是很重要的。</p>
<p>下面这种情况更遭：假设你认为 <tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;,</span> IUPAC.unambiguous_dna)</tt> 和
<tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;)</span></tt> （也就是默认的通用字母表）是等同的。那么依照逻辑，
<tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;,</span> IUPAC.protein)</tt> 和 <tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;)</span></tt> 也是等同的。现在从理
论上讲，如果 <em>A</em>=<em>B</em> ， <em>B</em>=<em>C</em> ，那么通过递延性，我们会期望
<em>A</em>=<em>C</em> 。因此遵从逻辑上的一致性我们需要将 <tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;,</span> IUPAC.unambiguous_dna)</tt>
和 <tt class="docutils literal"><span class="pre">Seq(&quot;ACGT&quot;,</span> IUPAC.protein)</tt> 等同起来，虽然大部分人会同意这一递延，
但是这是错误的。这一递延性的问题也会影响使用 <tt class="docutils literal">Seq</tt> 对象作为Python字典
的键值。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq1</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq2</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
</pre>
<p>那么接下来Biopython会怎么做？等同性测试是Python对象默认要做的测试。经过
检验查看内存中的对象是不是同一个。这是一个非常严格的测试：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq1</span> <span class="operator">==</span> <span class="name">seq2</span>
<span class="name builtin pseudo">False</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq1</span> <span class="operator">==</span> <span class="name">seq1</span>
<span class="name builtin pseudo">True</span>
</pre>
<p>如果你真想这么做，你可以更明确地使用Python中的 <tt class="docutils literal">id</tt> 函数，</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">id</span><span class="punctuation">(</span><span class="name">seq1</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="name builtin">id</span><span class="punctuation">(</span><span class="name">seq2</span><span class="punctuation">)</span>
<span class="name builtin pseudo">False</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">id</span><span class="punctuation">(</span><span class="name">seq1</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="name builtin">id</span><span class="punctuation">(</span><span class="name">seq1</span><span class="punctuation">)</span>
<span class="name builtin pseudo">True</span>
</pre>
<p>在日常使用中，你的所有序列可能都是同一个字母表，或者至少都是同一类型的序列
（都是DNA、RNA或者都是蛋白质）。你可能想要的只是以字符串的形式比较这些序列，
那么直接这么做：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">seq1</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">seq2</span><span class="punctuation">)</span>
<span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">seq1</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">seq1</span><span class="punctuation">)</span>
<span class="name builtin pseudo">True</span>
</pre>
<p>作为一个扩展，你可以建立一个Python字典，以 <tt class="docutils literal">Seq</tt> 对象作为键值。一般情况下，
将序列作为字符串赋予键值更有用。详见 <a class="reference external" href="#sec:seq-to-string">3.4</a> 部分。</p>
</div>
<div class="section" id="mutableseq">
<h2>3.12  MutableSeq对象</h2>
<p>就像正常的Python字符串， <tt class="docutils literal">Seq</tt> 对象是 “只读的” ，在Python术语上就是不可变的。
除了想要 <tt class="docutils literal">Seq</tt> 对象表现得向一个字符串之外，这是一个很有用的默认，因为在生
物学应用上你往往需要确保你没有改动你的序列数据：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GCCATTGTAATGGGCCGCTGAAAGGGTGCCCGA&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
</pre>
<p>当你尝试编辑序列是你看看会发生什么：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_seq</span><span class="punctuation">[</span><span class="literal number integer">5</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="literal string">&quot;G&quot;</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
<span class="operator">...</span>
<span class="name exception">TypeError</span><span class="punctuation">:</span> <span class="literal string">'Seq'</span> <span class="name builtin">object</span> <span class="name">does</span> <span class="operator word">not</span> <span class="name">support</span> <span class="name">item</span> <span class="name">assignment</span>
</pre>
<p>但是你可以使用 <tt class="docutils literal">MutableSeq</tt> 对象将它转换成可变的序列，然后做任何你想要做的。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span> <span class="operator">=</span> <span class="name">my_seq</span><span class="operator">.</span><span class="name">tomutable</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span>
<span class="name">MutableSeq</span><span class="punctuation">(</span><span class="literal string">'GCCATTGTAATGGGCCGCTGAAAGGGTGCCCGA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>或者你可以直接从字符串建立一个 <tt class="docutils literal">MutableSeq</tt> 对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">MutableSeq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span> <span class="operator">=</span> <span class="name">MutableSeq</span><span class="punctuation">(</span><span class="literal string">&quot;GCCATTGTAATGGGCCGCTGAAAGGGTGCCCGA&quot;</span><span class="punctuation">,</span> <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
</pre>
<p>这两种方式都可以将序列对象转换成可变的：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span>
<span class="name">MutableSeq</span><span class="punctuation">(</span><span class="literal string">'GCCATTGTAATGGGCCGCTGAAAGGGTGCCCGA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span><span class="punctuation">[</span><span class="literal number integer">5</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="literal string">&quot;C&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span>
<span class="name">MutableSeq</span><span class="punctuation">(</span><span class="literal string">'GCCATCGTAATGGGCCGCTGAAAGGGTGCCCGA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span><span class="operator">.</span><span class="name">remove</span><span class="punctuation">(</span><span class="literal string">&quot;T&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span>
<span class="name">MutableSeq</span><span class="punctuation">(</span><span class="literal string">'GCCACGTAATGGGCCGCTGAAAGGGTGCCCGA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span><span class="operator">.</span><span class="name">reverse</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mutable_seq</span>
<span class="name">MutableSeq</span><span class="punctuation">(</span><span class="literal string">'AGCCCGTGGGAAAGTCGCCGGGTAATGCACCG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<dl class="docutils">
<dt>请注意与 <tt class="docutils literal">Seq</tt> 对象不同的是， <tt class="docutils literal">MutableSeq</tt> 对象的各种方法都是实时呈现的，比如</dt>
<dd><tt class="docutils literal">reverse_complement()</tt> 和 <tt class="docutils literal">reverse()</tt> 方法！</dd>
</dl>
<p>Python中可变对象和不可变对象的一个重要的技术差别就是 <tt class="docutils literal">MutableSeq</tt> 对象不可以作为
字典的键值 ，但是Python字符串或者 <tt class="docutils literal">Seq</tt> 对象就可以。</p>
<p>一旦你的 <tt class="docutils literal">MutableSeq</tt> 对象编辑完成，很容易将它变回到只读的 <tt class="docutils literal">Seq</tt> 对象，你只需：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">new_seq</span> <span class="operator">=</span> <span class="name">mutable_seq</span><span class="operator">.</span><span class="name">toseq</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">new_seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'AGCCCGTGGGAAAGTCGCCGGGTAATGCACCG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<dl class="docutils">
<dt>就像你从 <tt class="docutils literal">Seq</tt> 对象中获取字符串一样，你也可以从 <tt class="docutils literal">MutableSeq</tt> 获得（参见</dt>
<dd><a class="reference external" href="#sec:seq-to-string">3.4</a> 章节）。</dd>
</dl>
</div>
<div class="section" id="unknownseq">
<h2>3.13  UnknownSeq对象</h2>
<p><tt class="docutils literal">UnknownSeq</tt> 对象是基本的 <tt class="docutils literal">Seq</tt> 对象中的一个子类，其目的是一个已知长度的
序列，但序列并不是由实际的字母组成的。在这种情况下，你当然可以将其作为一个
正常的 <tt class="docutils literal">Seq</tt> 对象，但是存储由一百万个 “N” 字母组成的字符串会浪费相当大量的内
存，这时你可以只存储一个 “N” 和序列所需的长度（整数）。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">UnknownSeq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk</span> <span class="operator">=</span> <span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">Alphabet</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'?'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">unk</span>
<span class="error">????????????????????</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">unk</span><span class="punctuation">)</span>
<span class="literal number integer">20</span>
</pre>
<p>当然你也可以指定一个字母，而不仅仅是 “?” 。一般核苷酸序列默认为 “N” ，蛋白质
序列默认为 “X” 。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">UnknownSeq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_dna</span> <span class="operator">=</span> <span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span><span class="operator">=</span><span class="name">IUPAC</span><span class="operator">.</span><span class="name">ambiguous_dna</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_dna</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'N'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">unk_dna</span>
<span class="name">NNNNNNNNNNNNNNNNNNNN</span>
</pre>
<p>你可以使用所有常规的 <tt class="docutils literal">Seq</tt> 对象，记住这些可以节省内存的 <tt class="docutils literal">UnknownSeq</tt> 对象，
如你所希望的那样在恰当的地方使用。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_dna</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'N'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_dna</span><span class="operator">.</span><span class="name">complement</span><span class="punctuation">()</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'N'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_dna</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'N'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_dna</span><span class="operator">.</span><span class="name">transcribe</span><span class="punctuation">()</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">IUPACAmbiguousRNA</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'N'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_protein</span> <span class="operator">=</span> <span class="name">unk_dna</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">unk_protein</span>
<span class="name">UnknownSeq</span><span class="punctuation">(</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="name">alphabet</span> <span class="operator">=</span> <span class="name">ProteinAlphabet</span><span class="punctuation">(),</span> <span class="name">character</span> <span class="operator">=</span> <span class="literal string">'X'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">unk_protein</span>
<span class="name">XXXXXX</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">unk_protein</span><span class="punctuation">)</span>
<span class="literal number integer">6</span>
</pre>
<p>你也许能够在自己的代码中找到 <tt class="docutils literal">UnknownSeq</tt> 对象的应用，但你更可能首先在由
<tt class="docutils literal">Bio.SeqIO</tt> 创建的 <tt class="docutils literal">SeqRecord</tt> 对象中遇到 <tt class="docutils literal">UnknownSeq</tt> 对象（参见第
<a class="reference external" href="#chapter:Bio.SeqIO">5</a> 章）。一些序列格式的文件不总是由实际的序列组成，
像GenBank和EMBL文件就可能包含各种特征的列表，而序列部分仅展示contig信息。
又或者在测序工作中的QUAL文件仅包含质量分数，而 <em>从未</em> 包含序列，取而代之的
和QUAL文件同时生成的FASTA格式文件 <em>确实</em> 是由序列构成。</p>
</div>
<div class="section" id="id23">
<h2>3.14  直接使用字符串</h2>
<p>在这一章的结尾，对于那些 <em>真的</em> 不想使用序列对象的人（或者那些更喜欢面向
对象的函数式编程风格的人）， <tt class="docutils literal">Bio.Seq</tt> 的模块级别的函数可以接受普通的
Python字符串，比如 <tt class="docutils literal">Seq</tt> 对象（包括 <tt class="docutils literal">UnknownSeq</tt> 对象）或者 <tt class="docutils literal">MutableSeq</tt> 对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">reverse_complement</span><span class="punctuation">,</span> <span class="name">transcribe</span><span class="punctuation">,</span> <span class="name">back_transcribe</span><span class="punctuation">,</span> <span class="name">translate</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_string</span> <span class="operator">=</span> <span class="literal string">&quot;GCTGTTATGGGTCGTTGGAAGGGTGGTCGTGCTGCTGGTTAG&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name">my_string</span><span class="punctuation">)</span>
<span class="literal string">'CTAACCAGCAGCACGACCACCCTTCCAACGACCCATAACAGC'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">transcribe</span><span class="punctuation">(</span><span class="name">my_string</span><span class="punctuation">)</span>
<span class="literal string">'GCUGUUAUGGGUCGUUGGAAGGGUGGUCGUGCUGCUGGUUAG'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">back_transcribe</span><span class="punctuation">(</span><span class="name">my_string</span><span class="punctuation">)</span>
<span class="literal string">'GCTGTTATGGGTCGTTGGAAGGGTGGTCGTGCTGCTGGTTAG'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">translate</span><span class="punctuation">(</span><span class="name">my_string</span><span class="punctuation">)</span>
<span class="literal string">'AVMGRWKGGRAAG*'</span>
</pre>
<p>尽管如此，我们鼓励你使用默认的 <tt class="docutils literal">Seq</tt> 对象。</p>
</div>
</div>
<div class="section" id="id24">
<h1>第4章  序列注释对象</h1>
<p>Chapter <a class="reference external" href="#chapter:Bio.Seq">3</a> 介绍了序列对象的基本情况。紧接上章的 <tt class="docutils literal">Seq</tt> 类，这章主要讲Sequence record 或称之为 <tt class="docutils literal">SeqRecord</tt> 类, 该类在 <tt class="docutils literal">Bio.SeqRecord</tt> 模块中有定义。 它（见 <tt class="docutils literal">SeqFeature</tt> 对象）可使序列与高级属性（如identifiers 和 features）关联。其应用贯穿序列输入/输出的交互界面 <tt class="docutils literal">Bio.SeqIO</tt> 过程中 （详见Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> ）。</p>
<p>如读者只需处理FASTA格式的序列文件等简单数据,可略过本章。如涉及带注释内容的数据（如 GenBank或EMBL格式文件）, 本章内容则非常重要。</p>
<p>尽管本章内容涵盖了 <tt class="docutils literal">SeqRecord</tt> 和 <tt class="docutils literal">SeqFeature</tt> 对象的大部分内容，但如需了解更多，读者可自行查阅 <tt class="docutils literal">SeqRecord</tt> wiki (<a class="reference external" href="http://biopython.org/wiki/SeqRecord">http://biopython.org/wiki/SeqRecord</a> ),和内置帮助文档 (或在线文档 <a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SeqRecord.SeqRecord-class.html">SeqRecord</a> 和 <a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SeqFeature.SeqFeature-class.html">SeqFeature</a> )，获取更多信息:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">SeqRecord</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<div class="section" id="seqrecord">
<h2>4.1  SeqRecord对象</h2>
<p><tt class="docutils literal">SeqRecord</tt> (Sequence Record) 类包含在 <tt class="docutils literal">Bio.SeqRecord</tt> 模块中。该类是 <tt class="docutils literal">Bio.SeqIO</tt> 序列输入/输出交互界面 (详见Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>)的基本数据类型。可以把identifiers 和features等高级属性与序列关联起来 (参见Chapter <a class="reference external" href="#chapter:Bio.Seq">3</a>)。</p>
<p><tt class="docutils literal">SeqRecord</tt> 类非常简单,包括下列属性:</p>
<dl class="docutils">
<dt><strong>.seq</strong></dt>
<dd>– 序列自身（即 <tt class="docutils literal">Seq</tt> 对象）。</dd>
<dt><strong>.id</strong></dt>
<dd>– 序列主ID（-字符串类型）。通常类同于accession number。</dd>
<dt><strong>.name</strong></dt>
<dd>– 序列名/id （-字符串类型）。 可以是accession number, 也可是clone名（类似GenBank record中的LOCUS id）。</dd>
<dt><strong>.description</strong></dt>
<dd>– 序列描述（-字符串类型）。</dd>
<dt><strong>.letter_annotations</strong></dt>
<dd>– 对照序列的每个字母逐字注释（per-letter-annotations），以信息名为键（keys），信息内容为值（value）所构成的字典。值与序列等长，用Python列表、元组或字符串表示。.letter_annotations可用于质量分数(如    Section <a class="reference external" href="#sec:FASTQ-filtering-example">18.1.6</a>) 或二级结构信息 (如 Stockholm/PFAM 比对文件)等数据的存储。</dd>
<dt><strong>.annotations</strong></dt>
<dd>– 用于储存附加信息的字典。信息名为键（keys），信息内容为值（value）。用于保存序列的零散信息（如unstructured information）。</dd>
<dt><strong>.features</strong></dt>
<dd>– <tt class="docutils literal">SeqFeature</tt> 对象列表，储存序列的结构化信息（structured information），如：基因位置, 蛋白结构域。features 详见本章第三节（ Section <a class="reference external" href="#sec:seq_features">4.3</a>）。</dd>
<dt><strong>.dbxrefs</strong></dt>
<dd>– 储存数据库交叉引用信息（cross-references）的字符串列表。</dd>
</dl>
</div>
<div class="section" id="id25">
<h2>4.2  创建 SeqRecord</h2>
<p>使用 <tt class="docutils literal">SeqRecord</tt> 对象非常简单，因为所有的信息都存储在该类的属性中；通常不必手动新建，用 <tt class="docutils literal">Bio.SeqIO</tt> 从序列文件读取即可（见Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>）。 当然新建 <tt class="docutils literal">SeqRecord</tt> 也不复杂。</p>
<div class="section" id="id26">
<h3>4.2.1  从头新建SeqRecord</h3>
<p><tt class="docutils literal">SeqRecord</tt> 最少只需包含 <tt class="docutils literal">Seq</tt> 对象:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GATC&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">simple_seq</span><span class="punctuation">)</span>
</pre>
<p>还可以通过初始化函数给 id, name和description赋值；反之，它们被设为默认值“unknown”（可随后编辑）:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">id</span>
<span class="literal string">'&lt;unknown id&gt;'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">id</span> <span class="operator">=</span> <span class="literal string">&quot;AC12345&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;Made up sequence I wish I could write a paper about&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">description</span>
<span class="name">Made</span> <span class="name">up</span> <span class="name">sequence</span> <span class="name">I</span> <span class="name">wish</span> <span class="name">I</span> <span class="name">could</span> <span class="name">write</span> <span class="name">a</span> <span class="name">paper</span> <span class="name">about</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATC'</span><span class="punctuation">,</span> <span class="name">Alphabet</span><span class="punctuation">())</span>
</pre>
<p>标识符对输出 <tt class="docutils literal">SeqRecord</tt> 内容到文件很重要，可随SeqRecord同时建立:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GATC&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">simple_seq</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;AC12345&quot;</span><span class="punctuation">)</span>
</pre>
<p>上述章节已提到，<tt class="docutils literal">SeqRecord</tt> 含有一个 <tt class="docutils literal">annotations</tt> 属性，用于储存各种杂乱注释的字典。添加annotations示例如下:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;evidence&quot;</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="literal string">&quot;None. I just made it up.&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">annotations</span>
<span class="punctuation">{</span><span class="literal string">'evidence'</span><span class="punctuation">:</span> <span class="literal string">'None. I just made it up.'</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;evidence&quot;</span><span class="punctuation">]</span>
<span class="name builtin pseudo">None</span><span class="operator">.</span> <span class="name">I</span> <span class="name">just</span> <span class="name">made</span> <span class="name">it</span> <span class="name">up</span><span class="operator">.</span>
</pre>
<p><tt class="docutils literal">letter_annotations</tt> 也是字典，其值为与序列等长的内置Python字符串、列表或元组:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">40</span><span class="punctuation">,</span><span class="literal number integer">40</span><span class="punctuation">,</span><span class="literal number integer">38</span><span class="punctuation">,</span><span class="literal number integer">30</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">letter_annotations</span>
<span class="punctuation">{</span><span class="literal string">'phred_quality'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">38</span><span class="punctuation">,</span> <span class="literal number integer">30</span><span class="punctuation">]}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">simple_seq_r</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">38</span><span class="punctuation">,</span> <span class="literal number integer">30</span><span class="punctuation">]</span>
</pre>
<p><tt class="docutils literal">dbxrefs</tt> 和 <tt class="docutils literal">features</tt> 分别是字符串和 <tt class="docutils literal">SeqFeature</tt> 对象的Python列表，将在后续章节讨论。</p>
</div>
<div class="section" id="fastaseqrecord">
<h3>4.2.2  根据FASTA文件创建SeqRecord对象</h3>
<p>本节以鼠疫耶尔森菌株（<em>Yersinia pestis biovar Microtus</em> str. 91001 ）的pPCP1质粒全长序列为例,说明从FASTA文件创建SeqRecord的过程。该序列原始文件来自NCBI，可在Biopython单元测试GenBank文件夹下找到，也可点击 <a class="reference external" href="http://biopython.org/SRC/biopython/Tests/GenBank/NC_005816.fna">NC_005816.fna</a> 下载。</p>
<p>序列以大于号开头，该文件只包含一条序列:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">45478711</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NC_005816</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="name">Yersinia</span> <span class="name">pestis</span> <span class="name">biovar</span> <span class="name">Microtus</span> <span class="operator">...</span> <span class="name">pPCP1</span><span class="punctuation">,</span> <span class="name">complete</span> <span class="name">sequence</span>
<span class="name">TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGGGGGTAATCTGCTCTCC</span>
<span class="operator">...</span>
</pre>
<p>回顾 Chapter <a class="reference external" href="#chapter:quick-start">2</a> 的内容，我们已经遇到过 <tt class="docutils literal"><span class="pre">Bio.SeqIO.parse(...)</span></tt> 函数，用于遍历 <tt class="docutils literal">SeqRecord</tt> 对象中的所有记录。 此处，我们介绍 <tt class="docutils literal">Bio.SeqIO</tt> 模块中的另一个类似函数Bio.SeqIO.read()，用于读取单条序列的文件 （详见 Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> ）:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.fna&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'</span><span class="punctuation">,</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|45478711|ref|NC_005816.1|'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'gi|45478711|ref|NC_005816.1|'</span><span class="punctuation">,</span>
<span class="name">description</span><span class="operator">=</span><span class="literal string">'gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus ... sequence'</span><span class="punctuation">,</span>
<span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[])</span>
</pre>
<p>现在让我们逐个介绍 <tt class="docutils literal">SeqRecord</tt> 对象中的主要属性，从给予我们序列属性的 <tt class="docutils literal">Seq</tt> 对象 开始:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'</span><span class="punctuation">,</span> <span class="name">SingleLetterAlphabet</span><span class="punctuation">())</span>
</pre>
<p>此处 <tt class="docutils literal">Bio.SeqIO</tt> 默认为通用字母表（generic alphabet）, 而非判断是否DNA序列。如果FASTA文件中序列类型已知，也可通过 <tt class="docutils literal">Bio.SeqIO</tt> 自行设定 (见Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> 用法)。</p>
<p>接下来介绍 identifiers 和 description:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span>
<span class="literal string">'gi|45478711|ref|NC_005816.1|'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'gi|45478711|ref|NC_005816.1|'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">description</span>
<span class="literal string">'gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus ... pPCP1, complete sequence'</span>
</pre>
<p>FASTA文件中序列名所在行的第一个单词(去除大于号后) 被当作 <tt class="docutils literal">id</tt> 和 <tt class="docutils literal">name</tt> ；而将整行 (去除大于号后) 作为 description。这样设定是为了向后兼容，同时也为了便于处理如下序列:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">Yersinia</span> <span class="name">pestis</span> <span class="name">biovar</span> <span class="name">Microtus</span> <span class="name builtin">str</span><span class="operator">.</span> <span class="literal number integer">91001</span> <span class="name">plasmid</span> <span class="name">pPCP1</span>
<span class="name">TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGGGGGTAATCTGCTCTCC</span>
<span class="operator">...</span>
</pre>
<p>Note: 读取FASTA 文件时其他注释属性为空:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="punctuation">[]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">annotations</span>
<span class="punctuation">{}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">letter_annotations</span>
<span class="punctuation">{}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span>
<span class="punctuation">[]</span>
</pre>
<p>本例中FASTA文件源于NCBI，其规范的格式，意味着我们可以方便的解析这些信息并选择提取GI和accession number等信息。然后，对于从其他来源获得的FASTA文件，并不能确保能获得这些信息。</p>
</div>
<div class="section" id="genbank-seqrecord">
<h3>4.2.3  从 GenBank文件创建 SeqRecord</h3>
<p>仍以疫耶尔森菌株pPCP1质粒全长序列（<em>Yersinia pestis biovar Microtus</em> str. 91001 plasmid pPCP1）为例，不同的是这次使用Genbank格式的文件，该文件同样包含在Biopython单元测试/GenBank文件夹下, 也可点击 <a class="reference external" href="http://biopython.org/SRC/biopython/Tests/GenBank/NC_005816.gb">NC_005816.gb</a>
下载。</p>
<p>该文件只含一条记录 (只有一个 LOCUS 行):</p>
<pre class="code python literal-block">
<span class="name">LOCUS</span>       <span class="name">NC_005816</span>               <span class="literal number integer">9609</span> <span class="name">bp</span>    <span class="name">DNA</span>     <span class="name">circular</span> <span class="name">BCT</span> <span class="literal number integer">21</span><span class="operator">-</span><span class="name">JUL</span><span class="operator">-</span><span class="literal number integer">2008</span>
<span class="name">DEFINITION</span>  <span class="name">Yersinia</span> <span class="name">pestis</span> <span class="name">biovar</span> <span class="name">Microtus</span> <span class="name builtin">str</span><span class="operator">.</span> <span class="literal number integer">91001</span> <span class="name">plasmid</span> <span class="name">pPCP1</span><span class="punctuation">,</span> <span class="name">complete</span>
            <span class="name">sequence</span><span class="operator">.</span>
<span class="name">ACCESSION</span>   <span class="name">NC_005816</span>
<span class="name">VERSION</span>     <span class="name">NC_005816</span><span class="operator">.</span><span class="literal number integer">1</span>  <span class="name">GI</span><span class="punctuation">:</span><span class="literal number integer">45478711</span>
<span class="name">PROJECT</span>     <span class="name">GenomeProject</span><span class="punctuation">:</span><span class="literal number integer">10638</span>
<span class="operator">...</span>
</pre>
<p>同样使用 <tt class="docutils literal">Bio.SeqIO</tt> 读取文件，代码跟处理FASTA 文件类似 (详见
Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> for details):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'</span><span class="punctuation">,</span>
<span class="name">IUPACAmbiguousDNA</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'NC_005816.1'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'NC_005816'</span><span class="punctuation">,</span>
<span class="name">description</span><span class="operator">=</span><span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span><span class="punctuation">,</span>
<span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">'Project:10638'</span><span class="punctuation">])</span>
</pre>
<p>你可能已经发现了一些不同之处，逐个环顾各个属性，序列字符串和上述类似，但此处 <tt class="docutils literal">Bio.SeqIO</tt> 可自动识别序列类型 （详见 chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> ）:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p><tt class="docutils literal">name</tt> 源于 LOCUS行, <tt class="docutils literal">id</tt> 附加了版本后缀。description源于DEFINITION 行:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span>
<span class="literal string">'NC_005816.1'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'NC_005816'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">description</span>
<span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span>
</pre>
<p>GenBank 文件中per-letter annotations为空:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">letter_annotations</span>
<span class="punctuation">{}</span>
</pre>
<p>多数注释信息储存在 <tt class="docutils literal">annotations</tt> 字典中，例如:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">)</span>
<span class="literal number integer">11</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;source&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Yersinia pestis biovar Microtus str. 91001'</span>
</pre>
<p><tt class="docutils literal">dbxrefs</tt> 列表中的数据来自 PROJECT 或DBLINK行:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="punctuation">[</span><span class="literal string">'Project:10638'</span><span class="punctuation">]</span>
</pre>
<p>最后也许也可能是最有意思的， <tt class="docutils literal">features</tt> 列表以 <tt class="docutils literal">SeqFeature</tt> 对象的形式保存了features table中的所有entries（如genes和CDS等）。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">)</span>
<span class="literal number integer">29</span>
</pre>
<p>接下来，我们将在 Section <a class="reference external" href="#sec:seq_features">4.3</a> 介绍 <tt class="docutils literal">SeqFeature</tt> 对象。</p>
</div>
</div>
<div class="section" id="feature-location-position">
<h2>4.3  Feature, location 和 position对象</h2>
<div class="section" id="seqfeature">
<h3>4.3.1  SeqFeature对象</h3>
<p>序列特征是描述一条序列不可或缺的部分。抛开序列本身，你需要一种方式去组织和获取关于这条序列的 “抽象” 信息。 尽管设计一个通用的类囊括序列的所有特征看似是不可能的，但是Biopython的 <tt class="docutils literal">SeqFeature</tt>
类试图尽可能多的囊括序列的所有特征。Biopython主要依据 GenBank/EMBL 特征表来设计相应的对象，认识到这一点，将有助于你更快更好的理解Biopython <tt class="docutils literal">SeqFeature</tt> 对象。</p>
<p><tt class="docutils literal">SeqFeature</tt> 对象的关键目的在于描述其相对于父序列（parent sequence，通常为 <tt class="docutils literal">SeqRecord</tt> 对象）所处的位置（location）, 通常是介于两个positions间的一个区域（region），后续Section <a class="reference external" href="#sec:locations">4.3.2</a> 将详细说明。</p>
<p><tt class="docutils literal">SeqFeature</tt> 对象含大量属性，首先一一例出，然后在后续章节举例说明其用法:</p>
<dl class="docutils">
<dt><strong>.type</strong></dt>
<dd>– 用文字描述的feature类型 (如 ‘CDS’ 或 ‘gene’)。</dd>
<dt><strong>.location</strong></dt>
<dd><p class="first">– <tt class="docutils literal">SeqFeature</tt> 在序列中所处的位置。见Section <a class="reference external" href="#sec:locations">4.3.2</a>。 <tt class="docutils literal">SeqFeature</tt> 设计了众多针对location对象的功能，包含一系列简写的属性。</p>
<dl class="last docutils">
<dt><strong>.ref</strong></dt>
<dd>– <tt class="docutils literal">.location.ref</tt> 简写 --location对象相关的参考序列。通常为空（None）。</dd>
<dt><strong>.ref_db</strong></dt>
<dd>– <tt class="docutils literal">.location.ref_db</tt> 简写 -- 指定 <tt class="docutils literal">.ref</tt> 相关数据库名称。通常为空（None）。</dd>
<dt><strong>.strand</strong></dt>
<dd>– <tt class="docutils literal">.location.strand</tt> 简写 -- 表示feature所处序列的strand。在双链核酸序列中，1表示正链, -1表示负链, 0 表示strand信息很重要但未知, None表示strand信息未知且不重要。蛋白和单链核酸序列为None。</dd>
</dl>
</dd>
<dt><strong>.qualifiers</strong></dt>
<dd>– 存储feature附加信息（Python字典）。键（key）为值（value）所存信息的单字简要描述，值为实际信息。比如，键为 “evidence” ，而值为 “computational (non-experimental)”。 这只是为了提醒人们注意，该feature没有被实验所证实（湿实验）。Note：为与GenBank/EMBL文件中的feature tables对应，规定.qualifiers 中值为字符串数组（即使只有一个字符串）。</dd>
<dt><strong>.sub_features</strong></dt>
<dd>– 只有在描述复杂位置时才使用，如 GenBank/EMBL文件中的 ‘joins’ 位置。 已被 <tt class="docutils literal">CompoundLocation</tt> 对象取代，因此略过不提。</dd>
</dl>
</div>
<div class="section" id="positionslocations">
<h3>4.3.2  Positions和locations</h3>
<p><tt class="docutils literal">SeqFeature</tt> 对象主要用于描述相对于父序列中的位置（region）信息。Region用location对象表示，通常是两个position间的范围。为了区分location和position，我们定义如下:</p>
<dl class="docutils">
<dt><strong>position</strong></dt>
<dd>– 表示位于序列中的单一位置, 可以是精确的也可以是不确定的位置（如5, 20, <tt class="docutils literal">&lt;100</tt> 和 <tt class="docutils literal">&gt;200</tt> ）。</dd>
<dt><strong>location</strong></dt>
<dd>– 介于两个positions间的区域。比如5..20 (5到20)。</dd>
</dl>
<p>之所以特意提及这两个概念是因为我经常混淆两者。</p>
<div class="section" id="featurelocation">
<h4>4.3.2.1  FeatureLocation 对象</h4>
<p>多数 <tt class="docutils literal">SeqFeature</tt> 特别简单（真核基因例外），只需起点、终点以及strand信息。最基本的 <tt class="docutils literal">FeatureLocation</tt> 对象中通常包括上述三点信息。</p>
<p>但实际情况未必如此简单，因为我们还需处理包含几个区域的复合locations，而且position本身很可能是不精确的。</p>
</div>
<div class="section" id="compoundlocation">
<h4>4.3.2.2  CompoundLocation 对象</h4>
<p>为了更方便的处理EMBL/GenBank文件中的 ‘join’ locations，Biopython 1.62引入 <tt class="docutils literal">CompoundLocation</tt> 对象。</p>
</div>
<div class="section" id="positions">
<h4>4.3.2.3  模糊Positions</h4>
<p>目前，我们只处理过简单position，feature location复杂因素之一就是由position本身
不准确所致。生物学中许多问题都是不确定的，比如：你通过双核苷酸priming证明了mRNA
的转录起始位点是这两个位点中的一个。这是十分有价值的发现，但困难来自于怎样表述这
个位点信息。为了处理类似情况，我们用模糊位点（fuzzy position）表示。根据fuzzy
position的不同，我们用5个类分别描述:</p>
<dl class="docutils">
<dt><strong>ExactPosition</strong></dt>
<dd>– 精确位点，用一个数字表示。从该对象的 <tt class="docutils literal">position</tt> 属性可得知精确位点信息。</dd>
<dt><strong>BeforePosition</strong></dt>
<dd>– 位于某个特定位点前。如 <tt class="docutils literal">`&lt;13'</tt> , 在GenBank/EMBL中代表实际位点位于13之前。从该对象的 <tt class="docutils literal">position</tt> 属性可得知上边界信息。</dd>
<dt><strong>AfterPosition</strong></dt>
<dd>– 与 <tt class="docutils literal">BeforePosition</tt> 相反,如 <tt class="docutils literal">`&gt;13'</tt> , 在GenBank/EMBL中代表实际位点位于13以后。从该对象的 <tt class="docutils literal">position</tt> 属性可获知下边界信息。</dd>
<dt><strong>WithinPosition</strong></dt>
<dd>– 介于两个特定位点之间，偶尔在GenBank/EMBL locations用到。如 ‘(1.5)’, GenBank/EMBL中代表实际位点位于1到5之间。该对象需要两个position属性表示，第一个 <tt class="docutils literal">position</tt> 表示下边界（本例为1）， <tt class="docutils literal">extension</tt> 表示上边界与下边界的差值（本例为4）。因此在WithinPosition中， <tt class="docutils literal">object.position</tt> 表示下边界， <tt class="docutils literal">object.position + object.extension</tt> 表示上边界。</dd>
<dt><strong>OneOfPosition</strong></dt>
<dd>– 表示几个位点中的一个（GenBank/EMBL文件中偶尔能看到），比如在基因起始位点不明确或者有两个候选位点的时候可以使用，或者用于明确表示两个相关基因特征时使用。</dd>
<dt><strong>UnknownPosition</strong></dt>
<dd>– 代表未知位点。在GenBank/EMBL文件中没有使用，对应 UniProt中的 ‘?’ feature坐标。</dd>
</dl>
<p>举例说明创建一个fuzzy end points:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqFeature</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">start_pos</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="operator">.</span><span class="name">AfterPosition</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">end_pos</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="operator">.</span><span class="name">BetweenPosition</span><span class="punctuation">(</span><span class="literal number integer">9</span><span class="punctuation">,</span> <span class="name">left</span><span class="operator">=</span><span class="literal number integer">8</span><span class="punctuation">,</span> <span class="name">right</span><span class="operator">=</span><span class="literal number integer">9</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_location</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="operator">.</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="name">start_pos</span><span class="punctuation">,</span> <span class="name">end_pos</span><span class="punctuation">)</span>
</pre>
<p>Note：Biopython 1.59以后，fuzzy-locations有修改, 特别是BetweenPosition和
WithinPosition，现在必须显示用整数表示。起点为较小值，终点则为较大值。</p>
<p>print输出 <tt class="docutils literal">FeatureLocation</tt> 对象，可看到简洁的结果:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_location</span>
<span class="punctuation">[</span><span class="operator">&gt;</span><span class="literal number integer">5</span><span class="punctuation">:(</span><span class="literal number integer">8</span><span class="operator">^</span><span class="literal number integer">9</span><span class="punctuation">)]</span>
</pre>
<p>也可通过start和end属性得到fuzzy position的起始/终止位点:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_location</span><span class="operator">.</span><span class="name">start</span>
<span class="name">AfterPosition</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_location</span><span class="operator">.</span><span class="name">start</span>
<span class="operator">&gt;</span><span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_location</span><span class="operator">.</span><span class="name">end</span>
<span class="name">BetweenPosition</span><span class="punctuation">(</span><span class="literal number integer">9</span><span class="punctuation">,</span> <span class="name">left</span><span class="operator">=</span><span class="literal number integer">8</span><span class="punctuation">,</span> <span class="name">right</span><span class="operator">=</span><span class="literal number integer">9</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_location</span><span class="operator">.</span><span class="name">end</span>
<span class="punctuation">(</span><span class="literal number integer">8</span><span class="operator">^</span><span class="literal number integer">9</span><span class="punctuation">)</span>
</pre>
<p>如果你只想获取数字，不理会模糊positions，则可将fuzzy position强制转换成一个整数:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">int</span><span class="punctuation">(</span><span class="name">my_location</span><span class="operator">.</span><span class="name">start</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">int</span><span class="punctuation">(</span><span class="name">my_location</span><span class="operator">.</span><span class="name">end</span><span class="punctuation">)</span>
<span class="literal number integer">9</span>
</pre>
<p>为了兼容旧版Biopython，保留了整数形式的 <tt class="docutils literal">nofuzzy_start</tt> and <tt class="docutils literal">nofuzzy_end</tt> :</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_location</span><span class="operator">.</span><span class="name">nofuzzy_start</span>
<span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_location</span><span class="operator">.</span><span class="name">nofuzzy_end</span>
<span class="literal number integer">9</span>
</pre>
<p>Notice：上述例子只是为了帮助你理解fuzzy locations。</p>
<p>相似的，如果要建立一个精确location，只需将整数传递给 <tt class="docutils literal">FeaturePosition</tt>
构造函数, 即可建立 <tt class="docutils literal">ExactPosition</tt> 对象:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exact_location</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="operator">.</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="literal number integer">9</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">exact_location</span>
<span class="punctuation">[</span><span class="literal number integer">5</span><span class="punctuation">:</span><span class="literal number integer">9</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exact_location</span><span class="operator">.</span><span class="name">start</span>
<span class="name">ExactPosition</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">int</span><span class="punctuation">(</span><span class="name">exact_location</span><span class="operator">.</span><span class="name">start</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exact_location</span><span class="operator">.</span><span class="name">nofuzzy_start</span>
<span class="literal number integer">5</span>
</pre>
<p>以上是Biopython处理fuzzy position的实现方法。希望读者能体会之所以这样设计，都是为了使用上的方便（至少不比精确位点复杂）</p>
</div>
<div class="section" id="location-testing">
<h4>4.3.2.4  Location testing</h4>
<p>可用Python关键词 <tt class="docutils literal">in</tt> 检验某个碱基或氨基酸残基的父坐标是否位于
feature/location中。</p>
<p>假定你想知道某个SNP位于哪个feature里，并知道该SNP的索引位置是4350（Python 计数）。一个简单的实现方案是用循环遍历所有features:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_snp</span> <span class="operator">=</span> <span class="literal number integer">4350</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">feature</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">my_snp</span> <span class="operator word">in</span> <span class="name">feature</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="name">feature</span><span class="operator">.</span><span class="name">type</span><span class="punctuation">,</span> <span class="name">feature</span><span class="operator">.</span><span class="name">qualifiers</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">'db_xref'</span><span class="punctuation">)</span>
<span class="operator">...</span>
<span class="name">source</span> <span class="punctuation">[</span><span class="literal string">'taxon:229193'</span><span class="punctuation">]</span>
<span class="name">gene</span> <span class="punctuation">[</span><span class="literal string">'GeneID:2767712'</span><span class="punctuation">]</span>
<span class="name">CDS</span> <span class="punctuation">[</span><span class="literal string">'GI:45478716'</span><span class="punctuation">,</span> <span class="literal string">'GeneID:2767712'</span><span class="punctuation">]</span>
</pre>
<p>Note： GenBank /EMBL 文件中的 gene 和CDS features（ <tt class="docutils literal">join</tt> ）只包含外显子，不含内含子。</p>
</div>
</div>
<div class="section" id="feature-location">
<h3>4.3.3  使用feature 或 location描述序列</h3>
<p><tt class="docutils literal">SeqFeature</tt> 或 location object对象并没有直接包含任何序列，只是可根据储存的location (见Section <a class="reference external" href="#sec:locations">4.3.2</a>)，从父序列中取得。例如：某一短基因位于负链5:18位置，由于GenBank/EMBL文件以1开始计数，Biopython中表示为 <tt class="docutils literal"><span class="pre">complement(6..18)</span></tt> :</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqFeature</span> <span class="keyword namespace">import</span> <span class="name">SeqFeature</span><span class="punctuation">,</span> <span class="name">FeatureLocation</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">example_parent</span> <span class="operator">=</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACCGAGACGGCAAAGGCTAGCATAGGTATGAGACTTCCTTCCTGCCAGTGCTGAGGAACTGGGAGCCTAC&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">example_feature</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="punctuation">(</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="literal number integer">18</span><span class="punctuation">),</span> <span class="name builtin">type</span><span class="operator">=</span><span class="literal string">&quot;gene&quot;</span><span class="punctuation">,</span> <span class="name">strand</span><span class="operator">=-</span><span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>你可以用切片从父序列截取5:18,然后取反向互补序列。如果是Biopython 1.59或以后版本，可使用如下方法:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">feature_seq</span> <span class="operator">=</span> <span class="name">example_parent</span><span class="punctuation">[</span><span class="name">example_feature</span><span class="operator">.</span><span class="name">location</span><span class="operator">.</span><span class="name">start</span><span class="punctuation">:</span><span class="name">example_feature</span><span class="operator">.</span><span class="name">location</span><span class="operator">.</span><span class="name">end</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">feature_seq</span>
<span class="name">AGCCTTTGCCGTC</span>
</pre>
<p>不过在处理复合 features (joins)时，此法相当繁琐。此时可以使用 <tt class="docutils literal">SeqFeature</tt> 对象的 <tt class="docutils literal">extract</tt> 方法处理:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">feature_seq</span> <span class="operator">=</span> <span class="name">example_feature</span><span class="operator">.</span><span class="name">extract</span><span class="punctuation">(</span><span class="name">example_parent</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">feature_seq</span>
<span class="name">AGCCTTTGCCGTC</span>
</pre>
<p><tt class="docutils literal">SeqFeature</tt> 或 location对象的长度等同于所表示序列的长度。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">example_feature</span><span class="operator">.</span><span class="name">extract</span><span class="punctuation">(</span><span class="name">example_parent</span><span class="punctuation">)</span>
<span class="name">AGCCTTTGCCGTC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">example_feature</span><span class="operator">.</span><span class="name">extract</span><span class="punctuation">(</span><span class="name">example_parent</span><span class="punctuation">))</span>
<span class="literal number integer">13</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">example_feature</span><span class="punctuation">)</span>
<span class="literal number integer">13</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">example_feature</span><span class="operator">.</span><span class="name">location</span><span class="punctuation">)</span>
<span class="literal number integer">13</span>
</pre>
<p>简单 <tt class="docutils literal">FeatureLocation</tt> 对象的长度等于终止osition减去起始position的差值；而 <tt class="docutils literal">CompoundLocation</tt> 的长度则为各片段长度之和。</p>
</div>
</div>
<div class="section" id="references">
<h2>4.4  References</h2>
<p>对一条序列的注释还包括参考文献（reference），Biopython通过
<tt class="docutils literal">Bio.SeqFeature.Reference</tt> 对象来储存相关的文献信息。</p>
<p>References属性储存了 <tt class="docutils literal">期刊名</tt> 、 <tt class="docutils literal">题名</tt> 、 <tt class="docutils literal">作者</tt> 等信息。此外还包括 <tt class="docutils literal">medline_id</tt> 、 <tt class="docutils literal">pubmed_id</tt> 以及 <tt class="docutils literal">comment</tt> 。</p>
<p>通常reference 也有 <tt class="docutils literal">location</tt> 对象，便于文献涉及研究对象在序列中的定位。该 <tt class="docutils literal">location</tt> 有可能是一个fuzzy location（见 section <a class="reference external" href="#sec:locations">4.3.2</a>）。</p>
<p>文献对象都以列表储存在 <tt class="docutils literal">SeqRecord</tt> 对象的 <tt class="docutils literal">annotations</tt> 字典中。 字典的键为 “references”。reference对象也是为了方便处理文献而设计，希望能满足各种使用需求。</p>
</div>
<div class="section" id="id27">
<h2>4.5  格式化方法</h2>
<p><tt class="docutils literal">SeqRecord</tt> 类中的 <tt class="docutils literal">format()</tt> 能将字符串转换成被 <tt class="docutils literal">Bio.SeqIO</tt> 支持的格式，如FASTA:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_protein</span>

<span class="name">record</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD&quot;</span> \
                      <span class="operator">+</span><span class="literal string">&quot;GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK&quot;</span> \
                      <span class="operator">+</span><span class="literal string">&quot;NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM&quot;</span> \
                      <span class="operator">+</span><span class="literal string">&quot;SSAC&quot;</span><span class="punctuation">,</span> <span class="name">generic_protein</span><span class="punctuation">),</span>
                   <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;gi|14150838|gb|AAK54648.1|AF376133_1&quot;</span><span class="punctuation">,</span>
                   <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;chalcone synthase [Cucumis sativus]&quot;</span><span class="punctuation">)</span>

<span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>输出为:</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">14150838</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AAK54648</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF376133_1</span> <span class="name">chalcone</span> <span class="name">synthase</span> <span class="punctuation">[</span><span class="name">Cucumis</span> <span class="name">sativus</span><span class="punctuation">]</span>
<span class="name">MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD</span>
<span class="name">GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK</span>
<span class="name">NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM</span>
<span class="name">SSAC</span>
</pre>
<p><tt class="docutils literal">format</tt> 方法接收单个必选参数，小写字母字符串是 <tt class="docutils literal">Bio.SeqIO</tt> 模块支持的输出格式 (见Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>)。然而，此 <tt class="docutils literal">format()</tt> 方法并不适用于包含多条序列的文件格式 (如多序列比对格式)（详见Section <a class="reference external" href="#sec:Bio.SeqIO-and-StringIO">5.5.4</a>）。</p>
</div>
<div class="section" id="id28">
<h2>4.6  SeqRecord切片</h2>
<p>通过切片截取 <tt class="docutils literal">SeqRecord</tt> 的部分序列可得到一条新的 <tt class="docutils literal">SeqRecord</tt> 。此处需引起注意的是per-letter annotations也被取切片, 但新序列中的features保持不变 (locations相应调整)。</p>
<p>以前述Genbank文件为例:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'</span><span class="punctuation">,</span>
<span class="name">IUPACAmbiguousDNA</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'NC_005816.1'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'NC_005816'</span><span class="punctuation">,</span>
<span class="name">description</span><span class="operator">=</span><span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span><span class="punctuation">,</span>
<span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">'Project:10638'</span><span class="punctuation">])</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="literal number integer">9609</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">)</span>
<span class="literal number integer">41</span>
</pre>
<p>本例中，我们关注 <tt class="docutils literal">YP_pPCP05</tt> 质粒上的 <tt class="docutils literal">pim</tt> 基因。从GenBank文件可直接看出 <tt class="docutils literal">pim</tt> gene/CDS location是 <tt class="docutils literal"><span class="pre">4343..4780</span></tt> （相应的Python 位置是 <tt class="docutils literal">4342:4780</tt> ）。Location信息位于GenBank文件第12和13 entries中, 由于python以0开始计数，因此python中，它们是 <tt class="docutils literal">features</tt> 列表中的 entries 11和12:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">[</span><span class="literal number integer">20</span><span class="punctuation">]</span>
<span class="name builtin">type</span><span class="punctuation">:</span> <span class="name">gene</span>
<span class="name">location</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">4342</span><span class="punctuation">:</span><span class="literal number integer">4780</span><span class="punctuation">](</span><span class="operator">+</span><span class="punctuation">)</span>
<span class="name">qualifiers</span><span class="punctuation">:</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">db_xref</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'GeneID:2767712'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'pim'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">locus_tag</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'YP_pPCP05'</span><span class="punctuation">]</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">[</span><span class="literal number integer">21</span><span class="punctuation">]</span>
<span class="name builtin">type</span><span class="punctuation">:</span> <span class="name">CDS</span>
<span class="name">location</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">4342</span><span class="punctuation">:</span><span class="literal number integer">4780</span><span class="punctuation">](</span><span class="operator">+</span><span class="punctuation">)</span>
<span class="name">qualifiers</span><span class="punctuation">:</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">codon_start</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'1'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">db_xref</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'GI:45478716'</span><span class="punctuation">,</span> <span class="literal string">'GeneID:2767712'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'pim'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">locus_tag</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'YP_pPCP05'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">note</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'similar to many previously sequenced pesticin immunity ...'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">product</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'pesticin immunity protein'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">protein_id</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'NP_995571.1'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">transl_table</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'11'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">translation</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'MGGGMISKLFCLALIFLSSSGLAEKNTYTAKDILQNLELNTFGNSLSH...'</span><span class="punctuation">]</span>
</pre>
<p>从父记录中取切片（4300 到 4800），观测所得到的features数量:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">4300</span><span class="punctuation">:</span><span class="literal number integer">4800</span><span class="punctuation">]</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ATAAATAGATTATTCCAAATAATTTATTTATGTAAGAACAGGATGGGAGGGGGA...TTA'</span><span class="punctuation">,</span>
<span class="name">IUPACAmbiguousDNA</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'NC_005816.1'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'NC_005816'</span><span class="punctuation">,</span>
<span class="name">description</span><span class="operator">=</span><span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span><span class="punctuation">,</span>
<span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[])</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">sub_record</span><span class="punctuation">)</span>
<span class="literal number integer">500</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">sub_record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">)</span>
<span class="literal number integer">2</span>
</pre>
<p>子记录（sub_record）只包括两个features, 分别是 <tt class="docutils literal">YP_pPCP05</tt> 质粒的gene和CDS:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="name builtin">type</span><span class="punctuation">:</span> <span class="name">gene</span>
<span class="name">location</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">42</span><span class="punctuation">:</span><span class="literal number integer">480</span><span class="punctuation">](</span><span class="operator">+</span><span class="punctuation">)</span>
<span class="name">qualifiers</span><span class="punctuation">:</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">db_xref</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'GeneID:2767712'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'pim'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">locus_tag</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'YP_pPCP05'</span><span class="punctuation">]</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">[</span><span class="literal number integer">20</span><span class="punctuation">]</span>
<span class="name builtin">type</span><span class="punctuation">:</span> <span class="name">CDS</span>
<span class="name">location</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">42</span><span class="punctuation">:</span><span class="literal number integer">480</span><span class="punctuation">](</span><span class="operator">+</span><span class="punctuation">)</span>
<span class="name">qualifiers</span><span class="punctuation">:</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">codon_start</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'1'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">db_xref</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'GI:45478716'</span><span class="punctuation">,</span> <span class="literal string">'GeneID:2767712'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'pim'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">locus_tag</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'YP_pPCP05'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">note</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'similar to many previously sequenced pesticin immunity ...'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">product</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'pesticin immunity protein'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">protein_id</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'NP_995571.1'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">transl_table</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'11'</span><span class="punctuation">]</span>
    <span class="name">Key</span><span class="punctuation">:</span> <span class="name">translation</span><span class="punctuation">,</span> <span class="name">Value</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'MGGGMISKLFCLALIFLSSSGLAEKNTYTAKDILQNLELNTFGNSLSH...'</span><span class="punctuation">]</span>
</pre>
<p>注意：locations已被调整至对应生成的新父序列!</p>
<p>尽可能灵敏和直观地获取子记录的相关特征（和任意的per-letter annotation），但是对于其余注释，Biopython无法判断是否仍然适用于子记录。因此子记录忽略了 <tt class="docutils literal">annotations</tt> 和 <tt class="docutils literal">dbxrefs</tt> 以避免引起歧义。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">annotations</span>
<span class="punctuation">{}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="punctuation">[]</span>
</pre>
<p>为了便于实际操作，子记录保留了 <tt class="docutils literal">id</tt> , <tt class="docutils literal">name</tt> 和 <tt class="docutils literal">description</tt> :</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">id</span>
<span class="literal string">'NC_005816.1'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'NC_005816'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">description</span>
<span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span>
</pre>
<p>上述例子很好的展示了问题，由于子记录不包括完整的质粒序列，因此description是错的。我们可以将子记录看做是截短版的GenBank文件，可用Section <cite>4.5 &lt;#sec:SeqRecord-format&gt;`__中所述 ``format`</cite> 方法纠正：
:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, partial.&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">sub_record</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>FASTQ例子参见 Sections <a class="reference external" href="#sec:FASTQ-slicing-off-primer">18.1.7</a>
和 <a class="reference external" href="#sec:FASTQ-slicing-off-adaptor">18.1.8</a> （此例中per-letter annotations (read质量分数) 也被取切片）。</p>
</div>
<div class="section" id="id29">
<h2>4.7  SeqRecord对象相加</h2>
<p><tt class="docutils literal">SeqRecord</tt> 对象可相加得到一个新的 <tt class="docutils literal">SeqRecord</tt> 。注意：per-letter annotations也相加, features (locations 调整)；而其它annotation 保持不变(如id、name和description)。</p>
<p>以FASTQ 文件中的第一条记录为例说明per-letter annotation （Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a> 详细介绍 <tt class="docutils literal">SeqIO</tt> 函数）:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;example.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="literal number integer">25</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">CCCTTCTTGTCTTCAGCGTTTCTCC</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">18</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">22</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span>
<span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">23</span><span class="punctuation">,</span> <span class="literal number integer">23</span><span class="punctuation">]</span>
</pre>
<p>假设上述序列数据来自Roche 454测序, 你根据其它信息得知 <tt class="docutils literal">TTT</tt> 应该是 <tt class="docutils literal">TT</tt> 。此时可分别用切片提取第三个 <tt class="docutils literal">T</tt> 前后的序列（ <tt class="docutils literal">SeqRecord</tt> ）:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">left</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[:</span><span class="literal number integer">20</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">left</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">CCCTTCTTGTCTTCAGCGTT</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">left</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">18</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">22</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">right</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">21</span><span class="punctuation">:]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">right</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">CTCC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">right</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">23</span><span class="punctuation">,</span> <span class="literal number integer">23</span><span class="punctuation">]</span>
</pre>
<p>两部分相加:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">edited</span> <span class="operator">=</span> <span class="name">left</span> <span class="operator">+</span> <span class="name">right</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">edited</span><span class="punctuation">)</span>
<span class="literal number integer">24</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">edited</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">CCCTTCTTGTCTTCAGCGTTCTCC</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">edited</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">18</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">22</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span>
<span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal number integer">23</span><span class="punctuation">,</span> <span class="literal number integer">23</span><span class="punctuation">]</span>
</pre>
<p>很容易和直观吧！上述两步可合并:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">edited</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[:</span><span class="literal number integer">20</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">21</span><span class="punctuation">:]</span>
</pre>
<p>现在以GenBank文件（假定是环状基因组）为例说明features:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG'</span><span class="punctuation">,</span>
<span class="name">IUPACAmbiguousDNA</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'NC_005816.1'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'NC_005816'</span><span class="punctuation">,</span>
<span class="name">description</span><span class="operator">=</span><span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span><span class="punctuation">,</span>
<span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">'Project:10638'</span><span class="punctuation">])</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="literal number integer">9609</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">)</span>
<span class="literal number integer">41</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="punctuation">[</span><span class="literal string">'Project:58037'</span><span class="punctuation">]</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">annotations</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'comment'</span><span class="punctuation">,</span> <span class="literal string">'sequence_version'</span><span class="punctuation">,</span> <span class="literal string">'source'</span><span class="punctuation">,</span> <span class="literal string">'taxonomy'</span><span class="punctuation">,</span> <span class="literal string">'keywords'</span><span class="punctuation">,</span> <span class="literal string">'references'</span><span class="punctuation">,</span>
<span class="literal string">'accessions'</span><span class="punctuation">,</span> <span class="literal string">'data_file_division'</span><span class="punctuation">,</span> <span class="literal string">'date'</span><span class="punctuation">,</span> <span class="literal string">'organism'</span><span class="punctuation">,</span> <span class="literal string">'gi'</span><span class="punctuation">]</span>
</pre>
<p>可改变起点:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">2000</span><span class="punctuation">:]</span> <span class="operator">+</span> <span class="name">record</span><span class="punctuation">[:</span><span class="literal number integer">2000</span><span class="punctuation">]</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GATACGCAGTCATATTTTTTACACAATTCTCTAATCCCGACAAGGTCGTAGGTC...GGA'</span><span class="punctuation">,</span>
<span class="name">IUPACAmbiguousDNA</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'NC_005816.1'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'NC_005816'</span><span class="punctuation">,</span>
<span class="name">description</span><span class="operator">=</span><span class="literal string">'Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.'</span><span class="punctuation">,</span>
<span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[])</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">shifted</span><span class="punctuation">)</span>
<span class="literal number integer">9609</span>
</pre>
<p>Note: 上述方法并不完美（丢失了数据库交叉引用dbxrefs 和源feature）:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">shifted</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">)</span>
<span class="literal number integer">40</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="punctuation">[]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span><span class="operator">.</span><span class="name">annotations</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[]</span>
</pre>
<p>这是因为 <tt class="docutils literal">SeqRecord</tt> 切片对 annotation 保留非常谨慎 (错误保留 annotation 可能引起大问题)。如果你想保留数据库的交叉引用dbxrefs和其余annotations 字典必须明确说明，才能得以保留:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span><span class="operator">.</span><span class="name">dbxrefs</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span><span class="punctuation">[:]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span><span class="operator">.</span><span class="name">annotations</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">annotations</span><span class="operator">.</span><span class="name">copy</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="punctuation">[</span><span class="literal string">'Project:10638'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shifted</span><span class="operator">.</span><span class="name">annotations</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'comment'</span><span class="punctuation">,</span> <span class="literal string">'sequence_version'</span><span class="punctuation">,</span> <span class="literal string">'source'</span><span class="punctuation">,</span> <span class="literal string">'taxonomy'</span><span class="punctuation">,</span> <span class="literal string">'keywords'</span><span class="punctuation">,</span> <span class="literal string">'references'</span><span class="punctuation">,</span>
<span class="literal string">'accessions'</span><span class="punctuation">,</span> <span class="literal string">'data_file_division'</span><span class="punctuation">,</span> <span class="literal string">'date'</span><span class="punctuation">,</span> <span class="literal string">'organism'</span><span class="punctuation">,</span> <span class="literal string">'gi'</span><span class="punctuation">]</span>
</pre>
<p>Note: 此例中序列record的identifiers也应调整（因为NCBI的reference链接的是未经修改的 <em>原始</em> 序列）。</p>
</div>
<div class="section" id="id30">
<h2>4.8  反向互补SeqRecord对象</h2>
<p>为消除序列反向互补后annotation改变带来的困难，Biopython 1.57 <tt class="docutils literal">SeqRecord</tt> 对象加入了 <tt class="docutils literal">reverse_complement</tt> 方法。这也成为Biopython 1.57的新特性之一。</p>
<p>序列用Seq对象中的reverse_complement方法反向互补。Features随location而改变，strand也被重新计算。复制并反转per-letter-annotation（通常情况下这种做法比较合适，如对质量分数注释的反转）。然而多数annotation的转变却存有问题。</p>
<p>比如record ID是accession号，该accession不应被用于反向互补序列。默认identifier转换可导致后续分析中的轻度数据损坏。因此 <tt class="docutils literal">SeqRecord</tt> 的id、
name、description、annotations和dbxrefs默认不变。</p>
<p><tt class="docutils literal">SeqRecord</tt> 对象的 <tt class="docutils literal">reverse_complement</tt> 法用多个可选参数以对应record的属性。将这些参数设为 <tt class="docutils literal">True</tt> 表示复制旧值；而 <tt class="docutils literal">False</tt> 意为用缺省值替换旧值。当然也可自定义新值。</p>
<p>举例:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">)</span>
<span class="name">NC_005816</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="literal number integer">9609</span> <span class="literal number integer">41</span> <span class="literal number integer">1</span> <span class="literal number integer">11</span>
</pre>
<p>反向互补该record并给ID赋予新值 - 注意：多数annotation丢失，而features仍在:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">rc</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;TESTING&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">rc</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rc</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rc</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rc</span><span class="operator">.</span><span class="name">dbxrefs</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rc</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">)</span>
<span class="name">TESTING</span> <span class="literal number integer">9609</span> <span class="literal number integer">41</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span>
</pre>
</div>
</div>
<div class="section" id="id31">
<h1>第5章  序列输入和输出</h1>
<p>本章将详细讨论 <tt class="docutils literal">Bio.SeqIO</tt> 模块，该模块在第 <a class="reference external" href="#chapter:quick-start">2</a> 章已经做过简单的介绍并在第 <a class="reference external" href="#chapter:SeqRecord">4</a> 章使用过，它旨在提供一个简单的接口，实现对各种不同格式序列文件进行统一的处理。详细信息请查阅 <tt class="docutils literal">Bio.SeqIO</tt> 维基页面（ <a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a> ）和内置文档（ <a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SeqIO-module.html">SeqIO</a> ）:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>学习本章的要领是学会使用 <tt class="docutils literal">SeqRecord</tt> 对象（请见第<a class="reference external" href="#chapter:SeqRecord">4</a> 章），该对象包含一个 <tt class="docutils literal">Seq</tt> 对象（请见第 <a class="reference external" href="#chapter:Bio.Seq">3</a> 章）和注释信息（如序列ID和描述信息）。</p>
<div class="section" id="id32">
<h2>5.1 解析/读取序列</h2>
<p>该模块的主要函数是 <tt class="docutils literal">Bio.SeqIO.parse()</tt> ，它用于读取序列文件生成 <tt class="docutils literal">SeqRecord</tt> 对象，包含两个参数：</p>
<ol class="arabic simple">
<li>第一个参数是一个文件名或者一个句柄（ <em>handle</em> ）。句柄可以是打开的文件，命令行程序的输出，或者来自下载的数据(请见第 <a class="reference external" href="#sec:SeqIO_Online">5.3</a> 节)。更多关于句柄的信息请见第 <a class="reference external" href="#sec:appendix-handles">22.1</a> 节。</li>
<li>第二个参数是一个小写字母字符串，用于指定序列格式（我们并不推测文件格式！），支持的文件格式请见 <a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a> 。</li>
</ol>
<p>还有一个用于指定字符集的 <tt class="docutils literal">alphabet</tt> 参数，这对FASTA这样的文件格式非常有用，在这里 <tt class="docutils literal">Bio.SeqIO</tt> 默认参数为字母表。</p>
<p><tt class="docutils literal">Bio.SeqIO.parse()</tt> 函数返回一个 <tt class="docutils literal">SeqRecord</tt> 对象迭代器（ <em>iterator</em> ），迭代器通常用在循环中。</p>
<p>有时你需要处理只包含一个序列条目的文件，此时请使用函数 <tt class="docutils literal">Bio.SeqIO.read()</tt> 。它使用与函数 <tt class="docutils literal">Bio.SeqIO.parse()</tt> 相同的参数，当文件有且仅有一个序列条目时返回一个 <tt class="docutils literal">SeqRecord</tt> 对象，否则触发异常。</p>
<div class="section" id="id33">
<h3>5.1.1 读取序列文件</h3>
<p>总的来说， <tt class="docutils literal">Bio.SeqIO.parse()</tt> 用于读取序列文件并返回 <tt class="docutils literal">SeqRecord</tt> 对象，并通常用在循环中，如：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span>
    <span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
    <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="punctuation">)</span>
</pre>
<p>上面的示例来自第 <a class="reference external" href="#sec:sequence-parsing">2.4</a> 节，它将读取来自FASTA格式文件 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta">ls_orchid.fasta</a> 的兰花DNA序列。如果你想读取GenBank格式文件，如 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk">ls_orchid.gbk</a> ，只需要更改文件名和格式字符串：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span>
    <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span>
    <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="punctuation">)</span>
</pre>
<p>同样地，如果需要读取其他格式文件，并且 <tt class="docutils literal">Bio.SeqIO.parse()</tt> 支持该文件格式，你只需要修改到相应的格式字符串，如“swiss”为SwissProt格式文件，“embl”为EMBL格式文本文件。详细的清单请见维基页面（ <a class="reference external" href="http://biopython.org/wiki/SeqIO">http://biopython.org/wiki/SeqIO</a> ）和内置文档（ <a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SeqIO-module.html">在线文档</a> ）。</p>
<p>另外一个非常常见的使用Python迭代器的地方是在列表解析（list comprehension，或者生成器表达式generator expression）。例如，如果需要从文件中提取序列ID列表，我们可以通过以下的列表推导很容易地实现：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">identifiers</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span> <span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">identifiers</span>
<span class="punctuation">[</span><span class="literal string">'Z78533.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78532.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78531.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78530.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78529.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78527.1'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Z78439.1'</span><span class="punctuation">]</span>
</pre>
<p>更多关于 <tt class="docutils literal">SeqIO.parse()</tt> 在列表推导中运用的示例请见第 <a class="reference external" href="#seq:sequence-parsing-plus-pylab">18.2</a> 节（e.g. 对序列长度或GC%作图）。</p>
</div>
<div class="section" id="id34">
<h3>5.1.2 遍历序列文件</h3>
<p>在上述示例中，我们通常使用for循环遍历所有的序列条目（records）。你可以对for循环使用所有类型的支持迭代接口的Python对象（包括列表，元组（tuple）和字符串）。</p>
<p><tt class="docutils literal">Bio.SeqIO</tt> 返回的对象实际上是一个返回 <tt class="docutils literal">SeqRecord</tt> 对象的迭代器。你将顺序地获得每个条目，但是有且仅有一次；优势是，当处理大文件时，迭代器可以有效地节约内存空间。</p>
<p>除了使用for循环，还可以使用迭代器的 <tt class="docutils literal">.next()</tt> 方法遍历序列条目，如：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">record_iterator</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>

<span class="name">first_record</span> <span class="operator">=</span> <span class="name">record_iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">id</span>
<span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">description</span>

<span class="name">second_record</span> <span class="operator">=</span> <span class="name">record_iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="name">second_record</span><span class="operator">.</span><span class="name">id</span>
<span class="keyword">print</span> <span class="name">second_record</span><span class="operator">.</span><span class="name">description</span>
</pre>
<p>注意：如果使用 <tt class="docutils literal">.next()</tt> 方法，当没有序列条目时，将抛出 <tt class="docutils literal">StopIteration</tt> 异常。</p>
<p>一种特殊情形是，序列文件包含多个序列条目，而你只需要第一个条目。在这种情况下，可使用以下代码，非常简洁：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">first_record</span>  <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
</pre>
<p>注意：像上述示例中使用 <tt class="docutils literal">.next()</tt> 方法将忽略文件中其余的序列。如果序列文件“有且仅有”一条序列条目，如本章后面的某些在线示例、包含单条染色体序列的GenBank文件，请使用 <tt class="docutils literal">Bio.SeqIO.read()</tt> 函数。该函数会检查文件是否包含额外的序列条目。</p>
</div>
<div class="section" id="id35">
<h3>5.1.3  获得序列文件中序列条目列表</h3>
<p>在上一节中，我们讨论了如何使用 <tt class="docutils literal">Bio.SeqIO.parse()</tt> 返回一个 <tt class="docutils literal">SeqRecord</tt> 迭代器，然后顺序地获取序列条目。往往我们需要以任意顺序获取序列条目，Python列表数据类型便可以达到这个目的。使用Python内置函数 <tt class="docutils literal">list()</tt> ，我们可以将序列条目迭代器转变成 <tt class="docutils literal">SeqRecord</tt> 对象列表，如下：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">))</span>

<span class="keyword">print</span> <span class="literal string">&quot;Found </span><span class="literal string interpol">%i</span><span class="literal string"> records&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">)</span>

<span class="keyword">print</span> <span class="literal string">&quot;The last record&quot;</span>
<span class="name">last_record</span> <span class="operator">=</span> <span class="name">records</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span> <span class="comment">#using Python's list tricks</span>
<span class="keyword">print</span> <span class="name">last_record</span><span class="operator">.</span><span class="name">id</span>
<span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">last_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">last_record</span><span class="punctuation">)</span>

<span class="keyword">print</span> <span class="literal string">&quot;The first record&quot;</span>
<span class="name">first_record</span> <span class="operator">=</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span> <span class="comment">#remember, Python counts from zero</span>
<span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">id</span>
<span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">first_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">first_record</span><span class="punctuation">)</span>
</pre>
<p>运行结果:</p>
<pre class="code python literal-block">
<span class="name">Found</span> <span class="literal number integer">94</span> <span class="name">records</span>
<span class="name">The</span> <span class="name">last</span> <span class="name">record</span>
<span class="name">Z78439</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
<span class="literal number integer">592</span>
<span class="name">The</span> <span class="name">first</span> <span class="name">record</span>
<span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
<span class="literal number integer">740</span>
</pre>
<p>当然，你仍然可以对 <tt class="docutils literal">SeqRecord</tt> 对象列表使用for循环。使用列表比使用迭代器灵活得多（例如，可以根据列表大小知道序列条目数量），但缺点是for循环要同时读取所有的内容，需要更多的内存空间。</p>
</div>
<div class="section" id="id36">
<h3>5.1.4 提取数据</h3>
<p><tt class="docutils literal">SeqRecord</tt> 对象及其注释信息在第 <a class="reference external" href="#chapter:SeqRecord">4</a> 章中有更详细的介绍。为了解释注释信息是如何存储的，我们从GenBank文件 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk">ls_orchid.gbk</a> 中解析出第一个序列条目，并将其输出：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">record_iterator</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="name">first_record</span> <span class="operator">=</span> <span class="name">record_iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="name">first_record</span>
</pre>
<p>输出结果:</p>
<pre class="code python literal-block">
<span class="name">ID</span><span class="punctuation">:</span> <span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">Name</span><span class="punctuation">:</span> <span class="name">Z78533</span>
<span class="name">Description</span><span class="punctuation">:</span> <span class="name">C</span><span class="operator">.</span><span class="name">irapeanum</span> <span class="literal number float">5.8</span><span class="name">S</span> <span class="name">rRNA</span> <span class="name">gene</span> <span class="operator word">and</span> <span class="name">ITS1</span> <span class="operator word">and</span> <span class="name">ITS2</span> <span class="name">DNA</span><span class="operator">.</span>
<span class="name">Number</span> <span class="name">of</span> <span class="name">features</span><span class="punctuation">:</span> <span class="literal number integer">5</span>
<span class="operator">/</span><span class="name">sequence_version</span><span class="operator">=</span><span class="literal number integer">1</span>
<span class="operator">/</span><span class="name">source</span><span class="operator">=</span><span class="name">Cypripedium</span> <span class="name">irapeanum</span>
<span class="operator">/</span><span class="name">taxonomy</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">'Eukaryota'</span><span class="punctuation">,</span> <span class="literal string">'Viridiplantae'</span><span class="punctuation">,</span> <span class="literal string">'Streptophyta'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Cypripedium'</span><span class="punctuation">]</span>
<span class="operator">/</span><span class="name">keywords</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">'5.8S ribosomal RNA'</span><span class="punctuation">,</span> <span class="literal string">'5.8S rRNA gene'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'ITS1'</span><span class="punctuation">,</span> <span class="literal string">'ITS2'</span><span class="punctuation">]</span>
<span class="operator">/</span><span class="name">references</span><span class="operator">=</span><span class="punctuation">[</span><span class="operator">...</span><span class="punctuation">]</span>
<span class="operator">/</span><span class="name">accessions</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">'Z78533'</span><span class="punctuation">]</span>
<span class="operator">/</span><span class="name">data_file_division</span><span class="operator">=</span><span class="name">PLN</span>
<span class="operator">/</span><span class="name">date</span><span class="operator">=</span><span class="literal number integer">30</span><span class="operator">-</span><span class="name">NOV</span><span class="operator">-</span><span class="literal number integer">2006</span>
<span class="operator">/</span><span class="name">organism</span><span class="operator">=</span><span class="name">Cypripedium</span> <span class="name">irapeanum</span>
<span class="operator">/</span><span class="name">gi</span><span class="operator">=</span><span class="literal number integer">2765658</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>这可以得到 <tt class="docutils literal">SeqRecord</tt> 大部分的易读的注释汇总信息。在此例中，我们将使用 <tt class="docutils literal">.annotations</tt> 属性-即Python字典（dictionary）。该注释字典的内容如上述示例结果，你也可以直接输出：</p>
<pre class="code python literal-block">
<span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">annotations</span>
</pre>
<p>与其他Python字典一样，你可以轻松地获得键列表：</p>
<pre class="code python literal-block">
<span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">annotations</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
</pre>
<p>或者值列表:</p>
<pre class="code python literal-block">
<span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">annotations</span><span class="operator">.</span><span class="name">values</span><span class="punctuation">()</span>
</pre>
<p>通常，注释值是字符串或者字符串列表。一个特例是，文件中的所有参考文献（references）都以引用（reference）对象方式存储。</p>
<p>例如你想从GenBank文件 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk">ls_orchid.gbk</a> 中提取出物种列表。我们需要的信息 <em>Cypripedium irapeanum</em> 被保存在这个注释字典的‘source’和‘organism’键中，我们可以用下面的方式获取：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;source&quot;</span><span class="punctuation">]</span>
<span class="name">Cypripedium</span> <span class="name">irapeanum</span>
</pre>
<p>或:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">first_record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;organism&quot;</span><span class="punctuation">]</span>
<span class="name">Cypripedium</span> <span class="name">irapeanum</span>
</pre>
<p>通常，‘organism’ 用于学名（拉丁名，e.g. <em>Arabidopsis thaliana</em> ），而 ‘source’ 用于俗名（common name）（e.g. thale cress）。在此例中，以及在通常情况下，这两个字段是相同的。</p>
<p>现在，让我们遍历所有的序列条目， 创建一个包含所有兰花序列的物种列表：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">all_species</span> <span class="operator">=</span> <span class="punctuation">[]</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">):</span>
    <span class="name">all_species</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;organism&quot;</span><span class="punctuation">])</span>
<span class="keyword">print</span> <span class="name">all_species</span>
</pre>
<p>另外一种方式是使用列表解析：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">all_species</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;organism&quot;</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> \
               <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)]</span>
<span class="keyword">print</span> <span class="name">all_species</span>
</pre>
<p>两种方式的输出结果相同：</p>
<pre class="code python literal-block">
<span class="punctuation">[</span><span class="literal string">'Cypripedium irapeanum'</span><span class="punctuation">,</span> <span class="literal string">'Cypripedium californicum'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Paphiopedilum barbatum'</span><span class="punctuation">]</span>
</pre>
<p>因为GenBank文件注释是以标准方式注释，所以相当简单。</p>
<p>现在，假设你需要从一个FASTA文件而不是GenBank文件提取出物种列表，那么你不得不多写一些代码，用以从序列条目的描述行提取需要的数据。使用的示例FASTA文件 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta">ls_orchid.fasta</a> 格式如下：</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">2765658</span><span class="operator">|</span><span class="name">emb</span><span class="operator">|</span><span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">CIZ78533</span> <span class="name">C</span><span class="operator">.</span><span class="name">irapeanum</span> <span class="literal number float">5.8</span><span class="name">S</span> <span class="name">rRNA</span> <span class="name">gene</span> <span class="operator word">and</span> <span class="name">ITS1</span> <span class="operator word">and</span> <span class="name">ITS2</span> <span class="name">DNA</span>
<span class="name">CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGGAATAAACGATCGAGTG</span>
<span class="name">AATCCGGAGGACCGGTGTACTCAGCTCACCGGGGGCATTGCTCCCGTGGTGACCCTGATTTGTTGTTGGG</span>
<span class="operator">...</span>
</pre>
<p>你可以手动检查，对于每一个序列条目，物种名都是描述行的第二个单词。这意味着如果我们以空白分割序列条目的 <tt class="docutils literal">.description</tt> ，物种名将会是第1个元素（第0个元素是序列ID），我们可以这样做：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">all_species</span> <span class="operator">=</span> <span class="punctuation">[]</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">):</span>
    <span class="name">all_species</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">description</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">()[</span><span class="literal number integer">1</span><span class="punctuation">])</span>
<span class="keyword">print</span> <span class="name">all_species</span>
</pre>
<p>将得到:</p>
<pre class="code python literal-block">
<span class="punctuation">[</span><span class="literal string">'C.irapeanum'</span><span class="punctuation">,</span> <span class="literal string">'C.californicum'</span><span class="punctuation">,</span> <span class="literal string">'C.fasciculatum'</span><span class="punctuation">,</span> <span class="literal string">'C.margaritaceum'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'P.barbatum'</span><span class="punctuation">]</span>
</pre>
<p>使用更简洁的列表解析：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">all_species</span> <span class="operator">==</span> <span class="punctuation">[</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">description</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">()[</span><span class="literal number integer">1</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> \
                <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)]</span>
<span class="keyword">print</span> <span class="name">all_species</span>
</pre>
<p>通常，对FASTA描述行提取信息不是那么方便。如果你能获得对目标序列注释很好的文件格式如GenBank或者EMBL，那么这类注释信息就很容易处理。</p>
</div>
</div>
<div class="section" id="id37">
<h2>5.2 从压缩文档读取解析序列信息</h2>
<p>在上一节中，我们研究了从文件中解析序列信息。除了使用文件名，你可以让 <tt class="docutils literal">Bio.SeqIO</tt> 使用文件句柄（请见第 <a class="reference external" href="#sec:appendix-handles">22.1</a> 节）。在这一节，我们将使用文件句柄从压缩文件中解析序列信息。</p>
<p>正如你上面看到的，我们可以使用文件名作为 <tt class="docutils literal">Bio.SeqIO.read()</tt> 或 <tt class="docutils literal">Bio.SeqIO.parse()</tt> 的参数 - 例如在这个例子中，我们利用生成器表达式计算GenBank文件中多条序列条目的总长：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">sum</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">))</span>
<span class="literal number integer">67518</span>
</pre>
<p>此处，我们使用文件句柄，并使用 <tt class="docutils literal">with</tt> 语句（Python 2.5及以上版本）自动关闭句柄：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">__future__</span> <span class="keyword namespace">import</span> <span class="name">with_statement</span> <span class="comment">#Needed on Python 2.5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">with</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">)</span> <span class="keyword">as</span> <span class="name">handle</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name builtin">sum</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">))</span>
<span class="literal number integer">67518</span>
</pre>
<p>或者，用旧版本的方式，手动关闭句柄：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">sum</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">))</span>
<span class="literal number integer">67518</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>现在，如果我们有一个gzip压缩的文件呢？这种类型的文件在Linux系统中被普遍使用。我们可以使用Python的 <tt class="docutils literal">gzip</tt> 模块打开压缩文档以读取数据 - 返回一个句柄对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">gzip</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">gzip</span><span class="operator">.</span><span class="name">open</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk.gz&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;r&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">sum</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">))</span>
<span class="literal number integer">67518</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>相同地，如果我们有一个bzip2压缩文件（遗憾的是与函数的名字是不太一致）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">bz2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">bz2</span><span class="operator">.</span><span class="name">BZ2File</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk.bz2&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;r&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">sum</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">))</span>
<span class="literal number integer">67518</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>如果你在使用Python2.7及以上版本， <tt class="docutils literal">with</tt> 也可以读取gzip和bz2文件。然而在这之前的版本中使用将中断程序(<a class="reference external" href="http://bugs.python.org/issue3860">Issue 3860</a> ), 抛出 <tt class="docutils literal">__exit__</tt> 缺失这类 <tt class="docutils literal">属性错误</tt> （ <tt class="docutils literal">AttributeError</tt> ）。</p>
<p>有一种gzip（GNU zip）变种称为BGZF（Blocked GNU Zip Format），它可以作为普通gzip文件被读取，但具有随机读取的优点，我们将在稍后的第 <a class="reference external" href="#sec:SeqIO-index-bgzf">5.4.4</a> 节讨论。</p>
</div>
<div class="section" id="id38">
<h2>5.3 解析来自网络的序列</h2>
<p>在上一节中，我们研究了从文件（使用文件名或者文件句柄）和压缩文件（使用文件句柄）解析序列数据。这里我们将使用 <tt class="docutils literal">Bio.SeqIO</tt> 的另一种类型句柄，网络连接，从网络下载和解析序列。</p>
<p>请注意，你可以一气呵成地下载序列并解析成为 <tt class="docutils literal">SeqRecord</tt> 对象，这并不意味这是一个好主意。通常，你可能需要下载序列并存入文件以重复使用。</p>
<div class="section" id="id39">
<h3>5.3.1 解析来自网络的GenBank序列条目</h3>
<p>第 <a class="reference external" href="#sec:efetch">9.6</a> 节将更详细地讨论Entrez EFetch接口，但是现在我们将通过它连接到NCBI，通过GI号从GenBank获得 <em>Opuntia</em> （刺梨）序列。</p>
<p>首先，我们只获取一条序列条目。如果你不关注注释和相关信息，下载FASTA文件是个不错的选择，因为他们相对紧凑。请记住，当你希望处理的对象包含有且仅有一条序列条目时，使用 <tt class="docutils literal">Bio.SeqIO.read()</tt> 函数：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;6273291&quot;</span><span class="punctuation">)</span>
<span class="name">seq_record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> with </span><span class="literal string interpol">%i</span><span class="literal string"> features&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">))</span>
</pre>
<p>输出结果为:</p>
<pre class="code python literal-block">
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273291</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191665</span> <span class="keyword">with</span> <span class="literal number integer">0</span> <span class="name">features</span>
</pre>
<p>NCBI也允许你获取其它格式文件，尤其是GenBank文件。直到2009年复活节，Entrez EFetch API使用“genbank”作为返回类型。然而NCBI现在坚持使用“gb” （蛋白使用“gp”）作为官方返回类型，具体描述参见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/efetchseq_help.html">EFetch for Sequence and other Molecular Biology Databases</a> 。因此，Biopython1.50及以后版本的 <tt class="docutils literal">Bio.SeqIO</tt> 中，我们支持“gb”作为“genbank”的别名。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;6273291&quot;</span><span class="punctuation">)</span>
<span class="name">seq_record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">)</span> <span class="comment">#using &quot;gb&quot; as an alias for &quot;genbank&quot;</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> with </span><span class="literal string interpol">%i</span><span class="literal string"> features&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">))</span>
</pre>
<p>输出结果为：</p>
<pre class="code python literal-block">
<span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
</pre>
<p>请注意，这次我们获得3个特征。</p>
<p>现在，让我们获取多个序列条目。这次句柄包含多条序列条目，因此我们必须使用 <tt class="docutils literal">Bio.SeqIO.parse()</tt> 函数：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">,</span> \
                       <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;6273291,6273290,6273289&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">seq_record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">description</span><span class="punctuation">[:</span><span class="literal number integer">50</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">&quot;...&quot;</span>
    <span class="keyword">print</span> <span class="literal string">&quot;Sequence length </span><span class="literal string interpol">%i</span><span class="literal string">,&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="punctuation">),</span>
    <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%i</span><span class="literal string"> features,&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">),</span>
    <span class="keyword">print</span> <span class="literal string">&quot;from: </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;source&quot;</span><span class="punctuation">]</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>输出结果为：</p>
<pre class="code python literal-block">
<span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="name">Opuntia</span> <span class="name">marenae</span> <span class="name">rpl16</span> <span class="name">gene</span><span class="punctuation">;</span> <span class="name">chloroplast</span> <span class="name">gene</span> <span class="keyword">for</span> <span class="name">c</span><span class="operator">...</span>
<span class="name">Sequence</span> <span class="name">length</span> <span class="literal number integer">902</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">features</span><span class="punctuation">,</span> <span class="name">from</span><span class="punctuation">:</span> <span class="name">chloroplast</span> <span class="name">Opuntia</span> <span class="name">marenae</span>
<span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="name">Opuntia</span> <span class="name">clavata</span> <span class="name">rpl16</span> <span class="name">gene</span><span class="punctuation">;</span> <span class="name">chloroplast</span> <span class="name">gene</span> <span class="keyword">for</span> <span class="name">c</span><span class="operator">...</span>
<span class="name">Sequence</span> <span class="name">length</span> <span class="literal number integer">899</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">features</span><span class="punctuation">,</span> <span class="name">from</span><span class="punctuation">:</span> <span class="name">chloroplast</span> <span class="name">Grusonia</span> <span class="name">clavata</span>
<span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="name">Opuntia</span> <span class="name">bradtiana</span> <span class="name">rpl16</span> <span class="name">gene</span><span class="punctuation">;</span> <span class="name">chloroplast</span> <span class="name">gene</span> <span class="keyword">for</span><span class="operator">...</span>
<span class="name">Sequence</span> <span class="name">length</span> <span class="literal number integer">899</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">features</span><span class="punctuation">,</span> <span class="name">from</span><span class="punctuation">:</span> <span class="name">chloroplast</span> <span class="name">Opuntia</span> <span class="name">bradtianaa</span>
</pre>
<p>更多关于 <tt class="docutils literal">Bio.Entrez</tt> 模块的信息请见第 <a class="reference external" href="#chapter:entrez">9</a> 章，并阅读NCBI Entrez使用指南（第 <a class="reference external" href="#sec:entrez-guidelines">9.1</a> 节）。</p>
</div>
<div class="section" id="swissprot">
<h3>5.3.2 解析来自网络的SwissProt序列条目</h3>
<p>现在我们使用句柄下载来自ExPASy的SwissProt文件，更深入的信息请见第 <a class="reference external" href="#chapter:swiss_prot">10</a> 章。如上面提到的，当你希望处理的对象包含有且仅有一条序列条目时，使用 <tt class="docutils literal">Bio.SeqIO.read()</tt> 函数：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_sprot_raw</span><span class="punctuation">(</span><span class="literal string">&quot;O23729&quot;</span><span class="punctuation">)</span>
<span class="name">seq_record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;swiss&quot;</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">id</span>
<span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">name</span>
<span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">description</span>
<span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Length </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">annotations</span><span class="punctuation">[</span><span class="literal string">&quot;keywords&quot;</span><span class="punctuation">]</span>
</pre>
<p>如果网络连接正常，你将会得到：</p>
<pre class="code python literal-block">
<span class="name">O23729</span>
<span class="name">CHS3_BROFI</span>
<span class="name">RecName</span><span class="punctuation">:</span> <span class="name">Full</span><span class="operator">=</span><span class="name">Chalcone</span> <span class="name">synthase</span> <span class="literal number integer">3</span><span class="punctuation">;</span> <span class="name">EC</span><span class="operator">=</span><span class="literal number float">2.3</span><span class="operator">.</span><span class="literal number float">1.74</span><span class="punctuation">;</span> <span class="name">AltName</span><span class="punctuation">:</span> <span class="name">Full</span><span class="operator">=</span><span class="name">Naringenin</span><span class="operator">-</span><span class="name">chalcone</span> <span class="name">synthase</span> <span class="literal number integer">3</span><span class="punctuation">;</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'MAPAMEEIRQAQRAEGPAAVLAIGTSTPPNALYQADYPDYYFRITKSEHLTELK...GAE'</span><span class="punctuation">,</span> <span class="name">ProteinAlphabet</span><span class="punctuation">())</span>
<span class="name">Length</span> <span class="literal number integer">394</span>
<span class="punctuation">[</span><span class="literal string">'Acyltransferase'</span><span class="punctuation">,</span> <span class="literal string">'Flavonoid biosynthesis'</span><span class="punctuation">,</span> <span class="literal string">'Transferase'</span><span class="punctuation">]</span>
</pre>
</div>
</div>
<div class="section" id="id40">
<h2>5.4 序列文件作为字典</h2>
<p>我们将介绍 <tt class="docutils literal">Bio.SeqIO</tt> 模块中3个相关函数，用于随机读取多序列文件。这里需要权衡灵活性和内存使用。总之：</p>
<ul class="simple">
<li><tt class="docutils literal">Bio.SeqIO.to_dict()</tt> 最灵活但内存占用最大 （请见第 <a class="reference external" href="#SeqIO:to_dict">5.4.1</a> 节）。这基本上是一个辅助函数，用于建立Python <tt class="docutils literal">字典</tt> ，每个条目以 <tt class="docutils literal">SeqRecord</tt> 对象形式存储在内存中，允许你修改这些条目。</li>
<li><tt class="docutils literal">Bio.SeqIO.index()</tt> 处于中间水平，类似于只读字典，当需要时解析序列到 <tt class="docutils literal">SeqRecord</tt> 对象（请见第 <a class="reference external" href="#sec:SeqIO-index">5.4.2</a> 节）。</li>
<li><tt class="docutils literal">Bio.SeqIO.index_db()</tt> 也类似于只读字典，但是将文件中的ID和文件偏移值存储到硬盘（SQLite3数据库），这意味着它对内存需求很低（请见第 <a class="reference external" href="#sec:SeqIO-index-db">5.4.3</a> 节），但会慢一点。</li>
</ul>
<p>全面的概述请见讨论部分（第 <a class="reference external" href="#sec:SeqIO-indexing-discussion">5.4.5</a> 节）。</p>
<div class="section" id="id41">
<h3>5.4.1 序列文件作为字典-在内存中</h3>
<p>我们对兰花数据文件接下来的处理将用于展示如何对他们建立索引，以及使用Python的 <tt class="docutils literal">dictionary</tt>  数量类型（与Perl中hash类似）以类似于数据库的方式读取数据。这常用于从中等大小的文件中读取某些特定元素，形成一个很好的快速数据库。如果处理较大的文件，内存将是个问题，请见下面第 <a class="reference external" href="#sec:SeqIO-index">5.4.2</a> 节。</p>
<p>你可以使用 <tt class="docutils literal">Bio.SeqIO.to_dict()</tt> 函数创建一个 <tt class="docutils literal">SeqRecord</tt> 字典（在内存中）。默认会使用每条序列条目的ID（i.e.  <tt class="docutils literal">.id</tt> 属性）作为键。让我们用GenBank文件试一试：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">to_dict</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">))</span>
</pre>
<p><tt class="docutils literal">Bio.SeqIO.to_dict()</tt> 仅需一个参数，即能够得到 <tt class="docutils literal">SeqRecord</tt> 对象的列表或生成器，这里我们使用 <tt class="docutils literal">SeqIO.parse</tt> 函数输出。顾名思义， <tt class="docutils literal">Bio.SeqIO.to_dict()</tt> 返回一个Python字典。</p>
<p>因为变量 <tt class="docutils literal">orchid_dict</tt>  是一个普通的Python字典，我们可以查看所有的键：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">orchid_dict</span><span class="punctuation">)</span>
<span class="literal number integer">94</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'Z78484.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78464.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78455.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78442.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78532.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78453.1'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Z78471.1'</span><span class="punctuation">]</span>
</pre>
<p>如果你确实需要，你甚至可以一次性查看所有的序列条目：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">values</span><span class="punctuation">()</span> <span class="comment">#lots of output!</span>
<span class="operator">...</span>
</pre>
<p>我们可以通过键读取单个 <tt class="docutils literal">SeqRecord</tt>  对象并操作改对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq_record</span> <span class="operator">=</span> <span class="name">orchid_dict</span><span class="punctuation">[</span><span class="literal string">&quot;Z78475.1&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">description</span>
<span class="name">P</span><span class="operator">.</span><span class="name">supardii</span> <span class="literal number float">5.8</span><span class="name">S</span> <span class="name">rRNA</span> <span class="name">gene</span> <span class="operator word">and</span> <span class="name">ITS1</span> <span class="operator word">and</span> <span class="name">ITS2</span> <span class="name">DNA</span><span class="operator">.</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">repr</span><span class="punctuation">(</span><span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGT'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>因此，可以用我们的GenBank序列条目轻松地在内存中创建一个数据库（in memory “database”）。接下来我们将尝试使用FASTA文件。</p>
<p>值得注意的是，对有Python使用经验的人来说，可以轻松地创建一个类似的字典。然而，典型的字典构建方法不能很好地处理重复键的情况。使用 <tt class="docutils literal">Bio.SeqIO.to_dict()</tt> 函数将明确检查重复键，如果发现任何重复键将引发异常并退出。</p>
<div class="section" id="id42">
<h4>5.4.1.1 指定字典键</h4>
<p>使用上述相同的代码，仅将文件改为FASTA文件：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">to_dict</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="keyword">print</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
</pre>
<p>这次键为：</p>
<pre class="code python literal-block">
<span class="punctuation">[</span><span class="literal string">'gi|2765596|emb|Z78471.1|PDZ78471'</span><span class="punctuation">,</span> <span class="literal string">'gi|2765646|emb|Z78521.1|CCZ78521'</span><span class="punctuation">,</span> <span class="operator">...</span>
 <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'gi|2765613|emb|Z78488.1|PTZ78488'</span><span class="punctuation">,</span> <span class="literal string">'gi|2765583|emb|Z78458.1|PHZ78458'</span><span class="punctuation">]</span>
</pre>
<p>这结果是之前在第 <a class="reference external" href="#sec:fasta-parsing">2.4.1</a> 节中我们解析的FASTA文件结果。如果你需要别的作为键，如登录号（Accession Number），可使用 <tt class="docutils literal">SeqIO.to_dict()</tt> 的可选参数 <tt class="docutils literal">key_function</tt> ，它允许你根据你的序列条目特点，自定义字典键。</p>
<p>首先，你必须写一个函数，当使用 <tt class="docutils literal">SeqRecord</tt> 对象作为参数时，可以返回你需要的键（字符串）。通常，函数的细节依赖于你要处理的序列条目的特点。但是对于我们的兰花数据，我们只需要使用“管道”符号（|）切分ID并返回第四个条目（第三个元素）：</p>
<pre class="code python literal-block">
<span class="keyword">def</span> <span class="name function">get_accession</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">):</span>
    <span class="literal string doc">&quot;&quot;&quot;&quot;Given a SeqRecord, return the accession number as a string.

    e.g. &quot;gi|2765613|emb|Z78488.1|PTZ78488&quot; -&gt; &quot;Z78488.1&quot;
    &quot;&quot;&quot;</span>
    <span class="name">parts</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">(</span><span class="literal string">&quot;|&quot;</span><span class="punctuation">)</span>
    <span class="keyword">assert</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">parts</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="literal number integer">5</span> <span class="operator word">and</span> <span class="name">parts</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="literal string">&quot;gi&quot;</span> <span class="operator word">and</span> <span class="name">parts</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="literal string">&quot;emb&quot;</span>
    <span class="keyword">return</span> <span class="name">parts</span><span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">]</span>
</pre>
<p>然后我们可以将此函数赋与 <tt class="docutils literal">SeqIO.to_dict()</tt> 函数用于构建字典：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">to_dict</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">),</span> <span class="name">key_function</span><span class="operator">=</span><span class="name">get_accession</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
</pre>
<p>最终可到到新的字典键：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'Z78484.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78464.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78455.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78442.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78532.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78453.1'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Z78471.1'</span><span class="punctuation">]</span>
</pre>
<p>不是太困难！</p>
</div>
<div class="section" id="seguid">
<h4>5.4.1.2 使用SEGUID校验和对字典建立索引</h4>
<p>为了介绍另外一个 <tt class="docutils literal">SeqRecord</tt> 对象字典的示例，我们将使用SEGUID校验和函数。这是一个相对较新的校验和，冲突非常罕见（i.e. 两条不同序列具有相同的校验和），相对CRC64校验和有所提升。</p>
<p>让我们再一次处理兰花GenBank文件：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqUtils.CheckSum</span> <span class="keyword namespace">import</span> <span class="name">seguid</span>
<span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">seguid</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
</pre>
<p>将得到：</p>
<pre class="code python literal-block">
<span class="name">Z78533</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="name">JUEoWn6DPhgZ9nAyowsgtoD9TTo</span>
<span class="name">Z78532</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="name">MN</span><span class="operator">/</span><span class="name">s0q9zDoCVEEc</span><span class="operator">+</span><span class="name">k</span><span class="operator">/</span><span class="name">IFwCNF2pY</span>
<span class="operator">...</span>
<span class="name">Z78439</span><span class="operator">.</span><span class="literal number integer">1</span> <span class="name">H</span><span class="operator">+</span><span class="name">JfaShya</span><span class="operator">/</span><span class="literal number integer">4</span><span class="name">yyAj7IbMqgNkxdxQ</span>
</pre>
<p>现在，再次调用 <tt class="docutils literal">Bio.SeqIO.to_dict()</tt> 函数 <tt class="docutils literal">key_function</tt> 参数， <tt class="docutils literal">key_function</tt> 参数需要一个函数将 <tt class="docutils literal">SeqRecord</tt> 转变为字符串。我们不能直接使用`seguid() `` 函数，因为它需要 <tt class="docutils literal">Seq</tt> 对象（或字符串）作为参数。不过，我们可以使用Python的 <tt class="docutils literal">lambda</tt> 特性创建一个一次性（“one off”）函数，然后传递给 <tt class="docutils literal">Bio.SeqIO.to_dict()</tt> ：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqUtils.CheckSum</span> <span class="keyword namespace">import</span> <span class="name">seguid</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seguid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">to_dict</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>                             <span class="keyword">lambda</span> <span class="name">rec</span> <span class="punctuation">:</span> <span class="name">seguid</span><span class="punctuation">(</span><span class="name">rec</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">seguid_dict</span><span class="punctuation">[</span><span class="literal string">&quot;MN/s0q9zDoCVEEc+k/IFwCNF2pY&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span>
<span class="name">Z78532</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">description</span>
<span class="name">C</span><span class="operator">.</span><span class="name">californicum</span> <span class="literal number float">5.8</span><span class="name">S</span> <span class="name">rRNA</span> <span class="name">gene</span> <span class="operator word">and</span> <span class="name">ITS1</span> <span class="operator word">and</span> <span class="name">ITS2</span> <span class="name">DNA</span><span class="operator">.</span>
</pre>
<p>将会返回文件中第二个序列条目 <tt class="docutils literal">Z78532.1</tt> 。</p>
</div>
</div>
<div class="section" id="id43">
<h3>5.4.2 序列文件作为字典 - 索引文件</h3>
<p>之前众多示例试图解释的是使用 <tt class="docutils literal">Bio.SeqIO.to_dict()</tt> 的灵活性。然而，因为它将所有的信息都存储在内存中，你能处理的文件大小受限于电脑的RAM。通常，这仅能处理一些小文件或中等大小文件。</p>
<p>对于更大的文件，应该考虑使用 <tt class="docutils literal">Bio.SeqIO.index()</tt> ，工作原理上略有不同。尽管仍然是返回一个类似于字典的对象，它并不将所有的信息存储在内存中。相反，它仅仅记录每条序列条目在文件中的位置 - 当你需要读取某条特定序列条目时，它才进行解析。</p>
<p>让我们使用之前相同的GenBank文件作为示例：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">orchid_dict</span><span class="punctuation">)</span>
<span class="literal number integer">94</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'Z78484.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78464.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78455.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78442.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78532.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78453.1'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Z78471.1'</span><span class="punctuation">]</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq_record</span> <span class="operator">=</span> <span class="name">orchid_dict</span><span class="punctuation">[</span><span class="literal string">&quot;Z78475.1&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">description</span>
<span class="name">P</span><span class="operator">.</span><span class="name">supardii</span> <span class="literal number float">5.8</span><span class="name">S</span> <span class="name">rRNA</span> <span class="name">gene</span> <span class="operator word">and</span> <span class="name">ITS1</span> <span class="operator word">and</span> <span class="name">ITS2</span> <span class="name">DNA</span><span class="operator">.</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq_record</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGT'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>注意： <tt class="docutils literal">Bio.SeqIO.index()</tt> 不接受句柄参数，仅仅接受文件名。这有充分的理由，但是过于技术性。第二个参数是文件格式（与其它 <tt class="docutils literal">Bio.SeqIO</tt> 函数一样的小写字符串）。你可以使用许多其他的简单的文件格式，包括FASTA和FASTQ文件（示例参见第 <a class="reference external" href="#sec:fastq-indexing">18.1.11</a> 节），但不支持比对文件格式，如PHYLIP或Clustal。最后有个可选参数，你可以指定字符集或者键函数。</p>
<p>下面是使用FASTA文件做的相同的示例 - 仅改变了文件名和格式：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">orchid_dict</span><span class="punctuation">)</span>
<span class="literal number integer">94</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'gi|2765596|emb|Z78471.1|PDZ78471'</span><span class="punctuation">,</span> <span class="literal string">'gi|2765646|emb|Z78521.1|CCZ78521'</span><span class="punctuation">,</span> <span class="operator">...</span>
 <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'gi|2765613|emb|Z78488.1|PTZ78488'</span><span class="punctuation">,</span> <span class="literal string">'gi|2765583|emb|Z78458.1|PHZ78458'</span><span class="punctuation">]</span>
</pre>
<div class="section" id="id44">
<h4>5.4.2.1 指定字典键</h4>
<p>如果想使用与之前一样的键，像第 <a class="reference external" href="#seq:seqio-todict-functionkey">5.4.1.1</a> 节 <tt class="docutils literal">Bio.SeqIO.to_dict()</tt> 示例，你需要写一个小函数，从FASTA ID（字符串）中匹配你想要的键：</p>
<pre class="code python literal-block">
<span class="keyword">def</span> <span class="name function">get_acc</span><span class="punctuation">(</span><span class="name">identifier</span><span class="punctuation">):</span>
    <span class="literal string doc">&quot;&quot;&quot;&quot;Given a SeqRecord identifier string, return the accession number as a string.

    e.g. &quot;gi|2765613|emb|Z78488.1|PTZ78488&quot; -&gt; &quot;Z78488.1&quot;
    &quot;&quot;&quot;</span>
    <span class="name">parts</span> <span class="operator">=</span> <span class="name">identifier</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">(</span><span class="literal string">&quot;|&quot;</span><span class="punctuation">)</span>
    <span class="keyword">assert</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">parts</span><span class="punctuation">)</span> <span class="operator">==</span> <span class="literal number integer">5</span> <span class="operator word">and</span> <span class="name">parts</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="literal string">&quot;gi&quot;</span> <span class="operator word">and</span> <span class="name">parts</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">]</span> <span class="operator">==</span> <span class="literal string">&quot;emb&quot;</span>
    <span class="keyword">return</span> <span class="name">parts</span><span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">]</span>
</pre>
<p>然后我们将此函数赋与 <tt class="docutils literal">Bio.SeqIO.index()</tt> 函数用于构建字典：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">,</span> <span class="name">key_function</span><span class="operator">=</span><span class="name">get_acc</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">orchid_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'Z78484.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78464.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78455.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78442.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78532.1'</span><span class="punctuation">,</span> <span class="literal string">'Z78453.1'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Z78471.1'</span><span class="punctuation">]</span>
</pre>
<p>当你知道怎样实现就变得很简单了。</p>
</div>
<div class="section" id="id45">
<h4>5.4.2.2 获取序列条目原始数据</h4>
<p>来自 <tt class="docutils literal">Bio.SeqIO.index()</tt> 的字典样对象以 <tt class="docutils literal">SeqRecord</tt> 对象形式返回序列条目。但是，有时候从文件中直接获取原始数据非常有用。对于此种情况，使用 <tt class="docutils literal">get_raw()</tt> 方法，它仅需要一个参数（序列ID），然后返回一个字符串（提取自文件的未处理数据）。</p>
<p>一个重要的例子就是从大文件中提取出一个序列子集，特别是当 <tt class="docutils literal">Bio.SeqIO.write()</tt> 还不支持这种输出格式（e.g. SwissProt文件格式的文本文件 ） 或者需要完整地保留源文件（Biopython的GenBank和EMBL格式输出并不会保留每一点注释信息）。</p>
<p>假如你已经从UniProt FTP站点下载了整个数据库的SwissPort格式文本文件（ <a class="reference external" href="ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz">ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz</a> ），并也已经解压为文件 <tt class="docutils literal">uniprot_sprot.dat</tt> ，你需要从中提取一部分序列条目：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">uniprot</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;uniprot_sprot.dat&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;swiss&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;selected.dat&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">acc</span> <span class="operator word">in</span> <span class="punctuation">[</span><span class="literal string">&quot;P33487&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;P19801&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;P13689&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;Q8JZQ5&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;Q9TRC7&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="name">handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">uniprot</span><span class="operator">.</span><span class="name">get_raw</span><span class="punctuation">(</span><span class="name">acc</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>在第 <a class="reference external" href="#sec:SeqIO-sort">18.1.5</a> 节有更多关于使用 <tt class="docutils literal">SeqIO.index()</tt> 函数对大文件序列排序的示例（不需要一次加载所有信息到内存）。</p>
</div>
</div>
<div class="section" id="id46">
<h3>5.4.3 序列文件作为字典 - 数据库索引文件</h3>
<p>Biopython 1.57引入一个替代的函数， <tt class="docutils literal">Bio.SeqIO.index_db()</tt> 。由于它将序列信息以文件方式存储在硬盘上（使用SQLite3数据库）而不是内存中，因此它可以处理超大文件。同时，你可以同时对多个文件建立索引（前提是所有序列条目的ID是唯一的）。</p>
<p><tt class="docutils literal">Bio.SeqIO.index()</tt> 函数有三个参数：</p>
<ul class="simple">
<li>索引文件名，我们建议使用以 <tt class="docutils literal">.idx</tt> 结尾的字符，改索引文件实质上是SQLite3数据库；</li>
<li>要建立索引的文件列表（或者单个文件名）；</li>
<li>文件格式（与 <tt class="docutils literal">SeqIO</tt> 模块中其它函数一样的小写字符串）。</li>
</ul>
<p>将以NCBI FTP站点 <a class="reference external" href="ftp://ftp.ncbi.nih.gov/genbank/">ftp://ftp.ncbi.nih.gov/genbank/</a> 的GenBank文本文件为例，这些文件为gzip压缩文件。对于GenBank版本182，病毒序列共包含16个文件， <tt class="docutils literal">gbvrl1.seq</tt>  -  <tt class="docutils literal">gbvrl16.seq</tt> ，共包含约一百万条序列条目。对这些文件，你可以像这样建立索引：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">files</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">&quot;gbvrl</span><span class="literal string interpol">%i</span><span class="literal string">.seq&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">16</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gb_vrl</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index_db</span><span class="punctuation">(</span><span class="literal string">&quot;gbvrl.idx&quot;</span><span class="punctuation">,</span> <span class="name">files</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%i</span><span class="literal string"> sequences indexed&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">gb_vrl</span><span class="punctuation">)</span>
<span class="literal number integer">958086</span> <span class="name">sequences</span> <span class="name">indexed</span>
</pre>
<p>在我个人电脑上，运行大约需要2分钟。如果你重新运行，索引文件（这里为 <tt class="docutils literal">gbvrl.idx</tt> ）将在不到一秒的时间内加载。你可以将这个索引作为一个只读的Python字典，并不需要去担心序列来自哪个文件，e.g.:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">gb_vrl</span><span class="punctuation">[</span><span class="literal string">&quot;GQ333173.1&quot;</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">description</span>
<span class="name">HIV</span><span class="operator">-</span><span class="literal number integer">1</span> <span class="name">isolate</span> <span class="name">F12279A1</span> <span class="keyword namespace">from</span> <span class="name namespace">Uganda</span> <span class="name namespace">gag</span> <span class="name namespace">protein</span> <span class="punctuation">(</span><span class="name">gag</span><span class="punctuation">)</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">partial</span> <span class="name">cds</span><span class="operator">.</span>
</pre>
<div class="section" id="id47">
<h4>5.4.3.1 获取序列条目原始数据</h4>
<p>与第 <a class="reference external" href="#sec:seqio-index-getraw">5.4.2.2</a> 节讨论的 <tt class="docutils literal">Bio.SeqIO.index()</tt> 函数一样，该字典样对象同样允许你获取每个序列条目的原始文件：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">gb_vrl</span><span class="operator">.</span><span class="name">get_raw</span><span class="punctuation">(</span><span class="literal string">&quot;GQ333173.1&quot;</span><span class="punctuation">)</span>
<span class="name">LOCUS</span>       <span class="name">GQ333173</span>                 <span class="literal number integer">459</span> <span class="name">bp</span>    <span class="name">DNA</span>     <span class="name">linear</span>   <span class="name">VRL</span> <span class="literal number integer">21</span><span class="operator">-</span><span class="name">OCT</span><span class="operator">-</span><span class="literal number integer">2009</span>
<span class="name">DEFINITION</span>  <span class="name">HIV</span><span class="operator">-</span><span class="literal number integer">1</span> <span class="name">isolate</span> <span class="name">F12279A1</span> <span class="keyword namespace">from</span> <span class="name namespace">Uganda</span> <span class="name namespace">gag</span> <span class="name namespace">protein</span> <span class="punctuation">(</span><span class="name">gag</span><span class="punctuation">)</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">partial</span>
            <span class="name">cds</span><span class="operator">.</span>
<span class="name">ACCESSION</span>   <span class="name">GQ333173</span>
<span class="operator">...</span>
<span class="operator">//</span>
</pre>
</div>
</div>
<div class="section" id="id48">
<h3>5.4.4 对压缩文件建立索引</h3>
<p>经常你要建立索引的文件可能非常大，因此你想对它进行压缩。不幸的是，对常规的文件格式如gzip和bzip2高效的随机读取通常很困难。在这种情况下，BGZF (Blocked GNU Zip Format)非常有用。它是gzip变体（也可以使用标准的gzip工具解压），因BAM文件格式得到推广，<a class="reference external" href="http://samtools.sourceforge.net/">samtools</a> 和 <a class="reference external" href="http://samtools.sourceforge.net/tabix.shtml">tabix</a> ；</p>
<p>你可以使用samtools的命令行工具 <tt class="docutils literal">bgzip</tt> 创建BGZF格式压缩文件。在我们的示例中，使用文件扩展名 <tt class="docutils literal">*.bgz</tt> ，以区分于普通的压缩文件（命名为 <tt class="docutils literal">*.gz</tt> ）。你也可以在Python中使用 <tt class="docutils literal">Bio.bgzf</tt> 模块读写BGZF文件。</p>
<p><tt class="docutils literal">Bio.SeqIO.index()</tt> 和 <tt class="docutils literal">Bio.SeqIO.index_db()</tt> 函数均可以用于BGZF压缩文件。例如，如果使用过未压缩的GenBank文件：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">orchid_dict</span><span class="punctuation">)</span>
<span class="literal number integer">94</span>
</pre>
<p>你可以使用如下的命令行命令压缩该文件（同时保留源文件） - 不需要担心，压缩文件和别的示例及已经包含：</p>
<pre class="code python literal-block">
<span class="error">$</span> <span class="name">bgzip</span> <span class="operator">-</span><span class="name">c</span> <span class="name">ls_orchid</span><span class="operator">.</span><span class="name">gbk</span> <span class="operator">&gt;</span> <span class="name">ls_orchid</span><span class="operator">.</span><span class="name">gbk</span><span class="operator">.</span><span class="name">bgz</span>
</pre>
<p>你可以用相同的方式使用压缩文件：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk.bgz&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">orchid_dict</span><span class="punctuation">)</span>
<span class="literal number integer">94</span>
</pre>
<p>或：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">orchid_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index_db</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk.bgz.idx&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;ls_orchid.gbk.bgz&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">orchid_dict</span><span class="punctuation">)</span>
<span class="literal number integer">94</span>
</pre>
<p><cite>SeqIO `</cite> 建立索引时自动检测是否为BGZF压缩格式。注意：压缩文件和未压缩文件不能使用相同的索引文件。</p>
</div>
<div class="section" id="id49">
<h3>5.4.5 讨论</h3>
<p>这些方法你该使用哪种及其原因，取决于你要做什么（以及你要处理的数据有多大）。然而，通常 <tt class="docutils literal">Bio.SeqIO.index()</tt> 是个不错的选择。如果你正在处理上百万条序列条目，多个文件，或者重复性分析，那么看看 <tt class="docutils literal">Bio.SeqIO.index_db()</tt> 。</p>
<p>选择 <tt class="docutils literal">Bio.SeqIO.to_dict()</tt> 而不选择 <tt class="docutils literal">Bio.SeqIO.index()</tt> 或 <tt class="docutils literal">Bio.SeqIO.index_db()</tt> 的原因主要是它的灵活性，尽管会占用更多内存。存储 <tt class="docutils literal">SeqRecord</tt> 对象到内存的优势在于可以随意被改变，添加或者删除。除了高内存消耗这个缺点外，建立索引也可能花费更长的时间，因为所有的条目都需要被完全解析。</p>
<p><tt class="docutils literal">Bio.SeqIO.index()</tt> 和 <tt class="docutils literal">Bio.SeqIO.index_db()</tt> 都是在需要时才解析序列条目。当建立索引时，他们扫描文件，寻找每个序列条目的起始，并做尽可能少的工作提取出ID信息。</p>
<p>选择 <tt class="docutils literal">Bio.SeqIO.index()</tt> 而不选择 <tt class="docutils literal">Bio.SeqIO.index_db()</tt> 的原因包括以下：</p>
<ul class="simple">
<li>建立索引更快（需要注意的是简单文件格式）</li>
<li>读取 <tt class="docutils literal">SeqRecord</tt> 对象稍快（但是这种差异只有在解析简单格式文件是才可见）</li>
<li>可以使用不可变的Python对象作为字典键而不仅仅是字符串（e.g. 如字符串元组、不可变容器（frozen set））</li>
<li>如果被建立索引的序列文件改变，不需要担心索引数据库过期。</li>
</ul>
<p>选择 <tt class="docutils literal">Bio.SeqIO.index_db()</tt> 而不选择 <tt class="docutils literal">Bio.SeqIO.index()</tt> 的原因包括以下：</p>
<ul class="simple">
<li>没有内存限制 - 这对通常多达10亿的二代测序文件来说非常重要，如果使用 <tt class="docutils literal">Bio.SeqIO.index()</tt> 可能需要超过4G的RAM和64位Python</li>
<li>索引数据量保存在硬盘上，可重复使用。尽管建立索引数据库需要花费更多的时间，但是从长远看来。如果你有个脚本重新运行这个相同的数据库，可以节约时间</li>
<li>可以同时对多个文件建立索引</li>
<li><cite>get_raw() `</cite> 方法可以快得多，因为对于大多数文件格式只需要存储序列条目的长度和偏移量（offset）</li>
</ul>
</div>
</div>
<div class="section" id="id50">
<h2>5.5 写入序列文件</h2>
<p>我们已经讨论了使用 <tt class="docutils literal">Bio.SeqIO.parse()</tt> 输入序列（读取文件），现在我们将研究使用 <tt class="docutils literal">Bio.SeqIO.write()</tt> 输出序列（写入文件）。该函数需要三个参数：某些 <tt class="docutils literal">SeqRecord</tt> 对象，要写入的句柄或文件名，和序列格式。</p>
<p>我们先用硬编码方式（手动创建而不是从文件中加载）创建一个些新的 <tt class="docutils literal">SeqRecord</tt> 对象，示例如下：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_protein</span>

<span class="name">rec1</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;SSAC&quot;</span><span class="punctuation">,</span> <span class="name">generic_protein</span><span class="punctuation">),</span>
                 <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;gi|14150838|gb|AAK54648.1|AF376133_1&quot;</span><span class="punctuation">,</span>
                 <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;chalcone synthase [Cucumis sativus]&quot;</span><span class="punctuation">)</span>

<span class="name">rec2</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;YPDYYFRITNREHKAELKEKFQRMCDKSMIKKRYMYLTEEILKENPSMCEYMAPSLDARQ&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;DMVVVEIPKLGKEAAVKAIKEWGQ&quot;</span><span class="punctuation">,</span> <span class="name">generic_protein</span><span class="punctuation">),</span>
                 <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;gi|13919613|gb|AAK33142.1|&quot;</span><span class="punctuation">,</span>
                 <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;chalcone synthase [Fragaria vesca subsp. bracteata]&quot;</span><span class="punctuation">)</span>

<span class="name">rec3</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;MVTVEEFRRAQCAEGPATVMAIGTATPSNCVDQSTYPDYYFRITNSEHKVELKEKFKRMC&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;EKSMIKKRYMHLTEEILKENPNICAYMAPSLDARQDIVVVEVPKLGKEAAQKAIKEWGQP&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;KSKITHLVFCTTSGVDMPGCDYQLTKLLGLRPSVKRFMMYQQGCFAGGTVLRMAKDLAEN&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;NKGARVLVVCSEITAVTFRGPNDTHLDSLVGQALFGDGAAAVIIGSDPIPEVERPLFELV&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;SAAQTLLPDSEGAIDGHLREVGLTFHLLKDVPGLISKNIEKSLVEAFQPLGISDWNSLFW&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;IAHPGGPAILDQVELKLGLKQEKLKATRKVLSNYGNMSSACVLFILDEMRKASAKEGLGT&quot;</span> \
                    <span class="operator">+</span><span class="literal string">&quot;TGEGLEWGVLFGFGPGLTVETVVLHSVAT&quot;</span><span class="punctuation">,</span> <span class="name">generic_protein</span><span class="punctuation">),</span>
                 <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;gi|13925890|gb|AAK49457.1|&quot;</span><span class="punctuation">,</span>
                 <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;chalcone synthase [Nicotiana tabacum]&quot;</span><span class="punctuation">)</span>

<span class="name">my_records</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">rec1</span><span class="punctuation">,</span> <span class="name">rec2</span><span class="punctuation">,</span> <span class="name">rec3</span><span class="punctuation">]</span>
</pre>
<p>现在我们得到一个 <tt class="docutils literal">SeqRecord</tt> 对象列表，将它写入一个FASTA格式文件：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">my_records</span><span class="punctuation">,</span> <span class="literal string">&quot;my_example.faa&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>如果用你喜欢的文本编辑软件打开，可得到：</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">14150838</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AAK54648</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF376133_1</span> <span class="name">chalcone</span> <span class="name">synthase</span> <span class="punctuation">[</span><span class="name">Cucumis</span> <span class="name">sativus</span><span class="punctuation">]</span>
<span class="name">MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD</span>
<span class="name">GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK</span>
<span class="name">NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM</span>
<span class="name">SSAC</span>
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">13919613</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AAK33142</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="name">chalcone</span> <span class="name">synthase</span> <span class="punctuation">[</span><span class="name">Fragaria</span> <span class="name">vesca</span> <span class="name">subsp</span><span class="operator">.</span> <span class="name">bracteata</span><span class="punctuation">]</span>
<span class="name">YPDYYFRITNREHKAELKEKFQRMCDKSMIKKRYMYLTEEILKENPSMCEYMAPSLDARQ</span>
<span class="name">DMVVVEIPKLGKEAAVKAIKEWGQ</span>
<span class="operator">&gt;</span><span class="name">gi</span><span class="operator">|</span><span class="literal number integer">13925890</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AAK49457</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="name">chalcone</span> <span class="name">synthase</span> <span class="punctuation">[</span><span class="name">Nicotiana</span> <span class="name">tabacum</span><span class="punctuation">]</span>
<span class="name">MVTVEEFRRAQCAEGPATVMAIGTATPSNCVDQSTYPDYYFRITNSEHKVELKEKFKRMC</span>
<span class="name">EKSMIKKRYMHLTEEILKENPNICAYMAPSLDARQDIVVVEVPKLGKEAAQKAIKEWGQP</span>
<span class="name">KSKITHLVFCTTSGVDMPGCDYQLTKLLGLRPSVKRFMMYQQGCFAGGTVLRMAKDLAEN</span>
<span class="name">NKGARVLVVCSEITAVTFRGPNDTHLDSLVGQALFGDGAAAVIIGSDPIPEVERPLFELV</span>
<span class="name">SAAQTLLPDSEGAIDGHLREVGLTFHLLKDVPGLISKNIEKSLVEAFQPLGISDWNSLFW</span>
<span class="name">IAHPGGPAILDQVELKLGLKQEKLKATRKVLSNYGNMSSACVLFILDEMRKASAKEGLGT</span>
<span class="name">TGEGLEWGVLFGFGPGLTVETVVLHSVAT</span>
</pre>
<p>怎样才能知道 <tt class="docutils literal">Bio.SeqIO.write()</tt> 函数写入了多少条序列条目到句柄呢？如果你的序列条目保存在一个列表中，只需要使用 <tt class="docutils literal">len(my_records)</tt> ，但是你不能对来自生成器/迭代器的序列条目。 <tt class="docutils literal">Bio.SeqIO.write()</tt> 函数本身就返回写入文件的 <tt class="docutils literal">SeqRecord</tt> 对象个数。</p>
<ul class="simple">
<li>注意 - 如果你 <tt class="docutils literal">Bio.SeqIO.write()</tt> 函数要写入的文件已经存在，旧文件将会被覆写，并且不会得到任何警告信息。</li>
</ul>
<div class="section" id="round-trips">
<h3>5.5.1 可逆读写（Round trips）</h3>
<p>某些人需要他们的解析器是“可逆”的，即当你读入某个文件后和可以按原样写回。这需要解析器提取足够多的信息用于 * 精确 * 还原原始文件， <tt class="docutils literal">Bio.SeqIO</tt> 不打算这么做。</p>
<p>一个简单的例子是，FASTA文件中，允许序列以任意字符数换行。解析以下两条序列得到一个相同的 <tt class="docutils literal">SeqRecord</tt> 对象，这两条序列仅在换行上不同：</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">YAL068C</span><span class="operator">-</span><span class="literal number float">7235.2170</span> <span class="name">Putative</span> <span class="name">promoter</span> <span class="name">sequence</span>
<span class="name">TACGAGAATAATTTCTCATCATCCAGCTTTAACACAAAATTCGCACAGTTTTCGTTAAGA</span>
<span class="name">GAACTTAACATTTTCTTATGACGTAAATGAAGTTTATATATAAATTTCCTTTTTATTGGA</span>

<span class="operator">&gt;</span><span class="name">YAL068C</span><span class="operator">-</span><span class="literal number float">7235.2170</span> <span class="name">Putative</span> <span class="name">promoter</span> <span class="name">sequence</span>
<span class="name">TACGAGAATAATTTCTCATCATCCAGCTTTAACACAAAATTCGCA</span>
<span class="name">CAGTTTTCGTTAAGAGAACTTAACATTTTCTTATGACGTAAATGA</span>
<span class="name">AGTTTATATATAAATTTCCTTTTTATTGGA</span>
</pre>
<p>为了创建一个可逆读写的FASTA解析器，需要记录序列换行发生的位置，而这些额外的信息通常毫无意义。因此，Biopython在输出时使用默认的60字符换行。空白字符在许多其他文件格式中运用也存在相同的问题。另外一个问题是，在某些情况下，Biopython并不能保存每一点注释信息（e.g. GenBank和EMBL）。</p>
<p>少数时候，重要的是保留原来的布局（这可能有点怪异），第 <a class="reference external" href="#sec:seqio-index-getraw">5.4.2.2</a> 节关于 <tt class="docutils literal">Bio.SeqIO.index()</tt> 字典样对象的 <tt class="docutils literal">get_raw()</tt> 方法提供了可能的解决方案。</p>
</div>
<div class="section" id="id51">
<h3>5.5.2 序列格式间的转换</h3>
<p>在之前的例子中我们使用 <tt class="docutils literal">SeqRecord</tt> 对象列表作为 <tt class="docutils literal">Bio.SeqIO.write()</tt> 函数的输入，但是它也接受如来自于 <tt class="docutils literal">Bio.SeqIO.parse()</tt> 的 <tt class="docutils literal">SeqRecord</tt> 迭代器 - 这允许我们通过结合使用这两个函数实现文件转换。</p>
<p>在这个例子中，我们将读取GenBank格式文件 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.gbk">ls_orchid.gbk</a> ，然后输出为FASTA格式文件：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;my_example.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Converted </span><span class="literal string interpol">%i</span><span class="literal string"> records&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>这仍然有点复杂，因为文件格式转换是比较常见的任务，有一个辅助函数可以替代上述代码：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;my_example.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Converted </span><span class="literal string interpol">%i</span><span class="literal string"> records&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p><tt class="docutils literal">Bio.SeqIO.convert()</tt> 函数可以使用句柄或文件名。然而需要注意的是，如果输出文件已存在，将覆写该文件。想了解更多信息，请使用内置帮助文档：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>原理上讲，只需要改变文件名和格式字符串，该代码即可实现Biopython支持的文件格式间的转换。然而，写入某种格式时需要某些特定的信息（e.g. 质量值），而其他格式文件不包含此信息。例如，你可以将FASTQ转化为FASTA文件，却不能进行逆操作。不同FASTQ格式间的相互转变请见cookbook章第 <a class="reference external" href="#sec:SeqIO-fastq-conversion">18.1.9</a> 节和第 <a class="reference external" href="#sec:SeqIO-fasta-qual-conversion">18.1.10</a> 节。</p>
<p>最后，使用 <tt class="docutils literal">Bio.SeqIO.convert()</tt> 函数额外的好处是更快，（最大的好处是代码会更短）原因是该转换函数可以利用几个文件格式特殊的优化条件和技巧。</p>
</div>
<div class="section" id="id52">
<h3>5.5.3 转化序列到反向互补序列</h3>
<p>假设你有一个核苷酸序列文件，需要转换成一个包含其反向互补的文件。这时，需要做些工作，将从文件得到的 <tt class="docutils literal">SeqRecord</tt> 对象转化为适合存储到输出文件的信息。</p>
<p>首先，我们将使用 <tt class="docutils literal">Bio.SeqIO.parse()</tt> 加载文件中的核酸序列，然后使用 <tt class="docutils literal">Seq</tt> 对象的内置方法 <tt class="docutils literal">.reverse_complement()</tt> 输出其反向互补序列（请见第 <a class="reference external" href="#sec:seq-reverse-complement">3.7</a> 节）。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
</pre>
<p>现在，如果我们想保存这些反向互补序列到某个文件，需要创建 <tt class="docutils literal">SeqRecord</tt> 对象。我们可以使用 <tt class="docutils literal">SeqRecord</tt> 对象的内置方法 <tt class="docutils literal">.reverse_complement()</tt> （请见第 <a class="reference external" href="#sec:SeqRecord-reverse-complement">4.8</a> 节），但是我们必须决定新的序列条目怎么命名。</p>
<p>这是一个绝好的展示列表解析效率地方，列表解析通过在内存中创建一个列表实现：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">rec</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;rc_&quot;</span><span class="operator">+</span><span class="name">rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;reverse complement&quot;</span><span class="punctuation">)</span> \
<span class="operator">...</span>            <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">)</span>
</pre>
<p>这时就用到了列表解析的绝妙之处，在其中添加一个条件语句：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">rec</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;rc_&quot;</span><span class="operator">+</span><span class="name">rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;reverse complement&quot;</span><span class="punctuation">)</span> \
<span class="operator">...</span>            <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span><span class="operator">&lt;</span><span class="literal number integer">700</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">)</span>
<span class="literal number integer">18</span>
</pre>
<p>这将在内存中创建一个序列小于700bp的反向互补序列列表。我们可以以相同的方式使用生成器表达式 - 但是更有优势的是，它不需要同时在内存中创建所有序列条目的列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;rc_&quot;</span><span class="operator">+</span><span class="name">rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;reverse complement&quot;</span><span class="punctuation">)</span> \
<span class="operator">...</span>           <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span><span class="operator">&lt;</span><span class="literal number integer">700</span><span class="punctuation">)</span>
</pre>
<p>完整的示例如下：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;rc_&quot;</span><span class="operator">+</span><span class="name">rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;reverse complement&quot;</span><span class="punctuation">)</span> \
<span class="operator">...</span>            <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span><span class="operator">&lt;</span><span class="literal number integer">700</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;rev_comp.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">18</span>
</pre>
<p>在第 <a class="reference external" href="#sec:SeqIO-translate">18.1.3</a> 节有一个相关的示例，将FASTA文件中核酸序列翻译为氨基酸。</p>
</div>
<div class="section" id="id53">
<h3>5.5.4 获得格式化为字符串的 <tt class="docutils literal">SeqRecord</tt> 对象</h3>
<p>有时你不需要将序列条目写入文件或者句柄，而是想获得包含特定格式序列条目的字符串。 <tt class="docutils literal">Bio.SeqIO</tt> 接口基于句柄，但是Python有一个有用的内置模块，提供基于字符串的句柄。</p>
<p>举个例子来说明如果使用这个功能，我们先从兰花GenBank文件加载一系列 <tt class="docutils literal">SeqRecord</tt> 对象，然后创建一个包含FASTA格式序列条目的字符串：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">StringIO</span> <span class="keyword namespace">import</span> <span class="name">StringIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="name">out_handle</span> <span class="operator">=</span> <span class="name">StringIO</span><span class="punctuation">()</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">out_handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="name">fasta_data</span> <span class="operator">=</span> <span class="name">out_handle</span><span class="operator">.</span><span class="name">getvalue</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="name">fasta_data</span>
</pre>
<p>当你第一次看到，会觉得这并不够简单明了。在特殊情况下，你希望得到一个只包含特定格式的单条序列条目的字符串，可以使用 <tt class="docutils literal">SeqRecord</tt> 类的 <tt class="docutils literal">format()</tt> （请见第 <a class="reference external" href="#sec:SeqRecord-format">4.5</a> 节）。</p>
<p>注意：尽管我们不鼓励这么做，你可以使用 <tt class="docutils literal">format()</tt> 方法写入文件，示例如下：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">out_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid_long.tab&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">):</span>
    <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="literal number integer">100</span><span class="punctuation">:</span>
        <span class="name">out_handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;tab&quot;</span><span class="punctuation">))</span>
<span class="name">out_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>这类代码可以处理顺序文件格式如FASTA或者此处使用的简单的制表符分割文件，但不能处理更复杂的或是交错式文件格式。这就是为什么我们仍然强调使用 <tt class="docutils literal">Bio.SeqIO.write()</tt> 的原因，如下面的示例：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="literal number integer">100</span><span class="punctuation">)</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;ls_orchid.tab&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;tab&quot;</span><span class="punctuation">)</span>
</pre>
<p>同时 ，单次调用 <tt class="docutils literal"><span class="pre">SeqIO.write(...)</span></tt> 也比多次调用 <tt class="docutils literal"><span class="pre">SeqRecord.format(...)</span></tt> 方法更快。</p>
</div>
</div>
</div>
<div class="section" id="id54">
<h1>第6章 多序列比对</h1>
<p>多序列比对（Multiple Sequence Alignment, MSA）是指对多个序列进行对位排列。 这通常需要保证序列间的等同位点处在同一列上，并通过引进小横线（-）以保证最终的序列具有相同的长度。这种序列比对可以视作是由字符组成的矩阵。在Biopython中，多序列比对中每一个序列是以 <tt class="docutils literal">SeqRecord</tt> 对象来表示的。</p>
<p>这里我们介绍一种新的对象 -- <tt class="docutils literal">MultipleSeqAlignment</tt> 来表示这样一类数据，我们还将介绍 <tt class="docutils literal">Bio.AlignIO</tt> 模块来读写不同格式的多序列比对数据（ <tt class="docutils literal">Bio.AlignIO</tt> 在设计上与之前介绍的 <tt class="docutils literal">Bio.SeqIO</tt> 模块是类似的）。Biopython中， <tt class="docutils literal">Bio.SeqIO</tt> 和 <tt class="docutils literal">Bio.AlignIO</tt> 都能读写各种格式的多序列比对数据。在实际处理中，使用哪一个模块取决于用户需要对数据进行何种操作。</p>
<p>本章的最后一部分是关于各种常用多序列比对程序（ClustalW和MUSCLE）的Biopython命令行封装。</p>
<div class="section" id="id55">
<h2>6.1 读取多序列比对数据</h2>
<p>在Biopython中，有两种方法读取多序列比对数据， <tt class="docutils literal">Bio.AlignIO.read()</tt> 和 <tt class="docutils literal">Bio.AlignIO.parse()</tt> 。这两种方法跟 <tt class="docutils literal">Bio.SeqIO</tt> 处理一个和多个数据的设计方式是一样的。 <tt class="docutils literal">Bio.AlignIO.read()</tt> 只能读取一个多序列比对而 <tt class="docutils literal">Bio.AlignIO.parse()</tt> 可以依次读取多个序列比对数据。</p>
<p>使用 <tt class="docutils literal">Bio.AlignIO.parse()</tt> 将会返回一个 <tt class="docutils literal">MultipleSeqAlignment</tt> 的 <em>迭代器（iterator）</em> 。迭代器往往在循环中使用。在实际数据分析过程中会时常处理包含有多个多序列比对的文件。例如PHYLIP中的 <tt class="docutils literal">seqboot</tt> ，EMBOSS工具箱中的 <tt class="docutils literal">water</tt> 和 <tt class="docutils literal">needle</tt>, 以及Bill Pearson的FASTA程序。</p>
<p>然而在大多数情况下，你所遇到的文件仅仅包括一个多序列比对。这时，你应该使用 <tt class="docutils literal">Bio.AlignIO.read()</tt> ，这将返回一个 <tt class="docutils literal">MultipleSeqAlignment</tt> 对象。</p>
<p>这两个函数都接受两个必须参数：</p>
<ol class="arabic simple">
<li>第一个参数为包含有多序列比对数据的 <em>句柄（handle）</em> 。在实际操作中，这往往是一个打开的文件（详细信息请见 <a class="reference external" href="#sec:appendix-handles">22.1</a> ）或者文件名。</li>
<li>第二个参数为多序列比对文件格式（小写）。与 <tt class="docutils literal">Bio.SeqIO</tt> 模块一样，Biopython不会对将读取的文件格式进行猜测。所有 <tt class="docutils literal">Bio.AlignIO</tt> 模块支持的多序列比对数据格式可以在 <a class="reference external" href="http://biopython.org/wiki/AlignIO">http://biopython.org/wiki/AlignIO</a> 中找到。</li>
</ol>
<p><tt class="docutils literal">Bio.AlignIO</tt> 模块还接受一个可选参数 <tt class="docutils literal">seq_count</tt> 。这一参数将在 <a class="reference external" href="#sec:AlignIO-count-argument">6.1.3</a> 中具体讨论。它可以处理不确定的多序列比对格式，或者包含有多个序列的排列。</p>
<p>另一个可选参数 <tt class="docutils literal">alphabet</tt> 允许用户指定序列比对文件的字符（alphabet），它可以用来说明序列比对的类型（DNA，RNA或蛋白质）。因为大多数序列比对格式并不区别序列的类型，因此指定这一参数可能会对后期的分析产生帮助。 <tt class="docutils literal">Bio.AlignIO</tt> 默认将使用一般字符（generic alphabet），这将不区分各种序列比对类型。</p>
<div class="section" id="id56">
<h3>6.1.1 单一的序列比对</h3>
<p>例如，请见以下PFAM（或者Stockholm）格式的蛋白序列比对文件。</p>
<pre class="code python literal-block">
<span class="comment"># STOCKHOLM 1.0</span>
<span class="comment">#=GS COATB_BPIKE/30-81  AC P03620.1</span>
<span class="comment">#=GS COATB_BPIKE/30-81  DR PDB; 1ifl ; 1-52;</span>
<span class="comment">#=GS Q9T0Q8_BPIKE/1-52  AC Q9T0Q8.1</span>
<span class="comment">#=GS COATB_BPI22/32-83  AC P15416.1</span>
<span class="comment">#=GS COATB_BPM13/24-72  AC P69541.1</span>
<span class="comment">#=GS COATB_BPM13/24-72  DR PDB; 2cpb ; 1-49;</span>
<span class="comment">#=GS COATB_BPM13/24-72  DR PDB; 2cps ; 1-49;</span>
<span class="comment">#=GS COATB_BPZJ2/1-49   AC P03618.1</span>
<span class="comment">#=GS Q9T0Q9_BPFD/1-49   AC Q9T0Q9.1</span>
<span class="comment">#=GS Q9T0Q9_BPFD/1-49   DR PDB; 1nh4 A; 1-49;</span>
<span class="comment">#=GS COATB_BPIF1/22-73  AC P03619.2</span>
<span class="comment">#=GS COATB_BPIF1/22-73  DR PDB; 1ifk ; 1-50;</span>
<span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>             <span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span>
<span class="comment">#=GR COATB_BPIKE/30-81  SS    -HHHHHHHHHHHHHH--HHHHHHHH--HHHHHHHHHHHHHHHHHHHHH----</span>
<span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>             <span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span>
<span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>             <span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span>
<span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>             <span class="name">AEGDDP</span><span class="operator">...</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span>
<span class="comment">#=GR COATB_BPM13/24-72  SS    ---S-T...CHCHHHHCCCCTCCCTTCHHHHHHHHHHHHHHHHHHHHCTT--</span>
<span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>              <span class="name">AEGDDP</span><span class="operator">...</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span>
<span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>              <span class="name">AEGDDP</span><span class="operator">...</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span>
<span class="comment">#=GR Q9T0Q9_BPFD/1-49   SS    ------...-HHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH--</span>
<span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>             <span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span>
<span class="comment">#=GR COATB_BPIF1/22-73  SS    XX-HHHH--HHHHHH--HHHHHHH--HHHHHHHHHHHHHHHHHHHHHHH---</span>
<span class="comment">#=GC SS_cons                  XHHHHHHHHHHHHHHHCHHHHHHHHCHHHHHHHHHHHHHHHHHHHHHHHC--</span>
<span class="comment">#=GC seq_cons                 AEssss...AptAhDSLpspAT-hIu.sWshVsslVsAsluIKLFKKFsSKA</span>
<span class="operator">//</span>
</pre>
<p>这是PFAM数据库中Phage_Coat_Gp8的种子排列（PF05371）。该排列下载于一个已经过期的PFAM数据库版本（ <a class="reference external" href="http://pfam.sanger.ac.uk/">http://pfam.sanger.ac.uk/</a> ），但这并不影响我们的例子（假设你已经将以上内容下载到一个名为''PF05371_seed.sth''的文件中，并在Python的当前工作目录下）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
</pre>
<p>这段代码将在屏幕上打印出该序列比对的概要信息：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">52</span> <span class="name">columns</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKL</span><span class="operator">...</span><span class="name">SRA</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKL</span><span class="operator">...</span><span class="name">SRA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>你会注意到，以上输出截短了中间一部分序列的内容。你也可以很容易地通过控制多序列比对中每一条序列（作为 <tt class="docutils literal">SeqRecord</tt> 对象）来输出你所喜欢的格式。例如：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;Alignment length </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">get_alignment_length</span><span class="punctuation">()</span>
<span class="name">Alignment</span> <span class="name">length</span> <span class="literal number integer">52</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> - </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span> <span class="operator">-</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span> <span class="operator">-</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span> <span class="operator">-</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="operator">-</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span> <span class="operator">-</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="operator">-</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span> <span class="operator">-</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>你也可以使用上面alignment对象的 <tt class="docutils literal">format</tt> 方法来以指定的格式显示它。具体信息可以参见 <a class="reference external" href="#sec:alignment-format-method">6.2.2</a> 。</p>
<p>你是否已经注意到以上原始数据文件中包含有蛋白数据库（PDB）交叉引用以及相关二级结构的信息？你可以尝试以下代码：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">dbxrefs</span>
<span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span> <span class="punctuation">[</span><span class="literal string">'PDB; 1ifl ; 1-52;'</span><span class="punctuation">]</span>
<span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span> <span class="punctuation">[</span><span class="literal string">'PDB; 2cpb ; 1-49;'</span><span class="punctuation">,</span> <span class="literal string">'PDB; 2cps ; 1-49;'</span><span class="punctuation">]</span>
<span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span> <span class="punctuation">[</span><span class="literal string">'PDB; 1nh4 A; 1-49;'</span><span class="punctuation">]</span>
<span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span> <span class="punctuation">[</span><span class="literal string">'PDB; 1ifk ; 1-50;'</span><span class="punctuation">]</span>
</pre>
<p>如果你希望显示所有的序列注释信息，请使用以下例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span>
</pre>
<p>Sanger网站
<a class="reference external" href="http://pfam.sanger.ac.uk/family?acc=PF05371">http://pfam.sanger.ac.uk/family?acc=PF05371</a>
可以让你下载各种不同的序列比对的格式。以下例子为FASTA格式：</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span>
<span class="operator">&gt;</span><span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span>
<span class="operator">&gt;</span><span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span>
<span class="operator">&gt;</span><span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span>
<span class="operator">&gt;</span><span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span>
<span class="operator">&gt;</span><span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span>
<span class="operator">&gt;</span><span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
<span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span>
</pre>
<p>注意Sanger网站有一个选项可以将序列比对中的间隔（gap）用小圆点或者是小横线表示。在以上例子中，序列间隔由小横线表示。假设你已经下载该文件，并保存为 “PF05371_seed.faa”。你可以使用以下代码来读入该序列比对。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.faa&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name">alignment</span>
</pre>
<p>你可能已经发现，以上代码中唯一的变化只是指定格式的参数。所返回的alignment对象将会包含同样的序列和序列名字。但是仔细的读者会发现，每一个alignment的SeqRecord中并不包含数据的引用注释。这是因为FASTA格式本身并没有包含这一类信息。</p>
<p>此外，除了使用Sanger网站，你也可以利用 <tt class="docutils literal">Bio.AlignIO</tt> 来将原始的Stockholm格式转换成FASTA文件格式（见下文）。</p>
<p>对于任何一种Biopython支持的格式，你都可以用同样的方式读取它（通过指定文件的格式）。例如，你可以使用“phylip”来表示PHYLIP格式文件，用&quot;nexus&quot;来指定NEXUS格式文件或者用“emboss”来指定EMBOSS工具箱的输出文件。读者可以在以下链接中找到所有支持的格式（ <a class="reference external" href="http://biopython.org/wiki/AlignIO">http://biopython.org/wiki/AlignIO</a> ），或者内置的帮助中（以及在线文档 <a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.AlignIO-module.html">online</a> ）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">AlignIO</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="id57">
<h3>6.1.2  多个序列比对</h3>
<p>在前一章中，我们旨在读取仅包含有一个序列比对的文件。然而，在很多情况下，文件可能包含有多个序列比对。这时，你可以使用 <tt class="docutils literal">Bio.AlignIO.parse()</tt> 来读取它们。</p>
<p>假设我们有一个PHYLIP格式的很小的序列比对：</p>
<pre class="code python literal-block">
    <span class="literal number integer">5</span>    <span class="literal number integer">6</span>
<span class="name">Alpha</span>     <span class="name">AACAAC</span>
<span class="name">Beta</span>      <span class="name">AACCCC</span>
<span class="name">Gamma</span>     <span class="name">ACCAAC</span>
<span class="name">Delta</span>     <span class="name">CCACCA</span>
<span class="name">Epsilon</span>   <span class="name">CCAAAC</span>
</pre>
<p>如果你想用PHYLIP工具包来bootstrap一个系统发生树，其中的一个步骤是用 <tt class="docutils literal">bootseq</tt> 程序来产生许多序列比对。这将给出类似于以下格式的序列比对：</p>
<pre class="code python literal-block">
    <span class="literal number integer">5</span>     <span class="literal number integer">6</span>
<span class="name">Alpha</span>     <span class="name">AAACCA</span>
<span class="name">Beta</span>      <span class="name">AAACCC</span>
<span class="name">Gamma</span>     <span class="name">ACCCCA</span>
<span class="name">Delta</span>     <span class="name">CCCAAC</span>
<span class="name">Epsilon</span>   <span class="name">CCCAAA</span>
    <span class="literal number integer">5</span>     <span class="literal number integer">6</span>
<span class="name">Alpha</span>     <span class="name">AAACAA</span>
<span class="name">Beta</span>      <span class="name">AAACCC</span>
<span class="name">Gamma</span>     <span class="name">ACCCAA</span>
<span class="name">Delta</span>     <span class="name">CCCACC</span>
<span class="name">Epsilon</span>   <span class="name">CCCAAA</span>
    <span class="literal number integer">5</span>     <span class="literal number integer">6</span>
<span class="name">Alpha</span>     <span class="name">AAAAAC</span>
<span class="name">Beta</span>      <span class="name">AAACCC</span>
<span class="name">Gamma</span>     <span class="name">AACAAC</span>
<span class="name">Delta</span>     <span class="name">CCCCCA</span>
<span class="name">Epsilon</span>   <span class="name">CCCAAC</span>
<span class="operator">...</span>
    <span class="literal number integer">5</span>     <span class="literal number integer">6</span>
<span class="name">Alpha</span>     <span class="name">AAAACC</span>
<span class="name">Beta</span>      <span class="name">ACCCCC</span>
<span class="name">Gamma</span>     <span class="name">AAAACC</span>
<span class="name">Delta</span>     <span class="name">CCCCAA</span>
<span class="name">Epsilon</span>   <span class="name">CAAACC</span>
</pre>
<p>如果你想用 <tt class="docutils literal">Bio.AlignIO</tt> 来读取这个文件，你可以使用：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignments</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;resampled.phy&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phylip&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">alignment</span> <span class="operator word">in</span> <span class="name">alignments</span><span class="punctuation">:</span>
    <span class="keyword">print</span> <span class="name">alignment</span>
    <span class="keyword">print</span>
</pre>
<p>这将给出以下的输出（这时只显示缩略的一部分）：</p>
<pre class="code python literal-block">
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">5</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">6</span> <span class="name">columns</span>
<span class="name">AAACCA</span> <span class="name">Alpha</span>
<span class="name">AAACCC</span> <span class="name">Beta</span>
<span class="name">ACCCCA</span> <span class="name">Gamma</span>
<span class="name">CCCAAC</span> <span class="name">Delta</span>
<span class="name">CCCAAA</span> <span class="name">Epsilon</span>

<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">5</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">6</span> <span class="name">columns</span>
<span class="name">AAACAA</span> <span class="name">Alpha</span>
<span class="name">AAACCC</span> <span class="name">Beta</span>
<span class="name">ACCCAA</span> <span class="name">Gamma</span>
<span class="name">CCCACC</span> <span class="name">Delta</span>
<span class="name">CCCAAA</span> <span class="name">Epsilon</span>

<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">5</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">6</span> <span class="name">columns</span>
<span class="name">AAAAAC</span> <span class="name">Alpha</span>
<span class="name">AAACCC</span> <span class="name">Beta</span>
<span class="name">AACAAC</span> <span class="name">Gamma</span>
<span class="name">CCCCCA</span> <span class="name">Delta</span>
<span class="name">CCCAAC</span> <span class="name">Epsilon</span>

<span class="operator">...</span>

<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">5</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">6</span> <span class="name">columns</span>
<span class="name">AAAACC</span> <span class="name">Alpha</span>
<span class="name">ACCCCC</span> <span class="name">Beta</span>
<span class="name">AAAACC</span> <span class="name">Gamma</span>
<span class="name">CCCCAA</span> <span class="name">Delta</span>
<span class="name">CAAACC</span> <span class="name">Epsilon</span>
</pre>
<p>与 <tt class="docutils literal">Bio.SeqIO.parse</tt> 一样， <tt class="docutils literal">Bio.SeqIO.parse()</tt> 将返回一个迭代器（iterator）。如果你希望把所有的序列比对都读取到内存中，以下代码将把它们储存在一个列表对象里。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignments</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">AlignIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;resampled.phy&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phylip&quot;</span><span class="punctuation">))</span>
<span class="name">last_align</span> <span class="operator">=</span> <span class="name">alignments</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="name">first_align</span> <span class="operator">=</span> <span class="name">alignments</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="id58">
<h3>6.1.3  含糊的序列比对</h3>
<p>许多序列比对的文件格式可以非常明确地储存多个序列比对。然而，例如FASTA一类的普通序列文件格式并没有很直接的分隔符来分开多个序列比对。读者可以见以下例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAG</span><span class="operator">--</span><span class="name">G</span>
<span class="operator">&gt;</span><span class="name">Beta</span>
<span class="name">ACTACCGCTAGCTCAGAAG</span>
<span class="operator">&gt;</span><span class="name">Gamma</span>
<span class="name">ACTACGGCTAGCACAGAAG</span>
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAGG</span><span class="operator">--</span>
<span class="operator">&gt;</span><span class="name">Beta</span>
<span class="name">ACTACCGCTAGCTCAGAAG</span>
<span class="operator">&gt;</span><span class="name">Gamma</span>
<span class="name">ACTACGGCTAGCACAGAAG</span>
</pre>
<p>以上FASTA格式文件可以认为是一个包含有6条序列的序列比对（有重复序列名）。或者从文件名来看，这很可能是两个序列比对，每一个包含有三个序列，只是这两个序列比对恰好具有相同的长度。</p>
<p>以下是另一个例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAG</span><span class="operator">--</span><span class="name">G</span>
<span class="operator">&gt;</span><span class="name">Beta</span>
<span class="name">ACTACCGCTAGCTCAGAAG</span>
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAGG</span><span class="operator">--</span>
<span class="operator">&gt;</span><span class="name">Gamma</span>
<span class="name">ACTACGGCTAGCACAGAAG</span>
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAGG</span><span class="operator">--</span>
<span class="operator">&gt;</span><span class="name">Delta</span>
<span class="name">ACTACGGCTAGCACAGAAG</span>
</pre>
<p>同样，这也可能是一个包含有六个序列的序列比对。然而，根据序列名判断，这很可能是三个两两间的序列比较，而且恰好有同样的长度。</p>
<p>最后一个例子也类似：</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAG</span><span class="operator">--</span><span class="name">G</span>
<span class="operator">&gt;</span><span class="name">XXX</span>
<span class="name">ACTACCGCTAGCTCAGAAG</span>
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="name">ACTACGACTAGCTCAGG</span>
<span class="operator">&gt;</span><span class="name">YYY</span>
<span class="name">ACTACGGCAAGCACAGG</span>
<span class="operator">&gt;</span><span class="name">Alpha</span>
<span class="operator">--</span><span class="name">ACTACGAC</span><span class="operator">--</span><span class="name">TAGCTCAGG</span>
<span class="operator">&gt;</span><span class="name">ZZZ</span>
<span class="name">GGACTACGACAATAGCTCAGG</span>
</pre>
<p>在这一个例子中，由于序列有不同的长度，这不能被当作是一个包含六个序列的单独的序列比对。很显然，这可以被看成是三个两两间的序列比对。</p>
<p>很明显，将多个序列比对以FASTA格式储存并不方便。然而，在某些情况下，如果你一定要这么做， <tt class="docutils literal">Bio.AlignIO</tt> 依然能够处理上述情形（但是所有的序列比对必须都含有相同的序列）。一个很常见的例子是，我们经常会使用EMBOSS工具箱中的 <tt class="docutils literal">needle</tt> 和 <tt class="docutils literal">water</tt> 来产生许多两两间的序列比对 —— 然而在这种情况下，你可以指定数据格式为“emboss”，<tt class="docutils literal">Bio.AlignIO</tt> 仍然能够识别这些原始输出。</p>
<p>为了处理这样的FASTA格式的数据，我们可以指定 <tt class="docutils literal">Bio.AlignIO.parse()</tt> 的第三个可选参数 <tt class="docutils literal">seq_count</tt> ，这一参数将告诉Biopython你所期望的每个序列比对中序列的个数。例如：</p>
<pre class="code python literal-block">
<span class="keyword">for</span> <span class="name">alignment</span> <span class="operator word">in</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">,</span> <span class="name">seq_count</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="literal string">&quot;Alignment length </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">get_alignment_length</span><span class="punctuation">()</span>
    <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">:</span>
        <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> - </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>
    <span class="keyword">print</span>
</pre>
<p>这将给出：</p>
<pre class="code python literal-block">
<span class="name">Alignment</span> <span class="name">length</span> <span class="literal number integer">19</span>
<span class="name">ACTACGACTAGCTCAG</span><span class="operator">--</span><span class="name">G</span> <span class="operator">-</span> <span class="name">Alpha</span>
<span class="name">ACTACCGCTAGCTCAGAAG</span> <span class="operator">-</span> <span class="name">XXX</span>

<span class="name">Alignment</span> <span class="name">length</span> <span class="literal number integer">17</span>
<span class="name">ACTACGACTAGCTCAGG</span> <span class="operator">-</span> <span class="name">Alpha</span>
<span class="name">ACTACGGCAAGCACAGG</span> <span class="operator">-</span> <span class="name">YYY</span>

<span class="name">Alignment</span> <span class="name">length</span> <span class="literal number integer">21</span>
<span class="operator">--</span><span class="name">ACTACGAC</span><span class="operator">--</span><span class="name">TAGCTCAGG</span> <span class="operator">-</span> <span class="name">Alpha</span>
<span class="name">GGACTACGACAATAGCTCAGG</span> <span class="operator">-</span> <span class="name">ZZZ</span>
</pre>
<p>如果你使用 <tt class="docutils literal">Bio.AlignIO.read()</tt> 或者 <tt class="docutils literal">Bio.AlignIO.parse()</tt> 而不指定 <tt class="docutils literal">seq_count</tt> ，这将返回一个包含有六条序列的序列比对。对于上面的第三个例子，由于序列长度不同，导致它们不能被解析为一个序列比对，Biopython将会抛出一个异常。</p>
<p>如果数据格式本身包含有分割符， <tt class="docutils literal">Bio.AlignIO</tt> 可以很聪明地自动确定文件中每一个序列比对，而无需指定 <tt class="docutils literal">seq_count</tt> 选项。如果你仍然指定 <tt class="docutils literal">seq_count</tt> 但是却与数据本身的分隔符相冲突，Biopython将产生一个错误。</p>
<p>注意指定这一可选的 <tt class="docutils literal">seq_count</tt> 参数将假设文件中所有的序列比对都包含相同数目的序列。假如你真的遇到每一个序列比对都有不同数目的序列， <tt class="docutils literal">Bio.AlignIO</tt> 将无法读取。这时，我们建议你使用 <tt class="docutils literal">Bio.SeqIO</tt> 来读取数据，然后将序列转换为序列比对。</p>
</div>
</div>
<div class="section" id="id59">
<h2>6.2  序列比对的写出</h2>
<p>我们已经讨论了 <tt class="docutils literal">Bio.AlignIO.read()</tt> 和 <tt class="docutils literal">Bio.AlignIO.parse()</tt> 来读取各种格式的序列比对，现在让我们来使用 <tt class="docutils literal">Bio.AlignIO.write()</tt> 写出序列比对文件。</p>
<p>这一函数接受三个参数：一个 <tt class="docutils literal">MultipleSeqAlignment</tt> 对象（或者是一个 <tt class="docutils literal">Alignment</tt> 对象），一个可写的文件句柄（handle）或者期望写出的文件名，以及写出文件的格式。</p>
<p>这里有一个手动构造一个 <tt class="docutils literal">MultipleSeqAlignment</tt> 对象的例子（注意 <tt class="docutils literal">MultipleSeqAlignment</tt> 是由若干个 <tt class="docutils literal">SeqRecord</tt> 组成的）：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">generic_dna</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Align</span> <span class="keyword namespace">import</span> <span class="name">MultipleSeqAlignment</span>

<span class="name">align1</span> <span class="operator">=</span> <span class="name">MultipleSeqAlignment</span><span class="punctuation">([</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACTGCTAGCTAG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Alpha&quot;</span><span class="punctuation">),</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACT-CTAGCTAG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Beta&quot;</span><span class="punctuation">),</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACTGCTAGDTAG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Gamma&quot;</span><span class="punctuation">),</span>
         <span class="punctuation">])</span>

<span class="name">align2</span> <span class="operator">=</span> <span class="name">MultipleSeqAlignment</span><span class="punctuation">([</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GTCAGC-AG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Delta&quot;</span><span class="punctuation">),</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GACAGCTAG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Epsilon&quot;</span><span class="punctuation">),</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;GTCAGCTAG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Zeta&quot;</span><span class="punctuation">),</span>
         <span class="punctuation">])</span>

<span class="name">align3</span> <span class="operator">=</span> <span class="name">MultipleSeqAlignment</span><span class="punctuation">([</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACTAGTACAGCTG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Eta&quot;</span><span class="punctuation">),</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;ACTAGTACAGCT-&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Theta&quot;</span><span class="punctuation">),</span>
             <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;-CTACTACAGGTG&quot;</span><span class="punctuation">,</span> <span class="name">generic_dna</span><span class="punctuation">),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Iota&quot;</span><span class="punctuation">),</span>
         <span class="punctuation">])</span>

<span class="name">my_alignments</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">align1</span><span class="punctuation">,</span> <span class="name">align2</span><span class="punctuation">,</span> <span class="name">align3</span><span class="punctuation">]</span>
</pre>
<p>现在我们有一个包含三个 <tt class="docutils literal">MultipleSeqAlignment</tt> 对象的列表（ <tt class="docutils literal">my_alignments</tt> ），现在我们将它写出为PHYLIP格式：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">AlignIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">my_alignments</span><span class="punctuation">,</span> <span class="literal string">&quot;my_example.phy&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phylip&quot;</span><span class="punctuation">)</span>
</pre>
<p>如果你用你喜欢的文本编辑器在你当前的工作目录下打开 <tt class="docutils literal">my_example.phy</tt> 文件，你会看到以下内容：</p>
<pre class="code python literal-block">
 <span class="literal number integer">3</span> <span class="literal number integer">12</span>
<span class="name">Alpha</span>      <span class="name">ACTGCTAGCT</span> <span class="name">AG</span>
<span class="name">Beta</span>       <span class="name">ACT</span><span class="operator">-</span><span class="name">CTAGCT</span> <span class="name">AG</span>
<span class="name">Gamma</span>      <span class="name">ACTGCTAGDT</span> <span class="name">AG</span>
 <span class="literal number integer">3</span> <span class="literal number integer">9</span>
<span class="name">Delta</span>      <span class="name">GTCAGC</span><span class="operator">-</span><span class="name">AG</span>
<span class="name">Epislon</span>    <span class="name">GACAGCTAG</span>
<span class="name">Zeta</span>       <span class="name">GTCAGCTAG</span>
 <span class="literal number integer">3</span> <span class="literal number integer">13</span>
<span class="name">Eta</span>        <span class="name">ACTAGTACAG</span> <span class="name">CTG</span>
<span class="name">Theta</span>      <span class="name">ACTAGTACAG</span> <span class="name">CT</span><span class="operator">-</span>
<span class="name">Iota</span>       <span class="operator">-</span><span class="name">CTACTACAG</span> <span class="name">GTG</span>
</pre>
<p>在更多情况下，你希望读取一个已经含有序列比对的文件，经过某些操作（例如去掉一些行和列）然后将它重新储存起来。</p>
<p>假如你希望知道有多少序列比对被 <tt class="docutils literal">Bio.AlignIO.write()</tt> 函数写入句柄中。如果你的序列比对都被放在一个列表中（如同以上的例子），你可以很容易地使用 <tt class="docutils literal">len(my_alignments)</tt> 来获得这一信息。然而，如果你的序列比对在一个生成器/迭代器对象中，你无法轻松地完成这件事情。为此， <tt class="docutils literal">Bio.AlignIO.write()</tt> 将会返回它所写出的序列比对个数。</p>
<p><em>注意</em> - 如果你所指定给 <tt class="docutils literal">Bio.AlignIO.write()</tt> 的文件已经存在在当前目录下，这一文件将被直接覆盖掉而不会有任何警告。</p>
<div class="section" id="id60">
<h3>6.2.1  序列比对的格式间转换</h3>
<p><tt class="docutils literal">Bio.AlignIO</tt> 模块中的序列比对格式转换功能与 <tt class="docutils literal">Bio.SeqIO</tt> （见 <a class="reference external" href="#sec:SeqIO-conversion">5.5.2</a> ）模块的格式转换是一样的。在通常情况下，我们建议使用 <tt class="docutils literal">Bio.AlignIO.parse()</tt> 来读取序列比对数据，然后使用 <tt class="docutils literal">Bio.AlignIO.write()</tt> 函数来写出。或者你也可以直接使用 <tt class="docutils literal">Bio.AlignIO.convert()</tt> 函数来实现格式的转换。</p>
<p>在本例中，我们将读取PFAM/Stockholm格式的序列比对，然后将其保存为Clustal格式：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PF05371_seed.aln&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Converted </span><span class="literal string interpol">%i</span><span class="literal string"> alignments&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>或者，使用 <tt class="docutils literal">Bio.AlignIO.parse()</tt> 和 <tt class="docutils literal">Bio.AlignIO.write()</tt> ：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignments</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">alignments</span><span class="punctuation">,</span> <span class="literal string">&quot;PF05371_seed.aln&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Converted </span><span class="literal string interpol">%i</span><span class="literal string"> alignments&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p><tt class="docutils literal">Bio.AlignIO.write()</tt> 函数默认处理的情形是一个包括有多个序列比对的对象。在以上例子中，我们给予 <tt class="docutils literal">Bio.AlignIO.write()</tt> 的参数是一个由 <tt class="docutils literal">Bio.AlignIO.parse()</tt> 函数返回的一个迭代器。</p>
<p>在以下例子中，我们知道序列比对文件中仅包含有一个序列比对，因此我们使用 <tt class="docutils literal">Bio.AlignIO.read()</tt> 函数来读取数据，然后使用 <tt class="docutils literal">Bio.AlignIO.write()</tt> 来将数据保存为另一种格式：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="name">AlignIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">([</span><span class="name">alignment</span><span class="punctuation">],</span> <span class="literal string">&quot;PF05371_seed.aln&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
</pre>
<p>使用以上两个例子，你都可以将PFAM/Stockholm格式的序列比对数据转换为Clustal格式：</p>
<pre class="code python literal-block">
<span class="name">CLUSTAL</span> <span class="name">X</span> <span class="punctuation">(</span><span class="literal number float">1.81</span><span class="punctuation">)</span> <span class="name">multiple</span> <span class="name">sequence</span> <span class="name">alignment</span>


<span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>                   <span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSS</span>
<span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>                   <span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVS</span>
<span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>                   <span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSS</span>
<span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>                   <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTS</span>
<span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>                    <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFAS</span>
<span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>                    <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTS</span>
<span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>                   <span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVS</span>

<span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>                   <span class="name">KA</span>
<span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>                   <span class="name">RA</span>
<span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>                   <span class="name">KA</span>
<span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>                   <span class="name">KA</span>
<span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>                    <span class="name">KA</span>
<span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>                    <span class="name">KA</span>
<span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>                   <span class="name">RA</span>
</pre>
<p>另外，你也可以使用以下代码将它保存为PHYLIP格式：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">AlignIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PF05371_seed.phy&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phylip&quot;</span><span class="punctuation">)</span>
</pre>
<p>你可以获得以下PHYLIP格式的文件输出：</p>
<pre class="code python literal-block">
 <span class="literal number integer">7</span> <span class="literal number integer">52</span>
<span class="name">COATB_BPIK</span> <span class="name">AEPNAATNYA</span> <span class="name">TEAMDSLKTQ</span> <span class="name">AIDLISQTWP</span> <span class="name">VVTTVVVAGL</span> <span class="name">VIRLFKKFSS</span>
<span class="name">Q9T0Q8_BPI</span> <span class="name">AEPNAATNYA</span> <span class="name">TEAMDSLKTQ</span> <span class="name">AIDLISQTWP</span> <span class="name">VVTTVVVAGL</span> <span class="name">VIKLFKKFVS</span>
<span class="name">COATB_BPI2</span> <span class="name">DGTSTATSYA</span> <span class="name">TEAMNSLKTQ</span> <span class="name">ATDLIDQTWP</span> <span class="name">VVTSVAVAGL</span> <span class="name">AIRLFKKFSS</span>
<span class="name">COATB_BPM1</span> <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">A</span> <span class="name">KAAFNSLQAS</span> <span class="name">ATEYIGYAWA</span> <span class="name">MVVVIVGATI</span> <span class="name">GIKLFKKFTS</span>
<span class="name">COATB_BPZJ</span> <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">A</span> <span class="name">KAAFDSLQAS</span> <span class="name">ATEYIGYAWA</span> <span class="name">MVVVIVGATI</span> <span class="name">GIKLFKKFAS</span>
<span class="name">Q9T0Q9_BPF</span> <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">A</span> <span class="name">KAAFDSLQAS</span> <span class="name">ATEYIGYAWA</span> <span class="name">MVVVIVGATI</span> <span class="name">GIKLFKKFTS</span>
<span class="name">COATB_BPIF</span> <span class="name">FAADDATSQA</span> <span class="name">KAAFDSLTAQ</span> <span class="name">ATEMSGYAWA</span> <span class="name">LVVLVVGATV</span> <span class="name">GIKLFKKFVS</span>

           <span class="name">KA</span>
           <span class="name">RA</span>
           <span class="name">KA</span>
           <span class="name">KA</span>
           <span class="name">KA</span>
           <span class="name">KA</span>
           <span class="name">RA</span>
</pre>
<p>PHYLIP格式最大的一个缺陷就是它严格地要求每一条序列的ID是都为10个字符（ID中多出的字符将被截短）。在这一个例子中，截短的序列ID依然是唯一的（只是缺少了可读性）。在某些情况下，我们并没有一个好的方式去压缩序列的ID。以下例子提供了另一种解决方案 —— 利用自定义的序列ID来代替原本的序列ID：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="name">name_mapping</span> <span class="operator">=</span> <span class="punctuation">{}</span>
<span class="keyword">for</span> <span class="name">i</span><span class="punctuation">,</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name builtin">enumerate</span><span class="punctuation">(</span><span class="name">alignment</span><span class="punctuation">):</span>
    <span class="name">name_mapping</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span>
    <span class="name">record</span><span class="operator">.</span><span class="name">id</span> <span class="operator">=</span> <span class="literal string">&quot;seq</span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">i</span>
<span class="keyword">print</span> <span class="name">name_mapping</span>

<span class="name">AlignIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">([</span><span class="name">alignment</span><span class="punctuation">],</span> <span class="literal string">&quot;PF05371_seed.phy&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phylip&quot;</span><span class="punctuation">)</span>
</pre>
<p>以上代码将会建立一个字典对象实现自定义的ID和原始ID的映射：</p>
<pre class="code python literal-block">
<span class="punctuation">{</span><span class="literal number integer">0</span><span class="punctuation">:</span> <span class="literal string">'COATB_BPIKE/30-81'</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">:</span> <span class="literal string">'Q9T0Q8_BPIKE/1-52'</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">:</span> <span class="literal string">'COATB_BPI22/32-83'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">}</span>
</pre>
<p>以下为PHYLIP的格式输出：</p>
<pre class="code python literal-block">
 <span class="literal number integer">7</span> <span class="literal number integer">52</span>
<span class="name">seq0</span>       <span class="name">AEPNAATNYA</span> <span class="name">TEAMDSLKTQ</span> <span class="name">AIDLISQTWP</span> <span class="name">VVTTVVVAGL</span> <span class="name">VIRLFKKFSS</span>
<span class="name">seq1</span>       <span class="name">AEPNAATNYA</span> <span class="name">TEAMDSLKTQ</span> <span class="name">AIDLISQTWP</span> <span class="name">VVTTVVVAGL</span> <span class="name">VIKLFKKFVS</span>
<span class="name">seq2</span>       <span class="name">DGTSTATSYA</span> <span class="name">TEAMNSLKTQ</span> <span class="name">ATDLIDQTWP</span> <span class="name">VVTSVAVAGL</span> <span class="name">AIRLFKKFSS</span>
<span class="name">seq3</span>       <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">A</span> <span class="name">KAAFNSLQAS</span> <span class="name">ATEYIGYAWA</span> <span class="name">MVVVIVGATI</span> <span class="name">GIKLFKKFTS</span>
<span class="name">seq4</span>       <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">A</span> <span class="name">KAAFDSLQAS</span> <span class="name">ATEYIGYAWA</span> <span class="name">MVVVIVGATI</span> <span class="name">GIKLFKKFAS</span>
<span class="name">seq5</span>       <span class="name">AEGDDP</span><span class="operator">---</span><span class="name">A</span> <span class="name">KAAFDSLQAS</span> <span class="name">ATEYIGYAWA</span> <span class="name">MVVVIVGATI</span> <span class="name">GIKLFKKFTS</span>
<span class="name">seq6</span>       <span class="name">FAADDATSQA</span> <span class="name">KAAFDSLTAQ</span> <span class="name">ATEMSGYAWA</span> <span class="name">LVVLVVGATV</span> <span class="name">GIKLFKKFVS</span>

           <span class="name">KA</span>
           <span class="name">RA</span>
           <span class="name">KA</span>
           <span class="name">KA</span>
           <span class="name">KA</span>
           <span class="name">KA</span>
           <span class="name">RA</span>
</pre>
<p>由于序列ID的限制性，PHYLIP格式不是储存序列比对的理想格式。我们建议你将数据储存成PFAM/Stockholm或者其它能对序列比对进行注释的格式来保存你的数据。</p>
</div>
<div class="section" id="formatted-strings">
<h3>6.2.2  将序列比对对象转换为格式化字符串（formatted strings）</h3>
<p>因为 <tt class="docutils literal">Bio.AlignIO</tt> 模块是基于文件句柄的，因此你如果想将序列比对读入为一个字符串对象，你需要做一些额外的工作。然而，我们提供一个 <tt class="docutils literal">format()</tt> 方法来帮助你实现这项任务。 <tt class="docutils literal">format()</tt> 方法需要用户提供一个小写的格式参数（这可以是任何 <tt class="docutils literal">AlignIO</tt> 支持的序列比对格式）。例如：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
</pre>
<p>我们在 <a class="reference external" href="#sec:SeqRecord-format">4.5</a> 中讲到， <tt class="docutils literal">Bio.SeqIO</tt> 也有一个对 <tt class="docutils literal">SeqRecord</tt> 输出的方法。</p>
<p><tt class="docutils literal">format()</tt> 方法是利用 <tt class="docutils literal">StringIO</tt> 以及 <tt class="docutils literal">Bio.AlignIO.write()</tt> 来实现以上输出的。如果你使用的是较老版本的Biopython，你可以使用以下代码来完成相同的工作：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">StringIO</span> <span class="keyword namespace">import</span> <span class="name">StringIO</span>

<span class="name">alignments</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>

<span class="name">out_handle</span> <span class="operator">=</span> <span class="name">StringIO</span><span class="punctuation">()</span>
<span class="name">AlignIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">alignments</span><span class="punctuation">,</span> <span class="name">out_handle</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
<span class="name">clustal_data</span> <span class="operator">=</span> <span class="name">out_handle</span><span class="operator">.</span><span class="name">getvalue</span><span class="punctuation">()</span>

<span class="keyword">print</span> <span class="name">clustal_data</span>
</pre>
</div>
</div>
<div class="section" id="id61">
<h2>6.3  序列比对的操纵</h2>
<p>现在我们已经了解了如何读入和写出序列比对。让我们继续看看如何对读入的序列比对进行操作。</p>
<div class="section" id="slice">
<h3>6.3.1  序列比对的切片（slice）操作</h3>
<p>首先，用户可以认为读入的序列比对是一个由 <tt class="docutils literal">SeqRecord</tt> 对象构成的Python列表（list）。有了这样一个印象以后，你可以使用 <tt class="docutils literal">len()</tt> 方法来得到行数（序列比对的个数），你也可以对序列比对进行迭代。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;Number of rows: </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">alignment</span><span class="punctuation">)</span>
<span class="name">Number</span> <span class="name">of</span> <span class="name">rows</span><span class="punctuation">:</span> <span class="literal number integer">7</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> - </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span> <span class="operator">-</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span> <span class="operator">-</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span> <span class="operator">-</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="operator">-</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span> <span class="operator">-</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="operator">-</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span> <span class="operator">-</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>你可以使用列表所拥有的 <tt class="docutils literal">append</tt> 和 <tt class="docutils literal">extend</tt> 方法来给序列比对增加序列。请读者一定要正确理解序列比对与其包含的序列的关系，这样你就可以使用切片操作来获得其中某些序列比对。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">52</span> <span class="name">columns</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAATNYATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKL</span><span class="operator">...</span><span class="name">SRA</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">DGTSTATSYATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKL</span><span class="operator">...</span><span class="name">SRA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">:</span><span class="literal number integer">7</span><span class="punctuation">]</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">4</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">52</span> <span class="name">columns</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span><span class="operator">---</span><span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKL</span><span class="operator">...</span><span class="name">SKA</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDATSQAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKL</span><span class="operator">...</span><span class="name">SRA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>假如你需要获得特定的列该怎么办呢？如果你接触过Numpy矩阵那么一定对下面的语法非常熟悉，使用双切片：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number integer">6</span><span class="punctuation">]</span>
<span class="name">T</span>
</pre>
<p>使用两个整数来获得序列比对中的一个字符，这其实是以下操作的简化方式：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">]</span>
<span class="name">T</span>
</pre>
<p>你可以用下面的代码来获取整列：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[:,</span><span class="literal number integer">6</span><span class="punctuation">]</span>
<span class="name">TTT</span><span class="operator">---</span><span class="name">T</span>
</pre>
<p>你也可以同时选择特定的行和列。例如，以下代码将打印出第3到6行的前6列：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">:</span><span class="literal number integer">6</span><span class="punctuation">,:</span><span class="literal number integer">6</span><span class="punctuation">]</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">6</span> <span class="name">columns</span>
<span class="name">AEGDDP</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
</pre>
<p>使用 <tt class="docutils literal">:</tt> 将打印出所有行：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[:,:</span><span class="literal number integer">6</span><span class="punctuation">]</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">6</span> <span class="name">columns</span>
<span class="name">AEPNAA</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAA</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">DGTSTA</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AEGDDP</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDP</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDP</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>切片给我们提供了一个简单的方式来去除一部分序列比对。在以下例子中，有三条序列的7，8，9三列为间隔（-）。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[:,</span><span class="literal number integer">6</span><span class="punctuation">:</span><span class="literal number integer">9</span><span class="punctuation">]</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">3</span> <span class="name">columns</span>
<span class="name">TNY</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">TNY</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">TSY</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="operator">---</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="operator">---</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="operator">---</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">TSQ</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>你也可以通过切片来获得第9列以后的所有序列：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">alignment</span><span class="punctuation">[:,</span><span class="literal number integer">9</span><span class="punctuation">:]</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">43</span> <span class="name">columns</span>
<span class="name">ATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">ATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">ATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>现在，你可以通过列来操纵序列比对。这也是你能够去除序列比对中的许多列。例如：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">edited</span> <span class="operator">=</span> <span class="name">alignment</span><span class="punctuation">[:,:</span><span class="literal number integer">6</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="name">alignment</span><span class="punctuation">[:,</span><span class="literal number integer">9</span><span class="punctuation">:]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">edited</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">49</span> <span class="name">columns</span>
<span class="name">AEPNAAATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEPNAAATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">DGTSTAATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">AEGDDPAKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDPAKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEGDDPAKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">FAADDAAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
</pre>
<p>另一个经常使用的序列比对操作是将多个基因的序列比对拼接成一个大的序列比对（meta-alignment）。
在进行这种操作时一定要注意序列的ID需要匹配（具体请见 <a class="reference external" href="#sec:SeqRecord-addition">4.7</a> 关于 <tt class="docutils literal">SeqRecord</tt>
的说明)。为了达到这种目的，用 <tt class="docutils literal">sort()</tt> 方法将序列ID按照字母顺序进行排列可能会有所帮助：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">edited</span><span class="operator">.</span><span class="name">sort</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">edited</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">49</span> <span class="name">columns</span>
<span class="name">DGTSTAATEAMNSLKTQATDLIDQTWPVVTSVAVAGLAIRLFKKFSSKA</span> <span class="name">COATB_BPI22</span><span class="operator">/</span><span class="literal number integer">32</span><span class="operator">-</span><span class="literal number integer">83</span>
<span class="name">FAADDAAKAAFDSLTAQATEMSGYAWALVVLVVGATVGIKLFKKFVSRA</span> <span class="name">COATB_BPIF1</span><span class="operator">/</span><span class="literal number integer">22</span><span class="operator">-</span><span class="literal number integer">73</span>
<span class="name">AEPNAAATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIRLFKKFSSKA</span> <span class="name">COATB_BPIKE</span><span class="operator">/</span><span class="literal number integer">30</span><span class="operator">-</span><span class="literal number integer">81</span>
<span class="name">AEGDDPAKAAFNSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="name">COATB_BPM13</span><span class="operator">/</span><span class="literal number integer">24</span><span class="operator">-</span><span class="literal number integer">72</span>
<span class="name">AEGDDPAKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFASKA</span> <span class="name">COATB_BPZJ2</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
<span class="name">AEPNAAATEAMDSLKTQAIDLISQTWPVVTTVVVAGLVIKLFKKFVSRA</span> <span class="name">Q9T0Q8_BPIKE</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">52</span>
<span class="name">AEGDDPAKAAFDSLQASATEYIGYAWAMVVVIVGATIGIKLFKKFTSKA</span> <span class="name">Q9T0Q9_BPFD</span><span class="operator">/</span><span class="literal number integer">1</span><span class="operator">-</span><span class="literal number integer">49</span>
</pre>
<p>注意：只有当两个序列比对拥有相同的行的时候才能进行序列比对的拼接。</p>
</div>
<div class="section" id="id62">
<h3>6.3.2  序列比对作为数组</h3>
<p>根据你的需要，有时将序列比对转换为字符数组是非常方便的。你可以用 <tt class="docutils literal">Numpy</tt> 来实现这一目的：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">numpy</span> <span class="keyword namespace">as</span> <span class="name namespace">np</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">alignment</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;PF05371_seed.sth&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;stockholm&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align_array</span> <span class="operator">=</span> <span class="name">np</span><span class="operator">.</span><span class="name">array</span><span class="punctuation">([</span><span class="name builtin">list</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">],</span> <span class="name">np</span><span class="operator">.</span><span class="name">character</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align_array</span><span class="operator">.</span><span class="name">shape</span>
<span class="punctuation">(</span><span class="literal number integer">7</span><span class="punctuation">,</span> <span class="literal number integer">52</span><span class="punctuation">)</span>
</pre>
<p>如果你需要频繁地使用列操作，你可以让 <tt class="docutils literal">Numpy</tt> 将序列比对以列的形式进行储存（与Fortran一样），而不是 <tt class="docutils literal">Numpy</tt> 默认形式（与C一样以行储存）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align_array</span> <span class="operator">=</span> <span class="name">np</span><span class="operator">.</span><span class="name">array</span><span class="punctuation">([</span><span class="name builtin">list</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="punctuation">],</span> <span class="name">np</span><span class="operator">.</span><span class="name">character</span><span class="punctuation">,</span> <span class="name">order</span><span class="operator">=</span><span class="literal string">&quot;F&quot;</span><span class="punctuation">)</span>
</pre>
<p>注意， <tt class="docutils literal">Numpy</tt> 的数组和Biopython默认的序列比对对象是分别储存在内存中的，编辑其中的一个不会更新另一个的值。</p>
</div>
</div>
<div class="section" id="id63">
<h2>6.4  构建序列比对的工具</h2>
<p>目前有非常多的算法来帮助你构建一个序列比对，包括两两间的比对和多序列比对。这些算法在计算上往往是非常慢的，你一定不会希望用Python来实现他们。然而，你可以使用Biopython来运行命令行程序。通常你需要：</p>
<ol class="arabic simple">
<li>准备一个包含未比对序列的输入文件，一般为FASTA格式的序列。你可以使用 <tt class="docutils literal">Bio.SeqIO</tt> 来创建一个 (具体见第5章 &lt;#chapter:Bio.SeqIO&gt;`__).</li>
<li>在Biopython中运行一个命令行程序来构建序列比对（我们将在这里详细介绍）。这需要通过Biopython的打包程序（wrapper）来实现。</li>
<li>读取以上程序的输出，也就是排列好的序列比对。这往往可以通过 <tt class="docutils literal">Bio.AlignIO</tt> 来实现（请看本章前部分内容）。</li>
</ol>
<p>本章所介绍的所有的命令行打包程序都将以同样的方式使用。你创造一个命令行对象来指定各种参数（例如：输入文件名，输出文件名等），然后通过Python的系统命令模块来运行这一程序（例如：使用 <tt class="docutils literal">subprocess</tt> 进程）。</p>
<p>大多数的打包程序都在 <tt class="docutils literal">Bio.Align.Applications</tt> 中定义：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">Bio.Align.Applications</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">dir</span><span class="punctuation">(</span><span class="name">Bio</span><span class="operator">.</span><span class="name">Align</span><span class="operator">.</span><span class="name">Applications</span><span class="punctuation">)</span>
<span class="operator">...</span>
<span class="punctuation">[</span><span class="literal string">'ClustalwCommandline'</span><span class="punctuation">,</span> <span class="literal string">'DialignCommandline'</span><span class="punctuation">,</span> <span class="literal string">'MafftCommandline'</span><span class="punctuation">,</span> <span class="literal string">'MuscleCommandline'</span><span class="punctuation">,</span>
<span class="literal string">'PrankCommandline'</span><span class="punctuation">,</span> <span class="literal string">'ProbconsCommandline'</span><span class="punctuation">,</span> <span class="literal string">'TCoffeeCommandline'</span> <span class="operator">...</span><span class="punctuation">]</span>
</pre>
<p>（以下划线开头的记录不是Biopython打包程序，这些变量在Python中有特殊的含义。） <tt class="docutils literal">Bio.Emboss.Applications</tt> 中包含对 <a class="reference external" href="http://emboss.sourceforge.net/">EMBOSS</a> 的打包程序（包括 <tt class="docutils literal">needle</tt> 和 <tt class="docutils literal">water</tt> ）。EMBOSS和PHYLIP的打包程序将在 <a class="reference external" href="#seq:emboss-needle-water">6.4.5</a> 节中详细介绍。在本章中，我们并不打算将所有的序列比对程序都予以介绍，但是Biopython中各种序列比对程序都具有相同的使用方式。</p>
<div class="section" id="clustalw">
<h3>6.4.1  ClustalW</h3>
<p>ClustalW是一个非常流行的进行多序列比对的命令行程序（其还有一个图形化的版本称之为ClustalX）。Biopython的 <tt class="docutils literal">Bio.Align.Applications</tt> 模块包含这一多序列比对程序的打包程序。</p>
<p>我们建议你在Python中使用ClustalW之前在命令行界面下手动使用ClustalW，这样能使你更清楚这一程序的参数。你会发现Biopython打包程序非常严格地遵循实际的命令行API：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">ClustalwCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">ClustalwCommandline</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>作为最简单的一个例子，你仅仅需要一个FASTA格式的序列文件作为输入，例如： <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/opuntia.fasta">opuntia.fasta</a> （你可以在线或者在Biopython/Doc/examples文件夹中找到该序列）。 <cite>opuntia.fasta</cite> 包含着7个prickly-pear的DNA序列（来自仙人掌科）。</p>
<p>ClustalW在默认情况下会产生一个包括所有输入序列的序列比对以及一个由输入序列名字构成的指导树（guide tree）。例如，用上述文件作为输入，ClustalW将会输出 <tt class="docutils literal">opuntia.aln</tt> 和 <tt class="docutils literal">opuntia.dnd</tt> 两个文件：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">ClustalwCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cline</span> <span class="operator">=</span> <span class="name">ClustalwCommandline</span><span class="punctuation">(</span><span class="literal string">&quot;clustalw2&quot;</span><span class="punctuation">,</span> <span class="name">infile</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">cline</span>
<span class="name">clustalw2</span> <span class="operator">-</span><span class="name">infile</span><span class="operator">=</span><span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span>
</pre>
<p>注意这里我们给出的执行文件名是 <tt class="docutils literal">clustalw2</tt> ，这是ClustalW的第二个版本（第一个版本的文件名为 <tt class="docutils literal">clustalw</tt> ）。ClustalW的这两个版本具有相同的参数，并且在功能上也是一致的。</p>
<p>你可能会发现，尽管你安装了ClustalW，以上的命令行却无法正确运行。你可能会得到“command not found”的错误信息（尤其是在Windows上）。这往往是由于ClustalW的运行程序并不在系统的工作目录PATH下（一个包含着运行程序路径的环境变量）。你既可以修改PATH，使其包括ClustalW的运行程序（不同系统需要以不同的方式修改），或者你也可以直接指定程序的绝对路径。例如：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">os</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">ClustalwCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clustalw_exe</span> <span class="operator">=</span> <span class="literal string">r&quot;C:\Program Files\new clustal\clustalw2.exe&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clustalw_cline</span> <span class="operator">=</span> <span class="name">ClustalwCommandline</span><span class="punctuation">(</span><span class="name">clustalw_exe</span><span class="punctuation">,</span> <span class="name">infile</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">)</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">assert</span> <span class="name">os</span><span class="operator">.</span><span class="name">path</span><span class="operator">.</span><span class="name">isfile</span><span class="punctuation">(</span><span class="name">clustalw_exe</span><span class="punctuation">),</span> <span class="literal string">&quot;Clustal W executable missing&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span> <span class="name">stderr</span> <span class="operator">=</span> <span class="name">clustalw_cline</span><span class="punctuation">()</span>
</pre>
<p>注意，Python中 <tt class="docutils literal">\n</tt> 和 <tt class="docutils literal">\t</tt> 会被解析为一个新行和制表空白（tab）。然而，如果你将一个小写的“r”放在字符串的前面，这一字符串就将保留原始状态，而不被解析。这种方式对于指定Windows风格的文件名来说是一种良好的习惯。</p>
<p>Biopython在内部使用较新的 <tt class="docutils literal">subprocess</tt> 模块来实现打包程序，而不是 <tt class="docutils literal">os.system()</tt> 和 <tt class="docutils literal">os.popen*</tt> 。</p>
<p>现在，我们有必要去了解命令行工具是如何工作的。当你使用一个命令行时，它往往会在屏幕上输出一些内容。这一输出可以被保存或重定向。在系统输出中，有两种管道（pipe）来区分不同的输出信息--标准输出（standard output）包含正常的输出内容，标准错误（standard error）显示错误和调试信息。同时，系统也接受标准输入（standard input）。这也是命令行工具如何读取数据文件的。当程序运行结束以后，它往往会返回一个整数。一般返回值为0意味着程序正常结束。</p>
<p>当你使用Biopython打包程序来调用命令行工具的时候，它将会等待程序结束，并检查程序的返回值。如果返回值不为0，Biopython将会提示一个错误信息。Biopython打包程序将会输出两个字符串，标准输出和标准错误。</p>
<p>在ClustalW的例子中，当你使用程序时，所有重要的输出都被保存到输出文件中。所有打印在屏幕上的内容（通过 stdout or stderr）可以被忽略掉（假设它已经成功运行）。</p>
<p>当运行ClustalW的时候，我们所关心的往往是输出的序列比对文件和指导树文件。ClustalW会自动根据输入数据的文件名来命名输出文件。在本例中，输出文件将是 <tt class="docutils literal">opuntia.aln</tt> 。当你成功运行完ClustalW以后，你可以使用 <tt class="docutils literal">Bio.AlignIO</tt> 来读取输出结果：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;opuntia.aln&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">align</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">906</span> <span class="name">columns</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273285</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191659</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273284</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191658</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273287</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191661</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
<span class="name">TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273286</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191660</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273290</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273289</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273291</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191</span>
</pre>
<p>另一个输出文件 <tt class="docutils literal">opuntia.dnd</tt> 中包含有一个newick格式的指导树，你可以使用Biopython中的 <tt class="docutils literal">Bio.Phylo</tt> 来读取它：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Phylo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;opuntia.dnd&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;newick&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_ascii</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
                             <span class="name">_______________</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273291</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191665</span>
  <span class="name">__________________________</span><span class="operator">|</span>
 <span class="operator">|</span>                          <span class="operator">|</span>   <span class="name">______</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273290</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191664</span>
 <span class="operator">|</span>                          <span class="operator">|</span><span class="name">__</span><span class="operator">|</span>
 <span class="operator">|</span>                             <span class="operator">|</span><span class="name">_____</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273289</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191663</span>
 <span class="operator">|</span>
<span class="name">_</span><span class="operator">|</span><span class="name">_________________</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273287</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191661</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191661</span>
 <span class="operator">|</span>
 <span class="operator">|</span><span class="name">__________</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273286</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191660</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191660</span>
 <span class="operator">|</span>
 <span class="operator">|</span>    <span class="name">__</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273285</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191659</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191659</span>
 <span class="operator">|</span><span class="name">___</span><span class="operator">|</span>
     <span class="operator">|</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273284</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191658</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191658</span>
</pre>
<p><a class="reference external" href="#sec:Phylo">13</a>  章中详细介绍了如何使用Biopython对进化树数据进行处理。</p>
</div>
<div class="section" id="muscle">
<h3>6.4.2  MUSCLE</h3>
<p>MUSCLE是另一个较新的序列比对工具，Biopython的 <tt class="docutils literal">Bio.Align.Applications</tt> 中也有针对Muscle的打包程序。与ClustalW一样，我们也建议你先在命令行界面下使用MUSCLE以后再使用Biopython打包程序。你会发现，Biopython的打包程序非常严格地包括了所有命令行输入参数：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">MuscleCommandline</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>作为最简单的例子，你只需要一个Fasta格式的数据文件作为输入。例如： <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/opuntia.fasta">opuntia.fasta</a> 然后你可以告诉MUSCLE来读取该FASTA文件，并将序列比对写出：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">,</span> <span class="name">out</span><span class="operator">=</span><span class="literal string">&quot;opuntia.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">cline</span>
<span class="name">muscle</span> <span class="operator">-</span><span class="operator word">in</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span> <span class="operator">-</span><span class="name">out</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">txt</span>
</pre>
<p>注意，MUSCLE使用“-in”和“-out”来指定输入和输出文件，而在Biopython中，我们使用“input”和“out”作为关键字来指定输入输出。这是由于“in”是Python的一个关键词而被保留。</p>
<p>默认情况下，MUSCLE的输出文件将是包含间隔（gap）的FASTA格式文件。 当你指定 <tt class="docutils literal">format=fasta</tt> 时， <tt class="docutils literal">Bio.AlignIO</tt> 能够读取该FASTA文件。你也可以告诉MUSCLE来输出ClustalW-like的文件结果：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">,</span> <span class="name">out</span><span class="operator">=</span><span class="literal string">&quot;opuntia.aln&quot;</span><span class="punctuation">,</span> <span class="name">clw</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">cline</span>
<span class="name">muscle</span> <span class="operator">-</span><span class="operator word">in</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span> <span class="operator">-</span><span class="name">out</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">aln</span> <span class="operator">-</span><span class="name">clw</span>
</pre>
<p>或者，严格的ClustalW的输出文件（这将输出原始的ClustalW的文件标签）。例如：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">,</span> <span class="name">out</span><span class="operator">=</span><span class="literal string">&quot;opuntia.aln&quot;</span><span class="punctuation">,</span> <span class="name">clwstrict</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">cline</span>
<span class="name">muscle</span> <span class="operator">-</span><span class="operator word">in</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span> <span class="operator">-</span><span class="name">out</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">aln</span> <span class="operator">-</span><span class="name">clwstrict</span>
</pre>
<p>你可以使用 <tt class="docutils literal">Bio.AlignIO</tt> 的 <tt class="docutils literal"><span class="pre">format=&quot;clustal&quot;</span></tt> 参数来读取这些序列比对输出。</p>
<p>MUSCLE也可以处理GCG和MSF（使用 <tt class="docutils literal">msf</tt> 参数）甚至HTML格式，但是目前Biopython并不能读取它们。</p>
<p>你也可以设置MUSCLE其它的可选参数，例如最大数目的迭代数。具体信息请查阅Biopython的内部帮助文档。</p>
</div>
<div class="section" id="id64">
<h3>6.4.3  MUSCLE标准输出</h3>
<p>使用以上的MUSCLE命令行将会把序列比对结果写出到一个文件中。然而MUSCLE也允许你将序列比对结果作为系统的标准输出。Biopython打包程序可以利用这一特性来避免创建一个临时文件。例如：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">muscle_cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">muscle_cline</span>
<span class="name">muscle</span> <span class="operator">-</span><span class="operator word">in</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span>
</pre>
<p>如果你使用打包程序运行上述命令，程序将返回一个字符串对象。为了读取它，我们可以使用 <tt class="docutils literal">StringIO</tt> 模块。记住MUSCLE将默认以FASTA格式输出序列比对：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">muscle_cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span> <span class="name">stderr</span> <span class="operator">=</span> <span class="name">muscle_cline</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">StringIO</span> <span class="keyword namespace">import</span> <span class="name">StringIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">StringIO</span><span class="punctuation">(</span><span class="name">stdout</span><span class="punctuation">),</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">align</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">906</span> <span class="name">columns</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273289</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191663</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273291</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191665</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273290</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191664</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273287</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191661</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191661</span>
<span class="name">TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273286</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191660</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191660</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273285</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191659</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191659</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273284</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191658</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191658</span>
</pre>
<p>以上是一个非常简单的例子，如果你希望处理较大的输出数据，我们并不建议你将它们全部读入内存中。对于这种情况， <tt class="docutils literal">subprocess</tt> 模块可以非常方便地处理。例如：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">subprocess</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">muscle_cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">child</span> <span class="operator">=</span> <span class="name">subprocess</span><span class="operator">.</span><span class="name">Popen</span><span class="punctuation">(</span><span class="name builtin">str</span><span class="punctuation">(</span><span class="name">muscle_cline</span><span class="punctuation">),</span>
<span class="operator">...</span>                          <span class="name">stdout</span><span class="operator">=</span><span class="name">subprocess</span><span class="operator">.</span><span class="name">PIPE</span><span class="punctuation">,</span>
<span class="operator">...</span>                          <span class="name">stderr</span><span class="operator">=</span><span class="name">subprocess</span><span class="operator">.</span><span class="name">PIPE</span><span class="punctuation">,</span>
<span class="operator">...</span>                          <span class="name">shell</span><span class="operator">=</span><span class="punctuation">(</span><span class="name">sys</span><span class="operator">.</span><span class="name">platform</span><span class="operator">!=</span><span class="literal string">&quot;win32&quot;</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">child</span><span class="operator">.</span><span class="name">stdout</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">align</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">7</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">906</span> <span class="name">columns</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273289</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191663</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273291</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191665</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191665</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273290</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191664</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273287</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191661</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191661</span>
<span class="name">TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273286</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191660</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191660</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273285</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191659</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191659</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273284</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191658</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF191658</span>
</pre>
</div>
<div class="section" id="id65">
<h3>6.4.4  以标准输入和标准输出使用MUSCLE</h3>
<p>事实上，我们并不需要将序列放在一个文件里来使用MUSCLE。MUSCLE可以读取系统标准输入的内容。注意，这有一点高级和繁琐，若非必须，你可以不用关心这个技术。</p>
<p>为了让MUSCLE读取标准输入的内容，我们首先需要将未排列的序列以 <tt class="docutils literal">SeqRecord</tt> 对象的形式读入到内存。在这里，我们将以一个规则来选择特定的序列（序列长度小于900bp的），使用生成器表达式。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">r</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="operator">&lt;</span> <span class="literal number integer">900</span><span class="punctuation">)</span>
</pre>
<p>随后，我们需要建立一个MUSCLE命令行，但是不指定输入和输出（MUSCLE默认为标准输入和标准输出）。这里，我们将指定输出格式为严格的Clustal格式：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align.Applications</span> <span class="keyword namespace">import</span> <span class="name">MuscleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">muscle_cline</span> <span class="operator">=</span> <span class="name">MuscleCommandline</span><span class="punctuation">(</span><span class="name">clwstrict</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">muscle_cline</span>
<span class="name">muscle</span> <span class="operator">-</span><span class="name">clwstrict</span>
</pre>
<p>我们使用Python的内置模块 <tt class="docutils literal">subprocess</tt> 来实现这一目的：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">subprocess</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">sys</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">child</span> <span class="operator">=</span> <span class="name">subprocess</span><span class="operator">.</span><span class="name">Popen</span><span class="punctuation">(</span><span class="name builtin">str</span><span class="punctuation">(</span><span class="name">cline</span><span class="punctuation">),</span>
<span class="operator">...</span>                          <span class="name">stdin</span><span class="operator">=</span><span class="name">subprocess</span><span class="operator">.</span><span class="name">PIPE</span><span class="punctuation">,</span>
<span class="operator">...</span>                          <span class="name">stdout</span><span class="operator">=</span><span class="name">subprocess</span><span class="operator">.</span><span class="name">PIPE</span><span class="punctuation">,</span>
<span class="operator">...</span>                          <span class="name">stderr</span><span class="operator">=</span><span class="name">subprocess</span><span class="operator">.</span><span class="name">PIPE</span><span class="punctuation">,</span>
<span class="operator">...</span>                          <span class="name">shell</span><span class="operator">=</span><span class="punctuation">(</span><span class="name">sys</span><span class="operator">.</span><span class="name">platform</span><span class="operator">!=</span><span class="literal string">&quot;win32&quot;</span><span class="punctuation">))</span>
</pre>
<p>这一命令将启动MUSCLE，但是它将会等待FASTA格式的输入数据。我们可以通过标准输入句柄来提供给它：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">child</span><span class="operator">.</span><span class="name">stdin</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">6</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">child</span><span class="operator">.</span><span class="name">stdin</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>在将6条序列写入句柄后，MUSCLE仍将会等待，判断是否所有的FASTA序列全部输入完毕了。我们可以关闭句柄来提示给MUSCLE。这时，MUSCLE将开始运行。最后，我们可以在标准输出中获得结果：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">child</span><span class="operator">.</span><span class="name">stdout</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">align</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">6</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">900</span> <span class="name">columns</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273290</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273289</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273287</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191661</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273286</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191660</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273285</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191659</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19165</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273284</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191658</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19165</span>
</pre>
<p>现在我们在没有创造一个FASTA文件的情况下获得了一个序列比对。然而，由于你没有在Biopython外运行MUSCLE，这会使调试程序的难度增大，而且存在程序跨平台使用的问题（Windows和Linux）。</p>
<p>如果你觉得 <tt class="docutils literal">subprocess</tt> 不方便使用，Biopython提供了另一种方式。如果你用 <tt class="docutils literal">muscle_cline()</tt> 来运行外部程序（如MUSCLE），你可以用一个字符串对象作为输入。例如，你可以以这种方式使用： <tt class="docutils literal"><span class="pre">muscle_cline(stdin=...)</span></tt> 。假如你的序列文件不大，你可以将其储存为 <tt class="docutils literal">StringIO</tt> 对象（具体见 <a class="reference external" href="#sec:appendix-handles">22.1</a>)：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">r</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">r</span><span class="punctuation">)</span> <span class="operator">&lt;</span> <span class="literal number integer">900</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">StringIO</span> <span class="keyword namespace">import</span> <span class="name">StringIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">StringIO</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">6</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">data</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">getvalue</span><span class="punctuation">()</span>
</pre>
<p>你可以以下方式运行外部程序和读取结果：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span> <span class="name">stderr</span> <span class="operator">=</span> <span class="name">muscle_cline</span><span class="punctuation">(</span><span class="name">stdin</span><span class="operator">=</span><span class="name">data</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">StringIO</span><span class="punctuation">(</span><span class="name">stdout</span><span class="punctuation">),</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">align</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">6</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">900</span> <span class="name">columns</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273290</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191664</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273289</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191663</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273287</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191661</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273286</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191660</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19166</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273285</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191659</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19165</span>
<span class="name">TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG</span><span class="operator">...</span><span class="name">AGA</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">6273284</span><span class="operator">|</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF191658</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF19165</span>
</pre>
<p>你可能觉得这种方式更便捷，但它需要更多的内存（这是由于我们是以字符串对象来储存输入的FASTA文件和输出的Clustal排列）。</p>
</div>
<div class="section" id="embossneedlewater">
<h3>6.4.5  EMBOSS包的序列比对工具——needle和water</h3>
<p><a class="reference external" href="http://emboss.sourceforge.net/">EMBOSS</a> 包有两个序列比对程序—— <tt class="docutils literal">water</tt> 和 <tt class="docutils literal">needle</tt> 来实现Smith-Waterman做局部序列比对（local alignment）和Needleman-Wunsch算法来做全局排列（global alignment）。这两个程序具有相同的使用方式，因此我们仅以 <tt class="docutils literal">needle</tt> 为例。</p>
<p>假设你希望做全局的序列两两排列，你可以将FASTA格式序列以如下方式储存：</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">HBA_HUMAN</span>
<span class="name">MVLSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTYFPHFDLSHGSAQVKGHG</span>
<span class="name">KKVADALTNAVAHVDDMPNALSALSDLHAHKLRVDPVNFKLLSHCLLVTLAAHLPAEFTP</span>
<span class="name">AVHASLDKFLASVSTVLTSKYR</span>
</pre>
<p>以上内容在 <tt class="docutils literal">alpha.fasta</tt> 文件中，另一个在 <tt class="docutils literal">beta.fasta</tt> 中如下：</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">HBB_HUMAN</span>
<span class="name">MVHLTPEEKSAVTALWGKVNVDEVGGEALGRLLVVYPWTQRFFESFGDLSTPDAVMGNPK</span>
<span class="name">VKAHGKKVLGAFSDGLAHLDNLKGTFATLSELHCDKLHVDPENFRLLGNVLVCVLAHHFG</span>
<span class="name">KEFTPPVQAAYQKVVAGVANALAHKYH</span>
</pre>
<p>让我们开始使用一个完整的 <tt class="docutils literal">needle</tt> 命令行对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Emboss.Applications</span> <span class="keyword namespace">import</span> <span class="name">NeedleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span> <span class="operator">=</span> <span class="name">NeedleCommandline</span><span class="punctuation">(</span><span class="name">asequence</span><span class="operator">=</span><span class="literal string">&quot;alpha.faa&quot;</span><span class="punctuation">,</span> <span class="name">bsequence</span><span class="operator">=</span><span class="literal string">&quot;beta.faa&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                  <span class="name">gapopen</span><span class="operator">=</span><span class="literal number integer">10</span><span class="punctuation">,</span> <span class="name">gapextend</span><span class="operator">=</span><span class="literal number float">0.5</span><span class="punctuation">,</span> <span class="name">outfile</span><span class="operator">=</span><span class="literal string">&quot;needle.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">needle_cline</span>
<span class="name">needle</span> <span class="operator">-</span><span class="name">outfile</span><span class="operator">=</span><span class="name">needle</span><span class="operator">.</span><span class="name">txt</span> <span class="operator">-</span><span class="name">asequence</span><span class="operator">=</span><span class="name">alpha</span><span class="operator">.</span><span class="name">faa</span> <span class="operator">-</span><span class="name">bsequence</span><span class="operator">=</span><span class="name">beta</span><span class="operator">.</span><span class="name">faa</span> <span class="operator">-</span><span class="name">gapopen</span><span class="operator">=</span><span class="literal number integer">10</span> <span class="operator">-</span><span class="name">gapextend</span><span class="operator">=</span><span class="literal number float">0.5</span>
</pre>
<p>你可能会有疑问，为什么不直接在终端里运行这一程序呢？你会发现，它将进行一个序列两两间的排列，并把结果记录在 <tt class="docutils literal">needle.txt</tt> 中（以EMBOSS默认的序列比对格式）。</p>
<p>即使你安装了EMBOSS，使用以上命令仍可能会出错，你可能获得一个错误消息“command not found”，尤其是在Windows环境中。这很可能是由于EMBOSS工具的安装目录并不在系统的PATH中。遇到这种情况，你既可以更新系统的环境变量，也可以在Biopython中指定EMBOSS的安装路径。例如：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Emboss.Applications</span> <span class="keyword namespace">import</span> <span class="name">NeedleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span> <span class="operator">=</span> <span class="name">NeedleCommandline</span><span class="punctuation">(</span><span class="literal string">r&quot;C:\EMBOSS\needle.exe&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                  <span class="name">asequence</span><span class="operator">=</span><span class="literal string">&quot;alpha.faa&quot;</span><span class="punctuation">,</span> <span class="name">bsequence</span><span class="operator">=</span><span class="literal string">&quot;beta.faa&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                  <span class="name">gapopen</span><span class="operator">=</span><span class="literal number integer">10</span><span class="punctuation">,</span> <span class="name">gapextend</span><span class="operator">=</span><span class="literal number float">0.5</span><span class="punctuation">,</span> <span class="name">outfile</span><span class="operator">=</span><span class="literal string">&quot;needle.txt&quot;</span><span class="punctuation">)</span>
</pre>
<p>在Python中， <tt class="docutils literal">\n</tt> 和 <tt class="docutils literal">\t</tt> 分别意味着换行符和制表符。而在字符串前有一个“r”代表着raw字符串（ <tt class="docutils literal">\n</tt> 和 <tt class="docutils literal">\t</tt> 将保持它们本来的状态）。</p>
<p>现在你可以自己尝试着手动运行EMBOSS工具箱中的程序，比较一下各个参数以及其对应的Biopython打包程序帮助文档：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Emboss.Applications</span> <span class="keyword namespace">import</span> <span class="name">NeedleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">NeedleCommandline</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>提示：你也可以指定特定的参数设置。例如：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Emboss.Applications</span> <span class="keyword namespace">import</span> <span class="name">NeedleCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span> <span class="operator">=</span> <span class="name">NeedleCommandline</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span><span class="operator">.</span><span class="name">asequence</span><span class="operator">=</span><span class="literal string">&quot;alpha.faa&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span><span class="operator">.</span><span class="name">bsequence</span><span class="operator">=</span><span class="literal string">&quot;beta.faa&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span><span class="operator">.</span><span class="name">gapopen</span><span class="operator">=</span><span class="literal number integer">10</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span><span class="operator">.</span><span class="name">gapextend</span><span class="operator">=</span><span class="literal number float">0.5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">needle_cline</span><span class="operator">.</span><span class="name">outfile</span><span class="operator">=</span><span class="literal string">&quot;needle.txt&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">needle_cline</span>
<span class="name">needle</span> <span class="operator">-</span><span class="name">outfile</span><span class="operator">=</span><span class="name">needle</span><span class="operator">.</span><span class="name">txt</span> <span class="operator">-</span><span class="name">asequence</span><span class="operator">=</span><span class="name">alpha</span><span class="operator">.</span><span class="name">faa</span> <span class="operator">-</span><span class="name">bsequence</span><span class="operator">=</span><span class="name">beta</span><span class="operator">.</span><span class="name">faa</span> <span class="operator">-</span><span class="name">gapopen</span><span class="operator">=</span><span class="literal number integer">10</span> <span class="operator">-</span><span class="name">gapextend</span><span class="operator">=</span><span class="literal number float">0.5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">needle_cline</span><span class="operator">.</span><span class="name">outfile</span>
<span class="name">needle</span><span class="operator">.</span><span class="name">txt</span>
</pre>
<p>现在我们获得了一个 <tt class="docutils literal">needle</tt> 命令行，并希望在Python中运行它。我们在之前解释过，如果你希望完全地控制这一过程， <tt class="docutils literal">subprocess</tt> 是最好的选择，但是如果你只是想尝试使用打包程序，以下命令足以达到目的：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span> <span class="name">stderr</span> <span class="operator">=</span> <span class="name">needle_cline</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">stdout</span> <span class="operator">+</span> <span class="name">stderr</span>
<span class="name">Needleman</span><span class="operator">-</span><span class="name">Wunsch</span> <span class="keyword">global</span> <span class="name">alignment</span> <span class="name">of</span> <span class="name">two</span> <span class="name">sequences</span>
</pre>
<p>随后，我们需要载入 <tt class="docutils literal">Bio.AlignIO</tt> 模块来读取needle输出（ <tt class="docutils literal">emboss</tt> 格式）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;needle.txt&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;emboss&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">align</span>
<span class="name">SingleLetterAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">2</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">149</span> <span class="name">columns</span>
<span class="name">MV</span><span class="operator">-</span><span class="name">LSPADKTNVKAAWGKVGAHAGEYGAEALERMFLSFPTTKTY</span><span class="operator">...</span><span class="name">KYR</span> <span class="name">HBA_HUMAN</span>
<span class="name">MVHLTPEEKSAVTALWGKV</span><span class="operator">--</span><span class="name">NVDEVGGEALGRLLVVYPWTQRF</span><span class="operator">...</span><span class="name">KYH</span> <span class="name">HBB_HUMAN</span>
</pre>
<p>在这个例子中，我们让EMBOSS将结果保存到一个输出文件中，但是你也可以让其写入标准输出中（这往往是在不需要临时文件的情况下的选择，你可以使用 <tt class="docutils literal">stdout=True</tt> 参数而不是 <tt class="docutils literal">outfile</tt> 参数）。与MUSCLE的例子一样，你也可以从标准输入里读取序列（ <tt class="docutils literal"><span class="pre">asequence=&quot;stdin&quot;</span></tt> 参数）。</p>
<p>以上例子仅仅介绍了 <tt class="docutils literal">needle</tt> 和 <tt class="docutils literal">water</tt> 最简单的使用。一个有用的小技巧是，第二个序列文件可以包含有多个序列，EMBOSS工具将将每一个序列与第一个文件进行两两序列比对。</p>
<p>注意，Biopython有它自己的两两比对模块 <tt class="docutils literal">Bio.pairwise2</tt> （用C语言编写）。但是它无法与序列比对对象一起工作，因此我们不在本章讨论它。具体信息请查阅模块的docstring（内部帮助文档）。</p>
</div>
</div>
</div>
<div class="section" id="blast">
<h1>第7章  BLAST</h1>
<p>嗨，每个人都喜欢BLAST，对吧？我是指，通过BLAST把你的序列和世界上已知的序列
比较是多么简单方便啊。不过，这章当然不是讲Blast有多么酷，因为我们都已经
知道了。这章是来解决使用Blast的一些麻烦地方——处理大量的BLAST比对结果数据
通常是困难的，还有怎么自动运行BLAST序列比对。</p>
<p>幸运的是，Biopython社区的人早就了解了这些难处。所以，他们已经发展了很多
工具来简化BlAST使用和结果处理。这章会具体讲解怎么用这些工具来做些有用的
事情。</p>
<p>使用BLAST通常可以分成2个步。这两步都可以用上Biopython。第一步，提交你的查询
序列，运行BLAST，并得到输出数据。第二步，用Python解析BLAST的输出，并作进一步
分析。</p>
<p>你第一次接触并运行BLAST也许就是通过NCBI的web服务。事实上，你可以通过多种方式
(这些方式可以分成几类)来使用BLAST。这些方式最重要的区别在于你是在你的自己
电脑上运行一个本地BLAST，还是在远程服务器（另外一台电脑，通常是NCBI的服务器）上运行。
我们将在一个Python脚本里调用NCBI在线BLAST服务来开始这章的内容。</p>
<p><em>注意</em>: 接下来的 <a class="reference external" href="#chapter:searchio">第8章</a> 介绍的 <tt class="docutils literal">Bio.SearchIO</tt> 是一个
Biopython实验性质的模块。我们准备最终用它来替换原来的 <tt class="docutils literal">Bio.Blast</tt> 模块。
因为它提供了一个更为通用的序列搜索相关的框架。不过，在这个模块的稳定版本发布之前，在实际工作中的代码里，请继续用 <tt class="docutils literal">Bio.Blast</tt>
模块来处理NCBI BLAST。</p>
<div class="section" id="internetblast">
<h2>7.1  通过Internet运行BLAST</h2>
<p>我们使用 <tt class="docutils literal">Bio.Blast.NCBIWWW</tt> 模块的函数 <tt class="docutils literal">qblast()</tt> 来调用在线版本的BLAST。
这个函数有3个必需的参数:</p>
<ul class="simple">
<li>第一个参数是用来搜索的blast程序，这是小写的字符串。对这个参数的选项和描述可以在
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/BLAST/blast_program.shtml">http://www.ncbi.nlm.nih.gov/BLAST/blast_program.shtml</a>.
查到。目前 <tt class="docutils literal">qblast</tt> 只支持 blastn, blastp, blastx, tblast 和 tblastx.</li>
<li>第二个参数指定了将要搜索的数据库。同样地，这个参数的选项也可以在
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/BLAST/blast_databases.shtml">http://www.ncbi.nlm.nih.gov/BLAST/blast_databases.shtml</a>.
查到</li>
<li>第三个参数是一个包含你要查询序列的字符串。这个字符串可以是序列的本身
（fasta格式的），或者是一个类似GI的id。</li>
</ul>
<p><tt class="docutils literal">qblast</tt> 函数还可以接受许多其他选项和参数。这些参数基本上类似于你在BLAST网站页面
能够设置的参数。在这里我们只强调其中的一些：</p>
<ul class="simple">
<li><tt class="docutils literal">qblast</tt> 函数可以返回多种格式的BLAST结果。你可以通过可选参数 <tt class="docutils literal">format_type</tt>
指定格式关键字为：<tt class="docutils literal">&quot;HTML&quot;</tt>, <tt class="docutils literal">&quot;Text&quot;</tt>, <tt class="docutils literal">&quot;ASN.1&quot;</tt>, 或 <tt class="docutils literal">&quot;XML&quot;</tt> 。默认
格式是 <tt class="docutils literal">&quot;XML&quot;</tt> ，这是解析器期望的格式，<a class="reference external" href="#sec:parsing-blast">7.3节</a> 对其有详细的描述。</li>
<li>参数 <tt class="docutils literal">expect</tt>  指定期望值，即阀值 e-value。</li>
</ul>
<p>更多可选的BLAST参数，请参照NCBI的文档，或者是Biopython内置的文档。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIWWW</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">NCBIWWW</span><span class="operator">.</span><span class="name">qblast</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>请注意，NCBI BLAST 网站上的默认参数和QBLAST的默认参数不完全相同。如果你得到了
不同的结果，你就需要检查下参数设置 （比如，e-value阈值和gap值）.</p>
<p>举个例子，如果你有条核酸序列，想使用BLAST对核酸数据库（nt）进行搜索，已知这条查询序列的GI号，
你可以这样做：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIWWW</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name">NCBIWWW</span><span class="operator">.</span><span class="name">qblast</span><span class="punctuation">(</span><span class="literal string">&quot;blastn&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;nt&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;8332116&quot;</span><span class="punctuation">)</span>
</pre>
<p>或者，我们想要查询的序列在FASTA文件中，那么我们只需打开这个文件并把这条记录读入到字符串，然后用这个字符串作为查询参数:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIWWW</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">fasta_string</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta&quot;</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name">NCBIWWW</span><span class="operator">.</span><span class="name">qblast</span><span class="punctuation">(</span><span class="literal string">&quot;blastn&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;nt&quot;</span><span class="punctuation">,</span> <span class="name">fasta_string</span><span class="punctuation">)</span>
</pre>
<p>我们同样可以读取FASTA文件为一个 <tt class="docutils literal">SeqRecord</tt> 序列对象，然后以这个序列自身作为参数：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIWWW</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta&quot;</span><span class="punctuation">,</span> <span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name">NCBIWWW</span><span class="operator">.</span><span class="name">qblast</span><span class="punctuation">(</span><span class="literal string">&quot;blastn&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;nt&quot;</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
</pre>
<p>只提供序列意味着BLAST会自动分配给你一个ID。你可能更喜欢用 <tt class="docutils literal">SeqRecord</tt>
对象的format方法来包装一个fasta字符串，因为这个对象会包含fasta文件中已有的ID</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIWWW</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta&quot;</span><span class="punctuation">,</span> <span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name">NCBIWWW</span><span class="operator">.</span><span class="name">qblast</span><span class="punctuation">(</span><span class="literal string">&quot;blastn&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;nt&quot;</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
</pre>
<p>如果你的序列在一个非FASTA格式的文件中并且你用 <tt class="docutils literal">Bio.SeqIO</tt> (看第5章`5 &lt;#chapter:Bio.SeqIO&gt;`__)
把序列取出来了，那么这个方法更有用。</p>
<p>不论你给 <tt class="docutils literal">qblast()</tt> 函数提供了什么参数，都应该返回一个handle object的结果(
默认是XML格式)。下一步就是将这个XML输出解析为代表BLAST搜索结果的Python
对象（<a class="reference external" href="#sec:parsing-blast">7.3</a>）。
不过，也许你想先把这个XML输出保存一个本地文件副本。当调试从BLAST结果提取信息的代码的时候，我发现这样做
尤其有用。(因为重新运行在线BLAST搜索很慢并且会浪费NCBI服务器的运行时间)。</p>
<p>这里我们需要注意下：因为用 <tt class="docutils literal">result_handle.read()</tt> 来读取BLAST结果只能用一次 -
再次调用 <tt class="docutils literal">result_handle.read()</tt> 会返回一个空的字符串.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">save_file</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;my_blast.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">save_file</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">save_file</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>这些做好后，结果已经存储在 <tt class="docutils literal">my_blast.xml</tt> 文件中了并且原先的handle中的数据
已经被全部提取出来了(所以我们把它关闭了)。但是，BLAST解析器的 <tt class="docutils literal">parse</tt> 函数（描述见<a class="reference external" href="#sec:parsing-blast">7.3</a>)
采用一个文件句柄类的对象，所以我们只需打开已经保存的文件作为输入。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;my_blast.xml&quot;</span><span class="punctuation">)</span>
</pre>
<p>既然现在已经把BLAST的结果又一次读回handle，我们可以分析下这些结果。所以我们正好可以去读
关于结果解析的章节（看下面  <a class="reference external" href="#sec:parsing-blast">7.3</a> ）。你现在也许想跳过去看吧 ...</p>
</div>
<div class="section" id="id66">
<h2>7.2  本地运行BLAST</h2>
<div class="section" id="id67">
<h3>7.2.1  介绍</h3>
<p>在本地运行BLAST（跟通过internet运行比，见 <a class="reference external" href="#sec:running-www-blast">7.1</a> ）
至少有2个主要优点：</p>
<ul class="simple">
<li>本地运行BLAST可能比通过internet运行更快；</li>
<li>本地运行可以让你建立自己的数据库来对序列进行搜索。</li>
</ul>
<p>处理有版权的或者没有发表的序列数据也许是本地运行BLAST的另一个原因。你也许
不能泄露这些序列数据，所以没法提交给NCBI来BLAST。</p>
<p>不幸的是，本地运行也有些缺点 - 安装所有的东东并成功运行需要花些力气：</p>
<ul class="simple">
<li>本地运行BLAST需要你安装相关命令行工具。</li>
<li>本地运行BLAST需要安装一个很大的BLAST的数据库（并且需要保持数据更新）.</li>
</ul>
<p>更令人困惑的是，至少有4种不同的BLAST安装程序包，并且还有其他的一些工具能
产生类似的BLAST 输出文件，比如BLAT。</p>
</div>
<div class="section" id="ncbiblast">
<h3>7.2.2  单机版的NCBI老版本BLAST</h3>
<p><a class="reference external" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">NCBI “老版本” BLAST</a>
包括命令行工具 <tt class="docutils literal">blastall</tt> ， <tt class="docutils literal">blastpgp</tt> 和 <tt class="docutils literal">rpsblast</tt> 。
这是NCBI发布它的替代品BLAST+ 前使用最为广泛的单机版BLAST工具。</p>
<p><tt class="docutils literal">Bio.Blast.Applications</tt> 模块有个对老版本NCBI BLAST 工具像 <tt class="docutils literal">blastall</tt> ， <tt class="docutils literal">blastpgp</tt>
和 <tt class="docutils literal">rpsblast</tt> 的封装， 并且在 <tt class="docutils literal">Bio.Blast.NCBIStandalone</tt> 还有个辅助函数。
这些东东现在都被认为是过时的，并且当用户们迁移到BLAST+程序套件后，这些都会被弃用，
最终从Biopython删除。</p>
<p>为了减少你的困惑，我们在这个指南中不会提到怎么从Biopython调用这些老版本的工具。
如果你有兴趣，可以看下在Biopython 1.52中包含的基本指南。（看下 <tt class="docutils literal"><span class="pre">biopython-1.52.tar.gz</span></tt>
或者 <tt class="docutils literal"><span class="pre">biopython-1.52.zip</span></tt> 中Doc目录下的指南的PDF文件 或者 HTML 文件）。</p>
</div>
<div class="section" id="ncbi-blast">
<h3>7.2.3  单机版 NCBI BLAST+</h3>
<p><a class="reference external" href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">NCBI “新版本”的
BLAST+</a>
在2009年发布。它替代了原来老版本的BLAST程序包。<tt class="docutils literal">Bio.Blast.Applications</tt> 模块
包装了这些新工具像 <tt class="docutils literal">blastn</tt> , <tt class="docutils literal">blastp</tt> , <tt class="docutils literal">blastx</tt> , <tt class="docutils literal">tblastn</tt> , <tt class="docutils literal">tblastx</tt>
(这些以前都是由 <tt class="docutils literal">blastall</tt>  处理)。而 <tt class="docutils literal">rpsblast</tt> 和 <tt class="docutils literal">rpstblastn</tt> （替代了
原来的 <tt class="docutils literal">rpsblast</tt> ）。我们这里不包括对 <tt class="docutils literal">makeblastdb</tt> 的包装，它在BLAST+中用于从FASTA文件
建立一个本地BLAST数据库，还有其在老版本BLAST中的等效工具 <tt class="docutils literal">formatdb</tt> 。</p>
<p>这节将简要地介绍怎样在Python中使用这些工具。如果你已经阅读了并试过
<a class="reference external" href="#sec:alignment-tools">6.4</a> 节的序列联配（alignment）工具，下面介绍
的方法应该是很简单直接的。首先，我们构建一个命令行字符串（就像你使用单机版
BLAST的时候，在终端打入命令行一样）。然后，我们在Python中运行这个命令。</p>
<p>举个例子，你有个FASTA格式的核酸序列文件，你想用它通过BLASTX（翻译）来搜索
非冗余（NR）蛋白质数据库。如果你（或者你的系统管理员）下载并安装好了这个数据库，
那么你只要运行：</p>
<pre class="code python literal-block">
<span class="name">blastx</span> <span class="operator">-</span><span class="name">query</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span> <span class="operator">-</span><span class="name">db</span> <span class="name">nr</span> <span class="operator">-</span><span class="name">out</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">xml</span> <span class="operator">-</span><span class="name">evalue</span> <span class="literal number float">0.001</span> <span class="operator">-</span><span class="name">outfmt</span> <span class="literal number integer">5</span>
</pre>
<p>这样就完成了运行BLASTX查找非冗余蛋白质数据库，用0.001的e值并产生XML格式的
输出结果文件（这样我们可以继续下一步解析）。在我的电脑上运行这条命令花了大约6分钟
- 这就是为什么我们需要保存输出到文件。这样我们就可以在需要时重复任何基于这个输出的分析。</p>
<p>在Biopython中，我们可以用NCBI BLASTX包装模块  <tt class="docutils literal">Bio.Blast.Applications</tt> 来构建
命令行字符串并运行它：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast.Applications</span> <span class="keyword namespace">import</span> <span class="name">NcbiblastxCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">NcbiblastxCommandline</span><span class="punctuation">)</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blastx_cline</span> <span class="operator">=</span> <span class="name">NcbiblastxCommandline</span><span class="punctuation">(</span><span class="name">query</span><span class="operator">=</span><span class="literal string">&quot;opuntia.fasta&quot;</span><span class="punctuation">,</span> <span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nr&quot;</span><span class="punctuation">,</span> <span class="name">evalue</span><span class="operator">=</span><span class="literal number float">0.001</span><span class="punctuation">,</span>
<span class="operator">...</span>                                      <span class="name">outfmt</span><span class="operator">=</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="name">out</span><span class="operator">=</span><span class="literal string">&quot;opuntia.xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blastx_cline</span>
<span class="name">NcbiblastxCommandline</span><span class="punctuation">(</span><span class="name">cmd</span><span class="operator">=</span><span class="literal string">'blastx'</span><span class="punctuation">,</span> <span class="name">out</span><span class="operator">=</span><span class="literal string">'opuntia.xml'</span><span class="punctuation">,</span> <span class="name">outfmt</span><span class="operator">=</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="name">query</span><span class="operator">=</span><span class="literal string">'opuntia.fasta'</span><span class="punctuation">,</span>
<span class="name">db</span><span class="operator">=</span><span class="literal string">'nr'</span><span class="punctuation">,</span> <span class="name">evalue</span><span class="operator">=</span><span class="literal number float">0.001</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blastx_cline</span>
<span class="name">blastx</span> <span class="operator">-</span><span class="name">out</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">xml</span> <span class="operator">-</span><span class="name">outfmt</span> <span class="literal number integer">5</span> <span class="operator">-</span><span class="name">query</span> <span class="name">opuntia</span><span class="operator">.</span><span class="name">fasta</span> <span class="operator">-</span><span class="name">db</span> <span class="name">nr</span> <span class="operator">-</span><span class="name">evalue</span> <span class="literal number float">0.001</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span> <span class="name">stderr</span> <span class="operator">=</span> <span class="name">blastx_cline</span><span class="punctuation">()</span>
</pre>
<p>在这个例子中，终端里应该没有任何从BLASTX的输出，所以stdout和stderr是空的。
你可能想要检查下输出文件 <tt class="docutils literal">opuntia.xml</tt> 是否已经创建。</p>
<p>如果你回想下这个指南的中的早先的例子，<tt class="docutils literal">opuntia.fasta</tt> 包含7条序列，
所以BLAST XML 格式的结果输出文件应该包括多个结果。因此，我们在
下面的`7.3 &lt;#sec:parsing-blast&gt;`__ 节将用 <tt class="docutils literal">Bio.Blast.NCBIXML.parse()</tt> 来
解析这个结果文件。</p>
</div>
<div class="section" id="wu-blast-ab-blast">
<h3>7.2.4  WU-BLAST 和 AB-BLAST</h3>
<p>你也许会碰到 <a class="reference external" href="http://blast.wustl.edu/">Washington University BLAST</a> (WU-BLAST)，
和它的后继版本`Advanced Biocomputing BLAST &lt;<a class="reference external" href="http://blast.advbiocomp.com">http://blast.advbiocomp.com</a>&gt;`__ (AB-BLAST,
在2009年发布，免费但是没有开源）。这些程序包包括了命令工具行
<tt class="docutils literal"><span class="pre">wu-blastall</span></tt> 和 <tt class="docutils literal"><span class="pre">ab-blastall</span></tt> 。</p>
<p>Biopython 目前还没有提供调用这些工具的包装程序，但是应该可以解析它们
与NCBI兼容的输出结果。</p>
</div>
</div>
<div class="section" id="id68">
<h2>7.3  解析BLAST 输出</h2>
<p>就像上面提过的那样，BLAST能生成多种格式的输出，比如 XML， HTML 和纯文本格式。
以前，Biopython有针对HTML 和纯文本格式输出文件的解析器，因为当时只有这两种
格式的输出结果文件。不幸的是，这两种方式的BLAST 输出结果一直在变动，而每次
变动就会导致解析器失效。 所以，我们删除了针对HTML格式的解析器，不过纯文本格式
的解析还可以用（见 <a class="reference external" href="#sec:parsing-blast-deprecated">7.5</a> ）。使用这个解析器
有一定的风险，它可能能工作也可能无效，依赖于你正在使用哪个BLAST版本。</p>
<p>跟上BLAST输出文件格式的改变很难，特别是当用户使用不同版本的BLAST的时候。
我们推荐使用XML格式的解析器。因为最近版本的BLAST能生成这种格式的输出结果。
XML格式的输出不仅比HTML 和纯文本格式的更稳定，而且解析起来更加容易自动化，
从而提高整个Biopython整体的稳定性。</p>
<p>你可以通过好几个途径来获得XML格式的BLAST输出文件。对解析器来说，不管你是
怎么生成输出的，只要是输出的格式是XML就行。</p>
<ul class="simple">
<li>你可以通过Biopython来运行因特网上的BLAST，就像 <a class="reference external" href="#sec:running-www-blast">7.1</a>
节描述的那样。</li>
<li>你可以通过Biopython来运行本地的BLAST，就像 <a class="reference external" href="#sec:running-local-blast">7.2</a>
节描述的那样。</li>
<li>你可以在通过浏览器在NCBI网站上进行BLAST搜索，然后保存结果文件。你需要选择输出
结果文件是XML格式的，并保存最终的结果网页（你知道，就是包含所有有趣结果的那个网页）
到文件。</li>
<li>你也可以直接运行本地电脑上的BlAST，不通过Biopython，保存输出结果到文件。
同样的你也需要选择输出文件格式为XML。</li>
</ul>
<p>关键点就是你不必用Biopython脚本来获取数据才能解析它。通过以上任何一种方式
获取了结果输出，你然后需要获得文件句柄来处理它。在Python中，一个文件句柄就是一种
用于描述到任何信息源的输入的良好通用的方式，以便于这些信息能够使用 <tt class="docutils literal">read()</tt> 和 <tt class="docutils literal">readline()</tt>
函数（见章节 sec:appendix-handles）来获取。</p>
<p>如果你一直跟着上几节用来和BLAST交互的代码的话，你已经有了个 <tt class="docutils literal">result_handle</tt>
，一个用来得到BLAST的结果文件句柄。 比如通过GI号来进行一个在线BLAST搜索：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIWWW</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name">NCBIWWW</span><span class="operator">.</span><span class="name">qblast</span><span class="punctuation">(</span><span class="literal string">&quot;blastn&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;nt&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;8332116&quot;</span><span class="punctuation">)</span>
</pre>
<p>如果你通过其他方式运行了BLAST，并且XML格式的BLAST结果输出文件是 <tt class="docutils literal">my_blast.xml</tt> ,
那么你只需要打开文件来读：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;my_blast.xml&quot;</span><span class="punctuation">)</span>
</pre>
<p>好的，现在我们已经有了个文件句柄，可以解析输出结果了。解析结果的代码
很短。如果你想要一条BLAST输出结果（就是说，你只用了一条序列去搜索）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIXML</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">NCBIXML</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">)</span>
</pre>
<p>或者， 你有许多搜索结果（就是说，你用了多条序列去BLAST搜索）</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIXML</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_records</span> <span class="operator">=</span> <span class="name">NCBIXML</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">)</span>
</pre>
<p>就像 <tt class="docutils literal">Bio.SeqIO</tt> 和 <tt class="docutils literal">Bio.AlignIO</tt> (参见 章节 <a class="reference external" href="#chapter:Bio.SeqIO">5</a>
和 <a class="reference external" href="#chapter:Bio.AlignIO">6</a>), 我们有一对输入函数， <tt class="docutils literal">read</tt> 和
<tt class="docutils literal">parse</tt> 。 当你只有一个输出结果的时候用 <tt class="docutils literal">read</tt> 。当你有许多
输出结果的时候，可以用 <tt class="docutils literal">parse</tt> 这个迭代器。 但是，我们调用函数获得结果
不是 <tt class="docutils literal">SeqRecord</tt> 或者 <tt class="docutils literal">MultipleSeqAlignment</tt> 对象，我们得到BLAST记录对象。</p>
<p>为了能处理BLAST结果文件很大有很多结果这种情况， <tt class="docutils literal">NCBIXML.parse()</tt>
返回一个迭代器。简单来说，一个迭代器可以让你一个接着一个地获得BLAST
的搜索结果。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIXML</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_records</span> <span class="operator">=</span> <span class="name">NCBIXML</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="comment"># ... do something with blast_record</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="comment"># ... do something with blast_record</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="comment"># ... do something with blast_record</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
  <span class="name">File</span> <span class="literal string">&quot;&lt;stdin&gt;&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="operator">&lt;</span><span class="name">module</span><span class="operator">&gt;</span>
<span class="name exception">StopIteration</span>
<span class="comment"># No further records</span>
</pre>
<p>或者，你也可以使用 <tt class="docutils literal">for</tt> - 循环</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">blast_record</span> <span class="operator word">in</span> <span class="name">blast_records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="comment"># Do something with blast_record</span>
</pre>
<p>注意对每个BLAST搜索结果只能迭代一次。通常，对于每个BLAST记录，你可能会保存你
感兴趣的信息。如果你想保存所有返回的BLAST记录，你可以把迭代
转换成列表。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_records</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">blast_records</span><span class="punctuation">)</span>
</pre>
<p>现在，你可以像通常的做法通过索引从这个列表中获得每一条BLAST结果。 如果你的BLAST输出
结果文件很大，那么当把它们全部放入一个列表时，你也许会遇到内存不够的情况。</p>
<p>一般来说，你会一次运行一个BLAST搜索。然后，你只需提取第一条BLAST 搜索记录到
<tt class="docutils literal">blast_records</tt> :</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIXML</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_records</span> <span class="operator">=</span> <span class="name">NCBIXML</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
</pre>
<p>or more elegantly:</p>
<p>或者更加优雅地：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIXML</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">NCBIXML</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">)</span>
</pre>
<p>我猜你现在在想BLAST搜索记录中到底有什么。</p>
</div>
<div class="section" id="id69">
<h2>7.4  BLAST 记录类</h2>
<p>一个BLAST搜索结果记录包括了所有你想要从中提取出来的信息。现在，我们将
用一个例子说明你怎么从BLAST搜索结果提取出一些信息。但是，如果你想从BLAST
搜索结果获得的信息没有在这里提到，你可以详细阅读BLAST搜索记录类，
并且可以参考下源代码 或者 是自动生成的文档 - 文档字符串里面包含了许多
关于各部分源代码是什么的很有用的信息。</p>
<p>继续我们的例子，让我们打印出所有大于某一特定阈值的BLAST命中结果的一些汇总信息。
代码如下：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">E_VALUE_THRESH</span> <span class="operator">=</span> <span class="literal number float">0.04</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">alignment</span> <span class="operator word">in</span> <span class="name">blast_record</span><span class="operator">.</span><span class="name">alignments</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">hsp</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">hsps</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">if</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">expect</span> <span class="operator">&lt;</span> <span class="name">E_VALUE_THRESH</span><span class="punctuation">:</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'****Alignment****'</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'sequence:'</span><span class="punctuation">,</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">title</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'length:'</span><span class="punctuation">,</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">length</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'e value:'</span><span class="punctuation">,</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">expect</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">query</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">'...'</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">match</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">'...'</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">sbjct</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">'...'</span>
</pre>
<p>上面代码会打印出如下图的总结报告：</p>
<pre class="code python literal-block">
<span class="operator">****</span><span class="name">Alignment</span><span class="operator">****</span>
<span class="name">sequence</span><span class="punctuation">:</span> <span class="operator">&gt;</span><span class="name">gb</span><span class="operator">|</span><span class="name">AF283004</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span><span class="name">AF283004</span> <span class="name">Arabidopsis</span> <span class="name">thaliana</span> <span class="name">cold</span> <span class="name">acclimation</span> <span class="name">protein</span> <span class="name">WCOR413</span><span class="operator">-</span><span class="name">like</span> <span class="name">protein</span>
<span class="name">alpha</span> <span class="name">form</span> <span class="name">mRNA</span><span class="punctuation">,</span> <span class="name">complete</span> <span class="name">cds</span>
<span class="name">length</span><span class="punctuation">:</span> <span class="literal number integer">783</span>
<span class="name">e</span> <span class="name">value</span><span class="punctuation">:</span> <span class="literal number float">0.034</span>
<span class="name">tacttgttgatattggatcgaacaaactggagaaccaacatgctcacgtcacttttagtcccttacatattcctc</span><span class="operator">...</span>
<span class="operator">|||||||||</span> <span class="operator">|</span> <span class="operator">|||||||||||</span> <span class="operator">||</span> <span class="operator">||||</span>  <span class="operator">||</span> <span class="operator">||</span> <span class="operator">||||||||</span> <span class="operator">||||||</span> <span class="operator">|</span>  <span class="operator">|</span> <span class="operator">||||||||</span> <span class="operator">|||</span> <span class="operator">||...</span>
<span class="name">tacttgttggtgttggatcgaaccaattggaagacgaatatgctcacatcacttctcattccttacatcttcttc</span><span class="operator">...</span>
</pre>
<p>基本上，一旦你解析了BLAST搜索结果文件，你可以提取任何你需要的信息。
当然，这取决于你想要获得什么信息。但是希望这里的例子能够帮助你开始工作。</p>
<p>在用Biopython提取BLAST搜索结果信息的时候，重要的是你需要考虑到信息存储在什么
（Biopython）对象中。在Biopython中，解析器返回 <tt class="docutils literal">Record</tt>  对象，这个对象
可以是 <tt class="docutils literal">Blast</tt> 类型的，也可以是 <tt class="docutils literal">PSIBlast</tt> 类型的，具体哪个取决你
解析什么。这些对象的定义都可以在  <tt class="docutils literal">Bio.Blast.Record</tt> 找到 并且很完整。</p>
<p>下面是 我尝试画的 <tt class="docutils literal">Blast</tt> 和 <tt class="docutils literal">PSIBlast</tt> 记录类的UML图。如果你对UML图很熟悉，不妨
看看下面的UML图是否有错误或者可以改进的地方，如果有，请联系我。
BLAST类图在这里 <a class="reference external" href="#fig:blastrecord">7.4</a> 。</p>
<p><img alt="image1" src="../images/BlastRecord.png" /></p>
<p>PSIBlast 记录类是类似的，但是支持用在迭代器中的rounds方法。PSIBlast类图在这里 <a class="reference external" href="#fig:psiblastrecord">7.4</a> 。</p>
<p><img alt="image2" src="../images/PSIBlastRecord.png" /></p>
</div>
<div class="section" id="id70">
<h2>7.5  废弃的BLAST 解析器</h2>
<p>老版本的Biopython 有针对纯文本和HTML格式输出结果的解析器。但是经过几年
我们发现维护这些解析器很困难。基本上，任何BLAST输出的任何小改变都会导致
这些解析器失效。所以我们推荐你解析XML格式的BLAST输出结果，就像在
<a class="reference external" href="#sec:parsing-blast">7.3</a> 描述的那样。</p>
<p>取决于你使用Biopython的版本，纯文本格式的解析器也许有效也许失效。
用这个解析器的所带来的风险由你自己承担。</p>
<div class="section" id="id71">
<h3>7.5.1  解析纯文本格式的BLAST输出</h3>
<p>纯文本格式的解析器在 <tt class="docutils literal">Bio.Blast.NCBIStandalone</tt> 。</p>
<p>和xml解析器类似， 我们也需要一个能够传给解析器的文件句柄。这个文件句柄必须
实现了 <tt class="docutils literal">readline()</tt> 方法 。通常要获得这样文件句柄，既可以用Biopython提供的
<tt class="docutils literal">blastall</tt> 或 <tt class="docutils literal">blastpgp</tt> 函数来调用本地的BLAST，或者从命令行运行本地的
BLAST， 并且如下处理：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;my_file_of_blast_output.txt&quot;</span><span class="punctuation">)</span>
</pre>
<p>好了，既然现在得到了个文件句柄（我们称它是 <tt class="docutils literal">result_handle</tt> ），
我们已经做好了解析它的准备。按下面的代码来解析：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIStandalone</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_parser</span> <span class="operator">=</span> <span class="name">NCBIStandalone</span><span class="operator">.</span><span class="name">BlastParser</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_parser</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">)</span>
</pre>
<p>这样就能把BALST的搜索结果报告解析到Blast记录类中（取决你于你解析的对象，
解析结果可能返回一条 Blast 或者 PSIBlast记录）。这样你就可以从中提取
信息了。在我们的例子里，我们来打印出大于某个阈值的所有比对的一个总结
信息。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">E_VALUE_THRESH</span> <span class="operator">=</span> <span class="literal number float">0.04</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">alignment</span> <span class="operator word">in</span> <span class="name">blast_record</span><span class="operator">.</span><span class="name">alignments</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">hsp</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">hsps</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">if</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">expect</span> <span class="operator">&lt;</span> <span class="name">E_VALUE_THRESH</span><span class="punctuation">:</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'****Alignment****'</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'sequence:'</span><span class="punctuation">,</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">title</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'length:'</span><span class="punctuation">,</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">length</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="literal string">'e value:'</span><span class="punctuation">,</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">expect</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">query</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">'...'</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">match</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">'...'</span>
<span class="operator">...</span>             <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">sbjct</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="literal string">'...'</span>
</pre>
<p>如果你已经读过 <a class="reference external" href="#sec:parsing-blast">7.3 节</a> 关于解析XML格式的部分，
你将会发现上面的代码和那个章节的是一样的。一旦你把输出文件解析到记录类中，
你就能处理信息，不管你原来的BLAST输出格式是什么。很赞吧。</p>
<p>好，解析一条记录是不错，那么如果我有一个包含许多记录的BLAST文件 -
我该怎么处理它们呢？好吧，不要害怕，答案就在下个章节中。</p>
</div>
<div class="section" id="blastblast">
<h3>7.5.2  解析包含多次BLAST结果的纯文本BLAST文件</h3>
<p>我们可以用BLAST迭代器解析多次结果。为了得到一个迭代器，我们首先需要创建一个解析器，来
解析BLAST的搜索结果报告为Blast记录对象。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIStandalone</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_parser</span> <span class="operator">=</span> <span class="name">NCBIStandalone</span><span class="operator">.</span><span class="name">BlastParser</span><span class="punctuation">()</span>
</pre>
<p>然后，我们假定我们有一个连接到一大堆blast记录的文件句柄，我们把这个文件句柄
叫做  <tt class="docutils literal">result_handle</tt> 。 怎么得到一个文件句柄在上面blast解析章节有详细
描述。</p>
<p>好了，我们现在有了一个解析器和一个文件句柄，我们可以用以下命令来创建
一个迭代器。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_iterator</span> <span class="operator">=</span> <span class="name">NCBIStandalone</span><span class="operator">.</span><span class="name">Iterator</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">,</span> <span class="name">blast_parser</span><span class="punctuation">)</span>
</pre>
<p>第二个参数，解析器，是可选的。如果我们没有提供一个解析器，那么迭代器将会
一次返回一个原始的BLAST搜索结果。</p>
<p>现在我们已经有了个迭代器，就可以开始通过 <tt class="docutils literal">next()</tt> 方法来获取BLAST
记录（由我们的解析器产生）。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_record</span> <span class="operator">=</span> <span class="name">blast_iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
</pre>
<p>每次调用next都会返回一条我们能处理的新记录。现在我们可以遍历所有记录，并打印一
个我们最爱、漂亮的、简洁的BLAST记录报告。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">blast_record</span> <span class="operator word">in</span> <span class="name">blast_iterator</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">E_VALUE_THRESH</span> <span class="operator">=</span> <span class="literal number float">0.04</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">alignment</span> <span class="operator word">in</span> <span class="name">blast_record</span><span class="operator">.</span><span class="name">alignments</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">for</span> <span class="name">hsp</span> <span class="operator word">in</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">hsps</span><span class="punctuation">:</span>
<span class="operator">...</span>             <span class="keyword">if</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">expect</span> <span class="operator">&lt;</span> <span class="name">E_VALUE_THRESH</span><span class="punctuation">:</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="literal string">'****Alignment****'</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="literal string">'sequence:'</span><span class="punctuation">,</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">title</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="literal string">'length:'</span><span class="punctuation">,</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">length</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="literal string">'e value:'</span><span class="punctuation">,</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">expect</span>
<span class="operator">...</span>                 <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">hsp</span><span class="operator">.</span><span class="name">query</span><span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="literal number integer">75</span><span class="punctuation">:</span>
<span class="operator">...</span>                     <span class="name">dots</span> <span class="operator">=</span> <span class="literal string">'...'</span>
<span class="operator">...</span>                 <span class="keyword">else</span><span class="punctuation">:</span>
<span class="operator">...</span>                     <span class="name">dots</span> <span class="operator">=</span> <span class="literal string">''</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">query</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="name">dots</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">match</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="name">dots</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="name">hsp</span><span class="operator">.</span><span class="name">sbjct</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">75</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="name">dots</span>
</pre>
<p>迭代器允许你处理很多blast记录而不出现内存不足的问题。因为，它使一次处理
一个记录。我曾经用大处理过一个非常巨大的文件，没有出过任何问题。</p>
</div>
<div class="section" id="id72">
<h3>7.5.3  在巨大的BLAST纯文本文件中发现不对的记录</h3>
<p>当我开始解析一个巨大的blast 文件，有时候会碰到一个郁闷的问题就是解析器以一个
ValueError异常终止了。这是个严肃的问题。因为你无法分辨导致ValueError异常的是
解析器的问题还是BLAST的问题。更加糟糕是，你不知道在哪一行解析器失效了。所以，
你不能忽略这个错误。不然，可能会忽视一个重要的数据。</p>
<p>我们以前必须写一些小脚本来解决这个问题。不过，现在 <tt class="docutils literal">Bio.Blast</tt> 模块包含了
<tt class="docutils literal">BlastErrorParser</tt> ，可以更加简单地来解决这个问题。 <tt class="docutils literal">BlastErrorParser</tt>
和常规的 <tt class="docutils literal">BlastParser</tt> 类似，但是它加了特别一层来捕获由解析器产生的ValueErrors
异常，并尝试来诊断这些错误。</p>
<p>让我们来看看怎样用这个解析器 - 首先我们定义我们准备解析的文件和报告错误情况的
输出文件。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">os</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_file</span> <span class="operator">=</span> <span class="name">os</span><span class="operator">.</span><span class="name">path</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">os</span><span class="operator">.</span><span class="name">getcwd</span><span class="punctuation">(),</span> <span class="literal string">&quot;blast_out&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;big_blast.out&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">error_file</span> <span class="operator">=</span> <span class="name">os</span><span class="operator">.</span><span class="name">path</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">os</span><span class="operator">.</span><span class="name">getcwd</span><span class="punctuation">(),</span> <span class="literal string">&quot;blast_out&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;big_blast.problems&quot;</span><span class="punctuation">)</span>
</pre>
<p>现在我们想要一个  <tt class="docutils literal">BlastErrorParser</tt> ：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Blast</span> <span class="keyword namespace">import</span> <span class="name">NCBIStandalone</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">error_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="name">error_file</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_error_parser</span> <span class="operator">=</span> <span class="name">NCBIStandalone</span><span class="operator">.</span><span class="name">BlastErrorParser</span><span class="punctuation">(</span><span class="name">error_handle</span><span class="punctuation">)</span>
</pre>
<p>注意，解析器有个关于文件句柄的可选参数。如果传递了这个参数，那么解析器就会
把产生ValueError异常的记录写到这个文件句柄中。不然的话，这些错误记录就不会
被记录下来。</p>
<p>现在，我们可以像用常规的blast解析器一样地用 <tt class="docutils literal">BlastErrorParser</tt> 。
特别的是，我们也许想要一个一次读入一个记录的迭代器并用 <tt class="docutils literal">BlastErrorParser</tt>
来解析它。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="name">blast_file</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">iterator</span> <span class="operator">=</span> <span class="name">NCBIStandalone</span><span class="operator">.</span><span class="name">Iterator</span><span class="punctuation">(</span><span class="name">result_handle</span><span class="punctuation">,</span> <span class="name">blast_error_parser</span><span class="punctuation">)</span>
</pre>
<p>我们可以一次读一个记录，并且我们现在可以捕获并处理那些因为Blast引起的、
不是解析器本身导致的错误。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">try</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">next_record</span> <span class="operator">=</span> <span class="name">iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="operator">...</span> <span class="keyword">except</span> <span class="name">NCBIStandalone</span><span class="operator">.</span><span class="name">LowQualityBlastError</span><span class="punctuation">,</span> <span class="name">info</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;LowQualityBlastError detected in id </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">info</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
</pre>
<p><tt class="docutils literal">.next()</tt> 方法通常被 <tt class="docutils literal">for</tt> 循环间接地调用。现在， <tt class="docutils literal">BlastErrorParser</tt>
能够捕获如下的错误：</p>
<ul class="simple">
<li><tt class="docutils literal">ValueError</tt> - 这就是和常规BlastParser产生的一样的错误。这个错误产生
是因为解析器不能解析某个文件。通常是因为解析器有bug， 或者是
因为你使用解析器的版本和你BLAST命令的版本不一致。</li>
<li><tt class="docutils literal">LowQualityBlastError</tt> - 当Blast一条低质量的序列时（比如，一条
只有1个核苷酸的短序列），似乎Blast会终止并屏蔽掉整个序列，所有就没有什么可以
解析了。 这种情况下，Blast就会产生一个不完整的报告导致解析器出现ValueError
错误。 <tt class="docutils literal">LowQualityBlastError</tt> 错误在这种情况下产生。这个错误返回如下
信息：<ul>
<li><tt class="docutils literal">item[0]</tt> – The error message</li>
<li><tt class="docutils literal">item[0]</tt> - 错误消息</li>
<li><tt class="docutils literal">item[1]</tt> – The id of the input record that caused the error.
This is really useful if you want to record all of the records
that are causing problems.</li>
<li><tt class="docutils literal">item[1]</tt> - 导致错误产生的输入记录id。如果你想记录所有导致问题
记录的时候很有用。</li>
</ul>
</li>
</ul>
<p>就像上面提到的那样，BlastErrorParser 将会把有问题的记录写到指定的``error_handle``。
然后，你可以排查这些有问题记录。你可以针对某条记录来调试解析器，或者找到
你运行blast中的问题。无论哪种方式，这些都是有用的经验。</p>
<p>希望 <tt class="docutils literal">BlastErrorParser</tt> 能帮你更简单的调试和处理一些数据巨大的Blast 文件。</p>
</div>
</div>
<div class="section" id="psi-blast">
<h2>7.6  处理PSI-BLAST</h2>
<p>你可以通过 <tt class="docutils literal">Bio.Blast.Applications</tt> 模块中的包装函数来运行单机版本的PSI-BLAST
（老版本的NCBI命令工具行 <tt class="docutils literal">blastpgp</tt> 或者它的替代程序 <tt class="docutils literal">psiblast</tt> ）。</p>
<p>在写这篇指南的时候，没有迹象表明NCBI将会支持通过internet来进行PSI-BLAST
搜索。</p>
<p>请注意 <tt class="docutils literal">Bio.Blast.NCBIXML</tt> 解析器能读入并解析当前版本PSI-BLAST的、XML格式的
输出，但是像哪条序列在每个迭代循环中是新的还是复用的信息在XML格式输出中是没有的。
如果，你需要这些信息你应该用纯文本输出和 <tt class="docutils literal">Bio.Blast.NCBIStandalone</tt> 模块的
<tt class="docutils literal">PSIBlastParser</tt> 。</p>
</div>
<div class="section" id="rps-blast">
<h2>7.7  处理 RPS-BLAST</h2>
<p>你可以通过 <tt class="docutils literal">Bio.Blast.Applications</tt> 模块中的包装函数来运行单机版本的RPS-BLAST
（或者老版本的NCBI命令工具行 <tt class="docutils literal">rpsblast</tt> 或者同样名字的替代程序 ）。</p>
<p>在写这篇指南的时候，没有迹象表明NCBI将会支持通过internet来进行RPS-BLAST
搜索</p>
<p>你可以通过 <tt class="docutils literal">Bio.Blast.NCBIXML</tt> 这个解析器来读入并解析当前版本的RPS-BLAST的
XML格式的输出。</p>
</div>
</div>
<div class="section" id="id73">
<h1>第8章  BLAST和其他序列搜索工具(<em>实验性质的代码</em>)</h1>
<p><em>WARNING</em>: 这章教程介绍了Biopython中一个 <em>实验的</em> 模块。它正在被加入到
Biopython中，并且以一个预尾期的状态整理到教程当中，这样在我们发布稳定版
的之前可以收到一系列的反馈和并作改进。那时有些细节可能会改变，并且用到
当前 <tt class="docutils literal">Bio.SearchIO</tt> 模块的脚本也需要更新。切记！为了与NCBI BLAST有关的
代码可以稳定工作，请继续使用第 <a class="reference external" href="#chapter:blast">7</a> 章介绍的Bio.Blast。</p>
<p>生物序列的鉴定是生物信息工作的主要部分。有几个工具像BLAST（可能是最流行
的），FASTA ，HMMER还有许多其它的都有这个功能，每个工具都有独特的算法和
途径。一般来说，这些工具都是用你的序列在数据库中搜索可能的匹配。随着序列
数量的增加（匹配数也会随之增加），将会有成百上千的可能匹配，解析这些结果
无疑变得越来越困难。理所当然，人工解析搜索结果变得不可能。而且你可能会同
时用几种不同的搜索工具，每种工具都有独特的统计方法、规则和输出格式。可以
想象，同时用多种工具搜索多条序列是多么恐怖的事。</p>
<p>我们对此非常了解，所以我们在Biopython构建了 <tt class="docutils literal">Bio.SearchIO</tt> 亚模块。<tt class="docutils literal">Bio.SearchIO</tt> 模块使从搜索结果中提取信息变得简单，并且可以处理不同工具
的不同标准和规则。<tt class="docutils literal">SearchIO</tt> 和BioPerl中模块名称一致。</p>
<p>在本章中，我们将学习 <tt class="docutils literal">Bio.SearchIO</tt> 的主要功能，了解它可以做什么。我
们将使用两个主要的搜索工具：BLAST和FASTA。它们只是用来阐明思路，你可以轻
易地把工作流程应用到 <tt class="docutils literal">Bio.SearchIO</tt> 支持的其他工具中。欢迎你使用我们将要
用到的搜索结果文件。BLAST搜索结果文件可以在
<a class="reference external" href="http://biopython.org/SRC/Tests/Tutorial/my_blast.xml">这里</a> 下载。
BLAT输出结果文件可以在
<a class="reference external" href="http://biopython.org/SRC/Tests/Tutorial/my_blat.psl">这里</a> 下载。两个结果
文件都是用下面这条序列搜索产生的：</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">mystery_seq</span>
<span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG</span>
</pre>
<p>BLAST的XML结果是用 <tt class="docutils literal">blastn</tt> 搜索NCBI的 <tt class="docutils literal">refseq_rna</tt> 数据库得到的。对于
BLAT，数据库是2009年2月份的 <tt class="docutils literal">hg19</tt> 人类基因组草图，输出格式是PSL。</p>
<p>我们从 <tt class="docutils literal">Bio.SearchIO</tt> 的对象模型的介绍开始。这个模型代表你的搜索结果，因
此它是 <tt class="docutils literal">Bio.SearchIO</tt> 的核心。然后，我们会介绍 <tt class="docutils literal">Bio.SearchIO</tt> 常用的主要
功能。</p>
<p>现在一切就绪，让我们开始第一步：介绍核心对象模型。</p>
<div class="section" id="searchio">
<h2>8.1  SearchIO对象模型</h2>
<p>尽管多数搜索工具的输出风格极为不同，但是它们蕴含的理念很相似：</p>
<ul class="simple">
<li>输出文件可能包含一条或更多的搜索查询的结果。</li>
<li>在每次查询中，你会在给定的数据库中得到一个或更多的hit（命中）。</li>
<li>在每个hit中，你会得到一个或更多包含query（查询)序列和数据库序列实际比对的区域。</li>
<li>一些工具如BLAT和Exonerate可能会把这些区域分成几个比对片段（或在BLAT中
称为区块，在Exonerate中称为可能外显子）。这并不是很常见，像BLAST和
HMMER就不这么做。</li>
</ul>
<p>知道这些共性之后，我们决定把它们作为创造 <tt class="docutils literal">Bio.SearchIO</tt> 对象模型的基础。对
象模型是Python对象组成的嵌套分级系统，每个对象都代表一个上面列出的概念。这些对
象是：</p>
<ul class="simple">
<li><tt class="docutils literal">QueryResult</tt>，代表单个搜索查询。</li>
<li><tt class="docutils literal">Hit</tt>，代表单个的数据库hit。<tt class="docutils literal">Hit</tt> 对象包含在 <tt class="docutils literal">QueryResult</tt> 中，
每个 <tt class="docutils literal">QueryResult</tt> 中有0个或多个 <tt class="docutils literal">Hit</tt> 对象。</li>
<li><tt class="docutils literal">HSP</tt> (high-scoring pair（高分片段）)，代表query和hit序列中有
意义比对的区域。<tt class="docutils literal">HSP</tt> 对象包含在 <tt class="docutils literal">Hit</tt> 对象中，而且每个 <tt class="docutils literal">Hit</tt> 有一个
或更多的 <tt class="docutils literal">HSP</tt> 对象。</li>
<li><tt class="docutils literal">HSPFragment</tt>，代表query和hit序列中单个的邻近比对。 <tt class="docutils literal">HSPFragment</tt>
对象包含在 <tt class="docutils literal">HSP</tt> 对象中。多数的搜索工具如BLAST和HMMER把 <tt class="docutils literal">HSP</tt> 和
<tt class="docutils literal">HSPFragment</tt> 合并，因为一个 <tt class="docutils literal">HSP</tt> 只含有一个 <tt class="docutils literal">HSPFragment</tt>。但是
像BLAT和Exonerate会产生含有多个 <tt class="docutils literal">HSPFragment</tt> 的 <tt class="docutils literal">HSP</tt> 。似乎有些困
惑？不要紧，稍后我们将详细介绍这两个对象。</li>
</ul>
<p>这四个对象是当你用 <tt class="docutils literal">Bio.SearchIO</tt> 会碰到的。 <tt class="docutils literal">Bio.SearchIO</tt> 四
个主要方法： <tt class="docutils literal">read</tt> ，<tt class="docutils literal">parse</tt>，<tt class="docutils literal">index</tt> 或 <tt class="docutils literal">index_db</tt> 中任意一个都可
以产生这四个对象。这些方法的会在后面的部分详细介绍。这部分只会用到 <tt class="docutils literal">read</tt> 和
<tt class="docutils literal">parse</tt> ，这两个方法和 <tt class="docutils literal">Bio.SeqIO</tt> 以及 <tt class="docutils literal">Bio.AlignIO</tt> 中的 <tt class="docutils literal">read</tt> 和 <tt class="docutils literal">parse</tt> 方法功
能相似：</p>
<ul class="simple">
<li><tt class="docutils literal">read</tt> 用于单query对输出文件进行搜索并且返回一个 <tt class="docutils literal">QueryResult</tt> 对象。</li>
<li><tt class="docutils literal">parse</tt> 用于多query对输出文件进行搜索并且返回一个可以yield <tt class="docutils literal">QueryResult</tt> 对象的generator。</li>
</ul>
<p>完成这些之后，让我们开始学习每个 <tt class="docutils literal">Bio.SearchIO</tt> 对象，从 <tt class="docutils literal">QueryResult</tt> 开始。</p>
<div class="section" id="queryresult">
<h3>8.1.1  QueryResult</h3>
<p><tt class="docutils literal">QueryResult</tt>，代表单query搜索，每个 <tt class="docutils literal">QueryResult</tt> 中有0个或多个 <tt class="docutils literal">Hit</tt> 对象。我们来看看BLAST文件是什么样的：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blast.xml'</span><span class="punctuation">,</span> <span class="literal string">'blast-xml'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blast_qresult</span>
<span class="name">Program</span><span class="punctuation">:</span> <span class="name">blastn</span> <span class="punctuation">(</span><span class="literal number float">2.2</span><span class="operator">.</span><span class="literal number integer">27</span><span class="operator">+</span><span class="punctuation">)</span>
  <span class="name">Query</span><span class="punctuation">:</span> <span class="literal number integer">42291</span> <span class="punctuation">(</span><span class="literal number integer">61</span><span class="punctuation">)</span>
         <span class="name">mystery_seq</span>
 <span class="name">Target</span><span class="punctuation">:</span> <span class="name">refseq_rna</span>
   <span class="name">Hits</span><span class="punctuation">:</span> <span class="operator">----</span>  <span class="operator">-----</span>  <span class="operator">----------------------------------------------------------</span>
            <span class="comment">#  # HSP  ID + description</span>
         <span class="operator">----</span>  <span class="operator">-----</span>  <span class="operator">----------------------------------------------------------</span>
            <span class="literal number integer">0</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
            <span class="literal number integer">1</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171311</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035856</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
            <span class="literal number integer">2</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270133242</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032573</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Macaca</span> <span class="name">mulatta</span> <span class="name">microRNA</span> <span class="operator">...</span>
            <span class="literal number integer">3</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171322</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035857</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
            <span class="literal number integer">4</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171267</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035851</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
            <span class="literal number integer">5</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205330</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030198</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
            <span class="literal number integer">6</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205302</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030191</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">51.</span><span class="operator">..</span>
            <span class="literal number integer">7</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171259</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035850</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
            <span class="literal number integer">8</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205451</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030222</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">51.</span><span class="operator">..</span>
            <span class="literal number integer">9</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171447</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035871</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">10</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171276</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035852</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">11</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205290</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030188</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">51.</span><span class="operator">..</span>
           <span class="literal number integer">12</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171354</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035860</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">13</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205281</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030186</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
           <span class="literal number integer">14</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205298</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030190</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
           <span class="literal number integer">15</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171394</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035865</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">16</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205429</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030218</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">51.</span><span class="operator">..</span>
           <span class="literal number integer">17</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205423</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030217</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
           <span class="literal number integer">18</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171401</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035866</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">19</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270133247</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032574</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Macaca</span> <span class="name">mulatta</span> <span class="name">microRNA</span> <span class="operator">...</span>
           <span class="literal number integer">20</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205309</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030193</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
           <span class="literal number integer">21</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270132717</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032716</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Macaca</span> <span class="name">mulatta</span> <span class="name">microRNA</span> <span class="operator">...</span>
           <span class="literal number integer">22</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171437</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035870</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">23</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270133306</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032587</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Macaca</span> <span class="name">mulatta</span> <span class="name">microRNA</span> <span class="operator">...</span>
           <span class="literal number integer">24</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171428</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035869</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">25</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171211</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035845</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">26</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171153</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035838</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">27</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171146</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035837</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
           <span class="literal number integer">28</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270133254</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032575</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Macaca</span> <span class="name">mulatta</span> <span class="name">microRNA</span> <span class="operator">...</span>
           <span class="literal number integer">29</span>      <span class="literal number integer">2</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205445</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030221</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">51.</span><span class="operator">..</span>
           <span class="operator">~~~</span>
           <span class="literal number integer">97</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">356517317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_003527287</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">PREDICTED</span><span class="punctuation">:</span> <span class="name">Glycine</span> <span class="name">ma</span><span class="operator">...</span>
           <span class="literal number integer">98</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">297814701</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_002875188</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Arabidopsis</span> <span class="name">lyrata</span> <span class="name">su</span><span class="operator">...</span>
           <span class="literal number integer">99</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">397513516</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_003827011</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">PREDICTED</span><span class="punctuation">:</span> <span class="name">Pan</span> <span class="name">panisc</span><span class="operator">...</span>
</pre>
<p>虽然我们才接触对象模型的皮毛，但是你已经可以看到一些有用的信息了。通过调用``QueryResult`` 对象的 <tt class="docutils literal">print</tt> 方法，你可以看到：</p>
<ul class="simple">
<li>程序的名称和版本 (blastn version 2.2.27+)</li>
<li>query的ID，描述和序列长度(ID是42291，描述是 ‘mystery_seq’，长度是61)</li>
<li>搜索的目标数据库 (refseq_rna)</li>
<li>hit结果的快速预览。对于我们的查询序列，有100个可能的hit（表格中表示的是
0-99）对于每个hit，我们可以看到它包含的高分比对片段（HSP)，ID和一个片
段描述。注意， <tt class="docutils literal">Bio.SearchIO</tt> 截断了表格，只显示0-29，然后是97-99。</li>
</ul>
<p>现在让我们用同样的步骤来检查BLAT的结果：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blat.psl'</span><span class="punctuation">,</span> <span class="literal string">'blat-psl'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blat_qresult</span>
<span class="name">Program</span><span class="punctuation">:</span> <span class="name">blat</span> <span class="punctuation">(</span><span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">version</span><span class="operator">&gt;</span><span class="punctuation">)</span>
  <span class="name">Query</span><span class="punctuation">:</span> <span class="name">mystery_seq</span> <span class="punctuation">(</span><span class="literal number integer">61</span><span class="punctuation">)</span>
         <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
 <span class="name">Target</span><span class="punctuation">:</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">target</span><span class="operator">&gt;</span>
   <span class="name">Hits</span><span class="punctuation">:</span> <span class="operator">----</span>  <span class="operator">-----</span>  <span class="operator">----------------------------------------------------------</span>
            <span class="comment">#  # HSP  ID + description</span>
         <span class="operator">----</span>  <span class="operator">-----</span>  <span class="operator">----------------------------------------------------------</span>
            <span class="literal number integer">0</span>     <span class="literal number integer">17</span>  <span class="name">chr19</span>  <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
</pre>
<p>马上可以看到一些不同点。有些是由于BLAT使用PSL格式储存它的信息，稍后会看
到。其余是由于BLAST和BLAT搜索的程序和数据库之间明显的差异造成的：</p>
<ul class="simple">
<li>程序名称和版本。 <tt class="docutils literal">Bio.SearchIO</tt> 知道程序是BLAST，但是在输出文件中没
有信息显示程序版本，所以默认是 ‘&lt;unknown version&gt;’。</li>
<li>query的ID，描述和序列的长度。注意，这些细节和BLAST的细节只有细小的差别，
ID是 ‘mystery_seq’ 而不是42991，缺少描述，但是序列长度仍是61。这
实际上是文件格式本身导致的差异。BLAST有时创建自己的query ID并且用你的原
始ID作为序列描述。</li>
<li>目标数据库是未知的，因为BLAT输出文件没提到相关信息。</li>
<li>最后，hit列表完全不同。这里，我们的查询序列只命中到 ‘chr19’ 数据库条
目，但是我们可以看到它含有17个HSP区域。这并不让人诧异，考虑到我们
使用的是不同的程序，并且这些程序都有自己的数据库。</li>
</ul>
<p>所有通过调用 <tt class="docutils literal">print</tt> 方法看到的信息都可以单独地用Python的对象属性获得（又叫点标记法）。同样还可以用相同的方法获得其他格式特有的属性。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">blast_qresult</span><span class="operator">.</span><span class="name">program</span><span class="punctuation">,</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">version</span><span class="punctuation">)</span>
<span class="name">blastn</span> <span class="literal number float">2.2</span><span class="operator">.</span><span class="literal number integer">27</span><span class="operator">+</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">blat_qresult</span><span class="operator">.</span><span class="name">program</span><span class="punctuation">,</span> <span class="name">blat_qresult</span><span class="operator">.</span><span class="name">version</span><span class="punctuation">)</span>
<span class="name">blat</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">version</span><span class="operator">&gt;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">param_evalue_threshold</span>    <span class="comment"># blast-xml specific</span>
<span class="literal number float">10.0</span>
</pre>
<p>想获得一个可访问属性的完整列表，可以查询每个格式特有的文档。这些是
<a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SearchIO.BlastIO-module.html">BLAST</a>
<a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SearchIO.BlatIO-module.html">BLAT</a>.</p>
<p>已经知道了在 <tt class="docutils literal">QueryResult</tt> 对象上可以调用 <tt class="docutils literal">print</tt> 方法，让我们研究的更深
一些。 <tt class="docutils literal">QueryResult</tt> 到底是什么？就Python对象来说， <tt class="docutils literal">QueryResult</tt> 混合
了列表和字典的特性。换句话说，也就是一个包含了列表和字典功能的容器对象。</p>
<p>和列表以及字典一样， <tt class="docutils literal">QueryResult</tt> 对象是可迭代的。每次迭代返回一个hit
对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">hit</span> <span class="operator word">in</span> <span class="name">blast_qresult</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">hit</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|262205317|ref|NR_030195.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|301171311|ref|NR_035856.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|270133242|ref|NR_032573.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|301171322|ref|NR_035857.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">2</span> <span class="name">hsps</span><span class="punctuation">)</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|301171267|ref|NR_035851.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>要得到 <tt class="docutils literal">QueryResult</tt> 对象有多少hit，可以简单调用Python的 <tt class="docutils literal">len</tt> 方法：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">blast_qresult</span><span class="punctuation">)</span>
<span class="literal number integer">100</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">blat_qresult</span><span class="punctuation">)</span>
<span class="literal number integer">1</span>
</pre>
<p>同列表类似，你可以用切片来获得 <tt class="docutils literal">QueryResult</tt> 对象的hit：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>        <span class="comment"># retrieves the top hit</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|262205317|ref|NR_030195.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span>       <span class="comment"># retrieves the last hit</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|397513516|ref|XM_003827011.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
</pre>
<p>要得到多个hit，你同样可以对 <tt class="docutils literal">QueryResult</tt> 对象作切片。这种情况下，返回一个包含被切hit的新 <tt class="docutils literal">QueryResult</tt> 对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_slice</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="punctuation">[:</span><span class="literal number integer">3</span><span class="punctuation">]</span>     <span class="comment"># slices the first three hits</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blast_slice</span>
<span class="name">Program</span><span class="punctuation">:</span> <span class="name">blastn</span> <span class="punctuation">(</span><span class="literal number float">2.2</span><span class="operator">.</span><span class="literal number integer">27</span><span class="operator">+</span><span class="punctuation">)</span>
  <span class="name">Query</span><span class="punctuation">:</span> <span class="literal number integer">42291</span> <span class="punctuation">(</span><span class="literal number integer">61</span><span class="punctuation">)</span>
         <span class="name">mystery_seq</span>
 <span class="name">Target</span><span class="punctuation">:</span> <span class="name">refseq_rna</span>
   <span class="name">Hits</span><span class="punctuation">:</span> <span class="operator">----</span>  <span class="operator">-----</span>  <span class="operator">----------------------------------------------------------</span>
            <span class="comment">#  # HSP  ID + description</span>
         <span class="operator">----</span>  <span class="operator">-----</span>  <span class="operator">----------------------------------------------------------</span>
            <span class="literal number integer">0</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number float">52.</span><span class="operator">..</span>
            <span class="literal number integer">1</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171311</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035856</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span><span class="operator">...</span>
            <span class="literal number integer">2</span>      <span class="literal number integer">1</span>  <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270133242</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032573</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>  <span class="name">Macaca</span> <span class="name">mulatta</span> <span class="name">microRNA</span> <span class="operator">...</span>
</pre>
<p>同字典类似，可以通过ID获取hit。如果你知道一个特定的hit ID存在于一个搜索结果中时，特别有用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="punctuation">[</span><span class="literal string">'gi|262205317|ref|NR_030195.1|'</span><span class="punctuation">]</span>
<span class="name">Hit</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|262205317|ref|NR_030195.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">hsps</span><span class="punctuation">)</span>
</pre>
<p>你可以用 <tt class="docutils literal">hits</tt> 方法获得完整的 <tt class="docutils literal">Hit</tt> 对象，也可以用 <tt class="docutils literal">hit_keys</tt> 方法获得完整的 <tt class="docutils literal">Hit</tt> ID：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">hits</span>
<span class="punctuation">[</span><span class="operator">...</span><span class="punctuation">]</span>       <span class="comment"># list of all hits</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">hit_keys</span>
<span class="punctuation">[</span><span class="operator">...</span><span class="punctuation">]</span>       <span class="comment"># list of all hit IDs</span>
</pre>
<p>如果你想确定一个特定的hit是否存在于查询结果中该怎么做呢？可以用 <tt class="docutils literal">in</tt> 关键字作一个简单的成员检验：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="literal string">'gi|262205317|ref|NR_030195.1|'</span> <span class="operator word">in</span> <span class="name">blast_qresult</span>
<span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="literal string">'gi|262205317|ref|NR_030194.1|'</span> <span class="operator word">in</span> <span class="name">blast_qresult</span>
<span class="name builtin pseudo">False</span>
</pre>
<p>有时候，只知道一个hit是否存在是不够的，你可能也会想知道hit的排名。 <tt class="docutils literal">index</tt> 方法可以帮助你：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">'gi|301171437|ref|NR_035870.1|'</span><span class="punctuation">)</span>
<span class="literal number integer">22</span>
</pre>
<p>记住，我们用的是Python风格的索引，是从0开始。这代表hit的排名是23而不是22。</p>
<p>同样，注意你看的hit排名是基于原始搜索输出文件的本来顺序。不同的搜索工具可
能会基于不同的标准排列hit。</p>
<p>如果原本的hit排序不合你意，可以用 <tt class="docutils literal">QueryResult</tt> 对象的 <tt class="docutils literal">sort</tt> 方法。
它和Python的 <tt class="docutils literal">list.sort</tt> 方法很相似，只是有个是否创建一个新的排序后的
<tt class="docutils literal">QueryResult</tt> 对象的选项。</p>
<p>这里有个用 <tt class="docutils literal">QueryResult.sort</tt> 方法对hit排序的例子，这个方法基于每个hit
的完整序列长度。对于这个特殊的排序，我们设置 <tt class="docutils literal">in_place</tt> 参数等于 <tt class="docutils literal">False</tt> ，
这样排序方法会返回一个新的 <tt class="docutils literal">QueryResult</tt> 对象，而原来的对象是未排序的。
我们同样可以设置 <tt class="docutils literal">reverse</tt> 参数等于 `` True `` 以递减排序。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">hit</span> <span class="operator word">in</span> <span class="name">blast_qresult</span><span class="punctuation">[:</span><span class="literal number integer">5</span><span class="punctuation">]:</span>   <span class="comment"># id and sequence length of the first five hits</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">hit</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">hit</span><span class="operator">.</span><span class="name">seq_len</span>
<span class="operator">...</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">61</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171311</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035856</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">60</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270133242</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032573</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">85</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171322</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035857</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">86</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171267</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035851</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">80</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="name">sort_key</span> <span class="operator">=</span> <span class="keyword">lambda</span> <span class="name">hit</span><span class="punctuation">:</span> <span class="name">hit</span><span class="operator">.</span><span class="name">seq_len</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sorted_qresult</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">sort</span><span class="punctuation">(</span><span class="name">key</span><span class="operator">=</span><span class="name">sort_key</span><span class="punctuation">,</span> <span class="name">reverse</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">in_place</span><span class="operator">=</span><span class="name builtin pseudo">False</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">hit</span> <span class="operator word">in</span> <span class="name">sorted_qresult</span><span class="punctuation">[:</span><span class="literal number integer">5</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">hit</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name">hit</span><span class="operator">.</span><span class="name">seq_len</span>
<span class="operator">...</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">397513516</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_003827011</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">6002</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">390332045</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_776818</span><span class="operator">.</span><span class="literal number integer">2</span><span class="operator">|</span> <span class="literal number integer">4082</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">390332043</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_003723358</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">4079</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">356517317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_003527287</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">3251</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">356543101</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">XM_003539954</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">2936</span>
</pre>
<p>有 <tt class="docutils literal">in_place</tt> 参数的好处是可以保留原本的顺序，后面会用到。注意这不是 <tt class="docutils literal">QueryResult.sort</tt> 的默认行为，需要我们明确地设置 <tt class="docutils literal">in_place</tt> 为 <tt class="docutils literal">True</tt> 。</p>
<p>现在，你已经知道使用 <tt class="docutils literal">QueryResult</tt> 对象。但是，在我们学习 <tt class="docutils literal">Bio.SearchIO</tt>
模块下个对象前，先了解下可以让 <tt class="docutils literal">QueryResult</tt> 对象更易使用的两个方法：
<tt class="docutils literal">filter</tt> 和 <tt class="docutils literal">map</tt> 方法。</p>
<p>如果你对Python的列表推导式、generator表达式或内建的 <tt class="docutils literal">filter</tt> 和 <tt class="docutils literal">map</tt>
很熟悉，就知道（不知道就是看看吧!)它们在处理list-like的对象时有多有用。
你可以用这些内建的方法来操作 <tt class="docutils literal">QueryResult</tt> 对象，但是这只对正常list有效，并且可操作性也会受到限制。</p>
<p>这就是为什么 <tt class="docutils literal">QueryResult</tt> 对象提供自己特有的 <tt class="docutils literal">filter</tt> 和 <tt class="docutils literal">map</tt>
方法。对于 <tt class="docutils literal">filter</tt> 有相似的 <tt class="docutils literal">hit_filter</tt> 和 <tt class="docutils literal">hsp_filter</tt> 方法，
从名称就可以看出，这些方法过滤 <tt class="docutils literal">QueryResult</tt> 对象的 <tt class="docutils literal">Hit</tt> 对象或者
<tt class="docutils literal">HSP</tt> 对象。同样的，对于 <tt class="docutils literal">map</tt> ， <tt class="docutils literal">QueryResult</tt> 对象同样提供相似
的  <tt class="docutils literal">hit_map</tt> 和 <tt class="docutils literal">hsp_map</tt> 方法。这些方法分别应用于 <tt class="docutils literal">QueryResult</tt> 对象 <tt class="docutils literal">hit</tt> 或者 <tt class="docutils literal">HSP</tt> 对象。</p>
<p>让我们来看看这些方法的功能，从 <tt class="docutils literal">hit_filter</tt> 开始。这个方法接受一个回调
函数，这个函数检验给定的 <tt class="docutils literal">Hit</tt> 是否符合你设定的条件。换句话说，这个方法
必须接受一个单独 <tt class="docutils literal">Hit</tt> 对象作为参数并且返回  <tt class="docutils literal">True</tt> 或 <tt class="docutils literal">False</tt> 。</p>
<p>这里有个用 <tt class="docutils literal">hit_filter</tt> 筛选出只有一个HSP的 <tt class="docutils literal">Hit</tt> 对象的例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">filter_func</span> <span class="operator">=</span> <span class="keyword">lambda</span> <span class="name">hit</span><span class="punctuation">:</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">hit</span><span class="operator">.</span><span class="name">hsps</span><span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="literal number integer">1</span>     <span class="comment"># the callback function</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">blast_qresult</span><span class="punctuation">)</span>      <span class="comment"># no. of hits before filtering</span>
<span class="literal number integer">100</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">filtered_qresult</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">hit_filter</span><span class="punctuation">(</span><span class="name">filter_func</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">filtered_qresult</span><span class="punctuation">)</span>   <span class="comment"># no. of hits after filtering</span>
<span class="literal number integer">37</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">hit</span> <span class="operator word">in</span> <span class="name">filtered_qresult</span><span class="punctuation">[:</span><span class="literal number integer">5</span><span class="punctuation">]:</span>    <span class="comment"># quick check for the hit lengths</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">hit</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">hit</span><span class="operator">.</span><span class="name">hsps</span><span class="punctuation">)</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171322</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035857</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">2</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205330</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030198</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">2</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171447</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035871</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">2</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205298</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030190</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">2</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">270132717</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_032716</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="literal number integer">2</span>
</pre>
<p><tt class="docutils literal">hsp_filter</tt> 和 <tt class="docutils literal">hit_filter</tt> 功能相同，只是它过滤每个hit中的 <tt class="docutils literal">HSP</tt> 对象，
而不是 <tt class="docutils literal">Hit</tt> 。</p>
<p>对于 <tt class="docutils literal">map</tt> 方法，同样接受一个回调函数作为参数。但是回调函数返回修改过的 <tt class="docutils literal">Hit</tt> 或 <tt class="docutils literal">HSP</tt> 对象（取决于你是否使用 <tt class="docutils literal">hit_map</tt> 或 <tt class="docutils literal">hsp_map</tt> 方法），
而不是返回 <tt class="docutils literal">True</tt> 或 <tt class="docutils literal">False</tt>。</p>
<p>来看一个用 <tt class="docutils literal">hit_map</tt> 方法来重命名hit ID的例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">def</span> <span class="name function">map_func</span><span class="punctuation">(</span><span class="name">hit</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="name">hit</span><span class="operator">.</span><span class="name">id</span> <span class="operator">=</span> <span class="name">hit</span><span class="operator">.</span><span class="name">id</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">(</span><span class="literal string">'|'</span><span class="punctuation">)[</span><span class="literal number integer">3</span><span class="punctuation">]</span>   <span class="comment"># renames 'gi|301171322|ref|NR_035857.1|' to 'NR_035857.1'</span>
<span class="operator">...</span>     <span class="keyword">return</span> <span class="name">hit</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mapped_qresult</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="operator">.</span><span class="name">hit_map</span><span class="punctuation">(</span><span class="name">map_func</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">hit</span> <span class="operator word">in</span> <span class="name">mapped_qresult</span><span class="punctuation">[:</span><span class="literal number integer">5</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">hit</span><span class="operator">.</span><span class="name">id</span>
<span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">NR_035856</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">NR_032573</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">NR_035857</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">NR_035851</span><span class="operator">.</span><span class="literal number integer">1</span>
</pre>
<p>同样的， <tt class="docutils literal">hsp_map</tt> 和 <tt class="docutils literal">hit_map</tt> 作用相似, 但是作用于 <tt class="docutils literal">HSP</tt> 对象而不是 <tt class="docutils literal">Hit</tt> 对象。</p>
</div>
<div class="section" id="hit">
<h3>8.1.2  Hit</h3>
<p><tt class="docutils literal">Hit</tt> 对象代表从单个数据库获得所有查询结果。在 <tt class="docutils literal">Bio.SearchIO</tt> 对象等级中是二级容器。它们被包含在 <tt class="docutils literal">QueryResult</tt> 对象中，同时它们又包含 <tt class="docutils literal">HSP</tt> 对象。</p>
<p>看看它们是什么样的，从我们的BLAST搜索开始：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blast.xml'</span><span class="punctuation">,</span> <span class="literal string">'blast-xml'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hit</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">]</span>    <span class="comment"># fourth hit from the query result</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blast_hit</span>
<span class="name">Query</span><span class="punctuation">:</span> <span class="literal number integer">42291</span>
       <span class="name">mystery_seq</span>
  <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">301171322</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_035857</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="punctuation">(</span><span class="literal number integer">86</span><span class="punctuation">)</span>
       <span class="name">Pan</span> <span class="name">troglodytes</span> <span class="name">microRNA</span> <span class="name">mir</span><span class="operator">-</span><span class="literal number integer">520</span><span class="name">c</span> <span class="punctuation">(</span><span class="name">MIR520C</span><span class="punctuation">),</span> <span class="name">microRNA</span>
 <span class="name">HSPs</span><span class="punctuation">:</span> <span class="operator">----</span>  <span class="operator">--------</span>  <span class="operator">---------</span>  <span class="operator">------</span>  <span class="operator">---------------</span>  <span class="operator">---------------------</span>
          <span class="comment">#   E-value  Bit score    Span      Query range              Hit range</span>
       <span class="operator">----</span>  <span class="operator">--------</span>  <span class="operator">---------</span>  <span class="operator">------</span>  <span class="operator">---------------</span>  <span class="operator">---------------------</span>
          <span class="literal number integer">0</span>   <span class="literal number float">8.9e-20</span>     <span class="literal number float">100.47</span>      <span class="literal number integer">60</span>           <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>                <span class="punctuation">[</span><span class="literal number integer">13</span><span class="punctuation">:</span><span class="literal number integer">73</span><span class="punctuation">]</span>
          <span class="literal number integer">1</span>   <span class="literal number float">3.3e-06</span>      <span class="literal number float">55.39</span>      <span class="literal number integer">60</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>                <span class="punctuation">[</span><span class="literal number integer">13</span><span class="punctuation">:</span><span class="literal number integer">73</span><span class="punctuation">]</span>
</pre>
<p>可以看到我们获得了必要的信息：</p>
<ul class="simple">
<li>query的ID和描述信息。一个hit总是和一个query绑定，所以我们同样希望记录原始
query。这些值可以通过 <tt class="docutils literal">query_id</tt> 和  <tt class="docutils literal">query_description</tt> 属性从hit
中获取。</li>
<li>我们同样得到了hit的ID、描述和序列全长。它们可以分别通过 <tt class="docutils literal">id</tt>，
<tt class="docutils literal">description</tt>，和 <tt class="docutils literal">seq_len</tt> 获取。</li>
<li>最后，有一个hit含有的HSP的简短信息表。在每行中，HSP重要信息被
列出来：HSP索引，e值，得分，长度（包括gap），query坐标和hit坐标。</li>
</ul>
<p>现在，和BLAT结果作对比。记住，在BLAT搜索结果中，我们发现有一个含有17HSP的hit。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blat.psl'</span><span class="punctuation">,</span> <span class="literal string">'blat-psl'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hit</span> <span class="operator">=</span> <span class="name">blat_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>      <span class="comment"># the only hit</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blat_hit</span>
<span class="name">Query</span><span class="punctuation">:</span> <span class="name">mystery_seq</span>
       <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
  <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">chr19</span> <span class="punctuation">(</span><span class="literal number integer">59128983</span><span class="punctuation">)</span>
       <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
 <span class="name">HSPs</span><span class="punctuation">:</span> <span class="operator">----</span>  <span class="operator">--------</span>  <span class="operator">---------</span>  <span class="operator">------</span>  <span class="operator">---------------</span>  <span class="operator">---------------------</span>
          <span class="comment">#   E-value  Bit score    Span      Query range              Hit range</span>
       <span class="operator">----</span>  <span class="operator">--------</span>  <span class="operator">---------</span>  <span class="operator">------</span>  <span class="operator">---------------</span>  <span class="operator">---------------------</span>
          <span class="literal number integer">0</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54204480</span><span class="punctuation">:</span><span class="literal number integer">54204541</span><span class="punctuation">]</span>
          <span class="literal number integer">1</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54233104</span><span class="punctuation">:</span><span class="literal number integer">54264463</span><span class="punctuation">]</span>
          <span class="literal number integer">2</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54254477</span><span class="punctuation">:</span><span class="literal number integer">54260071</span><span class="punctuation">]</span>
          <span class="literal number integer">3</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54210720</span><span class="punctuation">:</span><span class="literal number integer">54210780</span><span class="punctuation">]</span>
          <span class="literal number integer">4</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54198476</span><span class="punctuation">:</span><span class="literal number integer">54198536</span><span class="punctuation">]</span>
          <span class="literal number integer">5</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54265610</span><span class="punctuation">:</span><span class="literal number integer">54265671</span><span class="punctuation">]</span>
          <span class="literal number integer">6</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54238143</span><span class="punctuation">:</span><span class="literal number integer">54240175</span><span class="punctuation">]</span>
          <span class="literal number integer">7</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54189735</span><span class="punctuation">:</span><span class="literal number integer">54189795</span><span class="punctuation">]</span>
          <span class="literal number integer">8</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54185425</span><span class="punctuation">:</span><span class="literal number integer">54185486</span><span class="punctuation">]</span>
          <span class="literal number integer">9</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54197657</span><span class="punctuation">:</span><span class="literal number integer">54197717</span><span class="punctuation">]</span>
         <span class="literal number integer">10</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54255662</span><span class="punctuation">:</span><span class="literal number integer">54255723</span><span class="punctuation">]</span>
         <span class="literal number integer">11</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54201651</span><span class="punctuation">:</span><span class="literal number integer">54201712</span><span class="punctuation">]</span>
         <span class="literal number integer">12</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">8</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54206009</span><span class="punctuation">:</span><span class="literal number integer">54206061</span><span class="punctuation">]</span>
         <span class="literal number integer">13</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>          <span class="punctuation">[</span><span class="literal number integer">10</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54178987</span><span class="punctuation">:</span><span class="literal number integer">54179038</span><span class="punctuation">]</span>
         <span class="literal number integer">14</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">8</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54212018</span><span class="punctuation">:</span><span class="literal number integer">54212071</span><span class="punctuation">]</span>
         <span class="literal number integer">15</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">8</span><span class="punctuation">:</span><span class="literal number integer">51</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54234278</span><span class="punctuation">:</span><span class="literal number integer">54234321</span><span class="punctuation">]</span>
         <span class="literal number integer">16</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">8</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54238143</span><span class="punctuation">:</span><span class="literal number integer">54238196</span><span class="punctuation">]</span>
</pre>
<p>我们得到了和前面看到的BLAST hit详细程度相似的结果。但是有些不同点需要解释：</p>
<ul class="simple">
<li>e-value和bit score列的值。因为BLAT HSP没有e-values和bit scores，默
认显示‘?’.</li>
<li>span列是怎么回事呢？span值本来是显示完整的比对长度，包含所有的残基和
gap。但是PSL格式目前还不支持这些信息并且 <tt class="docutils literal">Bio.SearchIO</tt> 也不打算去猜它到底是多少，所有我们得到了和e-value以及bit score列相同的 ‘?’。</li>
</ul>
<p>就Python对象来说， <tt class="docutils literal">Hit</tt> 和列表行为最相似，但是额外含有 <tt class="docutils literal">HSP</tt> 。如果
你对列表熟悉，在使用 <tt class="docutils literal">Hit</tt> 对象是不会遇到困难。</p>
<p>和列表一样， <tt class="docutils literal">Hit</tt> 对象是可迭代的，并且每次迭代返回一个 <tt class="docutils literal">HSP</tt> 对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">hsp</span> <span class="operator word">in</span> <span class="name">blast_hit</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">hsp</span>
<span class="name">HSP</span><span class="punctuation">(</span><span class="name">hit_id</span><span class="operator">=</span><span class="literal string">'gi|301171322|ref|NR_035857.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">fragments</span><span class="punctuation">)</span>
<span class="name">HSP</span><span class="punctuation">(</span><span class="name">hit_id</span><span class="operator">=</span><span class="literal string">'gi|301171322|ref|NR_035857.1|'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">fragments</span><span class="punctuation">)</span>
</pre>
<p>你可以对 <tt class="docutils literal">Hit</tt> 对象调用 <tt class="docutils literal">len</tt> 方法查看它含有多少个 <tt class="docutils literal">HSP</tt> 对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">blast_hit</span><span class="punctuation">)</span>
<span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">blat_hit</span><span class="punctuation">)</span>
<span class="literal number integer">17</span>
</pre>
<p>你可以对 <tt class="docutils literal">Hit</tt> 对象作切片取得单个或多个 <tt class="docutils literal">HSP</tt> 对象，和 <tt class="docutils literal">QueryResult</tt>
一样，如果切取多个 <tt class="docutils literal">HSP</tt>  ，会返回包含被切片 <tt class="docutils literal">HSP</tt>  的一个新 <tt class="docutils literal">Hit</tt> 对象。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hit</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>                 <span class="comment"># retrieve single items</span>
<span class="name">HSP</span><span class="punctuation">(</span><span class="name">hit_id</span><span class="operator">=</span><span class="literal string">'chr19'</span><span class="punctuation">,</span> <span class="name">query_id</span><span class="operator">=</span><span class="literal string">'mystery_seq'</span><span class="punctuation">,</span> <span class="literal number integer">1</span> <span class="name">fragments</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sliced_hit</span> <span class="operator">=</span> <span class="name">blat_hit</span><span class="punctuation">[</span><span class="literal number integer">4</span><span class="punctuation">:</span><span class="literal number integer">9</span><span class="punctuation">]</span>  <span class="comment"># retrieve multiple items</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">sliced_hit</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">sliced_hit</span>
<span class="name">Query</span><span class="punctuation">:</span> <span class="name">mystery_seq</span>
       <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
  <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">chr19</span> <span class="punctuation">(</span><span class="literal number integer">59128983</span><span class="punctuation">)</span>
       <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
 <span class="name">HSPs</span><span class="punctuation">:</span> <span class="operator">----</span>  <span class="operator">--------</span>  <span class="operator">---------</span>  <span class="operator">------</span>  <span class="operator">---------------</span>  <span class="operator">---------------------</span>
          <span class="comment">#   E-value  Bit score    Span      Query range              Hit range</span>
       <span class="operator">----</span>  <span class="operator">--------</span>  <span class="operator">---------</span>  <span class="operator">------</span>  <span class="operator">---------------</span>  <span class="operator">---------------------</span>
          <span class="literal number integer">0</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54198476</span><span class="punctuation">:</span><span class="literal number integer">54198536</span><span class="punctuation">]</span>
          <span class="literal number integer">1</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54265610</span><span class="punctuation">:</span><span class="literal number integer">54265671</span><span class="punctuation">]</span>
          <span class="literal number integer">2</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54238143</span><span class="punctuation">:</span><span class="literal number integer">54240175</span><span class="punctuation">]</span>
          <span class="literal number integer">3</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">60</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54189735</span><span class="punctuation">:</span><span class="literal number integer">54189795</span><span class="punctuation">]</span>
          <span class="literal number integer">4</span>         <span class="error">?</span>          <span class="error">?</span>       <span class="error">?</span>           <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>    <span class="punctuation">[</span><span class="literal number integer">54185425</span><span class="punctuation">:</span><span class="literal number integer">54185486</span><span class="punctuation">]</span>
</pre>
<p>你同样可以对一个 <tt class="docutils literal">Hit</tt> 里的 <tt class="docutils literal">HSP</tt>  排序，和你在 <tt class="docutils literal">QueryResult</tt> 对象
中看到的方法一样。</p>
<p>最后，同样可以对 <tt class="docutils literal">Hit</tt> 对象使用 <tt class="docutils literal">filter</tt> 和 <tt class="docutils literal">map</tt> 方法。和 <tt class="docutils literal">QueryResult</tt>
不同， <tt class="docutils literal">Hit</tt> 对象只有一种 <tt class="docutils literal">filter</tt> (<tt class="docutils literal">Hit.filter</tt>) 和一种 <tt class="docutils literal">map</tt> (<tt class="docutils literal">Hit.map</tt>)。</p>
</div>
<div class="section" id="hsp">
<h3>8.1.3  HSP</h3>
<p><tt class="docutils literal">HSP</tt> (高分片段)代表hit序列中的一个区域，该区域包含对于查询序列有意义的
比对。它包含了你的查询序列和一个数据库条目之间精确的匹配。由于匹配取决于
序列搜索工具的算法， <tt class="docutils literal">HSP</tt>  含有大部分统计信息，这些统计是由搜索工具计
算得到的。这使得不同搜索工具的 <tt class="docutils literal">HSP</tt>  对象之间的差异和你在 <tt class="docutils literal">QueryResult</tt>
以及 <tt class="docutils literal">Hit</tt> 对象看到的差异更加明显。</p>
<p>我们来看看BLAST和BLAT搜索的例子。先看BLAST HSP：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blast.xml'</span><span class="punctuation">,</span> <span class="literal string">'blast-xml'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">]</span>    <span class="comment"># first hit, first hsp</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blast_hsp</span>
      <span class="name">Query</span><span class="punctuation">:</span> <span class="literal number integer">42291</span> <span class="name">mystery_seq</span>
        <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number integer">520</span><span class="name">b</span> <span class="punctuation">(</span><span class="name">MIR520</span><span class="operator">...</span>
<span class="name">Query</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Hit</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="name">Quick</span> <span class="name">stats</span><span class="punctuation">:</span> <span class="name">evalue</span> <span class="literal number float">4.9e-23</span><span class="punctuation">;</span> <span class="name">bitscore</span> <span class="literal number float">111.29</span>
  <span class="name">Fragments</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="punctuation">(</span><span class="literal number integer">61</span> <span class="name">columns</span><span class="punctuation">)</span>
     <span class="name">Query</span> <span class="operator">-</span> <span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG</span>
             <span class="operator">|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||</span>
       <span class="name">Hit</span> <span class="operator">-</span> <span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG</span>
</pre>
<p>和 <tt class="docutils literal">QueryResult</tt> 以及 <tt class="docutils literal">Hit</tt> 类似，调用 <tt class="docutils literal">HSP</tt>  的 <tt class="docutils literal">print</tt> 方法,
显示细节：</p>
<ul class="simple">
<li>有query和hit的ID以及描述。我们需要这些来识别我们的 <tt class="docutils literal">HSP</tt>  。</li>
<li>我们同样得到了query和hit序列的匹配范围。这里用的的切片标志着范围的表示
是使用Python的索引风格（从0开始，半开区间）。圆括号里的数字表示正负链。
这里，两条序列都是正链。</li>
<li>还有一些简短统计：e-value和bitscore。</li>
<li>还有一些HSP片段的信息。现在可以忽略，稍后会解释。</li>
<li>最后，还有query和hit的比对。</li>
</ul>
<p>这些信息可以用点标记从它们本身获得，和 <tt class="docutils literal">Hit</tt> 以及 <tt class="docutils literal">QueryResult</tt> 相同：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">query_range</span>
<span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">61</span><span class="punctuation">)</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">evalue</span>
<span class="literal number float">4.91307e-23</span>
</pre>
<p>它们并不是仅有的属性， <tt class="docutils literal">HSP</tt>  对象有一系列的属性，使得获得它们的具体信
息更加容易。下面是一些例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">hit_start</span>         <span class="comment"># start coordinate of the hit sequence</span>
<span class="literal number integer">0</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">query_span</span>        <span class="comment"># how many residues in the query sequence</span>
<span class="literal number integer">61</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">aln_span</span>          <span class="comment"># how long the alignment is</span>
<span class="literal number integer">61</span>
</pre>
<p>查看 <tt class="docutils literal">HSP</tt>
<a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SearchIO._model.hsp-module.html">文档</a>
获取完整的属性列表。</p>
<p>不仅如此，每个搜索工具通常会对它的 <tt class="docutils literal">HSP</tt>  对象作统计学或其他细节计算。例如，一个
XML BLAST搜索同样输出gap以及相同的残基数量。这些属性可以像这样被获取：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">gap_num</span>       <span class="comment"># number of gaps</span>
<span class="literal number integer">0</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">ident_num</span>     <span class="comment"># number of identical residues</span>
<span class="literal number integer">61</span>
</pre>
<p>这些细节是格式特异的；它们可能不会出现在其他的格式中。要知道哪些细节在给
定的序列搜索工具中是存在的，你应该查看那种格式的在 <tt class="docutils literal">Bio.SearchIO</tt> 中的
文档。或者可以用 <tt class="docutils literal">.__dict__.keys()</tt> 获得快速列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">__dict__</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">'bitscore'</span><span class="punctuation">,</span> <span class="literal string">'evalue'</span><span class="punctuation">,</span> <span class="literal string">'ident_num'</span><span class="punctuation">,</span> <span class="literal string">'gap_num'</span><span class="punctuation">,</span> <span class="literal string">'bitscore_raw'</span><span class="punctuation">,</span> <span class="literal string">'pos_num'</span><span class="punctuation">,</span> <span class="literal string">'_items'</span><span class="punctuation">]</span>
</pre>
<p>最后，你可能已经注意到了，我们HSP的 <tt class="docutils literal">query</tt> 和 <tt class="docutils literal">hit</tt> 属性不只是规律字符串：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">query</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTT...GGG'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'42291'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'aligned query sequence'</span><span class="punctuation">,</span> <span class="name">description</span><span class="operator">=</span><span class="literal string">'mystery_seq'</span><span class="punctuation">,</span> <span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[])</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">hit</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTT...GGG'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|262205317|ref|NR_030195.1|'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'aligned hit sequence'</span><span class="punctuation">,</span> <span class="name">description</span><span class="operator">=</span><span class="literal string">'Homo sapiens microRNA 520b (MIR520B), microRNA'</span><span class="punctuation">,</span> <span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[])</span>
</pre>
<p>它们是你已经在第 <a class="reference external" href="#chapter:SeqRecord">4</a> 章看到过的 <tt class="docutils literal">SeqRecord</tt> 对象！
意味着你可以对 <tt class="docutils literal">SeqRecord</tt> 对象做的各种有趣的事同样适用于 <tt class="docutils literal">HSP.query</tt> 和 <tt class="docutils literal">HSP.hit</tt> 对象。</p>
<p>现在 <tt class="docutils literal">HSP</tt>  对象有个 <tt class="docutils literal">alignment</tt> 属性（一个 <tt class="docutils literal">MultipleSeqAlignment</tt>
对象）应该不会让你感到惊讶：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blast_hsp</span><span class="operator">.</span><span class="name">aln</span>
<span class="name">DNAAlphabet</span><span class="punctuation">()</span> <span class="name">alignment</span> <span class="keyword">with</span> <span class="literal number integer">2</span> <span class="name">rows</span> <span class="operator word">and</span> <span class="literal number integer">61</span> <span class="name">columns</span>
<span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAG</span><span class="operator">...</span><span class="name">GGG</span> <span class="literal number integer">42291</span>
<span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAG</span><span class="operator">...</span><span class="name">GGG</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>
</pre>
<p>探索完BLAST HSP对象，让我们看看来自BLAT结果的不一样的HSP。我们将对它调用 <tt class="docutils literal">print</tt> 方法：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blat.psl'</span><span class="punctuation">,</span> <span class="literal string">'blat-psl'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span> <span class="operator">=</span> <span class="name">blat_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">]</span>       <span class="comment"># first hit, first hsp</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blat_hsp</span>
      <span class="name">Query</span><span class="punctuation">:</span> <span class="name">mystery_seq</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
        <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">chr19</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
<span class="name">Query</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Hit</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">54204480</span><span class="punctuation">:</span><span class="literal number integer">54204541</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="name">Quick</span> <span class="name">stats</span><span class="punctuation">:</span> <span class="name">evalue</span> <span class="error">?</span><span class="punctuation">;</span> <span class="name">bitscore</span> <span class="error">?</span>
  <span class="name">Fragments</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="punctuation">(</span><span class="error">?</span> <span class="name">columns</span><span class="punctuation">)</span>
</pre>
<p>一些输出你应该已经猜到了。我们得到了查询序列、hit ID、描述以及序列坐标。
evalue和bitscore的值是 ‘?’ ，因为BLAT HSP并没有这些属性。但是最大的不同
是你看不到任何的序列比对展示。如果你看的更仔细，PSL格式本身并没有任何的
hit和query序列，所以 <tt class="docutils literal">Bio.SearchIO</tt> 不会创建任何序列或者比对对象。如果
你尝试获取 <tt class="docutils literal">HSP.query</tt> ，<tt class="docutils literal">HSP.hit</tt> ， 或者 <tt class="docutils literal">HSP.aln</tt> 属性会怎么样
呢？你会得到这些属性的默认值 <tt class="docutils literal">None</tt> ：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">hit</span> <span class="operator word">is</span> <span class="name builtin pseudo">None</span>
<span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">query</span> <span class="operator word">is</span> <span class="name builtin pseudo">None</span>
<span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">aln</span> <span class="operator word">is</span> <span class="name builtin pseudo">None</span>
<span class="name builtin pseudo">True</span>
</pre>
<p>这并不影响其他的属性。例如，你仍然可以获取query和hit比对的长度。尽管不显
示任何的属性，但是PSL格式还是有这些信息的，所以 <tt class="docutils literal">Bio.SearchIO</tt> 可以抽
提出这些信息。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">query_span</span>     <span class="comment"># length of query match</span>
<span class="literal number integer">61</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">hit_span</span>       <span class="comment"># length of hit match</span>
<span class="literal number integer">61</span>
</pre>
<p>其他格式特异的属性同样被展示出来：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">score</span>          <span class="comment"># PSL score</span>
<span class="literal number integer">61</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">mismatch_num</span>   <span class="comment"># the mismatch column</span>
<span class="literal number integer">0</span>
</pre>
<p>到目前为止，一切还不错？当你看到BLAT结果中不同的HSP时，事情变得更有趣了。
你可能会回想起在BLAT搜索中，有时我们把结果分成 ‘blocks’ 。这些区块是必需比对片段，可能会有些内含子在它们之间。</p>
<p>让我们看看 <tt class="docutils literal">Bio.SearchIO</tt> 怎么处理包含多个区块的BLAT HSP：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span> <span class="operator">=</span> <span class="name">blat_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">1</span><span class="punctuation">]</span>      <span class="comment"># first hit, second hsp</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blat_hsp2</span>
      <span class="name">Query</span><span class="punctuation">:</span> <span class="name">mystery_seq</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
        <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">chr19</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
<span class="name">Query</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Hit</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">54233104</span><span class="punctuation">:</span><span class="literal number integer">54264463</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="name">Quick</span> <span class="name">stats</span><span class="punctuation">:</span> <span class="name">evalue</span> <span class="error">?</span><span class="punctuation">;</span> <span class="name">bitscore</span> <span class="error">?</span>
  <span class="name">Fragments</span><span class="punctuation">:</span> <span class="operator">---</span>  <span class="operator">--------------</span>  <span class="operator">----------------------</span>  <span class="operator">----------------------</span>
               <span class="comment">#            Span             Query range               Hit range</span>
             <span class="operator">---</span>  <span class="operator">--------------</span>  <span class="operator">----------------------</span>  <span class="operator">----------------------</span>
               <span class="literal number integer">0</span>               <span class="error">?</span>                  <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">18</span><span class="punctuation">]</span>     <span class="punctuation">[</span><span class="literal number integer">54233104</span><span class="punctuation">:</span><span class="literal number integer">54233122</span><span class="punctuation">]</span>
               <span class="literal number integer">1</span>               <span class="error">?</span>                 <span class="punctuation">[</span><span class="literal number integer">18</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span>     <span class="punctuation">[</span><span class="literal number integer">54264420</span><span class="punctuation">:</span><span class="literal number integer">54264463</span><span class="punctuation">]</span>
</pre>
<p>怎么回事？我们仍然得到了一些必要的信息：ID，描述信息，坐标和快速统计，和
你前面看到的一样。但是片段信息完全不同。我们得到了有两行数据的表格，而不是显示 ‘Fragment: 1’。</p>
<p>这就是 <tt class="docutils literal">Bio.SearchIO</tt> 处理含有多片段HSP的方式。和前面提到的一样，一个
HSP比对可能会被内含子分成多个片段。内含子不是query-hit匹配的一部分，所以
它们不能被当成query或hit序列的一部分。但是，它们确实影响我们处理序列坐标，
所以我们不能忽视。</p>
<p>看看上面的HSP的hit坐标。在 <tt class="docutils literal">Hit range</tt> 区域，我们看到坐标是
<tt class="docutils literal">[54233104:54264463]</tt>。但是看看表格中的行，我们发现不是坐标跨度的所有区域
都能匹配我们的query。特殊的是，间断区域从 <tt class="docutils literal">54233122</tt> 到 <tt class="docutils literal">54264420</tt> 。</p>
<p>你可能会问，为什么query坐标好像是邻近的?这是很好的。在这个例子中，query是连续的（无间断区域），但是hit却不是。</p>
<p>所有的这些属性都是可以直接从HSP获取的，通过这样的方式：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">hit_range</span>         <span class="comment"># hit start and end coordinates of the entire HSP</span>
<span class="punctuation">(</span><span class="literal number integer">54233104</span><span class="punctuation">,</span> <span class="literal number integer">54264463</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">hit_range_all</span>     <span class="comment"># hit start and end coordinates of each fragment</span>
<span class="punctuation">[(</span><span class="literal number integer">54233104</span><span class="punctuation">,</span> <span class="literal number integer">54233122</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="literal number integer">54264420</span><span class="punctuation">,</span> <span class="literal number integer">54264463</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">hit_span</span>          <span class="comment"># hit span of the entire HSP</span>
<span class="literal number integer">31359</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">hit_span_all</span>      <span class="comment"># hit span of each fragment</span>
<span class="punctuation">[</span><span class="literal number integer">18</span><span class="punctuation">,</span> <span class="literal number integer">43</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">hit_inter_ranges</span>  <span class="comment"># start and end coordinates of intervening regions in the hit sequence</span>
<span class="punctuation">[(</span><span class="literal number integer">54233122</span><span class="punctuation">,</span> <span class="literal number integer">54264420</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">hit_inter_spans</span>   <span class="comment"># span of intervening regions in the hit sequence</span>
<span class="punctuation">[</span><span class="literal number integer">31298</span><span class="punctuation">]</span>
</pre>
<p>这些属性中大多数都不能简单地从PSL文件获得，但是当你分析PSL文件时，
<tt class="docutils literal">Bio.SearchIO</tt> 会动态地帮你计算。它需要的只是每个片段的开始和结束坐标。</p>
<p><tt class="docutils literal">query</tt>， <tt class="docutils literal">hit</tt>， 和 <tt class="docutils literal">aln</tt> 属性又是什么情况？如果HSP含有多个片段，
你就不能使用这些属性，因为它们只取回单个 <tt class="docutils literal">SeqRecord</tt> 或
<tt class="docutils literal">MultipleSeqAlignment</tt> 对象。但是，你可以用相应的 <tt class="docutils literal">*_all</tt> 方法：
<tt class="docutils literal">query_all</tt>， <tt class="docutils literal">hit_all</tt>， 和 <tt class="docutils literal">aln_all</tt>。 这些属性会返回包含每个HSP
片段的 <tt class="docutils literal">SeqRecord</tt> 或 <tt class="docutils literal">MultipleSeqAlignment</tt> 对象的列表。还有其他相同
功能的属性，也就是只对只有一个片段的HSP有效。查看 <tt class="docutils literal">HSP</tt>
<a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SearchIO._model.hsp-module.html">文档</a>
获得完整的列表。</p>
<p>最后，想要检查是否是多片段HSP，你可以用 <tt class="docutils literal">is_fragmented</tt> 属性：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp2</span><span class="operator">.</span><span class="name">is_fragmented</span>     <span class="comment"># BLAT HSP with 2 fragments</span>
<span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_hsp</span><span class="operator">.</span><span class="name">is_fragmented</span>      <span class="comment"># BLAT HSP from earlier, with one fragment</span>
<span class="name builtin pseudo">False</span>
</pre>
<p>在进入下部分之前，你只需要了解我们可以对 <tt class="docutils literal">HSP</tt> 对象使用切片，和
<tt class="docutils literal">QueryResult</tt> 或 <tt class="docutils literal">Hit</tt> 对象一样。当你使用切片的时候，会返回一个
<tt class="docutils literal">HSPFragment</tt> 对象。</p>
</div>
<div class="section" id="id74">
<h3>8.1.4  HSP片段</h3>
<p><tt class="docutils literal">HSPFragment</tt> 代表query和hit之间单个连续匹配。应该把它当作对象模型
和搜索结果的核心，因为它决定你的搜索是否有结果。</p>
<p>在多数情况下，你不必直接处理 <tt class="docutils literal">HSPFragment</tt> 对象，因为没有那么多搜索工具
断裂它们的HSP。当你确实需要处理它们时，需要记住的是 <tt class="docutils literal">HSPFragment</tt> 对象
要被写地尽量压缩。在多数情况下，它们仅仅包含直接与序列有关的属性：正负链，
阅读框，字母表，位置坐标，序列本身以及它们的ID和描述。</p>
<p>当你对 <tt class="docutils literal">HSPFragment</tt> 对象调用 <tt class="docutils literal">print</tt> 方法时，这些属性可以非常简单地显示
出来。这里有个从我们BLAST搜索得到的例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blast.xml'</span><span class="punctuation">,</span> <span class="literal string">'blast-xml'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_frag</span> <span class="operator">=</span> <span class="name">blast_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">]</span>    <span class="comment"># first hit, first hsp, first fragment</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blast_frag</span>
      <span class="name">Query</span><span class="punctuation">:</span> <span class="literal number integer">42291</span> <span class="name">mystery_seq</span>
        <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">gi</span><span class="operator">|</span><span class="literal number integer">262205317</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NR_030195</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span> <span class="name">Homo</span> <span class="name">sapiens</span> <span class="name">microRNA</span> <span class="literal number integer">520</span><span class="name">b</span> <span class="punctuation">(</span><span class="name">MIR520</span><span class="operator">...</span>
<span class="name">Query</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Hit</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Fragments</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="punctuation">(</span><span class="literal number integer">61</span> <span class="name">columns</span><span class="punctuation">)</span>
     <span class="name">Query</span> <span class="operator">-</span> <span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG</span>
             <span class="operator">|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||</span>
       <span class="name">Hit</span> <span class="operator">-</span> <span class="name">CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTTTAGAGGG</span>
</pre>
<p>在这个水平上，BLAT和BLAST片段看起来非常相似，除了没有出现的query和hit序列：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'my_blat.psl'</span><span class="punctuation">,</span> <span class="literal string">'blat-psl'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blat_frag</span> <span class="operator">=</span> <span class="name">blat_qresult</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">]</span>    <span class="comment"># first hit, first hsp, first fragment</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">blat_frag</span>
      <span class="name">Query</span><span class="punctuation">:</span> <span class="name">mystery_seq</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
        <span class="name">Hit</span><span class="punctuation">:</span> <span class="name">chr19</span> <span class="operator">&lt;</span><span class="name">unknown</span> <span class="name">description</span><span class="operator">&gt;</span>
<span class="name">Query</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">:</span><span class="literal number integer">61</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Hit</span> <span class="name builtin">range</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">54204480</span><span class="punctuation">:</span><span class="literal number integer">54204541</span><span class="punctuation">]</span> <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
  <span class="name">Fragments</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="punctuation">(</span><span class="error">?</span> <span class="name">columns</span><span class="punctuation">)</span>
</pre>
<p>在所有情况下，这些属性都可以通过我们最爱的点标记访问。一些例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_frag</span><span class="operator">.</span><span class="name">query_start</span>      <span class="comment"># query start coordinate</span>
<span class="literal number integer">0</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_frag</span><span class="operator">.</span><span class="name">hit_strand</span>       <span class="comment"># hit sequence strand</span>
<span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">blast_frag</span><span class="operator">.</span><span class="name">hit</span>              <span class="comment"># hit sequence, as a SeqRecord object</span>
<span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CCCTCTACAGGGAAGCGCTTTCTGTTGTCTGAAAGAAAAGAAAGTGCTTCCTTT...GGG'</span><span class="punctuation">,</span> <span class="name">DNAAlphabet</span><span class="punctuation">()),</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|262205317|ref|NR_030195.1|'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'aligned hit sequence'</span><span class="punctuation">,</span> <span class="name">description</span><span class="operator">=</span><span class="literal string">'Homo sapiens microRNA 520b (MIR520B), microRNA'</span><span class="punctuation">,</span> <span class="name">dbxrefs</span><span class="operator">=</span><span class="punctuation">[])</span>
</pre>
</div>
</div>
<div class="section" id="id75">
<h2>8.2  一个关于标准和惯例的注意事项</h2>
<p>在我们进入到主要功能前，你需要知道 <tt class="docutils literal">Bio.SearchIO</tt> 使用的一些标准。如果
你已经接触过多序列搜索工具，你可能必须面对每个程序处理事情方式不同的问题，
如序列位置坐标。这可能不是一个令人高兴的经历，因为这些搜索工具通常有它们
自己的标准。例如，一种工具可能使用“从1开始”(one-based)的坐标，而其他工具
使用“从0开始”(zero-based)的坐标。或者，一种程序在处理负链时，可能会反转
开始和结束坐标，而其他程序确不会。简而言之，会产生一些必须要处理的混乱。</p>
<p>我们意识到这种问题，并且打算在 <tt class="docutils literal">Bio.SearchIO</tt> 中解决。毕竟， <tt class="docutils literal">Bio.SearchIO</tt> 的目标之一就是创建一个通用简单的接口来处理多种不同的搜索
输出文件。意味着要制定一个超越你所见的对象模型的标准。</p>
<p>现在，你可能抱怨，”不要又来一个标准“。好吧，最后我们必须选择一个标准，这
是必须的。并且，我们并不是创造一个全新的事物；只是采用一个我们觉得对Python
使用者最好的标准（这是Biopython，毕竟）。</p>
<p>在使用 <tt class="docutils literal">Bio.SearchIO</tt> 时你可以认为有个三个隐含的标准：</p>
<ul class="simple">
<li>第一个适用于序列坐标。在 <tt class="docutils literal">Bio.SearchIO</tt> 模块中，所有序列坐标遵循Python
的坐标风格：
从0开始，半开区间。例如，在一个BLAST XML输出文件中，HSP的起始和结束坐标
是10和28，它们在 <tt class="docutils literal">Bio.SearchIO</tt> 中将变成9和28。起始坐标变成9因为Python
中索引是从0开始，而结束坐标仍然是28因为Python索引删除了区间中最后一个
项目。</li>
<li>第二个是关于序列坐标顺序。在 <tt class="docutils literal">Bio.SearchIO</tt> 中，开始坐标总是小于或
等于结束坐标。但是这不是在所有的序列搜索工具中都始终适用。因为当序列
为负链时，起始坐标会更大一些。</li>
<li>最后一个标准是关于链和阅读框的值。对于链值，只有四个可选值： <tt class="docutils literal">1</tt> (正链)， <tt class="docutils literal"><span class="pre">-1</span></tt> (负链)， <tt class="docutils literal">0</tt> (蛋白序列)， 和 <tt class="docutils literal">None</tt> (无链)。对于阅读框，
可选值是从 <tt class="docutils literal"><span class="pre">-3</span></tt> 至 <tt class="docutils literal">3</tt> 的整型以及 <tt class="docutils literal">None</tt> 。</li>
</ul>
<p>注意，这些标准只是存在于 <tt class="docutils literal">Bio.SearchIO</tt> 对象中。如果你把 <tt class="docutils literal">Bio.SearchIO</tt>
对象写入一种输出格式， <tt class="docutils literal">Bio.SearchIO</tt> 会使用该格式的标准来输出。它并不
强加它的标准到你的输出文件。</p>
</div>
<div class="section" id="id76">
<h2>8.3  读取搜索输出文件</h2>
<p>有两个方法，你可以用来读取搜索输出文件到 <tt class="docutils literal">Bio.SearchIO</tt> 对象： <tt class="docutils literal">read</tt> 和 <tt class="docutils literal">parse</tt>。
它们和其他亚模块如 <tt class="docutils literal">Bio.SeqIO</tt> 或 <tt class="docutils literal">Bio.AlignIO</tt> 中的 <tt class="docutils literal">read</tt> 和 <tt class="docutils literal">parse</tt> 方法在
本质上是相似的。你都需要提供搜索输出文件名和文件格式名，都是Python字符串类型。你可以
查阅文档来获得 <tt class="docutils literal">Bio.SearchIO</tt> 可以识别的格式清单。</p>
<p><tt class="docutils literal">Bio.SearchIO.read</tt> 用于读取单query的搜索输出文件并且返回一个 <tt class="docutils literal">QueryResult</tt> 对象。你在前面的例子中已经看到过 <tt class="docutils literal">read</tt> 的使用了。
你没看到的是， <tt class="docutils literal">read</tt> 同样接受额外的关键字参数，取决于文件的格式。</p>
<p>这里有一些例子。在第一个例子中，我们和前面一样用 <tt class="docutils literal">read</tt> 读BLAST表格输出
文件。在第二个例子中，我们用一个关键字来修饰，所以它分析带有注释的BLAST
表格变量。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresult</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_003.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresult</span>
<span class="name">QueryResult</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">hits</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresult2</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_007.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">,</span> <span class="name">comments</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresult2</span>
<span class="name">QueryResult</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">hits</span><span class="punctuation">)</span>
</pre>
<p>这些关键字在不同的文件格式中是不一样的。查看格式文档，看看它是否有关键字参数来控制它的分析器行为。</p>
<p>对于 <tt class="docutils literal">Bio.SearchIO.parse</tt>，是用来读取含有任意数量query的搜索输出文件。
这个方法返回一个generator对象，在每次迭代中yield一个 <tt class="docutils literal">QueryResult</tt> 对象。
和 <tt class="docutils literal">Bio.SearchIO.read</tt> 一样，它同样接受格式特异的关键字参数：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresults</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_001.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">qresult</span> <span class="operator word">in</span> <span class="name">qresults</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">qresult</span><span class="operator">.</span><span class="name">id</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">16080617</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NP_391444</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">11464971</span><span class="punctuation">:</span><span class="literal number integer">4</span><span class="operator">-</span><span class="literal number integer">101</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresults2</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_005.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">,</span> <span class="name">comments</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">qresult</span> <span class="operator word">in</span> <span class="name">qresults2</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">qresult</span><span class="operator">.</span><span class="name">id</span>
<span class="name">random_s00</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">16080617</span><span class="operator">|</span><span class="name">ref</span><span class="operator">|</span><span class="name">NP_391444</span><span class="operator">.</span><span class="literal number integer">1</span><span class="operator">|</span>
<span class="name">gi</span><span class="operator">|</span><span class="literal number integer">11464971</span><span class="punctuation">:</span><span class="literal number integer">4</span><span class="operator">-</span><span class="literal number integer">101</span>
</pre>
</div>
<div class="section" id="id77">
<h2>8.4  用索引处理含有大量搜索输出的文件</h2>
<p>有时，你得到了一个包含成百上千个query的搜索输出文件要分析，你当然可以使用
<tt class="docutils literal">Bio.SearchIO.parse</tt> 来处理，但是如果你仅仅需要访问少数query的话，效率
是及其低下的。这是因为 <tt class="docutils literal">parse</tt> 会分析所有的query，直到找到你感兴趣。</p>
<p>在这种情况下，理想的选择是用 <tt class="docutils literal">Bio.SearchIO.index</tt> 或 <tt class="docutils literal">Bio.SearchIO.index_db</tt>
来索引文件。如果名字听起来很熟悉，是因为你之前已经见过了，在
Section <a class="reference external" href="#sec:SeqIO-index">5.4.2</a>。这些方法和 <tt class="docutils literal">Bio.SeqIO</tt>
中相应的方法行为很相似，只是多了些格式特异的关键字参数。</p>
<p>这里有一些例子。你可以只用文件名和格式名来 <tt class="docutils literal">index</tt></p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idx</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_001.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">sorted</span><span class="punctuation">(</span><span class="name">idx</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">())</span>
<span class="punctuation">[</span><span class="literal string">'gi|11464971:4-101'</span><span class="punctuation">,</span> <span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idx</span><span class="punctuation">[</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">]</span>
<span class="name">QueryResult</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">hits</span><span class="punctuation">)</span>
</pre>
<p>或者依旧使用格式特异的关键字参数：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idx</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_005.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">,</span> <span class="name">comments</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">sorted</span><span class="punctuation">(</span><span class="name">idx</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">())</span>
<span class="punctuation">[</span><span class="literal string">'gi|11464971:4-101'</span><span class="punctuation">,</span> <span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">,</span> <span class="literal string">'random_s00'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idx</span><span class="punctuation">[</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">]</span>
<span class="name">QueryResult</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">hits</span><span class="punctuation">)</span>
</pre>
<p>或者使用 <tt class="docutils literal">key_function</tt> 参数，和 <tt class="docutils literal">Bio.SeqIO</tt> 中一样：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">key_function</span> <span class="operator">=</span> <span class="keyword">lambda</span> <span class="name builtin">id</span><span class="punctuation">:</span> <span class="name builtin">id</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">()</span>    <span class="comment"># capitalizes the keys</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idx</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">'tab_2226_tblastn_001.txt'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">,</span> <span class="name">key_function</span><span class="operator">=</span><span class="name">key_function</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">sorted</span><span class="punctuation">(</span><span class="name">idx</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">())</span>
<span class="punctuation">[</span><span class="literal string">'GI|11464971:4-101'</span><span class="punctuation">,</span> <span class="literal string">'GI|16080617|REF|NP_391444.1|'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idx</span><span class="punctuation">[</span><span class="literal string">'GI|16080617|REF|NP_391444.1|'</span><span class="punctuation">]</span>
<span class="name">QueryResult</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">'gi|16080617|ref|NP_391444.1|'</span><span class="punctuation">,</span> <span class="literal number integer">3</span> <span class="name">hits</span><span class="punctuation">)</span>
</pre>
<p><tt class="docutils literal">Bio.SearchIO.index_db</tt> 和 <tt class="docutils literal">index</tt> 作用差不多，不同的只是它把query
偏移量写入一个SQLite数据库文件中。</p>
</div>
<div class="section" id="id78">
<h2>8.5  写入和转换搜索输出文件</h2>
<p>有时候，读取一个搜索输出文件，作些调整并写到一个新的文件是很有用的。
<tt class="docutils literal">Bio.SearchIO</tt> 提供了一个 <tt class="docutils literal">write</tt> 方法，让你可以准确地完成这种工作。
它需要的参数是：一个可迭代返回 <tt class="docutils literal">QueryResult</tt> 的对象，输出文件名，输出文件
格式和一些可选的格式特异的关键字参数。它返回一个4项目的元组，分别代表
被写入的 <tt class="docutils literal">QueryResult</tt>， <tt class="docutils literal">Hit</tt>， <tt class="docutils literal">HSP</tt>， 和 <tt class="docutils literal">HSPFragment</tt> 对象的数量。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">qresults</span> <span class="operator">=</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">'mirna.xml'</span><span class="punctuation">,</span> <span class="literal string">'blast-xml'</span><span class="punctuation">)</span>     <span class="comment"># read XML file</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">qresults</span><span class="punctuation">,</span> <span class="literal string">'results.tab'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">)</span>    <span class="comment"># write to tabular file</span>
<span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">239</span><span class="punctuation">,</span> <span class="literal number integer">277</span><span class="punctuation">,</span> <span class="literal number integer">277</span><span class="punctuation">)</span>
</pre>
<p>你应该注意，不同的文件格式需要 <tt class="docutils literal">QueryResult</tt>， <tt class="docutils literal">Hit</tt>， <tt class="docutils literal">HSP</tt> 和
<tt class="docutils literal">HSPFragment</tt> 对象的不同属性。如果这些属性不存在，那么将不能写入。
也就是，你想写入的格式可能有时也会失效。举个例子，如果你读取一个BLASTXML文件，
你就不能将结果写入PSL文件，因为PSL文件需要一些属性，而这些属性BLAST却不能
提供（如重复匹配的数量）。如果你确实想写到PSL，可以手工设置这些属性。</p>
<p>和 <tt class="docutils literal">read</tt>， <tt class="docutils literal">parse</tt>， <tt class="docutils literal">index</tt> 和 <tt class="docutils literal">index_db</tt> 相似， <tt class="docutils literal">write</tt> 同
样接受格式特异的关键字参数。查阅文档获得 <tt class="docutils literal">Bio.SearchIO</tt> 可写格式和这些
格式的参数的完整清单。</p>
<p>最后， <tt class="docutils literal">Bio.SearchIO</tt> 同样提供一个 <tt class="docutils literal">convert</tt> 方法，可以理解为 <tt class="docutils literal">Bio.SearchIO.parse</tt> 和 <tt class="docutils literal">Bio.SearchIO.write</tt> 的简单替代方法。使用 <tt class="docutils literal">convert</tt> 方法的例子如下：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SearchIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SearchIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">'mirna.xml'</span><span class="punctuation">,</span> <span class="literal string">'blast-xml'</span><span class="punctuation">,</span> <span class="literal string">'results.tab'</span><span class="punctuation">,</span> <span class="literal string">'blast-tab'</span><span class="punctuation">)</span>
<span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">239</span><span class="punctuation">,</span> <span class="literal number integer">277</span><span class="punctuation">,</span> <span class="literal number integer">277</span><span class="punctuation">)</span>
</pre>
<p>因为 <tt class="docutils literal">convert</tt> 使用 <tt class="docutils literal">write</tt> 方法，所以只有所有需要的属性都存在时，格式
转换才能正常工作。这里由于BLAST XML文件提供BLAST 表格文件所需的所有默认值，
格式转换才能正常完成。但是，其他格式转换就可能不会正常工作，因为你需要先手工指定所需的属性。</p>
</div>
</div>
<div class="section" id="ncbi-entrez">
<h1>第9章  访问NCBI Entrez数据库</h1>
<p>Entrez
(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/Entrez">http://www.ncbi.nlm.nih.gov/Entrez</a>)
是一个给客户提供NCBI各个数据库（如PubMed, GeneBank, GEO等等）访问的检索系统。
用户可以通过浏览器手动输入查询条目访问Entrez，也可以使用Biopython的 <tt class="docutils literal">Bio.Entrez</tt> 模块以编程方式访问来访问Entrez。
如果使用第二种方法，用户用一个Python脚本就可以实现在PubMed里面搜索或者从GenBank下载数据。</p>
<p><tt class="docutils literal">Bio.Entrez</tt> 模块利用了Entrez Programming Utilities（也称作EUtils），包含八个工具，详情请见NCBI的网站：
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/utils/">http://www.ncbi.nlm.nih.gov/entrez/utils/</a>.
每个工具都能在Python的 <tt class="docutils literal">Bio.Entrez</tt> 模块中找到对应函数，后面会详细讲到。这个模块可以保证用来查询的URL
的正确性，并且向NCBI要求的一样，每三秒钟查询的次数不超过一。</p>
<p>EUtils返回的输出结果通常是XML格式，我们有以下不同的方法来解析这种类型的输入文件：</p>
<ol class="arabic simple">
<li>使用 <tt class="docutils literal">Bio.Entrez</tt>解析器将XML输出的解析成Python对象;</li>
<li>使用Python标准库中的DOM (Document Object Model)解析器;</li>
<li>使用Python标准库中的SAX (Simple API for XML)解析器;</li>
<li>把XML输出当做原始的文本文件，通过字符串查找和处理来进行解析；</li>
</ol>
<p>对于DOM和SAX解析器，可以查看Python的文档. <tt class="docutils literal">Bio.Entrez</tt> 中使用到的解析器将会在下面讨论.</p>
<p>NCBI使用DTD (Document Type Definition)文件来描述XML文件中所包含信息的结构. 大多数NCBI使用的DTD文件
格式都包含在了Biopython发行包里。当NCBI Entrez读入一个XML格式的文件的时候，<tt class="docutils literal">Bio.Entrez</tt>
将会使用DTD文件。</p>
<p>有时候，你可能会发现与某种特殊的XML相关的DTD文件在Biopython发行包里面不存在。当NCBI升级它的
DTD文件的时候，这种情况可能发生。如果发生这种情况，<tt class="docutils literal">Entrez.read</tt> 将会显示丢失的DTD文件名字和URL的
警示信息。解析器会通过互联网获取缺失的DTD文件，让XML的分析继续正常进行。如果本地存在对应的DTD文件的
话，处理起来会更快。因此，为了更快的处理，我们可以通过警示信息里面的URL来下载对应的DTD文件，将文件放在DTD
文件默认存放的文件夹 <tt class="docutils literal"><span class="pre">...site-packages/Bio/Entrez/DTDs</span></tt> 。如果你没有权限进入这个文件夹，你也可以把
DTD文件放到 <tt class="docutils literal"><span class="pre">~/.biopython/Bio/Entrez/DTDs</span></tt> 这个目录，<tt class="docutils literal">~</tt> 表示的是你的Home目录。因为这个目录会先于
<tt class="docutils literal"><span class="pre">...site-packages/Bio/Entrez/DTDs</span></tt> 被解析器读取，所以当 <tt class="docutils literal"><span class="pre">...site-packages/Bio/Entrez/DTDs</span></tt>
下面的DTD文件过时的时候，你也可以将最新版本的DTD文件放到Home目录的那个文件夹下面。当然也有其他方案，如果你
是通过源码来安装的Biopython，你可以将DTD文件放到源码的 <tt class="docutils literal">Bio/Entrez/DTDs</tt> 文件夹下，然后重新安装Biopython。
这样会将新的DTD文件和之前的一样地安装到正确的位置。</p>
<p>Entrez Programming Utilities也可以生成其他格式的输出文件，比如Fasta、序列数据库里面的GenBank文件格式
或者文献数据库里面的MedLine格式，更多内容将会在 <a class="reference external" href="#sec:entrez-specialized-parsers">9.12</a> 中讨论。</p>
<div class="section" id="entrez">
<h2>9.1  Entrez 简介</h2>
<p>在我们通过Biopython访问NCBI的线上资源（通过 <tt class="docutils literal">Bio.Entrez</tt> 或者其他模块）的时候，请先阅读 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/books/NBK25497/#chapter2.Usage_Guidelines_and_Requiremen">NCBI的Entrez
用户规范</a>.
如果NCBI发现你在滥用他们的系统，他们会禁止你的访问。</p>
<p>详细规范如下：</p>
<ul>
<li><p class="first">对任何连续超过100次的访问请求，请在周末时间或者避开美国的使用高峰时间。这个取决于你是否遵从。</p>
</li>
<li><p class="first">使用这个网址 <a class="reference external" href="http://eutils.ncbi.nlm.nih.gov">http://eutils.ncbi.nlm.nih.gov</a> ，
而不是通常的NCBI网址。Biopython使用的是这个网址。</p>
</li>
<li><p class="first">每秒钟不要超过三次请求（比2009年年初的每三秒钟最多一次请求要宽松）。这个由Biopython自动强制实行。</p>
</li>
<li><p class="first">使用email参数，这样如果遇到什么问题，NCBI可以通过邮件联系到你。你可以在每次请求Entrez的时候明确的设置
这个参数（例如，在参数列表中包含 <tt class="docutils literal"><span class="pre">email=&quot;A.N.Other&#64;example.com&quot;</span></tt> ），或者你也可以设置一个全局的email
地址：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>
</pre>
<p><tt class="docutils literal">Bio.Entrez</tt> 将会在每次向Entrez请求的时候使用这个邮件地址。请千万不要胡乱的填写邮件地址，不填写都比
这要好。邮件的参数从2010年6月1日将是强制的参数。在过度使用的情况下，NCBI会在封锁用户访问E-utilities之前尝试通过
用户提供的邮件地址联系。</p>
</li>
<li><p class="first">如果你是在一个大的软件包里面使用Biopython的，请通过tool这个参数明确说明。你既可以在每次请求访问Entrez
的时候通过参数明确地指明使用的工具（例如，在参数列表中包含 <tt class="docutils literal"><span class="pre">tool=&quot;MyLocalScript&quot;</span></tt> ），或者你也可以
设置一个全局的tool名称：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">tool</span> <span class="operator">=</span> <span class="literal string">&quot;MyLocalScript&quot;</span>
</pre>
<p>默认的tool名称是Biopython。</p>
</li>
<li><p class="first">对于大规模的查询请求，NCBI也推荐使用他们的会话历史特性（ WebEnv会话cookie字符串，见
章节 <a class="reference external" href="#sec:entrez-webenv">9.15</a> ）。 只是这个稍微有点复杂。</p>
</li>
</ul>
<p>最后，根据你的使用情况选择不同的策略。如果你打算下载大量的数据，最好使用其他的方法。比如，你想得到所有人的
基因的数据，那么考虑通过FTP得到每个染色体的GenBank文件，然后将这些文件导入到你自己的BioSQL数据库里面去。
(请见Section <a class="reference external" href="#sec:BioSQL">18.5</a>).</p>
</div>
<div class="section" id="einfo-entrez">
<h2>9.2  EInfo: 获取Entrez数据库的信息</h2>
<p>EInfo为每个NCBI的数据库提供了条目索引，最近更新的时间以及可用的链接。此外，你可以很容易的使用EInfo通过
Entrez获取所有数据库名字的列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">einfo</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
</pre>
<p>变量 <tt class="docutils literal">result</tt> 现在包含了XML格式的数据库列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">result</span>
<span class="operator">&lt;</span><span class="error">?</span><span class="name">xml</span> <span class="name">version</span><span class="operator">=</span><span class="literal string">&quot;1.0&quot;</span><span class="error">?</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="error">!</span><span class="name">DOCTYPE</span> <span class="name">eInfoResult</span> <span class="name">PUBLIC</span> <span class="literal string">&quot;-//NLM//DTD eInfoResult, 11 May 2002//EN&quot;</span>
 <span class="literal string">&quot;http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd&quot;</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="name">eInfoResult</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="name">DbList</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pubmed</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">protein</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nucleotide</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nuccore</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nucgss</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nucest</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">structure</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">genome</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">books</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">cancerchromosomes</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">cdd</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">gap</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">domains</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">gene</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">genomeprj</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">gensat</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">geo</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">gds</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">homologene</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">journals</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">mesh</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">ncbisearch</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nlmcatalog</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">omia</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">omim</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pmc</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">popset</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">probe</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">proteinclusters</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pcassay</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pccompound</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pcsubstance</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">snp</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">taxonomy</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">toolkit</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">unigene</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">unists</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
<span class="operator">&lt;/</span><span class="name">DbList</span><span class="operator">&gt;</span>
<span class="operator">&lt;/</span><span class="name">eInfoResult</span><span class="operator">&gt;</span>
</pre>
<p>因为这是一个相当简单的XML文件，我们可以简单的通过字符串查找提取里面所包含的信息。使用 <tt class="docutils literal">Bio.Entrez</tt> 的解析器，
我们可以直接将这个XML读入到一个Python对象里面去：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">einfo</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>现在 <tt class="docutils literal">record</tt> 是拥有一个确定键值的字典：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">u'DbList'</span><span class="punctuation">]</span>
</pre>
<p>这个键对应的值存储了上面XML文件里面包含的数据库名字的列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DbList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'pubmed'</span><span class="punctuation">,</span> <span class="literal string">'protein'</span><span class="punctuation">,</span> <span class="literal string">'nucleotide'</span><span class="punctuation">,</span> <span class="literal string">'nuccore'</span><span class="punctuation">,</span> <span class="literal string">'nucgss'</span><span class="punctuation">,</span> <span class="literal string">'nucest'</span><span class="punctuation">,</span>
 <span class="literal string">'structure'</span><span class="punctuation">,</span> <span class="literal string">'genome'</span><span class="punctuation">,</span> <span class="literal string">'books'</span><span class="punctuation">,</span> <span class="literal string">'cancerchromosomes'</span><span class="punctuation">,</span> <span class="literal string">'cdd'</span><span class="punctuation">,</span> <span class="literal string">'gap'</span><span class="punctuation">,</span>
 <span class="literal string">'domains'</span><span class="punctuation">,</span> <span class="literal string">'gene'</span><span class="punctuation">,</span> <span class="literal string">'genomeprj'</span><span class="punctuation">,</span> <span class="literal string">'gensat'</span><span class="punctuation">,</span> <span class="literal string">'geo'</span><span class="punctuation">,</span> <span class="literal string">'gds'</span><span class="punctuation">,</span> <span class="literal string">'homologene'</span><span class="punctuation">,</span>
 <span class="literal string">'journals'</span><span class="punctuation">,</span> <span class="literal string">'mesh'</span><span class="punctuation">,</span> <span class="literal string">'ncbisearch'</span><span class="punctuation">,</span> <span class="literal string">'nlmcatalog'</span><span class="punctuation">,</span> <span class="literal string">'omia'</span><span class="punctuation">,</span> <span class="literal string">'omim'</span><span class="punctuation">,</span> <span class="literal string">'pmc'</span><span class="punctuation">,</span>
 <span class="literal string">'popset'</span><span class="punctuation">,</span> <span class="literal string">'probe'</span><span class="punctuation">,</span> <span class="literal string">'proteinclusters'</span><span class="punctuation">,</span> <span class="literal string">'pcassay'</span><span class="punctuation">,</span> <span class="literal string">'pccompound'</span><span class="punctuation">,</span>
 <span class="literal string">'pcsubstance'</span><span class="punctuation">,</span> <span class="literal string">'snp'</span><span class="punctuation">,</span> <span class="literal string">'taxonomy'</span><span class="punctuation">,</span> <span class="literal string">'toolkit'</span><span class="punctuation">,</span> <span class="literal string">'unigene'</span><span class="punctuation">,</span> <span class="literal string">'unists'</span><span class="punctuation">]</span>
</pre>
<p>对于这些数据库，我们可以使用EInfo获得更多的信息：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">einfo</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DbInfo&quot;</span><span class="punctuation">][</span><span class="literal string">&quot;Description&quot;</span><span class="punctuation">]</span>
<span class="literal string">'PubMed bibliographic record'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DbInfo&quot;</span><span class="punctuation">][</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal string">'17989604'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DbInfo&quot;</span><span class="punctuation">][</span><span class="literal string">&quot;LastUpdate&quot;</span><span class="punctuation">]</span>
<span class="literal string">'2008/05/24 06:45'</span>
</pre>
<p>通过 <tt class="docutils literal"><span class="pre">record[&quot;DbInfo&quot;].keys()</span></tt> 可以获取存储在这个记录里面的其他信息。这里面最有用的信息之一是一个ESearch可用的
搜索值列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">field</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DbInfo&quot;</span><span class="punctuation">][</span><span class="literal string">&quot;FieldList&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%(Name)s</span><span class="literal string">, </span><span class="literal string interpol">%(FullName)s</span><span class="literal string">, </span><span class="literal string interpol">%(Description)s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">field</span>
<span class="name">ALL</span><span class="punctuation">,</span> <span class="name">All</span> <span class="name">Fields</span><span class="punctuation">,</span> <span class="name">All</span> <span class="name">terms</span> <span class="keyword namespace">from</span> <span class="name namespace">all</span> <span class="name namespace">searchable</span> <span class="name namespace">fields</span>
<span class="name">UID</span><span class="punctuation">,</span> <span class="name">UID</span><span class="punctuation">,</span> <span class="name">Unique</span> <span class="name">number</span> <span class="name">assigned</span> <span class="name">to</span> <span class="name">publication</span>
<span class="name">FILT</span><span class="punctuation">,</span> <span class="name">Filter</span><span class="punctuation">,</span> <span class="name">Limits</span> <span class="name">the</span> <span class="name">records</span>
<span class="name">TITL</span><span class="punctuation">,</span> <span class="name">Title</span><span class="punctuation">,</span> <span class="name">Words</span> <span class="operator word">in</span> <span class="name">title</span> <span class="name">of</span> <span class="name">publication</span>
<span class="name">WORD</span><span class="punctuation">,</span> <span class="name">Text</span> <span class="name">Word</span><span class="punctuation">,</span> <span class="name">Free</span> <span class="name">text</span> <span class="name">associated</span> <span class="keyword">with</span> <span class="name">publication</span>
<span class="name">MESH</span><span class="punctuation">,</span> <span class="name">MeSH</span> <span class="name">Terms</span><span class="punctuation">,</span> <span class="name">Medical</span> <span class="name">Subject</span> <span class="name">Headings</span> <span class="name">assigned</span> <span class="name">to</span> <span class="name">publication</span>
<span class="name">MAJR</span><span class="punctuation">,</span> <span class="name">MeSH</span> <span class="name">Major</span> <span class="name">Topic</span><span class="punctuation">,</span> <span class="name">MeSH</span> <span class="name">terms</span> <span class="name">of</span> <span class="name">major</span> <span class="name">importance</span> <span class="name">to</span> <span class="name">publication</span>
<span class="name">AUTH</span><span class="punctuation">,</span> <span class="name">Author</span><span class="punctuation">,</span> <span class="name">Author</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span> <span class="name">of</span> <span class="name">publication</span>
<span class="name">JOUR</span><span class="punctuation">,</span> <span class="name">Journal</span><span class="punctuation">,</span> <span class="name">Journal</span> <span class="name">abbreviation</span> <span class="name">of</span> <span class="name">publication</span>
<span class="name">AFFL</span><span class="punctuation">,</span> <span class="name">Affiliation</span><span class="punctuation">,</span> <span class="name">Author</span><span class="literal string">'s institutional affiliation and address</span>
<span class="operator">...</span>
</pre>
<p>这是一个很长的列表，但是间接的告诉你在使用PubMed的时候，你可以通过 <tt class="docutils literal">Jones[AUTH]</tt> 搜索作者，或者通过
<tt class="docutils literal">Sanger[AFFL]</tt> 将作者范围限制在Sanger Centre。这个会非常方便，特别是在你对某个数据库不太熟悉的时候。</p>
</div>
<div class="section" id="esearch-entrez">
<h2>9.3  ESearch: 搜索Entrez数据库</h2>
<p>我们可以使用 <tt class="docutils literal">Bio.Entrez.esearch()</tt> 来搜索任意的数据库。例如，我们在PubMed中搜索跟Biopython相关的文献：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;biopython&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'19304878'</span><span class="punctuation">,</span> <span class="literal string">'18606172'</span><span class="punctuation">,</span> <span class="literal string">'16403221'</span><span class="punctuation">,</span> <span class="literal string">'16377612'</span><span class="punctuation">,</span> <span class="literal string">'14871861'</span><span class="punctuation">,</span> <span class="literal string">'14630660'</span><span class="punctuation">,</span> <span class="literal string">'12230038'</span><span class="punctuation">]</span>
</pre>
<p>在输出的结果中，我们可以看到七个PubMed IDs（包括19304878，这个是Biopython应用笔记的PMID），你可以通过
EFetch来获取这些文献（请见章节 <a class="reference external" href="#sec:efetch">9.6</a> ）。</p>
<p>你也可以通过ESearch来搜索GenBank。我们将以在*Cypripedioideae* orchids中搜索*matK*基因为例，快速展示
一下（请见章节 <a class="reference external" href="#sec:entrez-einfo">9.2</a> 关于EInfo：一种查明你可以在哪个Entrez数据库中搜索的方法）。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Cypripedioideae[Orgn] AND matK[Gene]&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal string">'25'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'126789333'</span><span class="punctuation">,</span> <span class="literal string">'37222967'</span><span class="punctuation">,</span> <span class="literal string">'37222966'</span><span class="punctuation">,</span> <span class="literal string">'37222965'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'61585492'</span><span class="punctuation">]</span>
</pre>
<p>每个IDs(126789333, 37222967, 37222966, …)是GenBank的一个标识。请见Section <a class="reference external" href="#sec:efetch">9.6</a>
此章包含了怎样下载这些GenBank的记录的信息。</p>
<p>注意，不是像 <tt class="docutils literal">Cypripedioideae[Orgn]</tt> 这样在搜索的时候加上特定的物种名字，而是需要在搜索的时候使用NCBI的
taxon ID，像 <tt class="docutils literal">txid158330[Orgn]</tt> 这样。这个并没有记录在ESearch的帮助页面上，NCBI通过邮件回复解释了这个
问题。你可以通过经常和Entrez的网站接口互动，来推断搜索条目的格式。例如，在基因组搜索的时候加上 <tt class="docutils literal">complete[prop]</tt>
可以把结果限制在完成的基因组上。</p>
<p>作为最后一个例子，让我们获取一个computational journal名字的列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;journals&quot;</span><span class="punctuation">,</span> <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;computational&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal string">'16'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'30367'</span><span class="punctuation">,</span> <span class="literal string">'33843'</span><span class="punctuation">,</span> <span class="literal string">'33823'</span><span class="punctuation">,</span> <span class="literal string">'32989'</span><span class="punctuation">,</span> <span class="literal string">'33190'</span><span class="punctuation">,</span> <span class="literal string">'33009'</span><span class="punctuation">,</span> <span class="literal string">'31986'</span><span class="punctuation">,</span>
 <span class="literal string">'34502'</span><span class="punctuation">,</span> <span class="literal string">'8799'</span><span class="punctuation">,</span> <span class="literal string">'22857'</span><span class="punctuation">,</span> <span class="literal string">'32675'</span><span class="punctuation">,</span> <span class="literal string">'20258'</span><span class="punctuation">,</span> <span class="literal string">'33859'</span><span class="punctuation">,</span> <span class="literal string">'32534'</span><span class="punctuation">,</span>
 <span class="literal string">'32357'</span><span class="punctuation">,</span> <span class="literal string">'32249'</span><span class="punctuation">]</span>
</pre>
<p>同样，我们可以通过EFetch来获得关于每个journal IDs更多的消息。</p>
<p>ESearch有很多有用的参数——参见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/esearch_help.html">ESearch 帮助页面</a>
来获取更多信息.</p>
</div>
<div class="section" id="epost-identifiers">
<h2>9.4  EPost: 上传identifiers的列表</h2>
<p>EPost上传在后续搜索中将会用到的IDs的列表，参见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/epost_help.html">EPost 帮助页面</a>
来获取更多信息. 通过 <tt class="docutils literal">Bio.Entrez.epost()</tt> 函数可以在Biopython中实现。</p>
<p>为了举一个关于此用法的例子，假设你有一个想通过EFetch下载的IDs的长长的列表（可能是序列，也有可能是引用的
其他内容）。当你通过EFetch发出下载请求的时候，你的IDs列表、数据库等，将会被转变成一个长的URL，然后被发送
到服务器。如果IDs列表很长，URL也会很长，长的URL可能会断掉（比如，一些代理不能复制全部的内容）。</p>
<p>另外，你也可以把以上分成两步来完成，首先用EPost来上传IDs的列表（这个使用了一个内部的 “HTML post” ，而不是
“HTML get” ， 避开了long URL可能产生的问题）。由于历史记录的支持，你可以使用EFetch来指向这个长的IDs列表，
并且下载相关的数据。</p>
<p>让我们通过下面一个简单的例子来看看EPost是如何工作的——上传了一些PubMed的IDs：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">id_list</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">&quot;19304878&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;18606172&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;16403221&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;16377612&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;14871861&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;14630660&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">epost</span><span class="punctuation">(</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;,&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">id_list</span><span class="punctuation">))</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&lt;</span><span class="error">?</span><span class="name">xml</span> <span class="name">version</span><span class="operator">=</span><span class="literal string">&quot;1.0&quot;</span><span class="error">?</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="error">!</span><span class="name">DOCTYPE</span> <span class="name">ePostResult</span> <span class="name">PUBLIC</span> <span class="literal string">&quot;-//NLM//DTD ePostResult, 11 May 2002//EN&quot;</span>
 <span class="literal string">&quot;http://www.ncbi.nlm.nih.gov/entrez/query/DTD/ePost_020511.dtd&quot;</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="name">ePostResult</span><span class="operator">&gt;</span>
 <span class="operator">&lt;</span><span class="name">QueryKey</span><span class="operator">&gt;</span><span class="literal number integer">1</span><span class="operator">&lt;/</span><span class="name">QueryKey</span><span class="operator">&gt;</span>
 <span class="operator">&lt;</span><span class="name">WebEnv</span><span class="operator">&gt;</span><span class="name">NCID_01_206841095_130</span><span class="operator">.</span><span class="literal number float">14.22</span><span class="operator">.</span><span class="literal number integer">101</span><span class="name">_9001_1242061629</span><span class="operator">&lt;/</span><span class="name">WebEnv</span><span class="operator">&gt;</span>
<span class="operator">&lt;/</span><span class="name">ePostResult</span><span class="operator">&gt;</span>
</pre>
<p>返回的XML包含了两个重要的字符串， <tt class="docutils literal">QueryKey</tt> 和 <tt class="docutils literal">WebEnv</tt> ，两个字符串一起确定了之前的历史记录。你可以
使用其他的Entrez工具，例如EFetch，来提取这些值：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">id_list</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">&quot;19304878&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;18606172&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;16403221&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;16377612&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;14871861&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;14630660&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">search_results</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">epost</span><span class="punctuation">(</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;,&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">id_list</span><span class="punctuation">)))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">webenv</span> <span class="operator">=</span> <span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;WebEnv&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">query_key</span> <span class="operator">=</span> <span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;QueryKey&quot;</span><span class="punctuation">]</span>
</pre>
<p>第 <a class="reference external" href="#sec:entrez-webenv">9.15</a> 章节讲述了如何使用历史的特性。</p>
</div>
<div class="section" id="esummary-ids">
<h2>9.5  ESummary: 通过主要的IDs来获取摘要</h2>
<p>ESummary可以通过一个primary IDs来获取文章的摘要（参见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/esummary_help.html">ESummary 帮助页面</a>
来获取更多信息）。在Biopython中，ESummary以 <tt class="docutils literal">Bio.Entrez.esummary()</tt> 的形式出现。根据上面的搜索结果，
我们可以获得ID为30367杂志相关的更多信息：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esummary</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;journals&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;30367&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Id&quot;</span><span class="punctuation">]</span>
<span class="literal string">'30367'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Title&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Computational biology and chemistry'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Publisher&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Pergamon,'</span>
</pre>
</div>
<div class="section" id="efetch-entrez">
<h2>9.6  EFetch: 从Entrez下载更多的记录</h2>
<p>当你想要从Entrez中提取完整的记录的时候，你可以使用EFetch。 在 <a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/entrez/query/static/efetch_help.html">EFetch的帮助页面</a>
可以查到EFetch可以起作用的数据库。</p>
<p>NCBI大部分的数据库都支持多种不同的文件格式。当使用 <tt class="docutils literal">Bio.Entrez.efetch()</tt> 从Entrez下载特定的某种格式的时候，
需要 <tt class="docutils literal">rettype</tt> 和或者 <tt class="docutils literal">retmode</tt> 这些可选的参数。对于不同数据库类型不同的搭配在下面的网页中有描述：
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/efetch_help.html">NCBI efetch webpage</a>
(例如：
<a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/corehtml/query/static/efetchlit_help.html">literature</a>,
<a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/corehtml/query/static/efetchseq_help.html">sequences</a>
and
<a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/corehtml/query/static/efetchtax_help.html">taxonomy</a>).</p>
<p>一种常用的用法是下载FASTA或者GenBank/GenPept的文本格式 (接着可以使用 <tt class="docutils literal">Bio.SeqIO</tt> 来解析, 参见 <a class="reference external" href="#sec:SeqIO_GenBank_Online">5.3.1</a>
和 <a class="reference external" href="#sec:efetch">9.6</a> ）。从上面 <em>Cypripedioideae</em> 的例子,我们可以通过 <tt class="docutils literal">Bio.Entrez.efetch</tt>
从GenBank下载记录186972394。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;186972394&quot;</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="name">LOCUS</span>       <span class="name">EU490707</span>                <span class="literal number integer">1302</span> <span class="name">bp</span>    <span class="name">DNA</span>     <span class="name">linear</span>   <span class="name">PLN</span> <span class="literal number oct">05</span><span class="operator">-</span><span class="name">MAY</span><span class="operator">-</span><span class="literal number integer">2008</span>
<span class="name">DEFINITION</span>  <span class="name">Selenipedium</span> <span class="name">aequinoctiale</span> <span class="name">maturase</span> <span class="name">K</span> <span class="punctuation">(</span><span class="name">matK</span><span class="punctuation">)</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">partial</span> <span class="name">cds</span><span class="punctuation">;</span>
            <span class="name">chloroplast</span><span class="operator">.</span>
<span class="name">ACCESSION</span>   <span class="name">EU490707</span>
<span class="name">VERSION</span>     <span class="name">EU490707</span><span class="operator">.</span><span class="literal number integer">1</span>  <span class="name">GI</span><span class="punctuation">:</span><span class="literal number integer">186972394</span>
<span class="name">KEYWORDS</span>    <span class="operator">.</span>
<span class="name">SOURCE</span>      <span class="name">chloroplast</span> <span class="name">Selenipedium</span> <span class="name">aequinoctiale</span>
  <span class="name">ORGANISM</span>  <span class="name">Selenipedium</span> <span class="name">aequinoctiale</span>
            <span class="name">Eukaryota</span><span class="punctuation">;</span> <span class="name">Viridiplantae</span><span class="punctuation">;</span> <span class="name">Streptophyta</span><span class="punctuation">;</span> <span class="name">Embryophyta</span><span class="punctuation">;</span> <span class="name">Tracheophyta</span><span class="punctuation">;</span>
            <span class="name">Spermatophyta</span><span class="punctuation">;</span> <span class="name">Magnoliophyta</span><span class="punctuation">;</span> <span class="name">Liliopsida</span><span class="punctuation">;</span> <span class="name">Asparagales</span><span class="punctuation">;</span> <span class="name">Orchidaceae</span><span class="punctuation">;</span>
            <span class="name">Cypripedioideae</span><span class="punctuation">;</span> <span class="name">Selenipedium</span><span class="operator">.</span>
<span class="name">REFERENCE</span>   <span class="literal number integer">1</span>  <span class="punctuation">(</span><span class="name">bases</span> <span class="literal number integer">1</span> <span class="name">to</span> <span class="literal number integer">1302</span><span class="punctuation">)</span>
  <span class="name">AUTHORS</span>   <span class="name">Neubig</span><span class="punctuation">,</span><span class="name">K</span><span class="operator">.</span><span class="name">M</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Whitten</span><span class="punctuation">,</span><span class="name">W</span><span class="operator">.</span><span class="name">M</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Carlsward</span><span class="punctuation">,</span><span class="name">B</span><span class="operator">.</span><span class="name">S</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Blanco</span><span class="punctuation">,</span><span class="name">M</span><span class="operator">.</span><span class="name">A</span><span class="operator">.</span><span class="punctuation">,</span>
            <span class="name">Endara</span><span class="punctuation">,</span><span class="name">C</span><span class="operator">.</span><span class="name">L</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Williams</span><span class="punctuation">,</span><span class="name">N</span><span class="operator">.</span><span class="name">H</span><span class="operator">.</span> <span class="operator word">and</span> <span class="name">Moore</span><span class="punctuation">,</span><span class="name">M</span><span class="operator">.</span><span class="name">J</span><span class="operator">.</span>
  <span class="name">TITLE</span>     <span class="name">Phylogenetic</span> <span class="name">utility</span> <span class="name">of</span> <span class="name">ycf1</span> <span class="operator word">in</span> <span class="name">orchids</span>
  <span class="name">JOURNAL</span>   <span class="name">Unpublished</span>
<span class="name">REFERENCE</span>   <span class="literal number integer">2</span>  <span class="punctuation">(</span><span class="name">bases</span> <span class="literal number integer">1</span> <span class="name">to</span> <span class="literal number integer">1302</span><span class="punctuation">)</span>
  <span class="name">AUTHORS</span>   <span class="name">Neubig</span><span class="punctuation">,</span><span class="name">K</span><span class="operator">.</span><span class="name">M</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Whitten</span><span class="punctuation">,</span><span class="name">W</span><span class="operator">.</span><span class="name">M</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Carlsward</span><span class="punctuation">,</span><span class="name">B</span><span class="operator">.</span><span class="name">S</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Blanco</span><span class="punctuation">,</span><span class="name">M</span><span class="operator">.</span><span class="name">A</span><span class="operator">.</span><span class="punctuation">,</span>
            <span class="name">Endara</span><span class="punctuation">,</span><span class="name">C</span><span class="operator">.</span><span class="name">L</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Williams</span><span class="punctuation">,</span><span class="name">N</span><span class="operator">.</span><span class="name">H</span><span class="operator">.</span> <span class="operator word">and</span> <span class="name">Moore</span><span class="punctuation">,</span><span class="name">M</span><span class="operator">.</span><span class="name">J</span><span class="operator">.</span>
  <span class="name">TITLE</span>     <span class="name">Direct</span> <span class="name">Submission</span>
  <span class="name">JOURNAL</span>   <span class="name">Submitted</span> <span class="punctuation">(</span><span class="literal number integer">14</span><span class="operator">-</span><span class="name">FEB</span><span class="operator">-</span><span class="literal number integer">2008</span><span class="punctuation">)</span> <span class="name">Department</span> <span class="name">of</span> <span class="name">Botany</span><span class="punctuation">,</span> <span class="name">University</span> <span class="name">of</span>
            <span class="name">Florida</span><span class="punctuation">,</span> <span class="literal number integer">220</span> <span class="name">Bartram</span> <span class="name">Hall</span><span class="punctuation">,</span> <span class="name">Gainesville</span><span class="punctuation">,</span> <span class="name">FL</span> <span class="literal number integer">32611</span><span class="operator">-</span><span class="literal number integer">8526</span><span class="punctuation">,</span> <span class="name">USA</span>
<span class="name">FEATURES</span>             <span class="name">Location</span><span class="operator">/</span><span class="name">Qualifiers</span>
     <span class="name">source</span>          <span class="literal number float">1.</span><span class="operator">.</span><span class="literal number integer">1302</span>
                     <span class="operator">/</span><span class="name">organism</span><span class="operator">=</span><span class="literal string">&quot;Selenipedium aequinoctiale&quot;</span>
                     <span class="operator">/</span><span class="name">organelle</span><span class="operator">=</span><span class="literal string">&quot;plastid:chloroplast&quot;</span>
                     <span class="operator">/</span><span class="name">mol_type</span><span class="operator">=</span><span class="literal string">&quot;genomic DNA&quot;</span>
                     <span class="operator">/</span><span class="name">specimen_voucher</span><span class="operator">=</span><span class="literal string">&quot;FLAS:Blanco 2475&quot;</span>
                     <span class="operator">/</span><span class="name">db_xref</span><span class="operator">=</span><span class="literal string">&quot;taxon:256374&quot;</span>
     <span class="name">gene</span>            <span class="operator">&lt;</span><span class="literal number float">1.</span><span class="operator">.&gt;</span><span class="literal number integer">1302</span>
                     <span class="operator">/</span><span class="name">gene</span><span class="operator">=</span><span class="literal string">&quot;matK&quot;</span>
     <span class="name">CDS</span>             <span class="operator">&lt;</span><span class="literal number float">1.</span><span class="operator">.&gt;</span><span class="literal number integer">1302</span>
                     <span class="operator">/</span><span class="name">gene</span><span class="operator">=</span><span class="literal string">&quot;matK&quot;</span>
                     <span class="operator">/</span><span class="name">codon_start</span><span class="operator">=</span><span class="literal number integer">1</span>
                     <span class="operator">/</span><span class="name">transl_table</span><span class="operator">=</span><span class="literal number integer">11</span>
                     <span class="operator">/</span><span class="name">product</span><span class="operator">=</span><span class="literal string">&quot;maturase K&quot;</span>
                     <span class="operator">/</span><span class="name">protein_id</span><span class="operator">=</span><span class="literal string">&quot;ACC99456.1&quot;</span>
                     <span class="operator">/</span><span class="name">db_xref</span><span class="operator">=</span><span class="literal string">&quot;GI:186972395&quot;</span>
                     <span class="operator">/</span><span class="name">translation</span><span class="operator">=</span><span class="literal string">&quot;IFYEPVEIFGYDNKSSLVLVKRLITRMYQQNFLISSVNDSNQKG</span>
                     <span class="name">FWGHKHFFSSHFSSQMVSEGFGVILEIPFSSQLVSSLEEKKIPKYQNLRSIHSIFPFL</span>
                     <span class="name">EDKFLHLNYVSDLLIPHPIHLEILVQILQCRIKDVPSLHLLRLLFHEYHNLNSLITSK</span>
                     <span class="name">KFIYAFSKRKKRFLWLLYNSYVYECEYLFQFLRKQSSYLRSTSSGVFLERTHLYVKIE</span>
                     <span class="name">HLLVVCCNSFQRILCFLKDPFMHYVRYQGKAILASKGTLILMKKWKFHLVNFWQSYFH</span>
                     <span class="name">FWSQPYRIHIKQLSNYSFSFLGYFSSVLENHLVVRNQMLENSFIINLLTKKFDTIAPV</span>
                     <span class="name">ISLIGSLSKAQFCTVLGHPISKPIWTDFSDSDILDRFCRICRNLCRYHSGSSKKQVLY</span>
                     <span class="name">RIKYILRLSCARTLARKHKSTVRTFMRRLGSGLLEEFFMEEE</span><span class="literal string">&quot;</span>
<span class="name">ORIGIN</span>
        <span class="literal number integer">1</span> <span class="name">attttttacg</span> <span class="name">aacctgtgga</span> <span class="name">aatttttggt</span> <span class="name">tatgacaata</span> <span class="name">aatctagttt</span> <span class="name">agtacttgtg</span>
       <span class="literal number integer">61</span> <span class="name">aaacgtttaa</span> <span class="name">ttactcgaat</span> <span class="name">gtatcaacag</span> <span class="name">aattttttga</span> <span class="name">tttcttcggt</span> <span class="name">taatgattct</span>
      <span class="literal number integer">121</span> <span class="name">aaccaaaaag</span> <span class="name">gattttgggg</span> <span class="name">gcacaagcat</span> <span class="name">tttttttctt</span> <span class="name">ctcatttttc</span> <span class="name">ttctcaaatg</span>
      <span class="literal number integer">181</span> <span class="name">gtatcagaag</span> <span class="name">gttttggagt</span> <span class="name">cattctggaa</span> <span class="name">attccattct</span> <span class="name">cgtcgcaatt</span> <span class="name">agtatcttct</span>
      <span class="literal number integer">241</span> <span class="name">cttgaagaaa</span> <span class="name">aaaaaatacc</span> <span class="name">aaaatatcag</span> <span class="name">aatttacgat</span> <span class="name">ctattcattc</span> <span class="name">aatatttccc</span>
      <span class="literal number integer">301</span> <span class="name">tttttagaag</span> <span class="name">acaaattttt</span> <span class="name">acatttgaat</span> <span class="name">tatgtgtcag</span> <span class="name">atctactaat</span> <span class="name">accccatccc</span>
      <span class="literal number integer">361</span> <span class="name">atccatctgg</span> <span class="name">aaatcttggt</span> <span class="name">tcaaatcctt</span> <span class="name">caatgccgga</span> <span class="name">tcaaggatgt</span> <span class="name">tccttctttg</span>
      <span class="literal number integer">421</span> <span class="name">catttattgc</span> <span class="name">gattgctttt</span> <span class="name">ccacgaatat</span> <span class="name">cataatttga</span> <span class="name">atagtctcat</span> <span class="name">tacttcaaag</span>
      <span class="literal number integer">481</span> <span class="name">aaattcattt</span> <span class="name">acgccttttc</span> <span class="name">aaaaagaaag</span> <span class="name">aaaagattcc</span> <span class="name">tttggttact</span> <span class="name">atataattct</span>
      <span class="literal number integer">541</span> <span class="name">tatgtatatg</span> <span class="name">aatgcgaata</span> <span class="name">tctattccag</span> <span class="name">tttcttcgta</span> <span class="name">aacagtcttc</span> <span class="name">ttatttacga</span>
      <span class="literal number integer">601</span> <span class="name">tcaacatctt</span> <span class="name">ctggagtctt</span> <span class="name">tcttgagcga</span> <span class="name">acacatttat</span> <span class="name">atgtaaaaat</span> <span class="name">agaacatctt</span>
      <span class="literal number integer">661</span> <span class="name">ctagtagtgt</span> <span class="name">gttgtaattc</span> <span class="name">ttttcagagg</span> <span class="name">atcctatgct</span> <span class="name">ttctcaagga</span> <span class="name">tcctttcatg</span>
      <span class="literal number integer">721</span> <span class="name">cattatgttc</span> <span class="name">gatatcaagg</span> <span class="name">aaaagcaatt</span> <span class="name">ctggcttcaa</span> <span class="name">agggaactct</span> <span class="name">tattctgatg</span>
      <span class="literal number integer">781</span> <span class="name">aagaaatgga</span> <span class="name">aatttcatct</span> <span class="name">tgtgaatttt</span> <span class="name">tggcaatctt</span> <span class="name">attttcactt</span> <span class="name">ttggtctcaa</span>
      <span class="literal number integer">841</span> <span class="name">ccgtatagga</span> <span class="name">ttcatataaa</span> <span class="name">gcaattatcc</span> <span class="name">aactattcct</span> <span class="name">tctcttttct</span> <span class="name">ggggtatttt</span>
      <span class="literal number integer">901</span> <span class="name">tcaagtgtac</span> <span class="name">tagaaaatca</span> <span class="name">tttggtagta</span> <span class="name">agaaatcaaa</span> <span class="name">tgctagagaa</span> <span class="name">ttcatttata</span>
      <span class="literal number integer">961</span> <span class="name">ataaatcttc</span> <span class="name">tgactaagaa</span> <span class="name">attcgatacc</span> <span class="name">atagccccag</span> <span class="name">ttatttctct</span> <span class="name">tattggatca</span>
     <span class="literal number integer">1021</span> <span class="name">ttgtcgaaag</span> <span class="name">ctcaattttg</span> <span class="name">tactgtattg</span> <span class="name">ggtcatccta</span> <span class="name">ttagtaaacc</span> <span class="name">gatctggacc</span>
     <span class="literal number integer">1081</span> <span class="name">gatttctcgg</span> <span class="name">attctgatat</span> <span class="name">tcttgatcga</span> <span class="name">ttttgccgga</span> <span class="name">tatgtagaaa</span> <span class="name">tctttgtcgt</span>
     <span class="literal number integer">1141</span> <span class="name">tatcacagcg</span> <span class="name">gatcctcaaa</span> <span class="name">aaaacaggtt</span> <span class="name">ttgtatcgta</span> <span class="name">taaaatatat</span> <span class="name">acttcgactt</span>
     <span class="literal number integer">1201</span> <span class="name">tcgtgtgcta</span> <span class="name">gaactttggc</span> <span class="name">acggaaacat</span> <span class="name">aaaagtacag</span> <span class="name">tacgcacttt</span> <span class="name">tatgcgaaga</span>
     <span class="literal number integer">1261</span> <span class="name">ttaggttcgg</span> <span class="name">gattattaga</span> <span class="name">agaattcttt</span> <span class="name">atggaagaag</span> <span class="name">aa</span>
<span class="operator">//</span>
</pre>
<p>参数 <tt class="docutils literal"><span class="pre">rettype=&quot;gb&quot;</span></tt> 和 <tt class="docutils literal"><span class="pre">retmode=&quot;text&quot;</span></tt> 让我们下载的数据为GenBank格式。</p>
<p>需要注意的是直到2009年，Entrez EFetch API要求使用 “genbank” 作为返回类型，然而现在NCBI坚持使用官方的
“gb” 或 “gbwithparts” （或者针对蛋白的“gp”) 返回类型。同样需要注意的是，直到2012年2月，
Entrez EFetch API默认的返回格式为纯文本格式文件，现在默认的为XML格式。</p>
<p>作为另外的选择，你也可以使用 <tt class="docutils literal"><span class="pre">rettype=&quot;fasta&quot;</span></tt> 来获取Fasta格式的文件；参见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/efetchseq_help.html">EFetch Sequences 帮助页面</a> 。
记住，可选的数据格式决定于你要下载的数据库——请参见 <a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/entrez/query/static/efetch_help.html">EFetch 帮助页面</a>.</p>
<p>如果你要获取记录的格式是 <tt class="docutils literal">Bio.SeqIO</tt> 所接受的一种格式(见章节 <a class="reference external" href="#chapter:Bio.SeqIO">5</a>),
你可以直接将其解析为一个 <tt class="docutils literal">SeqRecord</tt> ：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span><span class="punctuation">,</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;186972394&quot;</span><span class="punctuation">,</span><span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span>
<span class="name">ID</span><span class="punctuation">:</span> <span class="name">EU490707</span><span class="operator">.</span><span class="literal number integer">1</span>
<span class="name">Name</span><span class="punctuation">:</span> <span class="name">EU490707</span>
<span class="name">Description</span><span class="punctuation">:</span> <span class="name">Selenipedium</span> <span class="name">aequinoctiale</span> <span class="name">maturase</span> <span class="name">K</span> <span class="punctuation">(</span><span class="name">matK</span><span class="punctuation">)</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">partial</span> <span class="name">cds</span><span class="punctuation">;</span> <span class="name">chloroplast</span><span class="operator">.</span>
<span class="name">Number</span> <span class="name">of</span> <span class="name">features</span><span class="punctuation">:</span> <span class="literal number integer">3</span>
<span class="operator">...</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'ATTTTTTACGAACCTGTGGAAATTTTTGGTTATGACAATAAATCTAGTTTAGTA...GAA'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>需要注意的是，一种更加典型的用法是先把序列数据保存到一个本地文件，<em>然后</em> 使用 <tt class="docutils literal">Bio.SeqIO</tt> 来解析。这样就避免了
在运行脚本的时候需要重复的下载同样的文件，并减轻NCBI服务器的负载。例如：</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">os</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="name">filename</span> <span class="operator">=</span> <span class="literal string">&quot;gi_186972394.gbk&quot;</span>
<span class="keyword">if</span> <span class="operator word">not</span> <span class="name">os</span><span class="operator">.</span><span class="name">path</span><span class="operator">.</span><span class="name">isfile</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">):</span>
    <span class="comment"># Downloading...</span>
    <span class="name">net_handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span><span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;186972394&quot;</span><span class="punctuation">,</span><span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
    <span class="name">out_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
    <span class="name">out_handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">net_handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">())</span>
    <span class="name">out_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
    <span class="name">net_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
    <span class="keyword">print</span> <span class="literal string">&quot;Saved&quot;</span>

<span class="keyword">print</span> <span class="literal string">&quot;Parsing...&quot;</span>
<span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="name">record</span>
</pre>
<p>为了得到XML格式的输出，你可以使用 <tt class="docutils literal">Bio.Entrez.read()</tt> 函数和参数 <tt class="docutils literal"><span class="pre">retmode=&quot;xml&quot;</span></tt> 进行解析，：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;186972394&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;GBSeq_definition&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Selenipedium aequinoctiale maturase K (matK) gene, partial cds; chloroplast'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;GBSeq_source&quot;</span><span class="punctuation">]</span>
<span class="literal string">'chloroplast Selenipedium aequinoctiale'</span>
</pre>
<p>就像这样处理数据。例如解析其他数据库特异的文件格式（例如，PubMed中用到的 <tt class="docutils literal">MEDLINE</tt> 格式），请参见章节 <a class="reference external" href="#sec:entrez-specialized-parsers">9.12</a>.</p>
<p>如果你想使用 <tt class="docutils literal">Bio.Entrez.esearch()</tt> 进行搜索，然后用 <tt class="docutils literal">Bio.Entrez.efetch()</tt> 下载数据，那么你需要用到
WebEnv的历史特性，请参加见章节 <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
</div>
<div class="section" id="elink-ncbi-entrez">
<h2>9.7  ELink: 在NCBI Entrez中搜索相关的条目</h2>
<p>ELink，在Biopython中是 <tt class="docutils literal">Bio.Entrez.elink()</tt> ，可以用来在NCBI Entrez数据库中寻找相关的条目。例如，你
可以使用它在gene数据库中寻找核苷酸条目，或者其他很酷的事情。</p>
<p>让我们使用ELink来在2009年的 <em>Bioinformatics</em> 杂志中寻找与Biopython应用相关的文章。这篇文章的PubMed ID
是19304878：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pmid</span> <span class="operator">=</span> <span class="literal string">&quot;19304878&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">elink</span><span class="punctuation">(</span><span class="name">dbfrom</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="name">pmid</span><span class="punctuation">))</span>
</pre>
<p>变量 <tt class="docutils literal">record</tt> 包含了一个Python列表，列出了已经搜索过的数据库。因为我们特指了一个PubMed ID来搜索，所以
<tt class="docutils literal">record</tt> 只包含了一个条目。这个条目是一个字典变量，包含了我们需要寻找的条目的信息，以及能搜索到的所有相关
的内容：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;DbFrom&quot;</span><span class="punctuation">]</span>
<span class="literal string">'pubmed'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'19304878'</span><span class="punctuation">]</span>
</pre>
<p>键 <tt class="docutils literal">&quot;LinkSetDb&quot;</tt> 包含了搜索结果，将每个目标数据库保存为一个列表。在我们这个搜索中，我们只在PubMed数据库
中找到了结果（尽管已经被分到了不同的分类）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">])</span>
<span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">linksetdb</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">linksetdb</span><span class="punctuation">[</span><span class="literal string">&quot;DbTo&quot;</span><span class="punctuation">],</span> <span class="name">linksetdb</span><span class="punctuation">[</span><span class="literal string">&quot;LinkName&quot;</span><span class="punctuation">],</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">linksetdb</span><span class="punctuation">[</span><span class="literal string">&quot;Link&quot;</span><span class="punctuation">])</span>
<span class="operator">...</span>
<span class="name">pubmed</span> <span class="name">pubmed_pubmed</span> <span class="literal number integer">110</span>
<span class="name">pubmed</span> <span class="name">pubmed_pubmed_combined</span> <span class="literal number integer">6</span>
<span class="name">pubmed</span> <span class="name">pubmed_pubmed_five</span> <span class="literal number integer">6</span>
<span class="name">pubmed</span> <span class="name">pubmed_pubmed_reviews</span> <span class="literal number integer">5</span>
<span class="name">pubmed</span> <span class="name">pubmed_pubmed_reviews_five</span> <span class="literal number integer">5</span>
</pre>
<p>实际的搜索结果被保存在键值为 <tt class="docutils literal">&quot;Link&quot;</tt> 的字典下。在标准搜索下，总共找到了110个条目。让我们现在看看我们第一个
搜索结果：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Link&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">u'Id'</span><span class="punctuation">:</span> <span class="literal string">'19304878'</span><span class="punctuation">}</span>
</pre>
<p>这个就是我们搜索的文章，从中并不能看到更多的结果，所以让我们来看看我们的第二个搜索结果：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Link&quot;</span><span class="punctuation">][</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">u'Id'</span><span class="punctuation">:</span> <span class="literal string">'14630660'</span><span class="punctuation">}</span>
</pre>
<p>这个PubMed ID为14530660的文章是关于Biopython PDB解析器的。</p>
<p>我们通过一个循环来打印出所有的PubMed IDs：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">link</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Link&quot;</span><span class="punctuation">]</span> <span class="punctuation">:</span> <span class="keyword">print</span> <span class="name">link</span><span class="punctuation">[</span><span class="literal string">&quot;Id&quot;</span><span class="punctuation">]</span>
<span class="literal number integer">19304878</span>
<span class="literal number integer">14630660</span>
<span class="literal number integer">18689808</span>
<span class="literal number integer">17121776</span>
<span class="literal number integer">16377612</span>
<span class="literal number integer">12368254</span>
<span class="operator">......</span>
</pre>
<p>现在漂亮极了，但是对我个人而言，我对某篇文章是否被引用过更感兴趣。好吧，ELink也可以完成这个——至少对PubMed
Central的杂志来说是这样的（请见 Section <a class="reference external" href="#sec:elink-citations">9.15.3</a>）。</p>
<p>关于ELink的帮助，请见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/elink_help.html">ELink 帮助页面</a> 。
这是一个关于 <a class="reference external" href="http://eutils.ncbi.nlm.nih.gov/corehtml/query/static/entrezlinks.html">link names</a>
的整个的子页面， 描述了不同的数据库可以怎样交叉的索引。</p>
</div>
<div class="section" id="egquery">
<h2>9.8  EGQuery: 全局搜索- 统计搜索的条目</h2>
<p>EGQuery提供搜索字段在每个Entrez数据库中的数目。当我们只需要知道在每个数据库中能找到的条目的个数，
而不需要知道具体搜索结果的时候，这个非常的有用（请见例子`9.14.2 &lt;#subsec:entrez_example_genbank&gt;`__ below）。</p>
<p>在这个例子中，我们使用 <tt class="docutils literal">Bio.Entrez.egquery()</tt> 来获取跟 “Biopython” 相关的数目：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">egquery</span><span class="punctuation">(</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;biopython&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">row</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;eGQueryResult&quot;</span><span class="punctuation">]:</span> <span class="keyword">print</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;DbName&quot;</span><span class="punctuation">],</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="operator">...</span>
<span class="name">pubmed</span> <span class="literal number integer">6</span>
<span class="name">pmc</span> <span class="literal number integer">62</span>
<span class="name">journals</span> <span class="literal number integer">0</span>
<span class="operator">...</span>
</pre>
<p>请见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/egquery_help.html">EGQuery 帮助页面</a>
获得更多信息.</p>
</div>
<div class="section" id="espell">
<h2>9.9  ESpell: 获得拼写建议</h2>
<p>ESpell可以检索拼写建议。在这个例子中，我们使用 <tt class="docutils literal">Bio.Entrez.espell()</tt> 来获得Biopython正确的拼写：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">espell</span><span class="punctuation">(</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;biopythooon&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;Query&quot;</span><span class="punctuation">]</span>
<span class="literal string">'biopythooon'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;CorrectedQuery&quot;</span><span class="punctuation">]</span>
<span class="literal string">'biopython'</span>
</pre>
<p>请见 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/espell_help.html">ESpell 帮助页面</a>
获得更多信息. 这个的主要用法是在使用GUI工具的时候为搜索的条目自动的提供拼写建议。</p>
</div>
<div class="section" id="entrez-xml">
<h2>9.10  解析大的Entrez XML文件</h2>
<p><tt class="docutils literal">Entrez.read</tt> 函数将Entrez返回的结果读取到一个Python对象里面去，这个对象被保存在内存中。对于解析太大的
XML文件而内存不够时，可以使用 <tt class="docutils literal">Entrez.parse</tt> 这个函数。这是一个生成器函数，它将一个一个的读取XML文件里面的内容。只有XML
文件是一个列表对象的时候，这个函数才有用（换句话说，如果在一个内存无限的计算机上 <tt class="docutils literal">Entrez.read</tt> 将返回一个
Python列表）。</p>
<p>例如，你可以通过NCBI的FTP站点从Entrez Gene 数据库中下载某个物种全部的条目作为一个文件。这个文件可能很大。
作为一个例子，在2009年9月4日，文件 <tt class="docutils literal">Homo_sapiens.ags.gz</tt> 包含了Entrez Gene数据库中人的序列，文件大小
有116576kB。这个文件是 <tt class="docutils literal">ASN</tt> 格式，可以通过NCBI的 <tt class="docutils literal">gene2xml</tt> 程序转成XML格式（请到NCBI的FTP站点获取
更多的信息）：</p>
<pre class="code python literal-block">
<span class="name">gene2xml</span> <span class="operator">-</span><span class="name">b</span> <span class="name">T</span> <span class="operator">-</span><span class="name">i</span> <span class="name">Homo_sapiens</span><span class="operator">.</span><span class="name">ags</span> <span class="operator">-</span><span class="name">o</span> <span class="name">Homo_sapiens</span><span class="operator">.</span><span class="name">xml</span>
</pre>
<p>XML结果文件有6.1GB. 在大多数电脑上尝试 <tt class="docutils literal">Entrez.read</tt> 都会导致 <tt class="docutils literal">MemoryError</tt> 。</p>
<p>XML文件 <tt class="docutils literal">Homo_sapiens.xml</tt> 包含了一个Entrez gene记录的列表，每个对应于人的一个Entrez基因信息。 <tt class="docutils literal">Entrez.parse</tt>
将一个一个的读取这些记录。这样你可以通过遍历每个记录的方式打印或者存储每个记录相关的信息。例如，下面这个脚本
遍历了Entrez基因里面的记录，打印了每个基因的数目和名字：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;Homo_sapiens.xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">status</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">'Entrezgene_track-info'</span><span class="punctuation">][</span><span class="literal string">'Gene-track'</span><span class="punctuation">][</span><span class="literal string">'Gene-track_status'</span><span class="punctuation">]</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">status</span><span class="operator">.</span><span class="name">attributes</span><span class="punctuation">[</span><span class="literal string">'value'</span><span class="punctuation">]</span><span class="operator">==</span><span class="literal string">'discontinued'</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">continue</span>
<span class="operator">...</span>     <span class="name">geneid</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">'Entrezgene_track-info'</span><span class="punctuation">][</span><span class="literal string">'Gene-track'</span><span class="punctuation">][</span><span class="literal string">'Gene-track_geneid'</span><span class="punctuation">]</span>
<span class="operator">...</span>     <span class="name">genename</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">'Entrezgene_gene'</span><span class="punctuation">][</span><span class="literal string">'Gene-ref'</span><span class="punctuation">][</span><span class="literal string">'Gene-ref_locus'</span><span class="punctuation">]</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">geneid</span><span class="punctuation">,</span> <span class="name">genename</span>
</pre>
<p>将会打印以下内容:</p>
<pre class="code python literal-block">
<span class="literal number integer">1</span> <span class="name">A1BG</span>
<span class="literal number integer">2</span> <span class="name">A2M</span>
<span class="literal number integer">3</span> <span class="name">A2MP</span>
<span class="literal number integer">8</span> <span class="name">AA</span>
<span class="literal number integer">9</span> <span class="name">NAT1</span>
<span class="literal number integer">10</span> <span class="name">NAT2</span>
<span class="literal number integer">11</span> <span class="name">AACP</span>
<span class="literal number integer">12</span> <span class="name">SERPINA3</span>
<span class="literal number integer">13</span> <span class="name">AADAC</span>
<span class="literal number integer">14</span> <span class="name">AAMP</span>
<span class="literal number integer">15</span> <span class="name">AANAT</span>
<span class="literal number integer">16</span> <span class="name">AARS</span>
<span class="literal number integer">17</span> <span class="name">AAVS1</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="id79">
<h2>9.11  错误处理</h2>
<p>当解析XML文件的时候，可能出现一下三个错误：</p>
<ul class="simple">
<li>这个文件可能不是以常规的 XML 文件格式开头；</li>
<li>这个文件可能不完整或者包含一些非 XML 格式的内容；</li>
<li>这个文件是正常的 XML 文件，但是包含和相关 DTD 文件无关的条目。</li>
</ul>
<p>第一种情况会在，例如，你尝试把一个 Fasta 文件当做 XML 文件来处理时发生：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.fna&quot;</span><span class="punctuation">)</span> <span class="comment"># a Fasta file</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
  <span class="name">File</span> <span class="literal string">&quot;&lt;stdin&gt;&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="operator">&lt;</span><span class="name">module</span><span class="operator">&gt;</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/__init__.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">257</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">read</span>
    <span class="name">record</span> <span class="operator">=</span> <span class="name">handler</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/Parser.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">164</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">read</span>
    <span class="keyword">raise</span> <span class="name">NotXMLError</span><span class="punctuation">(</span><span class="name">e</span><span class="punctuation">)</span>
<span class="name">Bio</span><span class="operator">.</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">Parser</span><span class="operator">.</span><span class="name">NotXMLError</span><span class="punctuation">:</span> <span class="name">Failed</span> <span class="name">to</span> <span class="name">parse</span> <span class="name">the</span> <span class="name">XML</span> <span class="name">data</span> <span class="punctuation">(</span><span class="name">syntax</span> <span class="name">error</span><span class="punctuation">:</span> <span class="name">line</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">column</span> <span class="literal number integer">0</span><span class="punctuation">)</span><span class="operator">.</span> <span class="name">Please</span> <span class="name">make</span> <span class="name">sure</span> <span class="name">that</span> <span class="name">the</span> <span class="name builtin">input</span> <span class="name">data</span> <span class="name">are</span> <span class="operator word">in</span> <span class="name">XML</span> <span class="name">format</span><span class="operator">.</span>
</pre>
<p>这时候，解析器找不到 <tt class="docutils literal"><span class="pre">&lt;?xml</span> ...</tt> 标签，而这是一个 XML 文件开始的标志，那么可以确定这个文件不是 XML 文件。</p>
<p>当你的文件是XML格式，但是是不完整的（例如，提前结束了），那么解析器会报CorruptedXMLError错误。下面
这个是一个XML文件提前结束的例子：</p>
<pre class="code python literal-block">
<span class="operator">&lt;</span><span class="error">?</span><span class="name">xml</span> <span class="name">version</span><span class="operator">=</span><span class="literal string">&quot;1.0&quot;</span><span class="error">?</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="error">!</span><span class="name">DOCTYPE</span> <span class="name">eInfoResult</span> <span class="name">PUBLIC</span> <span class="literal string">&quot;-//NLM//DTD eInfoResult, 11 May 2002//EN&quot;</span> <span class="literal string">&quot;http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd&quot;</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="name">eInfoResult</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="name">DbList</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pubmed</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">protein</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nucleotide</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nuccore</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nucgss</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">nucest</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">structure</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">genome</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">books</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">cancerchromosomes</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">cdd</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
</pre>
<p>这个会生成以下的日志文件：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
  <span class="name">File</span> <span class="literal string">&quot;&lt;stdin&gt;&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="operator">&lt;</span><span class="name">module</span><span class="operator">&gt;</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/__init__.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">257</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">read</span>
    <span class="name">record</span> <span class="operator">=</span> <span class="name">handler</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/Parser.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">160</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">read</span>
    <span class="keyword">raise</span> <span class="name">CorruptedXMLError</span><span class="punctuation">(</span><span class="name">e</span><span class="punctuation">)</span>
<span class="name">Bio</span><span class="operator">.</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">Parser</span><span class="operator">.</span><span class="name">CorruptedXMLError</span><span class="punctuation">:</span> <span class="name">Failed</span> <span class="name">to</span> <span class="name">parse</span> <span class="name">the</span> <span class="name">XML</span> <span class="name">data</span> <span class="punctuation">(</span><span class="name">no</span> <span class="name">element</span> <span class="name">found</span><span class="punctuation">:</span> <span class="name">line</span> <span class="literal number integer">16</span><span class="punctuation">,</span> <span class="name">column</span> <span class="literal number integer">0</span><span class="punctuation">)</span><span class="operator">.</span> <span class="name">Please</span> <span class="name">make</span> <span class="name">sure</span> <span class="name">that</span> <span class="name">the</span> <span class="name builtin">input</span> <span class="name">data</span> <span class="name">are</span> <span class="operator word">not</span> <span class="name">corrupted</span><span class="operator">.</span>

<span class="operator">&gt;&gt;&gt;</span>
</pre>
<p>注意，报错信息告诉你在XML文件的什么位置检测到了错误。</p>
<p>如果XML文件当中包含有对应DTD文件中没有描述的标签的时候，会发生第三类错误。以下是这样一个XML文件的例子：</p>
<pre class="code python literal-block">
<span class="operator">&lt;</span><span class="error">?</span><span class="name">xml</span> <span class="name">version</span><span class="operator">=</span><span class="literal string">&quot;1.0&quot;</span><span class="error">?</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="error">!</span><span class="name">DOCTYPE</span> <span class="name">eInfoResult</span> <span class="name">PUBLIC</span> <span class="literal string">&quot;-//NLM//DTD eInfoResult, 11 May 2002//EN&quot;</span> <span class="literal string">&quot;http://www.ncbi.nlm.nih.gov/entrez/query/DTD/eInfo_020511.dtd&quot;</span><span class="operator">&gt;</span>
<span class="operator">&lt;</span><span class="name">eInfoResult</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbInfo</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DbName</span><span class="operator">&gt;</span><span class="name">pubmed</span><span class="operator">&lt;/</span><span class="name">DbName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">MenuName</span><span class="operator">&gt;</span><span class="name">PubMed</span><span class="operator">&lt;/</span><span class="name">MenuName</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">Description</span><span class="operator">&gt;</span><span class="name">PubMed</span> <span class="name">bibliographic</span> <span class="name">record</span><span class="operator">&lt;/</span><span class="name">Description</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">Count</span><span class="operator">&gt;</span><span class="literal number integer">20161961</span><span class="operator">&lt;/</span><span class="name">Count</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">LastUpdate</span><span class="operator">&gt;</span><span class="literal number integer">2010</span><span class="operator">/</span><span class="literal number integer">09</span><span class="operator">/</span><span class="literal number integer">10</span> <span class="literal number oct">04</span><span class="punctuation">:</span><span class="literal number integer">52</span><span class="operator">&lt;/</span><span class="name">LastUpdate</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">FieldList</span><span class="operator">&gt;</span>
                <span class="operator">&lt;</span><span class="name">Field</span><span class="operator">&gt;</span>
<span class="operator">...</span>
                <span class="operator">&lt;/</span><span class="name">Field</span><span class="operator">&gt;</span>
        <span class="operator">&lt;/</span><span class="name">FieldList</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">DocsumList</span><span class="operator">&gt;</span>
                <span class="operator">&lt;</span><span class="name">Docsum</span><span class="operator">&gt;</span>
                        <span class="operator">&lt;</span><span class="name">DsName</span><span class="operator">&gt;</span><span class="name">PubDate</span><span class="operator">&lt;/</span><span class="name">DsName</span><span class="operator">&gt;</span>
                        <span class="operator">&lt;</span><span class="name">DsType</span><span class="operator">&gt;</span><span class="literal number integer">4</span><span class="operator">&lt;/</span><span class="name">DsType</span><span class="operator">&gt;</span>
                        <span class="operator">&lt;</span><span class="name">DsTypeName</span><span class="operator">&gt;</span><span class="name">string</span><span class="operator">&lt;/</span><span class="name">DsTypeName</span><span class="operator">&gt;</span>
                <span class="operator">&lt;/</span><span class="name">Docsum</span><span class="operator">&gt;</span>
                <span class="operator">&lt;</span><span class="name">Docsum</span><span class="operator">&gt;</span>
                        <span class="operator">&lt;</span><span class="name">DsName</span><span class="operator">&gt;</span><span class="name">EPubDate</span><span class="operator">&lt;/</span><span class="name">DsName</span><span class="operator">&gt;</span>
<span class="operator">...</span>
        <span class="operator">&lt;/</span><span class="name">DbInfo</span><span class="operator">&gt;</span>
<span class="operator">&lt;/</span><span class="name">eInfoResult</span><span class="operator">&gt;</span>
</pre>
<p>在这个文件里面，因为一些原因，<tt class="docutils literal">&lt;DocsumList&gt;</tt> （还有一些其他的）标签没有在DTD文件 <tt class="docutils literal">eInfo_020511.dtd</tt>
中列出来，XML文件对应DTD文件的第二行会特别的描述出来。默认情况下，如果没有找到DTD文件中的标签，解析器
会中止并报ValidationError错误。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;einfo3.xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
  <span class="name">File</span> <span class="literal string">&quot;&lt;stdin&gt;&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="operator">&lt;</span><span class="name">module</span><span class="operator">&gt;</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/__init__.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">257</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">read</span>
    <span class="name">record</span> <span class="operator">=</span> <span class="name">handler</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/Parser.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">154</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">read</span>
    <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">parser</span><span class="operator">.</span><span class="name">ParseFile</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
  <span class="name">File</span> <span class="literal string">&quot;/usr/local/lib/python2.7/site-packages/Bio/Entrez/Parser.py&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">246</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="name">startElementHandler</span>
    <span class="keyword">raise</span> <span class="name">ValidationError</span><span class="punctuation">(</span><span class="name">name</span><span class="punctuation">)</span>
<span class="name">Bio</span><span class="operator">.</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">Parser</span><span class="operator">.</span><span class="name">ValidationError</span><span class="punctuation">:</span> <span class="name">Failed</span> <span class="name">to</span> <span class="name">find</span> <span class="name">tag</span> <span class="literal string">'DocsumList'</span> <span class="operator word">in</span> <span class="name">the</span> <span class="name">DTD</span><span class="operator">.</span> <span class="name">To</span> <span class="name">skip</span> <span class="name builtin">all</span> <span class="name">tags</span> <span class="name">that</span> <span class="name">are</span> <span class="operator word">not</span> <span class="name">represented</span> <span class="operator word">in</span> <span class="name">the</span> <span class="name">DTD</span><span class="punctuation">,</span> <span class="name">please</span> <span class="name">call</span> <span class="name">Bio</span><span class="operator">.</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span> <span class="operator word">or</span> <span class="name">Bio</span><span class="operator">.</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">parse</span> <span class="keyword">with</span> <span class="name">validate</span><span class="operator">=</span><span class="name builtin pseudo">False</span><span class="operator">.</span>
</pre>
<p>可选地，你可以让解析器跳过这样的标签，而不是报ValidationError错误。通过调用 <tt class="docutils literal">Entrez.read</tt> 或者
<tt class="docutils literal">Entrez.parse</tt> 并使参数 <tt class="docutils literal">validate</tt> 等于False可以实现这个功能：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;einfo3.xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span><span class="name">validate</span><span class="operator">=</span><span class="name builtin pseudo">False</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span>
</pre>
<p>当然，XML文件中的tag没有出现在对应DTD文件中的信息，将不会在 <tt class="docutils literal">Entrez.read</tt> 的返回记录中出现。</p>
</div>
<div class="section" id="id80">
<h2>9.12  专用的解析器</h2>
<p>函数 <tt class="docutils literal">Bio.Entrez.read()</tt> 可以处理大部分（如果不是所有的话）Entrez返回的XML文件。Entrez也可以
允许你通过其他格式来获取数据，有时候，这种方式在可读性上比XML文件格式更具优势（或者下载文件的大小）。</p>
<p>为了使用 <tt class="docutils literal">Bio.Entrez.efetch()</tt> 函数从Entrez中提取一种特有的文件格式，需要指明 <tt class="docutils literal">rettype</tt> 和或者或 <tt class="docutils literal">retmode</tt>
等可选参数。不同的组合在 <a class="reference external" href="http://www.ncbi.nlm.nih.gov/entrez/query/static/efetch_help.html">NCBI efetch的页面</a> 。
有对不同数据库的描述。</p>
<p>一个显然的例子是，你可能更想以FASTA或者 GenBank/GenPept ( 这些可以通过 <tt class="docutils literal">Bio.SeqIO</tt> 来处理, 请见 Sections <a class="reference external" href="#sec:SeqIO_GenBank_Online">5.3.1</a>
和 <a class="reference external" href="#sec:efetch">9.6</a> ） 纯文本形式下载序列。对于文献数据库，Biopython包含了一个处理PubMed中
使用的 <tt class="docutils literal">MEDLINE</tt> 格式的解析器。</p>
<div class="section" id="medline">
<h3>9.12.1  解析Medline记录</h3>
<p>你可以在 <tt class="docutils literal">Bio.Medline</tt> 中找到Medline的解析器。假设你想处理包含一个Medline记录的 <tt class="docutils literal">pubmed_result1.txt</tt>
文件。你可以在Biopython的 <tt class="docutils literal">Tests\Medline</tt> 目录下找到这个文件，这个文件内容如下所示：</p>
<pre class="code python literal-block">
<span class="name">PMID</span><span class="operator">-</span> <span class="literal number integer">12230038</span>
<span class="name">OWN</span> <span class="operator">-</span> <span class="name">NLM</span>
<span class="name">STAT</span><span class="operator">-</span> <span class="name">MEDLINE</span>
<span class="name">DA</span>  <span class="operator">-</span> <span class="literal number integer">20020916</span>
<span class="name">DCOM</span><span class="operator">-</span> <span class="literal number integer">20030606</span>
<span class="name">LR</span>  <span class="operator">-</span> <span class="literal number integer">20041117</span>
<span class="name">PUBM</span><span class="operator">-</span> <span class="name">Print</span>
<span class="name">IS</span>  <span class="operator">-</span> <span class="literal number integer">1467</span><span class="operator">-</span><span class="literal number integer">5463</span> <span class="punctuation">(</span><span class="name">Print</span><span class="punctuation">)</span>
<span class="name">VI</span>  <span class="operator">-</span> <span class="literal number integer">3</span>
<span class="name">IP</span>  <span class="operator">-</span> <span class="literal number integer">3</span>
<span class="name">DP</span>  <span class="operator">-</span> <span class="literal number integer">2002</span> <span class="name">Sep</span>
<span class="name">TI</span>  <span class="operator">-</span> <span class="name">The</span> <span class="name">Bio</span><span class="operator">*</span> <span class="name">toolkits</span><span class="operator">--</span><span class="name">a</span> <span class="name">brief</span> <span class="name">overview</span><span class="operator">.</span>
<span class="name">PG</span>  <span class="operator">-</span> <span class="literal number integer">296</span><span class="operator">-</span><span class="literal number integer">302</span>
<span class="name">AB</span>  <span class="operator">-</span> <span class="name">Bioinformatics</span> <span class="name">research</span> <span class="operator word">is</span> <span class="name">often</span> <span class="name">difficult</span> <span class="name">to</span> <span class="name">do</span> <span class="keyword">with</span> <span class="name">commercial</span> <span class="name">software</span><span class="operator">.</span> <span class="name">The</span>
      <span class="name">Open</span> <span class="name">Source</span> <span class="name">BioPerl</span><span class="punctuation">,</span> <span class="name">BioPython</span> <span class="operator word">and</span> <span class="name">Biojava</span> <span class="name">projects</span> <span class="name">provide</span> <span class="name">toolkits</span> <span class="keyword">with</span>
<span class="operator">...</span>
</pre>
<p>我们首先打开文件，然后解析它：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Medline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">input</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;pubmed_result1.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Medline</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="punctuation">)</span>
</pre>
<p>现在 <tt class="docutils literal">record</tt> 将 Medline记录以Python字典的形式保存起来：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;PMID&quot;</span><span class="punctuation">]</span>
<span class="literal string">'12230038'</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;AB&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Bioinformatics research is often difficult to do with commercial software.</span>
<span class="name">The</span> <span class="name">Open</span> <span class="name">Source</span> <span class="name">BioPerl</span><span class="punctuation">,</span> <span class="name">BioPython</span> <span class="operator word">and</span> <span class="name">Biojava</span> <span class="name">projects</span> <span class="name">provide</span> <span class="name">toolkits</span> <span class="keyword">with</span>
<span class="name">multiple</span> <span class="name">functionality</span> <span class="name">that</span> <span class="name">make</span> <span class="name">it</span> <span class="name">easier</span> <span class="name">to</span> <span class="name">create</span> <span class="name">customised</span> <span class="name">pipelines</span> <span class="operator word">or</span>
<span class="name">analysis</span><span class="operator">.</span> <span class="name">This</span> <span class="name">review</span> <span class="name">briefly</span> <span class="name">compares</span> <span class="name">the</span> <span class="name">quirks</span> <span class="name">of</span> <span class="name">the</span> <span class="name">underlying</span> <span class="name">languages</span>
<span class="operator word">and</span> <span class="name">the</span> <span class="name">functionality</span><span class="punctuation">,</span> <span class="name">documentation</span><span class="punctuation">,</span> <span class="name">utility</span> <span class="operator word">and</span> <span class="name">relative</span> <span class="name">advantages</span> <span class="name">of</span> <span class="name">the</span>
<span class="name">Bio</span> <span class="name">counterparts</span><span class="punctuation">,</span> <span class="name">particularly</span> <span class="keyword namespace">from</span> <span class="name namespace">the</span> <span class="name namespace">point</span> <span class="name namespace">of</span> <span class="name namespace">view</span> <span class="name namespace">of</span> <span class="name namespace">the</span> <span class="name namespace">beginning</span>
<span class="name">biologist</span> <span class="name">programmer</span><span class="operator">.</span><span class="literal string">'</span>
</pre>
<p>用于Medline记录的键值可以相当模糊，使用</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
</pre>
<p>可以做一个简单的总结。</p>
<p>为了解析包含多个Medline记录的文件，你可以使用 <tt class="docutils literal">parse</tt> 函数来代替：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Medline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">input</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;pubmed_result2.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Medline</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;TI&quot;</span><span class="punctuation">]</span>
<span class="name">A</span> <span class="name">high</span> <span class="name">level</span> <span class="name">interface</span> <span class="name">to</span> <span class="name">SCOP</span> <span class="operator word">and</span> <span class="name">ASTRAL</span> <span class="name">implemented</span> <span class="operator word">in</span> <span class="name">python</span><span class="operator">.</span>
<span class="name">GenomeDiagram</span><span class="punctuation">:</span> <span class="name">a</span> <span class="name">python</span> <span class="name">package</span> <span class="keyword">for</span> <span class="name">the</span> <span class="name">visualization</span> <span class="name">of</span> <span class="name">large</span><span class="operator">-</span><span class="name">scale</span> <span class="name">genomic</span> <span class="name">data</span><span class="operator">.</span>
<span class="name">Open</span> <span class="name">source</span> <span class="name">clustering</span> <span class="name">software</span><span class="operator">.</span>
<span class="name">PDB</span> <span class="name builtin">file</span> <span class="name">parser</span> <span class="operator word">and</span> <span class="name">structure</span> <span class="keyword">class</span> <span class="name class">implemented</span> <span class="operator word">in</span> <span class="name">Python</span><span class="operator">.</span>
</pre>
<p>你可以通过 <tt class="docutils literal">Bio.Entrez.efetch</tt> 来下载Medline记录，而不是保存在某个文件里。例如，让我们来查看PubMed
里面跟Biopython相关的所有所有Medline记录：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;biopython&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'19304878'</span><span class="punctuation">,</span> <span class="literal string">'18606172'</span><span class="punctuation">,</span> <span class="literal string">'16403221'</span><span class="punctuation">,</span> <span class="literal string">'16377612'</span><span class="punctuation">,</span> <span class="literal string">'14871861'</span><span class="punctuation">,</span> <span class="literal string">'14630660'</span><span class="punctuation">,</span> <span class="literal string">'12230038'</span><span class="punctuation">]</span>
</pre>
<p>现在我们使用 <tt class="docutils literal">Bio.Entrez.efetch</tt> 来下载这些Medline记录:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idlist</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span><span class="name builtin">id</span><span class="operator">=</span><span class="name">idlist</span><span class="punctuation">,</span><span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;medline&quot;</span><span class="punctuation">,</span><span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
</pre>
<p>这里，我们使 <tt class="docutils literal"><span class="pre">rettype=&quot;medline&quot;,</span> <span class="pre">retmode=&quot;text&quot;</span></tt> 来以纯文本形式的Medline格式来得到这些记录。现在
我们使用 <tt class="docutils literal">Bio.Medline</tt> 来解析这些记录：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Medline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Medline</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;AU&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Cock PJ'</span><span class="punctuation">,</span> <span class="literal string">'Antao T'</span><span class="punctuation">,</span> <span class="literal string">'Chang JT'</span><span class="punctuation">,</span> <span class="literal string">'Chapman BA'</span><span class="punctuation">,</span> <span class="literal string">'Cox CJ'</span><span class="punctuation">,</span> <span class="literal string">'Dalke A'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'de Hoon MJ'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Munteanu CR'</span><span class="punctuation">,</span> <span class="literal string">'Gonzalez-Diaz H'</span><span class="punctuation">,</span> <span class="literal string">'Magalhaes AL'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Casbon JA'</span><span class="punctuation">,</span> <span class="literal string">'Crooks GE'</span><span class="punctuation">,</span> <span class="literal string">'Saqi MA'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Pritchard L'</span><span class="punctuation">,</span> <span class="literal string">'White JA'</span><span class="punctuation">,</span> <span class="literal string">'Birch PR'</span><span class="punctuation">,</span> <span class="literal string">'Toth IK'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'de Hoon MJ'</span><span class="punctuation">,</span> <span class="literal string">'Imoto S'</span><span class="punctuation">,</span> <span class="literal string">'Nolan J'</span><span class="punctuation">,</span> <span class="literal string">'Miyano S'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Hamelryck T'</span><span class="punctuation">,</span> <span class="literal string">'Manderick B'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Mangalam H'</span><span class="punctuation">]</span>
</pre>
<p>为了比对，我们展示了一个XML格式的例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idlist</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span><span class="name builtin">id</span><span class="operator">=</span><span class="name">idlist</span><span class="punctuation">,</span><span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;medline&quot;</span><span class="punctuation">,</span><span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;MedlineCitation&quot;</span><span class="punctuation">][</span><span class="literal string">&quot;Article&quot;</span><span class="punctuation">][</span><span class="literal string">&quot;ArticleTitle&quot;</span><span class="punctuation">]</span>
<span class="name">Biopython</span><span class="punctuation">:</span> <span class="name">freely</span> <span class="name">available</span> <span class="name">Python</span> <span class="name">tools</span> <span class="keyword">for</span> <span class="name">computational</span> <span class="name">molecular</span> <span class="name">biology</span> <span class="operator word">and</span>
 <span class="name">bioinformatics</span><span class="operator">.</span>
<span class="name">Enzymes</span><span class="operator">/</span><span class="name">non</span><span class="operator">-</span><span class="name">enzymes</span> <span class="name">classification</span> <span class="name">model</span> <span class="name">complexity</span> <span class="name">based</span> <span class="name">on</span> <span class="name">composition</span><span class="punctuation">,</span> <span class="name">sequence</span><span class="punctuation">,</span>
 <span class="literal number integer">3</span><span class="name">D</span> <span class="operator word">and</span> <span class="name">topological</span> <span class="name">indices</span><span class="operator">.</span>
<span class="name">A</span> <span class="name">high</span> <span class="name">level</span> <span class="name">interface</span> <span class="name">to</span> <span class="name">SCOP</span> <span class="operator word">and</span> <span class="name">ASTRAL</span> <span class="name">implemented</span> <span class="operator word">in</span> <span class="name">python</span><span class="operator">.</span>
<span class="name">GenomeDiagram</span><span class="punctuation">:</span> <span class="name">a</span> <span class="name">python</span> <span class="name">package</span> <span class="keyword">for</span> <span class="name">the</span> <span class="name">visualization</span> <span class="name">of</span> <span class="name">large</span><span class="operator">-</span><span class="name">scale</span> <span class="name">genomic</span> <span class="name">data</span><span class="operator">.</span>
<span class="name">Open</span> <span class="name">source</span> <span class="name">clustering</span> <span class="name">software</span><span class="operator">.</span>
<span class="name">PDB</span> <span class="name builtin">file</span> <span class="name">parser</span> <span class="operator word">and</span> <span class="name">structure</span> <span class="keyword">class</span> <span class="name class">implemented</span> <span class="operator word">in</span> <span class="name">Python</span><span class="operator">.</span>
<span class="name">The</span> <span class="name">Bio</span><span class="operator">*</span> <span class="name">toolkits</span><span class="operator">--</span><span class="name">a</span> <span class="name">brief</span> <span class="name">overview</span><span class="operator">.</span>
</pre>
<p>需要注意的是，在上面这两个例子当中，为了简便我们混合使用了 ESearch 和 EFetch。在这种情形下，NCBI 希望你
使用他们的历史记录特性，在下面章节中会讲到Section <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
</div>
<div class="section" id="geo">
<h3>9.12.2  解析GEO记录</h3>
<p>GEO ( <a class="reference external" href="http://www.ncbi.nlm.nih.gov/geo/">Gene Expression Omnibus</a> ) 是高通量基因表达和杂交芯片
数据的数据库。 <tt class="docutils literal">Bio.Geo</tt> 模块可以用来解析GEO格式的数据。</p>
<p>下面的代码展示了怎样将一个名称为 <tt class="docutils literal">GSE16.txt</tt> 的GEO文件存进一个记录，并打印该记录：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Geo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;GSE16.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Geo</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span>
</pre>
<p>你可以使用 ESearch 来搜索 “gds” 数据库 (GEO 数据集) :</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span> <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;gds&quot;</span><span class="punctuation">,</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;GSE16&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'200000016'</span><span class="punctuation">,</span> <span class="literal string">'100000028'</span><span class="punctuation">]</span>
</pre>
<p>通过Entrez网站，UID “200000016” 是GDS16，其他的hit “100000028” 是相关的平台。不幸的是，在写
这份指南的时候，NCBI貌似还不支持通过Entrez下载GEO文件（不论XML文件，还是SOFT格式的文件）。</p>
<p>然而，可以相当直接的通过 FTP <a class="reference external" href="ftp://ftp.ncbi.nih.gov/pub/geo/">ftp://ftp.ncbi.nih.gov/pub/geo/</a> 来下载 GEO 文件。
在这个例子当中，你需要的文件应该是 <a class="reference external" href="ftp://ftp.ncbi.nih.gov/pub/geo/DATA/SOFT/by_series/GSE16/GSE16_family.soft.gz">ftp://ftp.ncbi.nih.gov/pub/geo/DATA/SOFT/by_series/GSE16/GSE16_family.soft.gz</a>
（一个压缩文件，参见Python的gzip 模块）。</p>
</div>
<div class="section" id="unigene">
<h3>9.12.3  解析UniGene记录</h3>
<p>UniGene是NCBI的转录组数据库，每个UniGene记录展示了该转录本在某个特定物种中相关的基因。一个典型的UniGene
记录如下所示：</p>
<pre class="code python literal-block">
<span class="name">ID</span>          <span class="name">Hs</span><span class="operator">.</span><span class="literal number integer">2</span>
<span class="name">TITLE</span>       <span class="name">N</span><span class="operator">-</span><span class="name">acetyltransferase</span> <span class="literal number integer">2</span> <span class="punctuation">(</span><span class="name">arylamine</span> <span class="name">N</span><span class="operator">-</span><span class="name">acetyltransferase</span><span class="punctuation">)</span>
<span class="name">GENE</span>        <span class="name">NAT2</span>
<span class="name">CYTOBAND</span>    <span class="literal number integer">8</span><span class="name">p22</span>
<span class="name">GENE_ID</span>     <span class="literal number integer">10</span>
<span class="name">LOCUSLINK</span>   <span class="literal number integer">10</span>
<span class="name">HOMOL</span>       <span class="name">YES</span>
<span class="name">EXPRESS</span>      <span class="name">bone</span><span class="operator">|</span> <span class="name">connective</span> <span class="name">tissue</span><span class="operator">|</span> <span class="name">intestine</span><span class="operator">|</span> <span class="name">liver</span><span class="operator">|</span> <span class="name">liver</span> <span class="name">tumor</span><span class="operator">|</span> <span class="name">normal</span><span class="operator">|</span> <span class="name">soft</span> <span class="name">tissue</span><span class="operator">/</span><span class="name">muscle</span> <span class="name">tissue</span> <span class="name">tumor</span><span class="operator">|</span> <span class="name">adult</span>
<span class="name">RESTR_EXPR</span>   <span class="name">adult</span>
<span class="name">CHROMOSOME</span>  <span class="literal number integer">8</span>
<span class="name">STS</span>         <span class="name">ACC</span><span class="operator">=</span><span class="name">PMC310725P3</span> <span class="name">UNISTS</span><span class="operator">=</span><span class="literal number integer">272646</span>
<span class="name">STS</span>         <span class="name">ACC</span><span class="operator">=</span><span class="name">WIAF</span><span class="operator">-</span><span class="literal number integer">2120</span> <span class="name">UNISTS</span><span class="operator">=</span><span class="literal number integer">44576</span>
<span class="name">STS</span>         <span class="name">ACC</span><span class="operator">=</span><span class="name">G59899</span> <span class="name">UNISTS</span><span class="operator">=</span><span class="literal number integer">137181</span>
<span class="operator">...</span>
<span class="name">STS</span>         <span class="name">ACC</span><span class="operator">=</span><span class="name">GDB</span><span class="punctuation">:</span><span class="literal number integer">187676</span> <span class="name">UNISTS</span><span class="operator">=</span><span class="literal number integer">155563</span>
<span class="name">PROTSIM</span>     <span class="name">ORG</span><span class="operator">=</span><span class="literal number integer">10090</span><span class="punctuation">;</span> <span class="name">PROTGI</span><span class="operator">=</span><span class="literal number integer">6754794</span><span class="punctuation">;</span> <span class="name">PROTID</span><span class="operator">=</span><span class="name">NP_035004</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">PCT</span><span class="operator">=</span><span class="literal number float">76.55</span><span class="punctuation">;</span> <span class="name">ALN</span><span class="operator">=</span><span class="literal number integer">288</span>
<span class="name">PROTSIM</span>     <span class="name">ORG</span><span class="operator">=</span><span class="literal number integer">9796</span><span class="punctuation">;</span> <span class="name">PROTGI</span><span class="operator">=</span><span class="literal number integer">149742490</span><span class="punctuation">;</span> <span class="name">PROTID</span><span class="operator">=</span><span class="name">XP_001487907</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">PCT</span><span class="operator">=</span><span class="literal number float">79.66</span><span class="punctuation">;</span> <span class="name">ALN</span><span class="operator">=</span><span class="literal number integer">288</span>
<span class="name">PROTSIM</span>     <span class="name">ORG</span><span class="operator">=</span><span class="literal number integer">9986</span><span class="punctuation">;</span> <span class="name">PROTGI</span><span class="operator">=</span><span class="literal number integer">126722851</span><span class="punctuation">;</span> <span class="name">PROTID</span><span class="operator">=</span><span class="name">NP_001075655</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">PCT</span><span class="operator">=</span><span class="literal number float">76.90</span><span class="punctuation">;</span> <span class="name">ALN</span><span class="operator">=</span><span class="literal number integer">288</span>
<span class="operator">...</span>
<span class="name">PROTSIM</span>     <span class="name">ORG</span><span class="operator">=</span><span class="literal number integer">9598</span><span class="punctuation">;</span> <span class="name">PROTGI</span><span class="operator">=</span><span class="literal number integer">114619004</span><span class="punctuation">;</span> <span class="name">PROTID</span><span class="operator">=</span><span class="name">XP_519631</span><span class="operator">.</span><span class="literal number integer">2</span><span class="punctuation">;</span> <span class="name">PCT</span><span class="operator">=</span><span class="literal number float">98.28</span><span class="punctuation">;</span> <span class="name">ALN</span><span class="operator">=</span><span class="literal number integer">288</span>

<span class="name">SCOUNT</span>      <span class="literal number integer">38</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">BC067218</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g45501306</span><span class="punctuation">;</span> <span class="name">PID</span><span class="operator">=</span><span class="name">g45501307</span><span class="punctuation">;</span> <span class="name">SEQTYPE</span><span class="operator">=</span><span class="name">mRNA</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">NM_000015</span><span class="operator">.</span><span class="literal number integer">2</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g116295259</span><span class="punctuation">;</span> <span class="name">PID</span><span class="operator">=</span><span class="name">g116295260</span><span class="punctuation">;</span> <span class="name">SEQTYPE</span><span class="operator">=</span><span class="name">mRNA</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">D90042</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g219415</span><span class="punctuation">;</span> <span class="name">PID</span><span class="operator">=</span><span class="name">g219416</span><span class="punctuation">;</span> <span class="name">SEQTYPE</span><span class="operator">=</span><span class="name">mRNA</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">D90040</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g219411</span><span class="punctuation">;</span> <span class="name">PID</span><span class="operator">=</span><span class="name">g219412</span><span class="punctuation">;</span> <span class="name">SEQTYPE</span><span class="operator">=</span><span class="name">mRNA</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">BC015878</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g16198419</span><span class="punctuation">;</span> <span class="name">PID</span><span class="operator">=</span><span class="name">g16198420</span><span class="punctuation">;</span> <span class="name">SEQTYPE</span><span class="operator">=</span><span class="name">mRNA</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">CR407631</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g47115198</span><span class="punctuation">;</span> <span class="name">PID</span><span class="operator">=</span><span class="name">g47115199</span><span class="punctuation">;</span> <span class="name">SEQTYPE</span><span class="operator">=</span><span class="name">mRNA</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">BG569293</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g13576946</span><span class="punctuation">;</span> <span class="name">CLONE</span><span class="operator">=</span><span class="name">IMAGE</span><span class="punctuation">:</span><span class="literal number integer">4722596</span><span class="punctuation">;</span> <span class="name">END</span><span class="operator">=</span><span class="literal number integer">5</span><span class="literal string">'; LID=6989; SEQTYPE=EST; TRACE=44157214</span>
<span class="operator">...</span>
<span class="name">SEQUENCE</span>    <span class="name">ACC</span><span class="operator">=</span><span class="name">AU099534</span><span class="operator">.</span><span class="literal number integer">1</span><span class="punctuation">;</span> <span class="name">NID</span><span class="operator">=</span><span class="name">g13550663</span><span class="punctuation">;</span> <span class="name">CLONE</span><span class="operator">=</span><span class="name">HSI08034</span><span class="punctuation">;</span> <span class="name">END</span><span class="operator">=</span><span class="literal number integer">5</span><span class="literal string">'; LID=8800; SEQTYPE=EST</span>
<span class="operator">//</span>
</pre>
<p>这个记录展示了这个转录本（如 <tt class="docutils literal">SEQUENCE</tt> 行展示）是来自人的NAT2基因，编码en N-acetyltransferase。
<tt class="docutils literal">PROTSIM</tt> 显示的是和NAT2显著相似的蛋白质， <tt class="docutils literal">STS</tt> 展示的是基因组当中的STS位点。</p>
<p>我们使用 <tt class="docutils literal">Bio.UniGene</tt> 模块来解析UniGene文件：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">UniGene</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">input</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;myunigenefile.data&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">UniGene</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="punctuation">)</span>
</pre>
<p><tt class="docutils literal">UniGene.read</tt> 返回的是一个包含一些和UniGene记录的字段相对应属性的Python对象。例如，</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">ID</span>
<span class="literal string">&quot;Hs.2&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">title</span>
<span class="literal string">&quot;N-acetyltransferase 2 (arylamine N-acetyltransferase)&quot;</span>
</pre>
<p><tt class="docutils literal">EXPRESS</tt> 和 <tt class="docutils literal">RESTR_EXPR</tt> 两行被存储为字符串的Python列表：</p>
<pre class="code python literal-block">
<span class="punctuation">[</span><span class="literal string">'bone'</span><span class="punctuation">,</span> <span class="literal string">'connective tissue'</span><span class="punctuation">,</span> <span class="literal string">'intestine'</span><span class="punctuation">,</span> <span class="literal string">'liver'</span><span class="punctuation">,</span> <span class="literal string">'liver tumor'</span><span class="punctuation">,</span> <span class="literal string">'normal'</span><span class="punctuation">,</span> <span class="literal string">'soft tissue/muscle tissue tumor'</span><span class="punctuation">,</span> <span class="literal string">'adult'</span><span class="punctuation">]</span>
</pre>
<p>跟 <tt class="docutils literal">STS</tt> , <tt class="docutils literal">PROTSIM</tt> , 和 <tt class="docutils literal">SEQUENCE</tt> 相关的特有的对象被保存在如下键所对应的字典中：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">sts</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">acc</span>
<span class="literal string">'PMC310725P3'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">sts</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">unists</span>
<span class="literal string">'272646'</span>
</pre>
<p>和 <tt class="docutils literal">PROTSIM</tt> 、 <tt class="docutils literal">SEQUENCE</tt> 这两行相似。</p>
<p>我们使用 <tt class="docutils literal">Bio.UniGene</tt> 中的 <tt class="docutils literal">parse</tt> 函数来处理一个文件中包含多个UniGene记录的情况：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">UniGene</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">input</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;unigenerecords.data&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">UniGene</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">ID</span>
</pre>
</div>
</div>
<div class="section" id="id81">
<h2>9.13  使用代理</h2>
<p>通常状况下，你不需要使用代理，但是如果你的网络有问题的时候，我们有以下应对方法。在内部， <tt class="docutils literal">Bio.Entrez</tt> 使用
一个标准的 Python 库 <tt class="docutils literal">urllib</tt> 来访问 NCBI的服务器。这个将检查叫做 <tt class="docutils literal">http_proxy</tt> 的环境变量来自动配置简单
的代理服务。不幸的是，这个模块不支持需要认证的代理。</p>
<p>你可以选择设定环境变量 <tt class="docutils literal">http_proxy</tt> 。同样，你可以在Python脚本开头的地方设置这个参数，例如：</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">os</span>
<span class="name">os</span><span class="operator">.</span><span class="name">environ</span><span class="punctuation">[</span><span class="literal string">&quot;http_proxy&quot;</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="literal string">&quot;http://proxyhost.example.com:8080&quot;</span>
</pre>
<p>参见 <a class="reference external" href="http://www.python.org/doc/lib/module-urllib.html">urllib
文档</a> 获得更多信息。</p>
</div>
<div class="section" id="id82">
<h2>9.14  实例</h2>
<div class="section" id="pubmedmedline">
<h3>9.14.1  PubMed和Medline</h3>
<p>如果你是在医药领域或者对人类的问题感兴趣（或者尽管并你不感兴趣，大多数情况下也适用!），PubMed(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/PubMed/">http://www.ncbi.nlm.nih.gov/PubMed/</a>)
是一个包含了各方面的非常优秀的资源。像其他的一样，我们希望能够通过 Python 脚本从中抓取一些信息。</p>
<p>在这个例子当中，我们要查询PubMed当中所有跟Orchids相关的文章(see section <a class="reference external" href="#sec:orchids">2.3</a> for our motivation)。
我们首先看看有多少这样的文章：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">egquery</span><span class="punctuation">(</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;orchid&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">row</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;eGQueryResult&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;DbName&quot;</span><span class="punctuation">]</span><span class="operator">==</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal number integer">463</span>
</pre>
<p>现在我们使用 <tt class="docutils literal">Bio.Entrez.efetch</tt> 这个函数来下载这463篇文章的PubMed IDs：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;orchid&quot;</span><span class="punctuation">,</span> <span class="name">retmax</span><span class="operator">=</span><span class="literal number integer">463</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idlist</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">idlist</span>
</pre>
<p>返回值是一个Python列表，包含了所有和orchids相关文章的PubMed IDs：</p>
<pre class="code python literal-block">
<span class="punctuation">[</span><span class="literal string">'18680603'</span><span class="punctuation">,</span> <span class="literal string">'18665331'</span><span class="punctuation">,</span> <span class="literal string">'18661158'</span><span class="punctuation">,</span> <span class="literal string">'18627489'</span><span class="punctuation">,</span> <span class="literal string">'18627452'</span><span class="punctuation">,</span> <span class="literal string">'18612381'</span><span class="punctuation">,</span>
<span class="literal string">'18594007'</span><span class="punctuation">,</span> <span class="literal string">'18591784'</span><span class="punctuation">,</span> <span class="literal string">'18589523'</span><span class="punctuation">,</span> <span class="literal string">'18579475'</span><span class="punctuation">,</span> <span class="literal string">'18575811'</span><span class="punctuation">,</span> <span class="literal string">'18575690'</span><span class="punctuation">,</span>
<span class="operator">...</span>
</pre>
<p>这样我们就得到了这些信息，显然我们想要得到对应的Medline records和更多额外的信息。这里，我们将以纯文本的
形式下载和Medline records相关的信息，然后使用 <tt class="docutils literal">Bio.Medline</tt> 模块来解析他们：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Medline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="name">idlist</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;medline&quot;</span><span class="punctuation">,</span>
                           <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Medline</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>注意 - 我们完成了一次搜索和获取，NCBI更希望你在这种情况下使用他们的历史记录支持。请见章节 <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
<p>请记住 <tt class="docutils literal">records</tt> 是一个迭代器，所以你只能访问这些records一次。如果你想保存这些records，你需要把他们转成列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">)</span>
</pre>
<p>现在让我们迭代这些records，然后分别打印每一个record的信息：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;title:&quot;</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;TI&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;?&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;authors:&quot;</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;AU&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;?&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;source:&quot;</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;SO&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;?&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="keyword">print</span>
</pre>
<p>这个的输出结果是这样的:</p>
<pre class="code python literal-block">
<span class="name">title</span><span class="punctuation">:</span> <span class="name">Sex</span> <span class="name">pheromone</span> <span class="name">mimicry</span> <span class="operator word">in</span> <span class="name">the</span> <span class="name">early</span> <span class="name">spider</span> <span class="name">orchid</span> <span class="punctuation">(</span><span class="name">ophrys</span> <span class="name">sphegodes</span><span class="punctuation">):</span>
<span class="name">patterns</span> <span class="name">of</span> <span class="name">hydrocarbons</span> <span class="keyword">as</span> <span class="name">the</span> <span class="name">key</span> <span class="name">mechanism</span> <span class="keyword">for</span> <span class="name">pollination</span> <span class="name">by</span> <span class="name">sexual</span>
<span class="name">deception</span> <span class="punctuation">[</span><span class="name">In</span> <span class="name">Process</span> <span class="name">Citation</span><span class="punctuation">]</span>
<span class="name">authors</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal string">'Schiestl FP'</span><span class="punctuation">,</span> <span class="literal string">'Ayasse M'</span><span class="punctuation">,</span> <span class="literal string">'Paulus HF'</span><span class="punctuation">,</span> <span class="literal string">'Lofstedt C'</span><span class="punctuation">,</span> <span class="literal string">'Hansson BS'</span><span class="punctuation">,</span>
<span class="literal string">'Ibarra F'</span><span class="punctuation">,</span> <span class="literal string">'Francke W'</span><span class="punctuation">]</span>
<span class="name">source</span><span class="punctuation">:</span> <span class="name">J</span> <span class="name">Comp</span> <span class="name">Physiol</span> <span class="punctuation">[</span><span class="name">A</span><span class="punctuation">]</span> <span class="literal number integer">2000</span> <span class="name">Jun</span><span class="punctuation">;</span><span class="literal number integer">186</span><span class="punctuation">(</span><span class="literal number integer">6</span><span class="punctuation">):</span><span class="literal number integer">567</span><span class="operator">-</span><span class="literal number integer">74</span>
</pre>
<p>特别有意思的是作者的列表，作者的列表会作为一个标准的Python列表返回。这使得用标准的Python工具操作和搜索
变得简单。例如，我们可以像下面的代码这样循环读取所有条目来查找某个特定的作者：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">search_author</span> <span class="operator">=</span> <span class="literal string">&quot;Waits T&quot;</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="operator word">not</span> <span class="literal string">&quot;AU&quot;</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">continue</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">search_author</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;AU&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="literal string">&quot;Author </span><span class="literal string interpol">%s</span><span class="literal string"> found: </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">search_author</span><span class="punctuation">,</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;SO&quot;</span><span class="punctuation">])</span>
</pre>
<p>希望这个章节可以让你知道Entrez和Medline借口的能力和便利性和怎样同时使用他们。</p>
</div>
<div class="section" id="id83">
<h3>9.14.2  搜索，下载，和解析Entrez核酸记录</h3>
<p>这里我们将展示一个关于远程Entrez查询的简单例子。在 <a class="reference external" href="#sec:orchids">2.3</a> 章节，我们讲到了使用NCBI
的Entrez网站来搜索 NCBI 的核酸数据库来获得关于Cypripedioideae的信息。现在我们看看如何使用Python脚本
自动的处理。在这个例子当中，我们仅仅展示如何使用Entrez模块来连接，获取结果，解析他们。</p>
<p>首先，我们在下载这些结果之前，使用EGQuery来计算结果的数目。EGQuery 将会告诉我们在每个数据库中分别有多少
搜索结果，但在我们这个例子当中，我们只对核苷酸感兴趣：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">egquery</span><span class="punctuation">(</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Cypripedioideae&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">row</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;eGQueryResult&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;DbName&quot;</span><span class="punctuation">]</span><span class="operator">==</span><span class="literal string">&quot;nuccore&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal number integer">814</span>
</pre>
<p>所以，我们预期能找到814个 Entrez 核酸记录（这是我在2008年得到的结果；在未来这个结果应该会增加）。如果你得
到了高的不可思议的结果数目时，你可能得重新考虑是否需要下载所有的这些结果，下载是我们的下一步：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Cypripedioideae&quot;</span><span class="punctuation">,</span> <span class="name">retmax</span><span class="operator">=</span><span class="literal number integer">814</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>在这里, <tt class="docutils literal">record</tt> 是一个包含了搜索结果和一些辅助信息的Python字典。仅仅作为参考信息，让我们看看在这些字典当中
究竟存储了些什么内容：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">u'Count'</span><span class="punctuation">,</span> <span class="literal string">u'RetMax'</span><span class="punctuation">,</span> <span class="literal string">u'IdList'</span><span class="punctuation">,</span> <span class="literal string">u'TranslationSet'</span><span class="punctuation">,</span> <span class="literal string">u'RetStart'</span><span class="punctuation">,</span> <span class="literal string">u'QueryTranslation'</span><span class="punctuation">]</span>
</pre>
<p>首先, 让我们检查看看我们得到了多少个结果:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="literal string">'814'</span>
</pre>
<p>这个结果是我们所期望的。这814个结果被存在了 <tt class="docutils literal"><span class="pre">record['IdList']</span></tt> 中:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">])</span>
<span class="literal number integer">814</span>
</pre>
<p>让我们看看前五个结果:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">][:</span><span class="literal number integer">5</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'187237168'</span><span class="punctuation">,</span> <span class="literal string">'187372713'</span><span class="punctuation">,</span> <span class="literal string">'187372690'</span><span class="punctuation">,</span> <span class="literal string">'187372688'</span><span class="punctuation">,</span> <span class="literal string">'187372686'</span><span class="punctuation">]</span>
</pre>
<p>我们可以使用 <tt class="docutils literal">efetch</tt> 来下载这些结果. 尽管你可以一个一个的下载这些记录，但为了减少 NCBI 服务器的负载，最好呢还是
一次性的下载所有的结果。然而在这个情况下，你应该完美的使用在后面章节 <a class="reference external" href="#sec:entrez-webenv">9.15</a> 中会要讲到的历史记录特性。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">idlist</span> <span class="operator">=</span> <span class="literal string">&quot;,&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">][:</span><span class="literal number integer">5</span><span class="punctuation">])</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">idlist</span>
<span class="literal number integer">187237168</span><span class="punctuation">,</span><span class="literal number integer">187372713</span><span class="punctuation">,</span><span class="literal number integer">187372690</span><span class="punctuation">,</span><span class="literal number integer">187372688</span><span class="punctuation">,</span><span class="literal number integer">187372686</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="name">idlist</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
</pre>
<p>每个这样的records对应一个GenBank record.</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">u'GBSeq_moltype'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_source'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_sequence'</span><span class="punctuation">,</span>
 <span class="literal string">u'GBSeq_primary-accession'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_definition'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_accession-version'</span><span class="punctuation">,</span>
 <span class="literal string">u'GBSeq_topology'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_length'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_feature-table'</span><span class="punctuation">,</span>
 <span class="literal string">u'GBSeq_create-date'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_other-seqids'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_division'</span><span class="punctuation">,</span>
 <span class="literal string">u'GBSeq_taxonomy'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_references'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_update-date'</span><span class="punctuation">,</span>
 <span class="literal string">u'GBSeq_organism'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_locus'</span><span class="punctuation">,</span> <span class="literal string">u'GBSeq_strandedness'</span><span class="punctuation">]</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;GBSeq_primary-accession&quot;</span><span class="punctuation">]</span>
<span class="name">DQ110336</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;GBSeq_other-seqids&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'gb|DQ110336.1|'</span><span class="punctuation">,</span> <span class="literal string">'gi|187237168'</span><span class="punctuation">]</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;GBSeq_definition&quot;</span><span class="punctuation">]</span>
<span class="name">Cypripedium</span> <span class="name">calceolus</span> <span class="name">voucher</span> <span class="name">Davis</span> <span class="literal number oct">03</span><span class="operator">-</span><span class="literal number oct">03</span> <span class="name">A</span> <span class="name">maturase</span> <span class="punctuation">(</span><span class="name">matR</span><span class="punctuation">)</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">partial</span> <span class="name">cds</span><span class="punctuation">;</span>
<span class="name">mitochondrial</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;GBSeq_organism&quot;</span><span class="punctuation">]</span>
<span class="name">Cypripedium</span> <span class="name">calceolus</span>
</pre>
<p>你可以用这个来快速的开始搜索 —— 但是对于频繁的使用请见  <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
</div>
<div class="section" id="genbank-record">
<h3>9.14.3  搜索、下载和解析GenBank record</h3>
<p>GenBank record 格式是保存序列信息、序列特征和其他相关信息非常普遍的一种方法。这种格式是从 NCBI 数据库
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/">http://www.ncbi.nlm.nih.gov/</a> 获取信息非常好的一种方式 .</p>
<p>在这个例子当中，我们将展示怎样去查询 NCBI 数据库，根据query提取记录，然后使用 <tt class="docutils literal">Bio.SeqIO</tt> 解析他们 ——
在 <a class="reference external" href="#sec:SeqIO_GenBank_Online">5.3.1</a> 中提到过这些。简单起见，这个例子*不会*使用 WebEnv 历史记录特性
—— 请到 Section <a class="reference external" href="#sec:entrez-webenv">9.15</a> 查看。</p>
<p>首先，我们想要查询找出要获取的记录的ID。这里我们快速的检索我们最喜欢的一个物种 <em>Opuntia</em> (多刺的梨型仙人掌)。我们
可以做一个快速的检索来获得所有满足要求的GIs（GenBank标志符）。首先我们看看有多少个记录：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">egquery</span><span class="punctuation">(</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Opuntia AND rpl16&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">row</span> <span class="operator word">in</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;eGQueryResult&quot;</span><span class="punctuation">]:</span>
<span class="operator">...</span>     <span class="keyword">if</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;DbName&quot;</span><span class="punctuation">]</span><span class="operator">==</span><span class="literal string">&quot;nuccore&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="name">row</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">]</span>
<span class="operator">...</span>
<span class="literal number integer">9</span>
</pre>
<p>现在我们下载GenBank identifiers的列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nuccore&quot;</span><span class="punctuation">,</span> <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Opuntia AND rpl16&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gi_list</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gi_list</span>
<span class="punctuation">[</span><span class="literal string">'57240072'</span><span class="punctuation">,</span> <span class="literal string">'57240071'</span><span class="punctuation">,</span> <span class="literal string">'6273287'</span><span class="punctuation">,</span> <span class="literal string">'6273291'</span><span class="punctuation">,</span> <span class="literal string">'6273290'</span><span class="punctuation">,</span> <span class="literal string">'6273289'</span><span class="punctuation">,</span> <span class="literal string">'6273286'</span><span class="punctuation">,</span>
<span class="literal string">'6273285'</span><span class="punctuation">,</span> <span class="literal string">'6273284'</span><span class="punctuation">]</span>
</pre>
<p>现在我们使用这些GIs来下载GenBank records —— 注意在老的Biopython版本中，你必须将GI号用逗号隔开传递给Entrez，例如
在 Biopython 1.59中，你可以传递一个列表，下面的内容会为你做转换：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gi_str</span> <span class="operator">=</span> <span class="literal string">&quot;,&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">gi_list</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nuccore&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="name">gi_str</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
</pre>
<p>如果你想看原始的 GenBank 文件，你可以从这个句柄中读取并打印结果：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">text</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">text</span>
<span class="name">LOCUS</span>       <span class="name">AY851612</span>                 <span class="literal number integer">892</span> <span class="name">bp</span>    <span class="name">DNA</span>     <span class="name">linear</span>   <span class="name">PLN</span> <span class="literal number integer">10</span><span class="operator">-</span><span class="name">APR</span><span class="operator">-</span><span class="literal number integer">2007</span>
<span class="name">DEFINITION</span>  <span class="name">Opuntia</span> <span class="name">subulata</span> <span class="name">rpl16</span> <span class="name">gene</span><span class="punctuation">,</span> <span class="name">intron</span><span class="punctuation">;</span> <span class="name">chloroplast</span><span class="operator">.</span>
<span class="name">ACCESSION</span>   <span class="name">AY851612</span>
<span class="name">VERSION</span>     <span class="name">AY851612</span><span class="operator">.</span><span class="literal number integer">1</span>  <span class="name">GI</span><span class="punctuation">:</span><span class="literal number integer">57240072</span>
<span class="name">KEYWORDS</span>    <span class="operator">.</span>
<span class="name">SOURCE</span>      <span class="name">chloroplast</span> <span class="name">Austrocylindropuntia</span> <span class="name">subulata</span>
  <span class="name">ORGANISM</span>  <span class="name">Austrocylindropuntia</span> <span class="name">subulata</span>
            <span class="name">Eukaryota</span><span class="punctuation">;</span> <span class="name">Viridiplantae</span><span class="punctuation">;</span> <span class="name">Streptophyta</span><span class="punctuation">;</span> <span class="name">Embryophyta</span><span class="punctuation">;</span> <span class="name">Tracheophyta</span><span class="punctuation">;</span>
            <span class="name">Spermatophyta</span><span class="punctuation">;</span> <span class="name">Magnoliophyta</span><span class="punctuation">;</span> <span class="name">eudicotyledons</span><span class="punctuation">;</span> <span class="name">core</span> <span class="name">eudicotyledons</span><span class="punctuation">;</span>
            <span class="name">Caryophyllales</span><span class="punctuation">;</span> <span class="name">Cactaceae</span><span class="punctuation">;</span> <span class="name">Opuntioideae</span><span class="punctuation">;</span> <span class="name">Austrocylindropuntia</span><span class="operator">.</span>
<span class="name">REFERENCE</span>   <span class="literal number integer">1</span>  <span class="punctuation">(</span><span class="name">bases</span> <span class="literal number integer">1</span> <span class="name">to</span> <span class="literal number integer">892</span><span class="punctuation">)</span>
  <span class="name">AUTHORS</span>   <span class="name">Butterworth</span><span class="punctuation">,</span><span class="name">C</span><span class="operator">.</span><span class="name">A</span><span class="operator">.</span> <span class="operator word">and</span> <span class="name">Wallace</span><span class="punctuation">,</span><span class="name">R</span><span class="operator">.</span><span class="name">S</span><span class="operator">.</span>
<span class="operator">...</span>
</pre>
<p>在这个例子当中，我们只是得到了原始的记录。为了得到对Python友好的格式，我们可以使用 <tt class="docutils literal">Bio.SeqIO</tt> 将GenBank
数据转化成 <tt class="docutils literal">SeqRecord</tt> 对象，包括 <tt class="docutils literal">SeqFeature</tt> 对象 (请见 Chapter <a class="reference external" href="#chapter:Bio.SeqIO">5</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nuccore&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="name">gi_str</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;gb&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">)</span>
</pre>
<p>我们现在可以逐个查看这些record来寻找我们感兴趣的信息：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="operator">...</span>    <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">, length </span><span class="literal string interpol">%i</span><span class="literal string">, with </span><span class="literal string interpol">%i</span><span class="literal string"> features&quot;</span> \
<span class="operator">&gt;&gt;&gt;</span> <span class="operator">...</span>           <span class="operator">%</span> <span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">name</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">),</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">))</span>
<span class="name">AY851612</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">892</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AY851611</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">881</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191661</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">895</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191665</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">902</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191664</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">899</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191663</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">899</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191660</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">893</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191659</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">894</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
<span class="name">AF191658</span><span class="punctuation">,</span> <span class="name">length</span> <span class="literal number integer">896</span><span class="punctuation">,</span> <span class="keyword">with</span> <span class="literal number integer">3</span> <span class="name">features</span>
</pre>
<p>使用这些自动的查询提取功能相对于手动处理是一个很大的进步。尽管这些模块需要遵守NCBI每秒钟最多三次的规则，然而NCBI
有其他像避开高峰时刻的建议。请见章节 <a class="reference external" href="#sec:entrez-guidelines">9.1</a>. 尤其需要注意的是，这个例子没有
用到 WebEnv 历史记录特性。你应该使用这个来完成一些琐碎的搜索和下载的工作，请见章节 <a class="reference external" href="#sec:entrez-webenv">9.15</a>.</p>
<p>最后，如果你计划重复你的分析，你应该下载这些record <em>一次</em> ，然后将他们保存在你的硬盘里，在本地进行分析；而不是
从 NCBI 下载之后就马上进行分析（像这个例子一样）。</p>
</div>
<div class="section" id="id84">
<h3>9.14.4  查看物种的谱系关系</h3>
<p>仍然以植物为例子，让我们找出Cyripedioideae兰花家族的谱系。首先让我们在Taxonomy数据库中查找跟Cypripedioideae
相关的记录，确实找到了一个确切的 NCBI taxonomy 标识号：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>     <span class="comment"># Always tell NCBI who you are</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;Taxonomy&quot;</span><span class="punctuation">,</span> <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Cypripedioideae&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'158330'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="literal string">'158330'</span>
</pre>
<p>现在，我们使用 <tt class="docutils literal">efetch</tt> 从 Taxonomy 数据库中下载这些条目，然后解析它：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;Taxonomy&quot;</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;158330&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;xml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>再次，这个record保存了许多的信息：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="literal string">u'Lineage'</span><span class="punctuation">,</span> <span class="literal string">u'Division'</span><span class="punctuation">,</span> <span class="literal string">u'ParentTaxId'</span><span class="punctuation">,</span> <span class="literal string">u'PubDate'</span><span class="punctuation">,</span> <span class="literal string">u'LineageEx'</span><span class="punctuation">,</span>
 <span class="literal string">u'CreateDate'</span><span class="punctuation">,</span> <span class="literal string">u'TaxId'</span><span class="punctuation">,</span> <span class="literal string">u'Rank'</span><span class="punctuation">,</span> <span class="literal string">u'GeneticCode'</span><span class="punctuation">,</span> <span class="literal string">u'ScientificName'</span><span class="punctuation">,</span>
 <span class="literal string">u'MitoGeneticCode'</span><span class="punctuation">,</span> <span class="literal string">u'UpdateDate'</span><span class="punctuation">]</span>
</pre>
<p>我们可以直接从这个record获得谱系信息：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Lineage&quot;</span><span class="punctuation">]</span>
<span class="literal string">'cellular organisms; Eukaryota; Viridiplantae; Streptophyta; Streptophytina;</span>
 <span class="name">Embryophyta</span><span class="punctuation">;</span> <span class="name">Tracheophyta</span><span class="punctuation">;</span> <span class="name">Euphyllophyta</span><span class="punctuation">;</span> <span class="name">Spermatophyta</span><span class="punctuation">;</span> <span class="name">Magnoliophyta</span><span class="punctuation">;</span>
 <span class="name">Liliopsida</span><span class="punctuation">;</span> <span class="name">Asparagales</span><span class="punctuation">;</span> <span class="name">Orchidaceae</span><span class="literal string">'</span>
</pre>
<p>这个record数据包含的信息远远超过在这里显示的 —— 例如查看 <tt class="docutils literal">&quot;LineageEx&quot;</tt> 而不是 <tt class="docutils literal">&quot;Lineage&quot;</tt> 相关的
信息，你也可以得到谱系里面的 NCBI taxon 标识号信息。</p>
</div>
</div>
<div class="section" id="webenv">
<h2>9.15  使用历史记录和WebEnv</h2>
<p>通常，你想做一系列相关的查询。最典型的是，进行一个搜索，精炼搜索，然后提取详细的搜索结果。你 <em>可以</em> 通过一系列
独立的调用Entrez来完成这些工作。然而，NCBI更希望你利用历史记录支持的优势来完成这个 - 例如将ESearch
和EFetch结合起来。</p>
<p>另外一个关于历史记录支持，典型的使用是结合EPost和EFetch。你可以使用EPost来上传一个标识号的
列表，这样就开始一些新的history session。接下来你就可以用EFetch指向这个session来下载这些数据（而不是那些
标识号）。</p>
<div class="section" id="history">
<h3>9.15.1  利用 history 来搜索和下载序列</h3>
<p>假设我们想搜索和下载所有的 <em>Opuntia</em> rpl16核酸序列，然后将它们保存到一个FASTA文件里。就像章节 <a class="reference external" href="#sec:entrez-search-fetch-genbank">9.14.3</a>
里一样, 我们可以简单的用 <tt class="docutils literal">Bio.Entrez.esearch()</tt> 得到一个GI号的列表，然后调用 <tt class="docutils literal">Bio.Entrez.efetch()</tt>
来下载他们。</p>
<p>然而，被认同的方法是使用历史记录特性来进行搜索。然后，我们可以通过指向这些搜索结果的引用来获取他们 -
NCBI 将会提前进行缓冲。</p>
<p>为此，调用 <tt class="docutils literal">Bio.Entrez.esearch()</tt> 是正常的，但是需要额外的 <tt class="docutils literal"><span class="pre">usehistory=&quot;y&quot;</span></tt> 参数，</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;history.user&#64;example.com&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">search_handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span><span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Opuntia[orgn] and rpl16&quot;</span><span class="punctuation">,</span>
                                   <span class="name">usehistory</span><span class="operator">=</span><span class="literal string">&quot;y&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">search_results</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">search_handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">search_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>当你得到XML输出的时候，它仍然包括了常见的搜索结果：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">gi_list</span> <span class="operator">=</span> <span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;IdList&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">count</span> <span class="operator">=</span> <span class="name builtin">int</span><span class="punctuation">(</span><span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">])</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">assert</span> <span class="name">count</span> <span class="operator">==</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">gi_list</span><span class="punctuation">)</span>
</pre>
<p>然而，你将得到两个额外的信息， <tt class="docutils literal">WebEnv</tt> 会话cookie 和 <tt class="docutils literal">QueryKey</tt> :</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">webenv</span> <span class="operator">=</span> <span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;WebEnv&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">query_key</span> <span class="operator">=</span> <span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;QueryKey&quot;</span><span class="punctuation">]</span>
</pre>
<p>将这些值保存到 <tt class="docutils literal">session_cookie</tt> 和 <tt class="docutils literal">query_key</tt> 后，我们可以使用它们作为 <tt class="docutils literal">Bio.Entrez.efetch()</tt> 的
参数，而不用提供GI numbers的identifiers。</p>
<p>对于小数据量你一次下载所有的数据也没有关系，但是最好能够分批下载。你可以使用 <tt class="docutils literal">restart</tt> 和 <tt class="docutils literal">retmax</tt> 来说明
哪一部分搜索结果是你想得到的（条目以0开始计算，返回结果的最大数目）。例如：</p>
<pre class="code python literal-block">
<span class="name">batch_size</span> <span class="operator">=</span> <span class="literal number integer">3</span>
<span class="name">out_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;orchid_rpl16.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">start</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="name">count</span><span class="punctuation">,</span><span class="name">batch_size</span><span class="punctuation">):</span>
    <span class="name">end</span> <span class="operator">=</span> <span class="name builtin">min</span><span class="punctuation">(</span><span class="name">count</span><span class="punctuation">,</span> <span class="name">start</span><span class="operator">+</span><span class="name">batch_size</span><span class="punctuation">)</span>
    <span class="keyword">print</span> <span class="literal string">&quot;Going to download record </span><span class="literal string interpol">%i</span><span class="literal string"> to </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">start</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">end</span><span class="punctuation">)</span>
    <span class="name">fetch_handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;nucleotide&quot;</span><span class="punctuation">,</span> <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">,</span>
                                 <span class="name">retstart</span><span class="operator">=</span><span class="name">start</span><span class="punctuation">,</span> <span class="name">retmax</span><span class="operator">=</span><span class="name">batch_size</span><span class="punctuation">,</span>
                                 <span class="name">webenv</span><span class="operator">=</span><span class="name">webenv</span><span class="punctuation">,</span> <span class="name">query_key</span><span class="operator">=</span><span class="name">query_key</span><span class="punctuation">)</span>
    <span class="name">data</span> <span class="operator">=</span> <span class="name">fetch_handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
    <span class="name">fetch_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
    <span class="name">out_handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
<span class="name">out_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>我们以此为例来说明，这个例子分三次来下载FASTA records。除非你是要下载基因组或者染色体数目，你最好选取一个
比较大的batch大小。</p>
</div>
<div class="section" id="id85">
<h3>9.15.2  利用history来搜索和下载综述</h3>
<p>这是另外一个history的例子，搜索过去几年当中发表的关于 <em>Opuntia</em> 的文章，然后下载到一个MedLine格式的文件里：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;history.user&#64;example.com&quot;</span>
<span class="name">search_results</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">esearch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span>
                                            <span class="name">term</span><span class="operator">=</span><span class="literal string">&quot;Opuntia[ORGN]&quot;</span><span class="punctuation">,</span>
                                            <span class="name">reldate</span><span class="operator">=</span><span class="literal number integer">365</span><span class="punctuation">,</span> <span class="name">datetype</span><span class="operator">=</span><span class="literal string">&quot;pdat&quot;</span><span class="punctuation">,</span>
                                            <span class="name">usehistory</span><span class="operator">=</span><span class="literal string">&quot;y&quot;</span><span class="punctuation">))</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name builtin">int</span><span class="punctuation">(</span><span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">])</span>
<span class="keyword">print</span> <span class="literal string">&quot;Found </span><span class="literal string interpol">%i</span><span class="literal string"> results&quot;</span> <span class="operator">%</span> <span class="name">count</span>

<span class="name">batch_size</span> <span class="operator">=</span> <span class="literal number integer">10</span>
<span class="name">out_handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;recent_orchid_papers.txt&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">start</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="name">count</span><span class="punctuation">,</span><span class="name">batch_size</span><span class="punctuation">):</span>
    <span class="name">end</span> <span class="operator">=</span> <span class="name builtin">min</span><span class="punctuation">(</span><span class="name">count</span><span class="punctuation">,</span> <span class="name">start</span><span class="operator">+</span><span class="name">batch_size</span><span class="punctuation">)</span>
    <span class="keyword">print</span> <span class="literal string">&quot;Going to download record </span><span class="literal string interpol">%i</span><span class="literal string"> to </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">start</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">end</span><span class="punctuation">)</span>
    <span class="name">fetch_handle</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">efetch</span><span class="punctuation">(</span><span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span>
                                 <span class="name">rettype</span><span class="operator">=</span><span class="literal string">&quot;medline&quot;</span><span class="punctuation">,</span> <span class="name">retmode</span><span class="operator">=</span><span class="literal string">&quot;text&quot;</span><span class="punctuation">,</span>
                                 <span class="name">retstart</span><span class="operator">=</span><span class="name">start</span><span class="punctuation">,</span> <span class="name">retmax</span><span class="operator">=</span><span class="name">batch_size</span><span class="punctuation">,</span>
                                 <span class="name">webenv</span><span class="operator">=</span><span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;WebEnv&quot;</span><span class="punctuation">],</span>
                                 <span class="name">query_key</span><span class="operator">=</span><span class="name">search_results</span><span class="punctuation">[</span><span class="literal string">&quot;QueryKey&quot;</span><span class="punctuation">])</span>
    <span class="name">data</span> <span class="operator">=</span> <span class="name">fetch_handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
    <span class="name">fetch_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
    <span class="name">out_handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
<span class="name">out_handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>在写这份文档的时候，这个搜索返回了28个匹配结果 - 但是因为这个是跟时间相关的搜索，因此返回结果会发生变化。
像在上面 <a class="reference external" href="#subsec:entrez-and-medline">9.12.1</a> 讲到的一样, 你可以使用 <tt class="docutils literal">Bio.Medline</tt> 来解析
保存下来的记录。</p>
</div>
<div class="section" id="id86">
<h3>9.15.3  搜索引用文章</h3>
<p>回到 Section <a class="reference external" href="#sec:elink">9.7</a> 我们提到可以使用ELink来搜索制定文章的引用。不幸的是，这个只包含
PubMed Central（为PubMed中所有文献来做这个事情，意味这NIH将要付出更多的工作）包含的那些杂志。让我们以
Biopython PDB parser文章为例来试试看， PubMed ID 14630660：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Entrez</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">email</span> <span class="operator">=</span> <span class="literal string">&quot;A.N.Other&#64;example.com&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pmid</span> <span class="operator">=</span> <span class="literal string">&quot;14630660&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">results</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">elink</span><span class="punctuation">(</span><span class="name">dbfrom</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pmc&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                    <span class="name">LinkName</span><span class="operator">=</span><span class="literal string">&quot;pubmed_pmc_refs&quot;</span><span class="punctuation">,</span> <span class="name">from_uid</span><span class="operator">=</span><span class="name">pmid</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pmc_ids</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">link</span><span class="punctuation">[</span><span class="literal string">&quot;Id&quot;</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name">link</span> <span class="operator word">in</span> <span class="name">results</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Link&quot;</span><span class="punctuation">]]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pmc_ids</span>
<span class="punctuation">[</span><span class="literal string">'2744707'</span><span class="punctuation">,</span> <span class="literal string">'2705363'</span><span class="punctuation">,</span> <span class="literal string">'2682512'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'1190160'</span><span class="punctuation">]</span>
</pre>
<p>好极了 - 11篇文章。但是为什么没有Biopython应用笔记（PubMed ID 19304878）呢？好吧，你可能已经从变量的
名称中猜到了，实际上他们不是PubMed IDs，而是PubMed Central IDs。我们的应用笔记是列表当中第三个引用的
文章， PMCID 2682512。</p>
<p>那么，如果（像我）你希望得到的是PubMed IDs的列表的话，该怎么做呢？好吧，你可以使用再次使用ELink来更改他们。
这将成为两步处理，所以你应该使用历史记录特性来完成这个工作（章节 <a class="reference external" href="#sec:entrez-webenv">9.15</a>）。</p>
<p>但是首先，让我们使用更直接的方法来进行第二次调用ELink：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">results2</span> <span class="operator">=</span> <span class="name">Entrez</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">Entrez</span><span class="operator">.</span><span class="name">elink</span><span class="punctuation">(</span><span class="name">dbfrom</span><span class="operator">=</span><span class="literal string">&quot;pmc&quot;</span><span class="punctuation">,</span> <span class="name">db</span><span class="operator">=</span><span class="literal string">&quot;pubmed&quot;</span><span class="punctuation">,</span> <span class="name">LinkName</span><span class="operator">=</span><span class="literal string">&quot;pmc_pubmed&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                     <span class="name">from_uid</span><span class="operator">=</span><span class="literal string">&quot;,&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">pmc_ids</span><span class="punctuation">)))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pubmed_ids</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">link</span><span class="punctuation">[</span><span class="literal string">&quot;Id&quot;</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name">link</span> <span class="operator word">in</span> <span class="name">results2</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;LinkSetDb&quot;</span><span class="punctuation">][</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">&quot;Link&quot;</span><span class="punctuation">]]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pubmed_ids</span>
<span class="punctuation">[</span><span class="literal string">'19698094'</span><span class="punctuation">,</span> <span class="literal string">'19450287'</span><span class="punctuation">,</span> <span class="literal string">'19304878'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'15985178'</span><span class="punctuation">]</span>
</pre>
<p>这次，你可以立即的看到Biopython应用笔记作为第三个hit（PubMed ID 19304878）。</p>
<p>现在，让我们重新使用历史记录再试一遍 … <em>TODO</em>.</p>
<p>最终，不要忘记在Entrez调用的时候，加上你 <em>自己</em> 的电子邮箱地址。</p>
</div>
</div>
</div>
<div class="section" id="swiss-protexpasy">
<h1>第10章 Swiss-Prot和ExPASy</h1>
<div class="section" id="swiss-prot">
<h2>10.1  解析Swiss-Prot文件</h2>
<p>Swiss-Prot
( <a class="reference external" href="http://www.expasy.org/sprot">http://www.expasy.org/sprot</a> )是一个
蛋白质序列数据库。 Biopython能够解析纯文本的Swiss-Prot文件,
这种格式也被Swiss-Prot、TrEMBL和PIRPSD的UniProt数据库使用。然而我们并
不支持UniProKB的XML格式文件。</p>
<div class="section" id="parsing-swiss-prot-records">
<h3>10.1.1  Parsing Swiss-Prot records</h3>
<p>在 <a class="reference external" href="#sec:SeqIO_ExPASy_and_SwissProt">5.3.2</a> 章节中, 我们描述过怎样将一个
Swiss-Prot记录中的序列提出来作为一个 <tt class="docutils literal">SeqRecord</tt> 对象。 此外，你可以将
Swiss-Prot记录存到  <tt class="docutils literal">Bio.SwissProt.Record</tt> 对象, 这实际上存储了Swiss-Prot记录
中所包含的的全部信息。在这部分我们将介绍怎样从一个Swiss-Prot文件中提
取 <tt class="docutils literal">Bio.SwissProt.Record</tt> 对象。</p>
<p>为了解析Swiss-Prot记录，我们首先需要得到一个Swiss-Prot记录文件。根据该Swiss-Prot
记录的储存位置和储存方式，获取该记录文件的方式也有所不同：</p>
<ul>
<li><p class="first">本地打开Swiss-Prot文件：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;myswissprotfile.dat&quot;</span><span class="punctuation">)</span>
</pre>
</li>
<li><p class="first">打开使用gzip压缩的Swiss-Prot文件：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">gzip</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">gzip</span><span class="operator">.</span><span class="name">open</span><span class="punctuation">(</span><span class="literal string">&quot;myswissprotfile.dat.gz&quot;</span><span class="punctuation">)</span>
</pre>
</li>
<li><p class="first">在线打开Swiss-Prot文件：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">urllib</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">urllib</span><span class="operator">.</span><span class="name">urlopen</span><span class="punctuation">(</span><span class="literal string">&quot;http://www.somelocation.org/data/someswissprotfile.dat&quot;</span><span class="punctuation">)</span>
</pre>
</li>
<li><p class="first">从ExPASy数据库在线打开Swiss-Prot文件
(见 <a class="reference external" href="#subsec:expasy_swissprot">10.5.1</a> 章节):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_sprot_raw</span><span class="punctuation">(</span><span class="name">myaccessionnumber</span><span class="punctuation">)</span>
</pre>
</li>
</ul>
<p>对于解析来说，关键点在于Swiss-Prot格式的数据，而不是获取它的方式。</p>
<p>我们可以用 <a class="reference external" href="#sec:SeqIO_ExPASy_and_SwissProt">5.3.2</a> 章节中描述的方式，
通过 <tt class="docutils literal">Bio.SeqIO</tt> 来获取格式未知的 <tt class="docutils literal">SeqRecord</tt> 对象。此外，我们也可以
用 <tt class="docutils literal">Bio.SwissProt</tt> 来获取更加匹配基本文件格式的 <tt class="docutils literal">Bio.SwissProt.Record</tt> 对象。</p>
<p>我们使用 <tt class="docutils literal">read()</tt> 函数来从文件中读取一个Swiss-Prot记录：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SwissProt</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SwissProt</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>该函数只适用于仅存储了一个Swiss-Prot记录的文件，而当文件中没有或存在
多个记录时使用该函数，会出现 <tt class="docutils literal">ValueError</tt> 提示。</p>
<p>现在我们可以输出一些与这些记录相关的信息：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">description</span>
<span class="literal string">'RecName: Full=Chalcone synthase 3; EC=2.3.1.74; AltName: Full=Naringenin-chalcone synthase 3;'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">ref</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">references</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;authors:&quot;</span><span class="punctuation">,</span> <span class="name">ref</span><span class="operator">.</span><span class="name">authors</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;title:&quot;</span><span class="punctuation">,</span> <span class="name">ref</span><span class="operator">.</span><span class="name">title</span>
<span class="operator">...</span>
<span class="name">authors</span><span class="punctuation">:</span> <span class="name">Liew</span> <span class="name">C</span><span class="operator">.</span><span class="name">F</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Lim</span> <span class="name">S</span><span class="operator">.</span><span class="name">H</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Loh</span> <span class="name">C</span><span class="operator">.</span><span class="name">S</span><span class="operator">.</span><span class="punctuation">,</span> <span class="name">Goh</span> <span class="name">C</span><span class="operator">.</span><span class="name">J</span><span class="operator">.</span><span class="punctuation">;</span>
<span class="name">title</span><span class="punctuation">:</span> <span class="literal string">&quot;Molecular cloning and sequence analysis of chalcone synthase cDNAs of</span>
<span class="name">Bromheadia</span> <span class="name">finlaysoniana</span><span class="operator">.</span><span class="literal string">&quot;;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">organism_classification</span>
<span class="punctuation">[</span><span class="literal string">'Eukaryota'</span><span class="punctuation">,</span> <span class="literal string">'Viridiplantae'</span><span class="punctuation">,</span> <span class="literal string">'Streptophyta'</span><span class="punctuation">,</span> <span class="literal string">'Embryophyta'</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal string">'Bromheadia'</span><span class="punctuation">]</span>
</pre>
<p>为了解析包含多个Swiss-Prot记录的文件，我们使用 <tt class="docutils literal">parse</tt> 函数。这个函数能够让我们对
文件中的记录进行循环迭代操作。</p>
<p>比如，我们要解析整个Swiss-Prot数据库并且收集所有的描述。你可以从
<a class="reference external" href="ftp://ftp.expasy.org/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.dat.gz">ExPAYs FTP site</a>
下载这些gzip压缩文件 <tt class="docutils literal">uniprot_sprot.dat.gz</tt> (大约 300MB)。文件中含有
<tt class="docutils literal">uniprot_sprot.dat</tt> 一个文件(至少1.5GB)。</p>
<p>如同这一部分刚开始所描述的，你可以按照如下所示的方法使用python
的 <tt class="docutils literal">gzip</tt> 模块打开并解压 <tt class="docutils literal">.gz</tt> 文件:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">gzip</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">gzip</span><span class="operator">.</span><span class="name">open</span><span class="punctuation">(</span><span class="literal string">&quot;uniprot_sprot.dat.gz&quot;</span><span class="punctuation">)</span>
</pre>
<p>然而，解压一个大文件比较耗时，而且每次用这种方式打开一个
文件都是比较慢的。所以，如果你有空闲的硬盘空间并且在
最开始就在硬盘里通过解压到来得到 <tt class="docutils literal">uniprot_sprot.dat</tt> ，这样能够在以后就可以像平常那样来打开文件：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;uniprot_sprot.dat&quot;</span><span class="punctuation">)</span>
</pre>
<p>到2009年6月为止，从ExPASy下载下来的整个Swiss-Prot数据库一共
有468851个Swiss-Prot记录，一种建立关于这些记录的描述列表的
间接方式就是使用一种列表解析：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SwissProt</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;uniprot_sprot.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">descriptions</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">record</span><span class="operator">.</span><span class="name">description</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SwissProt</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">descriptions</span><span class="punctuation">)</span>
<span class="literal number integer">468851</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">descriptions</span><span class="punctuation">[:</span><span class="literal number integer">5</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'RecName: Full=Protein MGF 100-1R;'</span><span class="punctuation">,</span>
 <span class="literal string">'RecName: Full=Protein MGF 100-1R;'</span><span class="punctuation">,</span>
 <span class="literal string">'RecName: Full=Protein MGF 100-1R;'</span><span class="punctuation">,</span>
 <span class="literal string">'RecName: Full=Protein MGF 100-1R;'</span><span class="punctuation">,</span>
 <span class="literal string">'RecName: Full=Protein MGF 100-2L;'</span><span class="punctuation">]</span>
</pre>
<p>或者对记录迭代器使用for循环：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SwissProt</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">descriptions</span> <span class="operator">=</span> <span class="punctuation">[]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;uniprot_sprot.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SwissProt</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="name">descriptions</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">description</span><span class="punctuation">)</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">descriptions</span><span class="punctuation">)</span>
<span class="literal number integer">468851</span>
</pre>
<p>由于输入文件太大，这两种方法在我的新台式机上花费大约十一分钟（用解压好的
<tt class="docutils literal">uniprot_sprot.dat</tt> 作为输入文件）。</p>
<p>从Swiss-Prot记录中提取任何你想要的信息也同样简单。比如你想看看一个
Swiss-Prot记录中的成员，就输入：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">dir</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="punctuation">[</span><span class="literal string">'__ doc__ '</span><span class="punctuation">,</span> <span class="literal string">'__ init__ '</span><span class="punctuation">,</span> <span class="literal string">'__ module__ '</span><span class="punctuation">,</span> <span class="literal string">'accessions'</span><span class="punctuation">,</span> <span class="literal string">'annotation_update'</span><span class="punctuation">,</span>
<span class="literal string">'comments'</span><span class="punctuation">,</span> <span class="literal string">'created'</span><span class="punctuation">,</span> <span class="literal string">'cross_references'</span><span class="punctuation">,</span> <span class="literal string">'data_class'</span><span class="punctuation">,</span> <span class="literal string">'description'</span><span class="punctuation">,</span>
<span class="literal string">'entry_name'</span><span class="punctuation">,</span> <span class="literal string">'features'</span><span class="punctuation">,</span> <span class="literal string">'gene_name'</span><span class="punctuation">,</span> <span class="literal string">'host_organism'</span><span class="punctuation">,</span> <span class="literal string">'keywords'</span><span class="punctuation">,</span>
<span class="literal string">'molecule_type'</span><span class="punctuation">,</span> <span class="literal string">'organelle'</span><span class="punctuation">,</span> <span class="literal string">'organism'</span><span class="punctuation">,</span> <span class="literal string">'organism_classification'</span><span class="punctuation">,</span>
<span class="literal string">'references'</span><span class="punctuation">,</span> <span class="literal string">'seqinfo'</span><span class="punctuation">,</span> <span class="literal string">'sequence'</span><span class="punctuation">,</span> <span class="literal string">'sequence_length'</span><span class="punctuation">,</span>
<span class="literal string">'sequence_update'</span><span class="punctuation">,</span> <span class="literal string">'taxonomy_id'</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="id87">
<h3>10.1.2  解析Swiss-Prot关键词和分类列表</h3>
<p>Swiss-Prot也会提供一个 <tt class="docutils literal">keywlist.txt</tt> 文件，该文件列出了Swiss-Prot中所用到
的关键词和分类。其中所包含的词条形式如下：</p>
<pre class="code python literal-block">
<span class="name">ID</span>   <span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span><span class="operator">.</span>
<span class="name">AC</span>   <span class="name">KW</span><span class="operator">-</span><span class="literal number oct">0001</span>
<span class="name">DE</span>   <span class="name">Protein</span> <span class="name">which</span> <span class="name">contains</span> <span class="name">at</span> <span class="name">least</span> <span class="name">one</span> <span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span> <span class="name">iron</span><span class="operator">-</span><span class="name">sulfur</span> <span class="name">cluster</span><span class="punctuation">:</span> <span class="literal number integer">2</span> <span class="name">iron</span>
<span class="name">DE</span>   <span class="name">atoms</span> <span class="name">complexed</span> <span class="name">to</span> <span class="literal number integer">2</span> <span class="name">inorganic</span> <span class="name">sulfides</span> <span class="operator word">and</span> <span class="literal number integer">4</span> <span class="name">sulfur</span> <span class="name">atoms</span> <span class="name">of</span>
<span class="name">DE</span>   <span class="name">cysteines</span> <span class="keyword namespace">from</span> <span class="name namespace">the</span> <span class="name namespace">protein.</span>
<span class="name">SY</span>   <span class="name">Fe2S2</span><span class="punctuation">;</span> <span class="punctuation">[</span><span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span><span class="punctuation">]</span> <span class="name">cluster</span><span class="punctuation">;</span> <span class="punctuation">[</span><span class="name">Fe2S2</span><span class="punctuation">]</span> <span class="name">cluster</span><span class="punctuation">;</span> <span class="name">Fe2</span><span class="operator">/</span><span class="name">S2</span> <span class="punctuation">(</span><span class="name">inorganic</span><span class="punctuation">)</span> <span class="name">cluster</span><span class="punctuation">;</span>
<span class="name">SY</span>   <span class="name">Di</span><span class="operator">-</span><span class="name">mu</span><span class="operator">-</span><span class="name">sulfido</span><span class="operator">-</span><span class="name">diiron</span><span class="punctuation">;</span> <span class="literal number integer">2</span> <span class="name">iron</span><span class="punctuation">,</span> <span class="literal number integer">2</span> <span class="name">sulfur</span> <span class="name">cluster</span> <span class="name">binding</span><span class="operator">.</span>
<span class="name">GO</span>   <span class="name">GO</span><span class="punctuation">:</span><span class="literal number oct">0051537</span><span class="punctuation">;</span> <span class="literal number integer">2</span> <span class="name">iron</span><span class="punctuation">,</span> <span class="literal number integer">2</span> <span class="name">sulfur</span> <span class="name">cluster</span> <span class="name">binding</span>
<span class="name">HI</span>   <span class="name">Ligand</span><span class="punctuation">:</span> <span class="name">Iron</span><span class="punctuation">;</span> <span class="name">Iron</span><span class="operator">-</span><span class="name">sulfur</span><span class="punctuation">;</span> <span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span><span class="operator">.</span>
<span class="name">HI</span>   <span class="name">Ligand</span><span class="punctuation">:</span> <span class="name">Metal</span><span class="operator">-</span><span class="name">binding</span><span class="punctuation">;</span> <span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span><span class="operator">.</span>
<span class="name">CA</span>   <span class="name">Ligand</span><span class="operator">.</span>
<span class="operator">//</span>
<span class="name">ID</span>   <span class="literal number integer">3</span><span class="name">D</span><span class="operator">-</span><span class="name">structure</span><span class="operator">.</span>
<span class="name">AC</span>   <span class="name">KW</span><span class="operator">-</span><span class="literal number oct">0002</span>
<span class="name">DE</span>   <span class="name">Protein</span><span class="punctuation">,</span> <span class="operator word">or</span> <span class="name">part</span> <span class="name">of</span> <span class="name">a</span> <span class="name">protein</span><span class="punctuation">,</span> <span class="name">whose</span> <span class="name">three</span><span class="operator">-</span><span class="name">dimensional</span> <span class="name">structure</span> <span class="name">has</span>
<span class="name">DE</span>   <span class="name">been</span> <span class="name">resolved</span> <span class="name">experimentally</span> <span class="punctuation">(</span><span class="keyword">for</span> <span class="name">example</span> <span class="name">by</span> <span class="name">X</span><span class="operator">-</span><span class="name">ray</span> <span class="name">crystallography</span> <span class="operator word">or</span>
<span class="name">DE</span>   <span class="name">NMR</span> <span class="name">spectroscopy</span><span class="punctuation">)</span> <span class="operator word">and</span> <span class="name">whose</span> <span class="name">coordinates</span> <span class="name">are</span> <span class="name">available</span> <span class="operator word">in</span> <span class="name">the</span> <span class="name">PDB</span>
<span class="name">DE</span>   <span class="name">database</span><span class="operator">.</span> <span class="name">Can</span> <span class="name">also</span> <span class="name">be</span> <span class="name">used</span> <span class="keyword">for</span> <span class="name">theoretical</span> <span class="name">models</span><span class="operator">.</span>
<span class="name">HI</span>   <span class="name">Technical</span> <span class="name">term</span><span class="punctuation">:</span> <span class="literal number integer">3</span><span class="name">D</span><span class="operator">-</span><span class="name">structure</span><span class="operator">.</span>
<span class="name">CA</span>   <span class="name">Technical</span> <span class="name">term</span><span class="operator">.</span>
<span class="operator">//</span>
<span class="name">ID</span>   <span class="literal number integer">3</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">4</span><span class="name">S</span><span class="operator">.</span>
<span class="operator">...</span>
</pre>
<p>文件中的词条可以通过使用 <tt class="docutils literal">Bio.SwissProt.KeyWList</tt> 模块中的 <tt class="docutils literal">parse</tt> 函数
来解析，并且每一个词条都会被存储在名为 <tt class="docutils literal">Bio.SwissProt.KeyWList.Record</tt> 的
python字典里。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SwissProt</span> <span class="keyword namespace">import</span> <span class="name">KeyWList</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;keywlist.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">KeyWList</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">'ID'</span><span class="punctuation">]</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">'DE'</span><span class="punctuation">]</span>
</pre>
<p>这些命令行将会输出：</p>
<pre class="code python literal-block">
<span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span><span class="operator">.</span>
<span class="name">Protein</span> <span class="name">which</span> <span class="name">contains</span> <span class="name">at</span> <span class="name">least</span> <span class="name">one</span> <span class="literal number integer">2</span><span class="name">Fe</span><span class="operator">-</span><span class="literal number integer">2</span><span class="name">S</span> <span class="name">iron</span><span class="operator">-</span><span class="name">sulfur</span> <span class="name">cluster</span><span class="punctuation">:</span> <span class="literal number integer">2</span> <span class="name">iron</span> <span class="name">atoms</span>
<span class="name">complexed</span> <span class="name">to</span> <span class="literal number integer">2</span> <span class="name">inorganic</span> <span class="name">sulfides</span> <span class="operator word">and</span> <span class="literal number integer">4</span> <span class="name">sulfur</span> <span class="name">atoms</span> <span class="name">of</span> <span class="name">cysteines</span> <span class="keyword namespace">from</span> <span class="name namespace">the</span>
<span class="name">protein</span><span class="operator">.</span>
<span class="operator">...</span>
</pre>
</div>
</div>
<div class="section" id="prosite">
<h2>10.2  解析Prosite记录</h2>
<p>Prosite是一个包含了蛋白质结构域、蛋白家族、功能位点以及识别它们的模式和图
谱，而且它是和Swiss-Prot同时开发出来的。
在Biopython中，Prosite记录是由 <tt class="docutils literal">Bio.ExPASy.Prosite.Record</tt> 类来表示的，
其中的成员与该Prosite记录中的不同区域相对应。</p>
<p>一般来说，一个Prosite文件可以包含多个Prosite记录。比如，从 <a class="reference external" href="ftp://ftp.expasy.org/databases/prosite/prosite.dat">ExPASy FTP
site</a> 网站下载
下来的、容纳了整个Prosite记录的 <tt class="docutils literal">prosite.dat</tt> 文件，含有2073条记录（2007年12月发布的第20.24版本）。
为了解析这样一个文件，我们再次使用一个迭代器：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Prosite</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;myprositefile.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Prosite</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>现在我们可以逐个提取这些记录并输出其中一些信息。比如，使用包含整个Prosite数据库的
文件将会使我们找到如下等信息：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Prosite</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;prosite.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Prosite</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">accession</span>
<span class="literal string">'PS00001'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'ASN_GLYCOSYLATION'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">pdoc</span>
<span class="literal string">'PDOC00001'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">accession</span>
<span class="literal string">'PS00004'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'CAMP_PHOSPHO_SITE'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">pdoc</span>
<span class="literal string">'PDOC00004'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">records</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">accession</span>
<span class="literal string">'PS00005'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'PKC_PHOSPHO_SITE'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">pdoc</span>
<span class="literal string">'PDOC00005'</span>
</pre>
<p>如果你想知道有多少条Prosite记录，你可以输入：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Prosite</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;prosite.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Prosite</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">n</span> <span class="operator">=</span> <span class="literal number integer">0</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span> <span class="name">n</span><span class="operator">+=</span><span class="literal number integer">1</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">n</span>
<span class="literal number integer">2073</span>
</pre>
<p>为了从这些数据中读取某一条特定的记录，可以使用 <tt class="docutils literal">read</tt> 函数：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Prosite</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;mysingleprositerecord.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Prosite</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>如果并不存在或存在多个你想要找的Prosite记录时，这个函数将会输出一个“ValueError”提示。</p>
</div>
<div class="section" id="id88">
<h2>10.3  解析Prosite文件记录</h2>
<p>在上述的Prosite示例中，像 <tt class="docutils literal">'PDOC00001'</tt> 、 <tt class="docutils literal">'PDOC00004'</tt> 、 <tt class="docutils literal">'PDOC00005'</tt> 等这样的编号指的就
是Prosite文件。Prosite文件记录可以以单个文件（ <tt class="docutils literal">prosite.doc</tt> ）的形式从ExPASy获取，并
且该文件包含了所有Prosite文档记录。</p>
<p>我们使用 <tt class="docutils literal">Bio.ExPASy.Prodoc</tt> 中的解析器来解析这些Prosite文档记录。比如，为了生成一个包含所有
Prosite文档记录的编号列表，你可以使用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Prodoc</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;prosite.doc&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Prodoc</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">accessions</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">record</span><span class="operator">.</span><span class="name">accession</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">]</span>
</pre>
<p>进一步可以使用 <tt class="docutils literal">read()</tt> 函数来对这些数据中具体某一条文档记录来进行查询。</p>
</div>
<div class="section" id="id89">
<h2>10.4  解析酶记录</h2>
<p>ExPASy的酶数据库是一个关于酶的系统命名信息的数据库。如下所示是一个比较典型的酶的记录</p>
<pre class="code python literal-block">
<span class="name">ID</span>   <span class="literal number float">3.1</span><span class="operator">.</span><span class="literal number float">1.34</span>
<span class="name">DE</span>   <span class="name">Lipoprotein</span> <span class="name">lipase</span><span class="operator">.</span>
<span class="name">AN</span>   <span class="name">Clearing</span> <span class="name">factor</span> <span class="name">lipase</span><span class="operator">.</span>
<span class="name">AN</span>   <span class="name">Diacylglycerol</span> <span class="name">lipase</span><span class="operator">.</span>
<span class="name">AN</span>   <span class="name">Diglyceride</span> <span class="name">lipase</span><span class="operator">.</span>
<span class="name">CA</span>   <span class="name">Triacylglycerol</span> <span class="operator">+</span> <span class="name">H</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">)</span><span class="name">O</span> <span class="operator">=</span> <span class="name">diacylglycerol</span> <span class="operator">+</span> <span class="name">a</span> <span class="name">carboxylate</span><span class="operator">.</span>
<span class="name">CC</span>   <span class="operator">-</span><span class="error">!</span><span class="operator">-</span> <span class="name">Hydrolyzes</span> <span class="name">triacylglycerols</span> <span class="operator word">in</span> <span class="name">chylomicrons</span> <span class="operator word">and</span> <span class="name">very</span> <span class="name">low</span><span class="operator">-</span><span class="name">density</span>
<span class="name">CC</span>       <span class="name">lipoproteins</span> <span class="punctuation">(</span><span class="name">VLDL</span><span class="punctuation">)</span><span class="operator">.</span>
<span class="name">CC</span>   <span class="operator">-</span><span class="error">!</span><span class="operator">-</span> <span class="name">Also</span> <span class="name">hydrolyzes</span> <span class="name">diacylglycerol</span><span class="operator">.</span>
<span class="name">PR</span>   <span class="name">PROSITE</span><span class="punctuation">;</span> <span class="name">PDOC00110</span><span class="punctuation">;</span>
<span class="name">DR</span>   <span class="name">P11151</span><span class="punctuation">,</span> <span class="name">LIPL_BOVIN</span> <span class="punctuation">;</span>  <span class="name">P11153</span><span class="punctuation">,</span> <span class="name">LIPL_CAVPO</span> <span class="punctuation">;</span>  <span class="name">P11602</span><span class="punctuation">,</span> <span class="name">LIPL_CHICK</span> <span class="punctuation">;</span>
<span class="name">DR</span>   <span class="name">P55031</span><span class="punctuation">,</span> <span class="name">LIPL_FELCA</span> <span class="punctuation">;</span>  <span class="name">P06858</span><span class="punctuation">,</span> <span class="name">LIPL_HUMAN</span> <span class="punctuation">;</span>  <span class="name">P11152</span><span class="punctuation">,</span> <span class="name">LIPL_MOUSE</span> <span class="punctuation">;</span>
<span class="name">DR</span>   <span class="name">O46647</span><span class="punctuation">,</span> <span class="name">LIPL_MUSVI</span> <span class="punctuation">;</span>  <span class="name">P49060</span><span class="punctuation">,</span> <span class="name">LIPL_PAPAN</span> <span class="punctuation">;</span>  <span class="name">P49923</span><span class="punctuation">,</span> <span class="name">LIPL_PIG</span>   <span class="punctuation">;</span>
<span class="name">DR</span>   <span class="name">Q06000</span><span class="punctuation">,</span> <span class="name">LIPL_RAT</span>   <span class="punctuation">;</span>  <span class="name">Q29524</span><span class="punctuation">,</span> <span class="name">LIPL_SHEEP</span> <span class="punctuation">;</span>
<span class="operator">//</span>
</pre>
<p>在这个例子中，第一行显示了脂蛋白脂肪酶（第二行）的酶编号(EC, Enzyme Commission)。
脂蛋白脂肪酶其他的名称有“清除因子脂肪酶”和“甘油二脂脂肪酶”（第三行至第五行）。
开头为“CA”的那一行显示了该酶的催化活性。评论行开头为“CC”。“PR”行显示了对应Prosite
文档记录的参考，以及“DR”行显示了Swiss-Prot记录的参考。
然而并不是所有的词条都必需出现在酶记录当中。</p>
<p>在Biopython中，一个酶记录由 <tt class="docutils literal">Bio.ExPASy.Enzyme.Record</tt> 类来代表。这个记录源于对应
于酶相关文件中所用到的双字母编码的python字典和哈希键。为了阅读含有一个酶记录的酶文件，
你可以使用 <tt class="docutils literal">Bio.ExPASy.Enzyme</tt> 中的 <tt class="docutils literal">read</tt> 函数：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Enzyme</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;lipoprotein.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Enzyme</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;ID&quot;</span><span class="punctuation">]</span>
<span class="literal string">'3.1.1.34'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DE&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Lipoprotein lipase.'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;AN&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Clearing factor lipase.'</span><span class="punctuation">,</span> <span class="literal string">'Diacylglycerol lipase.'</span><span class="punctuation">,</span> <span class="literal string">'Diglyceride lipase.'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;CA&quot;</span><span class="punctuation">]</span>
<span class="literal string">'Triacylglycerol + H(2)O = diacylglycerol + a carboxylate.'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;PR&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'PDOC00110'</span><span class="punctuation">]</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;CC&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'Hydrolyzes triacylglycerols in chylomicrons and very low-density lipoproteins</span>
<span class="punctuation">(</span><span class="name">VLDL</span><span class="punctuation">)</span><span class="operator">.</span><span class="literal string">', '</span><span class="name">Also</span> <span class="name">hydrolyzes</span> <span class="name">diacylglycerol</span><span class="operator">.</span><span class="literal string">']</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;DR&quot;</span><span class="punctuation">]</span>
<span class="punctuation">[[</span><span class="literal string">'P11151'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_BOVIN'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'P11153'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_CAVPO'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'P11602'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_CHICK'</span><span class="punctuation">],</span>
<span class="punctuation">[</span><span class="literal string">'P55031'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_FELCA'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'P06858'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_HUMAN'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'P11152'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_MOUSE'</span><span class="punctuation">],</span>
<span class="punctuation">[</span><span class="literal string">'O46647'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_MUSVI'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'P49060'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_PAPAN'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'P49923'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_PIG'</span><span class="punctuation">],</span>
<span class="punctuation">[</span><span class="literal string">'Q06000'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_RAT'</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="literal string">'Q29524'</span><span class="punctuation">,</span> <span class="literal string">'LIPL_SHEEP'</span><span class="punctuation">]]</span>
</pre>
<p>如果没有找到或者找到多个酶记录时， <tt class="docutils literal">read</tt> 函数会反馈一个ValueError提示。</p>
<p>所有酶记录都可以从 <a class="reference external" href="ftp://ftp.expasy.org/databases/enzyme/enzyme.dat">ExPASy FTP site</a> 网站下载
为单个文件（ <tt class="docutils literal">enzyme.dat</tt> ），该文件包含了4877个记录（2009年3月发布的第三版）。为了打开含有多个
酶记录的文件，你可以使用 <tt class="docutils literal">Bio.ExPASy.Enzyme</tt> 中的 <tt class="docutils literal">parse</tt> 函数来获得一个迭代器：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Enzyme</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;enzyme.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="name">Enzyme</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>我们现在每次都可以对这些记录进行迭代。比如我们可以对那些已有的酶记录做一个EC编号列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ecnumbers</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">record</span><span class="punctuation">[</span><span class="literal string">&quot;ID&quot;</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="accessing-the-expasy-server">
<h2>10.5  Accessing the ExPASy server</h2>
<p>Swiss-Prot、Prosite和Prosite文档记录可以从
<a class="reference external" href="http://www.expasy.org">http://www.expasy.org</a> 的ExPASy网络服务器下载到。在ExPASy服
务器上可以进行六种查询：</p>
<dl class="docutils">
<dt><strong>get_prodoc_entry</strong></dt>
<dd>下载一个HTML格式的Prosite文档记录</dd>
<dt><strong>get_prosite_entry</strong></dt>
<dd>下载一个HTML格式的Prosite记录</dd>
<dt><strong>get_prosite_raw</strong></dt>
<dd>下载一个原始格式的Prosite或Prosite文档记录</dd>
<dt><strong>get_sprot_raw</strong></dt>
<dd>下载一个原始格式的Swiss-Prot记录</dd>
<dt><strong>sprot_search_ful</strong></dt>
<dd>搜索一个Swiss-Prot记录</dd>
<dt><strong>sprot_search_de</strong></dt>
<dd>搜索一个Swiss-Prot记录</dd>
</dl>
<p>为了从python脚本来访问该网络服务器，我们可以使用 <tt class="docutils literal">Bio.ExPASy</tt> 模块。</p>
<div class="section" id="id90">
<h3>10.5.1  获取一个Swiss-Prot记录</h3>
<p>现在让我们来寻找一个关于兰花的查儿酮合成酶（对于寻找和兰花相关的有趣东西的理由
请看 <a class="reference external" href="#sec:orchids">2.3</a> 章节）。查儿酮合成酶参与了植物中类黄酮的生物合成，
类黄酮能够合成包含色素和UV保护分子等物质。</p>
<p>如果你要对Swiss-Prot进行搜索，你可以找到三个关于查儿酮合成酶的兰花蛋白，id编号
为O23729, O23730, O23731。现在我们要写一个能够获取这些蛋白并能够找到一些有趣
的信息的脚本。</p>
<p>首先，我们使用 <tt class="docutils literal">Bio.ExPASy</tt> 中的 <tt class="docutils literal">get_sprot_raw()</tt> 函数来获取这些记录。这个函
数非常棒，因为你可以给它提供一个id然后得到一个原始文本记录（不会受到HTML的干扰）。
然后我们可以使用 <tt class="docutils literal">Bio.SwissProt.read</tt> 来提取对应的Swiss-Prot记录，也可以使用 <tt class="docutils literal">Bio.SeqIO.read</tt> 来
得到一个序列记录SeqRecord。下列代码能够实现我刚刚提到的任务：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SwissProt</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="name">accessions</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">&quot;O23729&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;O23730&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;O23731&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">records</span> <span class="operator">=</span> <span class="punctuation">[]</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">accession</span> <span class="operator word">in</span> <span class="name">accessions</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_sprot_raw</span><span class="punctuation">(</span><span class="name">accession</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="name">record</span> <span class="operator">=</span> <span class="name">SwissProt</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="name">records</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
</pre>
<p>如果你提供给 <tt class="docutils literal">ExPASy.get_sprot_raw</tt> 的编号并不存在，那么 <tt class="docutils literal">SwissProt.read(handle)</tt> 会反
馈一个 <tt class="docutils literal">ValueError</tt> 提示。你可以根据 <tt class="docutils literal">ValueException</tt> 异常来找到无效的编号：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">accession</span> <span class="operator word">in</span> <span class="name">accessions</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_sprot_raw</span><span class="punctuation">(</span><span class="name">accession</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="keyword">try</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="name">record</span> <span class="operator">=</span> <span class="name">SwissProt</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">...</span>     <span class="keyword">except</span> <span class="name">ValueException</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">print</span> <span class="literal string">&quot;WARNING: Accession </span><span class="literal string interpol">%s</span><span class="literal string"> not found&quot;</span> <span class="operator">%</span> <span class="name">accession</span>
<span class="operator">...</span>     <span class="name">records</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="id91">
<h3>10.5.2  搜索Swiss-Prot</h3>
<p>现在，你可以察觉到我已经提前知道了这个记录的编号。的确， <tt class="docutils literal">get_sprot_raw()</tt> 需要一个词条或者编号。
当你并没有编号或者词条的时候，你可使用 <tt class="docutils literal">sprot_search_de()</tt> 或者 <tt class="docutils literal">sprot_search_ful()</tt> 函数来解决问题。</p>
<p><tt class="docutils literal">sprot_search_de()</tt> 在ID, DE, GN, OS和OG行进行搜索；
<tt class="docutils literal">sprot_search_ful()</tt> 则在所有行进行搜索。具体相关细节分别在
<a class="reference external" href="http://www.expasy.org/cgi-bin/sprot-search-de">http://www.expasy.org/cgi-bin/sprot-search-de</a>
和
<a class="reference external" href="http://www.expasy.org/cgi-bin/sprot-search-ful">http://www.expasy.org/cgi-bin/sprot-search-ful</a> 上有说明。
注意它们的默认情况下并不搜索TrEMBL（参数为 <tt class="docutils literal">trembl</tt> ）。还要注意它们返回的是html网页，然而编号却可以很容易从中得到：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">re</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">sprot_search_de</span><span class="punctuation">(</span><span class="literal string">&quot;Orchid Chalcone Synthase&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="comment"># or:</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="comment"># handle = ExPASy.sprot_search_ful(&quot;Orchid and {Chalcone Synthase}&quot;)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">html_results</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">if</span> <span class="literal string">&quot;Number of sequences found&quot;</span> <span class="operator word">in</span> <span class="name">html_results</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">ids</span> <span class="operator">=</span> <span class="name">re</span><span class="operator">.</span><span class="name">findall</span><span class="punctuation">(</span><span class="literal string">r'HREF=&quot;/uniprot/(\w+)&quot;'</span><span class="punctuation">,</span> <span class="name">html_results</span><span class="punctuation">)</span>
<span class="operator">...</span> <span class="keyword">else</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="name">ids</span> <span class="operator">=</span> <span class="name">re</span><span class="operator">.</span><span class="name">findall</span><span class="punctuation">(</span><span class="literal string">r'href=&quot;/cgi-bin/niceprot\.pl\?(\w+)&quot;'</span><span class="punctuation">,</span> <span class="name">html_results</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="prositeprosite">
<h3>10.5.3  获取Prosite和Prosite文档记录</h3>
<p>我们可以得到HTML格式和原始格式的Prosite和Prosite文档记录。为了用biopython解析Prosite和Prosite文档记录，
你应该使用原始格式的记录。而对于其他的目的，你或许会对HTML格式感兴趣。</p>
<p>为了获取一个原始格式的Prosite或者Prosite文档的记录，请使用 <tt class="docutils literal">get_prosite_raw()</tt> 。
例如，为了下载一个prosite记录并以原始格式输出，你可以使用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_prosite_raw</span><span class="punctuation">(</span><span class="literal string">'PS00001'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">text</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">text</span>
</pre>
<p>为了获取一个Prosite记录并将其解析成一个 <tt class="docutils literal">Bio.Prosite.Record</tt> 对象，请使用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Prosite</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_prosite_raw</span><span class="punctuation">(</span><span class="literal string">'PS00001'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Prosite</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>该函数也可以用于获取Prosite文档记录并解析到一个 <tt class="docutils literal">Bio.ExPASy.Prodoc.Record</tt> 对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">Prodoc</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_prosite_raw</span><span class="punctuation">(</span><span class="literal string">'PDOC00001'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Prodoc</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
</pre>
<p>对于不存在的编号， <tt class="docutils literal">ExPASy.get_prosite_raw</tt> 返回一个空字符串。当遇到空字符
串， <tt class="docutils literal">Prosite.read</tt> 和 <tt class="docutils literal">Prodoc.read</tt> 会反馈一个ValueError错误。你可以
根据这些错误异常提示来找到无效的编号。</p>
<p><tt class="docutils literal">get_prosite_entry()</tt> 和 <tt class="docutils literal">get_prodoc_entry()</tt> 函数可用于下载HTML格式的Prosite和Prosite文档记录。
为了生成展示单个Prosite记录的网页，你可以使用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_prosite_entry</span><span class="punctuation">(</span><span class="literal string">'PS00001'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">html</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">output</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;myprositerecord.html&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">output</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">html</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">output</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>类似地，Prosite文档文本的网页展示如下：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">ExPASy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ExPASy</span><span class="operator">.</span><span class="name">get_prodoc_entry</span><span class="punctuation">(</span><span class="literal string">'PDOC00001'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">html</span> <span class="operator">=</span> <span class="name">handle</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">output</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;myprodocrecord.html&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">output</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">html</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">output</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>对于这些函数，无效的编号会返回一个HTML格式的错误信息。</p>
</div>
</div>
<div class="section" id="id92">
<h2>10.6  浏览Prosite数据库</h2>
<p><a class="reference external" href="http://www.expasy.org/tools/scanprosite/">ScanProsite</a>  允许你通过向Prosite数据库提供一个
Uniprot或者PDB序列编号或序列来在线浏览蛋白序列。关于ScanProsite更多的信息，请阅
读 <a class="reference external" href="http://www.expasy.org/tools/scanprosite/scanprosite-doc.html">ScanProsite文档</a> 以及
<a class="reference external" href="http://www.expasy.org/tools/scanprosite/ScanPrositeREST.html">程序性访问ScanProsite说明文档</a> 。</p>
<p>你也可以使用Biopython的 <tt class="docutils literal">Bio.ExPASy.ScanProsite</tt> 模块来从python浏览Prosite数据库，这个模块既
能够帮你安全访问ScanProsite，也可以对ScanProsite返回的结果进行解析。为了查看下边序列中
的Prosite模式（pattern）：</p>
<pre class="code python literal-block">
<span class="name">MEHKEVVLLLLLFLKSGQGEPLDDYVNTQGASLFSVTKKQLGAGSIEECAAKCEEDEEFT</span>
<span class="name">CRAFQYHSKEQQCVIMAENRKSSIIIRMRDVVLFEKKVYLSECKTGNGKNYRGTMSKTKN</span>
</pre>
<p>你可以使用下边的代码：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sequence</span> <span class="operator">=</span> <span class="literal string">&quot;MEHKEVVLLLLLFLKSGQGEPLDDYVNTQGASLFSVTKKQLGAGSIEECAAKCEEDEEFT</span>
<span class="name">CRAFQYHSKEQQCVIMAENRKSSIIIRMRDVVLFEKKVYLSECKTGNGKNYRGTMSKTKN</span><span class="literal string">&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.ExPASy</span> <span class="keyword namespace">import</span> <span class="name">ScanProsite</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ScanProsite</span><span class="operator">.</span><span class="name">scan</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">sequence</span><span class="punctuation">)</span>
</pre>
<p>你可以通过执行 <tt class="docutils literal">handle.read()</tt> 获取原始XML格式的搜索结果。此外，我们可以使用 <tt class="docutils literal">Bio.ExPASy.ScanProsite.read</tt>
来将原始的XML数据解析到一个python对象：</p>
<pre class="code python literal-block">
   <span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span> <span class="operator">=</span> <span class="name">ScanProsite</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
   <span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">type</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">)</span>
   <span class="operator">&lt;</span><span class="keyword">class</span> <span class="error">'</span><span class="name class">Bio</span><span class="operator">.</span><span class="name">ExPASy</span><span class="operator">.</span><span class="name">ScanProsite</span><span class="operator">.</span><span class="name">Record</span><span class="literal string">'&gt;</span>

<span class="literal string backtick">``</span><span class="name">Bio</span><span class="operator">.</span><span class="name">ExPASy</span><span class="operator">.</span><span class="name">ScanProsite</span><span class="operator">.</span><span class="name">Record</span><span class="literal string backtick">``</span> <span class="error">对象源自一个由</span><span class="name">ScanProsite</span><span class="error">返回的包含了</span><span class="name">ScanProsite</span> <span class="name">hits</span><span class="error">的列表，这个对象也能够存储</span><span class="name">hits</span><span class="error">的数量以及所找到序列的数量。本次</span><span class="name">ScanProsite</span><span class="error">搜索找到了</span><span class="literal number integer">6</span><span class="error">个</span><span class="name">hits</span><span class="error">：</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="operator">.</span><span class="name">n_seq</span>
<span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="operator">.</span><span class="name">n_match</span>
<span class="literal number integer">6</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">)</span>
<span class="literal number integer">6</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'signature_ac'</span><span class="punctuation">:</span> <span class="literal string">u'PS50948'</span><span class="punctuation">,</span> <span class="literal string">'level'</span><span class="punctuation">:</span> <span class="literal string">u'0'</span><span class="punctuation">,</span> <span class="literal string">'stop'</span><span class="punctuation">:</span> <span class="literal number integer">98</span><span class="punctuation">,</span> <span class="literal string">'sequence_ac'</span><span class="punctuation">:</span> <span class="literal string">u'USERSEQ1'</span><span class="punctuation">,</span> <span class="literal string">'start'</span><span class="punctuation">:</span> <span class="literal number integer">16</span><span class="punctuation">,</span> <span class="literal string">'score'</span><span class="punctuation">:</span> <span class="literal string">u'8.873'</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'start'</span><span class="punctuation">:</span> <span class="literal number integer">37</span><span class="punctuation">,</span> <span class="literal string">'stop'</span><span class="punctuation">:</span> <span class="literal number integer">39</span><span class="punctuation">,</span> <span class="literal string">'sequence_ac'</span><span class="punctuation">:</span> <span class="literal string">u'USERSEQ1'</span><span class="punctuation">,</span> <span class="literal string">'signature_ac'</span><span class="punctuation">:</span> <span class="literal string">u'PS00005'</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'start'</span><span class="punctuation">:</span> <span class="literal number integer">45</span><span class="punctuation">,</span> <span class="literal string">'stop'</span><span class="punctuation">:</span> <span class="literal number integer">48</span><span class="punctuation">,</span> <span class="literal string">'sequence_ac'</span><span class="punctuation">:</span> <span class="literal string">u'USERSEQ1'</span><span class="punctuation">,</span> <span class="literal string">'signature_ac'</span><span class="punctuation">:</span> <span class="literal string">u'PS00006'</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'start'</span><span class="punctuation">:</span> <span class="literal number integer">60</span><span class="punctuation">,</span> <span class="literal string">'stop'</span><span class="punctuation">:</span> <span class="literal number integer">62</span><span class="punctuation">,</span> <span class="literal string">'sequence_ac'</span><span class="punctuation">:</span> <span class="literal string">u'USERSEQ1'</span><span class="punctuation">,</span> <span class="literal string">'signature_ac'</span><span class="punctuation">:</span> <span class="literal string">u'PS00005'</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="punctuation">[</span><span class="literal number integer">4</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'start'</span><span class="punctuation">:</span> <span class="literal number integer">80</span><span class="punctuation">,</span> <span class="literal string">'stop'</span><span class="punctuation">:</span> <span class="literal number integer">83</span><span class="punctuation">,</span> <span class="literal string">'sequence_ac'</span><span class="punctuation">:</span> <span class="literal string">u'USERSEQ1'</span><span class="punctuation">,</span> <span class="literal string">'signature_ac'</span><span class="punctuation">:</span> <span class="literal string">u'PS00004'</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="punctuation">[</span><span class="literal number integer">5</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'start'</span><span class="punctuation">:</span> <span class="literal number integer">106</span><span class="punctuation">,</span> <span class="literal string">'stop'</span><span class="punctuation">:</span> <span class="literal number integer">111</span><span class="punctuation">,</span> <span class="literal string">'sequence_ac'</span><span class="punctuation">:</span> <span class="literal string">u'USERSEQ1'</span><span class="punctuation">,</span> <span class="literal string">'signature_ac'</span><span class="punctuation">:</span> <span class="literal string">u'PS00008'</span><span class="punctuation">}</span>
</pre>
<p>其他的ScanProsite参数可以以关键词参数的形式被传递，更多的信息详见 <a class="reference external" href="http://www.expasy.org/tools/scanprosite/ScanPrositeREST.html">程序性访问
ScanProsite说明文档</a> 。
比如，传递 <tt class="docutils literal">lowscore=1</tt> 可以帮我们找到一个新的低分值hit：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">ScanProsite</span><span class="operator">.</span><span class="name">scan</span><span class="punctuation">(</span><span class="name">seq</span><span class="operator">=</span><span class="name">sequence</span><span class="punctuation">,</span> <span class="name">lowscore</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span> <span class="operator">=</span> <span class="name">ScanProsite</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">result</span><span class="operator">.</span><span class="name">n_match</span>
<span class="literal number integer">7</span>
</pre>
</div>
</div>
<div class="section" id="d-pdb">
<h1>第11章  走向3D：PDB模块</h1>
<p>Bio.PDB是Biopython中处理生物大分子晶体结构的模块。除了别的类之外，Bio.PDB包含PDBParser类，此类能够产生一个Structure对象，以一种较方便的方式获取文件中的原子数据。只是在处理PDB文件头所包含的信息时，该类有一定的局限性。</p>
<div class="section" id="id93">
<h2>11.1  晶体结构文件的读与写</h2>
<div class="section" id="pdb">
<h3>11.1.1  读取PDB文件</h3>
<p>首先，我们创建一个 <tt class="docutils literal">PDBParser</tt> 对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.PDB.PDBParser</span> <span class="keyword namespace">import</span> <span class="name">PDBParser</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">p</span> <span class="operator">=</span> <span class="name">PDBParser</span><span class="punctuation">(</span><span class="name">PERMISSIVE</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p><tt class="docutils literal">PERMISSIV</tt> 标签表示一些与PDB文件相关的问题（见 <a class="reference external" href="#problem%20structures">11.7.1</a> ）会被忽略（注意某些原子和/或残基会丢失）。如果没有这个标签，则会在解析器运行期间有问题被检测到的时候生成一个 <tt class="docutils literal">PDBConstructionException</tt> 标签。</p>
<p>接着通过 <tt class="docutils literal">PDBParser</tt> 解析PDB文件，就产生了Structure对象（在此例子中，PDB文件为'pdb1fat.ent'，'1fat'是用户定义的结构名称）:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">structure_id</span> <span class="operator">=</span> <span class="literal string">&quot;1fat&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">filename</span> <span class="operator">=</span> <span class="literal string">&quot;pdb1fat.ent&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">s</span> <span class="operator">=</span> <span class="name">p</span><span class="operator">.</span><span class="name">get_structure</span><span class="punctuation">(</span><span class="name">structure_id</span><span class="punctuation">,</span> <span class="name">filename</span><span class="punctuation">)</span>
</pre>
<p>你可以从PDBParser对象中用 <tt class="docutils literal">get_header</tt> 和 <tt class="docutils literal">get_trailer</tt> 方法来提取PDB文件中的文件头和文件尾（简单的字符串列表）。然而许多PDB文件头包含不完整或错误的信息。许多错误在等价的mmCIF格式文件中得到修正。* 因此，如果你对文件头信息感兴趣，可以用下面即将讲到的 <tt class="docutils literal">MMCIF2Dict</tt> 来提取信息，而不用处理PDB文件文件头。*</p>
<p>现在澄清了，让我们回到解析PDB文件头这件事上。结构对象有个属性叫 <tt class="docutils literal">header</tt> ，这是一个将头记录映射到其相应值的Python字典。</p>
<p>例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">resolution</span> <span class="operator">=</span> <span class="name">structure</span><span class="operator">.</span><span class="name">header</span><span class="punctuation">[</span><span class="literal string">'resolution'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">keywords</span> <span class="operator">=</span> <span class="name">structure</span><span class="operator">.</span><span class="name">header</span><span class="punctuation">[</span><span class="literal string">'keywords'</span><span class="punctuation">]</span>
</pre>
<p>在这个字典中可用的关键字有 <tt class="docutils literal">name</tt> 、 <tt class="docutils literal">head</tt> 、 <tt class="docutils literal">deposition_date</tt> 、 <tt class="docutils literal">release_date</tt> 、 <tt class="docutils literal">structure_method</tt> 、 <tt class="docutils literal">resolution</tt> 、 <tt class="docutils literal">structure_reference</tt> （映射到一个参考文献列表）、 <tt class="docutils literal">journal_reference</tt> 、 <tt class="docutils literal">author</tt> 、和 <tt class="docutils literal">compound</tt> （映射到一个字典，其中包含结晶化合物的各种信息）。</p>
<p>没有创建 <tt class="docutils literal">Structure</tt> 对象的时候，也可以创建这个字典，比如直接从PDB文件创建:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">file</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span><span class="literal string">'r'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">header_dict</span> <span class="operator">=</span> <span class="name">parse_pdb_header</span><span class="punctuation">(</span><span class="name builtin">file</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">file</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
</div>
<div class="section" id="mmcif">
<h3>11.1.2  读取mmCIF文件</h3>
<p>与PDB文件的情形类似，先创建一个 <tt class="docutils literal">MMCIFParser</tt> 对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.PDB.MMCIFParser</span> <span class="keyword namespace">import</span> <span class="name">MMCIFParser</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">parser</span> <span class="operator">=</span> <span class="name">MMCIFParser</span><span class="punctuation">()</span>
</pre>
<p>然后用这个解析器从mmCIF文件创建一个结构对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">structure</span> <span class="operator">=</span> <span class="name">parser</span><span class="operator">.</span><span class="name">get_structure</span><span class="punctuation">(</span><span class="literal string">'1fat'</span><span class="punctuation">,</span> <span class="literal string">'1fat.cif'</span><span class="punctuation">)</span>
</pre>
<p>为了尽量少访问mmCIF文件，可以用 <tt class="docutils literal">MMCIF2Dict</tt> 类创建一个Python字典来将所有mmCIF文件中各种标签映射到其对应的值上。若有多个值（像 <tt class="docutils literal">_atom_site.Cartn_y</tt> 标签，储存的是所有原子的*y*坐标值），则这个标签映射到一个值列表。从mmCIF文件创建字典如下：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.PDB.MMCIF2Dict</span> <span class="keyword namespace">import</span> <span class="name">MMCIF2Dict</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mmcif_dict</span> <span class="operator">=</span> <span class="name">MMCIF2Dict</span><span class="punctuation">(</span><span class="literal string">'1FAT.cif'</span><span class="punctuation">)</span>
</pre>
<p>例：从mmCIF文件获取溶剂含量:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sc</span> <span class="operator">=</span> <span class="name">mmcif_dict</span><span class="punctuation">[</span><span class="literal string">'_exptl_crystal.density_percent_sol'</span><span class="punctuation">]</span>
</pre>
<p>例：获取包含所有原子*y*坐标的列表:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">y_list</span> <span class="operator">=</span> <span class="name">mmcif_dict</span><span class="punctuation">[</span><span class="literal string">'_atom_site.Cartn_y'</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="pdb-xml">
<h3>11.1.3  读取PDB XML格式的文件</h3>
<p>这个功能暂时还不支持，不过我们确实计划在未来支持这个功能（这项任务并不大）。如果你需要的话联系Biopython开发人员（ <a class="reference external" href="mailto:biopython-dev&#64;biopython.org">biopython-dev&#64;biopython.org</a> ）。</p>
</div>
<div class="section" id="id94">
<h3>11.1.4  写PDB文件</h3>
<p>可以用PDBIO类实现。当然也可很方便地输出一个结构的特定部分。</p>
<p>例子：保存一个结构</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">io</span> <span class="operator">=</span> <span class="name">PDBIO</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">io</span><span class="operator">.</span><span class="name">set_structure</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">io</span><span class="operator">.</span><span class="name">save</span><span class="punctuation">(</span><span class="literal string">'out.pdb'</span><span class="punctuation">)</span>
</pre>
<p>如果你想写出结构的一部分，可以用 <cite>Select</cite> 类（也在 <tt class="docutils literal">PDBIO</tt> 中）来实现。 <cite>Select</cite> 有如下四种方法：</p>
<ul class="simple">
<li><tt class="docutils literal">accept_model(model)</tt></li>
<li><tt class="docutils literal">accept_chain(chain)</tt></li>
<li><tt class="docutils literal">accept_residue(residue)</tt></li>
<li><tt class="docutils literal">accept_atom(atom)</tt></li>
</ul>
<p>在默认情况下，每种方法的返回值都为1（表示model/chain/residue/atom被包含在输出结果中）。通过子类化 <tt class="docutils literal">Select</tt> 和返回值0，你可以从输出中排除model、chain等。也许麻烦，但很强大。接下来的代码将只输出甘氨酸残基：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">class</span> <span class="name class">GlySelect</span><span class="punctuation">(</span><span class="name">Select</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">def</span> <span class="name function">accept_residue</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">,</span> <span class="name">residue</span><span class="punctuation">):</span>
<span class="operator">...</span>         <span class="keyword">if</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_name</span><span class="punctuation">()</span><span class="operator">==</span><span class="literal string">'GLY'</span><span class="punctuation">:</span>
<span class="operator">...</span>             <span class="keyword">return</span> <span class="name builtin pseudo">True</span>
<span class="operator">...</span>         <span class="keyword">else</span><span class="punctuation">:</span>
<span class="operator">...</span>             <span class="keyword">return</span> <span class="name builtin pseudo">False</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">io</span> <span class="operator">=</span> <span class="name">PDBIO</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">io</span><span class="operator">.</span><span class="name">set_structure</span><span class="punctuation">(</span><span class="name">s</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">io</span><span class="operator">.</span><span class="name">save</span><span class="punctuation">(</span><span class="literal string">'gly_only.pdb'</span><span class="punctuation">,</span> <span class="name">GlySelect</span><span class="punctuation">())</span>
</pre>
<p>如果这部分对你来说太复杂，那么 <tt class="docutils literal">Dice</tt> 模块有一个很方便的 <tt class="docutils literal">extract</tt> 函数，它可以输出一条链中起始和终止氨基酸残基之间的所有氨基酸残基。</p>
</div>
</div>
<div class="section" id="id95">
<h2>11.2  结构的表示</h2>
<p>一个 <tt class="docutils literal">Structure</tt> 对象的整体布局遵循称为SMCRA（Structure/Model/Chain/Residue/Atom，结构/模型/链/残基/原子）的体系架构：</p>
<ul class="simple">
<li>结构由模型组成</li>
<li>模型由多条链组成</li>
<li>链由残基组成</li>
<li>多个原子构成残基</li>
</ul>
<p>这是很多结构生物学家/生物信息学家看待结构的方法，也是处理结构的一种简单而有效的方法。在需要的时候加上额外的材料。一个 <tt class="docutils literal">Structure</tt> 对象的UML图（暂时忘掉 <tt class="docutils literal">Disordered</tt> 吧）如下图所示 <a class="reference external" href="#fig:smcra">11.1</a> 。这样的数据结构不一定最适用于表示一个结构的生物大分子内容，但要很好地解释一个描述结构的文件中所呈现的数据（最典型的如PDB或MMCIF文件），这样的数据结构就是必要的了。如果这种层次结构不能表示一个结构文件的内容，那么可以相当确定是这个文件有错误或至少描述结构不够明确。一旦不能生成SMCRA数据结构，就有理由怀疑出了故障。因此，解析PDB文件可用于检测可能的故障。我们将在 <a class="reference external" href="#problem%20structures">11.7.1</a> 小节给出关于这一点的一些例子。</p>
<hr class="docutils" />
<p><img alt="image3" src="../images/smcra.png" /></p>
<p>图11.1：用来表示大分子结构的 <tt class="docutils literal">Structure</tt> 类的SMCRA体系的UML图。带方块的实线表示集合，带箭头的实线表示引用，带三角形的实线表示继承，带三角形的虚线表示接口实现。</p>
<hr class="docutils" />
<p>结构，模型，链，残基都是实体基类的子类。原子类仅仅（部分）实现了实体接口（因为原子类没有子类）。</p>
<p>对于每个实体子类，你可以用该子类的一个唯一标识符作为键来提取子类（比如，可以用原子名称作为键从残基对象中提取一个原子对象；用链的标识符作为键从域对象中提取链）。</p>
<p>紊乱原子和残基用DisorderedAtom和DisorderedResidue类来表示，二者都是DisorderedEntityWrapper基类的子类。它们隐藏了紊乱的复杂性，表现得与原子和残基对象无二。</p>
<p>一般地，一个实体子类（即原子，残基，链，模型）能通过标识符作为键来从父类（分别为残基，链，模型，结构）中提取。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">child_entity</span> <span class="operator">=</span> <span class="name">parent_entity</span><span class="punctuation">[</span><span class="name">child_id</span><span class="punctuation">]</span>
</pre>
<p>你可以从一个父实体对象获得所有子实体的列表。需要注意的是，这个列表以一种特定的方式排列（例如根据在模型对象中链对象的链标识符来排序）。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">child_list</span> <span class="operator">=</span> <span class="name">parent_entity</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">()</span>
</pre>
<p>你也可以从子类得到父类：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">parent_entity</span> <span class="operator">=</span> <span class="name">child_entity</span><span class="operator">.</span><span class="name">get_parent</span><span class="punctuation">()</span>
</pre>
<p>在SMCRA的所有层次水平，你还可以提取一个 <em>完整id</em> 。完整id是包含所有从顶层对象（结构）到当前对象的id的一个元组。一个残基对象的完整id可以这么得到：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">full_id</span> <span class="operator">=</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_full_id</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">full_id</span>
<span class="punctuation">(</span><span class="literal string">&quot;1abc&quot;</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal string">&quot;A&quot;</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">&quot;&quot;</span><span class="punctuation">,</span> <span class="literal number integer">10</span><span class="punctuation">,</span> <span class="literal string">&quot;A&quot;</span><span class="punctuation">))</span>
</pre>
<p>这对应于：</p>
<ul class="simple">
<li>id为&quot;1abc&quot;的结构</li>
<li>id为0的模型</li>
<li>id为&quot;A&quot;的链</li>
<li>id为(&quot; &quot;, 10, &quot;A&quot;)的残基</li>
</ul>
<p>这个残基id表示该残基不是异质残基（也不是水分子），因为其异质值为空；而序列标识符为10，插入码为&quot;A&quot;。</p>
<p>要得到实体的id，用 <tt class="docutils literal">get_id</tt> 方法即可：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">entity</span><span class="operator">.</span><span class="name">get_id</span><span class="punctuation">()</span>
</pre>
<p>可以用 <tt class="docutils literal">has_id</tt> 方法来检查这个实体是否有子类具有给定id：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">entity</span><span class="operator">.</span><span class="name">has_id</span><span class="punctuation">(</span><span class="name">entity_id</span><span class="punctuation">)</span>
</pre>
<p>实体的长度等于其子类的个数：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nr_children</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">entity</span><span class="punctuation">)</span>
</pre>
<p>对于从父实体得到的子实体，可以删除，重命名，添加等等，但这并不包含任何完整性检查（比如，有可能添加两个相同id的残基到同一条链上）。这就真的需要包含完整性检查的装饰类（Decorator）来完成了，但是如果你想使用原始接口的话可以查看源代码（Entity.py）。</p>
<div class="section" id="id96">
<h3>11.2.1  结构</h3>
<p>结构对象是层次中的最高层。其id是用户指定的一个字符串。结构包含一系列子模型。大部分晶体结构（但不是全部）含有一个单一模型，但是NMR结构通常由若干模型构成。晶体结构中大部分子的乱序也能导致多个模型。</p>
</div>
<div class="section" id="id97">
<h3>11.2.2  模型</h3>
<p>结构域对象的id是一个整数，源自该模型在所解析文件中的位置（自动从0开始）。晶体结构通常只有一个模型（id为0），而NMR文件通常含有多个模型。然而许多PDB解析器都假定只有一个结构域， <tt class="docutils literal">Bio.PDB</tt> 中的 <tt class="docutils literal">Structure</tt> 类就设计成能轻松处理含有不止一个模型的PDB文件。</p>
<p>举个例子，从一个结构对象中获取其第一个模型：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">first_model</span> <span class="operator">=</span> <span class="name">structure</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
</pre>
<p>模型对象存储着子链的列表。</p>
</div>
<div class="section" id="id98">
<h3>11.2.3  链</h3>
<p>链对象的id来自PDB/mmCIF文件中的链标识符，是个单字符（通常是一个字母）。模型中的每个链都具有唯一的id。例如，从一个模型对象中取出标识符为“A”的链对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">chain_A</span> <span class="operator">=</span> <span class="name">model</span><span class="punctuation">[</span><span class="literal string">&quot;A&quot;</span><span class="punctuation">]</span>
</pre>
<p>链对象储存着残基对象的列表。</p>
</div>
<div class="section" id="id99">
<h3>11.2.4  残基</h3>
<p>一个残基id是一个三元组：</p>
<ul>
<li><p class="first"><strong>异质域</strong> (hetfield)，即：</p>
<ul class="simple">
<li><tt class="docutils literal">'W'</tt> 代表水分子</li>
<li><tt class="docutils literal">'H_'</tt> 后面紧跟残基名称，代表其它异质残基（例如 <tt class="docutils literal">'H_GLC'</tt> 表示一个葡萄糖分子）</li>
<li>空值表示标准的氨基酸和核酸</li>
</ul>
<p>采用这种体制的理由在 <a class="reference external" href="#异质%20问题">11.4.1</a> 部分有叙述。</p>
</li>
<li><p class="first"><strong>序列标识符</strong> （resseq），一个描述该残基在链上的位置的整数（如100）；</p>
</li>
<li><p class="first"><strong>插入码</strong> （icode），一个字符串，如“A”。插入码有时用来保存某种特定的、想要的残基编号体制。一个Ser 80的插入突变（比如在Thr 80和Asn 81残基间插入）可能具有如下序列标识符和插入码：Thr 80 A, Ser 80 B, Asn 81。这样一来，残基编号体制保持与野生型结构一致。</p>
</li>
</ul>
<p>因此，上述的葡萄酸残基id就是 <tt class="docutils literal">(’H_GLC’, 100, ’A’)</tt> 。如果异质标签和插入码为空，那么可以只使用序列标识符：</p>
<pre class="code python literal-block">
<span class="comment"># Full id</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">=</span><span class="name">chain</span><span class="punctuation">[(</span><span class="literal string">' '</span><span class="punctuation">,</span> <span class="literal number integer">100</span><span class="punctuation">,</span> <span class="literal string">' '</span><span class="punctuation">)]</span>
<span class="comment"># Shortcut id</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">=</span><span class="name">chain</span><span class="punctuation">[</span><span class="literal number integer">100</span><span class="punctuation">]</span>
</pre>
<p>异质标签的起因是许许多多的PDB文件使用相同的序列标识符表示一个氨基酸和一个异质残基或一个水分子，这会产生一个很明显的问题，如果不使用异质标签的话。</p>
<p>毫不奇怪，一个残基对象存储着一个子原子集，它还包含一个表示残基名称的字符串（如 “ASN”）和残基的片段标识符（这对X-PLOR的用户来说很熟悉，但是在SMCRA数据结构的构建中没用到）。</p>
<p>让我们来看一些例子。插入码为空的Asn 10具有残基id <tt class="docutils literal">(’ ’, 10, ’ ’)</tt> ；Water 10，残基id <tt class="docutils literal">(’W’, 10, ’ ’)</tt>；一个序列标识符为10的葡萄糖分子（名称为GLC的异质残基），残基id为 <tt class="docutils literal">(’H_GLC’, 10, ’ ’)</tt> 。在这种情况下，三个残基（具有相同插入码和序列标识符）可以位于同一条链上，因为它们的残基id是不同的。</p>
<p>大多数情况下，hetflag和插入码均为空，如 <tt class="docutils literal">(’ ’, 10, ’ ’)</tt> 。在这些情况下，序列标识符可以用作完整id的快捷方式：</p>
<pre class="code python literal-block">
<span class="comment"># use full id</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">res10</span> <span class="operator">=</span> <span class="name">chain</span><span class="punctuation">[(</span><span class="literal string">' '</span><span class="punctuation">,</span> <span class="literal number integer">10</span><span class="punctuation">,</span> <span class="literal string">' '</span><span class="punctuation">)]</span>
<span class="comment"># use shortcut</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">res10</span> <span class="operator">=</span> <span class="name">chain</span><span class="punctuation">[</span><span class="literal number integer">10</span><span class="punctuation">]</span>
</pre>
<p>一个链对象中每个残基对象都应该具有唯一的id。但是对含紊乱原子的残基，要以一种特殊的方式来处理，详见 <a class="reference external" href="#point%20mutations">11.3.3</a> 。</p>
<p>一个残基对象还有大量其它方法：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_resname</span><span class="punctuation">()</span>       <span class="comment"># returns the residue name, e.g. &quot;ASN&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">.</span><span class="name">is_disordered</span><span class="punctuation">()</span>     <span class="comment"># returns 1 if the residue has disordered atoms</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_segid</span><span class="punctuation">()</span>         <span class="comment"># returns the SEGID, e.g. &quot;CHN1&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">.</span><span class="name">has_id</span><span class="punctuation">(</span><span class="name">name</span><span class="punctuation">)</span>        <span class="comment"># test if a residue has a certain atom</span>
</pre>
<p>你可以用 <tt class="docutils literal">is_aa(residue)</tt> 来检验一个残基对象是否为氨基酸。</p>
</div>
<div class="section" id="id100">
<h3>11.2.5  原子</h3>
<p>原子对象储存着所有与原子有关的数据，它没有子类。原子的id就是它的名称（如，“OG”代表Ser残基的侧链氧原子）。在残基中原子id必需是唯一的。此外，对于紊乱原子会产生异常，见 <a class="reference external" href="#disordered%20atoms">11.3.2</a> 小节的描述。</p>
<p>原子id就是原子名称（如 <tt class="docutils literal">’CA’</tt> ）。在实践中，原子名称是从PDB文件中原子名称去除所有空格而创建的。</p>
<p>但是在PDB文件中，空格可以是原子名称的一部分。通常，钙原子称为 <tt class="docutils literal"><span class="pre">’CA..’</span></tt> 是为了和Cα原子（叫做 <tt class="docutils literal">’.CA.’</tt> ）区分开。在这种情况下，如果去掉空格就会产生问题（如统一个残基中的两个原子都叫做 <tt class="docutils literal">’CA’</tt> ），所以保留空格。</p>
<p>在PDB文件中，一个原子名字由4个字符组成，通常头尾皆为空格。为了方便使用，空格通常可以去掉（在PDB文件中氨基酸的Cα原子标记为“.CA.”，点表示空格）。为了生成原子名称（然后是原子id），空格删掉了，除非会在一个残基中造成名字冲突（如两个原子对象有相同的名称和id）。对于后面这种情况，会尝试让原子名称包含空格。这种情况可能会发生在，比如残基包含名称为“.CA.”和“CA..”的原子，尽管这不怎么可能。</p>
<p>所存储的原子数据包括原子名称，原子坐标（如果有的话还包括标准差），B因子（包括各向异性B因子和可能存在的标准差），altloc标识符和完整的、包括空格的原子名称。较少用到的项如原子序号和原子电荷（有时在PDB文件中规定）也就没有存储。</p>
<p>为了处理原子坐标，可以用 <tt class="docutils literal">’Atom’</tt> 对象的 <tt class="docutils literal">transform</tt> 方法。用 <tt class="docutils literal">set_coord</tt> 方法可以直接设定原子坐标。</p>
<p>一个Atom对象还有如下其它方法：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_name</span><span class="punctuation">()</span>       <span class="comment"># atom name (spaces stripped, e.g. &quot;CA&quot;)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_id</span><span class="punctuation">()</span>         <span class="comment"># id (equals atom name)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_coord</span><span class="punctuation">()</span>      <span class="comment"># atomic coordinates</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>     <span class="comment"># atomic coordinates as Vector object</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_bfactor</span><span class="punctuation">()</span>    <span class="comment"># isotropic B factor</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_occupancy</span><span class="punctuation">()</span>  <span class="comment"># occupancy</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_altloc</span><span class="punctuation">()</span>     <span class="comment"># alternative location specifier</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_sigatm</span><span class="punctuation">()</span>     <span class="comment"># standard deviation of atomic parameters</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_siguij</span><span class="punctuation">()</span>     <span class="comment"># standard deviation of anisotropic B factor</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_anisou</span><span class="punctuation">()</span>     <span class="comment"># anisotropic B factor</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">a</span><span class="operator">.</span><span class="name">get_fullname</span><span class="punctuation">()</span>   <span class="comment"># atom name (with spaces, e.g. &quot;.CA.&quot;)</span>
</pre>
<p>siguij，各向异性B因子和sigatm Numpy阵列可以用来表示原子坐标。</p>
<p><tt class="docutils literal">get_vector</tt> 方法会返回一个代表 <tt class="docutils literal">Atom</tt>  对象坐标的 <tt class="docutils literal">Vector</tt> 对象，可以对原子坐标进行向量运算。 <tt class="docutils literal">Vector</tt> 实现了完整的三维向量运算、矩阵乘法（包括左乘和右乘）和一些高级的、与旋转相关的操作。</p>
<p>举个Bio.PDB的 <tt class="docutils literal">Vector</tt> 模块功能的例子，假设你要查找Gly残基的Cβ原子的位置，如果存在的话。将Gly残基的N原子沿Cα-C化学键旋转-120度，能大致将其放在一个真正的Cβ原子的位置上。怎么做呢？就是下面这样使用 <tt class="docutils literal">Vector</tt> 模块中的``rotaxis`` 方法（能用来构造一个绕特定坐标轴的旋转）：</p>
<pre class="code python literal-block">
<span class="comment"># get atom coordinates as vectors</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">n</span> <span class="operator">=</span> <span class="name">residue</span><span class="punctuation">[</span><span class="literal string">'N'</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">c</span> <span class="operator">=</span> <span class="name">residue</span><span class="punctuation">[</span><span class="literal string">'C'</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ca</span> <span class="operator">=</span> <span class="name">residue</span><span class="punctuation">[</span><span class="literal string">'CA'</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="comment"># center at origin</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">n</span> <span class="operator">=</span> <span class="name">n</span> <span class="operator">-</span> <span class="name">ca</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">c</span> <span class="operator">=</span> <span class="name">c</span> <span class="operator">-</span> <span class="name">ca</span>
<span class="comment"># find rotation matrix that rotates n</span>
<span class="comment"># -120 degrees along the ca-c vector</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">rot</span> <span class="operator">=</span> <span class="name">rotaxis</span><span class="punctuation">(</span><span class="operator">-</span><span class="name">pi</span> <span class="operator">*</span> <span class="literal number float">120.0</span><span class="operator">/</span><span class="literal number float">180.0</span><span class="punctuation">,</span> <span class="name">c</span><span class="punctuation">)</span>
<span class="comment"># apply rotation to ca-n vector</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cb_at_origin</span> <span class="operator">=</span> <span class="name">n</span><span class="operator">.</span><span class="name">left_multiply</span><span class="punctuation">(</span><span class="name">rot</span><span class="punctuation">)</span>
<span class="comment"># put on top of ca atom</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cb</span> <span class="operator">=</span> <span class="name">cb_at_origin</span><span class="operator">+</span><span class="name">ca</span>
</pre>
<p>这个例子展示了在原子数据上能进行一些相当不平凡的向量运算，这些运算会很有用。除了所有常用向量运算（叉积（用 <tt class="docutils literal">*</tt><tt class="docutils literal">*</tt> ），点积（用 <tt class="docutils literal">*</tt> ），角度， 取范数等）和上述提到的 <tt class="docutils literal">rotaxis</tt> 函数，<tt class="docutils literal">Vector</tt> 模块还有方法能旋转（ <tt class="docutils literal">rotmat</tt> ）或反射（ <tt class="docutils literal">refmat</tt> ）一个向量到另外一个向量上。</p>
</div>
<div class="section" id="atom-residue-chain-model">
<h3>11.2.6  从结构中提取指定的 <tt class="docutils literal">Atom/Residue/Chain/Model</tt></h3>
<p>举些例子如下：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">structure</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">chain</span> <span class="operator">=</span> <span class="name">model</span><span class="punctuation">[</span><span class="literal string">'A'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span> <span class="operator">=</span> <span class="name">chain</span><span class="punctuation">[</span><span class="literal number integer">100</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atom</span> <span class="operator">=</span> <span class="name">residue</span><span class="punctuation">[</span><span class="literal string">'CA'</span><span class="punctuation">]</span>
</pre>
<p>还可以用一个快捷方式：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atom</span> <span class="operator">=</span> <span class="name">structure</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">][</span><span class="literal string">'A'</span><span class="punctuation">][</span><span class="literal number integer">100</span><span class="punctuation">][</span><span class="literal string">'CA'</span><span class="punctuation">]</span>
</pre>
</div>
</div>
<div class="section" id="id101">
<h2>11.3  紊乱</h2>
<p>Bio.PDB能够处理紊乱原子和点突变（比如Gly和Ala残基在相同位置上）。</p>
<div class="section" id="id102">
<h3>11.3.1  一般性方法</h3>
<p>紊乱可以从两个角度来解决：原子和残基的角度。一般来说，我们尝试压缩所有由紊乱引起的复杂性。如果你仅仅想遍历所有Cα原子，那么你不必在意一些具有紊乱侧链的残基。另一方面，应该考虑在数据结构中完整地表示紊乱性。因此，紊乱原子或残基存储在特定的对象中，这些对象表现得就像毫无紊乱。这可以通过表示紊乱原子或残基的子集来完成。至于挑选哪个子集（例如使用Ser残基的哪两个紊乱OG侧链原子位置），由用户来决定。</p>
</div>
<div class="section" id="id103">
<h3>11.3.2  紊乱原子</h3>
<p>紊乱原子可以用普通的 <tt class="docutils literal">Atom</tt> 对象来表示，但是所有表示相同物理原子的 <tt class="docutils literal">Atom</tt> 对象都存储在一个 <tt class="docutils literal">DisorderedAtom</tt> 对象中（见图. <a class="reference external" href="#fig:smcra">11.1</a> ）。 <tt class="docutils literal">DisorderedAtom</tt> 对象中每个 <tt class="docutils literal">Atom</tt> 对象都能用它的altloc标识符来唯一地索引。 <tt class="docutils literal">DisorderedAtom</tt> 对象将所有未捕获方法的调用发送给选定的Atom对象，缺省对象是代表最高使用率的原子的那个。当然用户可以使用其altloc标识符来更改选定的 <tt class="docutils literal">Atom</tt> 对象。以这种方式，原子紊乱就正确地表示出来而没有很多额外的复杂性。换言之，如果你对原子紊乱不感兴趣，你也不会被它困扰。</p>
<p>每个紊乱原子都有一个特征性的altloc标识符。你可以设定：一个 <tt class="docutils literal">DisorderedAtom</tt> 对象表现得像与一个指定的altloc标识符相关的 <tt class="docutils literal">Atom</tt> 对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atom</span><span class="operator">.</span><span class="name">disordered_select</span><span class="punctuation">(</span><span class="literal string">'A'</span><span class="punctuation">)</span> <span class="comment"># select altloc A atom</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">atom</span><span class="operator">.</span><span class="name">get_altloc</span><span class="punctuation">()</span>
<span class="literal string">&quot;A&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atom</span><span class="operator">.</span><span class="name">disordered_select</span><span class="punctuation">(</span><span class="literal string">'B'</span><span class="punctuation">)</span> <span class="comment"># select altloc B atom</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">atom</span><span class="operator">.</span><span class="name">get_altloc</span><span class="punctuation">()</span>
<span class="literal string">&quot;B&quot;</span>
</pre>
</div>
<div class="section" id="id104">
<h3>11.3.3  紊乱残基</h3>
<div class="section" id="id105">
<h4>普通例子</h4>
<p>最常见的例子是一个残基包含一个或多个紊乱原子。这显然可以通过用DisorderedAtom对象表示这些紊乱原子来解决，并将DisorderedAtom对象存储在一个Residue对象中，就像正常的Atom对象那样。通过将所有未捕获方法调用发送给其中一个Atom对象（被选定的Atom对象），DisorderedAtom对象表现完全像一个正常的原子对象（事实上这个原子有最高的使用率）。</p>
</div>
<div class="section" id="id106">
<h4>点突变</h4>
<p>一个特殊的例子就是当紊乱是由点突变导致的时候，也就是说，在晶体结构中出现一条多肽的两或多个点突变。关于这一点，可以在PDB结构1EN2中找到一个例子。</p>
<p>既然这些残基属于不同的残基类型（举例说Ser 60 和Cys 60），那么它们不应该像通常情况一样存储在一个单一 <tt class="docutils literal">Residue</tt> 对象中。这种情况下每个残基用一个 <tt class="docutils literal">Residue</tt> 对象来表示，两种 <tt class="docutils literal">Residue</tt> 对象都保存在一个单一 <tt class="docutils literal">DisorderedResidue</tt> 对象中（见 图. <a class="reference external" href="#fig:smcra">11.1</a> ）。</p>
<p><tt class="docutils literal">DisorderedResidue</tt> 对象将所有未捕获方法发送给选定的 <tt class="docutils literal">Residue</tt> 对象（默认是所添加的最后一个 <tt class="docutils literal">Residue</tt> 对象），因此表现得像一个正常的残基。在 <tt class="docutils literal">DisorderedResidue</tt> 中每个 <tt class="docutils literal">Residue</tt> 对象可通过残基名称来唯一标识。在上述例子中，残基Ser 60在 <tt class="docutils literal">DisorderedResidue</tt> 对象中的id为“SER”，而残基Cys 60则是“CYS”。用户可以通过这个id选择在 <tt class="docutils literal">DisorderedResidue</tt> 中的有效 <tt class="docutils literal">Residue</tt> 对象。</p>
<p>例子：假设一个链在位置10有一个由Ser和Cys残基构成的点突变。确信这个链的残基10表现为Cys残基。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span> <span class="operator">=</span> <span class="name">chain</span><span class="punctuation">[</span><span class="literal number integer">10</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span><span class="operator">.</span><span class="name">disordered_select</span><span class="punctuation">(</span><span class="literal string">'CYS'</span><span class="punctuation">)</span>
</pre>
<p>另外，通过使用 <tt class="docutils literal">(Disordered)Residue</tt> 对象的 <tt class="docutils literal">get_unpacked_list</tt> 方法，你能获得所有 <tt class="docutils literal">Atom</tt> 对象的列表（也就是说，所有 <tt class="docutils literal">DisorderedAtom</tt> 对象解包到它们各自的 <tt class="docutils literal">Atom</tt> 对象）。</p>
</div>
</div>
</div>
<div class="section" id="id107">
<h2>11.4  异质残基</h2>
<div class="section" id="id108">
<h3>11.4.1  相关问题</h3>
<p>关于异质残基的一个很普遍的问题是同一条链中的若干异质和非异质残基有同样的序列标识符（和插入码）。因此，要为每个异质残基生成唯一的id，水分子和其他异质残基应该以不同的方式来对待。</p>
<p>记住Residue残基有一个元组（hetfield, resseq, icode）作为id。hetfield值为空(“ ”)表示为氨基酸和核酸；为一个字符串，则表示水分子和其他异质残基。hetfield的内容将在下面解释。</p>
</div>
<div class="section" id="id109">
<h3>11.4.2  水残基</h3>
<p>水残基的hetfield字符串由字母“W”构成。所以水分子的一个典型的残基id为(“W”, 1, “ ”)。</p>
</div>
<div class="section" id="id110">
<h3>11.4.3  其他异质残基</h3>
<p>其他异质残基的hetfield字符以“H_”起始，后接残基名称。一个葡萄糖分子，比如残基名称为“GLC”，则hetfield字符为“H_GLC”；它的残基id可以是(“H_GLC”, 1,
“ ”)。</p>
</div>
</div>
<div class="section" id="structure">
<h2>11.5  浏览Structure对象</h2>
<div class="section" id="pdb-modelchainresidueatom">
<h3>解析PDB文件，提取一些Model、Chain、Residue和Atom对象</h3>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.PDB.PDBParser</span> <span class="keyword namespace">import</span> <span class="name">PDBParser</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">parser</span> <span class="operator">=</span> <span class="name">PDBParser</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">structure</span> <span class="operator">=</span> <span class="name">parser</span><span class="operator">.</span><span class="name">get_structure</span><span class="punctuation">(</span><span class="literal string">&quot;test&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;1fat.pdb&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">structure</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">chain</span> <span class="operator">=</span> <span class="name">model</span><span class="punctuation">[</span><span class="literal string">&quot;A&quot;</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span> <span class="operator">=</span> <span class="name">chain</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atom</span> <span class="operator">=</span> <span class="name">residue</span><span class="punctuation">[</span><span class="literal string">&quot;CA&quot;</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="id111">
<h3>迭代遍历一个结构中的所有原子</h3>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">p</span> <span class="operator">=</span> <span class="name">PDBParser</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">structure</span> <span class="operator">=</span> <span class="name">p</span><span class="operator">.</span><span class="name">get_structure</span><span class="punctuation">(</span><span class="literal string">'X'</span><span class="punctuation">,</span> <span class="literal string">'pdb1fat.ent'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">model</span> <span class="operator word">in</span> <span class="name">structure</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">chain</span> <span class="operator word">in</span> <span class="name">model</span><span class="punctuation">:</span>
<span class="operator">...</span>         <span class="keyword">for</span> <span class="name">residue</span> <span class="operator word">in</span> <span class="name">chain</span><span class="punctuation">:</span>
<span class="operator">...</span>             <span class="keyword">for</span> <span class="name">atom</span> <span class="operator word">in</span> <span class="name">residue</span><span class="punctuation">:</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="name">atom</span>
<span class="operator">...</span>
</pre>
<p>有个快捷方式可以遍历一个结构中所有原子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atoms</span> <span class="operator">=</span> <span class="name">structure</span><span class="operator">.</span><span class="name">get_atoms</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">atom</span> <span class="operator word">in</span> <span class="name">atoms</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">atom</span>
<span class="operator">...</span>
</pre>
<p>类似地，遍历一条链中的所有原子，可以这么做：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atoms</span> <span class="operator">=</span> <span class="name">chain</span><span class="operator">.</span><span class="name">get_atoms</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">atom</span> <span class="operator word">in</span> <span class="name">atoms</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">atom</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="id112">
<h3>遍历模型中的所有残基</h3>
<p>或者，如果你想遍历在一条模型中的所有残基：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residues</span> <span class="operator">=</span> <span class="name">model</span><span class="operator">.</span><span class="name">get_residues</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">residue</span> <span class="operator word">in</span> <span class="name">residues</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">residue</span>
<span class="operator">...</span>
</pre>
<p>你也可以用 <tt class="docutils literal">Selection.unfold_entities</tt> 函数来获取一个结构的所有残基：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">res_list</span> <span class="operator">=</span> <span class="name">Selection</span><span class="operator">.</span><span class="name">unfold_entities</span><span class="punctuation">(</span><span class="name">structure</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">)</span>
</pre>
<p>或者获得链上的所有原子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">atom_list</span> <span class="operator">=</span> <span class="name">Selection</span><span class="operator">.</span><span class="name">unfold_entities</span><span class="punctuation">(</span><span class="name">chain</span><span class="punctuation">,</span> <span class="literal string">'A'</span><span class="punctuation">)</span>
</pre>
<p>明显的是， <tt class="docutils literal">A=atom, R=residue, C=chain, M=model, S=structure</tt> 。你可以用这种标记返回层次中的上层，如从一个 <tt class="docutils literal">Atoms</tt> 列表得到（唯一的） <tt class="docutils literal">Residue</tt> 或 <tt class="docutils literal">Chain</tt> 父类的列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue_list</span> <span class="operator">=</span> <span class="name">Selection</span><span class="operator">.</span><span class="name">unfold_entities</span><span class="punctuation">(</span><span class="name">atom_list</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">chain_list</span> <span class="operator">=</span> <span class="name">Selection</span><span class="operator">.</span><span class="name">unfold_entities</span><span class="punctuation">(</span><span class="name">atom_list</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">)</span>
</pre>
<p>更多信息详见API文档。</p>
</div>
<div class="section" id="resseq-10-glc">
<h3>从链中提取异质残基（如resseq 10的葡萄糖（GLC）部分）</h3>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue_id</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="literal string">&quot;H_GLC&quot;</span><span class="punctuation">,</span> <span class="literal number integer">10</span><span class="punctuation">,</span> <span class="literal string">&quot; &quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue</span> <span class="operator">=</span> <span class="name">chain</span><span class="punctuation">[</span><span class="name">residue_id</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="id113">
<h3>打印链中所有异质残基</h3>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">residue</span> <span class="operator word">in</span> <span class="name">chain</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>    <span class="name">residue_id</span> <span class="operator">=</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_id</span><span class="punctuation">()</span>
<span class="operator">...</span>    <span class="name">hetfield</span> <span class="operator">=</span> <span class="name">residue_id</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">...</span>    <span class="keyword">if</span> <span class="name">hetfield</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">==</span><span class="literal string">&quot;H&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>        <span class="keyword">print</span> <span class="name">residue_id</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="b50ca">
<h3>输出一个结构分子中所有B因子大于50的CA原子的坐标</h3>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">model</span> <span class="operator word">in</span> <span class="name">structure</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">chain</span> <span class="operator word">in</span> <span class="name">model</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>         <span class="keyword">for</span> <span class="name">residue</span> <span class="operator word">in</span> <span class="name">chain</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>             <span class="keyword">if</span> <span class="name">residue</span><span class="operator">.</span><span class="name">has_id</span><span class="punctuation">(</span><span class="literal string">&quot;CA&quot;</span><span class="punctuation">):</span>
<span class="operator">...</span>                 <span class="name">ca</span> <span class="operator">=</span> <span class="name">residue</span><span class="punctuation">[</span><span class="literal string">&quot;CA&quot;</span><span class="punctuation">]</span>
<span class="operator">...</span>                 <span class="keyword">if</span> <span class="name">ca</span><span class="operator">.</span><span class="name">get_bfactor</span><span class="punctuation">()</span> <span class="operator">&gt;</span> <span class="literal number float">50.0</span><span class="punctuation">:</span>
<span class="operator">...</span>                     <span class="keyword">print</span> <span class="name">ca</span><span class="operator">.</span><span class="name">get_coord</span><span class="punctuation">()</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="id114">
<h3>输出所有含紊乱原子的残基</h3>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">model</span> <span class="operator word">in</span> <span class="name">structure</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">chain</span> <span class="operator word">in</span> <span class="name">model</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>         <span class="keyword">for</span> <span class="name">residue</span> <span class="operator word">in</span> <span class="name">chain</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>             <span class="keyword">if</span> <span class="name">residue</span><span class="operator">.</span><span class="name">is_disordered</span><span class="punctuation">():</span>
<span class="operator">...</span>                 <span class="name">resseq</span> <span class="operator">=</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_id</span><span class="punctuation">()[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="operator">...</span>                 <span class="name">resname</span> <span class="operator">=</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_resname</span><span class="punctuation">()</span>
<span class="operator">...</span>                 <span class="name">model_id</span> <span class="operator">=</span> <span class="name">model</span><span class="operator">.</span><span class="name">get_id</span><span class="punctuation">()</span>
<span class="operator">...</span>                 <span class="name">chain_id</span> <span class="operator">=</span> <span class="name">chain</span><span class="operator">.</span><span class="name">get_id</span><span class="punctuation">()</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="name">model_id</span><span class="punctuation">,</span> <span class="name">chain_id</span><span class="punctuation">,</span> <span class="name">resname</span><span class="punctuation">,</span> <span class="name">resseq</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="altloc-a">
<h3>遍历所有紊乱原子，并选取所有具有altloc A的原子（如果有的话）</h3>
<p>这将会保证，SMCRA数据结构会表现得如同只存在altloc A原子一样。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">model</span> <span class="operator word">in</span> <span class="name">structure</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">chain</span> <span class="operator word">in</span> <span class="name">model</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>         <span class="keyword">for</span> <span class="name">residue</span> <span class="operator word">in</span> <span class="name">chain</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>             <span class="keyword">if</span> <span class="name">residue</span><span class="operator">.</span><span class="name">is_disordered</span><span class="punctuation">():</span>
<span class="operator">...</span>                 <span class="keyword">for</span> <span class="name">atom</span> <span class="operator word">in</span> <span class="name">residue</span><span class="operator">.</span><span class="name">get_list</span><span class="punctuation">():</span>
<span class="operator">...</span>                     <span class="keyword">if</span> <span class="name">atom</span><span class="operator">.</span><span class="name">is_disordered</span><span class="punctuation">():</span>
<span class="operator">...</span>                         <span class="keyword">if</span> <span class="name">atom</span><span class="operator">.</span><span class="name">disordered_has_id</span><span class="punctuation">(</span><span class="literal string">&quot;A&quot;</span><span class="punctuation">):</span>
<span class="operator">...</span>                             <span class="name">atom</span><span class="operator">.</span><span class="name">disordered_select</span><span class="punctuation">(</span><span class="literal string">&quot;A&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="id115">
<h3>从 <tt class="docutils literal">Structure</tt> 对象中提取多肽</h3>
<p>为了从一个结构中提取多肽，需要用 <tt class="docutils literal">PolypeptideBuilder</tt> 从 <tt class="docutils literal">Structure</tt> 构建一个 <tt class="docutils literal">Polypeptide</tt> 对象的列表，如下所示：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model_nr</span> <span class="operator">=</span> <span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">polypeptide_list</span> <span class="operator">=</span> <span class="name">build_peptides</span><span class="punctuation">(</span><span class="name">structure</span><span class="punctuation">,</span> <span class="name">model_nr</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">polypeptide</span> <span class="operator word">in</span> <span class="name">polypeptide_list</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">polypeptide</span>
<span class="operator">...</span>
</pre>
<p>Polypeptide对象正是Residue对象的一个UserList，总是从单结构域（在此例中为模型1）中创建而来。你可以用所得 <tt class="docutils literal">Polypeptide</tt> 对象来获取序列作为 <tt class="docutils literal">Seq</tt> 对象，或获得Cα原子的列表。多肽可以通过一个C-N 化学键或一个Cα-Cα化学键距离标准来建立。</p>
<p>例子：</p>
<pre class="code python literal-block">
<span class="comment"># Using C-N</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ppb</span><span class="operator">=</span><span class="name">PPBuilder</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pp</span> <span class="operator word">in</span> <span class="name">ppb</span><span class="operator">.</span><span class="name">build_peptides</span><span class="punctuation">(</span><span class="name">structure</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pp</span><span class="operator">.</span><span class="name">get_sequence</span><span class="punctuation">()</span>
<span class="operator">...</span>
<span class="comment"># Using CA-CA</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ppb</span><span class="operator">=</span><span class="name">CaPPBuilder</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pp</span> <span class="operator word">in</span> <span class="name">ppb</span><span class="operator">.</span><span class="name">build_peptides</span><span class="punctuation">(</span><span class="name">structure</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pp</span><span class="operator">.</span><span class="name">get_sequence</span><span class="punctuation">()</span>
<span class="operator">...</span>
</pre>
<p>需要注意的是，上例中通过 <tt class="docutils literal">PolypeptideBuilder</tt> 只考虑了结构的模型 0。尽管如此，还是可以用 <tt class="docutils literal">PolypeptideBuilder</tt> 从 <tt class="docutils literal">Model</tt> 和 <tt class="docutils literal">Chain</tt> 对象创建 <tt class="docutils literal">Polypeptide</tt> 对象。</p>
</div>
<div class="section" id="id116">
<h3>获取结构的序列</h3>
<p>要做的第一件事就是从结构中提取所有多肽（如上所述）。然后每条多肽的序列就容易从 <tt class="docutils literal">Polypeptide</tt> 对象获得。该序列表示为一个Biopython <tt class="docutils literal">Seq</tt> 对象，它的字母表由 <tt class="docutils literal">ProteinAlphabet</tt> 对象来定义。</p>
<p>例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">seq</span> <span class="operator">=</span> <span class="name">polypeptide</span><span class="operator">.</span><span class="name">get_sequence</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'SNVVE...'</span><span class="punctuation">,</span> <span class="operator">&lt;</span><span class="keyword">class</span> <span class="name class">Bio</span><span class="operator">.</span><span class="name">Alphabet</span><span class="operator">.</span><span class="name">ProteinAlphabet</span><span class="operator">&gt;</span><span class="punctuation">)</span>
</pre>
</div>
</div>
<div class="section" id="id117">
<h2>11.6  分析结构</h2>
<div class="section" id="id118">
<h3>11.6.1  度量距离</h3>
<p>重载原子的减法运算来返回两个原子之间的距离。</p>
<pre class="code python literal-block">
<span class="comment"># Get some atoms</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ca1</span> <span class="operator">=</span> <span class="name">residue1</span><span class="punctuation">[</span><span class="literal string">'CA'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ca2</span> <span class="operator">=</span> <span class="name">residue2</span><span class="punctuation">[</span><span class="literal string">'CA'</span><span class="punctuation">]</span>
<span class="comment"># Simply subtract the atoms to get their distance</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">distance</span> <span class="operator">=</span> <span class="name">ca1</span><span class="operator">-</span><span class="name">ca2</span>
</pre>
</div>
<div class="section" id="id119">
<h3>11.6.2  度量角度</h3>
<p>用原子坐标的向量表示，和 <tt class="docutils literal">Vector</tt> 模块中的 <tt class="docutils literal">calc_angle</tt> 函数可以计算角度。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector1</span> <span class="operator">=</span> <span class="name">atom1</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector2</span> <span class="operator">=</span> <span class="name">atom2</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector3</span> <span class="operator">=</span> <span class="name">atom3</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">angle</span> <span class="operator">=</span> <span class="name">calc_angle</span><span class="punctuation">(</span><span class="name">vector1</span><span class="punctuation">,</span> <span class="name">vector2</span><span class="punctuation">,</span> <span class="name">vector3</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="id120">
<h3>11.6.3  度量扭转角</h3>
<p>用原子坐标的向量表示，然后用 <tt class="docutils literal">Vector</tt> 模块中的 <tt class="docutils literal">calc_dihedral</tt> 函数可以计算角度。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector1</span> <span class="operator">=</span> <span class="name">atom1</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector2</span> <span class="operator">=</span> <span class="name">atom2</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector3</span> <span class="operator">=</span> <span class="name">atom3</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">vector4</span> <span class="operator">=</span> <span class="name">atom4</span><span class="operator">.</span><span class="name">get_vector</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">angle</span> <span class="operator">=</span> <span class="name">calc_dihedral</span><span class="punctuation">(</span><span class="name">vector1</span><span class="punctuation">,</span> <span class="name">vector2</span><span class="punctuation">,</span> <span class="name">vector3</span><span class="punctuation">,</span> <span class="name">vector4</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="id121">
<h3>11.6.4  确定原子-原子触点</h3>
<p>用 <tt class="docutils literal">NeighborSearch</tt> 来进行邻接查询。用C语言写的（使得运行很快）KD树模块（见 <tt class="docutils literal">Bio.KDTree</tt> ）可以用来完成邻接查询。它也包含了一个快速方法来找出相距一定距离的所有点对。</p>
</div>
<div class="section" id="id122">
<h3>11.6.5  叠加两个结构</h3>
<p>可以用 <tt class="docutils literal">Superimposer</tt> 对象将两个坐标集叠加。这个对象计算出旋转和平移矩阵，该矩阵旋转两个列表上相重叠的原子使其满足RMSD最小。当然这两个列表含有相同数目的原子。 <tt class="docutils literal">Superimposer</tt> 对象也可以将旋转/平移应用在一列原子上。旋转和平移作为一个元组储存在 <tt class="docutils literal">Superimposer</tt> 对象的 <tt class="docutils literal">rotran</tt> 属性中（注意，旋转是右乘），RMSD储存在属性 <tt class="docutils literal">rmsd</tt> 中。</p>
<p><tt class="docutils literal">Superimposer</tt> 使用的算法来自[<a class="reference external" href="#golub1989">17</a>,
Golub &amp; Van Loan]并使用了奇异值分解（这是通用 <tt class="docutils literal">Bio.SVDSuperimposer</tt> 模块中实现了的）。</p>
<p>例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sup</span> <span class="operator">=</span> <span class="name">Superimposer</span><span class="punctuation">()</span>
<span class="comment"># Specify the atom lists</span>
<span class="comment"># 'fixed' and 'moving' are lists of Atom objects</span>
<span class="comment"># The moving atoms will be put on the fixed atoms</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sup</span><span class="operator">.</span><span class="name">set_atoms</span><span class="punctuation">(</span><span class="name">fixed</span><span class="punctuation">,</span> <span class="name">moving</span><span class="punctuation">)</span>
<span class="comment"># Print rotation/translation/rmsd</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">sup</span><span class="operator">.</span><span class="name">rotran</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">sup</span><span class="operator">.</span><span class="name">rms</span>
<span class="comment"># Apply rotation/translation to the moving atoms</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sup</span><span class="operator">.</span><span class="name">apply</span><span class="punctuation">(</span><span class="name">moving</span><span class="punctuation">)</span>
</pre>
<p>为了基于有效位点来叠加两个结构，用有效位点的原子来计算旋转/平移矩阵（如上所述），并应用到整个分子。</p>
</div>
<div class="section" id="id123">
<h3>11.6.6  双向映射两个相关结构的残基</h3>
<p>首先，创建一个FASTA格式的比对文件，然后使用``StructureAlignment`` 类。这个类也可以用来比对两个以上的结构。</p>
</div>
<div class="section" id="hse">
<h3>11.6.7  计算半球暴露（HSE）</h3>
<p>半球暴露（Half Sphere Exposure，HSE）是对溶剂暴露 [<a class="reference external" href="#hamelryck2005">20</a>]的一种新的二维度量。根本上，它计数了围绕一个残基，在其侧链方向上及反方向（在13 Å范围内）的Cα原子。尽管简单，它表现得比溶剂暴露的其它度量都要好。</p>
<p>HSE有两种风味：HSEα和HSEβ。前者仅用到Cα原子的位置，而后者用到Cα和Cβ原子的位置。HSE度量是由 <tt class="docutils literal">HSExposure</tt> 类来计算的，这个类也能计算触点数目。后一个类有方法能返回一个字典，该字典将一个``Residue`` 对象映射到相应的HSEα,HSEβ和触点数目值。</p>
<p>例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">structure</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">hse</span> <span class="operator">=</span> <span class="name">HSExposure</span><span class="punctuation">()</span>
<span class="comment"># Calculate HSEalpha</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exp_ca</span> <span class="operator">=</span> <span class="name">hse</span><span class="operator">.</span><span class="name">calc_hs_exposure</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">option</span><span class="operator">=</span><span class="literal string">'CA3'</span><span class="punctuation">)</span>
<span class="comment"># Calculate HSEbeta</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exp_cb</span><span class="operator">=</span><span class="name">hse</span><span class="operator">.</span><span class="name">calc_hs_exposure</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">option</span><span class="operator">=</span><span class="literal string">'CB'</span><span class="punctuation">)</span>
<span class="comment"># Calculate classical coordination number</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exp_fs</span> <span class="operator">=</span> <span class="name">hse</span><span class="operator">.</span><span class="name">calc_fs_exposure</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">)</span>
<span class="comment"># Print HSEalpha for a residue</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">exp_ca</span><span class="punctuation">[</span><span class="name">some_residue</span><span class="punctuation">]</span>
</pre>
</div>
<div class="section" id="id124">
<h3>11.6.8  确定二级结构</h3>
<p>为了这个功能，你需要安装DSSP（并获得一个对学术性使用免费的证书，参见 <a class="reference external" href="http://www.cmbi.kun.nl/gv/dssp/">http://www.cmbi.kun.nl/gv/dssp/</a> ）。然后用 <tt class="docutils literal">DSSP</tt> 类，可以映射 <tt class="docutils literal">Residue</tt> 对象到其二级结构上（和溶剂可及表面区域）。DSSP代码如下表所列表 <a class="reference external" href="#cap:DSSP-codes">11.1</a> 。注意DSSP（程序及其相应的类）不能处理多个模型！</p>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="22%" />
<col width="78%" />
</colgroup>
<tbody valign="top">
<tr><td>Code</td>
<td>Secondary structure</td>
</tr>
<tr><td>H</td>
<td>α-helix</td>
</tr>
<tr><td>B</td>
<td>Isolated β-bridge residue</td>
</tr>
<tr><td>E</td>
<td>Strand</td>
</tr>
<tr><td>G</td>
<td>3-10 helix</td>
</tr>
<tr><td>I</td>
<td>Π-helix</td>
</tr>
<tr><td>T</td>
<td>Turn</td>
</tr>
<tr><td>S</td>
<td>Bend</td>
</tr>
<tr><td><ul class="first last simple">
<li></li>
</ul>
</td>
<td>Other</td>
</tr>
</tbody>
</table>
<p>Table 11.1: Bio.PDB中的DSSP代码。</p>
<hr class="docutils" />
<p><tt class="docutils literal">DSSP</tt> 类也可以用来计算残基的溶剂可及表面。还请参考 <a class="reference external" href="#subsec:residue_depth">11.6.9</a> 。</p>
</div>
<div class="section" id="id125">
<h3>11.6.9  计算残基深度</h3>
<p>残基深度是残基原子到溶剂可及表面的平均距离。它是溶剂可及性的一种相当新颖和非常强大的参数化。为了这个功能，你需要安装Michel Sanner的 MSMS程序（ <a class="reference external" href="http://www.scripps.edu/pub/olson-web/people/sanner/html/msms_home.html">http://www.scripps.edu/pub/olson-web/people/sanner/html/msms_home.html</a> ）。然后使用 <tt class="docutils literal">ResidueDepth</tt> 类。这个类像字典一样将 <tt class="docutils literal">Residue</tt> 对象映射到相应的（残基深度，Cα深度）元组。Cα深度是残基的Cα原子到溶剂可及表面的距离。</p>
<p>例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">structure</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">rd</span> <span class="operator">=</span> <span class="name">ResidueDepth</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">pdb_file</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">residue_depth</span><span class="punctuation">,</span> <span class="name">ca_depth</span><span class="operator">=</span><span class="name">rd</span><span class="punctuation">[</span><span class="name">some_residue</span><span class="punctuation">]</span>
</pre>
<p>你也可以以带有表面点的数值Python数组的形式获得分子表面本身（通过 <tt class="docutils literal">get_surface</tt> 函数）。</p>
</div>
</div>
<div class="section" id="id126">
<h2>11.7  PDB文件中的常见问题</h2>
<p>众所周知，很多PDB文件包含语义错误（不是结构本身的错误，而是在PDB文件中的表示）。Bio.PDB可以有两种方式来处理这个问题。PDBParser对象能表现出两种方式：严格方式和宽容方式（默认方式）：</p>
<p>例子:</p>
<pre class="code python literal-block">
<span class="comment"># Permissive parser</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">parser</span> <span class="operator">=</span> <span class="name">PDBParser</span><span class="punctuation">(</span><span class="name">PERMISSIVE</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">parser</span> <span class="operator">=</span> <span class="name">PDBParser</span><span class="punctuation">()</span> <span class="comment"># The same (default)</span>
<span class="comment"># Strict parser</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">strict_parser</span> <span class="operator">=</span> <span class="name">PDBParser</span><span class="punctuation">(</span><span class="name">PERMISSIVE</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">)</span>
</pre>
<p>在宽容状态（默认），明显包含错误的PDB文件会被“纠正”（比如说一些残基或原子丢失）。这些错误包括：</p>
<ul class="simple">
<li>多个残基使用同一个标识符</li>
<li>多个原子使用统一个标识符（考虑altloc识别符）</li>
</ul>
<p>这些错误暗示了PDB文件中确实存在错误（详情见 [<a class="reference external" href="#hamelryck2003a">18</a>, Hamelryck and Manderick, 2003] ）。在严格模式，带错的PDB文件会引发异常，这有助于发现PDB文件中的错误。</p>
<p>但是有些错误能自动修正。正常情况下，每个紊乱原子应该会有一个非空altloc标识符。可是很多结构没有遵循这个惯例，而在同一原子的两个紊乱位置存在一个空的和一个非空的标识符。这个错误会被以正确的方式自动解析。</p>
<p>有时候一个结构会有这样的情况：一部分残基属于A链，接下来一部分残基属于B链，然后又有一部分残基属于A链，也就是说，这种链是“断的”。这也能被自动正确解析。</p>
<div class="section" id="id127">
<h3>11.7.1  例子</h3>
<p>PDBParser/Structure类经过了将近800个结构（每个都属于不同的SCOP超家族）上的测试。测试总共耗时20分钟左右，或者说平均每个结构只需1.5秒。在一台1000 MHz的PC上只需10秒就可解析包含近64000个原子的大核糖体亚基（1FKK）的结构。</p>
<p>当不能建立明确的数据结构时会发生三类异常。在这三类异常中，可能的起因是PDB文件中一个本应修正的错误。这些情况下产生异常要比冒险地错误描述一个数据结构中的结构好得多。</p>
<div class="section" id="id128">
<h4>11.7.1.1  重复残基</h4>
<p>一个结构包含在一条链中具有相同的序列标识符（resseq 3）和icode的两个氨基酸残基。仔细观察可以发现这条链包含残基：Thr A3, …, Gly A202, Leu A3, Glu A204。很明显第二个Leu A3应该是Leu A203。类似的情况也存在于结构1FFK（比如它包含残基Gly B64, Met B65, Glu B65, Thr B67，也就是说Glu B65应该是Glu B66）上。</p>
</div>
<div class="section" id="id129">
<h4>11.7.1.2  重复原子</h4>
<p>结构1EJG含有在A链22位的一个Ser/Pro点突变。依次，Ser 22含一些紊乱原子。和期望的一样，所有属于 Ser 22的原子都有一个非空的altloc标识符（B或C）。所有Pro 22的原子都有altloc A，除了含空altloc的N原子。这会生成一个异常，因为一个点突变处属于两个残基的所有原子都应该有非空的altloc。结果这个原子很可能被Ser 和 Pro 22共用，而Ser22丢失了这个N原子。此外，这也点出了文件中的一个问题：这个N原子应该出现在Ser和Pro残基中，两种情形下都与合适的altloc标识符关联。</p>
</div>
</div>
<div class="section" id="id130">
<h3>11.7.2  自动纠正</h3>
<p>一些错误相当普遍且能够在没有太大误解风险的情况下容易地纠正过来。这些错误列在下面。</p>
<div class="section" id="altloc">
<h4>11.7.2.1  紊乱原子的空altloc</h4>
<p>正常情况下，每个紊乱原子应该会有一个非空altloc标识符，可是很多结构没有遵循这个惯例，而是在同一原子的两个紊乱位置存在一个空的和一个非空的标识符。这个错误会被以正确的方式自动解析。</p>
</div>
<div class="section" id="id131">
<h4>11.7.2.2  断链</h4>
<p>有时候一个结构会有这样的情况：一部分残基属于A链，接下来一部分残基属于B链，然后又有一部分残基属于A链，也就是说，链是“断的”，这也能被正确的解析。</p>
</div>
</div>
<div class="section" id="id132">
<h3>11.7.3  致命错误</h3>
<p>有时候一个PDB文件不能被明确解释。这会产生异常并等待用户去修正这个PDB文件，而不是猜测和冒出错的风险。这些异常列在下面。</p>
<div class="section" id="id133">
<h4>11.7.3.1  重复残基</h4>
<p>在一条链上的所有残基都应该有一个唯一的id。该id基于下述生成：</p>
<ul class="simple">
<li>序列标识符（resseq）</li>
<li>插入码（icode）</li>
<li>hetfield字符（“W”代表水，“H_”后面的残基名称代表其他异质残基）</li>
<li>发生点突变的残基的名称（在DisorderedResidue对象中存储Residue对象）</li>
</ul>
<p>如果这样还不能生成一个唯一的id，那么肯定是一些地方出了错，这时会生成一个异常。</p>
</div>
<div class="section" id="id134">
<h4>11.7.3.2  重复原子</h4>
<p>一个残基上所有原子应该有一个唯一的id，这个id基于下述产生：</p>
<ul class="simple">
<li>原子名称（不带空格，否则会报错）</li>
<li>altloc标识符</li>
</ul>
<p>如果这样还不能生成一个唯一的id，那么肯定是一些地方出了错，这时会生成一个异常。</p>
</div>
</div>
</div>
<div class="section" id="protein-data-bank">
<h2>11.8  访问Protein Data Bank</h2>
<div class="section" id="id135">
<h3>11.8.1  从Protein Data Bank下载结构</h3>
<p>结构可以从PDB（Protein Data Bank）通过 <tt class="docutils literal">PDBList</tt> 对象的 <tt class="docutils literal">retrieve_pdb_file</tt> 方法下载。这种方法的要点是结构的PDB标识符。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pdbl</span> <span class="operator">=</span> <span class="name">PDBList</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pdbl</span><span class="operator">.</span><span class="name">retrieve_pdb_file</span><span class="punctuation">(</span><span class="literal string">'1FAT'</span><span class="punctuation">)</span>
</pre>
<p><tt class="docutils literal">PDBList</tt> 类也能用作命令行工具：</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">PDBList</span><span class="operator">.</span><span class="name">py</span> <span class="literal number integer">1</span><span class="name">fat</span>
</pre>
<p>下载的文件将以 <tt class="docutils literal">pdb1fat.ent</tt> 为名保存在当前工作目录。注意 <tt class="docutils literal">retrieve_pdb_file</tt> 方法还有个可选参数 <tt class="docutils literal">pdir</tt> 用来指定一个特定的路径来保存所下载的PDB文件。</p>
<p><tt class="docutils literal">retrieve_pdb_file</tt> 方法还有其他选项可以指定下载所用的压缩格式（默认的 <tt class="docutils literal">.Z</tt> 格式和 <tt class="docutils literal">gunzip</tt> 格式）。另外，在创建 <tt class="docutils literal">PDBList</tt> 对象时还可以指定PDB ftp站点。默认使用Worldwide Protein Data Bank（ <a class="reference external" href="ftp://ftp.wwpdb.org/pub/pdb/data/structures/divided/pdb/">ftp://ftp.wwpdb.org/pub/pdb/data/structures/divided/pdb/</a> ）。详细内容参见API文档。再次感谢Kristian Rother对此模块的所做的贡献。</p>
</div>
<div class="section" id="id136">
<h3>11.8.2  下载整个PDB</h3>
<p>下面的命令将会保存所有PDB文件至 <tt class="docutils literal">/data/pdb</tt> 目录：</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">PDBList</span><span class="operator">.</span><span class="name">py</span> <span class="name builtin">all</span> <span class="operator">/</span><span class="name">data</span><span class="operator">/</span><span class="name">pdb</span>

<span class="name">python</span> <span class="name">PDBList</span><span class="operator">.</span><span class="name">py</span> <span class="name builtin">all</span> <span class="operator">/</span><span class="name">data</span><span class="operator">/</span><span class="name">pdb</span> <span class="operator">-</span><span class="name">d</span>
</pre>
<p>在API中这个方法叫做 <tt class="docutils literal">download_entire_pdb</tt> 。添加 <tt class="docutils literal"><span class="pre">-d</span></tt> 会在同一目录下保存所有文件。否则将分别保存至PDB风格的、与其PDB ID对应的子目录中。根据网速，完整的下载全部PDB文件大概需要2-4天。</p>
</div>
<div class="section" id="id137">
<h3>11.8.3  保持本地PDB拷贝的更新</h3>
<p>这也能通过 <tt class="docutils literal">PDBList</tt> 对象来完成。可以简单的创建一个 <tt class="docutils literal">PDBList</tt> 对象（指定本地PDB拷贝的目录），然后调用 <tt class="docutils literal">update_pdb</tt> 方法：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pl</span> <span class="operator">=</span> <span class="name">PDBList</span><span class="punctuation">(</span><span class="name">pdb</span><span class="operator">=</span><span class="literal string">'/data/pdb'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pl</span><span class="operator">.</span><span class="name">update_pdb</span><span class="punctuation">()</span>
</pre>
<p>当然还可以每周用 <tt class="docutils literal">cronjob</tt> 实现本地拷贝自动更新。还可以指定PDB ftp站点（详见API文档）。</p>
<p><tt class="docutils literal">PDBList</tt> 有其他许多其它方法可供调用。 <tt class="docutils literal">get_all_obsolete</tt> 方法可以获取所有已经废弃的PDB项的一个列表；  <tt class="docutils literal">changed_this_week</tt>  方法可以用于获得当前一周内新增加、修改或废弃的PDB项。更多 <tt class="docutils literal">PDBList</tt> 的用法参见API文档。</p>
</div>
</div>
<div class="section" id="id138">
<h2>11.9  常见问题</h2>
<div class="section" id="bio-pdb">
<h3>11.9.1  Bio.PDB测试得如何？</h3>
<p>事实上，相当好。Bio.PDB已经在从PDB获得的近5500个结构上广泛的测试过，所有结构都能正确地解析。更多细节可以参考在Bioinformatics上发表的关于Bio.PDB的文章。作为一个可靠的工具，Bio.PDB已经并正用于许多研究项目中。我几乎每天都在用它，出于研究目的、提升其性能和增加新属性。</p>
</div>
<div class="section" id="id139">
<h3>11.9.2  它有多快？</h3>
<p><tt class="docutils literal">PDBParser</tt> 的性能经过将近800个结构测试（每个都属于不同的SCOP超家族），总共花费20分钟左右，也就是说平均每个结构只需1.5秒。在一台1000 MHz的PC上解析巨大的包含近64000个原子的核糖体亚单位（1FKK）只需10秒。总而言之，它比很多应用程序都快得多。</p>
</div>
<div class="section" id="id140">
<h3>11.9.3  是否支持分子图形展示？</h3>
<p>不直接支持，很大程度上是因为已有相当多基于Python或Python-aware的解决方案，也可能会用到Bio.PDB。顺便说一下，我的选择是Pymol（我在Pymol中使用Bio.PDB非常成功，将来Bio.PDB中会有特定的PyMol模块）。基于Python或Python-aware的分子图形解决方案包括：</p>
<ul class="simple">
<li>PyMol:
<a class="reference external" href="http://pymol.sourceforge.net/">http://pymol.sourceforge.net/</a></li>
<li>Chimera:
<a class="reference external" href="http://www.cgl.ucsf.edu/chimera/">http://www.cgl.ucsf.edu/chimera/</a></li>
<li>PMV:
<a class="reference external" href="http://www.scripps.edu/~sanner/python/">http://www.scripps.edu/~sanner/python/</a></li>
<li>Coot:
<a class="reference external" href="http://www.ysbl.york.ac.uk/~emsley/coot/">http://www.ysbl.york.ac.uk/~emsley/coot/</a></li>
<li>CCP4mg:
<a class="reference external" href="http://www.ysbl.york.ac.uk/~lizp/molgraphics.html">http://www.ysbl.york.ac.uk/~lizp/molgraphics.html</a></li>
<li>mmLib: <a class="reference external" href="http://pymmlib.sourceforge.net/">http://pymmlib.sourceforge.net/</a></li>
<li>VMD:
<a class="reference external" href="http://www.ks.uiuc.edu/Research/vmd/">http://www.ks.uiuc.edu/Research/vmd/</a></li>
<li>MMTK:
<a class="reference external" href="http://starship.python.net/crew/hinsen/MMTK/">http://starship.python.net/crew/hinsen/MMTK/</a></li>
</ul>
</div>
<div class="section" id="id141">
<h3>11.9.4  谁在用Bio.PDB？</h3>
<p>Bio.PDB曾用于构建DISEMBL，一个能预测蛋白结构中的紊乱区域的web服务器( <a class="reference external" href="http://dis.embl.de/">http://dis.embl.de/</a> )；COLUMBA，一个提供注释过的蛋白结构的站点( <a class="reference external" href="http://www.columba-db.de/">http://www.columba-db.de/</a> )。Bio.PDB也用于进行PDB中蛋白质间有效位点的大规模相似性搜索[<a class="reference external" href="#hamelryck2003b">19</a>, Hamelryck, 2003]，用于开发新的算法来鉴别线性二级结构元件[<a class="reference external" href="#majumdar2005">26</a>, Majumdar <em>et al.</em>, 2005]。</p>
<p>从对特征和信息的需求判断，许多大型制药公司也使用Bio.PDB。</p>
</div>
</div>
</div>
<div class="section" id="bio-popgen">
<h1>第12章  Bio.PopGen：群体遗传学</h1>
<p>Bio.PopGen是一个群体遗传学相关的模块，在Biopython 1.44及以后的版本中可用。</p>
<p>该模块的中期目标是支持各种类型的数据格式、应用程序和数据库。目前，该模块正在紧张的开发中，并会快速实现对新特征的支持。这可能会带来一些不稳定的API，尤其是当你使用的是开发版。不过，我们正式公开发行的API应该更加稳定。</p>
<div class="section" id="genepop">
<h2>12.1  GenePop</h2>
<p>GenePop（ <a class="reference external" href="http://genepop.curtin.edu.au/">http://genepop.curtin.edu.au/</a>）是一款主流的群体遗传学软件包，支持Hardy-Weinberg检验、连锁不平衡、群体分化、基础统计计算、 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 和迁移率估计等等。GenePop并不支持基于序列的统计计算，因为它并不能处理序列数据。GenePop文件格式广泛用于多种其它的群体遗传学应用软件，因此成为群体遗传学领域重要格式。</p>
<p>Bio.PopGen提供GenePop文件格式解析器和生成器，同时也提供操作记录内容的小工具。此处有个关于怎样读取GenePop文件的示例（你可以在Biopython的Test/PopGen文件夹下找到GenePop示例文件）：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen</span> <span class="keyword namespace">import</span> <span class="name">GenePop</span>

<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;example.gen&quot;</span><span class="punctuation">)</span>
<span class="name">rec</span> <span class="operator">=</span> <span class="name">GenePop</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>它将读取名为example.gen的文件并解析。如果你输出rec，那么该记录将会以GenePop格式再次输出。</p>
<p>在rec中最重要的信息是基因座名称和群体信息（当然不止这些，请使用help(GenePop.Record)获得API帮助文档）。基因座名称可以在rec.loci_list中找到，群体信息可以在rec.populations中找到。群体信息是一个列表，每个群体（population）作为其中一个元素。每个元素本身又是包含个体（individual）的列表，每个个体包含个体名和等位基因列表（每个marker两个元素），下面是一个rec.populations的示例：</p>
<pre class="code python literal-block">
<span class="punctuation">[</span>
    <span class="punctuation">[</span>
        <span class="punctuation">(</span><span class="literal string">'Ind1'</span><span class="punctuation">,</span> <span class="punctuation">[(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">),</span>    <span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="literal number integer">200</span><span class="punctuation">,</span> <span class="literal number integer">201</span><span class="punctuation">)],</span>
        <span class="punctuation">(</span><span class="literal string">'Ind2'</span><span class="punctuation">,</span> <span class="punctuation">[(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name builtin pseudo">None</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="name builtin pseudo">None</span><span class="punctuation">,</span> <span class="name builtin pseudo">None</span><span class="punctuation">)],</span>
    <span class="punctuation">],</span>
    <span class="punctuation">[</span>
        <span class="punctuation">(</span><span class="literal string">'Other1'</span><span class="punctuation">,</span> <span class="punctuation">[(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">),</span>  <span class="punctuation">(</span><span class="literal number integer">4</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="literal number integer">200</span><span class="punctuation">,</span> <span class="literal number integer">200</span><span class="punctuation">)],</span>
    <span class="punctuation">]</span>
<span class="punctuation">]</span>
</pre>
<p>在上面的例子中，我们有两个群体，第一个群体包含两个个体，第二个群体只包含一个个体。第一个群体的第一个个体叫做Ind1，紧接着是3个基因座各自的等位基因信息。请注意，对于任何的基因座，信息可以缺失（如上述个体Ind2）。</p>
<p>有几个可用的工具函数可以处理GenePop记录，如下例：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen</span> <span class="keyword namespace">import</span> <span class="name">GenePop</span>

<span class="comment">#Imagine that you have loaded rec, as per the code snippet above...</span>

<span class="name">rec</span><span class="operator">.</span><span class="name">remove_population</span><span class="punctuation">(</span><span class="name">pos</span><span class="punctuation">)</span>
<span class="comment">#Removes a population from a record, pos is the population position in</span>
<span class="comment">#  rec.populations, remember that it starts on position 0.</span>
<span class="comment">#  rec is altered.</span>

<span class="name">rec</span><span class="operator">.</span><span class="name">remove_locus_by_position</span><span class="punctuation">(</span><span class="name">pos</span><span class="punctuation">)</span>
<span class="comment">#Removes a locus by its position, pos is the locus position in</span>
<span class="comment">#  rec.loci_list, remember that it starts on position 0.</span>
<span class="comment">#  rec is altered.</span>

<span class="name">rec</span><span class="operator">.</span><span class="name">remove_locus_by_name</span><span class="punctuation">(</span><span class="name">name</span><span class="punctuation">)</span>
<span class="comment">#Removes a locus by its name, name is the locus name as in</span>
<span class="comment">#  rec.loci_list. If the name doesn't exist the function fails</span>
<span class="comment">#  silently.</span>
<span class="comment">#  rec is altered.</span>

<span class="name">rec_loci</span> <span class="operator">=</span> <span class="name">rec</span><span class="operator">.</span><span class="name">split_in_loci</span><span class="punctuation">()</span>
<span class="comment">#Splits a record in loci, that is, for each loci, it creates a new</span>
<span class="comment">#  record, with a single loci and all populations.</span>
<span class="comment">#  The result is returned in a dictionary, being each key the locus name.</span>
<span class="comment">#  The value is the GenePop record.</span>
<span class="comment">#  rec is not altered.</span>

<span class="name">rec_pops</span> <span class="operator">=</span>  <span class="name">rec</span><span class="operator">.</span><span class="name">split_in_pops</span><span class="punctuation">(</span><span class="name">pop_names</span><span class="punctuation">)</span>
<span class="comment">#Splits a record in populations, that is, for each population, it creates</span>
<span class="comment">#  a new record, with a single population and all loci.</span>
<span class="comment">#  The result is returned in a dictionary, being each key</span>
<span class="comment">#  the population name. As population names are not available in GenePop,</span>
<span class="comment">#  they are passed in array (pop_names).</span>
<span class="comment">#  The value of each dictionary entry is the GenePop record.</span>
<span class="comment">#  rec is not altered.</span>
</pre>
<p>GenePop不支持群体名，这种限制有时会很麻烦。Biopython对群体名的支持正在规划中，这些功能扩展仍会保持对标准格式的兼容性。同时，中期目标是对GenePop网络服务的支持。</p>
</div>
<div class="section" id="coalescent-simulation">
<h2>12.2  溯祖模拟（Coalescent simulation）</h2>
<p>溯祖模拟是一种对群体遗传学信息根据时间向后推算的模型（backward model）。对祖先的模拟是通过寻找到最近共同祖先（Most Recent Common Ancestor，MRCA）完成。从MRCA到目前这一代样本间的血统关系有时称为家系（genealogy）。简单的情况是假定群体大小固定，单倍型，无群体结构，然后模拟无选择压的单个基因座的等位基因。</p>
<p>溯祖理论被广泛用于多种领域，如选择压力检测、真实群体的群体参数估计以及疾病基因图谱。</p>
<p>Biopython对溯祖的实现不是去创建一个新的内置模拟器，而是利用现有的SIMCOAL2（ <a class="reference external" href="http://cmpg.unibe.ch/software/simcoal2/">http://cmpg.unibe.ch/software/simcoal2/</a> ）。与其他相比，SIMCOAL2允许存在群体结构，多群体事件，多种类型的可发生重组的基因座（SNPs，序列，STRs/微卫星和RFLPs），具有多染色体的二倍体和测量偏倚（ascertainment bias）。注意，SIMCOAL2并不支持所有的选择模型。建议阅读上述链接中的SIMCOAL2帮助文档。</p>
<p>SIMCOAL2的输入是一个指定所需的群体和基因组的文件，输出是一系列文件（通常在1000个左右），它们包括每个亚群（subpopulation）中模拟的个体基因组。这些文件有多种途径，如计算某些统计数据（e.g. <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 或Tajima D）的置信区间以得到可信的范围。然后将真实的群体遗传数据统计结果与这些置信区间相比较。</p>
<p>Biopython溯祖模拟可以创建群体场景（demographic scenarios）和基因组，然后运行SIMCOAL2。</p>
<div class="section" id="scenario">
<h3>12.2.1  创建场景（scenario）</h3>
<p>创建场景包括创建群体及其染色体结构。多数情况下（如计算近似贝斯估计量（Approximate Bayesian Computations – ABC）），测试不同参数的变化很重要（如不同的有效群体大小 <span class="formula"><i>N</i><sub><i>e</i></sub></span>
 ， 从10，50，500到1000个体）。提供的代码可以很容易地模拟具有不同群体参数的场景。</p>
<p>下面我们将学习怎样创建场景，然后是怎样进行模拟。</p>
<div class="section" id="id142">
<h4>12.2.1.1  群体</h4>
<p>有一些内置的预定义群体，均包含两个共同的参数：同群种（deme）的样本大小（在模板中称为 sample_size ，其使用请见下文）和同群种大小，如亚群大小（pop_size）。所有的群体都可以作为模板，所有的参数也可以变化，每个模板都有自己的系统名称。这些预定义的群体/模板（template）包括：</p>
<dl class="docutils">
<dt><strong>Single population, constant size</strong></dt>
<dd>单一种群固定群体大小。标准的参数即可满足它，模板名称：simple.</dd>
<dt><strong>Single population, bottleneck</strong></dt>
<dd>单一群体瓶颈效应，如图 <a class="reference external" href="#fig:bottle">12.2.1.1</a> 所示。参数有当前种群大小（图中ne3模板的pop_size）、种群扩张时间 - 扩张发生后代数（expand_gen），瓶颈发生时的有效群体大小（ne2），种群收缩时间（contract_gen）以及原始种群大小（ne3）。模板名：bottle。</dd>
<dt><strong>Island model</strong></dt>
<dd>典型的岛屿模型。同群种（deme）总数表示为total_demes，迁移率表示为mig。模板名：island。</dd>
<dt><strong>Stepping stone model - 1 dimension</strong></dt>
<dd>一维脚踏石模型（Stepping stone model），极端状态下种群分布不连续。同群种（deme）总数表示为total_demes，迁移率表示为mig。模板名：ssm_1d。</dd>
<dt><strong>Stepping stone model - 2 dimensions</strong></dt>
<dd>二维脚踏石模型，极端状态下种群分布不连续。参数有表示水平维度的x和表示垂直维度的y（同群种总数即为x × y），以及表示迁移率的mig。模板名：ssm_2d。</dd>
</dl>
<p><img alt="image4" src="../images/bottle.png" /></p>
<p>在我们的第一个示例中，将生成一个单一种群固定群体大小（Single population, constant size）模板，样本大小（sample size）为30，同群种大小（deme size）为500。代码如下：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen.SimCoal.Template</span> <span class="keyword namespace">import</span> <span class="name">generate_simcoal_from_template</span>

<span class="name">generate_simcoal_from_template</span><span class="punctuation">(</span><span class="literal string">'simple'</span><span class="punctuation">,</span>
    <span class="punctuation">[(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="punctuation">[(</span><span class="literal string">'SNP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">24</span><span class="punctuation">,</span> <span class="literal number float">0.0005</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])])],</span>
    <span class="punctuation">[(</span><span class="literal string">'sample_size'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">30</span><span class="punctuation">]),</span>
    <span class="punctuation">(</span><span class="literal string">'pop_size'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">100</span><span class="punctuation">])])</span>
</pre>
<p>执行该段代码将会在当前目录生成一个名为simple_100_300.par的文件，该文件可作为SIMCOAL2的输入文件，用于模拟群体（下面将会展示Biopython是如何调用SIMCOAL2）。</p>
<p>这段代码仅包含一个函数的调用，让我们一个参数一个参数地讨论。</p>
<p>第一个参数是模板id（从上面的模板列表中选择）。我们使用 ’simple’，表示的是单一群体固定种群大小模板。</p>
<p>第二个参数是染色体结构，将在下一节详细阐述。</p>
<p>第三个参数是所有需要的参数列表及其所有可能的值（此列中所有的参数都只含有一个可能值）。</p>
<p>现在让我们看看岛屿模型示例。我们希望生成几个岛屿模型，并对不同大小的同群种感兴趣：10、50和100，迁移率为1%。样本大小和同群种大小与上一个示例一致，代码如下：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen.SimCoal.Template</span> <span class="keyword namespace">import</span> <span class="name">generate_simcoal_from_template</span>

<span class="name">generate_simcoal_from_template</span><span class="punctuation">(</span><span class="literal string">'island'</span><span class="punctuation">,</span>
    <span class="punctuation">[(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="punctuation">[(</span><span class="literal string">'SNP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">24</span><span class="punctuation">,</span> <span class="literal number float">0.0005</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])])],</span>
    <span class="punctuation">[(</span><span class="literal string">'sample_size'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">30</span><span class="punctuation">]),</span>
    <span class="punctuation">(</span><span class="literal string">'pop_size'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">100</span><span class="punctuation">]),</span>
    <span class="punctuation">(</span><span class="literal string">'mig'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number float">0.01</span><span class="punctuation">]),</span>
    <span class="punctuation">(</span><span class="literal string">'total_demes'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">10</span><span class="punctuation">,</span> <span class="literal number integer">50</span><span class="punctuation">,</span> <span class="literal number integer">100</span><span class="punctuation">])])</span>
</pre>
<p>此例将会生成3个文件：island_100_0.01_100_30.par，island_10_0.01_100_30.par 和 island_50_0.01_100_30.par。注意，生成文件名的规律是：模板名，然后是参数值逆序排列。</p>
<p>还有一些存在较多争议的群体模板（请见Biopython源代码中Bio/PopGen/SimCoal/data文件夹）。同时，用户可以创建新的模板，该功能将在以后的文档中讨论。</p>
</div>
<div class="section" id="id143">
<h4>12.2.1.2  染色体结构</h4>
<p>我们强烈建议你阅读SIMCOAL2文档，以完整理解染色体结构建模的各种使用。在本小节，我们只讨论如何使用Biopython接口实现指定的染色体结构，不会涉及SIMCOAL2可实现哪些染色体结构。</p>
<p>我们首先实现一条染色体，包含24个SNPs，每个相邻基因座的重组率为0.0005，次等位基因的最小频率为0。这些由以下列表指定（作为第二个参数传递给generate_simcoal_from_template函数）：</p>
<pre class="code python literal-block">
<span class="punctuation">[(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="punctuation">[(</span><span class="literal string">'SNP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">24</span><span class="punctuation">,</span> <span class="literal number float">0.0005</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])])]</span>
</pre>
<p>这实际上是上一个示例使用的染色体结构。</p>
<p>染色体结构表示为一个包含所有染色体的列表，每条染色体（即列表中的每个元素）由一个元组（tuple）组成，元组包括一对元素组成。元组的第一个元素是染色体被重复的次数（因为有可能需要多次重复同一条染色体）。元组的第二个元素是一个表示该染色体的实际组成的列表，每个列表元素又包括一对元素，第一个是基因座类型，第二个是该基因座的参数列表。是否有点混淆了呢？在我们展示示例之前，先让我们回顾下上一个示例：我们有一个列表（表示一条染色体），该染色体只有一个实例（因此不会被重复），它由24个SNPs组成，每个相邻SNP间的重组率为0.0005，次等位基因的最小频率为0.0（即它可以在某些染色体中缺失）。</p>
<p>现在让我们看看更复杂的示例：</p>
<pre class="code python literal-block">
<span class="punctuation">[</span>
  <span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="punctuation">[</span>
       <span class="punctuation">(</span><span class="literal string">'SNP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">24</span><span class="punctuation">,</span> <span class="literal number float">0.0005</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])</span>
      <span class="punctuation">]</span>
  <span class="punctuation">),</span>
  <span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="punctuation">[</span>
       <span class="punctuation">(</span><span class="literal string">'DNA'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">10</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.00005</span><span class="punctuation">,</span> <span class="literal number float">0.33</span><span class="punctuation">]),</span>
       <span class="punctuation">(</span><span class="literal string">'RFLP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.0001</span><span class="punctuation">]),</span>
       <span class="punctuation">(</span><span class="literal string">'MICROSAT'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.001</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])</span>
      <span class="punctuation">]</span>
  <span class="punctuation">)</span>
<span class="punctuation">]</span>
</pre>
<p>首先，我们有5条与上一示例具有相同结构组成的染色体（即24SNPs）。然后是2条这样的染色体：包含一段具有重组率为0.0、突变率为0.0005及置换率为0.33的10个核苷酸长度的DNA序列，一段具有重组率为0.0、突变率为0.0001的RFLP，一段具有重组率为0.0、突变率为0.001、几何参数为0.0、范围限制参数为0.0的微卫星（microsatellite，STR）序列（注意，因为这是单个微卫星，接下来没有基因座，因此这里的重组率没有任何影响，更多关于这些参数的信息请查阅SIMCOAL2文档，你可以使用它们模拟各种突变模型，包括典型的微卫星渐变突变模型）。</p>
</div>
</div>
<div class="section" id="simcoal2">
<h3>12.2.2  运行SIMCOAL2</h3>
<p>现在我们讨论如何从Biopython内部运行SIMCOAL2。这需要SIMCOAL2的可执行二进制文件名为simcoal2（在Windows平台下为simcoal2.exe），请注意，从官网下载的程序命名格式通常为simcoal2_x_y。因此，当安装SIMCOAL2时，需要重命名可执行文件，这样Biopython才能正确调用。</p>
<p>SIMCOAL2可以处理不是使用上诉方法生成的文件（如手动配置的参数文件），但是我们将使用上述方法得到的文件创建模型：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen.SimCoal.Template</span> <span class="keyword namespace">import</span> <span class="name">generate_simcoal_from_template</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen.SimCoal.Controller</span> <span class="keyword namespace">import</span> <span class="name">SimCoalController</span>


<span class="name">generate_simcoal_from_template</span><span class="punctuation">(</span><span class="literal string">'simple'</span><span class="punctuation">,</span>
    <span class="punctuation">[</span>
      <span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="punctuation">[</span>
           <span class="punctuation">(</span><span class="literal string">'SNP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">24</span><span class="punctuation">,</span> <span class="literal number float">0.0005</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])</span>
          <span class="punctuation">]</span>
      <span class="punctuation">),</span>
      <span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="punctuation">[</span>
           <span class="punctuation">(</span><span class="literal string">'DNA'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">10</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.00005</span><span class="punctuation">,</span> <span class="literal number float">0.33</span><span class="punctuation">]),</span>
           <span class="punctuation">(</span><span class="literal string">'RFLP'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.0001</span><span class="punctuation">]),</span>
           <span class="punctuation">(</span><span class="literal string">'MICROSAT'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.001</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">])</span>
          <span class="punctuation">]</span>
      <span class="punctuation">)</span>
    <span class="punctuation">],</span>
    <span class="punctuation">[(</span><span class="literal string">'sample_size'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">30</span><span class="punctuation">]),</span>
    <span class="punctuation">(</span><span class="literal string">'pop_size'</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">100</span><span class="punctuation">])])</span>

<span class="name">ctrl</span> <span class="operator">=</span> <span class="name">SimCoalController</span><span class="punctuation">(</span><span class="literal string">'.'</span><span class="punctuation">)</span>
<span class="name">ctrl</span><span class="operator">.</span><span class="name">run_simcoal</span><span class="punctuation">(</span><span class="literal string">'simple_100_30.par'</span><span class="punctuation">,</span> <span class="literal number integer">50</span><span class="punctuation">)</span>
</pre>
<p>需要注意的是最后两行（以及新增的import行）。首先是创建一个应用程序控制器对象，需要指定二进制可执行文件所在路径。</p>
<p>模拟器在最后一行运行：从上述阐述的规律可知，文件名为simple_100_30.par的输入文件是我们创建的模拟参数文件，然后我们指定了希望运行50次独立模拟。默认情况下，Biopython模拟二倍体数据，但是可以添加第三个参数用于模拟单倍体数据（字符串'0'）。然后，SIMCOAL2将会执行（这需要运行很长时间），并创建一个包含模拟结果的文件夹，结果文件可便可用于分析（尤其是研究Arlequin3数据）。在未来的Biopython版本中，可能会支持Arlequin3格式文件的读取，从而在Biopython中便能分析SIMCOAL2结果。</p>
</div>
</div>
<div class="section" id="id144">
<h2>12.3  其它应用程序</h2>
<p>这里我们讨论一些处理其它的群体遗传学中应用程序的接口和小工具，这些应用程序具有争议，使用得较少。</p>
<div class="section" id="fdist">
<h3>12.3.1  FDist：检测选择压力和分子适应</h3>
<p>FDist是一个选择压力检测的应用程序包，基于通过 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 和杂合度计算（即模拟）得到的“中性”（“neutral”）置信区间。“中性”置信区间外的Markers（可以是SNPs，微卫星，AFLPs等等）可以被认为是候选的受选择marker。</p>
<p>FDist主要运用在当marker数量足够用于估计平均 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 ，而不足以从数据集中计算出离群点 - 直接地或者在知道大多数marker在基因组中的相对位置的情况下使用基于如Extended Haplotype Heterozygosity （EHH）的方法。</p>
<p>典型的FDist的使用如下：</p>
<ol class="arabic simple">
<li>从其它格式读取数据为FDist格式；</li>
<li>计算平均 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 ，由FDist的datacal完成；</li>
<li>根据平均 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 和期望的总群体数模拟“中性”markers，这是核心部分，由FDist的fdist完成；</li>
<li>根据指定的置信范围（通常是95%或者是99%）计算置信区间，由cplot完成，主要用于对区间作图；</li>
<li>用模拟的“中性”置信区间评估每个Marker的状态，由pv完成，用于检测每个marker与模拟的相比的离群状态；</li>
</ol>
<p>我们将以示例代码讨论每一步（FDist可执行二进制文件需要在PATH环境变量中）。</p>
<p>FDist数据格式是该应用程序特有的，不被其它应用程序使用。因此你需要转化你的数据格式到FDist可使用的格式。Biopython可以帮助你完成这个过程。这里有一个将GenePop格式转换为FDist格式的示例（同时包括后面示例将用到的import语句）：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen</span> <span class="keyword namespace">import</span> <span class="name">GenePop</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen</span> <span class="keyword namespace">import</span> <span class="name">FDist</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen.FDist</span> <span class="keyword namespace">import</span> <span class="name">Controller</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.PopGen.FDist.Utils</span> <span class="keyword namespace">import</span> <span class="name">convert_genepop_to_fdist</span>

<span class="name">gp_rec</span> <span class="operator">=</span> <span class="name">GenePop</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;example.gen&quot;</span><span class="punctuation">))</span>
<span class="name">fd_rec</span> <span class="operator">=</span> <span class="name">convert_genepop_to_fdist</span><span class="punctuation">(</span><span class="name">gp_rec</span><span class="punctuation">)</span>
<span class="name">in_file</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;infile&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="name">in_file</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name builtin">str</span><span class="punctuation">(</span><span class="name">fd_rec</span><span class="punctuation">))</span>
<span class="name">in_file</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>在该段代码中，我们解析GenePop文件并转化为FDist记录（record）。</p>
<p>输出FDist记录将得到可以直接保存到可用于FDist的文件的字符串。FDist需要输入文件名为infile，因此我们将记录保存到文件名为infile的文件。</p>
<p>FDist记录最重要的字段（field）是：num_pops，群体数量；num_loci，基因座数量和loci_data，marker数据。记录的许多信息对用户来说可能没有用处，仅用于传递给FDist。</p>
<p>下一步是计算平均数据集的 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 （以及样本大小）：</p>
<pre class="code python literal-block">
<span class="name">ctrl</span> <span class="operator">=</span> <span class="name">Controller</span><span class="operator">.</span><span class="name">FDistController</span><span class="punctuation">()</span>
<span class="name">fst</span><span class="punctuation">,</span> <span class="name">samp_size</span> <span class="operator">=</span> <span class="name">ctrl</span><span class="operator">.</span><span class="name">run_datacal</span><span class="punctuation">()</span>
</pre>
<p>第一行我们创建了一个控制调用FDist软件包的对象，该对象被用于调用该包的其它应用程序。</p>
<p>第二行我们调用datacal应用程序，它用于计算 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 和样本大小。值得注意的是，用datacal计算得到的 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 是 Weir-Cockerham θ的“变种”（ <em>variation</em> ）。</p>
<p>现在我们可以调用主程序fdist模拟中性Markers。</p>
<pre class="code python literal-block">
<span class="name">sim_fst</span> <span class="operator">=</span> <span class="name">ctrl</span><span class="operator">.</span><span class="name">run_fdist</span><span class="punctuation">(</span><span class="name">npops</span> <span class="operator">=</span> <span class="literal number integer">15</span><span class="punctuation">,</span> <span class="name">nsamples</span> <span class="operator">=</span> <span class="name">fd_rec</span><span class="operator">.</span><span class="name">num_pops</span><span class="punctuation">,</span> <span class="name">fst</span> <span class="operator">=</span> <span class="name">fst</span><span class="punctuation">,</span>
    <span class="name">sample_size</span> <span class="operator">=</span> <span class="name">samp_size</span><span class="punctuation">,</span> <span class="name">mut</span> <span class="operator">=</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">num_sims</span> <span class="operator">=</span> <span class="literal number integer">40000</span><span class="punctuation">)</span>
</pre>
<dl class="docutils">
<dt><strong>npops</strong></dt>
<dd>现存自然群体数量，完全是一个“瞎猜值”（“guestimate”），必须小于100。</dd>
<dt><strong>nsamples</strong></dt>
<dd>抽样群体数量，需要小于npops。</dd>
<dt><strong>fst</strong></dt>
<dd>平均 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 。</dd>
<dt><strong>sample_size</strong></dt>
<dd>每个群体抽样个体平均数</dd>
<dt><strong>mut</strong></dt>
<dd>突变模型：0 - 无限等位基因突变模型；1 - 渐变突变模型</dd>
<dt><strong>num_sims</strong></dt>
<dd>执行模拟的次数。通常，40000左右的数值即可，但是如果得到的执行区间范围比较大（可以通过下面的置信区间作图检测到），可以上调此值（建议每次调整10000次模拟）。</dd>
</dl>
<p>样本数量和样本大小措辞上的混乱源于原始的应用程序。</p>
<p>将会得到一个名为out.dat的文件，它包含模拟的杂合度和 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 值，行数与模拟的次数相同。</p>
<p>注意，fdist返回它可以模拟的平均 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 ，关于此问题更多的细节，请阅读下面的“估计期望的平均 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 ”</p>
<p>下一步（可选步骤）是计算置信区间：</p>
<pre class="code python literal-block">
<span class="name">cpl_interval</span> <span class="operator">=</span> <span class="name">ctrl</span><span class="operator">.</span><span class="name">run_cplot</span><span class="punctuation">(</span><span class="name">ci</span><span class="operator">=</span><span class="literal number float">0.99</span><span class="punctuation">)</span>
</pre>
<p>只能在运行fdist之后才能调用cplot。</p>
<p>这将计算先前fdist结果的置信区间（此例中为99%）。第一个元素是杂合度，第二个是该杂合度的 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 置信下限，第三个是 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 平均值，第四个是置信上限。可以用于记录置信区间等高线。该列表也可以输出到out.cpl文件。</p>
<p>这步的主要目的是返回一系列的点用于对置信区间作图。如果只是需要根据模拟结果对每个marker的状态进行评估，可以跳过此步。</p>
<pre class="code python literal-block">
<span class="name">pv_data</span> <span class="operator">=</span> <span class="name">ctrl</span><span class="operator">.</span><span class="name">run_pv</span><span class="punctuation">()</span>
</pre>
<p>只能在运行fdist之后才能调用pv。</p>
<p>这将使用模拟marker对每个个体真实的marker进行评估，并返回一个列表，顺序与FDist记录中loci_list一致（loci_list又与GenePop顺序一致）。列表中每个元素包含四个元素，其中最重要的是最后一个元素（关于其他的元素，为了简单起见，我们不在这里讨论，请见pv帮助文档），它返回模拟的 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 低于 marker <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 的概率。较大值说明极有可能是正选择（positive selection）marker，较小值表明可能是平衡选择（balancing selection）marker，中间值则可能是中性marker。怎样的值是“较大值”、“较小值”或者“中间值”是一个很主观的问题，但当使用置信区间方法及95%的置信区间时，“较大值”在0.95 - 1.00之间，“较小值”在0.00 - 0.05之间，“中间值”在0.05 - 0.05之间。</p>
<div class="section" id="f-st">
<h4>12.3.1.1  估计期望的平均 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
</h4>
<p>FDist通过对由下例公式得到的迁移率进行溯祖模拟估计期望的平均 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 ：</p>
<div class="formula">
<i>N</i><sub><i>m</i></sub> = <span class="fraction"><span class="ignored">(</span><span class="numerator">1 − <i>F</i><sub><i>st</i></sub></span><span class="ignored">)/(</span><span class="denominator">4<i>F</i><sub><i>st</i></sub></span><span class="ignored">)</span></span>
</div>
<p>该公式有一些前提，比如种群大小无限大。</p>
<p>在实践中，当群体数量比较小，突变模型为渐进突变模型，样本大小增加，fdist将不能模拟得到可接受的近似平均 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 。</p>
<p>为了解决这个问题，Biopython提供了一个使用迭代方法的函数，通过依次运行几个fdist得到期望的值。该方法比运行单个fdist相比耗费更多计算资源，但是可以得到更好的结果。以下代码运行fdist得到期望的 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 ：</p>
<pre class="code python literal-block">
<span class="name">sim_fst</span> <span class="operator">=</span> <span class="name">ctrl</span><span class="operator">.</span><span class="name">run_fdist_force_fst</span><span class="punctuation">(</span><span class="name">npops</span> <span class="operator">=</span> <span class="literal number integer">15</span><span class="punctuation">,</span> <span class="name">nsamples</span> <span class="operator">=</span> <span class="name">fd_rec</span><span class="operator">.</span><span class="name">num_pops</span><span class="punctuation">,</span>
    <span class="name">fst</span> <span class="operator">=</span> <span class="name">fst</span><span class="punctuation">,</span> <span class="name">sample_size</span> <span class="operator">=</span> <span class="name">samp_size</span><span class="punctuation">,</span> <span class="name">mut</span> <span class="operator">=</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">num_sims</span> <span class="operator">=</span> <span class="literal number integer">40000</span><span class="punctuation">,</span>
    <span class="name">limit</span> <span class="operator">=</span> <span class="literal number float">0.05</span><span class="punctuation">)</span>
</pre>
<p>与run_fdist相比，唯一一个新的可选参数是limit，它表示期望的最大错误率。run_fdist可以（或许应该）由run_fdist_force_fst替代。</p>
</div>
<div class="section" id="id145">
<h4>12.3.1.2  说明</h4>
<p>计算平均 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 的过程可能比这里呈现的要复杂得多。更多的信息请查阅FDist README文件。同时，Biopython的代码也可用于实现更复杂的过程。</p>
</div>
</div>
</div>
<div class="section" id="id146">
<h2>12.4  未来发展</h2>
<p>最期望的是您的参与！</p>
<p>尽管如此，已经完成的功能模块正在逐步加入到Bio.PopGen，这些代码覆盖了程序FDist和SimCoal2，HapMap和UCSC Table Browser数据库，以及一些简单的统计计算，如 <span class="formula"><i>F</i><sub><i>st</i></sub></span>
 ， 或等位基因数。</p>
</div>
</div>
<div class="section" id="bio-phylo">
<h1>第13章  Bio.Phylo系统发育分析</h1>
<p>Biopython1.54开始引入了Bio.Phylo模块，与SeqIO和AlignIO类似，它的目的是提供
一个通用的独立于源数据格式的方法来使用系统进化树，同时提供一致的API来进行
I/O操作。</p>
<p>Bio.Phylo在一篇开放获取的期刊文章中有介绍
[<a class="reference external" href="#talevich2012">9</a>, Talevich <em>et al.</em>, 2012], 这可能对您也有所帮助。</p>
<div class="section" id="id147">
<h2>13.1  示例: 树中有什么？</h2>
<p>为了熟悉这个模块，让我们首先从一个已经创建好的树开始，从几个不同的角度来审视
它。接着我们将给树的分支上颜色，并使用一个特殊的phyloXML特性，最终保存树。</p>
<p>译者注：本翻译中， <strong>分支</strong> 对应原文中的 <strong>branch</strong> ，原文中一般代表
某一个节点的前导连线；而 <strong>进化枝</strong> 对应原文中的 <strong>clade</strong> ，代表某个节点所代表的整个
进化分支，包括本身和它所有的后代；若clade代表biopython中的对象则保留原文</p>
<p>在终端中使用你喜欢的编辑器创建一个简单的Newick文件：</p>
<pre class="code python literal-block">
<span class="operator">%</span> <span class="name">cat</span> <span class="operator">&gt;</span> <span class="name">simple</span><span class="operator">.</span><span class="name">dnd</span> <span class="operator">&lt;&lt;</span><span class="name">EOF</span>
<span class="operator">&gt;</span> <span class="punctuation">(((</span><span class="name">A</span><span class="punctuation">,</span><span class="name">B</span><span class="punctuation">),(</span><span class="name">C</span><span class="punctuation">,</span><span class="name">D</span><span class="punctuation">)),(</span><span class="name">E</span><span class="punctuation">,</span><span class="name">F</span><span class="punctuation">,</span><span class="name">G</span><span class="punctuation">));</span>
<span class="operator">&gt;</span> <span class="name">EOF</span>
</pre>
<p>这棵树没有分支长度，只有一个拓扑结构和标记的端点。（如果你有一个真正的树文件，
你也可以使用它来替代进行示例操作。）</p>
<p>选择启动你的Python解释器：</p>
<pre class="code python literal-block">
<span class="operator">%</span> <span class="name">ipython</span> <span class="operator">-</span><span class="name">pylab</span>
</pre>
<p>对于交互式操作，使用参数 <tt class="docutils literal"><span class="pre">-pylab</span></tt> 启动IPython解释器能启用 <strong>matplotlib</strong> 整合
功能，这样图像就能自动弹出来。我们将在这个示例中使用该功能。</p>
<p>现在，在Python终端中，读取树文件，给定文件名和格式名。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Phylo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;simple.dnd&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;newick&quot;</span><span class="punctuation">)</span>
</pre>
<p>以字符串打印该树对象我们将得到整个对象的层次结构概况。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">tree</span>

<span class="name">Tree</span><span class="punctuation">(</span><span class="name">weight</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">rooted</span><span class="operator">=</span><span class="name builtin pseudo">False</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;&quot;</span><span class="punctuation">)</span>
    <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">)</span>
        <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">)</span>
                <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;A&quot;</span><span class="punctuation">)</span>
                <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;B&quot;</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">)</span>
                <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;C&quot;</span><span class="punctuation">)</span>
                <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;D&quot;</span><span class="punctuation">)</span>
        <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;E&quot;</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;F&quot;</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;G&quot;</span><span class="punctuation">)</span>
</pre>
<p><tt class="docutils literal">Tree</tt> 对象包含树的全局信息，如树是有根树还是无根树。它包含一个根进化枝，
和以此往下以列表嵌套的所有进化枝，直至叶子分支。</p>
<p>函数 <tt class="docutils literal">draw_ascii</tt> 创建一个简单的ASCII-art(纯文本)系统发生图。在没有更好
图形工具的情况下，这对于交互研究来说是一个方便的可视化展示方式。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_ascii</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
                                                    <span class="name">________________________</span> <span class="name">A</span>
                           <span class="name">________________________</span><span class="operator">|</span>
                          <span class="operator">|</span>                        <span class="operator">|</span><span class="name">________________________</span> <span class="name">B</span>
  <span class="name">________________________</span><span class="operator">|</span>
 <span class="operator">|</span>                        <span class="operator">|</span>                         <span class="name">________________________</span> <span class="name">C</span>
 <span class="operator">|</span>                        <span class="operator">|</span><span class="name">________________________</span><span class="operator">|</span>
<span class="name">_</span><span class="operator">|</span>                                                 <span class="operator">|</span><span class="name">________________________</span> <span class="name">D</span>
 <span class="operator">|</span>
 <span class="operator">|</span>                         <span class="name">________________________</span> <span class="name">E</span>
 <span class="operator">|</span>                        <span class="operator">|</span>
 <span class="operator">|</span><span class="name">________________________</span><span class="operator">|</span><span class="name">________________________</span> <span class="name">F</span>
                          <span class="operator">|</span>
                          <span class="operator">|</span><span class="name">________________________</span> <span class="name">G</span>
</pre>
<p>如果你安装有 <strong>matplotlib</strong> 或者 <strong>pylab</strong>, 你可以使用 <tt class="docutils literal">draw</tt> 函数一个图像(见 Fig.
<a class="reference external" href="#fig:phylo-simple-draw">13.1</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">rooted</span> <span class="operator">=</span> <span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
</pre>
<p><img alt="image5" src="../images/phylo-simple-draw.png" /></p>
<div class="section" id="id148">
<h3>13.1.1  给树的分支上颜色</h3>
<p>函数 <tt class="docutils literal">draw</tt> 和 <tt class="docutils literal">draw_graphviz</tt> 支持在树中显示不同的颜色和分支宽度。
从Biopython 1.59开始，Clade对象就开始支持 <tt class="docutils literal">color</tt> 和 <tt class="docutils literal">width</tt> 属性，
且使用他们不需要额外支持。这两个属性都表示导向给定的进化枝前面的分支的
属性，并依次往下作用，所以所有的后代分支在显示时也都继承相同的宽度和颜
色。</p>
<p>在早期的Biopython版本中，PhyloXML树有些特殊的特性，使用这些属性需要首先
将这个树转换为一个基本树对象的子类Phylogeny，该类在Bio.Phylo.PhyloXML模
块中。</p>
<p>在Biopython 1.55和之后的版本中，这是一个很方便的树方法：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">tree</span><span class="operator">.</span><span class="name">as_phyloxml</span><span class="punctuation">()</span>
</pre>
<p>在Biopython 1.54中, 你能通过导入一个额外的模块实现相同的事情：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Phylo.PhyloXML</span> <span class="keyword namespace">import</span> <span class="name">Phylogeny</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylogeny</span><span class="operator">.</span><span class="name">from_tree</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
</pre>
<p>注意Newick和Nexus文件类型并不支持分支颜色和宽度，如果你在Bio.Phylo中使用
这些属性，你只能保存这些值到PhyloXML格式中。（你也可以保存成Newick或Nexus
格式，但是颜色和宽度信息在输出的文件时会被忽略掉。）</p>
<p>现在我们开始指定颜色。首先，我们将设置根进化枝为灰色。我们能通过赋值24位
的颜色值来实现，用三位数的RGB值、HTML格式的十六进制字符串、或者预先设置好的
颜色名称。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">root</span><span class="operator">.</span><span class="name">color</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="literal number integer">128</span><span class="punctuation">,</span> <span class="literal number integer">128</span><span class="punctuation">,</span> <span class="literal number integer">128</span><span class="punctuation">)</span>
</pre>
<p>Or:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">root</span><span class="operator">.</span><span class="name">color</span> <span class="operator">=</span> <span class="literal string">&quot;#808080&quot;</span>
</pre>
<p>Or:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">root</span><span class="operator">.</span><span class="name">color</span> <span class="operator">=</span> <span class="literal string">&quot;gray&quot;</span>
</pre>
<p>一个进化枝的颜色会被当作从上而下整个进化枝的颜色，所以我们这里设置根的
的颜色会将整个树的颜色变为灰色。我们能通过在树中下面分支赋值不同的颜色
来重新定义某个分支的颜色。</p>
<p>让我们先定位“E”和“F”最近祖先（MRCA）节点。方法 <tt class="docutils literal">common_ancestor</tt> 返回
原始树中这个进化枝的引用，所以当我们设置该进化枝为“salmon”颜色时，这个颜
色则会在原始的树中显示出来。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mrca</span> <span class="operator">=</span> <span class="name">tree</span><span class="operator">.</span><span class="name">common_ancestor</span><span class="punctuation">({</span><span class="literal string">&quot;name&quot;</span><span class="punctuation">:</span> <span class="literal string">&quot;E&quot;</span><span class="punctuation">},</span> <span class="punctuation">{</span><span class="literal string">&quot;name&quot;</span><span class="punctuation">:</span> <span class="literal string">&quot;F&quot;</span><span class="punctuation">})</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mrca</span><span class="operator">.</span><span class="name">color</span> <span class="operator">=</span> <span class="literal string">&quot;salmon&quot;</span>
</pre>
<p>当我们碰巧明确地知道某个进化枝在树中的位置，以嵌套列表的形式，我们就能
通过索引的方式直接跳到那个位置。这里，索引 <tt class="docutils literal">[0,1]</tt> 表示根节点的第一个
子代节点的第二个子代。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">clade</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="literal number integer">1</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">color</span> <span class="operator">=</span> <span class="literal string">&quot;blue&quot;</span>
</pre>
<p>最后，展示一下我们的工作结果 (see Fig. <a class="reference external" href="#fig:phylo-color-draw">13.1.1</a>):</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
</pre>
<p><img alt="image6" src="../images/phylo-color-draw.png" /></p>
<p>注意进化枝的颜色包括导向它的分支和它的子代的分支。E和F的共同祖先结果刚好
在根分支下面，而通过这样上色，我们能清楚的看出这个树的根在哪里。</p>
<p>我们已经完成了很多！现在让我们休息一下，保存一下我们的工作。使用一个文件
名或句柄（这里我们使用标准输出来查看将会输出什么）和 <tt class="docutils literal">phyloxml</tt> 格式来
调用 <tt class="docutils literal">write</tt> 函数。PhyloXML格式保存了我们设置的颜色，所以你能通过其他树
查看工具，如Archaeopteryx，打开这个phyloXML文件，这些颜色也会显示出来。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">sys</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">,</span> <span class="name">sys</span><span class="operator">.</span><span class="name">stdout</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>

<span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">phyloxml</span> <span class="name">xmlns</span><span class="punctuation">:</span><span class="name">phy</span><span class="operator">=</span><span class="literal string">&quot;http://www.phyloxml.org&quot;</span><span class="operator">&gt;</span>
  <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">phylogeny</span> <span class="name">rooted</span><span class="operator">=</span><span class="literal string">&quot;true&quot;</span><span class="operator">&gt;</span>
    <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">clade</span><span class="operator">&gt;</span>
      <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">branch_length</span><span class="operator">&gt;</span><span class="literal number float">1.0</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">branch_length</span><span class="operator">&gt;</span>
      <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">color</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">red</span><span class="operator">&gt;</span><span class="literal number integer">128</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">red</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">green</span><span class="operator">&gt;</span><span class="literal number integer">128</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">green</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">blue</span><span class="operator">&gt;</span><span class="literal number integer">128</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">blue</span><span class="operator">&gt;</span>
      <span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">color</span><span class="operator">&gt;</span>
      <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">clade</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">branch_length</span><span class="operator">&gt;</span><span class="literal number float">1.0</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">branch_length</span><span class="operator">&gt;</span>
        <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">clade</span><span class="operator">&gt;</span>
          <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">branch_length</span><span class="operator">&gt;</span><span class="literal number float">1.0</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">branch_length</span><span class="operator">&gt;</span>
          <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">clade</span><span class="operator">&gt;</span>
            <span class="operator">&lt;</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">name</span><span class="operator">&gt;</span><span class="name">A</span><span class="operator">&lt;/</span><span class="name">phy</span><span class="punctuation">:</span><span class="name">name</span><span class="operator">&gt;</span>
            <span class="operator">...</span>
</pre>
<p>本章的其余部分将更加细致的介绍Bio.Phylo核心功能。关于Bio.Phylo的更多例
子，请参见Biopython.org上的Cookbook手册页面。</p>
<p><a class="reference external" href="http://biopython.org/wiki/Phylo_cookbook">http://biopython.org/wiki/Phylo_cookbook</a></p>
</div>
</div>
<div class="section" id="i-o">
<h2>13.2  I/O 函数</h2>
<p>和SeqIO、AlignIO类似, Phylo使用四个函数处理文件的输入输出： <tt class="docutils literal">parse</tt> 、
<tt class="docutils literal">read</tt> 、 <tt class="docutils literal">write</tt> 和 <tt class="docutils literal">convert</tt> ，所有的函数都支持Newick、NEXUS、
phyloXML和NeXML等树文件格式。</p>
<p><tt class="docutils literal">read</tt> 函数解析并返回给定文件中的单个树。注意，如果文件中包含多个或不包含任何树，它将抛出一个错误。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Phylo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;Tests/Nexus/int_node_labels.nwk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;newick&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">tree</span>
</pre>
<p>（Biopython发布包的 <tt class="docutils literal">Tests/Nexus/</tt> 和 <tt class="docutils literal">Tests/PhyloXML/</tt> 文件夹中有相应的例子）</p>
<p>处理多个（或者未知个数）的树文件，需要使用 <tt class="docutils literal">parse</tt> 函数迭代给定文件中的每一个树。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">trees</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;Tests/PhyloXML/phyloxml_examples.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">tree</span> <span class="operator word">in</span> <span class="name">trees</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">tree</span>
</pre>
<p>使用 <tt class="docutils literal">write</tt> 函数输出一个或多个可迭代的树。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">trees</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">Phylo</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;phyloxml_examples.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree1</span> <span class="operator">=</span> <span class="name">trees</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">others</span> <span class="operator">=</span> <span class="name">trees</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">:]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">tree1</span><span class="punctuation">,</span> <span class="literal string">&quot;tree1.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">others</span><span class="punctuation">,</span> <span class="literal string">&quot;other_trees.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">12</span>
</pre>
<p>使用 <tt class="docutils literal">convert</tt> 函数转换任何支持的树格式。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;tree1.dnd&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;newick&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;tree1.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;nexml&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;other_trees.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;other_trees.nex&quot;</span><span class="punctuation">,</span> <span class="literal string">'nexus&quot;)</span>
<span class="literal number integer">12</span>
</pre>
<p>和SeqIO和AlignIO类似，当使用字符串而不是文件作为输入输出时，需要使用 ‵‵StringIO`` 函数。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Phylo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">StringIO</span> <span class="keyword namespace">import</span> <span class="name">StringIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">StringIO</span><span class="punctuation">(</span><span class="literal string">&quot;(((A,B),(C,D)),(E,F,G));&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;newick&quot;</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="id149">
<h2>13.3  查看和导出树</h2>
<p>了解一个 <tt class="docutils literal">Tree</tt> 对象概况的最简单的方法是用 <tt class="docutils literal">print</tt> 函数将它打印出来：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;Tests/PhyloXML/example.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">tree</span>
<span class="name">Phylogeny</span><span class="punctuation">(</span><span class="name">rooted</span><span class="operator">=</span><span class="literal string">'True'</span><span class="punctuation">,</span> <span class="name">description</span><span class="operator">=</span><span class="literal string">'phyloXML allows to use either a &quot;branch_length&quot;</span>
<span class="name">attribute</span><span class="operator">...</span><span class="literal string">', name='</span><span class="name">example</span> <span class="keyword namespace">from</span> <span class="name namespace">Prof.</span> <span class="name namespace">Joe</span> <span class="name namespace">Felsenstein</span><span class="literal string">'s book &quot;Inferring Phyl...'</span><span class="punctuation">)</span>
    <span class="name">Clade</span><span class="punctuation">()</span>
        <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal string">'0.06'</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal string">'0.102'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'A'</span><span class="punctuation">)</span>
            <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal string">'0.23'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'B'</span><span class="punctuation">)</span>
        <span class="name">Clade</span><span class="punctuation">(</span><span class="name">branch_length</span><span class="operator">=</span><span class="literal string">'0.4'</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">'C'</span><span class="punctuation">)</span>
</pre>
<p>上面实际上是Biopython的树对象层次结构的一个概况。然而更可能的情况是，你希望见到
画出树的形状，这里有三个函数来做这件事情。</p>
<p>如我们在demo中看到的一样， <tt class="docutils literal">draw_ascii</tt> 打印一个树的ascii-art图像（有根进化树）
到标准输出，或者一个打开的文件句柄，若有提供。不是所有关于树的信息被显示出来，但是它提供了一个
不依靠于任何外部依赖的快速查看树的方法。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;example.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_ascii</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
             <span class="name">__________________</span> <span class="name">A</span>
  <span class="name">__________</span><span class="operator">|</span>
<span class="name">_</span><span class="operator">|</span>          <span class="operator">|</span><span class="name">___________________________________________</span> <span class="name">B</span>
 <span class="operator">|</span>
 <span class="operator">|</span><span class="name">___________________________________________________________________________</span> <span class="name">C</span>
</pre>
<p><tt class="docutils literal">draw</tt> 函数则使用matplotlib类库画出一个更加好看的图像。查看API文档以获得关于它所接受的
用来定制输出的参数。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;example.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">,</span> <span class="name">branch_labels</span><span class="operator">=</span><span class="keyword">lambda</span> <span class="name">c</span><span class="punctuation">:</span> <span class="name">c</span><span class="operator">.</span><span class="name">branch_length</span><span class="punctuation">)</span>
</pre>
<p><img alt="image7" src="../images/phylo-draw-example.png" /></p>
<p><tt class="docutils literal">draw_graphviz</tt> 则画出一个无根的进化分枝图（cladogram），但是它要求你安装有Graphviz、
PyDot或PyGraphviz、Network和matplotlib（或pylab）。使用上面相同的例子，和Graphviz中的
<tt class="docutils literal">dot</tt> 程序，让我们来画一个有根树（见图. <a class="reference external" href="#fig:phylo-dot">13.3</a> ）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;example.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_graphviz</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">,</span> <span class="name">prog</span><span class="operator">=</span><span class="literal string">'dot'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">pylab</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pylab</span><span class="operator">.</span><span class="name">show</span><span class="punctuation">()</span>                    <span class="comment"># Displays the tree in an interactive viewer</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pylab</span><span class="operator">.</span><span class="name">savefig</span><span class="punctuation">(</span><span class="literal string">'phylo-dot.png'</span><span class="punctuation">)</span>  <span class="comment"># Creates a PNG file of the same graphic</span>
</pre>
<p><img alt="image8" src="../images/phylo-dot.png" /></p>
<p>（提示：如果你使用 <tt class="docutils literal"><span class="pre">-pylab</span></tt> 选项执行IPython，调用 <tt class="docutils literal">draw_graphviz</tt> 将导致matplotlib
查看器自动运行，而不需要手动的调用 <tt class="docutils literal">show()</tt> 方法。）</p>
<p>这将输出树对象到一个NetworkX图中，使用Graphviz来布局节点的位置，并使用matplotlib来显示
它。这里有几个关键词参数来修改结果图像，包括大多数被NetworkX函数 <tt class="docutils literal">networkx.draw</tt> 和
<tt class="docutils literal">networkx.draw_graphviz</tt> 所接受的参数。</p>
<p>最终的显示也受所提供的树对象的 <tt class="docutils literal">rooted</tt> 属性的影响。有根树在每个分支（branch）上显示
一个“head”来表明它的方向（见图. <a class="reference external" href="#fig:phylo-rooted">13.3</a> ）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;simple.dnd&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;newick&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">rooted</span> <span class="operator">=</span> <span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_graphiz</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
</pre>
<p><img alt="image9" src="../images/phylo-rooted.png" /></p>
<p>“prog”参数指定Graphviz的用来布局的引擎。默认的引擎 <tt class="docutils literal">twopi</tt> 对任何大小的树都表现很好，
很可靠的避免交叉的分支出现。<tt class="docutils literal">neato</tt> 程序可能画出更加好看的中等大小的树，但是有时候会
有交叉分支出现（见图. <a class="reference external" href="#fig:phylo-color">13.3</a> ）。 <tt class="docutils literal">dot</tt> 程序或许对小型的树有用，
但是对于大一点的树的布局易产生奇怪的事情。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_graphviz</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">,</span> <span class="name">prog</span><span class="operator">=</span><span class="literal string">&quot;neato&quot;</span><span class="punctuation">)</span>
</pre>
<p><img alt="image10" src="../images/phylo-color.png" /></p>
<p>这个查看方式非常方便研究大型的树，因为matplotlib查看器可以放大选择的区域，使得杂乱的图像
变得稀疏。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;apaf.xml&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;phyloxml&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_graphviz</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">,</span> <span class="name">prog</span><span class="operator">=</span><span class="literal string">&quot;neato&quot;</span><span class="punctuation">,</span> <span class="name">node_size</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">)</span>
</pre>
<p><img alt="image11" src="../images/phylo-apaf.png" /> <img alt="image12" src="../images/phylo-apaf-zoom.png" /></p>
<p>注意，分支长度并没有被正确地显示，因为Graphviz在布局时忽略了他们。然而，分支长度可以在输出
树为NetworkX图对象（ <tt class="docutils literal">to_networkx</tt> ）时重新获得。</p>
<p>查看Biopython维基的Phylo页面
(<a class="reference external" href="http://biopython.org/wiki/Phylo">http://biopython.org/wiki/Phylo</a>)
以获得关于 <tt class="docutils literal">draw_ascii</tt> 、 <tt class="docutils literal">draw_graphviz</tt> 和 <tt class="docutils literal">to_networkx</tt> 的更加高级的功能的描述
和例子。</p>
</div>
<div class="section" id="treeclade">
<h2>13.4  使用Tree和Clade对象</h2>
<p><tt class="docutils literal">parse</tt> 和 <tt class="docutils literal">read</tt> 方法产生的 <tt class="docutils literal">Tree</tt> 对象是一些包含递归的子树的容器，连接到 <tt class="docutils literal">Tree</tt>
对象的 <tt class="docutils literal">root</tt> 属性（不管进化树实际上被认为是否有根）。一个 <tt class="docutils literal">Tree</tt> 包含进化树的全局信息，
如有根性（rootedness）和指向一个单独的 <tt class="docutils literal">Clade</tt> 的引用; 一个 <tt class="docutils literal">Clade</tt> 包含节点和进化枝
特异性信息，如分支长度（branch length）和一个它自身后代 <tt class="docutils literal">Clade</tt> 实例的列表，附着在 <tt class="docutils literal">clades</tt>
属性上。</p>
<p>所以，这里 <tt class="docutils literal">tree</tt> 和 <tt class="docutils literal">tree.root</tt> 间是有区别的. 然而，实际操作中，你几乎不需要担心它。为了
缓和这个不同，<tt class="docutils literal">Tree</tt> 和 <tt class="docutils literal">Clade</tt> 两者都继承自 <tt class="docutils literal">TreeMixin</tt>，它包含常用的用来查找、审视和
修改树和任何它的进化枝的方法的实现。这意味着，所有 <tt class="docutils literal">tree</tt> 所支持的方法在 <tt class="docutils literal">tree.root</tt> 和
任何它下面的clade中都能用。（ <tt class="docutils literal">Clade</tt> 也有一个 <tt class="docutils literal">root</tt> 属性，它返回clade对象本身。）</p>
<div class="section" id="id150">
<h3>13.4.1  查找和遍历类方法</h3>
<p>为了方便起见，我们提供了两个简化的方法来直接返回所有的外部或内部节点为列表：</p>
<dl class="docutils">
<dt><strong>``get_terminals``</strong></dt>
<dd>创建一个包含树的所有末端（叶子）节点的列表。</dd>
<dt><strong>``get_nonterminals``</strong></dt>
<dd>创建一个包含树的所有非末端（内部）节点的列表。</dd>
</dl>
<p>这两个都包装了一个能完全控制树的遍历的方法 <tt class="docutils literal">find_clades</tt>。另外两个遍历方法 <tt class="docutils literal">find_elements</tt>
和 <tt class="docutils literal">find_any</tt> 依赖于同样的核心功能，也接受同样的参数，没有更好的描述我们就把这个参数叫做
“目标说明”（target specification）吧。它们指定哪些树中的对象将被匹配并在迭代过程中返回。
第一个参数可以是下面的任何类型：</p>
<ul>
<li><p class="first">一个 <strong>TreeElement 实例</strong> ，那个树的元素将根据一致性被匹配——这样，使用Clade实例作为目标将找到
树中的这个Clade；</p>
</li>
<li><p class="first">一个 <strong>string</strong> ，匹配树元素的字符串表示——特别地，Clade的 <tt class="docutils literal">name</tt> <em>(在Biopython 1.56中引入)</em>；</p>
</li>
<li><p class="first">一个 <strong>class</strong> 或 <strong>type</strong>，这样每一个类型（或子类型）相同的树元素都被匹配；</p>
</li>
<li><p class="first">一个 <strong>dictionary</strong> ，其中键（key）是树元素的属性名，值（value）将匹配到每个树元素相应的属性值。
它变得更加详细：</p>
<ul class="simple">
<li>如果提供的是 <tt class="docutils literal">int</tt> 类型，它将匹配数值上相等的属性，即，1将匹配1或者1.0</li>
<li>如果提供的是boolean类型（True或者False），对应的属性值将被当做boolean求值和检验</li>
<li><tt class="docutils literal">None</tt> 匹配 <tt class="docutils literal">None</tt></li>
<li>如果提供的是字符串，将被当做正则表达式对待（必须匹配对应元素属性的全部，不能只是前面的部分）。
提供没有特殊正则表达式字符的字符串将精准的匹配字符串属性，所以如果你不适用正则表达式，不用
担心它。例如，包含进化枝名称Foo1、Foo2和Foo3的一个树，
<tt class="docutils literal"><span class="pre">tree.find_clades({&quot;name&quot;:</span> <span class="pre">&quot;Foo1&quot;})</span></tt> 将匹配 Foo1，
<tt class="docutils literal">{&quot;name&quot;: <span class="pre">&quot;Foo.*&quot;}</span></tt> 匹配所有的三个进化枝，而
<tt class="docutils literal">{&quot;name&quot;: &quot;Foo&quot;}</tt> 并不匹配任何进化枝。</li>
</ul>
<p>由于浮点数值可能产生奇怪的行为，我们不支持直接匹配 <tt class="docutils literal">float</tt>s 类型。作为替代，使用boolean值
<tt class="docutils literal">True</tt> 来匹配每个元素中指定属性的非零值，然后再对这个属性用不等式（或精确地数值，如果你喜欢
危险地活着）进行手动过滤。</p>
<p>如果该字典包含多个条目，匹配的元素必须匹配所有给定的属性值——以“and”方式思考，而不是“or”。</p>
</li>
<li><p class="first">一个接受一个参数（它将应用于树中的每一个元素），返回True或False的函数 <strong>function</strong> 。为方便起见，
LookupError、AttributeError和ValueError被沉默，这样就提供了另外一个在树中查找浮点值的安全方式，
或者一些更加复杂的特性。</p>
</li>
</ul>
<p>在目标参数后面，有两个可选的关键词参数：</p>
<dl class="docutils">
<dt><strong>terminal</strong></dt>
<dd>— 用来选择或排除末端进化枝（或者叫叶子节点）的一个boolean值：True仅搜索末端进化枝，False则搜索
非末端（内部）进化枝，而默认为None，同时搜索末端和非末端进化枝，包括没有 <tt class="docutils literal">is_terminal</tt> 方法的
任何树元素。</dd>
<dt><strong>order</strong></dt>
<dd>— 树遍历的顺序：<tt class="docutils literal">&quot;preorder&quot;</tt> （默认值）是深度优先搜索（depth-first search，DFS）， <tt class="docutils literal">&quot;postorder&quot;</tt>
是子节点先于父节点的DFS搜索， <tt class="docutils literal">&quot;level&quot;</tt> 是宽度优先搜索（breadth-first search，BFS）。</dd>
</dl>
<p>最后，这些方法接受任意的关键词参数，这些参数将被以和词典“目标说明”相同的方式对待：键表示要搜索的元素
属性的名称，参数值（string、integer、None或者boolean）将和找到的每个属性的值进行比较。如果没有提供
关键词参数，则任何TreeElement类型将被匹配。这个的代码普遍比传入一个词典作为“目标说明”要短：
<tt class="docutils literal"><span class="pre">tree.find_clades({&quot;name&quot;:</span> <span class="pre">&quot;Foo1&quot;})</span></tt> 可以简化为 <tt class="docutils literal"><span class="pre">tree.find_clades(name=&quot;Foo1&quot;)</span></tt>。</p>
<p>（在Biopython 1.56和以后的版本中，这可以更短：<tt class="docutils literal"><span class="pre">tree.find_clades(&quot;Foo1&quot;)</span></tt> ）</p>
<p>现在我们已经掌握了“目标说明”，这里有一些遍历树的方法：</p>
<dl class="docutils">
<dt><strong>``find_clades``</strong></dt>
<dd><p class="first">查找每个包含匹配元素的进化枝。就是说，用 <tt class="docutils literal">find_elements</tt> 查找每个元素，然而返回对应的clade对象。
（这通常是你想要的。）</p>
<p class="last">最终的结果是一个包含所有匹配对象的迭代器，默认为深度优先搜索。这不一定是和Newick、Nexus或XML原文件
中显示的相同的顺序。</p>
</dd>
<dt><strong>``find_elements``</strong></dt>
<dd><dl class="first last docutils">
<dt>查找和给定属性匹配的所有树元素，返回匹配的元素本身。简单的Newick树没有复杂的子元素，所以它将和</dt>
<dd><tt class="docutils literal">find_clades</tt> 的行为一致。PhyloXML树通常在clade上附加有复杂的对象，所以这个方法对提取这些信息
非常有用。</dd>
</dl>
</dd>
<dt><strong>``find_any``</strong></dt>
<dd>返回 <tt class="docutils literal">find_elements()</tt> 所找到的第一个元素，或者None。这对于检测树中是否存在匹配的元素也非常有用，
可以在条件判断语句中使用。</dd>
</dl>
<p>另外两个用于帮助在树的节点间导航的方法：</p>
<dl class="docutils">
<dt><strong>``get_path``</strong></dt>
<dd>直接列出从树的根节点（或当前进化枝）到给定的目标间的所有clade。返回包含这个路径上所有clade对象的
列表，以给定目标为结尾，但不包含根进化枝。</dd>
<dt><strong>``trace``</strong></dt>
<dd>列出树中两个目标间的所有clade对象，不包含起始和结尾。</dd>
</dl>
</div>
<div class="section" id="id151">
<h3>13.4.2  信息类方法</h3>
<p>这些方法提供关于整个树（或任何进化枝）的信息。</p>
<dl class="docutils">
<dt><strong>``common_ancestor``</strong></dt>
<dd>查找所提供的所有目标的最近共同祖先（the most recent common ancestor）
（这将是一个Clade对象）。如果没有提供任何目标，将返回当前Clade（调用该
方法的那个）的根；如果提供一个目标，将返回目标本身。然而，如果有任何提供
的目标无法在当前tree（或clade）中找到，将引起一个异常。</dd>
<dt><strong>``count_terminals``</strong></dt>
<dd>计算树中末端（叶子）节点的个数。</dd>
<dt><strong>``depths``</strong></dt>
<dd>创建一个树中进化枝到其深度的映射。结果是一个字典，其中键是树中所有的Clade
实例，值是从根到每个clade（包含末端）的距离。默认距离是到这个clade的分支
长度累加，然而使用 <tt class="docutils literal">unit_branch_lengths=True</tt> 选项，将只计算分支的个数
（其在树中的级数）。</dd>
<dt><strong>``distance``</strong></dt>
<dd>计算两个目标间的分支长度总和。如果只指定一个目标，另一个则为该树的根。</dd>
<dt><strong>``total_branch_length``</strong></dt>
<dd>计算这个树中的分支长度总和。这在系统发生学中通常就称为树的长度“length”，
但是我们使用更加明确的名称，以避免和Python的术语混淆。</dd>
</dl>
<p>余下的方法是boolean检测方法：</p>
<dl class="docutils">
<dt><strong>``is_bifurcating``</strong></dt>
<dd>如果树是严格的二叉树；即，所有的节点有2个或者0个子代（对应的，内部或外部）。
根节点可能有三个后代，然而仍然被认为是二叉树的一部分。</dd>
<dt><strong>``is_monophyletic``</strong></dt>
<dd>检验给定的所有目标是否组成一个完成的子进化枝——即，存在一个进化枝满足：它的
末端节点和给定的目标是相同的集合。目标需要时树中的末端节点。为方便起见，若
给定目标是一个单系（monophyletic），这个方法将返回它们的共同祖先（MCRA）（
而不是 <tt class="docutils literal">True</tt> ），否则将返回 <tt class="docutils literal">False</tt> 。</dd>
<dt><strong>``is_parent_of``</strong></dt>
<dd>若目标是这个树的后代（descendant）则为True——不必为直接后代。检验一个进化枝的
直接后代，只需要用简单的列表成员检测方法： <tt class="docutils literal">if subclade in clade: ...</tt></dd>
<dt><strong>``is_preterminal``</strong></dt>
<dd>若所有的直接后代都为末端则为True；否则任何一个直接后代不为末端则为False。</dd>
</dl>
</div>
<div class="section" id="id152">
<h3>13.4.3  修改类方法</h3>
<p>这些方法都在原地对树进行修改，所以如果你想保持原来的树不变，你首先要使用Python的
<tt class="docutils literal">copy</tt> 模块对树进行完整的拷贝：</p>
<pre class="code python literal-block">
<span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'example.xml'</span><span class="punctuation">,</span> <span class="literal string">'phyloxml'</span><span class="punctuation">)</span>
<span class="keyword namespace">import</span> <span class="name namespace">copy</span>
<span class="name">newtree</span> <span class="operator">=</span> <span class="name">copy</span><span class="operator">.</span><span class="name">deepcopy</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
</pre>
<dl class="docutils">
<dt><strong>``collapse``</strong></dt>
<dd>从树中删除目标，重新连接它的子代（children）到它的父亲节点（parent）。</dd>
<dt><strong>``collapse_all``</strong></dt>
<dd>删除这个树的所有后代（descendants），只保留末端节点（terminals）。
分支长度被保留，即到每个末端节点的距离保持不变。如指定一个目标（见上），
只坍塌（collapses）和指定匹配的内部节点。</dd>
<dt><strong>``ladderize``</strong></dt>
<dd>根据末端节点的个数，在原地对进化枝（clades）进行排序。越深的进化枝默认被放到最后，
使用 <tt class="docutils literal">reverse=True</tt> 将其放到最前。</dd>
<dt><strong>``prune``</strong></dt>
<dd>从树中修剪末端进化枝（terminal clade）。如果分类名（taxon）来自一个二叉枝（bifurcation），
连接的节点将被坍塌，它的分支长度将被加到剩下的末端节点上。这可能不再是一个有意义的值。</dd>
<dt><strong>``root_with_outgroup``</strong></dt>
<dd><p class="first">使用包含给定目标的外群进化枝（outgroup clade）重新确定树的根节点，即外群的共同祖先。该方法
只在Tree对象中能用，不能用于Clade对象。</p>
<p>如果外群和self.root一致，将不发生改变。如果外群进化枝是末端（即一个末端节点被作为外群），一个
新的二叉根进化枝将被创建，且到给定外群的分支长度为0。否则，外群根部的内部节点变为整个树的一个
三叉根。如果原先的根是一个二叉，它将被从树中遗弃。</p>
<p class="last">在所有的情况下，树的分支长度总和保持不变。</p>
</dd>
<dt><strong>``root_at_midpoint``</strong></dt>
<dd>重新选择树中两个最远的节点的中点作为树的根。（这实际上是使用 <tt class="docutils literal">root_with_outgroup</tt> 函数。）</dd>
<dt><strong>``split``</strong></dt>
<dd>产生 <em>n</em> （默认为2）个 新的后代。在一个物种树中，这是一个物种形成事件。新的进化枝拥有给定的
<tt class="docutils literal">branch_length</tt> 以及和这个进化枝的根相同的名字，名字后面包含一个整数后缀（从0开始计数）——
例如，分割名为“A”的进化枝将生成子进化枝“A0”和“A1”。</dd>
</dl>
<p>查看Biopython维基的Phylo页面
(<a class="reference external" href="http://biopython.org/wiki/Phylo">http://biopython.org/wiki/Phylo</a>)
以获得更多已有方法的使用示例。</p>
</div>
<div class="section" id="phyloxml">
<h3>13.4.4  PhyloXML树的特性</h3>
<p>phyloXML文件格式包含用来注释树的，采用额外数据格式和图像提示的字段。</p>
<p>参加Biopython维基上的PhyloXML页面
(<a class="reference external" href="http://biopython.org/wiki/PhyloXML">http://biopython.org/wiki/PhyloXML</a>)
以查看关于使用PhyloXML提供的额外注释特性的描述和例子。</p>
</div>
</div>
<div class="section" id="id153">
<h2>13.5  运行外部程序</h2>
<p>尽管Bio.Phylo本身不从序列比对推断进化树，但这里有一些第三方的程序可以使用。
他们通过 <tt class="docutils literal">Bio.Phylo.Applications</tt> 模块获得支持，使用和 <tt class="docutils literal">Bio.Emboss.Applications</tt> 、
<tt class="docutils literal">Bio.Align.Applications</tt> 以及其他模块相同的通用框架。</p>
<p>Biopython 1.58引入了一个PhyML的打包程序（wrapper）
(<a class="reference external" href="http://www.atgc-montpellier.fr/phyml/">http://www.atgc-montpellier.fr/phyml/</a>)。
该程序接受一个 <tt class="docutils literal"><span class="pre">phylip-relaxed</span></tt> 格式（它是Phylip格式，然而没有对分类名称的10个字符的限制）
的比对输入和多种参数。一个快速的例子是：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Phylo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Phylo.Applications</span> <span class="keyword namespace">import</span> <span class="name">PhymlCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cmd</span> <span class="operator">=</span> <span class="name">PhymlCommandline</span><span class="punctuation">(</span><span class="name builtin">input</span><span class="operator">=</span><span class="literal string">'Tests/Phylip/random.phy'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">out_log</span><span class="punctuation">,</span> <span class="name">err_log</span> <span class="operator">=</span> <span class="name">cmd</span><span class="punctuation">()</span>
</pre>
<p>这生成一个树文件盒一个统计文件，名称为：
[<em>input filename</em>]<tt class="docutils literal">_phyml_tree.txt</tt> 和
[<em>input filename</em>]<tt class="docutils literal">_phyml_stats.txt</tt>. 树文件的格式是Newick格式：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">'Tests/Phylip/random.phy_phyml_tree.txt'</span><span class="punctuation">,</span> <span class="literal string">'newick'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Phylo</span><span class="operator">.</span><span class="name">draw_ascii</span><span class="punctuation">(</span><span class="name">tree</span><span class="punctuation">)</span>
</pre>
<p>一个类似的RAxML打包程序
(<a class="reference external" href="http://sco.h-its.org/exelixis/software.html">http://sco.h-its.org/exelixis/software.html</a>)
也已经被添加到Biopython 1.60中。</p>
<p>注意，如果你系统中已经安装了EMBOSS的Phylip扩展，一些常用的Phylip程序，包括 <tt class="docutils literal">dnaml</tt> 和 <tt class="docutils literal">protml</tt>
已经通过 <tt class="docutils literal">Bio.Emboss.Applications</tt> 中的EMBOSS打包程序被支持。参见章节 <a class="reference external" href="#sec:alignment-tools">6.4</a>
以查看使用这些程序的例子和提示。</p>
</div>
<div class="section" id="paml">
<h2>13.6  PAML整合</h2>
<p>Biopython 1.58引入了对PAML的支持
(<a class="reference external" href="http://abacus.gene.ucl.ac.uk/software/paml.html">http://abacus.gene.ucl.ac.uk/software/paml.html</a>),
它是一个采用最大似然法（maximum likelihood）进行系统进化分析的程序包。目前，对程序codeml、baseml和yn00的支持
已经实现。由于PAML使用控制文件而不是命令行参数来控制运行时选项，这个打包程序（wrapper）的使用格式和Biopython
的其他应用打包程序有些差异。</p>
<p>一个典型的流程是：初始化一个PAML对象，指定一个比对文件，一个树文件，一个输出文件和工作路径。下一步，运行时
选项通过 <tt class="docutils literal">set_options()</tt> 方法或者读入一个已有的控制文件来设定。最后，程序通过 <tt class="docutils literal">run()</tt> 方法来运行，输出文件
将自动被解析到一个结果目录。</p>
<p>下面是一个codeml典型用法的例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Phylo.PAML</span> <span class="keyword namespace">import</span> <span class="name">codeml</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cml</span> <span class="operator">=</span> <span class="name">codeml</span><span class="operator">.</span><span class="name">Codeml</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cml</span><span class="operator">.</span><span class="name">alignment</span> <span class="operator">=</span> <span class="literal string">&quot;Tests/PAML/alignment.phylip&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cml</span><span class="operator">.</span><span class="name">tree</span> <span class="operator">=</span> <span class="literal string">&quot;Tests/PAML/species.tree&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cml</span><span class="operator">.</span><span class="name">out_file</span> <span class="operator">=</span> <span class="literal string">&quot;results.out&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cml</span><span class="operator">.</span><span class="name">working_dir</span> <span class="operator">=</span> <span class="literal string">&quot;./scratch&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cml</span><span class="operator">.</span><span class="name">set_options</span><span class="punctuation">(</span><span class="name">seqtype</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">verbose</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">noisy</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">RateAncestor</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">model</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">NSsites</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">],</span>
<span class="operator">...</span>         <span class="name">CodonFreq</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">cleandata</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">fix_alpha</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">,</span>
<span class="operator">...</span>         <span class="name">kappa</span><span class="operator">=</span><span class="literal number float">4.54006</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">results</span> <span class="operator">=</span> <span class="name">cml</span><span class="operator">.</span><span class="name">run</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ns_sites</span> <span class="operator">=</span> <span class="name">results</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;NSsites&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m0</span> <span class="operator">=</span> <span class="name">ns_sites</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m0_params</span> <span class="operator">=</span> <span class="name">m0</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;parameters&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m0_params</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;omega&quot;</span><span class="punctuation">)</span>
</pre>
<p>已有的输出文件也可以通过模块的 <tt class="docutils literal">read()</tt> 方法来解析：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">results</span> <span class="operator">=</span> <span class="name">codeml</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;Tests/PAML/Results/codeml/codeml_NSsites_all.out&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">results</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="literal string">&quot;lnL max&quot;</span><span class="punctuation">)</span>
</pre>
<p>这个新模块的详细介绍目前在Biopython维基上可以看到：
<a class="reference external" href="http://biopython.org/wiki/PAML">http://biopython.org/wiki/PAML</a></p>
</div>
<div class="section" id="id154">
<h2>13.7  未来计划</h2>
<p>Bio.Phylo 目前还在开发中，下面是我们可能会在将来的发布版本中添加的特性：</p>
<dl class="docutils">
<dt><strong>新方法</strong></dt>
<dd>通常用来操作Tree和Clade对象的有用方法会首先出现在Biopython维基上，这样常规用户
就能在我们添加到Bio.Phylo之前测试这些方法，看看它们是否有用：
<a class="reference external" href="http://biopython.org/wiki/Phylo_cookbook">http://biopython.org/wiki/Phylo_cookbook</a></dd>
<dt><strong>Bio.Nexus port</strong></dt>
<dd><p class="first">这个模块的大部分是在2009年NESCent主办的谷歌编程夏令营中写的，作为实现Python对phyloXML数据格式（见
<a class="reference external" href="#sec:PhyloXML">13.4.4</a> ）支持的一个项目。对Newick和Nexus格式的支持，已经通过导入Bio.Nexus模块
的一部分被添加到Bio.Phylo使用的新类中。</p>
<p class="last">目前，Bio.Nexus包含一些还没有导入到Bio.Phylo类中的有用的特性——特别是，计算一致树（consensus tree）。
如果你发现某些功能Bio.Phylo中没有，试试在Bio.Nexus中能不能找到。</p>
</dd>
</dl>
<p>我们乐意接受任何增强该模块功能和使用性的建议；如果有，只需要通过邮件列表或我们的bug数据库让我们知道。</p>
</div>
</div>
<div class="section" id="bio-motifs">
<h1>第14章   使用Bio.motifs进行模体序列分析</h1>
<p>这章主要的介绍Biopython中的 <tt class="docutils literal">Bio.motifs</tt> 包。这个包是为了方便那些需要进行模体序列分析的人们而特意提供的，所以我想你们在使用时肯定对模体序列分析的一些相关要点都很熟悉。假如在使用中遇到不清楚的地方，请您查阅 <a class="reference external" href="#sec:links">14.8</a> 相关章节以获得有关的信息。</p>
<p>这章的大部分内容是介绍Biopython 1.61 之前版本中新加入的 <tt class="docutils literal">Bio.motifs</tt> 包，该包替代了Biopython 1.50版本中的 <tt class="docutils literal">Bio.Motif</tt> 包，而 <tt class="docutils literal">Bio.Motif</tt> 包是基于较早版本的Biopython 中的两个模块 <tt class="docutils literal">Bio.AlignAce</tt> 和 <tt class="docutils literal">Bio.MEME</tt> 。<tt class="docutils literal">Bio.motifs</tt> 包较好地综合了上述的几个模块的功能，做为一个统一模块工具。</p>
<p>说到其他库，看到这里，你或许会对 <a class="reference external" href="http://fraenkel.mit.edu/TAMO/">TAMO</a> 感兴趣，这是另一个分析模体序列的Python库。它能提供更多关于 <em>de-novo</em> 模体的查找方式，不过它并没有纳入到Biopython中，而且在商业用途上还有一些限制。</p>
<div class="section" id="id155">
<h2>14.1  模体对象</h2>
<p>由于我们感兴趣的是模体分析，所以我们需要先看看 <tt class="docutils literal">Motif</tt> 对象。对此我们需要先导入Bio.motifs包：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">motifs</span>
</pre>
<p>然后我们可以开始创建我们第一个模体对象。我们可以从模体的实例列表中创建一个 <tt class="docutils literal">Motif</tt> 对象，也可以通过读取模体数据库中或模体查找软件产生的文件来获得一个 <tt class="docutils literal">Motif</tt> 对象。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">motifs</span>
</pre>
<div class="section" id="id156">
<h3>14.1.1  从实例中创建一个模体</h3>
<p>假设我们有一些DNA模体的实例：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">instances</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TACAA&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>              <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TACGC&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>              <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TACAC&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>              <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TACCC&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>              <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AACCC&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>              <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AATGC&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>              <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;AATGC&quot;</span><span class="punctuation">),</span>
<span class="operator">...</span>             <span class="punctuation">]</span>
</pre>
<p>然后我们可以如下创建一个模体对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">create</span><span class="punctuation">(</span><span class="name">instances</span><span class="punctuation">)</span>
</pre>
<p>这些实例被存储在一个名为 <tt class="docutils literal">m.instances</tt> 的属性中，这个其实也就是一个Python的列表，只不过附加了一些功能，这些功能将在之后介绍。将这些模体对象打印出来后就可以看出这些实例是从哪构建出来的。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m</span>
<span class="name">TACAA</span>
<span class="name">TACGC</span>
<span class="name">TACAC</span>
<span class="name">TACCC</span>
<span class="name">AACCC</span>
<span class="name">AATGC</span>
<span class="name">AATGC</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>模体的长度像其他一些实例一些被定义为序列的长度：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">m</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
</pre>
<p>模体对象有一个 <tt class="docutils literal">.counts</tt> 属性，可以用来查看碱基在每个位置的数目。可以把这个统计表用易读的格式打印出来：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">3.00</span>   <span class="literal number float">7.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">2.00</span>   <span class="literal number float">1.00</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">5.00</span>   <span class="literal number float">2.00</span>   <span class="literal number float">6.00</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">3.00</span>   <span class="literal number float">0.00</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">4.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">2.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>你也可以像使用字典一样获取这些数目：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[</span><span class="literal string">'A'</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">7</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">]</span>
</pre>
<p>但是你也可以把它看成一个二维数列，核苷酸作为列，位置作为行：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[</span><span class="literal string">'T'</span><span class="punctuation">,</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="literal number integer">4</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[</span><span class="literal string">'T'</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">]</span>
<span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[</span><span class="literal string">'T'</span><span class="punctuation">,</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="literal number integer">0</span>
</pre>
<p>你还可以直接获得核苷酸数目矩阵中的列</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[:,</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number integer">3</span><span class="punctuation">}</span>
</pre>
<p>除了使用核苷酸本身，你还可以使用模体碱基序列按字符排序后的核苷酸索引：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span>
<span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="operator">.</span><span class="name">letters</span>
<span class="literal string">'GATC'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">sorted</span><span class="punctuation">(</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="operator">.</span><span class="name">letters</span><span class="punctuation">)</span>
<span class="punctuation">[</span><span class="literal string">'A'</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[</span><span class="literal string">'A'</span><span class="punctuation">,:]</span>
<span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">7</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">,:]</span>
<span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">7</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>模体有一个相关联的一致序列，这个序列被定义为由 <tt class="docutils literal">.counts</tt> 矩阵相应列中具有最大值的碱基，这些碱基是按模体序列排列的：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TACGC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>反一致序列也一样，只不过是由 <tt class="docutils literal">.counts</tt> 矩阵中相应列的最小值来选：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">anticonsensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GGGTG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>你也可以利用简并一致序列，用不确定核苷酸来表示序列某一位置的所有核苷酸：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">degenerate_consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'WACVC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>此处，W和R都是按照IUPAC不确定核苷酸表规定的：W代表A或T，V代表A，C或G [<a class="reference external" href="#cornish1985">10</a>] 。这些简并一致序列是按照Cavener指定的规则[<a class="reference external" href="#cavener1987">11</a>]来建立的。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">r</span> <span class="operator">=</span> <span class="name">m</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">r</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GCGTA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">r</span><span class="operator">.</span><span class="name">degenerate_consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GBGTW'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">r</span>
<span class="name">TTGTA</span>
<span class="name">GCGTA</span>
<span class="name">GTGTA</span>
<span class="name">GGGTA</span>
<span class="name">GGGTT</span>
<span class="name">GCATT</span>
<span class="name">GCATT</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>反向互补序列和简并一致序列都只在DNA模体中有。</p>
</div>
<div class="section" id="id157">
<h3>14.1.2  读取模体</h3>
<p>从实例手动创建一个模体确实有点无趣，所以用一些I/O函数来读写模体是很有用的。目前对于如何存储模体还没有一些真正的标准，不过有一些格式用得比其他更经常。这其中最重要的区别在于模体表示是基于实例还是某种PWM矩阵。</p>
<div class="section" id="jaspar">
<h4>JASPAR</h4>
<p>作为一个最流行的模体数据库 <a class="reference external" href="http://jaspar.genereg.net">JASPAR</a> 它不是以一系列的实例就是频率矩阵。比如，下面就是JASPAR <tt class="docutils literal">Arnt.sites</tt> 文件的开头和结尾行显示了老鼠螺旋-环-螺旋转录因子Arnt的结合位点：</p>
<pre class="code python literal-block">
<span class="operator">&gt;</span><span class="name">MA0004</span> <span class="name">ARNT</span>    <span class="literal number integer">1</span>
<span class="name">CACGTGatgtcctc</span>
<span class="operator">&gt;</span><span class="name">MA0004</span> <span class="name">ARNT</span>    <span class="literal number integer">2</span>
<span class="name">CACGTGggaggtac</span>
<span class="operator">&gt;</span><span class="name">MA0004</span> <span class="name">ARNT</span>    <span class="literal number integer">3</span>
<span class="name">CACGTGccgcgcgc</span>
<span class="operator">...</span>
<span class="operator">&gt;</span><span class="name">MA0004</span> <span class="name">ARNT</span>    <span class="literal number integer">18</span>
<span class="name">AACGTGacagccctcc</span>
<span class="operator">&gt;</span><span class="name">MA0004</span> <span class="name">ARNT</span>    <span class="literal number integer">19</span>
<span class="name">AACGTGcacatcgtcc</span>
<span class="operator">&gt;</span><span class="name">MA0004</span> <span class="name">ARNT</span>    <span class="literal number integer">20</span>
<span class="name">aggaatCGCGTGc</span>
</pre>
<p>那些用大字字母表示的序列的一部分就是被用来相互比对的模体实例。</p>
<p>我们可以从下面的实例创建一个 <tt class="docutils literal">Motif</tt> 对象：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">motifs</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">arnt</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;Arnt.sites&quot;</span><span class="punctuation">),</span> <span class="literal string">&quot;sites&quot;</span><span class="punctuation">)</span>
</pre>
<p>从这个模体创建的实例存储在该模体的 <tt class="docutils literal">.instances</tt> 属性：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[:</span><span class="literal number integer">3</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CACGTG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()),</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CACGTG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()),</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CACGTG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">instance</span> <span class="operator word">in</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">instance</span>
<span class="operator">...</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">CACGTG</span>
<span class="name">AACGTG</span>
<span class="name">AACGTG</span>
<span class="name">AACGTG</span>
<span class="name">AACGTG</span>
<span class="name">CGCGTG</span>
</pre>
<p>这个模体的计数矩阵可以从这些实例中自动计算出来：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">counts</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">4.00</span>  <span class="literal number float">19.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="name">C</span><span class="punctuation">:</span>  <span class="literal number float">16.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>   <span class="literal number float">0.00</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>JASPAR数据库也可以让模体像计数矩阵一样获得，不需要那些创建它们的实例。比如，下面这个JASPAR文件 <tt class="docutils literal">SRF.pfm</tt> 包含了人类SRF转录因子的计数矩阵：</p>
<pre class="code python literal-block">
 <span class="literal number integer">2</span>  <span class="literal number integer">9</span>  <span class="literal number integer">0</span>  <span class="literal number integer">1</span> <span class="literal number integer">32</span>  <span class="literal number integer">3</span> <span class="literal number integer">46</span>  <span class="literal number integer">1</span> <span class="literal number integer">43</span> <span class="literal number integer">15</span>  <span class="literal number integer">2</span>  <span class="literal number integer">2</span>
 <span class="literal number integer">1</span> <span class="literal number integer">33</span> <span class="literal number integer">45</span> <span class="literal number integer">45</span>  <span class="literal number integer">1</span>  <span class="literal number integer">1</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">1</span>  <span class="literal number integer">0</span>  <span class="literal number integer">1</span>
<span class="literal number integer">39</span>  <span class="literal number integer">2</span>  <span class="literal number integer">1</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span> <span class="literal number integer">44</span> <span class="literal number integer">43</span>
 <span class="literal number integer">4</span>  <span class="literal number integer">2</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span> <span class="literal number integer">13</span> <span class="literal number integer">42</span>  <span class="literal number integer">0</span> <span class="literal number integer">45</span>  <span class="literal number integer">3</span> <span class="literal number integer">30</span>  <span class="literal number integer">0</span>  <span class="literal number integer">0</span>
</pre>
<p>我们可以如下为计数矩阵创建一个模体：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">srf</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;SRF.pfm&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;pfm&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">srf</span><span class="operator">.</span><span class="name">counts</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>      <span class="literal number integer">6</span>      <span class="literal number integer">7</span>      <span class="literal number integer">8</span>      <span class="literal number integer">9</span>     <span class="literal number integer">10</span>     <span class="literal number integer">11</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">2.00</span>   <span class="literal number float">9.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>  <span class="literal number float">32.00</span>   <span class="literal number float">3.00</span>  <span class="literal number float">46.00</span>   <span class="literal number float">1.00</span>  <span class="literal number float">43.00</span>  <span class="literal number float">15.00</span>   <span class="literal number float">2.00</span>   <span class="literal number float">2.00</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">1.00</span>  <span class="literal number float">33.00</span>  <span class="literal number float">45.00</span>  <span class="literal number float">45.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>
<span class="name">G</span><span class="punctuation">:</span>  <span class="literal number float">39.00</span>   <span class="literal number float">2.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">44.00</span>  <span class="literal number float">43.00</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">4.00</span>   <span class="literal number float">2.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">13.00</span>  <span class="literal number float">42.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">45.00</span>   <span class="literal number float">3.00</span>  <span class="literal number float">30.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>由于这个模体是由计数矩阵直接创建的，所以它没有相关的实例：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">srf</span><span class="operator">.</span><span class="name">instances</span>
<span class="name builtin pseudo">None</span>
</pre>
<p>我们可以获得这两个模体的一致序列：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">counts</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">CACGTG</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">srf</span><span class="operator">.</span><span class="name">counts</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">GCCCATATATGG</span>
</pre>
</div>
<div class="section" id="meme">
<h4>MEME</h4>
<p>MEME [<a class="reference external" href="#bailey1994">12</a>] 是一个用来在一堆相关DNA或蛋白质序列中发现模体的工具。它输入一组相关DNA或蛋白质序列，输出所要求的模体。因此和JASPAR文件相比，MEME输出文件里面一般是含有多个模体。例子如下。</p>
<p>在输出文件的开头，有一些MEME生成的关于MEME和所用MEME版本的背景信息：</p>
<pre class="code python literal-block">
<span class="operator">********************************************************************************</span>
<span class="name">MEME</span> <span class="operator">-</span> <span class="name">Motif</span> <span class="name">discovery</span> <span class="name">tool</span>
<span class="operator">********************************************************************************</span>
<span class="name">MEME</span> <span class="name">version</span> <span class="literal number float">3.0</span> <span class="punctuation">(</span><span class="name">Release</span> <span class="name">date</span><span class="punctuation">:</span> <span class="literal number integer">2004</span><span class="operator">/</span><span class="literal number integer">08</span><span class="operator">/</span><span class="literal number integer">18</span> <span class="literal number integer">09</span><span class="punctuation">:</span><span class="literal number oct">07</span><span class="punctuation">:</span><span class="literal number oct">01</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>再往下，简要概括了输入的训练序列集：</p>
<pre class="code python literal-block">
<span class="operator">********************************************************************************</span>
<span class="name">TRAINING</span> <span class="name">SET</span>
<span class="operator">********************************************************************************</span>
<span class="name">DATAFILE</span><span class="operator">=</span> <span class="name">INO_up800</span><span class="operator">.</span><span class="name">s</span>
<span class="name">ALPHABET</span><span class="operator">=</span> <span class="name">ACGT</span>
<span class="name">Sequence</span> <span class="name">name</span>            <span class="name">Weight</span> <span class="name">Length</span>  <span class="name">Sequence</span> <span class="name">name</span>            <span class="name">Weight</span> <span class="name">Length</span>
<span class="operator">-------------</span>            <span class="operator">------</span> <span class="operator">------</span>  <span class="operator">-------------</span>            <span class="operator">------</span> <span class="operator">------</span>
<span class="name">CHO1</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>  <span class="name">CHO2</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>
<span class="name">FAS1</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>  <span class="name">FAS2</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>
<span class="name">ACC1</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>  <span class="name">INO1</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>
<span class="name">OPI3</span>                     <span class="literal number float">1.0000</span>    <span class="literal number integer">800</span>
<span class="operator">********************************************************************************</span>
</pre>
<p>以及所使用到的命令：</p>
<pre class="code python literal-block">
<span class="operator">********************************************************************************</span>
<span class="name">COMMAND</span> <span class="name">LINE</span> <span class="name">SUMMARY</span>
<span class="operator">********************************************************************************</span>
<span class="name">This</span> <span class="name">information</span> <span class="name">can</span> <span class="name">also</span> <span class="name">be</span> <span class="name">useful</span> <span class="operator word">in</span> <span class="name">the</span> <span class="name">event</span> <span class="name">you</span> <span class="name">wish</span> <span class="name">to</span> <span class="name">report</span> <span class="name">a</span>
<span class="name">problem</span> <span class="keyword">with</span> <span class="name">the</span> <span class="name">MEME</span> <span class="name">software</span><span class="operator">.</span>

<span class="name">command</span><span class="punctuation">:</span> <span class="name">meme</span> <span class="operator">-</span><span class="name">mod</span> <span class="name">oops</span> <span class="operator">-</span><span class="name">dna</span> <span class="operator">-</span><span class="name">revcomp</span> <span class="operator">-</span><span class="name">nmotifs</span> <span class="literal number integer">2</span> <span class="operator">-</span><span class="name">bfile</span> <span class="name">yeast</span><span class="operator">.</span><span class="name">nc</span><span class="operator">.</span><span class="literal number float">6.</span><span class="name">freq</span> <span class="name">INO_up800</span><span class="operator">.</span><span class="name">s</span>
<span class="operator">...</span>
</pre>
<p>接下来就是每个被发现模体的详细信息：</p>
<pre class="code python literal-block">
<span class="operator">********************************************************************************</span>
<span class="name">MOTIF</span>  <span class="literal number integer">1</span>        <span class="name">width</span> <span class="operator">=</span>   <span class="literal number integer">12</span>   <span class="name">sites</span> <span class="operator">=</span>   <span class="literal number integer">7</span>   <span class="name">llr</span> <span class="operator">=</span> <span class="literal number integer">95</span>   <span class="name">E</span><span class="operator">-</span><span class="name">value</span> <span class="operator">=</span> <span class="literal number float">2.0e-001</span>
<span class="operator">********************************************************************************</span>
<span class="operator">--------------------------------------------------------------------------------</span>
        <span class="name">Motif</span> <span class="literal number integer">1</span> <span class="name">Description</span>
<span class="operator">--------------------------------------------------------------------------------</span>
<span class="name">Simplified</span>        <span class="name">A</span>  <span class="punctuation">:::</span><span class="literal number integer">9</span><span class="punctuation">:</span><span class="name">a</span><span class="punctuation">::::</span><span class="literal number integer">3</span><span class="punctuation">:</span>
<span class="name">pos</span><span class="operator">.-</span><span class="name">specific</span>     <span class="name">C</span>  <span class="punctuation">::</span><span class="name">a</span><span class="punctuation">:</span><span class="literal number integer">9</span><span class="punctuation">:</span><span class="literal number integer">11691</span><span class="name">a</span>
<span class="name">probability</span>       <span class="name">G</span>  <span class="punctuation">::::</span><span class="literal number integer">1</span><span class="punctuation">::</span><span class="literal number integer">94</span><span class="punctuation">:</span><span class="literal number integer">4</span><span class="punctuation">:</span>
<span class="name">matrix</span>            <span class="name">T</span>  <span class="name">aa</span><span class="punctuation">:</span><span class="literal number integer">1</span><span class="punctuation">::</span><span class="literal number integer">9</span><span class="punctuation">::</span><span class="literal number integer">11</span><span class="punctuation">:</span>
</pre>
<p>使用下面的方法来读取这个文件（以 <tt class="docutils literal">meme.dna.oops.txt</tt> 存储）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;meme.dna.oops.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;meme&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p><tt class="docutils literal">motifs.parse</tt> 命令直接读取整个文件，所以在使用后可以关闭这个文件。其中头文件信息被存储于属性中</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">version</span>
<span class="literal string">'3.0'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">datafile</span>
<span class="literal string">'INO_up800.s'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">command</span>
<span class="literal string">'meme -mod oops -dna -revcomp -nmotifs 2 -bfile yeast.nc.6.freq INO_up800.s'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">alphabet</span>
<span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">sequences</span>
<span class="punctuation">[</span><span class="literal string">'CHO1'</span><span class="punctuation">,</span> <span class="literal string">'CHO2'</span><span class="punctuation">,</span> <span class="literal string">'FAS1'</span><span class="punctuation">,</span> <span class="literal string">'FAS2'</span><span class="punctuation">,</span> <span class="literal string">'ACC1'</span><span class="punctuation">,</span> <span class="literal string">'INO1'</span><span class="punctuation">,</span> <span class="literal string">'OPI3'</span><span class="punctuation">]</span>
</pre>
<p>这个数据记录是 <tt class="docutils literal">Bio.motifs.meme.Record</tt> 类的一个对象。这个类继承于列表（list），所以你可以把这个 <tt class="docutils literal">record</tt> 看成模体对象的一个列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">TTCACATGCCGC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">degenerate_consensus</span>
<span class="name">TTCACATGSCNC</span>
</pre>
<p>除了一般的模体属性外，每个模体还同时保存着它们由MEME计算的各自特异信息。例如：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">num_occurrences</span>
<span class="literal number integer">7</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">length</span>
<span class="literal number integer">12</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">evalue</span> <span class="operator">=</span> <span class="name">motif</span><span class="operator">.</span><span class="name">evalue</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%3.1g</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">evalue</span>
<span class="literal number float">0.2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">name</span>
<span class="literal string">'Motif 1'</span>
</pre>
<p>除了像上面所做的用索引来获得相关记录，你也可以用它的名称来找到这个记录：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal string">'Motif 1'</span><span class="punctuation">]</span>
</pre>
<p>每个模体都有一个 <tt class="docutils literal">.instances</tt> 属性与在这个被发现模体中的序列实例，能够为每个实例提供一些信息：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">)</span>
<span class="literal number integer">7</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="name">Instance</span><span class="punctuation">(</span><span class="literal string">'TTCACATGCCGC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">motif_name</span>
<span class="literal string">'Motif 1'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">sequence_name</span>
<span class="literal string">'INO1'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">start</span>
<span class="literal number integer">620</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">strand</span>
<span class="literal string">'-'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">length</span>
<span class="literal number integer">12</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pvalue</span> <span class="operator">=</span> <span class="name">motif</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">pvalue</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%5.3g</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">pvalue</span>
<span class="literal number float">1.85e-08</span>
</pre>
</div>
<div class="section" id="mast">
<h4>MAST</h4>
</div>
<div class="section" id="transfac">
<h4>TRANSFAC</h4>
<p>TRANSFAC是一个为转录因子手动创建的一个专业数据库，同时还包括染色体结合位点和DNA结合的描述 [<a class="reference external" href="#matys2003">27</a>]。TRANSFAC数据库中所用的文件格式至今还被其他工具所使用，我们下面将介绍TRANSFAC文件格式。</p>
<p>TRANSFAC文件格式简单概括如下：</p>
<pre class="code python literal-block">
<span class="name">ID</span>  <span class="name">motif1</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">S</span>
<span class="literal number oct">02</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">R</span>
<span class="literal number oct">03</span>      <span class="literal number integer">3</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">1</span>      <span class="name">A</span>
<span class="literal number oct">04</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">05</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">A</span>
<span class="literal number oct">06</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">4</span>      <span class="literal number integer">1</span>      <span class="name">G</span>
<span class="literal number oct">07</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">4</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">08</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="name">T</span>
<span class="literal number integer">09</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">10</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="name">K</span>
<span class="literal number integer">11</span>      <span class="literal number integer">0</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="name">Y</span>
<span class="literal number integer">12</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="literal number integer">1</span>      <span class="name">G</span>
<span class="operator">//</span>
</pre>
<p>这个文件显示了模体 <tt class="docutils literal">motif1</tt> 中12个核苷酸的频率矩阵。总的来说，一个TRANSFAC文件里面可以包含多个模体。以下是示例文件 <tt class="docutils literal">transfac.dat</tt> 的内容：</p>
<pre class="code python literal-block">
<span class="name">VV</span>  <span class="name">EXAMPLE</span> <span class="name">January</span> <span class="literal number integer">15</span><span class="punctuation">,</span> <span class="literal number integer">2013</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="name">ID</span>  <span class="name">motif1</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">S</span>
<span class="literal number oct">02</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">R</span>
<span class="literal number oct">03</span>      <span class="literal number integer">3</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">1</span>      <span class="name">A</span>
<span class="operator">...</span>
<span class="literal number integer">11</span>      <span class="literal number integer">0</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="name">Y</span>
<span class="literal number integer">12</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="literal number integer">1</span>      <span class="name">G</span>
<span class="operator">//</span>
<span class="name">ID</span>  <span class="name">motif2</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">R</span>
<span class="literal number oct">02</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">S</span>
<span class="operator">...</span>
<span class="literal number integer">09</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="name">T</span>
<span class="literal number integer">10</span>      <span class="literal number integer">0</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="name">Y</span>
<span class="operator">//</span>
</pre>
<p>可用如下方法读取TRANSFAC文件：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;transfac.dat&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;TRANSFAC&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>如果有总版本号的话，它是存储在 <tt class="docutils literal">record.version</tt> 中：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">version</span>
<span class="literal string">'EXAMPLE January 15, 2013'</span>
</pre>
<p>每个在 <tt class="docutils literal">record</tt> 中的模体都是 <tt class="docutils literal">Bio.motifs.transfac.Motif</tt> 类的实例，这些实例同时继承 <tt class="docutils literal">Bio.motifs.Motif</tt> 类和Python字典的属性。这些字典用双字母的键来存储关于这个模体的其他附加信息：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span> <span class="operator">=</span> <span class="name">record</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">degenerate_consensus</span> <span class="comment"># Using the Bio.motifs.Motif method</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'SRACAGGTGKYG'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="punctuation">[</span><span class="literal string">'ID'</span><span class="punctuation">]</span> <span class="comment"># Using motif as a dictionary</span>
<span class="literal string">'motif1'</span>
</pre>
<p>TRANSFAC文件一般比这些例子更详细，包含了许多关于模体的附加信息。表格 <a class="reference external" href="#table:transfaccodes">14.1.2</a> 列出了在TRANSFAC文件常见的双字母含义：</p>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr><td>Table 14.1: TRANSFAC文件中常见的字段</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="16%" />
<col width="84%" />
</colgroup>
<tbody valign="top">
<tr><td><tt class="docutils literal">AC</tt></td>
<td>Accession numbers 序列号</td>
</tr>
<tr><td><tt class="docutils literal">AS</tt></td>
<td>Accession numbers, secondary 第二序列号</td>
</tr>
<tr><td><tt class="docutils literal">BA</tt></td>
<td>Statistical basis 统计依据</td>
</tr>
<tr><td><tt class="docutils literal">BF</tt></td>
<td>Binding factors 结合因子</td>
</tr>
<tr><td><tt class="docutils literal">BS</tt></td>
<td>Factor binding sites underlying the matrix
基于矩阵的转录结合位点</td>
</tr>
<tr><td><tt class="docutils literal">CC</tt></td>
<td>Comments 注解</td>
</tr>
<tr><td><tt class="docutils literal">CO</tt></td>
<td>Copyright notice 版权事项</td>
</tr>
<tr><td><tt class="docutils literal">DE</tt></td>
<td>Short factor description 短因子说明</td>
</tr>
<tr><td><tt class="docutils literal">DR</tt></td>
<td>External databases 外部数据库</td>
</tr>
<tr><td><tt class="docutils literal">DT</tt></td>
<td>Date created/updated 创建或更新日期</td>
</tr>
<tr><td><tt class="docutils literal">HC</tt></td>
<td>Subfamilies 亚家庭名称</td>
</tr>
<tr><td><tt class="docutils literal">HP</tt></td>
<td>Superfamilies 超家庭名称</td>
</tr>
<tr><td><tt class="docutils literal">ID</tt></td>
<td>Identifier 身份证</td>
</tr>
<tr><td><tt class="docutils literal">NA</tt></td>
<td>Name of the binding factor 结合因子的名称</td>
</tr>
<tr><td><tt class="docutils literal">OC</tt></td>
<td>Taxonomic classification 分类</td>
</tr>
<tr><td><tt class="docutils literal">OS</tt></td>
<td>Species/Taxon 种类或分类</td>
</tr>
<tr><td><tt class="docutils literal">OV</tt></td>
<td>Older version 旧版本</td>
</tr>
<tr><td><tt class="docutils literal">PV</tt></td>
<td>Preferred version 首选版本</td>
</tr>
<tr><td><tt class="docutils literal">TY</tt></td>
<td>Type 类型</td>
</tr>
<tr><td><tt class="docutils literal">XX</tt></td>
<td>Empty line; these are not stored in the Record.
空白行;没在记录中存储的数据</td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<p>每个模体同时也有一个包含与这个模体相关参考资料的 <tt class="docutils literal">references</tt> 属性，用下面的双字母键来获得：</p>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr><td>Table 14.2: TRANSFAC文件中用来存储参考资料的字段</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="24%" />
<col width="76%" />
</colgroup>
<tbody valign="top">
<tr><td><tt class="docutils literal">RN</tt></td>
<td>Reference number 参考数目</td>
</tr>
<tr><td><tt class="docutils literal">RA</tt></td>
<td>Reference authors 参考资料作者</td>
</tr>
<tr><td><tt class="docutils literal">RL</tt></td>
<td>Reference data 参考数据</td>
</tr>
<tr><td><tt class="docutils literal">RT</tt></td>
<td>Reference title 参考标题</td>
</tr>
<tr><td><tt class="docutils literal">RX</tt></td>
<td>PubMed ID</td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<p>将TRANSFAC文件按原来格式打印出来：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">record</span>
<span class="name">VV</span>  <span class="name">EXAMPLE</span> <span class="name">January</span> <span class="literal number integer">15</span><span class="punctuation">,</span> <span class="literal number integer">2013</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="name">ID</span>  <span class="name">motif1</span>
<span class="name">XX</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">S</span>
<span class="literal number oct">02</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">R</span>
<span class="literal number oct">03</span>      <span class="literal number integer">3</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">1</span>      <span class="name">A</span>
<span class="literal number oct">04</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">05</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">A</span>
<span class="literal number oct">06</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">4</span>      <span class="literal number integer">1</span>      <span class="name">G</span>
<span class="literal number oct">07</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">4</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">08</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="name">T</span>
<span class="literal number integer">09</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">10</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="name">K</span>
<span class="literal number integer">11</span>      <span class="literal number integer">0</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="name">Y</span>
<span class="literal number integer">12</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="literal number integer">1</span>      <span class="name">G</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="name">ID</span>  <span class="name">motif2</span>
<span class="name">XX</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">R</span>
<span class="literal number oct">02</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="name">S</span>
<span class="literal number oct">03</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">04</span>      <span class="literal number integer">3</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">1</span>      <span class="name">A</span>
<span class="literal number oct">05</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">4</span>      <span class="literal number integer">1</span>      <span class="name">G</span>
<span class="literal number oct">06</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">A</span>
<span class="literal number oct">07</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">4</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">08</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">09</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="name">T</span>
<span class="literal number integer">10</span>      <span class="literal number integer">0</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="name">Y</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>通过用字符串形式来截取输出并且保存在文件中，你可以按TRANSFAC的格式导出这些模体：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">text</span> <span class="operator">=</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;mytransfacfile.dat&quot;</span><span class="punctuation">,</span> <span class="literal string">'w'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">text</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
</div>
</div>
<div class="section" id="id158">
<h3>14.1.3  模体写出</h3>
<p>说到导出，我们可以先看看导出函数。以JASPAR <tt class="docutils literal">.pfm</tt> 格式导出模体文件，可以用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;pfm&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">3</span>       <span class="literal number integer">7</span>       <span class="literal number integer">0</span>       <span class="literal number integer">2</span>       <span class="literal number integer">1</span>
<span class="literal number integer">0</span>       <span class="literal number integer">0</span>       <span class="literal number integer">5</span>       <span class="literal number integer">2</span>       <span class="literal number integer">6</span>
<span class="literal number integer">0</span>       <span class="literal number integer">0</span>       <span class="literal number integer">0</span>       <span class="literal number integer">3</span>       <span class="literal number integer">0</span>
<span class="literal number integer">4</span>       <span class="literal number integer">0</span>       <span class="literal number integer">2</span>       <span class="literal number integer">0</span>       <span class="literal number integer">0</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>用类似TRANSFAC的格式导出一个模体：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;transfac&quot;</span><span class="punctuation">)</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">3</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">4</span>      <span class="name">W</span>
<span class="literal number oct">02</span>      <span class="literal number integer">7</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">A</span>
<span class="literal number oct">03</span>      <span class="literal number integer">0</span>      <span class="literal number integer">5</span>      <span class="literal number integer">0</span>      <span class="literal number integer">2</span>      <span class="name">C</span>
<span class="literal number oct">04</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">0</span>      <span class="name">V</span>
<span class="literal number oct">05</span>      <span class="literal number integer">1</span>      <span class="literal number integer">6</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>你可以用 <tt class="docutils literal">motifs.write</tt> 来写出多个模体。这个函数在使用的时候不必担心这些模体来自于TRANSFAC文件。比如：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">two_motifs</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">arnt</span><span class="punctuation">,</span> <span class="name">srf</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">two_motifs</span><span class="punctuation">,</span> <span class="literal string">'transfac'</span><span class="punctuation">)</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">4</span>     <span class="literal number integer">16</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">02</span>     <span class="literal number integer">19</span>      <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>      <span class="name">A</span>
<span class="literal number oct">03</span>      <span class="literal number integer">0</span>     <span class="literal number integer">20</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">04</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>     <span class="literal number integer">20</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number oct">05</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>     <span class="literal number integer">20</span>      <span class="name">T</span>
<span class="literal number oct">06</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>     <span class="literal number integer">20</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="name">P0</span>      <span class="name">A</span>      <span class="name">C</span>      <span class="name">G</span>      <span class="name">T</span>
<span class="literal number oct">01</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>     <span class="literal number integer">39</span>      <span class="literal number integer">4</span>      <span class="name">G</span>
<span class="literal number oct">02</span>      <span class="literal number integer">9</span>     <span class="literal number integer">33</span>      <span class="literal number integer">2</span>      <span class="literal number integer">2</span>      <span class="name">C</span>
<span class="literal number oct">03</span>      <span class="literal number integer">0</span>     <span class="literal number integer">45</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">04</span>      <span class="literal number integer">1</span>     <span class="literal number integer">45</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">C</span>
<span class="literal number oct">05</span>     <span class="literal number integer">32</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>     <span class="literal number integer">13</span>      <span class="name">A</span>
<span class="literal number oct">06</span>      <span class="literal number integer">3</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>     <span class="literal number integer">42</span>      <span class="name">T</span>
<span class="literal number oct">07</span>     <span class="literal number integer">46</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="name">A</span>
<span class="literal number integer">08</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>     <span class="literal number integer">45</span>      <span class="name">T</span>
<span class="literal number integer">09</span>     <span class="literal number integer">43</span>      <span class="literal number integer">0</span>      <span class="literal number integer">0</span>      <span class="literal number integer">3</span>      <span class="name">A</span>
<span class="literal number integer">10</span>     <span class="literal number integer">15</span>      <span class="literal number integer">1</span>      <span class="literal number integer">0</span>     <span class="literal number integer">30</span>      <span class="name">T</span>
<span class="literal number integer">11</span>      <span class="literal number integer">2</span>      <span class="literal number integer">0</span>     <span class="literal number integer">44</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="literal number integer">12</span>      <span class="literal number integer">2</span>      <span class="literal number integer">1</span>     <span class="literal number integer">43</span>      <span class="literal number integer">0</span>      <span class="name">G</span>
<span class="name">XX</span>
<span class="operator">//</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
</div>
<div class="section" id="id159">
<h3>14.1.4  绘制序列标识图</h3>
<p>如果能够联网，我们可以创建一个 <a class="reference external" href="http://weblogo.berkeley.edu">weblogo</a> ：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">weblogo</span><span class="punctuation">(</span><span class="literal string">&quot;Arnt.png&quot;</span><span class="punctuation">)</span>
</pre>
<p>将得到的标识图存储成PNG格式。</p>
</div>
</div>
<div class="section" id="id160">
<h2>14.2  位置权重矩阵</h2>
<p>模体对象的 <tt class="docutils literal">.counts</tt> 属性能够显示在序列上每个位置核苷酸出现的次数。我们可以把这矩阵除以序列中的实例数目来标准化这矩阵，得到每个核苷酸在序列位置上出现概率。我们把这概率看作位置权重矩阵。不过，要知道在字面上，这个术语也可以用来说明位置特异性得分矩阵，这个我们将会在下面讨论。</p>
<p>通常来说，伪计数（pseudocounts）在归一化之前都已经加到每个位置中。这样可以避免在这序列上过度拟合位置权重矩阵以至趋向于模体的实例的有限数量，还可以避免概率为0。向每个位置的核苷酸添加一个固定的伪计数，可以为 <tt class="docutils literal">pseudocounts</tt> 参数指定一个数值：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pwm</span> <span class="operator">=</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="operator">.</span><span class="name">normalize</span><span class="punctuation">(</span><span class="name">pseudocounts</span><span class="operator">=</span><span class="literal number float">0.5</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">pwm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.39</span>   <span class="literal number float">0.83</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.28</span>   <span class="literal number float">0.17</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.61</span>   <span class="literal number float">0.28</span>   <span class="literal number float">0.72</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.39</span>   <span class="literal number float">0.06</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.50</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.28</span>   <span class="literal number float">0.06</span>   <span class="literal number float">0.06</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>另外， <tt class="docutils literal">pseudocounts</tt> 可以利用字典为每个核苷酸指定一个伪计数值。例如，由于在人类基因组中GC含量大概为40%,因此可以选择下面这些伪计数值：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pwm</span> <span class="operator">=</span> <span class="name">m</span><span class="operator">.</span><span class="name">counts</span><span class="operator">.</span><span class="name">normalize</span><span class="punctuation">(</span><span class="name">pseudocounts</span><span class="operator">=</span><span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span><span class="literal number float">0.6</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.4</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.4</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.6</span><span class="punctuation">})</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">pwm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.40</span>   <span class="literal number float">0.84</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.29</span>   <span class="literal number float">0.18</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.60</span>   <span class="literal number float">0.27</span>   <span class="literal number float">0.71</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.38</span>   <span class="literal number float">0.04</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.51</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.29</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.07</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>位置权重矩阵有它自己的方法计算一致序列、反向一致序列和简并一致序列：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pwm</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TACGC'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pwm</span><span class="operator">.</span><span class="name">anticonsensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GGGTG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pwm</span><span class="operator">.</span><span class="name">degenerate_consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'WACNC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>应当注意到由于伪计数的原因，由位置仅重矩阵计算得到的简并一致序列和由模体中实例计算得到的简并一致序列有一点不同：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">degenerate_consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'WACVC'</span><span class="punctuation">,</span> <span class="name">IUPACAmbiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>位置权重矩阵的反向互补矩阵可以直接用 <tt class="docutils literal">pwm</tt> 计算出来：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">rpwm</span> <span class="operator">=</span> <span class="name">pwm</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">rpwm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.29</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.51</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.38</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.04</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.71</span>   <span class="literal number float">0.27</span>   <span class="literal number float">0.60</span>   <span class="literal number float">0.04</span>   <span class="literal number float">0.04</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.18</span>   <span class="literal number float">0.29</span>   <span class="literal number float">0.07</span>   <span class="literal number float">0.84</span>   <span class="literal number float">0.40</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
</div>
<div class="section" id="id161">
<h2>14.3  位置特异性得分矩阵</h2>
<p>使用背景分布和加入伪计数的PWM，很容易就能计算出log-odds比率，提供特定标记的log odds值，这值来自于在这个背景的模体。我们可以用在位置仅重矩阵中 <tt class="docutils literal"><span class="pre">.log-odds()</span></tt> 方法：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pssm</span> <span class="operator">=</span> <span class="name">pwm</span><span class="operator">.</span><span class="name">log_odds</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">pssm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.68</span>   <span class="literal number float">1.76</span>  <span class="operator">-</span><span class="literal number float">1.91</span>   <span class="literal number float">0.21</span>  <span class="operator">-</span><span class="literal number float">0.49</span>
<span class="name">C</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">2.49</span>  <span class="operator">-</span><span class="literal number float">2.49</span>   <span class="literal number float">1.26</span>   <span class="literal number float">0.09</span>   <span class="literal number float">1.51</span>
<span class="name">G</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">2.49</span>  <span class="operator">-</span><span class="literal number float">2.49</span>  <span class="operator">-</span><span class="literal number float">2.49</span>   <span class="literal number float">0.60</span>  <span class="operator">-</span><span class="literal number float">2.49</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">1.03</span>  <span class="operator">-</span><span class="literal number float">1.91</span>   <span class="literal number float">0.21</span>  <span class="operator">-</span><span class="literal number float">1.91</span>  <span class="operator">-</span><span class="literal number float">1.91</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>这时我们可以更经常看到特定标记和背景下的正值和负值。0.0意味着在模体和背景中观察到一个标记有相等的可能性。</p>
<p>上面是假设A,C,G和T在背景中出现的概率是相同的。那在A,C,G和T出现概率不同的情况下，为了计算特定背景下的位置特异性得分矩阵，可以使用 <tt class="docutils literal">background</tt> 参数。例如，在40%GC含量的背景下，可以用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">background</span> <span class="operator">=</span> <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span><span class="literal number float">0.3</span><span class="punctuation">,</span><span class="literal string">'C'</span><span class="punctuation">:</span><span class="literal number float">0.2</span><span class="punctuation">,</span><span class="literal string">'G'</span><span class="punctuation">:</span><span class="literal number float">0.2</span><span class="punctuation">,</span><span class="literal string">'T'</span><span class="punctuation">:</span><span class="literal number float">0.3</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pssm</span> <span class="operator">=</span> <span class="name">pwm</span><span class="operator">.</span><span class="name">log_odds</span><span class="punctuation">(</span><span class="name">background</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">pssm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.42</span>   <span class="literal number float">1.49</span>  <span class="operator">-</span><span class="literal number float">2.17</span>  <span class="operator">-</span><span class="literal number float">0.05</span>  <span class="operator">-</span><span class="literal number float">0.75</span>
<span class="name">C</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">2.17</span>  <span class="operator">-</span><span class="literal number float">2.17</span>   <span class="literal number float">1.58</span>   <span class="literal number float">0.42</span>   <span class="literal number float">1.83</span>
<span class="name">G</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">2.17</span>  <span class="operator">-</span><span class="literal number float">2.17</span>  <span class="operator">-</span><span class="literal number float">2.17</span>   <span class="literal number float">0.92</span>  <span class="operator">-</span><span class="literal number float">2.17</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.77</span>  <span class="operator">-</span><span class="literal number float">2.17</span>  <span class="operator">-</span><span class="literal number float">0.05</span>  <span class="operator">-</span><span class="literal number float">2.17</span>  <span class="operator">-</span><span class="literal number float">2.17</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>从PSSM中得到的最大和最小值被存储在 <tt class="docutils literal">.max</tt> 和 <tt class="docutils literal">.min</tt> 属性中：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">max</span>
<span class="literal number float">6.59</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">min</span>
<span class="operator">-</span><span class="literal number float">10.85</span>
</pre>
<p>在特定背景下计算平均值和标准方差使用 <tt class="docutils literal">.mean</tt> 和 <tt class="docutils literal">.std</tt> 方法。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">mean</span> <span class="operator">=</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">mean</span><span class="punctuation">(</span><span class="name">background</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">std</span> <span class="operator">=</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">std</span><span class="punctuation">(</span><span class="name">background</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;mean = </span><span class="literal string interpol">%0.2f</span><span class="literal string">, standard deviation = </span><span class="literal string interpol">%0.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">mean</span><span class="punctuation">,</span> <span class="name">std</span><span class="punctuation">)</span>
<span class="name">mean</span> <span class="operator">=</span> <span class="literal number float">3.21</span><span class="punctuation">,</span> <span class="name">standard</span> <span class="name">deviation</span> <span class="operator">=</span> <span class="literal number float">2.59</span>
</pre>
<p>如果没有指定特定的背景，就会使用一个统一的背景。因为同KL散度或相对熵的值相同，所以平均值就显得特别重要，并且它也是同背景相比的模体信息含量的测量方法。由于在Biopython中用以2为底的对数来计算log-odds值，信息含量的的单位是bit。</p>
<p><tt class="docutils literal">.reverse_complement</tt>, <tt class="docutils literal">.consensus</tt>, <tt class="docutils literal">.anticonsensus</tt> 和 <tt class="docutils literal">.degenerate_consensus</tt> 方法可以直接对PSSM使用。</p>
</div>
<div class="section" id="id162">
<h2>14.4  搜索实例</h2>
<p>模体最常用的功能就是在序列中的查找它的实例。在这节，我们会用如下的序列作为例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">test_seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TACACTGCATTACAACCCAAGCATTA&quot;</span><span class="punctuation">,</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">)</span>
<span class="literal number integer">26</span>
</pre>
<div class="section" id="id163">
<h3>14.4.1  搜索准确匹配实例</h3>
<p>查找实例最简单的方法就是查找模体实例的准确匹配：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">seq</span> <span class="operator word">in</span> <span class="name">m</span><span class="operator">.</span><span class="name">instances</span><span class="operator">.</span><span class="name">search</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pos</span><span class="punctuation">,</span> <span class="name">seq</span>
<span class="operator">...</span>
<span class="literal number integer">0</span> <span class="name">TACAC</span>
<span class="literal number integer">10</span> <span class="name">TACAA</span>
<span class="literal number integer">13</span> <span class="name">AACCC</span>
</pre>
<p>我们可获得反向互补序列（找到互补链的实例）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">seq</span> <span class="operator word">in</span> <span class="name">r</span><span class="operator">.</span><span class="name">instances</span><span class="operator">.</span><span class="name">search</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pos</span><span class="punctuation">,</span> <span class="name">seq</span>
<span class="operator">...</span>
<span class="literal number integer">6</span> <span class="name">GCATT</span>
<span class="literal number integer">20</span> <span class="name">GCATT</span>
</pre>
</div>
<div class="section" id="pssm">
<h3>14.4.2  用PSSM得分搜索匹配实例</h3>
<p>在模体中很容易找出相应的位置,引起对模体的高log-odds值：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">position</span><span class="punctuation">,</span> <span class="name">score</span> <span class="operator word">in</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">search</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">,</span> <span class="name">threshold</span><span class="operator">=</span><span class="literal number float">3.0</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;Position </span><span class="literal string interpol">%d</span><span class="literal string">: score = </span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">position</span><span class="punctuation">,</span> <span class="name">score</span><span class="punctuation">)</span>
<span class="operator">...</span>
<span class="name">Position</span> <span class="literal number integer">0</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">5.622</span>
<span class="name">Position</span> <span class="operator">-</span><span class="literal number integer">20</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">4.601</span>
<span class="name">Position</span> <span class="literal number integer">10</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">3.037</span>
<span class="name">Position</span> <span class="literal number integer">13</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">5.738</span>
<span class="name">Position</span> <span class="operator">-</span><span class="literal number integer">6</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">4.601</span>
</pre>
<p>负值的位置是指在测试序列的反向链中找到的模体的实例，而且得力于Python的索引。在 <tt class="docutils literal">pos</tt> 的模体实例可以用 <tt class="docutils literal">test_seq[pos:pos+len(m)]</tt> 来定位，不管 <tt class="docutils literal">pos</tt> 值是正还是负。</p>
<p>你可能注意到阀值参数，在这里随意地设为3.0。这里是 <em>log</em><sub>2</sub> ，所以我们现在开始寻找那些在模体中出现概率为背景中出现概率8倍序列。默认的阀值是0.0,在此阀值下，会把所有比背景中出现概率大的模体实例都找出来。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">calculate</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">)</span>
<span class="name">array</span><span class="punctuation">([</span>  <span class="literal number float">5.62230396</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">5.6796999</span> <span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">3.43177247</span><span class="punctuation">,</span>   <span class="literal number float">0.93827754</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">6.84962511</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">2.04066086</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">10.84962463</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">3.65614533</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">0.03370807</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">3.91102552</span><span class="punctuation">,</span>   <span class="literal number float">3.03734159</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">2.14918518</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">0.6016975</span> <span class="punctuation">,</span>   <span class="literal number float">5.7381525</span> <span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">0.50977498</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">3.56422281</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">8.73414803</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">0.09919716</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">0.6016975</span> <span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">2.39429784</span><span class="punctuation">,</span>
       <span class="operator">-</span><span class="literal number float">10.84962463</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">3.65614533</span><span class="punctuation">],</span> <span class="name">dtype</span><span class="operator">=</span><span class="name">float32</span><span class="punctuation">)</span>
</pre>
<p>通常来说，上述是计算PSSM得分的最快方法。这些得分只能由前导链用 <tt class="docutils literal">pssm.calculate</tt> 计算得到。为了得到互补链的PSSM值，你可以利用PSSM的互补矩阵：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">rpssm</span> <span class="operator">=</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">rpssm</span><span class="operator">.</span><span class="name">calculate</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">)</span>
<span class="name">array</span><span class="punctuation">([</span> <span class="operator">-</span><span class="literal number float">9.43458748</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">3.06172252</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">7.18665981</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">7.76216221</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">2.04066086</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">4.26466274</span><span class="punctuation">,</span>   <span class="literal number float">4.60124254</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">4.2480607</span> <span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">8.73414803</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">2.26503372</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">6.49598789</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">5.64668512</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">8.73414803</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">10.84962463</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">4.82356262</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">4.82356262</span><span class="punctuation">,</span>
        <span class="operator">-</span><span class="literal number float">5.64668512</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">8.73414803</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">4.15613794</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">5.6796999</span> <span class="punctuation">,</span>
         <span class="literal number float">4.60124254</span><span class="punctuation">,</span>  <span class="operator">-</span><span class="literal number float">4.2480607</span> <span class="punctuation">],</span> <span class="name">dtype</span><span class="operator">=</span><span class="name">float32</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="id164">
<h3>14.4.3  选择得分阀值</h3>
<p>如果不想刚才那么随意设定一个阀值，你可以探究一下PSSM得分的分布。由于得分的空间分布随着模体长度而成倍增长，我们用一个近似于给定精度值来计算，如此可使计算成本更容易控制：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">distribution</span> <span class="operator">=</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">distribution</span><span class="punctuation">(</span><span class="name">background</span><span class="operator">=</span><span class="name">background</span><span class="punctuation">,</span> <span class="name">precision</span><span class="operator">=</span><span class="literal number integer">10</span><span class="operator">**</span><span class="literal number integer">4</span><span class="punctuation">)</span>
</pre>
<p><tt class="docutils literal">distribution</tt> 对象可以用来决定许多不同的阀值。我们可以指定一个需要的的假阳性率（找到一个由序列在此背景下产生的模体实例的概率）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">threshold</span> <span class="operator">=</span> <span class="name">distribution</span><span class="operator">.</span><span class="name">threshold_fpr</span><span class="punctuation">(</span><span class="literal number float">0.01</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">threshold</span>
<span class="literal number float">4.009</span>
</pre>
<p>或者假阴性率（找不到模体产生的实例概率）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">threshold</span> <span class="operator">=</span> <span class="name">distribution</span><span class="operator">.</span><span class="name">threshold_fnr</span><span class="punctuation">(</span><span class="literal number float">0.1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">threshold</span>
<span class="operator">-</span><span class="literal number float">0.510</span>
</pre>
<p>或者一个阀值（近似），满足假阳性率和假阴性率之间的关系（fnr/fpr≃ <em>t</em>)：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">threshold</span> <span class="operator">=</span> <span class="name">distribution</span><span class="operator">.</span><span class="name">threshold_balanced</span><span class="punctuation">(</span><span class="literal number integer">1000</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">threshold</span>
<span class="literal number float">6.241</span>
</pre>
<p>或者一个阀值能够大体满足假阳性率和信息含量的 −<em>log</em> 值之间的相等关系（与Hertz和Stormo的Patser软件所用的一样）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">threshold</span> <span class="operator">=</span> <span class="name">distribution</span><span class="operator">.</span><span class="name">threshold_patser</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">threshold</span>
<span class="literal number float">0.346</span>
</pre>
<p>比如在我们这个模体中，当以1000比率的平衡阀值查找实例，你可以得到一个让你获得相同结果的阀值（对这个序列来说）。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">threshold</span> <span class="operator">=</span> <span class="name">distribution</span><span class="operator">.</span><span class="name">threshold_fpr</span><span class="punctuation">(</span><span class="literal number float">0.01</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">threshold</span>
<span class="literal number float">4.009</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">position</span><span class="punctuation">,</span> <span class="name">score</span> <span class="operator word">in</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">search</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">,</span><span class="name">threshold</span><span class="operator">=</span><span class="name">threshold</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;Position </span><span class="literal string interpol">%d</span><span class="literal string">: score = </span><span class="literal string interpol">%5.3f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">position</span><span class="punctuation">,</span> <span class="name">score</span><span class="punctuation">)</span>
<span class="operator">...</span>
<span class="name">Position</span> <span class="literal number integer">0</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">5.622</span>
<span class="name">Position</span> <span class="operator">-</span><span class="literal number integer">20</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">4.601</span>
<span class="name">Position</span> <span class="literal number integer">13</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">5.738</span>
<span class="name">Position</span> <span class="operator">-</span><span class="literal number integer">6</span><span class="punctuation">:</span> <span class="name">score</span> <span class="operator">=</span> <span class="literal number float">4.601</span>
</pre>
</div>
</div>
<div class="section" id="id165">
<h2>14.5  模体对象自身相关的位置特异性得分矩阵</h2>
<p>为了更好的利用PSSMs来查找潜在的TFBSs，每个模体都同位置权重矩阵和位置特异性得分矩阵相关联。用Arnt模体来举个例子：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">motifs</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;Arnt.sites&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">'sites'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">counts</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">4.00</span>  <span class="literal number float">19.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="name">C</span><span class="punctuation">:</span>  <span class="literal number float">16.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">20.00</span>   <span class="literal number float">0.00</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pwm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.20</span>   <span class="literal number float">0.95</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.80</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.05</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">1.00</span>   <span class="literal number float">0.00</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">0.32</span>   <span class="literal number float">1.93</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">1.68</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="literal number float">2.00</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="operator">-</span><span class="name">inf</span>  <span class="operator">-</span><span class="literal number float">2.32</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="literal number float">2.00</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="literal number float">2.00</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="operator">-</span><span class="name">inf</span>   <span class="literal number float">2.00</span>   <span class="operator">-</span><span class="name">inf</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>在这出现的负无穷大是由于在频率矩阵中相关项的值为0,并且我们默认使用0作为伪计数：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">letter</span> <span class="operator word">in</span> <span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">: </span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">letter</span><span class="punctuation">,</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pseudocounts</span><span class="punctuation">[</span><span class="name">letter</span><span class="punctuation">])</span>
<span class="operator">...</span>
<span class="name">A</span><span class="punctuation">:</span> <span class="literal number float">0.00</span>
<span class="name">C</span><span class="punctuation">:</span> <span class="literal number float">0.00</span>
<span class="name">G</span><span class="punctuation">:</span> <span class="literal number float">0.00</span>
<span class="name">T</span><span class="punctuation">:</span> <span class="literal number float">0.00</span>
</pre>
<p>如果你更改了 <tt class="docutils literal">.pseudocouts</tt> 属性，那么位置频率矩阵和位置特异性得分矩阵就都会自动重新计算：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pseudocounts</span> <span class="operator">=</span> <span class="literal number float">3.0</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">letter</span> <span class="operator word">in</span> <span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">: </span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">letter</span><span class="punctuation">,</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pseudocounts</span><span class="punctuation">[</span><span class="name">letter</span><span class="punctuation">])</span>
<span class="operator">...</span>
<span class="name">A</span><span class="punctuation">:</span> <span class="literal number float">3.00</span>
<span class="name">C</span><span class="punctuation">:</span> <span class="literal number float">3.00</span>
<span class="name">G</span><span class="punctuation">:</span> <span class="literal number float">3.00</span>
<span class="name">T</span><span class="punctuation">:</span> <span class="literal number float">3.00</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pwm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.22</span>   <span class="literal number float">0.69</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.59</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.72</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.12</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.72</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.72</span>
<span class="name">T</span><span class="punctuation">:</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.09</span>   <span class="literal number float">0.72</span>   <span class="literal number float">0.09</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">0.19</span>   <span class="literal number float">1.46</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">1.25</span>  <span class="operator">-</span><span class="literal number float">1.42</span>   <span class="literal number float">1.52</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>
<span class="name">G</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.00</span>  <span class="operator">-</span><span class="literal number float">1.42</span>   <span class="literal number float">1.52</span>  <span class="operator">-</span><span class="literal number float">1.42</span>   <span class="literal number float">1.52</span>
<span class="name">T</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>  <span class="operator">-</span><span class="literal number float">1.42</span>   <span class="literal number float">1.52</span>  <span class="operator">-</span><span class="literal number float">1.42</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>如果你想对4个核苷酸使用不同的伪计数，可以使用字典来设定4个核苷酸的 <tt class="docutils literal">pseudocounts</tt> 。把 <tt class="docutils literal">motif.pseudocounts</tt> 设为 <tt class="docutils literal">None</tt> 会让伪计数重置为0的默认值。</p>
<p>位置特异性得分矩阵依赖于一个默认均一的背景分布：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">letter</span> <span class="operator word">in</span> <span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">: </span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">letter</span><span class="punctuation">,</span> <span class="name">motif</span><span class="operator">.</span><span class="name">background</span><span class="punctuation">[</span><span class="name">letter</span><span class="punctuation">])</span>
<span class="operator">...</span>
<span class="name">A</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
<span class="name">C</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
<span class="name">G</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
<span class="name">T</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
</pre>
<p>同样，如果你更改了背景分布，位置特异性得分矩阵也会重新计算：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">background</span> <span class="operator">=</span> <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.2</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.3</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.3</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.2</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.13</span>   <span class="literal number float">1.78</span>  <span class="operator">-</span><span class="literal number float">1.09</span>  <span class="operator">-</span><span class="literal number float">1.09</span>  <span class="operator">-</span><span class="literal number float">1.09</span>  <span class="operator">-</span><span class="literal number float">1.09</span>
<span class="name">C</span><span class="punctuation">:</span>   <span class="literal number float">0.98</span>  <span class="operator">-</span><span class="literal number float">1.68</span>   <span class="literal number float">1.26</span>  <span class="operator">-</span><span class="literal number float">1.68</span>  <span class="operator">-</span><span class="literal number float">1.68</span>  <span class="operator">-</span><span class="literal number float">1.68</span>
<span class="name">G</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">1.68</span>  <span class="operator">-</span><span class="literal number float">1.26</span>  <span class="operator">-</span><span class="literal number float">1.68</span>   <span class="literal number float">1.26</span>  <span class="operator">-</span><span class="literal number float">1.68</span>   <span class="literal number float">1.26</span>
<span class="name">T</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">1.09</span>  <span class="operator">-</span><span class="literal number float">1.09</span>  <span class="operator">-</span><span class="literal number float">1.09</span>  <span class="operator">-</span><span class="literal number float">1.09</span>   <span class="literal number float">1.85</span>  <span class="operator">-</span><span class="literal number float">1.09</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>把 <tt class="docutils literal">motif.backgroud</tt> 设为 <tt class="docutils literal">None</tt> 后会将其重置为均一的分布。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">background</span> <span class="operator">=</span> <span class="name builtin pseudo">None</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">letter</span> <span class="operator word">in</span> <span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">: </span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">letter</span><span class="punctuation">,</span> <span class="name">motif</span><span class="operator">.</span><span class="name">background</span><span class="punctuation">[</span><span class="name">letter</span><span class="punctuation">])</span>
<span class="operator">...</span>
<span class="name">A</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
<span class="name">C</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
<span class="name">G</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
<span class="name">T</span><span class="punctuation">:</span> <span class="literal number float">0.25</span>
</pre>
<p>如果你把 <tt class="docutils literal">motif.background</tt> 设为一个单一值，这个值将会被看成是GC含量：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motif</span><span class="operator">.</span><span class="name">background</span> <span class="operator">=</span> <span class="literal number float">0.8</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">letter</span> <span class="operator word">in</span> <span class="literal string">&quot;ACGT&quot;</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">: </span><span class="literal string interpol">%4.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">letter</span><span class="punctuation">,</span> <span class="name">motif</span><span class="operator">.</span><span class="name">background</span><span class="punctuation">[</span><span class="name">letter</span><span class="punctuation">])</span>
<span class="operator">...</span>
<span class="name">A</span><span class="punctuation">:</span> <span class="literal number float">0.10</span>
<span class="name">C</span><span class="punctuation">:</span> <span class="literal number float">0.40</span>
<span class="name">G</span><span class="punctuation">:</span> <span class="literal number float">0.40</span>
<span class="name">T</span><span class="punctuation">:</span> <span class="literal number float">0.10</span>
</pre>
<p>应当注意到你能够在当前计算背景下计算PSSM的平均值：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span><span class="operator">.</span><span class="name">mean</span><span class="punctuation">(</span><span class="name">motif</span><span class="operator">.</span><span class="name">background</span><span class="punctuation">)</span>
<span class="literal number float">4.703928</span>
</pre>
<p>它的标准方差也是一样：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span><span class="operator">.</span><span class="name">std</span><span class="punctuation">(</span><span class="name">motif</span><span class="operator">.</span><span class="name">background</span><span class="punctuation">)</span>
<span class="literal number float">3.290900</span>
</pre>
<p>和它的分布：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">distribution</span> <span class="operator">=</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span><span class="operator">.</span><span class="name">distribution</span><span class="punctuation">(</span><span class="name">background</span><span class="operator">=</span><span class="name">motif</span><span class="operator">.</span><span class="name">background</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">threshold</span> <span class="operator">=</span> <span class="name">distribution</span><span class="operator">.</span><span class="name">threshold_fpr</span><span class="punctuation">(</span><span class="literal number float">0.01</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">threshold</span>
<span class="literal number float">3.854375</span>
</pre>
<p>请注意，每当你调用 <tt class="docutils literal">motif.pwm</tt> 或 <tt class="docutils literal">motif.pssm</tt> ，位置仅重矩阵和位置特异性得分矩阵都会重新计算。如果看重速度并且需要重复用到PWM或PSSM时，你可以把他们保存成变量，如下所示：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pssm</span> <span class="operator">=</span> <span class="name">motif</span><span class="operator">.</span><span class="name">pssm</span>
</pre>
</div>
<div class="section" id="id166">
<h2>14.6  模体比较</h2>
<p>当有多个模体时，我们就会想去比较它们。</p>
<p>在我们开始比较之前，应当要指出模体的边界通常比较模糊。这也就是说我们需要比较不同长度的模体，因此这些比较也要涉及到相关的比对。所以我们需要考虑两个东西：</p>
<ul class="simple">
<li>模体比对</li>
<li>比较比对后模体的相关函数</li>
</ul>
<p>为了比对模体，我们使用PSSMs的不含间隔的比对，并且用0来代替矩阵开始和结束位置缺失的列。这说明我们能够有效地利用背景分布来代替PSSM中缺失的列。距离函数然后可以返回模体间最小的距离，以及比对中相应的偏移量。</p>
<p>举个例子，先导入和测试模体 <tt class="docutils literal">m</tt> 相似的模体：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m_reb1</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;REB1.pfm&quot;</span><span class="punctuation">),</span> <span class="literal string">&quot;pfm&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m_reb1</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GTTACCCGG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m_reb1</span><span class="operator">.</span><span class="name">counts</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>      <span class="literal number integer">6</span>      <span class="literal number integer">7</span>      <span class="literal number integer">8</span>
<span class="name">A</span><span class="punctuation">:</span>  <span class="literal number float">30.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span> <span class="literal number float">100.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">15.00</span>
<span class="name">C</span><span class="punctuation">:</span>  <span class="literal number float">10.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span> <span class="literal number float">100.00</span> <span class="literal number float">100.00</span> <span class="literal number float">100.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">15.00</span>
<span class="name">G</span><span class="punctuation">:</span>  <span class="literal number float">50.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">60.00</span>  <span class="literal number float">55.00</span>
<span class="name">T</span><span class="punctuation">:</span>  <span class="literal number float">10.00</span> <span class="literal number float">100.00</span> <span class="literal number float">100.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>   <span class="literal number float">0.00</span>  <span class="literal number float">40.00</span>  <span class="literal number float">15.00</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>为了让模体能够进行相互比较，选择和模体 <tt class="docutils literal">m</tt> 相同伪计数和背景值：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m_reb1</span><span class="operator">.</span><span class="name">pseudocounts</span> <span class="operator">=</span> <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span><span class="literal number float">0.6</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.4</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.4</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.6</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m_reb1</span><span class="operator">.</span><span class="name">background</span> <span class="operator">=</span> <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span><span class="literal number float">0.3</span><span class="punctuation">,</span><span class="literal string">'C'</span><span class="punctuation">:</span><span class="literal number float">0.2</span><span class="punctuation">,</span><span class="literal string">'G'</span><span class="punctuation">:</span><span class="literal number float">0.2</span><span class="punctuation">,</span><span class="literal string">'T'</span><span class="punctuation">:</span><span class="literal number float">0.3</span><span class="punctuation">}</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">pssm_reb1</span> <span class="operator">=</span> <span class="name">m_reb1</span><span class="operator">.</span><span class="name">pssm</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">pssm_reb1</span>
        <span class="literal number integer">0</span>      <span class="literal number integer">1</span>      <span class="literal number integer">2</span>      <span class="literal number integer">3</span>      <span class="literal number integer">4</span>      <span class="literal number integer">5</span>      <span class="literal number integer">6</span>      <span class="literal number integer">7</span>      <span class="literal number integer">8</span>
<span class="name">A</span><span class="punctuation">:</span>   <span class="literal number float">0.00</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>   <span class="literal number float">1.72</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">0.97</span>
<span class="name">C</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">0.97</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>   <span class="literal number float">2.30</span>   <span class="literal number float">2.30</span>   <span class="literal number float">2.30</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">0.41</span>
<span class="name">G</span><span class="punctuation">:</span>   <span class="literal number float">1.30</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>   <span class="literal number float">1.57</span>   <span class="literal number float">1.44</span>
<span class="name">T</span><span class="punctuation">:</span>  <span class="operator">-</span><span class="literal number float">1.53</span>   <span class="literal number float">1.72</span>   <span class="literal number float">1.72</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>  <span class="operator">-</span><span class="literal number float">5.67</span>   <span class="literal number float">0.41</span>  <span class="operator">-</span><span class="literal number float">0.97</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
</pre>
<p>我们将用皮尔逊相关（Pearson correlation）来比较这些模体。由于我们想要让它偏向于一个距离长度，我们实际上取1−<em>r</em> ，其中 <em>r</em> 是皮尔逊相关系数（Pearson correlation coefficient，PCC）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">distance</span><span class="punctuation">,</span> <span class="name">offset</span> <span class="operator">=</span> <span class="name">pssm</span><span class="operator">.</span><span class="name">dist_pearson</span><span class="punctuation">(</span><span class="name">pssm_reb1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;distance = </span><span class="literal string interpol">%5.3g</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">distance</span>
<span class="name">distance</span> <span class="operator">=</span> <span class="literal number float">0.239</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">offset</span>
<span class="operator">-</span><span class="literal number integer">2</span>
</pre>
<p>这意味着模体 <tt class="docutils literal">m</tt> 和 <tt class="docutils literal">m_reb1</tt> 间最佳PCC可以从下面的比对中获得：</p>
<pre class="code python literal-block">
<span class="name">m</span><span class="punctuation">:</span>      <span class="name">bbTACGCbb</span>
<span class="name">m_reb1</span><span class="punctuation">:</span> <span class="name">GTTACCCGG</span>
</pre>
<p>其中 <tt class="docutils literal">b</tt> 代表背景分布。PCC值大概为1−0.239=0.761。</p>
</div>
<div class="section" id="de-novo">
<h2>14.7  查找 <em>De novo</em> 模体</h2>
<p>如今，Biopython对 <em>De novo</em> 模体查找的支持是有限的。也就是说，我们支持AlignAce和MEME的运行和读取。由于模体查找工具如雨后春笋般出现，所以很欢迎新的分析程序加入进来。</p>
<div class="section" id="id167">
<h3>14.7.1  MEME</h3>
<p>假设用MEME以你喜欢的参数设置来跑序列，并把结果保存在文件 <tt class="docutils literal">meme.out</tt> 中。你可以用以下的命令来得到MEME输出的模体：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">motifs</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;meme.out&quot;</span><span class="punctuation">),</span> <span class="literal string">&quot;meme&quot;</span><span class="punctuation">)</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span>
<span class="punctuation">[</span><span class="operator">&lt;</span><span class="name">Bio</span><span class="operator">.</span><span class="name">motifs</span><span class="operator">.</span><span class="name">meme</span><span class="operator">.</span><span class="name">Motif</span> <span class="name builtin">object</span> <span class="name">at</span> <span class="literal number hex">0xc356b0</span><span class="operator">&gt;</span><span class="punctuation">]</span>
</pre>
<p>除了最想要的一系列模体外，结果中还包含了很多有用的信息，可以通过那些一目了然的属性名获得：</p>
<ul class="simple">
<li><tt class="docutils literal">.alphabet</tt></li>
<li><tt class="docutils literal">.datafile</tt></li>
<li><tt class="docutils literal">.sequence_names</tt></li>
<li><tt class="docutils literal">.version</tt></li>
<li><tt class="docutils literal">.command</tt></li>
</ul>
<p>由MEME解析得到的模体可以像平常的模体对象（有实例）一样处理，它们也提供了一些额外的功能，可以为实例增加额外的信息。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CTCAATCGTA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">sequence_name</span>
<span class="literal string">'SEQ10;'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">start</span>
<span class="literal number integer">3</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">strand</span>
<span class="literal string">'+'</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">pvalue</span>
<span class="literal number float">8.71e-07</span>
</pre>
</div>
<div class="section" id="alignace">
<h3>14.7.2  AlignAce</h3>
<p>我们可以用AlignACE程序实现类似的效果。假如，你把结果保存在 <tt class="docutils literal">alignace.out</tt> 文件中。你可以用下面的代码读取结果：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">motifs</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsA</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;alignace.out&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;alignace&quot;</span><span class="punctuation">)</span>
</pre>
<p>同样，你的模体也和正常的模体对象有相同的属性：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsA</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TCTACGATTGAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>事实上，你甚至可以观察到，AlignAce找到了一个和MEME非常相似的模体。下面只是MEME模体互补链的一个较长版本：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">consensus</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TACGATTGAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>如果你的机器上安装了AlignAce，你可以直接从Biopython中运行AlignAce。下面就是一个如何运行AlignAce的简单例子（其他参数可以用关键字参数来调用）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">command</span><span class="operator">=</span><span class="literal string">&quot;/opt/bin/AlignACE&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">input_file</span><span class="operator">=</span><span class="literal string">&quot;test.fa&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.motifs.applications</span> <span class="keyword namespace">import</span> <span class="name">AlignAceCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cmd</span> <span class="operator">=</span> <span class="name">AlignAceCommandline</span><span class="punctuation">(</span><span class="name">cmd</span><span class="operator">=</span><span class="name">command</span><span class="punctuation">,</span><span class="name builtin">input</span><span class="operator">=</span><span class="name">input_file</span><span class="punctuation">,</span><span class="name">gcback</span><span class="operator">=</span><span class="literal number float">0.6</span><span class="punctuation">,</span><span class="name">numcols</span><span class="operator">=</span><span class="literal number integer">10</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span><span class="name">stderr</span><span class="operator">=</span> <span class="name">cmd</span><span class="punctuation">()</span>
</pre>
<p>由于AlignAce把所有的结果输出到标准输出，所以你可以通过读取结果的第一部分来获得模体：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifs</span> <span class="operator">=</span> <span class="name">motifs</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">stdout</span><span class="punctuation">,</span><span class="literal string">&quot;alignace&quot;</span><span class="punctuation">)</span>
</pre>
</div>
</div>
<div class="section" id="id168">
<h2>14.8  相关链接</h2>
<ul class="simple">
<li><a class="reference external" href="http://en.wikipedia.org/wiki/Sequence_motif">Sequence motif</a> in
wikipedia</li>
<li><a class="reference external" href="http://en.wikipedia.org/wiki/Position_weight_matrix">PWM</a> in
wikipedia</li>
<li><a class="reference external" href="http://en.wikipedia.org/wiki/Consensus_sequence">Consensus
sequence</a> in
wikipedia</li>
<li><a class="reference external" href="http://bio.cs.washington.edu/assessment/">Comparison of different motif finding
programs</a></li>
</ul>
</div>
<div class="section" id="bio-motif">
<h2>14.9  旧版Bio.Motif模块</h2>
<p>本章剩下部分将介绍Biopython 1.61版本前的 <tt class="docutils literal">Bio.Motifs</tt> 模块，该模块取代了Biopython 1.50版本中基于两个早期Biopython模块—— <tt class="docutils literal">Bio.AlignAce</tt> 和 <tt class="docutils literal">Bio.MEME</tt> 的 <tt class="docutils literal">Bio.Motif</tt> 模块。</p>
<p>为了平滑的过渡，早期的 <tt class="docutils literal">Bio.Motif</tt> 模块将会和它的取代者 <tt class="docutils literal">Bio.Motifs</tt> 一同维护到至少发行两个版本，并且持续至少一年。</p>
<div class="section" id="id169">
<h3>14.9.1  模体对象</h3>
<p>由于我们对模体分析感兴趣，不过让我们首先看看 <tt class="docutils literal">Motif</tt> 对象。第一步要先导入模体库：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Motif</span>
</pre>
<p>然后可以开始创建第一个模体对象。创建一个DNA模体：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span> <span class="operator">=</span> <span class="name">Motif</span><span class="operator">.</span><span class="name">Motif</span><span class="punctuation">(</span><span class="name">alphabet</span><span class="operator">=</span><span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
</pre>
<p>现在这里面什么也没有，往新建的模体加入一些序列：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">add_instance</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TATAA&quot;</span><span class="punctuation">,</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">add_instance</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TATTA&quot;</span><span class="punctuation">,</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">add_instance</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TATAA&quot;</span><span class="punctuation">,</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">add_instance</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TATAA&quot;</span><span class="punctuation">,</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">))</span>
</pre>
<p>现在我们有了一个完整的 <tt class="docutils literal">Motif</tt> 实例，我们可以试着从中获取一些基本信息。先看看长度和一致序列：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">m</span><span class="punctuation">)</span>
<span class="literal number integer">5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">consensus</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TATAA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>对于DNA模体，我们还可以获得一个模体的反向互补序列：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">consensus</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TTATA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name">m</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">:</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">i</span>
<span class="name">TTATA</span>
<span class="name">TAATA</span>
<span class="name">TTATA</span>
<span class="name">TTATA</span>
</pre>
<p>我们也可以简单的调取模体的信息容量：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%0.2f</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">m</span><span class="operator">.</span><span class="name">ic</span><span class="punctuation">()</span>
<span class="literal number float">5.27</span>
</pre>
<p>这给我们提供了模体中信息容量的比特数，这指出和背景有多少不同。</p>
<p>展示模体最常用的就是PWM（位置仅重矩阵）。它概括了在模体上任意位置出现一个符号（这里指核苷酸）的概率。这个可以用 <tt class="docutils literal">.pwm()</tt> 方法来计算：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">pwm</span><span class="punctuation">()</span>
<span class="punctuation">[{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.85</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.85</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.85</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.65</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.25</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.85</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.05</span><span class="punctuation">}]</span>
</pre>
<p>模体的PWM中的概率是基于实例中的计数，但我们发现，虽然模体中没有出现G和C，可是它们的概率仍然是非0的。这主要是因为有伪计数的存在，简单地说，就是一种常用的方式来承认我们认知的不完备以及为了避免使用0进行对数运算而出现的技术问题。</p>
<p>我可以调整伪计数添加到模体对象两个属性的方式。 <tt class="docutils literal">.background</tt> 是我们假设代表背景分布的所有字符的概率分布，是非模体序列（通常基于各自基因组的GC含量）。在模体创建的时候，就默认的设置为一个统一分布：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">background</span>
<span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.25</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.25</span><span class="punctuation">,</span> <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.25</span><span class="punctuation">,</span> <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="literal number float">0.25</span><span class="punctuation">}</span>
</pre>
<p>另一个就是 <tt class="docutils literal">.beta</tt> ，这个参数可以说明我们应该给伪计数设定为何值。默认设定为1.0。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">beta</span>
<span class="literal number float">1.0</span>
</pre>
<p>所以输入伪计数的总量等于一个实例的输入总量。</p>
<p>使用背景分布和附加了伪计数的pwm，可以很容易的计算log-odd比率，这告诉我们在背景下，一个来自模体特定碱基的log-odd值。我们可以使用 <tt class="docutils literal">.log_odds()</tt> 方法：</p>
<pre class="code python literal-block">
 <span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">log_odds</span><span class="punctuation">()</span>
<span class="punctuation">[{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">1.7655347463629771</span><span class="punctuation">,</span>
  <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">1.7655347463629771</span><span class="punctuation">,</span>
  <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">1.7655347463629771</span><span class="punctuation">,</span>
  <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">1.3785116232537298</span><span class="punctuation">,</span>
  <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="literal number float">0.0</span><span class="punctuation">,</span>
  <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">},</span>
 <span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">1.7655347463629771</span><span class="punctuation">,</span>
  <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">,</span>
  <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.3219280948873622</span><span class="punctuation">}</span>
<span class="punctuation">]</span>
</pre>
<p>此处，我们可以看出如果模体中的碱基比背景中出现频率更高，其值为正值，反之则为负值。0.0说明在背景和模体中出现的概率是相同的（如第二个位置的“T”）。</p>
<div class="section" id="id170">
<h4>14.9.1.1  模体读写</h4>
<p>手动从实例创建一个模体确实没什么技术含量，所以很有必要有一些读写功能来读取和写出模体。对于如何存储模体还没有一个固定的标准，但是有一些格式比其他格式更流行。这些格式的主要区别在于模体的创建是基于实例还是一些PWM矩阵。其中一个最流行的模体数据库就是 <a class="reference external" href="http://jaspar.genereg.net">JASPAR</a> ，该数据库保存了上述两种类型的格式，所以让我们看看是如何从实例中导入JASPAR模体：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Motif</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">arnt</span> <span class="operator">=</span> <span class="name">Motif</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;Arnt.sites&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;jaspar-sites&quot;</span><span class="punctuation">)</span>
</pre>
<p>从一个计数矩阵中导入：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">srf</span> <span class="operator">=</span> <span class="name">Motif</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;SRF.pfm&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;jaspar-pfm&quot;</span><span class="punctuation">)</span>
</pre>
<p><tt class="docutils literal">arnt</tt> 和 <tt class="docutils literal">srf</tt> 模体可以为我们做相同的事情，但是它们使用不同的内部表现形式来展现模体。我们可以用 <tt class="docutils literal">has_counts</tt> 和 <tt class="docutils literal">has_instances</tt> 属性来区分它们：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">has_instances</span>
<span class="name builtin pseudo">True</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">srf</span><span class="operator">.</span><span class="name">has_instances</span>
<span class="name builtin pseudo">False</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">srf</span><span class="operator">.</span><span class="name">has_counts</span>
<span class="name builtin pseudo">True</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">srf</span><span class="operator">.</span><span class="name">counts</span>
<span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">9</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">32</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">46</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">43</span><span class="punctuation">,</span> <span class="literal number integer">15</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">],</span>
 <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">33</span><span class="punctuation">,</span> <span class="literal number integer">45</span><span class="punctuation">,</span> <span class="literal number integer">45</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">],</span>
 <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">39</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">44</span><span class="punctuation">,</span> <span class="literal number integer">43</span><span class="punctuation">],</span>
 <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">4</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">13</span><span class="punctuation">,</span> <span class="literal number integer">42</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">45</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">,</span> <span class="literal number integer">30</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">]}</span>
</pre>
<p>对于模体的不同表现形式，可以用转换功能来实现相互转换：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">make_counts_from_instances</span><span class="punctuation">()</span>
<span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">8</span><span class="punctuation">,</span> <span class="literal number integer">38</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">],</span>
 <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">32</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">],</span>
 <span class="literal string">'G'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">40</span><span class="punctuation">],</span>
 <span class="literal string">'T'</span><span class="punctuation">:</span> <span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">]}</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="name">srf</span><span class="operator">.</span><span class="name">make_instances_from_counts</span><span class="punctuation">()</span>
<span class="punctuation">[</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GGGAAAAAAAGG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()),</span>
 <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GGCCAAATAAGG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()),</span>
 <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GACCAAATAAGG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">()),</span>
<span class="operator">....</span>
</pre>
<p>在这里需要注意的是 <tt class="docutils literal">make_instances_from_counts()</tt> 方法创建的是假实例，因为按照相同的pwm能够得到许多不同的实例，所以不能反过来重建原来的矩阵。不过这对我们利用PWM来展现模体没有什么影响，但从基于计数的模体中导出实例时要小心。</p>
<p>说到导出，让我们看看导出函数。我们可以按fasta的格式导出：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;</span><span class="name">instance0</span>
<span class="name">TATAA</span>
<span class="operator">&gt;</span><span class="name">instance1</span>
<span class="name">TATTA</span>
<span class="operator">&gt;</span><span class="name">instance2</span>
<span class="name">TATAA</span>
<span class="operator">&gt;</span><span class="name">instance3</span>
<span class="name">TATAA</span>
</pre>
<p>或者是按TRANSFAC样的矩阵格式导出（能被一些处理软件识别）</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">m</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;transfac&quot;</span><span class="punctuation">)</span>
<span class="name">XX</span>
<span class="name">TY</span> <span class="name">Motif</span>
<span class="name">ID</span>
<span class="name">BF</span> <span class="name">undef</span>
<span class="name">P0</span> <span class="name">G</span> <span class="name">A</span> <span class="name">T</span> <span class="name">C</span>
<span class="literal number oct">01</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span> <span class="literal number integer">4</span> <span class="literal number integer">0</span>
<span class="literal number oct">02</span> <span class="literal number integer">0</span> <span class="literal number integer">4</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span>
<span class="literal number oct">03</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span> <span class="literal number integer">4</span> <span class="literal number integer">0</span>
<span class="literal number oct">04</span> <span class="literal number integer">0</span> <span class="literal number integer">3</span> <span class="literal number integer">1</span> <span class="literal number integer">0</span>
<span class="literal number oct">05</span> <span class="literal number integer">0</span> <span class="literal number integer">4</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span>
<span class="name">XX</span>
</pre>
<p>最后，如果能够联网，我们可以创建一个 <a class="reference external" href="http://weblogo.berkeley.edu">weblogo</a> ：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">arnt</span><span class="operator">.</span><span class="name">weblogo</span><span class="punctuation">(</span><span class="literal string">&quot;Arnt.png&quot;</span><span class="punctuation">)</span>
</pre>
<p>我们可以把得到的标识图以png的格式保存到特定的文件中。</p>
</div>
</div>
<div class="section" id="id171">
<h3>14.9.2  查找实例</h3>
<p>模体中最常用的就是在一些序列中查找实例。为解释这部分，我们将手动创建一个如下的序列：</p>
<pre class="code python literal-block">
<span class="name">test_seq</span><span class="operator">=</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;TATGATGTAGTATAATATAATTATAA&quot;</span><span class="punctuation">,</span><span class="name">m</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">)</span>
</pre>
<p>查找实例最简单的方法就是在模体中查找具体匹配的实例：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">seq</span> <span class="operator word">in</span> <span class="name">m</span><span class="operator">.</span><span class="name">search_instances</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">seq</span><span class="operator">.</span><span class="name">tostring</span><span class="punctuation">()</span>
<span class="operator">...</span>
<span class="literal number integer">10</span> <span class="name">TATAA</span>
<span class="literal number integer">15</span> <span class="name">TATAA</span>
<span class="literal number integer">21</span> <span class="name">TATAA</span>
</pre>
<p>对于互补序列，也可以用相同的方法（为了找到互补链上的实例）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">seq</span> <span class="operator word">in</span> <span class="name">m</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">search_instances</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">seq</span><span class="operator">.</span><span class="name">tostring</span><span class="punctuation">()</span>
<span class="operator">...</span>
<span class="literal number integer">12</span> <span class="name">TAATA</span>
<span class="literal number integer">20</span> <span class="name">TTATA</span>
</pre>
<p>提高模体的log-odds值能让查为位置更加简单:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">score</span> <span class="operator word">in</span> <span class="name">m</span><span class="operator">.</span><span class="name">search_pwm</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">,</span><span class="name">threshold</span><span class="operator">=</span><span class="literal number float">5.0</span><span class="punctuation">):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">score</span>
<span class="operator">...</span>
<span class="literal number integer">10</span> <span class="literal number float">8.44065060871</span>
<span class="operator">-</span><span class="literal number integer">12</span> <span class="literal number float">7.06213898545</span>
<span class="literal number integer">15</span> <span class="literal number float">8.44065060871</span>
<span class="operator">-</span><span class="literal number integer">20</span> <span class="literal number float">8.44065060871</span>
<span class="literal number integer">21</span> <span class="literal number float">8.44065060871</span>
</pre>
<p>你可能注意到阀值参数，在这里随意地设为5.0。按 <em>log</em><sub>2</sub> 来算，我们应当查找那些在模体中出现概率为背景中出现概率32倍的序列。默认的阀值是0.0,在些阀值下，会把所有比背景中出现概率大的模体实例都找出来。</p>
<p>如果不想那么随意的选择一个阀值，你可以研究一下 <tt class="docutils literal">Motif.score_distribution</tt> 类，它为模体提供一个相应的得分分布。由于得分的空间分布随着模体长度而成倍增长，我们正用一个近似于给定精度值计算，从而使计算成本易于控制：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sd</span> <span class="operator">=</span> <span class="name">Motif</span><span class="operator">.</span><span class="name">score_distribution</span><span class="punctuation">(</span><span class="name">m</span><span class="punctuation">,</span><span class="name">precision</span><span class="operator">=</span><span class="literal number integer">10</span><span class="operator">**</span><span class="literal number integer">4</span><span class="punctuation">)</span>
</pre>
<p>上面那个sd对象可以用来决定许多不同的阀值。</p>
<p>我们可以设定一个需要的假阳性率（找到一个由此序列在这个背景下产生的模体实例的概率）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sd</span><span class="operator">.</span><span class="name">threshold_fpr</span><span class="punctuation">(</span><span class="literal number float">0.01</span><span class="punctuation">)</span>
<span class="literal number float">4.3535838726139886</span>
</pre>
<p>或者假阴性率（找不到模体产生的实例的概率）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sd</span><span class="operator">.</span><span class="name">threshold_fnr</span><span class="punctuation">(</span><span class="literal number float">0.1</span><span class="punctuation">)</span>
<span class="literal number float">0.26651713652234044</span>
</pre>
<p>或者一个阀值（近似），满足假阳性率和假阴性率之间的关系（fnr/fpr≃ <em>t</em>)：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sd</span><span class="operator">.</span><span class="name">threshold_balanced</span><span class="punctuation">(</span><span class="literal number integer">1000</span><span class="punctuation">)</span>
<span class="literal number float">8.4406506087056368</span>
</pre>
<p>或者一个阀值能够大体满足假阳性率和信息含量的 −<em>log</em> 值之间的相等关系（像Hertz和Stormo的Patser软件所用的一样）：</p>
<p>在我们这个例子中，当以1000比率的平衡阀值查找实例时，你可以得到一个让你获得相同结果（对于这个序列来说）的阀值：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">score</span> <span class="operator word">in</span> <span class="name">m</span><span class="operator">.</span><span class="name">search_pwm</span><span class="punctuation">(</span><span class="name">test_seq</span><span class="punctuation">,</span><span class="name">threshold</span><span class="operator">=</span><span class="name">sd</span><span class="operator">.</span><span class="name">threshold_balanced</span><span class="punctuation">(</span><span class="literal number integer">1000</span><span class="punctuation">)):</span>
<span class="operator">...</span>     <span class="keyword">print</span> <span class="name">pos</span><span class="punctuation">,</span><span class="name">score</span>
<span class="operator">...</span>
<span class="literal number integer">10</span> <span class="literal number float">8.44065060871</span>
<span class="literal number integer">15</span> <span class="literal number float">8.44065060871</span>
<span class="operator">-</span><span class="literal number integer">20</span> <span class="literal number float">8.44065060871</span>
<span class="literal number integer">21</span> <span class="literal number float">8.44065060871</span>
</pre>
</div>
<div class="section" id="id172">
<h3>14.9.3  模体比较</h3>
<p>当有多个模体时，我们就会想去比较他们。对此， <tt class="docutils literal">Bio.Motif</tt> 有三种不同的方法来进行模体比较。</p>
<p>在我们开始比较之前，应当指出模体的边界通常是相当模糊的。也就是说我们经常需要比较不同长度的模体，因此这些比较涉及到相关的比对。所以我们需要考虑两个要点：</p>
<ul class="simple">
<li>模体比对</li>
<li>比较比对后模体的相关函数</li>
</ul>
<p>在 <tt class="docutils literal">Bio.Motif</tt> 中有三种比较方法，这些方法都是基于来源于模体比对的想法，而采用不同方式。简单来说，我们使用不含间隔的PSSMs比对，并且用0来代替矩阵同背景相比，在开始和结束位置出现缺失的列。这三种比较方法都可以解释成距离估量，但是只有一个（ <tt class="docutils literal"><span class="pre">dist——dpq</span></tt> ）满足三角不等式。这些方法都返回距离的最小值和模体相应的偏移量。</p>
<p>为了展示这些比较功能是如何实现的，导入和测试模体 <tt class="docutils literal">m</tt> 相似的其他模体：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ubx</span><span class="operator">=</span><span class="name">Motif</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;Ubx.pfm&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;jaspar-pfm&quot;</span><span class="punctuation">)</span>
<span class="operator">&lt;</span><span class="name">Bio</span><span class="operator">.</span><span class="name">Motif</span><span class="operator">.</span><span class="name">Motif</span><span class="operator">.</span><span class="name">Motif</span> <span class="name builtin">object</span> <span class="name">at</span> <span class="literal number hex">0xc29b90</span><span class="operator">&gt;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ubx</span><span class="operator">.</span><span class="name">consensus</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TAAT'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>第一个展示的功能是基于皮尔逊相关（Pearson correlation）的。因为我们想让它类似于一个距离估量，所以我们实际上取 1−<em>r</em> ，其中的 <em>r</em> 是皮尔逊相关系数（Pearson correlation coefficient，PCC）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">dist_pearson</span><span class="punctuation">(</span><span class="name">ubx</span><span class="punctuation">)</span>
<span class="punctuation">(</span><span class="literal number float">0.41740393308237722</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">)</span>
</pre>
<p>这意味着模体 <tt class="docutils literal">m</tt> 和 <tt class="docutils literal">Ubx</tt> 间最佳的PCC可以从下面的比对中获得：</p>
<pre class="code python literal-block">
<span class="name">bbTAAT</span>
<span class="name">TATAAb</span>
</pre>
<p>其中 <tt class="docutils literal">b</tt> 代表背景分布。PCC值大概为 1-0.42=0.58.如果我们尝试计算Ubx模体的互补序列：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">dist_pearson</span><span class="punctuation">(</span><span class="name">ubx</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">())</span>
<span class="punctuation">(</span><span class="literal number float">0.25784180151584823</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>我们可以发现更好的PCC值（大概为0.75），并且比对也是不同的：</p>
<pre class="code python literal-block">
<span class="name">bATTA</span>
<span class="name">TATAA</span>
</pre>
<p>还有两个其他的功能函数： <tt class="docutils literal">dist_dpq</tt> ,这是基于Kullback-Leibler散度的真正度量（满足三角不等式）。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">dist_dpq</span><span class="punctuation">(</span><span class="name">ubx</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">())</span>
<span class="punctuation">(</span><span class="literal number float">0.49292358382899853</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>还有 <tt class="docutils literal">dist_product</tt> 方法，它是基于概率的方法，这概率可以看成是两个模体独立产生两个相同实例的概率。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">m</span><span class="operator">.</span><span class="name">dist_product</span><span class="punctuation">(</span><span class="name">ubx</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">())</span>
<span class="punctuation">(</span><span class="literal number float">0.16224587301064275</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
</div>
<div class="section" id="id173">
<h3>14.9.4  <em>De novo</em> 模体查找</h3>
<p>目前，Biopython对 <em>de novo</em> 模体查找只有一些有限的支持。也就是说，我们只支持AlignAce和MEME的运行和读取。由于现模体查找工具发展如雨后春笋般，我们很欢迎有新的贡献者加入。</p>
<div class="section" id="id174">
<h4>14.9.4.1  MEME</h4>
<p>假如你以中意的参数用MEME来跑你自己的序列，并把得到的结果保存在 <tt class="docutils literal">meme.out</tt> 文件中。你可以用以下代码读取MEME产生的文件获得那些模体：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">Motif</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;meme.out&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;MEME&quot;</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span>
<span class="punctuation">[</span><span class="operator">&lt;</span><span class="name">Bio</span><span class="operator">.</span><span class="name">Motif</span><span class="operator">.</span><span class="name">MEMEMotif</span><span class="operator">.</span><span class="name">MEMEMotif</span> <span class="name builtin">object</span> <span class="name">at</span> <span class="literal number hex">0xc356b0</span><span class="operator">&gt;</span><span class="punctuation">]</span>
</pre>
<p>除了那一系列想要的模体外，结果对象中还有很多有用的信息，可以用那些一目了然的属性名来获取：</p>
<ul class="simple">
<li><tt class="docutils literal">.alphabet</tt></li>
<li><tt class="docutils literal">.datafile</tt></li>
<li><tt class="docutils literal">.sequence_names</tt></li>
<li><tt class="docutils literal">.version</tt></li>
<li><tt class="docutils literal">.command</tt></li>
</ul>
<p>MEME解析器得到的模体可以像通常模体（含有实例）一样进行处理，它们也可以通过对实例添加附加信息而提供一些额外的功能。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">consensus</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'CTCAATCGTA'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>

<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">pvalue</span>
<span class="literal number float">8.71e-07</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">sequence_name</span>
<span class="literal string">'SEQ10;'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">start</span>
<span class="literal number integer">3</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">instances</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">strand</span>
<span class="literal string">'+'</span>
</pre>
</div>
<div class="section" id="id175">
<h4>14.9.4.2  AlignAce</h4>
<p>对于AlignACE程序也可以做相同的事情。假如你把结果存储于文件 <tt class="docutils literal">alignace.out</tt> 文件中。你可以用以下代码读取结果：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsA</span><span class="operator">=</span><span class="name builtin">list</span><span class="punctuation">(</span><span class="name">Motif</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;alignace.out&quot;</span><span class="punctuation">),</span><span class="literal string">&quot;AlignAce&quot;</span><span class="punctuation">))</span>
</pre>
<p>同样，得到的模体也和平常的模体一样：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsA</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">consensus</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TCTACGATTGAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>事实上，你甚至可以发现AlignAce和MEME得到的模体十分相似，只不过AlignAce模体是MEME模体反向互补序列的加长版本而已：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">motifsM</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">consensus</span><span class="punctuation">()</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TACGATTGAG'</span><span class="punctuation">,</span> <span class="name">IUPACUnambiguousDNA</span><span class="punctuation">())</span>
</pre>
<p>如果你的机器上安装了AlignAce，你也可以直接从Biopython中启动。下面就是一个如何启动的小例子（其他参数可以用关键字参数指定）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">command</span><span class="operator">=</span><span class="literal string">&quot;/opt/bin/AlignACE&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">input_file</span><span class="operator">=</span><span class="literal string">&quot;test.fa&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Motif.Applications</span> <span class="keyword namespace">import</span> <span class="name">AlignAceCommandline</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cmd</span> <span class="operator">=</span> <span class="name">AlignAceCommandline</span><span class="punctuation">(</span><span class="name">cmd</span><span class="operator">=</span><span class="name">command</span><span class="punctuation">,</span><span class="name builtin">input</span><span class="operator">=</span><span class="name">input_file</span><span class="punctuation">,</span><span class="name">gcback</span><span class="operator">=</span><span class="literal number float">0.6</span><span class="punctuation">,</span><span class="name">numcols</span><span class="operator">=</span><span class="literal number integer">10</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">stdout</span><span class="punctuation">,</span><span class="name">stderr</span><span class="operator">=</span> <span class="name">cmd</span><span class="punctuation">()</span>
</pre>
<p>由于AlignAce把结果打印到标准输出，因此你可以通过读取结果的第一部分来获得你想要的模体：</p>
<pre class="code python literal-block">
<span class="name">motifs</span><span class="operator">=</span><span class="name builtin">list</span><span class="punctuation">(</span><span class="name">Motif</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">stdout</span><span class="punctuation">,</span><span class="literal string">&quot;AlignAce&quot;</span><span class="punctuation">))</span>
</pre>
</div>
</div>
</div>
</div>
<div class="section" id="id176">
<h1>第15章 聚类分析</h1>
<p>聚类分析是根据元素相似度，进行分组的过程。在生物信息学中，聚类分析广泛
用于基因表达数据分析，用来对具有相似表达谱的基因归类；从而鉴定功能相关的基
因，或预测未知基因的功能。</p>
<p>Biopython中的 <tt class="docutils literal">Bio.Cluster</tt> 模块提供了常用的聚类算法。虽然Bio.Cluster被设计用于
基因表达数据，它也可用于其他类型数据的聚类。 <tt class="docutils literal">Bio.Cluster</tt>
和其使用的C聚类库的说明见De Hoon <em>et al.</em> [<a class="reference external" href="#dehoon2004">14</a>].</p>
<p><tt class="docutils literal">Bio.Cluster</tt> 包含了以下四种聚类算法：</p>
<ul class="simple">
<li>系统聚类（成对重心法，最短距离，最大距离和平均连锁法);</li>
<li><em>k</em>-means, <em>k</em>-medians, 和 <em>k</em>-medoids 聚类;</li>
<li>自组织映射（Self-Organizing Maps）;</li>
<li>主成分分析</li>
</ul>
<div class="section" id="id177">
<h2>数据表示法</h2>
<p>用于聚类的输入为一个 <em>n</em> x <em>m</em> 的Python 数值矩阵 <tt class="docutils literal">data</tt>。在基因表达数据聚类中，
每一行表示不同的基因，每一列表示不同的实验条件。 <tt class="docutils literal">Bio.Cluster</tt> 既可以
针对每行（基因），也可以针对每列（实验条件）进行聚类。</p>
</div>
<div class="section" id="id178">
<h2>缺失值</h2>
<p>在芯片实验中，经常会有些缺失值，可以用一个额外的 <em>n</em> × <em>m</em> Numerical Python
整型矩阵 <tt class="docutils literal">mask</tt> 表示。例如 <tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt> ，表示 <tt class="docutils literal">data[i,j]</tt> 是个缺失值，
并且在分析中被忽略。</p>
</div>
<div class="section" id="id179">
<h2>随机数生成器</h2>
<p><em>k</em>-means/medians/medoids 聚类和 Self-Organizing
Maps (SOMs) 需要调用随机数生成器。在 <tt class="docutils literal">Bio.Cluster</tt> 中，正态分布随机数
生成器的算法是基于L’Ecuyer [<a class="reference external" href="#lecuyer1988">25</a>],二项分布的随机数
生成器算法是基于Kachitvichyanukul and Schmeiser [<a class="reference external" href="#kachitvichyanukul1988">23</a>]
开发的BTPE算法。随机数生成器在调用时会首先进行初始化。由于随机数生成器使用了
两个乘同余发生器（multiplicative linear congruential generators），所以初始化时需要两个整型的
种子。这两个种子可以调用系统提供的 <tt class="docutils literal">rand</tt> （C标准库）函数生成。在 <tt class="docutils literal">Bio.Cluster</tt> 中，
我们首先调用 <tt class="docutils literal">srand</tt> 使用以秒为单位的时间戳的值初始值，再用 <tt class="docutils literal">rand</tt> 随机产生两
个随机数作为种子来产生正态分布的随机数。</p>
</div>
<div class="section" id="id180">
<h2>15.1 距离函数</h2>
<p>为了对元素根据相似度进行聚类，第一步需要定义相似度。<tt class="docutils literal">Bio.Cluster</tt> 提供了八种不同
的距离函数来衡量相似度或者距离，分别用不同的字母代表：</p>
<ul class="simple">
<li><tt class="docutils literal">'e'</tt>: Euclidean 距离;</li>
<li><tt class="docutils literal">'b'</tt>: City-block 距离.</li>
<li><tt class="docutils literal">'c'</tt>: Pearson 相关系数;</li>
<li><tt class="docutils literal">'a'</tt>: Pearson相关系数的绝对值;</li>
<li><tt class="docutils literal">'u'</tt>: Uncentered Pearson correlation （相当于两个数据向量的夹角余弦值）</li>
<li><tt class="docutils literal">'x'</tt>: uncentered Pearson correlation的绝对值;</li>
<li><tt class="docutils literal">'s'</tt>: Spearman’s 秩相关系数;</li>
<li><tt class="docutils literal">'k'</tt>: Kendall’s τ.</li>
</ul>
<p>前两个距离函数满足三角形的两边和大于第三边的特点：</p>
<div class="formula">
<i>d</i><span class="symbol">(</span><u><i>u</i></u>, <u><i>v</i></u><span class="symbol">)</span> ≤ <i>d</i><span class="symbol">(</span><u><i>u</i></u>, <u><i>w</i></u><span class="symbol">)</span> + <i>d</i><span class="symbol">(</span><u><i>w</i></u>, <u><i>v</i></u><span class="symbol">)</span><span class="textrm"> for all </span><u><i>u</i></u>, <u><i>v</i></u>, <u><i>w</i></u>, 
</div>
<p>所以称之为 <em>metrics</em>。 在任何语言中，这个意味着两点之间直线最短。</p>
<p>剩余的六种距离函数跟相关系数有关，距离 <em>d</em> 是由相关性 <em>r</em> 确定： <em>d</em>=1−<em>r</em>。
请注意这类距离函数是 <em>semi-metrics</em> ，因此不满足三角形的两边之和大于第三边的
性质。例如</p>
<div class="formula">
<u><i>u</i></u> = <span class="symbol">(</span>1, 0,  − 1<span class="symbol">)</span>;
</div>
<div class="formula">
<u><i>v</i></u> = <span class="symbol">(</span>1, 1, 0<span class="symbol">)</span>;
</div>
<div class="formula">
<u><i>w</i></u> = <span class="symbol">(</span>0, 1, 1<span class="symbol">)</span>;
</div>
<p>通过计算Pearson距离，可以得到 <em>d</em>(<em>u</em>,<em>w</em>) = 1.8660, 而
<em>d</em>(<em>u</em>,<em>v</em>)+<em>d</em>(<em>v</em>,<em>w</em>) = 1.6340.</p>
<div class="section" id="euclidean">
<h3>Euclidean 距离</h3>
<p>在 <tt class="docutils literal">Bio.Cluster</tt> 中, Euclidean 距离被定义为</p>
<div class="formula">
<i>d</i> = <span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="limits"><sup class="limit"><i>n</i></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><span class="symbol">(</span><i>x</i><sub><i>i</i></sub> − <i>y</i><sub><i>i</i></sub><span class="symbol">)</span><sup>2</sup>.
</div>
<p>求和时只考虑*x*<sub>*i*</sub> 和 <em>y</em><sub>*i*</sub> 都存在的值, 分母 <em>n</em>
也相应的做出调整。当分析表达谱数据时，由于 <em>x</em><sub>*i*</sub> 和 <em>y</em><sub>*i*</sub>
会直接相减, 因此在使用Euclidean距离前，请对表达谱数据标准化处理。</p>
</div>
<div class="section" id="city-block-distance">
<h3>City-block distance</h3>
<p>city-block distance也称之为Manhattan 距离，跟Euclidean距离有一定的相似性。Euclidean距离
表示的是两点间最短的距离，而city-block距离则是两点在所有维度中距离的和。由于基因表达的数据
经常会有缺失数据，在 <tt class="docutils literal">Bio.Cluster</tt> 中，city-block距离定义为总距离除以总维度：</p>
<div class="formula">
<i>d</i> = <span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="limits"><sup class="limit"><i>n</i></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><span class="symbol">|</span><i>x</i><sub><i>i</i></sub> − <i>y</i><sub><i>i</i></sub><span class="symbol">|</span>.
</div>
<p>City-block distance类似于当你在从城市里一个位置到另一个位置时，所经过街道的距离。
而在Euclidean 距离中，表达谱的数据会直接相减，因此必须先对数据进行标准化。</p>
</div>
<div class="section" id="pearson">
<h3>Pearson 相关系数</h3>
<p>Pearson相关系数定义为：</p>
<div class="formula">
<i>r</i> = <span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="limits"><sup class="limit"><i>n</i></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="fraction"><span class="ignored">(</span><span class="numerator"><i>x</i><sub><i>i</i></sub> − <span class="bar"><i>x</i></span></span><span class="ignored">)/(</span><span class="denominator"><i>σ</i><sub><i>x</i></sub></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span><span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="fraction"><span class="ignored">(</span><span class="numerator"><i>y</i><sub><i>i</i></sub> − <span class="bar"><i>y</i></span></span><span class="ignored">)/(</span><span class="denominator"><i>σ</i><sub><i>y</i></sub></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>, 
</div>
<p>其中 x, ȳ 分别是 <em>x</em> 和 <em>y</em> 的样品均值, σ<sub>*x*</sub>, σ<sub>*y*</sub>
是 <em>x</em> 和 <em>y</em> 的样品标准差. Pearson相关系数是用于测量 <em>x</em> and <em>y</em> 散点图对直线的
拟合程度。如果所有的点都在直线上，那么Pearson相关系数为 +1 or -1, 取决于直线的斜率
是正还是负。如果Pearson 相关系数等于0，表明 <em>x</em> 和 <em>y</em> 之间没有相关性。</p>
<p><em>Pearson distance</em> 定义为</p>
<div class="formula">
<i>d</i><sub><span class="textrm">P</span></sub> ≡ 1 − <i>r</i>.
</div>
<p>由于Pearson 相关性的值介于 -1 和 1之间, Pearson 距离的范围为 0 和 2 之间.</p>
</div>
<div class="section" id="absolute-pearson-correlation">
<h3>Absolute Pearson correlation</h3>
<p>通过对Pearson相关系数取绝对值，可以得到一个0和1之间的数。如果绝对值是1，
所有的点都位于一条斜率为正或负直线上。当绝对值为0时，表明 <em>x</em> and <em>y</em> 没有相关性。</p>
<p>对应的距离定义为：</p>
<div class="formula">
<i>d</i><sub><span class="textrm"></span><i>A</i></sub> ≡ 1 − <span class="symbol">|</span><i>r</i><span class="symbol">|</span>, 
</div>
<p>其中 <em>r</em> 是 Pearson 相关系数. 由于Pearson的相关系数的绝对值介于 0 和 1之间, 对应的
距离也位于0和1之间。</p>
<p>在基因表达数据分析中，应当注意，当相关性的绝对值等于1时，表明两组基因的表达情况完全一样或者完全
相反。</p>
</div>
<div class="section" id="uncentered-correlation">
<h3>Uncentered correlation (夹角余弦)</h3>
<p>在某些情况下，使用 <em>uncentered correlation</em> 比常规的Pearson相关系数更合适。
uncentered correlation 定义为：</p>
<div class="formula">
<i>r</i><sub><span class="textrm"></span><i>U</i></sub> = <span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="limits"><sup class="limit"><i>n</i></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="fraction"><span class="ignored">(</span><span class="numerator"><i>x</i><sub><i>i</i></sub></span><span class="ignored">)/(</span><span class="denominator"><i>σ</i><span class="scripts"><sup class="script">(0)</sup><sub class="script"><i>x</i></sub></span></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span><span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="fraction"><span class="ignored">(</span><span class="numerator"><i>y</i><sub><i>i</i></sub></span><span class="ignored">)/(</span><span class="denominator"><i>σ</i><span class="scripts"><sup class="script">(0)</sup><sub class="script"><i>y</i></sub></span></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>, 
</div>
<p>其中</p>
<div class="formula">
<span class="environment"><span class="arrayrow">
<span class="arraycell align-r">
<i>σ</i><span class="scripts"><sup class="script">(0)</sup><sub class="script"><i>x</i></sub></span>
</span>
<span class="arraycell align-c">
 = 
</span>
<span class="arraycell align-l">
<span class="sqrt"><span class="radical">√</span><span class="ignored">(</span><span class="root"><span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="limits"><sup class="limit"><i>n</i></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><i>x</i><span class="scripts"><sup class="script">2</sup><sub class="script"><i>i</i></sub></span></span><span class="ignored">)</span></span>;
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-r">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-l">
 
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-r">
<i>σ</i><span class="scripts"><sup class="script">(0)</sup><sub class="script"><i>y</i></sub></span>
</span>
<span class="arraycell align-c">
 = 
</span>
<span class="arraycell align-l">
<span class="sqrt"><span class="radical">√</span><span class="ignored">(</span><span class="root"><span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="limits"><sup class="limit"><i>n</i></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><i>y</i><span class="scripts"><sup class="script">2</sup><sub class="script"><i>i</i></sub></span></span><span class="ignored">)</span></span>.
</span>

</span>
</span>
</div>
<p>这个公式同Pearson相关系数的公式形式一样，只是把样本均值 x, ȳ 设为0 。
uncentered correlation 适用于表达量基准为0的情况。例如，在对基因表达分析中，使用
比值对数时，当log-ratio 等于0 表明红绿信号强度相等，也意味着实验处理
不影响基因的表达量。</p>
<p>uncentered correlation 系数对应的距离计算方法为：</p>
<div class="formula">
<i>d</i><sub><span class="mbox">U</span></sub> ≡ 1 − <i>r</i><sub><span class="mbox">U</span></sub>, 
</div>
<p>其中 <em>r</em><sub>U</sub> 是uncentered 相关性系数。 由于uncentered系数位于-1 和 1
之间，对应的距离范围为 0 与 2之间。</p>
<p>由于 uncentered 相关系数值等同于两个数据向量在 <em>n</em> 维空间里的夹角余弦，因此也常称为夹角余弦。</p>
</div>
<div class="section" id="absolute-uncentered-correlation">
<h3>Absolute uncentered correlation</h3>
<p>与 Pearson 相关性类似, 也可以用uncentered correlation的绝对值来定义距离:</p>
<div class="formula">
<i>d</i><sub><span class="mbox">AU</span></sub> ≡ 1 − <span class="symbol">|</span><i>r</i><sub><span class="mbox">U</span></sub><span class="symbol">|</span>, 
</div>
<p>其中 <em>r</em><sub>U</sub> 是 uncentered相关系数。由于uncentered 相关系数的
绝对值位于 0 和 1 之间，对应的距离也为位于 0 和 1之间。</p>
<p>从几何学上来讲，uncentered相关系数的绝对值等于两个数据所在向量的支持线（supporting lines）
的角度余弦值（即不考虑向量的方向性）。</p>
</div>
<div class="section" id="spearman-rank-correlation">
<h3>Spearman rank correlation</h3>
<p>Spearman秩相关系数是一种非参的相关性测量方法，对于数据中的离群点，比Pearson相关系数
有更好的稳健性。</p>
<p>为了计算Spearman秩相关系数，首先对每个数据集里的数据按值排序，得到每个数据的对应的
秩。然后，计算对两个数据的秩集合计算Pearson相关系数，得到Spearman的相关系数。</p>
<p>同Pearson相关性类似，Spearman秩相关系数对应的距离定义为：</p>
<div class="formula">
<i>d</i><sub><span class="mbox">S</span></sub> ≡ 1 − <i>r</i><sub><span class="mbox">S</span></sub>, 
</div>
<p>其中 <em>r</em><sub>S</sub> 是Spearman秩相关系数。</p>
</div>
<div class="section" id="kendalls">
<h3>Kendall’s τ</h3>
<p>Kendall’s τ 是另一个非参的计算相关性的方法。它同Spearman秩相关系数类似，但它不对数据进行排序，
而是使用相对秩来计算  τ (see Snedecor &amp; Cochran [<a class="reference external" href="#snedecor1989">29</a>]) 。</p>
<p>Kendall’s τ 对应的距离计算为：</p>
<div class="formula">
<i>d</i><sub><span class="mbox">K</span></sub> ≡ 1 − <i>τ</i>.
</div>
<p>因为 Kendall’s τ 位于 -1 和 1之间, 对应的距离位于 0 和 2之间。</p>
</div>
<div class="section" id="weighting">
<h3>Weighting</h3>
<p>对于 <tt class="docutils literal">Bio.Cluster</tt> 中大部分距离函数，都可以使用加权向量。加权向量包含着
数据集中每个元素的权重。如果元素 <em>i</em> 的权重为 <em>w</em><sub>*i*</sub>，那么将会认为该元素
出现了 <em>w</em><sub>*i*</sub> 次 。权重值可以不为整数。对于 Spearman 秩相关系数
和Kendall’s τ, 权重没有太大的意义，因此不适用于这两个函数。</p>
</div>
<div class="section" id="id181">
<h3>计算距离矩阵</h3>
<p>距离矩阵是 <tt class="docutils literal">data</tt> 中，所有元素的两两间的距离的平方矩阵，可以用 <tt class="docutils literal">Bio.Cluster</tt> 模块中 <tt class="docutils literal">distancematrix</tt> 函数计算：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">distancematrix</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">matrix</span> <span class="operator">=</span> <span class="name">distancematrix</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>其中，包含以下参数：</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt> (必选)</dt>
<dd><p class="first last">包含所有元素的矩阵</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">缺失数据矩阵。若 <tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, 则 <tt class="docutils literal">data[i,j]</tt> 缺失。若 <tt class="docutils literal"><span class="pre">mask==None</span></tt>, 表明没有缺失数据。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">weight</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">权重矩阵。若 <tt class="docutils literal"><span class="pre">weight==None</span></tt>, 则假设所有的数据使用相同的权重。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">选择使用 <tt class="docutils literal">data</tt> 中的行 (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>), 或者列 (<tt class="docutils literal"><span class="pre">transpose==1</span></tt>)来计算距离.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (默认: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">选择距离函数 (具体见 <a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
</ul>
<p>为了节省内存，函数返回的距离矩阵是一个一维数组的列表。每行的列数等于行号。
因此，第一行有0个元素。例如：</p>
<pre class="code python literal-block">
<span class="punctuation">[</span><span class="name">array</span><span class="punctuation">([]),</span>
 <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">1.</span><span class="punctuation">]),</span>
 <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">7.</span><span class="punctuation">,</span> <span class="literal number float">3.</span><span class="punctuation">]),</span>
 <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">4.</span><span class="punctuation">,</span> <span class="literal number float">2.</span><span class="punctuation">,</span> <span class="literal number float">6.</span><span class="punctuation">])]</span>
</pre>
<p>对应的距离矩阵为：</p>
<div class="formula">
<span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎜</span></span><span class="arrayrow"><span class="bracket align-left">⎜</span></span><span class="arrayrow"><span class="bracket align-left">⎜</span></span><span class="arrayrow"><span class="bracket align-left">⎜</span></span><span class="arrayrow"><span class="bracket align-left">⎜</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="array"><span class="arrayrow">
<span class="arraycell align-c">
0
</span>
<span class="arraycell align-c">
1
</span>
<span class="arraycell align-c">
7
</span>
<span class="arraycell align-c">
4
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-c">
1
</span>
<span class="arraycell align-c">
0
</span>
<span class="arraycell align-c">
3
</span>
<span class="arraycell align-c">
2
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-c">
7
</span>
<span class="arraycell align-c">
3
</span>
<span class="arraycell align-c">
0
</span>
<span class="arraycell align-c">
6
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-c">
 
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-c">
4
</span>
<span class="arraycell align-c">
2
</span>
<span class="arraycell align-c">
6
</span>
<span class="arraycell align-c">
0
</span>

</span>
</span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎟</span></span><span class="arrayrow"><span class="bracket align-right">⎟</span></span><span class="arrayrow"><span class="bracket align-right">⎟</span></span><span class="arrayrow"><span class="bracket align-right">⎟</span></span><span class="arrayrow"><span class="bracket align-right">⎟</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>.
</div>
</div>
</div>
<div class="section" id="id182">
<h2>15.2  计算类的相关性质</h2>
<div class="section" id="id183">
<h3>计算类中心</h3>
<p>类中心可以定义为该类中在每个维度上所有元素的平均值或者中值，可以用 <tt class="docutils literal">Bio.Cluster</tt> 中的 <tt class="docutils literal">clustercentroids</tt>
函数计算：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">clustercentroids</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cdata</span><span class="punctuation">,</span> <span class="name">cmask</span> <span class="operator">=</span> <span class="name">clustercentroids</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>包含了以下参数:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt> (必选)</dt>
<dd><p class="first last">包含所有元素的矩阵。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">缺失数据矩阵。若 <tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, 则 <tt class="docutils literal">data[i,j]</tt> 缺失。若 <tt class="docutils literal"><span class="pre">mask==None</span></tt>, 则明没有缺失数据。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">clusterid</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">一个表示每个元素的所属类的整型向量。如果 <tt class="docutils literal">clusterid</tt> 是 <tt class="docutils literal">None</tt>, 表明所有的元素属于相同的类。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (默认: <tt class="docutils literal">'a'</tt>)</dt>
<dd><p class="first last">指定使用算术平方根 (<tt class="docutils literal"><span class="pre">method=='a'</span></tt>) 或者中值(<tt class="docutils literal"><span class="pre">method=='m'</span></tt>) 来计算类中心。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">选择使用 <tt class="docutils literal">data</tt> 中的行 (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>), 或者列 (<tt class="docutils literal"><span class="pre">transpose==1</span></tt>) 来计算类中心.</p>
</dd>
</dl>
</li>
</ul>
<p>这个函数返回值为元组 <tt class="docutils literal">(cdata, cmask)</tt>。 类中心的数据存储在一个二维的Numerical Python
数组 <tt class="docutils literal">cdata</tt> 中, 缺失值的结果存储在二维的Numerical Python整型数组 <tt class="docutils literal">cmask</tt> 中。 当 <tt class="docutils literal">transpose</tt> = <tt class="docutils literal">0</tt> 时，
这两个数组的维度是（类数，列数），当 <tt class="docutils literal">transpose</tt> = <tt class="docutils literal">1</tt> 时，数组的长度为 （行数，类数）。
其中每一行（当 <tt class="docutils literal">transpose</tt> = <tt class="docutils literal">0</tt>) 或者 每一列（当 <tt class="docutils literal">transpose</tt> = <tt class="docutils literal">1</tt> ）
包含着对应每类对应的数据的平均值。</p>
</div>
<div class="section" id="id184">
<h3>计算类间距离</h3>
<p>根据每个 <em>items</em> 的距离函数，我们可以计算出两个 <em>clusters</em> 的距离。两个类别的
算术平均值之间的距离通常用于重心法聚类和 <em>k</em>-means 聚类，而 <em>k</em>-medoids
聚类中，通常利用两类的中值进行计算。最短距离法利用的是两类间最近的元素之间的距离，
而最大距离法利用最长的元素之间的距离。在两两平均连锁聚类法中，
类间的距离定义为类内所有对应元素两两间距离的平均值。</p>
<p>为了计算两类之间的距离，可以利用:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">clusterdistance</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">distance</span> <span class="operator">=</span> <span class="name">clusterdistance</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>其中，包含的参数有：</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt> (必选)</dt>
<dd><p class="first last">包含所有元素的矩阵。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">缺失数据矩阵。若 <tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, 则 <tt class="docutils literal">data[i,j]</tt> 缺失。若 <tt class="docutils literal"><span class="pre">mask==None</span></tt>, 则明没有缺失数据。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">weight</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">权重矩阵。若 <tt class="docutils literal"><span class="pre">weight==None</span></tt>, 则假设所有的数据使用相同的权重。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">index1</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">第一个类所包含的元素索引的列表。如果一个类别只包含一个元素 <em>i</em> ，则数据类型
可以为一个列表 <tt class="docutils literal">[i]</tt>, 或者整数 <tt class="docutils literal">i</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">index2</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">第二个类所包含的元素的列表。如果一个类别只包含一个元素 <em>i</em> ，则数据类型
可以为一个列表 <tt class="docutils literal">[i]</tt>, 或者整数 <tt class="docutils literal">i</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (默认: <tt class="docutils literal">'a'</tt>)</dt>
<dd><p class="first last">选择计算类别间距离的方法:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal">'a'</tt>: 使用两个类中心的距离 (算术平均值);</li>
<li><tt class="docutils literal">'m'</tt>: 使用两个类中心的距离 (中值);</li>
<li><tt class="docutils literal">'s'</tt>: 使用两类中最短的两个元素之间的距离;</li>
<li><tt class="docutils literal">'x'</tt>: 使用两类中最长的两个元素之间的距离;</li>
<li><tt class="docutils literal">'v'</tt>: 使用两类中对应元素间的距离的平均值作为距离。</li>
</ul>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (默认: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">选择距离函数 (具体见 <a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">选择使用 <tt class="docutils literal">data</tt> 中的行 (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>), 或者列 (<tt class="docutils literal"><span class="pre">transpose==1</span></tt>)来计算距离.</p>
</dd>
</dl>
</li>
</ul>
</div>
</div>
<div class="section" id="id185">
<h2>15.3  划分算法</h2>
<p>划分算法依据所有元素到各自聚类中心距离之和最小化原则，
将元素分为 <em>k</em> 类。类的个数 <em>k</em> 由用户定义。 <tt class="docutils literal">Bio.Cluster</tt> 提供了三种不同
的算法:</p>
<ul class="simple">
<li><em>k</em>-means 聚类</li>
<li><em>k</em>-medians 聚类</li>
<li><em>k</em>-medoids 聚类</li>
</ul>
<p>这些算法的区别在于如何定义聚类中心。在 <em>k</em>-means 中, 聚类中心定义为该类中所有
元素的平均值。 在 <em>k</em>-medians 聚类中， 利用每个维度的中间值来计算。
最后， <em>k</em>-medoids 聚类中，聚类中心定义为该类中，距离其他所有元素距离之和最小的元素所在的位置。
这个方法适用于已知距离矩阵，但是原始数据矩阵未知的情况，例如根据结构相似度对蛋白进行聚类。</p>
<p>expectation-maximization (EM) 算法通常用于将数据分成 <em>k</em> 组。在 EM算法的起始阶段,
随机的把元素分配到不同的组。为了保证所有的类都包含元素，可以利用二项分布的方法随机
为每类挑选元素。然后，随机的对分组进行排列，保证每个元素有相同的概率被分到任何一个类别。
最终，保证每类中至少含有一个元素。</p>
<p>之后进行迭代:</p>
<ul class="simple">
<li>利用均值，中值或者medoid计算每类的中心;</li>
<li>计算每类的元素离各自中心的距离;</li>
<li>对于每个元素，判别其离哪个聚类中心最近;</li>
<li>将元素重新分配到最近的聚类，当不能进行调整时，迭代终止。</li>
</ul>
<p>为了避免迭代中产生空的类别，在 <em>k</em>-means 和 <em>k</em>-medians 聚类中，算法始终记录着每类中元素的
个数，并且阻止最后一个元素被分到其他的类别中。对于 <em>k</em>-medoids 聚类, 这种检查就是没有必要的，
因为当只剩最后一个元素时，它离中心的距离为0，所以不会被分配到其他的类别中。</p>
<p>由于起始阶段的每类中的元素分配是随机的，而通常当EM算法执行时，可能产生不同的聚类结果。为了找到最优的聚类结果，
可以对进行 <em>k</em>-means 算法重复多次，每次都以不同的随机分配作为起始。每次运行后，都会保存所有元素距离
其中心距离之和，并且选择总距离最小的运行结果最为最终的结果。</p>
<p>EM算法运行的次数取决于需要聚类元素的多少。一般而言，我们可以根据最优解被发现的次数来选择。
这个次数会作为划分算法的返回值。如果最优解被多次返回，那么不太可能存在比这个
更优的解。然后，如果最优解只被发现一次，那么可能存在着距离更小的解。但是，如果需要聚类的
元素过多的话（多余几百），那么很难找到一个全局最优解。</p>
<p>EM算法会在不能进行任何分配的时候停止。我们注意到，在某些随机的起始分配中，由于
相同的解会在迭代中周期性的重复，从而导致EM算法的失败。因此，我们在迭代中也会
检查是否有周期性出现的解存在。首先，在给定数目的迭代后，当前的聚类结果会保存作为一个参考。之后
继续迭代一定次数，比较该结果同之前保存的结果，可以确定之前的结果是否重复出现。
如果有重复出现，迭代会终止。如果没有出现，那么再次迭代后的结果会保存作为新的参考。
通常，会首先重复10次迭代，再保存结果为新的参考。之后，迭代的次数会翻倍，保证在长的周期中也可以
检测到该解。</p>
<div class="section" id="k-means-and-k-medians">
<h3><em>k</em>-means and <em>k</em>-medians</h3>
<p><em>k</em>-means 和 <em>k</em>-medians 算法可以利用 <tt class="docutils literal"><span class="pre">Bio.Cluster``中的</span> ``kcluster</tt> 实现:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">kcluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clusterid</span><span class="punctuation">,</span> <span class="name">error</span><span class="punctuation">,</span> <span class="name">nfound</span> <span class="operator">=</span> <span class="name">kcluster</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>其中，包含的参数有：</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt> (必选)</dt>
<dd><p class="first last">包含所有元素的矩阵。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">nclusters</tt> (默认: <tt class="docutils literal">2</tt>)</dt>
<dd><p class="first last">期望的类的数目 <em>k</em>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">缺失数据矩阵。若 <tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, 则 <tt class="docutils literal">data[i,j]</tt> 缺失。若 <tt class="docutils literal"><span class="pre">mask==None</span></tt>, 则明没有缺失数据。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">weight</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">权重矩阵。若 <tt class="docutils literal"><span class="pre">weight==None</span></tt>, 则假设所有的数据使用相同的权重。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">选择使用 <tt class="docutils literal">data</tt> 中的行 (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>), 或者列 (<tt class="docutils literal"><span class="pre">transpose==1</span></tt>)来计算距离.
-  <tt class="docutils literal">npass</tt> (默认: <tt class="docutils literal">1</tt>)
<em>k</em>-means/-medians 聚类算法运行的次数，每次运行使用不同的随机的起始值。
如果指定了 <tt class="docutils literal">initialid</tt> , 程序会忽略``npass`` 的值，并且聚类算法只会运行一次。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (默认: <tt class="docutils literal">a</tt>)</dt>
<dd><p class="first last">指定聚类中心计算方法:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">method=='a'</span></tt>: 算数平均值 (<em>k</em>-means clustering);</li>
<li><tt class="docutils literal"><span class="pre">method=='m'</span></tt>: 中值 (<em>k</em>-medians clustering).</li>
</ul>
<p>当指定 <tt class="docutils literal">method</tt> 使用其他值时，算法会采用算数平均值。</p>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (默认: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">选择距离函数 (具体见 <a class="reference external" href="#sec:distancefunctions">15.1</a>).
尽管八种距离都可以用于 <tt class="docutils literal">kcluster</tt> 计算,
但从经验上来讲，Euclidean 距离适合 <em>k</em>-means 算法, city-block 距离适合 <em>k</em>-medians.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">initialid</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">指定EM算法运行初始的聚类类别。如果 <tt class="docutils literal"><span class="pre">initialid==None</span></tt>, 那么每运行一次EM算法时，
都会采取不同的随机初始聚类，总共运行的次数由 <tt class="docutils literal">npass</tt> 决定。如果 <tt class="docutils literal">initialid</tt> 不是 <tt class="docutils literal">None</tt>,
那么它应该为一个长度为类别数的1维数组，每类中至少含有一个元素。通常当初始分类确定后，EM算法的结果也就确定了。</p>
</dd>
</dl>
</li>
</ul>
<p>这个函数的返回值为一个包含 <tt class="docutils literal">(clusterid, error, nfound)</tt> 的元组，其中 <tt class="docutils literal">clusterid</tt> 是
一个整型矩阵，为每行或列所在的类。 <tt class="docutils literal">error</tt> 是最优聚类解中，每类内距离的总和，
<tt class="docutils literal">nfound</tt> 指的是最优解出现的次数。</p>
</div>
<div class="section" id="k-medoids">
<h3><em>k</em>-medoids 聚类</h3>
<p><tt class="docutils literal">kmedoids</tt> 函数根据提供的距离矩阵和聚类数，来运行 <em>k</em>-medoids 聚类：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">kmedoids</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clusterid</span><span class="punctuation">,</span> <span class="name">error</span><span class="punctuation">,</span> <span class="name">nfound</span> <span class="operator">=</span> <span class="name">kmedoids</span><span class="punctuation">(</span><span class="name">distance</span><span class="punctuation">)</span>
</pre>
<p>其中，包含的参数有: , nclusters=2, npass=1,
initialid=None)|</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">distance</tt> (必选)</dt>
<dd><p class="first last">两两元素间的距离矩阵，可以通过三种不同的方法提供：</p>
</dd>
</dl>
<ul>
<li><p class="first">提供一个2D的 Numerical Python 数组 (函数只会使用矩阵里左下角数据):</p>
<pre class="code python literal-block">
<span class="name">distance</span> <span class="operator">=</span> <span class="name">array</span><span class="punctuation">([[</span><span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">1.1</span><span class="punctuation">,</span> <span class="literal number float">2.3</span><span class="punctuation">],</span>
                  <span class="punctuation">[</span><span class="literal number float">1.1</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">],</span>
                  <span class="punctuation">[</span><span class="literal number float">2.3</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">]])</span>
</pre>
</li>
<li><p class="first">输入一个一维的 Numerical Python 数组，包含了距离矩阵左下角的数据：</p>
<pre class="code python literal-block">
<span class="name">distance</span> <span class="operator">=</span> <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">1.1</span><span class="punctuation">,</span> <span class="literal number float">2.3</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">])</span>
</pre>
</li>
<li><p class="first">输入一个列表，包含距离矩阵左下角的数据：</p>
<pre class="code python literal-block">
<span class="name">distance</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">array</span><span class="punctuation">([]</span><span class="operator">|</span><span class="punctuation">,</span>
            <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">1.1</span><span class="punctuation">]),</span>
            <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">2.3</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">])</span>
           <span class="punctuation">]</span>
</pre>
</li>
</ul>
<p>三种方法对应着同样的距离矩阵。</p>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">nclusters</tt> (默认: <tt class="docutils literal">2</tt>)</dt>
<dd><p class="first last">期望的类的数目 <em>k</em>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">npass</tt> (默认: <tt class="docutils literal">1</tt>)</dt>
<dd><p class="first last"><em>k</em>-medoids 聚类算法运行的次数，每次运行使用不同的随机的起始值。
如果指定了 <tt class="docutils literal">initialid</tt> , <tt class="docutils literal">npass</tt> 的值会忽略，并且聚类算法只会运行一次。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">initialid</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">指定EM算法运行初始的聚类类别。如果 <tt class="docutils literal"><span class="pre">initialid==None</span></tt>, 那么每运行一次EM算法时，
都会采取不同的随机初始聚类，总共运行的次数由 <tt class="docutils literal">npass</tt> 决定。如果 <tt class="docutils literal">initialid</tt> 不是 <tt class="docutils literal">None</tt>,
那么它应该为一个长度为类别数的1维数组，每类中至少含有一个元素。通常当初始分类确定后，EM算法的结果也就确定了。</p>
</dd>
</dl>
</li>
</ul>
<p>函数返回值为一个 包含 <tt class="docutils literal">(clusterid, error, nfound)</tt> 的元组, 其中
<tt class="docutils literal">clusterid</tt> 一个整型矩阵，为每行或列类所在的类。<tt class="docutils literal">error</tt> 是在最优解中，类内距离的总和，
<tt class="docutils literal">nfound</tt> 指的是最优解出现的次数。需要注意的是， <tt class="docutils literal">clusterid</tt> 中的类号是指的是代表聚类中心的元素号。</p>
</div>
</div>
<div class="section" id="id186">
<h2>15.4  系统聚类</h2>
<p>系统聚类同 <em>k</em>-means 聚类有本质的不同。在系统聚类中，基因间或者实验条件间的相似度是通过
树的形式展现出来的。由于可以利用Treeview或者Java Treeview来查看这些树的结构，因此系统聚类在基因表达谱数据中得到普遍应用。</p>
<p>系统聚类的第一步是计算所有元素间的距离矩阵。之后，融合两个最近的元素成为一个节点。然后，不断的
通过融合相近的元素或者节点来形成新的节点，直到所有的元素都属于同一个节点。在追溯元素和节点融合
的过程的同时形成了树的结构。不同于 <em>k</em>-means 使用的EM算法，系统聚类的过程是固定的。</p>
<p>系统聚类也存在着几个不同的方法，他们区别在于如何计算子节点间的距离。在
<tt class="docutils literal">Bio.Cluster</tt> 中，提供了最短距离法（ pairwise single）,最长距离法（maximum）, 类平均法（average）,
和重心法（centroid linkage）。</p>
<ul class="simple">
<li>在最短距离法中，节点间的距离被定义两个节点最近样品间距离。</li>
<li>在最短距离法中，节点间的距离被定义两个节点最远样品间距离。</li>
<li>在类平均法中，节点间的距离被定义为所有样品对之间的平均距离。</li>
<li>在重心法中，节点间的距离被定义为两个节点重心间的距离。重心的计算是通过对
每类中所有元素进行计算的。由于每次都要计算新的节点与 其他元素和已存在节点的距离，
因此重心法的运行时间比其他系统聚类的方法更长。该方法另外一个特性是，当聚类树的
长大的时候，距离并不会增加，有时候反而减少。这是由于使用Pearson相关系数作为距离时，
对重心的计算和距离的计算不一致产生:因为Pearson相关系数在计算距离时会对数据进行有效归一化，，
但是重心的计算不会存在该种归一化。</li>
</ul>
<p>对于最短距离法，最长距离法和类平均法时，两个节点之间的距离是直接对类别里的元素计算得到的。
因此，聚类的算法在得到距离矩阵后，不一定需要提供最开始的基因表达数据。而对于重心法而言，
新生成的节点的中心必须依靠原始的数据，而不是仅仅依靠距离矩阵。</p>
<p>最短距离法的实现是根据 SLINK algorithm (R. Sibson, 1973), 这个算法具有快速和高效的特点。
并且这个方法聚类的结果同传统的方法结果一致。并且该算法，也可以有效的运用于大量的数据，而传统的
算法则需要大量的内存需求和运行时间。</p>
<div class="section" id="id187">
<h3>展示系统聚类的结果</h3>
<p>系统聚类的结果是用树的结构展示所有节点，每个节点包含两个元素或者子节点。通常，我们既关心那个元素
或者哪个子节点互相融合，也关心二者之间的距离（或者相似度）。我们可以调用 <tt class="docutils literal">Bio.Cluster</tt> 中的
<tt class="docutils literal">Node</tt> 类，来存储聚类树的一个节点。 <tt class="docutils literal">Node</tt> 的实例包含以下三个属性：</p>
<ul class="simple">
<li><tt class="docutils literal">left</tt></li>
<li><tt class="docutils literal">right</tt></li>
<li><tt class="docutils literal">distance</tt></li>
</ul>
<p>其中, <tt class="docutils literal">left</tt> 和 <tt class="docutils literal">right</tt> 是合并到该节点两个元素或子节点的编号。
<tt class="docutils literal">distance</tt> 指的是二者间的距离。其中元素的编号是从0到（元素数目-1），
而聚类的组别是从-1到-（元素数目-1）。请注意，节点的数目比元素的数目少一。</p>
<p>为了创建一个新的 <tt class="docutils literal">Node</tt> 对象,我们需要指定 <tt class="docutils literal">left</tt> 和 <tt class="docutils literal">right</tt>;
<tt class="docutils literal">distance</tt> 是可选的。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">Node</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number integer">3</span><span class="punctuation">)</span>
<span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">):</span> <span class="literal number integer">0</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number integer">3</span><span class="punctuation">,</span><span class="literal number float">0.91</span><span class="punctuation">)</span>
<span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">):</span> <span class="literal number float">0.91</span>
</pre>
<p>对于已存在 <tt class="docutils literal">Node</tt> 对象的 <tt class="docutils literal">left</tt>, <tt class="docutils literal">right</tt>, 和 <tt class="docutils literal">distance</tt> 都是可以直接修改的：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">node</span> <span class="operator">=</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">4</span><span class="punctuation">,</span><span class="literal number integer">5</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">node</span><span class="operator">.</span><span class="name">left</span> <span class="operator">=</span> <span class="literal number integer">6</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">node</span><span class="operator">.</span><span class="name">right</span> <span class="operator">=</span> <span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">node</span><span class="operator">.</span><span class="name">distance</span> <span class="operator">=</span> <span class="literal number float">0.73</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">node</span>
<span class="punctuation">(</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">):</span> <span class="literal number float">0.73</span>
</pre>
<p>当 <tt class="docutils literal">left</tt> 和 <tt class="docutils literal">right</tt> 不是整数的时候，或者 <tt class="docutils literal">distance</tt> 不能被转化成浮点值，会抛出错误。</p>
<blockquote>
Python的类 <tt class="docutils literal">Tree</tt> 包含着整个系统聚类的结果。 <tt class="docutils literal">Tree</tt> 的对象可以通过
一个 <tt class="docutils literal">Node</tt> 的列表创建:</blockquote>
<pre class="code python literal-block">
   <span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">Node</span><span class="punctuation">,</span> <span class="name">Tree</span>
   <span class="operator">&gt;&gt;&gt;</span> <span class="name">nodes</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number float">0.2</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="literal number integer">3</span><span class="punctuation">,</span><span class="literal number float">0.5</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number integer">4</span><span class="punctuation">,</span><span class="literal number float">0.6</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="operator">-</span><span class="literal number integer">3</span><span class="punctuation">,</span><span class="literal number float">0.9</span><span class="punctuation">)]</span>
   <span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Tree</span><span class="punctuation">(</span><span class="name">nodes</span><span class="punctuation">)</span>
   <span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">tree</span>
   <span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">):</span> <span class="literal number float">0.2</span>
   <span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">):</span> <span class="literal number float">0.5</span>
   <span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">4</span><span class="punctuation">):</span> <span class="literal number float">0.6</span>
   <span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">3</span><span class="punctuation">):</span> <span class="literal number float">0.9</span>

<span class="literal string backtick">``</span><span class="name">Tree</span><span class="literal string backtick">``</span> <span class="error">的初始器会检查包含节点的列表是否是一个正确的系统聚类树的结果</span><span class="punctuation">:</span>
</pre>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nodes</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number float">0.2</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number float">0.5</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">Tree</span><span class="punctuation">(</span><span class="name">nodes</span><span class="punctuation">)</span>
<span class="name">Traceback</span> <span class="punctuation">(</span><span class="name">most</span> <span class="name">recent</span> <span class="name">call</span> <span class="name">last</span><span class="punctuation">):</span>
  <span class="name">File</span> <span class="literal string">&quot;&lt;stdin&gt;&quot;</span><span class="punctuation">,</span> <span class="name">line</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="operator word">in</span> <span class="error">?</span>
<span class="name exception">ValueError</span><span class="punctuation">:</span> <span class="name">Inconsistent</span> <span class="name">tree</span>
</pre>
<p>也可以使用中括号来对 <tt class="docutils literal">Tree</tt> 对象进行检索：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nodes</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number float">0.2</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number float">0.5</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Tree</span><span class="punctuation">(</span><span class="name">nodes</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">):</span> <span class="literal number float">0.2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">):</span> <span class="literal number float">0.5</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">):</span> <span class="literal number float">0.5</span>
</pre>
<p>因为 <tt class="docutils literal">Tree</tt> 对象是只读的，我们不能对 <tt class="docutils literal">Tree</tt> 对象中任何一个节点进行改变。然而，我们可以将其
转换成一个节点的列表，对列表进行操作，最后创建新的树。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Tree</span><span class="punctuation">([</span><span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number float">0.1</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number float">0.5</span><span class="punctuation">),</span> <span class="name">Node</span><span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">2</span><span class="punctuation">,</span><span class="literal number integer">3</span><span class="punctuation">,</span><span class="literal number float">0.9</span><span class="punctuation">)])</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">tree</span>
<span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">):</span> <span class="literal number float">0.1</span>
<span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">):</span> <span class="literal number float">0.5</span>
<span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">):</span> <span class="literal number float">0.9</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nodes</span> <span class="operator">=</span> <span class="name">tree</span><span class="punctuation">[:]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nodes</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="name">Node</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number float">0.2</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nodes</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">left</span> <span class="operator">=</span> <span class="literal number integer">2</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">Tree</span><span class="punctuation">(</span><span class="name">nodes</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">tree</span>
<span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">):</span> <span class="literal number float">0.2</span>
<span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">):</span> <span class="literal number float">0.5</span>
<span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">):</span> <span class="literal number float">0.9</span>
</pre>
<p>这个性质保证了``Tree`` 结果的正确性。</p>
<p>为了利用可视化工具，例如Java Treeview，来查看系统聚类树，最好对所有节点的距离进行标准化，
使其位于0和1之间。可以通过对 <tt class="docutils literal">Tree</tt> 对象调用 <tt class="docutils literal">scale</tt> 方法来实现这个功能：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span><span class="operator">.</span><span class="name">scale</span><span class="punctuation">()</span>
</pre>
<p>这个方法不需要任何参数，返回值是 <tt class="docutils literal">None</tt>.</p>
<p>经过系统聚类后，可以对 <tt class="docutils literal">Tree</tt> 对象进行剪接，将所有的元素分为 <em>k</em> 类：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clusterid</span> <span class="operator">=</span> <span class="name">tree</span><span class="operator">.</span><span class="name">cut</span><span class="punctuation">(</span><span class="name">nclusters</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>其中 <tt class="docutils literal">nclusters</tt> (默认是 <tt class="docutils literal">1</tt>) 是期望的类别数 <em>k</em>。这个方法会忽略树结构里面的
最高的 <em>k</em>−1 节点，最终形成 <em>k</em> 个独立的类别。对于 <em>k</em> 必须为正数，并且小于或者等于
元素的数目。这个方法会返回一个数组 <tt class="docutils literal">clusterid</tt> ,包含着每个元素对应的类。</p>
</div>
<div class="section" id="id188">
<h3>运行系统聚类</h3>
<p>为了进行系统聚类，可以用 <tt class="docutils literal">Bio.Cluster</tt> 中的 <tt class="docutils literal">treecluster</tt> 函数。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">treecluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">treecluster</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>包括以下参数:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt></dt>
<dd><p class="first last">包含所有元素的矩阵。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">缺失数据矩阵。若 <tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, 则 <tt class="docutils literal">data[i,j]</tt> 缺失。若 <tt class="docutils literal"><span class="pre">mask==None</span></tt>, 则明没有缺失数据。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">weight</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">权重矩阵。若 <tt class="docutils literal"><span class="pre">weight==None</span></tt>, 则假设所有的数据使用相同的权重。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">选择使用 <tt class="docutils literal">data</tt> 中的行 (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>), 或者列 (<tt class="docutils literal"><span class="pre">transpose==1</span></tt>)来计算距离.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (默认: <tt class="docutils literal">'m'</tt>)</dt>
<dd><p class="first last">选择节点间距离计算方法:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">method=='s'</span></tt>: 最小距离法</li>
<li><tt class="docutils literal"><span class="pre">method=='m'</span></tt>: 最大距离法</li>
<li><tt class="docutils literal"><span class="pre">method=='c'</span></tt>: 重心法</li>
<li><tt class="docutils literal"><span class="pre">method=='a'</span></tt>: 类平均法</li>
</ul>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (默认: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">选择距离函数 (具体见 <a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
</ul>
<p>为了对距离矩阵进行系统聚类，可以在调用 <tt class="docutils literal">treecluster</tt> 时，
用 <tt class="docutils literal">distancematrix</tt> 参数来代替 <tt class="docutils literal">data</tt> 参数：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">treecluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">treecluster</span><span class="punctuation">(</span><span class="name">distancematrix</span><span class="operator">=</span><span class="name">distance</span><span class="punctuation">)</span>
</pre>
<p>这种情况下，需要定义下列参数：</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">distancematrix</tt></dt>
<dd><p class="first last">元素两两间的距离矩阵，可以通过三种不同的方法提供：</p>
</dd>
</dl>
<ul>
<li><p class="first">提供一个2D的 Numerical Python 数组 (函数只会使用矩阵里左下角数据):</p>
<pre class="code python literal-block">
<span class="name">distance</span> <span class="operator">=</span> <span class="name">array</span><span class="punctuation">([[</span><span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">1.1</span><span class="punctuation">,</span> <span class="literal number float">2.3</span><span class="punctuation">],</span>
                  <span class="punctuation">[</span><span class="literal number float">1.1</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">],</span>
                  <span class="punctuation">[</span><span class="literal number float">2.3</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">,</span> <span class="literal number float">0.0</span><span class="punctuation">]])</span>
</pre>
</li>
<li><p class="first">输入一个一维的 Numerical Python 数组，包含了距离矩阵左下角的数据：</p>
<pre class="code python literal-block">
<span class="name">distance</span> <span class="operator">=</span> <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">1.1</span><span class="punctuation">,</span> <span class="literal number float">2.3</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">])</span>
</pre>
</li>
<li><p class="first">输入一个列表，包含距离矩阵左下角的数据：</p>
<pre class="code python literal-block">
<span class="name">distance</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">array</span><span class="punctuation">([]),</span>
            <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">1.1</span><span class="punctuation">]),</span>
            <span class="name">array</span><span class="punctuation">([</span><span class="literal number float">2.3</span><span class="punctuation">,</span> <span class="literal number float">4.5</span><span class="punctuation">])</span>
</pre>
<p>三种方法对应着同样的距离矩阵。由于 <tt class="docutils literal">treecluster</tt> 会对距离矩阵中的值进行随机洗牌，
如果后面需要调用这个距离矩阵，请在调用 <tt class="docutils literal">treecluster</tt> 之情，事先存到一个新的变量</p>
</li>
</ul>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt></dt>
<dd><p class="first last">选择节点间距离计算方法:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">method=='s'</span></tt>: 最小距离法</li>
<li><tt class="docutils literal"><span class="pre">method=='m'</span></tt>: 最大距离法</li>
<li><tt class="docutils literal"><span class="pre">method=='a'</span></tt>: 类平均法</li>
</ul>
<p>其中，最小距离法、最大距离法和类平均法可以只通过距离矩阵计算，而重心法却不行。</p>
</li>
</ul>
<p>当调用 <tt class="docutils literal"><span class="pre">treecluster``时,</span>&nbsp; ``data</tt> 或者 <tt class="docutils literal">distancematrix</tt> 总有一个必须为 <tt class="docutils literal">None</tt>。</p>
<p>函数返回一个 <tt class="docutils literal">Tree</tt> 对象，该对象包含着 (元素数目-1）个节点，当选择行作为聚类时，元素的
数目同行数一致；当使用列作为聚类时，元素的数目同列数一致。每个节点都意味着一对相邻连锁的
事件，其中节点的性质 <tt class="docutils literal">left</tt> 和 <tt class="docutils literal">right</tt> 包含着每个合并的元素或者子节点的编号， <tt class="docutils literal">distance</tt>
是两个合并元素或者子节点的距离。元素编号是从 0 到 (元素数目 − 1) , 而类别是从 -1 到 −(元素
数目 -1 ）</p>
</div>
</div>
<div class="section" id="self-organizing-maps">
<h2>15.5  Self-Organizing Maps</h2>
<p>Self-Organizing Maps (SOMs) 是由 Kohonen 在描述神经网络的时候发明的 (see for instance Kohonen, 1997 [<a class="reference external" href="#kohonen1997">24</a>]).
Tamayo (1999) 第一次讲 Self-Organizing Maps 应用到基因表达数据上。
[<a class="reference external" href="#tamayo1999">30</a>].</p>
<p>SOMs 根据某种拓扑结果将元素进行分类。通常选用的是矩形的拓扑结构。在SOMs生成的类别中，相邻的
两个类的拓扑结构相似度高于他们对其他的相似度。</p>
<p>计算SOM的第一步是随机分配数据向量到每个类别中，如果使用行进行聚类，那么每个数据向量中的元素
个数等于列数。</p>
<p>一个SOM 会一次读入一行，并且找到该向量最近的拓扑聚类结构。之后利用找到的数据向量对
这个类别的数据向量和相邻的类别的数据向量进行调整。调整如下：</p>
<div class="formula">
Δ<u><i>x</i></u><sub><span class="textrm">cell</span></sub> = <i>τ</i>⋅<span class="symbol">(</span><u><i>x</i></u><sub><span class="textrm">row</span></sub> − <u><i>x</i></u><sub><span class="textrm">cell</span></sub><span class="symbol">)</span>.
</div>
<p>参数 τ 会随着迭代次数增加而减少。可以用一个简单的线性函数来定义其与迭代次数的关系：</p>
<div class="formula">
<i>τ</i> = <i>τ</i><sub><span class="textrm">init</span></sub>⋅<span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span>1 − <span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>, 
</div>
<p>τ<sub>init</sub> 是指定的起始的 τ 值， <em>i</em> 是当前迭代的次数， <em>n</em> 是总的需要迭代的次数。
在迭代开始时，τ变化很快，然而在迭代末尾，变化越来越小。</p>
<p>所有在半径 <em>R</em> 内的类别都会在每次迭代中进行调整。半径也会随着迭代的增加而减小：</p>
<div class="formula">
<i>R</i> = <i>R</i><sub><span class="textrm">max</span></sub>⋅<span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span>1 − <span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator"><i>n</i></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>, 
</div>
<p>其中最大的半径定义为：</p>
<div class="formula">
<i>R</i><sub><span class="textrm">max</span></sub> = <span class="sqrt"><span class="radical">√</span><span class="ignored">(</span><span class="root"><i>N</i><span class="scripts"><sup class="script">2</sup><sub class="script"><i>x</i></sub></span> + <i>N</i><span class="scripts"><sup class="script">2</sup><sub class="script"><i>y</i></sub></span></span><span class="ignored">)</span></span>, 
</div>
<p>其中 (<em>N</em><sub>*x*</sub>, <em>N</em><sub>*y*</sub>) 是定义拓扑结构的矩形维度。</p>
<p>函数 <tt class="docutils literal">somcluster</tt> 可以用来在一个矩形的网格里计算 Self-Organizing Map。
首先，初始化一个随机数产生器。利用随机化产生器来对节点数据进行初始化。在SOM中，
基因或者芯片的调整顺序同样是随机的。用户可以定义总的SOM迭代的次数。</p>
<p>运行 <tt class="docutils literal">somcluster</tt>, 例如：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">somcluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clusterid</span><span class="punctuation">,</span> <span class="name">celldata</span> <span class="operator">=</span> <span class="name">somcluster</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>其中，可以定义一下参数:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt> (required)</dt>
<dd><p class="first last">包含所有元素的矩阵。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">缺失数据矩阵。若 <tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, 则 <tt class="docutils literal">data[i,j]</tt> 缺失。若 <tt class="docutils literal"><span class="pre">mask==None</span></tt>, 则明没有缺失数据。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">weight</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">权重矩阵。若 <tt class="docutils literal"><span class="pre">weight==None</span></tt>, 则假设所有的数据使用相同的权重。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">选择使用 <tt class="docutils literal">data</tt> 中的行 (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>), 或者列 (<tt class="docutils literal"><span class="pre">transpose==1</span></tt>)来聚类.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">nxgrid, nygrid</tt> (默认: <tt class="docutils literal">2, 1</tt>)</dt>
<dd><p class="first last">当Self-Organizing Map计算的时候，矩形的网格所包含的横向和纵向的格子。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">inittau</tt> (默认: <tt class="docutils literal">0.02</tt>)</dt>
<dd><p class="first last">SOM算法中，参数 τ 的初始值，默认是 0.02。 这个初始值同Michael Eisen’s Cluster/TreeView 一致。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">niter</tt> (默认: <tt class="docutils literal">1</tt>)</dt>
<dd><p class="first last">迭代运行的次数。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (默认: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">选择距离函数 (具体见 <a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
</ul>
<p>这个函数返回的是一个元组 <tt class="docutils literal">(clusterid, celldata)</tt>:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">clusterid</tt>:</dt>
<dd><p class="first last">一个两列的数组，行的数目等于待聚类元素的个数。每行包含着在矩形SOM网格中，将每个元素分配到的
格子的 <em>x</em> 和 <em>y</em> 的坐标。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">celldata</tt>:</dt>
<dd><p class="first last">当以行进行聚类时，生成的矩阵维度为 (<tt class="docutils literal">nxgrid</tt>, <tt class="docutils literal">nygrid</tt>, number of columns)；
当以列进行聚类时，生成的矩阵维度为 (<tt class="docutils literal">nxgrid</tt>, <tt class="docutils literal">nygrid</tt>, number of  rows)。
在这个矩阵里， <tt class="docutils literal"><span class="pre">[ix][iy]</span></tt> 表示着一个一维向量，其中用于计算该类中心的这基因的表达谱数据.</p>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id189">
<h2>15.6  主成分分析</h2>
<p>主成分分析 (PCA) 被广泛的用于分析多维数据，一个将主成分分析应用于表达谱数据的请见
Yeung and Ruzzo (2001) [<a class="reference external" href="#yeung2001">33</a>].</p>
<p>简而言之，PCA是一种坐标转换的方法，转换后的基础向量成为主成分，变换前的每行可以用主成分的
线性关系显示。主成分的选择是基于是残差尽可能的小的原则。例如，一个 <em>n</em> × 3 的数据矩阵可以表示为三维
空间内的一个椭圆球形的点的云。第一主成分是这个椭圆球形的最长轴，第二主成分是次长轴，第三主成分
是最短的轴。矩阵中，每一行都可以用主成分的线性关系展示。一般而言，为了对数据进行降维，只保留最
重要的几个主成分。剩余的残差认为是不可解释的方差。</p>
<p>可以通过计算数据的协方差矩阵的特征向量来得到主成分。每个主成分对应的特征值决定了
其在数据中代表的方差的大小。</p>
<p>在进行主成分分析前，矩阵的数据每一列都要减去其平均值。在上面椭圆球形云的例子中，数据在3D
空间中，围绕着其中心分布，而主成分则显示着每个点对其中心的变化。</p>
<p>函数 <tt class="docutils literal">pca</tt> 首先使用奇异值分解（singular value decomposition）来计算矩阵的特征值和
特征向量。奇异值分解使用的是Algol写的C语言的 <tt class="docutils literal">svd</tt> [<a class="reference external" href="#golub1971">16</a>], 利用的是
Householder bidiagonalization 和 QR 算法的变异。主成分，每个数据在主成分上的坐标和主成分
对应的特征值都会被计算出来，并按照特征值的降序排列。如果需要数据中心，则需要在调用 <tt class="docutils literal">pca</tt>
前，对每列数据减去其平均值。</p>
<p>将主成分分析应用于二维矩阵 <tt class="docutils literal">data</tt>,可以：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Cluster</span> <span class="keyword namespace">import</span> <span class="name">pca</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">columnmean</span><span class="punctuation">,</span> <span class="name">coordinates</span><span class="punctuation">,</span> <span class="name">components</span><span class="punctuation">,</span> <span class="name">eigenvalues</span> <span class="operator">=</span> <span class="name">pca</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
</pre>
<p>函数会返回一个元组 <tt class="docutils literal">columnmean, coordinates, components, eigenvalues</tt> :</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">columnmean</tt></dt>
<dd><p class="first last">包含 <tt class="docutils literal">data</tt> 每列均值的数组 .</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">coordinates</tt></dt>
<dd><p class="first last"><tt class="docutils literal">data</tt> 中每行数据在主成分上对应的坐标。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">components</tt></dt>
<dd><p class="first last">主成分</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">eigenvalues</tt></dt>
<dd><p class="first last">每个主成分对应的特征值</p>
</dd>
</dl>
</li>
</ul>
<p>原始的数据 <tt class="docutils literal">data</tt> 可以通过计算 <tt class="docutils literal">columnmean +&nbsp; dot(coordinates, components)</tt> 得到。</p>
</div>
<div class="section" id="cluster-treeview-type">
<h2>15.7  处理 Cluster/TreeView-type 文件</h2>
<p>Cluster/TreeView 是一个对基因表达数据可视化的工具。他们最初由 <a class="reference external" href="http://rana.lbl.gov">Michael
Eisen</a> 在 Stanford University 完成。<tt class="docutils literal">Bio.Cluster</tt>
包含着读写 Cluster/TreeView 对应的文件格式的函数。因此，将结果保存为该格式后，
可以用Treeview对结果进行直接的查看。我们推荐使用 Alok Saldanha 的
<a class="reference external" href="http://jtreeview.sourceforge.net/">http://jtreeview.sourceforge.net/</a>Java
TreeView 程序。这个软件可以显示系统聚类和 <em>k</em>-means 聚类的结果。</p>
<p>类 <tt class="docutils literal">Record</tt> 的一个对象包含着一个 Cluster/TreeView-type数据文件需要的所有信息。
为了将结果保存到一个 <tt class="docutils literal">Record</tt> 对象中，首先需要打开一个文件，并读取：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Cluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;mydatafile.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Cluster</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>两步操作使得你可以较灵活地操作不同来源的数据，例如：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">gzip</span> <span class="comment"># Python standard library</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">gzip</span><span class="operator">.</span><span class="name">open</span><span class="punctuation">(</span><span class="literal string">&quot;mydatafile.txt.gz&quot;</span><span class="punctuation">)</span>
</pre>
<p>来打开一个gzipped文件，或者利用</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">urllib</span> <span class="comment"># Python standard library</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name">urllib</span><span class="operator">.</span><span class="name">urlopen</span><span class="punctuation">(</span><span class="literal string">&quot;http://somewhere.org/mydatafile.txt&quot;</span><span class="punctuation">)</span>
</pre>
<p>来打开一个网络文件，然后调用 <tt class="docutils literal">read</tt>.</p>
<p><tt class="docutils literal">read</tt> 命令会读取一个由制表符分割的文本文件 <tt class="docutils literal">mydatafile.txt</tt>，文件包含着
符合Michael Eisen’s Cluster/TreeView格式的基因表达数据。具体的格式说明，可以参见
Cluster/TreeView手册，链接见 <a class="reference external" href="http://rana.lbl.gov/manuals/ClusterTreeView.pdf">Michael Eisen’s lab
website</a> 或者 <a class="reference external" href="http://bonsai.ims.u-tokyo.ac.jp/~mdehoon/software/cluster/cluster3.pdf">our
website</a>.</p>
<p>一个 <tt class="docutils literal">Record</tt> 对象有以下的性质:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">data</tt></dt>
<dd><p class="first last">包含基因表达数据的矩阵，每行为基因，每列为芯片。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">mask</tt></dt>
<dd><p class="first last">缺失值的整型数组。如果 <tt class="docutils literal"><span class="pre">mask[i,j]==0</span></tt>, 则 <tt class="docutils literal">data[i,j]</tt> 是缺失的. 如果 <tt class="docutils literal"><span class="pre">mask==None</span></tt>,
那么没有数据缺失。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">geneid</tt></dt>
<dd><p class="first last">包含每个基因的独特说明的列表 (例如 ORF 数目).</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">genename</tt></dt>
<dd><p class="first last">包含每个基因说明的列表（例如基因名）。如果文件中不包含该数据，
那么 <tt class="docutils literal">genename</tt> 被设为 <tt class="docutils literal">None</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">gweight</tt></dt>
<dd><p class="first last">计算表达谱数据中，基因间的距离使用的权重。如果文件中不含该信息，则
<tt class="docutils literal">gweight</tt> 为 <tt class="docutils literal">None</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">gorder</tt></dt>
<dd><p class="first last">期望输出文件中基因的排列的顺序。如果文件中不含该信息，则
<tt class="docutils literal">gorder</tt> 为``None``.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">expid</tt></dt>
<dd><p class="first last">包含每个芯片说明的列表，例如实验条件。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">eweight</tt></dt>
<dd><p class="first last">计算表达谱数据中，不同芯片间的距离使用的权重。如果文件中不含该信息，则
<tt class="docutils literal">eweight</tt> 为 <tt class="docutils literal">None</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">eorder</tt></dt>
<dd><p class="first last">期望输出文件中基因的排列的顺序。如果文件中不含该信息，则 <tt class="docutils literal">eorder</tt> 为  <tt class="docutils literal">None</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">uniqid</tt></dt>
<dd><p class="first last">用于代替文件中 UNIQID 的字符串.</p>
</dd>
</dl>
</li>
</ul>
<p>在载入 <tt class="docutils literal">Record</tt> 对象后，上述的每个性质可以直接读取和修改。例如，可以对
<tt class="docutils literal">record.data</tt> 直接取对数来对数据进行log转换。</p>
<div class="section" id="id190">
<h3>计算距离矩阵</h3>
<p>为了计算record中存储元素的距离矩阵，可以用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">matrix</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">distancematrix</span><span class="punctuation">()</span>
</pre>
<p>其中，包含以下参数：</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">选择对 <tt class="docutils literal">data</tt> 的行 (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>), 或者列 (<tt class="docutils literal"><span class="pre">transpose==1</span></tt>)计算距离。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (默认: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">选择合适的元素距离算法 (见
<a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
</ul>
<p>函数会返回一个距离矩阵，每行的列数等于行数。(see section
<a class="reference external" href="#subsec:distancematrix">15.1</a>).</p>
</div>
<div class="section" id="id191">
<h3>计算聚类中心</h3>
<p>为了计算存储在record中的元素的聚类中心，利用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">cdata</span><span class="punctuation">,</span> <span class="name">cmask</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">clustercentroids</span><span class="punctuation">()</span>
</pre>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">clusterid</tt> (默认: <tt class="docutils literal">None</tt>)</dt>
<dd><p class="first last">展示每个元素所属类的整型向量。如果缺少 <tt class="docutils literal">clusterid</tt>,默认所有的元素属于同一类。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (默认: <tt class="docutils literal">'a'</tt>)</dt>
<dd><p class="first last">选择使用算术平均值 (<tt class="docutils literal"><span class="pre">method=='a'</span></tt>) 或者中值 (<tt class="docutils literal"><span class="pre">method=='m'</span></tt>)来计算聚类中心。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">选择计算``data`` 的行 (<tt class="docutils literal"><span class="pre">transpose==0</span></tt>), 或者列 (<tt class="docutils literal"><span class="pre">transpose==1</span></tt>)计算中心。</p>
</dd>
</dl>
</li>
</ul>
<p>函数返回元组 <tt class="docutils literal">cdata, cmask</tt> ; 见 section
<a class="reference external" href="#subsec:clustercentroids">15.2</a> for a description.</p>
</div>
<div class="section" id="id192">
<h3>计算两类间的距离</h3>
<p>为了计算存储在record中的两类的距离，利用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">distance</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">clusterdistance</span><span class="punctuation">()</span>
</pre>
<p>其中，包含以下参数：</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">index1</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">第一个类别所包含的元素的列表。如果一个类别只包含一个元素 <em>i</em>
可以为一个列表 <tt class="docutils literal">[i]</tt>, 或者整数 <tt class="docutils literal">i</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">index2</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">第二个类别所包含的元素的列表。如果一个类别只包含一个元素 <em>i</em>
可以为一个列表 <tt class="docutils literal">[i]</tt>, 或者整数 <tt class="docutils literal">i</tt>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (默认: <tt class="docutils literal">'a'</tt>)</dt>
<dd><p class="first last">选择计算类别间距离的方法:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal">'a'</tt>: 使用两个聚类中心的距离 (算术平均值);</li>
<li><tt class="docutils literal">'m'</tt>: 使用两个聚类中心的距离 (中值);</li>
<li><tt class="docutils literal">'s'</tt>: 使用两类中最短的两个元素之间的距离;</li>
<li><tt class="docutils literal">'x'</tt>: 使用两类中最长的两个元素之间的距离;</li>
<li><tt class="docutils literal">'v'</tt>: 使用两类中两两元素距离的平均值作为距离。</li>
</ul>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (默认: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">选择使用的距离函数 (见
<a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">选择 使用 <tt class="docutils literal">data</tt> 的行 ( <tt class="docutils literal"><span class="pre">transpose==0</span></tt> ), 或者列 ( <tt class="docutils literal"><span class="pre">transpose==1</span></tt> )计算距离。</p>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id193">
<h3>进行系统聚类</h3>
<p>为了对存储在record中的数据进行系统聚类，利用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">tree</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">treecluster</span><span class="punctuation">()</span>
</pre>
<p>包含以下参数:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">选择使用行 ( <tt class="docutils literal"><span class="pre">transpose==0</span></tt> ) 或者列 ( <tt class="docutils literal"><span class="pre">transpose==1</span></tt> ) 用于聚类</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (默认: <tt class="docutils literal">'m'</tt>)</dt>
<dd><p class="first last">选择合适的节点距离计算方法:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">method=='s'</span></tt>: 最小距离法</li>
<li><tt class="docutils literal"><span class="pre">method=='m'</span></tt>: 最大距离法</li>
<li><tt class="docutils literal"><span class="pre">method=='c'</span></tt>: 重心法</li>
<li><tt class="docutils literal"><span class="pre">method=='a'</span></tt>: 类平均法</li>
</ul>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (默认: <tt class="docutils literal">'e'</tt>, Euclidean distance)</dt>
<dd><p class="first last">选择使用的距离函数(见 <a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt></dt>
<dd><p class="first last">选择使用基因或者芯片进行聚类，如果是 <tt class="docutils literal"><span class="pre">transpose==0</span></tt> , 则使用基因 (行) 进行聚类，如果使用
<tt class="docutils literal"><span class="pre">transpose==1</span></tt>, 芯片 (列) 用于聚类.</p>
</dd>
</dl>
</li>
</ul>
<p>函数返回 <tt class="docutils literal">Tree</tt> 对象。对象包含 (元素数目 − 1） 节点, 如果使用行进行聚类时，元素数目为总行数；
当使用列进行聚类时，元素数目为总列数。每个节点描述着一对节点连接，然而节点的性质 <tt class="docutils literal">left</tt> 和
<tt class="docutils literal">right</tt> 包含着相邻节点所有的元素和子节点数， <tt class="docutils literal">distance</tt> 显示着左右节点的距离。
元素从 0 到 (元素数目 − 1) 进行索引, 而类别从 -1 to −(元素数目−1)进行索引。</p>
</div>
<div class="section" id="k-means-or-k-medians">
<h3>进行 <em>k</em>-means or <em>k</em>-medians 聚类</h3>
<p>为了对存储在record中的元素进行 <em>k</em>-means 或者 <em>k</em>-medians 聚类，可以使用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clusterid</span><span class="punctuation">,</span> <span class="name">error</span><span class="punctuation">,</span> <span class="name">nfound</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">kcluster</span><span class="punctuation">()</span>
</pre>
<p>包含以下参数:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">nclusters</tt> (默认: <tt class="docutils literal">2</tt>)</dt>
<dd><p class="first last">类的数目 <em>k</em>.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (默认: <tt class="docutils literal">0</tt>)</dt>
<dd><p class="first last">选择 使用 <tt class="docutils literal">data</tt> 的行 ( <tt class="docutils literal"><span class="pre">transpose==0</span></tt> ), 或者列 ( <tt class="docutils literal"><span class="pre">transpose==1</span></tt> )计算距离。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">npass</tt> (默认: <tt class="docutils literal">1</tt>)</dt>
<dd><p class="first last"><em>k</em>-means/-medians 聚类算法运行的次数，每次运行使用不同的随机的起始值。
如果指定了 <tt class="docutils literal">initialid</tt> , <tt class="docutils literal">npass</tt> 的值会忽略，并且聚类算法只会运行一次。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">method</tt> (默认: <tt class="docutils literal">a</tt>)</dt>
<dd><p class="first last">指定确定聚类中心的方法:</p>
</dd>
</dl>
<ul class="simple">
<li><tt class="docutils literal"><span class="pre">method=='a'</span></tt>: 算数平均值 (<em>k</em>-means clustering);</li>
<li><tt class="docutils literal"><span class="pre">method=='m'</span></tt>: 中间值 (<em>k</em>-medians clustering).</li>
</ul>
<p>当指定 <tt class="docutils literal">method</tt> 使用其他值时，算法会采用算数平均值。</p>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (默认: <tt class="docutils literal">'e'</tt> , Euclidean distance)</dt>
<dd><p class="first last">选择使用的距离函数 (see
<a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
</ul>
<p>这个函数返回的是一个元组 <tt class="docutils literal">(clusterid, error, nfound)</tt> , 其中 <tt class="docutils literal">clusterid</tt> 是一个每行或则列对应的类的编号。
<tt class="docutils literal">error</tt> 是最优解的类内的距离和， <tt class="docutils literal">nfound</tt> 是最优解被发现的次数。</p>
</div>
<div class="section" id="self-organizing-map">
<h3>计算Self-Organizing Map</h3>
<p>可以利用以下命令，计算对存储在record中元素计算 Self-Organizing Map ：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">clusterid</span><span class="punctuation">,</span> <span class="name">celldata</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">somcluster</span><span class="punctuation">()</span>
</pre>
<p>包含以下参数:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">transpose</tt> (默认: <tt class="docutils literal">0</tt> )</dt>
<dd><p class="first last">选择 使用 <tt class="docutils literal">data</tt> 的行 ( <tt class="docutils literal"><span class="pre">transpose==0</span></tt> ), 或者列 ( <tt class="docutils literal"><span class="pre">transpose==1</span></tt> )计算距离.</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">nxgrid, nygrid</tt> (默认: <tt class="docutils literal">2, 1</tt>)</dt>
<dd><p class="first last">当Self-Organizing Map计算时，在矩形网格里的横向和纵向格子数目</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">inittau</tt> (默认: <tt class="docutils literal">0.02</tt>)</dt>
<dd><p class="first last">用于SOM算法的参数 τ 的初始值。默认的 <tt class="docutils literal">inittau</tt> 是0.02，同Michael Eisen’s Cluster/TreeView 程序中
使用的参数一致。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">niter</tt> (默认: <tt class="docutils literal">1</tt> )</dt>
<dd><p class="first last">迭代运行的次数。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">dist</tt> (默认: <tt class="docutils literal">'e'</tt> , Euclidean distance)</dt>
<dd><p class="first last">选择使用的距离函数(见 <a class="reference external" href="#sec:distancefunctions">15.1</a>).</p>
</dd>
</dl>
</li>
</ul>
<p>函数返回一个元组 <tt class="docutils literal">(clusterid, celldata)</tt> :</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">clusterid</tt>:</dt>
<dd><p class="first last">一个二维数组，行数同待聚类的元素数目相同。每行的内容对应着该元素在矩形SOM方格内 <em>x</em> 和 <em>y</em> 的坐标。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">celldata</tt>:</dt>
<dd><p class="first last">格式为一个矩阵，如果是对行聚类，内容为 ( <tt class="docutils literal">nxgrid</tt> , <tt class="docutils literal">nygrid</tt> , 列数)，如果是对列聚类，
那么内容为 ( <tt class="docutils literal">nxgrid</tt> , <tt class="docutils literal">nygrid</tt> , 行数) 。矩阵中，坐标 <tt class="docutils literal"><span class="pre">[ix][iy]</span></tt> 对应的是该坐标的网格里的
基因表达数据的聚类中心的一维向量。</p>
</dd>
</dl>
</li>
</ul>
</div>
<div class="section" id="id194">
<h3>保存聚类结果</h3>
<p>为了保存聚类结果，可以利用：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">save</span><span class="punctuation">(</span><span class="name">jobname</span><span class="punctuation">,</span> <span class="name">geneclusters</span><span class="punctuation">,</span> <span class="name">expclusters</span><span class="punctuation">)</span>
</pre>
<p>包含以下参数:</p>
<ul>
<li><dl class="first docutils">
<dt><tt class="docutils literal">jobname</tt></dt>
<dd><p class="first last">字符串 <tt class="docutils literal">jobname</tt> 作为保存的文件名。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">geneclusters</tt></dt>
<dd><p class="first last">这个参数指的是基因（以行聚类）的结果。在 <em>k</em>-means 聚类中，这个参数是一个一维的数组，包含着
每个基因对应的类别，可以通过 <tt class="docutils literal">kcluster</tt> 得到。在系统聚类中， <tt class="docutils literal">geneclusters</tt> 是一个 <tt class="docutils literal">Tree</tt> 对象。</p>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt><tt class="docutils literal">expclusters</tt></dt>
<dd><p class="first last">这个参数指的是实验条件（以列聚类）的结果。在 <em>k</em>-means 聚类中，这个参数是一个一维的数组，包含着
每个实验条件对应的类别，可以通过 <tt class="docutils literal">kcluster</tt> 得到。在系统聚类中， <tt class="docutils literal">geneclusters</tt> 是一个``Tree`` 对象。</p>
</dd>
</dl>
</li>
</ul>
<p>这个方法会生成文本文件 <tt class="docutils literal">jobname.cdt</tt>, <tt class="docutils literal">jobname.gtr</tt>, <tt class="docutils literal">jobname.atr</tt>, <tt class="docutils literal"><span class="pre">jobname*.kgg</span></tt>,
和/或 <tt class="docutils literal"><span class="pre">jobname*.kag</span></tt> 。 这些文件可以用于后续分析。如果 <tt class="docutils literal">geneclusters</tt> 和 <tt class="docutils literal">expclusters</tt>
都是 <tt class="docutils literal">None</tt> , 那这个方法只会生成 <tt class="docutils literal">jobname.cdt</tt> ; 这个文件可以被读取，生成一个新的 <tt class="docutils literal">Record</tt> 对象.</p>
</div>
</div>
<div class="section" id="id195">
<h2>15.8  示例</h2>
<p>以下是一个系统聚类的例子，其中使用最短距离法对基因进行聚类，用最大距离法对实验条件进行聚类。
由于使用 Euclidean 距离对基因进行聚类，因此需要将节点距离 <tt class="docutils literal">genetree</tt> 进行调整，使其处于0和1之间。
这种调整对于Java TreeView正确显示树结构也是很必须的。同时使用 uncentered correlation 对实验条件进行聚类。
在这种情况下，不需要任何的调整，因为 <tt class="docutils literal">exptree</tt> 中的结果已经位于0和2之间。 示例中使用的
文件 <tt class="docutils literal">cyano.txt</tt> 可以从 <tt class="docutils literal">data</tt> 文件夹中找到。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Cluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;cyano.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Cluster</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">genetree</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">treecluster</span><span class="punctuation">(</span><span class="name">method</span><span class="operator">=</span><span class="literal string">'s'</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">genetree</span><span class="operator">.</span><span class="name">scale</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exptree</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">treecluster</span><span class="punctuation">(</span><span class="name">dist</span><span class="operator">=</span><span class="literal string">'u'</span><span class="punctuation">,</span> <span class="name">transpose</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">save</span><span class="punctuation">(</span><span class="literal string">&quot;cyano_result&quot;</span><span class="punctuation">,</span> <span class="name">genetree</span><span class="punctuation">,</span> <span class="name">exptree</span><span class="punctuation">)</span>
</pre>
<p>这个命令会生成 <tt class="docutils literal">cyano_result.cdt</tt> , <tt class="docutils literal">cyano_result.gtr</tt> , 和 <tt class="docutils literal">cyano_result.atr</tt> 等文件。</p>
<p>同样的，也可以保存一个 <em>k</em>-means 聚类的结果:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Cluster</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;cyano.txt&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">Cluster</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="punctuation">(</span><span class="name">geneclusters</span><span class="punctuation">,</span> <span class="name">error</span><span class="punctuation">,</span> <span class="name">ifound</span><span class="punctuation">)</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">kcluster</span><span class="punctuation">(</span><span class="name">nclusters</span><span class="operator">=</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="name">npass</span><span class="operator">=</span><span class="literal number integer">1000</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="punctuation">(</span><span class="name">expclusters</span><span class="punctuation">,</span> <span class="name">error</span><span class="punctuation">,</span> <span class="name">ifound</span><span class="punctuation">)</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">kcluster</span><span class="punctuation">(</span><span class="name">nclusters</span><span class="operator">=</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name">npass</span><span class="operator">=</span><span class="literal number integer">100</span><span class="punctuation">,</span> <span class="name">transpose</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span><span class="operator">.</span><span class="name">save</span><span class="punctuation">(</span><span class="literal string">&quot;cyano_result&quot;</span><span class="punctuation">,</span> <span class="name">geneclusters</span><span class="punctuation">,</span> <span class="name">expclusters</span><span class="punctuation">)</span>
</pre>
<p>上述代码将生成文件 <tt class="docutils literal">cyano_result_K_G2_A2.cdt</tt> , <tt class="docutils literal">cyano_result_K_G2.kgg</tt> , 和 <tt class="docutils literal">cyano_result_K_A2.kag</tt> 。</p>
</div>
<div class="section" id="id196">
<h2>15.9  附加函数</h2>
<p><tt class="docutils literal">median(data)</tt> 返回一维数组 <tt class="docutils literal">data</tt> 的中值</p>
<p><tt class="docutils literal">mean(data)</tt> 返回一维数组 <tt class="docutils literal">data</tt> 的均值。</p>
<p><tt class="docutils literal">version()</tt> 返回使用的C聚类库的版本号。</p>
</div>
</div>
<div class="section" id="id197">
<h1>第16章 监督学习方法</h1>
<p>注意本章介绍的所有监督学习方法都需要先安装Numerical Python （numpy）。</p>
<div class="section" id="logistic">
<h2>16.1 Logistic 回归模型</h2>
<div class="section" id="id198">
<h3>16.1.1 背景和目的</h3>
<p>Logistic回归是一种监督学习方法，通过若干预测变量 <em>x</em><sub>*i*</sub> 的加权和来尝试将样本划分为 <em>K</em> 个不同类别。Logistic回归模型可用来计算预测变量的权重 β<sub>*i*</sub> 。在Biopython中，logistic回归模型目前只实现了二类别（ <em>K</em> = 2 ）分类，而预测变量的数量没有限制。</p>
<p>作为一个例子，我们试着预测细菌中的操纵子结构。一个操纵子是在一条DNA链上许多相邻基因组成的一个集合，可以被共同转录为一条mRNA分子。这条mRNA分子经翻译后产生多个不同的蛋白质。我们将以枯草芽孢杆菌的操纵子数据进行说明，它的一个操纵子平均包含2.4个基因。</p>
<p>作为理解细菌的基因调节的第一步，我们需要知道其操纵子的结构。枯草芽孢杆菌大约10%的基因操纵子结构已经通过实验获知。剩下的90%的基因操纵子结构可以通过一种监督学习方法来预测。</p>
<p>在这种监督学习方法中，我们需要选择某些与操纵子结构有关的容易度量的预测变量 <em>x</em><sub>*i*</sub> 。例如可以选择基因间碱基对距离来来作为其中一个预测变量。同一个操纵子中的相邻基因往往距离相对较近，而位于不同操纵子的相邻基因间通常具有更大的空间来容纳启动子和终止子序列。另一个预测变量可以基于基因表达量度。根据操纵子的定义，属于同一个操纵子的基因有相同的基因表达谱，而不同操纵子的两个基因的表达谱也不相同。在实际操作中，由于存在测量误差，对相同操纵子的基因表达轮廓的测量不会完全一致。为了测量基因表达轮廓的相似性，我们假设测量误差服从正态分布，然后计算对应的对数似然分值。</p>
<p>现在我们有了两个预测变量，可以据此预测在同一条DNA链上两个相邻基因是否属于相同的操纵子：
-  <em>x</em><sub>1</sub> ：两基因间的碱基对数；
-  <em>x</em><sub>2</sub> ：两基因表达谱的相似度。</p>
<p>在logistic回归模型中，我们使用这两个预测变量的加权和来计算一个联合得分 <em>S</em>：</p>
<div class="formula">
<i>S</i> = <i>β</i><sub>0</sub> + <i>β</i><sub>1</sub><i>x</i><sub>1</sub> + <i>β</i><sub>2</sub><i>x</i><sub>2</sub>.
</div>
<p>根据下面两组示例基因，logistic回归模型对参数 β<sub>0</sub> ， β<sub>1</sub>, β<sub>2</sub> 给出合适的值：
-  OP: 相邻基因，相同DNA链，属于相同操纵子；
-  NOP: 相邻基因，相同DNA链，属于不同操纵子。</p>
<p>在logistic回归模型中，属于某个类别的概率依赖于通过logistic函数得出的分数。对于这两类OP和NOP，相应概率可如下表述：</p>
<div class="formula">
<span class="environment"><span class="arrayrow">
<span class="arraycell align-r">
Pr(<span class="mathrm">OP</span>|<i>x</i><sub>1</sub>, <i>x</i><sub>2</sub>)
</span>
<span class="arraycell align-c">
 = 
</span>
<span class="arraycell align-l">
<span class="fraction"><span class="ignored">(</span><span class="numerator">exp(<i>β</i><sub>0</sub> + <i>β</i><sub>1</sub><i>x</i><sub>1</sub> + <i>β</i><sub>2</sub><i>x</i><sub>2</sub>)</span><span class="ignored">)/(</span><span class="denominator">1 + exp(<i>β</i><sub>0</sub> + <i>β</i><sub>1</sub><i>x</i><sub>1</sub> + <i>β</i><sub>2</sub><i>x</i><sub>2</sub>)</span><span class="ignored">)</span></span> <a class="eqnumber" name="eq:OP"> </a>
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-r">
 
</span>
<span class="arraycell align-c">
 
</span>
<span class="arraycell align-l">
 
</span>

</span>
<span class="arrayrow">
<span class="arraycell align-r">
Pr(<span class="mathrm">NOP</span>|<i>x</i><sub>1</sub>, <i>x</i><sub>2</sub>)
</span>
<span class="arraycell align-c">
 = 
</span>
<span class="arraycell align-l">
<span class="fraction"><span class="ignored">(</span><span class="numerator">1</span><span class="ignored">)/(</span><span class="denominator">1 + exp(<i>β</i><sub>0</sub> + <i>β</i><sub>1</sub><i>x</i><sub>1</sub> + <i>β</i><sub>2</sub><i>x</i><sub>2</sub>)</span><span class="ignored">)</span></span> <a class="eqnumber" name="eq:NOP"> </a>
</span>

</span>
</span>
</div>
<p>使用一组已知是否属于相同操纵子（OP类别）或不同操纵子（NOP类别）的基因对，通过最大化相应概率函数的对数似然值，我们可以计算权重 β<sub>0</sub>, β<sub>1</sub>, β<sub>2</sub> 。
(<a class="reference external" href="#eq:OP">16.2</a>) 和 (<a class="reference external" href="#eq:NOP">16.3</a>).</p>
</div>
<div class="section" id="id199">
<h3>16.1.2 训练logistic回归模型</h3>
<div class="system-message">
<p class="system-message-title">System Message: ERROR/3 (<tt class="docutils">allinone.rst</tt>, line 13090)</p>
Document or section may not begin with a transition.</div>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr><td>表16.1： 已知类别(OP or NOP)的相邻基因对.如果两个基因相重叠，其基因间距离为负值</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="19%" />
<col width="35%" />
<col width="38%" />
<col width="8%" />
</colgroup>
<tbody valign="top">
<tr><td>基因对</td>
<td>基因间距离 (<em>x</em><sub>1</sub>)</td>
<td>基因表达得分 (<em>x</em><sub>2</sub>)</td>
<td>类别</td>
</tr>
<tr><td><em>cotJA</em> — <em>cotJB</em></td>
<td>-53</td>
<td>-200.78</td>
<td>OP</td>
</tr>
<tr><td><em>yesK</em> — <em>yesL</em></td>
<td>117</td>
<td>-267.14</td>
<td>OP</td>
</tr>
<tr><td><em>lplA</em> — <em>lplB</em></td>
<td>57</td>
<td>-163.47</td>
<td>OP</td>
</tr>
<tr><td><em>lplB</em> — <em>lplC</em></td>
<td>16</td>
<td>-190.30</td>
<td>OP</td>
</tr>
<tr><td><em>lplC</em> — <em>lplD</em></td>
<td>11</td>
<td>-220.94</td>
<td>OP</td>
</tr>
<tr><td><em>lplD</em> — <em>yetF</em></td>
<td>85</td>
<td>-193.94</td>
<td>OP</td>
</tr>
<tr><td><em>yfmT</em> — <em>yfmS</em></td>
<td>16</td>
<td>-182.71</td>
<td>OP</td>
</tr>
<tr><td><em>yfmF</em> — <em>yfmE</em></td>
<td>15</td>
<td>-180.41</td>
<td>OP</td>
</tr>
<tr><td><em>citS</em> — <em>citT</em></td>
<td>-26</td>
<td>-181.73</td>
<td>OP</td>
</tr>
<tr><td><em>citM</em> — <em>yflN</em></td>
<td>58</td>
<td>-259.87</td>
<td>OP</td>
</tr>
<tr><td><em>yfiI</em> — <em>yfiJ</em></td>
<td>126</td>
<td>-414.53</td>
<td>NOP</td>
</tr>
<tr><td><em>lipB</em> — <em>yfiQ</em></td>
<td>191</td>
<td>-249.57</td>
<td>NOP</td>
</tr>
<tr><td><em>yfiU</em> — <em>yfiV</em></td>
<td>113</td>
<td>-265.28</td>
<td>NOP</td>
</tr>
<tr><td><em>yfhH</em> — <em>yfhI</em></td>
<td>145</td>
<td>-312.99</td>
<td>NOP</td>
</tr>
<tr><td><em>cotY</em> — <em>cotX</em></td>
<td>154</td>
<td>-213.83</td>
<td>NOP</td>
</tr>
<tr><td><em>yjoB</em> — <em>rapA</em></td>
<td>147</td>
<td>-380.85</td>
<td>NOP</td>
</tr>
<tr><td><em>ptsI</em> — <em>splA</em></td>
<td>93</td>
<td>-291.13</td>
<td>NOP</td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<p>表`16.1 &lt;#table:training&gt;`__ 列出了枯草芽孢杆菌的一些基因对，这些基因的操纵子结构已知。让我们根据表中的这些数据来计算其logistic回归模型：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">LogisticRegression</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">xs</span> <span class="operator">=</span> <span class="punctuation">[[</span><span class="operator">-</span><span class="literal number integer">53</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">200.78</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">117</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">267.14</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">57</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">163.47</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">16</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">190.30</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">11</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">220.94</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">85</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">193.94</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">16</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">182.71</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">15</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">180.41</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">26</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">181.73</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">58</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">259.87</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">126</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">414.53</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">191</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">249.57</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">113</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">265.28</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">145</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">312.99</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">154</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">213.83</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">147</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">380.85</span><span class="punctuation">],</span>
          <span class="punctuation">[</span><span class="literal number integer">93</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">291.13</span><span class="punctuation">]]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ys</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">1</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">,</span>
          <span class="literal number integer">0</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">train</span><span class="punctuation">(</span><span class="name">xs</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">)</span>
</pre>
<p>这里， <tt class="docutils literal">xs</tt> 和 <tt class="docutils literal">ys</tt> 是训练数据： <tt class="docutils literal">xs</tt> 包含每个基因对的预测变量， <tt class="docutils literal">ys</tt> 指定是否这个基因对属于相同操纵子（ <tt class="docutils literal">1</tt> ，类别OP）或不同操纵子（<tt class="docutils literal">0</tt>，类别NOP）。Logistic回归模型结果存储在 <tt class="docutils literal">model</tt> 中，包含权重 β<sub>0</sub>, β<sub>1</sub>, and β<sub>2</sub>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span><span class="operator">.</span><span class="name">beta</span>
<span class="punctuation">[</span><span class="literal number float">8.9830290157144681</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">0.035968960444850887</span><span class="punctuation">,</span> <span class="literal number float">0.02181395662983519</span><span class="punctuation">]</span>
</pre>
<p>注意 β<sub>1</sub> 是负的，这是因为具有更短基因间距离的基因对有更高的概率属于相同操纵子（类别OP）。另一方面， β<sub>2</sub> 为正，因为属于相同操纵子的基因对通常有更高的基因表达谱相似性得分。参数 β<sub>0</sub> 是正值是因为在这个训练数据中操纵子基因对占据大多数。</p>
<p>函数 <tt class="docutils literal">train</tt> 有两个可选参数： <tt class="docutils literal">update_fn</tt> 和 <tt class="docutils literal">typecode</tt> 。 <tt class="docutils literal">update_fn</tt> 可用来指定一个回调函数，以迭代数和对数似然值做参数。在这个例子中，我们可以使用这个回调函数追踪模型计算（使用Newton-Raphson迭代来最大化logistic回归模型的对数似然函数）进度：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">def</span> <span class="name function">show_progress</span><span class="punctuation">(</span><span class="name">iteration</span><span class="punctuation">,</span> <span class="name">loglikelihood</span><span class="punctuation">):</span>
        <span class="keyword">print</span> <span class="literal string">&quot;Iteration:&quot;</span><span class="punctuation">,</span> <span class="name">iteration</span><span class="punctuation">,</span> <span class="literal string">&quot;Log-likelihood function:&quot;</span><span class="punctuation">,</span> <span class="name">loglikelihood</span>
<span class="operator">&gt;&gt;&gt;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">train</span><span class="punctuation">(</span><span class="name">xs</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">,</span> <span class="name">update_fn</span><span class="operator">=</span><span class="name">show_progress</span><span class="punctuation">)</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">11.7835020695</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">7.15886767672</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">2</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">5.76877209868</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">3</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">5.11362294338</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">4</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">4.74870642433</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">5</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">4.50026077146</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">6</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">4.31127773737</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">7</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">4.16015043396</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">8</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">4.03561719785</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">9</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.93073282192</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">10</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.84087660929</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">11</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.76282560605</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">12</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.69425027154</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">13</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.6334178602</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">14</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.57900855837</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">15</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.52999671386</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">16</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.48557145163</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">17</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.44508206139</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">18</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.40799948447</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">19</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.3738885624</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">20</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.3423876581</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">21</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.31319343769</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">22</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.2860493346</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">23</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.2607366863</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">24</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.23706784091</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">25</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.21488073614</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">26</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.19403459259</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">27</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.17440646052</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">28</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.15588842703</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">29</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.13838533947</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">30</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.12181293595</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">31</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.10609629966</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">32</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.09116857282</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">33</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.07696988017</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">34</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.06344642288</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">35</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.05054971191</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">36</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.03823591619</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">37</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.02646530573</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">38</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.01520177394</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">39</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">3.00441242601</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">40</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.99406722296</span>
<span class="name">Iteration</span><span class="punctuation">:</span> <span class="literal number integer">41</span> <span class="name">Log</span><span class="operator">-</span><span class="name">likelihood</span> <span class="name">function</span><span class="punctuation">:</span> <span class="operator">-</span><span class="literal number float">2.98413867259</span>
</pre>
<p>一旦对数似然函数得分增加值小于0.01，迭代将终止。如果在500次迭代后还没有到达收敛， <tt class="docutils literal">train</tt> 函数返回并抛出一个 <tt class="docutils literal">AssertionError</tt> 。</p>
<p>可选的关键字 <tt class="docutils literal">typecode</tt> 几乎可以一直忽略。这个关键字允许用户选择要使用的数值矩阵类型。当为了避免大数据计算的内存问题时，可能有必要使用单精度浮点数（Float8，Float16等等）而不是默认的double型。</p>
</div>
<div class="section" id="id200">
<h3>16.1.3 使用logistic回归模型进行分类</h3>
<p>调用 <tt class="docutils literal">classify</tt> 函数可以进行分类。给定一个logistic回归模型和 <em>x</em><sub>1</sub> 和 <em>x</em><sub>2</sub> 的值（例如，操纵子结构未知的基因对）， <tt class="docutils literal">classify</tt> 函数返回 <tt class="docutils literal">1</tt> 或 <tt class="docutils literal">0</tt> ，分别对应类别OP和NOP。例如，考虑基因对 <em>yxcE</em> ， <em>yxcD</em> 和 <em>yxiB</em> ， <em>yxiA</em> ：</p>
<hr class="docutils" />
<table border="1" class="docutils">
<colgroup>
<col width="100%" />
</colgroup>
<tbody valign="top">
<tr><td>表16.2：操纵子状态未知的相邻基因对</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="20%" />
<col width="38%" />
<col width="41%" />
</colgroup>
<tbody valign="top">
<tr><td>基因对</td>
<td>基因间距离 <em>x</em><sub>1</sub></td>
<td>基因表达得分 <em>x</em><sub>2</sub></td>
</tr>
<tr><td><em>yxcE</em> — <em>yxcD</em></td>
<td>6</td>
<td>-173.143442352</td>
</tr>
<tr><td><em>yxiB</em> — <em>yxiA</em></td>
<td>309</td>
<td>-271.005880394</td>
</tr>
</tbody>
</table>
<hr class="docutils" />
<p>Logistic回归模型预测 <em>yxcE</em> ， <em>yxcD</em> 属于相同操纵子（类别OP），而 <em>yxiB</em> ， <em>yxiA</em> 属于不同操纵子:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;yxcE, yxcD:&quot;</span><span class="punctuation">,</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">,</span><span class="operator">-</span><span class="literal number float">173.143442352</span><span class="punctuation">])</span>
<span class="name">yxcE</span><span class="punctuation">,</span> <span class="name">yxcD</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;yxiB, yxiA:&quot;</span><span class="punctuation">,</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">309</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">271.005880394</span><span class="punctuation">])</span>
<span class="name">yxiB</span><span class="punctuation">,</span> <span class="name">yxiA</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
</pre>
<p>（这个结果和生物学文献报道的一致）。</p>
<p>为了确定这个预测的可信度，我们可以调用 <tt class="docutils literal">calculate</tt> 函数来获得类别OP和NOP的概率(公式
(<a class="reference external" href="#eq:OP">16.2</a>) 和 (<a class="reference external" href="#eq:NOP">16.3</a>))。对于 <em>yxcE</em>, <em>yxcD</em> 我们发现</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">q</span><span class="punctuation">,</span> <span class="name">p</span> <span class="operator">=</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">calculate</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">,</span><span class="operator">-</span><span class="literal number float">173.143442352</span><span class="punctuation">])</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;class OP: probability =&quot;</span><span class="punctuation">,</span> <span class="name">p</span><span class="punctuation">,</span> <span class="literal string">&quot;class NOP: probability =&quot;</span><span class="punctuation">,</span> <span class="name">q</span>
<span class="keyword">class</span> <span class="name class">OP</span><span class="punctuation">:</span> <span class="name">probability</span> <span class="operator">=</span> <span class="literal number float">0.993242163503</span> <span class="keyword">class</span> <span class="name class">NOP</span><span class="punctuation">:</span> <span class="name">probability</span> <span class="operator">=</span> <span class="literal number float">0.00675783649744</span>
</pre>
<p>对于 <em>yxiB</em> ， <em>yxiA</em></p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">q</span><span class="punctuation">,</span> <span class="name">p</span> <span class="operator">=</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">calculate</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="punctuation">[</span><span class="literal number integer">309</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">271.005880394</span><span class="punctuation">])</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;class OP: probability =&quot;</span><span class="punctuation">,</span> <span class="name">p</span><span class="punctuation">,</span> <span class="literal string">&quot;class NOP: probability =&quot;</span><span class="punctuation">,</span> <span class="name">q</span>
<span class="keyword">class</span> <span class="name class">OP</span><span class="punctuation">:</span> <span class="name">probability</span> <span class="operator">=</span> <span class="literal number float">0.000321211251817</span> <span class="keyword">class</span> <span class="name class">NOP</span><span class="punctuation">:</span> <span class="name">probability</span> <span class="operator">=</span> <span class="literal number float">0.999678788748</span>
</pre>
<p>为了确定回归模型的预测精确性，我们可以把模型应用到训练数据上：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">ys</span><span class="punctuation">)):</span>
        <span class="keyword">print</span> <span class="literal string">&quot;True:&quot;</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">],</span> <span class="literal string">&quot;Predicted:&quot;</span><span class="punctuation">,</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">xs</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">])</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
</pre>
<p>这表示除一个基因对外其他所有预测都是正确的。Leave-one-out分析可以对预测精确性给出一个更可信的估计。Leave-one-out是指从训练数据中移除要预测的基因重新计算模型，再用该模型进行预测比对：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">ys</span><span class="punctuation">)):</span>
        <span class="name">model</span> <span class="operator">=</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">train</span><span class="punctuation">(</span><span class="name">xs</span><span class="punctuation">[:</span><span class="name">i</span><span class="punctuation">]</span><span class="operator">+</span><span class="name">xs</span><span class="punctuation">[</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">:],</span> <span class="name">ys</span><span class="punctuation">[:</span><span class="name">i</span><span class="punctuation">]</span><span class="operator">+</span><span class="name">ys</span><span class="punctuation">[</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">:])</span>
        <span class="keyword">print</span> <span class="literal string">&quot;True:&quot;</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">],</span> <span class="literal string">&quot;Predicted:&quot;</span><span class="punctuation">,</span> <span class="name">LogisticRegression</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">xs</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">])</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
</pre>
<p>Leave-one-out分析显示这个logistic回归模型的预测只对两个基因对不正确，对应预测精确度为88%。</p>
</div>
<div class="section" id="id201">
<h3>16.1.4 Logistic回归，线性判别分析和支持向量机</h3>
<p>Logistic回归模型类似于线性判别分析。在线性判别分析中，类别概率同样可由方程(<a class="reference external" href="#eq:OP">16.2</a>) and (<a class="reference external" href="#eq:NOP">16.3</a>)给出。但是，不是直接估计系数β，我们首先对预测变量 <em>x</em> 拟合一个正态分布。然后通过这个正态分布的平均值和方差计算系数β。如果 <em>x</em> 的分布确实是正态的，线性判别分析将比logistic回归模型有更好的性能。另一方面，logistic回归模型对于偏态到正态的广泛分布更加强健。</p>
<p>另一个相似的方法是应用线性核函数的支持向量机。这样的SVM也使用一个预测变量的线性组合，但是是从靠近类别之间的边界区域的预测变量 <em>x</em> 来估计系数β。如果logistic回归模型(公式 (<a class="reference external" href="#eq:OP">16.2</a>) 和 (<a class="reference external" href="#eq:NOP">16.3</a>))很好的描述了远离边界区域的 <em>x</em> ，我们可以期望logistic回归模型优于线性核函数SVM，因为它应用了更多数据。如果不是，SVM可能更好。</p>
<p>Trevor Hastie, Robert Tibshirani, and Jerome Friedman: The Elements of Statistical Learning. Data Mining, Inference, and Prediction.(统计学习基础:数据挖掘、推理与预测) Springer Series in Statistics, 2001. 4.4章.</p>
</div>
</div>
<div class="section" id="k-knn">
<h2>16.2 <em>k</em>-最近邻居法（ <em>KNN</em> ）</h2>
<div class="section" id="id202">
<h3>16.2.1 背景和目的</h3>
<p>最近邻居法是一种不需要将数据拟合到一个模型的监督学习算法。数据点是基于训练数据集的 <em>k</em> 个最近邻居类别进行分类的。</p>
<p>在Biopython中， <em>KNN</em> 方法可在 <tt class="docutils literal">Bio.KNN</tt> 中获得。我们使用 <cite>16.1 &lt;#sec:LogisticRegression&gt;</cite> 同样的操纵子数据集来说明Biopython中 <em>KNN</em> 方法的用法。</p>
</div>
<div class="section" id="knn">
<h3>16.2.2 初始化一个 <em>KNN</em> 模型</h3>
<p>使用表`16.1 &lt;#table:training&gt;`__中的数据，我们创建和初始化一个*KNN*模型：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">kNN</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">k</span> <span class="operator">=</span> <span class="literal number integer">3</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">model</span> <span class="operator">=</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">train</span><span class="punctuation">(</span><span class="name">xs</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">,</span> <span class="name">k</span><span class="punctuation">)</span>
</pre>
<p>这里 <tt class="docutils literal">xs</tt> 和 <tt class="docutils literal">ys</tt> 和 <a class="reference external" href="#subsec:LogisticRegressionTraining">16.1.2</a> 中的相同。 <tt class="docutils literal">k</tt> 是分类中的邻居数 <em>k</em> 。对于二分类，为 <em>k</em> 选择一个奇数可以避免tied votes。函数名 <tt class="docutils literal">train</tt> 在这里有点不合适，因为就没有训练模型：这个函数仅仅是用来存储模型变量 <tt class="docutils literal">xs</tt> ， <tt class="docutils literal">ys</tt> 和 <tt class="docutils literal">k</tt> 。</p>
</div>
<div class="section" id="id203">
<h3>16.2.3 使用*KNN* 模型来分类</h3>
<p>应用 <em>KNN</em> 模型对新数据进行分类，我们使用 <tt class="docutils literal">classify</tt> 函数。这个函数以一个数据点(<em>x</em><sub>1</sub>,<em>x</em><sub>2</sub>)为参数并在训练数据集 <tt class="docutils literal">xs</tt> 中寻找 <em>k</em> -最近邻居。然后基于在这 <em>k</em> 个邻居中出现最多的类别（ <tt class="docutils literal">ys</tt> ）来对数据点(<em>x</em><sub>1</sub>,<em>x</em><sub>2</sub>)进行分类。</p>
<p>对于基因对 <em>yxcE</em> 、 <em>yxcD</em> 和 <em>yxiB</em> 、 <em>yxiA</em> 的例子，我们发现：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">173.143442352</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;yxcE, yxcD:&quot;</span><span class="punctuation">,</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">x</span><span class="punctuation">)</span>
<span class="name">yxcE</span><span class="punctuation">,</span> <span class="name">yxcD</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">309</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">271.005880394</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;yxiB, yxiA:&quot;</span><span class="punctuation">,</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">x</span><span class="punctuation">)</span>
<span class="name">yxiB</span><span class="punctuation">,</span> <span class="name">yxiA</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
</pre>
<p>和logistic回归模型一致，<em>yxcE</em>,*yxcD*被归为一类（类别OP），<em>yxiB</em>,*yxiA*属于不同操纵子。</p>
<p>函数 <tt class="docutils literal">classify</tt> 可以指定距离函数和权重函数作为可选参数。距离函数影响作为最近邻居的 <em>k</em> 个类别的选择，因为这些到查询点（ <em>x</em> ， <em>y</em> ）有最小距离的类别被定义为邻居。默认使用欧几里德距离。另外，我们也可以如示例中的使用曼哈顿距离：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">def</span> <span class="name function">cityblock</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">,</span> <span class="name">x2</span><span class="punctuation">):</span>
<span class="operator">...</span>    <span class="keyword">assert</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">)</span><span class="operator">==</span><span class="literal number integer">2</span>
<span class="operator">...</span>    <span class="keyword">assert</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">x2</span><span class="punctuation">)</span><span class="operator">==</span><span class="literal number integer">2</span>
<span class="operator">...</span>    <span class="name">distance</span> <span class="operator">=</span> <span class="name builtin">abs</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">-</span><span class="name">x2</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">])</span> <span class="operator">+</span> <span class="name builtin">abs</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span><span class="operator">-</span><span class="name">x2</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">])</span>
<span class="operator">...</span>    <span class="keyword">return</span> <span class="name">distance</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">173.143442352</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;yxcE, yxcD:&quot;</span><span class="punctuation">,</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">x</span><span class="punctuation">,</span> <span class="name">distance_fn</span> <span class="operator">=</span> <span class="name">cityblock</span><span class="punctuation">)</span>
<span class="name">yxcE</span><span class="punctuation">,</span> <span class="name">yxcD</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
</pre>
<p>权重函数可以用于权重投票。例如，相比于相邻较远的邻居，我们可能想给更近的邻居一个更高的权重：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">def</span> <span class="name function">weight</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">,</span> <span class="name">x2</span><span class="punctuation">):</span>
<span class="operator">...</span>    <span class="keyword">assert</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">)</span><span class="operator">==</span><span class="literal number integer">2</span>
<span class="operator">...</span>    <span class="keyword">assert</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">x2</span><span class="punctuation">)</span><span class="operator">==</span><span class="literal number integer">2</span>
<span class="operator">...</span>    <span class="keyword">return</span> <span class="name">exp</span><span class="punctuation">(</span><span class="operator">-</span><span class="name builtin">abs</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">]</span><span class="operator">-</span><span class="name">x2</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">])</span> <span class="operator">-</span> <span class="name builtin">abs</span><span class="punctuation">(</span><span class="name">x1</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span><span class="operator">-</span><span class="name">x2</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]))</span>
<span class="operator">...</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">173.143442352</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;yxcE, yxcD:&quot;</span><span class="punctuation">,</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">x</span><span class="punctuation">,</span> <span class="name">weight_fn</span> <span class="operator">=</span> <span class="name">weight</span><span class="punctuation">)</span>
<span class="name">yxcE</span><span class="punctuation">,</span> <span class="name">yxcD</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
</pre>
<p>默认所有邻居有相同权重。</p>
<p>为了确定这些预测的置信度，我们可以调用函数 <tt class="docutils literal">calculate</tt> 来计算分配到类别OP和NOP的总权重。对于默认的加权方案，这样减少了每个分类的邻居数量。对于 <em>yxcE</em> ， <em>yxcD</em> ， 我们发现</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">173.143442352</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">weight</span> <span class="operator">=</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">calculate</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">x</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;class OP: weight =&quot;</span><span class="punctuation">,</span> <span class="name">weight</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">],</span> <span class="literal string">&quot;class NOP: weight =&quot;</span><span class="punctuation">,</span> <span class="name">weight</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="keyword">class</span> <span class="name class">OP</span><span class="punctuation">:</span> <span class="name">weight</span> <span class="operator">=</span> <span class="literal number float">0.0</span> <span class="keyword">class</span> <span class="name class">NOP</span><span class="punctuation">:</span> <span class="name">weight</span> <span class="operator">=</span> <span class="literal number float">3.0</span>
</pre>
<p>这意味着 <tt class="docutils literal">x1</tt> ， <tt class="docutils literal">x2</tt> 的所有三个邻居都属于NOP类别。对另一个例子 <em>yesK</em> ， <em>yesL</em> 我们发现</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal number integer">117</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">267.14</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">weight</span> <span class="operator">=</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">calculate</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">x</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="literal string">&quot;class OP: weight =&quot;</span><span class="punctuation">,</span> <span class="name">weight</span><span class="punctuation">[</span><span class="literal number integer">0</span><span class="punctuation">],</span> <span class="literal string">&quot;class NOP: weight =&quot;</span><span class="punctuation">,</span> <span class="name">weight</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">]</span>
<span class="keyword">class</span> <span class="name class">OP</span><span class="punctuation">:</span> <span class="name">weight</span> <span class="operator">=</span> <span class="literal number float">2.0</span> <span class="keyword">class</span> <span class="name class">NOP</span><span class="punctuation">:</span> <span class="name">weight</span> <span class="operator">=</span> <span class="literal number float">1.0</span>
</pre>
<p>这意思是两个邻居是操纵子对，另一个是非操纵子对</p>
<p>对于*KNN*方法的预测精确性，我们对训练数据应用模型：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">ys</span><span class="punctuation">)):</span>
        <span class="keyword">print</span> <span class="literal string">&quot;True:&quot;</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">],</span> <span class="literal string">&quot;Predicted:&quot;</span><span class="punctuation">,</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">xs</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">])</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
</pre>
<p>显示除了两个基因对所有预测都是正确的。Leave-one-out分析可以对预测精确性给出一个更可信的估计，这是通过从训练数据中移除要预测的基因，再重新计算模型实现：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">ys</span><span class="punctuation">)):</span>
        <span class="name">model</span> <span class="operator">=</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">train</span><span class="punctuation">(</span><span class="name">xs</span><span class="punctuation">[:</span><span class="name">i</span><span class="punctuation">]</span><span class="operator">+</span><span class="name">xs</span><span class="punctuation">[</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">:],</span> <span class="name">ys</span><span class="punctuation">[:</span><span class="name">i</span><span class="punctuation">]</span><span class="operator">+</span><span class="name">ys</span><span class="punctuation">[</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">:])</span>
        <span class="keyword">print</span> <span class="literal string">&quot;True:&quot;</span><span class="punctuation">,</span> <span class="name">ys</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">],</span> <span class="literal string">&quot;Predicted:&quot;</span><span class="punctuation">,</span> <span class="name">kNN</span><span class="operator">.</span><span class="name">classify</span><span class="punctuation">(</span><span class="name">model</span><span class="punctuation">,</span> <span class="name">xs</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">])</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">1</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">0</span>
<span class="name builtin pseudo">True</span><span class="punctuation">:</span> <span class="literal number integer">0</span> <span class="name">Predicted</span><span class="punctuation">:</span> <span class="literal number integer">1</span>
</pre>
<p>Leave-one-out分析显示这个 <em>KNN</em> 模型的预测正确17个基因对中的13个，对应预测精确度为76%。</p>
</div>
</div>
<div class="section" id="naive">
<h2>16.3 Naive贝叶斯</h2>
<p>这部分将描述模块 <tt class="docutils literal">Bio.NaiveBayes</tt> .</p>
</div>
<div class="section" id="id204">
<h2>16.4 最大熵</h2>
<p>这部分将描述模块 <tt class="docutils literal">Bio.MaximumEntropy</tt>.</p>
</div>
<div class="section" id="id205">
<h2>16.5  马尔科夫模型</h2>
<p>这部分将描述模块 <tt class="docutils literal">Bio.MarkovModel</tt> 和/或
<tt class="docutils literal">Bio.HMM.MarkovModel</tt> .</p>
</div>
</div>
<div class="section" id="graphicsgenomediagram">
<h1>第17章 Graphics模块中的基因组可视化包—GenomeDiagram</h1>
<p><tt class="docutils literal">Bio.Graphics</tt> 模块基于Python的第三方扩展库 <a class="reference external" href="http://www.reportlab.org">ReportLab</a> ，ReportLab主要生成PDF文件，同时也能生成EPS（Encapsulated Postscript）文件和SVG文件。ReportLa可以导出矢量图，如果安装依赖关系（Dependencies），比如 <a class="reference external" href="http://www.pythonware.com/products/pil/">PIL(Python Imaging Library)</a> ，ReportLab也可以导出JPEG, PNG, GIF, BMP和PICT格式的位图(Bitmap image)。</p>
<div class="section" id="genomediagram">
<h2>17.1  基因组可视化包—GenomeDiagram</h2>
<div class="section" id="id206">
<h3>17.1.1  GenomeDiagram简介</h3>
<p><tt class="docutils literal">Bio.Graphics.GenomeDiagram</tt> 包被整合到Biopython 1.50版之前，就已经是Biopython的独立模块。GenomeDiagram包首次出现在2006年Pritchard等人在Bioinformatics杂志的一篇文章 [<a class="reference external" href="#pritchard2006">2</a>] ，文中展示了一些图像示例，更多图像示例请查看GenomeDiagram手册 <a class="reference external" href="http://biopython.org/DIST/docs/GenomeDiagram/userguide.pdf">http://biopython.org/DIST/docs/GenomeDiagram/userguide.pdf</a> 。正如“GenomeDiagram”名称所指，它主要用于可视化全基因组(特别是原核生物基因组)，即可绘制线型图也可绘制环形图，Toth等人在2006年发表的文章 [<a class="reference external" href="#toth2006">3</a>] 中图2就是一个示例。Van der Auwera 等人在2009年发表的文章 [<a class="reference external" href="#vanderauwera2009">4</a>] 中图1和图2也进一步说明，GenomeDiagram适用于噬菌体、质粒和线粒体等微小基因组的可视化。</p>
<p>如果存储基因组信息的是从GenBank文件中下载的 <tt class="docutils literal">SeqRecord</tt> 话，它会包含许多 <tt class="docutils literal">SeqFeature</tt> ，那么用这个模块处理就很简单（详见
第 <a class="reference external" href="#chapter:SeqRecord">4</a> 章和第 <a class="reference external" href="#chapter:Bio.SeqIO">5</a> 章）。</p>
</div>
<div class="section" id="id207">
<h3>17.1.2 图形，轨迹,  特征集和特征</h3>
<p>GenomeDiagram使用一组嵌套的对象，图层中沿着水平轴或圆圈的图形对象（diagram object）表示一个序列（sequence）或序列区域（sequence region）。一个图形可以包含多个轨迹（track），呈现为横向排列或者环形放射图。这些轨迹的长度通常相等，代表相同的序列区域。可用一个轨迹表示基因的位置，另一个轨迹表示调节区域，第三个轨迹表示GC含量。可将最常用轨迹的特征打包为一个特征集（feature-sets）。CDS的特征可以用一个特征集，而tRNA的特征可以用另外一个特征集。这不是强制性的要求，你可以在diagram中用同样的特征集。如果diagram中用不同的特征集，修改一个特征会很容易，比如把所有tRNA的特征都变为红色，你只需选择tRNA的特征就行。</p>
<p>新建图形主要有两种方式。第一种是自上而下的方法（Top-Down），首先新建diagram对象，然后用diagram的方法来添加track(s)，最后用track的方法添加特征。第二种是自下而上的方法（Bottom-Up），首先单独新建对象，然后再将其进行组合。</p>
</div>
<div class="section" id="id208">
<h3>17.1.3 自上而下的实例</h3>
<p>我们用一个从GenBank文件中读取出来的 <tt class="docutils literal">SeqRecord</tt> 来绘制全基因组（详见第 <a class="reference external" href="#chapter:Bio.SeqIO">5</a> 章）。这里用鼠疫杆菌 <em>Yersinia pestis biovar Microtus</em> 的pPCP1质粒，元数据文件NC_005816.gb在Biopython中GenBank的tests目录下， <a class="reference external" href="http://biopython.org/SRC/biopython/Tests/GenBank/NC_005816.gb">NC_005816.gb</a> 也可下载</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib</span> <span class="keyword namespace">import</span> <span class="name">colors</span>
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.units</span> <span class="keyword namespace">import</span> <span class="name">cm</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">GenomeDiagram</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
</pre>
<p>这里用自上而下的方法，导入目标序列后，新建一个diagram，然后新建一个track，最后新建一个特征集（feature set）：</p>
<pre class="code python literal-block">
<span class="name">gd_diagram</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Diagram</span><span class="punctuation">(</span><span class="literal string">&quot;Yersinia pestis biovar Microtus plasmid pPCP1&quot;</span><span class="punctuation">)</span>
<span class="name">gd_track_for_features</span> <span class="operator">=</span> <span class="name">gd_diagram</span><span class="operator">.</span><span class="name">new_track</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;Annotated Features&quot;</span><span class="punctuation">)</span>
<span class="name">gd_feature_set</span> <span class="operator">=</span> <span class="name">gd_track_for_features</span><span class="operator">.</span><span class="name">new_set</span><span class="punctuation">()</span>
</pre>
<p>接下来的部分最有趣，提取 <tt class="docutils literal">SeqRecord</tt> 中每个基因的 <tt class="docutils literal">SeqFeature</tt> 对象，就会为diagram生成一个相应的特征（feature），将其颜色设置为蓝色，分别用深蓝和浅蓝表示。</p>
<pre class="code python literal-block">
<span class="keyword">for</span> <span class="name">feature</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">:</span>
    <span class="keyword">if</span> <span class="name">feature</span><span class="operator">.</span><span class="name">type</span> <span class="operator">!=</span> <span class="literal string">&quot;gene&quot;</span><span class="punctuation">:</span>
        <span class="comment">#Exclude this feature</span>
        <span class="keyword">continue</span>
    <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">gd_feature_set</span><span class="punctuation">)</span> <span class="operator">%</span> <span class="literal number integer">2</span> <span class="operator">==</span> <span class="literal number integer">0</span><span class="punctuation">:</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">blue</span>
    <span class="keyword">else</span><span class="punctuation">:</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">lightblue</span>
    <span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
</pre>
<p>创建导出文件需要两步，首先是 <tt class="docutils literal">draw</tt> 方法，它用ReportLab对象生成全部图形。然后是  <tt class="docutils literal">write</tt> 方法，将图形存储到格式文件。注意：输出文件格式不止一种。</p>
<pre class="code python literal-block">
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;linear&quot;</span><span class="punctuation">,</span> <span class="name">orientation</span><span class="operator">=</span><span class="literal string">&quot;landscape&quot;</span><span class="punctuation">,</span> <span class="name">pagesize</span><span class="operator">=</span><span class="literal string">'A4'</span><span class="punctuation">,</span>
                <span class="name">fragments</span><span class="operator">=</span><span class="literal number integer">4</span><span class="punctuation">,</span> <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">))</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PDF&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear.eps&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;EPS&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear.svg&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;SVG&quot;</span><span class="punctuation">)</span>
</pre>
<p>如果安装了依赖关系（Dependencies），也可以生成位图（Bitmap image），代码如下：</p>
<pre class="code python literal-block">
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear.png&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PNG&quot;</span><span class="punctuation">)</span>
</pre>
<p><img alt="image13" src="../images/plasmid_linear.png" /></p>
<p>注意，我们将代码中的  <tt class="docutils literal">fragments</tt> 变量设置为“4”，基因组就会被分为“4”个片段。</p>
<p>如果想要环形图，可以试试以下的代码：</p>
<pre class="code python literal-block">
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;circular&quot;</span><span class="punctuation">,</span> <span class="name">circular</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">pagesize</span><span class="operator">=</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">,</span><span class="literal number integer">20</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">),</span>
                <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">),</span> <span class="name">circle_core</span><span class="operator">=</span><span class="literal number float">0.7</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_circular.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PDF&quot;</span><span class="punctuation">)</span>
</pre>
<p><img alt="image14" src="../images/plasmid_circular.png" /></p>
<p>示例图不是非常精彩，但这仅仅是精彩的开始。</p>
</div>
<div class="section" id="id209">
<h3>17.1.4  自下而上的实例</h3>
<p>现在，用“自下而上”的方法来创建相同的图形。首先新建不同的对象（可以是任何顺序），然后将其组合。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib</span> <span class="keyword namespace">import</span> <span class="name">colors</span>
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.units</span> <span class="keyword namespace">import</span> <span class="name">cm</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">GenomeDiagram</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>

<span class="comment">#Create the feature set and its feature objects,</span>
<span class="name">gd_feature_set</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">FeatureSet</span><span class="punctuation">()</span>
<span class="keyword">for</span> <span class="name">feature</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">:</span>
    <span class="keyword">if</span> <span class="name">feature</span><span class="operator">.</span><span class="name">type</span> <span class="operator">!=</span> <span class="literal string">&quot;gene&quot;</span><span class="punctuation">:</span>
        <span class="comment">#Exclude this feature</span>
        <span class="keyword">continue</span>
    <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">gd_feature_set</span><span class="punctuation">)</span> <span class="operator">%</span> <span class="literal number integer">2</span> <span class="operator">==</span> <span class="literal number integer">0</span><span class="punctuation">:</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">blue</span>
    <span class="keyword">else</span><span class="punctuation">:</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">lightblue</span>
    <span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="comment">#(this for loop is the same as in the previous example)</span>

<span class="comment">#Create a track, and a diagram</span>
<span class="name">gd_track_for_features</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Track</span><span class="punctuation">(</span><span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;Annotated Features&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Diagram</span><span class="punctuation">(</span><span class="literal string">&quot;Yersinia pestis biovar Microtus plasmid pPCP1&quot;</span><span class="punctuation">)</span>

<span class="comment">#Now have to glue the bits together...</span>
<span class="name">gd_track_for_features</span><span class="operator">.</span><span class="name">add_set</span><span class="punctuation">(</span><span class="name">gd_feature_set</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">add_track</span><span class="punctuation">(</span><span class="name">gd_track_for_features</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>同样，利用 <tt class="docutils literal">draw</tt> 和 <tt class="docutils literal">write</tt> 方法来创建线形图或者环形图，结果应该完全相同（“draw”和“write”部分的代码见17.1.3）。</p>
</div>
<div class="section" id="feature">
<h3>17.1.5  简单的Feature</h3>
<p>以上示例中，创建diagram使用的 <tt class="docutils literal">SeqRecord</tt> 的 <tt class="docutils literal">SeqFeature</tt> 对象（ 详见 <a class="reference external" href="#sec:seq_features">4.3</a> 章节）。如果你不需要 <tt class="docutils literal">SeqFeature</tt> 对象，只将目标feature定位在坐标轴，仅需要创建minimal
<tt class="docutils literal">SeqFeature</tt> 对象，方法很简单，代码如下：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqFeature</span> <span class="keyword namespace">import</span> <span class="name">SeqFeature</span><span class="punctuation">,</span> <span class="name">FeatureLocation</span>
<span class="name">my_seq_feature</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="punctuation">(</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="literal number integer">50</span><span class="punctuation">,</span><span class="literal number integer">100</span><span class="punctuation">),</span><span class="name">strand</span><span class="operator">=+</span><span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>对于序列来说， <tt class="docutils literal">+1</tt> 代表正向， <tt class="docutils literal"><span class="pre">-1</span></tt> 代表反向，  <tt class="docutils literal">None</tt> 代表两者都有，下面举个简单的示例：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqFeature</span> <span class="keyword namespace">import</span> <span class="name">SeqFeature</span><span class="punctuation">,</span> <span class="name">FeatureLocation</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">GenomeDiagram</span>
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.units</span> <span class="keyword namespace">import</span> <span class="name">cm</span>

<span class="name">gdd</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Diagram</span><span class="punctuation">(</span><span class="literal string">'Test Diagram'</span><span class="punctuation">)</span>
<span class="name">gdt_features</span> <span class="operator">=</span> <span class="name">gdd</span><span class="operator">.</span><span class="name">new_track</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">greytrack</span><span class="operator">=</span><span class="name builtin pseudo">False</span><span class="punctuation">)</span>
<span class="name">gds_features</span> <span class="operator">=</span> <span class="name">gdt_features</span><span class="operator">.</span><span class="name">new_set</span><span class="punctuation">()</span>

<span class="comment">#Add three features to show the strand options,</span>
<span class="name">feature</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="punctuation">(</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="literal number integer">25</span><span class="punctuation">,</span> <span class="literal number integer">125</span><span class="punctuation">),</span> <span class="name">strand</span><span class="operator">=+</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="name">gds_features</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;Forward&quot;</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="name">feature</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="punctuation">(</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="literal number integer">150</span><span class="punctuation">,</span> <span class="literal number integer">250</span><span class="punctuation">),</span> <span class="name">strand</span><span class="operator">=</span><span class="name builtin pseudo">None</span><span class="punctuation">)</span>
<span class="name">gds_features</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;Strandless&quot;</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
<span class="name">feature</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="punctuation">(</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="literal number integer">275</span><span class="punctuation">,</span> <span class="literal number integer">375</span><span class="punctuation">),</span> <span class="name">strand</span><span class="operator">=-</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="name">gds_features</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;Reverse&quot;</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>

<span class="name">gdd</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">format</span><span class="operator">=</span><span class="literal string">'linear'</span><span class="punctuation">,</span> <span class="name">pagesize</span><span class="operator">=</span><span class="punctuation">(</span><span class="literal number integer">15</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">,</span><span class="literal number integer">4</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">),</span> <span class="name">fragments</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">,</span>
         <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="literal number integer">400</span><span class="punctuation">)</span>
<span class="name">gdd</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;GD_labels_default.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;pdf&quot;</span><span class="punctuation">)</span>
</pre>
<p>图形示例结果请见下一节图中的第一个图，缺省的feature为浅绿色。</p>
<p>注意，这里用 <tt class="docutils literal">name</tt> 参数作为feature的“说明文本”（caption text）。下文将会讲述更多细节。</p>
</div>
<div class="section" id="id210">
<h3>17.1.6  Feature说明</h3>
<p>下面代码中， <tt class="docutils literal">feature</tt> 作为 <tt class="docutils literal">SeqFeature</tt> 的对象添加到diagram。</p>
<pre class="code python literal-block">
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
</pre>
<p>前面的示例用 <tt class="docutils literal">SeqFeature</tt> 的注释为feature做了恰当的文字说明。 <tt class="docutils literal">SeqFeature</tt> 对象的限定符（qualifiers dictionary）缺省值是： <tt class="docutils literal">gene</tt>, <tt class="docutils literal">label</tt>, <tt class="docutils literal">name</tt>, <tt class="docutils literal">locus_tag</tt>, 和 <tt class="docutils literal">product</tt> 。简单地说，你可以定义一个名称：</p>
<pre class="code python literal-block">
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;My Gene&quot;</span><span class="punctuation">)</span>
</pre>
<p>每个feature标签的说明文本可以设置字体、位置和方向。说明文本默认的位置在图形符号（sigil）的左边，可选择在中间或者右边，线形图中文本的默认方向是45°旋转。</p>
<pre class="code python literal-block">
<span class="comment">#Large font, parallel with the track</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;green&quot;</span><span class="punctuation">,</span>
                           <span class="name">label_size</span><span class="operator">=</span><span class="literal number integer">25</span><span class="punctuation">,</span> <span class="name">label_angle</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">)</span>

<span class="comment">#Very small font, perpendicular to the track (towards it)</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;purple&quot;</span><span class="punctuation">,</span>
                           <span class="name">label_position</span><span class="operator">=</span><span class="literal string">&quot;end&quot;</span><span class="punctuation">,</span>
                           <span class="name">label_size</span><span class="operator">=</span><span class="literal number integer">4</span><span class="punctuation">,</span> <span class="name">label_angle</span><span class="operator">=</span><span class="literal number integer">90</span><span class="punctuation">)</span>

<span class="comment">#Small font, perpendicular to the track (away from it)</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;blue&quot;</span><span class="punctuation">,</span>
                           <span class="name">label_position</span><span class="operator">=</span><span class="literal string">&quot;middle&quot;</span><span class="punctuation">,</span>
                           <span class="name">label_size</span><span class="operator">=</span><span class="literal number integer">6</span><span class="punctuation">,</span> <span class="name">label_angle</span><span class="operator">=-</span><span class="literal number integer">90</span><span class="punctuation">)</span>
</pre>
<p>用前面示例的代码将这三个片段组合之后应该可以得到如下的结果：</p>
<p><img alt="image15" src="../images/GD_sigil_labels.png" /></p>
<p>除此之外，还可以设置“label_color”来调节标签的颜色（第 <a class="reference external" href="#sec:gd_nice_example">17.1.9</a> 章节也将用到这一步），这里没有进行演示。</p>
<p>示例中默认的字体很小，这是比较明智的，因为通常我们会把许多Feature同时展示，而不像这里只展示了几个比较大的feature。</p>
</div>
<div class="section" id="id211">
<h3>17.1.7  表示Feature的图形符号</h3>
<p>以上示例中Feature的图形符号（sigil）默认是一个方框（plain box），GenomeDiagram第一版中只有这一选项，后来GenomeDiagram被整合到Biopython1.50时，新增了箭头状的图形符号（sigil）。</p>
<pre class="code python literal-block">
<span class="comment">#Default uses a BOX sigil</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">)</span>

<span class="comment">#You can make this explicit:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;BOX&quot;</span><span class="punctuation">)</span>

<span class="comment">#Or opt for an arrow:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">)</span>
</pre>
<p>Biopython 1.61又新增3个图形形状（sigil）。</p>
<pre class="code python literal-block">
<span class="comment">#Box with corners cut off (making it an octagon)</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;OCTO&quot;</span><span class="punctuation">)</span>

<span class="comment">#Box with jagged edges (useful for showing breaks in contains)</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;JAGGY&quot;</span><span class="punctuation">)</span>

<span class="comment">#Arrow which spans the axis with strand used only for direction</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;BIGARROW&quot;</span><span class="punctuation">)</span>
</pre>
<p>下面就是这些新增的图形形状（sigil），多数的图形形状都在边界框（bounding box）内部，在坐标轴的上/下位置代表序列（Strand）方向的正/反向，或者上下跨越坐标轴，高度是其他图形形状的两倍。“BIGARROW”有所不同，它总是跨越坐标轴，方向由feature的序列决定。</p>
<p><img alt="image16" src="../images/GD_sigils.png" /></p>
</div>
<div class="section" id="id212">
<h3>17.1.8 箭头形状</h3>
<p>上一部分我们简单引出了箭头形状。还有两个选项可以对箭头形状进行设置：首先根据边界框的高度比例来设置箭杆宽度。</p>
<pre class="code python literal-block">
<span class="comment">#Full height shafts, giving pointed boxes:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;brown&quot;</span><span class="punctuation">,</span>
                           <span class="name">arrowshaft_height</span><span class="operator">=</span><span class="literal number float">1.0</span><span class="punctuation">)</span>
<span class="comment">#Or, thin shafts:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;teal&quot;</span><span class="punctuation">,</span>
                           <span class="name">arrowshaft_height</span><span class="operator">=</span><span class="literal number float">0.2</span><span class="punctuation">)</span>
<span class="comment">#Or, very thin shafts:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;darkgreen&quot;</span><span class="punctuation">,</span>
                           <span class="name">arrowshaft_height</span><span class="operator">=</span><span class="literal number float">0.1</span><span class="punctuation">)</span>
</pre>
<p>结果见下图：</p>
<p><img alt="image17" src="../images/GD_sigil_arrow_shafts.png" /></p>
<p>其次，根据边界框的高度比例设置箭头长度（默认为0.5或50%）：</p>
<pre class="code python literal-block">
<span class="comment">#Short arrow heads:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;blue&quot;</span><span class="punctuation">,</span>
                           <span class="name">arrowhead_length</span><span class="operator">=</span><span class="literal number float">0.25</span><span class="punctuation">)</span>
<span class="comment">#Or, longer arrow heads:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;orange&quot;</span><span class="punctuation">,</span>
                           <span class="name">arrowhead_length</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="comment">#Or, very very long arrow heads (i.e. all head, no shaft, so triangles):</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="literal string">&quot;red&quot;</span><span class="punctuation">,</span>
                           <span class="name">arrowhead_length</span><span class="operator">=</span><span class="literal number integer">10000</span><span class="punctuation">)</span>
</pre>
<p>结果见下图：</p>
<p><img alt="image18" src="../images/GD_sigil_arrow_heads.png" /></p>
<p>Biopython1.61新增 <tt class="docutils literal">BIGARROW</tt> 箭头形状，它经常跨越坐标轴，箭头指向”左边“代表”反向“，指向”右边“代表”正向“。</p>
<pre class="code python literal-block">
<span class="comment">#A large arrow straddling the axis:</span>
<span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;BIGARROW&quot;</span><span class="punctuation">)</span>
</pre>
<p>上述 <tt class="docutils literal">ARROW</tt> 形状中的箭杆和箭头设置选项都适用于 <tt class="docutils literal">BIGARROW</tt> 。</p>
</div>
<div class="section" id="id213">
<h3>17.1.9 完美示例</h3>
<p>回到”自上而下的示例 Section <a class="reference external" href="#sec:gd_top_down">17.1.3</a> 中鼠疫杆菌 <em>Yersinia pestis biovar
Microtus</em> 的pPCP1质粒，现在使用”图形符号“的高级选项。箭头表示基因，窄框穿越箭头表示限制性内切酶的切割位点。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib</span> <span class="keyword namespace">import</span> <span class="name">colors</span>
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.units</span> <span class="keyword namespace">import</span> <span class="name">cm</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">GenomeDiagram</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqFeature</span> <span class="keyword namespace">import</span> <span class="name">SeqFeature</span><span class="punctuation">,</span> <span class="name">FeatureLocation</span>

<span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>

<span class="name">gd_diagram</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Diagram</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>
<span class="name">gd_track_for_features</span> <span class="operator">=</span> <span class="name">gd_diagram</span><span class="operator">.</span><span class="name">new_track</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="literal string">&quot;Annotated Features&quot;</span><span class="punctuation">)</span>
<span class="name">gd_feature_set</span> <span class="operator">=</span> <span class="name">gd_track_for_features</span><span class="operator">.</span><span class="name">new_set</span><span class="punctuation">()</span>

<span class="keyword">for</span> <span class="name">feature</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">:</span>
    <span class="keyword">if</span> <span class="name">feature</span><span class="operator">.</span><span class="name">type</span> <span class="operator">!=</span> <span class="literal string">&quot;gene&quot;</span><span class="punctuation">:</span>
        <span class="comment">#Exclude this feature</span>
        <span class="keyword">continue</span>
    <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">gd_feature_set</span><span class="punctuation">)</span> <span class="operator">%</span> <span class="literal number integer">2</span> <span class="operator">==</span> <span class="literal number integer">0</span><span class="punctuation">:</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">blue</span>
    <span class="keyword">else</span><span class="punctuation">:</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">lightblue</span>
    <span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span>
                               <span class="name">color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">,</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span>
                               <span class="name">label_size</span> <span class="operator">=</span> <span class="literal number integer">14</span><span class="punctuation">,</span> <span class="name">label_angle</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">)</span>

<span class="comment">#I want to include some strandless features, so for an example</span>
<span class="comment">#will use EcoRI recognition sites etc.</span>
<span class="keyword">for</span> <span class="name">site</span><span class="punctuation">,</span> <span class="name">name</span><span class="punctuation">,</span> <span class="name">color</span> <span class="operator word">in</span> <span class="punctuation">[(</span><span class="literal string">&quot;GAATTC&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;EcoRI&quot;</span><span class="punctuation">,</span><span class="name">colors</span><span class="operator">.</span><span class="name">green</span><span class="punctuation">),</span>
                          <span class="punctuation">(</span><span class="literal string">&quot;CCCGGG&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;SmaI&quot;</span><span class="punctuation">,</span><span class="name">colors</span><span class="operator">.</span><span class="name">orange</span><span class="punctuation">),</span>
                          <span class="punctuation">(</span><span class="literal string">&quot;AAGCTT&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;HindIII&quot;</span><span class="punctuation">,</span><span class="name">colors</span><span class="operator">.</span><span class="name">red</span><span class="punctuation">),</span>
                          <span class="punctuation">(</span><span class="literal string">&quot;GGATCC&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;BamHI&quot;</span><span class="punctuation">,</span><span class="name">colors</span><span class="operator">.</span><span class="name">purple</span><span class="punctuation">)]:</span>
    <span class="name">index</span> <span class="operator">=</span> <span class="literal number integer">0</span>
    <span class="keyword">while</span> <span class="name builtin pseudo">True</span><span class="punctuation">:</span>
        <span class="name">index</span>  <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">find</span><span class="punctuation">(</span><span class="name">site</span><span class="punctuation">,</span> <span class="name">start</span><span class="operator">=</span><span class="name">index</span><span class="punctuation">)</span>
        <span class="keyword">if</span> <span class="name">index</span> <span class="operator">==</span> <span class="operator">-</span><span class="literal number integer">1</span> <span class="punctuation">:</span> <span class="keyword">break</span>
        <span class="name">feature</span> <span class="operator">=</span> <span class="name">SeqFeature</span><span class="punctuation">(</span><span class="name">FeatureLocation</span><span class="punctuation">(</span><span class="name">index</span><span class="punctuation">,</span> <span class="name">index</span><span class="operator">+</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">site</span><span class="punctuation">)))</span>
        <span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">,</span> <span class="name">name</span><span class="operator">=</span><span class="name">name</span><span class="punctuation">,</span>
                                   <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">label_size</span> <span class="operator">=</span> <span class="literal number integer">10</span><span class="punctuation">,</span>
                                   <span class="name">label_color</span><span class="operator">=</span><span class="name">color</span><span class="punctuation">)</span>
        <span class="name">index</span> <span class="operator">+=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">site</span><span class="punctuation">)</span>

<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;linear&quot;</span><span class="punctuation">,</span> <span class="name">pagesize</span><span class="operator">=</span><span class="literal string">'A4'</span><span class="punctuation">,</span> <span class="name">fragments</span><span class="operator">=</span><span class="literal number integer">4</span><span class="punctuation">,</span>
                <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">))</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear_nice.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PDF&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear_nice.eps&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;EPS&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_linear_nice.svg&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;SVG&quot;</span><span class="punctuation">)</span>

<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;circular&quot;</span><span class="punctuation">,</span> <span class="name">circular</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span> <span class="name">pagesize</span><span class="operator">=</span><span class="punctuation">(</span><span class="literal number integer">20</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">,</span><span class="literal number integer">20</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">),</span>
                <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">),</span> <span class="name">circle_core</span> <span class="operator">=</span> <span class="literal number float">0.5</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_circular_nice.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PDF&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_circular_nice.eps&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;EPS&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="literal string">&quot;plasmid_circular_nice.svg&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;SVG&quot;</span><span class="punctuation">)</span>
</pre>
<p>输出结果见下图：</p>
<p><img alt="image19" src="../images/plasmid_linear_nice.png" /></p>
<p><img alt="image20" src="../images/plasmid_circular_nice.png" /></p>
</div>
<div class="section" id="id214">
<h3>17.1.10 多重轨迹</h3>
<p>前面实例中都是单独的track，我们可以创建多个track，比如，一个track展示基因，另一个track展示重复序列。Proux等人2002年报道的文章 [<a class="reference external" href="#proux2002">5</a>] 中图6是一个很好的范例，下面我们将三个噬菌体基因组依次进行展示。首先需要三个噬菌体的GenBank文件。</p>
<ul class="simple">
<li><tt class="docutils literal">NC_002703</tt> – Lactococcus phage Tuc2009, 全基因组大小 (38347 bp)</li>
<li><tt class="docutils literal">AF323668</tt> – Bacteriophage bIL285, 全基因组大小(35538 bp)</li>
<li><tt class="docutils literal">NC_003212</tt> – <em>Listeria innocua</em> Clip11262,我们将仅关注前噬菌体5的全基因组 (长度大体相同).</li>
</ul>
<p>这三个文件可以从Entrez下载，详情请查阅 <a class="reference external" href="#sec:efetch">9.6</a> 。从三个噬菌体基因组文件中分离（slice）提取相关Features信息（请查阅 <a class="reference external" href="#sec:SeqRecord-slicing">4.6</a> ），保证前两个噬菌体的反向互补链与其起始点对齐，再次保存Feature(详情请查阅 <a class="reference external" href="#sec:SeqRecord-reverse-complement">4.8</a>)。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>

<span class="name">A_rec</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_002703.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">)</span>
<span class="name">B_rec</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;AF323668.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">)</span>
<span class="name">C_rec</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_003212.gbk&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;gb&quot;</span><span class="punctuation">)[</span><span class="literal number integer">2587879</span><span class="punctuation">:</span><span class="literal number integer">2625807</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">(</span><span class="name">name</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
</pre>
<p>图像中用不同颜色表示基因功能的差异。这需要编辑GenBank文件中每一个feature的颜色参数——就像用  <a class="reference external" href="http://www.sanger.ac.uk/resources/software/artemis/">Sanger’s Artemis
editor</a> 处理 ——才能被GenomeDiagram识别。但是，这里只需要硬编码（hard code）三个颜色列表。</p>
<p>上述GenBank文件中的注释信息与Proux所用的文件信息并不完全相同，他们还添加了一些未注释的基因。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.colors</span> <span class="keyword namespace">import</span> <span class="name">red</span><span class="punctuation">,</span> <span class="name">grey</span><span class="punctuation">,</span> <span class="name">orange</span><span class="punctuation">,</span> <span class="name">green</span><span class="punctuation">,</span> <span class="name">brown</span><span class="punctuation">,</span> <span class="name">blue</span><span class="punctuation">,</span> <span class="name">lightblue</span><span class="punctuation">,</span> <span class="name">purple</span>

<span class="name">A_colors</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">red</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">7</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">orange</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">orange</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">11</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">green</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">4</span> \
         <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">green</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">,</span> <span class="name">green</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">brown</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">blue</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">4</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">lightblue</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span> \
         <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">,</span> <span class="name">lightblue</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">purple</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span>
<span class="name">B_colors</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">red</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">6</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">8</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">orange</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">orange</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">21</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">green</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span> \
         <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">brown</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">4</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">blue</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">3</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">lightblue</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">3</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">purple</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span>
<span class="name">C_colors</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">30</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">green</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">brown</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">4</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">blue</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">,</span> <span class="name">blue</span><span class="punctuation">]</span> <span class="operator">+</span> <span class="punctuation">[</span><span class="name">lightblue</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">2</span> \
         <span class="operator">+</span> <span class="punctuation">[</span><span class="name">grey</span><span class="punctuation">]</span><span class="operator">*</span><span class="literal number integer">5</span>
</pre>
<p>接下来是“draw”方法，给diagram添加3个track。我们在示例中设置不同的开始/结束值来体现它们之间长度不等（Biopython 1.59及更高级的版本）。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">GenomeDiagram</span>

<span class="name">name</span> <span class="operator">=</span> <span class="literal string">&quot;Proux Fig 6&quot;</span>
<span class="name">gd_diagram</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Diagram</span><span class="punctuation">(</span><span class="name">name</span><span class="punctuation">)</span>
<span class="name">max_len</span> <span class="operator">=</span> <span class="literal number integer">0</span>
<span class="keyword">for</span> <span class="name">record</span><span class="punctuation">,</span> <span class="name">gene_colors</span> <span class="operator word">in</span> <span class="name builtin">zip</span><span class="punctuation">([</span><span class="name">A_rec</span><span class="punctuation">,</span> <span class="name">B_rec</span><span class="punctuation">,</span> <span class="name">C_rec</span><span class="punctuation">],</span> <span class="punctuation">[</span><span class="name">A_colors</span><span class="punctuation">,</span> <span class="name">B_colors</span><span class="punctuation">,</span> <span class="name">C_colors</span><span class="punctuation">]):</span>
    <span class="name">max_len</span> <span class="operator">=</span> <span class="name builtin">max</span><span class="punctuation">(</span><span class="name">max_len</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">))</span>
    <span class="name">gd_track_for_features</span> <span class="operator">=</span> <span class="name">gd_diagram</span><span class="operator">.</span><span class="name">new_track</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span>
                            <span class="name">name</span><span class="operator">=</span><span class="name">record</span><span class="operator">.</span><span class="name">name</span><span class="punctuation">,</span>
                            <span class="name">greytrack</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span>
                            <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">))</span>
    <span class="name">gd_feature_set</span> <span class="operator">=</span> <span class="name">gd_track_for_features</span><span class="operator">.</span><span class="name">new_set</span><span class="punctuation">()</span>

    <span class="name">i</span> <span class="operator">=</span> <span class="literal number integer">0</span>
    <span class="keyword">for</span> <span class="name">feature</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">:</span>
        <span class="keyword">if</span> <span class="name">feature</span><span class="operator">.</span><span class="name">type</span> <span class="operator">!=</span> <span class="literal string">&quot;gene&quot;</span><span class="punctuation">:</span>
            <span class="comment">#Exclude this feature</span>
            <span class="keyword">continue</span>
        <span class="name">gd_feature_set</span><span class="operator">.</span><span class="name">add_feature</span><span class="punctuation">(</span><span class="name">feature</span><span class="punctuation">,</span> <span class="name">sigil</span><span class="operator">=</span><span class="literal string">&quot;ARROW&quot;</span><span class="punctuation">,</span>
                                   <span class="name">color</span><span class="operator">=</span><span class="name">gene_colors</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">],</span> <span class="name">label</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">,</span>
                                   <span class="name">name</span> <span class="operator">=</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">),</span>
                                   <span class="name">label_position</span><span class="operator">=</span><span class="literal string">&quot;start&quot;</span><span class="punctuation">,</span>
                                   <span class="name">label_size</span> <span class="operator">=</span> <span class="literal number integer">6</span><span class="punctuation">,</span> <span class="name">label_angle</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">)</span>
        <span class="name">i</span><span class="operator">+=</span><span class="literal number integer">1</span>

<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;linear&quot;</span><span class="punctuation">,</span> <span class="name">pagesize</span><span class="operator">=</span><span class="literal string">'A4'</span><span class="punctuation">,</span> <span class="name">fragments</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">,</span>
                <span class="name">start</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name">end</span><span class="operator">=</span><span class="name">max_len</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">name</span> <span class="operator">+</span> <span class="literal string">&quot;.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;PDF&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">name</span> <span class="operator">+</span> <span class="literal string">&quot;.eps&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;EPS&quot;</span><span class="punctuation">)</span>
<span class="name">gd_diagram</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">name</span> <span class="operator">+</span> <span class="literal string">&quot;.svg&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;SVG&quot;</span><span class="punctuation">)</span>
</pre>
<p>结果如图所示：</p>
<p><img alt="image21" src="../images/three_track_simple.png" /></p>
<p>在示例图中底部的噬菌体没有红色或橙色的基因标记。另外，三个噬菌体可视化图的长度不同，这是因为它们的比例相同，长度却不同。</p>
<p>另外有一点不同，不同噬菌体的同源蛋白质之间用有颜色的links相连，下一部分将解决这个问题。</p>
</div>
<div class="section" id="trackcross-links">
<h3>17.1.11 不同Track之间的Cross-Links</h3>
<p>Biopython 1.59新增绘制不同track之间Cross-Links的功能，这个功能可用于将要展示的简单线形图中，也可用于将线形图分割为短片段（fragments）和环形图。</p>
<p>我们接着模仿Proux等人 [<a class="reference external" href="#proux2002">5</a>] 的图像，我们需要一个包含基因之间的“cross links”、“得分”或“颜色”的列表。 实际应用中，可以从BLAST文件自动提取这些信息，这里是手动输入的。</p>
<p>噬菌体的名称同样表示为A，B和C。这里将要展示的是A与B之间的links，噬菌体A和B基因的相似百分比存储在元组中。</p>
<pre class="code python literal-block">
<span class="comment">#Tuc2009 (NC_002703) vs bIL285 (AF323668)</span>
<span class="name">A_vs_B</span> <span class="operator">=</span> <span class="punctuation">[</span>
    <span class="punctuation">(</span><span class="literal number integer">99</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_01&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;int&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">33</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_03&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf4&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">94</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_05&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf6&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">100</span><span class="punctuation">,</span><span class="literal string">&quot;Tuc2009_06&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf7&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">97</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_07&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf8&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">98</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_08&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf9&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">98</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_09&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf10&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">100</span><span class="punctuation">,</span><span class="literal string">&quot;Tuc2009_10&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf12&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">100</span><span class="punctuation">,</span><span class="literal string">&quot;Tuc2009_11&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf13&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">94</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_12&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf14&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">87</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_13&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf15&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">94</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_14&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf16&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">94</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_15&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf17&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">88</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_17&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;rusA&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">91</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_18&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf20&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">93</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_19&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf22&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">71</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_20&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf23&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">51</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_22&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf27&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">97</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_23&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf28&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">88</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_24&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf29&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">26</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_26&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf38&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">19</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_46&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf52&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">77</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_48&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf54&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">91</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_49&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf55&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">95</span><span class="punctuation">,</span> <span class="literal string">&quot;Tuc2009_52&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;orf60&quot;</span><span class="punctuation">),</span>
<span class="punctuation">]</span>
</pre>
<p>对噬菌体B和C做同样的处理：</p>
<pre class="code python literal-block">
<span class="comment">#bIL285 (AF323668) vs Listeria innocua prophage 5 (in NC_003212)</span>
<span class="name">B_vs_C</span> <span class="operator">=</span> <span class="punctuation">[</span>
    <span class="punctuation">(</span><span class="literal number integer">42</span><span class="punctuation">,</span> <span class="literal string">&quot;orf39&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2581&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">31</span><span class="punctuation">,</span> <span class="literal string">&quot;orf40&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2580&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">49</span><span class="punctuation">,</span> <span class="literal string">&quot;orf41&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2579&quot;</span><span class="punctuation">),</span> <span class="comment">#terL</span>
    <span class="punctuation">(</span><span class="literal number integer">54</span><span class="punctuation">,</span> <span class="literal string">&quot;orf42&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2578&quot;</span><span class="punctuation">),</span> <span class="comment">#portal</span>
    <span class="punctuation">(</span><span class="literal number integer">55</span><span class="punctuation">,</span> <span class="literal string">&quot;orf43&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2577&quot;</span><span class="punctuation">),</span> <span class="comment">#protease</span>
    <span class="punctuation">(</span><span class="literal number integer">33</span><span class="punctuation">,</span> <span class="literal string">&quot;orf44&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2576&quot;</span><span class="punctuation">),</span> <span class="comment">#mhp</span>
    <span class="punctuation">(</span><span class="literal number integer">51</span><span class="punctuation">,</span> <span class="literal string">&quot;orf46&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2575&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">33</span><span class="punctuation">,</span> <span class="literal string">&quot;orf47&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2574&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">40</span><span class="punctuation">,</span> <span class="literal string">&quot;orf48&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2573&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">25</span><span class="punctuation">,</span> <span class="literal string">&quot;orf49&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2572&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">50</span><span class="punctuation">,</span> <span class="literal string">&quot;orf50&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2571&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">48</span><span class="punctuation">,</span> <span class="literal string">&quot;orf51&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2570&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">24</span><span class="punctuation">,</span> <span class="literal string">&quot;orf52&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2568&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">30</span><span class="punctuation">,</span> <span class="literal string">&quot;orf53&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2567&quot;</span><span class="punctuation">),</span>
    <span class="punctuation">(</span><span class="literal number integer">28</span><span class="punctuation">,</span> <span class="literal string">&quot;orf54&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;lin2566&quot;</span><span class="punctuation">),</span>
<span class="punctuation">]</span>
</pre>
<p>噬菌体A和C的标识符（Identifiers）是基因座标签（locus tags），噬菌体B没有基因座标签，这里用基因名称来代替。以下的辅助函数可用基因座标签或基因名称来寻找Feature。</p>
<pre class="code python literal-block">
<span class="keyword">def</span> <span class="name function">get_feature</span><span class="punctuation">(</span><span class="name">features</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="punctuation">,</span> <span class="name">tags</span><span class="operator">=</span><span class="punctuation">[</span><span class="literal string">&quot;locus_tag&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;gene&quot;</span><span class="punctuation">]):</span>
    <span class="literal string doc">&quot;&quot;&quot;Search list of SeqFeature objects for an identifier under the given tags.&quot;&quot;&quot;</span>
    <span class="keyword">for</span> <span class="name">f</span> <span class="operator word">in</span> <span class="name">features</span><span class="punctuation">:</span>
        <span class="keyword">for</span> <span class="name">key</span> <span class="operator word">in</span> <span class="name">tags</span><span class="punctuation">:</span>
            <span class="comment">#tag may not be present in this feature</span>
            <span class="keyword">for</span> <span class="name">x</span> <span class="operator word">in</span> <span class="name">f</span><span class="operator">.</span><span class="name">qualifiers</span><span class="operator">.</span><span class="name">get</span><span class="punctuation">(</span><span class="name">key</span><span class="punctuation">,</span> <span class="punctuation">[]):</span>
                <span class="keyword">if</span> <span class="name">x</span> <span class="operator">==</span> <span class="name builtin">id</span><span class="punctuation">:</span>
                     <span class="keyword">return</span> <span class="name">f</span>
    <span class="keyword">raise</span> <span class="name exception">KeyError</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="punctuation">)</span>
</pre>
<p>现在将这些标识符对（identifier pairs）的列表转换为“SeqFeature”列表，因此来查找它们的坐标定位。现在将下列代码添加到上段代码中 <tt class="docutils literal"><span class="pre">gd_diagram.draw(...)</span></tt> 这一行之前，将cross-links添加到图像中。示例中的脚本文件 <a class="reference external" href="http://biopython.org/SRC/biopython/Doc/examples/Proux_et_al_2002_Figure_6.py">Proux_et_al_2002_Figure_6.py</a> 在Biopython源程序文件夹的 <tt class="docutils literal">Doc/examples</tt> 目录下。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics.GenomeDiagram</span> <span class="keyword namespace">import</span> <span class="name">CrossLink</span>
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib</span> <span class="keyword namespace">import</span> <span class="name">colors</span>
<span class="comment">#Note it might have been clearer to assign the track numbers explicitly...</span>
<span class="keyword">for</span> <span class="name">rec_X</span><span class="punctuation">,</span> <span class="name">tn_X</span><span class="punctuation">,</span> <span class="name">rec_Y</span><span class="punctuation">,</span> <span class="name">tn_Y</span><span class="punctuation">,</span> <span class="name">X_vs_Y</span> <span class="operator word">in</span> <span class="punctuation">[(</span><span class="name">A_rec</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">,</span> <span class="name">B_rec</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name">A_vs_B</span><span class="punctuation">),</span>
                                         <span class="punctuation">(</span><span class="name">B_rec</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name">C_rec</span><span class="punctuation">,</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">B_vs_C</span><span class="punctuation">)]:</span>
    <span class="name">track_X</span> <span class="operator">=</span> <span class="name">gd_diagram</span><span class="operator">.</span><span class="name">tracks</span><span class="punctuation">[</span><span class="name">tn_X</span><span class="punctuation">]</span>
    <span class="name">track_Y</span> <span class="operator">=</span> <span class="name">gd_diagram</span><span class="operator">.</span><span class="name">tracks</span><span class="punctuation">[</span><span class="name">tn_Y</span><span class="punctuation">]</span>
    <span class="keyword">for</span> <span class="name">score</span><span class="punctuation">,</span> <span class="name">id_X</span><span class="punctuation">,</span> <span class="name">id_Y</span> <span class="operator word">in</span> <span class="name">X_vs_Y</span><span class="punctuation">:</span>
        <span class="name">feature_X</span> <span class="operator">=</span> <span class="name">get_feature</span><span class="punctuation">(</span><span class="name">rec_X</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">,</span> <span class="name">id_X</span><span class="punctuation">)</span>
        <span class="name">feature_Y</span> <span class="operator">=</span> <span class="name">get_feature</span><span class="punctuation">(</span><span class="name">rec_Y</span><span class="operator">.</span><span class="name">features</span><span class="punctuation">,</span> <span class="name">id_Y</span><span class="punctuation">)</span>
        <span class="name">color</span> <span class="operator">=</span> <span class="name">colors</span><span class="operator">.</span><span class="name">linearlyInterpolatedColor</span><span class="punctuation">(</span><span class="name">colors</span><span class="operator">.</span><span class="name">white</span><span class="punctuation">,</span> <span class="name">colors</span><span class="operator">.</span><span class="name">firebrick</span><span class="punctuation">,</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="literal number integer">100</span><span class="punctuation">,</span> <span class="name">score</span><span class="punctuation">)</span>
        <span class="name">link_xy</span> <span class="operator">=</span> <span class="name">CrossLink</span><span class="punctuation">((</span><span class="name">track_X</span><span class="punctuation">,</span> <span class="name">feature_X</span><span class="operator">.</span><span class="name">location</span><span class="operator">.</span><span class="name">start</span><span class="punctuation">,</span> <span class="name">feature_X</span><span class="operator">.</span><span class="name">location</span><span class="operator">.</span><span class="name">end</span><span class="punctuation">),</span>
                            <span class="punctuation">(</span><span class="name">track_Y</span><span class="punctuation">,</span> <span class="name">feature_Y</span><span class="operator">.</span><span class="name">location</span><span class="operator">.</span><span class="name">start</span><span class="punctuation">,</span> <span class="name">feature_Y</span><span class="operator">.</span><span class="name">location</span><span class="operator">.</span><span class="name">end</span><span class="punctuation">),</span>
                            <span class="name">color</span><span class="punctuation">,</span> <span class="name">colors</span><span class="operator">.</span><span class="name">lightgrey</span><span class="punctuation">)</span>
        <span class="name">gd_diagram</span><span class="operator">.</span><span class="name">cross_track_links</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">link_xy</span><span class="punctuation">)</span>
</pre>
<p>这段代码有几个要点，第一， <tt class="docutils literal">GenomeDiagram</tt> 对象有一个 <tt class="docutils literal">cross_track_links</tt> 属性，这个属性只是 <tt class="docutils literal">CrossLink</tt> 对象的一组数据。每个 <tt class="docutils literal">CrossLink</tt> 对象有两个track-specific坐标，示例中用元组（tuples）来展现，可用 <tt class="docutils literal">GenomeDiagram.Feature</tt> 对象来代替。可选择添加颜色和边框颜色，还可以说明这个link是否需要翻转，这个功能易于表现染色体异位。</p>
<p>你也可以看我们是如何将BLAST中特征百分比（Percentage Identity Score）转换为白-红的渐变色（白-0%，红-100%）。这个实例中没有cross-links的重叠，如果有links重叠可以用ReportLab库中的透明度（transparency）来解决，通过设置颜色的alpha通道来使用。然而，若同时使用边框阴影和叠加透明度会增加理解的难度。结果见下图：</p>
<p><img alt="image22" src="../images/three_track_cl.png" /></p>
<p>当然，Biopython还有很多增强图像效果的方法。首先，这个示例中的cross links是蛋白质之间的，被呈现在一个链的固定区域（strand specific manor）。可以在feature track上用 ‘BOX’ sigil添加背景区域（background region）来扩展cross link的效果。同样，可以缩短feature tracks之间的垂直高度，使用更多的links来代替——一种方法是为空的track分配空间。此外，在没有大规模基因重叠的情况下，可以用跨越轴线的&quot;BIGARROW&quot;,这样就为track进一步增加了垂直空间。详情请查看Biopython源程序的 <tt class="docutils literal">Doc/examples</tt> 目录下的示例脚本文件：<a class="reference external" href="http://biopython.org/SRC/biopython/Doc/examples/Proux_et_al_2002_Figure_6.py">Proux_et_al_2002_Figure_6.py</a> 。
结果见下图：</p>
<p><img alt="image23" src="../images/three_track_cl2a.png" /></p>
<p>除此之外，你可能希望在图像编辑软件里手动调整gene标签的位置，添加特定标识，比如强调某个特别的区域。</p>
<p>如果有多个叠加的links，使用ReportLab库里的颜色透明度（transparent color）是非常好的方法，由于这个示例没有cross-link的重叠，所以没有用到颜色透明度（transparent color）。然而，尽量避免在这个示例中使用边框阴影（shaded color scheme）。</p>
</div>
<div class="section" id="id215">
<h3>17.1.12 高级选项</h3>
<p>可以通过控制刻度线（tick marks）来调节展示比例（scale），毕竟每个图形应该包括基本单位和轴线标签的数目。</p>
<p>到目前为止，我们只使用了 <tt class="docutils literal">FeatureSet</tt> 。GenomeDiagram还可以用 <tt class="docutils literal">GraphSet</tt> 来制作线形图，饼状图和heatmap热图（例如在轨迹内展示feature中的GC含量）。</p>
<p>目前还没有添加这个选项，最后，推荐你去参考GenomeDiagram单机版 <a class="reference external" href="http://biopython.org/DIST/docs/GenomeDiagram/userguide.pdf">用户指南
（PDF）</a> 和文档字符串（docstrings）。</p>
</div>
<div class="section" id="id216">
<h3>17.1.13 转换旧代码</h3>
<p>如果你有用GenomeDiagram独立版本写的旧代码，想将其转换为Bippython和新版本可识别的代码，你需要做一些调整——主要是import语句。GenomeDiagram的旧版本中使用英式拼写“colour” 和 “centre”来表示“color” 和“center”。被Biopython整合后，参数名可以使用任意一种。但是将来可能会不支持英式的参数名。</p>
<p>如果你过去使用下面的方式：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">GenomeDiagram</span> <span class="keyword namespace">import</span> <span class="name">GDFeatureSet</span><span class="punctuation">,</span> <span class="name">GDDiagram</span>
<span class="name">gdd</span> <span class="operator">=</span> <span class="name">GDDiagram</span><span class="punctuation">(</span><span class="literal string">&quot;An example&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>你只需要将import语句转换成下面这样：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics.GenomeDiagram</span> <span class="keyword namespace">import</span> <span class="name">FeatureSet</span> <span class="keyword">as</span> <span class="name">GDFeatureSet</span><span class="punctuation">,</span> <span class="name">Diagram</span> <span class="keyword">as</span> <span class="name">GDDiagram</span>
<span class="name">gdd</span> <span class="operator">=</span> <span class="name">GDDiagram</span><span class="punctuation">(</span><span class="literal string">&quot;An example&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>希望能够顺利运行。将来你可能想换用新名称，你必须在更大程度上改变你编写代码的方式：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics.GenomeDiagram</span> <span class="keyword namespace">import</span> <span class="name">FeatureSet</span><span class="punctuation">,</span> <span class="name">Diagram</span>
<span class="name">gdd</span> <span class="operator">=</span> <span class="name">Diagram</span><span class="punctuation">(</span><span class="literal string">&quot;An example&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>or:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">GenomeDiagram</span>
<span class="name">gdd</span> <span class="operator">=</span> <span class="name">GenomeDiagram</span><span class="operator">.</span><span class="name">Diagram</span><span class="punctuation">(</span><span class="literal string">&quot;An example&quot;</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
<p>如果运行过程中出现问题，请到Biopython邮件列表中寻求帮助。唯一的缺点就是没有包括旧模块 <tt class="docutils literal">GenomeDiagram.GDUtilities</tt> ，这个模块有计算GC百分比含量的函数，这一部分将会合并到 <tt class="docutils literal">Bio.SeqUtils</tt> 模块。</p>
</div>
</div>
<div class="section" id="id217">
<h2>17.2 染色体</h2>
<p><tt class="docutils literal">Bio.Graphics.BasicChromosome</tt> 模块可以绘制染色体，Jupe等人在2012发表的文章 [<a class="reference external" href="#jupe2012">6</a>] 中利用不同的颜色来展示不同的基因家族。</p>
<div class="section" id="id218">
<h3>17.2.1 简单染色体</h3>
<p>我们用 <em>Arabidopsis
thaliana</em> 来展示一个简单示例。</p>
<p>首先从NCBI的FTP服务器 <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/Arabidopsis_thaliana">ftp://ftp.ncbi.nlm.nih.gov/genomes/Arabidopsis_thaliana</a> 下载拟南芥已测序的五个染色体文件，利用 <tt class="docutils literal">Bio.SeqIO</tt> 函数计算它们的长度。你可以利用GenBank文件，但是对于染色体来说，FASTA文件的处理速度会快点。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">entries</span> <span class="operator">=</span> <span class="punctuation">[(</span><span class="literal string">&quot;Chr I&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_I/NC_003070.fna&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr II&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_II/NC_003071.fna&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr III&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_III/NC_003074.fna&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr IV&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_IV/NC_003075.fna&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr V&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_V/NC_003076.fna&quot;</span><span class="punctuation">)]</span>
<span class="keyword">for</span> <span class="punctuation">(</span><span class="name">name</span><span class="punctuation">,</span> <span class="name">filename</span><span class="punctuation">)</span> <span class="operator word">in</span> <span class="name">entries</span><span class="punctuation">:</span>
   <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
   <span class="keyword">print</span> <span class="name">name</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
</pre>
<p>计算出5个染色体长度后，就可用 <tt class="docutils literal">BasicChromosome</tt> 模块对其作如下的处理：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.units</span> <span class="keyword namespace">import</span> <span class="name">cm</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">BasicChromosome</span>

<span class="name">entries</span> <span class="operator">=</span> <span class="punctuation">[(</span><span class="literal string">&quot;Chr I&quot;</span><span class="punctuation">,</span> <span class="literal number integer">30432563</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr II&quot;</span><span class="punctuation">,</span> <span class="literal number integer">19705359</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr III&quot;</span><span class="punctuation">,</span> <span class="literal number integer">23470805</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr IV&quot;</span><span class="punctuation">,</span> <span class="literal number integer">18585042</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr V&quot;</span><span class="punctuation">,</span> <span class="literal number integer">26992728</span><span class="punctuation">)]</span>

<span class="name">max_len</span> <span class="operator">=</span> <span class="literal number integer">30432563</span> <span class="comment">#Could compute this</span>
<span class="name">telomere_length</span> <span class="operator">=</span> <span class="literal number integer">1000000</span> <span class="comment">#For illustration</span>

<span class="name">chr_diagram</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">Organism</span><span class="punctuation">()</span>
<span class="name">chr_diagram</span><span class="operator">.</span><span class="name">page_size</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="literal number float">29.7</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">,</span> <span class="literal number integer">21</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">)</span> <span class="comment">#A4 landscape</span>

<span class="keyword">for</span> <span class="name">name</span><span class="punctuation">,</span> <span class="name">length</span> <span class="operator word">in</span> <span class="name">entries</span><span class="punctuation">:</span>
    <span class="name">cur_chromosome</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">Chromosome</span><span class="punctuation">(</span><span class="name">name</span><span class="punctuation">)</span>
    <span class="comment">#Set the scale to the MAXIMUM length plus the two telomeres in bp,</span>
    <span class="comment">#want the same scale used on all five chromosomes so they can be</span>
    <span class="comment">#compared to each other</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">scale_num</span> <span class="operator">=</span> <span class="name">max_len</span> <span class="operator">+</span> <span class="literal number integer">2</span> <span class="operator">*</span> <span class="name">telomere_length</span>

    <span class="comment">#Add an opening telomere</span>
    <span class="name">start</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">TelomereSegment</span><span class="punctuation">()</span>
    <span class="name">start</span><span class="operator">.</span><span class="name">scale</span> <span class="operator">=</span> <span class="name">telomere_length</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">start</span><span class="punctuation">)</span>

    <span class="comment">#Add a body - using bp as the scale length here.</span>
    <span class="name">body</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">ChromosomeSegment</span><span class="punctuation">()</span>
    <span class="name">body</span><span class="operator">.</span><span class="name">scale</span> <span class="operator">=</span> <span class="name">length</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">body</span><span class="punctuation">)</span>

    <span class="comment">#Add a closing telomere</span>
    <span class="name">end</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">TelomereSegment</span><span class="punctuation">(</span><span class="name">inverted</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
    <span class="name">end</span><span class="operator">.</span><span class="name">scale</span> <span class="operator">=</span> <span class="name">telomere_length</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">end</span><span class="punctuation">)</span>

    <span class="comment">#This chromosome is done</span>
    <span class="name">chr_diagram</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">cur_chromosome</span><span class="punctuation">)</span>

<span class="name">chr_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="literal string">&quot;simple_chrom.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;Arabidopsis thaliana&quot;</span><span class="punctuation">)</span>
</pre>
<p>新建的PDF文档如图所示：</p>
<p><img alt="image24" src="../images/simple_chrom.png" /></p>
<p>这个示例可以短小精悍，下面的示例可以展示目标feature的定位。</p>
</div>
<div class="section" id="id219">
<h3>17.2.2 染色体注释</h3>
<p>继续前面的示例，我们可以同时展示tRNA基因。通过解析 <em>Arabidopsis thaliana</em> 的5个染色体GenBank文件，我们可以对他们进行定位。你需要从NCBI的FTP服务器下载这些文件 <a class="reference external" href="ftp://ftp.ncbi.nlm.nih.gov/genomes/Arabidopsis_thaliana">ftp://ftp.ncbi.nlm.nih.gov/genomes/Arabidopsis_thaliana</a> ，也可以保存子目录名称或者添加如下的路径：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">reportlab.lib.units</span> <span class="keyword namespace">import</span> <span class="name">cm</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Graphics</span> <span class="keyword namespace">import</span> <span class="name">BasicChromosome</span>

<span class="name">entries</span> <span class="operator">=</span> <span class="punctuation">[(</span><span class="literal string">&quot;Chr I&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_I/NC_003070.gbk&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr II&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_II/NC_003071.gbk&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr III&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_III/NC_003074.gbk&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr IV&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_IV/NC_003075.gbk&quot;</span><span class="punctuation">),</span>
           <span class="punctuation">(</span><span class="literal string">&quot;Chr V&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;CHR_V/NC_003076.gbk&quot;</span><span class="punctuation">)]</span>

<span class="name">max_len</span> <span class="operator">=</span> <span class="literal number integer">30432563</span> <span class="comment">#Could compute this</span>
<span class="name">telomere_length</span> <span class="operator">=</span> <span class="literal number integer">1000000</span> <span class="comment">#For illustration</span>

<span class="name">chr_diagram</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">Organism</span><span class="punctuation">()</span>
<span class="name">chr_diagram</span><span class="operator">.</span><span class="name">page_size</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="literal number float">29.7</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">,</span> <span class="literal number integer">21</span><span class="operator">*</span><span class="name">cm</span><span class="punctuation">)</span> <span class="comment">#A4 landscape</span>

<span class="keyword">for</span> <span class="name">index</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="name">name</span><span class="punctuation">,</span> <span class="name">filename</span><span class="punctuation">)</span> <span class="operator word">in</span> <span class="name builtin">enumerate</span><span class="punctuation">(</span><span class="name">entries</span><span class="punctuation">):</span>
    <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span><span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
    <span class="name">length</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
    <span class="name">features</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">f</span> <span class="keyword">for</span> <span class="name">f</span> <span class="operator word">in</span> <span class="name">record</span><span class="operator">.</span><span class="name">features</span> <span class="keyword">if</span> <span class="name">f</span><span class="operator">.</span><span class="name">type</span><span class="operator">==</span><span class="literal string">&quot;tRNA&quot;</span><span class="punctuation">]</span>
    <span class="comment">#Record an Artemis style integer color in the feature's qualifiers,</span>
    <span class="comment">#1 = Black, 2 = Red, 3 = Green, 4 = blue, 5 =cyan, 6 = purple</span>
    <span class="keyword">for</span> <span class="name">f</span> <span class="operator word">in</span> <span class="name">features</span><span class="punctuation">:</span> <span class="name">f</span><span class="operator">.</span><span class="name">qualifiers</span><span class="punctuation">[</span><span class="literal string">&quot;color&quot;</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">index</span><span class="operator">+</span><span class="literal number integer">2</span><span class="punctuation">]</span>

    <span class="name">cur_chromosome</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">Chromosome</span><span class="punctuation">(</span><span class="name">name</span><span class="punctuation">)</span>
    <span class="comment">#Set the scale to the MAXIMUM length plus the two telomeres in bp,</span>
    <span class="comment">#want the same scale used on all five chromosomes so they can be</span>
    <span class="comment">#compared to each other</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">scale_num</span> <span class="operator">=</span> <span class="name">max_len</span> <span class="operator">+</span> <span class="literal number integer">2</span> <span class="operator">*</span> <span class="name">telomere_length</span>

    <span class="comment">#Add an opening telomere</span>
    <span class="name">start</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">TelomereSegment</span><span class="punctuation">()</span>
    <span class="name">start</span><span class="operator">.</span><span class="name">scale</span> <span class="operator">=</span> <span class="name">telomere_length</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">start</span><span class="punctuation">)</span>

    <span class="comment">#Add a body - again using bp as the scale length here.</span>
    <span class="name">body</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">AnnotatedChromosomeSegment</span><span class="punctuation">(</span><span class="name">length</span><span class="punctuation">,</span> <span class="name">features</span><span class="punctuation">)</span>
    <span class="name">body</span><span class="operator">.</span><span class="name">scale</span> <span class="operator">=</span> <span class="name">length</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">body</span><span class="punctuation">)</span>

    <span class="comment">#Add a closing telomere</span>
    <span class="name">end</span> <span class="operator">=</span> <span class="name">BasicChromosome</span><span class="operator">.</span><span class="name">TelomereSegment</span><span class="punctuation">(</span><span class="name">inverted</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">)</span>
    <span class="name">end</span><span class="operator">.</span><span class="name">scale</span> <span class="operator">=</span> <span class="name">telomere_length</span>
    <span class="name">cur_chromosome</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">end</span><span class="punctuation">)</span>

    <span class="comment">#This chromosome is done</span>
    <span class="name">chr_diagram</span><span class="operator">.</span><span class="name">add</span><span class="punctuation">(</span><span class="name">cur_chromosome</span><span class="punctuation">)</span>

<span class="name">chr_diagram</span><span class="operator">.</span><span class="name">draw</span><span class="punctuation">(</span><span class="literal string">&quot;tRNA_chrom.pdf&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;Arabidopsis thaliana&quot;</span><span class="punctuation">)</span>
</pre>
<p>如果标签之间太紧密会发出警告，所以要注意第一条染色体的的前导链（左手边），可以创建一个彩色的PDF文件，如下图所示：</p>
<p><img alt="image25" src="../images/tRNA_chrom.png" /></p>
</div>
</div>
</div>
<div class="section" id="cookbook">
<h1>第18章  Cookbook – 用它做一些很酷的事情</h1>
<p>Biopython目前有两个版本的“cookbook”示例——本章（本章包含在教程中许多年，并渐渐成熟），和在Biopython维基上的由用户贡献的集合： <a class="reference external" href="http://biopython.org/wiki/Category:Cookbook">http://biopython.org/wiki/Category:Cookbook</a> 。</p>
<p>我们在试着鼓励Biopython用户在维基上贡献他们自己的示例。除了能帮助社区之外，分享像这样的示例的一个直接的好处是，你也能从其他Biopython用户和开发者中获得一些关于代码的意见反馈——这或许能帮助你改进自己的Python代码。</p>
<p>长期来说，我们可能最终会将这一章所有的示例都转移到维基上，或者本教程其他的地方。</p>
<div class="section" id="id220">
<h2>18.1  操作序列文件</h2>
<p>这部分将展示更多使用第 <a class="reference external" href="#chapter:Bio.SeqIO">5</a> 章所描述的 <tt class="docutils literal">Bio.SeqIO</tt> 模块来进行序列输入/输出操作的例子。</p>
<div class="section" id="id221">
<h3>18.1.1  过滤文件中的序列</h3>
<p>通常你会拥有一个包含许多序列的大文件（例如，FASTA基因文件，或者FASTQ或SFF读长文件），和一个包含你所感兴趣的序列的ID列表，而你希望创建一个由这一ID列表里的序列构成的文件。</p>
<p>让我们假设这个ID列表在一个简单的文本文件中，作为每一行的第一个词。这可能是一个表格文件，其第一列是序列ID。尝试下面的代码：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">input_file</span> <span class="operator">=</span> <span class="literal string">&quot;big_file.sff&quot;</span>
<span class="name">id_file</span> <span class="operator">=</span> <span class="literal string">&quot;short_list.txt&quot;</span>
<span class="name">output_file</span> <span class="operator">=</span> <span class="literal string">&quot;short_list.sff&quot;</span>
<span class="name">wanted</span> <span class="operator">=</span> <span class="name builtin">set</span><span class="punctuation">(</span><span class="name">line</span><span class="operator">.</span><span class="name">rstrip</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string escape">\n</span><span class="literal string">&quot;</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">(</span><span class="name builtin pseudo">None</span><span class="punctuation">,</span><span class="literal number integer">1</span><span class="punctuation">)[</span><span class="literal number integer">0</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name">line</span> <span class="operator word">in</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="name">id_file</span><span class="punctuation">))</span>
<span class="keyword">print</span> <span class="literal string">&quot;Found </span><span class="literal string interpol">%i</span><span class="literal string"> unique identifiers in </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">wanted</span><span class="punctuation">),</span> <span class="name">id_file</span><span class="punctuation">)</span>
<span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">r</span> <span class="keyword">for</span> <span class="name">r</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">input_file</span><span class="punctuation">,</span> <span class="literal string">&quot;sff&quot;</span><span class="punctuation">)</span> <span class="keyword">if</span> <span class="name">r</span><span class="operator">.</span><span class="name">id</span> <span class="operator word">in</span> <span class="name">wanted</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">output_file</span><span class="punctuation">,</span> <span class="literal string">&quot;sff&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> records from </span><span class="literal string interpol">%s</span><span class="literal string"> to </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">count</span><span class="punctuation">,</span> <span class="name">input_file</span><span class="punctuation">,</span> <span class="name">output_file</span><span class="punctuation">)</span>
<span class="keyword">if</span> <span class="name">count</span> <span class="operator">&lt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">wanted</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="literal string">&quot;Warning </span><span class="literal string interpol">%i</span><span class="literal string"> IDs not found in </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">wanted</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">count</span><span class="punctuation">,</span> <span class="name">input_file</span><span class="punctuation">)</span>
</pre>
<p>注意，我们使用Python的 <tt class="docutils literal">set</tt> 类型而不是 <tt class="docutils literal">list</tt>，这会使得检测成员关系更快。</p>
</div>
<div class="section" id="id222">
<h3>18.1.2  生成随机基因组</h3>
<p>假设你在检视一个基因组序列，寻找一些序列特征——或许是极端局部GC含量偏差，或者可能的限制性酶切位点。一旦你使你的Python代码在真实的基因组上运行后，尝试在相同的随机化版本基因组上运行，并进行统计分析或许是明智的选择（毕竟，任何你发现的“特性”都可能只是偶然事件）。</p>
<p>在这一讨论中，我们将使用来自 <em>Yersinia pestis biovar Microtus</em> 的pPCP1质粒的GenBank文件。该文件包含在Biopython单元测试的GenBank文件夹中，或者你可以从我们的网站上得到， <a class="reference external" href="http://biopython.org/SRC/biopython/Tests/GenBank/NC_005816.gb">NC_005816.gb</a>. 该文件仅有一个记录，所以我们能用 <tt class="docutils literal">Bio.SeqIO.read()</tt> 函数把它当做 <tt class="docutils literal">SeqRecord</tt> 读入：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">original_rec</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
</pre>
<p>那么，我们怎么生成一个原始序列重排后的版本能？我会使用Python内置的 <tt class="docutils literal">random</tt> 模块来做这个，特别是 <tt class="docutils literal">random.shuffle</tt> 函数——但是这个只作用于Python列表。我们的序列是一个 <tt class="docutils literal">Seq</tt> 对象，所以为了重排它，我们需要将它转换为一个列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">import</span> <span class="name namespace">random</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">nuc_list</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">original_rec</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">random</span><span class="operator">.</span><span class="name">shuffle</span><span class="punctuation">(</span><span class="name">nuc_list</span><span class="punctuation">)</span> <span class="comment">#acts in situ!</span>
</pre>
<p>现在，为了使用 <tt class="docutils literal">Bio.SeqIO</tt> 输出重排的序列，我们需要使用重排后的列表重新创建包含一个新的 <tt class="docutils literal">SeqRecord</tt> 包含随即化后的 <tt class="docutils literal">Seq</tt> 。要实现这个，我们需要将核苷酸（单字母字符串）列表转换为长字符串——在Python中，一般使用字符串对象的join方法来实现它。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">shuffled_rec</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">nuc_list</span><span class="punctuation">),</span> <span class="name">original_rec</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">),</span>
<span class="operator">...</span>                          <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Shuffled&quot;</span><span class="punctuation">,</span> <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;Based on </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">original_rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>
</pre>
<p>让我们将所有这些片段放到一起来组成一个完整的Python脚本，这个脚本将生成一个FASTA序列文件，其包含30个原始序列的随机重排版本。</p>
<p>第一个版本只是使用一个大的for循环，并一个一个的输出记录（使用章节 <a class="reference external" href="#sec:Bio.SeqIO-and-StringIO">5.5.4</a> 描述的``SeqRecord`` 的格式化方法）：</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">random</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>

<span class="name">original_rec</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>

<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;shuffled.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">30</span><span class="punctuation">):</span>
    <span class="name">nuc_list</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">original_rec</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
    <span class="name">random</span><span class="operator">.</span><span class="name">shuffle</span><span class="punctuation">(</span><span class="name">nuc_list</span><span class="punctuation">)</span>
    <span class="name">shuffled_rec</span> <span class="operator">=</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">nuc_list</span><span class="punctuation">),</span> <span class="name">original_rec</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">),</span> \
                             <span class="name builtin">id</span><span class="operator">=</span><span class="literal string">&quot;Shuffled</span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">),</span> \
                             <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;Based on </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">original_rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>
    <span class="name">handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">shuffled_rec</span><span class="operator">.</span><span class="name">format</span><span class="punctuation">(</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>我个人更喜欢下面的版本（不使用for循环），而使用一个函数来重排记录以及一个生成表达式：</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">random</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Seq</span> <span class="keyword namespace">import</span> <span class="name">Seq</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>

<span class="keyword">def</span> <span class="name function">make_shuffle_record</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">,</span> <span class="name">new_id</span><span class="punctuation">):</span>
    <span class="name">nuc_list</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span>
    <span class="name">random</span><span class="operator">.</span><span class="name">shuffle</span><span class="punctuation">(</span><span class="name">nuc_list</span><span class="punctuation">)</span>
    <span class="keyword">return</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">&quot;&quot;</span><span class="operator">.</span><span class="name">join</span><span class="punctuation">(</span><span class="name">nuc_list</span><span class="punctuation">),</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">alphabet</span><span class="punctuation">),</span> \
           <span class="name builtin">id</span><span class="operator">=</span><span class="name">new_id</span><span class="punctuation">,</span> <span class="name">description</span><span class="operator">=</span><span class="literal string">&quot;Based on </span><span class="literal string interpol">%s</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="name">original_rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span>

<span class="name">original_rec</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="name">shuffled_recs</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">make_shuffle_record</span><span class="punctuation">(</span><span class="name">original_rec</span><span class="punctuation">,</span> <span class="literal string">&quot;Shuffled</span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">i</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">))</span> \
                 <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">30</span><span class="punctuation">))</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;shuffled.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">shuffled_recs</span><span class="punctuation">,</span> <span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
</div>
<div class="section" id="cdsfasta">
<h3>18.1.3  翻译CDS条目为FASTA文件</h3>
<p>假设你有一个包含某个物种的CDS条目作为输入文件，你想生成一个由它们的蛋白序列组成的FASTA文件。也就是，从原始文件中取出每一个核苷酸序列，并翻译它。回到章节 <a class="reference external" href="#sec:translation">3.9</a> 我们了解了怎么使用 <tt class="docutils literal">Seq</tt> 对象的 <tt class="docutils literal">translate</tt> 方法，和可选的 <tt class="docutils literal">cds</tt> 参数来使得不同的起始密码子能正确翻译。</p>
<p>就像章节 <a class="reference external" href="#sec:SeqIO-reverse-complement">5.5.3</a> 中反向互补的例子中展示的那样，我们可以用 <tt class="docutils literal">Bio.SeqIO</tt> 将与翻译步骤结合起来。对于每一个核苷酸 <tt class="docutils literal">SeqRecord</tt> ，我们需要创建一个蛋白的 <tt class="docutils literal">SeqRecord</tt> —— 并对它命名。</p>
<p>你能编写自己的函数来做这个事情，为你的序列选择合适的蛋白标识和恰当的密码表。在本例中，我们仅使用默认的密码表，并给序列ID加一个前缀。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqRecord</span> <span class="keyword namespace">import</span> <span class="name">SeqRecord</span>
<span class="keyword">def</span> <span class="name function">make_protein_record</span><span class="punctuation">(</span><span class="name">nuc_record</span><span class="punctuation">):</span>
    <span class="literal string doc">&quot;&quot;&quot;Returns a new SeqRecord with the translated sequence (default table).&quot;&quot;&quot;</span>
    <span class="keyword">return</span> <span class="name">SeqRecord</span><span class="punctuation">(</span><span class="name">seq</span> <span class="operator">=</span> <span class="name">nuc_record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">cds</span><span class="operator">=</span><span class="name builtin pseudo">True</span><span class="punctuation">),</span> \
                     <span class="name builtin">id</span> <span class="operator">=</span> <span class="literal string">&quot;trans_&quot;</span> <span class="operator">+</span> <span class="name">nuc_record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> \
                     <span class="name">description</span> <span class="operator">=</span> <span class="literal string">&quot;translation of CDS, using default table&quot;</span><span class="punctuation">)</span>
</pre>
<p>我们能用这个函数将核苷酸记录转换为蛋白记录，作为输出。一个优雅且内存高效的方式是使用生成表达式（generator expression）：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">proteins</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">make_protein_record</span><span class="punctuation">(</span><span class="name">nuc_rec</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">nuc_rec</span> <span class="operator word">in</span> \
            <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;coding_sequences.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">proteins</span><span class="punctuation">,</span> <span class="literal string">&quot;translations.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>以上代码适用于完整编码序列的FASTA文件。如果你使用部分编码序列，你可能希望在以上的例子中使用 <tt class="docutils literal">nuc_record.seq.translate(to_stop=True)</tt> ，这会告诉Biopython不检查起始密码的有效性，等等。</p>
</div>
<div class="section" id="id223">
<h3>18.1.4  将FASTA文件中的序列变为大写</h3>
<p>通常你会从合作者那里得到FASTA文件的数据，有时候这些序列可能是大小写混合的。在某些情况下，这些可能是有意为之的（例如，小写的作为低质量的区域），但通常大小写并不重要。你可能希望编辑这个文件以使所有的序列都变得一致（如，都为大写），你可以使用 <tt class="docutils literal">SeqRecord</tt> 对象的 <tt class="docutils literal">upper()</tt> 方法轻易的实现（Biopython 1.55中引入）：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">()</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;mixed.fas&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;upper.fas&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Converted </span><span class="literal string interpol">%i</span><span class="literal string"> records to upper case&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>这是怎么工作的呢？第一行只是导入 <tt class="docutils literal">Bio.SeqIO</tt> 模块。第二行是最有趣的——这是一个Python生成器表达式，它提供 <tt class="docutils literal">mixed.fas</tt> 里每个记录的大写版本。第三行中，我们把这个生成器表达式传给 <tt class="docutils literal">Bio.SeqIO.write()</tt> 函数，它会把大写的序列写出到 <tt class="docutils literal">upper.fas</tt> 输出文件。</p>
<p>我们使用生成器（而不是一个列表或列表解析式）的原因是，前一方式每次仅有一个记录保存在内存中。当你在处理包含成千上万的条目的文件时，这可能非常重要。</p>
</div>
<div class="section" id="id224">
<h3>18.1.5  对序列文件排序</h3>
<p>假设你想对一个序列文件按序列长度排序（例如，一个序列拼接的重叠群(contig)集合），而你工作的文件格式可能是像FASTA或FASTQ这样 <tt class="docutils literal">Bio.SeqIO</tt> 能读写（和索引）的格式。</p>
<p>如果文件足够小，你能将它都一次读入内存为一个 <tt class="docutils literal">SeqRecord</tt> 对象列表，对列表进行排序，并保存它：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">records</span><span class="operator">.</span><span class="name">sort</span><span class="punctuation">(</span><span class="name builtin">cmp</span><span class="operator">=</span><span class="keyword">lambda</span> <span class="name">x</span><span class="punctuation">,</span><span class="name">y</span><span class="punctuation">:</span> <span class="name builtin">cmp</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">x</span><span class="punctuation">),</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">y</span><span class="punctuation">)))</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;sorted_orchids.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>唯一巧妙的地方是指明一个比较函数来说明怎样对记录进行排序（这里我们按长度对他们排序）。如果你希望最长的记录在第一个，你可以交换比对，或者使用reverse参数：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">records</span> <span class="operator">=</span> <span class="name builtin">list</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">records</span><span class="operator">.</span><span class="name">sort</span><span class="punctuation">(</span><span class="name builtin">cmp</span><span class="operator">=</span><span class="keyword">lambda</span> <span class="name">x</span><span class="punctuation">,</span><span class="name">y</span><span class="punctuation">:</span> <span class="name builtin">cmp</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">y</span><span class="punctuation">),</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">x</span><span class="punctuation">)))</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;sorted_orchids.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>现在这一实现是非常直接的——但是如果你的文件非常大，你不能像这样把它整个加载到内存中应该怎么办呢？例如，你可能有一些二代测序的读长要根据长度排序。这时你可以使用 <tt class="docutils literal">Bio.SeqIO.index()</tt> 函数解决。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="comment">#Get the lengths and ids, and sort on length</span>
<span class="name">len_and_ids</span> <span class="operator">=</span> <span class="name builtin">sorted</span><span class="punctuation">((</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">),</span> <span class="name">rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> \
                     <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">ids</span> <span class="operator">=</span> <span class="name builtin">reversed</span><span class="punctuation">([</span><span class="name builtin">id</span> <span class="keyword">for</span> <span class="punctuation">(</span><span class="name">length</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="punctuation">)</span> <span class="operator word">in</span> <span class="name">len_and_ids</span><span class="punctuation">])</span>
<span class="keyword">del</span> <span class="name">len_and_ids</span> <span class="comment">#free this memory</span>
<span class="name">record_index</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="name">records</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">record_index</span><span class="punctuation">[</span><span class="name builtin">id</span><span class="punctuation">]</span> <span class="keyword">for</span> <span class="name builtin">id</span> <span class="operator word">in</span> <span class="name">ids</span><span class="punctuation">)</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="literal string">&quot;sorted.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>首先我们使用 <tt class="docutils literal">Bio.SeqIO.parse()</tt> 来将整个文件扫描一遍，并将所有记录的标识和它们的长度保存在一个元组（tuple）中。接着我们对这个元组按照序列长度进行排序，并舍弃这些长度。有了这一排列后的标识列表， <tt class="docutils literal">Bio.SeqIO.index()</tt> 允许我们一个一个获取这些记录，我们把它们传给 <tt class="docutils literal">Bio.SeqIO.write()</tt> 输出。</p>
<p>这些例子都使用 <tt class="docutils literal">Bio.SeqIO</tt> 来解析记录为 <tt class="docutils literal">SeqRecord</tt> 对象，并通过 <tt class="docutils literal">Bio.SeqIO.write()</tt> 输出。当你想排序的文件格式 <tt class="docutils literal">Bio.SeqIO.write()</tt> 不支持应该怎么办呢？如纯文本的SwissProt格式。这里有一个额外的解决方法——使用在 Biopython 1.54 (见 <a class="reference external" href="#sec:seqio-index-getraw">5.4.2.2</a>) 中 <tt class="docutils literal">Bio.SeqIO.index()</tt> 添加的 <tt class="docutils literal">get_raw()</tt> 方法。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="comment">#Get the lengths and ids, and sort on length</span>
<span class="name">len_and_ids</span> <span class="operator">=</span> <span class="name builtin">sorted</span><span class="punctuation">((</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">),</span> <span class="name">rec</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> \
                     <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
<span class="name">ids</span> <span class="operator">=</span> <span class="name builtin">reversed</span><span class="punctuation">([</span><span class="name builtin">id</span> <span class="keyword">for</span> <span class="punctuation">(</span><span class="name">length</span><span class="punctuation">,</span> <span class="name builtin">id</span><span class="punctuation">)</span> <span class="operator word">in</span> <span class="name">len_and_ids</span><span class="punctuation">])</span>
<span class="keyword">del</span> <span class="name">len_and_ids</span> <span class="comment">#free this memory</span>
<span class="name">record_index</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;sorted.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name builtin">id</span> <span class="operator word">in</span> <span class="name">ids</span><span class="punctuation">:</span>
    <span class="name">handle</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">record_index</span><span class="operator">.</span><span class="name">get_raw</span><span class="punctuation">(</span><span class="name builtin">id</span><span class="punctuation">))</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>作为一个奖励，由于以上例子不重复将数据解析为 <tt class="docutils literal">SeqRecord</tt> 对象，所以它会更快。</p>
</div>
<div class="section" id="fastq">
<h3>18.1.6  FASTQ文件的简单质量过滤</h3>
<p>FASTQ文件格式在Sanger被引入，目前被广泛用来存储核苷酸序列（reads）和它们的测序质量。FASTQ文件（和相关的QUAL文件）是单字母注释（per-letter-annotation）的最好的例子，因为序列中每一个核苷酸都有一个相对应的质量分数。任何单字母注释都以list、tuple或string被保存在 <tt class="docutils literal">SeqRecord</tt> 的 <tt class="docutils literal">letter_annotations</tt> 字典中（单字符注释具有和序列长度相同个数的元素）。</p>
<p>一个常见的工作是输入一个大的测序读长集合，并根据它们的质量分数过滤它们（或修剪它们）。下面的例子非常简单，然而足以展示处理 <tt class="docutils literal">SeqRecord</tt> 对象中质量数据的基本用法。我们所有要做的事情是读入一个FASTQ文件，过滤并取出那些PHRED质量分数在某个阈值（这里为20）以上的序列。</p>
<p>在这个例子中，我们将使用从ENA序列读长存档下载的真实数据， <a class="reference external" href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz">ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz</a> (2MB) ，解压后为19MB的文件 <tt class="docutils literal">SRR020192.fastq</tt> 。这是在Roche 454 GS FLX测序平台生成的感染加利福利亚海狮的病毒单端数据（参见 <a class="reference external" href="http://www.ebi.ac.uk/ena/data/view/SRS004476">http://www.ebi.ac.uk/ena/data/view/SRS004476</a> ）。</p>
<p>首先，让我们来统计reads的数目：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">count</span> <span class="operator">=</span> <span class="literal number integer">0</span>
<span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">):</span>
    <span class="name">count</span> <span class="operator">+=</span> <span class="literal number integer">1</span>
<span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>现在让我们做一个简单的过滤，PHRED质量不小于20：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">good_reads</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> \
              <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span> \
              <span class="keyword">if</span> <span class="name builtin">min</span><span class="punctuation">(</span><span class="name">rec</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">])</span> <span class="operator">&gt;=</span> <span class="literal number integer">20</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">good_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;good_quality.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>这只取出了41892条读长中的14580条。一个更有意义的做法是根据质量来裁剪reads，但是这里只是作为一个例子。</p>
<p>FASTQ文件可以包含上百万的记录，所以最好避免一次全部加载它们到内存。这个例子使用一个生成器表达式，这意味着每次只有内存里只有一个 <tt class="docutils literal">SeqRecord</tt> 被创建 —— 避免内存限制。</p>
</div>
<div class="section" id="id225">
<h3>18.1.7  切除引物序列</h3>
<p>在这个例子中，假设我们需要寻找一个FASTQ数据中以 <tt class="docutils literal">GATGACGGTGT</tt> 为5’端的引物序列的reads。同上面的例子一样，我们将使用从ENA下载的 <tt class="docutils literal">SRR020192.fastq</tt> 文件（ <a class="reference external" href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz">ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz</a> ）。该方式同样适用于任何其他Biopython支持的格式（例如FASTA文件）。</p>
<p>这个代码使用 <tt class="docutils literal">Bio.SeqIO</tt> 和一个生成器表达式（避免一次加载所有的序列到内存中），以及 <tt class="docutils literal">Seq</tt> 对象的 <tt class="docutils literal">startswith</tt> 方法来检查读长是否以引物序列开始：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">primer_reads</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> \
                <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span> \
                <span class="keyword">if</span> <span class="name">rec</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">startswith</span><span class="punctuation">(</span><span class="literal string">&quot;GATGACGGTGT&quot;</span><span class="punctuation">))</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">primer_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;with_primer.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>这将从 <tt class="docutils literal">SRR014849.fastq</tt> 找到13819条读长记录，并保存为一个新的FASTQ文件——<tt class="docutils literal">with_primer.fastq</tt>。</p>
<p>现在，假设你希望创建一个包含这些读长，但去除了所有引物序列的FASTQ文件。只需要很小的修改，我们就能对 <tt class="docutils literal">SeqRecord</tt> 进行切片（参见章节 <a class="reference external" href="#sec:SeqRecord-slicing">4.6</a> ）以移除前11个字母（我们的引物长度）：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">trimmed_primer_reads</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">[</span><span class="literal number integer">11</span><span class="punctuation">:]</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> \
                        <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span> \
                        <span class="keyword">if</span> <span class="name">rec</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">startswith</span><span class="punctuation">(</span><span class="literal string">&quot;GATGACGGTGT&quot;</span><span class="punctuation">))</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">trimmed_primer_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;with_primer_trimmed.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>这也将从 <tt class="docutils literal">SRR020192.fastq</tt> 取出13819条读长，但是移除了前十个字符，并将它们保存为另一个新的FASTQ文件， <tt class="docutils literal">with_primer_trimmed.fastq</tt> 。</p>
<p>最后，假设你想移除部分reads中的引物并创建一个新的FASTQ文件，而其他的reads保持不变。如果我们仍然希望使用生成器表达式，声明我们自己的修剪（trim）函数可能更加清楚：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">def</span> <span class="name function">trim_primer</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">,</span> <span class="name">primer</span><span class="punctuation">):</span>
    <span class="keyword">if</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">startswith</span><span class="punctuation">(</span><span class="name">primer</span><span class="punctuation">):</span>
        <span class="keyword">return</span> <span class="name">record</span><span class="punctuation">[</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">primer</span><span class="punctuation">):]</span>
    <span class="keyword">else</span><span class="punctuation">:</span>
        <span class="keyword">return</span> <span class="name">record</span>

<span class="name">trimmed_reads</span> <span class="operator">=</span> <span class="punctuation">(</span><span class="name">trim_primer</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">,</span> <span class="literal string">&quot;GATGACGGTGT&quot;</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> \
                 <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">))</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">trimmed_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>以上代码会运行较长的时间，因为这次输出文件包含所有41892个reads。再次，我们将使用生成器表达式来避免内存问题。你也可以使用一个生成器函数来替代生成器表达式。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">def</span> <span class="name function">trim_primers</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">primer</span><span class="punctuation">):</span>
    <span class="literal string doc">&quot;&quot;&quot;Removes perfect primer sequences at start of reads.

    This is a generator function, the records argument should
    be a list or iterator returning SeqRecord objects.
    &quot;&quot;&quot;</span>
    <span class="name">len_primer</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">primer</span><span class="punctuation">)</span> <span class="comment">#cache this for later</span>
    <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
        <span class="keyword">if</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">startswith</span><span class="punctuation">(</span><span class="name">primer</span><span class="punctuation">):</span>
            <span class="keyword">yield</span> <span class="name">record</span><span class="punctuation">[</span><span class="name">len_primer</span><span class="punctuation">:]</span>
        <span class="keyword">else</span><span class="punctuation">:</span>
            <span class="keyword">yield</span> <span class="name">record</span>

<span class="name">original_reads</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="name">trimmed_reads</span> <span class="operator">=</span> <span class="name">trim_primers</span><span class="punctuation">(</span><span class="name">original_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;GATGACGGTGT&quot;</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">trimmed_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>这种形式非常灵活，如果你想做一些更复杂的事情，譬如只保留部分记录 —— 像下一个例子中展示的那样。</p>
</div>
<div class="section" id="id226">
<h3>18.1.8  切除接头序列</h3>
<p>这实际上是前面例子的一个简单扩展。我们将假设 <tt class="docutils literal">GATGACGGTGT</tt> 是某个FASTQ格式数据的一个接头序列，并再次使用来自NCBI的 <tt class="docutils literal">SRR020192.fastq</tt> 文件 （ <a class="reference external" href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz">ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz</a> ）。</p>
<p>然而在本例中，我们将在读长的 <em>任何位置</em> 查找序列，不仅仅是最开始：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>

<span class="keyword">def</span> <span class="name function">trim_adaptors</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">adaptor</span><span class="punctuation">):</span>
    <span class="literal string doc">&quot;&quot;&quot;Trims perfect adaptor sequences.

    This is a generator function, the records argument should
    be a list or iterator returning SeqRecord objects.
    &quot;&quot;&quot;</span>
    <span class="name">len_adaptor</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">adaptor</span><span class="punctuation">)</span> <span class="comment">#cache this for later</span>
    <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
        <span class="name">index</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">find</span><span class="punctuation">(</span><span class="name">adaptor</span><span class="punctuation">)</span>
        <span class="keyword">if</span> <span class="name">index</span> <span class="operator">==</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">:</span>
            <span class="comment">#adaptor not found, so won't trim</span>
            <span class="keyword">yield</span> <span class="name">record</span>
        <span class="keyword">else</span><span class="punctuation">:</span>
            <span class="comment">#trim off the adaptor</span>
            <span class="keyword">yield</span> <span class="name">record</span><span class="punctuation">[</span><span class="name">index</span><span class="operator">+</span><span class="name">len_adaptor</span><span class="punctuation">:]</span>

<span class="name">original_reads</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="name">trimmed_reads</span> <span class="operator">=</span> <span class="name">trim_adaptors</span><span class="punctuation">(</span><span class="name">original_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;GATGACGGTGT&quot;</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">trimmed_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>因为我们在这个例子中使用的是FASTQ文件， <tt class="docutils literal">SeqRecord</tt> 对象包括reads质量分数的单字母注释（per-letter-annotation）。我们可以通过对具有一定质量分数的 <tt class="docutils literal">SeqRecord</tt> 对象进行切片，并将返回的结果保存到一个FASTQ文件。</p>
<p>和上面的例子（只在每个读长的开始查找引物/接头）相比，你会发现有些reads剪切后非常短（例如，如果接头序列在读长的中部发现，而不是开始附近）。所以，让我们再加入一个最低长度要求：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>

<span class="keyword">def</span> <span class="name function">trim_adaptors</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">adaptor</span><span class="punctuation">,</span> <span class="name">min_len</span><span class="punctuation">):</span>
    <span class="literal string doc">&quot;&quot;&quot;Trims perfect adaptor sequences, checks read length.

    This is a generator function, the records argument should
    be a list or iterator returning SeqRecord objects.
    &quot;&quot;&quot;</span>
    <span class="name">len_adaptor</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">adaptor</span><span class="punctuation">)</span> <span class="comment">#cache this for later</span>
    <span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">records</span><span class="punctuation">:</span>
        <span class="name">len_record</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span> <span class="comment">#cache this for later</span>
        <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span> <span class="operator">&lt;</span> <span class="name">min_len</span><span class="punctuation">:</span>
           <span class="comment">#Too short to keep</span>
           <span class="keyword">continue</span>
        <span class="name">index</span> <span class="operator">=</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">find</span><span class="punctuation">(</span><span class="name">adaptor</span><span class="punctuation">)</span>
        <span class="keyword">if</span> <span class="name">index</span> <span class="operator">==</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">:</span>
            <span class="comment">#adaptor not found, so won't trim</span>
            <span class="keyword">yield</span> <span class="name">record</span>
        <span class="keyword">elif</span> <span class="name">len_record</span> <span class="operator">-</span> <span class="name">index</span> <span class="operator">-</span> <span class="name">len_adaptor</span> <span class="operator">&gt;=</span> <span class="name">min_len</span><span class="punctuation">:</span>
            <span class="comment">#after trimming this will still be long enough</span>
            <span class="keyword">yield</span> <span class="name">record</span><span class="punctuation">[</span><span class="name">index</span><span class="operator">+</span><span class="name">len_adaptor</span><span class="punctuation">:]</span>

<span class="name">original_reads</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="name">trimmed_reads</span> <span class="operator">=</span> <span class="name">trim_adaptors</span><span class="punctuation">(</span><span class="name">original_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;GATGACGGTGT&quot;</span><span class="punctuation">,</span> <span class="literal number integer">100</span><span class="punctuation">)</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">trimmed_reads</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Saved </span><span class="literal string interpol">%i</span><span class="literal string"> reads&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
<p>通过改变格式名称，你也可以将这个应用于FASTA文件。该代码也可以扩展为模糊匹配，而非绝对匹配（或许用一个两两比对，或者考虑读长的质量分数），但是这会使代码变得更慢。</p>
</div>
<div class="section" id="id227">
<h3>18.1.9  转换FASTQ文件</h3>
<p>回到章节 <a class="reference external" href="#sec:SeqIO-conversion">5.5.2</a> ，我们展示了怎样使用 <tt class="docutils literal">Bio.SeqIO</tt> 来实现两个文件格式间的转换。这里，我们将更进一步探讨二代DNA测序中使用的FASTQ文件。更加详细的介绍可以参加 Cock <em>et al.</em> (2009) [<a class="reference external" href="#cock2010">7</a>] 。FASTQ文件同时存储DNA序列（以Python字符串的形式）和相应的读长质量。</p>
<p>PHRED分数（在大多数FASTQ文件中使用，也存在于QUAL、ACE和SFF文件中）已经成为一个用来表示某个给定碱基测序错误概率（这里用 <em>P</em><sub>*e*</sub> 表示）的 <em>实际</em> 标准（使用一个以10为底的对数转换）：</p>
<div class="formula">
<i>Q</i><sub><span class="textrm">PHRED</span></sub> =  − 10 × <span class="textrm"> log</span><sub>10</sub>(<i>P</i><sub><i>e</i></sub>)
</div>
<p>这意味着一个错误的读长（ <em>P</em><sub>*e*</sub> = 1 ）得到的PHRED质量为0，而一个非常好的 <em>P</em><sub>*e*</sub> = 0.00001 的读长得到的PHRED质量为50。在实际的测序数据中，质量比这个要高的非常稀少，通过后期处理，如读长映射（mapping）和组装，PHRED质量到达90是可能的（确实，MAQ工具允许PHRED分数在0到93之间）。</p>
<p>FASTQ格式有潜力成为以单文件纯文本方式存储测序读长的字符和质量分数的 <em>实际</em> 的标准。 唯一的美中不足是，目前至少有三个FASTQ格式版本，它们相互并不兼容，且难以区分...</p>
<ol class="arabic simple">
<li>原始的Sanger FASTQ格式将PHRED质量分数和33个ASCII字符偏移进行编码。NCBI目前在它们的Short Read Archive中使用这种格式。我们在 <tt class="docutils literal">Bio.SeqIO</tt> 中称之为 <tt class="docutils literal">fastq</tt> （或 <tt class="docutils literal"><span class="pre">fastq-sanger</span></tt> ）格式。</li>
<li>Solexa（后来由Illumina收购）引入了他们自己的版本，使用Solexa质量分数和64个ASCII字符偏移进行编码。我们叫做 <tt class="docutils literal"><span class="pre">fastq-solexa</span></tt> 格式。</li>
<li>Illumina工作流1.3进一步推出了PHRED质量分数（更为一致的版本）的FASTQ文件，但是却以64个ASCII字符偏移编码。我们叫做 <tt class="docutils literal"><span class="pre">fastq-illumina</span></tt> 格式。</li>
</ol>
<p>Solexa质量分数采用一种不同的对数转换：</p>
<div class="formula">
<i>Q</i><sub><span class="textrm">Solexa</span></sub> =  − 10 × <span class="textrm"> log</span><sub>10</sub><span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="fraction"><span class="ignored">(</span><span class="numerator"><i>P</i><sub><i>e</i></sub></span><span class="ignored">)/(</span><span class="denominator">1 − <i>P</i><sub><i>e</i></sub></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>
</div>
<p>由于Solexa/Illumina目前在他们的1.3版本的工作流程中已迁移到使用PHRED分数，Solexa质量分数将逐渐淡出使用。如果你将错误估值取等号（ <em>P</em><sub>*e*</sub> ），这两个等式允许在两个评分系统之间进行转换 —— Biopython在 <tt class="docutils literal">Bio.SeqIO.QualityIO</tt> 模块中有函数可以实现。这一模块在使用 <tt class="docutils literal">Bio.SeqIO</tt> 进行从Solexa/Illumina老文件格式到标准Sanger FASTQ文件格式转换时被调用：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;solexa.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq-solexa&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;standard.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
</pre>
<p>如果你想转换新的Illumina 1.3+ FASTQ文件，改变只会导致ASCII码的整体偏移。因为尽管编码不同，所有的质量分数都是PHRED分数：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;illumina.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq-illumina&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;standard.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
</pre>
<p>注意，像这样使用 <tt class="docutils literal">Bio.SeqIO.convert()</tt> 会比 <tt class="docutils literal">Bio.SeqIO.parse()</tt> 和 <tt class="docutils literal">Bio.SeqIO.write()</tt> 组合快得 <em>多</em> ，因为转换FASTQ（包括FASTQ到FASTA的转换）的代码经过优化。</p>
<p>对于质量好的读长，PHRED和Solexa分数几乎相等，这意味着，因为 <tt class="docutils literal"><span class="pre">fasta-solexa</span></tt> 和 <tt class="docutils literal"><span class="pre">fastq-illumina</span></tt> 都使用64个ASCII字符偏移，它们的文件几乎相同。这是Illumina有意设计的，也意味着使用老版本 <tt class="docutils literal"><span class="pre">fasta-solexa</span></tt> 格式文件的应用或许也能接受新版本 <tt class="docutils literal"><span class="pre">fastq-illumina</span></tt> 格式文件（在高质量的数据上）。当然，两个版本和原始的，由Sanger、NCBI和其他地方使用的FASTQ标准有很大不同（格式名为 <tt class="docutils literal">fastq</tt> 或 <tt class="docutils literal"><span class="pre">fastq-sanger</span></tt> ）。</p>
<p>了解更多细节，请参见内置的帮助（或 <a class="reference external" href="http://www.biopython.org/DIST/docs/api/Bio.SeqIO.QualityIO-module.html">在线帮助</a> ）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqIO</span> <span class="keyword namespace">import</span> <span class="name">QualityIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">QualityIO</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="fastaqualfastq">
<h3>18.1.10  转换FASTA和QUAL文件为FASTQ文件</h3>
<p>FASTQ <em>同时</em> 包含序列和他们的质量信息字符串。FASTA文件 <em>只</em> 包含序列，而QUAL文件 <em>只</em> 包含质量。因此，一个单独的FASTQ文件可以转换为 <em>成对的</em> FASTA和QUAL文件，FASTQ文件也可以由成对的FASTA和QUAL文件生成。</p>
<p>从FASTQ到FASTA很简单：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;example.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;example.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
</pre>
<p>从FASTQ到QUAL也很简单：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;example.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;example.qual&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;qual&quot;</span><span class="punctuation">)</span>
</pre>
<p>然而，反向则有一点复杂。你可以使用 <tt class="docutils literal">Bio.SeqIO.parse()</tt> 迭代一个 <em>单独</em> 文件中的所有记录，但是这里我们有两个输入文件。有几个可能的策略，然而这里假设两个文件是真的完全匹配的，最内存高效的方式是同时循环两个文件。代码有些巧妙，所以在 <tt class="docutils literal">Bio.SeqIO.QualityIO</tt> 模块中我们提供一个函数来实现，叫做 <tt class="docutils literal">PairedFastaQualIterator</tt>。它接受两个句柄（FASTA文件和QUAL文件）并返回一个 <tt class="docutils literal">SeqRecord</tt> 迭代器：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqIO.QualityIO</span> <span class="keyword namespace">import</span> <span class="name">PairedFastaQualIterator</span>
<span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">PairedFastaQualIterator</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;example.fasta&quot;</span><span class="punctuation">),</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;example.qual&quot;</span><span class="punctuation">)):</span>
   <span class="keyword">print</span> <span class="name">record</span>
</pre>
<p>这个函数将检查FASTA和QUAL文件是否一致（例如，记录顺序是相同的，并且序列长度一致）。你可以和 <tt class="docutils literal">Bio.SeqIO.write()</tt> 函数结合使用，转换一对FASTA和QUAL文件为单独的FASTQ文件：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqIO.QualityIO</span> <span class="keyword namespace">import</span> <span class="name">PairedFastaQualIterator</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;temp.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;w&quot;</span><span class="punctuation">)</span> <span class="comment">#w=write</span>
<span class="name">records</span> <span class="operator">=</span> <span class="name">PairedFastaQualIterator</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;example.fasta&quot;</span><span class="punctuation">),</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;example.qual&quot;</span><span class="punctuation">))</span>
<span class="name">count</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">write</span><span class="punctuation">(</span><span class="name">records</span><span class="punctuation">,</span> <span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
<span class="keyword">print</span> <span class="literal string">&quot;Converted </span><span class="literal string interpol">%i</span><span class="literal string"> records&quot;</span> <span class="operator">%</span> <span class="name">count</span>
</pre>
</div>
<div class="section" id="id228">
<h3>18.1.11  索引FASTQ文件</h3>
<p>FASTQ文件通常非常大，包含上百万的读长。由于数据量的原因，你不能一次将所有的记录加载到内存中。这就是为什么上面的例子（过滤和剪切）以迭代的方式遍历整个文件，每次只查看一个 <tt class="docutils literal">SeqRecord</tt> 。</p>
<p>然而，有时候你不能使用一个大的循环或迭代器 —— 你或许需要随机获取读长。这里 <tt class="docutils literal">Bio.SeqIO.index()</tt> 函数被证明非常有用，它允许你使用名字获取FASTQ中的任何读长（参见章节 <a class="reference external" href="#sec:SeqIO-index">5.4.2</a> ）。</p>
<p>我们将再次使用来自 ENA (<a class="reference external" href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz">ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz</a>) 的文件 <tt class="docutils literal">SRR020192.fastq</tt> ，尽管这是一个非常小的FASTQ文件，只有不到50,000读长：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">fq_dict</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">index</span><span class="punctuation">(</span><span class="literal string">&quot;SRR020192.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">fq_dict</span><span class="punctuation">)</span>
<span class="literal number integer">41892</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">fq_dict</span><span class="operator">.</span><span class="name">keys</span><span class="punctuation">()[:</span><span class="literal number integer">4</span><span class="punctuation">]</span>
<span class="punctuation">[</span><span class="literal string">'SRR020192.38240'</span><span class="punctuation">,</span> <span class="literal string">'SRR020192.23181'</span><span class="punctuation">,</span> <span class="literal string">'SRR020192.40568'</span><span class="punctuation">,</span> <span class="literal string">'SRR020192.23186'</span><span class="punctuation">]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">fq_dict</span><span class="punctuation">[</span><span class="literal string">&quot;SRR020192.23186&quot;</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">seq</span>
<span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'GTCCCAGTATTCGGATTTGTCTGCCAAAACAATGAAATTGACACAGTTTACAAC...CCG'</span><span class="punctuation">,</span> <span class="name">SingleLetterAlphabet</span><span class="punctuation">())</span>
</pre>
<p>当在包含7百万读长的FASTQ文件上测试时，索引大概需要花费1分钟，然而获取记录几乎是瞬间完成的。</p>
<p>章节 <a class="reference external" href="#sec:SeqIO-sort">18.1.5</a> 的例子展示了如何使用 <tt class="docutils literal">Bio.SeqIO.index()</tt> 函数来对FASTA文件进行排序 —— 这也可以用在FASTQ文件上。</p>
</div>
<div class="section" id="sff">
<h3>18.1.12  转换SFF文件</h3>
<p>如果你处理454(Roche)序列数据，你可能会接触Standard Flowgram Format (SFF)原始数据。这包括序列读长（called bases）、质量分数和原始流信息。</p>
<p>一个最常见的工作是转换SFF文件为一对FASTA和QUAL文件，或者一个单独的FASTQ文件。这可以使用 <tt class="docutils literal">Bio.SeqIO.convert()</tt> 来轻松实现（参见 <a class="reference external" href="#sec:SeqIO-conversion">5.5.2</a> ）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;E3MFGYR02_random_10_reads.sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;reads.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">10</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;E3MFGYR02_random_10_reads.sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;reads.qual&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;qual&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">10</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;E3MFGYR02_random_10_reads.sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;reads.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">10</span>
</pre>
<p>注意这个转换函数返回记录的条数，在这个例子中为10。这将给你 <em>未裁剪</em> 的读长，其中先导和跟随链中低质量的序列，或接头序列将以小写字母显示。如果你希望得到 <em>裁剪</em> 后的读长（使用SFF文件中的剪切信息），可以使用下面的代码：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;E3MFGYR02_random_10_reads.sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;sff-trim&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">10</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;E3MFGYR02_random_10_reads.sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;sff-trim&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.qual&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;qual&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">10</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">convert</span><span class="punctuation">(</span><span class="literal string">&quot;E3MFGYR02_random_10_reads.sff&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;sff-trim&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;trimmed.fastq&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)</span>
<span class="literal number integer">10</span>
</pre>
<p>如果你使用Linux，你可以向Roche请求一份“脱离仪器（off instrument）”的工具（通常叫做Newbler工具）。它提供了另一种的方式来在命令行实现SFF到FASTA或QUAL的转换（但并不支持FASTQ输出）。</p>
<pre class="code python literal-block">
<span class="error">$</span> <span class="name">sffinfo</span> <span class="operator">-</span><span class="name">seq</span> <span class="operator">-</span><span class="name">notrim</span> <span class="name">E3MFGYR02_random_10_reads</span><span class="operator">.</span><span class="name">sff</span> <span class="operator">&gt;</span> <span class="name">reads</span><span class="operator">.</span><span class="name">fasta</span>
<span class="error">$</span> <span class="name">sffinfo</span> <span class="operator">-</span><span class="name">qual</span> <span class="operator">-</span><span class="name">notrim</span> <span class="name">E3MFGYR02_random_10_reads</span><span class="operator">.</span><span class="name">sff</span> <span class="operator">&gt;</span> <span class="name">reads</span><span class="operator">.</span><span class="name">qual</span>
<span class="error">$</span> <span class="name">sffinfo</span> <span class="operator">-</span><span class="name">seq</span> <span class="operator">-</span><span class="name">trim</span> <span class="name">E3MFGYR02_random_10_reads</span><span class="operator">.</span><span class="name">sff</span> <span class="operator">&gt;</span> <span class="name">trimmed</span><span class="operator">.</span><span class="name">fasta</span>
<span class="error">$</span> <span class="name">sffinfo</span> <span class="operator">-</span><span class="name">qual</span> <span class="operator">-</span><span class="name">trim</span> <span class="name">E3MFGYR02_random_10_reads</span><span class="operator">.</span><span class="name">sff</span> <span class="operator">&gt;</span> <span class="name">trimmed</span><span class="operator">.</span><span class="name">qual</span>
</pre>
<p>Biopython以大小写混合的方式来表示剪切位点，这是有意模拟Roche工具的做法。</p>
<p>要获得Biopython对SFF支持的更多信息，请参考内部帮助：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqIO</span> <span class="keyword namespace">import</span> <span class="name">SffIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">help</span><span class="punctuation">(</span><span class="name">SffIO</span><span class="punctuation">)</span>
<span class="operator">...</span>
</pre>
</div>
<div class="section" id="id229">
<h3>18.1.13  识别开放读码框</h3>
<p>在识别可能的基因中一个非常简单的第一步是寻找开放读码框（Open Reading Frame，ORF）。这里我们的意思是寻找六个编码框中所有的没有终止密码子的长区域 —— 一个ORF是一个不包含任何框内终止密码子的核苷酸区域。</p>
<p>当然，为了发现基因，你也需要确定起始密码子、可能的启动子的位置 —— 而且在真核生物中，你也需要关心内含子。然而，这种方法在病毒和原核生物中仍然有效。</p>
<p>为了展示怎样用Biopython实现这个目的，我们首先需要一个序列来查找。作为例子，我们再次使用细菌的质粒 —— 尽管这次我们将以没有任何基因标记的纯文本FASTA文件开始： <a class="reference external" href="http://biopython.org/SRC/biopython/Tests/GenBank/NC_005816.fna">NC_005816.fna</a> 。这是一个细菌序列，所以我们需要使用NCBI密码子表11（参见章节 <a class="reference external" href="#sec:translation">3.9</a> 关于翻译的介绍）。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.fna&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">table</span> <span class="operator">=</span> <span class="literal number integer">11</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">min_pro_len</span> <span class="operator">=</span> <span class="literal number integer">100</span>
</pre>
<p>这里有一个巧妙的技巧，使用 <tt class="docutils literal">Seq</tt> 对象的 <tt class="docutils literal">split</tt> 方法获得一个包含六个读码框中所有可能的ORF翻译的列表：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">for</span> <span class="name">strand</span><span class="punctuation">,</span> <span class="name">nuc</span> <span class="operator word">in</span> <span class="punctuation">[(</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">())]:</span>
<span class="operator">...</span>     <span class="keyword">for</span> <span class="name">frame</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">):</span>
<span class="operator">...</span>         <span class="name">length</span> <span class="operator">=</span> <span class="literal number integer">3</span> <span class="operator">*</span> <span class="punctuation">((</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">frame</span><span class="punctuation">)</span> <span class="operator">//</span> <span class="literal number integer">3</span><span class="punctuation">)</span> <span class="comment">#Multiple of three</span>
<span class="operator">...</span>         <span class="keyword">for</span> <span class="name">pro</span> <span class="operator word">in</span> <span class="name">nuc</span><span class="punctuation">[</span><span class="name">frame</span><span class="punctuation">:</span><span class="name">frame</span><span class="operator">+</span><span class="name">length</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">table</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">split</span><span class="punctuation">(</span><span class="literal string">&quot;*&quot;</span><span class="punctuation">):</span>
<span class="operator">...</span>             <span class="keyword">if</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">pro</span><span class="punctuation">)</span> <span class="operator">&gt;=</span> <span class="name">min_pro_len</span><span class="punctuation">:</span>
<span class="operator">...</span>                 <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">...</span><span class="literal string interpol">%s</span><span class="literal string"> - length </span><span class="literal string interpol">%i</span><span class="literal string">, strand </span><span class="literal string interpol">%i</span><span class="literal string">, frame </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> \
<span class="operator">...</span>                       <span class="operator">%</span> <span class="punctuation">(</span><span class="name">pro</span><span class="punctuation">[:</span><span class="literal number integer">30</span><span class="punctuation">],</span> <span class="name">pro</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">3</span><span class="punctuation">:],</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">pro</span><span class="punctuation">),</span> <span class="name">strand</span><span class="punctuation">,</span> <span class="name">frame</span><span class="punctuation">)</span>
<span class="name">GCLMKKSSIVATIITILSGSANAASSQLIP</span><span class="operator">...</span><span class="name">YRF</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">315</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">0</span>
<span class="name">KSGELRQTPPASSTLHLRLILQRSGVMMEL</span><span class="operator">...</span><span class="name">NPE</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">285</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">1</span>
<span class="name">GLNCSFFSICNWKFIDYINRLFQIIYLCKN</span><span class="operator">...</span><span class="name">YYH</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">176</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">1</span>
<span class="name">VKKILYIKALFLCTVIKLRRFIFSVNNMKF</span><span class="operator">...</span><span class="name">DLP</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">165</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">1</span>
<span class="name">NQIQGVICSPDSGEFMVTFETVMEIKILHK</span><span class="operator">...</span><span class="name">GVA</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">355</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">2</span>
<span class="name">RRKEHVSKKRRPQKRPRRRRFFHRLRPPDE</span><span class="operator">...</span><span class="name">PTR</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">128</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">2</span>
<span class="name">TGKQNSCQMSAIWQLRQNTATKTRQNRARI</span><span class="operator">...</span><span class="name">AIK</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">100</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">2</span>
<span class="name">QGSGYAFPHASILSGIAMSHFYFLVLHAVK</span><span class="operator">...</span><span class="name">CSD</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">114</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">0</span>
<span class="name">IYSTSEHTGEQVMRTLDEVIASRSPESQTR</span><span class="operator">...</span><span class="name">FHV</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">111</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">0</span>
<span class="name">WGKLQVIGLSMWMVLFSQRFDDWLNEQEDA</span><span class="operator">...</span><span class="name">ESK</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">125</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">1</span>
<span class="name">RGIFMSDTMVVNGSGGVPAFLFSGSTLSSY</span><span class="operator">...</span><span class="name">LLK</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">361</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">1</span>
<span class="name">WDVKTVTGVLHHPFHLTFSLCPEGATQSGR</span><span class="operator">...</span><span class="name">VKR</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">111</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">1</span>
<span class="name">LSHTVTDFTDQMAQVGLCQCVNVFLDEVTG</span><span class="operator">...</span><span class="name">KAA</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">107</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">2</span>
<span class="name">RALTGLSAPGIRSQTSCDRLRELRYVPVSL</span><span class="operator">...</span><span class="name">PLQ</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">119</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">frame</span> <span class="literal number integer">2</span>
</pre>
<p>注意，这里我们从 <em>每条</em> 序列的5’末（起始）端开始计算读码框。对 <em>正向</em> 链一直从5’末（起始）端开始计算有时更加容易。</p>
<p>你可以轻易编辑上面的循环代码，来创建一个待选蛋白列表，或者将它转换为一个列表解析。现在，这个代码所不能做的一个事情是记录蛋白的位置信息。</p>
<p>你可以用几种方式来处理。例如，下面的代码以蛋白计数的方式记录位置信息，并通过乘以三倍来转换为父序列（parent sequence），并记录编码框和链的信息：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">record</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="literal string">&quot;NC_005816.gb&quot;</span><span class="punctuation">,</span><span class="literal string">&quot;genbank&quot;</span><span class="punctuation">)</span>
<span class="name">table</span> <span class="operator">=</span> <span class="literal number integer">11</span>
<span class="name">min_pro_len</span> <span class="operator">=</span> <span class="literal number integer">100</span>

<span class="keyword">def</span> <span class="name function">find_orfs_with_trans</span><span class="punctuation">(</span><span class="name">seq</span><span class="punctuation">,</span> <span class="name">trans_table</span><span class="punctuation">,</span> <span class="name">min_protein_length</span><span class="punctuation">):</span>
    <span class="name">answer</span> <span class="operator">=</span> <span class="punctuation">[]</span>
    <span class="name">seq_len</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq</span><span class="punctuation">)</span>
    <span class="keyword">for</span> <span class="name">strand</span><span class="punctuation">,</span> <span class="name">nuc</span> <span class="operator word">in</span> <span class="punctuation">[(</span><span class="operator">+</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">seq</span><span class="punctuation">),</span> <span class="punctuation">(</span><span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="name">seq</span><span class="operator">.</span><span class="name">reverse_complement</span><span class="punctuation">())]:</span>
        <span class="keyword">for</span> <span class="name">frame</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="literal number integer">3</span><span class="punctuation">):</span>
            <span class="name">trans</span> <span class="operator">=</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">nuc</span><span class="punctuation">[</span><span class="name">frame</span><span class="punctuation">:]</span><span class="operator">.</span><span class="name">translate</span><span class="punctuation">(</span><span class="name">trans_table</span><span class="punctuation">))</span>
            <span class="name">trans_len</span> <span class="operator">=</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">trans</span><span class="punctuation">)</span>
            <span class="name">aa_start</span> <span class="operator">=</span> <span class="literal number integer">0</span>
            <span class="name">aa_end</span> <span class="operator">=</span> <span class="literal number integer">0</span>
            <span class="keyword">while</span> <span class="name">aa_start</span> <span class="operator">&lt;</span> <span class="name">trans_len</span><span class="punctuation">:</span>
                <span class="name">aa_end</span> <span class="operator">=</span> <span class="name">trans</span><span class="operator">.</span><span class="name">find</span><span class="punctuation">(</span><span class="literal string">&quot;*&quot;</span><span class="punctuation">,</span> <span class="name">aa_start</span><span class="punctuation">)</span>
                <span class="keyword">if</span> <span class="name">aa_end</span> <span class="operator">==</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">:</span>
                    <span class="name">aa_end</span> <span class="operator">=</span> <span class="name">trans_len</span>
                <span class="keyword">if</span> <span class="name">aa_end</span><span class="operator">-</span><span class="name">aa_start</span> <span class="operator">&gt;=</span> <span class="name">min_protein_length</span><span class="punctuation">:</span>
                    <span class="keyword">if</span> <span class="name">strand</span> <span class="operator">==</span> <span class="literal number integer">1</span><span class="punctuation">:</span>
                        <span class="name">start</span> <span class="operator">=</span> <span class="name">frame</span><span class="operator">+</span><span class="name">aa_start</span><span class="operator">*</span><span class="literal number integer">3</span>
                        <span class="name">end</span> <span class="operator">=</span> <span class="name builtin">min</span><span class="punctuation">(</span><span class="name">seq_len</span><span class="punctuation">,</span><span class="name">frame</span><span class="operator">+</span><span class="name">aa_end</span><span class="operator">*</span><span class="literal number integer">3</span><span class="operator">+</span><span class="literal number integer">3</span><span class="punctuation">)</span>
                    <span class="keyword">else</span><span class="punctuation">:</span>
                        <span class="name">start</span> <span class="operator">=</span> <span class="name">seq_len</span><span class="operator">-</span><span class="name">frame</span><span class="operator">-</span><span class="name">aa_end</span><span class="operator">*</span><span class="literal number integer">3</span><span class="operator">-</span><span class="literal number integer">3</span>
                        <span class="name">end</span> <span class="operator">=</span> <span class="name">seq_len</span><span class="operator">-</span><span class="name">frame</span><span class="operator">-</span><span class="name">aa_start</span><span class="operator">*</span><span class="literal number integer">3</span>
                    <span class="name">answer</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">((</span><span class="name">start</span><span class="punctuation">,</span> <span class="name">end</span><span class="punctuation">,</span> <span class="name">strand</span><span class="punctuation">,</span>
                                   <span class="name">trans</span><span class="punctuation">[</span><span class="name">aa_start</span><span class="punctuation">:</span><span class="name">aa_end</span><span class="punctuation">]))</span>
                <span class="name">aa_start</span> <span class="operator">=</span> <span class="name">aa_end</span><span class="operator">+</span><span class="literal number integer">1</span>
    <span class="name">answer</span><span class="operator">.</span><span class="name">sort</span><span class="punctuation">()</span>
    <span class="keyword">return</span> <span class="name">answer</span>

<span class="name">orf_list</span> <span class="operator">=</span> <span class="name">find_orfs_with_trans</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">,</span> <span class="name">table</span><span class="punctuation">,</span> <span class="name">min_pro_len</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">start</span><span class="punctuation">,</span> <span class="name">end</span><span class="punctuation">,</span> <span class="name">strand</span><span class="punctuation">,</span> <span class="name">pro</span> <span class="operator word">in</span> <span class="name">orf_list</span><span class="punctuation">:</span>
    <span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string">...</span><span class="literal string interpol">%s</span><span class="literal string"> - length </span><span class="literal string interpol">%i</span><span class="literal string">, strand </span><span class="literal string interpol">%i</span><span class="literal string">, </span><span class="literal string interpol">%i</span><span class="literal string">:</span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> \
          <span class="operator">%</span> <span class="punctuation">(</span><span class="name">pro</span><span class="punctuation">[:</span><span class="literal number integer">30</span><span class="punctuation">],</span> <span class="name">pro</span><span class="punctuation">[</span><span class="operator">-</span><span class="literal number integer">3</span><span class="punctuation">:],</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">pro</span><span class="punctuation">),</span> <span class="name">strand</span><span class="punctuation">,</span> <span class="name">start</span><span class="punctuation">,</span> <span class="name">end</span><span class="punctuation">)</span>
</pre>
<p>输出是：</p>
<pre class="code python literal-block">
<span class="name">NQIQGVICSPDSGEFMVTFETVMEIKILHK</span><span class="operator">...</span><span class="name">GVA</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">355</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">41</span><span class="punctuation">:</span><span class="literal number integer">1109</span>
<span class="name">WDVKTVTGVLHHPFHLTFSLCPEGATQSGR</span><span class="operator">...</span><span class="name">VKR</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">111</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">491</span><span class="punctuation">:</span><span class="literal number integer">827</span>
<span class="name">KSGELRQTPPASSTLHLRLILQRSGVMMEL</span><span class="operator">...</span><span class="name">NPE</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">285</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">1030</span><span class="punctuation">:</span><span class="literal number integer">1888</span>
<span class="name">RALTGLSAPGIRSQTSCDRLRELRYVPVSL</span><span class="operator">...</span><span class="name">PLQ</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">119</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2830</span><span class="punctuation">:</span><span class="literal number integer">3190</span>
<span class="name">RRKEHVSKKRRPQKRPRRRRFFHRLRPPDE</span><span class="operator">...</span><span class="name">PTR</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">128</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">3470</span><span class="punctuation">:</span><span class="literal number integer">3857</span>
<span class="name">GLNCSFFSICNWKFIDYINRLFQIIYLCKN</span><span class="operator">...</span><span class="name">YYH</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">176</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">4249</span><span class="punctuation">:</span><span class="literal number integer">4780</span>
<span class="name">RGIFMSDTMVVNGSGGVPAFLFSGSTLSSY</span><span class="operator">...</span><span class="name">LLK</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">361</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">4814</span><span class="punctuation">:</span><span class="literal number integer">5900</span>
<span class="name">VKKILYIKALFLCTVIKLRRFIFSVNNMKF</span><span class="operator">...</span><span class="name">DLP</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">165</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">5923</span><span class="punctuation">:</span><span class="literal number integer">6421</span>
<span class="name">LSHTVTDFTDQMAQVGLCQCVNVFLDEVTG</span><span class="operator">...</span><span class="name">KAA</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">107</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">5974</span><span class="punctuation">:</span><span class="literal number integer">6298</span>
<span class="name">GCLMKKSSIVATIITILSGSANAASSQLIP</span><span class="operator">...</span><span class="name">YRF</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">315</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">6654</span><span class="punctuation">:</span><span class="literal number integer">7602</span>
<span class="name">IYSTSEHTGEQVMRTLDEVIASRSPESQTR</span><span class="operator">...</span><span class="name">FHV</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">111</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">7788</span><span class="punctuation">:</span><span class="literal number integer">8124</span>
<span class="name">WGKLQVIGLSMWMVLFSQRFDDWLNEQEDA</span><span class="operator">...</span><span class="name">ESK</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">125</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">8087</span><span class="punctuation">:</span><span class="literal number integer">8465</span>
<span class="name">TGKQNSCQMSAIWQLRQNTATKTRQNRARI</span><span class="operator">...</span><span class="name">AIK</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">100</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">8741</span><span class="punctuation">:</span><span class="literal number integer">9044</span>
<span class="name">QGSGYAFPHASILSGIAMSHFYFLVLHAVK</span><span class="operator">...</span><span class="name">CSD</span> <span class="operator">-</span> <span class="name">length</span> <span class="literal number integer">114</span><span class="punctuation">,</span> <span class="name">strand</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">9264</span><span class="punctuation">:</span><span class="literal number integer">9609</span>
</pre>
<p>如果你注释掉排序语句，那么蛋白序列将和之前显示的顺序一样，所以你能确定这是在做相同的事情。这里，我们可以按位置进行排序，使得和GenBank文件中的实际注释相比对较更加容易（就像章节 <a class="reference external" href="#sec:gd_nice_example">17.1.9</a> 中显示的那样）。</p>
<p>然而，如果你想要的只是所有开放读码框的位置，翻译每一个可能的密码子将是很浪费时间的，包括转换和查找反向互补链。那么，你所要做的所有事情是查找可能的终止密码子(和他们反向互补)。使用正则表达式是一个很直接的方式（参见Python中的 <tt class="docutils literal">re</tt> 模块）。这是描述查找字符串的一个非常强大的模块（然而非常复杂），也被许多编程语言和命令行工具，如 <tt class="docutils literal">grep</tt>，所支持。你能找到一本书来描述它的使用！</p>
</div>
</div>
<div class="section" id="id230">
<h2>18.2  序列解析与简单作图</h2>
<p>这一部分展示更多使用第 <a class="reference external" href="#chapter:Bio.SeqIO">5</a> 章介绍的 <tt class="docutils literal">Bio.SeqIO</tt> 模块进行序列解析的例子，以及Python类库matplotlib中 <tt class="docutils literal">pylab</tt> 的作图接口（参见 <a class="reference external" href="http://matplotlib.sourceforge.net/">matplotlib 主页的教程</a> ）。注意，跟随这些例子，你需要安装matplotlib - 但是即使没有它，你依然可以尝试数据的解析的内容。</p>
<div class="section" id="id231">
<h3>18.2.1  序列长度柱状图</h3>
<p>在很多时候，你可能想要将某个数据集中的序列长度分布进行可视化 —— 例如，基因组组装项目中的contig的大小范围。在这个例子中，我们将再次使用我们的兰花FASTA文件 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta">ls_orchid.fasta</a> ，它只包含94条序列。</p>
<p>首先，我们使用 <tt class="docutils literal">Bio.SeqIO</tt> 来解析这个FASTA文件，并创建一个序列长度的列表。你可以用一个for循环来实现，然而我觉得列表解析（list comprehension）更简洁：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sizes</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)]</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">),</span> <span class="name builtin">min</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">),</span> <span class="name builtin">max</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">)</span>
<span class="punctuation">(</span><span class="literal number integer">94</span><span class="punctuation">,</span> <span class="literal number integer">572</span><span class="punctuation">,</span> <span class="literal number integer">789</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">sizes</span>
<span class="punctuation">[</span><span class="literal number integer">740</span><span class="punctuation">,</span> <span class="literal number integer">753</span><span class="punctuation">,</span> <span class="literal number integer">748</span><span class="punctuation">,</span> <span class="literal number integer">744</span><span class="punctuation">,</span> <span class="literal number integer">733</span><span class="punctuation">,</span> <span class="literal number integer">718</span><span class="punctuation">,</span> <span class="literal number integer">730</span><span class="punctuation">,</span> <span class="literal number integer">704</span><span class="punctuation">,</span> <span class="literal number integer">740</span><span class="punctuation">,</span> <span class="literal number integer">709</span><span class="punctuation">,</span> <span class="literal number integer">700</span><span class="punctuation">,</span> <span class="literal number integer">726</span><span class="punctuation">,</span> <span class="operator">...</span><span class="punctuation">,</span> <span class="literal number integer">592</span><span class="punctuation">]</span>
</pre>
<p>现在我们得到了所有基因的长度（以整数列表的形式），我们可以用matplotlib的柱状图功能来显示它。</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">sizes</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)]</span>

<span class="keyword namespace">import</span> <span class="name namespace">pylab</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">hist</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">,</span> <span class="name">bins</span><span class="operator">=</span><span class="literal number integer">20</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">title</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string interpol">%i</span><span class="literal string"> orchid sequences</span><span class="literal string escape">\n</span><span class="literal string">Lengths </span><span class="literal string interpol">%i</span><span class="literal string"> to </span><span class="literal string interpol">%i</span><span class="literal string">&quot;</span> \
            <span class="operator">%</span> <span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">),</span><span class="name builtin">min</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">),</span><span class="name builtin">max</span><span class="punctuation">(</span><span class="name">sizes</span><span class="punctuation">)))</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">xlabel</span><span class="punctuation">(</span><span class="literal string">&quot;Sequence length (bp)&quot;</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">ylabel</span><span class="punctuation">(</span><span class="literal string">&quot;Count&quot;</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">show</span><span class="punctuation">()</span>
</pre>
<p>这将弹出一个包含如下图形的新的窗口：</p>
<p><img alt="image26" src="../images/hist_plot.png" /></p>
<p>注意，这些兰花序列的长度大多数大约在740bp左右，这里有可能有两个不同长度的序列分类，其中包含一个较短的序列子集。</p>
<p><em>提示：</em> 除了使用 <tt class="docutils literal">pylab.show()</tt> 在窗口中显示图像以外，你也可以使用 <tt class="docutils literal"><span class="pre">pylab.savefig(...)</span></tt> 来将图像保存为图像文件中（例如PNG或PDF文件）。</p>
</div>
<div class="section" id="gc">
<h3>18.2.2  序列GC%含量作图</h3>
<p>对于核酸序列，另一个经常计算的值是GC含量。例如，你可能想要查看一个细菌基因组中所有基因的GC%，并研究任何离群值来确定可能最近通过基因水平转移而获得的基因。同样，对于这个例子，我们再次使用兰花FASTA文件 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta">ls_orchid.fasta</a> 。</p>
<p>首先，我们使用 <tt class="docutils literal">Bio.SeqIO</tt> 解析这个FASTA文件并创建一个GC百分含量的列表。你可以使用for循环，但我更喜欢这样：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SeqUtils</span> <span class="keyword namespace">import</span> <span class="name">GC</span>

<span class="name">gc_values</span> <span class="operator">=</span> <span class="name builtin">sorted</span><span class="punctuation">(</span><span class="name">GC</span><span class="punctuation">(</span><span class="name">rec</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span> <span class="keyword">for</span> <span class="name">rec</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">))</span>
</pre>
<p>读取完每个序列并计算了GC百分比，我们接着将它们按升序排列。现在，我们用matplotlib来对这个浮点数列表进行可视化：</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">pylab</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">plot</span><span class="punctuation">(</span><span class="name">gc_values</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">title</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string interpol">%i</span><span class="literal string"> orchid sequences</span><span class="literal string escape">\n</span><span class="literal string">GC</span><span class="literal string interpol">%%</span><span class="literal string"> </span><span class="literal string interpol">%0.1f</span><span class="literal string"> to </span><span class="literal string interpol">%0.1f</span><span class="literal string">&quot;</span> \
            <span class="operator">%</span> <span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">gc_values</span><span class="punctuation">),</span><span class="name builtin">min</span><span class="punctuation">(</span><span class="name">gc_values</span><span class="punctuation">),</span><span class="name builtin">max</span><span class="punctuation">(</span><span class="name">gc_values</span><span class="punctuation">)))</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">xlabel</span><span class="punctuation">(</span><span class="literal string">&quot;Genes&quot;</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">ylabel</span><span class="punctuation">(</span><span class="literal string">&quot;GC%&quot;</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">show</span><span class="punctuation">()</span>
</pre>
<p>像之前的例子一样，弹出一个窗口中将包含如下图形：</p>
<p><img alt="image27" src="../images/gc_plot.png" /></p>
<p>如果你使用的是一个物种中的所有基因集，你可能会得到一个更加平滑的图。</p>
</div>
<div class="section" id="id232">
<h3>18.2.3  核苷酸点线图</h3>
<p>点线图是可视化比较两条核苷酸序列的相似性的一种方式。采用一个滑动窗来相互比较较短的子序列（比较通常根据一个不匹配阈值来实现）。为了简单起见，此处我们将只查找完全匹配（如下图黑色所示）。</p>
<p>我们需要两条序列开始。为了论证，我们只取兰花FASTA文件中的前两条序列。<a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/ls_orchid.fasta">ls_orchid.fasta</a>:</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;ls_orchid.fasta&quot;</span><span class="punctuation">)</span>
<span class="name">record_iterator</span> <span class="operator">=</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">)</span>
<span class="name">rec_one</span> <span class="operator">=</span> <span class="name">record_iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="name">rec_two</span> <span class="operator">=</span> <span class="name">record_iterator</span><span class="operator">.</span><span class="name">next</span><span class="punctuation">()</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>我们将展示两种方式。首先，一个简单的实现，它将所有滑动窗大小的子序列相互比较，并生产一个相似性矩阵。你可以创建一个矩阵或数组对象，而在这儿，我们只用一个用嵌套的列表解析生成的布尔值列表的列表。</p>
<pre class="code python literal-block">
<span class="name">window</span> <span class="operator">=</span> <span class="literal number integer">7</span>
<span class="name">seq_one</span> <span class="operator">=</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">rec_one</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">()</span>
<span class="name">seq_two</span> <span class="operator">=</span> <span class="name builtin">str</span><span class="punctuation">(</span><span class="name">rec_two</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">()</span>
<span class="name">data</span> <span class="operator">=</span> <span class="punctuation">[[(</span><span class="name">seq_one</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">:</span><span class="name">i</span><span class="operator">+</span><span class="name">window</span><span class="punctuation">]</span> <span class="operator">&lt;&gt;</span> <span class="name">seq_two</span><span class="punctuation">[</span><span class="name">j</span><span class="punctuation">:</span><span class="name">j</span><span class="operator">+</span><span class="name">window</span><span class="punctuation">])</span> \
        <span class="keyword">for</span> <span class="name">j</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_one</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">window</span><span class="punctuation">)]</span> \
       <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq_two</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">window</span><span class="punctuation">)]</span>
</pre>
<p>注意，我们在这里并 <em>没有</em> 检查反向的互补匹配。现在我们将使用matplotlib的 <tt class="docutils literal">pylab.imshow()</tt> 函数来显示这个数据，首先启用灰度模式，以保证这是在黑白颜色下完成的：</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">pylab</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">gray</span><span class="punctuation">()</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">imshow</span><span class="punctuation">(</span><span class="name">data</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">xlabel</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> (length </span><span class="literal string interpol">%i</span><span class="literal string"> bp)&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">rec_one</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec_one</span><span class="punctuation">)))</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">ylabel</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> (length </span><span class="literal string interpol">%i</span><span class="literal string"> bp)&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">rec_two</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec_two</span><span class="punctuation">)))</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">title</span><span class="punctuation">(</span><span class="literal string">&quot;Dot plot using window size </span><span class="literal string interpol">%i</span><span class="literal string escape">\n</span><span class="literal string">(allowing no mis-matches)&quot;</span> <span class="operator">%</span> <span class="name">window</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">show</span><span class="punctuation">()</span>
</pre>
<p>这将弹出一个新的窗口，包含类似这样的图形：</p>
<p><img alt="image28" src="../images/dot_plot.png" /></p>
<p>可能如您所料，这两条序列非常相似，图中部分滑动窗大小的线沿着对角线匹配。图中没有对角线外的点，这意味着序列中并没有倒位或其他有趣的偏离对角线匹配。</p>
<p>上面的代码在小的例子中工作得很好，但是应用到大的序列时，这里有两个问题。首先，这种以穷举地方式进行所有可能的两两比对非常缓慢。作为替代，我们将创建一个词典来映射所有滑动窗大小的子序列的位置，然后取两者的交集来获得两条序列中都发现的子序列。这将占用更多的内存，然而速度 <em>更</em> 快。另外， <tt class="docutils literal">pylab.imshow()</tt> 函数只能显示较小的矩阵。作为替代，我们将使用 <tt class="docutils literal">pylab.scatter()</tt> 函数。</p>
<p>我们从创建，从滑动窗大小的子序列到其位置的字典映射开始：</p>
<pre class="code python literal-block">
<span class="name">window</span> <span class="operator">=</span> <span class="literal number integer">7</span>
<span class="name">dict_one</span> <span class="operator">=</span> <span class="punctuation">{}</span>
<span class="name">dict_two</span> <span class="operator">=</span> <span class="punctuation">{}</span>
<span class="keyword">for</span> <span class="punctuation">(</span><span class="name">seq</span><span class="punctuation">,</span> <span class="name">section_dict</span><span class="punctuation">)</span> <span class="operator word">in</span> <span class="punctuation">[(</span><span class="name builtin">str</span><span class="punctuation">(</span><span class="name">rec_one</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">(),</span> <span class="name">dict_one</span><span class="punctuation">),</span>
                            <span class="punctuation">(</span><span class="name builtin">str</span><span class="punctuation">(</span><span class="name">rec_two</span><span class="operator">.</span><span class="name">seq</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">upper</span><span class="punctuation">(),</span> <span class="name">dict_two</span><span class="punctuation">)]:</span>
    <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name builtin">range</span><span class="punctuation">(</span><span class="name builtin">len</span><span class="punctuation">(</span><span class="name">seq</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">window</span><span class="punctuation">):</span>
        <span class="name">section</span> <span class="operator">=</span> <span class="name">seq</span><span class="punctuation">[</span><span class="name">i</span><span class="punctuation">:</span><span class="name">i</span><span class="operator">+</span><span class="name">window</span><span class="punctuation">]</span>
        <span class="keyword">try</span><span class="punctuation">:</span>
            <span class="name">section_dict</span><span class="punctuation">[</span><span class="name">section</span><span class="punctuation">]</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">i</span><span class="punctuation">)</span>
        <span class="keyword">except</span> <span class="name exception">KeyError</span><span class="punctuation">:</span>
            <span class="name">section_dict</span><span class="punctuation">[</span><span class="name">section</span><span class="punctuation">]</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="name">i</span><span class="punctuation">]</span>
<span class="comment">#Now find any sub-sequences found in both sequences</span>
<span class="comment">#(Python 2.3 would require slightly different code here)</span>
<span class="name">matches</span> <span class="operator">=</span> <span class="name builtin">set</span><span class="punctuation">(</span><span class="name">dict_one</span><span class="punctuation">)</span><span class="operator">.</span><span class="name">intersection</span><span class="punctuation">(</span><span class="name">dict_two</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;</span><span class="literal string interpol">%i</span><span class="literal string"> unique matches&quot;</span> <span class="operator">%</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">matches</span><span class="punctuation">)</span>
</pre>
<p>为了使用 <tt class="docutils literal">pylab.scatter()</tt> 函数，我们需要两个分别对应 <em>x</em> 和 <em>y</em> 轴的列表：</p>
<pre class="code python literal-block">
<span class="comment">#Create lists of x and y co-ordinates for scatter plot</span>
<span class="name">x</span> <span class="operator">=</span> <span class="punctuation">[]</span>
<span class="name">y</span> <span class="operator">=</span> <span class="punctuation">[]</span>
<span class="keyword">for</span> <span class="name">section</span> <span class="operator word">in</span> <span class="name">matches</span><span class="punctuation">:</span>
    <span class="keyword">for</span> <span class="name">i</span> <span class="operator word">in</span> <span class="name">dict_one</span><span class="punctuation">[</span><span class="name">section</span><span class="punctuation">]:</span>
        <span class="keyword">for</span> <span class="name">j</span> <span class="operator word">in</span> <span class="name">dict_two</span><span class="punctuation">[</span><span class="name">section</span><span class="punctuation">]:</span>
            <span class="name">x</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">i</span><span class="punctuation">)</span>
            <span class="name">y</span><span class="operator">.</span><span class="name">append</span><span class="punctuation">(</span><span class="name">j</span><span class="punctuation">)</span>
</pre>
<p>现在我们能以散点图的形式画出优化后的点线图：</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">pylab</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">cla</span><span class="punctuation">()</span> <span class="comment">#clear any prior graph</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">gray</span><span class="punctuation">()</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">scatter</span><span class="punctuation">(</span><span class="name">x</span><span class="punctuation">,</span><span class="name">y</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">xlim</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec_one</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">window</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">ylim</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec_two</span><span class="punctuation">)</span><span class="operator">-</span><span class="name">window</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">xlabel</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> (length </span><span class="literal string interpol">%i</span><span class="literal string"> bp)&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">rec_one</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec_one</span><span class="punctuation">)))</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">ylabel</span><span class="punctuation">(</span><span class="literal string">&quot;</span><span class="literal string interpol">%s</span><span class="literal string"> (length </span><span class="literal string interpol">%i</span><span class="literal string"> bp)&quot;</span> <span class="operator">%</span> <span class="punctuation">(</span><span class="name">rec_two</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">rec_two</span><span class="punctuation">)))</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">title</span><span class="punctuation">(</span><span class="literal string">&quot;Dot plot using window size </span><span class="literal string interpol">%i</span><span class="literal string escape">\n</span><span class="literal string">(allowing no mis-matches)&quot;</span> <span class="operator">%</span> <span class="name">window</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">show</span><span class="punctuation">()</span>
</pre>
<p>这将弹出一个新的窗口，包含如下图形：</p>
<p><img alt="image29" src="../images/dot_plot_scatter.png" /></p>
<p>我个人认为第二个图更加易读！再次注意，我们在这里 <em>没有</em> 检查反向互补匹配 —— 你可以扩展这个例子来实现它，或许可以以一种颜色显示正向匹配，另一种显示反向匹配。</p>
</div>
<div class="section" id="id233">
<h3>18.2.4  绘制序列读长数据的质量图</h3>
<p>如果你在处理二代测序数据，你可能希望绘制数据的质量图。这里使用两个包含双末端（paired end）读长的FASTQ文件作为例子，其中 <tt class="docutils literal">SRR001666_1.fastq</tt> 为正向读长， <tt class="docutils literal">SRR001666_2.fastq</tt> 为反向读长。它们可以从ENA序列读长档案的FTP站点下载（ <a class="reference external" href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR001/SRR001666/SRR001666_1.fastq.gz">ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR001/SRR001666/SRR001666_1.fastq.gz</a> 和 <a class="reference external" href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR001/SRR001666/SRR001666_2.fastq.gz">ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR001/SRR001666/SRR001666_2.fastq.gz</a> ）， 且来自 <em>E. coli</em> —— 参见 <a class="reference external" href="http://www.ebi.ac.uk/ena/data/view/SRR001666">http://www.ebi.ac.uk/ena/data/view/SRR001666</a> 的详细介绍。在下面的代码中， <tt class="docutils literal"><span class="pre">pylab.subplot(...)</span></tt> 函数被用来在两个子图中展示正向和反向的质量。这里也有少量的代码来保证仅仅展示前50个读长的质量。</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">pylab</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">for</span> <span class="name">subfigure</span> <span class="operator word">in</span> <span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">,</span><span class="literal number integer">2</span><span class="punctuation">]:</span>
    <span class="name">filename</span> <span class="operator">=</span> <span class="literal string">&quot;SRR001666_</span><span class="literal string interpol">%i</span><span class="literal string">.fastq&quot;</span> <span class="operator">%</span> <span class="name">subfigure</span>
    <span class="name">pylab</span><span class="operator">.</span><span class="name">subplot</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">,</span> <span class="literal number integer">2</span><span class="punctuation">,</span> <span class="name">subfigure</span><span class="punctuation">)</span>
    <span class="keyword">for</span> <span class="name">i</span><span class="punctuation">,</span><span class="name">record</span> <span class="operator word">in</span> <span class="name builtin">enumerate</span><span class="punctuation">(</span><span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span> <span class="literal string">&quot;fastq&quot;</span><span class="punctuation">)):</span>
        <span class="keyword">if</span> <span class="name">i</span> <span class="operator">&gt;=</span> <span class="literal number integer">50</span> <span class="punctuation">:</span> <span class="keyword">break</span> <span class="comment">#trick!</span>
        <span class="name">pylab</span><span class="operator">.</span><span class="name">plot</span><span class="punctuation">(</span><span class="name">record</span><span class="operator">.</span><span class="name">letter_annotations</span><span class="punctuation">[</span><span class="literal string">&quot;phred_quality&quot;</span><span class="punctuation">])</span>
    <span class="name">pylab</span><span class="operator">.</span><span class="name">ylim</span><span class="punctuation">(</span><span class="literal number integer">0</span><span class="punctuation">,</span><span class="literal number integer">45</span><span class="punctuation">)</span>
    <span class="name">pylab</span><span class="operator">.</span><span class="name">ylabel</span><span class="punctuation">(</span><span class="literal string">&quot;PHRED quality score&quot;</span><span class="punctuation">)</span>
    <span class="name">pylab</span><span class="operator">.</span><span class="name">xlabel</span><span class="punctuation">(</span><span class="literal string">&quot;Position&quot;</span><span class="punctuation">)</span>
<span class="name">pylab</span><span class="operator">.</span><span class="name">savefig</span><span class="punctuation">(</span><span class="literal string">&quot;SRR001666.png&quot;</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;Done&quot;</span>
</pre>
<p>你应该注意到，这里我们使用了 <tt class="docutils literal">Bio.SeqIO</tt> 的格式名称 <tt class="docutils literal">fastq</tt> ，因为NCBI使用标准Sanger FASTQ和PHRED分数的存储这些读长。然而，你可能从读长的长度中猜到，这些数据来自Illumina Genome Analyzer，而且可能最初是以Solexa/Illumina FASTQ两种格式变种中的一种存在。</p>
<p>这个例子使用 <tt class="docutils literal"><span class="pre">pylab.savefig(...)</span></tt> 函数，而不是``pylab.show(...)`` ，然而就像前面提到的一样，它们两者都非常有用。下面是得到的结果：</p>
<p><img alt="image30" src="../images/SRR001666.png" /></p>
</div>
</div>
<div class="section" id="id234">
<h2>18.3  处理序列比对</h2>
<p>这部分可以看做是第 <a class="reference external" href="#chapter:Bio.AlignIO">6</a> 章的继续。</p>
<div class="section" id="id235">
<h3>18.3.1  计算摘要信息</h3>
<p>一旦你有一个比对，你很可能希望找出关于它的一些信息。我们尽力将这些功能分离到单独的能作用于比对对象的类中，而不是将所有的能生成比对信息的函数都放入比对对象本身。</p>
<p>准备计算比对对象的摘要信息非常快捷。假设我们已经得到了一个比对对象 <tt class="docutils literal">alignment</tt> ，例如由在第 <a class="reference external" href="#chapter:Bio.AlignIO">6</a> 章介绍的 <tt class="docutils literal"><span class="pre">Bio.AlignIO.read(...)</span></tt> 读入。我们获得该对象的摘要信息所要做的所有事情是：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Align</span> <span class="keyword namespace">import</span> <span class="name">AlignInfo</span>
<span class="name">summary_align</span> <span class="operator">=</span> <span class="name">AlignInfo</span><span class="operator">.</span><span class="name">SummaryInfo</span><span class="punctuation">(</span><span class="name">alignment</span><span class="punctuation">)</span>
</pre>
<p><tt class="docutils literal">summary_align</tt> 对象非常有用，它将帮你做以下巧妙的事情：</p>
<ol class="arabic simple">
<li>计算一个快速一致序列 – 参见章节 <a class="reference external" href="#sec:consensus">18.3.2</a></li>
<li>获取一个针对该比对的位点特异性打分矩阵 – 参见章节 <a class="reference external" href="#sec:pssm">18.3.3</a></li>
<li>计算比对的信息量 – 参见章节 <a class="reference external" href="#sec:getting_info_content">18.3.4</a></li>
<li>生成该比对中的替换信息 – 章节 <a class="reference external" href="#sec:sub_matrix">18.4</a> 详细描述了使用该方法生成一个替换矩阵</li>
</ol>
</div>
<div class="section" id="id236">
<h3>18.3.2  计算一个快速一致序列</h3>
<p>在章节 <a class="reference external" href="#sec:summary_info">18.3.1</a> 中描述的 <tt class="docutils literal">SummaryInfo</tt> 对象提供了一个可以快速计算比对的保守（consensus）序列的功能。假设我们有一个 <tt class="docutils literal">SummaryInfo</tt> 对象，叫做 <tt class="docutils literal">summary_align</tt>，我们能通过下面的方法计算一个保守序列：</p>
<pre class="code python literal-block">
<span class="name">consensus</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">dumb_consensus</span><span class="punctuation">()</span>
</pre>
<p>就行名字显示的那样，这是一个非常简单的保守序列计算器，它将只是在保守序列中累加每个位点的所有残基，如果最普遍的残基数大于某个阈值时，这个最普遍的残基将被添加到保守序列中。如果它没有到达这个阈值，将添加一个“不确定字符”。最终返回的保守序列对象是一个Seq对象，它的字母表是从组成保守序列所有序列的字母表中推断出来的。所以使用 <tt class="docutils literal">print consensus</tt> 将给出如下信息：</p>
<pre class="code python literal-block">
<span class="name">consensus</span> <span class="name">Seq</span><span class="punctuation">(</span><span class="literal string">'TATACATNAAAGNAGGGGGATGCGGATAAATGGAAAGGCGAAAGAAAGAAAAAAATGAAT</span>
<span class="operator">...</span><span class="literal string">', IUPACAmbiguousDNA())</span>
</pre>
<p>你可以通过传入可选参数来调整 <tt class="docutils literal">dumb_consensus</tt> 的工作方式：</p>
<dl class="docutils">
<dt><strong>the threshold</strong></dt>
<dd>这是用来设定某个残基在某个位点出现的频率超过一定阈值，才将其添加到保守序列。默认为0.7（即70%）。</dd>
<dt><strong>the ambiguous character</strong></dt>
<dd>指定保守序列中的不确定字符。默认为’N’。</dd>
<dt><strong>the consensus alphabet</strong></dt>
<dd>指定保守序列的字母表。如果没有提供，我们将从比对序列的字母表基础上推断该字母表。</dd>
</dl>
</div>
<div class="section" id="id237">
<h3>18.3.3  位点特异性评分矩阵</h3>
<p>位点特异性评分矩阵（Position specific score matrices，PSSMs）以另一种总结比对信息的方式（与刚才介绍的保守序列不同），这或许在某些情况下更为有用。简单来说，PSSM是一个计数矩阵。对于比对中的每一列，将所有可能出现的字母进行计数并加和。这些加和值将和一个代表序列（默认为比对中的第一条序列）一起显示出来。这个序列可能是保守序列，但也可以是比对中的任何序列。例如，对于比对，</p>
<pre class="code python literal-block">
<span class="name">GTATC</span>
<span class="name">AT</span><span class="operator">--</span><span class="name">C</span>
<span class="name">CTGTC</span>
</pre>
<p>它的PSSM是：</p>
<pre class="code python literal-block">
  <span class="name">G</span> <span class="name">A</span> <span class="name">T</span> <span class="name">C</span>
<span class="name">G</span> <span class="literal number integer">1</span> <span class="literal number integer">1</span> <span class="literal number integer">0</span> <span class="literal number integer">1</span>
<span class="name">T</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span> <span class="literal number integer">3</span> <span class="literal number integer">0</span>
<span class="name">A</span> <span class="literal number integer">1</span> <span class="literal number integer">1</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span>
<span class="name">T</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span> <span class="literal number integer">2</span> <span class="literal number integer">0</span>
<span class="name">C</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span> <span class="literal number integer">0</span> <span class="literal number integer">3</span>
</pre>
<p>假设我们有一个比对对象叫做 <tt class="docutils literal">c_align</tt> ，为了获得PSSM和保守序列，我们首先得到一个摘要对象（summary object），并计算一致序列：</p>
<pre class="code python literal-block">
<span class="name">summary_align</span> <span class="operator">=</span> <span class="name">AlignInfo</span><span class="operator">.</span><span class="name">SummaryInfo</span><span class="punctuation">(</span><span class="name">c_align</span><span class="punctuation">)</span>
<span class="name">consensus</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">dumb_consensus</span><span class="punctuation">()</span>
</pre>
<p>现在，我们想创建PSSM，但是在计算中忽略任何 <tt class="docutils literal">N</tt> 不确定残基：</p>
<pre class="code python literal-block">
<span class="name">my_pssm</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">pos_specific_score_matrix</span><span class="punctuation">(</span><span class="name">consensus</span><span class="punctuation">,</span>
                                                  <span class="name">chars_to_ignore</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">'N'</span><span class="punctuation">])</span>
</pre>
<p>关于此有亮点需要说明：</p>
<ol class="arabic">
<li><p class="first">为了维持字母表的严格性，你可以在PSSM的顶部显示比对对象字母表中规定的字符。空白字符（Gaps）并不包含在PSSM的顶轴中。</p>
</li>
<li><p class="first">传入并显示在左侧轴的序列可以不是保守序列。例如，你如果想要在PSSM左边显示比对中的第二条序列，你只需要：</p>
<pre class="code python literal-block">
<span class="name">second_seq</span> <span class="operator">=</span> <span class="name">alignment</span><span class="operator">.</span><span class="name">get_seq_by_num</span><span class="punctuation">(</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="name">my_pssm</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">pos_specific_score_matrix</span><span class="punctuation">(</span><span class="name">second_seq</span>
                                                  <span class="name">chars_to_ignore</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">'N'</span><span class="punctuation">])</span>
</pre>
</li>
</ol>
<p>以上的命令将返回一个 <tt class="docutils literal">PSSM</tt> 对象。为了显示出PSSM，我们只需 <tt class="docutils literal">print my_pssm</tt>，结果如下：</p>
<pre class="code python literal-block">
    <span class="name">A</span>   <span class="name">C</span>   <span class="name">G</span>   <span class="name">T</span>
<span class="name">T</span>  <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">7.0</span>
<span class="name">A</span>  <span class="literal number float">7.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span>
<span class="name">T</span>  <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">7.0</span>
<span class="name">A</span>  <span class="literal number float">7.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span>
<span class="name">C</span>  <span class="literal number float">0.0</span> <span class="literal number float">7.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span>
<span class="name">A</span>  <span class="literal number float">7.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span>
<span class="name">T</span>  <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">7.0</span>
<span class="name">T</span>  <span class="literal number float">1.0</span> <span class="literal number float">0.0</span> <span class="literal number float">0.0</span> <span class="literal number float">6.0</span>
<span class="operator">...</span>
</pre>
<p>你可以用 <tt class="docutils literal"><span class="pre">your_pssm[sequence_number][residue_count_name]</span></tt> 获得任何PSSM的元素。例如，获取上面PSSM中第二个元素的‘A’残基的计数，你可以：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_pssm</span><span class="punctuation">[</span><span class="literal number integer">1</span><span class="punctuation">][</span><span class="literal string">&quot;A&quot;</span><span class="punctuation">]</span>
<span class="literal number float">7.0</span>
</pre>
<p>PSSM类的结构有望使得获取元素和打印漂亮的矩阵都很方便。</p>
</div>
<div class="section" id="id238">
<h3>18.3.4  信息量</h3>
<p>一个潜在而有用的衡量进化保守性的测度是序列的信息量。</p>
<p>一个有用的针对分子生物学家的信息论的介绍可以在这里找到： <a class="reference external" href="http://www.lecb.ncifcrf.gov/~toms/paper/primer/">http://www.lecb.ncifcrf.gov/~toms/paper/primer/</a> 。对于我们的目地，我们将查看保守序列或其部分序列的信息量。我们使用下面的公式计算多序列比对中某个特定的列的信息量：</p>
<div class="formula">
<i>IC</i><sub><i>j</i></sub> = <span class="limits"><sup class="limit"><i>N</i><sub><i>a</i></sub></sup><span class="limit"><span class="symbol">∑</span></span><sub class="limit"><i>i</i> = 1</sub></span><i>P</i><sub><i>ij</i></sub><span class="mathrm">log</span><span class="array"><span class="arrayrow"><span class="bracket align-left">⎛</span></span><span class="arrayrow"><span class="bracket align-left">⎝</span></span></span><span class="fraction"><span class="ignored">(</span><span class="numerator"><i>P</i><sub><i>ij</i></sub></span><span class="ignored">)/(</span><span class="denominator"><i>Q</i><sub><i>i</i></sub></span><span class="ignored">)</span></span><span class="array"><span class="arrayrow"><span class="bracket align-right">⎞</span></span><span class="arrayrow"><span class="bracket align-right">⎠</span></span></span>
</div>
<p>其中：</p>
<ul class="simple">
<li><em>IC</em><sub>*j*</sub> – 比对中第 <em>j</em> 列的信息量。</li>
<li><em>N</em><sub>*a*</sub> – 字母表中字母的个数。</li>
<li><em>P</em><sub>*ij*</sub> – 第 <em>j</em> 列的某个特定字母 <em>i</em> 的频率（即，如果G在包含有6个序列的比对中有3次出现，则该列G的信息量为0.5）</li>
<li><em>Q</em><sub>*i*</sub> – 字母 <em>i</em> 的期望频率。这是一个可选参数，由用户自行决定使用。默认情况下，它被自动赋值为0.05 = 1/20，若为蛋白字母表；或0.25 = 1/4 ，若为核酸字母表。这是在没有先验分布假设的情况下计算信息量。而在假设先验分布或使用非标准字母表时，你需要提供 <em>Q</em><sub>*i*</sub> 的值。</li>
</ul>
<p>好了，现在我们知道Biopython如何计算了序列比对的信息量，让我们看看怎么对部分比对区域进行计算。</p>
<p>首先，我们需要使用我们的比对来获得一个比对摘要对象，我们假设它叫做 <tt class="docutils literal">summary_align</tt> （参见章节<a class="reference external" href="#sec:summary_info">18.3.1</a> 来了解怎样得到它）。一旦我们得到这个对象，计算某个区域的信息量就像下面一样简单：</p>
<pre class="code python literal-block">
<span class="name">info_content</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">information_content</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="literal number integer">30</span><span class="punctuation">,</span>
                                                 <span class="name">chars_to_ignore</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">'N'</span><span class="punctuation">])</span>
</pre>
<p>哇哦，这比上面的公式看起来要简单多了！变量 <tt class="docutils literal">info_content</tt> 现在含有一个浮点数来表示指定区域（比对中的5到30）的信息量。我们在计算信息量时特意忽略了不确定残基’N’，因为这个值没有包括在我们的字母表中（因而我们不必要关心它！）。</p>
<p>像上面提到的一样，我们同样能通过提供期望频率计算相对信息量：</p>
<pre class="code python literal-block">
<span class="name">expect_freq</span> <span class="operator">=</span> <span class="punctuation">{</span>
    <span class="literal string">'A'</span> <span class="punctuation">:</span> <span class="operator">.</span><span class="literal number integer">3</span><span class="punctuation">,</span>
    <span class="literal string">'G'</span> <span class="punctuation">:</span> <span class="operator">.</span><span class="literal number integer">2</span><span class="punctuation">,</span>
    <span class="literal string">'T'</span> <span class="punctuation">:</span> <span class="operator">.</span><span class="literal number integer">3</span><span class="punctuation">,</span>
    <span class="literal string">'C'</span> <span class="punctuation">:</span> <span class="operator">.</span><span class="literal number integer">2</span><span class="punctuation">}</span>
</pre>
<p>期望值不能以原始的字典传入，而需要作为 <tt class="docutils literal">SubsMat.FreqTable</tt> 对象传入（参见章节<a class="reference external" href="#sec:freq_table">20.2.2</a> 以获得关于FreqTables的更多信息）。FreqTable对象提供了一个关联字典和字母表的标准，这和Biopython中Seq类的工作方式类似。</p>
<p>要从频率字典创建一个FreqTable对象，你只需要：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio.SubsMat</span> <span class="keyword namespace">import</span> <span class="name">FreqTable</span>

<span class="name">e_freq_table</span> <span class="operator">=</span> <span class="name">FreqTable</span><span class="operator">.</span><span class="name">FreqTable</span><span class="punctuation">(</span><span class="name">expect_freq</span><span class="punctuation">,</span> <span class="name">FreqTable</span><span class="operator">.</span><span class="name">FREQ</span><span class="punctuation">,</span>
                                   <span class="name">IUPAC</span><span class="operator">.</span><span class="name">unambiguous_dna</span><span class="punctuation">)</span>
</pre>
<p>现在我们得到了它，计算我们比对区域的相对信息量就像下面一样简单：</p>
<pre class="code python literal-block">
<span class="name">info_content</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">information_content</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="literal number integer">30</span><span class="punctuation">,</span>
                                                 <span class="name">e_freq_table</span> <span class="operator">=</span> <span class="name">e_freq_table</span><span class="punctuation">,</span>
                                                 <span class="name">chars_to_ignore</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">'N'</span><span class="punctuation">])</span>
</pre>
<p>现在，<tt class="docutils literal">info_content</tt> 将包含与期望频率相关的该区域的相对信息量。</p>
<p>返回值是按上面的公式以2为对底数计算的。你可以通过传入 <tt class="docutils literal">log_base</tt> 参数来改变成你想要的底数：</p>
<pre class="code python literal-block">
<span class="name">info_content</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">information_content</span><span class="punctuation">(</span><span class="literal number integer">5</span><span class="punctuation">,</span> <span class="literal number integer">30</span><span class="punctuation">,</span> <span class="name">log_base</span> <span class="operator">=</span> <span class="literal number integer">10</span><span class="punctuation">,</span>
                                                 <span class="name">chars_to_ignore</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">'N'</span><span class="punctuation">])</span>
</pre>
<p>好了，现在你已经知道怎么计算信息量了。如果你想要在实际的生命科学问题中应用它，最好找一些关于
信息量的文献钻研，以了解它是怎样用的。希望你的钻研不会发现任何有关这个函数的编码错误。</p>
</div>
</div>
<div class="section" id="id239">
<h2>18.4  替换矩阵</h2>
<p>替换矩阵是每天的生物信息学工作中的极端重要的一部分。它们提供决定两个不同的残基有多少相互替换的可能性的得分规则。这在序列比较中必不可少。Durbin等的“Biological Biological Sequence Analysis” 一书中提供了对替换矩阵以及它们的用法的非常好的介绍。一些非常有名的替换矩阵是PAM和BLOSUM系列矩阵。</p>
<p>Biopython提供了大量的常见替换矩阵，也提供了创建你自己的替换矩阵的功能。</p>
<div class="section" id="id240">
<h3>18.4.1  使用常见替换矩阵</h3>
</div>
<div class="section" id="id241">
<h3>18.4.2  从序列比对创建你自己的替换矩阵</h3>
<p>使用替换矩阵类能轻易做出的一个非常酷的事情，是从序列比对创建出你自己的替换矩阵。实际中，通常是使用蛋白比对来做。在这个例子中，我们将首先得到一个Biopython比对对象，然后得到一个摘要对象来计算关于这个比对的相关信息。文件 <a class="reference external" href="examples/protein.aln">protein.aln</a> （也可在 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/protein.aln">这里</a> 获取）包含Clustalw格式的比对输出。</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">AlignIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Alphabet</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Alphabet</span> <span class="keyword namespace">import</span> <span class="name">IUPAC</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio.Align</span> <span class="keyword namespace">import</span> <span class="name">AlignInfo</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">filename</span> <span class="operator">=</span> <span class="literal string">&quot;protein.aln&quot;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">alpha</span> <span class="operator">=</span> <span class="name">Alphabet</span><span class="operator">.</span><span class="name">Gapped</span><span class="punctuation">(</span><span class="name">IUPAC</span><span class="operator">.</span><span class="name">protein</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">c_align</span> <span class="operator">=</span> <span class="name">AlignIO</span><span class="operator">.</span><span class="name">read</span><span class="punctuation">(</span><span class="name">filename</span><span class="punctuation">,</span> <span class="literal string">&quot;clustal&quot;</span><span class="punctuation">,</span> <span class="name">alphabet</span><span class="operator">=</span><span class="name">alpha</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">summary_align</span> <span class="operator">=</span> <span class="name">AlignInfo</span><span class="operator">.</span><span class="name">SummaryInfo</span><span class="punctuation">(</span><span class="name">c_align</span><span class="punctuation">)</span>
</pre>
<p>章节 <a class="reference external" href="#sec:align_clustal">6.4.1</a> 和 <a class="reference external" href="#sec:summary_info">18.3.1</a> 包含关于此类做法的更多信息。</p>
<p>现在我们得到了我们的 <tt class="docutils literal">summary_align</tt> 对象，我们想使用它来找出不同的残基相互替换的次数。为了使例子的可读性更强，我们将只关注那些有极性电荷侧链的氨基酸。幸运的是，这能在生成替代字典时轻松实现，通过传入所有需要被忽略的字符。这样我们将能创建一个只包含带电荷的极性氨基酸的替代字典：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">replace_info</span> <span class="operator">=</span> <span class="name">summary_align</span><span class="operator">.</span><span class="name">replacement_dictionary</span><span class="punctuation">([</span><span class="literal string">&quot;G&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;A&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;V&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;L&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;I&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                                      <span class="literal string">&quot;M&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;P&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;F&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;W&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;S&quot;</span><span class="punctuation">,</span>
<span class="operator">...</span>                                                      <span class="literal string">&quot;T&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;N&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;Q&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;Y&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;C&quot;</span><span class="punctuation">])</span>
</pre>
<p>这个关于氨基酸替代的信息以python字典的形式展示出来将会像如下的样子（顺序可能有所差异）：</p>
<pre class="code python literal-block">
<span class="punctuation">{(</span><span class="literal string">'R'</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">):</span> <span class="literal number float">2079.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'R'</span><span class="punctuation">,</span> <span class="literal string">'H'</span><span class="punctuation">):</span> <span class="literal number float">17.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'R'</span><span class="punctuation">,</span> <span class="literal string">'K'</span><span class="punctuation">):</span> <span class="literal number float">103.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'R'</span><span class="punctuation">,</span> <span class="literal string">'E'</span><span class="punctuation">):</span> <span class="literal number float">2.0</span><span class="punctuation">,</span>
<span class="punctuation">(</span><span class="literal string">'R'</span><span class="punctuation">,</span> <span class="literal string">'D'</span><span class="punctuation">):</span> <span class="literal number float">2.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'H'</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">):</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'D'</span><span class="punctuation">,</span> <span class="literal string">'H'</span><span class="punctuation">):</span> <span class="literal number float">15.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'K'</span><span class="punctuation">,</span> <span class="literal string">'K'</span><span class="punctuation">):</span> <span class="literal number float">3218.0</span><span class="punctuation">,</span>
<span class="punctuation">(</span><span class="literal string">'K'</span><span class="punctuation">,</span> <span class="literal string">'H'</span><span class="punctuation">):</span> <span class="literal number float">24.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'H'</span><span class="punctuation">,</span> <span class="literal string">'K'</span><span class="punctuation">):</span> <span class="literal number float">8.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'E'</span><span class="punctuation">,</span> <span class="literal string">'H'</span><span class="punctuation">):</span> <span class="literal number float">15.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'H'</span><span class="punctuation">,</span> <span class="literal string">'H'</span><span class="punctuation">):</span> <span class="literal number float">1235.0</span><span class="punctuation">,</span>
<span class="punctuation">(</span><span class="literal string">'H'</span><span class="punctuation">,</span> <span class="literal string">'E'</span><span class="punctuation">):</span> <span class="literal number float">18.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'H'</span><span class="punctuation">,</span> <span class="literal string">'D'</span><span class="punctuation">):</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'K'</span><span class="punctuation">,</span> <span class="literal string">'D'</span><span class="punctuation">):</span> <span class="literal number integer">0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'K'</span><span class="punctuation">,</span> <span class="literal string">'E'</span><span class="punctuation">):</span> <span class="literal number float">9.0</span><span class="punctuation">,</span>
<span class="punctuation">(</span><span class="literal string">'D'</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">):</span> <span class="literal number float">48.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'E'</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">):</span> <span class="literal number float">2.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'D'</span><span class="punctuation">,</span> <span class="literal string">'K'</span><span class="punctuation">):</span> <span class="literal number float">1.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'E'</span><span class="punctuation">,</span> <span class="literal string">'K'</span><span class="punctuation">):</span> <span class="literal number float">45.0</span><span class="punctuation">,</span>
<span class="punctuation">(</span><span class="literal string">'K'</span><span class="punctuation">,</span> <span class="literal string">'R'</span><span class="punctuation">):</span> <span class="literal number float">130.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'E'</span><span class="punctuation">,</span> <span class="literal string">'D'</span><span class="punctuation">):</span> <span class="literal number float">241.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'E'</span><span class="punctuation">,</span> <span class="literal string">'E'</span><span class="punctuation">):</span> <span class="literal number float">3305.0</span><span class="punctuation">,</span>
<span class="punctuation">(</span><span class="literal string">'D'</span><span class="punctuation">,</span> <span class="literal string">'E'</span><span class="punctuation">):</span> <span class="literal number float">270.0</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'D'</span><span class="punctuation">,</span> <span class="literal string">'D'</span><span class="punctuation">):</span> <span class="literal number float">2360.0</span><span class="punctuation">}</span>
</pre>
<p>这个信息提供了我们所需要的替换次数，或者说我们期望的不同的事情相互替换有多么频繁。事实上，（你可能会感到惊奇）这就是我们继续创建替代矩阵所需要的全部信息。首先，我们使用替代字典信息创建一个“接受替换矩阵”（Accepted Replacement Matrix，ARM）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SubsMat</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_arm</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">SeqMat</span><span class="punctuation">(</span><span class="name">replace_info</span><span class="punctuation">)</span>
</pre>
<p>使用这个“接受替换矩阵”，我们能继续创建我们的对数矩阵（即一个标准类型的替换举证）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_lom</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">make_log_odds_matrix</span><span class="punctuation">(</span><span class="name">my_arm</span><span class="punctuation">)</span>
</pre>
<p>在创建的这个对数矩阵时有以下可选参数：</p>
<ul class="simple">
<li><tt class="docutils literal">exp_freq_table</tt> – 你可以传入一个每个字母的期望频率的表格。如果提供，在计算期望替换时，这将替代传入的“接收替换矩阵”。</li>
<li><tt class="docutils literal">logbase</tt> - 用来创建对数奇数矩阵的对数底数。默认为10。</li>
<li><tt class="docutils literal">factor</tt> - 用来乘以每个矩阵元素的因数。默认为10，这样通常可以使得矩阵的数据容易处理。</li>
<li><tt class="docutils literal">round_digit</tt> - 矩阵中四舍五入所取的小数位数，默认为0（即没有小数）。</li>
</ul>
<p>一旦你获得了你的对数矩阵，你可以使用函数 <tt class="docutils literal">print_mat</tt> 很漂亮的显示出来。使用我们创建的矩阵可以得到：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_lom</span><span class="operator">.</span><span class="name">print_mat</span><span class="punctuation">()</span>
<span class="name">D</span>   <span class="literal number integer">2</span>
<span class="name">E</span>  <span class="operator">-</span><span class="literal number integer">1</span>   <span class="literal number integer">1</span>
<span class="name">H</span>  <span class="operator">-</span><span class="literal number integer">5</span>  <span class="operator">-</span><span class="literal number integer">4</span>   <span class="literal number integer">3</span>
<span class="name">K</span> <span class="operator">-</span><span class="literal number integer">10</span>  <span class="operator">-</span><span class="literal number integer">5</span>  <span class="operator">-</span><span class="literal number integer">4</span>   <span class="literal number integer">1</span>
<span class="name">R</span>  <span class="operator">-</span><span class="literal number integer">4</span>  <span class="operator">-</span><span class="literal number integer">8</span>  <span class="operator">-</span><span class="literal number integer">4</span>  <span class="operator">-</span><span class="literal number integer">2</span>   <span class="literal number integer">2</span>
   <span class="name">D</span>   <span class="name">E</span>   <span class="name">H</span>   <span class="name">K</span>   <span class="name">R</span>
</pre>
<p>很好。我们现在得到了自己的替换矩阵！</p>
</div>
</div>
<div class="section" id="biosql">
<h2>18.5  BioSQL – 存储序列到关系数据库中</h2>
<p><a class="reference external" href="http://www.biosql.org/">BioSQL</a> 是 <a class="reference external" href="http://open-bio.org/">OBF</a> 多个项目（BioPerl、 BioJava等）为了支持共享的存储序列数据的数据库架构而共同努力的结果。理论上，你可以用BioPerl加载GenBank文件到数据库中，然后用Biopython从数据库中提取出来为一个包含Feature的Record对象 —— 并获得或多或少和直接用 <tt class="docutils literal">Bio.SeqIO</tt> （第 <a class="reference external" href="#chapter:Bio.SeqIO">5</a> 章）加载GenBank文件为SeqRecord相同的东西。</p>
<p>Biopython中BioSQL模块的文档目前放在 <a class="reference external" href="http://biopython.org/wiki/BioSQL">http://biopython.org/wiki/BioSQL</a> ，是我们维基页面的一部分。</p>
</div>
</div>
<div class="section" id="id242">
<h1>第19章 Biopython测试框架</h1>
<p>Biopython具有一个基于Python标准单元测试框架 <cite>unittest&lt;http://docs.python.org/library/unittest.html&gt;</cite>
的回归测试框架（文件 <tt class="docutils literal">run_tests.py</tt>）。而为模块提供全面测试，
是确保Biopython代码在使用期内尽可能无bug的一个最重要的方面。
也经常是最被轻视的方面之一。本章旨在使运行Biopython测试和编
写测试代码尽可能容易。理想情况下，进入Biopython的每个模块都
应该有一个测试（还应该有对应文档！）。强烈鼓励我们所有开发
者，以及任何从源码安装Biopython的人运行单元测试。</p>
<div class="section" id="id243">
<h2>19.1  运行测试</h2>
<p>在你下载Biopython源码或者从我们的源码仓库签出时，你会发现一
个子目录调用 <tt class="docutils literal">Tests</tt>。 这包括关键脚本 <tt class="docutils literal">run_tests.py</tt>、
名为 <tt class="docutils literal">test_XXX.py</tt> 的很多独立脚本、一个叫 <tt class="docutils literal">output</tt> 的子目录和
很多其他包含测试套件输入文件的子目录。</p>
<p>作为构建和安装Biopython的一部分，你通常会在命令行上从Biopython
源码顶层目录运行整个测试套件如下：</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">setup</span><span class="operator">.</span><span class="name">py</span> <span class="name">test</span>
</pre>
<p>这事实上等价于转到 <tt class="docutils literal">Tests</tt> 子目录，并运行：</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">run_tests</span><span class="operator">.</span><span class="name">py</span>
</pre>
<p>你通常会想要只运行测试的一部分，这可以如下来操作：</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">run_tests</span><span class="operator">.</span><span class="name">py</span> <span class="name">test_SeqIO</span><span class="operator">.</span><span class="name">py</span> <span class="name">test_AlignIO</span><span class="operator">.</span><span class="name">py</span>
</pre>
<p>当给出测试列表时， <tt class="docutils literal">.py</tt> 扩展名是可选的，所以你可以只需打字：</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">run_tests</span><span class="operator">.</span><span class="name">py</span> <span class="name">test_SeqIO</span> <span class="name">test_AlignIO</span>
</pre>
<p>要运行 docstring 测试（见 <a class="reference external" href="#section:doctest">19.3</a> 节）的话，
你可以用</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">run_tests</span><span class="operator">.</span><span class="name">py</span> <span class="name">doctest</span>
</pre>
<p>缺省情况下， <tt class="docutils literal">run_tests.py</tt> 运行所有测试，包括docstring测试。</p>
<p>如果一单个测试失败了，你还可以尝试直接运行它，它会给出更多信息。</p>
<p>重要的是，要注意单个单元测试有两类作用：</p>
<ul class="simple">
<li>简单打印和比较脚本。 这些单元测试本质上是简短的 Python 示例
程序，它们会打印出各种输出文本。对于一个名为 <tt class="docutils literal">test_XXX.py</tt>
的测试文件，在 <tt class="docutils literal">output</tt> 子目录（包含期望的输出）下会有一个
叫做 <tt class="docutils literal">test_XXX</tt> 的匹配文本文件。测试框架所做的全部就是运行
脚本并检查输出的一致性。</li>
<li>基于 <tt class="docutils literal">unittest</tt> 的标准测试。 这些会 <tt class="docutils literal">import unittest</tt> ，然
后定义 <tt class="docutils literal">unittest.TestCase</tt> 类，这些类的每一个都带有一或多个
像以 <tt class="docutils literal">test_</tt> 开头、检查代码的某些方面的方法那样的子测试。这
些测试不应该直接打印任何输出。</li>
</ul>
<p>目前，大约一半的 Biopython 测试是 <tt class="docutils literal">unittest</tt> 风格的测试，另一半
是 print-and-compare 测试。</p>
<p>直接运行一个简单的 print-and-compare 测试通常会在屏幕上给出大量
输出，但是并不检查输出是否跟期望的一样。如果测试以一个意外的错误
而失败，那么应该很容易精确定位脚本失败的位置。例如，对于一个print-and-compare 测试，试一下：</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">test_SeqIO</span><span class="operator">.</span><span class="name">py</span>
</pre>
<p>基于 <tt class="docutils literal">unittest</tt> 的测试反倒是精确地显示你测试的哪一个小块失败了。
例如，</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">test_Cluster</span><span class="operator">.</span><span class="name">py</span>
</pre>
</div>
<div class="section" id="id244">
<h2>19.2  编写测试</h2>
<p>假如说你想为一个叫做 <tt class="docutils literal">Biospam</tt> 的模块写一些测试。这可以是你写的
一个模块，或者是一个还没有任何测试的现存模块。在下面的例子中，我们
假设 <tt class="docutils literal">Biospam</tt> 是一个做简单数学的模块。</p>
<p>每个 Biopython 测试都可以有三个重要的文件和相关目录：</p>
<ol class="arabic simple">
<li><tt class="docutils literal">test_Biospam.py</tt> – 关于你的模块的真正测试代码。</li>
<li><tt class="docutils literal">Biospam</tt> [optional]– 一个包含任何必要输入文件的目录。任何会
生成的输出文件也应该写在这里（并且最好在测试结束后打扫干净)以防
堵塞主 Tests 目录。</li>
<li><tt class="docutils literal">output/Biospam</tt> – [只针对 print-and-compare 测试] 这个文件
包括运行 <tt class="docutils literal">test_Biospam.py</tt> 的期望输出。这个文件对于 <tt class="docutils literal">unittest</tt>
风格的测试不是必须的，因为测试脚本 <tt class="docutils literal">test_Biospam.py</tt> 会自己做验证。</li>
</ol>
<p>你要自己决定你是想编写一个 print-and-compare 测试脚本还是一个 <tt class="docutils literal">unittest</tt>
风格的测试脚本。重要的是你不能把这两种风格混合在一个
测试脚本中。尤其是，不要在一个 print-and-compare 测试中使用``unittest``
特征。</p>
<p><tt class="docutils literal">Tests</tt> 目录中任何具有 <tt class="docutils literal">test_</tt> 前缀的脚本都会被 <tt class="docutils literal">run_tests.py</tt>
找到并运行。下面，我们展示一个示例测试脚本 <tt class="docutils literal">test_Biospam.py</tt> ，针对
一个 print-and-compare 测试和一个基于 <tt class="docutils literal">unittest</tt> 的测试。如果你把这个
脚本放进 Biopython的 <tt class="docutils literal">Tests</tt> 目录，那么 <tt class="docutils literal">run_tests.py</tt> 就会找到它并
执行其中包含的测试：</p>
<pre class="code python literal-block">
<span class="error">$</span> <span class="name">python</span> <span class="name">run_tests</span><span class="operator">.</span><span class="name">py</span>
<span class="name">test_Ace</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_AlignIO</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_BioSQL</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_BioSQL_SeqIO</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_Biospam</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_CAPS</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_Clustalw</span> <span class="operator">...</span> <span class="name">ok</span>
</pre>
<p>…</p>
<pre class="code python literal-block">
<span class="operator">----------------------------------------------------------------------</span>
<span class="name">Ran</span> <span class="literal number integer">107</span> <span class="name">tests</span> <span class="operator word">in</span> <span class="literal number float">86.127</span> <span class="name">seconds</span>
</pre>
<div class="section" id="print-and-compare">
<h3>19.2.1  编写一个 print-and-compare 测试</h3>
<p>一个 print-and-compare 风格的测试对于初学者和新手来说是很容易写的- 本质上它只是一个使用你的模块的示例脚本。</p>
<p>为了写一个关于 <tt class="docutils literal">Biospam</tt> 的 print-and-compare 测试，这是你应该
做的：</p>
<ol class="arabic">
<li><p class="first">编写一个叫 <tt class="docutils literal">test_Biospam.py</tt> 的脚本</p>
<ul class="simple">
<li>这个脚本应该位于 Tests 目录</li>
<li>脚本应该测试模块的所有重要功能（当然，你测试的越多、你的测试就
越好！）。</li>
<li>尽量避免任何平台特异的东西，例如打印浮点数而不用显式格式字符串
来避免有太多小数位（不同的平台会给出稍微不同的值）。</li>
</ul>
</li>
<li><p class="first">如果脚本需要文件来进行测试，这些应转到目录 Tests/Biospam 中进行
（如果你只需一些通用的东西，像一个 FASTA 序列文件，或者一条
GenBank 记录，试着用一个现存的样品输入文件来代替）。</p>
</li>
<li><p class="first">写出测试输出并验证输出是正确的。</p>
<p>有两种方法可以做到这一点：</p>
<ol class="arabic simple">
<li>长期方法：<ul>
<li>运行脚本并将输出写到一个文件中。在 UNIX （包括 Linux 和 Mac OS X
）机器上，你可以这样做： <tt class="docutils literal">python test_Biospam.py &gt; test_Biospam</tt>
这会把输出写到文件 <tt class="docutils literal">test_Biospam</tt> 中。</li>
<li>手动查看文件 <tt class="docutils literal">test_Biospam</tt> 来确保输出正确。当你确定都没问
题、没有bug后，你需要快速编辑 <tt class="docutils literal">test_Biospam</tt> 文件使其首行为：
‘<tt class="docutils literal">test_Biospam</tt>’  （不含引号）。</li>
<li>复制文件 <tt class="docutils literal">test_Biospam</tt> 到目录 Tests/output 中。</li>
</ul>
</li>
<li>快速方法:<ul>
<li>运行 <tt class="docutils literal">python run_tests.py <span class="pre">-g</span> test_Biospam.py</tt> 。回归测试框架
很聪明的会以他喜欢的方式把输出放在恰当的地方。</li>
<li>转到输出（应该在 <tt class="docutils literal">Tests/output/test_Biospam</tt>）并复查输出以确
保其完全正确。</li>
</ul>
</li>
</ol>
</li>
<li><p class="first">现在改换到 Tests 目录并运行 <tt class="docutils literal">python run_tests.py</tt> 进行回归测试。
这会运行所有测试，而你会看到你的测试也在运行（并通过）。</p>
</li>
<li><p class="first">好了！这样你就得到了可用于签入或提交到Biopython的、关于你的模块的
一个友好的测试。恭喜你！</p>
</li>
</ol>
<p>例如，测试 <tt class="docutils literal">Biospam</tt> 模块中的 <tt class="docutils literal">addition</tt> 和 <tt class="docutils literal">multiplication</tt> 功
能的测试脚本 <tt class="docutils literal">test_Biospam.py</tt> 也许看起来是下面这个样子：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Biospam</span>

<span class="keyword">print</span> <span class="literal string">&quot;2 + 3 =&quot;</span><span class="punctuation">,</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">addition</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;9 - 1 =&quot;</span><span class="punctuation">,</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">addition</span><span class="punctuation">(</span><span class="literal number integer">9</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;2 * 3 =&quot;</span><span class="punctuation">,</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">multiplication</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">)</span>
<span class="keyword">print</span> <span class="literal string">&quot;9 * (- 1) =&quot;</span><span class="punctuation">,</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">multiplication</span><span class="punctuation">(</span><span class="literal number integer">9</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">)</span>
</pre>
<p>我们用 <tt class="docutils literal">python run_tests.py <span class="pre">-g</span> test_Biospam.py</tt> 来生成对应的输出，
并检查输出文件 <tt class="docutils literal">output/test_Biospam</tt> ：</p>
<pre class="code python literal-block">
<span class="name">test_Biospam</span>
<span class="literal number integer">2</span> <span class="operator">+</span> <span class="literal number integer">3</span> <span class="operator">=</span> <span class="literal number integer">5</span>
<span class="literal number integer">9</span> <span class="operator">-</span> <span class="literal number integer">1</span> <span class="operator">=</span> <span class="literal number integer">8</span>
<span class="literal number integer">2</span> <span class="operator">*</span> <span class="literal number integer">3</span> <span class="operator">=</span> <span class="literal number integer">6</span>
<span class="literal number integer">9</span> <span class="operator">*</span> <span class="punctuation">(</span><span class="operator">-</span> <span class="literal number integer">1</span><span class="punctuation">)</span> <span class="operator">=</span> <span class="operator">-</span><span class="literal number integer">9</span>
</pre>
<p>通常，更大的 print-and-compare 测试的困难在于追踪输出行与测试脚本
命令之间的对应关系。为此，打印出一些标记是很重要的，这些标记帮助你
把输入脚本按行和产生的输出匹配起来。</p>
</div>
<div class="section" id="unittest">
<h3>19.2.2  编写一个基于 unittest 的测试</h3>
<p>我们想要Biopython中的所有模块都具有单元测试，并且一个简单的
print-and-compare 测试比一点儿测试都没有要好。不过，尽管有一个陡峭的
学习曲线，使用 <tt class="docutils literal">unittest</tt> 框架能给出一个更结构化的结果，并且如果有
一个测试失败，这能够清晰准确地指出测试的哪部分出了问题。子测试也可以
单独运行，这对于测试和排错很有帮助。</p>
<p>从2.1版开始 <tt class="docutils literal">unittest</tt> 框架就包含在Python中了，并且存档在
Python Library Reference （就是所推荐的你的枕边书）。也有 <a class="reference external" href="http://docs.python.org/library/unittest.html">关于unittest
的在线文档</a>。如果你
熟悉 <tt class="docutils literal">unittest</tt> 系统（或类似于某些噪音测试框架的东西），你应该不会有
什么麻烦。你也许发现，寻找Biopython中的现成例子很有帮助。</p>
<p>这是关于 <tt class="docutils literal">Biospam</tt> 的一个 <tt class="docutils literal">unittest</tt> 风格的极小测试脚本，你可以
复制粘贴过去运行它：</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">unittest</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Biospam</span>

<span class="keyword">class</span> <span class="name class">BiospamTestAddition</span><span class="punctuation">(</span><span class="name">unittest</span><span class="operator">.</span><span class="name">TestCase</span><span class="punctuation">):</span>

    <span class="keyword">def</span> <span class="name function">test_addition1</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">addition</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="literal number integer">5</span><span class="punctuation">)</span>

    <span class="keyword">def</span> <span class="name function">test_addition2</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">addition</span><span class="punctuation">(</span><span class="literal number integer">9</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="literal number integer">8</span><span class="punctuation">)</span>

<span class="keyword">class</span> <span class="name class">BiospamTestDivision</span><span class="punctuation">(</span><span class="name">unittest</span><span class="operator">.</span><span class="name">TestCase</span><span class="punctuation">):</span>

    <span class="keyword">def</span> <span class="name function">test_division1</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">division</span><span class="punctuation">(</span><span class="literal number float">3.0</span><span class="punctuation">,</span> <span class="literal number float">2.0</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertAlmostEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="literal number float">1.5</span><span class="punctuation">)</span>

    <span class="keyword">def</span> <span class="name function">test_division2</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">division</span><span class="punctuation">(</span><span class="literal number float">10.0</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">2.0</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertAlmostEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">5.0</span><span class="punctuation">)</span>


<span class="keyword">if</span> <span class="name">__name__</span> <span class="operator">==</span> <span class="literal string">&quot;__main__&quot;</span><span class="punctuation">:</span>
    <span class="name">runner</span> <span class="operator">=</span> <span class="name">unittest</span><span class="operator">.</span><span class="name">TextTestRunner</span><span class="punctuation">(</span><span class="name">verbosity</span> <span class="operator">=</span> <span class="literal number integer">2</span><span class="punctuation">)</span>
    <span class="name">unittest</span><span class="operator">.</span><span class="name">main</span><span class="punctuation">(</span><span class="name">testRunner</span><span class="operator">=</span><span class="name">runner</span><span class="punctuation">)</span>
</pre>
<p>在分割测试中，我们使用 <tt class="docutils literal">assertAlmostEqual</tt> 而不是 <tt class="docutils literal">assertEqual</tt>
以免因舍入误差造成的测试失败；详情以及 <tt class="docutils literal">unittest</tt> 中的其他可用功能
参见Python文档中的 <tt class="docutils literal">unittest</tt> 章节（<a class="reference external" href="http://docs.python.org/library/unittest.html">在线参考</a>）。</p>
<p>这里是基于 <tt class="docutils literal">unittest</tt> 的测试的一些关键点：</p>
<ul>
<li><p class="first">测试实例存储在 <tt class="docutils literal">unittest.TestCase</tt> 的子类中并涵盖了你的代码
的一个基本方面。</p>
</li>
<li><p class="first">对于任何在每个测试方法前后都要运行的重复代码，你可以使用方法
<tt class="docutils literal">setUp</tt> 和 <tt class="docutils literal">tearDown</tt> 。例如 <tt class="docutils literal">setUp</tt> 方法可用于创建你正在
测试的对象的实例，或打开一个文件句柄。 <tt class="docutils literal">tearDown</tt> 可做任何整理，
例如关闭文件句柄。</p>
</li>
<li><p class="first">测试以 <tt class="docutils literal">test_</tt> 为前缀并且每项测试应覆盖你所想要测试的内容的一个
具体部分。一个类中你想包含多少个测试都行。</p>
</li>
<li><p class="first">在测试脚本的末尾，你可以用</p>
<pre class="code python literal-block">
<span class="keyword">if</span> <span class="name">__name__</span> <span class="operator">==</span> <span class="literal string">&quot;__main__&quot;</span><span class="punctuation">:</span>
    <span class="name">runner</span> <span class="operator">=</span> <span class="name">unittest</span><span class="operator">.</span><span class="name">TextTestRunner</span><span class="punctuation">(</span><span class="name">verbosity</span> <span class="operator">=</span> <span class="literal number integer">2</span><span class="punctuation">)</span>
    <span class="name">unittest</span><span class="operator">.</span><span class="name">main</span><span class="punctuation">(</span><span class="name">testRunner</span><span class="operator">=</span><span class="name">runner</span><span class="punctuation">)</span>
</pre>
<p>来执行测试脚本，当脚本是从        自己运行（而不是从 <tt class="docutils literal">run_tests.py</tt> 导入）时。
如果你运行该脚本，那么你会见到类似下面的东西:</p>
<pre class="code python literal-block">
<span class="error">$</span> <span class="name">python</span> <span class="name">test_BiospamMyModule</span><span class="operator">.</span><span class="name">py</span>
<span class="name">test_addition1</span> <span class="punctuation">(</span><span class="name">__main__</span><span class="operator">.</span><span class="name">TestAddition</span><span class="punctuation">)</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_addition2</span> <span class="punctuation">(</span><span class="name">__main__</span><span class="operator">.</span><span class="name">TestAddition</span><span class="punctuation">)</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_division1</span> <span class="punctuation">(</span><span class="name">__main__</span><span class="operator">.</span><span class="name">TestDivision</span><span class="punctuation">)</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">test_division2</span> <span class="punctuation">(</span><span class="name">__main__</span><span class="operator">.</span><span class="name">TestDivision</span><span class="punctuation">)</span> <span class="operator">...</span> <span class="name">ok</span>

<span class="operator">----------------------------------------------------------------------</span>
<span class="name">Ran</span> <span class="literal number integer">4</span> <span class="name">tests</span> <span class="operator word">in</span> <span class="literal number float">0.059</span><span class="name">s</span>

<span class="name">OK</span>
</pre>
</li>
<li><p class="first">为了更清晰地表明每个测试都干了什么，你可以给每个测试加上 docstrings 。
它们会在运行测试的时候显示出来，如果一个测试失败这会是有用的信息。</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">unittest</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">Biospam</span>

<span class="keyword">class</span> <span class="name class">BiospamTestAddition</span><span class="punctuation">(</span><span class="name">unittest</span><span class="operator">.</span><span class="name">TestCase</span><span class="punctuation">):</span>

    <span class="keyword">def</span> <span class="name function">test_addition1</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="literal string doc">&quot;&quot;&quot;An addition test&quot;&quot;&quot;</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">addition</span><span class="punctuation">(</span><span class="literal number integer">2</span><span class="punctuation">,</span> <span class="literal number integer">3</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="literal number integer">5</span><span class="punctuation">)</span>

    <span class="keyword">def</span> <span class="name function">test_addition2</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="literal string doc">&quot;&quot;&quot;A second addition test&quot;&quot;&quot;</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">addition</span><span class="punctuation">(</span><span class="literal number integer">9</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number integer">1</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="literal number integer">8</span><span class="punctuation">)</span>

<span class="keyword">class</span> <span class="name class">BiospamTestDivision</span><span class="punctuation">(</span><span class="name">unittest</span><span class="operator">.</span><span class="name">TestCase</span><span class="punctuation">):</span>

    <span class="keyword">def</span> <span class="name function">test_division1</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="literal string doc">&quot;&quot;&quot;Now let's check division&quot;&quot;&quot;</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">division</span><span class="punctuation">(</span><span class="literal number float">3.0</span><span class="punctuation">,</span> <span class="literal number float">2.0</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertAlmostEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="literal number float">1.5</span><span class="punctuation">)</span>

    <span class="keyword">def</span> <span class="name function">test_division2</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">):</span>
        <span class="literal string doc">&quot;&quot;&quot;A second division test&quot;&quot;&quot;</span>
        <span class="name">result</span> <span class="operator">=</span> <span class="name">Biospam</span><span class="operator">.</span><span class="name">division</span><span class="punctuation">(</span><span class="literal number float">10.0</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">2.0</span><span class="punctuation">)</span>
        <span class="name builtin pseudo">self</span><span class="operator">.</span><span class="name">assertAlmostEqual</span><span class="punctuation">(</span><span class="name">result</span><span class="punctuation">,</span> <span class="operator">-</span><span class="literal number float">5.0</span><span class="punctuation">)</span>


<span class="keyword">if</span> <span class="name">__name__</span> <span class="operator">==</span> <span class="literal string">&quot;__main__&quot;</span><span class="punctuation">:</span>
    <span class="name">runner</span> <span class="operator">=</span> <span class="name">unittest</span><span class="operator">.</span><span class="name">TextTestRunner</span><span class="punctuation">(</span><span class="name">verbosity</span> <span class="operator">=</span> <span class="literal number integer">2</span><span class="punctuation">)</span>
    <span class="name">unittest</span><span class="operator">.</span><span class="name">main</span><span class="punctuation">(</span><span class="name">testRunner</span><span class="operator">=</span><span class="name">runner</span><span class="punctuation">)</span>
</pre>
<p>运行脚本你就会看到：</p>
<pre class="code python literal-block">
<span class="error">$</span> <span class="name">python</span> <span class="name">test_BiospamMyModule</span><span class="operator">.</span><span class="name">py</span>
<span class="name">An</span> <span class="name">addition</span> <span class="name">test</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">A</span> <span class="name">second</span> <span class="name">addition</span> <span class="name">test</span> <span class="operator">...</span> <span class="name">ok</span>
<span class="name">Now</span> <span class="name">let</span><span class="literal string">'s check division ... ok</span>
<span class="name">A</span> <span class="name">second</span> <span class="name">division</span> <span class="name">test</span> <span class="operator">...</span> <span class="name">ok</span>

<span class="operator">----------------------------------------------------------------------</span>
<span class="name">Ran</span> <span class="literal number integer">4</span> <span class="name">tests</span> <span class="operator word">in</span> <span class="literal number float">0.001</span><span class="name">s</span>

<span class="name">OK</span>
</pre>
</li>
</ul>
<p>如果你的模块包含 docstring 测试（见`19.3 &lt;#section:doctest&gt;`__小节），
你也许想在要运行的测试中包含这些。你可以修改 <tt class="docutils literal">if __name__ == &quot;__main__&quot;:</tt>
下面的代码如下面这样：</p>
<pre class="code python literal-block">
<span class="keyword">if</span> <span class="name">__name__</span> <span class="operator">==</span> <span class="literal string">&quot;__main__&quot;</span><span class="punctuation">:</span>
    <span class="name">unittest_suite</span> <span class="operator">=</span> <span class="name">unittest</span><span class="operator">.</span><span class="name">TestLoader</span><span class="punctuation">()</span><span class="operator">.</span><span class="name">loadTestsFromName</span><span class="punctuation">(</span><span class="literal string">&quot;test_Biospam&quot;</span><span class="punctuation">)</span>
    <span class="name">doctest_suite</span> <span class="operator">=</span> <span class="name">doctest</span><span class="operator">.</span><span class="name">DocTestSuite</span><span class="punctuation">(</span><span class="name">Biospam</span><span class="punctuation">)</span>
    <span class="name">suite</span> <span class="operator">=</span> <span class="name">unittest</span><span class="operator">.</span><span class="name">TestSuite</span><span class="punctuation">((</span><span class="name">unittest_suite</span><span class="punctuation">,</span> <span class="name">doctest_suite</span><span class="punctuation">))</span>
    <span class="name">runner</span> <span class="operator">=</span> <span class="name">unittest</span><span class="operator">.</span><span class="name">TextTestRunner</span><span class="punctuation">(</span><span class="name">sys</span><span class="operator">.</span><span class="name">stdout</span><span class="punctuation">,</span> <span class="name">verbosity</span> <span class="operator">=</span> <span class="literal number integer">2</span><span class="punctuation">)</span>
    <span class="name">runner</span><span class="operator">.</span><span class="name">run</span><span class="punctuation">(</span><span class="name">suite</span><span class="punctuation">)</span>
</pre>
<p>这只与你执行 <tt class="docutils literal">python test_Biospam.py</tt> 时是否想要运行 docstring 测试
有关；运行 <tt class="docutils literal">python run_tests.py</tt> ，docstring 测试会自动运行（假设他们
被包含在 <tt class="docutils literal">run_tests.py</tt> 中的 docstring 测试列表中，见下面的小节）。</p>
</div>
</div>
<div class="section" id="doctests">
<h2>19.3  编写 doctests</h2>
<p>Python 模块、类和函数支持使用 docstrings 创建文档。 <a class="reference external" href="http://docs.python.org/library/doctest.html">doctest 框架</a> （包含在Python中）
允许开发者将工作例子嵌入在 docstrings 中，并自动测试这些例子。</p>
<p>目前只有一小部分 Biopython 包含 doctests 。 <tt class="docutils literal">run_tests.py</tt> 脚本
看护着 doctests 的运行。为此， <tt class="docutils literal">run_tests.py</tt> 脚本开头是要测试
的模块的一个手动编译列表，该列表允许我们跳过那些可能没有安装可选
外部依赖库的模块（例如 Reportlab 和 NumPy 库）。所以如果你在 Biopython
模块中加一些针对 dostrings 的 doctests ，为了把它们包含在 Biopython
套件中，你必须更新 <tt class="docutils literal">run_tests.py</tt> 以包含你的模块。现在，
<tt class="docutils literal">run_tests.py</tt> 的相关部分看起来像下面这样：</p>
<pre class="code python literal-block">
<span class="comment"># This is the list of modules containing docstring tests.</span>
<span class="comment"># If you develop docstring tests for other modules, please add</span>
<span class="comment"># those modules here.</span>
<span class="name">DOCTEST_MODULES</span> <span class="operator">=</span> <span class="punctuation">[</span><span class="literal string">&quot;Bio.Seq&quot;</span><span class="punctuation">,</span>
                   <span class="literal string">&quot;Bio.SeqRecord&quot;</span><span class="punctuation">,</span>
                   <span class="literal string">&quot;Bio.SeqIO&quot;</span><span class="punctuation">,</span>
                   <span class="literal string">&quot;...&quot;</span><span class="punctuation">,</span>
                  <span class="punctuation">]</span>
<span class="comment">#Silently ignore any doctests for modules requiring numpy!</span>
<span class="keyword">try</span><span class="punctuation">:</span>
    <span class="keyword namespace">import</span> <span class="name namespace">numpy</span>
    <span class="name">DOCTEST_MODULES</span><span class="operator">.</span><span class="name">extend</span><span class="punctuation">([</span><span class="literal string">&quot;Bio.Statistics.lowess&quot;</span><span class="punctuation">])</span>
<span class="keyword">except</span> <span class="name exception">ImportError</span><span class="punctuation">:</span>
    <span class="keyword">pass</span>
</pre>
<p>注意我们首先把 doctests 看做文档，所以你应该坚持典型用法。通常处理错
误条件等诸如此类的复杂例子最好留给一个专门的单元测试。</p>
<p>注意，如果你想编写涉及文件解析的 doctests ，定义文件位置复杂性是很要
紧的。理想情况下，假设代码会从 <tt class="docutils literal">Tests</tt> 目录运行，使用相对路径即可，
关于这一点的一个例子参见 <tt class="docutils literal">Bio.SeqIO</tt> doctests 。</p>
<p>要想只运行 docstring 测试，使用</p>
<pre class="code python literal-block">
<span class="error">$</span> <span class="name">python</span> <span class="name">run_tests</span><span class="operator">.</span><span class="name">py</span> <span class="name">doctest</span>
</pre>
</div>
</div>
<div class="section" id="id245">
<h1>第20章 高级</h1>
<div class="section" id="id246">
<h2>20.1  解析器的设计</h2>
<p>过去很多Biopython解析器都是根据面向事件设计出来的，包括Scanner和Consumer。</p>
<p>Scanners是将输入的数据源进行逐行分析，只要识别出数据中的信息就会发送一个事件。
例如，如果数据中包含物种名信息，Scanner只要读到某行包含名称信息时就会产生一个 <tt class="docutils literal">organism_name</tt> 事件。</p>
<p>Consumers是用来接收Scanners所发出事件的对象。
接着上面的例子，当Consumer收到了 <tt class="docutils literal">organism_name</tt> 事件，在当前应用程序中无论以何种方式都会运行。</p>
<p>这是一个非常灵活的构架，如果你想要将一个文件解析成多种其他格式的，这将会很有优势。
例如， <tt class="docutils literal">Bio.GenBank</tt> 模块可以运用这种方式构建 <tt class="docutils literal">SeqRecord</tt> 或者其他独特的文件格式记录对象。</p>
<p>最近，很多添加了 <tt class="docutils literal">Bio.SeqIO</tt> 和 <tt class="docutils literal">Bio.AlignIO</tt> 的解析器使用了一种更为简单的方法，
但是只能产生单一形式的文件格式（分别是 <tt class="docutils literal">SeqRecord</tt> and <tt class="docutils literal">MultipleSeqAlignment</tt> ）。
在某些情况，<tt class="docutils literal">Bio.SeqIO</tt> 解析器实际上包含了另一种Biopython解析器 - 例如， <tt class="docutils literal">Bio.SwissProt</tt> 解析器产生了特定的SwissProt格式对象，又转换成了 <tt class="docutils literal">SeqRecord</tt> 格式对象。</p>
</div>
<div class="section" id="id247">
<h2>20.2  替换矩阵</h2>
<div class="section" id="subsmat">
<h3>20.2.1  SubsMat</h3>
<p>这个模块提供了一个类和一些固定的方法来产生替换矩阵，类似于BLOSUM或者PAM矩阵，但是是基于用户提供的数据。
此外，你还可以从已建立的替换矩阵集合MatrixInfo.py中选择一个矩阵。
<tt class="docutils literal">SeqMat</tt> 类来自于一个字典（dictionary）:</p>
<pre class="code python literal-block">
<span class="keyword">class</span> <span class="name class">SeqMat</span><span class="punctuation">(</span><span class="name builtin">dict</span><span class="punctuation">)</span>
</pre>
<p>这个字典的格式是 <tt class="docutils literal"><span class="pre">{(i1,j1):n1,</span> <span class="pre">(i1,j2):n2,...,(ik,jk):nk}</span></tt> ， i和j是字母集，而n是一个值。</p>
<ol class="arabic">
<li><p class="first">属性</p>
<ol class="arabic simple">
<li><tt class="docutils literal">self.alphabet</tt>: Bio.Alphabet中定义的一个类</li>
<li><tt class="docutils literal">self.ab_list</tt>: 排列好的字母列表。主要是内部需求。</li>
</ol>
</li>
<li><p class="first">方法</p>
<ol class="arabic">
<li><pre class="code python first literal-block">
<span class="name">__init__</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">,</span><span class="name">data</span><span class="operator">=</span><span class="name builtin pseudo">None</span><span class="punctuation">,</span><span class="name">alphabet</span><span class="operator">=</span><span class="name builtin pseudo">None</span><span class="punctuation">,</span> <span class="name">mat_name</span><span class="operator">=</span><span class="literal string">''</span><span class="punctuation">,</span> <span class="name">build_later</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">):</span>
</pre>
<ol class="arabic simple">
<li><tt class="docutils literal">data</tt>: 可以是一个字典，也可以是另一个SeqMat实例。</li>
<li><tt class="docutils literal">alphabet</tt>: 一个Bio.Alphabet的实例。如果没有提供，将从数据构建一个alphabet。</li>
<li><tt class="docutils literal">mat_name</tt>: 矩阵名，例如  <tt class="docutils literal">BLOSUM62</tt> 或者 <tt class="docutils literal">PAM250</tt></li>
<li><tt class="docutils literal">build_later</tt>: 默认值为false。如果为true，用户应该只提供alphabet和空字典。如果想要之后再构建矩阵，这样会跳过alphabet大小和矩阵大小的检查。</li>
</ol>
</li>
<li><pre class="code python first literal-block">
<span class="name">entropy</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">,</span><span class="name">obs_freq_mat</span><span class="punctuation">)</span>
</pre>
<ol class="arabic simple">
<li><tt class="docutils literal">obs_freq_mat</tt>: 一个观测频率矩阵。基于“obs_freq_mat”的频率返回矩阵的熵值。矩阵实例须为LO或者SUBS。</li>
</ol>
</li>
<li><pre class="code python first literal-block">
<span class="name builtin">sum</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">)</span>
</pre>
<p>计算矩阵的字母表中每个字母值的总和，返回值是字典的形式 <tt class="docutils literal">{i1: s1, i2: <span class="pre">s2,...,in:sn}</span></tt>, 其中:</p>
<ul class="simple">
<li>i: 一个字母;</li>
<li>s: 半矩阵中某个字母值的总和;</li>
<li>n: alphabet中字母的个数。</li>
</ul>
</li>
<li><pre class="code python first literal-block">
<span class="name">print_mat</span><span class="punctuation">(</span><span class="name builtin pseudo">self</span><span class="punctuation">,</span><span class="name">f</span><span class="punctuation">,</span><span class="name">format</span><span class="operator">=</span><span class="literal string">&quot;</span><span class="literal string interpol">%4d</span><span class="literal string">&quot;</span><span class="punctuation">,</span><span class="name">bottomformat</span><span class="operator">=</span><span class="literal string">&quot;</span><span class="literal string interpol">%4s</span><span class="literal string">&quot;</span><span class="punctuation">,</span><span class="name">alphabet</span><span class="operator">=</span><span class="name builtin pseudo">None</span><span class="punctuation">)</span>
</pre>
<p>将矩阵打印到文件句柄f。 <tt class="docutils literal">format</tt> 是矩阵值的格式； <tt class="docutils literal">bottomformat</tt> 是底部行（矩阵字母）的格式。下面是一个3字母矩阵的例子：</p>
<pre class="code python literal-block">
<span class="name">A</span> <span class="literal number integer">23</span>
<span class="name">B</span> <span class="literal number integer">12</span> <span class="literal number integer">34</span>
<span class="name">C</span> <span class="literal number integer">7</span>  <span class="literal number integer">22</span>  <span class="literal number integer">27</span>
  <span class="name">A</span>   <span class="name">B</span>   <span class="name">C</span>
</pre>
<p><tt class="docutils literal">alphabet</tt> 可选参数是alphabet中所有字符的一个字符串。如果用户提供了数据，则轴上字母的顺序是根据字符串中的顺序，而不是字母表的顺序。</p>
</li>
</ol>
</li>
<li><p class="first">用法</p>
<p>安排下面这部分是因为大多数读者希望知道如何产生一个对数机率矩阵（log-odds matrix）。
当然，也可以生成和研究过渡矩阵（log-odds matrix）。
但是大部分的人只是想要一个对数机率矩阵，仅此而已。</p>
<ol class="arabic">
<li><p class="first">产生一个可接受的替代矩阵</p>
<p>首先，你应该从数据中产生出一个可接受替代矩阵（accepted replacement matrix，ARM）。
ARM中的数值是根据你的数据中替换的个数决定的。数据可以是一对或者多对的序列比对结果。
例如，丙氨酸被半胱氨酸替换了10次，而半胱氨酸被丙氨酸替换了12次，其相对应的ARM为：</p>
<pre class="code python literal-block">
<span class="punctuation">(</span><span class="literal string">'A'</span><span class="punctuation">,</span><span class="literal string">'C'</span><span class="punctuation">):</span> <span class="literal number integer">10</span><span class="punctuation">,</span> <span class="punctuation">(</span><span class="literal string">'C'</span><span class="punctuation">,</span><span class="literal string">'A'</span><span class="punctuation">):</span> <span class="literal number integer">12</span>
</pre>
<p>由于顺序并不重要，用户也可以只用一个输入:</p>
<pre class="code python literal-block">
<span class="punctuation">(</span><span class="literal string">'A'</span><span class="punctuation">,</span><span class="literal string">'C'</span><span class="punctuation">):</span> <span class="literal number integer">22</span>
</pre>
<p>一个SeqMat实例的初始化可以用全矩阵（第一种计数方法：10,12），也可以用半矩阵（后一种方法，22）。
一个蛋白字母全矩阵的大小应该是20x20 = 400。而一个这样的半矩阵大小是20x20/2 + 20/2 = 210。
这是因为相同字母的输入并没有改变（矩阵的对角线）。如果一个大小为N的alphabet：</p>
<ol class="arabic simple">
<li>全矩阵大小: N*N</li>
<li>半矩阵大小: N(N+1)/2</li>
</ol>
<p>如果传递的是全矩阵，SeqMat的构造函数会自动产生半矩阵。
如果传递的是半矩阵，则键的字母将按照字母表顺序排列('A','C')，而不是('C','A')。</p>
<p>讲到这里，如果你想知道的仅仅只是怎样产生一个对数机率矩阵的话，请直接看用法示例那个章节。对于想要更加深入地知道核苷酸/氨基酸频率数据的读者，接下来要讲的是内部函数的细节。</p>
</li>
<li><p class="first">生成观测频率矩阵(observed frequency matrix，OFM)</p>
<p>用法:</p>
<pre class="code python literal-block">
<span class="name">OFM</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">_build_obs_freq_mat</span><span class="punctuation">(</span><span class="name">ARM</span><span class="punctuation">)</span>
</pre>
<p>OFM是由ARM产生的，只是将替换的个数换成了替换频率。</p>
</li>
<li><p class="first">生成期望频率矩阵(expected frequency matrix，EFM)</p>
<p>用法:</p>
<pre class="code python literal-block">
<span class="name">EFM</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">_build_exp_freq_mat</span><span class="punctuation">(</span><span class="name">OFM</span><span class="punctuation">,</span><span class="name">exp_freq_table</span><span class="punctuation">)</span>
</pre>
<ol class="arabic simple">
<li><tt class="docutils literal">exp_freq_table</tt>: 为一个FreqTable的实例。有关FreqTable更多信息请见第 <a class="reference external" href="#sec:freq_table">20.2.2</a> 节。简单地说，期望频率表表示字母表中每个元素显示的频率。这个表相当于一个字典，字母是键，字母对应的频率是值。总和为1。</li>
</ol>
<p>期望频率表可以（理论上说也应该可以）从OFM得到。所以大多数情况你可以用下面的代码产生 <tt class="docutils literal">exp_freq_table</tt>:</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">exp_freq_table</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">_exp_freq_table_from_obs_freq</span><span class="punctuation">(</span><span class="name">OFM</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">EFM</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">_build_exp_freq_mat</span><span class="punctuation">(</span><span class="name">OFM</span><span class="punctuation">,</span><span class="name">exp_freq_table</span><span class="punctuation">)</span>
</pre>
<p>如果需要，你也可以使用自己提供的 <tt class="docutils literal">exp_freq_table</tt> 。</p>
</li>
<li><p class="first">生成替换频率矩阵(substitution frequency matrix，SFM)</p>
<p>用法:</p>
<pre class="code python literal-block">
<span class="name">SFM</span> <span class="operator">=</span> <span class="name">SubsMat</span><span class="operator">.</span><span class="name">_build_subs_mat</span><span class="punctuation">(</span><span class="name">OFM</span><span class="punctuation">,</span><span class="name">EFM</span><span class="punctuation">)</span>
</pre>
<p>使用观察频率矩阵(OFM)和期望频率矩阵(EFM)，得到相应值的除法结果。</p>
</li>
<li><p class="first">生成对数机率矩阵(log-odds matrix， LOM)</p>
<p>用法:</p>
<pre class="code python literal-block">
<span class="name">LOM</span><span class="operator">=</span><span class="name">SubsMat</span><span class="operator">.</span><span class="name">_build_log_odds_mat</span><span class="punctuation">(</span><span class="name">SFM</span><span class="punctuation">[,</span><span class="name">logbase</span><span class="operator">=</span><span class="literal number integer">10</span><span class="punctuation">,</span><span class="name">factor</span><span class="operator">=</span><span class="literal number float">10.0</span><span class="punctuation">,</span><span class="name">round_digit</span><span class="operator">=</span><span class="literal number integer">1</span><span class="punctuation">])</span>
</pre>
<ol class="arabic simple">
<li>使用替换频率矩阵(SFM)。</li>
<li><tt class="docutils literal">logbase</tt>: 用来产生对数机率值的对数的底。</li>
<li><tt class="docutils literal">factor</tt>: 对数机率值的乘数因子。
每个数通过log(LOM[key])*factor产生，如果需要，还可以四舍五入到 <tt class="docutils literal">round_digit</tt> 指定的小数点位数。</li>
</ol>
</li>
</ol>
</li>
<li><p class="first">用法示例</p>
<p>因为大部分人都想用最简单的方法产生对数机率矩阵（LOM），SubsMat提供了一个可以完成所有需求的函数：</p>
<pre class="code python literal-block">
<span class="name">make_log_odds_matrix</span><span class="punctuation">(</span><span class="name">acc_rep_mat</span><span class="punctuation">,</span><span class="name">exp_freq_table</span><span class="operator">=</span><span class="name builtin pseudo">None</span><span class="punctuation">,</span><span class="name">logbase</span><span class="operator">=</span><span class="literal number integer">10</span><span class="punctuation">,</span>
                      <span class="name">factor</span><span class="operator">=</span><span class="literal number float">10.0</span><span class="punctuation">,</span><span class="name">round_digit</span><span class="operator">=</span><span class="literal number integer">0</span><span class="punctuation">):</span>
</pre>
<ol class="arabic simple">
<li><tt class="docutils literal">acc_rep_mat</tt>: 用户提供可接受替代矩阵（ARM）</li>
<li><tt class="docutils literal">exp_freq_table</tt>: 期望频率表。如果用户没有提供，就从 <tt class="docutils literal">acc_rep_mat</tt> 产生。</li>
<li><tt class="docutils literal">logbase</tt>: LOM的对数的底。默认为10。</li>
<li><tt class="docutils literal">round_digit</tt>: 四舍五入的小数点位数。默认为0。</li>
</ol>
</li>
</ol>
</div>
<div class="section" id="freqtable">
<h3>20.2.2  FreqTable</h3>
<pre class="code python literal-block">
<span class="name">FreqTable</span><span class="operator">.</span><span class="name">FreqTable</span><span class="punctuation">(</span><span class="name">UserDict</span><span class="operator">.</span><span class="name">UserDict</span><span class="punctuation">)</span>
</pre>
<ol class="arabic">
<li><p class="first">属性:</p>
<ol class="arabic simple">
<li><tt class="docutils literal">alphabet</tt>: 一个Bio.Alphabet实例。</li>
<li><tt class="docutils literal">data</tt>: 频率字典</li>
<li><tt class="docutils literal">count</tt>: 计数字典(如果有计数的话)。</li>
</ol>
</li>
<li><p class="first">功能:</p>
<ol class="arabic simple">
<li><tt class="docutils literal">read_count(f)</tt>: 从f读入一个计数文件。然后将其转换成频率。</li>
<li><tt class="docutils literal">read_freq(f)</tt>: 从f读入一个频率数据文件。当然，我们不用计数，我们感兴趣的是字母频率。</li>
</ol>
</li>
<li><p class="first">用法示例: 文件中有残基的个数，用空格分格，形式如下（以3个字母为例）：</p>
<pre class="code python literal-block">
<span class="name">A</span>   <span class="literal number integer">35</span>
<span class="name">B</span>   <span class="literal number integer">65</span>
<span class="name">C</span>   <span class="literal number integer">100</span>
</pre>
<p>用 <tt class="docutils literal">FreqTable.read_count(file_handle)</tt> 函数读入。</p>
<p>一个等价的频率文件:</p>
<pre class="code python literal-block">
<span class="name">A</span>  <span class="literal number float">0.175</span>
<span class="name">B</span>  <span class="literal number float">0.325</span>
<span class="name">C</span>  <span class="literal number float">0.5</span>
</pre>
<p>反之，残基频率或者计数也可以作为字典输入。
一个计数字典的例子（3个字母）：</p>
<pre class="code python literal-block">
<span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number integer">35</span><span class="punctuation">,</span> <span class="literal string">'B'</span><span class="punctuation">:</span> <span class="literal number integer">65</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number integer">100</span><span class="punctuation">}</span>
</pre>
<p>这也意味着'C'的频率是0.5，'B'的频率是0.325，'A'的频率是0.175。A、B、C的总和为200。</p>
<p>一个相同数据的频率字典如下：</p>
<pre class="code python literal-block">
<span class="punctuation">{</span><span class="literal string">'A'</span><span class="punctuation">:</span> <span class="literal number float">0.175</span><span class="punctuation">,</span> <span class="literal string">'B'</span><span class="punctuation">:</span> <span class="literal number float">0.325</span><span class="punctuation">,</span> <span class="literal string">'C'</span><span class="punctuation">:</span> <span class="literal number float">0.5</span><span class="punctuation">}</span>
</pre>
<p>总和为1。</p>
<p>当传入一个字典数据作为参数，应该指出这是一个计数还是频率的字典。因此FreqTable类的构造函数需要两个参数：字典本身和FreqTable.COUNT或者FreqTable.FREQ，分别代表计数或者频率。</p>
<p>读入期望的计数，readCount会产生频率。下面的任意一个都可以用来产生频率表（ftab）：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">SubsMat</span> <span class="keyword namespace">import</span> <span class="operator">*</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ftab</span> <span class="operator">=</span> <span class="name">FreqTable</span><span class="operator">.</span><span class="name">FreqTable</span><span class="punctuation">(</span><span class="name">my_frequency_dictionary</span><span class="punctuation">,</span><span class="name">FreqTable</span><span class="operator">.</span><span class="name">FREQ</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ftab</span> <span class="operator">=</span> <span class="name">FreqTable</span><span class="operator">.</span><span class="name">FreqTable</span><span class="punctuation">(</span><span class="name">my_count_dictionary</span><span class="punctuation">,</span><span class="name">FreqTable</span><span class="operator">.</span><span class="name">COUNT</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ftab</span> <span class="operator">=</span> <span class="name">FreqTable</span><span class="operator">.</span><span class="name">read_count</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">'myCountFile'</span><span class="punctuation">))</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">ftab</span> <span class="operator">=</span> <span class="name">FreqTable</span><span class="operator">.</span><span class="name">read_frequency</span><span class="punctuation">(</span><span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">'myFrequencyFile'</span><span class="punctuation">))</span>
</pre>
</li>
</ol>
</div>
</div>
</div>
<div class="section" id="id248">
<h1>第21章 为Biopython做贡献</h1>
<div class="section" id="id249">
<h2>21.1  错误报告与功能需求</h2>
<p>对于biopython的开发者来说，从使用者得到反馈是非常重要的；来自于众多使用者的问题反馈、错误报告（以及补丁）对于biopython这样的开源项目有着莫大的帮助。</p>
<p><a class="reference external" href="http://biopython.org/wiki/Mailing_lists">Biopython邮件列表</a> 是一个讨论功能需求和潜在bugs的主要论坛：</p>
<ul class="simple">
<li><a class="reference external" href="mailto:biopython&#64;biopython.org">biopython&#64;biopython.org</a> – 讨论任何跟biopython有关的问题。</li>
<li><a class="reference external" href="mailto:biopython-dev&#64;biopython.org">biopython-dev&#64;biopython.org</a>
– 开发者用于讨论biopython开发的邮件列表（其他人员同样可以做出贡献）</li>
</ul>
<p>除此之外，如果你发现bug， 请提交到我们的bug监控页面（<a class="reference external" href="http://redmine.open-bio.org/projects/biopython">http://redmine.open-bio.org/projects/biopython</a>）
这样，这些bug才不至于淹没于浩如烟海的收件箱中被遗忘。</p>
</div>
<div class="section" id="id250">
<h2>21.2  邮件列表与帮助新手</h2>
<p>我们希望所有的biopython使用者都在biopython主邮件列表注册并订阅相关邮件。同时使用者在对某一领域有一定了解之后，也希望他们能够帮助入门者并对他们的问题进行解答，毕竟每个人开始的时候都是菜鸟。</p>
</div>
<div class="section" id="id251">
<h2>21.3  贡献文档</h2>
<p>我们很高兴能从用户那里收到反馈，不论是通过bug报告还是邮件列表的形式。或许在你阅读这一教程的时候，你可能会发现某些你所感兴趣的话题没被涉及或者没有解释清楚。甚至biopython的内建文档（docstrings）以及一些在线文档（<a class="reference external" href="http://biopython.org/DIST/docs/api">http://biopython.org/DIST/docs/api</a>）有所缺失，有能力的用户都可以帮助我们填补这些空白。</p>
</div>
<div class="section" id="id252">
<h2>21.4  贡献学习手册示例</h2>
<p>如18章中用户手册解释的那样，如今biopython有一个汇集用户贡献的使用文档的wiki文库（<a class="reference external" href="http://biopython.org/wiki/Category:Cookbook">http://biopython.org/wiki/Category:Cookbook</a>），也许你也可以把做的一些使用说明放在这里。</p>
</div>
<div class="section" id="id253">
<h2>21.5  维护跨平台发行版本</h2>
<p>目前我们提供biopython的源代码安装（适用于任何操作系统，如果你安装了正确的编译工具的话），同时还提供有点击运行的windows安装工具。这涵盖了当今当多数操作系统。</p>
<p>对于大多数linux操作系统来说，都有志愿者将源代码进行编译成包，同时也解决了包依赖等相关问题，linux用户可以很容易的安装。对于这些工作，我们非常感激。如果你想要对于这一工作有所贡献的话，请阅读你的linux发行版关于这一事项的详细说明。</p>
<p>对于想要从事这一工作的用户，以下是一些关于主要操作系统平台的建议：</p>
<dl class="docutils">
<dt><strong>Windows</strong></dt>
<dd><p class="first">– Windows 产品通常有一个图形安装工具，能够将biopython所有的基本组建安装到相应的目录。我们使用
Distutils 工具来创建能完成这一任务的installer。</p>
<p>首先，你要确定你的Windows操作系统中安装了C编译器，而且你能够正确编译并安装（通常这是困难的地方），可以参考Biopython安装说明。</p>
<p>在你配置好C编译器之后，安装工作就很简单：</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">setup</span><span class="operator">.</span><span class="name">py</span> <span class="name">bdist_wininst</span>
</pre>
<p class="last">目前，在32位Windows操作系统上安装biopython没有任何问题。只是64位操作系统的安装还没能实现，要是谁能在这方面做点贡献就太好了。</p>
</dd>
<dt><strong>RPMs</strong></dt>
<dd><p class="first">– RPMs在一些linux发行版中是非常流行的包管理系统，在RPMs主页（<a class="reference external" href="http://www.rpm.org">http://www.rpm.org</a>）有很多关于RPMs的文档说明帮助你快速的开始。为你的系统创建一个RMP是很容易的一件事。你仅仅需要将源代码编译成包（需要一个能正常工作的C编译器），更多信息可以参考Biopython安装说明。</p>
<p>创建一个RPM,仅需做以下工作：</p>
<pre class="code python literal-block">
<span class="name">python</span> <span class="name">setup</span><span class="operator">.</span><span class="name">py</span> <span class="name">bdist_rpm</span>
</pre>
<p class="last">在终端运行以上命令后，会在 <tt class="docutils literal">dist</tt> 目录下创建一个跟你使用的操作系统相关的RPM，以及一个RPM源文件。一般情况下，这个RPM能正常工作。简洁而实用！</p>
</dd>
<dt><strong>Macintosh</strong></dt>
<dd>– 由于Apple迁移到Mac OS X, 很多工作就简单多了。 一般来说，Mac操作系统相当于一个Unix变体，Biopython源码的安装同在linux上一样容易。安装Apple's X-code是安装所有GCC编译器最简易的方法。
我们或许会为Mac操作系统提供点击-安装的工具，但目前为止还没必要。</dd>
</dl>
<p>一旦得到一个安装包，务必在你的系统上检测并确保所有文件能够正确安装并能正常工作。在这一工作完成之后，你的工作就结束了，剩下的就是Biopython开发者的事了（如果不确定该发送给谁，可以给Biopython主邮件列表发送邮件），谢谢！</p>
</div>
<div class="section" id="unit-tests">
<h2>21.6  贡献单元测试（Unit Tests）</h2>
<p>即使你没有任何新的功能添加到Biopython，但仍然想写一些代码，请考虑增加我们的单元测试。我们整个第19 <a class="reference external" href="#sec:regr_test">19</a> 章都在讲这个内容。</p>
</div>
<div class="section" id="id254">
<h2>21.7  贡献源码</h2>
<p>除了使用Python语言开发生物学相关的程序外，任何人都可以没有限制地加入Biopython源码的开发。任何人若对某方面的编程感兴趣，Biopython邮件列表是讨论此事最合适的地方——只需告知我们你的兴趣所在或工作内容。通常来讲，在开发某个模块之前，我们会在邮件列表里讨论此事，因为这样做会有助于产生好的想法，讨论完成之后，就剩下编程了！</p>
<p>主要的Biopython发布版本会尽量做到统一和通用，以方便用户的使用。在附带文档（<a class="reference external" href="http://biopython.org/wiki/Contributing">http://biopython.org/wiki/Contributing</a>）中，你可以获取在Biopython中用到的编程方式的原则。同时，我们也尽量在发行版和文档中加入源码和测试（关于regression tesing framework详见19章 <a class="reference external" href="#sec:regr_test">#sec:regr_test</a> 小节），以使得各方面能保持一致并正常工作。</p>
<p>值得注意的是，你需要有合法的权利去贡献源码并且在Biopython发行许可下发布。当然了，要是你的程序完全是由你自己编写，没有任何其他的代码，就不要为此担心了。另外，在贡献衍生版本的时候，会有些问题——比如说一些给予GPL或者LPGL的程序与Biopython许可不相容。如果你有什么疑问，请在biopython—dev邮件列表里讨论。</p>
<p>另外一个关于向Biopython贡献源码的问题涉及到开发和运行时依赖问题。一般来讲，编写程序调用像BLAST、EMBOSS或者ClustalW这样的独立程序没什么问题。但是，任何依赖于其他文库的程序——即使是Python文库（尤其是像NumPy这样用于编译和安装Biopython的文库）就需要做进一步的讨论。</p>
<p>除此之外，如果你手头有某些代码，而你又觉得不适合发行版，却又想共享出来，你可以将它们放在一个专门收集生物信息学Python代码的地方（<a class="reference external" href="http://biopython.org/wiki/Scriptcentral">http://biopython.org/wiki/Scriptcentral</a>），</p>
<p>希望这个文档能在你使用Biopython的过程中带给你想要的信息，当然了，最重要的就是贡献。</p>
</div>
</div>
<div class="section" id="python">
<h1>第22章 附录：Python相关知识</h1>
<p>如果你对Python编程还不熟练，那么你在使用Biopython过程中遇到的问题通常与Python本身有关。本章节主要向读者提供一些使用Biopython文库过程中有用的建议和一些常用代码（至少对我们来说很常用）。关于这一章的内容，你要是有什么好的建议的话，请告诉我们。</p>
<div class="section" id="handle">
<h2>22.1  到底什么是句柄（handle）？</h2>
<p>这个文档中，句柄常被提及，而且也比较难理解（至少对我来说）。一般说来，你可以把句柄想象成一个对文本信息的“封装”。</p>
<p>相比普通文本信息，使用句柄至少有两个好处：</p>
<ol class="arabic simple">
<li>对于以不同方式存储的信息，句柄提供了一个标准的处理方法。这些文本信息可能来自文件、内存中的一个字符串、命令行指令的输出或者来自于远程网站信息，但是句柄提供了一种通用的方式处理这些不同格式和来源的文本信息。</li>
<li>句柄可以依次读取文本信息，而不是一次读取所有信息。这点在处理超大文件时尤为有用，因为一次载入一个大文件可能会占去你所有的内存。</li>
</ol>
<p>不论是从文件读取文本信息还是将文本信息写入文件，句柄都能胜任。在读取文件时，常用的函数有 <tt class="docutils literal">read()</tt> 和 <tt class="docutils literal">readline()</tt> , 前者可以通过句柄读取所有文本信息，而后者则每次读取一行；对于文本信息的写入，则通常使用 <tt class="docutils literal">write()</tt> 函数。</p>
<p>句柄最常见的使用就是从文件读取信息，这可以通过Python内置函数 <tt class="docutils literal">open</tt> 来完成。 下面示例中，我们打开一个指向文件 <a class="reference external" href="examples/m_cold.fasta">m_cold.fasta</a> （可通过网址 <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/examples/m_cold.fasta">http://biopython.org/DIST/docs/tutorial/examples/m_cold.fasta</a> 获取）的句柄：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;r&quot;</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">handle</span><span class="operator">.</span><span class="name">readline</span><span class="punctuation">()</span>
<span class="literal string">&quot;&gt;gi|8332116|gb|BE037100.1|BE037100 MP14H09 MP Mesembryanthemum ...</span><span class="literal string escape">\n</span><span class="literal string">&quot;</span>
</pre>
<p>Biopython中句柄常用来向解析器（parsers）传递信息。比如说，自Biopython1.54版本后， <tt class="docutils literal">Bio.SeqIO</tt> 和 <tt class="docutils literal">Bio.AlignIO</tt> 模块中的主要函数都可以使用文件名来代替句柄使用：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
</pre>
<p>在比较早的BioPython版本中，必须使用句柄：</p>
<pre class="code python literal-block">
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name builtin">open</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta&quot;</span><span class="punctuation">,</span> <span class="literal string">&quot;r&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>这种操作方式仍有其用武之地，比如在解析一个gzip压缩的FASTA文件中：</p>
<pre class="code python literal-block">
<span class="keyword namespace">import</span> <span class="name namespace">gzip</span>
<span class="keyword namespace">from</span> <span class="name namespace">Bio</span> <span class="keyword namespace">import</span> <span class="name">SeqIO</span>
<span class="name">handle</span> <span class="operator">=</span> <span class="name">gzip</span><span class="operator">.</span><span class="name">open</span><span class="punctuation">(</span><span class="literal string">&quot;m_cold.fasta.gz&quot;</span><span class="punctuation">)</span>
<span class="keyword">for</span> <span class="name">record</span> <span class="operator word">in</span> <span class="name">SeqIO</span><span class="operator">.</span><span class="name">parse</span><span class="punctuation">(</span><span class="name">handle</span><span class="punctuation">,</span> <span class="literal string">&quot;fasta&quot;</span><span class="punctuation">):</span>
    <span class="keyword">print</span> <span class="name">record</span><span class="operator">.</span><span class="name">id</span><span class="punctuation">,</span> <span class="name builtin">len</span><span class="punctuation">(</span><span class="name">record</span><span class="punctuation">)</span>
<span class="name">handle</span><span class="operator">.</span><span class="name">close</span><span class="punctuation">()</span>
</pre>
<p>在第 <a class="reference external" href="#sec:SeqIO_compressed">5.2</a> 节中有更多此类示例可供参考，其中包括bzip2压缩文件的读取。</p>
<div class="section" id="id255">
<h3>22.1.1  从字符串创建句柄</h3>
<p>一个比较有用的工具是将字符串中包含的文本信息传递给一个句柄。以下示例是使用Python标准文库 <tt class="docutils literal">cStringIO</tt> 来展示如何实现的：</p>
<pre class="code python literal-block">
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_info</span> <span class="operator">=</span> <span class="literal string">'A string</span><span class="literal string escape">\n</span><span class="literal string"> with multiple lines.'</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">my_info</span>
<span class="name">A</span> <span class="name">string</span>
 <span class="keyword">with</span> <span class="name">multiple</span> <span class="name">lines</span><span class="operator">.</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword namespace">from</span> <span class="name namespace">StringIO</span> <span class="keyword namespace">import</span> <span class="name">StringIO</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">my_info_handle</span> <span class="operator">=</span> <span class="name">StringIO</span><span class="punctuation">(</span><span class="name">my_info</span><span class="punctuation">)</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">first_line</span> <span class="operator">=</span> <span class="name">my_info_handle</span><span class="operator">.</span><span class="name">readline</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">first_line</span>
<span class="name">A</span> <span class="name">string</span>
<span class="operator">&lt;</span><span class="name">BLANKLINE</span><span class="operator">&gt;</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="name">second_line</span> <span class="operator">=</span> <span class="name">my_info_handle</span><span class="operator">.</span><span class="name">readline</span><span class="punctuation">()</span>
<span class="operator">&gt;&gt;&gt;</span> <span class="keyword">print</span> <span class="name">second_line</span>
 <span class="keyword">with</span> <span class="name">multiple</span> <span class="name">lines</span><span class="operator">.</span>
</pre>
</div>
</div>
</div>
<div class="section" id="id256">
<h1>References</h1>
<dl class="docutils">
<dt>[1]</dt>
<dd>Peter J. A. Cock, Tiago Antao, Jeffrey T. Chang, Brad A. Chapman,
Cymon J. Cox, Andrew Dalke, Iddo Friedberg, Thomas Hamelryck, Frank
Kauff, Bartek Wilczynski, Michiel J. L. de Hoon: “Biopython: freely
available Python tools for computational molecular biology and
bioinformatics”. <em>Bioinformatics</em> <strong>25</strong> (11), 1422–1423 (2009).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/btp163">doi:10.1093/bioinformatics/btp163</a>,</dd>
<dt>[2]</dt>
<dd>Leighton Pritchard, Jennifer A. White, Paul R.J. Birch, Ian K. Toth:
“GenomeDiagram: a python package for the visualization of
large-scale genomic data”. <em>Bioinformatics</em> <strong>22</strong> (5): 616–617
(2006).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/btk021">doi:10.1093/bioinformatics/btk021</a>,</dd>
<dt>[3]</dt>
<dd>Ian K. Toth, Leighton Pritchard, Paul R. J. Birch: “Comparative
genomics reveals what makes an enterobacterial plant pathogen”.
<em>Annual Review of Phytopathology</em> <strong>44</strong>: 305–336 (2006).
<a class="reference external" href="http://dx.doi.org/10.1146/annurev.phyto.44.070505.143444">doi:10.1146/annurev.phyto.44.070505.143444</a>,</dd>
<dt>[4]</dt>
<dd>Géraldine A. van der Auwera, Jaroslaw E. Król, Haruo Suzuki, Brian
Foster, Rob van Houdt, Celeste J. Brown, Max Mergeay, Eva M. Top:
“Plasmids captured in C. metallidurans CH34: defining the PromA
family of broad-host-range plasmids”. <em>Antonie van Leeuwenhoek</em>
<strong>96</strong> (2): 193–204 (2009).
<a class="reference external" href="http://dx.doi.org/10.1007/s10482-009-9316-9">doi:10.1007/s10482-009-9316-9</a></dd>
<dt>[5]</dt>
<dd>Caroline Proux, Douwe van Sinderen, Juan Suarez, Pilar Garcia,
Victor Ladero, Gerald F. Fitzgerald, Frank Desiere, Harald Brüssow:
“The dilemma of phage taxonomy illustrated by comparative genomics
of Sfi21-Like Siphoviridae in lactic acid bacteria”. <em>Journal of
Bacteriology</em> <strong>184</strong> (21): 6026–6036 (2002).
<a class="reference external" href="http://dx.doi.org/10.1128/JB.184.21.6026-6036.2002">http://dx.doi.org/10.1128/JB.184.21.6026-6036.2002</a></dd>
<dt>[6]</dt>
<dd>Florian Jupe, Leighton Pritchard, Graham J. Etherington, Katrin
MacKenzie, Peter JA Cock, Frank Wright, Sanjeev Kumar Sharma1, Dan
Bolser, Glenn J Bryan, Jonathan DG Jones, Ingo Hein: “Identification
and localisation of the NB-LRR gene family within the potato
genome”. <em>BMC Genomics</em> <strong>13</strong>: 75 (2012).
<a class="reference external" href="http://dx.doi.org/10.1186/1471-2164-13-75">http://dx.doi.org/10.1186/1471-2164-13-75</a></dd>
<dt>[7]</dt>
<dd>Peter J. A. Cock, Christopher J. Fields, Naohisa Goto, Michael L.
Heuer, Peter M. Rice: “The Sanger FASTQ file format for sequences
with quality scores, and the Solexa/Illumina FASTQ variants”.
<em>Nucleic Acids Research</em> <strong>38</strong> (6): 1767–1771 (2010).
<a class="reference external" href="http://dx.doi.org/10.1093/nar/gkp1137">doi:10.1093/nar/gkp1137</a></dd>
<dt>[8]</dt>
<dd>Patrick O. Brown, David Botstein: “Exploring the new world of the
genome with DNA microarrays”. <em>Nature Genetics</em> <strong>21</strong> (Supplement
1), 33–37 (1999).
<a class="reference external" href="http://dx.doi.org/10.1038/4462">doi:10.1038/4462</a></dd>
<dt>[9]</dt>
<dd>Eric Talevich, Brandon M. Invergo, Peter J.A. Cock, Brad A. Chapman:
“Bio.Phylo: A unified toolkit for processing, analyzing and
visualizing phylogenetic trees in Biopython”. <em>BMC Bioinformatics</em>
<strong>13</strong>: 209 (2012).
<a class="reference external" href="http://dx.doi.org/10.1186/1471-2105-13-209">doi:10.1186/1471-2105-13-209</a></dd>
<dt>[10]</dt>
<dd>Athel Cornish-Bowden: “Nomenclature for incompletely specified bases
in nucleic acid sequences: Recommendations 1984.” <em>Nucleic Acids
Research</em> <strong>13</strong> (9): 3021–3030 (1985).
<a class="reference external" href="http://dx.doi.org/10.1093/nar/13.9.3021">doi:10.1093/nar/13.9.3021</a></dd>
<dt>[11]</dt>
<dd>Douglas R. Cavener: “Comparison of the consensus sequence flanking
translational start sites in Drosophila and vertebrates.” <em>Nucleic
Acids Research</em> <strong>15</strong> (4): 1353–1361 (1987).
<a class="reference external" href="http://dx.doi.org/10.1093/nar/15.4.1353">doi:10.1093/nar/15.4.1353</a></dd>
<dt>[12]</dt>
<dd>Timothy L. Bailey and Charles Elkan: “Fitting a mixture model by
expectation maximization to discover motifs in biopolymers”,
<em>Proceedings of the Second International Conference on Intelligent
Systems for Molecular Biology</em> 28–36. AAAI Press, Menlo Park,
California (1994).</dd>
<dt>[13]</dt>
<dd>Brad Chapman and Jeff Chang: “Biopython: Python tools for
computational biology”. <em>ACM SIGBIO Newsletter</em> <strong>20</strong> (2): 15–19
(August 2000).</dd>
<dt>[14]</dt>
<dd>Michiel J. L. de Hoon, Seiya Imoto, John Nolan, Satoru Miyano: “Open
source clustering software”. <em>Bioinformatics</em> <strong>20</strong> (9): 1453–1454
(2004).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/bth078">doi:10.1093/bioinformatics/bth078</a></dd>
<dt>[15]</dt>
<dd>Michiel B. Eisen, Paul T. Spellman, Patrick O. Brown, David
Botstein: “Cluster analysis and display of genome-wide expression
patterns”. <em>Proceedings of the National Academy of Science USA</em>
<strong>95</strong> (25): 14863–14868 (1998).
<a class="reference external" href="http://dx.doi.org/10.1073/pnas.96.19.10943-c">doi:10.1073/pnas.96.19.10943-c</a></dd>
<dt>[16]</dt>
<dd>Gene H. Golub, Christian Reinsch: “Singular value decomposition and
least squares solutions”. In <em>Handbook for Automatic Computation</em>,
<strong>2</strong>, (Linear Algebra) (J. H. Wilkinson and C. Reinsch, eds),
134–151. New York: Springer-Verlag (1971).</dd>
<dt>[17]</dt>
<dd>Gene H. Golub, Charles F. Van Loan: <em>Matrix computations</em>, 2nd
edition (1989).</dd>
<dt>[18]</dt>
<dd>Thomas Hamelryck and Bernard Manderick: 11PDB parser and structure
class implemented in Python”. <em>Bioinformatics</em>, <strong>19</strong> (17):
2308–2310 (2003) <a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/btg299">doi:
10.1093/bioinformatics/btg299</a>.</dd>
<dt>[19]</dt>
<dd>Thomas Hamelryck: “Efficient identification of side-chain patterns
using a multidimensional index tree”. <em>Proteins</em> <strong>51</strong> (1): 96–108
(2003).
<a class="reference external" href="http://dx.doi.org/10.1002/prot.10338">doi:10.1002/prot.10338</a></dd>
<dt>[20]</dt>
<dd>Thomas Hamelryck: “An amino acid has two sides; A new 2D measure
provides a different view of solvent exposure”. <em>Proteins</em> <strong>59</strong>
(1): 29–48 (2005).
<a class="reference external" href="http://dx.doi.org/10.1002/prot.20379">doi:10.1002/prot.20379</a>.</dd>
<dt>[21]</dt>
<dd>John A. Hartiga. <em>Clustering algorithms</em>. New York: Wiley (1975).</dd>
<dt>[22]</dt>
<dd>Anil L. Jain, Richard C. Dubes: <em>Algorithms for clustering data</em>.
Englewood Cliffs, N.J.: Prentice Hall (1988).</dd>
<dt>[23]</dt>
<dd>Voratas Kachitvichyanukul, Bruce W. Schmeiser: Binomial Random
Variate Generation. <em>Communications of the ACM</em> <strong>31</strong> (2): 216–222
(1988).
<a class="reference external" href="http://dx.doi.org/10.1145/42372.42381">doi:10.1145/42372.42381</a></dd>
<dt>[24]</dt>
<dd>Teuvo Kohonen: “Self-organizing maps”, 2nd Edition. Berlin; New
York: Springer-Verlag (1997).</dd>
<dt>[25]</dt>
<dd>Pierre L’Ecuyer: “Efficient and Portable Combined Random Number
Generators.” <em>Communications of the ACM</em> <strong>31</strong> (6): 742–749,774
(1988).
<a class="reference external" href="http://dx.doi.org/10.1145/62959.62969">doi:10.1145/62959.62969</a></dd>
<dt>[26]</dt>
<dd>Indraneel Majumdar, S. Sri Krishna, Nick V. Grishin: “PALSSE: A
program to delineate linear secondary structural elements from
protein structures.” <em>BMC Bioinformatics</em>, <strong>6</strong>: 202 (2005).
<a class="reference external" href="http://dx.doi.org/10.1186/1471-2105-6-202">doi:10.1186/1471-2105-6-202</a>.</dd>
<dt>[27]</dt>
<dd>V. Matys, E. Fricke, R. Geffers, E. G?ssling, M. Haubrock, R. Hehl,
K. Hornischer, D. Karas, A.E. Kel, O.V. Kel-Margoulis, D.U. Kloos,
S. Land, B. Lewicki-Potapov, H. Michael, R. Münch, I. Reuter, S.
Rotert, H. Saxel, M. Scheer, S. Thiele, E. Wingender E: “TRANSFAC:
transcriptional regulation, from patterns to profiles.” Nucleic
Acids Research <strong>31</strong> (1): 374–378 (2003).
<a class="reference external" href="http://dx.doi.org/10.1093/nar/gkg108">doi:10.1093/nar/gkg108</a></dd>
<dt>[28]</dt>
<dd>Robin Sibson: “SLINK: An optimally efficient algorithm for the
single-link cluster method”. <em>The Computer Journal</em> <strong>16</strong> (1):
30–34 (1973).
<a class="reference external" href="http://dx.doi.org/10.1093/comjnl/16.1.30">doi:10.1093/comjnl/16.1.30</a></dd>
<dt>[29]</dt>
<dd>George W. Snedecor, William G. Cochran: <em>Statistical methods</em>. Ames,
Iowa: Iowa State University Press (1989).</dd>
<dt>[30]</dt>
<dd>Pablo Tamayo, Donna Slonim, Jill Mesirov, Qing Zhu, Sutisak
Kitareewan, Ethan Dmitrovsky, Eric S. Lander, Todd R. Golub:
“Interpreting patterns of gene expression with self-organizing maps:
Methods and application to hematopoietic differentiation”.
<em>Proceedings of the National Academy of Science USA</em> <strong>96</strong> (6):
2907–2912 (1999).
<a class="reference external" href="http://dx.doi.org/10.1073/pnas.96.6.2907">doi:10.1073/pnas.96.6.2907</a></dd>
<dt>[31]</dt>
<dd>Robert C. Tryon, Daniel E. Bailey: <em>Cluster analysis</em>. New York:
McGraw-Hill (1970).</dd>
<dt>[32]</dt>
<dd>John W. Tukey: “Exploratory data analysis”. Reading, Mass.:
Addison-Wesley Pub. Co. (1977).</dd>
<dt>[33]</dt>
<dd>Ka Yee Yeung, Walter L. Ruzzo: “Principal Component Analysis for
clustering gene expression data”. <em>Bioinformatics</em> <strong>17</strong> (9):
763–774 (2001).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/17.9.763">doi:10.1093/bioinformatics/17.9.763</a></dd>
<dt>[34]</dt>
<dd>Alok Saldanha: “Java Treeview—extensible visualization of microarray
data”. <em>Bioinformatics</em> <strong>20</strong> (17): 3246–3248 (2004).
<a class="reference external" href="http://dx.doi.org/10.1093/bioinformatics/bth349">http://dx.doi.org/10.1093/bioinformatics/bth349</a></dd>
</dl>
</div>
</div>
</body>
</html>
